
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Large scale inference &#8212; Tsinfer manual</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=5349f25f" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=9c3e77be" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=1ae7504c"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'large_scale';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="API Documentation" href="api.html" />
    <link rel="prev" title="Inference overview" href="inference.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <script data-goatcounter="https://tskit.goatcounter.com/count" async src="//gc.zgo.at/count.js"></script>
</head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/tsinfer_logo.svg" class="logo__image only-light" alt=""/>
    <script>document.write(`<img src="_static/tsinfer_logo.svg" class="logo__image only-dark" alt=""/>`);</script>
  
  
    <p class="title logo__title">Version 0.5.1.dev22</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Installation</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Usage</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Inference</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="inference.html">Inference overview</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Large scale inference</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Interfaces</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="api.html">API Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">Command line interface</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">File Formats</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="file_formats.html">File formats</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Miscellaneous</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="development.html">Developer documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="CITATION.html">Citing tsinfer</a></li>
<li class="toctree-l1"><a class="reference internal" href="CHANGELOG.html">Changelog</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/large_scale.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Large scale inference</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-preparation">Data preparation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#ancestor-generation">Ancestor generation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#ancestor-matching">Ancestor matching</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#sample-matching">Sample matching</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="large-scale-inference">
<span id="sec-large-scale"></span><h1>Large scale inference<a class="headerlink" href="#large-scale-inference" title="Link to this heading">#</a></h1>
<p>Generally, for up to a few thousand samples a single multi-core machine
can infer a tree sequence in a few days, hours, or even minutes.
However, <em>tsinfer</em> has been successfully used with datasets up to half a million
samples, where ancestor and sample matching can take several CPU-years.
At this scale inference must be scaled across many machines.
<em>Tsinfer</em> provides specific APIs to enable this.
Here we detail considerations and tips for each step of the
inference process to help you scale up your analysis. A snakemake pipeline
which implements this parallelisation scheme is available as
<a class="reference external" href="https://github.com/benjeffery/tsinfer-snakemake">tsinfer-snakemake</a>.</p>
<section id="data-preparation">
<span id="sec-large-scale-ancestor-generation"></span><h2>Data preparation<a class="headerlink" href="#data-preparation" title="Link to this heading">#</a></h2>
<p>For large scale inference the data must be in <a class="reference external" href="https://github.com/sgkit-dev/vcf-zarr-spec">VCF Zarr</a>
format, read by the <a class="reference internal" href="api.html#tsinfer.VariantData" title="tsinfer.VariantData"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantData</span></code></a> class. <a class="reference external" href="https://github.com/sgkit-dev/bio2zarr">Bio2zarr</a>
is recommended for conversion from VCF, and <a class="reference external" href="https://github.com/sgkit-dev/sgkit">sgkit</a> can then
be used to perform initial filtering.</p>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>An upcoming tutorial will detail conversion from VCF to a VCF Zarr suitable for tsinfer.</p>
</div>
</section>
<section id="ancestor-generation">
<h2>Ancestor generation<a class="headerlink" href="#ancestor-generation" title="Link to this heading">#</a></h2>
<p>Ancestor generation is generally the fastest step in inference. It is not yet
parallelised out-of-core in tsinfer and must be performed on a single machine.
However it scales well on machines with
many cores and hyperthreading via the <code class="docutils literal notranslate"><span class="pre">num_threads</span></code> argument to
<a class="reference internal" href="api.html#tsinfer.generate_ancestors" title="tsinfer.generate_ancestors"><code class="xref py py-meth docutils literal notranslate"><span class="pre">generate_ancestors()</span></code></a>. The limiting factor is often that the
entire genotype array for the contig being inferred needs to fit in RAM.
This is the high-water mark for memory usage in tsinfer.</p>
<p>If your data consists of only biallelic sites, with no missingness,
the <code class="docutils literal notranslate"><span class="pre">genotype_encoding</span></code> argument can be set to
<a class="reference internal" href="api.html#tsinfer.GenotypeEncoding.ONE_BIT" title="tsinfer.GenotypeEncoding.ONE_BIT"><code class="xref py py-class docutils literal notranslate"><span class="pre">tsinfer.GenotypeEncoding.ONE_BIT</span></code></a> which reduces the memory footprint of
the genotype array by a factor of 8, such that the RAM needed is roughly
<code class="docutils literal notranslate"><span class="pre">num_sites</span> <span class="pre">*</span> <span class="pre">num_samples</span> <span class="pre">*</span> <span class="pre">ploidy</span> <span class="pre">/</span> <span class="pre">8</span> <span class="pre">bytes</span></code>. This memory optimisation
results in a surprisingly small increase in runtime.</p>
</section>
<section id="ancestor-matching">
<h2>Ancestor matching<a class="headerlink" href="#ancestor-matching" title="Link to this heading">#</a></h2>
<p>Ancestor matching is one of the more time consuming steps of inference. It
proceeds in groups, progressively growing the tree sequence with younger
ancestors. At each stage the parallelism is limited to the number of ancestors
whose possible inheritors are already matched, as all possible inheritors
of a sample must be matched in an earlier group. For a typical human data set
the number of samples per group varies from single digits up to approximately
the number of samples.
The plot below shows the number of ancestors matched in each group for a typical
human data set, earlier groups are older ancestors:</p>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/ancestor_grouping.png"><img alt="_images/ancestor_grouping.png" src="_images/ancestor_grouping.png" style="width: 80%;" />
</a>
</figure>
<p>There are five tsinfer API methods that can be used to parallelise ancestor
matching.</p>
<p>The five methods are:</p>
<ol class="arabic simple">
<li><p><a class="reference internal" href="api.html#tsinfer.match_ancestors_batch_init" title="tsinfer.match_ancestors_batch_init"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_init()</span></code></a></p></li>
<li><p><a class="reference internal" href="api.html#tsinfer.match_ancestors_batch_groups" title="tsinfer.match_ancestors_batch_groups"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_groups()</span></code></a></p></li>
<li><p><a class="reference internal" href="api.html#tsinfer.match_ancestors_batch_group_partition" title="tsinfer.match_ancestors_batch_group_partition"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_group_partition()</span></code></a></p></li>
<li><p><a class="reference internal" href="api.html#tsinfer.match_ancestors_batch_group_finalise" title="tsinfer.match_ancestors_batch_group_finalise"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_group_finalise()</span></code></a></p></li>
<li><p><a class="reference internal" href="api.html#tsinfer.match_ancestors_batch_finalise" title="tsinfer.match_ancestors_batch_finalise"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_finalise()</span></code></a></p></li>
</ol>
<p>Initially <a class="reference internal" href="api.html#tsinfer.match_ancestors_batch_init" title="tsinfer.match_ancestors_batch_init"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_init()</span></code></a> should be called to
set up the batch matching and to determine the groupings of ancestors.
This method writes a file <code class="docutils literal notranslate"><span class="pre">metadata.json</span></code> to the <code class="docutils literal notranslate"><span class="pre">work_dir</span></code> that contains
a JSON encoded dictionary with configuration for later steps, and a key
<code class="docutils literal notranslate"><span class="pre">ancestor_grouping</span></code> which is a list of dictionaries, each containing the
list of ancestors in that group (key:<code class="docutils literal notranslate"><span class="pre">ancestors</span></code>) and a proposed partioning of
those ancestors into sets that can be matched in parallel (key:<code class="docutils literal notranslate"><span class="pre">partitions</span></code>).
The dictionary is also returned by the method.
The partitioning is controlled by the <code class="docutils literal notranslate"><span class="pre">min_work_per_job</span></code> and <code class="docutils literal notranslate"><span class="pre">max_num_partitions</span></code>
arguments. For each group, ancestors are placed in a partition until the sum of their
lengths exceeds <code class="docutils literal notranslate"><span class="pre">min_work_per_job</span></code>, when a new partition is started. However, the
number of partitions is not allowed to exceed <code class="docutils literal notranslate"><span class="pre">max_num_partitions</span></code>. It is suggested
to set <code class="docutils literal notranslate"><span class="pre">max_num_partitions</span></code> to around 3-4x the number of worker nodes available,
and <code class="docutils literal notranslate"><span class="pre">min_work_per_job</span></code> to around 2,000,000 for a typical human data set.</p>
<p>Groups vs partitions is a point of common confusion. Note that groups of ancestors
are matched serially, and each group is split into partitions that can be
matched in parallel.</p>
<p>Each group is matched in turn, either by calling <a class="reference internal" href="api.html#tsinfer.match_ancestors_batch_groups" title="tsinfer.match_ancestors_batch_groups"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_groups()</span></code></a>
to match without partitioning, or by calling <a class="reference internal" href="api.html#tsinfer.match_ancestors_batch_group_partition" title="tsinfer.match_ancestors_batch_group_partition"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_group_partition()</span></code></a>
many times in parallel followed by a single call to <a class="reference internal" href="api.html#tsinfer.match_ancestors_batch_group_finalise" title="tsinfer.match_ancestors_batch_group_finalise"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_group_finalise()</span></code></a>.
Each call to <a class="reference internal" href="api.html#tsinfer.match_ancestors_batch_groups" title="tsinfer.match_ancestors_batch_groups"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_groups()</span></code></a> or <a class="reference internal" href="api.html#tsinfer.match_ancestors_batch_group_finalise" title="tsinfer.match_ancestors_batch_group_finalise"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_group_finalise()</span></code></a>
outputs the tree sequence to <code class="docutils literal notranslate"><span class="pre">work_dir</span></code>, which is then used by the next group. The length of
the <code class="docutils literal notranslate"><span class="pre">ancestor_grouping</span></code> in the metadata dictionary determines the group numbers that these methods
will need to be called for, and the length of the <code class="docutils literal notranslate"><span class="pre">partitions</span></code> list in each group determines
the number of calls to <a class="reference internal" href="api.html#tsinfer.match_ancestors_batch_group_partition" title="tsinfer.match_ancestors_batch_group_partition"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_group_partition()</span></code></a> that are needed (if any).</p>
<p><a class="reference internal" href="api.html#tsinfer.match_ancestors_batch_groups" title="tsinfer.match_ancestors_batch_groups"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_groups()</span></code></a> matches groups, without partitioning, from
<code class="docutils literal notranslate"><span class="pre">group_index_start</span></code> (inclusively) to <code class="docutils literal notranslate"><span class="pre">group_index_end</span></code> (exclusively). Combining
many groups into one call reduces the overhead from job submission and start
up times, but note on job failure the process can only be resumed from the
last <code class="docutils literal notranslate"><span class="pre">group_index_end</span></code>.</p>
<p>To match a single group in parallel, call <a class="reference internal" href="api.html#tsinfer.match_ancestors_batch_group_partition" title="tsinfer.match_ancestors_batch_group_partition"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_group_partition()</span></code></a>
once for each partition listed in the <code class="docutils literal notranslate"><span class="pre">ancestor_grouping[group_index]['partitions']</span></code> list,
incrementing <code class="docutils literal notranslate"><span class="pre">partition_index</span></code>. This will match the ancestors, placing the match data in
the <code class="docutils literal notranslate"><span class="pre">working_dir</span></code>. Once all are complete a single call to
<a class="reference internal" href="api.html#tsinfer.match_ancestors_batch_group_finalise" title="tsinfer.match_ancestors_batch_group_finalise"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_group_finalise()</span></code></a> will then insert the matches and
output the tree sequence to <code class="docutils literal notranslate"><span class="pre">work_dir</span></code>.</p>
<p>Each call to <a class="reference internal" href="api.html#tsinfer.match_ancestors_batch_groups" title="tsinfer.match_ancestors_batch_groups"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_groups()</span></code></a> and <a class="reference internal" href="api.html#tsinfer.match_ancestors_batch_group_finalise" title="tsinfer.match_ancestors_batch_group_finalise"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_group_finalise()</span></code></a> results in a tree sequence being written to <code class="docutils literal notranslate"><span class="pre">work_dir</span></code>.
These tree sequences are essentially checkpoints from with the batch matching workflow
can be resumed on job failure.</p>
<p>Finally after the final group, call <a class="reference internal" href="api.html#tsinfer.match_ancestors_batch_finalise" title="tsinfer.match_ancestors_batch_finalise"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_finalise()</span></code></a> to
combine the groups into a single tree sequence.</p>
<p>The partitioning in <code class="docutils literal notranslate"><span class="pre">metadata.json</span></code> does not have to be used for every group. As early groups are
not matching to a large tree sequence it is often faster to not partition the first half of the
groups, depending on job set up and queueing time on your cluster.</p>
<p>Calls to <a class="reference internal" href="api.html#tsinfer.match_ancestors_batch_group_partition" title="tsinfer.match_ancestors_batch_group_partition"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_group_partition()</span></code></a> will only use a single core, but
<a class="reference internal" href="api.html#tsinfer.match_ancestors_batch_groups" title="tsinfer.match_ancestors_batch_groups"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_groups()</span></code></a> will use as many cores as <code class="docutils literal notranslate"><span class="pre">num_threads</span></code> is set to.
Therefore this value and cluster resources requested should be scaled with the number of ancestors,
which can be read from the metadata dictionary.</p>
<p>As an example of how the API methods can be used together, suppose the metadata dictionary
created by <a class="reference internal" href="api.html#tsinfer.match_ancestors_batch_init" title="tsinfer.match_ancestors_batch_init"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_init()</span></code></a> contains the following:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="p">{</span>
    <span class="s2">&quot;ancestor_grouping&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="p">{</span><span class="s2">&quot;ancestors&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="o">...</span> <span class="mi">9</span><span class="p">],</span> <span class="s2">&quot;partitions&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">},</span>
        <span class="p">{</span>
          <span class="s2">&quot;ancestors&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">10</span><span class="p">,</span> <span class="o">...</span> <span class="mi">15</span><span class="p">],</span>
          <span class="s2">&quot;partitions&quot;</span><span class="p">:</span> <span class="p">[[</span><span class="mi">10</span><span class="p">,</span> <span class="mi">11</span><span class="p">,</span> <span class="mi">12</span><span class="p">],</span> <span class="p">[</span><span class="mi">13</span><span class="p">,</span> <span class="mi">14</span><span class="p">,</span> <span class="mi">15</span><span class="p">]]</span>
        <span class="p">},</span>
        <span class="p">{</span><span class="s2">&quot;ancestors&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">16</span><span class="p">,</span> <span class="o">...</span> <span class="mi">19</span><span class="p">],</span> <span class="s2">&quot;partitions&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">},</span>
        <span class="p">{</span><span class="s2">&quot;ancestors&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">20</span><span class="p">,</span> <span class="o">...</span> <span class="mi">25</span><span class="p">],</span> <span class="s2">&quot;partitions&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">},</span>
        <span class="p">{</span><span class="s2">&quot;ancestors&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">26</span><span class="p">,</span> <span class="o">...</span> <span class="mi">30</span><span class="p">],</span> <span class="s2">&quot;partitions&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">},</span>
        <span class="p">{</span>
          <span class="s2">&quot;ancestors&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">31</span><span class="p">,</span> <span class="o">...</span> <span class="mi">41</span><span class="p">],</span>
          <span class="s2">&quot;partitions&quot;</span><span class="p">:</span> <span class="p">[[</span><span class="mi">31</span><span class="p">,</span> <span class="mi">32</span><span class="p">,</span> <span class="mi">33</span><span class="p">,</span> <span class="mi">34</span><span class="p">,</span> <span class="mi">35</span><span class="p">,</span> <span class="mi">36</span><span class="p">],</span> <span class="p">[</span><span class="mi">37</span><span class="p">,</span> <span class="mi">38</span><span class="p">,</span> <span class="mi">39</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">41</span><span class="p">]]</span>
        <span class="p">},</span>
        <span class="p">{</span><span class="s2">&quot;ancestors&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">42</span><span class="p">,</span> <span class="o">...</span> <span class="mi">45</span><span class="p">],</span> <span class="s2">&quot;partitions&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">},</span>
        <span class="p">{</span><span class="s2">&quot;ancestors&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">46</span><span class="p">,</span> <span class="o">...</span> <span class="mi">50</span><span class="p">],</span> <span class="s2">&quot;partitions&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">},</span>
        <span class="p">{</span>
          <span class="s2">&quot;ancestors&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">51</span><span class="p">,</span> <span class="o">...</span> <span class="mi">65</span><span class="p">],</span>
          <span class="s2">&quot;partitions&quot;</span><span class="p">:</span> <span class="p">[</span>
            <span class="p">[</span><span class="mi">51</span><span class="p">,</span> <span class="mi">52</span><span class="p">,</span> <span class="mi">53</span><span class="p">,</span> <span class="mi">54</span><span class="p">],</span>
            <span class="p">[</span><span class="mi">55</span><span class="p">,</span> <span class="mi">56</span><span class="p">,</span> <span class="mi">57</span><span class="p">,</span> <span class="mi">58</span><span class="p">],</span>
            <span class="p">[</span><span class="mi">59</span><span class="p">,</span> <span class="mi">60</span><span class="p">,</span> <span class="mi">61</span><span class="p">,</span> <span class="mi">62</span><span class="p">,</span> <span class="mi">63</span><span class="p">,</span> <span class="mi">64</span><span class="p">,</span> <span class="mi">65</span><span class="p">]</span>
          <span class="p">]</span>
        <span class="p">},</span>
    <span class="p">]</span>
<span class="p">}</span>
</pre></div>
</div>
<p>Then the flow could look like the following diagram: (calls on the same horizontal line can be
done in parallel, note that method names are shortened):</p>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/example_flow.svg"><img alt="_images/example_flow.svg" src="_images/example_flow.svg" style="width: 80%;" />
</a>
</figure>
<p>Note that groups 1, 5 and 8 can be partitioned, but only groups 5 and 8 are actually partitioned in this example, as stated above partitioning for groups is optional. Groups 0-4 are matched in one call, groups 6 and 7 are matched in two calls, but
could have been matched in one. By splitting 6 and 7 the flow makes an additional resume point in the case of job failure at the cost of job start up and queueing time.</p>
</section>
<section id="sample-matching">
<h2>Sample matching<a class="headerlink" href="#sample-matching" title="Link to this heading">#</a></h2>
<p>Sample matching is far simpler than ancestor matching as it is essentially the same as a single group
of ancestors. There are three API methods that work together to enable distributed sample matching.</p>
<ol class="arabic simple">
<li><p><a class="reference internal" href="api.html#tsinfer.match_samples_batch_init" title="tsinfer.match_samples_batch_init"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_samples_batch_init()</span></code></a></p></li>
<li><p><a class="reference internal" href="api.html#tsinfer.match_samples_batch_partition" title="tsinfer.match_samples_batch_partition"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_samples_batch_partition()</span></code></a></p></li>
<li><p><a class="reference internal" href="api.html#tsinfer.match_samples_batch_finalise" title="tsinfer.match_samples_batch_finalise"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_samples_batch_finalise()</span></code></a></p></li>
</ol>
<p><a class="reference internal" href="api.html#tsinfer.match_samples_batch_init" title="tsinfer.match_samples_batch_init"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_samples_batch_init()</span></code></a> should be called to set up the batch matching and to determine the
groupings of samples. Similar to <a class="reference internal" href="api.html#tsinfer.match_ancestors_batch_init" title="tsinfer.match_ancestors_batch_init"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_init()</span></code></a> it has a <code class="docutils literal notranslate"><span class="pre">min_work_per_job</span></code>  argument to control the level of parallelism. The method writes a file
<code class="docutils literal notranslate"><span class="pre">metadata.json</span></code> to the directory <code class="docutils literal notranslate"><span class="pre">work_dir</span></code> that contains a JSON encoded dictionary with
configuration for later steps. This is also returned by the call. The <code class="docutils literal notranslate"><span class="pre">num_partitions</span></code> key in
this dictionary is the number of times <a class="reference internal" href="api.html#tsinfer.match_samples_batch_partition" title="tsinfer.match_samples_batch_partition"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_samples_batch_partition()</span></code></a> will need
to be called, with each partition index as the <code class="docutils literal notranslate"><span class="pre">partition_index</span></code> argument. These calls can happen
in parallel and write match data to the <code class="docutils literal notranslate"><span class="pre">work_dir</span></code> which is then used by
<a class="reference internal" href="api.html#tsinfer.match_samples_batch_finalise" title="tsinfer.match_samples_batch_finalise"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_samples_batch_finalise()</span></code></a> to output the tree sequence.</p>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="inference.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Inference overview</p>
      </div>
    </a>
    <a class="right-next"
       href="api.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">API Documentation</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-preparation">Data preparation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#ancestor-generation">Ancestor generation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#ancestor-matching">Ancestor matching</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#sample-matching">Sample matching</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Tskit Developers
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      Â© Copyright 2018.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>