
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Trait simulation &#8212; Tstrait manual</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=9c3e77be" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=1ae7504c"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'effect-size';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Genetic value" href="genetic.html" />
    <link rel="prev" title="Quick start" href="quick-start.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <script data-goatcounter="https://tskit.goatcounter.com/count" async src="//gc.zgo.at/count.js"></script>
</head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
  
    <p class="title logo__title">tstrait<br/>
version undefined
</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="intro.html">Introduction to tstrait</a></li>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installing tstrait</a></li>
<li class="toctree-l1"><a class="reference internal" href="quick-start.html">Quick start</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Simulation framework</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Trait simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="genetic.html">Genetic value</a></li>



<li class="toctree-l1"><a class="reference internal" href="environment.html">Environmental noise</a></li>
<li class="toctree-l1"><a class="reference internal" href="multi-trait.html">Multi-trait simulation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Interfaces</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="api.html">API Reference</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Extras</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="citation.html">Citing tstrait</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing to tstrait</a></li>
<li class="toctree-l1"><a class="reference internal" href="CHANGELOG.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference external" href="https://tskit.dev/community/">Community</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/effect-size.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Trait simulation</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#phenotype-model">Phenotype Model</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#effect-size-distribution">Effect Size Distribution</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#effect-size-simulation">Effect Size Simulation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#frequency-dependence">Frequency Dependence</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#specifying-causal-sites">Specifying Causal Sites</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="trait-simulation">
<span id="sim-trait-doc"></span><h1>Trait simulation<a class="headerlink" href="#trait-simulation" title="Link to this heading">#</a></h1>
<p>This page describes how tstrait simulates traits and how to specify the effect size
distribution.</p>
<p><strong>Learning Objectives</strong></p>
<p>After this effect size page, you will be able to:</p>
<ul class="simple">
<li><p>Understand the mathematical details of the phenotype model in tstrait</p></li>
<li><p>Understand how to specify distributions of simulated effect sizes</p></li>
<li><p>Understand how to simulate effect size in tstrait and read its output</p></li>
<li><p>Understand how to simulate effect sizes from the frequency dependence architecture</p></li>
</ul>
<section id="phenotype-model">
<span id="id1"></span><h2>Phenotype Model<a class="headerlink" href="#phenotype-model" title="Link to this heading">#</a></h2>
<p>tstrait simulates a vector of quantitative trait <span class="math notranslate nohighlight">\(y\)</span> from the following additive model,</p>
<div class="math notranslate nohighlight" id="equation-eq-phenotype-model">
<span class="eqno">(1)<a class="headerlink" href="#equation-eq-phenotype-model" title="Link to this equation">#</a></span>\[y = X\beta+\epsilon,\]</div>
<p>where <span class="math notranslate nohighlight">\(X\)</span> is the matrix that describes the number of causal alleles in each individual (the values
in each row will be <span class="math notranslate nohighlight">\(0\)</span>, <span class="math notranslate nohighlight">\(1\)</span>, or <span class="math notranslate nohighlight">\(2\)</span> in the diploid setting, for example), <span class="math notranslate nohighlight">\(\beta\)</span>
is the vector of effect sizes, and <span class="math notranslate nohighlight">\(\epsilon\)</span> is the vector of environmental noise. Environmental
noise is simulated from the following distribution,</p>
<div class="math notranslate nohighlight">
\[\epsilon\sim N\left(0,V_G\cdot\frac{1-h^2}{h^2} \right),\]</div>
<p>where <span class="math notranslate nohighlight">\(V_G=Var(X\beta)\)</span> and <span class="math notranslate nohighlight">\(h^2\)</span> is the narrow-sense heritability that is defined by the
user.</p>
<p>The genetic values (<span class="math notranslate nohighlight">\(X\beta\)</span>) are obtained by simply adding up over all the genomes in each individual,
regardless of ploidy.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<ul class="simple">
<li><p><a class="reference internal" href="genetic.html#genetic-value-doc"><span class="std std-ref">Genetic value</span></a> for obtaining the genetic value <span class="math notranslate nohighlight">\(X\beta\)</span>.</p></li>
<li><p><a class="reference internal" href="environment.html#environment-noise"><span class="std std-ref">Environmental noise</span></a> for simulating environmental noise <span class="math notranslate nohighlight">\(\epsilon\)</span>.</p></li>
<li><p><a class="reference internal" href="genetic.html#numericalise-genotype"><span class="std std-ref">Numericalisation of Genotypes</span></a> for modifying the numericalisation of genotypes.</p></li>
</ul>
</div>
<p>In this documentation, we will be describing how to simulate effect sizes in tstrait.</p>
</section>
<section id="effect-size-distribution">
<span id="effect-size-dist"></span><h2>Effect Size Distribution<a class="headerlink" href="#effect-size-distribution" title="Link to this heading">#</a></h2>
<p>The first step of trait simulation is to specify the distribution where the effect sizes will be
simulated. It can be specified in <code class="docutils literal notranslate"><span class="pre">distribution</span></code> input of <a class="reference internal" href="api.html#tstrait.trait_model" title="tstrait.trait_model"><code class="xref py py-func docutils literal notranslate"><span class="pre">trait_model()</span></code></a>. We also specify
other parameters of the distribution in the function as well. For example,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">tstrait</span>

<span class="n">model</span> <span class="o">=</span> <span class="n">tstrait</span><span class="o">.</span><span class="n">trait_model</span><span class="p">(</span><span class="n">distribution</span><span class="o">=</span><span class="s2">&quot;normal&quot;</span><span class="p">,</span> <span class="n">mean</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">var</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>sets a trait model, where the effect sizes are simulated from a normal distribution with
mean <span class="math notranslate nohighlight">\(0\)</span> and variance <span class="math notranslate nohighlight">\(1\)</span>. We can check the distribution name by using <code class="docutils literal notranslate"><span class="pre">.name</span></code> attribute
of a model instance.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model</span><span class="o">.</span><span class="n">name</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;normal&#39;
</pre></div>
</div>
</div>
</div>
<p>tstrait uses this distribution to simulate effect size <span class="math notranslate nohighlight">\(\beta\)</span> in the phenotype model described
in <a class="reference internal" href="#equation-eq-phenotype-model">(1)</a>.</p>
<p>The following effect size distributions are supported in tstrait, and please refer to links under
<strong>Details</strong> for details on the input and distribution.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="api.html#effect-size-distribution"><span class="std std-ref">Effect size distributions</span></a> for details on the supported distributions.</p>
</div>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Distribution</p></th>
<th class="head"><p>Input</p></th>
<th class="head"><p>Details</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">&quot;normal&quot;</span></code></p></td>
<td><p>Normal distribution</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">mean,</span> <span class="pre">var</span></code></p></td>
<td><p><a class="reference internal" href="api.html#tstrait.TraitModelNormal" title="tstrait.TraitModelNormal"><code class="xref py py-class docutils literal notranslate"><span class="pre">TraitModelNormal</span></code></a></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">&quot;t&quot;</span></code></p></td>
<td><p>Studentâ€™s t distribution</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">mean,</span> <span class="pre">var,</span> <span class="pre">df</span></code></p></td>
<td><p><a class="reference internal" href="api.html#tstrait.TraitModelT" title="tstrait.TraitModelT"><code class="xref py py-class docutils literal notranslate"><span class="pre">TraitModelT</span></code></a></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">&quot;fixed&quot;</span></code></p></td>
<td><p>Fixed value</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">value,</span> <span class="pre">random_sign</span></code></p></td>
<td><p><a class="reference internal" href="api.html#tstrait.TraitModelFixed" title="tstrait.TraitModelFixed"><code class="xref py py-class docutils literal notranslate"><span class="pre">TraitModelFixed</span></code></a></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">&quot;exponential&quot;</span></code></p></td>
<td><p>Exponential distribution</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">scale,</span> <span class="pre">random_sign</span></code></p></td>
<td><p><a class="reference internal" href="api.html#tstrait.TraitModelExponential" title="tstrait.TraitModelExponential"><code class="xref py py-class docutils literal notranslate"><span class="pre">TraitModelExponential</span></code></a></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">&quot;gamma&quot;</span></code></p></td>
<td><p>Gamma distribution</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">shape,</span> <span class="pre">scale,</span> <span class="pre">random_sign</span></code></p></td>
<td><p><a class="reference internal" href="api.html#tstrait.TraitModelGamma" title="tstrait.TraitModelGamma"><code class="xref py py-class docutils literal notranslate"><span class="pre">TraitModelGamma</span></code></a></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">&quot;multi_normal&quot;</span></code></p></td>
<td><p>Multivariate normal distribution</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">mean,</span> <span class="pre">cov</span></code></p></td>
<td><p><a class="reference internal" href="api.html#tstrait.TraitModelMultivariateNormal" title="tstrait.TraitModelMultivariateNormal"><code class="xref py py-class docutils literal notranslate"><span class="pre">tstrait.TraitModelMultivariateNormal</span></code></a></p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="effect-size-simulation">
<span id="effect-size-sim"></span><h2>Effect Size Simulation<a class="headerlink" href="#effect-size-simulation" title="Link to this heading">#</a></h2>
<p>Effect sizes can be simulated in tstrait by using <a class="reference internal" href="api.html#tstrait.sim_trait" title="tstrait.sim_trait"><code class="xref py py-func docutils literal notranslate"><span class="pre">tstrait.sim_trait()</span></code></a>. In the example below,
we will be simulating effect sizes of 5 causal sites from a simulated tree sequence data in
<a class="reference external" href="https://tskit.dev/msprime/docs/stable/intro.html#sec-intro" title="(in Project name not set)"><span class="xref std std-ref">msprime</span></a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">msprime</span>

<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="mi">10_000</span><span class="p">,</span>
    <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span>
    <span class="n">sequence_length</span><span class="o">=</span><span class="mi">100_000</span><span class="p">,</span>
    <span class="n">population_size</span><span class="o">=</span><span class="mi">10_000</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_mutations</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">1e-6</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>

<span class="n">trait_df</span> <span class="o">=</span> <span class="n">tstrait</span><span class="o">.</span><span class="n">sim_trait</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">num_causal</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">trait_df</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>position</th>
      <th>site_id</th>
      <th>effect_size</th>
      <th>causal_allele</th>
      <th>allele_freq</th>
      <th>trait_id</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>3440</td>
      <td>1123</td>
      <td>-0.536953</td>
      <td>C</td>
      <td>0.00005</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>47072</td>
      <td>15257</td>
      <td>0.581118</td>
      <td>G</td>
      <td>0.00375</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>50840</td>
      <td>16503</td>
      <td>0.364572</td>
      <td>T</td>
      <td>0.00700</td>
      <td>0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>75539</td>
      <td>24351</td>
      <td>0.294132</td>
      <td>T</td>
      <td>0.00095</td>
      <td>0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>95010</td>
      <td>30649</td>
      <td>0.028422</td>
      <td>G</td>
      <td>0.00510</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>The trait dataframe has 6 columns:</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>position</strong>: Position of sites that have causal allele in genome coordinates.</p></li>
<li><p><strong>site_id</strong>: Site IDs that have causal allele.</p></li>
<li><p><strong>effect_size</strong>: Simulated effect size of causal allele.</p></li>
<li><p><strong>causal_allele</strong>: Causal allele.</p></li>
<li><p><strong>allele_freq</strong>: Allele frequency of causal allele.</p></li>
<li><p><strong>trait_id</strong>: Trait ID that will be used in multi-trait simulation (See
<a class="reference internal" href="multi-trait.html#multi-trait"><span class="std std-ref">Multi-trait simulation</span></a> for details).</p></li>
</ul>
</div></blockquote>
<p>Next, we will be showing the distribution of simulated effect sizes by using
a normal distribution trait model with 1,000 causal sites.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">tstrait</span><span class="o">.</span><span class="n">trait_model</span><span class="p">(</span><span class="n">distribution</span><span class="o">=</span><span class="s2">&quot;normal&quot;</span><span class="p">,</span> <span class="n">mean</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">var</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">trait_df</span> <span class="o">=</span> <span class="n">tstrait</span><span class="o">.</span><span class="n">sim_trait</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">num_causal</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">trait_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">trait_df</span><span class="p">[</span><span class="s2">&quot;effect_size&quot;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">40</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Simulated Effect Size&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/e46dd90d007db2e71a388d92b9b4a93f2410abebf887970d28e9dc000064350c.png" src="_images/e46dd90d007db2e71a388d92b9b4a93f2410abebf887970d28e9dc000064350c.png" />
</div>
</div>
<p>We see that the simulated effect sizes are approximately following a <span class="math notranslate nohighlight">\(N(0,1)\)</span>
distribution, which coincides with the fact that we had specified a normal
distribution trait model with mean 0 and variance 1.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The site ID represents the IDs of causal sites, and information regarding the site can be
extracted by using <code class="docutils literal notranslate"><span class="pre">.site()</span></code>.</p>
</div>
<p>The below code is used to extract information of site with ID 0 from <code class="docutils literal notranslate"><span class="pre">ts</span></code> tree sequence.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Extract information of site with ID 0</span>
<span class="n">ts</span><span class="o">.</span><span class="n">site</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Site(id=0, position=1.0, ancestral_state=&#39;C&#39;, mutations=[Mutation(id=0, site=0, node=40280, derived_state=&#39;T&#39;, parent=-1, metadata=b&#39;&#39;, time=18036.72830221663, edge=41324)], metadata=b&#39;&#39;)
</pre></div>
</div>
</div>
</div>
<p>The details of sites in tree sequences can be found <a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-site-table-definition" title="(in Project name not set)"><span class="xref std std-ref">here</span></a>.</p>
</section>
<section id="frequency-dependence">
<span id="trait-frequency-dependence"></span><h2>Frequency Dependence<a class="headerlink" href="#frequency-dependence" title="Link to this heading">#</a></h2>
<p>Tstrait supports frequency dependence simulation. It has been shown that rare variants
have increased effect sizes compared with common variants
<a class="reference external" href="https://doi.org/10.1038/ng.3865">Speed et al. (2017)</a>, so more realistic simulations
can be made possible by increasing the effect size on rarer variants. The <code class="docutils literal notranslate"><span class="pre">alpha</span></code>
parameter in <a class="reference internal" href="api.html#tstrait.sim_phenotype" title="tstrait.sim_phenotype"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_phenotype()</span></code></a> and <a class="reference internal" href="api.html#tstrait.sim_trait" title="tstrait.sim_trait"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_trait()</span></code></a> are used to control
the degree of frequency dependence on simulated effect sizes.</p>
<p>In the frequency dependence model, the following value is multiplied to the effect size:</p>
<div class="math notranslate nohighlight" id="equation-eq-freq-dep">
<span class="eqno">(2)<a class="headerlink" href="#equation-eq-freq-dep" title="Link to this equation">#</a></span>\[\Big[\sqrt{2p(1-p)}\Big]^\alpha\]</div>
<p>In the above expression, <span class="math notranslate nohighlight">\(p\)</span> is the frequency of the causal allele, and
<span class="math notranslate nohighlight">\(\alpha\)</span> is the <code class="docutils literal notranslate"><span class="pre">alpha</span></code> input of <a class="reference internal" href="api.html#tstrait.sim_phenotype" title="tstrait.sim_phenotype"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_phenotype()</span></code></a> and
<a class="reference internal" href="api.html#tstrait.sim_trait" title="tstrait.sim_trait"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_trait()</span></code></a>. Putting a negative <span class="math notranslate nohighlight">\(\alpha\)</span> value increases the
magnitude of effect sizes on rare variants.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The default <code class="docutils literal notranslate"><span class="pre">alpha</span></code> parameter in <a class="reference internal" href="api.html#tstrait.sim_phenotype" title="tstrait.sim_phenotype"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_phenotype()</span></code></a> and
<a class="reference internal" href="api.html#tstrait.sim_trait" title="tstrait.sim_trait"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_trait()</span></code></a> are 0, and frequency dependent model is not used. Please
ignore the <code class="docutils literal notranslate"><span class="pre">alpha</span></code> parameter if you are not interested in implementing the
frequency dependent model.</p>
</div>
<p>In the below example, we will be demonstrating how <code class="docutils literal notranslate"><span class="pre">alpha</span></code> influences the simulated
effect sizes by using a simulated tree sequence with 10,000 individuals.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="mi">10_000</span><span class="p">,</span>
    <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span>
    <span class="n">sequence_length</span><span class="o">=</span><span class="mi">1_000_000</span><span class="p">,</span>
    <span class="n">population_size</span><span class="o">=</span><span class="mi">10_000</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_mutations</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">303</span><span class="p">)</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">tstrait</span><span class="o">.</span><span class="n">trait_model</span><span class="p">(</span><span class="n">distribution</span><span class="o">=</span><span class="s2">&quot;normal&quot;</span><span class="p">,</span> <span class="n">mean</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">var</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We will first simulate effect sizes by using the non-frequency dependent model
(<code class="docutils literal notranslate"><span class="pre">alpha</span></code> = 0), and visualize the simulated effect sizes. This is the default
<code class="docutils literal notranslate"><span class="pre">alpha</span></code> parameter in <a class="reference internal" href="api.html#tstrait.sim_trait" title="tstrait.sim_trait"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_trait()</span></code></a>, so there is no need for you to specify
it in your trait simulation.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># trait.sim_trait(ts, num_causal=1000, model=model, random_seed=1)</span>
<span class="c1"># also works here</span>
<span class="n">trait_df</span> <span class="o">=</span> <span class="n">tstrait</span><span class="o">.</span><span class="n">sim_trait</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">num_causal</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span> 
                             <span class="n">alpha</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">trait_df</span><span class="o">.</span><span class="n">allele_freq</span><span class="p">,</span> <span class="n">trait_df</span><span class="o">.</span><span class="n">effect_size</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Allele frequency&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Effect size&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Non-frequency dependent model, alpha = 0&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/0337615f46c70720c5598427117ece18a115a00070246b4759a0deba5098831a.png" src="_images/0337615f46c70720c5598427117ece18a115a00070246b4759a0deba5098831a.png" />
</div>
</div>
<p>We see no relationship between allele frequency and effect size. As a next example,
we will be simulating effect sizes with <code class="docutils literal notranslate"><span class="pre">alpha</span></code> = -1/2.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">trait_df</span> <span class="o">=</span> <span class="n">tstrait</span><span class="o">.</span><span class="n">sim_trait</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">num_causal</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span> 
                             <span class="n">alpha</span><span class="o">=-</span><span class="mi">1</span><span class="o">/</span><span class="mi">2</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">trait_df</span><span class="o">.</span><span class="n">allele_freq</span><span class="p">,</span> <span class="n">trait_df</span><span class="o">.</span><span class="n">effect_size</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Allele frequency&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Effect size&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Frequency dependent model, alpha = -1/2&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/77ce91557449d59da87f69c498c78c23de90ef759be3f321f25618632a1f7137.png" src="_images/77ce91557449d59da87f69c498c78c23de90ef759be3f321f25618632a1f7137.png" />
</div>
</div>
<p>We see that rarer variants have increased effect sizes, as expected from
the mathematical expression given in <a class="reference internal" href="#equation-eq-freq-dep">(2)</a>.</p>
</section>
<section id="specifying-causal-sites">
<span id="trait-causal-sites"></span><h2>Specifying Causal Sites<a class="headerlink" href="#specifying-causal-sites" title="Link to this heading">#</a></h2>
<p>Instead of specifying the number of causal sites, users can directly provide the
causal sites as an input of <a class="reference internal" href="api.html#tstrait.sim_phenotype" title="tstrait.sim_phenotype"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_phenotype()</span></code></a> and <a class="reference internal" href="api.html#tstrait.sim_trait" title="tstrait.sim_trait"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_trait()</span></code></a>.
When <code class="docutils literal notranslate"><span class="pre">causal_sites</span></code> argument of these functions are provided, tstrait will use
the inputted site IDs as causal sites, instead of randomly selecting causal
site IDs.</p>
<p>Example:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">trait_df</span> <span class="o">=</span> <span class="n">tstrait</span><span class="o">.</span><span class="n">sim_trait</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">causal_sites</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span> <span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span> 
                             <span class="n">alpha</span><span class="o">=-</span><span class="mi">1</span><span class="o">/</span><span class="mi">2</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">trait_df</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>position</th>
      <th>site_id</th>
      <th>effect_size</th>
      <th>causal_allele</th>
      <th>allele_freq</th>
      <th>trait_id</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>175</td>
      <td>0</td>
      <td>0.775715</td>
      <td>G</td>
      <td>0.02010</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1567</td>
      <td>3</td>
      <td>2.634851</td>
      <td>A</td>
      <td>0.00475</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>1713</td>
      <td>4</td>
      <td>0.910507</td>
      <td>G</td>
      <td>0.00875</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="quick-start.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Quick start</p>
      </div>
    </a>
    <a class="right-next"
       href="genetic.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Genetic value</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#phenotype-model">Phenotype Model</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#effect-size-distribution">Effect Size Distribution</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#effect-size-simulation">Effect Size Simulation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#frequency-dependence">Frequency Dependence</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#specifying-causal-sites">Specifying Causal Sites</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Tskit Developers
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      Â© Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>