
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Genetic value &#8212; Tstrait manual</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'genetic';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Environmental noise" href="environment.html" />
    <link rel="prev" title="Trait simulation" href="effect-size.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <script data-goatcounter="https://tskit.goatcounter.com/count" async src="//gc.zgo.at/count.js"></script>
</head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
  
    <p class="title logo__title">tstrait<br/>
version undefined
</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="intro.html">Introduction to tstrait</a></li>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installing tstrait</a></li>
<li class="toctree-l1"><a class="reference internal" href="quick-start.html">Quick start</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Simulation framework</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="effect-size.html">Trait simulation</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Genetic value</a></li>



<li class="toctree-l1"><a class="reference internal" href="environment.html">Environmental noise</a></li>
<li class="toctree-l1"><a class="reference internal" href="multi-trait.html">Multi-trait simulation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Interfaces</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="api.html">API Reference</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Extras</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="citation.html">Citing tstrait</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing to tstrait</a></li>
<li class="toctree-l1"><a class="reference internal" href="CHANGELOG.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference external" href="https://tskit.dev/community/">Community</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/genetic.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Genetic value</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Genetic value</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#algorithm-overview">Algorithm Overview</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#genetic-value-computation">Genetic value computation</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#user-defined-trait-dataframe">User defined trait dataframe</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#dataframe-requirements">Dataframe requirements</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#columns">Columns</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#data-requirements">Data requirements</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#numericalisation-of-genotypes">Numericalisation of Genotypes</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#normalise-genetic-value">Normalise Genetic Value</a></li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="genetic-value">
<span id="genetic-value-doc"></span><h1>Genetic value<a class="headerlink" href="#genetic-value" title="Link to this heading">#</a></h1>
<p>This page will describe how to simulate genetic values by using <a class="reference internal" href="api.html#tstrait.genetic_value" title="tstrait.genetic_value"><code class="xref py py-func docutils literal notranslate"><span class="pre">genetic_value()</span></code></a> function. Genetic
values of individuals are described by <span class="math notranslate nohighlight">\(X\beta\)</span>, where <span class="math notranslate nohighlight">\(X\)</span> is the matrix that describes
the number of causal alleles in each individual and <span class="math notranslate nohighlight">\(\beta\)</span> is the vector of effect sizes. See
<a class="reference internal" href="effect-size.html#phenotype-model"><span class="std std-ref">Phenotype Model</span></a> for mathematical details on the quantitative trait model.</p>
<p><strong>Learning Objectives</strong></p>
<p>After this genetic value page, you will be able to:</p>
<ul class="simple">
<li><p>Understand how to generate genetic value in tstrait</p></li>
<li><p>Understand how to use the user’s defined effect sizes to generate genetic values</p></li>
</ul>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="algorithm-overview">
<h1>Algorithm Overview<a class="headerlink" href="#algorithm-overview" title="Link to this heading">#</a></h1>
<p>Genetic value is computed in tstrait by using the trait information in the input trait dataframe.
It uses a tree traversal algorithm to count the number of causal allele in each individual and adds
the corresponding effect size to individual’s genetic value.</p>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Link to this heading">#</a></h2>
<p>The input of <a class="reference internal" href="api.html#tstrait.genetic_value" title="tstrait.genetic_value"><code class="xref py py-func docutils literal notranslate"><span class="pre">genetic_value()</span></code></a> function is the following:</p>
<dl class="simple myst">
<dt>ts</dt><dd><p>Tree sequence input.</p>
</dd>
<dt>trait_df</dt><dd><p>Trait dataframe that is described in <a class="reference internal" href="effect-size.html#effect-size-sim"><span class="std std-ref">Effect Size Simulation</span></a>. There are some requirements for the trait
dataframe input, which are described in <a class="reference internal" href="#req-trait-df"><span class="std std-ref">Dataframe requirements</span></a>.</p>
</dd>
</dl>
<p>The details of the parameters and how they influence the genetic value simulation are described in detail
below.</p>
</section>
<section id="genetic-value-computation">
<h2>Genetic value computation<a class="headerlink" href="#genetic-value-computation" title="Link to this heading">#</a></h2>
<p>In this example, we will be showing how to compute genetic values by using the simulated trait dataframe
in <a class="reference internal" href="api.html#tstrait.genetic_value" title="tstrait.genetic_value"><code class="xref py py-func docutils literal notranslate"><span class="pre">genetic_value()</span></code></a>.</p>
<p>We will be simulating a sample tree sequence data with 5 individuals by using <a class="reference external" href="https://tskit.dev/msprime/docs/stable/intro.html#sec-intro" title="(in Project name not set)"><span class="xref std std-ref">msprime</span></a>,
and we will simulate a trait dataframe with 3 causal sites. The dataframe output of <a class="reference internal" href="api.html#tstrait.sim_trait" title="tstrait.sim_trait"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_trait()</span></code></a>
can be automatically used as an input of <a class="reference internal" href="api.html#tstrait.genetic_value" title="tstrait.genetic_value"><code class="xref py py-func docutils literal notranslate"><span class="pre">genetic_value()</span></code></a>, so there is no need to worry about input
data requirements.
data requirements.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<ul class="simple">
<li><p><a class="reference external" href="https://tskit.dev/msprime/docs/stable/intro.html#sec-intro" title="(in Project name not set)"><span class="xref std std-ref">msprime</span></a> for simulating whole genome in tree sequence data.</p></li>
<li><p><a class="reference internal" href="effect-size.html#sim-trait-doc"><span class="std std-ref">Trait simulation</span></a> for simulating trait dataframe in tstrait.</p></li>
</ul>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">msprime</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">tstrait</span>

<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
    <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span>
    <span class="n">sequence_length</span><span class="o">=</span><span class="mi">100_000</span><span class="p">,</span>
    <span class="n">population_size</span><span class="o">=</span><span class="mi">10_000</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_mutations</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">101</span><span class="p">)</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">tstrait</span><span class="o">.</span><span class="n">trait_model</span><span class="p">(</span><span class="n">distribution</span><span class="o">=</span><span class="s2">&quot;normal&quot;</span><span class="p">,</span> <span class="n">mean</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">var</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">trait_df</span> <span class="o">=</span> <span class="n">tstrait</span><span class="o">.</span><span class="n">sim_trait</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">num_causal</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="n">trait_df</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>position</th>
      <th>site_id</th>
      <th>effect_size</th>
      <th>causal_allele</th>
      <th>allele_freq</th>
      <th>trait_id</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>1901</td>
      <td>3</td>
      <td>0.420445</td>
      <td>G</td>
      <td>0.2</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>69303</td>
      <td>89</td>
      <td>1.136047</td>
      <td>G</td>
      <td>0.6</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>82287</td>
      <td>107</td>
      <td>0.109706</td>
      <td>A</td>
      <td>0.1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Please refer to <a class="reference internal" href="effect-size.html#effect-size-sim"><span class="std std-ref">Effect Size Simulation</span></a> for details on the trait dataframe output and effect size
simulation by using <a class="reference internal" href="api.html#tstrait.sim_trait" title="tstrait.sim_trait"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_trait()</span></code></a>.</p>
<p>Next, we will be using the trait dataframe to generate genetic values by using <a class="reference internal" href="api.html#tstrait.genetic_value" title="tstrait.genetic_value"><code class="xref py py-func docutils literal notranslate"><span class="pre">genetic_value()</span></code></a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">genetic_df</span> <span class="o">=</span> <span class="n">tstrait</span><span class="o">.</span><span class="n">genetic_value</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">trait_df</span><span class="p">)</span>
<span class="n">genetic_df</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>trait_id</th>
      <th>individual_id</th>
      <th>genetic_value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>0</td>
      <td>2.692538</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0</td>
      <td>1</td>
      <td>2.381799</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0</td>
      <td>2</td>
      <td>0.420445</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0</td>
      <td>3</td>
      <td>1.136047</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0</td>
      <td>4</td>
      <td>1.136047</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>The output of <a class="reference internal" href="api.html#tstrait.genetic_value" title="tstrait.genetic_value"><code class="xref py py-func docutils literal notranslate"><span class="pre">genetic_value()</span></code></a> is a <a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a> object that describes
the genetic values of individuals.</p>
<p>The genetic value dataframe includes the following columns:</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>trait_id</strong>: Trait ID that will be used in multi-trait simulation.</p></li>
<li><p><strong>individual_id</strong>: Individual ID inside the tree sequence input.</p></li>
<li><p><strong>genetic_value</strong>: Simulated genetic values.</p></li>
</ul>
</div></blockquote>
<section id="user-defined-trait-dataframe">
<span id="genetic-user"></span><h3>User defined trait dataframe<a class="headerlink" href="#user-defined-trait-dataframe" title="Link to this heading">#</a></h3>
<p>We will be demonstrating how to simulate genetic values from a user defined trait dataframe.</p>
</section>
<section id="dataframe-requirements">
<span id="req-trait-df"></span><h3>Dataframe requirements<a class="headerlink" href="#dataframe-requirements" title="Link to this heading">#</a></h3>
<p>There are some requirements for the trait dataframe input:</p>
<section id="columns">
<h4>Columns<a class="headerlink" href="#columns" title="Link to this heading">#</a></h4>
<p>The following columns must be included in <code class="docutils literal notranslate"><span class="pre">trait_df</span></code>:</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>site_id</strong>: Site IDs that have causal allele.</p></li>
<li><p><strong>effect_size</strong>: Simulated effect size of causal allele.</p></li>
<li><p><strong>causal_allele</strong>: Causal allele.</p></li>
<li><p><strong>trait_id</strong>: Trait ID.</p></li>
</ul>
</div></blockquote>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The IDs in the trait_id column must start from 0 and be a consecutive integer. The IDs in the site_id
column must be non-decreasing.</p>
</div>
<p>If you are simulating single traits, <code class="docutils literal notranslate"><span class="pre">trait_id</span></code> must be an array of zeros. The method
<a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html#pandas.DataFrame.sort_values" title="(in pandas v2.3.3)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pandas.DataFrame.sort_values()</span></code></a> can be used to sort the dataframe based on <code class="docutils literal notranslate"><span class="pre">site_id</span></code>.</p>
</section>
<section id="data-requirements">
<h4>Data requirements<a class="headerlink" href="#data-requirements" title="Link to this heading">#</a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Site IDs in <strong>site_id</strong> column must be sorted in an ascending order. Please refer to
<a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html#pandas.DataFrame.sort_values" title="(in pandas v2.3.3)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pandas.DataFrame.sort_values()</span></code></a> for details on sorting values in a
<a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a>.</p></li>
<li><p>Trait IDs in <strong>trait_id</strong> column must start from zero and be consecutive.</p></li>
</ul>
</div></blockquote>
<p>We will be demonstrating how to simulate genetic values with 1 causal site.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>

<span class="n">data</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;site_id&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s2">&quot;effect_size&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;trait_id&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">],</span>
        <span class="s2">&quot;causal_allele&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;1&quot;</span><span class="p">]}</span>
<span class="n">trait_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="n">trait_df</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>site_id</th>
      <th>effect_size</th>
      <th>trait_id</th>
      <th>causal_allele</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
    <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span>
    <span class="n">sequence_length</span><span class="o">=</span><span class="mi">100_000</span><span class="p">,</span>
    <span class="n">population_size</span><span class="o">=</span><span class="mi">10_000</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_mutations</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">1e-6</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s2">&quot;binary&quot;</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">genetic_df</span> <span class="o">=</span> <span class="n">tstrait</span><span class="o">.</span><span class="n">genetic_value</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">trait_df</span><span class="p">)</span>
<span class="n">genetic_df</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>trait_id</th>
      <th>individual_id</th>
      <th>genetic_value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0</td>
      <td>1</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0</td>
      <td>2</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0</td>
      <td>3</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0</td>
      <td>4</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>As seen in the genetic value dataframe, the genetic values of individuals are multiples
of 1. This is because 1 is the only effect size in the effect size dataframe, and the
<a class="reference internal" href="api.html#tstrait.genetic_value" title="tstrait.genetic_value"><code class="xref py py-func docutils literal notranslate"><span class="pre">genetic_value()</span></code></a> function is counting the number of causal alleles of site 0
inside the individual.</p>
<p>Next, we will be inputting an effect size dataframe with multiple causal sites.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;site_id&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span> <span class="s2">&quot;effect_size&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">100</span><span class="p">],</span>
        <span class="s2">&quot;causal_allele&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;1&quot;</span><span class="p">],</span> <span class="s2">&quot;trait_id&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]}</span>
<span class="n">trait_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="n">trait_df</span> <span class="o">=</span> <span class="n">trait_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;site_id&quot;</span><span class="p">])</span>
<span class="n">trait_df</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>site_id</th>
      <th>effect_size</th>
      <th>causal_allele</th>
      <th>trait_id</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2</td>
      <td>10</td>
      <td>1</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>4</td>
      <td>100</td>
      <td>1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>The site IDs in <code class="docutils literal notranslate"><span class="pre">&quot;site_id&quot;</span></code> column in the trait dataframe must be sorted in an
ascending order, so it is suggested to use <code class="docutils literal notranslate"><span class="pre">.sort_values(by=['site_id'])</span></code>
code before inputting it inside the <a class="reference internal" href="api.html#tstrait.genetic_value" title="tstrait.genetic_value"><code class="xref py py-func docutils literal notranslate"><span class="pre">genetic_value()</span></code></a> function.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">genetic_df</span> <span class="o">=</span> <span class="n">tstrait</span><span class="o">.</span><span class="n">genetic_value</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">trait_df</span><span class="p">)</span>
<span class="n">genetic_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>trait_id</th>
      <th>individual_id</th>
      <th>genetic_value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0</td>
      <td>1</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0</td>
      <td>2</td>
      <td>10.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0</td>
      <td>3</td>
      <td>100.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0</td>
      <td>4</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
</section>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="numericalisation-of-genotypes">
<span id="numericalise-genotype"></span><h1>Numericalisation of Genotypes<a class="headerlink" href="#numericalisation-of-genotypes" title="Link to this heading">#</a></h1>
<p>The genotypes are numericalised as the number of causal alleles in each
individual (Please see <a class="reference internal" href="effect-size.html#phenotype-model"><span class="std std-ref">Phenotype Model</span></a> for mathematical details on the phenotype
model), but it would be possible to change the numericalisation by modifying the
genetic value dataframe based on the effect size dataframe. For example, in the
diploid setting, if you are interested in simulating phenotypes from the genotype
<span class="math notranslate nohighlight">\((aa=-1, Aa=0, AA=1)\)</span>, where <span class="math notranslate nohighlight">\(A\)</span> is the causal allele, we can simply subtract the
sum of effect sizes from the genetic value.</p>
<p>In the following, we will provide a quick example on how genetic values are being computed
based on the mutation information and effect sizes. We will assume that the <code class="docutils literal notranslate"><span class="pre">A</span></code> allele
represents the causal allele in site 1 and the <code class="docutils literal notranslate"><span class="pre">B</span></code> allele represents the causal allele in
site 2. The effect sizes will be encoded as <span class="math notranslate nohighlight">\(\beta_1\)</span> and <span class="math notranslate nohighlight">\(\beta_2\)</span> for sites 1 and 2,
respectively. The genotype and genetic value of individuals based on the tstrait’s
numericalisation of genotypes <span class="math notranslate nohighlight">\((aa=0, Aa=1, AA=2)\)</span> are shown in the table below:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p><strong>Individual ID</strong></p></th>
<th class="head"><p><strong>Site 1</strong></p></th>
<th class="head"><p><strong>Site 2</strong></p></th>
<th class="head"><p><strong>Genetic Value</strong></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>Aa</p></td>
<td><p>BB</p></td>
<td><p><span class="math notranslate nohighlight">\(\beta_1+2\beta_2\)</span></p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>aa</p></td>
<td><p>Bb</p></td>
<td><p><span class="math notranslate nohighlight">\(\beta_2\)</span></p></td>
</tr>
</tbody>
</table>
</div>
<p>If we modify the numericalisation of genotypes to be <span class="math notranslate nohighlight">\((aa=-1, Aa=0, AA=1)\)</span>, we get the following:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p><strong>Individual ID</strong></p></th>
<th class="head"><p><strong>Site 1</strong></p></th>
<th class="head"><p><strong>Site 2</strong></p></th>
<th class="head"><p><strong>Genetic Value</strong></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>Aa</p></td>
<td><p>BB</p></td>
<td><p><span class="math notranslate nohighlight">\(\beta_2\)</span></p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>aa</p></td>
<td><p>Bb</p></td>
<td><p><span class="math notranslate nohighlight">\(-\beta_1\)</span></p></td>
</tr>
</tbody>
</table>
</div>
<p>When we compare these outputs, we see that the genetic value of individuals in the new encoding
<span class="math notranslate nohighlight">\((aa=-1, Aa=0, AA=1)\)</span> is obtained by subtracting the sum of effect sizes <span class="math notranslate nohighlight">\((\beta_1+\beta_2)\)</span>
from the original genetic value.</p>
<p>This can be done in the following example:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">trait_df</span> <span class="o">=</span> <span class="n">tstrait</span><span class="o">.</span><span class="n">sim_trait</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">num_causal</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="n">genetic_df</span> <span class="o">=</span> <span class="n">tstrait</span><span class="o">.</span><span class="n">genetic_value</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">trait_df</span><span class="p">)</span>

<span class="c1"># The original dataframe</span>
<span class="n">genetic_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>trait_id</th>
      <th>individual_id</th>
      <th>genetic_value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>0</td>
      <td>0.639858</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0</td>
      <td>1</td>
      <td>1.976937</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0</td>
      <td>2</td>
      <td>2.802245</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0</td>
      <td>3</td>
      <td>3.112984</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0</td>
      <td>4</td>
      <td>3.222690</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">genetic_df</span><span class="p">[</span><span class="s2">&quot;genetic_value&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genetic_df</span><span class="p">[</span><span class="s2">&quot;genetic_value&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">trait_df</span><span class="p">[</span><span class="s2">&quot;effect_size&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="c1"># New dataframe</span>
<span class="n">genetic_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>trait_id</th>
      <th>individual_id</th>
      <th>genetic_value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>0</td>
      <td>-1.026340</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0</td>
      <td>1</td>
      <td>0.310739</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0</td>
      <td>2</td>
      <td>1.136047</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0</td>
      <td>3</td>
      <td>1.446785</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0</td>
      <td>4</td>
      <td>1.556492</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>The new genetic value dataframe can be used in <a class="reference internal" href="api.html#tstrait.sim_env" title="tstrait.sim_env"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_env()</span></code></a> to simulate phenotypes.</p>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="normalise-genetic-value">
<span id="id1"></span><h1>Normalise Genetic Value<a class="headerlink" href="#normalise-genetic-value" title="Link to this heading">#</a></h1>
<p>The computed genetic values can be scaled by using the <a class="reference internal" href="api.html#tstrait.normalise_genetic_value" title="tstrait.normalise_genetic_value"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalise_genetic_value()</span></code></a> function. The function
will first normalise the genetic value by subtracting the mean of the input genetic value and divide it
by the standard devitation of the input genetic value.
Afterwards, it scales the normalised genetic value based on the mean and variance input.
The output of <a class="reference internal" href="api.html#tstrait.normalise_genetic_value" title="tstrait.normalise_genetic_value"><code class="xref py py-func docutils literal notranslate"><span class="pre">normalise_genetic_value()</span></code></a> is a <a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a> object with the scaled genetic values.
It is suggested to use this function on the genetic value dataframe that is obtained by
<a class="reference internal" href="api.html#tstrait.genetic_value" title="tstrait.genetic_value"><code class="xref py py-func docutils literal notranslate"><span class="pre">genetic_value()</span></code></a>, and use the normalised genetic value dataframe to simulate phenotypes
by using <a class="reference internal" href="api.html#tstrait.sim_env" title="tstrait.sim_env"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_env()</span></code></a>.</p>
<p>An example usage of this function is shown below:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mean</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">var</span> <span class="o">=</span> <span class="mi">1</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="mi">10_000</span><span class="p">,</span>
    <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span>
    <span class="n">sequence_length</span><span class="o">=</span><span class="mi">100_000</span><span class="p">,</span>
    <span class="n">population_size</span><span class="o">=</span><span class="mi">10_000</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_mutations</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">1e-7</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1001</span><span class="p">)</span>
<span class="n">trait_df</span> <span class="o">=</span> <span class="n">tstrait</span><span class="o">.</span><span class="n">sim_trait</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">num_causal</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">500</span><span class="p">)</span>
<span class="n">genetic_df</span> <span class="o">=</span> <span class="n">tstrait</span><span class="o">.</span><span class="n">genetic_value</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">trait_df</span><span class="p">)</span>
<span class="n">normalised_df</span> <span class="o">=</span> <span class="n">tstrait</span><span class="o">.</span><span class="n">normalise_genetic_value</span><span class="p">(</span><span class="n">genetic_df</span><span class="p">,</span> <span class="n">mean</span><span class="o">=</span><span class="n">mean</span><span class="p">,</span> <span class="n">var</span><span class="o">=</span><span class="n">var</span><span class="p">)</span>
<span class="n">normalised_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>individual_id</th>
      <th>trait_id</th>
      <th>genetic_value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>0</td>
      <td>0.651249</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>0</td>
      <td>-0.530608</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>0</td>
      <td>0.534444</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>0</td>
      <td>-0.637727</td>
    </tr>
    <tr>
      <th>4</th>
      <td>4</td>
      <td>0</td>
      <td>-1.807479</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>The distribution of the normalised genetic value is shown below, and wee that the mean and variance
of the normalised genetic values are 0 and 1.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">normalised_df</span><span class="p">[</span><span class="s2">&quot;genetic_value&quot;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">40</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Normalised Genetic Value&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/d4850a066136b40e973116376bd408c07fe90dcab1532b63acba719710cbaa39.png" src="_images/d4850a066136b40e973116376bd408c07fe90dcab1532b63acba719710cbaa39.png" />
</div>
</div>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="effect-size.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Trait simulation</p>
      </div>
    </a>
    <a class="right-next"
       href="environment.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Environmental noise</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Genetic value</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#algorithm-overview">Algorithm Overview</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#genetic-value-computation">Genetic value computation</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#user-defined-trait-dataframe">User defined trait dataframe</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#dataframe-requirements">Dataframe requirements</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#columns">Columns</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#data-requirements">Data requirements</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#numericalisation-of-genotypes">Numericalisation of Genotypes</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#normalise-genetic-value">Normalise Genetic Value</a></li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Tskit Developers
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>