
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Ancestry simulations &#8212; Msprime manual</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'ancestry';</script>
    <link rel="icon" href="_static/favicon.ico"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Mutation simulations" href="mutations.html" />
    <link rel="prev" title="Installation" href="installation.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <script data-goatcounter="https://tskit.goatcounter.com/count" async src="//gc.zgo.at/count.js"></script>
</head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/msprime_logo.svg" class="logo__image only-light" alt="Msprime manual - Home"/>
    <script>document.write(`<img src="_static/msprime_logo.svg" class="logo__image only-dark" alt="Msprime manual - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Running simulations</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Ancestry simulations</a></li>
<li class="toctree-l1"><a class="reference internal" href="mutations.html">Mutation simulations</a></li>
<li class="toctree-l1"><a class="reference internal" href="demography.html">Demographic models</a></li>
<li class="toctree-l1"><a class="reference internal" href="replication.html">Randomness and replication</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Interfaces</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="api.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">Command line interface</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Utilities</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="rate_maps.html">Rate Maps</a></li>
<li class="toctree-l1"><a class="reference internal" href="pedigrees.html">Pedigrees</a></li>
<li class="toctree-l1"><a class="reference internal" href="likelihoods.html">Computing likelihoods</a></li>
<li class="toctree-l1"><a class="reference internal" href="logging.html">Logging</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Miscellaneous</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="legacy.html">Legacy (version 0.x) APIs</a></li>
<li class="toctree-l1"><a class="reference internal" href="switch_from_other_simulators.html">Switching from other simulators</a></li>
<li class="toctree-l1"><a class="reference internal" href="development.html">Development</a></li>
<li class="toctree-l1"><a class="reference internal" href="CITATION.html">Citing msprime</a></li>
<li class="toctree-l1"><a class="reference internal" href="CHANGELOG.html">Changelog</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/ancestry.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Ancestry simulations</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quick-reference">Quick reference</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#specifying-samples">Specifying samples</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#populations">Populations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sampling-time">Sampling time</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sample-details">Sample details</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#demography">Demography</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#ploidy">Ploidy</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#default-number-of-nodes-per-sample-individual">Default number of nodes per sample individual</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#coalescent-time-scales">Coalescent time scales</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#genome-properties">Genome properties</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sequence-length">Sequence length</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#recombination">Recombination</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-conversion">Gene conversion</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#discrete-or-continuous">Discrete or continuous?</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#multiple-chromosomes">Multiple chromosomes</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#do-i-need-to-do-this">Do I need to do this?</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#using-a-fixed-pedigree">Using a fixed pedigree</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#simulations-without-a-fixed-pedigree">Simulations without a fixed pedigree</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#example-1-simulating-with-a-pedigree">Example 1: simulating with a pedigree</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#example-2-simulating-without-a-pedigree">Example 2: simulating without a pedigree</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#recording-more-information">Recording more information</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#additional-nodes">Additional nodes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#coalescing-segments-only">Coalescing segments only</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#common-ancestor-events">Common ancestor events</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#recombination-events">Recombination events</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#migration-events">Migration events</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#census-events">Census events</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#using-the-dtwf-model">Using the DTWF model</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ancestral-recombination-graph">Ancestral recombination graph</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#manipulating-simulation-time">Manipulating simulation time</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#stopping-simulations-early">Stopping simulations early</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#setting-the-start-time">Setting the start time</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#order-of-event-execution">Order of event execution</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#specifying-the-initial-state">Specifying the initial state</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#combining-backward-time-simulations">Combining backward-time simulations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#interaction-with-demography">Interaction with demography</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#continuing-forwards-time-simulations-recapitating">Continuing forwards-time simulations (“recapitating”)</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#models">Models</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#specifying-ancestry-models">Specifying ancestry models</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#single-models">Single models</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#model-duration">Model duration</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#model-completion">Model completion</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#multiple-models">Multiple models</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#hudson-coalescent">Hudson coalescent</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#smc-approximations">SMC approximations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#discrete-time-wright-fisher">Discrete Time Wright-Fisher</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#multiple-merger-coalescents">Multiple merger coalescents</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#selective-sweeps">Selective sweeps</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#hard-sweeps">Hard sweeps</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#multiple-sweeps">Multiple sweeps</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#fixed-pedigree">Fixed pedigree</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#simple-example">Simple example</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#censoring-pedigree-information">Censoring pedigree information</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#completing-fixed-pedigree-simulations">Completing fixed pedigree simulations</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#missing-pedigree-data">Missing pedigree data</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pedigrees-and-demography">Pedigrees and demography</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#missing-data">Missing data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#common-errors">Common errors</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#infinite-waiting-time">Infinite waiting time</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="ancestry-simulations">
<span id="sec-ancestry"></span><h1>Ancestry simulations<a class="headerlink" href="#ancestry-simulations" title="Link to this heading">#</a></h1>
<p>Msprime simulates the ancestry of sampled genomes under a number of
different backwards-in-time population genetic
<a class="reference internal" href="#sec-ancestry-models"><span class="std std-ref">models</span></a>. The simulated ancestry is
represented as a <a class="reference external" href="https://tskit.dev">succinct tree sequence</a> using
the <a class="reference external" href="https://tskit.dev/tskit">tskit</a> library, which provides an
extensive suite of operations for analysing genealogical trees
and DNA sequence data. See the <a class="reference external" href="https://tskit.dev/tutorials/getting_started.html#sec-tskit-getting-started" title="(in Project name not set)"><span>Getting started with tskit</span></a>
tutorial for more information on how to process data using tskit.</p>
<p>Here we run a simple simulation and show a summary of the resulting
tree sequence:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
<span class="n">ts</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
              <style>
                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}
                .tskit-table tbody tr td {padding: 0.5em 0.5em;}
                .tskit-table tbody tr td:first-of-type {text-align: left;}
                .tskit-details-label {vertical-align: top; padding-right:5px;}
                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}
                .tskit-table-set-table {margin: 12px 0 0 12px;}
                details {display: inline-block;}
                summary {cursor: pointer; outline: 0; display: list-item;}
              </style>
              <div class="tskit-table-set">
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="padding:0;line-height:21px;">
                          <img style="height: 32px;display: inline-block;padding: 3px 5px 3px 0;" src="https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg"/>
                          <a target="_blank" href="https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class"> Tree Sequence </a>
                        </th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr><td>Trees</td><td>1</td></tr>
                      <tr><td>Sequence Length</td><td>1.0</td></tr>
                      <tr><td>Time Units</td><td>generations</td></tr>
                      <tr><td>Sample Nodes</td><td>4</td></tr>
                      <tr><td>Total Size</td><td>1.8 KiB</td></tr>
                      <tr>
                        <td>Metadata</td><td style="text-align: left;">No Metadata</td></tr>
                    </tbody>
                  </table>
                </div>
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="line-height:21px;">Table</th>
                        <th>Rows</th>
                        <th>Size</th>
                        <th>Has Metadata</th>
                      </tr>
                    </thead>
                    <tbody>
                    
                  <tr>
                    <td>Edges</td>
                      <td>6</td>
                      <td>200 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Individuals</td>
                      <td>2</td>
                      <td>80 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Migrations</td>
                      <td>0</td>
                      <td>8 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Mutations</td>
                      <td>0</td>
                      <td>16 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Nodes</td>
                      <td>7</td>
                      <td>204 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Populations</td>
                      <td>1</td>
                      <td>224 Bytes</td>
                      <td style="text-align: center;">
                        ✅
                      </td>
                    </tr>
                
                  <tr>
                    <td>Provenances</td>
                      <td>1</td>
                      <td>1021 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Sites</td>
                      <td>0</td>
                      <td>16 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
            </div></div>
</div>
<p>The <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a> function has many parameters to specify
the details of the simulations. The <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">API</span> <span class="pre">documentation</span></code></a>
contains a precise description of the parameter; the sections
in this page provide some explanation and examples of how to use these
parameters.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>It’s important to note that <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a> only simulates
the ancestral trees for samples: if we want actual sequence data
then we also need to simulate mutations on these trees.
See the <a class="reference internal" href="mutations.html#sec-mutations"><span class="std std-ref">Mutation simulations</span></a> section for more information on how to do
this.</p>
</div>
<hr class="docutils" />
<section id="quick-reference">
<span id="sec-ancestry-quickref"></span><h2>Quick reference<a class="headerlink" href="#quick-reference" title="Link to this heading">#</a></h2>
<dl class="simple myst">
<dt><a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a></dt><dd><p>Simulate ancestral tree sequence topology</p>
</dd>
</dl>
<p><strong>Models</strong></p>
<dl class="simple myst">
<dt><a class="reference internal" href="api.html#msprime.StandardCoalescent" title="msprime.StandardCoalescent"><code class="xref py py-class docutils literal notranslate"><span class="pre">StandardCoalescent</span></code></a></dt><dd><p>Coalescent with recombination (“hudson”)</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.SmcApproxCoalescent" title="msprime.SmcApproxCoalescent"><code class="xref py py-class docutils literal notranslate"><span class="pre">SmcApproxCoalescent</span></code></a></dt><dd><p>Sequentially Markov Coalescent (“smc”)</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.SmcPrimeApproxCoalescent" title="msprime.SmcPrimeApproxCoalescent"><code class="xref py py-class docutils literal notranslate"><span class="pre">SmcPrimeApproxCoalescent</span></code></a></dt><dd><p>SMC’(“smc_prime”)</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.DiscreteTimeWrightFisher" title="msprime.DiscreteTimeWrightFisher"><code class="xref py py-class docutils literal notranslate"><span class="pre">DiscreteTimeWrightFisher</span></code></a></dt><dd><p>Generation-by-generation Wright-Fisher</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.FixedPedigree" title="msprime.FixedPedigree"><code class="xref py py-class docutils literal notranslate"><span class="pre">FixedPedigree</span></code></a></dt><dd><p>Simulated conditioned on a given pedigree</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.BetaCoalescent" title="msprime.BetaCoalescent"><code class="xref py py-class docutils literal notranslate"><span class="pre">BetaCoalescent</span></code></a></dt><dd><p>Beta coalescent multiple-merger</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.DiracCoalescent" title="msprime.DiracCoalescent"><code class="xref py py-class docutils literal notranslate"><span class="pre">DiracCoalescent</span></code></a></dt><dd><p>Dirac coalescent multiple-merger</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.SweepGenicSelection" title="msprime.SweepGenicSelection"><code class="xref py py-class docutils literal notranslate"><span class="pre">SweepGenicSelection</span></code></a></dt><dd><p>Selective sweep at a linked locus</p>
</dd>
</dl>
<hr class="docutils" />
</section>
<section id="specifying-samples">
<span id="sec-ancestry-samples"></span><h2>Specifying samples<a class="headerlink" href="#specifying-samples" title="Link to this heading">#</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">samples</span></code> argument to <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a> defines the number
of sampled individuals whose history will be simulated. There are three different
forms; the <code class="docutils literal notranslate"><span class="pre">samples</span></code> argument can be:</p>
<ul class="simple">
<li><p>an <strong>integer</strong>, interpreted as the number of sampled individuals to draw in
a single population model;</p></li>
<li><p>a <strong>dictionary</strong> mapping population references (either integer IDs or
names) to the number of sampled individuals for that population;</p></li>
<li><p>a list of <a class="reference internal" href="api.html#msprime.SampleSet" title="msprime.SampleSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">SampleSet</span></code></a> objects, which provide more flexibility
in how groups of similar samples are drawn from populations.</p></li>
</ul>
<p>In the simplest case, provide a single integer which defines the number
of samples:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1234</span><span class="p">)</span>
<span class="n">ts</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
              <style>
                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}
                .tskit-table tbody tr td {padding: 0.5em 0.5em;}
                .tskit-table tbody tr td:first-of-type {text-align: left;}
                .tskit-details-label {vertical-align: top; padding-right:5px;}
                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}
                .tskit-table-set-table {margin: 12px 0 0 12px;}
                details {display: inline-block;}
                summary {cursor: pointer; outline: 0; display: list-item;}
              </style>
              <div class="tskit-table-set">
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="padding:0;line-height:21px;">
                          <img style="height: 32px;display: inline-block;padding: 3px 5px 3px 0;" src="https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg"/>
                          <a target="_blank" href="https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class"> Tree Sequence </a>
                        </th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr><td>Trees</td><td>1</td></tr>
                      <tr><td>Sequence Length</td><td>1.0</td></tr>
                      <tr><td>Time Units</td><td>generations</td></tr>
                      <tr><td>Sample Nodes</td><td>4</td></tr>
                      <tr><td>Total Size</td><td>1.8 KiB</td></tr>
                      <tr>
                        <td>Metadata</td><td style="text-align: left;">No Metadata</td></tr>
                    </tbody>
                  </table>
                </div>
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="line-height:21px;">Table</th>
                        <th>Rows</th>
                        <th>Size</th>
                        <th>Has Metadata</th>
                      </tr>
                    </thead>
                    <tbody>
                    
                  <tr>
                    <td>Edges</td>
                      <td>6</td>
                      <td>200 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Individuals</td>
                      <td>2</td>
                      <td>80 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Migrations</td>
                      <td>0</td>
                      <td>8 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Mutations</td>
                      <td>0</td>
                      <td>16 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Nodes</td>
                      <td>7</td>
                      <td>204 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Populations</td>
                      <td>1</td>
                      <td>224 Bytes</td>
                      <td style="text-align: center;">
                        ✅
                      </td>
                    </tr>
                
                  <tr>
                    <td>Provenances</td>
                      <td>1</td>
                      <td>1016 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Sites</td>
                      <td>0</td>
                      <td>16 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
            </div></div>
</div>
<p>Here we specify two sample <strong>individuals</strong>, which at the default
ploidy of two, gives us four sample <strong>nodes</strong>.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>It is important to note that the number of samples
refers to the number of <em>individuals</em> not the number of <em>nodes</em>
(monoploid genomes). See the <a class="reference internal" href="#sec-ancestry-ploidy"><span class="std std-ref">Ploidy</span></a>
section for details.</p>
</div>
<p>This integer form for the <code class="docutils literal notranslate"><span class="pre">samples</span></code> argument can only be used
when the <code class="docutils literal notranslate"><span class="pre">demography</span></code> is a single population model.</p>
<section id="populations">
<h3>Populations<a class="headerlink" href="#populations" title="Link to this heading">#</a></h3>
<p>The next example illustrates one usage of the dictionary form of the <code class="docutils literal notranslate"><span class="pre">samples</span></code>
argument, in which the keys refer to populations in the
<a class="reference internal" href="demography.html#sec-demography"><span class="std std-ref">demographic model</span></a> and
the values are the number of samples to draw.
(See the <a class="reference internal" href="#sec-ancestry-demography"><span class="std std-ref">Demography</span></a> section for more details on
the <code class="docutils literal notranslate"><span class="pre">demography</span></code> argument to <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a>.)
We first create a <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> object representing
a 10 deme linear stepping stone model. Then, we run the simulation
with 3 diploid samples each drawn from the first and last demes in this
linear habitat.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">N</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">stepping_stone_model</span><span class="p">(</span>
    <span class="p">[</span><span class="mi">100</span><span class="p">]</span> <span class="o">*</span> <span class="n">N</span><span class="p">,</span>
    <span class="n">migration_rate</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
    <span class="n">boundaries</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">({</span><span class="mi">0</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span> <span class="n">N</span> <span class="o">-</span> <span class="mi">1</span><span class="p">:</span> <span class="mi">3</span><span class="p">},</span> <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">)</span>
<span class="n">ts</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
              <style>
                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}
                .tskit-table tbody tr td {padding: 0.5em 0.5em;}
                .tskit-table tbody tr td:first-of-type {text-align: left;}
                .tskit-details-label {vertical-align: top; padding-right:5px;}
                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}
                .tskit-table-set-table {margin: 12px 0 0 12px;}
                details {display: inline-block;}
                summary {cursor: pointer; outline: 0; display: list-item;}
              </style>
              <div class="tskit-table-set">
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="padding:0;line-height:21px;">
                          <img style="height: 32px;display: inline-block;padding: 3px 5px 3px 0;" src="https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg"/>
                          <a target="_blank" href="https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class"> Tree Sequence </a>
                        </th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr><td>Trees</td><td>1</td></tr>
                      <tr><td>Sequence Length</td><td>1.0</td></tr>
                      <tr><td>Time Units</td><td>generations</td></tr>
                      <tr><td>Sample Nodes</td><td>12</td></tr>
                      <tr><td>Total Size</td><td>5.6 KiB</td></tr>
                      <tr>
                        <td>Metadata</td><td style="text-align: left;">No Metadata</td></tr>
                    </tbody>
                  </table>
                </div>
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="line-height:21px;">Table</th>
                        <th>Rows</th>
                        <th>Size</th>
                        <th>Has Metadata</th>
                      </tr>
                    </thead>
                    <tbody>
                    
                  <tr>
                    <td>Edges</td>
                      <td>22</td>
                      <td>712 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Individuals</td>
                      <td>6</td>
                      <td>192 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Migrations</td>
                      <td>0</td>
                      <td>8 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Mutations</td>
                      <td>0</td>
                      <td>16 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Nodes</td>
                      <td>23</td>
                      <td>652 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Populations</td>
                      <td>10</td>
                      <td>593 Bytes</td>
                      <td style="text-align: center;">
                        ✅
                      </td>
                    </tr>
                
                  <tr>
                    <td>Provenances</td>
                      <td>1</td>
                      <td>3.3 KiB</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Sites</td>
                      <td>0</td>
                      <td>16 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
            </div></div>
</div>
<p>The keys in the dictionary can also be the string names of the
population (see the <a class="reference internal" href="demography.html#sec-demography-populations-identifiers"><span class="std std-ref">Identifying populations</span></a> section),
which is useful when we are simulating from empirically
estimated models. For example, here create a <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> object
based on a species tree, and then draw samples using the species names.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">from_species_tree</span><span class="p">(</span>
    <span class="s2">&quot;(((human:5.6,chimpanzee:5.6):3.0,gorilla:8.6):9.4,orangutan:18.0)&quot;</span><span class="p">,</span>
    <span class="n">time_units</span><span class="o">=</span><span class="s2">&quot;myr&quot;</span><span class="p">,</span>
    <span class="n">initial_size</span><span class="o">=</span><span class="mi">10</span><span class="o">**</span><span class="mi">4</span><span class="p">,</span>
    <span class="n">generation_time</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">({</span><span class="s2">&quot;gorilla&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;human&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">},</span> <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">)</span>
<span class="n">ts</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
              <style>
                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}
                .tskit-table tbody tr td {padding: 0.5em 0.5em;}
                .tskit-table tbody tr td:first-of-type {text-align: left;}
                .tskit-details-label {vertical-align: top; padding-right:5px;}
                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}
                .tskit-table-set-table {margin: 12px 0 0 12px;}
                details {display: inline-block;}
                summary {cursor: pointer; outline: 0; display: list-item;}
              </style>
              <div class="tskit-table-set">
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="padding:0;line-height:21px;">
                          <img style="height: 32px;display: inline-block;padding: 3px 5px 3px 0;" src="https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg"/>
                          <a target="_blank" href="https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class"> Tree Sequence </a>
                        </th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr><td>Trees</td><td>1</td></tr>
                      <tr><td>Sequence Length</td><td>1.0</td></tr>
                      <tr><td>Time Units</td><td>generations</td></tr>
                      <tr><td>Sample Nodes</td><td>12</td></tr>
                      <tr><td>Total Size</td><td>5.0 KiB</td></tr>
                      <tr>
                        <td>Metadata</td><td style="text-align: left;">No Metadata</td></tr>
                    </tbody>
                  </table>
                </div>
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="line-height:21px;">Table</th>
                        <th>Rows</th>
                        <th>Size</th>
                        <th>Has Metadata</th>
                      </tr>
                    </thead>
                    <tbody>
                    
                  <tr>
                    <td>Edges</td>
                      <td>22</td>
                      <td>712 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Individuals</td>
                      <td>6</td>
                      <td>192 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Migrations</td>
                      <td>0</td>
                      <td>8 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Mutations</td>
                      <td>0</td>
                      <td>16 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Nodes</td>
                      <td>23</td>
                      <td>652 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Populations</td>
                      <td>7</td>
                      <td>481 Bytes</td>
                      <td style="text-align: center;">
                        ✅
                      </td>
                    </tr>
                
                  <tr>
                    <td>Provenances</td>
                      <td>1</td>
                      <td>2.8 KiB</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Sites</td>
                      <td>0</td>
                      <td>16 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
            </div></div>
</div>
<p>When samples are drawn using this dictionary form it is always
at the population’s <a class="reference internal" href="demography.html#sec-demography-populations-default-sampling-time"><span class="std std-ref">Default sampling time</span></a>.
If you wish to draw samples at a different time, then you must use the
more general form of a list of <a class="reference internal" href="api.html#msprime.SampleSet" title="msprime.SampleSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">SampleSet</span></code></a> objects.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>See the <a class="reference internal" href="#sec-ancestry-demography"><span class="std std-ref">Demography</span></a> section for more information
on how to specify demographic models in a simulation.</p>
</div>
</section>
<section id="sampling-time">
<span id="sec-ancestry-samples-sampling-time"></span><h3>Sampling time<a class="headerlink" href="#sampling-time" title="Link to this heading">#</a></h3>
<p>By default the samples that we draw from a <a class="reference internal" href="api.html#msprime.Population" title="msprime.Population"><code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code></a> are
at the population’s <a class="reference internal" href="api.html#msprime.Population.default_sampling_time" title="msprime.Population.default_sampling_time"><code class="xref py py-attr docutils literal notranslate"><span class="pre">default_sampling_time</span></code></a>.
See the <a class="reference internal" href="demography.html#sec-demography-populations-default-sampling-time"><span class="std std-ref">Default sampling time</span></a>
section for more information about how this is
set for populations in a demographic model.</p>
<p>We can manually control the time at which samples
are drawn using list of <a class="reference internal" href="api.html#msprime.SampleSet" title="msprime.SampleSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">SampleSet</span></code></a> objects form for the
samples argument.</p>
<p>In this example we create two diploid sample individuals,
one at the present time
and one taken 50 generations ago, representing one modern
and one ancient individual. By default, when we draw the
trees, the nodes that belong to samples are drawn as squares
whereas nodes that are not samples are drawn as circles.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">msprime</span><span class="o">.</span><span class="n">SampleSet</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span>
        <span class="n">msprime</span><span class="o">.</span><span class="n">SampleSet</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">time</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
    <span class="p">],</span>
    <span class="n">population_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span>
<span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/91b787063b98d7bb80d819d1fb202b5f3a02056a7ba56f6f4b91b82448880bb2.svg" src="_images/91b787063b98d7bb80d819d1fb202b5f3a02056a7ba56f6f4b91b82448880bb2.svg" />
</div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>See the <a class="reference internal" href="#sec-ancestry-event-order"><span class="std std-ref">Order of event execution</span></a> section for details
of the relative order in which sampling events occur.</p>
</div>
</section>
<section id="sample-details">
<span id="sec-ancestry-samples-sampling-details"></span><h3>Sample details<a class="headerlink" href="#sample-details" title="Link to this heading">#</a></h3>
<p>Sample individuals and nodes are allocated sequentially in the order that
they are specified. For example:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">island_model</span><span class="p">([</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">],</span> <span class="n">migration_rate</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="p">[</span>
        <span class="n">msprime</span><span class="o">.</span><span class="n">SampleSet</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">population</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
        <span class="n">msprime</span><span class="o">.</span><span class="n">SampleSet</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">population</span><span class="o">=</span><span class="mi">0</span><span class="p">)],</span>
    <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">,</span>
    <span class="n">end_time</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>╔══╤═════╤════════╤═══════╤════════╗
║id│flags│location│parents│metadata║
╠══╪═════╪════════╪═══════╪════════╣
║0 │    0│        │       │        ║
║1 │    0│        │       │        ║
║2 │    0│        │       │        ║
╚══╧═════╧════════╧═══════╧════════╝

╔══╤═════╤══════════╤══════════╤════╤════════╗
║id│flags│population│individual│time│metadata║
╠══╪═════╪══════════╪══════════╪════╪════════╣
║0 │    1│         1│         0│   0│        ║
║1 │    1│         1│         0│   0│        ║
║2 │    1│         0│         1│   0│        ║
║3 │    1│         0│         1│   0│        ║
║4 │    1│         0│         2│   0│        ║
║5 │    1│         0│         2│   0│        ║
╚══╧═════╧══════════╧══════════╧════╧════════╝
</pre></div>
</div>
</div>
</div>
<p>(Because we’re only interested in the sampled nodes and individuals we
stopped the simulation from actually doing anything by setting
<code class="docutils literal notranslate"><span class="pre">end_time=0</span></code> — see the <a class="reference internal" href="#sec-ancestry-start-time"><span class="std std-ref">Setting the start time</span></a> section
for more information.) Here we define three sample individuals,
and we therefore have three rows in the individual table.
Because these are diploid individuals, the node table contains
six sample nodes. If we look at the <code class="docutils literal notranslate"><span class="pre">individual</span></code> column in the
node table we can see that the first two nodes correspond to individual
<code class="docutils literal notranslate"><span class="pre">0</span></code>, the next two nodes individual <code class="docutils literal notranslate"><span class="pre">1</span></code>, etc. The sample configuration
stated that the first sample should come from population <code class="docutils literal notranslate"><span class="pre">1</span></code> and
the other two from population <code class="docutils literal notranslate"><span class="pre">0</span></code>, and we can see this reflected
in the <code class="docutils literal notranslate"><span class="pre">population</span></code> column of the node table. (Somewhat confusingly,
population values are associated with nodes rather than individuals;
this is mostly for historical reasons.)</p>
<p>The <a class="reference internal" href="api.html#msprime.SampleSet" title="msprime.SampleSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">SampleSet</span></code></a> class has a number of attributes which default
to <code class="docutils literal notranslate"><span class="pre">None</span></code>. If these are set they will <strong>override</strong> the values
which might be specified elsewhere. For example, we can specify
mixed ploidy samples via the <code class="docutils literal notranslate"><span class="pre">ploidy</span></code> attribute of <a class="reference internal" href="api.html#msprime.SampleSet" title="msprime.SampleSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">SampleSet</span></code></a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="p">[</span>
        <span class="n">msprime</span><span class="o">.</span><span class="n">SampleSet</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">=</span><span class="mi">3</span><span class="p">),</span>
        <span class="n">msprime</span><span class="o">.</span><span class="n">SampleSet</span><span class="p">(</span><span class="mi">2</span><span class="p">)],</span>
    <span class="n">ploidy</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">end_time</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>╔══╤═════╤════════╤═══════╤════════╗
║id│flags│location│parents│metadata║
╠══╪═════╪════════╪═══════╪════════╣
║0 │    0│        │       │        ║
║1 │    0│        │       │        ║
║2 │    0│        │       │        ║
╚══╧═════╧════════╧═══════╧════════╝

╔══╤═════╤══════════╤══════════╤════╤════════╗
║id│flags│population│individual│time│metadata║
╠══╪═════╪══════════╪══════════╪════╪════════╣
║0 │    1│         0│         0│   0│        ║
║1 │    1│         0│         0│   0│        ║
║2 │    1│         0│         0│   0│        ║
║3 │    1│         0│         1│   0│        ║
║4 │    1│         0│         2│   0│        ║
╚══╧═════╧══════════╧══════════╧════╧════════╝
</pre></div>
</div>
</div>
</div>
<p>(Again, we stop the simulation immediately because we’re only interested
in the initial samples.) Here we have three sampled individuals again
but in this they are mixed ploidy: the first individual is triploid
and the other two are haploid.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>It is vital to note that setting the <code class="docutils literal notranslate"><span class="pre">ploidy</span></code> value
of <a class="reference internal" href="api.html#msprime.SampleSet" title="msprime.SampleSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">SampleSet</span></code></a> objects only affects sampling and does
not affect the actual simulation. In this case, the simulation
will be run on the haploid time scale. See the
<a class="reference internal" href="#sec-ancestry-ploidy"><span class="std std-ref">Ploidy</span></a> section for more details.</p>
</div>
<p>If you wish to set up the node and individual IDs in some other way,
it is possible to do so by using the <code class="docutils literal notranslate"><span class="pre">initial_state</span></code> parameter.
See the  <a class="reference internal" href="#sec-ancestry-initial-state"><span class="std std-ref">Specifying the initial state</span></a> for more information
on how to use this (advanced) feature.</p>
</section>
</section>
<section id="demography">
<span id="sec-ancestry-demography"></span><h2>Demography<a class="headerlink" href="#demography" title="Link to this heading">#</a></h2>
<p>A <a class="reference internal" href="demography.html#sec-demography"><span class="std std-ref">demographic model</span></a> is a description of a
set of populations, the migration rates between then and the
events that modify the populations over time. Ancestry simulations
can take one of these models as input via the <code class="docutils literal notranslate"><span class="pre">demography</span></code>
parameter to <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a>.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>Please see the <a class="reference internal" href="demography.html#sec-demography"><span class="std std-ref">Demographic models</span></a> section for details on how
to create and debug demographic models, and the
<a class="reference internal" href="#sec-ancestry-samples"><span class="std std-ref">Specifying samples</span></a> section for help on how to
define the locations and times of sampled individuals.</p>
</div>
<p>We don’t need to explicitly specify a <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a>
for every ancestry simulation, however. By default, we assume a
single fixed-size population. For the default
<a class="reference internal" href="#sec-ancestry-models-hudson"><span class="std std-ref">standard coalescent</span></a> ancestry
model we assume a population size of 1, so that time is measured
in “coalescent units”. This is useful for theoretical work where
we usually work in scaled coalescent time.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>By default msprime uses the <strong>diploid</strong> coalescent time scale;
see the <a class="reference internal" href="#sec-ancestry-ploidy"><span class="std std-ref">Ploidy</span></a> section for more details.</p>
</div>
<p>Outside of theoretical analysis, working with scaled
coalescent time can be confusing and it’s usually better to
explicitly set the population size so that the output
times are directly interpretable in generations.
The <code class="docutils literal notranslate"><span class="pre">population_size</span></code> parameter to <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a>
allows us to set the size of the single constant
sized population in the default demography.</p>
<p>For example, here we run the same simulation twice, one with
the default population size and one with a population size
of 100 (this is the diploid population size because the
default <a class="reference internal" href="#sec-ancestry-ploidy"><span class="std std-ref">Ploidy</span></a> is 2):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1234</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/3073f40af0467850c4f2006fbad70613c396bbd8eeae384cdd7ae9b21a27b558.svg" src="_images/3073f40af0467850c4f2006fbad70613c396bbd8eeae384cdd7ae9b21a27b558.svg" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">population_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1234</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/c13d5f56f1cf9823556e342d1a0efe138f205272a68781d6f897a9a0d84a98f1.svg" src="_images/c13d5f56f1cf9823556e342d1a0efe138f205272a68781d6f897a9a0d84a98f1.svg" />
</div>
</div>
<p>We can see that the results are identical, except the node times have
been scaled by the population size.</p>
<p>The <a class="reference internal" href="#sec-ancestry-models-dtwf"><span class="std std-ref">Discrete Time Wright-Fisher</span></a> model is different. Because it
explicitly works with a population of N individuals we get an
error if we don’t specify a population size:</p>
<div class="cell tag_raises-exception docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s2">&quot;dtwf&quot;</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1234</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">ValueError</span><span class="g g-Whitespace">                                </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">11</span><span class="p">],</span> <span class="n">line</span> <span class="mi">1</span>
<span class="ne">----&gt; </span><span class="mi">1</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s2">&quot;dtwf&quot;</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1234</span><span class="p">)</span>

<span class="nn">File ~/work/tskit-site/tskit-site/msprime/ancestry.py:1290,</span> in <span class="ni">sim_ancestry</span><span class="nt">(samples, demography, sequence_length, discrete_genome, recombination_rate, gene_conversion_rate, gene_conversion_tract_length, population_size, ploidy, model, initial_state, start_time, end_time, record_migrations, record_full_arg, additional_nodes, coalescing_segments_only, num_labels, random_seed, num_replicates, replicate_index, record_provenance)</span>
<span class="g g-Whitespace">   </span><span class="mi">1265</span>     <span class="n">parameters</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1266</span>         <span class="n">command</span><span class="o">=</span><span class="s2">&quot;sim_ancestry&quot;</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1267</span>         <span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span>
   <span class="p">(</span><span class="o">...</span><span class="p">)</span>   <span class="mi">1287</span>         <span class="c1"># replicate index is excluded as it is inserted for each replicate</span>
<span class="g g-Whitespace">   </span><span class="mi">1288</span>     <span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1289</span>     <span class="n">provenance_dict</span> <span class="o">=</span> <span class="n">provenance</span><span class="o">.</span><span class="n">get_provenance_dict</span><span class="p">(</span><span class="n">parameters</span><span class="p">)</span>
<span class="ne">-&gt; </span><span class="mi">1290</span> <span class="n">sim</span> <span class="o">=</span> <span class="n">_parse_sim_ancestry</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1291</span>     <span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1292</span>     <span class="n">sequence_length</span><span class="o">=</span><span class="n">sequence_length</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1293</span>     <span class="n">recombination_rate</span><span class="o">=</span><span class="n">recombination_rate</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1294</span>     <span class="n">gene_conversion_rate</span><span class="o">=</span><span class="n">gene_conversion_rate</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1295</span>     <span class="n">gene_conversion_tract_length</span><span class="o">=</span><span class="n">gene_conversion_tract_length</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1296</span>     <span class="n">discrete_genome</span><span class="o">=</span><span class="n">discrete_genome</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1297</span>     <span class="n">population_size</span><span class="o">=</span><span class="n">population_size</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1298</span>     <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1299</span>     <span class="n">ploidy</span><span class="o">=</span><span class="n">ploidy</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1300</span>     <span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1301</span>     <span class="n">initial_state</span><span class="o">=</span><span class="n">initial_state</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1302</span>     <span class="n">start_time</span><span class="o">=</span><span class="n">start_time</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1303</span>     <span class="n">end_time</span><span class="o">=</span><span class="n">end_time</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1304</span>     <span class="n">record_migrations</span><span class="o">=</span><span class="n">record_migrations</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1305</span>     <span class="n">record_full_arg</span><span class="o">=</span><span class="n">record_full_arg</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1306</span>     <span class="n">additional_nodes</span><span class="o">=</span><span class="n">additional_nodes</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1307</span>     <span class="n">coalescing_segments_only</span><span class="o">=</span><span class="n">coalescing_segments_only</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1308</span>     <span class="n">num_labels</span><span class="o">=</span><span class="n">num_labels</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1309</span>     <span class="n">random_seed</span><span class="o">=</span><span class="n">random_seed</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1310</span> <span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1311</span> <span class="k">return</span> <span class="n">_wrap_replicates</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1312</span>     <span class="n">sim</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1313</span>     <span class="n">num_replicates</span><span class="o">=</span><span class="n">num_replicates</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1314</span>     <span class="n">replicate_index</span><span class="o">=</span><span class="n">replicate_index</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1315</span>     <span class="n">provenance_dict</span><span class="o">=</span><span class="n">provenance_dict</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1316</span> <span class="p">)</span>

<span class="nn">File ~/work/tskit-site/tskit-site/msprime/ancestry.py:1009,</span> in <span class="ni">_parse_sim_ancestry</span><span class="nt">(samples, sequence_length, recombination_rate, gene_conversion_rate, gene_conversion_tract_length, discrete_genome, population_size, demography, ploidy, model, initial_state, start_time, end_time, record_migrations, record_full_arg, additional_nodes, coalescing_segments_only, num_labels, random_seed, init_for_debugger)</span>
<span class="g g-Whitespace">   </span><span class="mi">1005</span> <span class="k">if</span> <span class="n">is_dtwf</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">1006</span>     <span class="c1"># A default size of 1 isn&#39;t so smart for DTWF and almost certainly</span>
<span class="g g-Whitespace">   </span><span class="mi">1007</span>     <span class="c1"># an error.</span>
<span class="g g-Whitespace">   </span><span class="mi">1008</span>     <span class="k">if</span> <span class="n">population_size</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<span class="ne">-&gt; </span><span class="mi">1009</span>         <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1010</span>             <span class="s2">&quot;When using the DTWF model, the population size must be set &quot;</span>
<span class="g g-Whitespace">   </span><span class="mi">1011</span>             <span class="s2">&quot;explicitly, either using the population_size or demography &quot;</span>
<span class="g g-Whitespace">   </span><span class="mi">1012</span>             <span class="s2">&quot;arguments.&quot;</span>
<span class="g g-Whitespace">   </span><span class="mi">1013</span>         <span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1014</span> <span class="k">if</span> <span class="n">initial_state</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">1015</span>     <span class="k">if</span> <span class="n">population_size</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>

<span class="ne">ValueError</span>: When using the DTWF model, the population size must be set explicitly, either using the population_size or demography arguments.
</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Because the DTWF models populations of individuals rather than
using continuous
time approximations, simulations with different population sizes
will yield entirely different results; we are <strong>not</strong> simply
rescaling time by using different population sizes. Furthermore,
simulations with large population sizes will take relatively
more CPU time and memory.</p>
</div>
<hr class="docutils" />
</section>
<section id="ploidy">
<span id="sec-ancestry-ploidy"></span><h2>Ploidy<a class="headerlink" href="#ploidy" title="Link to this heading">#</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">ploidy</span></code> argument to <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a> has two effects:</p>
<ol class="arabic simple">
<li><p>It sets the default ploidy, that is, the number of nodes per sample individual.</p></li>
<li><p>For continuous time coalescent models, it sets the time scale.</p></li>
</ol>
<p>Both of these are somewhat confusing, so let’s look at them one at a time.</p>
<section id="default-number-of-nodes-per-sample-individual">
<span id="sec-ancestry-ploidy-default-num-nodes"></span><h3>Default number of nodes per sample individual<a class="headerlink" href="#default-number-of-nodes-per-sample-individual" title="Link to this heading">#</a></h3>
<p>As discussed in the <a class="reference internal" href="#sec-ancestry-samples-sampling-details"><span class="std std-ref">Sample details</span></a> section,
the number of nodes (i.e., monoploid genomes) that we allocate per
sample individual is determined by the ploidy. Here, we specify
2 sample individuals with a ploidy of 3:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="n">samples</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">end_time</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>╔══╤═════╤════════╤═══════╤════════╗
║id│flags│location│parents│metadata║
╠══╪═════╪════════╪═══════╪════════╣
║0 │    0│        │       │        ║
║1 │    0│        │       │        ║
╚══╧═════╧════════╧═══════╧════════╝

╔══╤═════╤══════════╤══════════╤════╤════════╗
║id│flags│population│individual│time│metadata║
╠══╪═════╪══════════╪══════════╪════╪════════╣
║0 │    1│         0│         0│   0│        ║
║1 │    1│         0│         0│   0│        ║
║2 │    1│         0│         0│   0│        ║
║3 │    1│         0│         1│   0│        ║
║4 │    1│         0│         1│   0│        ║
║5 │    1│         0│         1│   0│        ║
╚══╧═════╧══════════╧══════════╧════╧════════╝
</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Because we’re only interested in the initial conditions here we
set the <code class="docutils literal notranslate"><span class="pre">end_time</span></code> to zero: see the <a class="reference internal" href="#sec-ancestry-start-time"><span class="std std-ref">Setting the start time</span></a>
section for more details.</p>
</div>
<p>Our individual table has two rows (the number of <code class="docutils literal notranslate"><span class="pre">samples</span></code>)
and node table has 6 rows: three for each of the sampled
individuals.</p>
<p>Note that this is the <strong>default</strong>, and can be overridden by the
<a class="reference internal" href="api.html#msprime.SampleSet.ploidy" title="msprime.SampleSet.ploidy"><code class="xref py py-attr docutils literal notranslate"><span class="pre">SampleSet.ploidy</span></code></a> value, as described in the
<a class="reference internal" href="#sec-ancestry-samples-sampling-details"><span class="std std-ref">Sample details</span></a> section.</p>
</section>
<section id="coalescent-time-scales">
<span id="sec-ancestry-ploidy-coalescent-time-scales"></span><h3>Coalescent time scales<a class="headerlink" href="#coalescent-time-scales" title="Link to this heading">#</a></h3>
<p>The ploidy argument also
affects the time scale on which coalescence occurs,
since it affects the total number of genomes in the population
(see also <a class="reference internal" href="#sec-ancestry-demography"><span class="std std-ref">population size</span></a>).
For example, consider the following two simulations, which
are identical except for different <code class="docutils literal notranslate"><span class="pre">ploidy</span></code> values
(note we use the <a class="reference internal" href="api.html#msprime.SampleSet" title="msprime.SampleSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">SampleSet</span></code></a> object to make
sure we allocate the same number of <strong>nodes</strong>, as discussed
in the <a class="reference internal" href="#sec-ancestry-ploidy-default-num-nodes"><span class="std std-ref">Default number of nodes per sample individual</span></a> section):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="p">[</span><span class="n">msprime</span><span class="o">.</span><span class="n">SampleSet</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">=</span><span class="mi">1</span><span class="p">)],</span>
    <span class="n">population_size</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">ploidy</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">1234</span>
<span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/a8deb3d001de8a2923285df52ab3b49b8fda363032c8ca2d2c8a8a40d67d4413.svg" src="_images/a8deb3d001de8a2923285df52ab3b49b8fda363032c8ca2d2c8a8a40d67d4413.svg" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="p">[</span><span class="n">msprime</span><span class="o">.</span><span class="n">SampleSet</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">=</span><span class="mi">1</span><span class="p">)],</span>
    <span class="n">population_size</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">ploidy</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">1234</span>
<span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/ee03351ca5d012b8344d6a0063aedc2dbb5ee54265ea1c0bc954d052aecefd4b.svg" src="_images/ee03351ca5d012b8344d6a0063aedc2dbb5ee54265ea1c0bc954d052aecefd4b.svg" />
</div>
</div>
<p>The resulting tree sequences are identical, except that node time differ by a
factor of 2, because in the underlying coalescent algorithm the total number of
genomes in the population acts as a time scaling.</p>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>Add a discussion here about the Lambda coalescents an how
things are more complicated for them.</p>
</div>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>The <a class="reference internal" href="#sec-ancestry-models-dtwf"><span class="std std-ref">Discrete Time Wright-Fisher</span></a> model is an explicitly diploid model
and it is therefore
an error to run it with any <code class="docutils literal notranslate"><span class="pre">ploidy</span></code> value other than 2.</p>
</div>
</section>
</section>
<section id="genome-properties">
<span id="sec-ancestry-genome-properties"></span><h2>Genome properties<a class="headerlink" href="#genome-properties" title="Link to this heading">#</a></h2>
<section id="sequence-length">
<span id="sec-ancestry-sequence-length"></span><h3>Sequence length<a class="headerlink" href="#sequence-length" title="Link to this heading">#</a></h3>
<p>There are a number of ways to specify the length of the
chromosome that we want to simulate. In the absence of recombination
and gene conversion, we assume a genome of length 1:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/72b253cb7e0e1e17186d87b5f5ca481d4e0ada890a8b97cf7f7805a1fd790dd8.svg" src="_images/72b253cb7e0e1e17186d87b5f5ca481d4e0ada890a8b97cf7f7805a1fd790dd8.svg" />
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> parameter determines the length of the
simulated sequence:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">sequence_length</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/71e574ece6fa8b4d2a6f7a215d0da7e5ae299964931c4a97db053a92668c6426.svg" src="_images/71e574ece6fa8b4d2a6f7a215d0da7e5ae299964931c4a97db053a92668c6426.svg" />
</div>
</div>
<p>By default there is no recombination or gene conversion, and so we
therefore still have only one tree defining the ancestry of the
samples. The sequence length makes little difference if we are
considering only the ancestry of a non-recombining sequence, however,
this will be important if we are interested in adding
<a class="reference internal" href="mutations.html#sec-mutations"><span class="std std-ref">mutations</span></a> later.</p>
</section>
<section id="recombination">
<span id="sec-ancestry-recombination"></span><h3>Recombination<a class="headerlink" href="#recombination" title="Link to this heading">#</a></h3>
<p>As we trace the ancestry of a sample of genomes backwards in time,
recombination has the effect of creating segments of genome
that have different (but highly correlated) genealogical trees.
The <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a> function has a <code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code>
argument that allows us to flexibly specify both uniform and
non-uniform rates of recombination along the genome.</p>
<p>The simplest way to run simulations with recombination is to
use the <code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code> and <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> parameters:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">3</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">sequence_length</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/f69d27abed270fb36dffa44791dcf9c7e7571da1d2708c05bb57affa5172f87a.svg" src="_images/f69d27abed270fb36dffa44791dcf9c7e7571da1d2708c05bb57affa5172f87a.svg" />
</div>
</div>
<p>Here, we have four distinct trees along the genome,
created by recombination events.</p>
<p>We can also provide a <a class="reference internal" href="api.html#msprime.RateMap" title="msprime.RateMap"><code class="xref py py-class docutils literal notranslate"><span class="pre">RateMap</span></code></a> argument to <code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rate_map</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">RateMap</span><span class="o">.</span><span class="n">uniform</span><span class="p">(</span><span class="n">sequence_length</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
<span class="n">other_ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="n">rate_map</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="k">assert</span> <span class="n">ts</span><span class="o">.</span><span class="n">equals</span><span class="p">(</span><span class="n">other_ts</span><span class="p">,</span> <span class="n">ignore_provenance</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>There are two things to note here:</p>
<ol class="arabic simple">
<li><p>We didn’t need to specify a <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> argument to
<a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a>, as this information is encapsulated by the <code class="docutils literal notranslate"><span class="pre">rate_map</span></code>
instance. (We can provide a value for <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> if we wish, but
it must be equal to the <a class="reference internal" href="api.html#msprime.RateMap.sequence_length" title="msprime.RateMap.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">sequence_length</span></code></a> property of the
<a class="reference internal" href="api.html#msprime.RateMap" title="msprime.RateMap"><code class="xref py py-class docutils literal notranslate"><span class="pre">RateMap</span></code></a>).</p></li>
<li><p>The result of the simulation for a <a class="reference internal" href="rate_maps.html#sec-rate-maps-creating-uniform"><span class="std std-ref">Uniform rate maps</span></a>
rate map with the same rate and <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> is identical to
the first simulation above.</p></li>
</ol>
<p>We can simulate non-uniform recombination by defining our <a class="reference internal" href="api.html#msprime.RateMap" title="msprime.RateMap"><code class="xref py py-class docutils literal notranslate"><span class="pre">RateMap</span></code></a>
appropriately. For example, here we define a map in which the
rate over the first half of the sequence is one tenth of the second half:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rate_map</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">RateMap</span><span class="p">(</span><span class="n">position</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">],</span> <span class="n">rate</span><span class="o">=</span><span class="p">[</span><span class="mf">0.01</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">])</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="n">rate_map</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/f2b3bab632ebb05954382d982094949c53e9fa82031d511d73342d3e4e5f5c6f.svg" src="_images/f2b3bab632ebb05954382d982094949c53e9fa82031d511d73342d3e4e5f5c6f.svg" />
</div>
</div>
<p>Because the recombination rate in from position 10 to 20 is so much
higher than the rate from 0 to 10, we can see that all the recombinations
have fallen in the high recombination region.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<ul class="simple">
<li><p>See the <a class="reference internal" href="rate_maps.html#sec-rate-maps-creating"><span class="std std-ref">Creating rate maps</span></a> section for more examples of
creating <a class="reference internal" href="api.html#msprime.RateMap" title="msprime.RateMap"><code class="xref py py-class docutils literal notranslate"><span class="pre">RateMap</span></code></a> instances.</p></li>
<li><p>See the <a class="reference internal" href="api.html#msprime.RateMap.read_hapmap" title="msprime.RateMap.read_hapmap"><code class="xref py py-meth docutils literal notranslate"><span class="pre">RateMap.read_hapmap()</span></code></a> method for reading in
genetic maps from text files.</p></li>
</ul>
</div>
</section>
<section id="gene-conversion">
<span id="sec-ancestry-gene-conversion"></span><h3>Gene conversion<a class="headerlink" href="#gene-conversion" title="Link to this heading">#</a></h3>
<p>Gene conversion events are defined by two parameters: the rate at which gene
conversion events are initiated and the distribution of tract lengths.
In the default case of discrete genome coordinates, tract lengths are drawn
from a geometric distribution with mean <code class="docutils literal notranslate"><span class="pre">gene_conversion_tract_length</span></code> (which
must be at least 1). Note that if we specify a tract length of 1, then all
gene conversion tracts will have length exactly 1.
In the following example one gene conversion event of length 1 has occurred:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">3</span><span class="p">,</span> <span class="n">gene_conversion_rate</span><span class="o">=</span><span class="mf">0.02</span><span class="p">,</span> <span class="n">gene_conversion_tract_length</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">sequence_length</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/594551ab5793662ae1acf51593b79289332c0fefb3e3a0283827e60b8fc736ee.svg" src="_images/594551ab5793662ae1acf51593b79289332c0fefb3e3a0283827e60b8fc736ee.svg" />
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>We can see the specific effect of gene conversion here in this simple
example, because the trees at either side of the event are identical.
However, this will not be true in general.</p>
</div>
<p>Continuous genomes can also be used. In this case the parameters define
the rate at which gene conversion events are initiated per unit of sequence
length and the mean of the exponentially distributed gene conversion tract
lengths. The following example shows the same simulation as above but for a
continuous genome:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">3</span><span class="p">,</span> <span class="n">gene_conversion_rate</span><span class="o">=</span><span class="mf">0.02</span><span class="p">,</span> <span class="n">gene_conversion_tract_length</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">sequence_length</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">discrete_genome</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/30585c2bc8d366ac5833b3976974443f5bccc20c9b8417a08998426492cb1392.svg" src="_images/30585c2bc8d366ac5833b3976974443f5bccc20c9b8417a08998426492cb1392.svg" />
</div>
</div>
<p>Recombination and gene conversion at constant rates can be simulated alongside.
In the following example recombinations at site 60 and 97 have occurred in addition
to a gene conversion event covering the tract from site 76 to site 80.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">3</span><span class="p">,</span> <span class="n">sequence_length</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">0.003</span><span class="p">,</span>
    <span class="n">gene_conversion_rate</span><span class="o">=</span><span class="mf">0.002</span><span class="p">,</span> <span class="n">gene_conversion_tract_length</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/13ee32edb33acbfd6e9764818c860f4699c7345a9619cb507dd98474ee670b90.svg" src="_images/13ee32edb33acbfd6e9764818c860f4699c7345a9619cb507dd98474ee670b90.svg" />
</div>
</div>
<p>Variable recombination rates and constant gene conversion rates
can also be combined.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Variable rates of gene conversion breakpoint initiation are not currently
supported, but are planned for a future release. If you are interested
in this feature please let us know!</p>
</div>
</section>
<section id="discrete-or-continuous">
<span id="sec-ancestry-discrete-genome"></span><h3>Discrete or continuous?<a class="headerlink" href="#discrete-or-continuous" title="Link to this heading">#</a></h3>
<p>By default, we assume that the genome we are simulating is <em>discrete</em>
so that genome coordinates are at integer positions:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">3</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">sequence_length</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/f69d27abed270fb36dffa44791dcf9c7e7571da1d2708c05bb57affa5172f87a.svg" src="_images/f69d27abed270fb36dffa44791dcf9c7e7571da1d2708c05bb57affa5172f87a.svg" />
</div>
</div>
<p>Alternatively, we can simulate a continous genome by setting
<code class="docutils literal notranslate"><span class="pre">discrete_genome=False</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">3</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">0.25</span><span class="p">,</span> <span class="n">sequence_length</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">discrete_genome</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">33</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/788cd898c2563579023b291cedd69ab74ae578c66ba63e3b1a8d71eda966ad35.svg" src="_images/788cd898c2563579023b291cedd69ab74ae578c66ba63e3b1a8d71eda966ad35.svg" />
</div>
</div>
<p>Here we see that the breakpoints along the genome occur at floating point
positions. Simulating a continuous genome sequence can be useful for
theoretical work, but we recommend using discrete coordinates for most
purposes.</p>
</section>
<section id="multiple-chromosomes">
<span id="sec-ancestry-multiple-chromosomes"></span><h3>Multiple chromosomes<a class="headerlink" href="#multiple-chromosomes" title="Link to this heading">#</a></h3>
<section id="do-i-need-to-do-this">
<h4>Do I need to do this?<a class="headerlink" href="#do-i-need-to-do-this" title="Link to this heading">#</a></h4>
<p>The main purpose of simulating multiple chromosomes together (instead of
running independent simulations for each chromosome) is to capture
correlations between trees on distinct chromosomes. In many scenarios,
<strong>simulating multiple chromosomes may not be necessary</strong>, and independent
simulations may be preferable as they will be much more efficient.</p>
<p>That being said, this section describes how to simulate data over
multiple chromosomes, or more generally, over multiple regions
with free recombination between them.</p>
</section>
<section id="using-a-fixed-pedigree">
<h4>Using a fixed pedigree<a class="headerlink" href="#using-a-fixed-pedigree" title="Link to this heading">#</a></h4>
<p>The most natural way to simulate multiple regions with free recombination for
a set of sampled individuals is by simulating through a known pedigree. The
<a class="reference internal" href="#sec-ancestry-models-fixed-pedigree"><span class="std std-ref">fixed pedigree</span></a> simulation model
constrains simulated genomic regions by an input pedigree, which specifies
parent-offspring relationships. Since autosomal genetic material in an
offspring has equal chance of being inherited from either their mother or their
father, independent simulations through the same pedigree ensures that those
regions are inherited from different parents or from the same parent each with
probability 1/2. This ensures free recombination between independent
simulations.</p>
<p>We often do not have a full pedigree, but instead have a pedigree that
extends some number of generations into the past. Because a recombination
fraction of 1/2 between regions causes correlations between chromosomes to
decay rapidly, even shallow pedigrees containing just a handful of generations
should broadly capture expected levels of correlations. At the termination of
the pedigree, the simulation can be completed using the DTWF, Hudson, or other
model. See <a class="reference internal" href="#sec-multi-chrom-example-1"><span class="std std-ref">example 1</span></a> below for a small
example of this procedure, and more details on completing pedigree simulations
can be found in the <a class="reference internal" href="#sec-ancestry-models-fixed-pedigree"><span class="std std-ref">Fixed pedigree</span></a>
section.</p>
<p>Why might simulating multiple chromosomes using a pedigree be preferable to
other simulation options? First, there is no need to concatenate recombination
maps for multiple chromosomes, as each region is simulated independently. This
also results in simpler, reusable code which can be parallelized. Finally, this
approach improves efficiency over a single larger simulation of multiple
concatenated chromosomes, as computational costs scale as the square of the
sequence length.</p>
</section>
<section id="simulations-without-a-fixed-pedigree">
<h4>Simulations without a fixed pedigree<a class="headerlink" href="#simulations-without-a-fixed-pedigree" title="Link to this heading">#</a></h4>
<p>Without a pedigree, we can use other simulation models in msprime such as
<a class="reference internal" href="#sec-ancestry-models-dtwf"><span class="std std-ref">DTWF</span></a> and the
<a class="reference internal" href="#sec-ancestry-models-multiple-mergers"><span class="std std-ref">multiple merger coalescent</span></a>.
These models do not directly support simulating multiple chromosomes
simultaneously, but we can emulate it using a single linear genome split into
chromosome segments. In this case, multiple chromosomes are modelled by specifying
a recombination map with single base-pair segments with recombination
probability 1/2 separating adjacent chromosomes.
The probability that two chromosomes are co-inherited across <span class="math notranslate nohighlight">\(t\)</span> generations
is <span class="math notranslate nohighlight">\(2^{-t}\)</span>. In the Poisson model of recombination used by msprime, the
probability of co-inheritance across <span class="math notranslate nohighlight">\(t\)</span> generations of two adjacent base
pairs separated by a recombination rate <span class="math notranslate nohighlight">\(r\)</span> per base pair is <span class="math notranslate nohighlight">\(e^{-rt}\)</span>.
To make these agree, we set the recombination rate in the base pair segments
separating chromosomes to <span class="math notranslate nohighlight">\(\log(2)\)</span>.</p>
<p><a class="reference external" href="https://doi.org/10.1371/journal.pgen.1008619">Nelson et al. 2020</a> describes
a few cases where simulating multiple chromosomes using the DTWF is important
for accurately modeling shared patterns of variation between samples.
In large simulated cohorts, where many samples can be close
relatives, we need the DTWF model and multiple chromosomes to obtain the
correct distribution of the number and total length of shared IBD segments. We
similarly require multiple chromosomes and the DTWF model to get the correct
ancestry patterns for very recently admixed populations.</p>
<p>Additionally,
<a class="reference internal" href="#sec-ancestry-models-multiple-mergers"><span class="std std-ref">Multiple merger coalescents</span></a>
result in positively correlated ancestries between unlinked chromosomes
(see <a class="reference external" href="https://www.genetics.org/content/193/1/255">Birkner et al. 2013</a>).
This correlation does not break down in time and multiple chromosomes should
not be simulated independently under the multiple merger models. Unlinked
chromosomes should scatter into distinct ancestors instantaneously under
multiple merger coalescents, not with probability 1/2 per generation
as in discrete population models. In msprime, this waiting time of length zero
is approximated by an exponential distribution with rate <span class="math notranslate nohighlight">\(r\)</span>, i.e. the
recombination rate at the base pair separating the chromosomes, which
should be set to a high enough value to obtain an acceptable approximation.
We recommend considering in particular the relative error in comparison to
the magnitudes other waiting times which are likely to arise in the simulation.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>Simulations of multiple chromosomes under either DTWF or the multiple merger
models should use a <a class="reference internal" href="#sec-ancestry-discrete-genome"><span class="std std-ref">discrete genome</span></a>.
While it is possible to set up such a simulation using a continuous genome,
we <em>strongly</em> recommend against it, as chromosome divisions cannot be defined
as discrete break points and the simulation will be very inefficient.</p>
</div>
</section>
<section id="example-1-simulating-with-a-pedigree">
<span id="sec-multi-chrom-example-1"></span><h4>Example 1: simulating with a pedigree<a class="headerlink" href="#example-1-simulating-with-a-pedigree" title="Link to this heading">#</a></h4>
<p>In this example, we simulate genomic data for a pair of first cousins in
a small pedigree extending 3 generations into the past. See
<a class="reference internal" href="#sec-ancestry-models-fixed-pedigree"><span class="std std-ref">here</span></a> for details on specifying,
importing, and simulating using pedigrees.</p>
<p>Let’s first specify our pedigree:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">io</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">tskit</span>
<span class="n">ped_txt</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span><span class="se">\</span>
<span class="s2"># id parent0 parent1 time is_sample</span>
<span class="s2">0   2   3   0.0 1</span>
<span class="s2">1   4   5   0.0 1</span>
<span class="s2">2   6   7   1.0 0</span>
<span class="s2">3   8   9   1.0 0</span>
<span class="s2">4   6   7   1.0 0</span>
<span class="s2">5   10  11  1.0 0</span>
<span class="s2">6   .   .   2.0 0</span>
<span class="s2">7   .   .   2.0 0</span>
<span class="s2">8   .   .   2.0 0</span>
<span class="s2">9   .   .   2.0 0</span>
<span class="s2">10  .   .   2.0 0</span>
<span class="s2">11  .   .   2.0 0</span>
<span class="s2">&quot;&quot;&quot;</span>
<span class="n">pedigree</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">parse_pedigree</span><span class="p">(</span><span class="n">io</span><span class="o">.</span><span class="n">StringIO</span><span class="p">(</span><span class="n">ped_txt</span><span class="p">),</span> <span class="n">sequence_length</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>

<span class="n">draw_pedigree</span><span class="p">(</span><span class="n">pedigree</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/fa85920be4a013a610d5a8614ade033bea6f66c9b07449105e12b32a9a3ae727.svg" src="_images/fa85920be4a013a610d5a8614ade033bea6f66c9b07449105e12b32a9a3ae727.svg" />
</div>
</div>
<p>Then suppose we would like to simulate three chromosomes of differing lengths
and recombination rates. In practice, we might have recombination
maps for each chromosome, but the general strategy remains the same.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Ls</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1000000</span><span class="p">,</span> <span class="mi">2000000</span><span class="p">,</span> <span class="mi">3000000</span><span class="p">]</span>
<span class="n">rs</span> <span class="o">=</span> <span class="p">[</span><span class="mf">1e-8</span><span class="p">,</span> <span class="mf">2e-8</span><span class="p">,</span> <span class="mf">3e-8</span><span class="p">]</span>

<span class="n">ts_chroms</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">pedigree</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">parse_pedigree</span><span class="p">(</span><span class="n">io</span><span class="o">.</span><span class="n">StringIO</span><span class="p">(</span><span class="n">ped_txt</span><span class="p">),</span> <span class="n">sequence_length</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">L</span><span class="p">,</span> <span class="n">r</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">Ls</span><span class="p">,</span> <span class="n">rs</span><span class="p">)):</span>
    <span class="n">pedigree</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">=</span> <span class="n">L</span>
    <span class="n">ped_ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
        <span class="n">initial_state</span><span class="o">=</span><span class="n">pedigree</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s2">&quot;fixed_pedigree&quot;</span><span class="p">,</span>
        <span class="n">recombination_rate</span><span class="o">=</span><span class="n">r</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">ts_chroms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
        <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
            <span class="n">initial_state</span><span class="o">=</span><span class="n">ped_ts</span><span class="p">,</span> <span class="n">population_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
            <span class="n">recombination_rate</span><span class="o">=</span><span class="n">r</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s2">&quot;dtwf&quot;</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="n">i</span><span class="o">+</span><span class="mi">100</span><span class="p">))</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">ts</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ts_chroms</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;chromosome </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2"> has length </span><span class="si">{</span><span class="n">ts</span><span class="o">.</span><span class="n">sequence_length</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">ts</span><span class="o">.</span><span class="n">num_trees</span><span class="si">}</span><span class="s2"> trees&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>chromosome 0 has length 1000000.0 and 48 trees
chromosome 1 has length 2000000.0 and 155 trees
chromosome 2 has length 3000000.0 and 377 trees
</pre></div>
</div>
</div>
</div>
<p>This results in three chromosomes of the desired lengths.</p>
</section>
<section id="example-2-simulating-without-a-pedigree">
<span id="sec-multi-chrom-example-2"></span><h4>Example 2: simulating without a pedigree<a class="headerlink" href="#example-2-simulating-without-a-pedigree" title="Link to this heading">#</a></h4>
<p>We first set up the chromosome coordinates, recombination rates, and the
corresponding recombination map. The following defines a recombination map for
three chromosomes each 1 cM in length:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">math</span>

<span class="n">r_chrom</span> <span class="o">=</span> <span class="mf">1e-8</span>
<span class="n">r_break</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
<span class="n">chrom_positions</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mf">1e6</span><span class="p">,</span> <span class="mf">2e6</span><span class="p">,</span> <span class="mf">3e6</span><span class="p">]</span>
<span class="n">map_positions</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">chrom_positions</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
    <span class="n">chrom_positions</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
    <span class="n">chrom_positions</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
    <span class="n">chrom_positions</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span>
    <span class="n">chrom_positions</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
    <span class="n">chrom_positions</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
<span class="p">]</span>
<span class="n">rates</span> <span class="o">=</span> <span class="p">[</span><span class="n">r_chrom</span><span class="p">,</span> <span class="n">r_break</span><span class="p">,</span> <span class="n">r_chrom</span><span class="p">,</span> <span class="n">r_break</span><span class="p">,</span> <span class="n">r_chrom</span><span class="p">]</span>
<span class="n">rate_map</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">RateMap</span><span class="p">(</span><span class="n">position</span><span class="o">=</span><span class="n">map_positions</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="n">rates</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We cannot use the default <a class="reference internal" href="api.html#msprime.StandardCoalescent" title="msprime.StandardCoalescent"><code class="xref py py-class docutils literal notranslate"><span class="pre">StandardCoalescent</span></code></a> model to run simulations
of multiple chromosomes; doing so would be equivalent to running independent
replicate simulations and <strong>much</strong> less efficient. To get the correct
correlation between chromosomes, we must use a model like
<a class="reference internal" href="#sec-ancestry-models-dtwf"><span class="std std-ref">discrete-time Wright-Fisher</span></a>.
However, because correlations
between chromosomes break down very rapidly, we only need to simulate using
<code class="docutils literal notranslate"><span class="pre">dtwf</span></code> for roughly 10-20 generations, after which we can switch to the <code class="docutils literal notranslate"><span class="pre">hudson</span></code>
model to finish the simulation more efficiently (see the
<a class="reference internal" href="#sec-ancestry-models-specifying"><span class="std std-ref">Specifying ancestry models</span></a> section for more information on switching
between ancestry models).</p>
<p>In this example, we simulate 10 sampled diploid individuals using the
recombination map defined above for three chromosomes and switching from the
<code class="docutils literal notranslate"><span class="pre">dtwf</span></code> to <code class="docutils literal notranslate"><span class="pre">hudson</span></code> model 20 generations ago:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">10</span><span class="p">,</span>
    <span class="n">population_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
    <span class="n">recombination_rate</span><span class="o">=</span><span class="n">rate_map</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="p">[</span>
        <span class="n">msprime</span><span class="o">.</span><span class="n">DiscreteTimeWrightFisher</span><span class="p">(</span><span class="n">duration</span><span class="o">=</span><span class="mi">20</span><span class="p">),</span>
        <span class="n">msprime</span><span class="o">.</span><span class="n">StandardCoalescent</span><span class="p">(),</span>
    <span class="p">],</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">1234</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">ts</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
              <style>
                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}
                .tskit-table tbody tr td {padding: 0.5em 0.5em;}
                .tskit-table tbody tr td:first-of-type {text-align: left;}
                .tskit-details-label {vertical-align: top; padding-right:5px;}
                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}
                .tskit-table-set-table {margin: 12px 0 0 12px;}
                details {display: inline-block;}
                summary {cursor: pointer; outline: 0; display: list-item;}
              </style>
              <div class="tskit-table-set">
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="padding:0;line-height:21px;">
                          <img style="height: 32px;display: inline-block;padding: 3px 5px 3px 0;" src="https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg"/>
                          <a target="_blank" href="https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class"> Tree Sequence </a>
                        </th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr><td>Trees</td><td>351</td></tr>
                      <tr><td>Sequence Length</td><td>3000000.0</td></tr>
                      <tr><td>Time Units</td><td>generations</td></tr>
                      <tr><td>Sample Nodes</td><td>20</td></tr>
                      <tr><td>Total Size</td><td>60.3 KiB</td></tr>
                      <tr>
                        <td>Metadata</td><td style="text-align: left;">No Metadata</td></tr>
                    </tbody>
                  </table>
                </div>
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="line-height:21px;">Table</th>
                        <th>Rows</th>
                        <th>Size</th>
                        <th>Has Metadata</th>
                      </tr>
                    </thead>
                    <tbody>
                    
                  <tr>
                    <td>Edges</td>
                      <td>1254</td>
                      <td>39.2 KiB</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Individuals</td>
                      <td>10</td>
                      <td>304 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Migrations</td>
                      <td>0</td>
                      <td>8 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Mutations</td>
                      <td>0</td>
                      <td>16 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Nodes</td>
                      <td>342</td>
                      <td>9.4 KiB</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Populations</td>
                      <td>1</td>
                      <td>224 Bytes</td>
                      <td style="text-align: center;">
                        ✅
                      </td>
                    </tr>
                
                  <tr>
                    <td>Provenances</td>
                      <td>1</td>
                      <td>1.4 KiB</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Sites</td>
                      <td>0</td>
                      <td>16 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
            </div></div>
</div>
<p>To place each chromosome in its own separate tree sequence, we can trim the full tree
sequence for each chromosome by defining the end points of each chromosome and
using the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.keep_intervals" title="(in Project name not set)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">keep_intervals()</span></code></a> method. It is important to
specify <code class="docutils literal notranslate"><span class="pre">simplify=False</span></code> so that node indices remain consistent across
chromosomes. The <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.trim" title="(in Project name not set)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trim()</span></code></a> method is then used to trim
the tree sequences to the focal chromosome, so that position indices begin at
0 within each chromosome tree sequence.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts_chroms</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">chrom_positions</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
    <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">chrom_positions</span><span class="p">[</span><span class="n">j</span><span class="p">:</span> <span class="n">j</span> <span class="o">+</span> <span class="mi">2</span><span class="p">]</span>
    <span class="n">chrom_ts</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">keep_intervals</span><span class="p">([[</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">]],</span> <span class="n">simplify</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">trim</span><span class="p">()</span>
    <span class="n">ts_chroms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">chrom_ts</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">chrom_ts</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">)</span>

<span class="c1"># the third chromosome</span>
<span class="n">chrom_ts</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>1000000.0
1000000.0
1000000.0
</pre></div>
</div>
<div class="output text_html">
            <div>
              <style>
                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}
                .tskit-table tbody tr td {padding: 0.5em 0.5em;}
                .tskit-table tbody tr td:first-of-type {text-align: left;}
                .tskit-details-label {vertical-align: top; padding-right:5px;}
                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}
                .tskit-table-set-table {margin: 12px 0 0 12px;}
                details {display: inline-block;}
                summary {cursor: pointer; outline: 0; display: list-item;}
              </style>
              <div class="tskit-table-set">
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="padding:0;line-height:21px;">
                          <img style="height: 32px;display: inline-block;padding: 3px 5px 3px 0;" src="https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg"/>
                          <a target="_blank" href="https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class"> Tree Sequence </a>
                        </th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr><td>Trees</td><td>102</td></tr>
                      <tr><td>Sequence Length</td><td>1000000.0</td></tr>
                      <tr><td>Time Units</td><td>generations</td></tr>
                      <tr><td>Sample Nodes</td><td>20</td></tr>
                      <tr><td>Total Size</td><td>26.7 KiB</td></tr>
                      <tr>
                        <td>Metadata</td><td style="text-align: left;">No Metadata</td></tr>
                    </tbody>
                  </table>
                </div>
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="line-height:21px;">Table</th>
                        <th>Rows</th>
                        <th>Size</th>
                        <th>Has Metadata</th>
                      </tr>
                    </thead>
                    <tbody>
                    
                  <tr>
                    <td>Edges</td>
                      <td>371</td>
                      <td>11.6 KiB</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Individuals</td>
                      <td>10</td>
                      <td>304 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Migrations</td>
                      <td>0</td>
                      <td>8 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Mutations</td>
                      <td>0</td>
                      <td>16 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Nodes</td>
                      <td>342</td>
                      <td>9.4 KiB</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Populations</td>
                      <td>1</td>
                      <td>224 Bytes</td>
                      <td style="text-align: center;">
                        ✅
                      </td>
                    </tr>
                
                  <tr>
                    <td>Provenances</td>
                      <td>3</td>
                      <td>2.3 KiB</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Sites</td>
                      <td>0</td>
                      <td>16 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
            </div></div>
</div>
<p>This gives us a list of tree sequences, one for each chromosome, in the order
that they were stitched together in the initial recombination map.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Many recombination maps inferred from data will contain regions at the start
and end of the chromosome where recombination rates could not be inferred.
These regions typically are assigned rate values of NaN. If we were to
concatenate multiple recombination maps that have flanking NaN regions, the
resulting maps would have regions with NaN rates in the interior of the map,
which will produce an error when using such a map in simulations.</p>
<p>These regions typically represent unassay-able segments of the chromosome, and
simulated genetic material in these regions should be discarded in downstream
analyses. Thus, the suggested approach is to modify the recombination maps to
set rates to zero in these flanking regions, concatenate the maps as shown in
this example, and then remove those regions of the resulting tree sequences by
calling <code class="docutils literal notranslate"><span class="pre">ts.delete_intervals(nan_intervals)</span></code>, where <code class="docutils literal notranslate"><span class="pre">nan_intervals</span></code> is a list
containing <code class="docutils literal notranslate"><span class="pre">[start,</span> <span class="pre">end]</span></code> values for each NaN flanking region.</p>
</div>
</section>
</section>
</section>
<section id="recording-more-information">
<h2>Recording more information<a class="headerlink" href="#recording-more-information" title="Link to this heading">#</a></h2>
<p>By default <code class="docutils literal notranslate"><span class="pre">msprime</span></code> stores the minimum amount of information required
to represent the simulated genealogical history of the samples. Sometimes
we are interested in more detailed information; this section gives details
of options that allow us to do this.</p>
<section id="additional-nodes">
<span id="sec-ancestry-additional-nodes"></span><h3>Additional nodes<a class="headerlink" href="#additional-nodes" title="Link to this heading">#</a></h3>
<p>In <code class="docutils literal notranslate"><span class="pre">msprime</span></code> we usually want to simulate the coalescent with recombination
and represent the output as efficiently as possible. As a result, we don’t
store individual recombination events, but rather their effects on the output
tree sequence. We also do not explicitly store common ancestor events that
do not result in marginal coalescences. For some purposes, however,
we want record information on other events of interest, not just the mimimal
representation of its outcome.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">additional_nodes</span></code> and
<a class="reference internal" href="#sec-additional-nodes-cso"><span class="std std-ref">coalescing_segments_only</span></a> options serve
this exact purpose. These options allow us to record the nodes associated with
a custom subset of all events we might observe in the history of a sample.
Besides samples and coalescence events, nodes can now also represent
<a class="reference internal" href="#sec-additional-nodes-ca"><span class="std std-ref">common ancestor events</span></a>,
<a class="reference internal" href="#sec-additional-nodes-re"><span class="std std-ref">recombination</span></a>,
<a class="reference internal" href="#sec-additional-nodes-re"><span class="std std-ref">gene conversion</span></a>,
<a class="reference internal" href="#sec-ancestry-record-migrations"><span class="std std-ref">migration</span></a>,
and <a class="reference internal" href="#sec-additional-nodes-re"><span class="std std-ref">pass through</span></a> events. The example below
and the next few paragraphs provide a guide on how
to record additional nodes and interpret the resulting node tables.</p>
<p>We first set up a simple pedigree simulation. Pedigrees are an
interesting starting point as in contrast to the coalescent models a single
node might be associated with more than one type of event. Conversely,
for the continuous-time coalescent model simulated by default by msprime,
each event is registered as a new node.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pb</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">PedigreeBuilder</span><span class="p">()</span>
<span class="n">mp_21</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">add_individual</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">dp_21</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">add_individual</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">mp_11</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">add_individual</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="p">[</span><span class="n">mp_21</span><span class="p">,</span> <span class="n">dp_21</span><span class="p">])</span>
<span class="n">dp_11</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">add_individual</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="p">[</span><span class="n">mp_21</span><span class="p">,</span> <span class="n">dp_21</span><span class="p">])</span>
<span class="n">mp_12</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">add_individual</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="p">[</span><span class="n">mp_21</span><span class="p">,</span> <span class="n">dp_21</span><span class="p">])</span>
<span class="n">dp_12</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">add_individual</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="p">[</span><span class="n">mp_21</span><span class="p">,</span> <span class="n">dp_21</span><span class="p">])</span>
<span class="n">pb</span><span class="o">.</span><span class="n">add_individual</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="p">[</span><span class="n">mp_11</span><span class="p">,</span> <span class="n">dp_12</span><span class="p">],</span> <span class="n">is_sample</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">pb</span><span class="o">.</span><span class="n">add_individual</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="p">[</span><span class="n">mp_11</span><span class="p">,</span> <span class="n">dp_12</span><span class="p">],</span> <span class="n">is_sample</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">pb</span><span class="o">.</span><span class="n">add_individual</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="p">[</span><span class="n">mp_12</span><span class="p">,</span> <span class="n">dp_11</span><span class="p">],</span> <span class="n">is_sample</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">pedigree_tables</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">finalise</span><span class="p">(</span><span class="n">sequence_length</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="n">draw_pedigree</span><span class="p">(</span><span class="n">pedigree_tables</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/b222f92352dbe777e0aa454eb03bc7995e9a18795f62a3bacd540fd2ad432523.svg" src="_images/b222f92352dbe777e0aa454eb03bc7995e9a18795f62a3bacd540fd2ad432523.svg" />
</div>
</div>
<p>To specify the particular node types we want to store information on, pass a
<a class="reference internal" href="api.html#msprime.NodeType" title="msprime.NodeType"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeType</span></code></a> object to the <code class="docutils literal notranslate"><span class="pre">additional_nodes</span></code> option of
<a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a>. This class extends <a class="reference external" href="https://docs.python.org/3/library/enum.html#enum.Flag" title="(in Python v3.13)"><code class="docutils literal notranslate"><span class="pre">enum.Flag</span></code></a>.
Each node type in the enumeration is associated with a numerical constant.
Different node types can be combined and compared using
<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html?highlight=bitwise">bitwise operations</a>.
At the same time, these constant also function as the flags identifying
the type of each node in the nodes table. Here, we take the bitwise union
of three members of the enumeration:
<a class="reference internal" href="api.html#msprime.NodeType.RECOMBINANT" title="msprime.NodeType.RECOMBINANT"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.NodeType.RECOMBINANT</span></code></a>, <a class="reference internal" href="api.html#msprime.NodeType.PASS_THROUGH" title="msprime.NodeType.PASS_THROUGH"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.NodeType.PASS_THROUGH</span></code></a>
and <a class="reference internal" href="api.html#msprime.NodeType.COMMON_ANCESTOR" title="msprime.NodeType.COMMON_ANCESTOR"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.NodeType.COMMON_ANCESTOR</span></code></a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">initial_state</span><span class="o">=</span><span class="n">pedigree_tables</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="s1">&#39;fixed_pedigree&#39;</span><span class="p">,</span>
    <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span>
    <span class="n">additional_nodes</span><span class="o">=</span><span class="p">(</span>
         <span class="n">msprime</span><span class="o">.</span><span class="n">NodeType</span><span class="o">.</span><span class="n">RECOMBINANT</span> <span class="o">|</span>
         <span class="n">msprime</span><span class="o">.</span><span class="n">NodeType</span><span class="o">.</span><span class="n">PASS_THROUGH</span> <span class="o">|</span> 
         <span class="n">msprime</span><span class="o">.</span><span class="n">NodeType</span><span class="o">.</span><span class="n">COMMON_ANCESTOR</span>
         <span class="p">),</span>
    <span class="n">coalescing_segments_only</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>

<span class="k">def</span><span class="w"> </span><span class="nf">count_flags</span><span class="p">(</span><span class="n">ts</span><span class="p">):</span>
    <span class="n">counter</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">nodetype</span> <span class="ow">in</span> <span class="n">msprime</span><span class="o">.</span><span class="n">NodeType</span><span class="p">:</span>
        <span class="n">flags</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">bitwise_and</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">nodes_flags</span><span class="p">,</span> <span class="n">nodetype</span><span class="o">.</span><span class="n">value</span><span class="p">)</span>
        <span class="n">counter</span><span class="p">[</span><span class="n">nodetype</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">flags</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">counter</span>

<span class="nb">print</span><span class="p">(</span><span class="n">count_flags</span><span class="p">(</span><span class="n">ts</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{&#39;RECOMBINANT&#39;: np.int64(4), &#39;COMMON_ANCESTOR&#39;: np.int64(0), &#39;MIGRANT&#39;: np.int64(0), &#39;CENSUS&#39;: np.int64(0), &#39;GENE_CONVERSION&#39;: np.int64(0), &#39;PASS_THROUGH&#39;: np.int64(7)}
</pre></div>
</div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">count_flags</span></code> function demonstrates the use of
bitwise operations to determine the type of a node. This function iterates
across all node types and checks for all rows in the nodes table whether
the intersection of the basic node type and the node within the nodes table
is different from 0. It returns a <code class="docutils literal notranslate"><span class="pre">dict</span></code> containing the counts of all possible
node types.</p>
<p>Note that <a class="reference internal" href="#sec-additional-nodes-re"><span class="std std-ref">recombination events</span></a> are
associated with two nodes,
one for both ends that are created by recombination backwards in time. The
number of recombination events is thus half the number of recombination nodes.
<code class="docutils literal notranslate"><span class="pre">PASS_THROUGH</span></code> events occur when the ancestral material of only a single
lineage passes through the genome of an ancestor, making coalescence
impossible. This event can only be observed in models that track individuals:
<a class="reference internal" href="api.html#msprime.DiscreteTimeWrightFisher" title="msprime.DiscreteTimeWrightFisher"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.DiscreteTimeWrightFisher</span></code></a> and
<a class="reference internal" href="api.html#msprime.FixedPedigree" title="msprime.FixedPedigree"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.FixedPedigree</span></code></a> models.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">node_style_map</span> <span class="o">=</span> <span class="p">{</span>
    <span class="n">msprime</span><span class="o">.</span><span class="n">NodeType</span><span class="o">.</span><span class="n">RECOMBINANT</span><span class="o">.</span><span class="n">value</span><span class="p">:</span> <span class="s1">&#39;yellow&#39;</span><span class="p">,</span>
    <span class="n">msprime</span><span class="o">.</span><span class="n">NodeType</span><span class="o">.</span><span class="n">COMMON_ANCESTOR</span><span class="o">.</span><span class="n">value</span><span class="p">:</span> <span class="s1">&#39;red&#39;</span><span class="p">,</span>
    <span class="n">msprime</span><span class="o">.</span><span class="n">NodeType</span><span class="o">.</span><span class="n">PASS_THROUGH</span><span class="o">.</span><span class="n">value</span><span class="p">:</span> <span class="s1">&#39;blue&#39;</span><span class="p">,</span>
    <span class="p">(</span><span class="n">msprime</span><span class="o">.</span><span class="n">NodeType</span><span class="o">.</span><span class="n">RECOMBINANT</span> <span class="o">|</span> <span class="n">msprime</span><span class="o">.</span><span class="n">NodeType</span><span class="o">.</span><span class="n">PASS_THROUGH</span><span class="p">)</span><span class="o">.</span><span class="n">value</span> <span class="p">:</span> <span class="s1">&#39;green&#39;</span><span class="p">,</span>
    <span class="p">(</span><span class="n">msprime</span><span class="o">.</span><span class="n">NodeType</span><span class="o">.</span><span class="n">RECOMBINANT</span> <span class="o">|</span> <span class="n">msprime</span><span class="o">.</span><span class="n">NodeType</span><span class="o">.</span><span class="n">COMMON_ANCESTOR</span><span class="p">)</span><span class="o">.</span><span class="n">value</span> <span class="p">:</span> <span class="s1">&#39;orange&#39;</span>
<span class="p">}</span>

<span class="n">css_string</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">nodes</span><span class="p">():</span>
    <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">flags</span> <span class="ow">in</span> <span class="n">node_style_map</span><span class="p">:</span>
        <span class="n">css_string</span> <span class="o">+=</span> <span class="s2">&quot;.n</span><span class="si">%s</span><span class="s2"> &gt; .sym {fill: </span><span class="si">%s</span><span class="s2">}&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">node_style_map</span><span class="p">[</span><span class="n">node</span><span class="o">.</span><span class="n">flags</span><span class="p">])</span>
<span class="n">wide_fmt</span> <span class="o">=</span> <span class="p">(</span><span class="mi">800</span><span class="p">,</span> <span class="mi">250</span><span class="p">)</span>
<span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="n">css_string</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">wide_fmt</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/43a71e8958273618c96963c07b05d847ea8e6ab4f6aed060f4ef215516f65cdb.svg" src="_images/43a71e8958273618c96963c07b05d847ea8e6ab4f6aed060f4ef215516f65cdb.svg" />
</div>
</div>
<p>Similarly, the bitwise operations logic can be used to color the different nodes
in the tree sequence according to their <a class="reference internal" href="api.html#msprime.NodeType" title="msprime.NodeType"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeType</span></code></a>. Node 5 is both
<a class="reference internal" href="api.html#msprime.NodeType.RECOMBINANT" title="msprime.NodeType.RECOMBINANT"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.NodeType.RECOMBINANT</span></code></a> and
<a class="reference internal" href="api.html#msprime.NodeType.PASS_THROUGH" title="msprime.NodeType.PASS_THROUGH"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.NodeType.PASS_THROUGH</span></code></a> (green). Nodes 0, 1, 4 are
recombinant nodes (yellow). All other nodes are only associated with a
<code class="docutils literal notranslate"><span class="pre">PASS_THROUGH</span></code> event (blue). Note that we do not observe any <code class="docutils literal notranslate"><span class="pre">COMMON_ANCESTOR</span></code>
events. This means that all yellow nodes are also associated with a
coalesence event (see <a class="reference internal" href="#sec-additional-nodes-ca"><span class="std std-ref">common ancestor events</span></a>).
Note that in order to verify whether a node is <code class="docutils literal notranslate"><span class="pre">RECOMBINANT</span></code> <strong>and</strong> <code class="docutils literal notranslate"><span class="pre">PASS_THROUGH</span></code>,
we use the bitwise <strong>OR</strong> to define its associated <code class="docutils literal notranslate"><span class="pre">NodeType</span></code> constant.</p>
<p>To summarize: <code class="docutils literal notranslate"><span class="pre">additional_nodes</span></code> are specified using bitwise flags. Multiple
basic node types can be combined by taking the bitwise <code class="docutils literal notranslate"><span class="pre">OR</span></code> (|) and then
passed to the <code class="docutils literal notranslate"><span class="pre">additional_nodes</span></code> option. This enables us to track the specified
nodes across the genealogical history of the sample. By means of bitwise <code class="docutils literal notranslate"><span class="pre">AND</span></code>
(&amp;) we can then query the nodes table and check the type of each of the recorded
nodes.</p>
<p>If we wish to reduce these trees down to the minimal representation, we can use
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.simplify" title="(in Project name not set)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">tskit.TreeSequence.simplify()</span></code></a>. The resulting tree sequence will have
all of these unary nodes removed and will be equivalent to (but not identical,
due to stochastic effects) calling <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a> without the
<code class="docutils literal notranslate"><span class="pre">additional_nodes</span></code> or <code class="docutils literal notranslate"><span class="pre">coalescing_segment_only</span></code> argument(s).</p>
</section>
<section id="coalescing-segments-only">
<span id="sec-additional-nodes-cso"></span><h3>Coalescing segments only<a class="headerlink" href="#coalescing-segments-only" title="Link to this heading">#</a></h3>
<p>Note that the <code class="docutils literal notranslate"><span class="pre">coalescing_segments_only</span></code> option should be set to <code class="docutils literal notranslate"><span class="pre">False</span></code> when
recording additional nodes. Setting <code class="docutils literal notranslate"><span class="pre">coalescing_segments_only</span></code> to <code class="docutils literal notranslate"><span class="pre">False</span></code>
allows us to switch off the default simulator behaviour of only recording the
relationships between overlapping ancestry segments. Instead, you can now
record edges along the full extent of any of the ancestral lineages
that was involved in any of the events as specified by the <code class="docutils literal notranslate"><span class="pre">additional_nodes</span></code> flag.
This option can also be set to <code class="docutils literal notranslate"><span class="pre">False</span></code> without specifying any additional nodes.
When <code class="docutils literal notranslate"><span class="pre">coalescing_segments_only</span></code> is set to <code class="docutils literal notranslate"><span class="pre">False</span></code>,
edges that are a result of a coalescent event will record
the full length of the tracked ancestral material,
not just the overlapping segments of genome (as is the default behavior.
As a result, this option will produce
in nodes with only one child (unary nodes) along parts (unary regions) of
the genome.</p>
<p>For instance: suppose an ancestral segment ancestral to node <code class="docutils literal notranslate"><span class="pre">m</span></code> spans from <code class="docutils literal notranslate"><span class="pre">a</span></code> to <code class="docutils literal notranslate"><span class="pre">b</span></code> along the genome,
and another overlapping segment ancestral to node <code class="docutils literal notranslate"><span class="pre">n</span></code> spans from <code class="docutils literal notranslate"><span class="pre">c</span></code> to <code class="docutils literal notranslate"><span class="pre">d</span></code>, with <code class="docutils literal notranslate"><span class="pre">a</span> <span class="pre">&lt;</span> <span class="pre">c</span> <span class="pre">&lt;</span> <span class="pre">b</span> <span class="pre">&lt;</span> <span class="pre">d</span></code>.
If the two coalesce to create node <code class="docutils literal notranslate"><span class="pre">p</span></code>, if <code class="docutils literal notranslate"><span class="pre">coalescing_segments_only=True</span></code> (the default)
then edges from <code class="docutils literal notranslate"><span class="pre">m</span></code> and <code class="docutils literal notranslate"><span class="pre">n</span></code> to <code class="docutils literal notranslate"><span class="pre">p</span></code> over the genomic segment <code class="docutils literal notranslate"><span class="pre">[c,b)</span></code> would be created,
and the ancestry of the remaining segments (i.e., <code class="docutils literal notranslate"><span class="pre">[a,c)</span></code> for <code class="docutils literal notranslate"><span class="pre">m</span></code> and <code class="docutils literal notranslate"><span class="pre">[b,d)</span></code> for <code class="docutils literal notranslate"><span class="pre">n</span></code>) would be left unspecified
until further coalescent events.
However, if <code class="docutils literal notranslate"><span class="pre">coalescing_segments_only=False</span></code>, then the edges recorded
would be from <code class="docutils literal notranslate"><span class="pre">m</span></code> to <code class="docutils literal notranslate"><span class="pre">p</span></code> over the entire segment <code class="docutils literal notranslate"><span class="pre">[a,</span> <span class="pre">b)</span></code>, and from <code class="docutils literal notranslate"><span class="pre">n</span></code> to <code class="docutils literal notranslate"><span class="pre">p</span></code> over the entire segment <code class="docutils literal notranslate"><span class="pre">[c,d)</span></code>.
The nodes <code class="docutils literal notranslate"><span class="pre">m</span></code> and <code class="docutils literal notranslate"><span class="pre">n</span></code> coalesce (in <code class="docutils literal notranslate"><span class="pre">p</span></code>) on only the overlapping segment <code class="docutils literal notranslate"><span class="pre">[c,b)</span></code>, and so the node <code class="docutils literal notranslate"><span class="pre">p</span></code> will be a unary node in the flanking regions: above <code class="docutils literal notranslate"><span class="pre">m</span></code> on the segment <code class="docutils literal notranslate"><span class="pre">[a,c)</span></code> and above <code class="docutils literal notranslate"><span class="pre">n</span></code> on the segment <code class="docutils literal notranslate"><span class="pre">[b,d)</span></code>.</p>
</section>
<section id="common-ancestor-events">
<span id="sec-additional-nodes-ca"></span><h3>Common ancestor events<a class="headerlink" href="#common-ancestor-events" title="Link to this heading">#</a></h3>
<p>We distinguish two types of common ancestor events: coalescence and common
ancestor (in the strict sense) events.</p>
<p>Coalescence events are the default node type (associated with value 0) and are
therefore only implicitly part of the <a class="reference internal" href="api.html#msprime.NodeType" title="msprime.NodeType"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeType</span></code></a> enumeration.
In contrast, whenever two nonoverlapping ancestral segments
are found to have inherited from the same ancestral genome,
there is no coalescence event, and so we register this in the node table as
<a class="reference internal" href="api.html#msprime.NodeType.COMMON_ANCESTOR" title="msprime.NodeType.COMMON_ANCESTOR"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.NodeType.COMMON_ANCESTOR</span></code></a>.
Finally, when keeping track of individuals
(<a class="reference internal" href="api.html#msprime.DiscreteTimeWrightFisher" title="msprime.DiscreteTimeWrightFisher"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.DiscreteTimeWrightFisher</span></code></a>,
<a class="reference internal" href="api.html#msprime.FixedPedigree" title="msprime.FixedPedigree"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.FixedPedigree</span></code></a>), we might observe genomes (ploids) through
which the ancestral material of only a single lineage passes.
Such genomes can be registered in the table as
<a class="reference internal" href="api.html#msprime.NodeType.PASS_THROUGH" title="msprime.NodeType.PASS_THROUGH"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.NodeType.PASS_THROUGH</span></code></a> nodes.
Although this is obviously not a common ancestor event, note that for both
of these models each node has to carry at least one of these three flags.</p>
</section>
<section id="recombination-events">
<span id="sec-additional-nodes-re"></span><h3>Recombination events<a class="headerlink" href="#recombination-events" title="Link to this heading">#</a></h3>
<p>Additional nodes can also be used to mark recombination events (cross over) as well
as gene conversion. A seperate <a class="reference internal" href="api.html#msprime.NodeType" title="msprime.NodeType"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.NodeType</span></code></a> exists for each:
<a class="reference internal" href="api.html#msprime.NodeType.RECOMBINANT" title="msprime.NodeType.RECOMBINANT"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.NodeType.RECOMBINANT</span></code></a> and <a class="reference internal" href="api.html#msprime.NodeType.GENE_CONVERSION" title="msprime.NodeType.GENE_CONVERSION"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.NodeType.GENE_CONVERSION</span></code></a>.</p>
<p>A recombination event results in two extra nodes being recorded, one identifying
the genome providing the genetic material to the left of the breakpoint and
the other identifying the genome providing the genetic material to the
right. For more information on how to set the recombination rate or specify a
recombination map: see <a class="reference internal" href="#sec-ancestry-recombination"><span class="std std-ref">Recombination</span></a>. More information on how
to set up a simulation with gene conversion can be found here:
<a class="reference internal" href="#sec-ancestry-gene-conversion"><span class="std std-ref">Gene conversion</span></a>.</p>
</section>
<section id="migration-events">
<span id="sec-ancestry-record-migrations"></span><h3>Migration events<a class="headerlink" href="#migration-events" title="Link to this heading">#</a></h3>
<p>When running simulations with <a class="reference internal" href="demography.html#sec-demography"><span class="std std-ref">demography</span></a>, by
default, migration events are not recorded in the resulting tree sequence.
While the population associated with each node is determined by the simulated migration
events, knowledge of the
precise sequence of migrations is useful in a number of situations: for example see
the <code class="docutils literal notranslate"><span class="pre">tskit</span></code>
<a class="reference external" href="https://tskit-dev.github.io/tutorials/introgression.html">tutorial on introgression.</a></p>
<p>The <code class="docutils literal notranslate"><span class="pre">record_migrations</span></code> parameter allows us to record these events in the
simulated tree sequence outputted by <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a>. Briefly, a migration record
encodes the node associated with the migration, the time of the event, the source and
destination populations (backwards in time), and the left and right coordinates of the
migrating segement. For more details on migration records, see the
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-migration-table-definition" title="(in Project name not set)"><span class="xref std std-ref">migration table definition.</span></a></p>
<p>Here, we provide a simple example of the effect of setting <code class="docutils literal notranslate"><span class="pre">record_migrations=True</span></code>
using the <a class="reference internal" href="api.html#msprime.Demography.stepping_stone_model" title="msprime.Demography.stepping_stone_model"><code class="xref py py-meth docutils literal notranslate"><span class="pre">stepping</span> <span class="pre">stone</span> <span class="pre">model</span></code></a>
where migration is permitted between adjacent populations. Additionally, we
set <code class="docutils literal notranslate"><span class="pre">additional_nodes=msprime.NodeType.MIGRANT</span></code>
(see <a class="reference internal" href="#sec-ancestry-additional-nodes"><span class="std std-ref">previous section</span></a>)
to record nodes corresponding to migration events. This is not necessary, but will be
helpful to visualise the result.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">N</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">stepping_stone_model</span><span class="p">(</span>
    <span class="p">[</span><span class="mi">20</span><span class="p">]</span> <span class="o">*</span> <span class="n">N</span><span class="p">,</span> <span class="n">migration_rate</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">boundaries</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="n">N</span> <span class="o">-</span> <span class="mi">1</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span>
    <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">,</span>
    <span class="n">record_migrations</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">additional_nodes</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">NodeType</span><span class="o">.</span><span class="n">MIGRANT</span><span class="p">,</span>
    <span class="n">coalescing_segments_only</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>In the following visualisation, we trace the migration history of two samples backwards
in time to their most recent common ancestor. Each one of the ten populations in the
model is assigned a
colour and a position on the x axis. The time of migration events is recorded on the y
axis.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">tskit</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>

<span class="n">cmap</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">colormaps</span><span class="p">[</span><span class="s1">&#39;viridis&#39;</span><span class="p">]</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">first</span><span class="p">()</span>
<span class="n">mig_color_dict</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">u</span><span class="p">,</span> <span class="n">linestyle</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">()[::</span><span class="mi">2</span><span class="p">],</span> <span class="p">[</span><span class="s2">&quot;dotted&quot;</span><span class="p">,</span> <span class="s2">&quot;dashdot&quot;</span><span class="p">]):</span>
    <span class="n">mig</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">migrations</span>
    <span class="n">loc</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">time</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">while</span> <span class="n">u</span> <span class="o">!=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span><span class="p">:</span>
        <span class="n">migs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">mig</span><span class="o">.</span><span class="n">node</span> <span class="o">==</span> <span class="n">u</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">cur_mig</span> <span class="ow">in</span> <span class="n">migs</span><span class="p">:</span>
            <span class="n">cur_mig</span> <span class="o">=</span> <span class="n">mig</span><span class="p">[</span><span class="n">cur_mig</span><span class="p">]</span>
            <span class="n">loc</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_mig</span><span class="o">.</span><span class="n">dest</span> <span class="o">+</span> <span class="mf">0.5</span><span class="p">)</span>
            <span class="n">time</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cur_mig</span><span class="o">.</span><span class="n">time</span><span class="p">)</span>
        <span class="n">u</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">parent</span><span class="p">(</span><span class="n">u</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">loc</span><span class="p">,</span> <span class="n">time</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="n">linestyle</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">N</span><span class="p">):</span>
    <span class="n">rgba</span> <span class="o">=</span> <span class="n">cmap</span><span class="p">(</span><span class="n">i</span> <span class="o">/</span> <span class="n">N</span><span class="p">)</span>
    <span class="n">mig_color_dict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">rgba</span><span class="p">)</span><span class="o">*</span><span class="mi">255</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">axvspan</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="n">rgba</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Time of migration event&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Population&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/e720a0e5dd99dc708939bb7a19b3c9406985297af5db51c99f8dc7aa77ba3d43.svg" src="_images/e720a0e5dd99dc708939bb7a19b3c9406985297af5db51c99f8dc7aa77ba3d43.svg" />
</div>
</div>
<p>Note also how many migration events occurred before the two
samples coalesced, even in this small example.</p>
<p>The next visualisation shows where each of these migrations fall on the edges
of the simulated tree. We show the nodes associated with each migration
event, colouring nodes and edges by the population with which they are associated
(using the same colour scheme as the previous visualisation).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mig_color_dict</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">cmap</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">colormaps</span><span class="p">[</span><span class="s1">&#39;viridis&#39;</span><span class="p">]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">N</span><span class="p">):</span>
    <span class="n">rgba</span> <span class="o">=</span> <span class="n">cmap</span><span class="p">(</span><span class="n">i</span> <span class="o">/</span> <span class="n">N</span><span class="p">)</span>
    <span class="n">mig_color_dict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">rgba</span><span class="p">)</span> <span class="o">*</span> <span class="mi">255</span>

<span class="n">edge_colors</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">string</span> <span class="o">=</span> <span class="s2">&quot;.tree .edge {stroke-width:4px}&quot;</span>
<span class="n">pop</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">population</span>
<span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">nodes</span><span class="p">():</span>
    <span class="n">node</span> <span class="o">=</span> <span class="n">node</span><span class="o">.</span><span class="n">id</span>
    <span class="n">edge_colors</span><span class="p">[</span><span class="n">node</span><span class="p">]</span> <span class="o">=</span> <span class="n">mig_color_dict</span><span class="p">[</span><span class="n">pop</span><span class="p">[</span><span class="n">node</span><span class="p">]]</span>
    <span class="n">string</span> <span class="o">+=</span> <span class="p">(</span>
        <span class="s2">&quot;.node.n&quot;</span>
        <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">node</span><span class="p">)</span>
        <span class="o">+</span> <span class="s2">&quot;&gt; .sym {fill: rgb&quot;</span>
        <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">tuple</span><span class="p">(</span><span class="n">mig_color_dict</span><span class="p">[</span><span class="n">pop</span><span class="p">[</span><span class="n">node</span><span class="p">]][</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]))</span>
        <span class="o">+</span> <span class="s2">&quot;}&quot;</span>
    <span class="p">)</span>
    <span class="n">string</span> <span class="o">+=</span> <span class="p">(</span>
        <span class="s2">&quot;.node.n&quot;</span>
        <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">node</span><span class="p">)</span>
        <span class="o">+</span> <span class="s2">&quot; &gt; .edge {stroke: rgb&quot;</span>
        <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">tuple</span><span class="p">(</span><span class="n">mig_color_dict</span><span class="p">[</span><span class="n">pop</span><span class="p">[</span><span class="n">node</span><span class="p">]][</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]))</span>
        <span class="o">+</span> <span class="s2">&quot;}&quot;</span>
    <span class="p">)</span>
<span class="k">for</span> <span class="n">mig</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">migrations</span><span class="p">():</span>
    <span class="n">edge_colors</span><span class="p">[</span><span class="n">mig</span><span class="o">.</span><span class="n">node</span><span class="p">]</span> <span class="o">=</span> <span class="n">mig_color_dict</span><span class="p">[</span><span class="n">mig</span><span class="o">.</span><span class="n">source</span><span class="p">]</span>
    <span class="n">string</span> <span class="o">+=</span> <span class="p">(</span>
        <span class="s2">&quot;.node.n&quot;</span>
        <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">mig</span><span class="o">.</span><span class="n">node</span><span class="p">)</span>
        <span class="o">+</span> <span class="s2">&quot;&gt; .sym {fill: rgb&quot;</span>
        <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">tuple</span><span class="p">(</span><span class="n">mig_color_dict</span><span class="p">[</span><span class="n">pop</span><span class="p">[</span><span class="n">mig</span><span class="o">.</span><span class="n">node</span><span class="p">]][</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]))</span>
        <span class="o">+</span> <span class="s2">&quot;}&quot;</span>
    <span class="p">)</span>
    <span class="n">string</span> <span class="o">+=</span> <span class="p">(</span>
        <span class="s2">&quot;.node.n&quot;</span>
        <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">mig</span><span class="o">.</span><span class="n">node</span><span class="p">)</span>
        <span class="o">+</span> <span class="s2">&quot; &gt; .edge {stroke: rgb&quot;</span>
        <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">tuple</span><span class="p">(</span><span class="n">mig_color_dict</span><span class="p">[</span><span class="n">mig</span><span class="o">.</span><span class="n">dest</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]))</span>
        <span class="o">+</span> <span class="s2">&quot;}&quot;</span>
    <span class="p">)</span>
<span class="n">node_labels</span> <span class="o">=</span> <span class="p">{</span><span class="n">node</span><span class="o">.</span><span class="n">id</span><span class="p">:</span> <span class="s2">&quot;&quot;</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">nodes</span><span class="p">()}</span>
<span class="n">SVG</span><span class="p">(</span>
    <span class="n">ts</span><span class="o">.</span><span class="n">first</span><span class="p">()</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span>
        <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">500</span><span class="p">,</span> <span class="mi">400</span><span class="p">),</span> <span class="n">style</span><span class="o">=</span><span class="n">string</span><span class="p">,</span> <span class="n">node_labels</span><span class="o">=</span><span class="n">node_labels</span><span class="p">,</span> <span class="n">symbol_size</span><span class="o">=</span><span class="mi">10</span>
    <span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/613cde047c647c5929f6ee7ca76bdd6fff8a41cde8d17b54658eb95a6f7ae04e.svg" src="_images/613cde047c647c5929f6ee7ca76bdd6fff8a41cde8d17b54658eb95a6f7ae04e.svg" />
</div>
</div>
</section>
<section id="census-events">
<span id="sec-ancestry-census-events"></span><h3>Census events<a class="headerlink" href="#census-events" title="Link to this heading">#</a></h3>
<p>Census events allow you to add a node to each branch of the tree sequence at a given time
during the simulation. This can be useful when you wish to study haplotypes that are
ancestral to your simulated sample, or when you wish to know which lineages were present in
which populations at specified times.</p>
<p>For instance, the following example creates a two population island
model <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a>, with the populations exchanging migrants
at rate 0.05. At generation 5000, we add a census event using the
<a class="reference internal" href="api.html#msprime.Demography.add_census" title="msprime.Demography.add_census"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.add_census()</span></code></a> method
to determine where each of the lineages is at that time:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">island_model</span><span class="p">([</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">1000</span><span class="p">],</span> <span class="n">migration_rate</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_census</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">5000</span><span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span>
    <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">,</span>
    <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">1e-7</span><span class="p">,</span>
    <span class="n">sequence_length</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">141</span>
<span class="p">)</span>
<span class="n">css_string</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">nodes</span><span class="p">():</span>
    <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">time</span> <span class="o">==</span> <span class="mi">5000</span><span class="p">:</span>
        <span class="n">css_string</span> <span class="o">+=</span> <span class="s2">&quot;.n</span><span class="si">%s</span><span class="s2"> &gt; .sym {fill: </span><span class="si">%s</span><span class="s2">}&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s1">&#39;green&#39;</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="n">css_string</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/9334ead1a084ecf7d215a95f3f13071ceb6d8dc8b23d0b8ce8cd6ce5913304f2.svg" src="_images/9334ead1a084ecf7d215a95f3f13071ceb6d8dc8b23d0b8ce8cd6ce5913304f2.svg" />
</div>
</div>
<p>The resulting tree sequence has nodes on each tree at the specified census time.
These are the nodes with IDs 8, 9, 10, 11, 12 and 13.</p>
<p>This tells us that the genetic material ancestral to the present day sample was
held within 5 haplotypes at time 5000. The node table shows us that four of
these haplotypes (nodes 8, 9, 10 and 11) were in population 0 at this time, and
two of these haplotypes (nodes 12 and 13) were in population 1 at this time.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="p">)</span>
<span class="n">census_nodes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">bitwise_and</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">nodes_flags</span><span class="p">,</span> <span class="n">msprime</span><span class="o">.</span><span class="n">NodeType</span><span class="o">.</span><span class="n">CENSUS</span><span class="o">.</span><span class="n">value</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span>  
<span class="nb">print</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="p">[</span><span class="n">census_nodes</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>╔══╤═══════╤══════════╤══════════╤═════════════╤════════╗
║id│flags  │population│individual│time         │metadata║
╠══╪═══════╪══════════╪══════════╪═════════════╪════════╣
║0 │      1│         0│         0│   0.00000000│        ║
║1 │      1│         0│         0│   0.00000000│        ║
║2 │      1│         1│         1│   0.00000000│        ║
║3 │      1│         1│         1│   0.00000000│        ║
║4 │      0│         1│        -1│2331.91109774│        ║
║5 │      0│         1│        -1│3741.02811877│        ║
║6 │      0│         0│        -1│4216.80416680│        ║
║7 │      0│         1│        -1│4580.66488183│        ║
║8 │1048576│         0│        -1│5000.00000000│        ║
║9 │1048576│         0│        -1│5000.00000000│        ║
║10│1048576│         0│        -1│5000.00000000│        ║
║11│1048576│         0│        -1│5000.00000000│        ║
║12│1048576│         1│        -1│5000.00000000│        ║
║13│1048576│         1│        -1│5000.00000000│        ║
║14│      0│         1│        -1│5232.65267392│        ║
║15│      0│         0│        -1│8192.48105714│        ║
╚══╧═══════╧══════════╧══════════╧═════════════╧════════╝

╔══╤═══════╤══════════╤══════════╤════╤════════╗
║id│flags  │population│individual│time│metadata║
╠══╪═══════╪══════════╪══════════╪════╪════════╣
║0 │1048576│         0│        -1│5000│        ║
║1 │1048576│         0│        -1│5000│        ║
║2 │1048576│         0│        -1│5000│        ║
║3 │1048576│         0│        -1│5000│        ║
║4 │1048576│         1│        -1│5000│        ║
║5 │1048576│         1│        -1│5000│        ║
╚══╧═══════╧══════════╧══════════╧════╧════════╝
</pre></div>
</div>
</div>
</div>
<p>If we wish to study these ancestral haplotypes further, we can simplify the tree sequence
with respect to the census nodes and perform subsequent analyses on this simplified tree
sequence.
In this example, <code class="docutils literal notranslate"><span class="pre">ts_anc</span></code> is a tree sequence obtained from the original tree sequence
<code class="docutils literal notranslate"><span class="pre">ts</span></code> by labelling the census nodes as samples and removing all nodes and edges that are
not ancestral to these census nodes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">nodes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">census_nodes</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">nodes</span><span class="p">)</span>
<span class="n">ts_anc</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">simplify</span><span class="p">(</span><span class="n">samples</span><span class="o">=</span><span class="n">nodes</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[ 8  9 10 11 12 13]
</pre></div>
</div>
</div>
</div>
<section id="using-the-dtwf-model">
<span id="sec-ancestry-census-events-dtwf"></span><h4>Using the DTWF model<a class="headerlink" href="#using-the-dtwf-model" title="Link to this heading">#</a></h4>
<p>Census events assume that they are being used as part of a continuous time
simulation process, in which multiple events are guaranteed not to happen
at the same time. However, the <a class="reference internal" href="api.html#msprime.DiscreteTimeWrightFisher" title="msprime.DiscreteTimeWrightFisher"><code class="xref py py-class docutils literal notranslate"><span class="pre">DiscreteTimeWrightFisher</span></code></a> model does
not meet these assumptions, and allows multiple events (such as coalescences,
migrations etc) to all occur at the same time. This includes census events,
which can lead to errors when used in conjunction with the DTWF model.</p>
<p>Consider the following example:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">({</span><span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">},</span> <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s2">&quot;dtwf&quot;</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">33</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">time_scale</span><span class="o">=</span><span class="s2">&quot;log_time&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/4dfa5f19369876df91ccdd4094618b4d9eb9d8ab36d4fea30a0047e35beab8cb.svg" src="_images/4dfa5f19369876df91ccdd4094618b4d9eb9d8ab36d4fea30a0047e35beab8cb.svg" />
</div>
</div>
<p>We can see that a coalescence happens at time 3, giving node 7. Now, let’s
try to perform a census at time 3:</p>
<div class="cell tag_raises-exception tag_output_scroll docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_census</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">({</span><span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">},</span> <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s2">&quot;dtwf&quot;</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">33</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">LibraryError</span><span class="g g-Whitespace">                              </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">47</span><span class="p">],</span> <span class="n">line</span> <span class="mi">4</span>
<span class="g g-Whitespace">      </span><span class="mi">2</span> <span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="g g-Whitespace">      </span><span class="mi">3</span> <span class="n">demography</span><span class="o">.</span><span class="n">add_census</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="ne">----&gt; </span><span class="mi">4</span> <span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">({</span><span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">},</span> <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s2">&quot;dtwf&quot;</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">33</span><span class="p">)</span>

<span class="nn">File ~/work/tskit-site/tskit-site/msprime/ancestry.py:1311,</span> in <span class="ni">sim_ancestry</span><span class="nt">(samples, demography, sequence_length, discrete_genome, recombination_rate, gene_conversion_rate, gene_conversion_tract_length, population_size, ploidy, model, initial_state, start_time, end_time, record_migrations, record_full_arg, additional_nodes, coalescing_segments_only, num_labels, random_seed, num_replicates, replicate_index, record_provenance)</span>
<span class="g g-Whitespace">   </span><span class="mi">1289</span>     <span class="n">provenance_dict</span> <span class="o">=</span> <span class="n">provenance</span><span class="o">.</span><span class="n">get_provenance_dict</span><span class="p">(</span><span class="n">parameters</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1290</span> <span class="n">sim</span> <span class="o">=</span> <span class="n">_parse_sim_ancestry</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1291</span>     <span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1292</span>     <span class="n">sequence_length</span><span class="o">=</span><span class="n">sequence_length</span><span class="p">,</span>
   <span class="p">(</span><span class="o">...</span><span class="p">)</span>   <span class="mi">1309</span>     <span class="n">random_seed</span><span class="o">=</span><span class="n">random_seed</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1310</span> <span class="p">)</span>
<span class="ne">-&gt; </span><span class="mi">1311</span> <span class="k">return</span> <span class="n">_wrap_replicates</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1312</span>     <span class="n">sim</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1313</span>     <span class="n">num_replicates</span><span class="o">=</span><span class="n">num_replicates</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1314</span>     <span class="n">replicate_index</span><span class="o">=</span><span class="n">replicate_index</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1315</span>     <span class="n">provenance_dict</span><span class="o">=</span><span class="n">provenance_dict</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1316</span> <span class="p">)</span>

<span class="nn">File ~/work/tskit-site/tskit-site/msprime/ancestry.py:681,</span> in <span class="ni">_wrap_replicates</span><span class="nt">(simulator, num_replicates, replicate_index, provenance_dict, mutation_rate)</span>
<span class="g g-Whitespace">    </span><span class="mi">675</span> <span class="n">iterator</span> <span class="o">=</span> <span class="n">simulator</span><span class="o">.</span><span class="n">run_replicates</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">676</span>     <span class="n">num_replicates</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">677</span>     <span class="n">mutation_rate</span><span class="o">=</span><span class="n">mutation_rate</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">678</span>     <span class="n">provenance_dict</span><span class="o">=</span><span class="n">provenance_dict</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">679</span> <span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">680</span> <span class="k">if</span> <span class="n">replicate_index</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<span class="ne">--&gt; </span><span class="mi">681</span>     <span class="n">deque</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">deque</span><span class="p">(</span><span class="n">iterator</span><span class="p">,</span> <span class="n">maxlen</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">682</span>     <span class="k">return</span> <span class="n">deque</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
<span class="g g-Whitespace">    </span><span class="mi">683</span> <span class="k">else</span><span class="p">:</span>

<span class="nn">File ~/work/tskit-site/tskit-site/msprime/ancestry.py:1606,</span> in <span class="ni">Simulator.run_replicates</span><span class="nt">(self, num_replicates, mutation_rate, provenance_dict)</span>
<span class="g g-Whitespace">   </span><span class="mi">1604</span> <span class="k">for</span> <span class="n">replicate_index</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_replicates</span><span class="p">):</span>
<span class="g g-Whitespace">   </span><span class="mi">1605</span>     <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Starting replicate </span><span class="si">%d</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">replicate_index</span><span class="p">)</span>
<span class="ne">-&gt; </span><span class="mi">1606</span>     <span class="bp">self</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
<span class="g g-Whitespace">   </span><span class="mi">1607</span>     <span class="k">if</span> <span class="n">mutation_rate</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">1608</span>         <span class="c1"># This is only called from simulate() or the ms interface,</span>
<span class="g g-Whitespace">   </span><span class="mi">1609</span>         <span class="c1"># so does not need any further parameters.</span>
<span class="g g-Whitespace">   </span><span class="mi">1610</span>         <span class="n">mutations</span><span class="o">.</span><span class="n">_simple_mutate</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1611</span>             <span class="n">tables</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">tables</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1612</span>             <span class="n">random_generator</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">random_generator</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1613</span>             <span class="n">sequence_length</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1614</span>             <span class="n">rate</span><span class="o">=</span><span class="n">mutation_rate</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1615</span>         <span class="p">)</span>

<span class="nn">File ~/work/tskit-site/tskit-site/msprime/ancestry.py:1571,</span> in <span class="ni">Simulator.run</span><span class="nt">(self, event_chunk, debug_func)</span>
<span class="g g-Whitespace">   </span><span class="mi">1569</span>     <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Model durations must be &gt;= 0&quot;</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1570</span> <span class="n">end_time</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">time</span> <span class="o">+</span> <span class="n">model_duration</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">end_time</span><span class="p">)</span>
<span class="ne">-&gt; </span><span class="mi">1571</span> <span class="n">exit_reason</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_run_until</span><span class="p">(</span><span class="n">end_time</span><span class="p">,</span> <span class="n">event_chunk</span><span class="p">,</span> <span class="n">debug_func</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1572</span> <span class="k">if</span> <span class="n">exit_reason</span> <span class="o">==</span> <span class="n">ExitReason</span><span class="o">.</span><span class="n">COALESCENCE</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">time</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">end_time</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">1573</span>     <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Skipping remaining </span><span class="si">%d</span><span class="s2"> models&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">models</span><span class="p">)</span> <span class="o">-</span> <span class="n">j</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>

<span class="nn">File ~/work/tskit-site/tskit-site/msprime/ancestry.py:1536,</span> in <span class="ni">Simulator._run_until</span><span class="nt">(self, end_time, event_chunk, debug_func)</span>
<span class="g g-Whitespace">   </span><span class="mi">1534</span> <span class="n">ret</span> <span class="o">=</span> <span class="n">ExitReason</span><span class="o">.</span><span class="n">MAX_EVENTS</span>
<span class="g g-Whitespace">   </span><span class="mi">1535</span> <span class="k">while</span> <span class="n">ret</span> <span class="o">==</span> <span class="n">ExitReason</span><span class="o">.</span><span class="n">MAX_EVENTS</span><span class="p">:</span>
<span class="ne">-&gt; </span><span class="mi">1536</span>     <span class="n">ret</span> <span class="o">=</span> <span class="n">ExitReason</span><span class="p">(</span><span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">end_time</span><span class="p">,</span> <span class="n">event_chunk</span><span class="p">))</span>
<span class="g g-Whitespace">   </span><span class="mi">1537</span>     <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">time</span> <span class="o">&gt;</span> <span class="n">end_time</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">1538</span>         <span class="c1"># Currently the Pedigree and Sweeps models are &quot;non-reentrant&quot;</span>
<span class="g g-Whitespace">   </span><span class="mi">1539</span>         <span class="c1"># We can change this to an assertion once these have been fixed.</span>
<span class="g g-Whitespace">   </span><span class="mi">1540</span>         <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1541</span>             <span class="sa">f</span><span class="s2">&quot;Model </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> does not support interruption. &quot;</span>
<span class="g g-Whitespace">   </span><span class="mi">1542</span>             <span class="s2">&quot;Please open an issue on GitHub&quot;</span>
<span class="g g-Whitespace">   </span><span class="mi">1543</span>         <span class="p">)</span>

<span class="ne">LibraryError</span>: The simulation model supplied resulted in a parent node having a time value &lt;= to its child. This can occur as a result of multiple bottlenecks happening at the same time, multiple census events at the same time or numerical imprecision with very smallpopulation sizes.
</pre></div>
</div>
</div>
</div>
<p>We got an error from msprime about a parent node having a time value &lt;= to its child,
which occurred because we tried to add set of census nodes at time 3, but node 7 was
already at time 3, giving a zero branch length (which is disallowed by tskit).</p>
<p>The solution is to use non-integer time values when performing a census
under the DTWF:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_census</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mf">3.5</span><span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">({</span><span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">},</span> <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s2">&quot;dtwf&quot;</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">33</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">time_scale</span><span class="o">=</span><span class="s2">&quot;log_time&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/743025a50bda06241781b0002622a2ba0c56bbf6218c4e9b3fa80d2cdc27a12b.svg" src="_images/743025a50bda06241781b0002622a2ba0c56bbf6218c4e9b3fa80d2cdc27a12b.svg" />
</div>
</div>
</section>
</section>
<section id="ancestral-recombination-graph">
<span id="sec-ancestry-full-arg"></span><h3>Ancestral recombination graph<a class="headerlink" href="#ancestral-recombination-graph" title="Link to this heading">#</a></h3>
<p>This is a legacy option and identical to setting the <code class="docutils literal notranslate"><span class="pre">additional_nodes</span></code> option to store
common_ancestor events, recombinants, (and migrants if applicable). This is illustrated
in the following example:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">3</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">sequence_length</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="n">record_full_arg</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span>
<span class="n">additional_nodes</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">NodeType</span><span class="o">.</span><span class="n">COMMON_ANCESTOR</span> <span class="o">|</span> <span class="n">msprime</span><span class="o">.</span><span class="n">NodeType</span><span class="o">.</span><span class="n">RECOMBINANT</span>
<span class="n">ts_other</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">3</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">sequence_length</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="n">additional_nodes</span><span class="o">=</span><span class="n">additional_nodes</span><span class="p">,</span>
    <span class="n">coalescing_segments_only</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">equals</span><span class="p">(</span><span class="n">ts_other</span><span class="o">.</span><span class="n">tables</span><span class="p">,</span> <span class="n">ignore_provenance</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>True
╔══╤══════╤══════════╤══════════╤══════════╤════════╗
║id│flags │population│individual│time      │metadata║
╠══╪══════╪══════════╪══════════╪══════════╪════════╣
║0 │     1│         0│         0│0.00000000│        ║
║1 │     1│         0│         0│0.00000000│        ║
║2 │     1│         0│         1│0.00000000│        ║
║3 │     1│         0│         1│0.00000000│        ║
║4 │     1│         0│         2│0.00000000│        ║
║5 │     1│         0│         2│0.00000000│        ║
║6 │     0│         0│        -1│0.21230674│        ║
║7 │     0│         0│        -1│0.51485157│        ║
║8 │     0│         0│        -1│0.71161449│        ║
║9 │131072│         0│        -1│0.91107705│        ║
║10│131072│         0│        -1│0.91107705│        ║
║11│262144│         0│        -1│1.21743776│        ║
║12│     0│         0│        -1│1.23130397│        ║
║13│     0│         0│        -1│4.80416300│        ║
╚══╧══════╧══════════╧══════════╧══════════╧════════╝
</pre></div>
</div>
<img alt="_images/2a6bc0f3f39d8fccdb7c46270f3ce2ce6d441ffc52e2a92b58588206de572fb1.svg" src="_images/2a6bc0f3f39d8fccdb7c46270f3ce2ce6d441ffc52e2a92b58588206de572fb1.svg" />
</div>
</div>
<p>After running the simulation we first print out the
<a class="reference external" href="https://tskit.readthedocs.io/en/stable/data-model.html#node-table">node table</a>, which
contains information on all the nodes in the tree sequence. Note that <code class="docutils literal notranslate"><span class="pre">flags</span></code>
column contains several different values: all of the sample nodes (at time 0)
have a flag value of <code class="docutils literal notranslate"><span class="pre">1</span></code> (<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.NODE_IS_SAMPLE" title="(in Project name not set)"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NODE_IS_SAMPLE</span></code></a>). Most other
nodes have a flag value of <code class="docutils literal notranslate"><span class="pre">0</span></code>, which is the standard for internal nodes
in a coalescent simulations.</p>
<p>Nodes 9 and 10 have flags equal to 131072 (<code class="docutils literal notranslate"><span class="pre">NodeType.RECOMBINANT</span></code>), which
tells us that they correspond to a recombination event in the ARG. A
recombination event results in two extra nodes being recorded, one identifying
the individual providing the genetic material to the left of the breakpoint and
the other identifying the individuals providing the genetic material to the
right. The effect of this extra node can be seen in the trees: node 9 is
present as a ‘unary’ node in the left hand tree and node 10 in the right.</p>
<p>In this particular case, the first thing that happens to these two lineages
as we go up the tree sequence is that they coalesce straight back together again
at node 11, forming a (normally undetectable) “diamond event” in the ARG, and
explaining why the topology of both these trees appears the same. In a full ARG,
this is one of many ways that a coalescent event can occur without resulting in a
marginal coalescence, and which instead results in an additional unary node in the
trees. Such nodes are given a flags value of 262144 (<code class="docutils literal notranslate"><span class="pre">NodeType.COMMON_ANCESTOR</span></code>).</p>
<p>If we wish to reduce these trees down to the minimal representation, we can
use <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.simplify" title="(in Project name not set)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">tskit.TreeSequence.simplify()</span></code></a>. The resulting tree sequence will have
all of these unary nodes removed and will be equivalent to (but not identical, due to
stochastic effects) calling <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a> without the <code class="docutils literal notranslate"><span class="pre">record_full_arg</span></code>
argument.</p>
<p>Migrations nodes are also recorded in the ARG using the
<code class="docutils literal notranslate"><span class="pre">NodeType.MIGRANT</span></code> flag. See the <a class="reference internal" href="api.html#sec-api-node-flags"><span class="std std-ref">Node flags</span></a>
section for more details.</p>
</section>
</section>
<section id="manipulating-simulation-time">
<h2>Manipulating simulation time<a class="headerlink" href="#manipulating-simulation-time" title="Link to this heading">#</a></h2>
<section id="stopping-simulations-early">
<span id="sec-ancestry-end-time"></span><h3>Stopping simulations early<a class="headerlink" href="#stopping-simulations-early" title="Link to this heading">#</a></h3>
<p>In most simulations we want to simulate a complete history for our samples:
that is, to when we have a common ancestor at every point along the
sequence. For example, here we generate a complete ancestral history
for two diploid samples:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">sequence_length</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">y_ticks</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">]))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/3a3ca8f5bc0fc7f88afca76ada5a80fafc477749103217b4fe30086cea588250.svg" src="_images/3a3ca8f5bc0fc7f88afca76ada5a80fafc477749103217b4fe30086cea588250.svg" />
</div>
</div>
<p>Sometimes we would like to stop the simulation early, <strong>before</strong> complete
coalescence has occurred; perhaps we would like to
<a class="reference internal" href="#sec-ancestry-initial-state-combining"><span class="std std-ref">combine several simulations</span></a>
with different properties, or we are really only interested in the
recent past and simulating more ancient processes would be a waste
of computational resources.</p>
<p>We can do this by specifying a <code class="docutils literal notranslate"><span class="pre">end_time</span></code> value.
Let’s repeat the simulation as above with an <code class="docutils literal notranslate"><span class="pre">end_time</span></code> of 2:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">2</span><span class="p">,</span> <span class="n">sequence_length</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">end_time</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">y_ticks</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">]))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/c9b6d1230dac330d151705fe1cb64c1eb94798126d41db355a05aa6815723bbf.svg" src="_images/c9b6d1230dac330d151705fe1cb64c1eb94798126d41db355a05aa6815723bbf.svg" />
</div>
</div>
<p>There are a number of important things to observe about this
tree sequence:</p>
<ol class="arabic simple">
<li><p>The history up until time 2 is <strong>identical</strong> to the original simulation.</p></li>
<li><p>The first tree has not fully coalesced, and we therefore have <strong>multiple
roots</strong>. Trees with multiple roots are fully supported in tskit.</p></li>
<li><p>The nodes 8 and 9 are at time 2 (our <code class="docutils literal notranslate"><span class="pre">end_time</span></code> value), and these
connect to the uncoalesced portions of the tree. These “unary” nodes
are very important for ensuring the correct statistical properties
when we
<a class="reference internal" href="#sec-ancestry-initial-state-combining"><span class="std std-ref">combine multiple simulations</span></a>.</p></li>
</ol>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>The ancestry model <a class="reference internal" href="#sec-ancestry-models-specifying-duration"><span class="std std-ref">duration</span></a>
can also be used to stop simulations before coalescence.</p>
</div>
<p>For discrete time models, the exact interpretation
of the <code class="docutils literal notranslate"><span class="pre">end_time</span></code> is important — we continue to simulate
the process while the time is &lt;= <code class="docutils literal notranslate"><span class="pre">end_time</span></code>. For example here
we use an <code class="docutils literal notranslate"><span class="pre">end_time</span></code> of 1 generation in a <a class="reference internal" href="#sec-ancestry-models-dtwf"><span class="std std-ref">Discrete Time Wright-Fisher</span></a>
simulation:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="n">population_size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s2">&quot;dtwf&quot;</span><span class="p">,</span> <span class="n">end_time</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/0fc20412839f693c52b075b4769b3fb8bce1c39d7fff232ebe050a716a3e8611.svg" src="_images/0fc20412839f693c52b075b4769b3fb8bce1c39d7fff232ebe050a716a3e8611.svg" />
</div>
</div>
<p>We have run 1 generation of the DTWF, resulting in the coalescence
at node 9.</p>
</section>
<section id="setting-the-start-time">
<span id="sec-ancestry-start-time"></span><h3>Setting the start time<a class="headerlink" href="#setting-the-start-time" title="Link to this heading">#</a></h3>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>Setting simulation start time is an advanced feature which is rarely
needed in practice.</p>
</div>
<p>Sometimes we need control when the simulation process starts. For
example, given the initial configuration of at time zero we want to
“jump ahead” to a later time point before we start generating
random events. Consider the following example:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span>
<span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
        <div>
            <style scoped="">
                .tskit-table tbody tr th:only-of-type {vertical-align: middle;}
                .tskit-table tbody tr th {vertical-align: top;}
                .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
                .tskit-table tbody th {padding: 0.5em 0.5em;}
            </style>
            <table border="1" class="tskit-table">
                <thead>
                    <tr>
                        <th>id</th><th>flags</th><th>population</th><th>individual</th><th>time</th><th>metadata</th>
                    </tr>
                </thead>
                <tbody>
                    <tr><td>0</td><td>1</td><td>0</td><td>0</td><td>0.00000000</td><td></td></tr>
<tr><td>1</td><td>1</td><td>0</td><td>0</td><td>0.00000000</td><td></td></tr>
<tr><td>2</td><td>1</td><td>0</td><td>1</td><td>0.00000000</td><td></td></tr>
<tr><td>3</td><td>1</td><td>0</td><td>1</td><td>0.00000000</td><td></td></tr>
<tr><td>4</td><td>0</td><td>0</td><td>-1</td><td>0.15642269</td><td></td></tr>
<tr><td>5</td><td>0</td><td>0</td><td>-1</td><td>0.29152170</td><td></td></tr>
<tr><td>6</td><td>0</td><td>0</td><td>-1</td><td>2.11740681</td><td></td></tr>

                </tbody>
            </table>
        </div>
    </div></div>
</div>
<p>Now we run the same simulation with a <code class="docutils literal notranslate"><span class="pre">start_time</span></code> of 10:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">start_time</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span>
<span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
        <div>
            <style scoped="">
                .tskit-table tbody tr th:only-of-type {vertical-align: middle;}
                .tskit-table tbody tr th {vertical-align: top;}
                .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
                .tskit-table tbody th {padding: 0.5em 0.5em;}
            </style>
            <table border="1" class="tskit-table">
                <thead>
                    <tr>
                        <th>id</th><th>flags</th><th>population</th><th>individual</th><th>time</th><th>metadata</th>
                    </tr>
                </thead>
                <tbody>
                    <tr><td>0</td><td>1</td><td>0</td><td>0</td><td>0.00000000</td><td></td></tr>
<tr><td>1</td><td>1</td><td>0</td><td>0</td><td>0.00000000</td><td></td></tr>
<tr><td>2</td><td>1</td><td>0</td><td>1</td><td>0.00000000</td><td></td></tr>
<tr><td>3</td><td>1</td><td>0</td><td>1</td><td>0.00000000</td><td></td></tr>
<tr><td>4</td><td>0</td><td>0</td><td>-1</td><td>10.15642269</td><td></td></tr>
<tr><td>5</td><td>0</td><td>0</td><td>-1</td><td>10.29152170</td><td></td></tr>
<tr><td>6</td><td>0</td><td>0</td><td>-1</td><td>12.11740681</td><td></td></tr>

                </tbody>
            </table>
        </div>
    </div></div>
</div>
<p>We can see that the results are identical except that 10 has
been added to all of the internal node times.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>It is important to note that the <code class="docutils literal notranslate"><span class="pre">start_time</span></code> only controls the
“random” events that happen within the simulation model: events
that are specified to occur at a given time (such as sampling
or demographic events) will occur at that time, even if this
is before the <code class="docutils literal notranslate"><span class="pre">start_time</span></code>.</p>
</div>
<p>Consider the following example in which we create a simple
<a class="reference internal" href="demography.html#sec-demography-examples-population-tree"><span class="std std-ref">population tree</span></a>
demography, and then take samples at different times all <em>before</em>
<code class="docutils literal notranslate"><span class="pre">start_time</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population_split</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">],</span> <span class="n">ancestral</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">)</span>

<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="p">[</span>
        <span class="n">msprime</span><span class="o">.</span><span class="n">SampleSet</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">population</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">time</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
        <span class="n">msprime</span><span class="o">.</span><span class="n">SampleSet</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">population</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">time</span><span class="o">=</span><span class="mi">2</span><span class="p">),</span>
        <span class="n">msprime</span><span class="o">.</span><span class="n">SampleSet</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">population</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="n">time</span><span class="o">=</span><span class="mi">6</span><span class="p">),</span>
    <span class="p">],</span>
    <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">,</span>
    <span class="n">ploidy</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">start_time</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
        <div>
            <style scoped="">
                .tskit-table tbody tr th:only-of-type {vertical-align: middle;}
                .tskit-table tbody tr th {vertical-align: top;}
                .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
                .tskit-table tbody th {padding: 0.5em 0.5em;}
            </style>
            <table border="1" class="tskit-table">
                <thead>
                    <tr>
                        <th>id</th><th>flags</th><th>population</th><th>individual</th><th>time</th><th>metadata</th>
                    </tr>
                </thead>
                <tbody>
                    <tr><td>0</td><td>1</td><td>0</td><td>0</td><td>1.00000000</td><td></td></tr>
<tr><td>1</td><td>1</td><td>1</td><td>1</td><td>2.00000000</td><td></td></tr>
<tr><td>2</td><td>1</td><td>2</td><td>2</td><td>6.00000000</td><td></td></tr>
<tr><td>3</td><td>0</td><td>2</td><td>-1</td><td>52.47084127</td><td></td></tr>
<tr><td>4</td><td>0</td><td>2</td><td>-1</td><td>66.18169933</td><td></td></tr>

                </tbody>
            </table>
        </div>
    </div></div>
</div>
<p>We can see that the samples we specify are in the correct populations
and their node times are at the sampling time. Note that all the
coalescences happen in population ID 2 (“C”), indicating that the
lineages from “A” and “B” moved there at the time of the split
(which we can observe directly using the
<a class="reference internal" href="#sec-ancestry-record-migrations"><span class="std std-ref">record_migrations</span></a> option if we wish).
Note also that we drew a sample from population “C” at time 6,
which is <a class="reference internal" href="demography.html#sec-demography-direction-of-time"><span class="std std-ref">after</span></a> the split;
if we tried to draw a sample before the split, we would get an
error because the population would be
<a class="reference internal" href="demography.html#sec-demography-populations-life-cycle"><span class="std std-ref">inactive</span></a>.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>See the <a class="reference internal" href="#sec-ancestry-event-order"><span class="std std-ref">Order of event execution</span></a> section for details
of the precise order in which fixed-time events occur.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The same considerations apply when we are running simulations based
on a given <a class="reference internal" href="#sec-ancestry-initial-state"><span class="std std-ref">initial state</span></a>; in
this case, the <code class="docutils literal notranslate"><span class="pre">start_time</span></code> is by default set to the time
of the youngest root in the input trees.</p>
</div>
</section>
<section id="order-of-event-execution">
<span id="sec-ancestry-event-order"></span><h3>Order of event execution<a class="headerlink" href="#order-of-event-execution" title="Link to this heading">#</a></h3>
<p>There are two different classes of event that happen in ancestry
simulations: “fixed” and “random” events. Random events are
generated by the ancestry model, and fixed events are those
that occur at a time specified by the user. Fixed events
are <a class="reference internal" href="demography.html#sec-demography-events"><span class="std std-ref">demographic events</span></a> or
<a class="reference internal" href="#sec-ancestry-samples"><span class="std std-ref">sampling events</span></a>.
If multiple events are specified to occur at exactly the same time, then:</p>
<ol class="arabic simple">
<li><p>All demographic events at this time are run, <em>in the order they
were specified</em>.</p></li>
<li><p>Sampling events occur immediately after all the demographic events
have been executed. (See the <a class="reference internal" href="#sec-ancestry-samples-sampling-time"><span class="std std-ref">Sampling time</span></a>
section for more details on ancient sampling events.)</p></li>
</ol>
<p>These details are slightly different for the <a class="reference internal" href="#sec-ancestry-models-dtwf"><span class="std std-ref">Discrete Time Wright-Fisher</span></a>
model. See the <a class="reference internal" href="api.html#msprime.DiscreteTimeWrightFisher" title="msprime.DiscreteTimeWrightFisher"><code class="xref py py-class docutils literal notranslate"><span class="pre">API</span> <span class="pre">documentation</span></code></a> for
details.</p>
</section>
</section>
<section id="specifying-the-initial-state">
<span id="sec-ancestry-initial-state"></span><h2>Specifying the initial state<a class="headerlink" href="#specifying-the-initial-state" title="Link to this heading">#</a></h2>
<p>By default <code class="docutils literal notranslate"><span class="pre">msprime</span></code> simulations are initialised by specifying a set of
<a class="reference internal" href="#sec-ancestry-samples"><span class="std std-ref">samples</span></a>,
which sets up the simulation with segments of ancestral material covering the
whole sequence. Internally, this initial state is implemented by
setting up a <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TableCollection" title="(in Project name not set)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TableCollection</span></code></a> which contains the
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-population-table-definition" title="(in Project name not set)"><span class="xref std std-ref">population</span></a>,
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-node-table-definition" title="(in Project name not set)"><span class="xref std std-ref">node</span></a>,
and <a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-individual-table-definition" title="(in Project name not set)"><span class="xref std std-ref">individual</span></a> information.
The low-level simulation code then reads these tables and sets up the
simulation accordingly. We can see what this initial state looks like
by setting up a simple simulation and stopping it immediately:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">initial_ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">end_time</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">initial_ts</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
              <style>
                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}
                .tskit-table tbody tr td {padding: 0.5em 0.5em;}
                .tskit-table tbody tr td:first-of-type {text-align: left;}
                .tskit-details-label {vertical-align: top; padding-right:5px;}
                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}
                .tskit-table-set-table {margin: 12px 0 0 12px;}
                details {display: inline-block;}
                summary {cursor: pointer; outline: 0; display: list-item;}
              </style>
              <div class="tskit-table-set">
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="padding:0;line-height:21px;">
                          <img style="height: 32px;display: inline-block;padding: 3px 5px 3px 0;" src="https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg"/>
                          <a target="_blank" href="https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class"> Tree Sequence </a>
                        </th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr><td>Trees</td><td>1</td></tr>
                      <tr><td>Sequence Length</td><td>1.0</td></tr>
                      <tr><td>Time Units</td><td>generations</td></tr>
                      <tr><td>Sample Nodes</td><td>2</td></tr>
                      <tr><td>Total Size</td><td>1.4 KiB</td></tr>
                      <tr>
                        <td>Metadata</td><td style="text-align: left;">No Metadata</td></tr>
                    </tbody>
                  </table>
                </div>
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="line-height:21px;">Table</th>
                        <th>Rows</th>
                        <th>Size</th>
                        <th>Has Metadata</th>
                      </tr>
                    </thead>
                    <tbody>
                    
                  <tr>
                    <td>Edges</td>
                      <td>0</td>
                      <td>8 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Individuals</td>
                      <td>1</td>
                      <td>52 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Migrations</td>
                      <td>0</td>
                      <td>8 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Mutations</td>
                      <td>0</td>
                      <td>16 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Nodes</td>
                      <td>2</td>
                      <td>64 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Populations</td>
                      <td>1</td>
                      <td>224 Bytes</td>
                      <td style="text-align: center;">
                        ✅
                      </td>
                    </tr>
                
                  <tr>
                    <td>Provenances</td>
                      <td>1</td>
                      <td>1019 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Sites</td>
                      <td>0</td>
                      <td>16 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
            </div></div>
</div>
<p>We can see that the returned tree sequence has one population,
two sample nodes and one individual (since sample individuals are
<a class="reference internal" href="#sec-ancestry-ploidy"><span class="std std-ref">diploid</span></a> by default). There are no
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-edge-table-definition" title="(in Project name not set)"><span class="xref std std-ref">edges</span></a> because the simulation
didn’t run for any period of time: we’re essentially just returning
the initial state used to start the simulation.</p>
<p>Then, running a simulation in which the <code class="docutils literal notranslate"><span class="pre">initial_state</span></code> is equal to
<code class="docutils literal notranslate"><span class="pre">initial_ts</span></code> is <em>precisely</em> the same as running the simulation
with the same arguments as we used above:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts1</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="n">initial_state</span><span class="o">=</span><span class="n">initial_ts</span><span class="p">,</span> <span class="n">population_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">ts2</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">population_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="k">assert</span> <span class="n">ts1</span><span class="o">.</span><span class="n">equals</span><span class="p">(</span><span class="n">ts2</span><span class="p">,</span> <span class="n">ignore_provenance</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Note that the <code class="docutils literal notranslate"><span class="pre">initial_state</span></code> tree sequence only contains the
information about the ancestral histories, and not any of the
simulation parameters. In the previous example we used default
parameters to keep things simple, but in general the full
simulation model will need to be specified as well as the initial state.
In particular, see the <a class="reference internal" href="#sec-ancestry-initial-state-demography"><span class="std std-ref">Interaction with demography</span></a> section.</p>
</div>
<p>While it is possible to specify the initial state manually by
constructing the input tree sequence,
it is an advanced topic and rarely necessary. More often we are
interested in specifying the initial state of an <code class="docutils literal notranslate"><span class="pre">msprime</span></code>
simulation using the output of <em>another</em> simulation.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <code class="docutils literal notranslate"><span class="pre">initial_state</span></code> tree sequence is quite literally the initial
state of the tree sequence that is returned at the end of a simulation,
and that any information present (including metadata) will not be
altered.</p>
</div>
<section id="combining-backward-time-simulations">
<span id="sec-ancestry-initial-state-combining"></span><h3>Combining backward-time simulations<a class="headerlink" href="#combining-backward-time-simulations" title="Link to this heading">#</a></h3>
<p>Suppose that we wished to simulate a scenario
in which we have recombination happening at some rate in the recent past,
and in the more distant past there is no recombination.
We can do this by combining the simulations,
using one as the <code class="docutils literal notranslate"><span class="pre">initial_state</span></code> for the other.</p>
<p>We first run the simulation of the recent past:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts1</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">2</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">sequence_length</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">end_time</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">population_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts1</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">time_scale</span><span class="o">=</span><span class="s2">&quot;log_time&quot;</span><span class="p">,</span> <span class="n">y_ticks</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">]))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/4d0ac5d95b21d9a759b4853b85ecf2f5af7b8afd0d086184a4fdd046458ea661.svg" src="_images/4d0ac5d95b21d9a759b4853b85ecf2f5af7b8afd0d086184a4fdd046458ea661.svg" />
</div>
</div>
<p>Some recombination and coalescence has happened and we have three
marginal trees along the genome. None of these trees has fully coalesced,
and so each of the trees has multiple roots.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>The set of nodes at time 10 connecting to all the internal nodes in the
trees are very important: without these nodes and unary edges, the
statistical properties of the combined simulations would not be correct!
(See also section below on
<a class="reference internal" href="#sec-ancestry-initial-state-forward-simulations"><span class="std std-ref">forward simulations</span></a>.)</p>
</div>
<p>We then run the next phase of the simulation in which there is
<em>no</em> recombination by passing <code class="docutils literal notranslate"><span class="pre">ts1</span></code> as the argument to <code class="docutils literal notranslate"><span class="pre">initial_state</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts2</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="n">initial_state</span><span class="o">=</span><span class="n">ts1</span><span class="p">,</span> <span class="n">population_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts2</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">time_scale</span><span class="o">=</span><span class="s2">&quot;log_time&quot;</span><span class="p">,</span> <span class="n">y_ticks</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">200</span><span class="p">,</span> <span class="mi">500</span><span class="p">]))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/26e54ba45161d29820cc774283e47072cd9bd53f64b856bccbc1c9a571ccb10c.svg" src="_images/26e54ba45161d29820cc774283e47072cd9bd53f64b856bccbc1c9a571ccb10c.svg" />
</div>
</div>
<p>Since there is no recombination in this more ancient simulation,
node <code class="docutils literal notranslate"><span class="pre">12</span></code> is the MRCA in all three trees. The unary nodes
<code class="docutils literal notranslate"><span class="pre">6</span></code>, <code class="docutils literal notranslate"><span class="pre">7</span></code>, <code class="docutils literal notranslate"><span class="pre">8</span></code>, and <code class="docutils literal notranslate"><span class="pre">9</span></code> marking the transition between the
two simulation regimes are still present in the trees. If you
wish to remove these, we can use the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.simplify" title="(in Project name not set)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">tskit.TreeSequence.simplify()</span></code></a>
method:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts_simplified</span> <span class="o">=</span> <span class="n">ts2</span><span class="o">.</span><span class="n">simplify</span><span class="p">()</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts_simplified</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/3301261e3dc90899f82a3489c7615e6bf57e1d36e5865cfe1445c7eaa84f3488.svg" src="_images/3301261e3dc90899f82a3489c7615e6bf57e1d36e5865cfe1445c7eaa84f3488.svg" />
</div>
</div>
</section>
<section id="interaction-with-demography">
<span id="sec-ancestry-initial-state-demography"></span><h3>Interaction with demography<a class="headerlink" href="#interaction-with-demography" title="Link to this heading">#</a></h3>
<p>In the previous example we saw how to combine two single population simulations.
We can also combine more complex simulations involving
<a class="reference internal" href="demography.html#sec-demography"><span class="std std-ref">demography</span></a>, but care needs to be taken to ensure that
the correct models are simulated in each part of the simulation.
Suppose we have simple model with a
<a class="reference internal" href="demography.html#sec-demography-events-population-split"><span class="std std-ref">population split</span></a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Contemporary population A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Contemporary population B&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Ancestral population&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population_split</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">ancestral</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">])</span>
<span class="n">demography</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (3)</b><table border="1" class="tskit-table"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>A</td><td>Contemporary population A</td><td>100.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>B</td><td>Contemporary population B</td><td>200.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>2</td><td>C</td><td>Ancestral population</td><td>300.0</td><td>0</td><td>1e+02</td><td>{}</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Migration matrix (all zero)</b><table border="1" class="tskit-table"><thead><tr></tr></thead><tbody></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events (1)</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>100</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[A, B], ancestral=C</td><td>Moves all lineages from derived populations &#x27;A&#x27; and &#x27;B&#x27; to the ancestral &#x27;C&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr></tbody></table></div></div></div></div>
</div>
<p>Then, we run a simulation of this model for 50 generations:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts1</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span> <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">,</span> <span class="n">end_time</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1234</span><span class="p">)</span>
<span class="n">node_labels</span> <span class="o">=</span> <span class="p">{</span>
    <span class="n">node</span><span class="o">.</span><span class="n">id</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">node</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">ts1</span><span class="o">.</span><span class="n">population</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">population</span><span class="p">)</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">ts1</span><span class="o">.</span><span class="n">nodes</span><span class="p">()}</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts1</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">node_labels</span><span class="o">=</span><span class="n">node_labels</span><span class="p">,</span> <span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/86f4798c35fd20d41285b6e1285e39704df98d9a374c84ec25daf47710b0b052.svg" src="_images/86f4798c35fd20d41285b6e1285e39704df98d9a374c84ec25daf47710b0b052.svg" />
</div>
</div>
<p>After 50 generations we can see there has been a coalescence in population <code class="docutils literal notranslate"><span class="pre">A</span></code>
but none in <code class="docutils literal notranslate"><span class="pre">B</span></code> and we therefore have three lineages left when the simulation
finishes. We can then continue the simulation based on this initial state:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts2</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">initial_state</span><span class="o">=</span><span class="n">ts1</span><span class="p">,</span> <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">5678</span><span class="p">)</span>
<span class="n">node_labels</span> <span class="o">=</span> <span class="p">{</span>
    <span class="n">node</span><span class="o">.</span><span class="n">id</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">node</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">ts2</span><span class="o">.</span><span class="n">population</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">population</span><span class="p">)</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">ts2</span><span class="o">.</span><span class="n">nodes</span><span class="p">()}</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts2</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">node_labels</span><span class="o">=</span><span class="n">node_labels</span><span class="p">,</span> <span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/dbd12c1c430e5aa4629cedd42e15073f72aff886527ada0b7ad970cb350fc4e5.svg" src="_images/dbd12c1c430e5aa4629cedd42e15073f72aff886527ada0b7ad970cb350fc4e5.svg" />
</div>
</div>
<p>Note that we use the <strong>same</strong> demography object, and so the lineages
migrate and ultimately coalesce in population <code class="docutils literal notranslate"><span class="pre">C</span></code>, as we’d expect.
This is the simplest case, in which we already have the demography
object which we can use to model the entire simulation from end-to-end.</p>
<p>In other cases (e.g. when working with simulations from a different
program), we don’t have the <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> object at hand
and we need to create one that is both compatible with the
<code class="docutils literal notranslate"><span class="pre">initial_state</span></code> tree sequence and reflects the demographic
model that we are interested in using. The best way to do this
is to use the <a class="reference internal" href="api.html#msprime.Demography.from_tree_sequence" title="msprime.Demography.from_tree_sequence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.from_tree_sequence()</span></code></a> method to
first get a base demography that is based on the
tree sequence <a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-population-table-definition" title="(in Project name not set)"><span class="xref std std-ref">population table</span></a>
and <a class="reference external" href="https://tskit.dev/tskit/docs/stable/metadata.html#sec-metadata" title="(in Project name not set)"><span class="xref std std-ref">metadata</span></a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">from_tree_sequence</span><span class="p">(</span><span class="n">ts1</span><span class="p">)</span>
<span class="n">demography</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (3)</b><table border="1" class="tskit-table"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>A</td><td>Contemporary population A</td><td>0.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>B</td><td>Contemporary population B</td><td>0.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>2</td><td>C</td><td>Ancestral population</td><td>0.0</td><td>0</td><td>0</td><td>{}</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Migration matrix (all zero)</b><table border="1" class="tskit-table"><thead><tr></tr></thead><tbody></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events (0)</b><table border="1" class="tskit-table"><thead><tr></tr></thead><tbody></tbody></table></div></div></div></div>
</div>
<p>We can see that the population names and descriptions have been recovered
from the tree sequence metadata but <strong>no other information</strong>: the
population sizes are all zero and we no longer have a population split
event. If we try to run a simulation using this demography we get an
error:</p>
<div class="cell tag_raises-exception docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="n">initial_state</span><span class="o">=</span><span class="n">ts1</span><span class="p">,</span> <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">InputError</span><span class="g g-Whitespace">                                </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">65</span><span class="p">],</span> <span class="n">line</span> <span class="mi">1</span>
<span class="ne">----&gt; </span><span class="mi">1</span> <span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="n">initial_state</span><span class="o">=</span><span class="n">ts1</span><span class="p">,</span> <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">)</span>

<span class="nn">File ~/work/tskit-site/tskit-site/msprime/ancestry.py:1290,</span> in <span class="ni">sim_ancestry</span><span class="nt">(samples, demography, sequence_length, discrete_genome, recombination_rate, gene_conversion_rate, gene_conversion_tract_length, population_size, ploidy, model, initial_state, start_time, end_time, record_migrations, record_full_arg, additional_nodes, coalescing_segments_only, num_labels, random_seed, num_replicates, replicate_index, record_provenance)</span>
<span class="g g-Whitespace">   </span><span class="mi">1265</span>     <span class="n">parameters</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1266</span>         <span class="n">command</span><span class="o">=</span><span class="s2">&quot;sim_ancestry&quot;</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1267</span>         <span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span>
   <span class="p">(</span><span class="o">...</span><span class="p">)</span>   <span class="mi">1287</span>         <span class="c1"># replicate index is excluded as it is inserted for each replicate</span>
<span class="g g-Whitespace">   </span><span class="mi">1288</span>     <span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1289</span>     <span class="n">provenance_dict</span> <span class="o">=</span> <span class="n">provenance</span><span class="o">.</span><span class="n">get_provenance_dict</span><span class="p">(</span><span class="n">parameters</span><span class="p">)</span>
<span class="ne">-&gt; </span><span class="mi">1290</span> <span class="n">sim</span> <span class="o">=</span> <span class="n">_parse_sim_ancestry</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1291</span>     <span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1292</span>     <span class="n">sequence_length</span><span class="o">=</span><span class="n">sequence_length</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1293</span>     <span class="n">recombination_rate</span><span class="o">=</span><span class="n">recombination_rate</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1294</span>     <span class="n">gene_conversion_rate</span><span class="o">=</span><span class="n">gene_conversion_rate</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1295</span>     <span class="n">gene_conversion_tract_length</span><span class="o">=</span><span class="n">gene_conversion_tract_length</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1296</span>     <span class="n">discrete_genome</span><span class="o">=</span><span class="n">discrete_genome</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1297</span>     <span class="n">population_size</span><span class="o">=</span><span class="n">population_size</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1298</span>     <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1299</span>     <span class="n">ploidy</span><span class="o">=</span><span class="n">ploidy</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1300</span>     <span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1301</span>     <span class="n">initial_state</span><span class="o">=</span><span class="n">initial_state</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1302</span>     <span class="n">start_time</span><span class="o">=</span><span class="n">start_time</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1303</span>     <span class="n">end_time</span><span class="o">=</span><span class="n">end_time</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1304</span>     <span class="n">record_migrations</span><span class="o">=</span><span class="n">record_migrations</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1305</span>     <span class="n">record_full_arg</span><span class="o">=</span><span class="n">record_full_arg</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1306</span>     <span class="n">additional_nodes</span><span class="o">=</span><span class="n">additional_nodes</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1307</span>     <span class="n">coalescing_segments_only</span><span class="o">=</span><span class="n">coalescing_segments_only</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1308</span>     <span class="n">num_labels</span><span class="o">=</span><span class="n">num_labels</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1309</span>     <span class="n">random_seed</span><span class="o">=</span><span class="n">random_seed</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1310</span> <span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1311</span> <span class="k">return</span> <span class="n">_wrap_replicates</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1312</span>     <span class="n">sim</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1313</span>     <span class="n">num_replicates</span><span class="o">=</span><span class="n">num_replicates</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1314</span>     <span class="n">replicate_index</span><span class="o">=</span><span class="n">replicate_index</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1315</span>     <span class="n">provenance_dict</span><span class="o">=</span><span class="n">provenance_dict</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1316</span> <span class="p">)</span>

<span class="nn">File ~/work/tskit-site/tskit-site/msprime/ancestry.py:1081,</span> in <span class="ni">_parse_sim_ancestry</span><span class="nt">(samples, sequence_length, recombination_rate, gene_conversion_rate, gene_conversion_tract_length, discrete_genome, population_size, demography, ploidy, model, initial_state, start_time, end_time, record_migrations, record_full_arg, additional_nodes, coalescing_segments_only, num_labels, random_seed, init_for_debugger)</span>
<span class="g g-Whitespace">   </span><span class="mi">1078</span> <span class="n">random_seed</span> <span class="o">=</span> <span class="n">_parse_random_seed</span><span class="p">(</span><span class="n">random_seed</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1079</span> <span class="n">random_generator</span> <span class="o">=</span> <span class="n">_msprime</span><span class="o">.</span><span class="n">RandomGenerator</span><span class="p">(</span><span class="n">random_seed</span><span class="p">)</span>
<span class="ne">-&gt; </span><span class="mi">1081</span> <span class="k">return</span> <span class="n">Simulator</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1082</span>     <span class="n">tables</span><span class="o">=</span><span class="n">initial_state</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1083</span>     <span class="n">recombination_map</span><span class="o">=</span><span class="n">recombination_map</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1084</span>     <span class="n">gene_conversion_map</span><span class="o">=</span><span class="n">gene_conversion_map</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1085</span>     <span class="n">gene_conversion_tract_length</span><span class="o">=</span><span class="n">gene_conversion_tract_length</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1086</span>     <span class="n">discrete_genome</span><span class="o">=</span><span class="n">discrete_genome</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1087</span>     <span class="n">ploidy</span><span class="o">=</span><span class="n">ploidy</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1088</span>     <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1089</span>     <span class="n">models</span><span class="o">=</span><span class="n">models</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1090</span>     <span class="n">store_migrations</span><span class="o">=</span><span class="n">record_migrations</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1091</span>     <span class="n">additional_nodes</span><span class="o">=</span><span class="n">additional_nodes</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1092</span>     <span class="n">coalescing_segments_only</span><span class="o">=</span><span class="n">coalescing_segments_only</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1093</span>     <span class="n">start_time</span><span class="o">=</span><span class="n">start_time</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1094</span>     <span class="n">end_time</span><span class="o">=</span><span class="n">end_time</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1095</span>     <span class="n">num_labels</span><span class="o">=</span><span class="n">num_labels</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1096</span>     <span class="n">random_generator</span><span class="o">=</span><span class="n">random_generator</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1097</span> <span class="p">)</span>

<span class="nn">File ~/work/tskit-site/tskit-site/msprime/ancestry.py:1426,</span> in <span class="ni">Simulator.__init__</span><span class="nt">(self, tables, recombination_map, gene_conversion_map, gene_conversion_tract_length, discrete_genome, ploidy, demography, random_generator, models, store_migrations, additional_nodes, coalescing_segments_only, start_time, end_time, num_labels)</span>
<span class="g g-Whitespace">   </span><span class="mi">1423</span> <span class="n">gene_conversion_rate</span> <span class="o">=</span> <span class="n">gene_conversion_map</span><span class="o">.</span><span class="n">rate</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="g g-Whitespace">   </span><span class="mi">1425</span> <span class="n">start_time</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span> <span class="k">if</span> <span class="n">start_time</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">start_time</span>
<span class="ne">-&gt; </span><span class="mi">1426</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1427</span>     <span class="n">tables</span><span class="o">=</span><span class="n">ll_tables</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1428</span>     <span class="n">recombination_map</span><span class="o">=</span><span class="n">ll_recomb_map</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1429</span>     <span class="n">start_time</span><span class="o">=</span><span class="n">start_time</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1430</span>     <span class="n">random_generator</span><span class="o">=</span><span class="n">random_generator</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1431</span>     <span class="n">migration_matrix</span><span class="o">=</span><span class="n">demography</span><span class="o">.</span><span class="n">migration_matrix</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1432</span>     <span class="n">population_configuration</span><span class="o">=</span><span class="n">ll_population_configuration</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1433</span>     <span class="n">demographic_events</span><span class="o">=</span><span class="n">ll_demographic_events</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1434</span>     <span class="n">store_migrations</span><span class="o">=</span><span class="n">store_migrations</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1435</span>     <span class="n">additional_nodes</span><span class="o">=</span><span class="n">additional_nodes</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1436</span>     <span class="n">coalescing_segments_only</span><span class="o">=</span><span class="n">coalescing_segments_only</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1437</span>     <span class="n">num_labels</span><span class="o">=</span><span class="n">num_labels</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1438</span>     <span class="n">segment_block_size</span><span class="o">=</span><span class="n">segment_block_size</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1439</span>     <span class="n">avl_node_block_size</span><span class="o">=</span><span class="n">avl_node_block_size</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1440</span>     <span class="n">node_mapping_block_size</span><span class="o">=</span><span class="n">node_mapping_block_size</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1441</span>     <span class="n">gene_conversion_rate</span><span class="o">=</span><span class="n">gene_conversion_rate</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1442</span>     <span class="n">gene_conversion_tract_length</span><span class="o">=</span><span class="n">gene_conversion_tract_length</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1443</span>     <span class="n">discrete_genome</span><span class="o">=</span><span class="n">discrete_genome</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1444</span>     <span class="n">ploidy</span><span class="o">=</span><span class="n">ploidy</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1445</span> <span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1446</span> <span class="c1"># Highlevel attributes used externally that have no lowlevel equivalent</span>
<span class="g g-Whitespace">   </span><span class="mi">1447</span> <span class="bp">self</span><span class="o">.</span><span class="n">end_time</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">inf</span> <span class="k">if</span> <span class="n">end_time</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">end_time</span>

<span class="ne">InputError</span>: Input error in initialise: Attempt to sample lineage in a population with size=0
</pre></div>
</div>
</div>
</div>
<p>To recover the original model we must update the <a class="reference internal" href="api.html#msprime.Population" title="msprime.Population"><code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code></a>
objects in place and add the
<a class="reference internal" href="demography.html#sec-demography-events-population-split"><span class="std std-ref">population split</span></a> event:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">initial_size</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">demography</span><span class="p">[</span><span class="s2">&quot;B&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">initial_size</span> <span class="o">=</span> <span class="mi">200</span>
<span class="n">demography</span><span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">initial_size</span> <span class="o">=</span> <span class="mi">300</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population_split</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">ancestral</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">])</span>
<span class="n">demography</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (3)</b><table border="1" class="tskit-table"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>A</td><td>Contemporary population A</td><td>100.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>B</td><td>Contemporary population B</td><td>200.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>2</td><td>C</td><td>Ancestral population</td><td>300.0</td><td>0</td><td>1e+02</td><td>{}</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Migration matrix (all zero)</b><table border="1" class="tskit-table"><thead><tr></tr></thead><tbody></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events (1)</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>100</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[A, B], ancestral=C</td><td>Moves all lineages from derived populations &#x27;A&#x27; and &#x27;B&#x27; to the ancestral &#x27;C&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr></tbody></table></div></div></div></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>See the <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> documentation for details on how to access
<a class="reference internal" href="api.html#msprime.Population" title="msprime.Population"><code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code></a> objects.</p>
</div>
<p>The demography that we obtain from <a class="reference internal" href="api.html#msprime.Demography.from_tree_sequence" title="msprime.Demography.from_tree_sequence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.from_tree_sequence()</span></code></a>
is really just a template, which we update and elaborate as we need.
In particular, we are free to add more populations as needed.</p>
</section>
<section id="continuing-forwards-time-simulations-recapitating">
<span id="sec-ancestry-initial-state-forward-simulations"></span><h3>Continuing forwards-time simulations (“recapitating”)<a class="headerlink" href="#continuing-forwards-time-simulations-recapitating" title="Link to this heading">#</a></h3>
<p>The most common use for the <code class="docutils literal notranslate"><span class="pre">initial_state</span></code> argument is to
start the simulation from the output of a forwards-time simulation.
Informally, we take an ‘unfinished’
tree sequence as a parameter to <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a>, initialise the simulation
from the state of this tree sequence and then run the simulation until
coalescence. The returned tree sequence is then the result of taking the
input tree sequence and completing the trees using the coalescent.</p>
<p>This is useful for forwards-time simulators such as
<a class="reference external" href="https://messerlab.org/slim/">SLiM</a> and
<a class="reference external" href="https://pypi.org/project/fwdpy11/">fwdpy11</a>
that can output tree sequences. By running
forward-time simulation for a certain number of generations we obtain a
tree sequence, but these trees may not have had sufficient time to
reach a most recent common ancestor. By using the <code class="docutils literal notranslate"><span class="pre">initial_state</span></code> argument
to <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a> we can combine the best of both forwards- and
backwards-time simulators. The recent past can be simulated forwards
in time and the ancient past by the coalescent. The coalescent
simulation is initialised by the root segments of the
input tree sequence, ensuring that the minimal amount of ancestral
material possible is simulated.</p>
<p>Any tree sequence can be provided as input to this process, but there is a
specific topological requirement that must be met for the simulations to be
statistically correct. To ensure that ancestral segments are correctly
associated within chromosomes when constructing the initial conditions for the
coalescent simulation, forward-time simulators <strong>must</strong> retain the nodes
corresponding to the initial generation. Furthermore, for every sample in the
final generation (i.e. the extant population at the present time) there must be
a path to one of the founder population nodes.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>See the <a class="reference external" href="https://tskit.dev/pyslim/docs/latest/tutorial.html#sec-tutorial-recapitation" title="(in Project name not set)"><span class="xref std std-ref">recapitation tutorial</span></a>
from the <a class="reference external" href="https://github.com/tskit-dev/pyslim">pyslim</a> documentation
for a detailed explanation of this process when
using <a class="reference external" href="https://messerlab.org/slim/">SLiM</a>.</p>
</div>
</section>
</section>
<section id="models">
<span id="sec-ancestry-models"></span><h2>Models<a class="headerlink" href="#models" title="Link to this heading">#</a></h2>
<p>The ancestry model determines the model under which the ancestral
history of the sample is generated. It is concerned with the
basic processes of how (for example) common ancestor and
recombination events occur. For example, the default
“standard” or <a class="reference internal" href="#sec-ancestry-models-hudson"><span class="std std-ref">Hudson coalescent</span></a> is an
efficient continuous time model, and the
<a class="reference internal" href="#sec-ancestry-models-dtwf"><span class="std std-ref">Discrete Time Wright-Fisher</span></a> model allows
for more detailed (if less efficient) simulations by
making fewer mathematical approximations.
We can combine multiple ancestry models in a simulation
using a flexible <a class="reference internal" href="#sec-ancestry-models-specifying"><span class="std std-ref">notation</span></a>.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>See the <a class="reference internal" href="#sec-ancestry-quickref"><span class="std std-ref">Quick reference</span></a> for a summary of the available
ancestry models</p>
</div>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>The ancestry model is distinct from the <a class="reference internal" href="demography.html#sec-demography"><span class="std std-ref">demographic model</span></a>,
which is mostly independent of the chosen ancestry model.</p>
</div>
<section id="specifying-ancestry-models">
<span id="sec-ancestry-models-specifying"></span><h3>Specifying ancestry models<a class="headerlink" href="#specifying-ancestry-models" title="Link to this heading">#</a></h3>
<p>The ancestry models used during a simulation are specified using
the <code class="docutils literal notranslate"><span class="pre">model</span></code> parameter to <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a>. Each model is a
subclass of the <a class="reference internal" href="api.html#msprime.AncestryModel" title="msprime.AncestryModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">AncestryModel</span></code></a> class (see the following
subsections for the available models, and examples of their usage).</p>
<section id="single-models">
<h4>Single models<a class="headerlink" href="#single-models" title="Link to this heading">#</a></h4>
<p>Most of the time we want to run a simulation under a single
ancestry model. By default, we run simulations under the
<a class="reference internal" href="api.html#msprime.StandardCoalescent" title="msprime.StandardCoalescent"><code class="xref py py-class docutils literal notranslate"><span class="pre">StandardCoalescent</span></code></a> model. If we wish to run
under a different model, we use the <code class="docutils literal notranslate"><span class="pre">model</span></code> argument to
<a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a>. For example, here we use the
<a class="reference internal" href="api.html#msprime.SmcApproxCoalescent" title="msprime.SmcApproxCoalescent"><code class="xref py py-class docutils literal notranslate"><span class="pre">SMC</span></code></a> model instead of the
standard coalescent:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts1</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">10</span><span class="p">,</span>
    <span class="n">sequence_length</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">SmcApproxCoalescent</span><span class="p">(),</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">1234</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We specify the model as an instance of one of the ancestry model classes.
For nonparametric models, there is also a useful shorthand
of the model name:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts2</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">10</span><span class="p">,</span>
    <span class="n">sequence_length</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="s2">&quot;smc&quot;</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">1234</span><span class="p">)</span>
<span class="k">assert</span> <span class="n">ts1</span><span class="o">.</span><span class="n">equals</span><span class="p">(</span><span class="n">ts2</span><span class="p">,</span> <span class="n">ignore_provenance</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>This string form is useful for single model simulations;
however, when working with
<a class="reference internal" href="#sec-ancestry-models-specifying-multiple"><span class="std std-ref">Multiple models</span></a> it is better to use
the explicit class form so that the
<a class="reference internal" href="#sec-ancestry-models-specifying-duration"><span class="std std-ref">Model duration</span></a> can be set.</p>
</section>
<section id="model-duration">
<span id="sec-ancestry-models-specifying-duration"></span><h4>Model duration<a class="headerlink" href="#model-duration" title="Link to this heading">#</a></h4>
<p>Each ancestry model instance has a <a class="reference internal" href="api.html#msprime.AncestryModel.duration" title="msprime.AncestryModel.duration"><code class="xref py py-attr docutils literal notranslate"><span class="pre">duration</span></code></a> associated
with it. This is the maximum amount of time that this model can run for. Thus,
if we wanted to run 10 generations of the <a class="reference internal" href="api.html#msprime.DiscreteTimeWrightFisher" title="msprime.DiscreteTimeWrightFisher"><code class="xref py py-class docutils literal notranslate"><span class="pre">DiscreteTimeWrightFisher</span></code></a>
model we could write:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">3</span><span class="p">,</span>
    <span class="n">population_size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">DiscreteTimeWrightFisher</span><span class="p">(</span><span class="n">duration</span><span class="o">=</span><span class="mi">10</span><span class="p">),</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">1234</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/271f15721480677eafe98ea4c70e9792e42882bbeeca02207344f9a1be7e4f0f.svg" src="_images/271f15721480677eafe98ea4c70e9792e42882bbeeca02207344f9a1be7e4f0f.svg" />
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Using the duration value for a single model like this is identical to
specifying an <code class="docutils literal notranslate"><span class="pre">end_time</span></code> value: see the <a class="reference internal" href="#sec-ancestry-end-time"><span class="std std-ref">Stopping simulations early</span></a>
section for more information.</p>
</div>
<p>It is vital to understand that the <code class="docutils literal notranslate"><span class="pre">duration</span></code> value here is <strong>not</strong>
an absolute time at which the simulation must stop, but rather
the number of generations that the model should run for. To illustrate
this we can use the <a class="reference internal" href="#sec-ancestry-start-time"><span class="std std-ref">start_time</span></a> parameter:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">3</span><span class="p">,</span>
    <span class="n">population_size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">start_time</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">DiscreteTimeWrightFisher</span><span class="p">(</span><span class="n">duration</span><span class="o">=</span><span class="mi">10</span><span class="p">),</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">1234</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/5708332a0b6dfd2080cc9ead69b2ad7bcf0abec7dc53b03f58f46c9aaf3d2248.svg" src="_images/5708332a0b6dfd2080cc9ead69b2ad7bcf0abec7dc53b03f58f46c9aaf3d2248.svg" />
</div>
</div>
<p>We can see that the simulation continued until 15 generations ago,
because it didn’t start until 5 generations ago and we then specified
that the model should run for at most 10 generations.</p>
<p>The model activity time is specified as a duration in this way
as this is a natural way to specify
<a class="reference internal" href="#sec-ancestry-models-specifying-multiple"><span class="std std-ref">Multiple models</span></a>, in which
some of the models may run for random periods of time.</p>
</section>
<section id="model-completion">
<span id="sec-ancestry-models-specifying-completion"></span><h4>Model completion<a class="headerlink" href="#model-completion" title="Link to this heading">#</a></h4>
<p>Most ancestry models will run until the overall simulation has
completed; that is, when we have simulated back to the most
recent common ancestor at every position along the sequence.
For example, the <a class="reference internal" href="api.html#msprime.StandardCoalescent" title="msprime.StandardCoalescent"><code class="xref py py-class docutils literal notranslate"><span class="pre">StandardCoalescent</span></code></a> and
<a class="reference internal" href="api.html#msprime.DiscreteTimeWrightFisher" title="msprime.DiscreteTimeWrightFisher"><code class="xref py py-class docutils literal notranslate"><span class="pre">DiscreteTimeWrightFisher</span></code></a> models will simulate
until coalescence, unless we explicitly indicate that we wish
to stop the simulation early via the
<a class="reference internal" href="#sec-ancestry-models-specifying-duration"><span class="std std-ref">model duration</span></a>
or the <a class="reference internal" href="#sec-ancestry-end-time"><span class="std std-ref">end_time</span></a> parameter.</p>
<p>Models such as <a class="reference internal" href="api.html#msprime.SweepGenicSelection" title="msprime.SweepGenicSelection"><code class="xref py py-class docutils literal notranslate"><span class="pre">SweepGenicSelection</span></code></a> are different though:
they are only defined for a limited amount of time and
may not run until coalescence (see the
<a class="reference internal" href="#sec-ancestry-models-selective-sweeps"><span class="std std-ref">Selective sweeps</span></a>
for more information on the model itself).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sweep</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">SweepGenicSelection</span><span class="p">(</span>
    <span class="n">position</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
    <span class="n">start_frequency</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
    <span class="n">end_frequency</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span>
    <span class="n">s</span><span class="o">=</span><span class="mf">0.25</span><span class="p">,</span>
    <span class="n">dt</span><span class="o">=</span><span class="mf">1e-6</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">3</span><span class="p">,</span>
    <span class="n">sequence_length</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">population_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="n">sweep</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">1234</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/eefbe1cdfe74833d8a81d33add0367e68b89cff56e17ac7c6e3c5f9aec5048cf.svg" src="_images/eefbe1cdfe74833d8a81d33add0367e68b89cff56e17ac7c6e3c5f9aec5048cf.svg" />
</div>
</div>
<p>Here, we see that even though we didn’t specify
a <a class="reference internal" href="#sec-ancestry-models-specifying-duration"><span class="std std-ref">model duration</span></a>
or <a class="reference internal" href="#sec-ancestry-end-time"><span class="std std-ref">end_time</span></a> parameter the simulation
has ended before we reached coalescence. In addition, the
model duration is <strong>random</strong>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">3</span><span class="p">,</span>
    <span class="n">sequence_length</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">population_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="n">sweep</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">12345</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/3dab8c2a0b030a4e487e73dfaf84fd6c34f5a049ae119f8aef5d4bfaec7430c4.svg" src="_images/3dab8c2a0b030a4e487e73dfaf84fd6c34f5a049ae119f8aef5d4bfaec7430c4.svg" />
</div>
</div>
<p>Here we ran the same simulation with a different random seed, and
we can see that the time the model completed was different in
the two cases (30.22 vs 28.15 generations ago).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Models such as <a class="reference internal" href="api.html#msprime.SweepGenicSelection" title="msprime.SweepGenicSelection"><code class="xref py py-class docutils literal notranslate"><span class="pre">SweepGenicSelection</span></code></a> can also complete due
to full coalescence.</p>
</div>
</section>
<section id="multiple-models">
<span id="sec-ancestry-models-specifying-multiple"></span><h4>Multiple models<a class="headerlink" href="#multiple-models" title="Link to this heading">#</a></h4>
<p>Sometimes we are interested in combining different models in a simulation.
To do this we provide a <strong>list</strong> of ancestry model instances as the
<code class="docutils literal notranslate"><span class="pre">model</span></code> argument to <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a>. The models
are run in the order they are specified. For each model, if its
<a class="reference internal" href="api.html#msprime.AncestryModel.duration" title="msprime.AncestryModel.duration"><code class="xref py py-attr docutils literal notranslate"><span class="pre">duration</span></code></a> is set, then the simulation will run
for at most that additional time duration. If the <code class="docutils literal notranslate"><span class="pre">duration</span></code> is
not set (or <code class="docutils literal notranslate"><span class="pre">None</span></code>), then the model will run until
<a class="reference internal" href="#sec-ancestry-models-specifying-completion"><span class="std std-ref">completion</span></a>.</p>
<p>For example, here we run a hybrid DTWF and coalescent
simulation (see the <a class="reference internal" href="#sec-ancestry-models-dtwf"><span class="std std-ref">Discrete Time Wright-Fisher</span></a> section for more
details):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">2</span><span class="p">,</span>
    <span class="n">population_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="p">[</span>
        <span class="n">msprime</span><span class="o">.</span><span class="n">DiscreteTimeWrightFisher</span><span class="p">(</span><span class="n">duration</span><span class="o">=</span><span class="mi">500</span><span class="p">),</span>
        <span class="n">msprime</span><span class="o">.</span><span class="n">StandardCoalescent</span><span class="p">(),</span>
    <span class="p">],</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">2</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>It is very imporant to remember to set a <code class="docutils literal notranslate"><span class="pre">duration</span></code> value in the first
model here! Otherwise, the entire simulation will run under the DTWF model.</p>
</div>
<p>In this example we’ve run the <a class="reference internal" href="api.html#msprime.DiscreteTimeWrightFisher" title="msprime.DiscreteTimeWrightFisher"><code class="xref py py-class docutils literal notranslate"><span class="pre">DiscreteTimeWrightFisher</span></code></a>
model for the first 500 generations, and then switched to the
<a class="reference internal" href="api.html#msprime.StandardCoalescent" title="msprime.StandardCoalescent"><code class="xref py py-class docutils literal notranslate"><span class="pre">StandardCoalescent</span></code></a>. Because we did not specify a
<code class="docutils literal notranslate"><span class="pre">duration</span></code> in the second model, it will run until coalescence.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Switching models at a fixed time point like this is equivalent to
first running the DTWF phase of the simulation with <code class="docutils literal notranslate"><span class="pre">end_time=500</span></code>
and then using the output as the
<a class="reference internal" href="#sec-ancestry-initial-state"><span class="std std-ref">initial state</span></a> for the simulation
of the Hudson phase. The model switching syntax here is a little
more convenient and efficient, however.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>See the <a class="reference internal" href="#sec-ancestry-models-selective-sweeps-multiple"><span class="std std-ref">Multiple sweeps</span></a> section for
an example of running many models with random durations.</p>
</div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>The logging output of msprime can be very useful when working with
multiple models. See the <a class="reference internal" href="logging.html#sec-logging"><span class="std std-ref">Logging</span></a> section for more
details.</p>
</div>
</section>
</section>
<section id="hudson-coalescent">
<span id="sec-ancestry-models-hudson"></span><h3>Hudson coalescent<a class="headerlink" href="#hudson-coalescent" title="Link to this heading">#</a></h3>
<p>The <a class="reference internal" href="api.html#msprime.StandardCoalescent" title="msprime.StandardCoalescent"><code class="xref py py-class docutils literal notranslate"><span class="pre">standard</span> <span class="pre">coalescent</span></code></a>
is the default model in msprime. The algorithm
for simulating the coalescent with recombination was developed
by <a class="reference external" href="https://doi.org/10.1016/0040-5809(83)90013-8">Hudson (1983)</a>,
and so we refer to it as the “Hudson” coalescent.</p>
<p>Running a simulation without specifying a <code class="docutils literal notranslate"><span class="pre">model</span></code> is the
same as running with <code class="docutils literal notranslate"><span class="pre">model=&quot;hudson&quot;</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts1</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">ts2</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s2">&quot;hudson&quot;</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="c1"># This is the same simulation so tree sequences are equal</span>
<span class="k">assert</span> <span class="n">ts1</span><span class="o">.</span><span class="n">equals</span><span class="p">(</span><span class="n">ts2</span><span class="p">,</span> <span class="n">ignore_provenance</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Time is measured in units of generations ago. This is a continuous time
model, in which the simulation moves event-by-event backwards
in time (contrasting with the <a class="reference internal" href="#sec-ancestry-models-dtwf"><span class="std std-ref">Discrete Time Wright-Fisher</span></a> model, which
works generation-by-generation). Thus, we can have fractional
generations, as in this example where the MRCA of the sample
occurs 10.59 “generations” ago:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">234</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/0ffdc8a306ca020f4de59b96dba11378e0b5983b036e1e1166cabe5636147ca7.svg" src="_images/0ffdc8a306ca020f4de59b96dba11378e0b5983b036e1e1166cabe5636147ca7.svg" />
</div>
</div>
<p>This occurs because time is scaled to be proportional to
the population size; if we run the same simulation with a
different <a class="reference internal" href="#sec-ancestry-demography"><span class="std std-ref">population size</span></a>, we can see that
the time values are simply scaled up accordingly
(the default <code class="docutils literal notranslate"><span class="pre">population_size</span></code> is 1):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="n">population_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">234</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/525d3a255df07178fb494f851fe89da738423a7db212eecea0e8688f062479bf.svg" src="_images/525d3a255df07178fb494f851fe89da738423a7db212eecea0e8688f062479bf.svg" />
</div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>The <a class="reference internal" href="#sec-ancestry-ploidy"><span class="std std-ref">ploidy</span></a> parameter also controls
the time scale; the
default here is the diploid time scale where time is measured
in units of 4N generations.</p>
</div>
</section>
<section id="smc-approximations">
<span id="sec-ancestry-models-smc"></span><h3>SMC approximations<a class="headerlink" href="#smc-approximations" title="Link to this heading">#</a></h3>
<p>The <a class="reference internal" href="api.html#msprime.SmcApproxCoalescent" title="msprime.SmcApproxCoalescent"><code class="xref py py-class docutils literal notranslate"><span class="pre">SMC</span></code></a> and <a class="reference internal" href="api.html#msprime.SmcPrimeApproxCoalescent" title="msprime.SmcPrimeApproxCoalescent"><code class="xref py py-class docutils literal notranslate"><span class="pre">SMC'</span></code></a>
are approximations of the continuous time
<a class="reference internal" href="#sec-ancestry-models-hudson"><span class="std std-ref">Hudson coalescent</span></a> model. These were originally
motivated largely by the need to simulate coalescent processes more efficiently
than was possible using the software available at the time; however,
<a class="reference external" href="https://doi.org/10.1371/journal.pcbi.1004842">improved algorithms</a>
mean that such approximations are now mostly unnecessary for simulations.</p>
<p>The SMC and SMC’ are however very important for inference, as the approximations
have made many analytical advances possible.</p>
<p>Since the SMC approximations are not required for simulation efficiency, these
models are implemented using a naive rejection sampling approach in msprime.
The implementation is intended to facilitate the study of the
SMC approximations, rather than to be used in a general-purpose way.</p>
</section>
<section id="discrete-time-wright-fisher">
<span id="sec-ancestry-models-dtwf"></span><h3>Discrete Time Wright-Fisher<a class="headerlink" href="#discrete-time-wright-fisher" title="Link to this heading">#</a></h3>
<p>Msprime provides the option to perform discrete-time Wright-Fisher (DTWF)
simulations for scenarios when the coalescent model is not appropriate, including large
sample sizes, multiple chromosomes, or recent migration.
Please see the <a class="reference internal" href="api.html#msprime.DiscreteTimeWrightFisher" title="msprime.DiscreteTimeWrightFisher"><code class="xref py py-class docutils literal notranslate"><span class="pre">API</span> <span class="pre">documentation</span></code></a>
for a formal description of the model,
and <a class="reference external" href="https://doi.org/10.1371/journal.pgen.1008619">Nelson et al. 2020</a> for
more details and background information.</p>
<p>To run DTWF simulations, we need to provide the <code class="docutils literal notranslate"><span class="pre">model=&quot;dtwf&quot;</span></code> argument
to <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">population_size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s2">&quot;dtwf&quot;</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">time_scale</span><span class="o">=</span><span class="s2">&quot;log_time&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/8564e6bd95f1de9dcd330624ee44f5034d25f0b28d1ed55688926f4fe188b7e9.svg" src="_images/8564e6bd95f1de9dcd330624ee44f5034d25f0b28d1ed55688926f4fe188b7e9.svg" />
</div>
</div>
<p>There are a few important points to note here:</p>
<ol class="arabic simple">
<li><p>All node times are integers (this is discrete time counted in generations);</p></li>
<li><p>We can have <strong>non-binary</strong> nodes (see node 10);</p></li>
<li><p>We can have <strong>simultaneous</strong> events (see nodes 8 and 9).</p></li>
</ol>
<p>Most features can be used with the DTWF model, and an error will be raised if
you attempt to use a feature that is not implemented for DTWF. (Please let
us know if this feature is important to you!)</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>For DTWF simulations with population structure, each row of the migration
matrix must sum to one or less.</p>
</div>
<p>Because DTWF simulations trace the history of a sample by going backwards
in time generation-by-generation rather than event-by-event like the
continuous time models, it is significantly slower to simulate than
the standard coalescent.</p>
<p>“Hybrid” simulations can be a good solution if DTWF simulations are
slow. In this case we can use DTWF simulations for the recent past, and using
the standard coalescent for the more distant past (where it is an excellent
approximation of the discrete time model). Hybrid simulations
of this type provide the best of computational efficiency and accuracy.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>See <a class="reference external" href="https://doi.org/10.1371/journal.pgen.1008619">Nelson et al. 2020</a> for
more details and examples where the DTWF model is more realistic than
the coalescent, and an assessment of the accuracy of hybrid simulations.</p>
</div>
<p>For example, here we simulate with the DTWF model for 500 generations,
before switching to the standard (Hudson) coalescent:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">2</span><span class="p">,</span>
    <span class="n">population_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="p">[</span>
        <span class="n">msprime</span><span class="o">.</span><span class="n">DiscreteTimeWrightFisher</span><span class="p">(</span><span class="n">duration</span><span class="o">=</span><span class="mi">500</span><span class="p">),</span>
        <span class="n">msprime</span><span class="o">.</span><span class="n">StandardCoalescent</span><span class="p">(),</span>
    <span class="p">],</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">4</span>
<span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">time_scale</span><span class="o">=</span><span class="s2">&quot;log_time&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/5fa3d77c916ed4e2d2ee52149e035cb6ad9b540acedbe10e2ba8a765b9a7ac63.svg" src="_images/5fa3d77c916ed4e2d2ee52149e035cb6ad9b540acedbe10e2ba8a765b9a7ac63.svg" />
</div>
</div>
<p>Because of the integer node times, we can see here that most of the coalescent
happened during the Wright-Fisher phase of the simulation, and as-of 500
generations in the past, there were only two lineages left. The continuous
time standard coalescent model was then used to simulate the ancient past of
these two lineages.</p>
<p>See the <a class="reference internal" href="#sec-ancestry-models-specifying"><span class="std std-ref">Specifying ancestry models</span></a> section for more details on
how the use the <code class="docutils literal notranslate"><span class="pre">model</span></code> argument to <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a> to run
multiple models.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>See the <a class="reference internal" href="#sec-ancestry-multiple-chromosomes"><span class="std std-ref">Multiple chromosomes</span></a> section for information
on how to simulate multiple chromosomes using the DTWF.</p>
</div>
</section>
<section id="multiple-merger-coalescents">
<span id="sec-ancestry-models-multiple-mergers"></span><h3>Multiple merger coalescents<a class="headerlink" href="#multiple-merger-coalescents" title="Link to this heading">#</a></h3>
<p>Some evolutionary scenarios, such as a skewed offspring distribution
combined with a type III survivorship curve, range expansion, and
rapid adaptation, can predict genealogies with up to four simultaneous
multiple mergers. Msprime provides the option to simulate from two classes
of such genealogical processes: the <a class="reference internal" href="api.html#msprime.BetaCoalescent" title="msprime.BetaCoalescent"><code class="xref py py-class docutils literal notranslate"><span class="pre">BetaCoalescent</span></code></a> and
the <a class="reference internal" href="api.html#msprime.DiracCoalescent" title="msprime.DiracCoalescent"><code class="xref py py-class docutils literal notranslate"><span class="pre">DiracCoalescent</span></code></a>. Please see the API documentation for formal
details of the models.</p>
<p>For haploid organisms, both models result in genealogies in which
any number of lineages can merge into a common ancestor,
but only one merger event can take place at a given time. For <span class="math notranslate nohighlight">\(p\)</span>-ploids,
up to <span class="math notranslate nohighlight">\(2p\)</span> simultaneous mergers can take place, corresponding to the
<span class="math notranslate nohighlight">\(2p\)</span> available parental chromosome copies.</p>
<p>The diploid Beta-Xi-coalescent can be simulated as follows:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">BetaCoalescent</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">1.001</span><span class="p">))</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/3296b5c225cf50eeacd08937ee84b6dcbe2b17bb41905b34f5488cae216122d6.svg" src="_images/3296b5c225cf50eeacd08937ee84b6dcbe2b17bb41905b34f5488cae216122d6.svg" />
</div>
</div>
<p>The specified value of <span class="math notranslate nohighlight">\(\alpha = 1.001\)</span> corresponds to a heavily skewed
offspring distribution. Values closer to <span class="math notranslate nohighlight">\(\alpha = 2\)</span> result in trees
whose distribution is closer to that of the standard coalescent, often featuring
no multiple mergers for small sample sizes:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">BetaCoalescent</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">1.8</span><span class="p">))</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/190ee68f4cd1f07df3f5d7cee840e2495411a3563d91099618c903b406ed6129.svg" src="_images/190ee68f4cd1f07df3f5d7cee840e2495411a3563d91099618c903b406ed6129.svg" />
</div>
</div>
<p>Multiple mergers still take place in a haploid simulation, but only one merger
can take place at a given time:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">BetaCoalescent</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">1.001</span><span class="p">))</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/a4bcdb2ea8089ba0cbe23f10e0f4cf3bba110206cd7cb2f75607f74860b2cb0b.svg" src="_images/a4bcdb2ea8089ba0cbe23f10e0f4cf3bba110206cd7cb2f75607f74860b2cb0b.svg" />
</div>
</div>
<p>A haploid simulation results in larger individual mergers than a polyploid simulation
because large mergers typically get broken up into multiple simultaneous mergers
in the polyploid model.</p>
<p>The number of generations between merger events in the Beta-coalescent depends
nonlinearly on both <span class="math notranslate nohighlight">\(\alpha\)</span> and the population size <span class="math notranslate nohighlight">\(N\)</span>
as detailed above.
For a fixed <span class="math notranslate nohighlight">\(\alpha\)</span>, the number of generations between common ancestor events
is proportional to <span class="math notranslate nohighlight">\(N^{\alpha - 1}\)</span>, albeit with a complicated constant of
proportionality that depends on <span class="math notranslate nohighlight">\(\alpha\)</span>. The dependence on <span class="math notranslate nohighlight">\(\alpha\)</span>
for fixed <span class="math notranslate nohighlight">\(N\)</span> is not monotone. Thus, branch lengths and the number of
generations until a most recent common ancestor depend on both of these parameters.</p>
<p>To illustrate, for <span class="math notranslate nohighlight">\(\alpha\)</span> close to 2 the relationship between effective
population size and number of generations is almost linear:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">population_size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">BetaCoalescent</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">1.99</span><span class="p">))</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">first</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">tmrca</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">population_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">BetaCoalescent</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">1.99</span><span class="p">))</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">first</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">tmrca</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0.14959691919068155
14.286394871874865
</pre></div>
</div>
</div>
</div>
<p>For <span class="math notranslate nohighlight">\(\alpha\)</span> close to 1 the effective population size has little effect:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">population_size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">BetaCoalescent</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">1.1</span><span class="p">))</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">first</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">tmrca</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">population_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">BetaCoalescent</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">1.1</span><span class="p">))</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">first</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">tmrca</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>16.311807036386615
25.85247192870844
</pre></div>
</div>
</div>
</div>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>Both the haploid and polyploid timescales collapse to zero as <span class="math notranslate nohighlight">\(\alpha\)</span>
approaches 2. For reasons of numerical stability, values above
<span class="math notranslate nohighlight">\(\alpha = 1.991\)</span> have been prohibited. The
<a class="reference internal" href="#sec-ancestry-models-hudson"><span class="std std-ref">Hudson coalescent</span></a> provides a very good
approximation for <span class="math notranslate nohighlight">\(1.991 &lt;= \alpha &lt;= 2\)</span>.</p>
</div>
<p>The Dirac-coalescent is simulated similarly in both the diploid case:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">DiracCoalescent</span><span class="p">(</span><span class="n">psi</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="mi">10</span><span class="p">))</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/552b1bb797f284e5e516a2b89cf011312aa5e97392652bf72a4755f18b1b9f52.svg" src="_images/552b1bb797f284e5e516a2b89cf011312aa5e97392652bf72a4755f18b1b9f52.svg" />
</div>
</div>
<p>and in the haploid case:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">DiracCoalescent</span><span class="p">(</span><span class="n">psi</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="mi">10</span><span class="p">))</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/453f1926088591bfbe29bb75cf92cc6b8dfc754b9172c90c8d48d974f4b73f8b.svg" src="_images/453f1926088591bfbe29bb75cf92cc6b8dfc754b9172c90c8d48d974f4b73f8b.svg" />
</div>
</div>
<p>As with the Beta-coalescent, a haploid simulation results in larger individual
mergers than a polyploid simulation because large mergers typically get broken
up into multiple simultaneous mergers in the polyploid model. Larger values
of the parameter <span class="math notranslate nohighlight">\(c &gt; 0\)</span> result in more frequent multiple mergers,
while larger values of <span class="math notranslate nohighlight">\(0 &lt; \psi \leq 1\)</span> result in multiple mergers
with more participating lineages. Setting either parameter to 0 would correspond
to the standard coalescent.</p>
<p>The Dirac-coalescent is obtained as the infinite population scaling limit of
Moran models, and therefore branch lengths are proportional to <span class="math notranslate nohighlight">\(N^2\)</span>
generations, as opposed to <span class="math notranslate nohighlight">\(N\)</span> generations under the standard coalescent.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">population_size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">DiracCoalescent</span><span class="p">(</span><span class="n">psi</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">first</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">tmrca</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">population_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">DiracCoalescent</span><span class="p">(</span><span class="n">psi</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">first</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">tmrca</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>161.80418322293528
16180.418322293526
</pre></div>
</div>
</div>
</div>
<p>Because branch lengths are proportional to <span class="math notranslate nohighlight">\(N^2\)</span> generations, the number
of events simulated along branches (such as mutations or recombinations) will
also be proportional to <span class="math notranslate nohighlight">\(N^2\)</span>. For example, if a tree sequence generated
using a given census size <span class="math notranslate nohighlight">\(N\)</span> is used as input for simulating segregating
sites (see <a class="reference internal" href="mutations.html#sec-mutations"><span class="std std-ref">Mutation simulations</span></a> for details on how to do
this), the expected number of segregating sites obtained will be proportional
to <span class="math notranslate nohighlight">\(N^2\)</span>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">population_size</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span><span class="n">recombination_rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span>
    <span class="n">sequence_length</span><span class="o">=</span><span class="mi">10</span><span class="o">**</span><span class="mi">4</span><span class="p">,</span><span class="n">model</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">DiracCoalescent</span><span class="p">(</span><span class="n">psi</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="mi">0</span><span class="p">))</span>
<span class="n">mts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_mutations</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">mts</span><span class="o">.</span><span class="n">get_num_sites</span><span class="p">()))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>6536
</pre></div>
</div>
</div>
</div>
<p>If the desired amount of diversity (as measured by the number of segregating
sites) is proportional to <span class="math notranslate nohighlight">\(N\)</span>, then <span class="math notranslate nohighlight">\(N^{1/2}\)</span> should be used for
the population size under the Dirac-coalescent.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">population_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span><span class="n">recombination_rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span>
    <span class="n">sequence_length</span><span class="o">=</span><span class="mi">10</span><span class="o">**</span><span class="mi">4</span><span class="p">,</span><span class="n">model</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">DiracCoalescent</span><span class="p">(</span><span class="n">psi</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="mi">0</span><span class="p">))</span>
<span class="n">mts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_mutations</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">mts</span><span class="o">.</span><span class="n">get_num_sites</span><span class="p">()))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>5
</pre></div>
</div>
</div>
</div>
<p>Therefore the population size must be carefully chosen (and
potentially rescaled) to obtain the desired simulated data in these models.</p>
</section>
<section id="selective-sweeps">
<span id="sec-ancestry-models-selective-sweeps"></span><h3>Selective sweeps<a class="headerlink" href="#selective-sweeps" title="Link to this heading">#</a></h3>
<p>Although the coalescent assumes that lineages are exchangable, (and therefore
evolving neutrally), approximations to some forms of selective sweep
(beneficial mutation moves through the population), have been derived. This is
done through the use of a structured coalescent model in the spirit of
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1206652/">Braverman et al. (1995)</a>.</p>
<p>In the <a class="reference internal" href="api.html#msprime.SweepGenicSelection" title="msprime.SweepGenicSelection"><code class="xref py py-class docutils literal notranslate"><span class="pre">SweepGenicSelection</span></code></a> model,
the population is split between two classes of selective backgrounds,
those linked to the beneficial allele, call it <span class="math notranslate nohighlight">\(B\)</span>, and those not,
<span class="math notranslate nohighlight">\(b\)</span>. In this model we track competing
rates of coalescence and recombination on the <span class="math notranslate nohighlight">\(B\)</span> and <span class="math notranslate nohighlight">\(b\)</span>
backgrounds.
The user supplies a final allele frequency and a starting
allele frequency, between which msprime simulates a stochastic
sweep trajectory according to a conditional diffusion model <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/15465123/">(Coop
and Griffiths, 2004</a>;
<a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/11861577/">Kim and Stephan 2002</a>).</p>
<p>Beyond the start and end frequencies of the sweep trajectory, the user
must specify the selection coefficient of the beneficial mutation
<span class="math notranslate nohighlight">\(s\)</span> the selective
advantage of the <span class="math notranslate nohighlight">\(B\)</span> homozygote over the <span class="math notranslate nohighlight">\(b\)</span> homozygote
and <span class="math notranslate nohighlight">\(h=0.5\)</span>. The position represents the location along
the chromosome where the beneficial allele occurs.
All other parameters can be set as usual.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The implementation of the structured coalescent is quite general, but there are
some limitations in the current implementation of the sweeps model (e.g., no
change of population size during a sweep). Please let us know if there are
related features you would like to see (or if you would be interested in
helping to create this functionality!)</p>
</div>
<section id="hard-sweeps">
<h4>Hard sweeps<a class="headerlink" href="#hard-sweeps" title="Link to this heading">#</a></h4>
<div class="admonition-todo admonition" id="id2">
<p class="admonition-title">Todo</p>
<p>What’s a hard sweep? Some intro sentences here.</p>
</div>
<p>In this example we run some replicate hard sweep simulations
and plot the mean
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.diversity" title="(in Project name not set)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pairwise</span> <span class="pre">diversity</span></code></a> in windows
across the simulated region, showing the characteristic
valley of diversity.</p>
<p>First we set up some basic parameters, and define the sweep model:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Ne</span> <span class="o">=</span> <span class="mf">1e3</span>
<span class="n">L</span> <span class="o">=</span> <span class="mf">1e6</span>  <span class="c1"># Length of simulated region</span>
<span class="n">num_reps</span> <span class="o">=</span> <span class="mi">100</span>

<span class="c1"># define hard sweep model</span>
<span class="n">sweep_model</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">SweepGenicSelection</span><span class="p">(</span>
    <span class="n">position</span><span class="o">=</span><span class="n">L</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span>  <span class="c1"># middle of chrom</span>
    <span class="n">start_frequency</span><span class="o">=</span><span class="mf">1.0</span> <span class="o">/</span> <span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">Ne</span><span class="p">),</span>
    <span class="n">end_frequency</span><span class="o">=</span><span class="mf">1.0</span> <span class="o">-</span> <span class="p">(</span><span class="mf">1.0</span> <span class="o">/</span> <span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">Ne</span><span class="p">)),</span>
    <span class="n">s</span><span class="o">=</span><span class="mf">0.25</span><span class="p">,</span>
    <span class="n">dt</span><span class="o">=</span><span class="mf">1e-6</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="admonition-todo admonition" id="id3">
<p class="admonition-title">Todo</p>
<p>Explain why these parameters give a hard sweep.</p>
</div>
<p>Next we set up the replicate simulations. As described
in the <a class="reference internal" href="#sec-ancestry-models-specifying-multiple"><span class="std std-ref">Multiple models</span></a> section,
the <code class="docutils literal notranslate"><span class="pre">model</span></code> parameter is a list when we want to run multiple
ancestry models. In this example, we have a sweep that occurs
in the immediate past, and is then followed by the
standard coalescent for the rest of time:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">reps</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">5</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="p">[</span><span class="n">sweep_model</span><span class="p">,</span> <span class="n">msprime</span><span class="o">.</span><span class="n">StandardCoalescent</span><span class="p">()],</span>
    <span class="n">population_size</span><span class="o">=</span><span class="n">Ne</span><span class="p">,</span>
    <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">1e-7</span><span class="p">,</span>
    <span class="n">sequence_length</span><span class="o">=</span><span class="n">L</span><span class="p">,</span>
    <span class="n">num_replicates</span><span class="o">=</span><span class="n">num_reps</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Because the <a class="reference internal" href="api.html#msprime.SweepGenicSelection" title="msprime.SweepGenicSelection"><code class="xref py py-class docutils literal notranslate"><span class="pre">SweepGenicSelection</span></code></a> model has a random
<a class="reference internal" href="#sec-ancestry-models-specifying-duration"><span class="std std-ref">duration</span></a> we do
not set this value in the model. Please see the
<a class="reference internal" href="#sec-ancestry-models-specifying-completion"><span class="std std-ref">Model completion</span></a> section for
more discussion on this important point.</p>
</div>
<p>Once we’ve set up the replicate simulations we can compute the
windows for plotting, run the actual simulations
(see the <a class="reference internal" href="replication.html#sec-randomness-replication"><span class="std std-ref">Running replicate simulations</span></a> section for more details)
and compute the
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.diversity" title="(in Project name not set)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pairwise</span> <span class="pre">diversity</span></code></a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">wins</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">L</span><span class="p">,</span> <span class="mi">21</span><span class="p">)</span>
<span class="n">mids</span> <span class="o">=</span> <span class="p">(</span><span class="n">wins</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span> <span class="o">+</span> <span class="n">wins</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span>
<span class="n">diversity</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">num_reps</span><span class="p">,</span> <span class="n">mids</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
<span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">ts</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">reps</span><span class="p">):</span>
    <span class="n">diversity</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">diversity</span><span class="p">(</span><span class="n">windows</span><span class="o">=</span><span class="n">wins</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;branch&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Finally, we can plot the observed mean diversity across the replicates
and compare it to the neutral expectation:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">matplotlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">pyplot</span> <span class="k">as</span> <span class="n">plt</span>

<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">mids</span><span class="p">,</span> <span class="n">diversity</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Simulations&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="mi">4</span> <span class="o">*</span> <span class="n">Ne</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;:&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">r</span><span class="s1">&#39;Neutral expectation&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Branch $\pi$&#39;</span><span class="p">);</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Position (bp)&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">();</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/46b1c97aecd85949d8073187323444f46be38a0846cfb51e45e24c9abba959b7.svg" src="_images/46b1c97aecd85949d8073187323444f46be38a0846cfb51e45e24c9abba959b7.svg" />
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>We use the “branch” measure of pairwise diversity which is
defined in terms of the trees rather than nucleotide sequences. See
<a class="reference external" href="https://doi.org/10.1534/genetics.120.303253">Ralph et al. 2020</a>
for more information.</p>
</div>
<p>As we can see, the selective sweep has reduced variation in the region
most closely linked to the beneficial allele and then diversity
increases with distance to each side.</p>
</section>
<section id="multiple-sweeps">
<span id="sec-ancestry-models-selective-sweeps-multiple"></span><h4>Multiple sweeps<a class="headerlink" href="#multiple-sweeps" title="Link to this heading">#</a></h4>
<div class="admonition-todo admonition" id="id4">
<p class="admonition-title">Todo</p>
<p>This example is very artificial; it would be much better to illustrate
with a meaningful example from the literature. Pull requests welcome!</p>
</div>
<p>To illustrate the idea that we can simulate large numbers of different
models, in this example we create a simulation in which we have
sweeps happening at random points along the genome, which
are separated by a random duration of the standard coalescent.</p>
<p>First we set up our models:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">random</span>
<span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">1234</span><span class="p">)</span>
<span class="n">L</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">Ne</span> <span class="o">=</span> <span class="mf">1e6</span>

<span class="n">models</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">100</span><span class="p">):</span>
    <span class="n">models</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">msprime</span><span class="o">.</span><span class="n">StandardCoalescent</span><span class="p">(</span><span class="n">duration</span><span class="o">=</span><span class="n">random</span><span class="o">.</span><span class="n">uniform</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)))</span>
    <span class="n">models</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
        <span class="n">msprime</span><span class="o">.</span><span class="n">SweepGenicSelection</span><span class="p">(</span>
            <span class="n">position</span><span class="o">=</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">L</span> <span class="o">-</span> <span class="mi">1</span><span class="p">),</span>
            <span class="n">start_frequency</span><span class="o">=</span><span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">Ne</span><span class="p">),</span>
            <span class="n">end_frequency</span><span class="o">=</span><span class="mf">1.0</span> <span class="o">-</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">Ne</span><span class="p">),</span>
            <span class="n">s</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span>
            <span class="n">dt</span><span class="o">=</span><span class="mf">1e-6</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="p">)</span>
<span class="n">models</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">msprime</span><span class="o">.</span><span class="n">StandardCoalescent</span><span class="p">())</span>
<span class="n">models</span><span class="p">[:</span><span class="mi">5</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[StandardCoalescent(duration=0.9664535356921388),
 SweepGenicSelection(duration=None, position=57, start_frequency=5e-07, end_frequency=0.9999995, s=0.01, dt=1e-06),
 StandardCoalescent(duration=0.11685051774599753),
 SweepGenicSelection(duration=None, position=12, start_frequency=5e-07, end_frequency=0.9999995, s=0.01, dt=1e-06),
 StandardCoalescent(duration=0.9109759624491242)]
</pre></div>
</div>
</div>
</div>
<p>We’ve created a list of models that have random parameters using the
standard Python tools. There are lots of sweeps (many more than we
will actually run, in all probability), each at a random position
on the genome and each separated by a random amount of time running
the standard coalescent. Finally, we add an instance of the standard
coalescent without a <a class="reference internal" href="#sec-ancestry-models-specifying-duration"><span class="std std-ref">duration</span></a>
to guarantee that our simulation will always coalesce fully
(See the <a class="reference internal" href="#sec-ancestry-models-specifying-completion"><span class="std std-ref">Model completion</span></a> section
for more information.) Finally, we print out the first 5 models
so we can inspect them.</p>
<p>Then, we can run our simulation as usual:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="mi">10</span><span class="p">,</span>
    <span class="n">population_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
    <span class="n">sequence_length</span><span class="o">=</span><span class="n">L</span><span class="p">,</span>
    <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="n">models</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">6789</span>
<span class="p">)</span>
<span class="n">ts</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
              <style>
                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}
                .tskit-table tbody tr td {padding: 0.5em 0.5em;}
                .tskit-table tbody tr td:first-of-type {text-align: left;}
                .tskit-details-label {vertical-align: top; padding-right:5px;}
                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}
                .tskit-table-set-table {margin: 12px 0 0 12px;}
                details {display: inline-block;}
                summary {cursor: pointer; outline: 0; display: list-item;}
              </style>
              <div class="tskit-table-set">
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="padding:0;line-height:21px;">
                          <img style="height: 32px;display: inline-block;padding: 3px 5px 3px 0;" src="https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg"/>
                          <a target="_blank" href="https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class"> Tree Sequence </a>
                        </th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr><td>Trees</td><td>100</td></tr>
                      <tr><td>Sequence Length</td><td>100.0</td></tr>
                      <tr><td>Time Units</td><td>generations</td></tr>
                      <tr><td>Sample Nodes</td><td>20</td></tr>
                      <tr><td>Total Size</td><td>102.7 KiB</td></tr>
                      <tr>
                        <td>Metadata</td><td style="text-align: left;">No Metadata</td></tr>
                    </tbody>
                  </table>
                </div>
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="line-height:21px;">Table</th>
                        <th>Rows</th>
                        <th>Size</th>
                        <th>Has Metadata</th>
                      </tr>
                    </thead>
                    <tbody>
                    
                  <tr>
                    <td>Edges</td>
                      <td>1588</td>
                      <td>49.6 KiB</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Individuals</td>
                      <td>10</td>
                      <td>304 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Migrations</td>
                      <td>0</td>
                      <td>8 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Mutations</td>
                      <td>0</td>
                      <td>16 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Nodes</td>
                      <td>523</td>
                      <td>14.3 KiB</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Populations</td>
                      <td>1</td>
                      <td>224 Bytes</td>
                      <td style="text-align: center;">
                        ✅
                      </td>
                    </tr>
                
                  <tr>
                    <td>Provenances</td>
                      <td>1</td>
                      <td>25.8 KiB</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Sites</td>
                      <td>0</td>
                      <td>16 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
            </div></div>
</div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>The logging output of msprime can be very useful when working with
multiple models. See the <a class="reference internal" href="logging.html#sec-logging"><span class="std std-ref">Logging</span></a> section for more
details.</p>
</div>
</section>
</section>
<section id="fixed-pedigree">
<span id="sec-ancestry-models-fixed-pedigree"></span><h3>Fixed pedigree<a class="headerlink" href="#fixed-pedigree" title="Link to this heading">#</a></h3>
<p>A pedigree describes parent-offspring relationships between individuals, and
these relationships constrain how genomes are inherited. We can simulate
ancestry conditioned on a fixed pedigree using the <a class="reference internal" href="api.html#msprime.FixedPedigree" title="msprime.FixedPedigree"><code class="xref py py-class docutils literal notranslate"><span class="pre">FixedPedigree</span></code></a>
ancestry model. In this model the transfer of genetic material from parent to
offspring is determined by the pedigree. Recombination occurs randomly such that
each of a diploid individual’s two genomes is a mixture of two corresponding
parental genomes, but everything else is determined by the pedigree.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>It is important to note that there is no concept of a population genetic
model “outside” of the pedigree in this model, and there
are significant caveats when dealing with incomplete pedigree information
and the treatment of founders.
Please see the <a class="reference internal" href="#sec-ancestry-models-fixed-pedigree-completing"><span class="std std-ref">Completing fixed pedigree simulations</span></a>
section for more information.</p>
</div>
<section id="simple-example">
<h4>Simple example<a class="headerlink" href="#simple-example" title="Link to this heading">#</a></h4>
<p>First we define a small pedigree, which follows a sample of three
individuals through four generations:</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>See the <a class="reference internal" href="pedigrees.html#sec-pedigrees"><span class="std std-ref">Pedigrees</span></a> section for details on this pedigree
text format and information on how pedigrees are encoded in msprime
via the individual and node tables.</p>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">io</span>
<span class="n">ped_txt</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span><span class="se">\</span>
<span class="s2"># id parent0 parent1 time is_sample</span>
<span class="s2">0   5   4   0.0 1</span>
<span class="s2">1   3   3   0.0 1</span>
<span class="s2">2   3   4   0.0 1</span>
<span class="s2">3   7   7   1.0 0</span>
<span class="s2">4   8   8   1.0 0</span>
<span class="s2">5   8   6   1.0 0</span>
<span class="s2">6   9   10  2.0 0</span>
<span class="s2">7   10  10  2.0 0</span>
<span class="s2">8   9   9   2.0 0</span>
<span class="s2">9   11  12  3.0 0</span>
<span class="s2">10  11  12  3.0 0</span>
<span class="s2">11  .   .   4.0 0</span>
<span class="s2">12  .   .   4.0 0</span>
<span class="s2">&quot;&quot;&quot;</span>
<span class="n">pedigree</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">parse_pedigree</span><span class="p">(</span><span class="n">io</span><span class="o">.</span><span class="n">StringIO</span><span class="p">(</span><span class="n">ped_txt</span><span class="p">),</span> <span class="n">sequence_length</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>

<span class="n">draw_pedigree</span><span class="p">(</span><span class="n">pedigree</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/e5dcd1f5ad29b4f370817eed2e382edc430d98dec17bbf022a3db401fa82a804.svg" src="_images/e5dcd1f5ad29b4f370817eed2e382edc430d98dec17bbf022a3db401fa82a804.svg" />
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>See the <a class="reference internal" href="pedigrees.html#sec-pedigrees-visualisation"><span class="std std-ref">Visualising pedigrees</span></a> section for a
definition of the <code class="docutils literal notranslate"><span class="pre">draw_pedigree</span></code> function used here.</p>
</div>
<p>We then use this pedigree information as the
<a class="reference internal" href="#sec-ancestry-initial-state"><span class="std std-ref">initial state</span></a> for the simulation
(note that we set the sequence length on the pedigree tables
when we called <a class="reference internal" href="api.html#msprime.parse_pedigree" title="msprime.parse_pedigree"><code class="xref py py-func docutils literal notranslate"><span class="pre">parse_pedigree()</span></code></a>):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ped_ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">initial_state</span><span class="o">=</span><span class="n">pedigree</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s2">&quot;fixed_pedigree&quot;</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">41</span><span class="p">,</span>
    <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">0.001</span><span class="p">)</span>
<span class="n">node_labels</span> <span class="o">=</span> <span class="p">{</span><span class="n">node</span><span class="o">.</span><span class="n">id</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">node</span><span class="o">.</span><span class="n">individual</span><span class="si">}</span><span class="s2">(</span><span class="si">{</span><span class="n">node</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">)&quot;</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">ped_ts</span><span class="o">.</span><span class="n">nodes</span><span class="p">()}</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ped_ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="n">node_labels</span><span class="o">=</span><span class="n">node_labels</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">600</span><span class="p">,</span><span class="mi">200</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/333e8540bdf715f615459642765add579328ab962edcc053d5fc563c601705bf.svg" src="_images/333e8540bdf715f615459642765add579328ab962edcc053d5fc563c601705bf.svg" />
</div>
</div>
<p>The output tree sequence contains the simulated genetic ancestry conditioned
on this input pedigree, as shown in the trees above. Each node is labelled
with it the individual ID it’s associated with, and its ID (as individual(node)).
The simulation stops at time 4, and we are left with multiple
roots, which correspond to the founder individuals. If you only
need simulations of the genetic ancestry through the pedigree,
then these trees can be used perfectly well in downstream
applications (such as <a class="reference internal" href="mutations.html#sec-mutations"><span class="std std-ref">Mutation simulations</span></a>).
See the <a class="reference internal" href="#sec-ancestry-models-fixed-pedigree-completing"><span class="std std-ref">Completing fixed pedigree simulations</span></a>
section for information (and caveats about) completing these simulations.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Because the input pedigree fully describes the simulation many features such as
<a class="reference internal" href="demography.html#sec-demography"><span class="std std-ref">demography</span></a> are not available when we use the
<a class="reference internal" href="api.html#msprime.FixedPedigree" title="msprime.FixedPedigree"><code class="xref py py-class docutils literal notranslate"><span class="pre">FixedPedigree</span></code></a> model. The <a class="reference internal" href="api.html#msprime.FixedPedigree" title="msprime.FixedPedigree"><code class="xref py py-class docutils literal notranslate"><span class="pre">FixedPedigree</span></code></a> model also cannot be
combined with other <a class="reference internal" href="#sec-ancestry-models"><span class="std std-ref">ancestry models</span></a>.</p>
</div>
</section>
<section id="censoring-pedigree-information">
<h4>Censoring pedigree information<a class="headerlink" href="#censoring-pedigree-information" title="Link to this heading">#</a></h4>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The output tree sequence also contains the full pedigree information
provided as input. This may be important if you are performing
simulations on pedigrees that are not freely available.</p>
</div>
<p>It is straightforward to remove the pedigree information from an
output tree sequence, if required:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tables</span> <span class="o">=</span> <span class="n">ped_ts</span><span class="o">.</span><span class="n">dump_tables</span><span class="p">()</span>
<span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>
<span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">individual</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full_like</span><span class="p">(</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">individual</span><span class="p">,</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span><span class="p">)</span>
<span class="n">censored_ts</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">()</span>
<span class="n">censored_ts</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
              <style>
                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}
                .tskit-table tbody tr td {padding: 0.5em 0.5em;}
                .tskit-table tbody tr td:first-of-type {text-align: left;}
                .tskit-details-label {vertical-align: top; padding-right:5px;}
                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}
                .tskit-table-set-table {margin: 12px 0 0 12px;}
                details {display: inline-block;}
                summary {cursor: pointer; outline: 0; display: list-item;}
              </style>
              <div class="tskit-table-set">
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="padding:0;line-height:21px;">
                          <img style="height: 32px;display: inline-block;padding: 3px 5px 3px 0;" src="https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg"/>
                          <a target="_blank" href="https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class"> Tree Sequence </a>
                        </th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr><td>Trees</td><td>2</td></tr>
                      <tr><td>Sequence Length</td><td>100.0</td></tr>
                      <tr><td>Time Units</td><td>generations</td></tr>
                      <tr><td>Sample Nodes</td><td>6</td></tr>
                      <tr><td>Total Size</td><td>2.6 KiB</td></tr>
                      <tr>
                        <td>Metadata</td><td style="text-align: left;">No Metadata</td></tr>
                    </tbody>
                  </table>
                </div>
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="line-height:21px;">Table</th>
                        <th>Rows</th>
                        <th>Size</th>
                        <th>Has Metadata</th>
                      </tr>
                    </thead>
                    <tbody>
                    
                  <tr>
                    <td>Edges</td>
                      <td>13</td>
                      <td>424 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Individuals</td>
                      <td>0</td>
                      <td>40 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Migrations</td>
                      <td>0</td>
                      <td>8 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Mutations</td>
                      <td>0</td>
                      <td>16 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Nodes</td>
                      <td>26</td>
                      <td>736 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Populations</td>
                      <td>1</td>
                      <td>224 Bytes</td>
                      <td style="text-align: center;">
                        ✅
                      </td>
                    </tr>
                
                  <tr>
                    <td>Provenances</td>
                      <td>1</td>
                      <td>1.0 KiB</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Sites</td>
                      <td>0</td>
                      <td>16 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
            </div></div>
</div>
</section>
<section id="completing-fixed-pedigree-simulations">
<span id="sec-ancestry-models-fixed-pedigree-completing"></span><h4>Completing fixed pedigree simulations<a class="headerlink" href="#completing-fixed-pedigree-simulations" title="Link to this heading">#</a></h4>
<p>If you wish to simulate ancestry older than what is defined in the
pedigree, you can “complete” (or “recapitate”) the within-pedigree
simulation using the standard methods described in the
<a class="reference internal" href="#sec-ancestry-initial-state"><span class="std std-ref">Specifying the initial state</span></a> section.</p>
<p>Continuing the example in the previous section, we provide the simulated
ancestry in <code class="docutils literal notranslate"><span class="pre">ped_ts</span></code> as the initial state for a <a class="reference internal" href="api.html#msprime.DiscreteTimeWrightFisher" title="msprime.DiscreteTimeWrightFisher"><code class="xref py py-class docutils literal notranslate"><span class="pre">DiscreteTimeWrightFisher</span></code></a>
simulation with a population size of 10:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">final_ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">initial_state</span><span class="o">=</span><span class="n">ped_ts</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s2">&quot;dtwf&quot;</span><span class="p">,</span> <span class="n">population_size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">final_ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">600</span><span class="p">,</span><span class="mi">200</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/9beb59d45a2f5c6ce88f0bb2486f15eaeb270809cd35627d88f8894778ffc533.svg" src="_images/9beb59d45a2f5c6ce88f0bb2486f15eaeb270809cd35627d88f8894778ffc533.svg" />
</div>
</div>
<p>We can see that the nodes 22 and 23 representing the founder genomes
from the input pedigree are used as the starting point for the DTWF simulation,
which then finishes at generation 6 when the MRCA (node 26) is found.</p>
</section>
<section id="missing-pedigree-data">
<span id="sec-ancestry-models-fixed-pedigree-missing-data"></span><h4>Missing pedigree data<a class="headerlink" href="#missing-pedigree-data" title="Link to this heading">#</a></h4>
<p>The previous example is very straightforward because the pedigree is
complete and we can trace all individuals back to the same
generation in the pedigree. The pedigrees we have in practice
are usually not like this, and there is a great deal of missing data.
The consequence of this for the <a class="reference internal" href="api.html#msprime.FixedPedigree" title="msprime.FixedPedigree"><code class="xref py py-class docutils literal notranslate"><span class="pre">FixedPedigree</span></code></a> simulations
is that genetic material will often be at a “dead end” at many
levels of the pedigree during a simulation.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ped_txt</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span><span class="se">\</span>
<span class="s2"># id parent0 parent1 time is_sample</span>
<span class="s2">0   5   4   0.0 1</span>
<span class="s2">1   3   3   0.0 1</span>
<span class="s2">2   3   4   0.0 1</span>
<span class="s2">3   7   7   1.0 0</span>
<span class="s2">4   8   8   1.0 0</span>
<span class="s2">5   8   6   1.0 0</span>
<span class="s2">6   9   10  2.0 0</span>
<span class="s2">7   .   .   2.0 0</span>
<span class="s2">8   9   9   2.0 0</span>
<span class="s2">9   11  12  3.0 0</span>
<span class="s2">10  11  12  3.0 0</span>
<span class="s2">11  .   .   4.0 0</span>
<span class="s2">12  .   .   4.0 0</span>
<span class="s2">&quot;&quot;&quot;</span>
<span class="n">pedigree</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">parse_pedigree</span><span class="p">(</span><span class="n">io</span><span class="o">.</span><span class="n">StringIO</span><span class="p">(</span><span class="n">ped_txt</span><span class="p">),</span> <span class="n">sequence_length</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>

<span class="n">draw_pedigree</span><span class="p">(</span><span class="n">pedigree</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/c7a1fea7b6311369d056bb6a42235f56cd233e0d0709911c52dbc769b4216e23.svg" src="_images/c7a1fea7b6311369d056bb6a42235f56cd233e0d0709911c52dbc769b4216e23.svg" />
</div>
</div>
<p>This pedigree has a “dead end” for individual 7, which is
reflected in the trees that we output from the <a class="reference internal" href="api.html#msprime.FixedPedigree" title="msprime.FixedPedigree"><code class="xref py py-class docutils literal notranslate"><span class="pre">FixedPedigree</span></code></a> simulation:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ped_ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">initial_state</span><span class="o">=</span><span class="n">pedigree</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s2">&quot;fixed_pedigree&quot;</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">41</span><span class="p">,</span>
    <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">0.001</span><span class="p">)</span>
<span class="n">node_labels</span> <span class="o">=</span> <span class="p">{</span><span class="n">node</span><span class="o">.</span><span class="n">id</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">node</span><span class="o">.</span><span class="n">individual</span><span class="si">}</span><span class="s2">(</span><span class="si">{</span><span class="n">node</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">)&quot;</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">ped_ts</span><span class="o">.</span><span class="n">nodes</span><span class="p">()}</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ped_ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="n">node_labels</span><span class="o">=</span><span class="n">node_labels</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">600</span><span class="p">,</span><span class="mi">200</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/7bdd5525233b4c744556f2db31d59cacec2d365bb735305044a20b58100a2277.svg" src="_images/7bdd5525233b4c744556f2db31d59cacec2d365bb735305044a20b58100a2277.svg" />
</div>
</div>
<p>When we provide this tree sequence as the
<a class="reference internal" href="#sec-ancestry-initial-state"><span class="std std-ref">initial state</span></a> of another simulation, we
then effectively start the simulation at the earlier time point corresponding to
individual 7. Ancestral material in older parts of the pedigree is then
brought into the simulation in the same manner as
<a class="reference internal" href="#sec-ancestry-samples-sampling-time"><span class="std std-ref">ancient samples</span></a> as we proceed backwards
in time.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>It is important to note that although ancestral material for pedigree
founders is introduced into the recapitation simulation at the correct time,
no account is taken of the number of lineages present in the pedigree
when calculating population sizes. Thus, the pedigree must be seen as
entirely <strong>external</strong> to the population model simulated during recapitation.</p>
</div>
</section>
<section id="pedigrees-and-demography">
<span id="sec-ancestry-models-fixed-pedigree-demography"></span><h4>Pedigrees and demography<a class="headerlink" href="#pedigrees-and-demography" title="Link to this heading">#</a></h4>
<p>For complex simulations in which our pedigree individuals (and
founders, in particular) are drawn from different populations,
we will want our follow-up recapitation simulations to use
a <a class="reference internal" href="demography.html#sec-demography"><span class="std std-ref">demographic model</span></a>. In order to do this,
we must define the demographic model <em>before</em> creating
the input pedigree. For example, consider the following
simple demography:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population_split</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">],</span> <span class="n">ancestral</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">);</span>
</pre></div>
</div>
</div>
</div>
<p>We define a pedigree in which the individuals belong
to the two “leaf” populations, A and B:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ped_txt</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span><span class="se">\</span>
<span class="s2"># id    parent0 parent1 time    is_sample   population</span>
<span class="s2">0   9   11  0.0 1   A</span>
<span class="s2">1   10  10  0.0 1   A</span>
<span class="s2">2   6   7   0.0 1   A</span>
<span class="s2">3   7   7   0.0 1   A</span>
<span class="s2">4   8   9   0.0 1   A</span>
<span class="s2">5   10  10  0.0 1   A</span>
<span class="s2">6   14  14  1.0 0   A</span>
<span class="s2">7   12  12  1.0 0   A</span>
<span class="s2">8   16  17  1.0 0   A</span>
<span class="s2">9   12  13  1.0 0   A</span>
<span class="s2">10  14  12  1.0 0   A</span>
<span class="s2">11  15  16  1.0 0   A</span>
<span class="s2">12  .   .   2.0 0   A</span>
<span class="s2">13  .   .   2.0 0   A</span>
<span class="s2">14  .   .   2.0 0   A</span>
<span class="s2">15  .   .   2.0 0   A</span>
<span class="s2">16  .   .   2.0 0   A</span>
<span class="s2">17  .   .   2.0 0   A</span>
<span class="s2">18  23  22  0.0 1   B</span>
<span class="s2">19  22  22  0.0 1   B</span>
<span class="s2">20  21  22  0.0 1   B</span>
<span class="s2">21  25  25  1.0 0   B</span>
<span class="s2">22  24  25  1.0 0   B</span>
<span class="s2">23  24  24  1.0 0   B</span>
<span class="s2">24  27  26  2.0 0   B</span>
<span class="s2">25  27  26  2.0 0   B</span>
<span class="s2">26  30  28  3.0 0   B</span>
<span class="s2">27  30  29  3.0 0   B</span>
<span class="s2">28  31  32  4.0 0   B</span>
<span class="s2">29  31  32  4.0 0   B</span>
<span class="s2">30  32  31  4.0 0   B</span>
<span class="s2">31  .   .   5.0 0   B</span>
<span class="s2">32  .   .   5.0 0   B</span>
<span class="s2">&quot;&quot;&quot;</span>
<span class="n">pedigree</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">parse_pedigree</span><span class="p">(</span>
    <span class="n">io</span><span class="o">.</span><span class="n">StringIO</span><span class="p">(</span><span class="n">ped_txt</span><span class="p">),</span> <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">,</span> <span class="n">sequence_length</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>

<span class="n">draw_pedigree</span><span class="p">(</span><span class="n">pedigree</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/b779191a9470ae5d60836e815ab7320350c705894102e723114118a4d9d202d4.svg" src="_images/b779191a9470ae5d60836e815ab7320350c705894102e723114118a4d9d202d4.svg" />
</div>
</div>
<p>We can see that the pedigree is split into two sub-pedigrees for the
two populations, and the founders for the two populations exist
at different times (2 and 5 generations ago for A and B, respectively).
Running the <a class="reference internal" href="api.html#msprime.FixedPedigree" title="msprime.FixedPedigree"><code class="xref py py-class docutils literal notranslate"><span class="pre">FixedPedigree</span></code></a> simulation, we can see that the
trees reflect this structure:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ped_ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">initial_state</span><span class="o">=</span><span class="n">pedigree</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s2">&quot;fixed_pedigree&quot;</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">41</span><span class="p">)</span>
<span class="n">node_labels</span> <span class="o">=</span> <span class="p">{</span><span class="n">node</span><span class="o">.</span><span class="n">id</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">node</span><span class="o">.</span><span class="n">individual</span><span class="si">}</span><span class="s2">(</span><span class="si">{</span><span class="n">node</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">)&quot;</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">ped_ts</span><span class="o">.</span><span class="n">nodes</span><span class="p">()}</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ped_ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="n">node_labels</span><span class="o">=</span><span class="n">node_labels</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">600</span><span class="p">,</span><span class="mi">200</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/1278265a496ec4a4a5f75488ecd354f2e29b716495f47ed99a1597312e10c988.svg" src="_images/1278265a496ec4a4a5f75488ecd354f2e29b716495f47ed99a1597312e10c988.svg" />
</div>
</div>
<p>When we complete this simulation we must use a <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> object
that matches the populations defined in the original pedigree tables. The
simplest way to do this is define the demography first, and use the same
object throughout as we have done here:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">full_ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">initial_state</span><span class="o">=</span><span class="n">ped_ts</span><span class="p">,</span> <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s2">&quot;dtwf&quot;</span><span class="p">)</span>
<span class="n">node_labels</span> <span class="o">=</span> <span class="p">{</span>
    <span class="n">node</span><span class="o">.</span><span class="n">id</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">node</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">full_ts</span><span class="o">.</span><span class="n">population</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">population</span><span class="p">)</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">full_ts</span><span class="o">.</span><span class="n">nodes</span><span class="p">()}</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">full_ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span>
    <span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="n">node_labels</span><span class="o">=</span><span class="n">node_labels</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">600</span><span class="p">,</span><span class="mi">300</span><span class="p">),</span> <span class="n">time_scale</span><span class="o">=</span><span class="s2">&quot;rank&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/f6e12f1096add1cdf197eb2c7957fc20b625f00a5a3ef3e4fccadbf4e9bb1776.svg" src="_images/f6e12f1096add1cdf197eb2c7957fc20b625f00a5a3ef3e4fccadbf4e9bb1776.svg" />
</div>
</div>
<p>Here we have changed the node labelling in the trees to show the name of the
population of each node. We can see that all of the early nodes are in the
populations defined from the pedigree. When the recapitation simulation begins,
all nodes are assigned into the relevant populations and the simulation proceeds
from there under the defined demography. In this demographic model, populations
A and B split from C 10 generations ago, and so the remaining lineages in populations
A and B evolve independently until this time. We can see that older coalescences
all happen within population C, as stipulated by the demographic model.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The same caveats that we noted
<a class="reference internal" href="#sec-ancestry-models-fixed-pedigree-missing-data"><span class="std std-ref">above</span></a> about missing
data in the pedigree apply: although the remaining ancestral lineages
will join the simulation at the correct times and in the
specified population, <em>no account</em> is taken
of the number of extant lineages from the pedigree when computing
population sizes.</p>
</div>
</section>
</section>
</section>
<section id="missing-data">
<span id="sec-ancestry-missing-data"></span><h2>Missing data<a class="headerlink" href="#missing-data" title="Link to this heading">#</a></h2>
<div class="admonition-todo admonition" id="id5">
<p class="admonition-title">Todo</p>
<p>Give an example of doing a simulation with unknown rates on the flanks
and show that the resulting tree sequence has missing data in these
regions.</p>
</div>
</section>
<section id="common-errors">
<span id="sec-ancestry-errors"></span><h2>Common errors<a class="headerlink" href="#common-errors" title="Link to this heading">#</a></h2>
<section id="infinite-waiting-time">
<span id="sec-ancestry-errors-infinite-waiting-time"></span><h3>Infinite waiting time<a class="headerlink" href="#infinite-waiting-time" title="Link to this heading">#</a></h3>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>explain this, why it happens and give examples of when
we don’t detect it. Mention the possible_lineage_locations method.</p>
</div>
<div class="cell tag_raises-exception docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="n">samples</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span> <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">LibraryError</span><span class="g g-Whitespace">                              </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">105</span><span class="p">],</span> <span class="n">line</span> <span class="mi">4</span>
<span class="g g-Whitespace">      </span><span class="mi">2</span> <span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="g g-Whitespace">      </span><span class="mi">3</span> <span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="ne">----&gt; </span><span class="mi">4</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="n">samples</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span> <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">)</span>

<span class="nn">File ~/work/tskit-site/tskit-site/msprime/ancestry.py:1311,</span> in <span class="ni">sim_ancestry</span><span class="nt">(samples, demography, sequence_length, discrete_genome, recombination_rate, gene_conversion_rate, gene_conversion_tract_length, population_size, ploidy, model, initial_state, start_time, end_time, record_migrations, record_full_arg, additional_nodes, coalescing_segments_only, num_labels, random_seed, num_replicates, replicate_index, record_provenance)</span>
<span class="g g-Whitespace">   </span><span class="mi">1289</span>     <span class="n">provenance_dict</span> <span class="o">=</span> <span class="n">provenance</span><span class="o">.</span><span class="n">get_provenance_dict</span><span class="p">(</span><span class="n">parameters</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1290</span> <span class="n">sim</span> <span class="o">=</span> <span class="n">_parse_sim_ancestry</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1291</span>     <span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1292</span>     <span class="n">sequence_length</span><span class="o">=</span><span class="n">sequence_length</span><span class="p">,</span>
   <span class="p">(</span><span class="o">...</span><span class="p">)</span>   <span class="mi">1309</span>     <span class="n">random_seed</span><span class="o">=</span><span class="n">random_seed</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1310</span> <span class="p">)</span>
<span class="ne">-&gt; </span><span class="mi">1311</span> <span class="k">return</span> <span class="n">_wrap_replicates</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1312</span>     <span class="n">sim</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1313</span>     <span class="n">num_replicates</span><span class="o">=</span><span class="n">num_replicates</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1314</span>     <span class="n">replicate_index</span><span class="o">=</span><span class="n">replicate_index</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1315</span>     <span class="n">provenance_dict</span><span class="o">=</span><span class="n">provenance_dict</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1316</span> <span class="p">)</span>

<span class="nn">File ~/work/tskit-site/tskit-site/msprime/ancestry.py:681,</span> in <span class="ni">_wrap_replicates</span><span class="nt">(simulator, num_replicates, replicate_index, provenance_dict, mutation_rate)</span>
<span class="g g-Whitespace">    </span><span class="mi">675</span> <span class="n">iterator</span> <span class="o">=</span> <span class="n">simulator</span><span class="o">.</span><span class="n">run_replicates</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">676</span>     <span class="n">num_replicates</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">677</span>     <span class="n">mutation_rate</span><span class="o">=</span><span class="n">mutation_rate</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">678</span>     <span class="n">provenance_dict</span><span class="o">=</span><span class="n">provenance_dict</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">679</span> <span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">680</span> <span class="k">if</span> <span class="n">replicate_index</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<span class="ne">--&gt; </span><span class="mi">681</span>     <span class="n">deque</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">deque</span><span class="p">(</span><span class="n">iterator</span><span class="p">,</span> <span class="n">maxlen</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">682</span>     <span class="k">return</span> <span class="n">deque</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
<span class="g g-Whitespace">    </span><span class="mi">683</span> <span class="k">else</span><span class="p">:</span>

<span class="nn">File ~/work/tskit-site/tskit-site/msprime/ancestry.py:1606,</span> in <span class="ni">Simulator.run_replicates</span><span class="nt">(self, num_replicates, mutation_rate, provenance_dict)</span>
<span class="g g-Whitespace">   </span><span class="mi">1604</span> <span class="k">for</span> <span class="n">replicate_index</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_replicates</span><span class="p">):</span>
<span class="g g-Whitespace">   </span><span class="mi">1605</span>     <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Starting replicate </span><span class="si">%d</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">replicate_index</span><span class="p">)</span>
<span class="ne">-&gt; </span><span class="mi">1606</span>     <span class="bp">self</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
<span class="g g-Whitespace">   </span><span class="mi">1607</span>     <span class="k">if</span> <span class="n">mutation_rate</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">1608</span>         <span class="c1"># This is only called from simulate() or the ms interface,</span>
<span class="g g-Whitespace">   </span><span class="mi">1609</span>         <span class="c1"># so does not need any further parameters.</span>
<span class="g g-Whitespace">   </span><span class="mi">1610</span>         <span class="n">mutations</span><span class="o">.</span><span class="n">_simple_mutate</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1611</span>             <span class="n">tables</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">tables</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1612</span>             <span class="n">random_generator</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">random_generator</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1613</span>             <span class="n">sequence_length</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1614</span>             <span class="n">rate</span><span class="o">=</span><span class="n">mutation_rate</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">1615</span>         <span class="p">)</span>

<span class="nn">File ~/work/tskit-site/tskit-site/msprime/ancestry.py:1571,</span> in <span class="ni">Simulator.run</span><span class="nt">(self, event_chunk, debug_func)</span>
<span class="g g-Whitespace">   </span><span class="mi">1569</span>     <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Model durations must be &gt;= 0&quot;</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1570</span> <span class="n">end_time</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">time</span> <span class="o">+</span> <span class="n">model_duration</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">end_time</span><span class="p">)</span>
<span class="ne">-&gt; </span><span class="mi">1571</span> <span class="n">exit_reason</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_run_until</span><span class="p">(</span><span class="n">end_time</span><span class="p">,</span> <span class="n">event_chunk</span><span class="p">,</span> <span class="n">debug_func</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1572</span> <span class="k">if</span> <span class="n">exit_reason</span> <span class="o">==</span> <span class="n">ExitReason</span><span class="o">.</span><span class="n">COALESCENCE</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">time</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">end_time</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">1573</span>     <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Skipping remaining </span><span class="si">%d</span><span class="s2"> models&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">models</span><span class="p">)</span> <span class="o">-</span> <span class="n">j</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>

<span class="nn">File ~/work/tskit-site/tskit-site/msprime/ancestry.py:1536,</span> in <span class="ni">Simulator._run_until</span><span class="nt">(self, end_time, event_chunk, debug_func)</span>
<span class="g g-Whitespace">   </span><span class="mi">1534</span> <span class="n">ret</span> <span class="o">=</span> <span class="n">ExitReason</span><span class="o">.</span><span class="n">MAX_EVENTS</span>
<span class="g g-Whitespace">   </span><span class="mi">1535</span> <span class="k">while</span> <span class="n">ret</span> <span class="o">==</span> <span class="n">ExitReason</span><span class="o">.</span><span class="n">MAX_EVENTS</span><span class="p">:</span>
<span class="ne">-&gt; </span><span class="mi">1536</span>     <span class="n">ret</span> <span class="o">=</span> <span class="n">ExitReason</span><span class="p">(</span><span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">end_time</span><span class="p">,</span> <span class="n">event_chunk</span><span class="p">))</span>
<span class="g g-Whitespace">   </span><span class="mi">1537</span>     <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">time</span> <span class="o">&gt;</span> <span class="n">end_time</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">1538</span>         <span class="c1"># Currently the Pedigree and Sweeps models are &quot;non-reentrant&quot;</span>
<span class="g g-Whitespace">   </span><span class="mi">1539</span>         <span class="c1"># We can change this to an assertion once these have been fixed.</span>
<span class="g g-Whitespace">   </span><span class="mi">1540</span>         <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">1541</span>             <span class="sa">f</span><span class="s2">&quot;Model </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> does not support interruption. &quot;</span>
<span class="g g-Whitespace">   </span><span class="mi">1542</span>             <span class="s2">&quot;Please open an issue on GitHub&quot;</span>
<span class="g g-Whitespace">   </span><span class="mi">1543</span>         <span class="p">)</span>

<span class="ne">LibraryError</span>: Infinite waiting time until next simulation event.
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="installation.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Installation</p>
      </div>
    </a>
    <a class="right-next"
       href="mutations.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Mutation simulations</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quick-reference">Quick reference</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#specifying-samples">Specifying samples</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#populations">Populations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sampling-time">Sampling time</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sample-details">Sample details</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#demography">Demography</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#ploidy">Ploidy</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#default-number-of-nodes-per-sample-individual">Default number of nodes per sample individual</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#coalescent-time-scales">Coalescent time scales</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#genome-properties">Genome properties</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sequence-length">Sequence length</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#recombination">Recombination</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-conversion">Gene conversion</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#discrete-or-continuous">Discrete or continuous?</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#multiple-chromosomes">Multiple chromosomes</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#do-i-need-to-do-this">Do I need to do this?</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#using-a-fixed-pedigree">Using a fixed pedigree</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#simulations-without-a-fixed-pedigree">Simulations without a fixed pedigree</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#example-1-simulating-with-a-pedigree">Example 1: simulating with a pedigree</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#example-2-simulating-without-a-pedigree">Example 2: simulating without a pedigree</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#recording-more-information">Recording more information</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#additional-nodes">Additional nodes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#coalescing-segments-only">Coalescing segments only</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#common-ancestor-events">Common ancestor events</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#recombination-events">Recombination events</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#migration-events">Migration events</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#census-events">Census events</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#using-the-dtwf-model">Using the DTWF model</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ancestral-recombination-graph">Ancestral recombination graph</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#manipulating-simulation-time">Manipulating simulation time</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#stopping-simulations-early">Stopping simulations early</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#setting-the-start-time">Setting the start time</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#order-of-event-execution">Order of event execution</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#specifying-the-initial-state">Specifying the initial state</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#combining-backward-time-simulations">Combining backward-time simulations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#interaction-with-demography">Interaction with demography</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#continuing-forwards-time-simulations-recapitating">Continuing forwards-time simulations (“recapitating”)</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#models">Models</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#specifying-ancestry-models">Specifying ancestry models</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#single-models">Single models</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#model-duration">Model duration</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#model-completion">Model completion</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#multiple-models">Multiple models</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#hudson-coalescent">Hudson coalescent</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#smc-approximations">SMC approximations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#discrete-time-wright-fisher">Discrete Time Wright-Fisher</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#multiple-merger-coalescents">Multiple merger coalescents</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#selective-sweeps">Selective sweeps</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#hard-sweeps">Hard sweeps</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#multiple-sweeps">Multiple sweeps</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#fixed-pedigree">Fixed pedigree</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#simple-example">Simple example</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#censoring-pedigree-information">Censoring pedigree information</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#completing-fixed-pedigree-simulations">Completing fixed pedigree simulations</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#missing-pedigree-data">Missing pedigree data</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pedigrees-and-demography">Pedigrees and demography</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#missing-data">Missing data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#common-errors">Common errors</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#infinite-waiting-time">Infinite waiting time</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Tskit Developers
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>