
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Demographic models &#8212; Msprime manual</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'demography';</script>
    <link rel="icon" href="_static/favicon.ico"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Randomness and replication" href="replication.html" />
    <link rel="prev" title="Mutation simulations" href="mutations.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <script data-goatcounter="https://tskit.goatcounter.com/count" async src="//gc.zgo.at/count.js"></script>
</head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/msprime_logo.svg" class="logo__image only-light" alt="Msprime manual - Home"/>
    <script>document.write(`<img src="_static/msprime_logo.svg" class="logo__image only-dark" alt="Msprime manual - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Running simulations</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="ancestry.html">Ancestry simulations</a></li>
<li class="toctree-l1"><a class="reference internal" href="mutations.html">Mutation simulations</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Demographic models</a></li>
<li class="toctree-l1"><a class="reference internal" href="replication.html">Randomness and replication</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Interfaces</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="api.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">Command line interface</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Utilities</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="rate_maps.html">Rate Maps</a></li>
<li class="toctree-l1"><a class="reference internal" href="pedigrees.html">Pedigrees</a></li>
<li class="toctree-l1"><a class="reference internal" href="likelihoods.html">Computing likelihoods</a></li>
<li class="toctree-l1"><a class="reference internal" href="logging.html">Logging</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Miscellaneous</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="legacy.html">Legacy (version 0.x) APIs</a></li>
<li class="toctree-l1"><a class="reference internal" href="switch_from_other_simulators.html">Switching from other simulators</a></li>
<li class="toctree-l1"><a class="reference internal" href="development.html">Development</a></li>
<li class="toctree-l1"><a class="reference internal" href="CITATION.html">Citing msprime</a></li>
<li class="toctree-l1"><a class="reference internal" href="CHANGELOG.html">Changelog</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/demography.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Demographic models</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quick-reference">Quick reference</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#model">Model</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#direction-of-time">Direction of time</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#definitions">Definitions</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#populations">Populations</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#sec-demography-populations-identifiers">Identifying populations</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#initial-size">Initial size</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#growth-rate">Growth rate</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#metadata">Metadata</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#life-cycle">Life cycle</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#default-sampling-time">Default sampling time</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#migration">Migration</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#events">Events</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#population-parameters-change">Population parameters change</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#migration-rate-change">Migration rate change</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#population-split">Population split</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#admixture">Admixture</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pulse-mass-migration">Pulse (mass) migration</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#instantaneous-bottleneck">Instantaneous bottleneck</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#simple-bottleneck">Simple bottleneck</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#demography-objects">Demography objects</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#debugging-tools">Debugging tools</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Demography objects</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#demographydebugger">DemographyDebugger</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-models">Example models</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#population-tree">Population tree</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sec-demography-examples-admixture">Admixture</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#trunk-population-models">Trunk population models</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#importing-exporting-model-definitions">Importing/exporting model definitions</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#demes">Demes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#species-trees">Species trees</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#numerical-utilities">Numerical utilities</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#population-sizes-over-time">Population sizes over time</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#possible-locations-of-lineages">Possible locations of lineages</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#lineage-probabilities">Lineage probabilities</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#coalescence-rates-and-mean-times">Coalescence rates and mean times</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#inverse-instantaneous-coalescence-rates">Inverse instantaneous coalescence rates</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="demographic-models">
<span id="sec-demography"></span><h1>Demographic models<a class="headerlink" href="#demographic-models" title="Link to this heading">#</a></h1>
<p>By default, msprime assumes a single randomly mating population of a fixed
size for <a class="reference internal" href="ancestry.html#sec-ancestry"><span class="std std-ref">ancestry simulations</span></a>, which is unrealistic
for most purposes. To enable more realistic and complex simulations,
msprime models population structure
by defining a set of discrete <a class="reference internal" href="#sec-demography-populations"><span class="std std-ref">Populations</span></a>,
with <a class="reference internal" href="#sec-demography-migration"><span class="std std-ref">Migration</span></a> between these populations occurring
at different rates. Populations and their parameters can
change over time, via <a class="reference internal" href="#sec-demography-events"><span class="std std-ref">demographic events</span></a>.
Please see the <a class="reference internal" href="#sec-demography-definitions"><span class="std std-ref">Definitions</span></a>
section for mathematical details.</p>
<p>The information required to define a demographic model is encapsulated
by the <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> object. To run
<a class="reference internal" href="ancestry.html#sec-ancestry"><span class="std std-ref">ancestry simulations</span></a> for a given demography
we use the <code class="docutils literal notranslate"><span class="pre">demography</span></code> parameter to <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a> and define
what populations our <a class="reference internal" href="ancestry.html#sec-ancestry-samples"><span class="std std-ref">samples</span></a> are drawn from.
For example, here we create a three-population model in which
populations <code class="docutils literal notranslate"><span class="pre">A</span></code> and <code class="docutils literal notranslate"><span class="pre">B</span></code> split from <code class="docutils literal notranslate"><span class="pre">C</span></code> 1000 generations ago.
We then run an ancestry simulation based on this demography;
here we draw one diploid sample each from <code class="docutils literal notranslate"><span class="pre">A</span></code> and <code class="docutils literal notranslate"><span class="pre">B</span></code> at time zero:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">10_000</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">5_000</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">1_000</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population_split</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">],</span> <span class="n">ancestral</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="n">samples</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span> <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ts</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
              <style>
                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}
                .tskit-table tbody tr td {padding: 0.5em 0.5em;}
                .tskit-table tbody tr td:first-of-type {text-align: left;}
                .tskit-details-label {vertical-align: top; padding-right:5px;}
                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}
                .tskit-table-set-table {margin: 12px 0 0 12px;}
                details {display: inline-block;}
                summary {cursor: pointer; outline: 0; display: list-item;}
              </style>
              <div class="tskit-table-set">
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="padding:0;line-height:21px;">
                          <img style="height: 32px;display: inline-block;padding: 3px 5px 3px 0;" src="https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg"/>
                          <a target="_blank" href="https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class"> Tree Sequence </a>
                        </th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr><td>Trees</td><td>1</td></tr>
                      <tr><td>Sequence Length</td><td>1.0</td></tr>
                      <tr><td>Time Units</td><td>generations</td></tr>
                      <tr><td>Sample Nodes</td><td>4</td></tr>
                      <tr><td>Total Size</td><td>2.5 KiB</td></tr>
                      <tr>
                        <td>Metadata</td><td style="text-align: left;">No Metadata</td></tr>
                    </tbody>
                  </table>
                </div>
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="line-height:21px;">Table</th>
                        <th>Rows</th>
                        <th>Size</th>
                        <th>Has Metadata</th>
                      </tr>
                    </thead>
                    <tbody>
                    
                  <tr>
                    <td>Edges</td>
                      <td>6</td>
                      <td>200 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Individuals</td>
                      <td>2</td>
                      <td>80 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Migrations</td>
                      <td>0</td>
                      <td>8 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Mutations</td>
                      <td>0</td>
                      <td>16 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Nodes</td>
                      <td>7</td>
                      <td>204 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Populations</td>
                      <td>3</td>
                      <td>294 Bytes</td>
                      <td style="text-align: center;">
                        ✅
                      </td>
                    </tr>
                
                  <tr>
                    <td>Provenances</td>
                      <td>1</td>
                      <td>1.7 KiB</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Sites</td>
                      <td>0</td>
                      <td>16 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
            </div></div>
</div>
<p>Please see the <a class="reference internal" href="#sec-demography-examples"><span class="std std-ref">Example models</span></a> section for
more detailed examples of models from the literature.</p>
<hr class="docutils" />
<section id="quick-reference">
<h2>Quick reference<a class="headerlink" href="#quick-reference" title="Link to this heading">#</a></h2>
<dl class="simple myst">
<dt><a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a></dt><dd><p>Description of a demographic model</p>
</dd>
</dl>
<p><strong>Building models</strong></p>
<dl class="simple myst">
<dt><a class="reference internal" href="api.html#msprime.Demography.add_population" title="msprime.Demography.add_population"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.add_population()</span></code></a></dt><dd><p>Add a population to the model</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.Demography.set_symmetric_migration_rate" title="msprime.Demography.set_symmetric_migration_rate"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.set_symmetric_migration_rate()</span></code></a></dt><dd><p>Sets the symmetric migration rates among a set of populations</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.Demography.set_migration_rate" title="msprime.Demography.set_migration_rate"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.set_migration_rate()</span></code></a></dt><dd><p>Sets the migration rate from one population to another</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.Demography.add_population_parameters_change" title="msprime.Demography.add_population_parameters_change"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.add_population_parameters_change()</span></code></a></dt><dd><p>Change population size and growth rate values at some time</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.Demography.add_migration_rate_change" title="msprime.Demography.add_migration_rate_change"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.add_migration_rate_change()</span></code></a></dt><dd><p>Change the migration rate from one population to another at some time</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.Demography.add_symmetric_migration_rate_change" title="msprime.Demography.add_symmetric_migration_rate_change"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.add_symmetric_migration_rate_change()</span></code></a></dt><dd><p>Change the migration rate among a set of populations at some time</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.Demography.add_population_split" title="msprime.Demography.add_population_split"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.add_population_split()</span></code></a></dt><dd><p>One or more derived populations split from an ancestral population at a given time</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.Demography.add_admixture" title="msprime.Demography.add_admixture"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.add_admixture()</span></code></a></dt><dd><p>A derived population where the given fractions of the individuals are
migrants from the corresponding ancestral populations.</p>
</dd>
</dl>
<p><strong>Predefined theoretical models</strong></p>
<dl class="simple myst">
<dt><a class="reference internal" href="api.html#msprime.Demography.isolated_model" title="msprime.Demography.isolated_model"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.isolated_model()</span></code></a></dt><dd><p>A model of isolated populations</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.Demography.island_model" title="msprime.Demography.island_model"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.island_model()</span></code></a></dt><dd><p>An interconnected web of populations</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.Demography.stepping_stone_model" title="msprime.Demography.stepping_stone_model"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.stepping_stone_model()</span></code></a></dt><dd><p>Simple spatial model</p>
</dd>
</dl>
<p><strong>Constructing models from existing definitions</strong></p>
<dl class="simple myst">
<dt><a class="reference internal" href="api.html#msprime.Demography.from_species_tree" title="msprime.Demography.from_species_tree"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.from_species_tree()</span></code></a></dt><dd><p>Parse a newick species tree</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.Demography.from_starbeast" title="msprime.Demography.from_starbeast"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.from_starbeast()</span></code></a></dt><dd><p>Parse StarBeast output</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.Demography.from_old_style" title="msprime.Demography.from_old_style"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.from_old_style()</span></code></a></dt><dd><p>Demography from msprime legacy 0.x inputs</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.Demography.from_demes" title="msprime.Demography.from_demes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.from_demes()</span></code></a></dt><dd><p>Convert a Demes demographic model to an msprime Demography</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.Demography.to_demes" title="msprime.Demography.to_demes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.to_demes()</span></code></a></dt><dd><p>Convert an msprime Demography to a Demes demographic model</p>
</dd>
</dl>
<p><strong>Debugging</strong></p>
<dl class="simple myst">
<dt><a class="reference internal" href="api.html#msprime.Demography.debug" title="msprime.Demography.debug"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.debug()</span></code></a></dt><dd><p>Get a Debugger for a demography</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.DemographyDebugger" title="msprime.DemographyDebugger"><code class="xref py py-class docutils literal notranslate"><span class="pre">DemographyDebugger</span></code></a></dt><dd><p>Debugger for demographic models</p>
</dd>
</dl>
<p><strong>Numerical methods</strong></p>
<dl class="simple myst">
<dt><a class="reference internal" href="api.html#msprime.DemographyDebugger.coalescence_rate_trajectory" title="msprime.DemographyDebugger.coalescence_rate_trajectory"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DemographyDebugger.coalescence_rate_trajectory()</span></code></a></dt><dd><p>Compute mean coalescence rate and fraction of uncoalesced lineages</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.DemographyDebugger.mean_coalescence_time" title="msprime.DemographyDebugger.mean_coalescence_time"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DemographyDebugger.mean_coalescence_time()</span></code></a></dt><dd><p>Compute the mean time until coalescence</p>
</dd>
<dt><a class="reference internal" href="api.html#msprime.DemographyDebugger.lineage_probabilities" title="msprime.DemographyDebugger.lineage_probabilities"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DemographyDebugger.lineage_probabilities()</span></code></a></dt><dd><p>Probability of the location of lineages over time</p>
</dd>
</dl>
</section>
<hr class="docutils" />
<section id="model">
<h2>Model<a class="headerlink" href="#model" title="Link to this heading">#</a></h2>
<p>In this section we define the model used to describe demography
in msprime, as encapsulated by a <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> object.</p>
<section id="direction-of-time">
<span id="sec-demography-direction-of-time"></span><h3>Direction of time<a class="headerlink" href="#direction-of-time" title="Link to this heading">#</a></h3>
<p>Msprime simulates retrospective population models like the coalescent
in which we begin with a sample of lineages at the present time
and then simulate the historical events that affect these lineages.
The process moves <strong>backwards in time</strong>, and time is measured in
units of <strong>generations ago</strong>. Everything in msprime is considered from
this backwards in time perspective of the simulation.
So here, and throughout the documentation,
when we say something happens “before” or “after” a particular event
this is in terms of the simulation itself, and therefore usually
the <strong>opposite</strong> of what you would mean if thinking about things
forwards in time.</p>
<p>This is initially very confusing, but getting used to thinking “backwards in
time” is essential for coalescent modelling and working with ancestral
histories in general, so it’s worth the effort of getting used to it. For
example, understanding how different patterns of lineages moving from one
population to another backwards in time corresponds to
<a class="reference internal" href="#sec-demography-events-population-split"><span class="std std-ref">population splits</span></a> and
<a class="reference internal" href="#sec-demography-events-admixture"><span class="std std-ref">admixture</span></a> events is a
great way to deepen your understanding of the coalescent in general.</p>
<p>Mostly it doesn’t matter, and it’s just a point of clarification
for reading this documentation and understanding why the models
we define are usually expressed backwards in time. However,
it is essential to be remember that <a class="reference internal" href="#sec-demography-migration"><span class="std std-ref">migration</span></a>
is described from the perspective of the coalescent, so that when describing
migrations in msprime, the following reversed terminology holds:</p>
<dl class="simple myst" id="sec-demography-direction-of-time-dest">
<dt>destination (or <code class="docutils literal notranslate"><span class="pre">dest</span></code>)</dt><dd><p>The population containing the immediate ancestors of the migrant(s). Forwards in time,
this is where migrants come from. In reverse time, a lineage that is traced
backwards in time (i.e. from the current day into the past) will appear to jump
<em>into</em> this ancestral destination at the precise time of the migration event.</p>
</dd>
</dl>
<dl class="simple myst" id="sec-demography-direction-of-time-source">
<dt><code class="docutils literal notranslate"><span class="pre">source</span></code></dt><dd><p>The population containing the immediate descendants of the migrants. Forwards in time,
this is where migrants end up. In reverse time, a lineage traced backwards
from the current day will appear to jump <em>from</em> this population into the ancestral
(“destination”) population.</p>
</dd>
</dl>
<p>We try to flag up the meaning of these definitions in the documentation,
where it’s important.</p>
</section>
<section id="definitions">
<span id="sec-demography-definitions"></span><h3>Definitions<a class="headerlink" href="#definitions" title="Link to this heading">#</a></h3>
<p>Formally, population structure in msprime is modelled by specifying a fixed
number of subpopulations <span class="math notranslate nohighlight">\(d\)</span>.
Each population has an initial absolute population size <span class="math notranslate nohighlight">\(s\)</span>
and a per generation exponential growth rate <span class="math notranslate nohighlight">\(\alpha\)</span>. The size of a
given population at time <span class="math notranslate nohighlight">\(t\)</span> in the past (measured in generations) is
therefore given by <span class="math notranslate nohighlight">\(s e^{-\alpha t}\)</span>. Demographic events that occur in
the history of the simulated population alter some aspect of this population
configuration at a particular time in the past.</p>
<p>Continuous migration between populations is modelled by a
<span class="math notranslate nohighlight">\(d \times d\)</span> matrix <span class="math notranslate nohighlight">\(M\)</span> of per-generation migration rates.
The <span class="math notranslate nohighlight">\((j,k)^{th}\)</span> entry
of <span class="math notranslate nohighlight">\(M\)</span> is the expected number of migrants moving from population
<span class="math notranslate nohighlight">\(k\)</span> to population <span class="math notranslate nohighlight">\(j\)</span> per generation, divided by the size of
population <span class="math notranslate nohighlight">\(j\)</span>. In terms of the coalescent process, <span class="math notranslate nohighlight">\(M_{j,k}\)</span>
gives the rate at which an ancestral lineage moves from population
<span class="math notranslate nohighlight">\(j\)</span> to population <span class="math notranslate nohighlight">\(k\)</span>, as one follows it back through time. In
continuous-time models, when <span class="math notranslate nohighlight">\(M_{j,k}\)</span> is close to zero, this rate is
approximately equivalent to the fraction of population <span class="math notranslate nohighlight">\(j\)</span> that is
replaced each generation by migrants from population <span class="math notranslate nohighlight">\(k\)</span>. In
discrete-time models, the equivalence is exact and each row of <span class="math notranslate nohighlight">\(M\)</span>
has the constraint <span class="math notranslate nohighlight">\(\sum_{k \neq j} M_{j,k} \leq 1\)</span>. This differs
from the migration matrix one usually uses in population demography: if
<span class="math notranslate nohighlight">\(m_{k,j}\)</span> is the proportion of individuals (in the usual sense; not
lineages) in population <span class="math notranslate nohighlight">\(k\)</span> that move to population <span class="math notranslate nohighlight">\(j\)</span> per
generation, then translating this proportion of population <span class="math notranslate nohighlight">\(k\)</span> to a
proportion of population <span class="math notranslate nohighlight">\(j\)</span>, we have
<span class="math notranslate nohighlight">\(M_{j,k} = m_{k,j} \times N_k / N_j\)</span>.</p>
<p>The details of population structure in msprime closely follow the model
used in the classical
<a class="reference external" href="https://academic.oup.com/bioinformatics/article/18/2/337/225783">ms</a>
program.</p>
<hr class="docutils" />
</section>
<section id="populations">
<span id="sec-demography-populations"></span><h3>Populations<a class="headerlink" href="#populations" title="Link to this heading">#</a></h3>
<p>A <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> contains a list of <a class="reference internal" href="api.html#msprime.Population" title="msprime.Population"><code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code></a> objects.
A population represents some randomly mating set of individuals that
optionally exchange migrants with other populations at some rate.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Msprime version 0.x performed poorly when there were large numbers of
populations, even if most of those populations had no lineages present
for much of the simulation.
Version 1 has a much better support for large numbers of populations,
and there is now no performance cost for having inactive populations.</p>
</div>
<p>Populations essentially have two purposes:</p>
<ol class="arabic simple">
<li><p>To define the state of the population at the start of the simulation (i.e.,
the <a class="reference internal" href="#sec-demography-populations-initial-size"><span class="std std-ref">Initial size</span></a> and
the <a class="reference internal" href="#sec-demography-populations-growth-rate"><span class="std std-ref">Growth rate</span></a>)</p></li>
<li><p>To define <a class="reference internal" href="#sec-demography-populations-metadata"><span class="std std-ref">Metadata</span></a> which is
associated with the corresponding <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.Population" title="(in Project name not set)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.Population</span></code></a> objects
in simulated tree sequences.</p></li>
</ol>
<p>Populations additionally have a “state” which determines whether lineages
can be present within them or not at a given time, as described in
the <a class="reference internal" href="#sec-demography-populations-life-cycle"><span class="std std-ref">Life cycle</span></a> section.</p>
<section id="sec-demography-populations-identifiers">
<span id="identifying-populations"></span><h4>Identifying populations<a class="headerlink" href="#sec-demography-populations-identifiers" title="Link to this heading">#</a></h4>
<p>When used as part of the <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a>, each population has an integer
ID (its zero-based index in the <code class="docutils literal notranslate"><span class="pre">populations</span></code> list) and a <code class="docutils literal notranslate"><span class="pre">name</span></code> attribute.
By default, msprime assigns the name <code class="docutils literal notranslate"><span class="pre">pop_j</span></code> to the population at index
<code class="docutils literal notranslate"><span class="pre">j</span></code> in a demographic model. These default names can be overridden, and
users should give populations more descriptive names when building complex models.</p>
<div class="admonition attention">
<p class="admonition-title">Attention</p>
<p>Population names must be unique within a single <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> and be valid
<a class="reference external" href="https://docs.python.org/3/reference/lexical_analysis.html#identifiers">Python identifiers</a>.
For example, this means that names like “my_pop_1” is valid, but “my-pop-1” and “my pop 1”
are not.</p>
</div>
<p>When defining empirical models (such as those in the <a class="reference internal" href="#sec-demography-examples"><span class="std std-ref">Example models</span></a>
section) names should be chosen for clarity.
We also encourage the use
of <a class="reference internal" href="#sec-demography-events-population-split"><span class="std std-ref">population split</span></a>
events to explicitly name and describe the different phases of a demographic
history rather than use anonymous population size changes (as encouraged
by the <a class="reference internal" href="legacy.html#sec-legacy-0x"><span class="std std-ref">legacy 0.x API</span></a>). This makes it easier
to interpret the resulting tree sequence files (since detailed metadata
about the populations in which coalescences occurred is maintained)
and makes it easier to avoid
<a class="reference external" href="http://dx.doi.org/10.1016/j.ajhg.2020.08.017">modelling errors</a>.</p>
</section>
<section id="initial-size">
<span id="sec-demography-populations-initial-size"></span><h4>Initial size<a class="headerlink" href="#initial-size" title="Link to this heading">#</a></h4>
<p>The <code class="docutils literal notranslate"><span class="pre">initial_size</span></code> of a population is its size, as a number of individuals,
at the start of a simulation
(looking <a class="reference internal" href="#sec-demography-direction-of-time"><span class="std std-ref">backwards in time</span></a>).
In other words, an initial size of <code class="docutils literal notranslate"><span class="pre">n</span></code> means that the population had
<code class="docutils literal notranslate"><span class="pre">n</span> <span class="pre">*</span> <span class="pre">ploidy</span></code> chromosomes at the most recent time.
If the population’s
<a class="reference internal" href="#sec-demography-populations-growth-rate"><span class="std std-ref">Growth rate</span></a> is zero, then
the population will have the same size for all time (unless there are
some <a class="reference internal" href="#sec-demography-events"><span class="std std-ref">Events</span></a> that change these parameters).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <code class="docutils literal notranslate"><span class="pre">initial_size</span></code> of a population refers to its size at time zero,
and is <strong>not</strong> affected by the simulation <a class="reference internal" href="ancestry.html#sec-ancestry-start-time"><span class="std std-ref">start_time</span></a>.</p>
</div>
<p>When using methods like <a class="reference internal" href="api.html#msprime.Demography.island_model" title="msprime.Demography.island_model"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.island_model()</span></code></a> to describe
simple multipopulation models, the <code class="docutils literal notranslate"><span class="pre">initial_size</span></code> parameter takes
a <strong>list</strong> of size values which defines both the number of
populations and their sizes. For example, here we define two isolated
populations (i.e., which have no migration) with sizes 100 and 200:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">isolated_model</span><span class="p">([</span><span class="mi">100</span><span class="p">,</span> <span class="mi">200</span><span class="p">])</span>
<span class="n">demography</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (2)</b><table border="1" class="tskit-table"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>pop_0</td><td></td><td>100.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>pop_1</td><td></td><td>200.0</td><td>0</td><td>0</td><td>{}</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Migration matrix (all zero)</b><table border="1" class="tskit-table"><thead><tr></tr></thead><tbody></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events (0)</b><table border="1" class="tskit-table"><thead><tr></tr></thead><tbody></tbody></table></div></div></div></div>
</div>
<p>When we have multiple populations that are the same size we can use
some Python tricks to avoid code duplication. For example, here
we create three populations with size 100:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">isolated_model</span><span class="p">([</span><span class="mi">100</span><span class="p">]</span> <span class="o">*</span> <span class="mi">3</span><span class="p">)</span>
<span class="n">demography</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (3)</b><table border="1" class="tskit-table"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>pop_0</td><td></td><td>100.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>pop_1</td><td></td><td>100.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>2</td><td>pop_2</td><td></td><td>100.0</td><td>0</td><td>0</td><td>{}</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Migration matrix (all zero)</b><table border="1" class="tskit-table"><thead><tr></tr></thead><tbody></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events (0)</b><table border="1" class="tskit-table"><thead><tr></tr></thead><tbody></tbody></table></div></div></div></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>We use the <a class="reference internal" href="api.html#msprime.Demography.isolated_model" title="msprime.Demography.isolated_model"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.isolated_model()</span></code></a> function here as a
way to create example demographies. However, it’s
important to note that simulations in which we sample from
multiple isolated populations will fail with an error because
the lineages in question can never coalesce. See
the <a class="reference internal" href="ancestry.html#sec-ancestry-errors-infinite-waiting-time"><span class="std std-ref">Infinite waiting time</span></a> section
for more details.</p>
</div>
</section>
<section id="growth-rate">
<span id="sec-demography-populations-growth-rate"></span><h4>Growth rate<a class="headerlink" href="#growth-rate" title="Link to this heading">#</a></h4>
<p>Each population has an exponential growth rate so that the
size of a population with initial size <span class="math notranslate nohighlight">\(s\)</span>
and growth rate <span class="math notranslate nohighlight">\(\alpha\)</span> is <span class="math notranslate nohighlight">\(s e^{-\alpha t}\)</span>
at time <span class="math notranslate nohighlight">\(t\)</span> generations in the past (see the
<a class="reference internal" href="#sec-demography-definitions"><span class="std std-ref">Definitions</span></a> section for more details).</p>
<p>Growth rates for functions like <a class="reference internal" href="api.html#msprime.Demography.island_model" title="msprime.Demography.island_model"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.island_model()</span></code></a>
that construct a <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> are specified in a similar
way to <a class="reference internal" href="#sec-demography-populations-initial-size"><span class="std std-ref">Initial size</span></a>: we provide
a list of sizes equal to the number of populations. For
example, here we define 2 populations with different
population sizes and growth rates:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">isolated_model</span><span class="p">([</span><span class="mi">100</span><span class="p">,</span> <span class="mi">200</span><span class="p">],</span> <span class="n">growth_rate</span><span class="o">=</span><span class="p">[</span><span class="mf">0.01</span><span class="p">,</span> <span class="mf">0.02</span><span class="p">])</span>
<span class="n">demography</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (2)</b><table border="1" class="tskit-table"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>pop_0</td><td></td><td>100.0</td><td>0.01</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>pop_1</td><td></td><td>200.0</td><td>0.02</td><td>0</td><td>{}</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Migration matrix (all zero)</b><table border="1" class="tskit-table"><thead><tr></tr></thead><tbody></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events (0)</b><table border="1" class="tskit-table"><thead><tr></tr></thead><tbody></tbody></table></div></div></div></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <a class="reference internal" href="api.html#msprime.DemographyDebugger" title="msprime.DemographyDebugger"><code class="xref py py-class docutils literal notranslate"><span class="pre">DemographyDebugger</span></code></a> is a useful tool to help understand
how population sizes change over time. See the
<a class="reference internal" href="#sec-demography-debugging-tools"><span class="std std-ref">Debugging tools</span></a> section for details.</p>
</div>
</section>
<section id="metadata">
<span id="sec-demography-populations-metadata"></span><h4>Metadata<a class="headerlink" href="#metadata" title="Link to this heading">#</a></h4>
<p>In <a class="reference external" href="https://tskit.dev/tskit">tskit</a> the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.Population" title="(in Project name not set)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code></a> class
largely exists as a container for <a class="reference external" href="https://tskit.dev/tskit/docs/stable/metadata.html#sec-metadata" title="(in Project name not set)"><span>Metadata</span></a>. Metadata is
important and useful: it lets us associate information about our simulated
populations with the output tree sequence, which we can then use
when we are analysing the data.</p>
<p>Msprime associates two pieces of mandatory metadata with every population:
their <code class="docutils literal notranslate"><span class="pre">name</span></code> and <code class="docutils literal notranslate"><span class="pre">description</span></code>. For example,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">stepping_stone_model</span><span class="p">([</span><span class="mi">100</span><span class="p">,</span> <span class="mi">100</span><span class="p">],</span> <span class="n">migration_rate</span><span class="o">=</span><span class="mf">0.01</span><span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">({</span><span class="mi">0</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span> <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">)</span>
<span class="nb">print</span><span class="p">([</span><span class="n">population</span><span class="o">.</span><span class="n">metadata</span> <span class="k">for</span> <span class="n">population</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">populations</span><span class="p">()])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[{&#39;description&#39;: &#39;&#39;, &#39;name&#39;: &#39;pop_0&#39;}, {&#39;description&#39;: &#39;&#39;, &#39;name&#39;: &#39;pop_1&#39;}]
</pre></div>
</div>
</div>
</div>
<p>Here we have two populations in the output tree sequence, and the metadata
for each population is a dictionary containing the keys <code class="docutils literal notranslate"><span class="pre">name</span></code> and
<code class="docutils literal notranslate"><span class="pre">description</span></code>. These correspond to the same attributes on the msprime
<a class="reference internal" href="api.html#msprime.Population" title="msprime.Population"><code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code></a> class. We don’t have to just use the defaults
for these values: we can set them to (more or less) anything we like.
For example,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">stepping_stone_model</span><span class="p">([</span><span class="mi">100</span><span class="p">,</span> <span class="mi">100</span><span class="p">],</span> <span class="n">migration_rate</span><span class="o">=</span><span class="mf">0.01</span><span class="p">)</span>
<span class="n">demography</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;awesome_pop&quot;</span>
<span class="n">demography</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="s2">&quot;This population is totally awesome&quot;</span>
<span class="n">demography</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;more_awesome_pop&quot;</span>
<span class="n">demography</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="s2">&quot;This population is even more awesome&quot;</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">({</span><span class="mi">0</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span> <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">)</span>
<span class="k">for</span> <span class="n">population</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">populations</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;id=</span><span class="si">{</span><span class="n">population</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">population</span><span class="o">.</span><span class="n">metadata</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>id=0: {&#39;description&#39;: &#39;This population is totally awesome&#39;, &#39;name&#39;: &#39;awesome_pop&#39;}
id=1: {&#39;description&#39;: &#39;This population is even more awesome&#39;, &#39;name&#39;: &#39;more_awesome_pop&#39;}
</pre></div>
</div>
</div>
</div>
<p>As well as the default <code class="docutils literal notranslate"><span class="pre">name</span></code> and <code class="docutils literal notranslate"><span class="pre">description</span></code> metadata keys
we can also associate additional metadata with population objects using
the <code class="docutils literal notranslate"><span class="pre">extra_metadata</span></code> attribute of the msprime <a class="reference internal" href="api.html#msprime.Population" title="msprime.Population"><code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code></a>
object. For example,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">stepping_stone_model</span><span class="p">([</span><span class="mi">100</span><span class="p">,</span> <span class="mi">100</span><span class="p">],</span> <span class="n">migration_rate</span><span class="o">=</span><span class="mf">0.01</span><span class="p">)</span>
<span class="n">demography</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;awesome_pop&quot;</span>
<span class="n">demography</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">extra_metadata</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;emoji&quot;</span><span class="p">:</span> <span class="s2">&quot;👍&quot;</span><span class="p">}</span>
<span class="n">demography</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;more_awesome_pop&quot;</span>
<span class="n">demography</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">extra_metadata</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;emoji&quot;</span><span class="p">:</span> <span class="s2">&quot;🤘&quot;</span><span class="p">}</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">({</span><span class="mi">0</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span> <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">)</span>
<span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">populations</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">pop</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;emoji&quot;</span><span class="p">],</span> <span class="n">pop</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0 	 👍 awesome_pop
1 	 🤘 more_awesome_pop
</pre></div>
</div>
</div>
</div>
</section>
<section id="life-cycle">
<span id="sec-demography-populations-life-cycle"></span><h4>Life cycle<a class="headerlink" href="#life-cycle" title="Link to this heading">#</a></h4>
<p>Each population has a <code class="docutils literal notranslate"><span class="pre">state</span></code> associated with it during the simulation,
which can be one of <code class="docutils literal notranslate"><span class="pre">INACTIVE</span></code>, <code class="docutils literal notranslate"><span class="pre">ACTIVE</span></code> or <code class="docutils literal notranslate"><span class="pre">PREVIOUSLY_ACTIVE</span></code>.
Populations can follow one of two state progressions:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>INACTIVE → ACTIVE
</pre></div>
</div>
<p>and</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>ACTIVE → PREVIOUSLY_ACTIVE
</pre></div>
</div>
<p>All populations are <code class="docutils literal notranslate"><span class="pre">ACTIVE</span></code> by default at the start of the simulation
(looking <a class="reference internal" href="#sec-demography-direction-of-time"><span class="std std-ref">backwards in time</span></a>)
except if they are <code class="docutils literal notranslate"><span class="pre">ancestral</span></code> in a
<a class="reference internal" href="#sec-demography-events-population-split"><span class="std std-ref">Population split</span></a>.
In this case populations are initially <code class="docutils literal notranslate"><span class="pre">INACTIVE</span></code>, by default.
An <code class="docutils literal notranslate"><span class="pre">ancestral</span></code> population then transitions
from <code class="docutils literal notranslate"><span class="pre">INACTIVE</span></code> → <code class="docutils literal notranslate"><span class="pre">ACTIVE</span></code> when the corresponding
population split event occurs.</p>
<p>It is possible to override the default initial state of the population
that is ancestral in a population split event using the <code class="docutils literal notranslate"><span class="pre">initially_active</span></code>
boolean flag in the <a class="reference internal" href="api.html#msprime.Population" title="msprime.Population"><code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code></a> class. This is to facilitate
“trunk population” models, as shown in
<a class="reference internal" href="#sec-demography-examples-trunk-population"><span class="std std-ref">this example</span></a>.</p>
<p>Populations transition from <code class="docutils literal notranslate"><span class="pre">ACTIVE</span></code> → <code class="docutils literal notranslate"><span class="pre">PREVIOUSLY_ACTIVE</span></code> when they
are <code class="docutils literal notranslate"><span class="pre">derived</span></code> in either population split or admixture events.</p>
<p>Any attempts to move lineages into a population that is not <code class="docutils literal notranslate"><span class="pre">ACTIVE</span></code>
via sampling, continuous or pulse migration will result in an error
during the simulation.</p>
</section>
<section id="default-sampling-time">
<span id="sec-demography-populations-default-sampling-time"></span><h4>Default sampling time<a class="headerlink" href="#default-sampling-time" title="Link to this heading">#</a></h4>
<p>The <code class="docutils literal notranslate"><span class="pre">default_sampling_time</span></code> property of a <a class="reference internal" href="api.html#msprime.Population" title="msprime.Population"><code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code></a>
defines the default time at which samples will be drawn from a population
in <a class="reference internal" href="api.html#msprime.sim_ancestry" title="msprime.sim_ancestry"><code class="xref py py-func docutils literal notranslate"><span class="pre">sim_ancestry()</span></code></a>. See the
<a class="reference internal" href="ancestry.html#sec-ancestry-samples"><span class="std std-ref">Specifying samples</span></a> section for details on how this time
is used when running simulations, and how sampling time can also
be specified directly, overriding the population default.</p>
<p>The default sampling time can be specified directly using either
the <a class="reference internal" href="api.html#msprime.Demography.add_population" title="msprime.Demography.add_population"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.add_population()</span></code></a> method, or by setting the
<code class="docutils literal notranslate"><span class="pre">default_sampling_time</span></code> attribute of the <a class="reference internal" href="api.html#msprime.Population" title="msprime.Population"><code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code></a>
later. The default value is 0, <strong>unless</strong> the population is
ancestral in a
<a class="reference internal" href="#sec-demography-events-population-split"><span class="std std-ref">Population split</span></a> event added
using the <a class="reference internal" href="api.html#msprime.Demography.add_population_split" title="msprime.Demography.add_population_split"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.add_population_split()</span></code></a> method. In
this case the default sampling time is set to the
time of the event (i.e., when the population
<a class="reference internal" href="#sec-demography-direction-of-time"><span class="std std-ref">first</span></a>
becomes <a class="reference internal" href="#sec-demography-populations-life-cycle"><span class="std std-ref">active</span></a>),
unless the <code class="docutils literal notranslate"><span class="pre">initially_active</span></code> flag has been set.</p>
</section>
</section>
<section id="migration">
<span id="sec-demography-migration"></span><h3>Migration<a class="headerlink" href="#migration" title="Link to this heading">#</a></h3>
<p>Migration is the process of lineages moving from one population to another
during the simulation.
This either happens through continuous migration, where a rate of
migration between each pair of populations is defined, or through
<a class="reference internal" href="#sec-demography-events-mass-migration"><span class="std std-ref">Pulse (mass) migration</span></a> events. In this section
we are concerned with continuous migration.</p>
<p>As described in the <a class="reference internal" href="#sec-demography-definitions"><span class="std std-ref">Definitions</span></a> section,
continuous migration between populations is modelled by a matrix
of rates, so that <code class="docutils literal notranslate"><span class="pre">M[j,</span> <span class="pre">k]</span></code> is the rate at which lineages move from
population <code class="docutils literal notranslate"><span class="pre">j</span></code> to population <code class="docutils literal notranslate"><span class="pre">k</span></code> in the <strong>coalescent process</strong>,
that is, <strong>backwards in time</strong>
(see the <a class="reference internal" href="#sec-demography-direction-of-time"><span class="std std-ref">Direction of time</span></a> section for more details).
Lineages that move from population
<code class="docutils literal notranslate"><span class="pre">j</span></code> to <code class="docutils literal notranslate"><span class="pre">k</span></code> backwards in time actually correspond to individuals
migrating from population <code class="docutils literal notranslate"><span class="pre">k</span></code> to <code class="docutils literal notranslate"><span class="pre">j</span></code> <strong>forwards in time</strong>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you’re confused at this point, don’t worry. Everyone is confused by
this.</p>
</div>
<p>Let’s take an example to clarify. Suppose we have a two population model
in which we have migration at rate 0.1 from <code class="docutils literal notranslate"><span class="pre">source</span></code> population <code class="docutils literal notranslate"><span class="pre">0</span></code> to
<code class="docutils literal notranslate"><span class="pre">dest</span></code> population <code class="docutils literal notranslate"><span class="pre">1</span></code> and no
migration in the other direction. We’ll then take one haploid sample
from each population:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">isolated_model</span><span class="p">([</span><span class="mi">100</span><span class="p">,</span> <span class="mi">100</span><span class="p">])</span>
<span class="n">demography</span><span class="o">.</span><span class="n">set_migration_rate</span><span class="p">(</span><span class="n">source</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="n">samples</span><span class="o">=</span><span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span>
    <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">,</span>
    <span class="n">ploidy</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">12345</span><span class="p">)</span>
<span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
        <div>
            <style scoped="">
                .tskit-table tbody tr th:only-of-type {vertical-align: middle;}
                .tskit-table tbody tr th {vertical-align: top;}
                .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
                .tskit-table tbody th {padding: 0.5em 0.5em;}
            </style>
            <table border="1" class="tskit-table">
                <thead>
                    <tr>
                        <th>id</th><th>flags</th><th>population</th><th>individual</th><th>time</th><th>metadata</th>
                    </tr>
                </thead>
                <tbody>
                    <tr><td>0</td><td>1</td><td>0</td><td>0</td><td>0.00000000</td><td></td></tr>
<tr><td>1</td><td>1</td><td>1</td><td>1</td><td>0.00000000</td><td></td></tr>
<tr><td>2</td><td>0</td><td>1</td><td>-1</td><td>64.57256316</td><td></td></tr>

                </tbody>
            </table>
        </div>
    </div></div>
</div>
<p>Because we have only two samples and no recombination we have
only one coalescent event in the simulation, and therefore
only one ancestral node. We are interested in what
populations nodes are associated with, which is shown in
the <code class="docutils literal notranslate"><span class="pre">population</span></code> column of the node table. We can see
that our samples, nodes 0 and 1, are associated with populations 0 and 1,
respectively. Node 2, which is the ancestor of nodes
0 and 1, is from population 1 which means that the <strong>lineage</strong>
for node 0 moved from (<code class="docutils literal notranslate"><span class="pre">source</span></code>) population 0 to
(<code class="docutils literal notranslate"><span class="pre">dest</span></code>) population 1 as we traced
its history back through time, so that it could ultimately
coalesce with lineage 1 in population 1. However, <strong>forwards</strong> in
time, this must mean that one of the individuals along node 0’s lineage
(each lineage is a succession of individuals, passing on genetic
information through the generations) must have migrated
from population <code class="docutils literal notranslate"><span class="pre">1</span></code> to population <code class="docutils literal notranslate"><span class="pre">0</span></code>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you’re still confused, don’t worry, it’s still OK. Just remember
that migration rates are confusing and come back to check the
documentation whenever you need to work with them.</p>
</div>
<p>Looking back through time,
a lineage moves from one population to another
when the individual that lineage is in was born,
at which point the lineage moves to the parent.
So, the migration rate of lineages from <code class="docutils literal notranslate"><span class="pre">j</span></code> to <code class="docutils literal notranslate"><span class="pre">k</span></code>
is the proportion of population <code class="docutils literal notranslate"><span class="pre">j</span></code> that is replaced
by offspring of individuals in population <code class="docutils literal notranslate"><span class="pre">k</span></code>, per generation
(since this is the probability that the lineage we are
following happens to have a parent in <code class="docutils literal notranslate"><span class="pre">k</span></code>).</p>
<p>If you have symmetric rates you needn’t worry about mixing things up;
set these with the <a class="reference internal" href="api.html#msprime.Demography.set_symmetric_migration_rate" title="msprime.Demography.set_symmetric_migration_rate"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.set_symmetric_migration_rate()</span></code></a> and
<a class="reference internal" href="api.html#msprime.Demography.add_symmetric_migration_rate_change" title="msprime.Demography.add_symmetric_migration_rate_change"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.add_symmetric_migration_rate_change()</span></code></a> methods
to set the migration rate in <em>both</em> directions. This has the
advantage of reducing duplication in our code, as well as
meaning that we don’t need to remember what <code class="docutils literal notranslate"><span class="pre">source</span></code> and <code class="docutils literal notranslate"><span class="pre">dest</span></code>
mean. See the <a class="reference internal" href="#sec-demography-examples"><span class="std std-ref">Example models</span></a> section for
examples of setting symmetric migration rates.</p>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>Clarify that demes does things the other way around.</p>
</div>
</section>
<section id="events">
<span id="sec-demography-events"></span><h3>Events<a class="headerlink" href="#events" title="Link to this heading">#</a></h3>
<p>Setting the population parameters and migration matrix in the <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a>
object define the state of demographic model at the
<a class="reference internal" href="#sec-demography-direction-of-time"><span class="std std-ref">start</span></a> of the simulation
(i.e., the present). We are often interested in population models in which
these parameters change over time; these are implemented through a set
of demographic events, which make some instantaneous changes to the state
of the simulation.</p>
<p>The <a class="reference internal" href="api.html#msprime.DemographyDebugger" title="msprime.DemographyDebugger"><code class="xref py py-class docutils literal notranslate"><span class="pre">DemographyDebugger</span></code></a> is very useful for inspecting the
state of the simulation at each of the epochs defined by these
events—see the <a class="reference internal" href="#sec-demography-debugging-tools"><span class="std std-ref">Debugging tools</span></a> for examples.</p>
<section id="population-parameters-change">
<span id="sec-demography-events-population-parameters-change"></span><h4>Population parameters change<a class="headerlink" href="#population-parameters-change" title="Link to this heading">#</a></h4>
<p>The <a class="reference internal" href="api.html#msprime.Demography.add_population_parameters_change" title="msprime.Demography.add_population_parameters_change"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.add_population_parameters_change()</span></code></a> method
adds an event to change the
<code class="docutils literal notranslate"><span class="pre">initial_size</span></code> or <code class="docutils literal notranslate"><span class="pre">growth_rate</span></code> of a
<a class="reference internal" href="#sec-demography-populations"><span class="std std-ref">population</span></a> (or populations)
at a given time. For example,
here we create a two-population model and add two events that
change the parameters of these populations over time. The first
event changes the size of population <code class="docutils literal notranslate"><span class="pre">A</span></code> to 200, 10 generations
in the past. The second event changes the size of <em>both</em> populations
to 10 after 20 generations using the <code class="docutils literal notranslate"><span class="pre">population=None</span></code> shorthand.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population_parameters_change</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">population</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population_parameters_change</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">population</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">debug</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div><details open="true"><summary>Epoch[0]: [0, 10) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=2 active=2)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>A</th><th>B</th></tr></thead><tbody><tr><td>A</td><td> 100.0</td><td> 100.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population A to B = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from B to A per
generation, divided by the size of A.'>0</td></tr><tr><td>B</td><td> 100.0</td><td> 100.0</td><td> 0</td><td title='Backwards in time migration rate from population B to A = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from A to B per
generation, divided by the size of B.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 10</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>10</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_parameters_change'>Population parameter change</a></td><td>population=A, initial_size=200</td><td>initial_size → 2e+02 for population A</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[1]: [10, 20) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=2 active=2)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>A</th><th>B</th></tr></thead><tbody><tr><td>A</td><td> 200.0</td><td> 200.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population A to B = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from B to A per
generation, divided by the size of A.'>0</td></tr><tr><td>B</td><td> 100.0</td><td> 100.0</td><td> 0</td><td title='Backwards in time migration rate from population B to A = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from A to B per
generation, divided by the size of B.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 20</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>20</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_parameters_change'>Population parameter change</a></td><td>population=-1, initial_size=10</td><td>initial_size → 10 for all populations</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[2]: [20, inf) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=2 active=2)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>A</th><th>B</th></tr></thead><tbody><tr><td>A</td><td> 10.0</td><td> 10.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population A to B = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from B to A per
generation, divided by the size of A.'>0</td></tr><tr><td>B</td><td> 10.0</td><td> 10.0</td><td> 0</td><td title='Backwards in time migration rate from population B to A = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from A to B per
generation, divided by the size of B.'>0</td><td>0</td></tr></tbody></table></div></div></details></div></div></div>
</div>
</section>
<section id="migration-rate-change">
<span id="sec-demography-events-migration-rate-change"></span><h4>Migration rate change<a class="headerlink" href="#migration-rate-change" title="Link to this heading">#</a></h4>
<p>The <a class="reference internal" href="api.html#msprime.Demography.add_symmetric_migration_rate_change" title="msprime.Demography.add_symmetric_migration_rate_change"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.add_symmetric_migration_rate_change()</span></code></a>
and <a class="reference internal" href="api.html#msprime.Demography.add_migration_rate_change" title="msprime.Demography.add_migration_rate_change"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.add_migration_rate_change()</span></code></a> methods
are used to change the state of the
<a class="reference internal" href="#sec-demography-migration"><span class="std std-ref">migration matrix</span></a> at a given time.
Here we create a two population model with no migration initially,
and then add symmetric migration between the two populations
at rate 0.1 from generation 10.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_symmetric_migration_rate_change</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">populations</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span><span class="s2">&quot;B&quot;</span><span class="p">],</span> <span class="n">rate</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">debug</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div><details open="true"><summary>Epoch[0]: [0, 10) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=2 active=2)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>A</th><th>B</th></tr></thead><tbody><tr><td>A</td><td> 100.0</td><td> 100.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population A to B = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from B to A per
generation, divided by the size of A.'>0</td></tr><tr><td>B</td><td> 100.0</td><td> 100.0</td><td> 0</td><td title='Backwards in time migration rate from population B to A = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from A to B per
generation, divided by the size of B.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 10</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>10</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_symmetric_migration_rate_change'>Symmetric migration rate change</a></td><td>populations=[A, B], rate=0.1</td><td>Sets the symmetric migration rate between A and B to 0.1 per generation</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[1]: [10, inf) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=2 active=2)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>A</th><th>B</th></tr></thead><tbody><tr><td>A</td><td> 100.0</td><td> 100.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population A to B = 0.1 per generation.
Forwards in time, this is the expected number of migrants moving from B to A per
generation, divided by the size of A.'>0.1</td></tr><tr><td>B</td><td> 100.0</td><td> 100.0</td><td> 0</td><td title='Backwards in time migration rate from population B to A = 0.1 per generation.
Forwards in time, this is the expected number of migrants moving from A to B per
generation, divided by the size of B.'>0.1</td><td>0</td></tr></tbody></table></div></div></details></div></div></div>
</div>
</section>
<section id="population-split">
<span id="sec-demography-events-population-split"></span><h4>Population split<a class="headerlink" href="#population-split" title="Link to this heading">#</a></h4>
<p>The <a class="reference internal" href="api.html#msprime.Demography.add_population_split" title="msprime.Demography.add_population_split"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.add_population_split()</span></code></a> method is used to declare
that one or more populations derive from a given ancestral
population at some time. Consider the following population tree
(much like a <a class="reference internal" href="#sec-demography-importing-species-trees"><span class="std std-ref">species tree</span></a>)
in which we have three present-day populations <code class="docutils literal notranslate"><span class="pre">A</span></code>, <code class="docutils literal notranslate"><span class="pre">B</span></code>
and <code class="docutils literal notranslate"><span class="pre">C</span></code>, which merge into ancestral populations <code class="docutils literal notranslate"><span class="pre">AB</span></code> and <code class="docutils literal notranslate"><span class="pre">ABC</span></code>:</p>
<div class="cell tag_hide-input docutils container">
<details class="admonition hide above-input">
<summary aria-label="Toggle hidden content">
<p class="collapsed admonition-title">Show code cell source</p>
<p class="expanded admonition-title">Hide code cell source</p>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">tskit</span>

<span class="n">tables</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">TableCollection</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">3</span><span class="p">):</span>
    <span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">flags</span><span class="o">=</span><span class="n">tskit</span><span class="o">.</span><span class="n">NODE_IS_SAMPLE</span><span class="p">)</span>
<span class="n">AB</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">ABC</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">tables</span><span class="o">.</span><span class="n">edges</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">AB</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
<span class="n">tables</span><span class="o">.</span><span class="n">edges</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">AB</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">tables</span><span class="o">.</span><span class="n">edges</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">ABC</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">tables</span><span class="o">.</span><span class="n">edges</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">ABC</span><span class="p">,</span> <span class="n">AB</span><span class="p">)</span>
<span class="n">tables</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="n">tree</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">()</span><span class="o">.</span><span class="n">first</span><span class="p">()</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">node_labels</span><span class="o">=</span><span class="p">{</span><span class="mi">0</span><span class="p">:</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">:</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="mi">3</span><span class="p">:</span><span class="s2">&quot;AB&quot;</span><span class="p">,</span> <span class="mi">4</span><span class="p">:</span><span class="s2">&quot;ABC&quot;</span><span class="p">}))</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<img alt="_images/a3acf4cf4ead0484e442b5d14d49b9ece55b942394b6f29ede335e1a23aeb2e0.svg" src="_images/a3acf4cf4ead0484e442b5d14d49b9ece55b942394b6f29ede335e1a23aeb2e0.svg" />
</div>
</div>
<p>We can describe this population topology as follows:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;AB&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;ABC&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population_split</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">],</span> <span class="n">ancestral</span><span class="o">=</span><span class="s2">&quot;AB&quot;</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population_split</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;AB&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">],</span> <span class="n">ancestral</span><span class="o">=</span><span class="s2">&quot;ABC&quot;</span><span class="p">)</span>
<span class="n">demography</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (5)</b><table border="1" class="tskit-table"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>A</td><td></td><td>100.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>B</td><td></td><td>100.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>2</td><td>C</td><td></td><td>100.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>3</td><td>AB</td><td></td><td>100.0</td><td>0</td><td>10</td><td>{}</td></tr><tr><td>4</td><td>ABC</td><td></td><td>100.0</td><td>0</td><td>20</td><td>{}</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Migration matrix (all zero)</b><table border="1" class="tskit-table"><thead><tr></tr></thead><tbody></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events (2)</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>10</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[A, B], ancestral=AB</td><td>Moves all lineages from derived populations &#x27;A&#x27; and &#x27;B&#x27; to the ancestral &#x27;AB&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr><tr><td>20</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[AB, C], ancestral=ABC</td><td>Moves all lineages from derived populations &#x27;AB&#x27; and &#x27;C&#x27; to the ancestral &#x27;ABC&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr></tbody></table></div></div></div></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Note that the
<a class="reference internal" href="#sec-demography-populations-default-sampling-time"><span class="std std-ref">default_sampling_time</span></a>
value for the ancestral populations is set to the time of the split event.</p>
</div>
<p>The <a class="reference internal" href="#sec-demography-debugging-tools"><span class="std std-ref">debug output</span></a> for this demography
shows how the various populations become
<a class="reference internal" href="#sec-demography-populations-life-cycle"><span class="std std-ref">active and inactive</span></a>
over time:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span><span class="o">.</span><span class="n">debug</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div><details open="true"><summary>Epoch[0]: [0, 10) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=5 active=3)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>A</th><th>B</th><th>C</th></tr></thead><tbody><tr><td>A</td><td> 100.0</td><td> 100.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population A to B = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from B to A per
generation, divided by the size of A.'>0</td><td title='Backwards in time migration rate from population A to C = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from C to A per
generation, divided by the size of A.'>0</td></tr><tr><td>B</td><td> 100.0</td><td> 100.0</td><td> 0</td><td title='Backwards in time migration rate from population B to A = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from A to B per
generation, divided by the size of B.'>0</td><td>0</td><td title='Backwards in time migration rate from population B to C = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from C to B per
generation, divided by the size of B.'>0</td></tr><tr><td>C</td><td> 100.0</td><td> 100.0</td><td> 0</td><td title='Backwards in time migration rate from population C to A = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from A to C per
generation, divided by the size of C.'>0</td><td title='Backwards in time migration rate from population C to B = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from B to C per
generation, divided by the size of C.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 10</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>10</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[A, B], ancestral=AB</td><td>Moves all lineages from derived populations &#x27;A&#x27; and &#x27;B&#x27; to the ancestral &#x27;AB&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[1]: [10, 20) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=5 active=2)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>C</th><th>AB</th></tr></thead><tbody><tr><td>C</td><td> 100.0</td><td> 100.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population C to AB = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from AB to C
per generation, divided by the size of C.'>0</td></tr><tr><td>AB</td><td> 100.0</td><td> 100.0</td><td> 0</td><td title='Backwards in time migration rate from population AB to C = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from C to AB
per generation, divided by the size of AB.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 20</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>20</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[AB, C], ancestral=ABC</td><td>Moves all lineages from derived populations &#x27;AB&#x27; and &#x27;C&#x27; to the ancestral &#x27;ABC&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[2]: [20, inf) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=5 active=1)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th></tr></thead><tbody><tr><td>ABC</td><td> 100.0</td><td> 100.0</td><td> 0</td></tr></tbody></table></div></div></details></div></div></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>See the population tree <a class="reference internal" href="#sec-demography-examples-population-tree"><span class="std std-ref">example</span></a>
for a Human out-of-Africa model from the literature.</p>
</div>
<p>It is also possible to have less tree-like population topologies, in which
several populations successively merge into one “trunk” population. For
example, here we define three populations, where <code class="docutils literal notranslate"><span class="pre">B</span></code> and <code class="docutils literal notranslate"><span class="pre">C</span></code> merge
into <code class="docutils literal notranslate"><span class="pre">A</span></code> 10 and 20 generations ago, respectively:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">initially_active</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population_split</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;B&quot;</span><span class="p">],</span> <span class="n">ancestral</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population_split</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">],</span> <span class="n">ancestral</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">debug</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div><details open="true"><summary>Epoch[0]: [0, 10) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=3 active=3)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>A</th><th>B</th><th>C</th></tr></thead><tbody><tr><td>A</td><td> 100.0</td><td> 100.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population A to B = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from B to A per
generation, divided by the size of A.'>0</td><td title='Backwards in time migration rate from population A to C = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from C to A per
generation, divided by the size of A.'>0</td></tr><tr><td>B</td><td> 100.0</td><td> 100.0</td><td> 0</td><td title='Backwards in time migration rate from population B to A = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from A to B per
generation, divided by the size of B.'>0</td><td>0</td><td title='Backwards in time migration rate from population B to C = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from C to B per
generation, divided by the size of B.'>0</td></tr><tr><td>C</td><td> 100.0</td><td> 100.0</td><td> 0</td><td title='Backwards in time migration rate from population C to A = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from A to C per
generation, divided by the size of C.'>0</td><td title='Backwards in time migration rate from population C to B = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from B to C per
generation, divided by the size of C.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 10</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>10</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[B], ancestral=A</td><td>Moves all lineages from the &#x27;B&#x27; derived population to the ancestral &#x27;A&#x27; population. Also set &#x27;B&#x27; to inactive, and all migration rates to and from the derived population to zero.</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[1]: [10, 20) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=3 active=2)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>A</th><th>C</th></tr></thead><tbody><tr><td>A</td><td> 100.0</td><td> 100.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population A to C = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from C to A per
generation, divided by the size of A.'>0</td></tr><tr><td>C</td><td> 100.0</td><td> 100.0</td><td> 0</td><td title='Backwards in time migration rate from population C to A = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from A to C per
generation, divided by the size of C.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 20</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>20</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[C], ancestral=A</td><td>Moves all lineages from the &#x27;C&#x27; derived population to the ancestral &#x27;A&#x27; population. Also set &#x27;C&#x27; to inactive, and all migration rates to and from the derived population to zero.</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[2]: [20, inf) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=3 active=1)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th></tr></thead><tbody><tr><td>A</td><td> 100.0</td><td> 100.0</td><td> 0</td></tr></tbody></table></div></div></details></div></div></div>
</div>
<p>We can see that the <code class="docutils literal notranslate"><span class="pre">B</span></code> and <code class="docutils literal notranslate"><span class="pre">C</span></code> populations successively become
<a class="reference internal" href="#sec-demography-populations-life-cycle"><span class="std std-ref">inactive</span></a> as lineages
move into <code class="docutils literal notranslate"><span class="pre">A</span></code>.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>For trunk populations like <code class="docutils literal notranslate"><span class="pre">A</span></code> here we <strong>must</strong> set the <code class="docutils literal notranslate"><span class="pre">initially_active</span></code>
flag if we want to draw samples from it at the present day.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>See the <a class="reference internal" href="#sec-demography-examples-trunk-population"><span class="std std-ref">trunk population example</span></a>
of this style of model from the literature.</p>
</div>
</section>
<section id="admixture">
<span id="sec-demography-events-admixture"></span><h4>Admixture<a class="headerlink" href="#admixture" title="Link to this heading">#</a></h4>
<p>The <a class="reference internal" href="api.html#msprime.Demography.add_admixture" title="msprime.Demography.add_admixture"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.add_admixture()</span></code></a> method is used to declare
that a derived population is admixed from a set of ancestral populations.
In the following example we create a demography with 4 populations:
<code class="docutils literal notranslate"><span class="pre">A</span></code>, <code class="docutils literal notranslate"><span class="pre">B</span></code> and <code class="docutils literal notranslate"><span class="pre">ADMIX</span></code> are contemporary populations, and
<code class="docutils literal notranslate"><span class="pre">ANC</span></code> is an ancestral root population. We add an admixture event
in which <code class="docutils literal notranslate"><span class="pre">ADMIX</span></code> is derived from <code class="docutils literal notranslate"><span class="pre">A</span></code> and <code class="docutils literal notranslate"><span class="pre">B</span></code> 10 generations ago,
and <code class="docutils literal notranslate"><span class="pre">A</span></code> and <code class="docutils literal notranslate"><span class="pre">B</span></code> then merge into <code class="docutils literal notranslate"><span class="pre">ANC</span></code>
20 generations ago. The <a class="reference internal" href="#sec-demography-debugging-tools"><span class="std std-ref">debug output</span></a>
can help us understand what this means:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;ADMIX&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;ANC&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_admixture</span><span class="p">(</span>
    <span class="n">time</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="s2">&quot;ADMIX&quot;</span><span class="p">,</span> <span class="n">ancestral</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">],</span> <span class="n">proportions</span><span class="o">=</span><span class="p">[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">])</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population_split</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">],</span> <span class="n">ancestral</span><span class="o">=</span><span class="s2">&quot;ANC&quot;</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">debug</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div><details open="true"><summary>Epoch[0]: [0, 10) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=4 active=3)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>A</th><th>B</th><th>ADMIX</th></tr></thead><tbody><tr><td>A</td><td> 100.0</td><td> 100.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population A to B = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from B to A per
generation, divided by the size of A.'>0</td><td title='Backwards in time migration rate from population A to ADMIX = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ADMIX to A per generation, divided by the size of A.'>0</td></tr><tr><td>B</td><td> 100.0</td><td> 100.0</td><td> 0</td><td title='Backwards in time migration rate from population B to A = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from A to B per
generation, divided by the size of B.'>0</td><td>0</td><td title='Backwards in time migration rate from population B to ADMIX = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ADMIX to B per generation, divided by the size of B.'>0</td></tr><tr><td>ADMIX</td><td> 100.0</td><td> 100.0</td><td> 0</td><td title='Backwards in time migration rate from population ADMIX to A = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from A to ADMIX per generation, divided by the size of ADMIX.'>0</td><td title='Backwards in time migration rate from population ADMIX to B = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from B to ADMIX per generation, divided by the size of ADMIX.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 10</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>10</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_admixture'>Admixture</a></td><td>derived=ADMIX ancestral=[A, B] proportions=[0.25, 0.75]</td><td>Moves all lineages from admixed population &#x27;ADMIX&#x27; to ancestral populations. Lineages move to &#x27;A&#x27; with proba 0.25; &#x27;B&#x27; with proba 0.75. Set &#x27;ADMIX&#x27; to inactive, and all migration rates to and from &#x27;ADMIX&#x27; to zero.</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[1]: [10, 20) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=4 active=2)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>A</th><th>B</th></tr></thead><tbody><tr><td>A</td><td> 100.0</td><td> 100.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population A to B = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from B to A per
generation, divided by the size of A.'>0</td></tr><tr><td>B</td><td> 100.0</td><td> 100.0</td><td> 0</td><td title='Backwards in time migration rate from population B to A = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from A to B per
generation, divided by the size of B.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 20</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>20</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[A, B], ancestral=ANC</td><td>Moves all lineages from derived populations &#x27;A&#x27; and &#x27;B&#x27; to the ancestral &#x27;ANC&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[2]: [20, inf) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=4 active=1)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th></tr></thead><tbody><tr><td>ANC</td><td> 100.0</td><td> 100.0</td><td> 0</td></tr></tbody></table></div></div></details></div></div></div>
</div>
<p>In the first epoch (from 0 to 10 generations ago) we have three active
populations: <code class="docutils literal notranslate"><span class="pre">A</span></code>, <code class="docutils literal notranslate"><span class="pre">B</span></code> and <code class="docutils literal notranslate"><span class="pre">ADMIX</span></code>. Then the admixture event
happens: <a class="reference internal" href="#sec-demography-direction-of-time"><span class="std std-ref">backwards in time</span></a>
we take all the lineages that are in <code class="docutils literal notranslate"><span class="pre">ADMIX</span></code>, and for each one
we move it to <code class="docutils literal notranslate"><span class="pre">A</span></code> with probability 1/4 and to <code class="docutils literal notranslate"><span class="pre">B</span></code> with probability
3/4. In the next epoch (10 to 20 generations ago), we just have
populations <code class="docutils literal notranslate"><span class="pre">A</span></code> and <code class="docutils literal notranslate"><span class="pre">B</span></code> active because the derived <code class="docutils literal notranslate"><span class="pre">ADMIX</span></code>
population is now inactive.</p>
<p>What does this mean forwards in time? In this case, rather than thinking
about sampled lineages moving from population to population we must think
about whole populations and individuals migrating.
So forwards in time, 10 generations ago
the <code class="docutils literal notranslate"><span class="pre">ADMIX</span></code> population suddenly appeared and it contained
100 individuals (note the initial size): 25 of these individuals were
migrants from <code class="docutils literal notranslate"><span class="pre">A</span></code> and the other 75 were migrants from <code class="docutils literal notranslate"><span class="pre">B</span></code>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <code class="docutils literal notranslate"><span class="pre">ANC</span></code> population in this model isn’t important, it’s there
to ensure that it’s possible for lineages to coalesce and
we don’t get an <a class="reference internal" href="ancestry.html#sec-ancestry-errors-infinite-waiting-time"><span class="std std-ref">Infinite waiting time</span></a>.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>See the <a class="reference internal" href="#sec-demography-examples-admixture"><span class="std std-ref">American Admixture example</span></a>
below.</p>
</div>
</section>
<section id="pulse-mass-migration">
<span id="sec-demography-events-mass-migration"></span><h4>Pulse (mass) migration<a class="headerlink" href="#pulse-mass-migration" title="Link to this heading">#</a></h4>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>We do not recommend using mass migration events unless it is
really necessary; most of the time, the desired effect can
be achieved using <a class="reference internal" href="#sec-demography-events-population-split"><span class="std std-ref">Population split</span></a>
and <a class="reference internal" href="#sec-demography-events-admixture"><span class="std std-ref">Admixture</span></a> events. If you are
<a class="reference internal" href="legacy.html#sec-legacy-0x"><span class="std std-ref">msprime 0.x</span></a> user who is accustomed to
working with mass migrations or updating legacy code, please
consider “migrating” to using these new, more specific, events.</p>
</div>
<p>The <a class="reference internal" href="api.html#msprime.Demography.add_mass_migration" title="msprime.Demography.add_mass_migration"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.add_mass_migration()</span></code></a> method is used to
move a proportion of the lineages currently
in one population (the <code class="docutils literal notranslate"><span class="pre">source</span></code>) to another (the <code class="docutils literal notranslate"><span class="pre">dest</span></code>).</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>There are two things it’s vitally important to realise about mass migrations:</p>
<ol class="arabic simple">
<li><p>Like <a class="reference internal" href="#sec-demography-migration"><span class="std std-ref">continuous migration</span></a> the source
and destination populations are from the perspective of the coalescent
process, that is <a class="reference internal" href="#sec-demography-direction-of-time"><span class="std std-ref">backwards in time</span></a>.
So, if a lineage is moved from
<code class="docutils literal notranslate"><span class="pre">source</span></code> population <code class="docutils literal notranslate"><span class="pre">A</span></code> to <code class="docutils literal notranslate"><span class="pre">dest</span></code> population <code class="docutils literal notranslate"><span class="pre">B</span></code> by msprime, this
corresponds to an individual moving <strong>from</strong> <code class="docutils literal notranslate"><span class="pre">B</span></code> <strong>to</strong> <code class="docutils literal notranslate"><span class="pre">A</span></code>,
forwards in time. (Yes, this is
<a class="reference internal" href="#sec-demography-migration"><span class="std std-ref">confusing</span></a>!)</p></li>
<li><p>Mass migration events <strong>do not</strong> alter the migration rates between
populations. So, even if all lineages are moved out of a
particular population (i.e., if <code class="docutils literal notranslate"><span class="pre">proportion=1</span></code>) that population
may still have lineages migrating into it after the event. This
can easily lead to errors when modelling population splits using
<a class="reference internal" href="legacy.html#msprime.MassMigration" title="msprime.MassMigration"><code class="xref py py-class docutils literal notranslate"><span class="pre">MassMigration</span></code></a> events. This error was present in the
documentation for an older version of msprime, which lead to an
incorrect version of an example demographic model being used
in a number of publications. See
<a class="reference external" href="http://dx.doi.org/10.1016/j.ajhg.2020.08.017">this paper</a> for
more details.</p></li>
</ol>
</div>
<p>The effect of mass migration events are summarised in the
<a class="reference internal" href="api.html#msprime.DemographyDebugger" title="msprime.DemographyDebugger"><code class="xref py py-class docutils literal notranslate"><span class="pre">DemographyDebugger</span></code></a> output:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">island_model</span><span class="p">([</span><span class="mi">100</span><span class="p">,</span> <span class="mi">100</span><span class="p">],</span> <span class="n">migration_rate</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_mass_migration</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">source</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">proportion</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">debug</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div><details open="true"><summary>Epoch[0]: [0, 10) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=2 active=2)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>pop_0</th><th>pop_1</th></tr></thead><tbody><tr><td>pop_0</td><td> 100.0</td><td> 100.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population pop_0 to pop_1 = 0.1 per
generation. Forwards in time, this is the expected number of migrants moving
from pop_1 to pop_0 per generation, divided by the size of pop_0.'>0.1</td></tr><tr><td>pop_1</td><td> 100.0</td><td> 100.0</td><td> 0</td><td title='Backwards in time migration rate from population pop_1 to pop_0 = 0.1 per
generation. Forwards in time, this is the expected number of migrants moving
from pop_0 to pop_1 per generation, divided by the size of pop_1.'>0.1</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 10</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>10</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_mass_migration'>Mass Migration</a></td><td>source=0, dest=1, proportion=0.5</td><td>Lineages currently in population 0 move to 1 with probability 0.5 (equivalent to individuals migrating from 1 to 0 forwards in time)</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[1]: [10, inf) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=2 active=2)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>pop_0</th><th>pop_1</th></tr></thead><tbody><tr><td>pop_0</td><td> 100.0</td><td> 100.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population pop_0 to pop_1 = 0.1 per
generation. Forwards in time, this is the expected number of migrants moving
from pop_1 to pop_0 per generation, divided by the size of pop_0.'>0.1</td></tr><tr><td>pop_1</td><td> 100.0</td><td> 100.0</td><td> 0</td><td title='Backwards in time migration rate from population pop_1 to pop_0 = 0.1 per
generation. Forwards in time, this is the expected number of migrants moving
from pop_0 to pop_1 per generation, divided by the size of pop_1.'>0.1</td><td>0</td></tr></tbody></table></div></div></details></div></div></div>
</div>
</section>
<section id="instantaneous-bottleneck">
<span id="sec-demography-events-instantaneous-bottleneck"></span><h4>Instantaneous bottleneck<a class="headerlink" href="#instantaneous-bottleneck" title="Link to this heading">#</a></h4>
<p>A common approach to modelling the effect of demographic history on genealogies
is to assume that effective population size (<span class="math notranslate nohighlight">\(N_e\)</span>) changes in discrete steps
which define a series of epochs.
In this setting of piece-wise constant
<span class="math notranslate nohighlight">\(N_e\)</span>, capturing a population bottleneck requires three epochs: <span class="math notranslate nohighlight">\(N_e\)</span> is
reduced by some fraction <span class="math notranslate nohighlight">\(b\)</span> at the start of the bottleneck,
<span class="math notranslate nohighlight">\(T_{start}\)</span>, and
recovers to its initial value at time <span class="math notranslate nohighlight">\(T_{end}\)</span> (Marth et al 2004).
If bottlenecks are short both on the timescale of coalescence and mutations, one
may expect little information about the duration of a bottleneck
<span class="math notranslate nohighlight">\((T_{end}-T_{start})\)</span> in sequence data.
Thus a simpler, alternative model is to assume that
bottlenecks are instantaneous
(<span class="math notranslate nohighlight">\(T_{end}-T_{start} \rightarrow 0\)</span>) and generate
a sudden burst of coalescence events in the genealogy.</p>
<p>The strength of the bottleneck <span class="math notranslate nohighlight">\(B\)</span> can be thought of as an (imaginary)
time period during which coalescence events are collapsed, i.e. there is no
growth in genealogical branches during <span class="math notranslate nohighlight">\(B\)</span> and the probability
that a single pair
of lineages entering the bottleneck coalesce during the bottleneck is
<span class="math notranslate nohighlight">\(1-e^{-B}\)</span>.</p>
<p>In the following example we create a population of size 100,
and then 25 generations ago create a bottleneck equivalent to
200 generations of the coalescent using the
<a class="reference internal" href="api.html#msprime.Demography.add_instantaneous_bottleneck" title="msprime.Demography.add_instantaneous_bottleneck"><code class="xref py py-meth docutils literal notranslate"><span class="pre">add_instantaneous_bottleneck()</span></code></a>
method:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_instantaneous_bottleneck</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">strength</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">population</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">1234</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">300</span><span class="p">,</span> <span class="mi">200</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/03ac76ff75eed3070a286efbdbc8c703025e77beac79a9aa251c912610f224d3.svg" src="_images/03ac76ff75eed3070a286efbdbc8c703025e77beac79a9aa251c912610f224d3.svg" />
</div>
</div>
<p>We can see that there is a strong burst of coalescence 25 generations ago.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>Note that there is no possibility of recombination, migration or any other
events occuring during the bottleneck: it is strictly <em>instantaneous</em>
and can only result in coalescences within the trees.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>See <a class="reference external" href="https://doi.org/10.1534/genetics.115.179861">Bunnefeld et al. (2015)</a>
and <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1461106/">Galtier et al. (2000)</a>
for more details on this model of population bottlenecks.</p>
</div>
</section>
<section id="simple-bottleneck">
<span id="sec-demography-events-simple-bottleneck"></span><h4>Simple bottleneck<a class="headerlink" href="#simple-bottleneck" title="Link to this heading">#</a></h4>
<p>In a <a class="reference internal" href="api.html#msprime.Demography.add_simple_bottleneck" title="msprime.Demography.add_simple_bottleneck"><code class="xref py py-meth docutils literal notranslate"><span class="pre">simple</span> <span class="pre">bottleneck</span></code></a>
each lineage in a given population has a specified
probability of coalescing into a single ancestral lineage and a given
time. For example, here we create a bottleneck in which lineages
have a probability of 0.8 of coalescing at time 1:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_simple_bottleneck</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">population</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">proportion</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>

<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">243</span><span class="p">)</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/7d2dd87f134a0fba8f77c0175b0f037e43482553d822a921cafd61ed97145d84.svg" src="_images/7d2dd87f134a0fba8f77c0175b0f037e43482553d822a921cafd61ed97145d84.svg" />
</div>
</div>
<p>We can see that at time 1 we had 7 extant lineages, and all of these
happened to coalesce, leading to a large polytomy at the root.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <a class="reference internal" href="api.html#msprime.Demography.add_simple_bottleneck" title="msprime.Demography.add_simple_bottleneck"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.add_simple_bottleneck()</span></code></a> method can be useful
for creating trees with particular properties, but it is not
based on any particular theoretical model and is unlikely to be
useful in modelling real populations.</p>
</div>
</section>
</section>
</section>
<section id="demography-objects">
<span id="sec-demography-demography-objects"></span><h2>Demography objects<a class="headerlink" href="#demography-objects" title="Link to this heading">#</a></h2>
<p><a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> objects provide a number of convenient ways of
accessing individual population data. Consider the following three
population example:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">demography</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (3)</b><table border="1" class="tskit-table"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>A</td><td></td><td>1.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>B</td><td></td><td>2.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>2</td><td>C</td><td></td><td>3.0</td><td>0</td><td>0</td><td>{}</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Migration matrix (all zero)</b><table border="1" class="tskit-table"><thead><tr></tr></thead><tbody></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events (0)</b><table border="1" class="tskit-table"><thead><tr></tr></thead><tbody></tbody></table></div></div></div></div>
</div>
<p>Demography objects are essentially a mapping from population identifiers
to the corresponding population objects, and behave like
a <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><code class="docutils literal notranslate"><span class="pre">dict</span></code></a> with some extra methods. The additional
feature provided by the <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> class is that populations
can be found either by their integer ID their string names
(see the <a class="reference internal" href="#sec-demography-populations-identifiers"><span class="std std-ref">Identifying populations</span></a> section for more
details):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">demography</span><span class="p">[</span><span class="s2">&quot;B&quot;</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">demography</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Population(initial_size=2, growth_rate=0, name=&#39;B&#39;, description=&#39;&#39;, extra_metadata={}, default_sampling_time=None, initially_active=None, id=1)
Population(initial_size=2, growth_rate=0, name=&#39;B&#39;, description=&#39;&#39;, extra_metadata={}, default_sampling_time=None, initially_active=None, id=1)
</pre></div>
</div>
</div>
</div>
<p>Note that the <code class="docutils literal notranslate"><span class="pre">keys</span></code> for a Demography are the names:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">list</span><span class="p">(</span><span class="n">demography</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&#39;A&#39;, &#39;B&#39;, &#39;C&#39;]
</pre></div>
</div>
</div>
</div>
<p>The values for population objects can be updated in-place. For example,
suppose we wanted to change the <a class="reference internal" href="#sec-demography-populations-growth-rate"><span class="std std-ref">Growth rate</span></a>
for <code class="docutils literal notranslate"><span class="pre">A</span></code> to 0.001:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">growth_rate</span> <span class="o">=</span> <span class="mf">0.001</span>
<span class="n">demography</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (3)</b><table border="1" class="tskit-table"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>A</td><td></td><td>1.0</td><td>0.001</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>B</td><td></td><td>2.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>2</td><td>C</td><td></td><td>3.0</td><td>0</td><td>0</td><td>{}</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Migration matrix (all zero)</b><table border="1" class="tskit-table"><thead><tr></tr></thead><tbody></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events (0)</b><table border="1" class="tskit-table"><thead><tr></tr></thead><tbody></tbody></table></div></div></div></div>
</div>
<p>We can also loop over all populations using the standard dict protocols.
For example, if we wanted to multiply all population sizes by 100 we
could do:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">demography</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
    <span class="n">pop</span><span class="o">.</span><span class="n">initial_size</span> <span class="o">*=</span> <span class="mi">100</span>
<span class="n">demography</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (3)</b><table border="1" class="tskit-table"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>A</td><td></td><td>100.0</td><td>0.001</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>B</td><td></td><td>200.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>2</td><td>C</td><td></td><td>300.0</td><td>0</td><td>0</td><td>{}</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Migration matrix (all zero)</b><table border="1" class="tskit-table"><thead><tr></tr></thead><tbody></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events (0)</b><table border="1" class="tskit-table"><thead><tr></tr></thead><tbody></tbody></table></div></div></div></div>
</div>
</section>
<section id="debugging-tools">
<span id="sec-demography-debugging-tools"></span><h2>Debugging tools<a class="headerlink" href="#debugging-tools" title="Link to this heading">#</a></h2>
<p>Creating demographic models is a difficult and
<a class="reference external" href="http://dx.doi.org/10.1016/j.ajhg.2020.08.017">error prone</a> process
and we need all the help we can get to avoid making mistakes.
Msprime provides some basic debugging utilities to help understand
and inspect the contents of <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> objects, as well
as understand how the model parameters change over time.</p>
<section id="id2">
<h3>Demography objects<a class="headerlink" href="#id2" title="Link to this heading">#</a></h3>
<p>When working in a Jupyter notebook environment it is very easy to
get a summary of the state of a Demography object: using Jupyter
magic, we write the variable at the end of a cell and it gets
summarised automatically:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define a demography</span>
<span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">island_model</span><span class="p">([</span><span class="mi">100</span><span class="p">,</span> <span class="mi">100</span><span class="p">],</span> <span class="n">migration_rate</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
<span class="c1"># Ask Jupyter to print out a summary of it.</span>
<span class="n">demography</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (2)</b><table border="1" class="tskit-table"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>pop_0</td><td></td><td>100.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>pop_1</td><td></td><td>100.0</td><td>0</td><td>0</td><td>{}</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Migration matrix</b><table border="1" class="tskit-table"><thead><tr><th></th><th>pop_0</th><th>pop_1</th></tr></thead><tbody><tr><td>pop_0</td><td>0</td><td title='Backwards in time migration rate from population pop_0 to pop_1 = 0.1 per
generation. Forwards in time, this is the expected number of migrants moving
from pop_1 to pop_0 per generation, divided by the size of pop_0.'>0.1</td></tr><tr><td>pop_1</td><td title='Backwards in time migration rate from population pop_1 to pop_0 = 0.1 per
generation. Forwards in time, this is the expected number of migrants moving
from pop_0 to pop_1 per generation, divided by the size of pop_1.'>0.1</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events (0)</b><table border="1" class="tskit-table"><thead><tr></tr></thead><tbody></tbody></table></div></div></div></div>
</div>
<p>If you prefer to work in a terminal or other non-notebook environment
the same information can be accessed by converting the demography object
to a string, e.g., by calling <code class="docutils literal notranslate"><span class="pre">print</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">demography</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Demography
╟  Populations
║  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
║  │ id │name   │description  │initial_size  │ growth_rate │  default_sampling_time│extra_metadata  │
║  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
║  │ 0  │pop_0  │             │100.0         │      0      │                      0│{}              │
║  │ 1  │pop_1  │             │100.0         │      0      │                      0│{}              │
║  └────────────────────────────────────────────────────────────────────────────────────────────────┘
╟  Migration Matrix
║  ┌───────────────────────┐
║  │       │ pop_0 │ pop_1 │
║  ├───────────────────────┤
║  │  pop_0│   0   │  0.1  │
║  │  pop_1│  0.1  │   0   │
║  └───────────────────────┘
╟  Events
║  ┌───────────────────────────────────┐
║  │  time│type  │parameters  │effect  │
║  ├───────────────────────────────────┤
║  └───────────────────────────────────┘
</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This may not render correctly in all browsers because of the lack of
support for fixed width fonts.</p>
</div>
</section>
<section id="demographydebugger">
<h3>DemographyDebugger<a class="headerlink" href="#demographydebugger" title="Link to this heading">#</a></h3>
<p>The <a class="reference internal" href="api.html#msprime.DemographyDebugger" title="msprime.DemographyDebugger"><code class="xref py py-class docutils literal notranslate"><span class="pre">DemographyDebugger</span></code></a> is a very useful tool for understanding
how a model changes over time. For example, let’s consider a simple
population tree:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">10_000</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">5_000</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">1_000</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population_split</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">],</span> <span class="n">ancestral</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">debug</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div><details open="true"><summary>Epoch[0]: [0, 1e+03) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=3 active=2)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>A</th><th>B</th></tr></thead><tbody><tr><td>A</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population A to B = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from B to A per
generation, divided by the size of A.'>0</td></tr><tr><td>B</td><td> 5000.0</td><td> 5000.0</td><td> 0</td><td title='Backwards in time migration rate from population B to A = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from A to B per
generation, divided by the size of B.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 1e+03</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>1000</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[A, B], ancestral=C</td><td>Moves all lineages from derived populations &#x27;A&#x27; and &#x27;B&#x27; to the ancestral &#x27;C&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[1]: [1e+03, inf) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=3 active=1)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th></tr></thead><tbody><tr><td>C</td><td> 1000.0</td><td> 1000.0</td><td> 0</td></tr></tbody></table></div></div></details></div></div></div>
</div>
<p>Here we define a <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> object and then get a visual
summary of the evolution of the model over time by calling
the <a class="reference internal" href="api.html#msprime.Demography.debug" title="msprime.Demography.debug"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.debug()</span></code></a> method (which returns a
<a class="reference internal" href="api.html#msprime.DemographyDebugger" title="msprime.DemographyDebugger"><code class="xref py py-class docutils literal notranslate"><span class="pre">DemographyDebugger</span></code></a> instance).</p>
<p>The debugger splits time up in to simulation “epochs” during which all the
parameters are the same, and no events happen (as everywhere in msprime,
we work <a class="reference internal" href="#sec-demography-direction-of-time"><span class="std std-ref">backwards in time</span></a>). In
each epoch we give the state of the population parameters and migration
rates for all <a class="reference internal" href="#sec-demography-populations-life-cycle"><span class="std std-ref">active</span></a>
populations. We also show the events that signal the end of the epoch,
which create the next epoch.</p>
<p>Again, the same information is available for non-notebook users by
converting the debugger object to string:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">demography</span><span class="o">.</span><span class="n">debug</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>DemographyDebugger
╠══════════════════════════════════╗
║ Epoch[0]: [0, 1e+03) generations ║
╠══════════════════════════════════╝
╟    Populations (total=3 active=2)
║    ┌───────────────────────────────────────────────┐
║    │   │     start│       end│growth_rate  │ A │ B │
║    ├───────────────────────────────────────────────┤
║    │  A│   10000.0│   10000.0│ 0           │ 0 │ 0 │
║    │  B│    5000.0│    5000.0│ 0           │ 0 │ 0 │
║    └───────────────────────────────────────────────┘
╟    Events @ generation 1e+03
║    ┌──────────────────────────────────────────────────────────────────────────────┐
║    │  time│type        │parameters       │effect                                  │
║    ├──────────────────────────────────────────────────────────────────────────────┤
║    │  1000│Population  │derived=[A, B],  │Moves all lineages from derived         │
║    │      │Split       │ancestral=C      │populations &#39;A&#39; and &#39;B&#39; to the          │
║    │      │            │                 │ancestral &#39;C&#39; population. Also set the  │
║    │      │            │                 │derived populations to inactive, and    │
║    │      │            │                 │all migration rates to and from the     │
║    │      │            │                 │derived populations to zero.            │
║    └──────────────────────────────────────────────────────────────────────────────┘
╠════════════════════════════════════╗
║ Epoch[1]: [1e+03, inf) generations ║
╠════════════════════════════════════╝
╟    Populations (total=3 active=1)
║    ┌─────────────────────────────────────┐
║    │   │    start│      end│growth_rate  │
║    ├─────────────────────────────────────┤
║    │  C│   1000.0│   1000.0│ 0           │
║    └─────────────────────────────────────┘
</pre></div>
</div>
</div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>The <a class="reference internal" href="api.html#msprime.DemographyDebugger" title="msprime.DemographyDebugger"><code class="xref py py-class docutils literal notranslate"><span class="pre">DemographyDebugger</span></code></a> has a number of methods to
perform calculations based on a given demography. See the
<a class="reference internal" href="#sec-demography-numerical"><span class="std std-ref">Numerical utilities</span></a> section for details and examples.</p>
</div>
</section>
</section>
<section id="example-models">
<span id="sec-demography-examples"></span><h2>Example models<a class="headerlink" href="#example-models" title="Link to this heading">#</a></h2>
<p>In this section we show how to build demographic models by working
through some example models from the literature.</p>
<section id="population-tree">
<span id="sec-demography-examples-population-tree"></span><h3>Population tree<a class="headerlink" href="#population-tree" title="Link to this heading">#</a></h3>
<p>The first example we examine is the
<a class="reference external" href="http://dx.doi.org/10.1371/journal.pgen.1000695">Gutenkunst et al.</a>
out-of-Africa model, in which the history of three present-day populations
is modelled by a tree of ancestral populations.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>If you want to use this precise model in your analyses
we strongly recommend using <a class="reference external" href="https://popsim-consortium.github.io/stdpopsim-docs/stable/introduction.html#sec-introduction" title="(in stdpopsim v0.3)"><span class="xref std std-ref">stdpopsim</span></a>,
which provides a community maintained <a class="reference external" href="https://popsim-consortium.github.io/stdpopsim-docs/stable/catalog.html#sec-catalog" title="(in stdpopsim v0.3)"><span class="xref std std-ref">catalog</span></a>
of species information and demographic models for simulation. The
model given here is equivalent to the
<a class="reference external" href="https://popsim-consortium.github.io/stdpopsim-docs/stable/catalog.html#sec_catalog_homsap_models_outofafrica_3g09" title="(in stdpopsim v0.3)"><span class="xref std std-ref">HomSap/OutOfAfrica_3G09</span></a>
model.</p>
</div>
<p>First we declare some variables and compute some quantities that will
come in useful:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">math</span>

<span class="c1"># Times are provided in years, so we convert into generations.</span>
<span class="n">generation_time</span> <span class="o">=</span> <span class="mi">25</span>
<span class="n">T_OOA</span> <span class="o">=</span> <span class="mf">21.2e3</span> <span class="o">/</span> <span class="n">generation_time</span>
<span class="n">T_AMH</span> <span class="o">=</span> <span class="mf">140e3</span> <span class="o">/</span> <span class="n">generation_time</span>
<span class="n">T_ANC</span> <span class="o">=</span> <span class="mf">220e3</span> <span class="o">/</span> <span class="n">generation_time</span>
<span class="c1"># We need to work out the starting population sizes based on</span>
<span class="c1"># the growth rates provided for these two populations</span>
<span class="n">r_CEU</span> <span class="o">=</span> <span class="mf">0.004</span>
<span class="n">r_CHB</span> <span class="o">=</span> <span class="mf">0.0055</span>
<span class="n">N_CEU</span> <span class="o">=</span> <span class="mi">1000</span> <span class="o">/</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="n">r_CEU</span> <span class="o">*</span> <span class="n">T_OOA</span><span class="p">)</span>
<span class="n">N_CHB</span> <span class="o">=</span> <span class="mi">510</span> <span class="o">/</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="n">r_CHB</span> <span class="o">*</span> <span class="n">T_OOA</span><span class="p">)</span>
<span class="n">N_AFR</span> <span class="o">=</span> <span class="mi">12300</span>
</pre></div>
</div>
</div>
</div>
<p>The time and other variables are declared here
following the <a class="reference external" href="https://en.wikipedia.org/wiki/Don't_repeat_yourself">DRY</a>
principle: when we have to use the same numerical value more than
once it’s a very good idea to use a variable to avoid errors. The
time conversions are needed because time in msprime is always measured
in generations, and the paper reported times in units of years.
The size computations are required because the paper reported
the European and Asian population sizes at the time they split
from the Out-of-Africa population, but we need the present-day
sizes.</p>
<p>Once we have computed these quantities we can declare the
<a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> object representing the model we are
building and add some <a class="reference internal" href="#sec-demography-populations"><span class="std std-ref">Populations</span></a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;YRI&quot;</span><span class="p">,</span>
    <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Yoruba in Ibadan, Nigeria&quot;</span><span class="p">,</span>
    <span class="n">initial_size</span><span class="o">=</span><span class="n">N_AFR</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;CEU&quot;</span><span class="p">,</span>
    <span class="n">description</span><span class="o">=</span><span class="p">(</span>
        <span class="s2">&quot;Utah Residents (CEPH) with Northern and Western European Ancestry&quot;</span>
    <span class="p">),</span>
    <span class="n">initial_size</span><span class="o">=</span><span class="n">N_CEU</span><span class="p">,</span>
    <span class="n">growth_rate</span><span class="o">=</span><span class="n">r_CEU</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;CHB&quot;</span><span class="p">,</span>
    <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Han Chinese in Beijing, China&quot;</span><span class="p">,</span>
    <span class="n">initial_size</span><span class="o">=</span><span class="n">N_CHB</span><span class="p">,</span>
    <span class="n">growth_rate</span><span class="o">=</span><span class="n">r_CHB</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;OOA&quot;</span><span class="p">,</span>
    <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Bottleneck out-of-Africa population&quot;</span><span class="p">,</span>
    <span class="n">initial_size</span><span class="o">=</span><span class="mi">2100</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;AMH&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Anatomically modern humans&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="n">N_AFR</span>
<span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;ANC&quot;</span><span class="p">,</span>
    <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Ancestral equilibrium population&quot;</span><span class="p">,</span>
    <span class="n">initial_size</span><span class="o">=</span><span class="mi">7300</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">demography</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (6)</b><table border="1" class="tskit-table"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>YRI</td><td>Yoruba in Ibadan, Nigeria</td><td>12300.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>CEU</td><td>Utah Residents (CEPH) with Northern and Western European Ancestry</td><td>29725.3</td><td>0.004</td><td>0</td><td>{}</td></tr><tr><td>2</td><td>CHB</td><td>Han Chinese in Beijing, China</td><td>54090.3</td><td>0.0055</td><td>0</td><td>{}</td></tr><tr><td>3</td><td>OOA</td><td>Bottleneck out-of-Africa population</td><td>2100.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>4</td><td>AMH</td><td>Anatomically modern humans</td><td>12300.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>5</td><td>ANC</td><td>Ancestral equilibrium population</td><td>7300.0</td><td>0</td><td>0</td><td>{}</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Migration matrix (all zero)</b><table border="1" class="tskit-table"><thead><tr></tr></thead><tbody></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events (0)</b><table border="1" class="tskit-table"><thead><tr></tr></thead><tbody></tbody></table></div></div></div></div>
</div>
<p>Here we create a new <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> object and add six populations
using the <a class="reference internal" href="api.html#msprime.Demography.add_population" title="msprime.Demography.add_population"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.add_population()</span></code></a> method. The first three
populations represent present-day populations and the others ancestral
populations. For convenience, we define the
<a class="reference internal" href="#sec-demography-populations-initial-size"><span class="std std-ref">Initial size</span></a> of the ancestral populations
here, but technically those populations won’t be
<a class="reference internal" href="#sec-demography-populations-life-cycle"><span class="std std-ref">active</span></a> until later in the
simulation (i.e., further back in time). To check that we have correctly
set up the starting states of the populations, we can look at the
HTML representation using Jupyter magic.</p>
<p>This sets up our populations, but there is no <a class="reference internal" href="#sec-demography-migration"><span class="std std-ref">Migration</span></a>
between any of them, as we can see in the output migration matrix.
We can add migration to the model using the
<a class="reference internal" href="api.html#msprime.Demography.set_symmetric_migration_rate" title="msprime.Demography.set_symmetric_migration_rate"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.set_symmetric_migration_rate()</span></code></a> method (we can also
set asymmetric migration rates using <a class="reference internal" href="api.html#msprime.Demography.set_migration_rate" title="msprime.Demography.set_migration_rate"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.set_migration_rate()</span></code></a>,
but beware of the <a class="reference internal" href="#sec-demography-migration"><span class="std std-ref">confusion</span></a> around the
meaning of <code class="docutils literal notranslate"><span class="pre">source</span></code> and <code class="docutils literal notranslate"><span class="pre">dest</span></code>).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Set the migration rates between extant populations</span>
<span class="n">demography</span><span class="o">.</span><span class="n">set_symmetric_migration_rate</span><span class="p">([</span><span class="s2">&quot;CEU&quot;</span><span class="p">,</span> <span class="s2">&quot;CHB&quot;</span><span class="p">],</span> <span class="mf">9.6e-5</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">set_symmetric_migration_rate</span><span class="p">([</span><span class="s2">&quot;YRI&quot;</span><span class="p">,</span> <span class="s2">&quot;CHB&quot;</span><span class="p">],</span> <span class="mf">1.9e-5</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">set_symmetric_migration_rate</span><span class="p">([</span><span class="s2">&quot;YRI&quot;</span><span class="p">,</span> <span class="s2">&quot;CEU&quot;</span><span class="p">],</span> <span class="mf">3e-5</span><span class="p">)</span>
<span class="n">demography</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (6)</b><table border="1" class="tskit-table"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>YRI</td><td>Yoruba in Ibadan, Nigeria</td><td>12300.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>CEU</td><td>Utah Residents (CEPH) with Northern and Western European Ancestry</td><td>29725.3</td><td>0.004</td><td>0</td><td>{}</td></tr><tr><td>2</td><td>CHB</td><td>Han Chinese in Beijing, China</td><td>54090.3</td><td>0.0055</td><td>0</td><td>{}</td></tr><tr><td>3</td><td>OOA</td><td>Bottleneck out-of-Africa population</td><td>2100.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>4</td><td>AMH</td><td>Anatomically modern humans</td><td>12300.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>5</td><td>ANC</td><td>Ancestral equilibrium population</td><td>7300.0</td><td>0</td><td>0</td><td>{}</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Migration matrix</b><table border="1" class="tskit-table"><thead><tr><th></th><th>YRI</th><th>CEU</th><th>CHB</th><th>OOA</th><th>AMH</th><th>ANC</th></tr></thead><tbody><tr><td>YRI</td><td>0</td><td title='Backwards in time migration rate from population YRI to CEU = 3e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from CEU to YRI per generation, divided by the size of YRI.'>3e-05</td><td title='Backwards in time migration rate from population YRI to CHB = 1.9e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from CHB to YRI per generation, divided by the size of YRI.'>1.9e-05</td><td title='Backwards in time migration rate from population YRI to OOA = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from OOA to YRI per generation, divided by the size of YRI.'>0</td><td title='Backwards in time migration rate from population YRI to AMH = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from AMH to YRI per generation, divided by the size of YRI.'>0</td><td title='Backwards in time migration rate from population YRI to ANC = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ANC to YRI per generation, divided by the size of YRI.'>0</td></tr><tr><td>CEU</td><td title='Backwards in time migration rate from population CEU to YRI = 3e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from YRI to CEU per generation, divided by the size of CEU.'>3e-05</td><td>0</td><td title='Backwards in time migration rate from population CEU to CHB = 9.6e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from CHB to CEU per generation, divided by the size of CEU.'>9.6e-05</td><td title='Backwards in time migration rate from population CEU to OOA = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from OOA to CEU per generation, divided by the size of CEU.'>0</td><td title='Backwards in time migration rate from population CEU to AMH = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from AMH to CEU per generation, divided by the size of CEU.'>0</td><td title='Backwards in time migration rate from population CEU to ANC = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ANC to CEU per generation, divided by the size of CEU.'>0</td></tr><tr><td>CHB</td><td title='Backwards in time migration rate from population CHB to YRI = 1.9e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from YRI to CHB per generation, divided by the size of CHB.'>1.9e-05</td><td title='Backwards in time migration rate from population CHB to CEU = 9.6e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from CEU to CHB per generation, divided by the size of CHB.'>9.6e-05</td><td>0</td><td title='Backwards in time migration rate from population CHB to OOA = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from OOA to CHB per generation, divided by the size of CHB.'>0</td><td title='Backwards in time migration rate from population CHB to AMH = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from AMH to CHB per generation, divided by the size of CHB.'>0</td><td title='Backwards in time migration rate from population CHB to ANC = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ANC to CHB per generation, divided by the size of CHB.'>0</td></tr><tr><td>OOA</td><td title='Backwards in time migration rate from population OOA to YRI = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from YRI to OOA per generation, divided by the size of OOA.'>0</td><td title='Backwards in time migration rate from population OOA to CEU = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from CEU to OOA per generation, divided by the size of OOA.'>0</td><td title='Backwards in time migration rate from population OOA to CHB = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from CHB to OOA per generation, divided by the size of OOA.'>0</td><td>0</td><td title='Backwards in time migration rate from population OOA to AMH = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from AMH to OOA per generation, divided by the size of OOA.'>0</td><td title='Backwards in time migration rate from population OOA to ANC = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ANC to OOA per generation, divided by the size of OOA.'>0</td></tr><tr><td>AMH</td><td title='Backwards in time migration rate from population AMH to YRI = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from YRI to AMH per generation, divided by the size of AMH.'>0</td><td title='Backwards in time migration rate from population AMH to CEU = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from CEU to AMH per generation, divided by the size of AMH.'>0</td><td title='Backwards in time migration rate from population AMH to CHB = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from CHB to AMH per generation, divided by the size of AMH.'>0</td><td title='Backwards in time migration rate from population AMH to OOA = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from OOA to AMH per generation, divided by the size of AMH.'>0</td><td>0</td><td title='Backwards in time migration rate from population AMH to ANC = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ANC to AMH per generation, divided by the size of AMH.'>0</td></tr><tr><td>ANC</td><td title='Backwards in time migration rate from population ANC to YRI = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from YRI to ANC per generation, divided by the size of ANC.'>0</td><td title='Backwards in time migration rate from population ANC to CEU = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from CEU to ANC per generation, divided by the size of ANC.'>0</td><td title='Backwards in time migration rate from population ANC to CHB = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from CHB to ANC per generation, divided by the size of ANC.'>0</td><td title='Backwards in time migration rate from population ANC to OOA = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from OOA to ANC per generation, divided by the size of ANC.'>0</td><td title='Backwards in time migration rate from population ANC to AMH = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from AMH to ANC per generation, divided by the size of ANC.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events (0)</b><table border="1" class="tskit-table"><thead><tr></tr></thead><tbody></tbody></table></div></div></div></div>
</div>
<p>Now we can see that the migration matrix shows the
appropriate rates between the extant populations.</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>If you hover your mouse over an entry in the migration matrix
it will tell which populations are involved, and how the rate
can be interpreted.</p>
</div>
<p>The model still does not have any sense of how the modern
and ancient populations relate to each other, or how
parameters change over time. To do this, we need to declare
some <a class="reference internal" href="#sec-demography-events"><span class="std std-ref">Events</span></a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span><span class="o">.</span><span class="n">add_population_split</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="n">T_OOA</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CEU&quot;</span><span class="p">,</span> <span class="s2">&quot;CHB&quot;</span><span class="p">],</span> <span class="n">ancestral</span><span class="o">=</span><span class="s2">&quot;OOA&quot;</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_symmetric_migration_rate_change</span><span class="p">(</span>
    <span class="n">time</span><span class="o">=</span><span class="n">T_OOA</span><span class="p">,</span> <span class="n">populations</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;YRI&quot;</span><span class="p">,</span> <span class="s2">&quot;OOA&quot;</span><span class="p">],</span> <span class="n">rate</span><span class="o">=</span><span class="mf">25e-5</span>
<span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population_split</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="n">T_AMH</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;YRI&quot;</span><span class="p">,</span> <span class="s2">&quot;OOA&quot;</span><span class="p">],</span> <span class="n">ancestral</span><span class="o">=</span><span class="s2">&quot;AMH&quot;</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population_split</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="n">T_ANC</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;AMH&quot;</span><span class="p">],</span> <span class="n">ancestral</span><span class="o">=</span><span class="s2">&quot;ANC&quot;</span><span class="p">)</span>
<span class="n">demography</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (6)</b><table border="1" class="tskit-table"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>YRI</td><td>Yoruba in Ibadan, Nigeria</td><td>12300.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>CEU</td><td>Utah Residents (CEPH) with Northern and Western European Ancestry</td><td>29725.3</td><td>0.004</td><td>0</td><td>{}</td></tr><tr><td>2</td><td>CHB</td><td>Han Chinese in Beijing, China</td><td>54090.3</td><td>0.0055</td><td>0</td><td>{}</td></tr><tr><td>3</td><td>OOA</td><td>Bottleneck out-of-Africa population</td><td>2100.0</td><td>0</td><td>8.5e+02</td><td>{}</td></tr><tr><td>4</td><td>AMH</td><td>Anatomically modern humans</td><td>12300.0</td><td>0</td><td>5.6e+03</td><td>{}</td></tr><tr><td>5</td><td>ANC</td><td>Ancestral equilibrium population</td><td>7300.0</td><td>0</td><td>8.8e+03</td><td>{}</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Migration matrix</b><table border="1" class="tskit-table"><thead><tr><th></th><th>YRI</th><th>CEU</th><th>CHB</th><th>OOA</th><th>AMH</th><th>ANC</th></tr></thead><tbody><tr><td>YRI</td><td>0</td><td title='Backwards in time migration rate from population YRI to CEU = 3e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from CEU to YRI per generation, divided by the size of YRI.'>3e-05</td><td title='Backwards in time migration rate from population YRI to CHB = 1.9e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from CHB to YRI per generation, divided by the size of YRI.'>1.9e-05</td><td title='Backwards in time migration rate from population YRI to OOA = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from OOA to YRI per generation, divided by the size of YRI.'>0</td><td title='Backwards in time migration rate from population YRI to AMH = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from AMH to YRI per generation, divided by the size of YRI.'>0</td><td title='Backwards in time migration rate from population YRI to ANC = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ANC to YRI per generation, divided by the size of YRI.'>0</td></tr><tr><td>CEU</td><td title='Backwards in time migration rate from population CEU to YRI = 3e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from YRI to CEU per generation, divided by the size of CEU.'>3e-05</td><td>0</td><td title='Backwards in time migration rate from population CEU to CHB = 9.6e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from CHB to CEU per generation, divided by the size of CEU.'>9.6e-05</td><td title='Backwards in time migration rate from population CEU to OOA = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from OOA to CEU per generation, divided by the size of CEU.'>0</td><td title='Backwards in time migration rate from population CEU to AMH = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from AMH to CEU per generation, divided by the size of CEU.'>0</td><td title='Backwards in time migration rate from population CEU to ANC = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ANC to CEU per generation, divided by the size of CEU.'>0</td></tr><tr><td>CHB</td><td title='Backwards in time migration rate from population CHB to YRI = 1.9e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from YRI to CHB per generation, divided by the size of CHB.'>1.9e-05</td><td title='Backwards in time migration rate from population CHB to CEU = 9.6e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from CEU to CHB per generation, divided by the size of CHB.'>9.6e-05</td><td>0</td><td title='Backwards in time migration rate from population CHB to OOA = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from OOA to CHB per generation, divided by the size of CHB.'>0</td><td title='Backwards in time migration rate from population CHB to AMH = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from AMH to CHB per generation, divided by the size of CHB.'>0</td><td title='Backwards in time migration rate from population CHB to ANC = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ANC to CHB per generation, divided by the size of CHB.'>0</td></tr><tr><td>OOA</td><td title='Backwards in time migration rate from population OOA to YRI = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from YRI to OOA per generation, divided by the size of OOA.'>0</td><td title='Backwards in time migration rate from population OOA to CEU = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from CEU to OOA per generation, divided by the size of OOA.'>0</td><td title='Backwards in time migration rate from population OOA to CHB = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from CHB to OOA per generation, divided by the size of OOA.'>0</td><td>0</td><td title='Backwards in time migration rate from population OOA to AMH = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from AMH to OOA per generation, divided by the size of OOA.'>0</td><td title='Backwards in time migration rate from population OOA to ANC = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ANC to OOA per generation, divided by the size of OOA.'>0</td></tr><tr><td>AMH</td><td title='Backwards in time migration rate from population AMH to YRI = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from YRI to AMH per generation, divided by the size of AMH.'>0</td><td title='Backwards in time migration rate from population AMH to CEU = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from CEU to AMH per generation, divided by the size of AMH.'>0</td><td title='Backwards in time migration rate from population AMH to CHB = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from CHB to AMH per generation, divided by the size of AMH.'>0</td><td title='Backwards in time migration rate from population AMH to OOA = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from OOA to AMH per generation, divided by the size of AMH.'>0</td><td>0</td><td title='Backwards in time migration rate from population AMH to ANC = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ANC to AMH per generation, divided by the size of AMH.'>0</td></tr><tr><td>ANC</td><td title='Backwards in time migration rate from population ANC to YRI = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from YRI to ANC per generation, divided by the size of ANC.'>0</td><td title='Backwards in time migration rate from population ANC to CEU = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from CEU to ANC per generation, divided by the size of ANC.'>0</td><td title='Backwards in time migration rate from population ANC to CHB = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from CHB to ANC per generation, divided by the size of ANC.'>0</td><td title='Backwards in time migration rate from population ANC to OOA = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from OOA to ANC per generation, divided by the size of ANC.'>0</td><td title='Backwards in time migration rate from population ANC to AMH = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from AMH to ANC per generation, divided by the size of ANC.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events (4)</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>848</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[CEU, CHB], ancestral=OOA</td><td>Moves all lineages from derived populations &#x27;CEU&#x27; and &#x27;CHB&#x27; to the ancestral &#x27;OOA&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr><tr><td>848</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_symmetric_migration_rate_change'>Symmetric migration rate change</a></td><td>populations=[YRI, OOA], rate=0.00025</td><td>Sets the symmetric migration rate between YRI and OOA to 0.00025 per generation</td></tr><tr><td>5600</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[YRI, OOA], ancestral=AMH</td><td>Moves all lineages from derived populations &#x27;YRI&#x27; and &#x27;OOA&#x27; to the ancestral &#x27;AMH&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr><tr><td>8800</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[AMH], ancestral=ANC</td><td>Moves all lineages from the &#x27;AMH&#x27; derived population to the ancestral &#x27;ANC&#x27; population. Also set &#x27;AMH&#x27; to inactive, and all migration rates to and from the derived population to zero.</td></tr></tbody></table></div></div></div></div>
</div>
<p>The first event we added was a <a class="reference internal" href="#sec-demography-events-population-split"><span class="std std-ref">Population split</span></a>
using the <a class="reference internal" href="api.html#msprime.Demography.add_population_split" title="msprime.Demography.add_population_split"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.add_population_split()</span></code></a> method. This has
three effects:</p>
<ol class="arabic simple">
<li><p>At the time of the split, all lineages currently in the
<code class="docutils literal notranslate"><span class="pre">CEU</span></code> and <code class="docutils literal notranslate"><span class="pre">CHB</span></code> populations are moved into ancestral
<code class="docutils literal notranslate"><span class="pre">OOA</span></code> population (remember, this is
<a class="reference internal" href="#sec-demography-direction-of-time"><span class="std std-ref">backwards in time</span></a>).</p></li>
<li><p>It marks the <code class="docutils literal notranslate"><span class="pre">OOA</span></code> population as
<a class="reference internal" href="#sec-demography-populations-life-cycle"><span class="std std-ref">inactive</span></a>
“before”  the event, and sets its
<a class="reference internal" href="#sec-demography-populations-default-sampling-time"><span class="std std-ref">sampling time</span></a>
to the time of the population split (as can be seen in the Populations table
above).</p></li>
<li><p>It marks the <code class="docutils literal notranslate"><span class="pre">CEU</span></code> and <code class="docutils literal notranslate"><span class="pre">CHB</span></code> populations as
<a class="reference internal" href="#sec-demography-populations-life-cycle"><span class="std std-ref">previously active</span></a>
“after” the event. Any subsequent attempts to move lineages into these
populations will result in an error.</p></li>
</ol>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>The <strong>effect</strong> column in the events output summarises the effect of
each event.</p>
</div>
<p>The next event is a <a class="reference internal" href="#sec-demography-events-migration-rate-change"><span class="std std-ref">Migration rate change</span></a>
using the <a class="reference internal" href="api.html#msprime.Demography.add_symmetric_migration_rate_change" title="msprime.Demography.add_symmetric_migration_rate_change"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.add_symmetric_migration_rate_change()</span></code></a> method,
that also happens at the Out-of-Africa split time. This sets the migration
rate between the <code class="docutils literal notranslate"><span class="pre">OOA</span></code> and <code class="docutils literal notranslate"><span class="pre">YRI</span></code> populations.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Any number of events can occur at the same time in msprime, and they are
guaranteed to be executed in the order that they are specified.</p>
</div>
<p>The remaining events are then both population splits: 5600 generations
ago, the <code class="docutils literal notranslate"><span class="pre">OOA</span></code> and <code class="docutils literal notranslate"><span class="pre">YRI</span></code> split from the <code class="docutils literal notranslate"><span class="pre">AMH</span></code> population,
and finally 8800 generation ago the <code class="docutils literal notranslate"><span class="pre">AMH</span></code> population “split” from
the <code class="docutils literal notranslate"><span class="pre">ANC</span></code> population. This last population split may seem unnecessary,
since we could have achieved the same effect by a population
<a class="reference internal" href="#sec-demography-events-population-parameters-change"><span class="std std-ref">size change</span></a>,
but from an interpretation perspective it is much better to explicitly
model distinct phases as separate populations.</p>
<p>We can now see the model as a set of populations, their initial state
and migration rates and a set of events. To help understand how this
model evolves over time the <a class="reference internal" href="api.html#msprime.DemographyDebugger" title="msprime.DemographyDebugger"><code class="xref py py-class docutils literal notranslate"><span class="pre">DemographyDebugger</span></code></a> is extremely
useful. To get a debugger instance for a given demography use the
<a class="reference internal" href="api.html#msprime.Demography.debug" title="msprime.Demography.debug"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.debug()</span></code></a> method. This will then display the
state of the demography as a set of “epochs”, where we see the
state of the populations and the migration matrix during different
time periods:</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Only populations that are <strong>active</strong> during a particular epoch
are displayed in the debugger output.</p>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span><span class="o">.</span><span class="n">debug</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div><details open="true"><summary>Epoch[0]: [0, 848) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=6 active=3)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>YRI</th><th>CEU</th><th>CHB</th></tr></thead><tbody><tr><td>YRI</td><td> 12300.0</td><td> 12300.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population YRI to CEU = 3e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from CEU to YRI per generation, divided by the size of YRI.'>3e-05</td><td title='Backwards in time migration rate from population YRI to CHB = 1.9e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from CHB to YRI per generation, divided by the size of YRI.'>1.9e-05</td></tr><tr><td>CEU</td><td> 29725.3</td><td> 1000.0</td><td> 0.004</td><td title='Backwards in time migration rate from population CEU to YRI = 3e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from YRI to CEU per generation, divided by the size of CEU.'>3e-05</td><td>0</td><td title='Backwards in time migration rate from population CEU to CHB = 9.6e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from CHB to CEU per generation, divided by the size of CEU.'>9.6e-05</td></tr><tr><td>CHB</td><td> 54090.3</td><td> 510.0</td><td> 0.0055</td><td title='Backwards in time migration rate from population CHB to YRI = 1.9e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from YRI to CHB per generation, divided by the size of CHB.'>1.9e-05</td><td title='Backwards in time migration rate from population CHB to CEU = 9.6e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from CEU to CHB per generation, divided by the size of CHB.'>9.6e-05</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 848</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>848</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[CEU, CHB], ancestral=OOA</td><td>Moves all lineages from derived populations &#x27;CEU&#x27; and &#x27;CHB&#x27; to the ancestral &#x27;OOA&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr><tr><td>848</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_symmetric_migration_rate_change'>Symmetric migration rate change</a></td><td>populations=[YRI, OOA], rate=0.00025</td><td>Sets the symmetric migration rate between YRI and OOA to 0.00025 per generation</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[1]: [848, 5.6e+03) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=6 active=2)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>YRI</th><th>OOA</th></tr></thead><tbody><tr><td>YRI</td><td> 12300.0</td><td> 12300.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population YRI to OOA = 0.00025 per
generation. Forwards in time, this is the expected number of migrants moving
from OOA to YRI per generation, divided by the size of YRI.'>0.00025</td></tr><tr><td>OOA</td><td> 2100.0</td><td> 2100.0</td><td> 0</td><td title='Backwards in time migration rate from population OOA to YRI = 0.00025 per
generation. Forwards in time, this is the expected number of migrants moving
from YRI to OOA per generation, divided by the size of OOA.'>0.00025</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 5.6e+03</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>5600</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[YRI, OOA], ancestral=AMH</td><td>Moves all lineages from derived populations &#x27;YRI&#x27; and &#x27;OOA&#x27; to the ancestral &#x27;AMH&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[2]: [5.6e+03, 8.8e+03) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=6 active=1)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th></tr></thead><tbody><tr><td>AMH</td><td> 12300.0</td><td> 12300.0</td><td> 0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 8.8e+03</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>8800</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[AMH], ancestral=ANC</td><td>Moves all lineages from the &#x27;AMH&#x27; derived population to the ancestral &#x27;ANC&#x27; population. Also set &#x27;AMH&#x27; to inactive, and all migration rates to and from the derived population to zero.</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[3]: [8.8e+03, inf) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=6 active=1)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th></tr></thead><tbody><tr><td>ANC</td><td> 7300.0</td><td> 7300.0</td><td> 0</td></tr></tbody></table></div></div></details></div></div></div>
</div>
</section>
<section id="sec-demography-examples-admixture">
<span id="id3"></span><h3>Admixture<a class="headerlink" href="#sec-demography-examples-admixture" title="Link to this heading">#</a></h3>
<p>Many models of populations over time cannot be described as a simple
tree. In this example we implement the American Admixture model
from <a class="reference external" href="https://doi.org/10.1371/journal.pgen.1007385">Browning et al. 2011</a>.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>If you want to use this precise model in your analyses
we strongly recommend using the <a class="reference external" href="https://popsim-consortium.github.io/stdpopsim-docs/stable/introduction.html#sec-introduction" title="(in stdpopsim v0.3)"><span class="xref std std-ref">stdpopsim</span></a>
<a class="reference external" href="https://popsim-consortium.github.io/stdpopsim-docs/stable/catalog.html#sec_catalog_homsap_models_americanadmixture_4b18" title="(in stdpopsim v0.3)"><span class="xref std std-ref">HomSap/AmericanAdmixture_4B18</span></a>
model.</p>
</div>
<p>We start again by assigning some values that we use repeatedly
to variables, declaring some populations, and setting up the
initial migration rates between the extant populations:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">T_OOA</span> <span class="o">=</span> <span class="mi">920</span>

<span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;AFR&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;African&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">14474</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;EUR&quot;</span><span class="p">,</span>
    <span class="n">description</span><span class="o">=</span><span class="s2">&quot;European&quot;</span><span class="p">,</span>
    <span class="n">initial_size</span><span class="o">=</span><span class="mi">34039</span><span class="p">,</span>
    <span class="n">growth_rate</span><span class="o">=</span><span class="mf">0.0038</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;EAS&quot;</span><span class="p">,</span>
    <span class="n">description</span><span class="o">=</span><span class="s2">&quot;East Asian&quot;</span><span class="p">,</span>
    <span class="n">initial_size</span><span class="o">=</span><span class="mi">45852</span><span class="p">,</span>
    <span class="n">growth_rate</span><span class="o">=</span><span class="mf">0.0048</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;ADMIX&quot;</span><span class="p">,</span>
    <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Admixed America&quot;</span><span class="p">,</span>
    <span class="n">initial_size</span><span class="o">=</span><span class="mi">54664</span><span class="p">,</span>
    <span class="n">growth_rate</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;OOA&quot;</span><span class="p">,</span>
    <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Bottleneck out-of-Africa&quot;</span><span class="p">,</span>
    <span class="n">initial_size</span><span class="o">=</span><span class="mi">1861</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;AMH&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Anatomically modern humans&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">14474</span>
<span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;ANC&quot;</span><span class="p">,</span>
    <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Ancestral equilibrium&quot;</span><span class="p">,</span>
    <span class="n">initial_size</span><span class="o">=</span><span class="mi">7310</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">set_symmetric_migration_rate</span><span class="p">([</span><span class="s2">&quot;AFR&quot;</span><span class="p">,</span> <span class="s2">&quot;EUR&quot;</span><span class="p">],</span> <span class="mf">2.5e-5</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">set_symmetric_migration_rate</span><span class="p">([</span><span class="s2">&quot;AFR&quot;</span><span class="p">,</span> <span class="s2">&quot;EAS&quot;</span><span class="p">],</span> <span class="mf">0.78e-5</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">set_symmetric_migration_rate</span><span class="p">([</span><span class="s2">&quot;EUR&quot;</span><span class="p">,</span> <span class="s2">&quot;EAS&quot;</span><span class="p">],</span> <span class="mf">3.11e-5</span><span class="p">)</span>

<span class="n">demography</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (7)</b><table border="1" class="tskit-table"><thead><tr><th>id</th><th>name</th><th>description</th><th>initial_size</th><th>growth_rate</th><th>default_sampling_time</th><th>extra_metadata</th></tr></thead><tbody><tr><td>0</td><td>AFR</td><td>African</td><td>14474.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>1</td><td>EUR</td><td>European</td><td>34039.0</td><td>0.0038</td><td>0</td><td>{}</td></tr><tr><td>2</td><td>EAS</td><td>East Asian</td><td>45852.0</td><td>0.0048</td><td>0</td><td>{}</td></tr><tr><td>3</td><td>ADMIX</td><td>Admixed America</td><td>54664.0</td><td>0.05</td><td>0</td><td>{}</td></tr><tr><td>4</td><td>OOA</td><td>Bottleneck out-of-Africa</td><td>1861.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>5</td><td>AMH</td><td>Anatomically modern humans</td><td>14474.0</td><td>0</td><td>0</td><td>{}</td></tr><tr><td>6</td><td>ANC</td><td>Ancestral equilibrium</td><td>7310.0</td><td>0</td><td>0</td><td>{}</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Migration matrix</b><table border="1" class="tskit-table"><thead><tr><th></th><th>AFR</th><th>EUR</th><th>EAS</th><th>ADMIX</th><th>OOA</th><th>AMH</th><th>ANC</th></tr></thead><tbody><tr><td>AFR</td><td>0</td><td title='Backwards in time migration rate from population AFR to EUR = 2.5e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from EUR to AFR per generation, divided by the size of AFR.'>2.5e-05</td><td title='Backwards in time migration rate from population AFR to EAS = 7.8e-06 per
generation. Forwards in time, this is the expected number of migrants moving
from EAS to AFR per generation, divided by the size of AFR.'>7.8e-06</td><td title='Backwards in time migration rate from population AFR to ADMIX = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ADMIX to AFR per generation, divided by the size of AFR.'>0</td><td title='Backwards in time migration rate from population AFR to OOA = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from OOA to AFR per generation, divided by the size of AFR.'>0</td><td title='Backwards in time migration rate from population AFR to AMH = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from AMH to AFR per generation, divided by the size of AFR.'>0</td><td title='Backwards in time migration rate from population AFR to ANC = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ANC to AFR per generation, divided by the size of AFR.'>0</td></tr><tr><td>EUR</td><td title='Backwards in time migration rate from population EUR to AFR = 2.5e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from AFR to EUR per generation, divided by the size of EUR.'>2.5e-05</td><td>0</td><td title='Backwards in time migration rate from population EUR to EAS = 3.11e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from EAS to EUR per generation, divided by the size of EUR.'>3.11e-05</td><td title='Backwards in time migration rate from population EUR to ADMIX = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ADMIX to EUR per generation, divided by the size of EUR.'>0</td><td title='Backwards in time migration rate from population EUR to OOA = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from OOA to EUR per generation, divided by the size of EUR.'>0</td><td title='Backwards in time migration rate from population EUR to AMH = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from AMH to EUR per generation, divided by the size of EUR.'>0</td><td title='Backwards in time migration rate from population EUR to ANC = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ANC to EUR per generation, divided by the size of EUR.'>0</td></tr><tr><td>EAS</td><td title='Backwards in time migration rate from population EAS to AFR = 7.8e-06 per
generation. Forwards in time, this is the expected number of migrants moving
from AFR to EAS per generation, divided by the size of EAS.'>7.8e-06</td><td title='Backwards in time migration rate from population EAS to EUR = 3.11e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from EUR to EAS per generation, divided by the size of EAS.'>3.11e-05</td><td>0</td><td title='Backwards in time migration rate from population EAS to ADMIX = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ADMIX to EAS per generation, divided by the size of EAS.'>0</td><td title='Backwards in time migration rate from population EAS to OOA = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from OOA to EAS per generation, divided by the size of EAS.'>0</td><td title='Backwards in time migration rate from population EAS to AMH = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from AMH to EAS per generation, divided by the size of EAS.'>0</td><td title='Backwards in time migration rate from population EAS to ANC = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ANC to EAS per generation, divided by the size of EAS.'>0</td></tr><tr><td>ADMIX</td><td title='Backwards in time migration rate from population ADMIX to AFR = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from AFR to ADMIX per generation, divided by the size of ADMIX.'>0</td><td title='Backwards in time migration rate from population ADMIX to EUR = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from EUR to ADMIX per generation, divided by the size of ADMIX.'>0</td><td title='Backwards in time migration rate from population ADMIX to EAS = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from EAS to ADMIX per generation, divided by the size of ADMIX.'>0</td><td>0</td><td title='Backwards in time migration rate from population ADMIX to OOA = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from OOA to ADMIX per generation, divided by the size of ADMIX.'>0</td><td title='Backwards in time migration rate from population ADMIX to AMH = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from AMH to ADMIX per generation, divided by the size of ADMIX.'>0</td><td title='Backwards in time migration rate from population ADMIX to ANC = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ANC to ADMIX per generation, divided by the size of ADMIX.'>0</td></tr><tr><td>OOA</td><td title='Backwards in time migration rate from population OOA to AFR = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from AFR to OOA per generation, divided by the size of OOA.'>0</td><td title='Backwards in time migration rate from population OOA to EUR = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from EUR to OOA per generation, divided by the size of OOA.'>0</td><td title='Backwards in time migration rate from population OOA to EAS = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from EAS to OOA per generation, divided by the size of OOA.'>0</td><td title='Backwards in time migration rate from population OOA to ADMIX = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ADMIX to OOA per generation, divided by the size of OOA.'>0</td><td>0</td><td title='Backwards in time migration rate from population OOA to AMH = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from AMH to OOA per generation, divided by the size of OOA.'>0</td><td title='Backwards in time migration rate from population OOA to ANC = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ANC to OOA per generation, divided by the size of OOA.'>0</td></tr><tr><td>AMH</td><td title='Backwards in time migration rate from population AMH to AFR = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from AFR to AMH per generation, divided by the size of AMH.'>0</td><td title='Backwards in time migration rate from population AMH to EUR = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from EUR to AMH per generation, divided by the size of AMH.'>0</td><td title='Backwards in time migration rate from population AMH to EAS = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from EAS to AMH per generation, divided by the size of AMH.'>0</td><td title='Backwards in time migration rate from population AMH to ADMIX = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ADMIX to AMH per generation, divided by the size of AMH.'>0</td><td title='Backwards in time migration rate from population AMH to OOA = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from OOA to AMH per generation, divided by the size of AMH.'>0</td><td>0</td><td title='Backwards in time migration rate from population AMH to ANC = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ANC to AMH per generation, divided by the size of AMH.'>0</td></tr><tr><td>ANC</td><td title='Backwards in time migration rate from population ANC to AFR = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from AFR to ANC per generation, divided by the size of ANC.'>0</td><td title='Backwards in time migration rate from population ANC to EUR = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from EUR to ANC per generation, divided by the size of ANC.'>0</td><td title='Backwards in time migration rate from population ANC to EAS = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from EAS to ANC per generation, divided by the size of ANC.'>0</td><td title='Backwards in time migration rate from population ANC to ADMIX = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ADMIX to ANC per generation, divided by the size of ANC.'>0</td><td title='Backwards in time migration rate from population ANC to OOA = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from OOA to ANC per generation, divided by the size of ANC.'>0</td><td title='Backwards in time migration rate from population ANC to AMH = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from AMH to ANC per generation, divided by the size of ANC.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events (0)</b><table border="1" class="tskit-table"><thead><tr></tr></thead><tbody></tbody></table></div></div></div></div>
</div>
<p>Here we set up seven <a class="reference internal" href="#sec-demography-populations"><span class="std std-ref">populations</span></a>,
their initial states and the migration rates between them. The
next thing we do is add an <a class="reference internal" href="#sec-demography-events-admixture"><span class="std std-ref">admixture</span></a>
event 12 generations in the past:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span><span class="o">.</span><span class="n">add_admixture</span><span class="p">(</span>
    <span class="mi">12</span><span class="p">,</span>
    <span class="n">derived</span><span class="o">=</span><span class="s2">&quot;ADMIX&quot;</span><span class="p">,</span>
    <span class="n">ancestral</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;AFR&quot;</span><span class="p">,</span> <span class="s2">&quot;EUR&quot;</span><span class="p">,</span> <span class="s2">&quot;EAS&quot;</span><span class="p">],</span>
    <span class="n">proportions</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span> <span class="o">/</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">2</span> <span class="o">/</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">3</span> <span class="o">/</span> <span class="mi">6</span><span class="p">],</span>
<span class="p">);</span>
</pre></div>
</div>
</div>
</div>
<p>During a simulation, in an admixture event all of the lineages
present in the derived population move to one of the ancestral
populations with the appropriate probability. Forwards in time,
this corresponds to the initial state of the admixed population
being composed of these fractions of individuals from the
ancestral populations.</p>
<p>We can then add population split events, similar to the
<a class="reference internal" href="#sec-demography-examples-population-tree"><span class="std std-ref">out of Africa model</span></a>.
The <a class="reference internal" href="api.html#msprime.DemographyDebugger" title="msprime.DemographyDebugger"><code class="xref py py-class docutils literal notranslate"><span class="pre">DemographyDebugger</span></code></a> output then shows how the
populations change over time:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span><span class="o">.</span><span class="n">add_population_split</span><span class="p">(</span><span class="n">T_OOA</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;EUR&quot;</span><span class="p">,</span> <span class="s2">&quot;EAS&quot;</span><span class="p">],</span> <span class="n">ancestral</span><span class="o">=</span><span class="s2">&quot;OOA&quot;</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_symmetric_migration_rate_change</span><span class="p">(</span>
    <span class="n">time</span><span class="o">=</span><span class="n">T_OOA</span><span class="p">,</span> <span class="n">populations</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;AFR&quot;</span><span class="p">,</span> <span class="s2">&quot;OOA&quot;</span><span class="p">],</span> <span class="n">rate</span><span class="o">=</span><span class="mf">15e-5</span>
<span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population_split</span><span class="p">(</span><span class="mi">2040</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;OOA&quot;</span><span class="p">,</span> <span class="s2">&quot;AFR&quot;</span><span class="p">],</span> <span class="n">ancestral</span><span class="o">=</span><span class="s2">&quot;AMH&quot;</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population_split</span><span class="p">(</span><span class="mi">5920</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;AMH&quot;</span><span class="p">],</span> <span class="n">ancestral</span><span class="o">=</span><span class="s2">&quot;ANC&quot;</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">debug</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div><details open="true"><summary>Epoch[0]: [0, 12) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=7 active=4)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>AFR</th><th>EUR</th><th>EAS</th><th>ADMIX</th></tr></thead><tbody><tr><td>AFR</td><td> 14474.0</td><td> 14474.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population AFR to EUR = 2.5e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from EUR to AFR per generation, divided by the size of AFR.'>2.5e-05</td><td title='Backwards in time migration rate from population AFR to EAS = 7.8e-06 per
generation. Forwards in time, this is the expected number of migrants moving
from EAS to AFR per generation, divided by the size of AFR.'>7.8e-06</td><td title='Backwards in time migration rate from population AFR to ADMIX = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ADMIX to AFR per generation, divided by the size of AFR.'>0</td></tr><tr><td>EUR</td><td> 34039.0</td><td> 32521.7</td><td> 0.0038</td><td title='Backwards in time migration rate from population EUR to AFR = 2.5e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from AFR to EUR per generation, divided by the size of EUR.'>2.5e-05</td><td>0</td><td title='Backwards in time migration rate from population EUR to EAS = 3.11e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from EAS to EUR per generation, divided by the size of EUR.'>3.11e-05</td><td title='Backwards in time migration rate from population EUR to ADMIX = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ADMIX to EUR per generation, divided by the size of EUR.'>0</td></tr><tr><td>EAS</td><td> 45852.0</td><td> 43285.5</td><td> 0.0048</td><td title='Backwards in time migration rate from population EAS to AFR = 7.8e-06 per
generation. Forwards in time, this is the expected number of migrants moving
from AFR to EAS per generation, divided by the size of EAS.'>7.8e-06</td><td title='Backwards in time migration rate from population EAS to EUR = 3.11e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from EUR to EAS per generation, divided by the size of EAS.'>3.11e-05</td><td>0</td><td title='Backwards in time migration rate from population EAS to ADMIX = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from ADMIX to EAS per generation, divided by the size of EAS.'>0</td></tr><tr><td>ADMIX</td><td> 54664.0</td><td> 30000.2</td><td> 0.05</td><td title='Backwards in time migration rate from population ADMIX to AFR = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from AFR to ADMIX per generation, divided by the size of ADMIX.'>0</td><td title='Backwards in time migration rate from population ADMIX to EUR = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from EUR to ADMIX per generation, divided by the size of ADMIX.'>0</td><td title='Backwards in time migration rate from population ADMIX to EAS = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from EAS to ADMIX per generation, divided by the size of ADMIX.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 12</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>12</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_admixture'>Admixture</a></td><td>derived=ADMIX ancestral=[AFR, EUR, EAS] proportions=[0.17, 0.33, 0.50]</td><td>Moves all lineages from admixed population &#x27;ADMIX&#x27; to ancestral populations. Lineages move to &#x27;AFR&#x27; with proba 0.167; &#x27;EUR&#x27; with proba 0.333; &#x27;EAS&#x27; with proba 0.5. Set &#x27;ADMIX&#x27; to inactive, and all migration rates to and from &#x27;ADMIX&#x27; to zero.</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[1]: [12, 920) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=7 active=3)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>AFR</th><th>EUR</th><th>EAS</th></tr></thead><tbody><tr><td>AFR</td><td> 14474.0</td><td> 14474.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population AFR to EUR = 2.5e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from EUR to AFR per generation, divided by the size of AFR.'>2.5e-05</td><td title='Backwards in time migration rate from population AFR to EAS = 7.8e-06 per
generation. Forwards in time, this is the expected number of migrants moving
from EAS to AFR per generation, divided by the size of AFR.'>7.8e-06</td></tr><tr><td>EUR</td><td> 32521.7</td><td> 1032.0</td><td> 0.0038</td><td title='Backwards in time migration rate from population EUR to AFR = 2.5e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from AFR to EUR per generation, divided by the size of EUR.'>2.5e-05</td><td>0</td><td title='Backwards in time migration rate from population EUR to EAS = 3.11e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from EAS to EUR per generation, divided by the size of EUR.'>3.11e-05</td></tr><tr><td>EAS</td><td> 43285.5</td><td> 554.0</td><td> 0.0048</td><td title='Backwards in time migration rate from population EAS to AFR = 7.8e-06 per
generation. Forwards in time, this is the expected number of migrants moving
from AFR to EAS per generation, divided by the size of EAS.'>7.8e-06</td><td title='Backwards in time migration rate from population EAS to EUR = 3.11e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from EUR to EAS per generation, divided by the size of EAS.'>3.11e-05</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 920</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>920</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[EUR, EAS], ancestral=OOA</td><td>Moves all lineages from derived populations &#x27;EUR&#x27; and &#x27;EAS&#x27; to the ancestral &#x27;OOA&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr><tr><td>920</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_symmetric_migration_rate_change'>Symmetric migration rate change</a></td><td>populations=[AFR, OOA], rate=0.00015</td><td>Sets the symmetric migration rate between AFR and OOA to 0.00015 per generation</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[2]: [920, 2.04e+03) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=7 active=2)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>AFR</th><th>OOA</th></tr></thead><tbody><tr><td>AFR</td><td> 14474.0</td><td> 14474.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population AFR to OOA = 0.00015 per
generation. Forwards in time, this is the expected number of migrants moving
from OOA to AFR per generation, divided by the size of AFR.'>0.00015</td></tr><tr><td>OOA</td><td> 1861.0</td><td> 1861.0</td><td> 0</td><td title='Backwards in time migration rate from population OOA to AFR = 0.00015 per
generation. Forwards in time, this is the expected number of migrants moving
from AFR to OOA per generation, divided by the size of OOA.'>0.00015</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 2.04e+03</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>2040</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[OOA, AFR], ancestral=AMH</td><td>Moves all lineages from derived populations &#x27;OOA&#x27; and &#x27;AFR&#x27; to the ancestral &#x27;AMH&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[3]: [2.04e+03, 5.92e+03) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=7 active=1)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th></tr></thead><tbody><tr><td>AMH</td><td> 14474.0</td><td> 14474.0</td><td> 0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 5.92e+03</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>5920</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[AMH], ancestral=ANC</td><td>Moves all lineages from the &#x27;AMH&#x27; derived population to the ancestral &#x27;ANC&#x27; population. Also set &#x27;AMH&#x27; to inactive, and all migration rates to and from the derived population to zero.</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[4]: [5.92e+03, inf) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=7 active=1)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th></tr></thead><tbody><tr><td>ANC</td><td> 7310.0</td><td> 7310.0</td><td> 0</td></tr></tbody></table></div></div></details></div></div></div>
</div>
</section>
<section id="trunk-population-models">
<span id="sec-demography-examples-trunk-population"></span><h3>Trunk population models<a class="headerlink" href="#trunk-population-models" title="Link to this heading">#</a></h3>
<p>For many empirical models we want to sequentially merge derived populations
into a “trunk” population. We illustrate this with a slightly different
version of the <a class="reference internal" href="#sec-demography-examples-population-tree"><span class="std std-ref">example</span></a> above,
this time defining only
four populations, as in this illustration of the model using
the <a class="reference external" href="https://github.com/apragsdale/demography">demography package</a>
(see also <a class="reference external" href="http://dx.doi.org/10.1371/journal.pgen.1000695.g002">Figure 2B</a>
of the Gutenkunst et al. paper):</p>
<a class="reference internal image-reference" href="_images/Gutenkunst_OOA_diagram.svg"><img alt="Schematic of Gutenkunst et al. (2009) out-of-Africa model." class="align-center" src="_images/Gutenkunst_OOA_diagram.svg" style="width: 500px;" /></a>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The version of this model in this documentation from 31 May 2016 to 29 May 2020
(on the stable branch) was <strong>incorrect</strong>. Specifically, it mistakenly
allowed for migration to continue beyond the merger of the African and
Eurasian bottleneck populations. This has now been fixed, but if you had
copied this model from the tutorial for your own analyses, you should
update your model code or use the implementation that has been verified in
<a class="reference external" href="https://popsim-consortium.github.io/stdpopsim-docs/stable/introduction.html#sec-introduction" title="(in stdpopsim v0.3)"><span class="xref std std-ref">stdpopsim project</span></a>. See
<a class="reference external" href="https://github.com/jeromekelleher/msprime-model-errors">here</a> for more
details on the faulty model and its likely effects on downstream analyses.</p>
</div>
<p>We proceed as before, calculating some quantities that we need
for the model, and creating a <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Times are provided in years, so we convert into generations.</span>
<span class="n">generation_time</span> <span class="o">=</span> <span class="mi">25</span>
<span class="n">T_OOA</span> <span class="o">=</span> <span class="mf">21.2e3</span> <span class="o">/</span> <span class="n">generation_time</span>
<span class="n">T_AMH</span> <span class="o">=</span> <span class="mf">140e3</span> <span class="o">/</span> <span class="n">generation_time</span>
<span class="n">T_ANC</span> <span class="o">=</span> <span class="mf">220e3</span> <span class="o">/</span> <span class="n">generation_time</span>
<span class="c1"># We need to work out the starting population sizes based on</span>
<span class="c1"># the growth rates provided for these two populations</span>
<span class="n">r_CEU</span> <span class="o">=</span> <span class="mf">0.004</span>
<span class="n">r_CHB</span> <span class="o">=</span> <span class="mf">0.0055</span>
<span class="n">N_CEU</span> <span class="o">=</span> <span class="mi">1000</span> <span class="o">/</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="n">r_CEU</span> <span class="o">*</span> <span class="n">T_OOA</span><span class="p">)</span>
<span class="n">N_CHB</span> <span class="o">=</span> <span class="mi">510</span> <span class="o">/</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="n">r_CHB</span> <span class="o">*</span> <span class="n">T_OOA</span><span class="p">)</span>

<span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="c1"># This is the &quot;trunk&quot; population that we merge other populations into</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;YRI&quot;</span><span class="p">,</span>
    <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Africa&quot;</span><span class="p">,</span>
    <span class="n">initial_size</span><span class="o">=</span><span class="mi">12300</span><span class="p">,</span>
    <span class="c1"># NOTE: we have to set this flag if we have a population that&#39;s</span>
    <span class="c1"># ancestral in a population split *and* is contemporary population.</span>
    <span class="n">initially_active</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">);</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;CEU&quot;</span><span class="p">,</span>
    <span class="n">description</span><span class="o">=</span><span class="s2">&quot;European&quot;</span><span class="p">,</span>
    <span class="n">initial_size</span><span class="o">=</span><span class="n">N_CEU</span><span class="p">,</span>
    <span class="n">growth_rate</span><span class="o">=</span><span class="n">r_CEU</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;CHB&quot;</span><span class="p">,</span>
    <span class="n">description</span><span class="o">=</span><span class="s2">&quot;East Asian&quot;</span><span class="p">,</span>
    <span class="n">initial_size</span><span class="o">=</span><span class="n">N_CHB</span><span class="p">,</span>
    <span class="n">growth_rate</span><span class="o">=</span><span class="n">r_CHB</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;OOA&quot;</span><span class="p">,</span>
    <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Bottleneck out-of-Africa population&quot;</span><span class="p">,</span>
    <span class="n">initial_size</span><span class="o">=</span><span class="mi">2100</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Set the migration rates between extant populations</span>
<span class="n">demography</span><span class="o">.</span><span class="n">set_symmetric_migration_rate</span><span class="p">([</span><span class="s2">&quot;CEU&quot;</span><span class="p">,</span> <span class="s2">&quot;CHB&quot;</span><span class="p">],</span> <span class="mf">9.6e-5</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">set_symmetric_migration_rate</span><span class="p">([</span><span class="s2">&quot;YRI&quot;</span><span class="p">,</span> <span class="s2">&quot;CHB&quot;</span><span class="p">],</span> <span class="mf">1.9e-5</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">set_symmetric_migration_rate</span><span class="p">([</span><span class="s2">&quot;YRI&quot;</span><span class="p">,</span> <span class="s2">&quot;CEU&quot;</span><span class="p">],</span> <span class="mf">3e-5</span><span class="p">)</span>

<span class="n">demography</span><span class="o">.</span><span class="n">add_population_split</span><span class="p">(</span>
    <span class="n">time</span><span class="o">=</span><span class="n">T_OOA</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CEU&quot;</span><span class="p">,</span> <span class="s2">&quot;CHB&quot;</span><span class="p">],</span> <span class="n">ancestral</span><span class="o">=</span><span class="s2">&quot;OOA&quot;</span>
<span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_symmetric_migration_rate_change</span><span class="p">(</span>
    <span class="n">time</span><span class="o">=</span><span class="n">T_OOA</span><span class="p">,</span> <span class="n">populations</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;YRI&quot;</span><span class="p">,</span> <span class="s2">&quot;OOA&quot;</span><span class="p">],</span> <span class="n">rate</span><span class="o">=</span><span class="mf">25e-5</span>
<span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population_split</span><span class="p">(</span>
    <span class="n">time</span><span class="o">=</span><span class="n">T_AMH</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;OOA&quot;</span><span class="p">],</span> <span class="n">ancestral</span><span class="o">=</span><span class="s2">&quot;YRI&quot;</span>
<span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population_parameters_change</span><span class="p">(</span>
    <span class="n">time</span><span class="o">=</span><span class="n">T_ANC</span><span class="p">,</span> <span class="n">population</span><span class="o">=</span><span class="s2">&quot;YRI&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">7300</span>
<span class="p">);</span>
</pre></div>
</div>
</div>
</div>
<p>The first population we add is <code class="docutils literal notranslate"><span class="pre">YRI</span></code>, which represents both contemporary
African population as well as ancient populations (which are highly misleading
to label as <code class="docutils literal notranslate"><span class="pre">YRI</span></code> — see the note below). This is the “trunk” population
that we merge other populations into over time, which we do via
<a class="reference internal" href="#sec-demography-events-population-split"><span class="std std-ref">Population split</span></a> events. Because these events are
intended to conveniently model “population trees”, we assume that the ancestral
population will be <a class="reference internal" href="#sec-demography-populations-life-cycle"><span class="std std-ref">inactive</span></a> before
the split (looking backwards). However, in this case, we will be merging the
<code class="docutils literal notranslate"><span class="pre">OOA</span></code> population in to the <code class="docutils literal notranslate"><span class="pre">YRI</span></code>, which we wish to also be
<a class="reference internal" href="#sec-demography-populations-life-cycle"><span class="std std-ref">active</span></a> at the start of the
simulation. To do this, we use the <code class="docutils literal notranslate"><span class="pre">initially_active=True</span></code> flag to
make sure that the <code class="docutils literal notranslate"><span class="pre">YRI</span></code> population is active, and can therefore
represent contemporary Africans as well as ancient populations.</p>
<p>We can see how the <code class="docutils literal notranslate"><span class="pre">YRI</span></code> population then remains active through
all epochs from the debug output:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span><span class="o">.</span><span class="n">debug</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div><details open="true"><summary>Epoch[0]: [0, 848) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=4 active=3)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>YRI</th><th>CEU</th><th>CHB</th></tr></thead><tbody><tr><td>YRI</td><td> 12300.0</td><td> 12300.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population YRI to CEU = 3e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from CEU to YRI per generation, divided by the size of YRI.'>3e-05</td><td title='Backwards in time migration rate from population YRI to CHB = 1.9e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from CHB to YRI per generation, divided by the size of YRI.'>1.9e-05</td></tr><tr><td>CEU</td><td> 29725.3</td><td> 1000.0</td><td> 0.004</td><td title='Backwards in time migration rate from population CEU to YRI = 3e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from YRI to CEU per generation, divided by the size of CEU.'>3e-05</td><td>0</td><td title='Backwards in time migration rate from population CEU to CHB = 9.6e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from CHB to CEU per generation, divided by the size of CEU.'>9.6e-05</td></tr><tr><td>CHB</td><td> 54090.3</td><td> 510.0</td><td> 0.0055</td><td title='Backwards in time migration rate from population CHB to YRI = 1.9e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from YRI to CHB per generation, divided by the size of CHB.'>1.9e-05</td><td title='Backwards in time migration rate from population CHB to CEU = 9.6e-05 per
generation. Forwards in time, this is the expected number of migrants moving
from CEU to CHB per generation, divided by the size of CHB.'>9.6e-05</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 848</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>848</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[CEU, CHB], ancestral=OOA</td><td>Moves all lineages from derived populations &#x27;CEU&#x27; and &#x27;CHB&#x27; to the ancestral &#x27;OOA&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr><tr><td>848</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_symmetric_migration_rate_change'>Symmetric migration rate change</a></td><td>populations=[YRI, OOA], rate=0.00025</td><td>Sets the symmetric migration rate between YRI and OOA to 0.00025 per generation</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[1]: [848, 5.6e+03) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=4 active=2)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>YRI</th><th>OOA</th></tr></thead><tbody><tr><td>YRI</td><td> 12300.0</td><td> 12300.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population YRI to OOA = 0.00025 per
generation. Forwards in time, this is the expected number of migrants moving
from OOA to YRI per generation, divided by the size of YRI.'>0.00025</td></tr><tr><td>OOA</td><td> 2100.0</td><td> 2100.0</td><td> 0</td><td title='Backwards in time migration rate from population OOA to YRI = 0.00025 per
generation. Forwards in time, this is the expected number of migrants moving
from YRI to OOA per generation, divided by the size of OOA.'>0.00025</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 5.6e+03</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>5600</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[OOA], ancestral=YRI</td><td>Moves all lineages from the &#x27;OOA&#x27; derived population to the ancestral &#x27;YRI&#x27; population. Also set &#x27;OOA&#x27; to inactive, and all migration rates to and from the derived population to zero.</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[2]: [5.6e+03, 8.8e+03) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=4 active=1)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th></tr></thead><tbody><tr><td>YRI</td><td> 12300.0</td><td> 12300.0</td><td> 0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 8.8e+03</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>8800</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_parameters_change'>Population parameter change</a></td><td>population=YRI, initial_size=7300</td><td>initial_size → 7.3e+03 for population YRI</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[3]: [8.8e+03, inf) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=4 active=1)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th></tr></thead><tbody><tr><td>YRI</td><td> 7300.0</td><td> 7300.0</td><td> 0</td></tr></tbody></table></div></div></details></div></div></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>We do not recommend using “trunk” populations in this way as it is
misleading to think of contemporary populations like the Yoruba being
in any way representative of the very ancient populations being
modelled in the oldest epochs.</p>
</div>
</section>
</section>
<section id="importing-exporting-model-definitions">
<span id="sec-demography-importing"></span><h2>Importing/exporting model definitions<a class="headerlink" href="#importing-exporting-model-definitions" title="Link to this heading">#</a></h2>
<section id="demes">
<span id="sec-demography-importing-demes"></span><h3>Demes<a class="headerlink" href="#demes" title="Link to this heading">#</a></h3>
<p>The <a class="reference external" href="https://popsim-consortium.github.io/demes-spec-docs/">Demes format</a>
is intended to become a standard for writing demographic models,
to facilitate interchange between simulators, as well as between
demographic inference software and simulators.
Demes models are typically written as a YAML file (see the
<a class="reference external" href="https://popsim-consortium.github.io/demes-spec-docs/main/tutorial.html#sec-tutorial" title="(in Project name not set)"><span class="xref std std-ref">Demes tutorial</span></a> for details), and loaded into Python
using the <a class="reference external" href="https://popsim-consortium.github.io/demes-docs/main/introduction.html#sec-introduction" title="(in Python)"><span class="xref std std-ref">demes</span></a> library.
The <a class="reference internal" href="api.html#msprime.Demography.from_demes" title="msprime.Demography.from_demes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.from_demes()</span></code></a> method converts a <a class="reference external" href="https://popsim-consortium.github.io/demes-docs/main/api.html#demes.Graph" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">demes.Graph</span></code></a>
object into a <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> object, and the <a class="reference internal" href="api.html#msprime.Demography.to_demes" title="msprime.Demography.to_demes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.to_demes()</span></code></a>
method does the opposite conversion.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Demes uses similar terminology to msprime, but uses different conventions.
In particular, Demes uses forwards-time conventions as opposed to the
<a class="reference internal" href="#sec-demography-direction-of-time"><span class="std std-ref">backwards-time</span></a> conventions used
in msprime</p>
</div>
<p>Demes models have time in units of “time ago”, and when using generations as
the time units, the times will directly correspond to msprime’s notion of time:
time 0 is “now” and time values increase towards the past.
However, in contrast to msprime, Demes uses forwards-time conventions.
Noteworthy differences between Demes and msprime include:</p>
<ul class="simple">
<li><p>Each deme has its own list of epochs, whereas in msprime’s
<a class="reference internal" href="api.html#msprime.DemographyDebugger" title="msprime.DemographyDebugger"><code class="xref py py-class docutils literal notranslate"><span class="pre">DemographyDebugger</span></code></a>, epochs are time intervals that apply across
the entire demography.</p></li>
<li><p>Demes uses <code class="docutils literal notranslate"><span class="pre">start_time</span></code> and <code class="docutils literal notranslate"><span class="pre">end_time</span></code> to indicate time intervals,
but the <code class="docutils literal notranslate"><span class="pre">start_time</span></code> has a value which is larger (more ancient)
than the <code class="docutils literal notranslate"><span class="pre">end_time</span></code>.</p></li>
<li><p>For migrations and pulses, <code class="docutils literal notranslate"><span class="pre">source</span></code> indicates the deme in which a migrant
is born and the <code class="docutils literal notranslate"><span class="pre">dest</span></code> indicates the deme in which a migrant has offspring.
This is opposite to the
<a class="reference internal" href="#sec-demography-direction-of-time"><span class="std std-ref">convention used in msprime</span></a>.</p></li>
</ul>
<p>Consider the following Demes model that describes two extant demes <code class="docutils literal notranslate"><span class="pre">A</span></code> and <code class="docutils literal notranslate"><span class="pre">B</span></code>,
which have a common ancestor <code class="docutils literal notranslate"><span class="pre">X</span></code>.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="c1"># Comments start with a hash.</span>
<span class="nt">description</span><span class="p">:</span>
<span class="w">  </span><span class="l l-Scalar l-Scalar-Plain">Asymmetric migration between two extant demes.</span>
<span class="nt">time_units</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">generations</span>
<span class="nt">defaults</span><span class="p">:</span>
<span class="w">  </span><span class="nt">epoch</span><span class="p">:</span>
<span class="w">    </span><span class="nt">start_size</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">5000</span>
<span class="nt">demes</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">X</span>
<span class="w">    </span><span class="nt">epochs</span><span class="p">:</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">end_time</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1000</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">A</span>
<span class="w">    </span><span class="nt">ancestors</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">X</span><span class="p p-Indicator">]</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">B</span>
<span class="w">    </span><span class="nt">ancestors</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">X</span><span class="p p-Indicator">]</span>
<span class="w">    </span><span class="nt">epochs</span><span class="p">:</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">start_size</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2000</span>
<span class="w">        </span><span class="nt">end_time</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">500</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">start_size</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">400</span>
<span class="w">        </span><span class="nt">end_size</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">10000</span>
<span class="w">        </span><span class="nt">end_time</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0</span>
<span class="nt">migrations</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">source</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">A</span>
<span class="w">    </span><span class="nt">dest</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">B</span>
<span class="w">    </span><span class="nt">rate</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1e-4</span>
</pre></div>
</div>
<p>When visualised using <a class="reference external" href="https://github.com/grahamgower/demesdraw">demesdraw</a>,
the demographic model looks like this (where arrows indicate movement of
individuals as time proceeds from the past to the present):</p>
<div class="cell tag_remove-input docutils container">
<div class="cell_output docutils container">
<img alt="_images/8cc205808a1453630e79646f319c0f28485e34333620425e81fcfa0e21500ffd.svg" src="_images/8cc205808a1453630e79646f319c0f28485e34333620425e81fcfa0e21500ffd.svg" />
</div>
</div>
<p>Let’s load this model using <a class="reference external" href="https://popsim-consortium.github.io/demes-docs/main/api.html#demes.load" title="(in Python)"><code class="xref py py-func docutils literal notranslate"><span class="pre">demes.load()</span></code></a>, convert it to
a <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> object with <a class="reference internal" href="api.html#msprime.Demography.from_demes" title="msprime.Demography.from_demes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.from_demes()</span></code></a>
and then look at the <a class="reference internal" href="api.html#msprime.DemographyDebugger" title="msprime.DemographyDebugger"><code class="xref py py-class docutils literal notranslate"><span class="pre">DemographyDebugger</span></code></a> output.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">demes</span>

<span class="n">graph</span> <span class="o">=</span> <span class="n">demes</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;demes_example_01.yaml&quot;</span><span class="p">)</span>
<span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">from_demes</span><span class="p">(</span><span class="n">graph</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">debug</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div><details open="true"><summary>Epoch[0]: [0, 500) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=3 active=2)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>A</th><th>B</th></tr></thead><tbody><tr><td>A</td><td> 5000.0</td><td> 5000.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population A to B = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from B to A per
generation, divided by the size of A.'>0</td></tr><tr><td>B</td><td> 10000.0</td><td> 400.0</td><td> 0.00644</td><td title='Backwards in time migration rate from population B to A = 0.0001 per generation.
Forwards in time, this is the expected number of migrants moving from A to B per
generation, divided by the size of B.'>0.0001</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 500</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>500</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_parameters_change'>Population parameter change</a></td><td>population=B, initial_size=2000, growth_rate=0</td><td>initial_size → 2e+03 and growth_rate → 0 for population B</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[1]: [500, 1e+03) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=3 active=2)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>A</th><th>B</th></tr></thead><tbody><tr><td>A</td><td> 5000.0</td><td> 5000.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population A to B = 0.0 per generation.
Forwards in time, this is the expected number of migrants moving from B to A per
generation, divided by the size of A.'>0</td></tr><tr><td>B</td><td> 2000.0</td><td> 2000.0</td><td> 0</td><td title='Backwards in time migration rate from population B to A = 0.0001 per generation.
Forwards in time, this is the expected number of migrants moving from A to B per
generation, divided by the size of B.'>0.0001</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 1e+03</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>1000</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_parameters_change'>Population parameter change</a></td><td>population=X, initial_size=5000</td><td>initial_size → 5e+03 for population X</td></tr><tr><td>1000</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[B, A], ancestral=X</td><td>Moves all lineages from derived populations &#x27;B&#x27; and &#x27;A&#x27; to the ancestral &#x27;X&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr><tr><td>1000</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_migration_rate_change'>Migration rate change</a></td><td>source=B, dest=A, rate=0</td><td>Backwards-time migration rate from B to A → 0</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[2]: [1e+03, inf) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=3 active=1)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th></tr></thead><tbody><tr><td>X</td><td> 5000.0</td><td> 5000.0</td><td> 0</td></tr></tbody></table></div></div></details></div></div></div>
</div>
<p>The demography debugger output shows three epochs. <code class="docutils literal notranslate"><span class="pre">Epoch[0]</span></code>
corresponds to the time interval [0, 500) generations ago, in which
deme <code class="docutils literal notranslate"><span class="pre">B</span></code> has an initial size of 10000 at time 0 and backwards in time,
shrinks in size to 400 individuals at time 500. Lineages sampled
from deme <code class="docutils literal notranslate"><span class="pre">B</span></code> migrate into deme <code class="docutils literal notranslate"><span class="pre">A</span></code> at a rate of 0.0001 per generation.
In <code class="docutils literal notranslate"><span class="pre">Epoch[1]</span></code>, [500, 1000) generations ago, both <code class="docutils literal notranslate"><span class="pre">A</span></code> and <code class="docutils literal notranslate"><span class="pre">B</span></code> have constant
sizes, and migration is the same as in <code class="docutils literal notranslate"><span class="pre">Epoch[0]</span></code>.
At the end of <code class="docutils literal notranslate"><span class="pre">Epoch[1]</span></code>, at generation 1000, lineages in demes <code class="docutils literal notranslate"><span class="pre">A</span></code> and <code class="docutils literal notranslate"><span class="pre">B</span></code>
are moved into deme <code class="docutils literal notranslate"><span class="pre">X</span></code> and migrations are turned off.
Finally, <code class="docutils literal notranslate"><span class="pre">Epoch[2]</span></code> shows that only deme <code class="docutils literal notranslate"><span class="pre">X</span></code> remains.</p>
<p>A <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> object can also be exported to a <a class="reference external" href="https://popsim-consortium.github.io/demes-docs/main/api.html#demes.Graph" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">demes.Graph</span></code></a>,
using the <a class="reference internal" href="api.html#msprime.Demography.to_demes" title="msprime.Demography.to_demes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.to_demes()</span></code></a> method. We’ll use this to make a visual
representation of a modified <a class="reference internal" href="api.html#msprime.Demography.island_model" title="msprime.Demography.island_model"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.island_model()</span></code></a> using
<a class="reference external" href="https://github.com/grahamgower/demesdraw">demesdraw</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">demesdraw</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>

<span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">island_model</span><span class="p">([</span><span class="mi">300</span><span class="p">]</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span> <span class="mf">1e-5</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;ancestral&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population_split</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="n">ancestral</span><span class="o">=</span><span class="s2">&quot;ancestral&quot;</span><span class="p">)</span>
<span class="n">graph</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">to_demes</span><span class="p">(</span><span class="n">demography</span><span class="p">)</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>  <span class="c1"># use plt.rcParams[&quot;figure.figsize&quot;]</span>
<span class="n">demesdraw</span><span class="o">.</span><span class="n">tubes</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/6369f7dfa89a43363a3b67b54d768ee7731c31a0113393f2f46da2cb67b6b619.svg" src="_images/6369f7dfa89a43363a3b67b54d768ee7731c31a0113393f2f46da2cb67b6b619.svg" />
</div>
</div>
<p>The <a class="reference external" href="https://popsim-consortium.github.io/demes-docs/main/api.html#demes.Graph" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">demes.Graph</span></code></a> can also be written to a YAML file,
e.g. for use with other simulation software.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demes</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="s2">&quot;/tmp/my-island-model.yaml&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="species-trees">
<span id="sec-demography-importing-species-trees"></span><h3>Species trees<a class="headerlink" href="#species-trees" title="Link to this heading">#</a></h3>
<p>Species trees hold information about the sequence and the times at which species
diverged from each other, which we model via a series of
<a class="reference internal" href="#sec-demography-events-population-split"><span class="std std-ref">Population split</span></a>.
The <a class="reference internal" href="api.html#msprime.Demography.from_species_tree" title="msprime.Demography.from_species_tree"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.from_species_tree()</span></code></a> method parses a
species tree in <a class="reference external" href="https://en.wikipedia.org/wiki/Newick_format">Newick format</a>
and returns a <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> object. These species trees do not contain
any information on the sizes of the relevant populations, however, and so these
must be specified separately using the <code class="docutils literal notranslate"><span class="pre">initial_size</span></code> argument
(see <a class="reference internal" href="#sec-demography-populations-initial-size"><span class="std std-ref">Initial size</span></a>).
When species trees are estimated with a program like
<a class="reference external" href="https://academic.oup.com/mbe/article/34/8/2101/3738283">StarBEAST</a> they can
contain estimates on the population sizes of extant and ancestral species.
The <a class="reference internal" href="api.html#msprime.Demography.from_starbeast" title="msprime.Demography.from_starbeast"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.from_starbeast()</span></code></a> method parses species trees estimated
with StarBEAST and sets the population sizes accordingly.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>When a species tree has branch lengths in units of years or millions of years
rather than generations (which is common), a generation time in years
must be specified.</p>
</div>
<p>Species trees must be encoded in
<a class="reference external" href="https://en.wikipedia.org/wiki/Newick_format">Newick</a> format, with named leaves and
branch lengths. Consider the following species tree of four primates, for example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">(((</span><span class="n">human</span><span class="p">:</span><span class="mf">5.6</span><span class="p">,</span><span class="n">chimpanzee</span><span class="p">:</span><span class="mf">5.6</span><span class="p">):</span><span class="mf">3.0</span><span class="p">,</span><span class="n">gorilla</span><span class="p">:</span><span class="mf">8.6</span><span class="p">):</span><span class="mf">9.4</span><span class="p">,</span><span class="n">orangutan</span><span class="p">:</span><span class="mf">18.0</span><span class="p">)</span>
</pre></div>
</div>
<p>When visualized in software like
<a class="reference external" href="http://tree.bio.ed.ac.uk/software/figtree/">FigTree</a>, the tree
looks like this:</p>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/primates.svg"><img alt="A species tree with branch lengths." src="_images/primates.svg" style="width: 400px;" /></a>
</figure>
<p>The numbers written on branches indicate the lengths of these branches,
which in this case is expressed in millions of years. We can then
convert this species tree into a <a class="reference internal" href="api.html#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> using the
<a class="reference internal" href="api.html#msprime.Demography.from_species_tree" title="msprime.Demography.from_species_tree"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Demography.from_species_tree()</span></code></a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">msprime</span>

<span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">from_species_tree</span><span class="p">(</span>
    <span class="s2">&quot;(((human:5.6,chimpanzee:5.6):3.0,gorilla:8.6):9.4,orangutan:18.0)&quot;</span><span class="p">,</span>
    <span class="n">initial_size</span><span class="o">=</span><span class="mi">10_000</span><span class="p">,</span>
    <span class="n">time_units</span><span class="o">=</span><span class="s2">&quot;myr&quot;</span><span class="p">,</span>
    <span class="n">generation_time</span><span class="o">=</span><span class="mi">28</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">debug</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div><details open="true"><summary>Epoch[0]: [0, 2e+05) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=7 active=4)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>human</th><th>chimpanzee</th><th>gorilla</th><th>orangutan</th></tr></thead><tbody><tr><td>human</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population human to chimpanzee = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from chimpanzee to human per generation, divided by the size of human.'>0</td><td title='Backwards in time migration rate from population human to gorilla = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from gorilla to human per generation, divided by the size of human.'>0</td><td title='Backwards in time migration rate from population human to orangutan = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from orangutan to human per generation, divided by the size of human.'>0</td></tr><tr><td>chimpanzee</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td title='Backwards in time migration rate from population chimpanzee to human = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from human to chimpanzee per generation, divided by the size of chimpanzee.'>0</td><td>0</td><td title='Backwards in time migration rate from population chimpanzee to gorilla = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from gorilla to chimpanzee per generation, divided by the size of chimpanzee.'>0</td><td title='Backwards in time migration rate from population chimpanzee to orangutan = 0.0
per generation. Forwards in time, this is the expected number of migrants moving
from orangutan to chimpanzee per generation, divided by the size of chimpanzee.'>0</td></tr><tr><td>gorilla</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td title='Backwards in time migration rate from population gorilla to human = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from human to gorilla per generation, divided by the size of gorilla.'>0</td><td title='Backwards in time migration rate from population gorilla to chimpanzee = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from chimpanzee to gorilla per generation, divided by the size of gorilla.'>0</td><td>0</td><td title='Backwards in time migration rate from population gorilla to orangutan = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from orangutan to gorilla per generation, divided by the size of gorilla.'>0</td></tr><tr><td>orangutan</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td title='Backwards in time migration rate from population orangutan to human = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from human to orangutan per generation, divided by the size of orangutan.'>0</td><td title='Backwards in time migration rate from population orangutan to chimpanzee = 0.0
per generation. Forwards in time, this is the expected number of migrants moving
from chimpanzee to orangutan per generation, divided by the size of orangutan.'>0</td><td title='Backwards in time migration rate from population orangutan to gorilla = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from gorilla to orangutan per generation, divided by the size of orangutan.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 2e+05</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>2e+05</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[human, chimpanzee], ancestral=pop_4</td><td>Moves all lineages from derived populations &#x27;human&#x27; and &#x27;chimpanzee&#x27; to the ancestral &#x27;pop_4&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[1]: [2e+05, 3.07e+05) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=7 active=3)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>gorilla</th><th>orangutan</th><th>pop_4</th></tr></thead><tbody><tr><td>gorilla</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population gorilla to orangutan = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from orangutan to gorilla per generation, divided by the size of gorilla.'>0</td><td title='Backwards in time migration rate from population gorilla to pop_4 = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from pop_4 to gorilla per generation, divided by the size of gorilla.'>0</td></tr><tr><td>orangutan</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td title='Backwards in time migration rate from population orangutan to gorilla = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from gorilla to orangutan per generation, divided by the size of orangutan.'>0</td><td>0</td><td title='Backwards in time migration rate from population orangutan to pop_4 = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from pop_4 to orangutan per generation, divided by the size of orangutan.'>0</td></tr><tr><td>pop_4</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td title='Backwards in time migration rate from population pop_4 to gorilla = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from gorilla to pop_4 per generation, divided by the size of pop_4.'>0</td><td title='Backwards in time migration rate from population pop_4 to orangutan = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from orangutan to pop_4 per generation, divided by the size of pop_4.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 3.07e+05</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>3.071e+05</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[pop_4, gorilla], ancestral=pop_5</td><td>Moves all lineages from derived populations &#x27;pop_4&#x27; and &#x27;gorilla&#x27; to the ancestral &#x27;pop_5&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[2]: [3.07e+05, 6.43e+05) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=7 active=2)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>orangutan</th><th>pop_5</th></tr></thead><tbody><tr><td>orangutan</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population orangutan to pop_5 = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from pop_5 to orangutan per generation, divided by the size of orangutan.'>0</td></tr><tr><td>pop_5</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td title='Backwards in time migration rate from population pop_5 to orangutan = 0.0 per
generation. Forwards in time, this is the expected number of migrants moving
from orangutan to pop_5 per generation, divided by the size of pop_5.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Events @ generation 6.43e+05</b><table border="1" class="tskit-table"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>6.429e+05</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[pop_5, orangutan], ancestral=pop_6</td><td>Moves all lineages from derived populations &#x27;pop_5&#x27; and &#x27;orangutan&#x27; to the ancestral &#x27;pop_6&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr></tbody></table></div></div></details><details open="true"><summary>Epoch[3]: [6.43e+05, inf) generations</summary><div style="margin-left:20px"><div><style scoped="">
            .tskit-table thead tr th:only-of-type {vertical-align: middle;}
            .tskit-table thead tr th {text-align: center;vertical-align: top;}
            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
            .tskit-table tbody th {padding: 0.5em 0.5em;}
        </style><b>Populations (total=7 active=1)</b><table border="1" class="tskit-table"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th></tr></thead><tbody><tr><td>pop_6</td><td> 10000.0</td><td> 10000.0</td><td> 0</td></tr></tbody></table></div></div></details></div></div></div>
</div>
<p>Because the species tree does not contain complete information about the
demographic model, we must provide some extra information. The
<code class="docutils literal notranslate"><span class="pre">initial_size</span></code> parameter lets us specify the size of each of the
populations; here we give all the populations a fixed size of 10,000
(there is much more flexibility here, however). Because the branch
lengths in the species tree are given in millions of years, we also
need provide a <code class="docutils literal notranslate"><span class="pre">time_units</span></code> parameter and a <code class="docutils literal notranslate"><span class="pre">generation_time</span></code>.</p>
<p>The epoch boundaries 200000, 307142.9, and 642857.1 correspond to the species
divergence times 5.6, 8.6, and 18.0 after converting the branch length units
of the species tree from millions of years to generations with the specified
generation time of 28 years.</p>
<p>Running the simulation is then straightforward:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
    <span class="p">{</span><span class="s2">&quot;human&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;orangutan&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">},</span> <span class="n">demography</span><span class="o">=</span><span class="n">demography</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">ts</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
              <style>
                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}
                .tskit-table tbody tr td {padding: 0.5em 0.5em;}
                .tskit-table tbody tr td:first-of-type {text-align: left;}
                .tskit-details-label {vertical-align: top; padding-right:5px;}
                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}
                .tskit-table-set-table {margin: 12px 0 0 12px;}
                details {display: inline-block;}
                summary {cursor: pointer; outline: 0; display: list-item;}
              </style>
              <div class="tskit-table-set">
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="padding:0;line-height:21px;">
                          <img style="height: 32px;display: inline-block;padding: 3px 5px 3px 0;" src="https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg"/>
                          <a target="_blank" href="https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class"> Tree Sequence </a>
                        </th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr><td>Trees</td><td>1</td></tr>
                      <tr><td>Sequence Length</td><td>1.0</td></tr>
                      <tr><td>Time Units</td><td>generations</td></tr>
                      <tr><td>Sample Nodes</td><td>8</td></tr>
                      <tr><td>Total Size</td><td>4.4 KiB</td></tr>
                      <tr>
                        <td>Metadata</td><td style="text-align: left;">No Metadata</td></tr>
                    </tbody>
                  </table>
                </div>
                <div class="tskit-table-set-table">
                  <table class="tskit-table">
                    <thead>
                      <tr>
                        <th style="line-height:21px;">Table</th>
                        <th>Rows</th>
                        <th>Size</th>
                        <th>Has Metadata</th>
                      </tr>
                    </thead>
                    <tbody>
                    
                  <tr>
                    <td>Edges</td>
                      <td>14</td>
                      <td>456 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Individuals</td>
                      <td>4</td>
                      <td>136 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Migrations</td>
                      <td>0</td>
                      <td>8 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Mutations</td>
                      <td>0</td>
                      <td>16 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Nodes</td>
                      <td>15</td>
                      <td>428 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Populations</td>
                      <td>7</td>
                      <td>481 Bytes</td>
                      <td style="text-align: center;">
                        ✅
                      </td>
                    </tr>
                
                  <tr>
                    <td>Provenances</td>
                      <td>1</td>
                      <td>2.8 KiB</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                  <tr>
                    <td>Sites</td>
                      <td>0</td>
                      <td>16 Bytes</td>
                      <td style="text-align: center;">
                        
                      </td>
                    </tr>
                
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
            </div></div>
</div>
<p>Note that the names of the populations are recorded in the population
metadata:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">population</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">populations</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">population</span><span class="o">.</span><span class="n">metadata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{&#39;description&#39;: &#39;&#39;, &#39;name&#39;: &#39;human&#39;}
{&#39;description&#39;: &#39;&#39;, &#39;name&#39;: &#39;chimpanzee&#39;}
{&#39;description&#39;: &#39;&#39;, &#39;name&#39;: &#39;gorilla&#39;}
{&#39;description&#39;: &#39;&#39;, &#39;name&#39;: &#39;orangutan&#39;}
{&#39;description&#39;: &#39;&#39;, &#39;name&#39;: &#39;pop_4&#39;}
{&#39;description&#39;: &#39;&#39;, &#39;name&#39;: &#39;pop_5&#39;}
{&#39;description&#39;: &#39;&#39;, &#39;name&#39;: &#39;pop_6&#39;}
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="numerical-utilities">
<span id="sec-demography-numerical"></span><h2>Numerical utilities<a class="headerlink" href="#numerical-utilities" title="Link to this heading">#</a></h2>
<section id="population-sizes-over-time">
<h3>Population sizes over time<a class="headerlink" href="#population-sizes-over-time" title="Link to this heading">#</a></h3>
<p>When working with exponential growth models it is useful to be able to compute
the sizes of the populations at given times. The
<a class="reference internal" href="api.html#msprime.DemographyDebugger.population_size_trajectory" title="msprime.DemographyDebugger.population_size_trajectory"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DemographyDebugger.population_size_trajectory()</span></code></a> gives a straightforward
way of doing this. For example, here we create two exponentially growing populations
and show how their sizes change over time:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span> <span class="n">growth_rate</span><span class="o">=</span><span class="mf">0.01</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span> <span class="n">growth_rate</span><span class="o">=</span><span class="mf">0.03</span><span class="p">)</span>

<span class="n">debug</span> <span class="o">=</span> <span class="n">demography</span><span class="o">.</span><span class="n">debug</span><span class="p">()</span>

<span class="n">t</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="n">num</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">S</span> <span class="o">=</span> <span class="n">debug</span><span class="o">.</span><span class="n">population_size_trajectory</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">S</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Time ago&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Population size&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">();</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/a1cd392329847c9e8a2bd8a7966470edcf3a5507d82c5134a78770c8204ab70a.svg" src="_images/a1cd392329847c9e8a2bd8a7966470edcf3a5507d82c5134a78770c8204ab70a.svg" />
</div>
</div>
</section>
<section id="possible-locations-of-lineages">
<h3>Possible locations of lineages<a class="headerlink" href="#possible-locations-of-lineages" title="Link to this heading">#</a></h3>
<p>When debugging a demographic model it can be helpful to track the possible
location of ancestral lineages over time, as determined by the model
structure and the initial composition of the sample. For example,
here we create a simple model involving a
<a class="reference internal" href="#sec-demography-events-population-split"><span class="std std-ref">population split</span></a>, and then
use the <a class="reference internal" href="api.html#msprime.DemographyDebugger.possible_lineage_locations" title="msprime.DemographyDebugger.possible_lineage_locations"><code class="xref py py-meth docutils literal notranslate"><span class="pre">possible_lineage_locations()</span></code></a> method:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population_split</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">],</span> <span class="n">ancestral</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">)</span>

<span class="n">debug</span> <span class="o">=</span> <span class="n">demography</span><span class="o">.</span><span class="n">debug</span><span class="p">()</span>
<span class="n">debug</span><span class="o">.</span><span class="n">possible_lineage_locations</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{(np.float64(0.0), np.float64(500.0)): array([ True,  True, False]),
 (np.float64(500.0), inf): array([False, False,  True])}
</pre></div>
</div>
</div>
</div>
<p>The method returns a mapping of time intervals to numpy boolean arrays.
The keys are <code class="docutils literal notranslate"><span class="pre">(start,</span> <span class="pre">end]</span></code> intervals during which the possible location
of samples are the same, and the values are arrays telling us whether
a given population can have lineages present during this epoch or not.
By default,
<a class="reference internal" href="api.html#msprime.DemographyDebugger.possible_lineage_locations" title="msprime.DemographyDebugger.possible_lineage_locations"><code class="xref py py-meth docutils literal notranslate"><span class="pre">possible_lineage_locations()</span></code></a> assumes that we
draw samples from all populations that have a
<a class="reference internal" href="#sec-demography-populations-default-sampling-time"><span class="std std-ref">default sampling time</span></a>
of zero, and so for the initial period between time 0 and 500 we
can have samples in population “A” and “B”, but not in “C”. The next time
slice, from 500 generations ago onwards, we can only
have lineages in population “C” (because after the population split all
lineages will be in the ancestral population).</p>
<p>We can also specify that we are interested in sampling from specific
populations at particular times. Suppose we take samples from “A” at
time 0 and from “B” at time 20:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">debug</span><span class="o">.</span><span class="n">possible_lineage_locations</span><span class="p">([</span>
    <span class="n">msprime</span><span class="o">.</span><span class="n">SampleSet</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">population</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">time</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span>
    <span class="n">msprime</span><span class="o">.</span><span class="n">SampleSet</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">population</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">time</span><span class="o">=</span><span class="mi">20</span><span class="p">),</span>
<span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{(np.float64(0.0), 20): array([ True, False, False]),
 (20, np.float64(500.0)): array([ True,  True, False]),
 (np.float64(500.0), inf): array([False, False,  True])}
</pre></div>
</div>
</div>
</div>
<p>Now, we can have lineages in “A” from 0 to 20 generations ago, but not
in “B”.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The number of samples we supply per population isn’t important
(once it’s greater than 0).</p>
</div>
<p>The effects of migration are also taken into account by this method.
For example, consider a different model in which we have two initially
isolated populations, with migration between then introduced 50
generations ago:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_symmetric_migration_rate_change</span><span class="p">(</span>
    <span class="n">time</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">populations</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">],</span> <span class="n">rate</span><span class="o">=</span><span class="mf">0.01</span><span class="p">)</span>

<span class="n">debug</span> <span class="o">=</span> <span class="n">demography</span><span class="o">.</span><span class="n">debug</span><span class="p">()</span>
<span class="n">debug</span><span class="o">.</span><span class="n">possible_lineage_locations</span><span class="p">([</span><span class="s2">&quot;A&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{(np.float64(0.0), np.float64(50.0)): array([ True, False]),
 (np.float64(50.0), inf): array([ True,  True])}
</pre></div>
</div>
</div>
</div>
<p>Because we only sampled lineages in “A”, we cannot have
any lineages in “B” until after the migration 50 generations ago.</p>
</section>
<section id="lineage-probabilities">
<h3>Lineage probabilities<a class="headerlink" href="#lineage-probabilities" title="Link to this heading">#</a></h3>
<p>The <a class="reference internal" href="api.html#msprime.DemographyDebugger.lineage_probabilities" title="msprime.DemographyDebugger.lineage_probabilities"><code class="xref py py-meth docutils literal notranslate"><span class="pre">lineage_probabilities()</span></code></a> method computes
the probability that a lineage sampled at a given time in a each population
is present in any other population at a set of times in the past.
Consider a simple
<a class="reference internal" href="#sec-demography-events-population-split"><span class="std std-ref">population split</span></a> model:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population_split</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">derived</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">],</span> <span class="n">ancestral</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">)</span>

<span class="n">debug</span> <span class="o">=</span> <span class="n">demography</span><span class="o">.</span><span class="n">debug</span><span class="p">()</span>
<span class="n">debug</span><span class="o">.</span><span class="n">lineage_probabilities</span><span class="p">([</span><span class="mi">250</span><span class="p">,</span> <span class="mi">750</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([[[1., 0., 0.],
        [0., 1., 0.],
        [0., 0., 1.]],

       [[0., 0., 1.],
        [0., 0., 1.],
        [0., 0., 1.]]])
</pre></div>
</div>
</div>
</div>
<p>Here we ask to compute the matrix of lineage probabilities 250 and
750 generations ago, before and after the population split. Before the
split, any lineages sampled in a population can only still be in
that population so the diagonal of the matrix is 1. After the
split, all lineages must be in population “C”, and so the
corresponding column is 1.</p>
<p>Migration is taken into account in these calculations.
If we add some migration between “A” and “B”,
we can see how lineages equilibrate between the two:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">set_migration_rate</span><span class="p">(</span><span class="n">source</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">0.005</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">set_migration_rate</span><span class="p">(</span><span class="n">source</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">0.02</span><span class="p">)</span>

<span class="n">debug</span> <span class="o">=</span> <span class="n">demography</span><span class="o">.</span><span class="n">debug</span><span class="p">()</span>
<span class="n">T</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">500</span><span class="p">,</span> <span class="mi">51</span><span class="p">)</span>
<span class="n">X</span> <span class="o">=</span> <span class="n">debug</span><span class="o">.</span><span class="n">lineage_probabilities</span><span class="p">(</span><span class="n">T</span><span class="p">)</span>

<span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">pop</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">demography</span><span class="o">.</span><span class="n">populations</span><span class="p">):</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">j</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">dest_pop</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">demography</span><span class="o">.</span><span class="n">populations</span><span class="p">):</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">T</span><span class="p">,</span> <span class="n">X</span><span class="p">[:,</span> <span class="n">j</span><span class="p">,</span> <span class="n">k</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="n">dest_pop</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Time ago&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Sample from </span><span class="si">{</span><span class="n">pop</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">j</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Probability of being in population&quot;</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/ce2226cb6edf853b645ad9da5a96469f1a066cee9bff67b3d4e4353827b69c54.svg" src="_images/ce2226cb6edf853b645ad9da5a96469f1a066cee9bff67b3d4e4353827b69c54.svg" />
</div>
</div>
<p>Since migration is asymmetric, lineages are more likely
to be found in population “A”.
We’ve set the migration rate for <em>lineages</em>
to be higher from “B” to “A”, so lineages tend to spend more time in “A”.
Since the two populations have the same size,
this corresponds (in forwards time) to a higher flux of migrants from “A” to “B”,
as would happen if mean fecundity in “A” is higher than in “B”.</p>
<p>We can also use this to see how lineages spread out over discrete
space in a 1D <a class="reference internal" href="api.html#msprime.Demography.stepping_stone_model" title="msprime.Demography.stepping_stone_model"><code class="xref py py-meth docutils literal notranslate"><span class="pre">stepping_stone_model()</span></code></a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">N</span> <span class="o">=</span> <span class="mi">32</span>
<span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span><span class="o">.</span><span class="n">stepping_stone_model</span><span class="p">(</span>
    <span class="n">initial_size</span><span class="o">=</span><span class="p">[</span><span class="mi">100</span><span class="p">]</span> <span class="o">*</span> <span class="n">N</span><span class="p">,</span> <span class="n">migration_rate</span><span class="o">=</span><span class="mf">0.01</span><span class="p">)</span>

<span class="n">debug</span> <span class="o">=</span> <span class="n">demography</span><span class="o">.</span><span class="n">debug</span><span class="p">()</span>
<span class="n">T</span> <span class="o">=</span> <span class="p">[</span><span class="mi">10</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">1000</span><span class="p">]</span>
<span class="n">X</span> <span class="o">=</span> <span class="n">debug</span><span class="o">.</span><span class="n">lineage_probabilities</span><span class="p">(</span><span class="n">T</span><span class="p">)</span>

<span class="k">for</span> <span class="n">t</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">T</span><span class="p">,</span> <span class="n">X</span><span class="p">):</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="n">N</span> <span class="o">//</span> <span class="mi">2</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">t</span><span class="si">}</span><span class="s2"> generations ago&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Population ID&quot;</span><span class="p">);</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Probability of being in population&quot;</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/a57566f510e83b1867c9a868dc9cbaeacdc4702a096a3814f182529517597d9d.svg" src="_images/a57566f510e83b1867c9a868dc9cbaeacdc4702a096a3814f182529517597d9d.svg" />
</div>
</div>
<p>Here we create a 32 population stepping stone model, and
compute the matrix of lineage probabilities at 10, 100
and 1000 generations ago. We then plot the probability
of lineages sampled in the middle population
being present in the linearly-connected populations.
As we can see, 10 generations ago, the probability
of lineages being present falls off quickly with
distance, but as we go further back into the past
the possible location of lineages becomes more evenly
distributed in space.</p>
</section>
<section id="coalescence-rates-and-mean-times">
<span id="sec-demography-numerical-coalescence"></span><h3>Coalescence rates and mean times<a class="headerlink" href="#coalescence-rates-and-mean-times" title="Link to this heading">#</a></h3>
<p>Lineage probabilities tell us where lineages might possibly be,
which uses only information about migration rates and population splits.
Using information about population sizes,
we can also compute the <em>coalescence rates</em>,
using <a class="reference internal" href="api.html#msprime.DemographyDebugger.coalescence_rate_trajectory" title="msprime.DemographyDebugger.coalescence_rate_trajectory"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DemographyDebugger.coalescence_rate_trajectory()</span></code></a>.
For instance, here are the three pairwise coalescence rate trajectories
in the two-population model with migration above:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">demography</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="p">()</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">add_population</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">initial_size</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">set_migration_rate</span><span class="p">(</span><span class="n">source</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">0.005</span><span class="p">)</span>
<span class="n">demography</span><span class="o">.</span><span class="n">set_migration_rate</span><span class="p">(</span><span class="n">source</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">0.02</span><span class="p">)</span>

<span class="n">debug</span> <span class="o">=</span> <span class="n">demography</span><span class="o">.</span><span class="n">debug</span><span class="p">()</span>
<span class="n">T</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">500</span><span class="p">,</span> <span class="mi">51</span><span class="p">)</span>
<span class="n">RAA</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">debug</span><span class="o">.</span><span class="n">coalescence_rate_trajectory</span><span class="p">(</span><span class="n">T</span><span class="p">,</span> <span class="p">{</span><span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">})</span>
<span class="n">RBB</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">debug</span><span class="o">.</span><span class="n">coalescence_rate_trajectory</span><span class="p">(</span><span class="n">T</span><span class="p">,</span> <span class="p">{</span><span class="s2">&quot;B&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">})</span>
<span class="n">RAB</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">debug</span><span class="o">.</span><span class="n">coalescence_rate_trajectory</span><span class="p">(</span><span class="n">T</span><span class="p">,</span> <span class="p">{</span><span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">})</span>

<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">T</span><span class="p">,</span> <span class="n">RAA</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;two A samples&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">T</span><span class="p">,</span> <span class="n">RBB</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;two B samples&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">T</span><span class="p">,</span> <span class="n">RAB</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;one of each&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;time ago&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;coalecence rate&quot;</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/10089ec8a07a24c76054ec956d928bfc3619e7259063b962f393472b86f2f85f.svg" src="_images/10089ec8a07a24c76054ec956d928bfc3619e7259063b962f393472b86f2f85f.svg" />
</div>
</div>
<p>The coalescence rate of lineages sampled from different populations begins at zero,
and increases as the chance of them migrating to the same population increases.
Interestingly, the coalescence rate of two lineages in population “B”
is nonmonotonic, due to the asymmetric migration rates.</p>
<p>Many properties of genomes (e.g., mean genetic diversity) are determined
by mean coalescence times, which we can compute with
<a class="reference internal" href="api.html#msprime.DemographyDebugger.mean_coalescence_time" title="msprime.DemographyDebugger.mean_coalescence_time"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DemographyDebugger.mean_coalescence_time()</span></code></a>.
For instance, we might want to know how the three pairwise mean coalescence times
change as we make the migration rate less asymmetric:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">get_mean_coaltimes</span><span class="p">(</span><span class="n">m</span><span class="p">):</span>
    <span class="n">demography</span><span class="o">.</span><span class="n">set_migration_rate</span><span class="p">(</span><span class="n">source</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="n">m</span><span class="p">)</span>
    <span class="n">debug</span> <span class="o">=</span> <span class="n">demography</span><span class="o">.</span><span class="n">debug</span><span class="p">()</span>
    <span class="n">TAA</span> <span class="o">=</span> <span class="n">debug</span><span class="o">.</span><span class="n">mean_coalescence_time</span><span class="p">({</span><span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">})</span>
    <span class="n">TBB</span> <span class="o">=</span> <span class="n">debug</span><span class="o">.</span><span class="n">mean_coalescence_time</span><span class="p">({</span><span class="s2">&quot;B&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">})</span>
    <span class="n">TAB</span> <span class="o">=</span> <span class="n">debug</span><span class="o">.</span><span class="n">mean_coalescence_time</span><span class="p">({</span><span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">})</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">TAA</span><span class="p">,</span> <span class="n">TBB</span><span class="p">,</span> <span class="n">TAB</span><span class="p">]</span>

<span class="n">M</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mf">0.001</span><span class="p">,</span> <span class="mf">0.02</span><span class="p">,</span> <span class="mi">11</span><span class="p">)</span>
<span class="n">T</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">M</span><span class="p">),</span> <span class="mi">3</span><span class="p">))</span>
<span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">m</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">M</span><span class="p">):</span>
    <span class="n">T</span><span class="p">[</span><span class="n">j</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">get_mean_coaltimes</span><span class="p">(</span><span class="n">m</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">M</span><span class="p">,</span> <span class="n">T</span><span class="p">[:,</span><span class="mi">0</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;two A samples&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">M</span><span class="p">,</span> <span class="n">T</span><span class="p">[:,</span><span class="mi">1</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;two B samples&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">M</span><span class="p">,</span> <span class="n">T</span><span class="p">[:,</span><span class="mi">2</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;one of each&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;A→B migration rate&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;coalecence rate&quot;</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/8e9b59be79ab98ecb7aa0e803314e7a11cac5d021d65d77b9551aad5ca3c2a8f.svg" src="_images/8e9b59be79ab98ecb7aa0e803314e7a11cac5d021d65d77b9551aad5ca3c2a8f.svg" />
</div>
</div>
<p>From this, we can see that genetic diversity increases as the migration rate increases,
that divergence between the populations is always greater than diversity within either,
and that diversity within “B” goes from a level close to that found between populations
to that within “A”, as the “A→B” migration rate approaches the “B→A” rate.</p>
</section>
<section id="inverse-instantaneous-coalescence-rates">
<span id="sec-demography-numerical-trajectories"></span><h3>Inverse instantaneous coalescence rates<a class="headerlink" href="#inverse-instantaneous-coalescence-rates" title="Link to this heading">#</a></h3>
<p>Coalescence rates are analogous to effective population sizes,
in that if <span class="math notranslate nohighlight">\(r(t)\)</span> is the coalescence rate at time <span class="math notranslate nohighlight">\(t\)</span> ago
between two lineages from a population in <em>any</em> population model,
then a single population of changing size,
with <span class="math notranslate nohighlight">\(2 N(t) = 1/r(t)\)</span> genomes at time <span class="math notranslate nohighlight">\(t\)</span> ago,
has exactly the same distribution of coalescence times.
(The factor of two assumes the organisms are diploid,
and this assumes a continuous-time model.)
One reason this is helpful is because we often fit simplified demographic models
to more complex reality,
and we might expect models that fit a single population-size-over-time
to infer something closer to the inverse coalescence rate.
For more discussion, see
<a class="reference external" href="https://doi.org/10.1038/s41437-017-0005-6">Chikhi et al., 2018</a> and
<a class="reference external" href="https://doi.org/10.7554/eLife.54967">Adrion et al., 2020</a>.</p>
<p>For instance, let’s take the Gutenkunst et al 2009 “Out of Africa” model
we used above in the <a class="reference internal" href="#sec-demography-examples-population-tree"><span class="std std-ref">Population tree</span></a> section,
and compare inverse coalescence rates to actual population sizes.
To follow along, first you should define <code class="docutils literal notranslate"><span class="pre">demography</span></code> as in the section above.
Next we make a grid of times, <code class="docutils literal notranslate"><span class="pre">T</span></code>,  spanning the last 10,000 generations,
and calculate coalescence rates at these times:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># first define Gutenkunst et al model as above</span>
<span class="n">debug</span> <span class="o">=</span> <span class="n">demography</span><span class="o">.</span><span class="n">debug</span><span class="p">()</span>

<span class="n">T</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span>
    <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1000</span><span class="p">,</span> <span class="mi">2001</span><span class="p">),</span>
    <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">1000</span><span class="p">,</span> <span class="mf">1e4</span><span class="p">,</span> <span class="mi">401</span><span class="p">)[</span><span class="mi">1</span><span class="p">:]</span>
<span class="p">])</span>

<span class="n">R</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">T</span><span class="p">),</span> <span class="mi">3</span><span class="p">))</span>
<span class="n">R</span><span class="p">[:,</span><span class="mi">0</span><span class="p">],</span> <span class="n">_</span> <span class="o">=</span> <span class="n">debug</span><span class="o">.</span><span class="n">coalescence_rate_trajectory</span><span class="p">(</span><span class="n">T</span><span class="p">,</span> <span class="p">{</span><span class="s2">&quot;YRI&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">})</span>
<span class="n">R</span><span class="p">[:,</span><span class="mi">1</span><span class="p">],</span> <span class="n">_</span> <span class="o">=</span> <span class="n">debug</span><span class="o">.</span><span class="n">coalescence_rate_trajectory</span><span class="p">(</span><span class="n">T</span><span class="p">,</span> <span class="p">{</span><span class="s2">&quot;CEU&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">})</span>
<span class="n">R</span><span class="p">[:,</span><span class="mi">2</span><span class="p">],</span> <span class="n">_</span> <span class="o">=</span> <span class="n">debug</span><span class="o">.</span><span class="n">coalescence_rate_trajectory</span><span class="p">(</span><span class="n">T</span><span class="p">,</span> <span class="p">{</span><span class="s2">&quot;CHB&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">})</span>
</pre></div>
</div>
</div>
</div>
<p>The time values look a bit weird because when at first we set them to have
equal spacing across the whole time period
we got a warning about possible numerical inaccuracy
(and inaccurate trajectories).
This is because <a class="reference internal" href="api.html#msprime.DemographyDebugger.coalescence_rate_trajectory" title="msprime.DemographyDebugger.coalescence_rate_trajectory"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DemographyDebugger.coalescence_rate_trajectory()</span></code></a>
uses the provided time steps to calculate a discrete approximation
to the trajectory, and for this to be a good approximation
we need to take small time steps when the population is changing in size rapidly.
In this model, the rapid growth occurs recently, so we need a finer grid
of time steps in the last 1,000 generations.</p>
<p>Next we compute the “census sizes”, and plot these
along with the inverse instantaneous coalescence rates.
Since <a class="reference internal" href="api.html#msprime.DemographyDebugger.population_size_trajectory" title="msprime.DemographyDebugger.population_size_trajectory"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DemographyDebugger.population_size_trajectory()</span></code></a>
gives us population sizes for each of the six populations
(three modern and three ancient),
we need to do some work to copy the right population sizes over
(e.g., prior to the most recent split we should use sizes for the
shared “OOA” population for both the non-African populations).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">N</span> <span class="o">=</span> <span class="n">debug</span><span class="o">.</span><span class="n">population_size_trajectory</span><span class="p">(</span><span class="n">T</span><span class="p">)</span>
<span class="c1"># CEU-CHB merge</span>
<span class="n">which_T</span> <span class="o">=</span> <span class="p">(</span><span class="n">T</span> <span class="o">&gt;=</span> <span class="n">debug</span><span class="o">.</span><span class="n">epoch_start_time</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">):</span>
    <span class="n">N</span><span class="p">[</span><span class="n">which_T</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">N</span><span class="p">[</span><span class="n">which_T</span><span class="p">,</span> <span class="mi">3</span><span class="p">]</span>
<span class="c1"># OOA-Afr merge</span>
<span class="n">which_T</span> <span class="o">=</span> <span class="p">(</span><span class="n">T</span> <span class="o">&gt;=</span> <span class="n">debug</span><span class="o">.</span><span class="n">epoch_start_time</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
<span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">):</span>
    <span class="n">N</span><span class="p">[</span><span class="n">which_T</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">N</span><span class="p">[</span><span class="n">which_T</span><span class="p">,</span> <span class="mi">4</span><span class="p">]</span>
<span class="n">which_T</span> <span class="o">=</span> <span class="p">(</span><span class="n">T</span> <span class="o">&gt;=</span> <span class="n">debug</span><span class="o">.</span><span class="n">epoch_start_time</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
<span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">):</span>
    <span class="n">N</span><span class="p">[</span><span class="n">which_T</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">N</span><span class="p">[</span><span class="n">which_T</span><span class="p">,</span> <span class="mi">5</span><span class="p">]</span>

<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">T</span><span class="p">,</span> <span class="mi">1</span><span class="o">/</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">R</span><span class="p">),</span> <span class="n">label</span><span class="o">=</span><span class="p">[</span><span class="n">demography</span><span class="o">.</span><span class="n">populations</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">3</span><span class="p">)])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">T</span><span class="p">,</span> <span class="n">N</span><span class="p">[:,:</span><span class="mi">3</span><span class="p">],</span> <span class="s2">&quot;:&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;actual population sizes&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;time ago (generations)&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;population size&quot;</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/310f0f3a2ed3ef666ea612802098b28402c6ff760394aef154559631f57ac00c.svg" src="_images/310f0f3a2ed3ef666ea612802098b28402c6ff760394aef154559631f57ac00c.svg" />
</div>
</div>
<p>We can see that the inverse instantaneous coalescence rates
do not respond instantly to population size changes,
when there’s more than one population with migration between them.
(If we had a <em>single</em> population, the inverse coalescence rates
would exactly match the population size.)
So, if the true history matched this model
and we inferred the population-size-through-time using only “CEU” samples,
we might expect the inferred history to have a gradual increase back towards the out-of-Africa bottleneck,
instead of the actual sudden change at the bottleneck,
which is due to subsequent population structure.</p>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="mutations.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Mutation simulations</p>
      </div>
    </a>
    <a class="right-next"
       href="replication.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Randomness and replication</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quick-reference">Quick reference</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#model">Model</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#direction-of-time">Direction of time</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#definitions">Definitions</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#populations">Populations</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#sec-demography-populations-identifiers">Identifying populations</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#initial-size">Initial size</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#growth-rate">Growth rate</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#metadata">Metadata</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#life-cycle">Life cycle</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#default-sampling-time">Default sampling time</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#migration">Migration</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#events">Events</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#population-parameters-change">Population parameters change</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#migration-rate-change">Migration rate change</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#population-split">Population split</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#admixture">Admixture</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pulse-mass-migration">Pulse (mass) migration</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#instantaneous-bottleneck">Instantaneous bottleneck</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#simple-bottleneck">Simple bottleneck</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#demography-objects">Demography objects</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#debugging-tools">Debugging tools</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Demography objects</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#demographydebugger">DemographyDebugger</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-models">Example models</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#population-tree">Population tree</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sec-demography-examples-admixture">Admixture</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#trunk-population-models">Trunk population models</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#importing-exporting-model-definitions">Importing/exporting model definitions</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#demes">Demes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#species-trees">Species trees</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#numerical-utilities">Numerical utilities</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#population-sizes-over-time">Population sizes over time</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#possible-locations-of-lineages">Possible locations of lineages</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#lineage-probabilities">Lineage probabilities</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#coalescence-rates-and-mean-times">Coalescence rates and mean times</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#inverse-instantaneous-coalescence-rates">Inverse instantaneous coalescence rates</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Tskit Developers
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>