
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>pyslim.slim_metadata &#8212; PySLiM manual</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/pyslim/slim_metadata';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <script data-goatcounter="https://tskit.goatcounter.com/count" async src="//gc.zgo.at/count.js"></script>
</head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../introduction.html">
  
  
  
  
  
  
    <p class="title logo__title">pyslim<br/>
version 1.1.0
</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../overview.html">Overview</a></li>

<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Using pyslim</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../vignette_space.html">Vignette: A spatial simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../vignette_continuing.html">Vignette: Following up with more coalescent simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../vignette_coalescent_diversity.html">Vignette: Starting with diversity generated by coalescent simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../vignette_parallel_phylo.html">Vignette: Parallelizing SLiM simulations in a phylogenetic tree</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../time_units.html">Time units</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../metadata.html">Metadata</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../previous_versions.html">Migrating from previous versions of pyslim</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">pyslim reference</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../python_api.html">Python API</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Miscellaneous</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../development.html">Development</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">



<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1></h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <h1>Source code for pyslim.slim_metadata</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span><span class="w"> </span><span class="nn">tskit</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">json</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">._version</span><span class="w"> </span><span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">.provenance</span><span class="w"> </span><span class="kn">import</span> <span class="n">slim_provenance_version</span><span class="p">,</span> <span class="n">get_environment</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">pyslim</span><span class="w"> </span><span class="kn">import</span> <span class="n">INDIVIDUAL_ALIVE</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pyslim</span><span class="w"> </span><span class="kn">import</span> <span class="n">GENOME_TYPE_AUTOSOME</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pyslim</span><span class="w"> </span><span class="kn">import</span> <span class="n">GENOME_TYPE_X</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pyslim</span><span class="w"> </span><span class="kn">import</span> <span class="n">GENOME_TYPE_Y</span>

<span class="k">def</span><span class="w"> </span><span class="nf">is_vacant_num_bytes</span><span class="p">(</span><span class="n">num_chromosomes</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    TODO document</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># see _is_chrom_vacant</span>
    <span class="k">return</span> <span class="nb">int</span><span class="p">((</span><span class="n">num_chromosomes</span> <span class="o">+</span> <span class="mi">7</span><span class="p">)</span><span class="o">/</span><span class="mi">8</span><span class="p">)</span>


<span class="c1"># def _isvacant_values(vacancy):</span>
<span class="c1">#     &quot;&quot;&quot;</span>
<span class="c1">#     Returns the correct is_vacant metadata for the given vacancy pattern.</span>
<span class="c1"># </span>
<span class="c1">#     :param list vacancy: A list of booleans.</span>
<span class="c1">#     &quot;&quot;&quot;</span>
<span class="c1">#     out = [0 for _ in is_vacant_num_bytes(len(vacancy))]</span>
<span class="c1">#     powers = [1, 2, 4, 8, 16, 32, 64, 128]</span>
<span class="c1">#     for k, v in enumerate(vacancy):</span>
<span class="c1">#         if v:</span>
<span class="c1">#             n = k % 8</span>
<span class="c1">#             m = int(k / 8)</span>
<span class="c1">#             out[m] += powers[n]</span>
<span class="c1">#     return out</span>


<span class="c1"># These are copied from slim_globals.cpp, modified to be python not json</span>
<span class="c1"># by changing false-&gt;False and true-&gt;True, then printed with</span>
<span class="c1"># json.dump(..., indent=True), then lightly edited.</span>

<span class="n">_raw_slim_metadata_schemas</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;tree_sequence&quot;</span> <span class="p">:</span> <span class="p">{</span>
         <span class="s2">&quot;$schema&quot;</span><span class="p">:</span> <span class="s2">&quot;http://json-schema.org/schema#&quot;</span><span class="p">,</span>
         <span class="s2">&quot;codec&quot;</span><span class="p">:</span> <span class="s2">&quot;json&quot;</span><span class="p">,</span>
         <span class="s2">&quot;examples&quot;</span><span class="p">:</span> <span class="p">[</span>
          <span class="p">{</span>
           <span class="s2">&quot;SLiM&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;file_version&quot;</span><span class="p">:</span> <span class="s2">&quot;0.9&quot;</span><span class="p">,</span>
            <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;fox&quot;</span><span class="p">,</span>
            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;foxes on Catalina island&quot;</span><span class="p">,</span>
            <span class="s2">&quot;cycle&quot;</span><span class="p">:</span> <span class="mi">123</span><span class="p">,</span>
            <span class="s2">&quot;tick&quot;</span><span class="p">:</span> <span class="mi">123</span><span class="p">,</span>
            <span class="s2">&quot;model_type&quot;</span><span class="p">:</span> <span class="s2">&quot;WF&quot;</span><span class="p">,</span>
            <span class="s2">&quot;this_chromosome&quot;</span><span class="p">:</span> <span class="p">{</span>
             <span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
             <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
             <span class="s2">&quot;symbol&quot;</span><span class="p">:</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span>
             <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;autosome_1&quot;</span><span class="p">,</span>
             <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;chromosomes&quot;</span><span class="p">:</span> <span class="p">[</span>
             <span class="p">{</span>
              <span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
              <span class="s2">&quot;symbol&quot;</span><span class="p">:</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span>
              <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;autosome_1&quot;</span><span class="p">,</span>
              <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span>
             <span class="p">},</span>
             <span class="p">{</span>
              <span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="mi">35</span><span class="p">,</span>
              <span class="s2">&quot;symbol&quot;</span><span class="p">:</span> <span class="s2">&quot;MT&quot;</span><span class="p">,</span>
              <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;mtDNA&quot;</span><span class="p">,</span>
              <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;HF&quot;</span>
             <span class="p">}</span>
            <span class="p">],</span>
            <span class="s2">&quot;nucleotide_based&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="s2">&quot;separate_sexes&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
            <span class="s2">&quot;spatial_dimensionality&quot;</span><span class="p">:</span> <span class="s2">&quot;xy&quot;</span><span class="p">,</span>
            <span class="s2">&quot;spatial_periodicity&quot;</span><span class="p">:</span> <span class="s2">&quot;x&quot;</span>
           <span class="p">}</span>
          <span class="p">}</span>
         <span class="p">],</span>
         <span class="s2">&quot;properties&quot;</span><span class="p">:</span> <span class="p">{</span>
          <span class="s2">&quot;SLiM&quot;</span><span class="p">:</span> <span class="p">{</span>
           <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Top-level metadata for a SLiM tree sequence, file format version 0.9&quot;</span><span class="p">,</span>
           <span class="s2">&quot;properties&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;file_version&quot;</span><span class="p">:</span> <span class="p">{</span>
             <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The SLiM &#39;file format version&#39; of this tree sequence.&quot;</span><span class="p">,</span>
             <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="p">{</span>
             <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The SLiM species name represented by this tree sequence.&quot;</span><span class="p">,</span>
             <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="p">{</span>
             <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;A user-configurable description of the species represented by this tree sequence.&quot;</span><span class="p">,</span>
             <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;cycle&quot;</span><span class="p">:</span> <span class="p">{</span>
             <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The &#39;SLiM cycle&#39; counter when this tree sequence was recorded.&quot;</span><span class="p">,</span>
             <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;tick&quot;</span><span class="p">:</span> <span class="p">{</span>
             <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The &#39;SLiM tick&#39; counter when this tree sequence was recorded.&quot;</span><span class="p">,</span>
             <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;model_type&quot;</span><span class="p">:</span> <span class="p">{</span>
             <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The model type used for the last part of this simulation (WF or nonWF).&quot;</span><span class="p">,</span>
             <span class="s2">&quot;enum&quot;</span><span class="p">:</span> <span class="p">[</span>
              <span class="s2">&quot;WF&quot;</span><span class="p">,</span>
              <span class="s2">&quot;nonWF&quot;</span>
             <span class="p">],</span>
             <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;this_chromosome&quot;</span><span class="p">:</span> <span class="p">{</span>
             <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The chromosome represented by the tree sequence in this file.&quot;</span><span class="p">,</span>
             <span class="s2">&quot;properties&quot;</span><span class="p">:</span> <span class="p">{</span>
              <span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="p">{</span>
               <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;An integer identifier for the chromosome, unique within this set of tree sequences; often the chromosome number in the organism being represented, such as 1.&quot;</span><span class="p">,</span>
               <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
              <span class="p">},</span>
              <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="p">{</span>
               <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The (zero-based) index of this chromosome in the chromosomes metadata array (if present), which should match the information given here.&quot;</span><span class="p">,</span>
               <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
              <span class="p">},</span>
              <span class="s2">&quot;symbol&quot;</span><span class="p">:</span> <span class="p">{</span>
               <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;A short string symbol for the chromosome, unique within this set of tree sequences, such as </span><span class="se">\&quot;</span><span class="s2">1</span><span class="se">\&quot;</span><span class="s2"> or </span><span class="se">\&quot;</span><span class="s2">MT</span><span class="se">\&quot;</span><span class="s2">.&quot;</span><span class="p">,</span>
               <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span>
              <span class="p">},</span>
              <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="p">{</span>
               <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;A user-specified name for the chromosome, such as an accession identifier.&quot;</span><span class="p">,</span>
               <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span>
              <span class="p">},</span>
              <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="p">{</span>
               <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The type of chromosome, as specified by SLiM.&quot;</span><span class="p">,</span>
               <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span>
              <span class="p">}</span>
             <span class="p">},</span>
             <span class="s2">&quot;required&quot;</span><span class="p">:</span> <span class="p">[</span>
              <span class="s2">&quot;id&quot;</span><span class="p">,</span>
              <span class="s2">&quot;index&quot;</span><span class="p">,</span>
              <span class="s2">&quot;symbol&quot;</span><span class="p">,</span>
              <span class="s2">&quot;type&quot;</span>
             <span class="p">],</span>
             <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;object&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;chromosomes&quot;</span><span class="p">:</span> <span class="p">{</span>
             <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The chromosomes represented by the collection of tree sequences, of which this tree sequence is one member.&quot;</span><span class="p">,</span>
             <span class="s2">&quot;items&quot;</span><span class="p">:</span> <span class="p">{</span>
              <span class="s2">&quot;properties&quot;</span><span class="p">:</span> <span class="p">{</span>
               <span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;An integer identifier for the chromosome, unique within this set of tree sequences; often the chromosome number in the organism being represented, such as 1.&quot;</span><span class="p">,</span>
                <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
               <span class="p">},</span>
               <span class="s2">&quot;symbol&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;A short string symbol for the chromosome, unique within this set of tree sequences, such as </span><span class="se">\&quot;</span><span class="s2">1</span><span class="se">\&quot;</span><span class="s2"> or </span><span class="se">\&quot;</span><span class="s2">MT</span><span class="se">\&quot;</span><span class="s2">.&quot;</span><span class="p">,</span>
                <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span>
               <span class="p">},</span>
               <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;A user-specified name for the chromosome, such as an accession identifier.&quot;</span><span class="p">,</span>
                <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span>
               <span class="p">},</span>
               <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The type of chromosome, as specified by SLiM.&quot;</span><span class="p">,</span>
                <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span>
               <span class="p">}</span>
              <span class="p">},</span>
              <span class="s2">&quot;required&quot;</span><span class="p">:</span> <span class="p">[</span>
               <span class="s2">&quot;id&quot;</span><span class="p">,</span>
               <span class="s2">&quot;symbol&quot;</span><span class="p">,</span>
               <span class="s2">&quot;type&quot;</span>
              <span class="p">],</span>
              <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;object&quot;</span>
             <span class="p">},</span>
             <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;array&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;nucleotide_based&quot;</span><span class="p">:</span> <span class="p">{</span>
             <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Whether the simulation was nucleotide-based.&quot;</span><span class="p">,</span>
             <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;boolean&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;separate_sexes&quot;</span><span class="p">:</span> <span class="p">{</span>
             <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Whether the simulation had separate sexes.&quot;</span><span class="p">,</span>
             <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;boolean&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;spatial_dimensionality&quot;</span><span class="p">:</span> <span class="p">{</span>
             <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The spatial dimensionality of the simulation.&quot;</span><span class="p">,</span>
             <span class="s2">&quot;enum&quot;</span><span class="p">:</span> <span class="p">[</span>
              <span class="s2">&quot;&quot;</span><span class="p">,</span>
              <span class="s2">&quot;x&quot;</span><span class="p">,</span>
              <span class="s2">&quot;xy&quot;</span><span class="p">,</span>
              <span class="s2">&quot;xyz&quot;</span>
             <span class="p">],</span>
             <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;spatial_periodicity&quot;</span><span class="p">:</span> <span class="p">{</span>
             <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The spatial periodicity of the simulation.&quot;</span><span class="p">,</span>
             <span class="s2">&quot;enum&quot;</span><span class="p">:</span> <span class="p">[</span>
              <span class="s2">&quot;&quot;</span><span class="p">,</span>
              <span class="s2">&quot;x&quot;</span><span class="p">,</span>
              <span class="s2">&quot;y&quot;</span><span class="p">,</span>
              <span class="s2">&quot;z&quot;</span><span class="p">,</span>
              <span class="s2">&quot;xy&quot;</span><span class="p">,</span>
              <span class="s2">&quot;xz&quot;</span><span class="p">,</span>
              <span class="s2">&quot;yz&quot;</span><span class="p">,</span>
              <span class="s2">&quot;xyz&quot;</span>
             <span class="p">],</span>
             <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;stage&quot;</span><span class="p">:</span> <span class="p">{</span>
             <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The stage of the SLiM life cycle when this tree sequence was recorded.&quot;</span><span class="p">,</span>
             <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span>
            <span class="p">}</span>
           <span class="p">},</span>
           <span class="s2">&quot;required&quot;</span><span class="p">:</span> <span class="p">[</span>
            <span class="s2">&quot;model_type&quot;</span><span class="p">,</span>
            <span class="s2">&quot;tick&quot;</span><span class="p">,</span>
            <span class="s2">&quot;file_version&quot;</span><span class="p">,</span>
            <span class="s2">&quot;spatial_dimensionality&quot;</span><span class="p">,</span>
            <span class="s2">&quot;spatial_periodicity&quot;</span><span class="p">,</span>
            <span class="s2">&quot;this_chromosome&quot;</span><span class="p">,</span>
            <span class="s2">&quot;separate_sexes&quot;</span><span class="p">,</span>
            <span class="s2">&quot;nucleotide_based&quot;</span>
           <span class="p">],</span>
           <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;object&quot;</span>
          <span class="p">}</span>
         <span class="p">},</span>
         <span class="s2">&quot;required&quot;</span><span class="p">:</span> <span class="p">[</span>
          <span class="s2">&quot;SLiM&quot;</span>
         <span class="p">],</span>
         <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;object&quot;</span>
    <span class="p">}</span>
    <span class="p">,</span>
    <span class="s2">&quot;edge&quot;</span> <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
    <span class="s2">&quot;site&quot;</span> <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
    <span class="s2">&quot;mutation&quot;</span> <span class="p">:</span> 
    <span class="p">{</span>
        <span class="s2">&quot;$schema&quot;</span><span class="p">:</span> <span class="s2">&quot;http://json-schema.org/schema#&quot;</span><span class="p">,</span>
        <span class="s2">&quot;additionalProperties&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="s2">&quot;codec&quot;</span><span class="p">:</span> <span class="s2">&quot;struct&quot;</span><span class="p">,</span>
        <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;SLiM schema for mutation metadata.&quot;</span><span class="p">,</span>
        <span class="s2">&quot;examples&quot;</span><span class="p">:</span> <span class="p">[</span>
            <span class="p">{</span>
                <span class="s2">&quot;mutation_list&quot;</span><span class="p">:</span> <span class="p">[</span>
                    <span class="p">{</span>
                        <span class="s2">&quot;mutation_type&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
                        <span class="s2">&quot;nucleotide&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
                        <span class="s2">&quot;selection_coeff&quot;</span><span class="p">:</span> <span class="o">-</span><span class="mf">0.2</span><span class="p">,</span>
                        <span class="s2">&quot;slim_time&quot;</span><span class="p">:</span> <span class="mi">243</span><span class="p">,</span>
                        <span class="s2">&quot;subpopulation&quot;</span><span class="p">:</span> <span class="mi">0</span>
                    <span class="p">}</span>
                <span class="p">]</span>
            <span class="p">}</span>
        <span class="p">],</span>
        <span class="s2">&quot;properties&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;mutation_list&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;items&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;additionalProperties&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                    <span class="s2">&quot;properties&quot;</span><span class="p">:</span> <span class="p">{</span>
                        <span class="s2">&quot;mutation_type&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;i&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The index of this mutation&#39;s mutationType.&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;nucleotide&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;b&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The nucleotide for this mutation (0=A , 1=C , 2=G, 3=T, or -1 for none)&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">5</span><span class="p">,</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;selection_coeff&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;f&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;This mutation&#39;s selection coefficient.&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;slim_time&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;i&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The SLiM tick counter when this mutation occurred.&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;subpopulation&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;i&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The ID of the subpopulation this mutation occurred in.&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
                        <span class="p">}</span>
                    <span class="p">},</span>
                    <span class="s2">&quot;required&quot;</span><span class="p">:</span> <span class="p">[</span>
                        <span class="s2">&quot;mutation_type&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;selection_coeff&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;subpopulation&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;slim_time&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;nucleotide&quot;</span>
                    <span class="p">],</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;object&quot;</span>
                <span class="p">},</span>
                <span class="s2">&quot;noLengthEncodingExhaustBuffer&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;array&quot;</span>
            <span class="p">}</span>
        <span class="p">},</span>
        <span class="s2">&quot;required&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;mutation_list&quot;</span>
        <span class="p">],</span>
        <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;object&quot;</span>
    <span class="p">},</span>
    <span class="s2">&quot;node&quot;</span> <span class="p">:</span> 
    <span class="p">{</span>
     <span class="s2">&quot;$schema&quot;</span><span class="p">:</span> <span class="s2">&quot;http://json-schema.org/schema#&quot;</span><span class="p">,</span>
     <span class="s2">&quot;additionalProperties&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
     <span class="s2">&quot;codec&quot;</span><span class="p">:</span> <span class="s2">&quot;struct&quot;</span><span class="p">,</span>
     <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;SLiM schema for node metadata.&quot;</span><span class="p">,</span>
     <span class="s2">&quot;examples&quot;</span><span class="p">:</span> <span class="p">[</span>
      <span class="p">{</span>
       <span class="s2">&quot;slim_id&quot;</span><span class="p">:</span> <span class="mi">123</span><span class="p">,</span>
       <span class="s2">&quot;is_vacant&quot;</span><span class="p">:</span> <span class="mi">0</span>
      <span class="p">}</span>
     <span class="p">],</span>
     <span class="s2">&quot;properties&quot;</span><span class="p">:</span> <span class="p">{</span>
      <span class="s2">&quot;slim_id&quot;</span><span class="p">:</span> <span class="p">{</span>
       <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;q&quot;</span><span class="p">,</span>
       <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The &#39;pedigree ID&#39; of the haplosomes associated with this node in SLiM.&quot;</span><span class="p">,</span>
       <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
       <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
      <span class="p">},</span>
      <span class="s2">&quot;is_vacant&quot;</span><span class="p">:</span> <span class="p">{</span>
       <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;A vector of byte (uint8_t) values, with each bit representing whether the node represents a vacant position, either unused or a null haplosome (1), or a non-null haplosome (0), in the corresponding chromosome. This field encodes vacancy for all of the chromosomes in the model, not just the chromosome represented in this file (so that the node table is identical across all chromosomes for a multi-chromosome model). Each chromosome receives one bit here; there are two node table entries per individual, used for the two haplosomes of every chromosome, so only one bit is needed in each entry (making two bits total per chromosome, across the two node table entries). The least significant bit of the first byte is used first (for one haplosome of the first chromosome); the most significant bit of the last byte is used last. The number of bytes present in this field is indicated by this schema&#39;s &#39;binaryFormat&#39; field, which is variable (!), and can also be deduced from the number of chromosomes in the model as given in the top-level &#39;chromosomes&#39; metadata key, which should always be present if this metadata is present.&quot;</span><span class="p">,</span>
       <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
       <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;array&quot;</span><span class="p">,</span>
       <span class="s2">&quot;length&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>  <span class="c1"># MAY NEED TO BE CHANGED (in SLiM code is &quot;%d&quot;)</span>
       <span class="s2">&quot;items&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span><span class="p">,</span>
        <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;B&quot;</span>
       <span class="p">}</span>
      <span class="p">}</span>
     <span class="p">},</span>
     <span class="s2">&quot;required&quot;</span><span class="p">:</span> <span class="p">[</span>
      <span class="s2">&quot;slim_id&quot;</span><span class="p">,</span>
      <span class="s2">&quot;is_vacant&quot;</span>
     <span class="p">],</span>
     <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="p">[</span>
      <span class="s2">&quot;object&quot;</span><span class="p">,</span>
      <span class="s2">&quot;null&quot;</span>
     <span class="p">]</span>
    <span class="p">},</span>
    <span class="s2">&quot;individual&quot;</span> <span class="p">:</span>
    <span class="p">{</span>
        <span class="s2">&quot;$schema&quot;</span><span class="p">:</span> <span class="s2">&quot;http://json-schema.org/schema#&quot;</span><span class="p">,</span>
        <span class="s2">&quot;additionalProperties&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="s2">&quot;codec&quot;</span><span class="p">:</span> <span class="s2">&quot;struct&quot;</span><span class="p">,</span>
        <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;SLiM schema for individual metadata.&quot;</span><span class="p">,</span>
        <span class="s2">&quot;examples&quot;</span><span class="p">:</span> <span class="p">[</span>
            <span class="p">{</span>
                <span class="s2">&quot;age&quot;</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span>
                <span class="s2">&quot;flags&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
                <span class="s2">&quot;pedigree_id&quot;</span><span class="p">:</span> <span class="mi">123</span><span class="p">,</span>
                <span class="s2">&quot;pedigree_p1&quot;</span><span class="p">:</span> <span class="mi">12</span><span class="p">,</span>
                <span class="s2">&quot;pedigree_p2&quot;</span><span class="p">:</span> <span class="mi">23</span><span class="p">,</span>
                <span class="s2">&quot;sex&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
                <span class="s2">&quot;subpopulation&quot;</span><span class="p">:</span> <span class="mi">0</span>
            <span class="p">}</span>
        <span class="p">],</span>
        <span class="s2">&quot;flags&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;SLIM_INDIVIDUAL_METADATA_MIGRATED&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Whether this individual was a migrant, either in the tick when the tree sequence &quot;</span>
                               <span class="s2">&quot;was written out (if the individual was alive then), or in the tick of the last time &quot;</span>
                               <span class="s2">&quot;they were Remembered (if not).&quot;</span><span class="p">,</span>
                <span class="s2">&quot;value&quot;</span><span class="p">:</span> <span class="mi">1</span>
            <span class="p">}</span>
        <span class="p">},</span>
        <span class="s2">&quot;properties&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;age&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;i&quot;</span><span class="p">,</span>
                <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The age of this individual, either when the tree sequence was written out &quot;</span>
                               <span class="s2">&quot;(if the individual was alive then), or the last time they were Remembered (if not).&quot;</span><span class="p">,</span>
                               <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span>
                               <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;flags&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;I&quot;</span><span class="p">,</span>
                <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Other information about the individual: see &#39;flags&#39;.&quot;</span><span class="p">,</span>
                <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">7</span><span class="p">,</span>
                <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;pedigree_id&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;q&quot;</span><span class="p">,</span>
                <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The &#39;pedigree ID&#39; of this individual in SLiM.&quot;</span><span class="p">,</span>
                <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
                <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;pedigree_p1&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;q&quot;</span><span class="p">,</span>
                <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The &#39;pedigree ID&#39; of this individual&#39;s first parent in SLiM.&quot;</span><span class="p">,</span>
                <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span>
                <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;pedigree_p2&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;q&quot;</span><span class="p">,</span>
                <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The &#39;pedigree ID&#39; of this individual&#39;s second parent in SLiM.&quot;</span><span class="p">,</span>
                <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
                <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;sex&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;i&quot;</span><span class="p">,</span>
                <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The sex of the individual (0 for female, 1 for male, -1 for hermaphrodite).&quot;</span><span class="p">,</span>
                <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">6</span><span class="p">,</span>
                <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;subpopulation&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;i&quot;</span><span class="p">,</span>
                <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The ID of the subpopulation the individual was part of, either when the tree sequence &quot;</span>
                               <span class="s2">&quot;was written out (if the individual was alive then), or the last time they were Remembered (if not).&quot;</span><span class="p">,</span>
                               <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">5</span><span class="p">,</span>
                               <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
            <span class="p">}</span>
        <span class="p">},</span>
        <span class="s2">&quot;required&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;pedigree_id&quot;</span><span class="p">,</span>
                <span class="s2">&quot;pedigree_p1&quot;</span><span class="p">,</span>
                <span class="s2">&quot;pedigree_p2&quot;</span><span class="p">,</span>
                <span class="s2">&quot;age&quot;</span><span class="p">,</span>
                <span class="s2">&quot;subpopulation&quot;</span><span class="p">,</span>
                <span class="s2">&quot;sex&quot;</span><span class="p">,</span>
                <span class="s2">&quot;flags&quot;</span>
            <span class="p">],</span>
        <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;object&quot;</span>
    <span class="p">},</span>
    <span class="s2">&quot;population&quot;</span><span class="p">:</span> 
    <span class="p">{</span>
            <span class="s2">&quot;$schema&quot;</span><span class="p">:</span> <span class="s2">&quot;http://json-schema.org/schema#&quot;</span><span class="p">,</span>
            <span class="s2">&quot;additionalProperties&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
            <span class="s2">&quot;codec&quot;</span><span class="p">:</span> <span class="s2">&quot;json&quot;</span><span class="p">,</span>
            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;SLiM schema for population metadata.&quot;</span><span class="p">,</span>
            <span class="s2">&quot;examples&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="p">{</span>
                    <span class="s2">&quot;bounds_x0&quot;</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span>
                    <span class="s2">&quot;bounds_x1&quot;</span><span class="p">:</span> <span class="mf">100.0</span><span class="p">,</span>
                    <span class="s2">&quot;bounds_y0&quot;</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span>
                    <span class="s2">&quot;bounds_y1&quot;</span><span class="p">:</span> <span class="mf">100.0</span><span class="p">,</span>
                    <span class="s2">&quot;female_cloning_fraction&quot;</span><span class="p">:</span> <span class="mf">0.25</span><span class="p">,</span>
                    <span class="s2">&quot;male_cloning_fraction&quot;</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span>
                    <span class="s2">&quot;migration_records&quot;</span><span class="p">:</span> <span class="p">[</span>
                        <span class="p">{</span>
                            <span class="s2">&quot;migration_rate&quot;</span><span class="p">:</span> <span class="mf">0.9</span><span class="p">,</span>
                            <span class="s2">&quot;source_subpop&quot;</span><span class="p">:</span> <span class="mi">1</span>
                        <span class="p">},</span>
                        <span class="p">{</span>
                            <span class="s2">&quot;migration_rate&quot;</span><span class="p">:</span> <span class="mf">0.1</span><span class="p">,</span>
                            <span class="s2">&quot;source_subpop&quot;</span><span class="p">:</span> <span class="mi">2</span>
                        <span class="p">}</span>
                    <span class="p">],</span>
                    <span class="s2">&quot;selfing_fraction&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">,</span>
                    <span class="s2">&quot;sex_ratio&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">,</span>
                    <span class="s2">&quot;slim_id&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span>
                    <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;p2&quot;</span>
                <span class="p">}</span>
            <span class="p">],</span>
            <span class="s2">&quot;properties&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;bounds_x0&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The minimum x-coordinate in this subpopulation.&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span>
                <span class="p">},</span>
                <span class="s2">&quot;bounds_x1&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The maximum x-coordinate in this subpopulation.&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span>
                <span class="p">},</span>
                <span class="s2">&quot;bounds_y0&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The minimum y-coordinate in this subpopulation.&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span>
                <span class="p">},</span>
                <span class="s2">&quot;bounds_y1&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The maximum y-coordinate in this subpopulation.&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span>
                <span class="p">},</span>
                <span class="s2">&quot;bounds_z0&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The minimum z-coordinate in this subpopulation.&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span>
                <span class="p">},</span>
                <span class="s2">&quot;bounds_z1&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The maximum z-coordinate in this subpopulation.&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span>
                <span class="p">},</span>
                <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;A description of this subpopulation.&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span>
                <span class="p">},</span>
                <span class="s2">&quot;female_cloning_fraction&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The frequency with which females in this subpopulation reproduce clonally (for WF models).&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span>
                <span class="p">},</span>
                <span class="s2">&quot;male_cloning_fraction&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The frequency with which males in this subpopulation reproduce clonally (for WF models).&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span>
                <span class="p">},</span>
                <span class="s2">&quot;migration_records&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;items&quot;</span><span class="p">:</span> <span class="p">{</span>
                        <span class="s2">&quot;properties&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;migration_rate&quot;</span><span class="p">:</span> <span class="p">{</span>
                                <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The fraction of children in this subpopulation that are composed of &#39;migrants&#39; from the source subpopulation (in WF models).&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span>
                            <span class="p">},</span>
                            <span class="s2">&quot;source_subpop&quot;</span><span class="p">:</span> <span class="p">{</span>
                                <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The ID of the subpopulation migrants come from (in WF models).&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
                            <span class="p">}</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;required&quot;</span><span class="p">:</span> <span class="p">[</span>
                            <span class="s2">&quot;source_subpop&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;migration_rate&quot;</span>
                        <span class="p">],</span>
                        <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;object&quot;</span>
                    <span class="p">},</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;array&quot;</span>
                <span class="p">},</span>
                <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;A human-readable name for this subpopulation.&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span>
                <span class="p">},</span>
                <span class="s2">&quot;selfing_fraction&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The frequency with which individuals in this subpopulation self (for WF models).&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span>
                <span class="p">},</span>
                <span class="s2">&quot;sex_ratio&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;This subpopulation&#39;s sex ratio (for WF models).&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span>
                <span class="p">},</span>
                <span class="s2">&quot;slim_id&quot;</span><span class="p">:</span> <span class="p">{</span>
                        <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The ID of this population in SLiM. Note that this is called a &#39;subpopulation&#39; in SLiM.&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
                    <span class="p">}</span>
        <span class="p">},</span>
        <span class="s2">&quot;required&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;object&quot;</span><span class="p">,</span>
                <span class="s2">&quot;null&quot;</span>
        <span class="p">]</span>
    <span class="p">}</span>
<span class="p">}</span>


<span class="k">def</span><span class="w"> </span><span class="nf">slim_node_metadata_schema</span><span class="p">(</span><span class="n">num_chromosomes</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Unlike other schema, the node metadata schema depends on the number of</span>
<span class="sd">    chromosomes in a multichromosome simulation, and</span>
<span class="sd">    {data}`.slim_metadata_schemas`</span>
<span class="sd">    returns the schema for a single-chromosome simulation. This function</span>
<span class="sd">    returns the correct schema for a simulation with arbitrary number of</span>
<span class="sd">    chromosomes. (The resulting schemas only differ in the</span>
<span class="sd">    ``schema[&quot;properties&quot;][&quot;is_vacant&quot;][&quot;length&quot;]`` property,</span>
<span class="sd">    which is set to ``floor(num_chromosomes+7)/8``, as described in</span>
<span class="sd">    the SLiM manual.)</span>

<span class="sd">    :param int num_chromosomes: The number of chromosomes in the model.</span>
<span class="sd">    :return tskit.MetadataSchema: The metadata schema to be used</span>
<span class="sd">        in the node table.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># From the SLiM manual on is_vacant:</span>
    <span class="c1"># M bytes (uint8_t): a series of bytes comprising a bitfield of is_vacant</span>
    <span class="c1"># values, true (1) if this node represents a vacant haplosome for a given</span>
    <span class="c1"># chromosome, false (0) otherwise. For chromosomes with indices 0...N−1, the</span>
    <span class="c1"># chromosome with index k has its is_vacant bit in bit k%8 of byte k/8, where</span>
    <span class="c1"># byte 0 is the first byte in the series of bytes provided, and bit 0 is the</span>
    <span class="c1"># least-significant bit, the one with value 0x01 (hexadecimal 1). The number</span>
    <span class="c1"># of bytes present, M, is equal to (N+7)/8, the minimum number of bytes</span>
    <span class="c1"># necessary. The operators / and % here are integer divide (rounding down)</span>
    <span class="c1"># and integer modulo, respectively.</span>
    <span class="n">num_bytes</span> <span class="o">=</span> <span class="n">is_vacant_num_bytes</span><span class="p">(</span><span class="n">num_chromosomes</span><span class="p">)</span>
    <span class="n">schema</span> <span class="o">=</span> <span class="n">_raw_slim_metadata_schemas</span><span class="p">[</span><span class="s2">&quot;node&quot;</span><span class="p">]</span>
    <span class="n">schema</span><span class="p">[</span><span class="s2">&quot;properties&quot;</span><span class="p">][</span><span class="s2">&quot;is_vacant&quot;</span><span class="p">][</span><span class="s2">&quot;length&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">num_bytes</span>
    <span class="k">return</span> <span class="n">tskit</span><span class="o">.</span><span class="n">MetadataSchema</span><span class="p">(</span><span class="n">schema</span><span class="p">)</span>


<span class="n">slim_metadata_schemas</span> <span class="o">=</span> <span class="p">{</span>
    <span class="n">k</span><span class="p">:</span> <span class="n">tskit</span><span class="o">.</span><span class="n">MetadataSchema</span><span class="p">(</span><span class="n">_raw_slim_metadata_schemas</span><span class="p">[</span><span class="n">k</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">_raw_slim_metadata_schemas</span>
    <span class="k">if</span> <span class="n">k</span> <span class="o">!=</span> <span class="s2">&quot;node&quot;</span>
<span class="p">}</span>
<span class="n">slim_metadata_schemas</span><span class="p">[</span><span class="s2">&quot;node&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">slim_node_metadata_schema</span><span class="p">()</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">A dictionary containing the metadata schemas used by SLiM for each of the tables</span>
<span class="sd">and for top-level metadata. **Warning:** node metadata schema depends on the</span>
<span class="sd">number of chromosomes, and so `slim_metadata_schemas[&quot;node&quot;]` is not valid for</span>
<span class="sd">a SLiM simulation with more than 8 chromosomes.</span>
<span class="sd">&quot;&quot;&quot;</span>


<div class="viewcode-block" id="default_slim_metadata">
<a class="viewcode-back" href="../../metadata.html#pyslim.default_slim_metadata">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">default_slim_metadata</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">num_chromosomes</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns default metadata of type ``name``, where ``name`` is one of</span>
<span class="sd">    &quot;tree_sequence&quot;, &quot;edge&quot;, &quot;site&quot;, &quot;mutation&quot;, &quot;mutation_list_entry&quot;,</span>
<span class="sd">    &quot;node&quot;, &quot;individual&quot;, or &quot;population&quot;.</span>

<span class="sd">    :param str name: The type of metadata requested.</span>
<span class="sd">    :rtype dict:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;tree_sequence&quot;</span><span class="p">:</span>
        <span class="n">out</span> <span class="o">=</span> <span class="p">{</span>
             <span class="s2">&quot;SLiM&quot;</span> <span class="p">:</span> <span class="p">{</span>
                 <span class="s2">&quot;model_type&quot;</span> <span class="p">:</span> <span class="s2">&quot;nonWF&quot;</span><span class="p">,</span>
                 <span class="s2">&quot;cycle&quot;</span> <span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
                 <span class="s2">&quot;tick&quot;</span> <span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
                 <span class="s2">&quot;file_version&quot;</span> <span class="p">:</span> <span class="n">slim_file_version</span><span class="p">,</span>
                 <span class="s2">&quot;spatial_dimensionality&quot;</span> <span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
                 <span class="s2">&quot;spatial_periodicity&quot;</span> <span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
                 <span class="s2">&quot;separate_sexes&quot;</span> <span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                 <span class="s2">&quot;nucleotide_based&quot;</span> <span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                 <span class="s2">&quot;stage&quot;</span> <span class="p">:</span> <span class="s2">&quot;late&quot;</span><span class="p">,</span>
                 <span class="s2">&quot;name&quot;</span> <span class="p">:</span> <span class="s2">&quot;sim&quot;</span><span class="p">,</span>
                 <span class="s2">&quot;description&quot;</span> <span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
                 <span class="s2">&quot;this_chromosome&quot;</span> <span class="p">:</span> <span class="p">{</span>
                     <span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
                     <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
                     <span class="s2">&quot;symbol&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                     <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span>
                 <span class="p">},</span>
                 <span class="s2">&quot;chromosomes&quot;</span><span class="p">:</span> <span class="p">[{</span>
                     <span class="s1">&#39;id&#39;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
                     <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
                     <span class="s1">&#39;symbol&#39;</span><span class="p">:</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span>
                     <span class="s1">&#39;type&#39;</span><span class="p">:</span> <span class="s1">&#39;A&#39;</span>
                 <span class="p">}],</span>
             <span class="p">}</span>
         <span class="p">}</span>
    <span class="k">elif</span> <span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;edge&quot;</span><span class="p">:</span>
        <span class="n">out</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">elif</span> <span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;site&quot;</span><span class="p">:</span>
        <span class="n">out</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">elif</span> <span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;mutation&quot;</span><span class="p">:</span>
        <span class="n">out</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;mutation_list&quot;</span><span class="p">:</span> <span class="p">[]</span>
        <span class="p">}</span>
    <span class="k">elif</span> <span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;mutation_list_entry&quot;</span><span class="p">:</span>
        <span class="n">out</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;mutation_type&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;selection_coeff&quot;</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span>
            <span class="s2">&quot;subpopulation&quot;</span><span class="p">:</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span><span class="p">,</span>
            <span class="s2">&quot;slim_time&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;nucleotide&quot;</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span>
        <span class="p">}</span>
    <span class="k">elif</span> <span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;node&quot;</span><span class="p">:</span>
        <span class="n">out</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;slim_id&quot;</span><span class="p">:</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span><span class="p">,</span>
            <span class="s2">&quot;is_vacant&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">is_vacant_num_bytes</span><span class="p">(</span><span class="n">num_chromosomes</span><span class="p">))],</span>
        <span class="p">}</span>
    <span class="k">elif</span> <span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;individual&quot;</span><span class="p">:</span>
        <span class="n">out</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;pedigree_id&quot;</span><span class="p">:</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span><span class="p">,</span>
            <span class="s2">&quot;age&quot;</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span>
            <span class="s2">&quot;subpopulation&quot;</span><span class="p">:</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span><span class="p">,</span>
            <span class="s2">&quot;sex&quot;</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span>
            <span class="s2">&quot;flags&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;pedigree_p1&quot;</span><span class="p">:</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span><span class="p">,</span>
            <span class="s2">&quot;pedigree_p2&quot;</span><span class="p">:</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span><span class="p">,</span>
        <span class="p">}</span>
    <span class="k">elif</span> <span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;population&quot;</span><span class="p">:</span>
        <span class="n">out</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;slim_id&quot;</span><span class="p">:</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span><span class="p">,</span>
            <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;default&quot;</span><span class="p">,</span>
            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
            <span class="s2">&quot;selfing_fraction&quot;</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span>
            <span class="s2">&quot;female_cloning_fraction&quot;</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span>
            <span class="s2">&quot;male_cloning_fraction&quot;</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span>
            <span class="s2">&quot;sex_ratio&quot;</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span>
            <span class="s2">&quot;bounds_x0&quot;</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span>
            <span class="s2">&quot;bounds_x1&quot;</span><span class="p">:</span> <span class="mf">1.0</span><span class="p">,</span>
            <span class="s2">&quot;bounds_y0&quot;</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span>
            <span class="s2">&quot;bounds_y1&quot;</span><span class="p">:</span> <span class="mf">1.0</span><span class="p">,</span>
            <span class="s2">&quot;bounds_z0&quot;</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span>
            <span class="s2">&quot;bounds_z1&quot;</span><span class="p">:</span> <span class="mf">1.0</span><span class="p">,</span>
            <span class="s2">&quot;migration_records&quot;</span><span class="p">:</span> <span class="p">[]</span>
        <span class="p">}</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;Unknown metadata request: name should be one of &#39;tree_sequence&#39;, &quot;</span>
            <span class="s2">&quot;&#39;edge&#39;, &#39;site&#39;, &#39;mutation&#39;, &#39;mutation_list_entry&#39;, &#39;node&#39;, &quot;</span>
            <span class="s2">&quot;&#39;individual&#39;, or &#39;population&#39;.&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">out</span></div>



<span class="c1">###########</span>
<span class="c1"># Top-level, a.k.a., tree sequence metadata</span>
<span class="c1">###########</span>

<span class="k">def</span><span class="w"> </span><span class="nf">set_tree_sequence_metadata</span><span class="p">(</span><span class="n">tables</span><span class="p">,</span>
        <span class="n">model_type</span><span class="p">,</span>
        <span class="n">tick</span><span class="p">,</span>
        <span class="n">cycle</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">spatial_dimensionality</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span>
        <span class="n">spatial_periodicity</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span>
        <span class="n">separate_sexes</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">nucleotide_based</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">stage</span><span class="o">=</span><span class="s1">&#39;late&#39;</span><span class="p">,</span>
        <span class="n">name</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span>
        <span class="n">description</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span>
        <span class="n">this_chromosome</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">chromosomes</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">file_version</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">set_table_schemas</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">file_version</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">file_version</span> <span class="o">=</span> <span class="n">slim_file_version</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">tables</span><span class="o">.</span><span class="n">metadata</span><span class="p">,</span> <span class="nb">bytes</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">tables</span><span class="o">.</span><span class="n">metadata</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Tree sequence has top-level metadata but no schema: this is a problem &quot;</span>
                             <span class="s2">&quot;since pyslim is trying to add to the metadata.&quot;</span><span class="p">)</span>
        <span class="n">schema_dict</span> <span class="o">=</span> <span class="n">slim_metadata_schemas</span><span class="p">[</span><span class="s1">&#39;tree_sequence&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">schema</span>
        <span class="n">metadata_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># we need to keep other keys in the metadata (and schema) if there are any</span>
        <span class="n">schema_dict</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">metadata_schema</span><span class="o">.</span><span class="n">schema</span>
        <span class="n">metadata_dict</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">metadata</span>
    <span class="k">if</span> <span class="n">cycle</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">cycle</span> <span class="o">=</span> <span class="n">tick</span>
    <span class="n">defaults</span> <span class="o">=</span> <span class="n">default_slim_metadata</span><span class="p">(</span><span class="s2">&quot;tree_sequence&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">this_chromosome</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">this_chromosome</span> <span class="o">=</span> <span class="n">defaults</span><span class="p">[</span><span class="s1">&#39;SLiM&#39;</span><span class="p">][</span><span class="s1">&#39;this_chromosome&#39;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">chromosomes</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">chromosomes</span> <span class="o">=</span> <span class="n">defaults</span><span class="p">[</span><span class="s1">&#39;SLiM&#39;</span><span class="p">][</span><span class="s1">&#39;chromosomes&#39;</span><span class="p">]</span>
    <span class="k">assert</span><span class="p">(</span><span class="n">schema_dict</span><span class="p">[</span><span class="s1">&#39;codec&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;json&#39;</span><span class="p">)</span>
    <span class="k">assert</span><span class="p">(</span><span class="n">schema_dict</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;object&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="s2">&quot;properties&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">schema_dict</span><span class="p">:</span>
        <span class="n">schema_dict</span><span class="p">[</span><span class="s2">&quot;properties&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">schema_dict</span><span class="p">[</span><span class="s1">&#39;properties&#39;</span><span class="p">][</span><span class="s1">&#39;SLiM&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">slim_metadata_schemas</span><span class="p">[</span><span class="s1">&#39;tree_sequence&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">schema</span><span class="p">[</span><span class="s1">&#39;properties&#39;</span><span class="p">][</span><span class="s1">&#39;SLiM&#39;</span><span class="p">]</span>
    <span class="n">tables</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">MetadataSchema</span><span class="p">(</span><span class="n">schema_dict</span><span class="p">)</span>
    <span class="n">metadata_dict</span><span class="p">[</span><span class="s1">&#39;SLiM&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;model_type&quot;</span><span class="p">:</span> <span class="n">model_type</span><span class="p">,</span>
            <span class="s2">&quot;tick&quot;</span><span class="p">:</span> <span class="n">tick</span><span class="p">,</span>
            <span class="s2">&quot;cycle&quot;</span><span class="p">:</span> <span class="n">cycle</span><span class="p">,</span>
            <span class="s2">&quot;file_version&quot;</span><span class="p">:</span> <span class="n">file_version</span><span class="p">,</span>
            <span class="s2">&quot;spatial_dimensionality&quot;</span><span class="p">:</span> <span class="n">spatial_dimensionality</span><span class="p">,</span>
            <span class="s2">&quot;spatial_periodicity&quot;</span><span class="p">:</span> <span class="n">spatial_periodicity</span><span class="p">,</span>
            <span class="s2">&quot;separate_sexes&quot;</span><span class="p">:</span> <span class="n">separate_sexes</span><span class="p">,</span>
            <span class="s2">&quot;nucleotide_based&quot;</span><span class="p">:</span> <span class="n">nucleotide_based</span><span class="p">,</span>
            <span class="s2">&quot;stage&quot;</span><span class="p">:</span> <span class="n">stage</span><span class="p">,</span>
            <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">name</span><span class="p">,</span>
            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="n">description</span><span class="p">,</span>
            <span class="s2">&quot;this_chromosome&quot;</span><span class="p">:</span> <span class="n">this_chromosome</span><span class="p">,</span>
            <span class="s2">&quot;chromosomes&quot;</span><span class="p">:</span> <span class="n">chromosomes</span><span class="p">,</span>
            <span class="p">}</span>
    <span class="n">tables</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="n">metadata_dict</span>


<span class="k">def</span><span class="w"> </span><span class="nf">set_metadata_schemas</span><span class="p">(</span><span class="n">tables</span><span class="p">,</span> <span class="n">num_chromosomes</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
    <span class="n">tables</span><span class="o">.</span><span class="n">edges</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">=</span> <span class="n">slim_metadata_schemas</span><span class="p">[</span><span class="s1">&#39;edge&#39;</span><span class="p">]</span>
    <span class="n">tables</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">=</span> <span class="n">slim_metadata_schemas</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">]</span>
    <span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">=</span> <span class="n">slim_metadata_schemas</span><span class="p">[</span><span class="s1">&#39;mutation&#39;</span><span class="p">]</span>
    <span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">=</span> <span class="n">slim_node_metadata_schema</span><span class="p">(</span><span class="n">num_chromosomes</span><span class="p">)</span>
    <span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">=</span> <span class="n">slim_metadata_schemas</span><span class="p">[</span><span class="s1">&#39;individual&#39;</span><span class="p">]</span>
    <span class="n">tables</span><span class="o">.</span><span class="n">populations</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">=</span> <span class="n">slim_metadata_schemas</span><span class="p">[</span><span class="s1">&#39;population&#39;</span><span class="p">]</span>


<span class="c1">################################</span>
<span class="c1"># Previous versions of metadata schema:</span>

<span class="k">def</span><span class="w"> </span><span class="nf">_old_metadata_schema</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">file_version</span><span class="p">):</span>
    <span class="c1"># Returns a metadata schema *if the format has changed*,</span>
    <span class="c1"># and None otherwise.</span>
    <span class="n">ms</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;tree_sequence&quot;</span> <span class="ow">and</span> <span class="n">file_version</span> <span class="o">==</span> <span class="s2">&quot;0.8&quot;</span><span class="p">:</span>
        <span class="n">pre_0_9_tree_sequence</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;$schema&quot;</span><span class="p">:</span> <span class="s2">&quot;http://json-schema.org/schema#&quot;</span><span class="p">,</span>
            <span class="s2">&quot;codec&quot;</span><span class="p">:</span> <span class="s2">&quot;json&quot;</span><span class="p">,</span>
            <span class="s2">&quot;examples&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="p">{</span>
                    <span class="s2">&quot;SLiM&quot;</span><span class="p">:</span> <span class="p">{</span>
                        <span class="s2">&quot;file_version&quot;</span><span class="p">:</span> <span class="s2">&quot;0.8&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;fox&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;foxes on Catalina island&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;cycle&quot;</span><span class="p">:</span> <span class="mi">123</span><span class="p">,</span>
                        <span class="s2">&quot;tick&quot;</span><span class="p">:</span> <span class="mi">123</span><span class="p">,</span>
                        <span class="s2">&quot;model_type&quot;</span><span class="p">:</span> <span class="s2">&quot;WF&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;nucleotide_based&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                        <span class="s2">&quot;separate_sexes&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                        <span class="s2">&quot;spatial_dimensionality&quot;</span><span class="p">:</span> <span class="s2">&quot;xy&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;spatial_periodicity&quot;</span><span class="p">:</span> <span class="s2">&quot;x&quot;</span>
                    <span class="p">}</span>
                <span class="p">}</span>
            <span class="p">],</span>
            <span class="s2">&quot;properties&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;SLiM&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Top-level metadata for a SLiM tree sequence, file format version 0.8&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;properties&quot;</span><span class="p">:</span> <span class="p">{</span>
                        <span class="s2">&quot;file_version&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The SLiM &#39;file format version&#39; of this tree sequence.&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The SLiM species name represented by this tree sequence.&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;A user-configurable description of the species represented by this tree sequence.&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;cycle&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The &#39;SLiM cycle&#39; counter when this tree sequence was recorded.&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;tick&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The &#39;SLiM tick&#39; counter when this tree sequence was recorded.&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;model_type&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The model type used for the last part of this simulation (WF or nonWF).&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;enum&quot;</span><span class="p">:</span> <span class="p">[</span> <span class="s2">&quot;WF&quot;</span><span class="p">,</span> <span class="s2">&quot;nonWF&quot;</span> <span class="p">],</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;nucleotide_based&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Whether the simulation was nucleotide-based.&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;boolean&quot;</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;separate_sexes&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Whether the simulation had separate sexes.&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;boolean&quot;</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;spatial_dimensionality&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The spatial dimensionality of the simulation.&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;enum&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;xy&quot;</span><span class="p">,</span> <span class="s2">&quot;xyz&quot;</span><span class="p">],</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;spatial_periodicity&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The spatial periodicity of the simulation.&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;enum&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;y&quot;</span><span class="p">,</span> <span class="s2">&quot;z&quot;</span><span class="p">,</span> <span class="s2">&quot;xy&quot;</span><span class="p">,</span> <span class="s2">&quot;xz&quot;</span><span class="p">,</span> <span class="s2">&quot;yz&quot;</span><span class="p">,</span> <span class="s2">&quot;xyz&quot;</span><span class="p">],</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;stage&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The stage of the SLiM life cycle when this tree sequence was recorded.&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span>
                        <span class="p">}</span>
                    <span class="p">},</span>
                    <span class="s2">&quot;required&quot;</span><span class="p">:</span> <span class="p">[</span>
                            <span class="s2">&quot;model_type&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;tick&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;file_version&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;spatial_dimensionality&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;spatial_periodicity&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;separate_sexes&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;nucleotide_based&quot;</span>
                    <span class="p">],</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;object&quot;</span>
                <span class="p">}</span>
            <span class="p">},</span>
            <span class="s2">&quot;required&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;SLiM&quot;</span><span class="p">],</span>
            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;object&quot;</span>
        <span class="p">}</span>

        <span class="n">ms</span> <span class="o">=</span> <span class="n">pre_0_9_tree_sequence</span>

    <span class="k">if</span> <span class="p">(</span><span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;tree_sequence&quot;</span>
        <span class="ow">and</span> <span class="n">file_version</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;0.1&quot;</span><span class="p">,</span> <span class="s2">&quot;0.2&quot;</span><span class="p">,</span> <span class="s2">&quot;0.3&quot;</span><span class="p">,</span> <span class="s2">&quot;0.4&quot;</span><span class="p">,</span> <span class="s2">&quot;0.5&quot;</span><span class="p">,</span> <span class="s2">&quot;0.6&quot;</span><span class="p">,</span> <span class="s2">&quot;0.7&quot;</span><span class="p">]):</span>
        <span class="n">pre_0_8_tree_sequence</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;$schema&quot;</span><span class="p">:</span> <span class="s2">&quot;http://json-schema.org/schema#&quot;</span><span class="p">,</span>
            <span class="s2">&quot;codec&quot;</span><span class="p">:</span> <span class="s2">&quot;json&quot;</span><span class="p">,</span>
            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;object&quot;</span><span class="p">,</span>
            <span class="s2">&quot;properties&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;SLiM&quot;</span> <span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Top-level metadata for a SLiM tree sequence, file format version 0.7&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;object&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;properties&quot;</span><span class="p">:</span> <span class="p">{</span>
                        <span class="s2">&quot;model_type&quot;</span> <span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;enum&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;WF&quot;</span><span class="p">,</span> <span class="s2">&quot;nonWF&quot;</span><span class="p">],</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The model type used for the last part of this simulation (WF or nonWF).&quot;</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;generation&quot;</span> <span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The &#39;SLiM generation&#39; counter when this tree sequence was recorded.&quot;</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;stage&quot;</span> <span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The stage of the SLiM life cycle when this tree sequence was recorded.&quot;</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;file_version&quot;</span> <span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The SLiM &#39;file format version&#39; of this tree sequence.&quot;</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;spatial_dimensionality&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;enum&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;xy&quot;</span><span class="p">,</span> <span class="s2">&quot;xyz&quot;</span><span class="p">],</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The spatial dimensionality of the simulation.&quot;</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;spatial_periodicity&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;string&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;enum&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;y&quot;</span><span class="p">,</span> <span class="s2">&quot;z&quot;</span><span class="p">,</span> <span class="s2">&quot;xy&quot;</span><span class="p">,</span> <span class="s2">&quot;xz&quot;</span><span class="p">,</span> <span class="s2">&quot;yz&quot;</span><span class="p">,</span> <span class="s2">&quot;xyz&quot;</span><span class="p">],</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The spatial periodicity of the simulation.&quot;</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;separate_sexes&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;boolean&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Whether the simulation had separate sexes.&quot;</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;nucleotide_based&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;boolean&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Whether the simulation was nucleotide-based.&quot;</span>
                        <span class="p">}</span>
                    <span class="p">},</span>
                    <span class="s2">&quot;required&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;model_type&quot;</span><span class="p">,</span> <span class="s2">&quot;generation&quot;</span><span class="p">,</span> <span class="s2">&quot;file_version&quot;</span><span class="p">,</span> <span class="s2">&quot;spatial_dimensionality&quot;</span><span class="p">,</span> <span class="s2">&quot;spatial_periodicity&quot;</span><span class="p">,</span> <span class="s2">&quot;separate_sexes&quot;</span><span class="p">,</span> <span class="s2">&quot;nucleotide_based&quot;</span><span class="p">]</span>
                <span class="p">}</span>
            <span class="p">},</span>
            <span class="s2">&quot;required&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;SLiM&quot;</span><span class="p">],</span>
            <span class="s2">&quot;examples&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="p">{</span>
                 <span class="s2">&quot;SLiM&quot;</span> <span class="p">:</span> <span class="p">{</span>
                     <span class="s2">&quot;model_type&quot;</span> <span class="p">:</span> <span class="s2">&quot;WF&quot;</span><span class="p">,</span>
                     <span class="s2">&quot;generation&quot;</span> <span class="p">:</span> <span class="mi">123</span><span class="p">,</span>
                     <span class="s2">&quot;file_version&quot;</span> <span class="p">:</span> <span class="s2">&quot;0.7&quot;</span><span class="p">,</span>
                     <span class="s2">&quot;spatial_dimensionality&quot;</span> <span class="p">:</span> <span class="s2">&quot;xy&quot;</span><span class="p">,</span>
                     <span class="s2">&quot;spatial_periodicity&quot;</span> <span class="p">:</span> <span class="s2">&quot;x&quot;</span><span class="p">,</span>
                     <span class="s2">&quot;separate_sexes&quot;</span> <span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                     <span class="s2">&quot;nucleotide_based&quot;</span> <span class="p">:</span> <span class="kc">False</span>
                     <span class="p">}</span>
                <span class="p">}</span>
            <span class="p">]</span>
        <span class="p">}</span>
        <span class="n">ms</span> <span class="o">=</span> <span class="n">pre_0_8_tree_sequence</span> 

    <span class="k">if</span> <span class="p">(</span><span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;population&quot;</span>
        <span class="ow">and</span> <span class="n">file_version</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;0.1&quot;</span><span class="p">,</span> <span class="s2">&quot;0.2&quot;</span><span class="p">,</span> <span class="s2">&quot;0.3&quot;</span><span class="p">,</span> <span class="s2">&quot;0.4&quot;</span><span class="p">,</span> <span class="s2">&quot;0.5&quot;</span><span class="p">,</span> <span class="s2">&quot;0.6&quot;</span><span class="p">]):</span>
        <span class="n">pre_0_7_population</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;$schema&quot;</span><span class="p">:</span> <span class="s2">&quot;http://json-schema.org/schema#&quot;</span><span class="p">,</span>
            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;SLiM schema for population metadata.&quot;</span><span class="p">,</span>
            <span class="s2">&quot;codec&quot;</span><span class="p">:</span> <span class="s2">&quot;struct&quot;</span><span class="p">,</span>
            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;object&quot;</span><span class="p">,</span> <span class="s2">&quot;null&quot;</span><span class="p">],</span>
            <span class="s2">&quot;properties&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;slim_id&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The ID of this population in SLiM. Note that this is called a &#39;subpopulation&#39; in SLiM.&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;i&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span>
                <span class="s2">&quot;selfing_fraction&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The frequency with which individuals in this subpopulation self (for WF models).&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;d&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">},</span>
                <span class="s2">&quot;female_cloning_fraction&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The frequency with which females in this subpopulation reproduce clonally (for WF models).&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;d&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">},</span>
                <span class="s2">&quot;male_cloning_fraction&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The frequency with which males in this subpopulation reproduce clonally (for WF models).&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;d&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">},</span>
                <span class="s2">&quot;sex_ratio&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;This subpopulation&#39;s sex ratio (for WF models).&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;d&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">5</span><span class="p">},</span>
                <span class="s2">&quot;bounds_x0&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The minimum x-coordinate in this subpopulation.&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;d&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">6</span><span class="p">},</span>
                <span class="s2">&quot;bounds_x1&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The maximum x-coordinate in this subpopulation.&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;d&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">7</span><span class="p">},</span>
                <span class="s2">&quot;bounds_y0&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The minimum y-coordinate in this subpopulation.&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;d&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">8</span><span class="p">},</span>
                <span class="s2">&quot;bounds_y1&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The maximum y-coordinate in this subpopulation.&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;d&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">9</span><span class="p">},</span>
                <span class="s2">&quot;bounds_z0&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The minimum z-coordinate in this subpopulation.&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;d&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">10</span><span class="p">},</span>
                <span class="s2">&quot;bounds_z1&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The maximum z-coordinate in this subpopulation.&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;d&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">11</span><span class="p">},</span>
                <span class="s2">&quot;migration_records&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;array&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">13</span><span class="p">,</span>
                    <span class="s2">&quot;arrayLengthFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;I&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;items&quot;</span><span class="p">:</span> <span class="p">{</span>
                        <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;object&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;properties&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;source_subpop&quot;</span><span class="p">:</span> <span class="p">{</span>
                                <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The ID of the subpopulation migrants come from (in WF models).&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;i&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span>
                            <span class="s2">&quot;migration_rate&quot;</span><span class="p">:</span> <span class="p">{</span>
                                <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The fraction of children in this subpopulation that are composed of &#39;migrants&#39; from the source subpopulation (in WF models).&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;d&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">}</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;required&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;source_subpop&quot;</span><span class="p">,</span> <span class="s2">&quot;migration_rate&quot;</span><span class="p">],</span>
                        <span class="s2">&quot;additionalProperties&quot;</span><span class="p">:</span> <span class="kc">False</span>
                    <span class="p">}</span>
                <span class="p">}</span>
            <span class="p">}</span>
        <span class="p">}</span>
        <span class="n">ms</span> <span class="o">=</span> <span class="n">pre_0_7_population</span> 

    <span class="k">if</span> <span class="p">(</span><span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;individual&quot;</span>
        <span class="ow">and</span> <span class="n">file_version</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;0.1&quot;</span><span class="p">,</span> <span class="s2">&quot;0.2&quot;</span><span class="p">,</span> <span class="s2">&quot;0.3&quot;</span><span class="p">,</span> <span class="s2">&quot;0.4&quot;</span><span class="p">,</span> <span class="s2">&quot;0.5&quot;</span><span class="p">,</span> <span class="s2">&quot;0.6&quot;</span><span class="p">]):</span>
        <span class="n">pre_0_7_individual</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;$schema&quot;</span><span class="p">:</span> <span class="s2">&quot;http://json-schema.org/schema#&quot;</span><span class="p">,</span>
            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;SLiM schema for individual metadata.&quot;</span><span class="p">,</span>
            <span class="s2">&quot;codec&quot;</span><span class="p">:</span> <span class="s2">&quot;struct&quot;</span><span class="p">,</span>
            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;object&quot;</span><span class="p">,</span>
            <span class="s2">&quot;properties&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;pedigree_id&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The &#39;pedigree ID&#39; of this individual in SLiM.&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;q&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">1</span>
                <span class="p">},</span>
                <span class="s2">&quot;age&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The age of this individual, either when the tree sequence was written out (if the individual was alive then), or the last time they were Remembered (if not).&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;i&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">2</span>
                <span class="p">},</span>
                <span class="s2">&quot;subpopulation&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The ID of the subpopulation the individual was part of, either when the tree sequence was written out (if the individual was alive then), or the last time they were Remembered (if not).&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;i&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">3</span>
                <span class="p">},</span>
                <span class="s2">&quot;sex&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The sex of the individual (0 for female, 1 for male, -1 for hermaphrodite).&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;i&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">4</span>
                <span class="p">},</span>
                <span class="s2">&quot;flags&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Other information about the individual: see &#39;flags&#39;.&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;I&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">5</span>
                <span class="p">}</span>
            <span class="p">},</span>
            <span class="s2">&quot;required&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;pedigree_id&quot;</span><span class="p">,</span> <span class="s2">&quot;age&quot;</span><span class="p">,</span> <span class="s2">&quot;subpopulation&quot;</span><span class="p">,</span> <span class="s2">&quot;sex&quot;</span><span class="p">,</span> <span class="s2">&quot;flags&quot;</span><span class="p">],</span>
            <span class="s2">&quot;additionalProperties&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="s2">&quot;flags&quot;</span><span class="p">:</span> <span class="p">{</span>
                 <span class="s2">&quot;SLIM_INDIVIDUAL_METADATA_MIGRATED&quot;</span><span class="p">:</span> <span class="p">{</span>
                     <span class="s2">&quot;value&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
                     <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Whether this individual was a migrant, either in the generation when the tree sequence was written out (if the individual was alive then), or in the generation of the last time they were Remembered (if not).&quot;</span>
                 <span class="p">}</span>
            <span class="p">},</span>
        <span class="p">}</span>
        <span class="n">ms</span> <span class="o">=</span> <span class="n">pre_0_7_individual</span> 

    <span class="k">if</span> <span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;mutation&quot;</span> <span class="ow">and</span> <span class="n">file_version</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;0.1&quot;</span><span class="p">,</span> <span class="s2">&quot;0.2&quot;</span><span class="p">]:</span>
        <span class="n">mutation_pre_0_3</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;$schema&quot;</span><span class="p">:</span> <span class="s2">&quot;http://json-schema.org/schema#&quot;</span><span class="p">,</span>
            <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;SLiM schema for mutation metadata.&quot;</span><span class="p">,</span>
            <span class="s2">&quot;codec&quot;</span><span class="p">:</span> <span class="s2">&quot;struct&quot;</span><span class="p">,</span>
            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;object&quot;</span><span class="p">,</span>
            <span class="s2">&quot;properties&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;mutation_list&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;array&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;noLengthEncodingExhaustBuffer&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                    <span class="s2">&quot;items&quot;</span><span class="p">:</span> <span class="p">{</span>
                        <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;object&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;properties&quot;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="s2">&quot;mutation_type&quot;</span><span class="p">:</span> <span class="p">{</span>
                                <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The index of this mutation&#39;s mutationType.&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;i&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">1</span>
                            <span class="p">},</span>
                            <span class="s2">&quot;selection_coeff&quot;</span><span class="p">:</span> <span class="p">{</span>
                                <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;number&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;This mutation&#39;s selection coefficient.&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;f&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">2</span>
                            <span class="p">},</span>
                            <span class="s2">&quot;subpopulation&quot;</span><span class="p">:</span> <span class="p">{</span>
                                <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The ID of the subpopulation this mutation occurred in.&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;i&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">3</span>
                            <span class="p">},</span>
                            <span class="s2">&quot;slim_time&quot;</span><span class="p">:</span> <span class="p">{</span>
                                <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The SLiM generation counter when this mutation occurred.&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;i&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">4</span>
                            <span class="p">},</span>
                        <span class="p">},</span>
                        <span class="s2">&quot;required&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;mutation_type&quot;</span><span class="p">,</span> <span class="s2">&quot;selection_coeff&quot;</span><span class="p">,</span> <span class="s2">&quot;subpopulation&quot;</span><span class="p">,</span>
                                     <span class="s2">&quot;slim_time&quot;</span><span class="p">],</span>
                        <span class="s2">&quot;additionalProperties&quot;</span><span class="p">:</span> <span class="kc">False</span>
                    <span class="p">}</span>
                <span class="p">}</span>
            <span class="p">},</span>
            <span class="s2">&quot;required&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;mutation_list&quot;</span><span class="p">],</span>
            <span class="s2">&quot;additionalProperties&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="p">}</span>
        <span class="n">ms</span> <span class="o">=</span> <span class="n">mutation_pre_0_3</span>

    <span class="k">if</span> <span class="p">(</span><span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;node&quot;</span>
        <span class="ow">and</span> <span class="n">file_version</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;0.1&quot;</span><span class="p">,</span> <span class="s2">&quot;0.2&quot;</span><span class="p">,</span> <span class="s2">&quot;0.3&quot;</span><span class="p">,</span> <span class="s2">&quot;0.4&quot;</span><span class="p">,</span> <span class="s2">&quot;0.5&quot;</span><span class="p">,</span> <span class="s2">&quot;0.6&quot;</span><span class="p">,</span> <span class="s2">&quot;0.7&quot;</span><span class="p">,</span> <span class="s2">&quot;0.8&quot;</span><span class="p">]):</span>
        <span class="n">node_pre_0_9</span> <span class="o">=</span> <span class="p">{</span>
           <span class="s2">&quot;$schema&quot;</span><span class="p">:</span> <span class="s2">&quot;http://json-schema.org/schema#&quot;</span><span class="p">,</span>
           <span class="s2">&quot;additionalProperties&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
           <span class="s2">&quot;codec&quot;</span><span class="p">:</span> <span class="s2">&quot;struct&quot;</span><span class="p">,</span>
           <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;SLiM schema for node metadata.&quot;</span><span class="p">,</span>
           <span class="s2">&quot;examples&quot;</span><span class="p">:</span> <span class="p">[</span>
               <span class="p">{</span>
                   <span class="s2">&quot;genome_type&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
                   <span class="s2">&quot;is_null&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                   <span class="s2">&quot;slim_id&quot;</span><span class="p">:</span> <span class="mi">123</span>
                <span class="p">}</span>
            <span class="p">],</span>
           <span class="s2">&quot;properties&quot;</span><span class="p">:</span> <span class="p">{</span>
               <span class="s2">&quot;genome_type&quot;</span><span class="p">:</span> <span class="p">{</span>
                   <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span>
                   <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The &#39;type&#39; of this genome (0 for autosome, 1 for X, 2 for Y).&quot;</span><span class="p">,</span>
                   <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span>
                   <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
                <span class="p">},</span>
               <span class="s2">&quot;is_null&quot;</span><span class="p">:</span> <span class="p">{</span>
                   <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;?&quot;</span><span class="p">,</span>
                   <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Whether this node describes a &#39;null&#39; (non-existant) chromosome.&quot;</span><span class="p">,</span>
                   <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
                   <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;boolean&quot;</span>
                <span class="p">},</span>
               <span class="s2">&quot;slim_id&quot;</span><span class="p">:</span> <span class="p">{</span>
                   <span class="s2">&quot;binaryFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;q&quot;</span><span class="p">,</span>
                   <span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;The &#39;pedigree ID&#39; of this chromosome in SLiM.&quot;</span><span class="p">,</span>
                   <span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
                   <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;integer&quot;</span>
                <span class="p">}</span>
            <span class="p">},</span>
           <span class="s2">&quot;required&quot;</span><span class="p">:</span> <span class="p">[</span>
               <span class="s2">&quot;slim_id&quot;</span><span class="p">,</span>
               <span class="s2">&quot;is_null&quot;</span><span class="p">,</span>
               <span class="s2">&quot;genome_type&quot;</span>
            <span class="p">],</span>
           <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="p">[</span>
               <span class="s2">&quot;object&quot;</span><span class="p">,</span>
               <span class="s2">&quot;null&quot;</span>
            <span class="p">]</span>
        <span class="p">}</span>
        <span class="n">ms</span> <span class="o">=</span> <span class="n">node_pre_0_9</span>

    <span class="c1"># everything else&#39;s format has remained unchanged</span>
    <span class="k">if</span> <span class="n">ms</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">ms</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">MetadataSchema</span><span class="p">(</span><span class="n">ms</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">ms</span>


<span class="k">def</span><span class="w"> </span><span class="nf">is_current_version</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">_warn</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Tests whether the tree sequence or table collection provided is the current</span>
<span class="sd">    SLiM file format or not. If not, use `pyslim.update( )` to bring it up to</span>
<span class="sd">    date.</span>

<span class="sd">    :param TreeSequence ts: The tree sequence or table collection.</span>
<span class="sd">    :return bool: Whether the tree sequence is the current version.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">out</span> <span class="o">=</span> <span class="p">(</span>
        <span class="nb">isinstance</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">metadata</span><span class="p">,</span> <span class="nb">dict</span><span class="p">)</span>
        <span class="ow">and</span>
        <span class="p">(</span><span class="s1">&#39;SLiM&#39;</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">metadata</span><span class="p">)</span>
        <span class="ow">and</span>
        <span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;SLiM&#39;</span><span class="p">][</span><span class="s1">&#39;file_version&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">slim_file_version</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="k">if</span> <span class="n">_warn</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">out</span><span class="p">:</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
                <span class="s2">&quot;This tree sequence is not the current SLiM format, &quot;</span>
                <span class="s2">&quot;so some operations may not work. &quot;</span>
                <span class="s2">&quot;Use `pyslim.update( )` to update the tree sequence.&quot;</span>
        <span class="p">)</span>
    <span class="k">return</span> <span class="n">out</span>


<div class="viewcode-block" id="update">
<a class="viewcode-back" href="../../python_api.html#pyslim.update">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">update</span><span class="p">(</span><span class="n">ts</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Update a tree sequence produced by a previous version of SLiM</span>
<span class="sd">    to the current file version.</span>

<span class="sd">    :return TreeSequence: The updated tree sequence.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">tables</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">dump_tables</span><span class="p">()</span>
    <span class="n">update_tables</span><span class="p">(</span><span class="n">tables</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">tables</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">()</span></div>



<span class="k">def</span><span class="w"> </span><span class="nf">update_tables</span><span class="p">(</span><span class="n">tables</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Update tables produced by a previous version of SLiM to the current file version.</span>
<span class="sd">    Modifies the tables in place.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># First we ensure we can find the file format version number</span>
    <span class="c1"># in top-level metadata. Then we proceed to fix up the tables as necessary.</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">tables</span><span class="o">.</span><span class="n">metadata</span><span class="p">,</span> <span class="nb">dict</span><span class="p">)</span> <span class="ow">and</span> <span class="s1">&#39;SLiM&#39;</span> <span class="ow">in</span> <span class="n">tables</span><span class="o">.</span><span class="n">metadata</span><span class="p">):</span>
        <span class="c1"># Old versions kept information in provenance, not top-level metadata.</span>
        <span class="c1"># Note this uses defaults on keys not present in provenance,</span>
        <span class="c1"># which prior to 0.5 was everything but generation and model_type.</span>
        <span class="n">values</span> <span class="o">=</span> <span class="n">default_slim_metadata</span><span class="p">(</span><span class="s1">&#39;tree_sequence&#39;</span><span class="p">)[</span><span class="s1">&#39;SLiM&#39;</span><span class="p">]</span>
        <span class="n">prov</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">file_version</span> <span class="o">=</span> <span class="s1">&#39;unknown&#39;</span>
        <span class="c1"># use only the last SLiM provenance</span>
        <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">tables</span><span class="o">.</span><span class="n">provenances</span><span class="p">:</span>
            <span class="n">is_slim</span><span class="p">,</span> <span class="n">this_file_version</span> <span class="o">=</span> <span class="n">slim_provenance_version</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> 
            <span class="k">if</span> <span class="n">is_slim</span><span class="p">:</span>
                <span class="n">prov</span> <span class="o">=</span> <span class="n">p</span>
                <span class="n">file_version</span> <span class="o">=</span> <span class="n">this_file_version</span>
        <span class="n">values</span><span class="p">[</span><span class="s1">&#39;file_version&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_version</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">record</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">prov</span><span class="o">.</span><span class="n">record</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">file_version</span> <span class="o">==</span> <span class="s2">&quot;0.1&quot;</span><span class="p">:</span>
                <span class="n">values</span><span class="p">[</span><span class="s1">&#39;model_type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s1">&#39;model_type&#39;</span><span class="p">]</span>
                <span class="n">values</span><span class="p">[</span><span class="s1">&#39;tick&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s1">&#39;generation&#39;</span><span class="p">]</span>
                <span class="n">values</span><span class="p">[</span><span class="s1">&#39;cycle&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s1">&#39;generation&#39;</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="s1">&#39;generation&#39;</span> <span class="ow">in</span> <span class="n">record</span><span class="p">[</span><span class="s1">&#39;slim&#39;</span><span class="p">]:</span>
                    <span class="n">values</span><span class="p">[</span><span class="s1">&#39;tick&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s1">&#39;slim&#39;</span><span class="p">][</span><span class="s1">&#39;generation&#39;</span><span class="p">]</span>
                    <span class="n">values</span><span class="p">[</span><span class="s1">&#39;cycle&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s1">&#39;slim&#39;</span><span class="p">][</span><span class="s1">&#39;generation&#39;</span><span class="p">]</span>
                <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">values</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">record</span><span class="p">[</span><span class="s1">&#39;parameters&#39;</span><span class="p">]:</span>
                        <span class="n">values</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s1">&#39;parameters&#39;</span><span class="p">][</span><span class="n">k</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">record</span><span class="p">[</span><span class="s1">&#39;slim&#39;</span><span class="p">]:</span>
                        <span class="n">values</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s1">&#39;slim&#39;</span><span class="p">][</span><span class="n">k</span><span class="p">]</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Failed to obtain metadata from provenance.&quot;</span><span class="p">)</span>
        <span class="n">set_tree_sequence_metadata</span><span class="p">(</span><span class="n">tables</span><span class="p">,</span> <span class="o">**</span><span class="n">values</span><span class="p">)</span>

    <span class="n">file_version</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;SLiM&#39;</span><span class="p">][</span><span class="s1">&#39;file_version&#39;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">file_version</span> <span class="o">!=</span> <span class="n">slim_file_version</span><span class="p">:</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;This is a version </span><span class="si">{}</span><span class="s2"> SLiM tree sequence.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">file_version</span><span class="p">)</span> <span class="o">+</span>
                      <span class="s2">&quot; If you write this out to a file, &quot;</span> <span class="o">+</span>
                      <span class="s2">&quot;it will be converted to version </span><span class="si">{}</span><span class="s2">.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">slim_file_version</span><span class="p">))</span>

        <span class="c1"># the only tables to have metadata schema changed thus far</span>
        <span class="c1"># are nodes, populations, individuals, mutations, and top-level:</span>
        <span class="n">old_schema</span> <span class="o">=</span> <span class="n">_old_metadata_schema</span><span class="p">(</span><span class="s2">&quot;tree_sequence&quot;</span><span class="p">,</span> <span class="n">file_version</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">old_schema</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">md</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">metadata</span>
            <span class="n">new_schema</span> <span class="o">=</span> <span class="n">slim_metadata_schemas</span><span class="p">[</span><span class="s2">&quot;tree_sequence&quot;</span><span class="p">]</span>
            <span class="n">new_properties</span> <span class="o">=</span> <span class="n">new_schema</span><span class="o">.</span><span class="n">asdict</span><span class="p">()[</span><span class="s1">&#39;properties&#39;</span><span class="p">][</span><span class="s1">&#39;SLiM&#39;</span><span class="p">][</span><span class="s1">&#39;required&#39;</span><span class="p">]</span>
            <span class="n">tables</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">=</span> <span class="n">new_schema</span>
            <span class="n">defaults</span> <span class="o">=</span> <span class="n">default_slim_metadata</span><span class="p">(</span><span class="s2">&quot;tree_sequence&quot;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">new_properties</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">k</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">md</span><span class="p">[</span><span class="s1">&#39;SLiM&#39;</span><span class="p">]:</span>
                    <span class="k">if</span> <span class="n">k</span> <span class="o">==</span> <span class="s2">&quot;tick&quot;</span><span class="p">:</span>
                        <span class="n">md</span><span class="p">[</span><span class="s1">&#39;SLiM&#39;</span><span class="p">][</span><span class="s1">&#39;tick&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">md</span><span class="p">[</span><span class="s1">&#39;SLiM&#39;</span><span class="p">][</span><span class="s1">&#39;generation&#39;</span><span class="p">]</span>
                        <span class="n">md</span><span class="p">[</span><span class="s1">&#39;SLiM&#39;</span><span class="p">][</span><span class="s1">&#39;cycle&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">md</span><span class="p">[</span><span class="s1">&#39;SLiM&#39;</span><span class="p">][</span><span class="s1">&#39;generation&#39;</span><span class="p">]</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">md</span><span class="p">[</span><span class="s1">&#39;SLiM&#39;</span><span class="p">][</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">defaults</span><span class="p">[</span><span class="s1">&#39;SLiM&#39;</span><span class="p">][</span><span class="n">k</span><span class="p">]</span>
            <span class="n">tables</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="n">md</span>

        <span class="n">old_schema</span> <span class="o">=</span> <span class="n">_old_metadata_schema</span><span class="p">(</span><span class="s2">&quot;node&quot;</span><span class="p">,</span> <span class="n">file_version</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">old_schema</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">nodes</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">nodes</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">==</span> <span class="n">tskit</span><span class="o">.</span><span class="n">MetadataSchema</span><span class="p">(</span><span class="kc">None</span><span class="p">):</span>
                <span class="n">nodes</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">=</span> <span class="n">old_schema</span>
            <span class="k">assert</span> <span class="s2">&quot;chromosomes&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">tables</span><span class="o">.</span><span class="n">metadata</span>
            <span class="c1"># if chromosomes was in metadata</span>
            <span class="c1"># we should use its length to get num_chroms,</span>
            <span class="c1"># but old file versions did not have this.</span>
            <span class="n">num_chroms</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">new_schema</span> <span class="o">=</span> <span class="n">slim_node_metadata_schema</span><span class="p">(</span><span class="n">num_chroms</span><span class="p">)</span>
            <span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">=</span> <span class="n">new_schema</span>
            <span class="n">not_vacant</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span>  <span class="c1"># single chromosome</span>
            <span class="n">yes_vacant</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">gt</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">nodes</span><span class="p">:</span>
                <span class="n">md</span> <span class="o">=</span> <span class="n">n</span><span class="o">.</span><span class="n">metadata</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">md</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">md</span><span class="p">[</span><span class="s1">&#39;is_vacant&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">yes_vacant</span> <span class="k">if</span> <span class="n">md</span><span class="p">[</span><span class="s1">&#39;is_null&#39;</span><span class="p">]</span> <span class="k">else</span> <span class="n">not_vacant</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">md</span><span class="p">[</span><span class="s1">&#39;is_null&#39;</span><span class="p">]:</span>
                        <span class="k">if</span> <span class="n">gt</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                            <span class="n">gt</span> <span class="o">=</span> <span class="n">md</span><span class="p">[</span><span class="s1">&#39;genome_type&#39;</span><span class="p">]</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="k">assert</span> <span class="n">md</span><span class="p">[</span><span class="s1">&#39;genome_type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">gt</span><span class="p">,</span> <span class="p">(</span><span class="s2">&quot;Inconsistent tables: &quot;</span>
                                            <span class="sa">f</span><span class="s2">&quot;mismatching genome types </span><span class="si">{</span><span class="n">gt</span><span class="si">}</span><span class="s2"> and &quot;</span>
                                            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">md</span><span class="p">[</span><span class="s1">&#39;genome_type&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> in node metadata.&quot;</span><span class="p">)</span>
                    <span class="k">del</span> <span class="n">md</span><span class="p">[</span><span class="s1">&#39;is_null&#39;</span><span class="p">]</span>
                    <span class="k">del</span> <span class="n">md</span><span class="p">[</span><span class="s1">&#39;genome_type&#39;</span><span class="p">]</span>
                <span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">n</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">metadata</span><span class="o">=</span><span class="n">md</span><span class="p">))</span>
            <span class="c1"># flags for node genome type pre-0.9:</span>
            <span class="c1"># confusingly and sub-optimally, these were redundant:</span>
            <span class="c1"># all non-null nodes in the same sim would have the same genome type</span>
            <span class="k">assert</span> <span class="n">gt</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
            <span class="n">GENOME_TYPE_AUTOSOME</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">GENOME_TYPE_X</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">GENOME_TYPE_Y</span> <span class="o">=</span> <span class="mi">2</span>
            <span class="n">top_md</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">metadata</span>
            <span class="n">i</span> <span class="o">=</span> <span class="n">top_md</span><span class="p">[</span><span class="s1">&#39;SLiM&#39;</span><span class="p">][</span><span class="s1">&#39;this_chromosome&#39;</span><span class="p">][</span><span class="s1">&#39;index&#39;</span><span class="p">]</span>
            <span class="c1"># &#39;chromosomes&#39; is not required so won&#39;t be inserted,</span>
            <span class="c1"># so we won&#39;t update it also</span>
            <span class="k">assert</span> <span class="s2">&quot;chromosomes&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">top_md</span><span class="p">[</span><span class="s1">&#39;SLiM&#39;</span><span class="p">][</span><span class="s1">&#39;this_chromosome&#39;</span><span class="p">],</span> <span class="p">(</span><span class="s2">&quot;&quot;</span>
                    <span class="s2">&quot;This is an unexpected result: if you hit this, &quot;</span>
                    <span class="s2">&quot;please file a bug at &quot;</span>
                    <span class="s2">&quot;https://github.com/tskit-dev/pyslim.&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">gt</span> <span class="o">==</span> <span class="n">GENOME_TYPE_X</span><span class="p">:</span>
                <span class="n">top_md</span><span class="p">[</span><span class="s1">&#39;SLiM&#39;</span><span class="p">][</span><span class="s1">&#39;this_chromosome&#39;</span><span class="p">][</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;X&quot;</span>
                <span class="n">top_md</span><span class="p">[</span><span class="s1">&#39;SLiM&#39;</span><span class="p">][</span><span class="s1">&#39;this_chromosome&#39;</span><span class="p">][</span><span class="s1">&#39;symbol&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;X&quot;</span>
                <span class="c1"># if &quot;chromosomes&quot; in tables.metadata[&#39;SLiM&#39;]:</span>
                <span class="c1">#     top_md[&#39;SLiM&#39;][&#39;chromosomes&#39;][i][&#39;type&#39;] = &quot;X&quot;</span>
                <span class="c1">#     top_md[&#39;SLiM&#39;][&#39;chromosomes&#39;][i][&#39;symbol&#39;] = &quot;X&quot;</span>
            <span class="k">elif</span> <span class="n">gt</span> <span class="o">==</span> <span class="n">GENOME_TYPE_Y</span><span class="p">:</span>
                <span class="n">top_md</span><span class="p">[</span><span class="s1">&#39;SLiM&#39;</span><span class="p">][</span><span class="s1">&#39;this_chromosome&#39;</span><span class="p">][</span><span class="s1">&#39;type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;-Y&quot;</span>
                <span class="n">top_md</span><span class="p">[</span><span class="s1">&#39;SLiM&#39;</span><span class="p">][</span><span class="s1">&#39;this_chromosome&#39;</span><span class="p">][</span><span class="s1">&#39;symbol&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Y&quot;</span>
                <span class="c1">#     top_md[&#39;SLiM&#39;][&#39;chromosomes&#39;][i][&#39;type&#39;] = &quot;Y&quot;</span>
                <span class="c1">#     top_md[&#39;SLiM&#39;][&#39;chromosomes&#39;][i][&#39;symbol&#39;] = &quot;Y&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">assert</span> <span class="n">gt</span> <span class="o">==</span> <span class="n">GENOME_TYPE_AUTOSOME</span>
            <span class="n">tables</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="n">top_md</span>

        <span class="n">old_schema</span> <span class="o">=</span> <span class="n">_old_metadata_schema</span><span class="p">(</span><span class="s2">&quot;population&quot;</span><span class="p">,</span> <span class="n">file_version</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">old_schema</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">pops</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">populations</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">tables</span><span class="o">.</span><span class="n">populations</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">pops</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">==</span> <span class="n">tskit</span><span class="o">.</span><span class="n">MetadataSchema</span><span class="p">(</span><span class="kc">None</span><span class="p">):</span>
                <span class="n">pops</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">=</span> <span class="n">old_schema</span>
            <span class="n">new_schema</span> <span class="o">=</span> <span class="n">slim_metadata_schemas</span><span class="p">[</span><span class="s2">&quot;population&quot;</span><span class="p">]</span>
            <span class="n">tables</span><span class="o">.</span><span class="n">populations</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">=</span> <span class="n">new_schema</span>
            <span class="c1"># just needs recoding</span>
            <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">pops</span><span class="p">:</span>
                <span class="n">tables</span><span class="o">.</span><span class="n">populations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span>

        <span class="n">old_schema</span> <span class="o">=</span> <span class="n">_old_metadata_schema</span><span class="p">(</span><span class="s2">&quot;individual&quot;</span><span class="p">,</span> <span class="n">file_version</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">old_schema</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">inds</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">inds</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">==</span> <span class="n">tskit</span><span class="o">.</span><span class="n">MetadataSchema</span><span class="p">(</span><span class="kc">None</span><span class="p">):</span>
                <span class="n">inds</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">=</span> <span class="n">old_schema</span>
            <span class="n">new_schema</span> <span class="o">=</span> <span class="n">slim_metadata_schemas</span><span class="p">[</span><span class="s2">&quot;individual&quot;</span><span class="p">]</span>
            <span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">=</span> <span class="n">new_schema</span>
            <span class="n">defaults</span> <span class="o">=</span> <span class="n">default_slim_metadata</span><span class="p">(</span><span class="s2">&quot;individual&quot;</span><span class="p">)</span>
            <span class="n">d</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;pedigree_p1&quot;</span><span class="p">,</span> <span class="s2">&quot;pedigree_p2&quot;</span><span class="p">]:</span>
                <span class="n">d</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">defaults</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">ind</span> <span class="ow">in</span> <span class="n">inds</span><span class="p">:</span>
                <span class="n">md</span> <span class="o">=</span> <span class="n">ind</span><span class="o">.</span><span class="n">metadata</span>
                <span class="n">md</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
                <span class="n">tables</span><span class="o">.</span><span class="n">individuals</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ind</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">metadata</span><span class="o">=</span><span class="n">md</span><span class="p">))</span>

        <span class="n">old_schema</span> <span class="o">=</span> <span class="n">_old_metadata_schema</span><span class="p">(</span><span class="s2">&quot;mutation&quot;</span><span class="p">,</span> <span class="n">file_version</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">old_schema</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">muts</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">clear</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">muts</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">==</span> <span class="n">tskit</span><span class="o">.</span><span class="n">MetadataSchema</span><span class="p">(</span><span class="kc">None</span><span class="p">):</span>
                <span class="n">muts</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">=</span> <span class="n">old_schema</span>
            <span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">metadata_schema</span> <span class="o">=</span> <span class="n">slim_metadata_schemas</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">muts</span><span class="p">:</span>
                <span class="n">md</span> <span class="o">=</span> <span class="n">mut</span><span class="o">.</span><span class="n">metadata</span>
                <span class="k">for</span> <span class="n">ml</span> <span class="ow">in</span> <span class="n">md</span><span class="p">[</span><span class="s1">&#39;mutation_list&#39;</span><span class="p">]:</span>
                    <span class="n">ml</span><span class="p">[</span><span class="s1">&#39;nucleotide&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
                <span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mut</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">metadata</span><span class="o">=</span><span class="n">md</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">file_version</span> <span class="o">==</span> <span class="s2">&quot;0.1&quot;</span><span class="p">:</span>
            <span class="c1"># shift times</span>
            <span class="n">slim_generation</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;SLiM&#39;</span><span class="p">][</span><span class="s1">&#39;tick&#39;</span><span class="p">]</span>
            <span class="n">node_times</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">time</span> <span class="o">+</span> <span class="n">slim_generation</span>
            <span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">set_columns</span><span class="p">(</span>
                    <span class="n">flags</span><span class="o">=</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">flags</span><span class="p">,</span>
                    <span class="n">time</span><span class="o">=</span><span class="n">node_times</span><span class="p">,</span>
                    <span class="n">population</span><span class="o">=</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">population</span><span class="p">,</span>
                    <span class="n">individual</span><span class="o">=</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">individual</span><span class="p">,</span>
                    <span class="n">metadata</span><span class="o">=</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">metadata</span><span class="p">,</span>
                    <span class="n">metadata_offset</span><span class="o">=</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">metadata_offset</span><span class="p">)</span>
            <span class="n">migration_times</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">migrations</span><span class="o">.</span><span class="n">time</span> <span class="o">+</span> <span class="n">slim_generation</span>
            <span class="n">tables</span><span class="o">.</span><span class="n">migrations</span><span class="o">.</span><span class="n">set_columns</span><span class="p">(</span>
                    <span class="n">left</span><span class="o">=</span><span class="n">tables</span><span class="o">.</span><span class="n">migrations</span><span class="o">.</span><span class="n">left</span><span class="p">,</span>
                    <span class="n">right</span><span class="o">=</span><span class="n">tables</span><span class="o">.</span><span class="n">migrations</span><span class="o">.</span><span class="n">right</span><span class="p">,</span>
                    <span class="n">node</span><span class="o">=</span><span class="n">tables</span><span class="o">.</span><span class="n">migrations</span><span class="o">.</span><span class="n">node</span><span class="p">,</span>
                    <span class="n">source</span><span class="o">=</span><span class="n">tables</span><span class="o">.</span><span class="n">migrations</span><span class="o">.</span><span class="n">source</span><span class="p">,</span>
                    <span class="n">dest</span><span class="o">=</span><span class="n">tables</span><span class="o">.</span><span class="n">migrations</span><span class="o">.</span><span class="n">dest</span><span class="p">,</span>
                    <span class="n">time</span><span class="o">=</span><span class="n">migration_times</span><span class="p">)</span>

        <span class="n">new_record</span> <span class="o">=</span> <span class="p">{</span>
                    <span class="s2">&quot;schema_version&quot;</span><span class="p">:</span> <span class="s2">&quot;1.0.0&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;software&quot;</span><span class="p">:</span> <span class="p">{</span>
                        <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;pyslim&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;version&quot;</span><span class="p">:</span> <span class="n">pyslim_version</span><span class="p">,</span>
                        <span class="p">},</span>
                    <span class="s2">&quot;parameters&quot;</span><span class="p">:</span> <span class="p">{</span>
                        <span class="s2">&quot;command&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;updrade_tables&quot;</span><span class="p">],</span>
                        <span class="s2">&quot;old_file_version&quot;</span><span class="p">:</span> <span class="n">file_version</span><span class="p">,</span>
                        <span class="s2">&quot;new_file_version&quot;</span><span class="p">:</span> <span class="n">slim_file_version</span><span class="p">,</span>
                        <span class="p">},</span>
                    <span class="s2">&quot;environment&quot;</span><span class="p">:</span> <span class="n">get_environment</span><span class="p">(),</span>
                <span class="p">}</span>
        <span class="n">tskit</span><span class="o">.</span><span class="n">validate_provenance</span><span class="p">(</span><span class="n">new_record</span><span class="p">)</span>
        <span class="n">tables</span><span class="o">.</span><span class="n">provenances</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">new_record</span><span class="p">))</span>

        <span class="n">set_metadata_schemas</span><span class="p">(</span><span class="n">tables</span><span class="p">)</span>
        <span class="n">md</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">metadata</span>
        <span class="n">md</span><span class="p">[</span><span class="s1">&#39;SLiM&#39;</span><span class="p">][</span><span class="s1">&#39;file_version&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">slim_file_version</span>
        <span class="n">tables</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="n">md</span>
</pre></div>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Tskit Developers
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2021.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>