<!DOCTYPE html>
<html lang="en-GB">

<!-- START: html-header -->
<head>
    
    <meta charset="utf-8">
    <meta name=”robots” content="index, follow">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>
        tskit-dev - Learn
    </title>
    <meta property="og:type" content="website">
    <meta content="summary_large_image" name="twitter:card">
    <meta content="" name="description">
    <meta content="Videos, tutorials and articles about tree sequences." property="og:description">
    <meta content="Videos, tutorials and articles about tree sequences." property="twitter:description">
    
        <meta content="https://i.imgur.com/I9PBjvU.png" property="og:image">
        <meta content="https://i.imgur.com/I9PBjvU.png" property="twitter:image">
    

    <meta content="tskit-dev - Learn" property="og:title">
    <meta content="tskit-dev - Learn" property="twitter:title">
    <script>
      document.documentElement.className = 'js'
    </script>
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <link rel="profile" href="http://gmpg.org/xfn/11">

    <link rel="shortcut icon" href="/assets/favicons/favicon.ico">
    <link rel="apple-touch-icon" sizes="180x180" href="/assets/favicons/apple-touch-icon.png">
    <link rel="icon" type="image/png" sizes="32x32" href="/assets/favicons/favicon-32x32.png">
    <link rel="icon" type="image/png" sizes="16x16" href="/assets/favicons/favicon-16x16.png">
    <link rel="manifest" href="/assets/favicons/site.webmanifest" crossorigin="use-credentials">
    <link rel="mask-icon" href="/assets/favicons/safari-pinned-tab.png" color="#085167">
    <meta name="msapplication-TileColor" content="#085167">
    <meta name="theme-color" content="#ffffff">

    <link rel="stylesheet" href="https://use.typekit.net/uhf4ono.css">
    <link rel="stylesheet" href="/assets/dist/css/style.min.css">
<script data-goatcounter="https://tskit.goatcounter.com/count" async src="//gc.zgo.at/count.js"></script>
</head>
<!-- END: html-header -->

<body>
<div class="page-wrapper">
    <!-- START: header -->
    <header class="header">
        <!-- START: navbar -->
        <nav class="nav navbar navbar-expand-lg">
    <div class="container">
        <a class="navbar-brand" href="/">
            <!-- START: icon-logo -->
            <svg class="icon icon--logo" xmlns="http://www.w3.org/2000/svg" x="0" y="0" viewBox="0 0 84.8 68">
                <style>
                    .logo-st0 {
                        fill: #085167
                    }

                    .logo-st1 {
                        fill: #14e2a8
                    }
                </style>
                <path class="logo-st0"
                      d="M2 61.4v-8.2H0v-3.4h2v-5h4.8v5h5v3.4h-5v8.2c0 1.9.8 2.9 2.5 2.9.9 0 1.7-.2 2.4-.6v3.5c-1.1.5-2.4.8-3.7.8-4.2 0-6-2.8-6-6.6"></path>
                <path class="logo-st0"
                      d="M62.7 67.5v-3.4h1.9V53.2h-1.9v-3.4h6.8v14.3h1.9v3.4h-8.7zm1.1-23.3c-.1-1.6 1.2-2.9 2.7-3 1.6-.1 2.9 1.2 3 2.7.1 1.6-1.2 2.9-2.7 3h-.1c-1.5.1-2.8-1-2.9-2.5v-.2"></path>
                <path class="logo-st0"
                      d="M39.6 67.5v-3.4h1.9V45.4h-1.9V42h6.8v14.4h4l1.7-3.2h-2.5v-3.4h9.6v3.4h-2.6L53.7 58l4.2 6h2.8v3.4h-9.9V64h1.7l-2.8-4.6h-3.4V64h1.9v3.4l-8.6.1z"></path>
                <path class="logo-st0"
                      d="M23.4 57c4.5 1.3 6.4 2.7 6.4 5.9 0 3.3-2.5 5.1-6.4 5.1-2.6 0-4.3-.7-5.4-1.9l-.1 1.3h-3.4v-6.1h3.7c.7 1.9 2.5 3.2 4.5 3.1 1.3 0 2.3-.5 2.3-1.6s-1-1.5-1.7-1.7l-2.6-.7c-2.8-.7-6.2-1.7-6.2-6 0-3.4 2.4-5.2 6-5.2 1.8 0 3.6.6 5 1.8l.1-1.3H29v5.9h-3.7c-.4-1.7-1.9-2.9-3.6-2.9-1.3 0-2.3.5-2.3 1.7 0 1 .6 1.4 1.8 1.7l2.2.9z"></path>
                <path class="logo-st0"
                      d="M75 61.4v-8.2h-2v-3.4h2v-5h4.8v5h5v3.4h-5v8.2c0 1.9.8 2.9 2.5 2.9.9 0 1.7-.2 2.4-.6v3.5c-1.1.5-2.4.8-3.7.8-4.2 0-6-2.8-6-6.6"></path>
                <path class="logo-st1" d="M75 28.8h4.8v12.5H75z"></path>
                <path class="logo-st1" d="M41.5 28.8h4.8v9.6h-4.8z"></path>
                <path class="logo-st1" d="M41.5 24h38.3v4.8H41.5z"></path>
                <path class="logo-st1" d="M58.3 14.4h4.8V24h-4.8z"></path>
                <path class="logo-st1" d="M2 14.4h4.8v26.9H2z"></path>
                <path class="logo-st1" d="M2 9.6h61.1v4.8H2z"></path>
                <path class="logo-st1" d="M34.6 0h4.8v9.6h-4.8z"></path>
            </svg>
            <!-- END: icon-logo -->
            <span class="visually-hidden">
                        tskit
                    </span>
        </a>
        <button class="navbar-toggler collapsed"
                type="button"
                data-bs-toggle="collapse"
                data-bs-target="#navbarContent"
                aria-controls="navbarContent"
                aria-expanded="false"
                aria-label="Toggle navigation"
        >
            <span class="navbar-toggler-icon"></span>
            <span class="navbar-toggler-icon"></span>
            <span class="navbar-toggler-icon"></span>
        </button>

        <div class="collapse navbar-collapse" id="navbarContent">
            <!-- START: menu -->
            <ul class="navbar-nav">
                <li class="nav__item">
                    <a href="/genetics-research" class=" nav__item__link">Genetics research</a>
                </li>
                <li class="nav__item">
                    <a href="/learn" class=" nav__item__link">Learn</a>
                </li>
                <li class="nav__item">
                    <a href="/software" class=" nav__item__link">Software</a>
                </li>
                <li class="nav__item">
                    <a href="/news" class=" nav__item__link">News</a>
                </li>
                <li class="nav__item">
                    <a href="/citation" class=" nav__item__link">Citation</a>
                </li>
                <li class="nav__item">
                    <a href="/community" class=" nav__item__link">Community</a>
                </li>
            </ul>
            <!-- END: menu -->
        </div>
    </div>
</nav>
        <!-- END: navbar -->

        <!-- START: heading -->
        
        <div class="page-heading page-heading--type-2">
            <div class="container">
                <div class="row">
                    <div class="col-lg-8">
                        <h1 class="page-heading__text">
                            Learn
                        </h1>
                    </div>
                </div>
            </div>
        </div>
        
        <!-- END: heading -->
    </header>
    <!-- END: header -->

    <!-- START: content -->
    <main class="page-content" id="content">
        <!-- START: hero-type-2 -->
<div class="hero hero--type-2">
    <div class="container">
        <div class="row">
            <div class="col-md-8 col-lg-6">
                <h2 id="what-is" class="hero__heading">
                    What is a tree sequence?
                </h2>
                <div class="hero__content wysiwyg">
                    <p>
                        A <em>succinct tree sequence</em>, or tree sequence for short, represents the relationships
                        between a set of DNA sequences (also known as an
                        <a href="/tutorials/args.html">ARG</a>). Tree sequences can be used to
                        store genetic data efficiently, and enable powerful analysis of millions of whole genomes
                        at a time. They can be created by simulation or by inferring relationships from genetic
                        variation.
                    </p>
                    <p>
                        For a fully functional taster of what tree sequences are and how they work, try our
                        <a href="/explore/lab/?path=tskit.ipynb"> interactive workbook</a>. 
                    </p>
                </div>
            </div>
        </div>
    </div>
</div>
<!-- END: hero-type-2 -->

<!-- START: cards-type-2 -->
<section class="section section--cards section--cards--type-2">
    <div class="container">
        <div class="row">
            <div class="col-md-8">
                <h3 class="section__heading">
                    Tree sequences provide:
                </h3>
            </div>
        </div>
        <div class="row">
            <div class="col-lg-4">
                <!-- START: card -->
                <a href="/tutorials/what_is.html#sec-what-is-ancestry" class="card card--type-2">
                    <h4 id="genetic-ancestry" class="card__heading">
                        A full record of genetic ancestry
                    </h4>
                    <p class="card__text">
                        A tree sequence, or ancestral recombination graph (ARG), concisely captures the genealogical history
                        of a set of genomes by representing the branches shared among adjacent local trees along the genome.
                    </p>
                    <span class="list__item__link link link--type-1">
                                <span class="link__text">
                                    Read more
                                </span>
                            </span>
                </a>
                <!-- END: card -->
            </div>
            <div class="col-lg-4">
                <!-- START: card -->
                <a href="/tutorials/what_is.html#sec-what-is-dna-data" class="card card--type-2">
                    <h4 id="DNA-encoding" class="card__heading">
                        An encoding of DNA data
                    </h4>
                    <p class="card__text">
                        Placing mutations on a tree sequence allows lossless representation and compression of
                        DNA datasets
                    </p>
                    <span class="list__item__link link link--type-1">
                                <span class="link__text">
                                    Read more
                                </span>
                            </span>
                </a>
                <!-- END: card -->
            </div>
            <div class="col-lg-4">
                <!-- START: card -->
                <a href="/tutorials/what_is.html#sec-what-is-analysis" class="card card--type-2">
                    <h4 id="efficient-analysis" class="card__heading">
                        An efficient analysis framework
                    </h4>
                    <p class="card__text">
                        Many of the analyses performed on DNA data can be expressed naturally as operations on
                        the ancestral trees, leading to highly efficient algorithms
                    </p>
                    <span class="list__item__link link link--type-1">
                                <span class="link__text">
                                    Read more
                                </span>
                            </span>
                </a>
                <!-- END: card -->
            </div>
        </div>
    </div>
</section>
<!-- END: cards-type-2 -->

<!-- START: filters -->
<section class="section section--filters">
    <div class="container">
        <div class="row">
            <div class="col-md-8">
                <h3 id="tutorials-publications-videos" class="section__heading">
                    Browse tutorials, publications and videos:
                </h3>
            </div>
        </div>
        <div class="row">
            <div class="col">
                <div class="btn-group" role="group" aria-label="Filters">
                    <a data-r-selector="all" onclick="switch_filter('all')" class="btn btn-outline-primary active" aria-current="page">All</a>
                    <a data-r-selector="paper" onclick="switch_filter('paper')" class="btn btn-outline-primary">Publications</a>
                    <a data-r-selector="tutorial" onclick="switch_filter('tutorial')" class="btn btn-outline-primary">Tutorials</a>
                    <a data-r-selector="video" onclick="switch_filter('video')" class="btn btn-outline-primary">Videos</a>
<!--                    <a onclick="switch_filter" class="btn btn-outline-primary">Applications</a>-->
                </div>
            </div>
        </div>
<!--        <div class="row">-->
<!--            <div class="col-md-8 col-lg-5">-->
<!--                <select class="form-select form-select-lg">-->
<!--                    <option>Choose resource type</option>-->
<!--                    <option value="publication">publication</option>-->
<!--                    <option value="tutorial">tutorial</option>-->
<!--                    <option value="video">video</option>-->
<!--                </select>-->
<!--            </div>-->
<!--        </div>-->
    </div>
    <div class="container">
        <div class="row">
            
            
            <div class="col-lg-4" data-r-type="paper">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Paper</span>
                    </div>
                    <h5 class="card__heading">
                        On ARGs, pedigrees, and genetic relatedness matrices
                    </h5>
                    
                        <div class="wysiwyg">
                            <p>
                                02 December, 2025
                            </p>
                            <p>
                                GENETICS (2025) Lehmann <em>et al.</em>
                            </p>
                            <p>
                                doi: <a href="https://doi.org/10.1093/genetics/iyaf219">10.1093/genetics/iyaf219</a>
                            </p>
                            <p>
                                <p>This paper discusses branch relatedness in the context of tree sequences,
and introduces a method to use the tree sequence encoding to efficiently
multiply the genetic relatedness matrix (GRM) by a vector, even for huge
sample sizes. This is used to implement a fast way of performing
principle component analysis (PCA) on ARGs and other tree sequences
(now implemented as <code class="language-plaintext highlighter-rouge">ts.pca()</code> in tskit)</p>

                            </p>
                        </div>
                        <a href="https://doi.org/10.1093/genetics/iyaf219" class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        Read the article
                                    </span>
                        </a>
                    
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="paper">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Paper</span>
                    </div>
                    <h5 class="card__heading">
                        A forest is more than its trees: haplotypes and ancestral recombination graphs
                    </h5>
                    
                        <div class="wysiwyg">
                            <p>
                                02 December, 2025
                            </p>
                            <p>
                                GENETICS (2025) Fritze <em>et al.</em>
                            </p>
                            <p>
                                doi: <a href="https://doi.org/10.1093/genetics/iyaf198">10.1093/genetics/iyaf198</a>
                            </p>
                            <p>
                                <p>This paper discusses how non-coalescent segments of nodes in an
ancestral recombination graph (ARG) are important, shows how
to infer them, and discusses why this can make tree sequence
encodings of ARGs more efficient. The functionality is
implemented in the <em>tskit</em> <code class="language-plaintext highlighter-rouge">ts.extend_haplotypes()</code> method.
The paper also describes haplotype-aware methods to compare
tree sequences, implemented in the <em>tscompare</em> library.</p>

                            </p>
                        </div>
                        <a href="https://doi.org/10.1093/genetics/iyaf198" class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        Read the article
                                    </span>
                        </a>
                    
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="video">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Video</span>
                    </div>
                    <h5 class="card__heading">
                        Untangling ancestral recombination graphs to infer complex demographic scenarios
                    </h5>
                    
                    
                        <div class="wysiwyg">
                            <p>
                                12 November, 2024
                            </p>
                            <p>
                                Nate Pope at Phyloseminar.org</br>
                                <p>A talk at <a href="https://phyloseminar.org">phyloseminar.org</a>,
that explains calculation and use of empirical pair and trio coalescence rate distributions from ARGs.
Demonstrates fitting of multipopulation demographic models from inferred coalescent rate distributions, including from inferred ARGs.</p>

<!--
*The ancestral recombination graph (ARG) encodes the complete genealogical relationships among a set of recombinant DNA sequences, and is a high-dimensional sufficient statistic for many evolutionary and population genetic models. Over the past decade multiple methods have been developed to infer an ARG from phased SNPs, typically using heuristics for the sake of scalability. There are two major obstacles to the use of inferred ARGs for subsequent model-based inference; for example, to fit and evaluate complex demographic models. First, error in ARG reconstruction is inevitable and difficult to diagnose; second, the likelihood of a parameterized model given an ARG is challenging to compute even in the absence of error. We describe a statistical framework that views ARGs as collections of dated events from which simpler time-to-event distributions (coalescences of labelled pairs, trios, quartets, etc.) may be efficiently approximated via weighted empirical distribution functions. Heavily parameterized demographic models can be fit by matching the empirical and expected hazard functions of many marginal time-to-event distributions simultaneously, using standard methods from the theory of continuous-time Markov processes. More generally, we discuss projecting an empirical ARG to a flexible set of time-indexed summary statistics that are amenable to modelling while remaining robust to reconstruction error.*
-->

                            </p>

                        </div>
                        
                            <a href="https://www.youtube.com/watch?v=cjSYaNuQwTs" class="video-trigger"
                               data-bs-toggle="modal" data-bs-target="#globalModal" data-video-id="cjSYaNuQwTs"
                               data-video-aspect-ratio="16x9">
                                <span class="visually-hidden">
                                    Play video
                                </span>
                                <div class="video-trigger__image-wrapper">
                                    <img alt="video thumbnail" class="img-fluid"
                                         src="https://img.youtube.com/vi/cjSYaNuQwTs/maxresdefault.jpg">
                                </div>
                            </a>
                        
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="video">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Video</span>
                    </div>
                    <h5 class="card__heading">
                        Bayesian inference of Ancestral Recombination Graphs- progress and challenges
                    </h5>
                    
                    
                        <div class="wysiwyg">
                            <p>
                                15 October, 2024
                            </p>
                            <p>
                                Yun Deng at Phyloseminar.org</br>
                                <p>A talk at <a href="https://phyloseminar.org">phyloseminar.org</a>,
introducing <a href="https://github.com/popgenmethods/SINGER">SINGER</a>,
an MCMC sampler for ARGs that scales to hundreds of samples,
and applying it to demographic inference an identifying selection.</p>

<!--
Ancestral Recombination Graphs (ARG), or sometimes known as Genome-wide Genealogies, describe the full genealogical history of the genomes and are richly informative about the evolutionary history. Recent years we have witnessed great progress in scalable inference of ARG on thousands or more genomes. However, many of them lack accuracy and can be sensitive to model mis-specification from demographic histories or selection. Moreover, they reconstruct only a single ARG topology and cannot quantify the considerable estimation uncertainty in ARG inference. To address these challenges, we introduce SINGER, a novel method which accelerates posterior sampling of ARG by highly optimized MCMC for at least hundreds of genomes. In this talk I will demonstrate the enhanced accuracy and robustness to model mis-specification of SINGER, and give examples of applications to real data. These examples include various aspects of evolutionary biology, such as demography, positive selection, balancing selection, and introgression, etc. Last but not least, I will discuss possible directions of pushing Bayesian inference of ARGs even further.*
-->

                            </p>

                        </div>
                        
                            <a href="https://www.youtube.com/watch?v=KlAgdID8oOU" class="video-trigger"
                               data-bs-toggle="modal" data-bs-target="#globalModal" data-video-id="KlAgdID8oOU"
                               data-video-aspect-ratio="16x9">
                                <span class="visually-hidden">
                                    Play video
                                </span>
                                <div class="video-trigger__image-wrapper">
                                    <img alt="video thumbnail" class="img-fluid"
                                         src="https://img.youtube.com/vi/KlAgdID8oOU/maxresdefault.jpg">
                                </div>
                            </a>
                        
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="video">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Video</span>
                    </div>
                    <h5 class="card__heading">
                        Scalable approaches to inference and analysis of genome-wide genealogies
                    </h5>
                    
                    
                        <div class="wysiwyg">
                            <p>
                                10 September, 2024
                            </p>
                            <p>
                                Arni Gunnarsson at Phyloseminar.org</br>
                                <p>A talk at <a href="https://phyloseminar.org">phyloseminar.org</a>,
discussing inferring ARGs
using <a href="https://www.biorxiv.org/content/10.1101/2024.08.31.610248v1">Threads</a>,
which uses an operation called “threading” and scales to biobank-scale samples.
Shows some applications to imputation and GWAS.</p>

<!--
*The ancestral recombination graph (ARG) is a graph-like structure that encodes a detailed genealogical history of a set of individuals along the genome. ARGs that are accurately reconstructed from genomic data sets are useful for a range of applications in statistical and population genetics, but inference from data sets comprising millions of samples and variants remains computationally challenging. In this talk, I will introduce a novel ARG inference algorithm, called Threads, and show how ARG inference can be applied to bionbank-scale data sets using the algorithmic paradigm of “threading”. Using inferred ARGs, I will then explore applications of inferred ARGs to three familiar tasks in statistical genetics. First, I will show how threading algorithms can be used to improve upon traditional genotype compression methods by identifying long identical-by-descent segments. Second, I will show how careful modeling of allele ages can help improve imputation of ultra-rare variants. Finally, I will discuss how inferred ARGs can complement or improve upon traditional genetic association studies.*
-->

                            </p>

                        </div>
                        
                            <a href="https://www.youtube.com/watch?v=E0JKlCgxzlo" class="video-trigger"
                               data-bs-toggle="modal" data-bs-target="#globalModal" data-video-id="E0JKlCgxzlo"
                               data-video-aspect-ratio="16x9">
                                <span class="visually-hidden">
                                    Play video
                                </span>
                                <div class="video-trigger__image-wrapper">
                                    <img alt="video thumbnail" class="img-fluid"
                                         src="https://img.youtube.com/vi/E0JKlCgxzlo/maxresdefault.jpg">
                                </div>
                            </a>
                        
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="paper">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Paper</span>
                    </div>
                    <h5 class="card__heading">
                        A general and efficient representation of ancestral recombination graphs
                    </h5>
                    
                        <div class="wysiwyg">
                            <p>
                                06 September, 2024
                            </p>
                            <p>
                                GENETICS (2024) Wong <em>et al.</em>
                            </p>
                            <p>
                                doi: <a href="https://doi.org/10.1093/genetics/iyae100">10.1093/genetics/iyae100</a>
                            </p>
                            <p>
                                <p>This paper recounts recent developments in competing representations of Ancestral Recombination Graphs.
It presents a simple formalism that defines an ARG in terms of specific genomes and their intervals of
genetic inheritance, showing how this generalizes to encompasses the outputs of recent inference methods.
Also see the <a href="/news/20240906-arg.html">tskit news item</a> linking to a layman’s summary.</p>

                            </p>
                        </div>
                        <a href="https://doi.org/10.1093/genetics/iyae100" class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        Read the article
                                    </span>
                        </a>
                    
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="paper">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Paper</span>
                    </div>
                    <h5 class="card__heading">
                        <tt>tstrait</tt>
                    : a quantitative trait simulator for ancestral recombination graphs
                    </h5>
                    
                        <div class="wysiwyg">
                            <p>
                                25 May, 2024
                            </p>
                            <p>
                                Bioinformatics (2024) Tagami <em>et al.</em>
                            </p>
                            <p>
                                doi: <a href="https://doi.org/10.1093/bioinformatics/btae334">10.1093/bioinformatics/btae334</a>
                            </p>
                            <p>
                                <p>This describes a new tskit-based simulator that maps genotypes to phenotypes efficiently for large numbers
of genomes. The new software, <code class="language-plaintext highlighter-rouge">tstrait</code>, takes an existing tree sequence containing mutations,
and simulates a quantitative phenotypic trait whose value is partially determined by a selected
set of causal mutations in the tree sequence. Also see the
<a href="/news/20230920-tstrait-0.0.1.html">tskit news item</a> about the associated software release.</p>

                            </p>
                        </div>
                        <a href="https://doi.org/10.1093/bioinformatics/btae334" class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        Read the article
                                    </span>
                        </a>
                    
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="paper">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Paper</span>
                    </div>
                    <h5 class="card__heading">
                        The era of the ARG: An introduction to ancestral recombination graphs and their significance in empirical evolutionary genomics
                    </h5>
                    
                        <div class="wysiwyg">
                            <p>
                                18 January, 2024
                            </p>
                            <p>
                                PLOS Genetics (2024) Lewanski <em>et al.</em>
                            </p>
                            <p>
                                doi: <a href="https://doi.org/10.1371/journal.pgen.1011110">10.1371/journal.pgen.1011110</a>
                            </p>
                            <p>
                                <p>Although not written by anyone from the core <code class="language-plaintext highlighter-rouge">tskit</code> development community, this review gives a good general introduction to Ancestral Recombination Graphs, and discusses how tree sequences can be used to represent them.</p>

                            </p>
                        </div>
                        <a href="https://doi.org/10.1371/journal.pgen.1011110" class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        Read the article
                                    </span>
                        </a>
                    
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="paper">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Paper</span>
                    </div>
                    <h5 class="card__heading">
                        link-ancestors: fast simulation of local ancestry with tree sequence software
                    </h5>
                    
                        <div class="wysiwyg">
                            <p>
                                20 November, 2023
                            </p>
                            <p>
                                Bioinformatics Advances (2023) Tsambos <em>et al.</em>
                            </p>
                            <p>
                                doi: <a href="https://doi.org/10.1093/bioadv/vbad163">10.1093/bioadv/vbad163</a>
                            </p>
                            <p>
                                <p>Describes an efficient algorithm for calculating “local ancestry” from tree sequences simulated by e.g. <code class="language-plaintext highlighter-rouge">msprime</code> or <code class="language-plaintext highlighter-rouge">SLiM</code>. This takes a tree sequence and a set of ancestors, and returns a record of the corresponding local ancestry.</p>

                            </p>
                        </div>
                        <a href="https://doi.org/10.1093/bioadv/vbad163" class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        Read the article
                                    </span>
                        </a>
                    
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="tutorial">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Tutorial</span>
                    </div>
                    <h5 class="card__heading">
                        ARGs as tree sequences
                    </h5>
                    
                    
                    
                        <div class="wysiwyg">
                            <p>
                                10 November, 2023
                            </p>
                            <p> <p>The <code class="language-plaintext highlighter-rouge">tskit</code> library provides a convenient way to encode and work with
ancestral recombination graphs (ARGs). This tutorial introduces the
concept of “full ARGs” in tskit, the link between these ARGs and
more simplified graphical representations, and how to use <code class="language-plaintext highlighter-rouge">tskit</code>
to work with inheritance graphs in general.</p>
 </p>
                        </div>
                        <a
                            
                                href=/tutorials/args.html
                            
                           class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        See the tutorial
                                    </span>
                        </a>
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="paper">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Paper</span>
                    </div>
                    <h5 class="card__heading">
                        On the genes, genealogies, and geographies of Quebec
                    </h5>
                    
                        <div class="wysiwyg">
                            <p>
                                25 May, 2023
                            </p>
                            <p>
                                Science (2023) Anderson-Trocmé <em>et al.</em>
                            </p>
                            <p>
                                doi: <a href="https://doi.org/10.1126/science.add5300">10.1126/science.add5300</a>
                            </p>
                            <p>
                                <p>This paper uses a new feature implemented in msprime to simulate genetic transmission through a known pedigree.
In this case, a human pedigree based on 4 million Canadian genealogical records was used to highlight the
multiple impacts of geography on human migration. The paper is also associated with a freely available 
<a href="https://zenodo.org/records/7702392">simulated whole-genome dataset</a> of almost 1.5 million individuals.</p>

                            </p>
                        </div>
                        <a href="https://doi.org/10.1126/science.add5300" class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        Read the article
                                    </span>
                        </a>
                    
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="tutorial">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Tutorial</span>
                    </div>
                    <h5 class="card__heading">
                        Tskit Rust
                    </h5>
                    
                    
                    
                        <div class="wysiwyg">
                            <p>
                                15 November, 2022
                            </p>
                            <p> <p>A walkthrough of using tskit the tskit API from Rust.</p>
 </p>
                        </div>
                        <a
                            
                                href=https://tskit.dev/tutorials/rust
                            
                           class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        See the tutorial
                                    </span>
                        </a>
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="video">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Video</span>
                    </div>
                    <h5 class="card__heading">
                        The ARG revolution in population and statistical genetics
                    </h5>
                    
                    
                        <div class="wysiwyg">
                            <p>
                                30 September, 2022
                            </p>
                            <p>
                                Wilder Wohns at SMBE Everywhere</br>
                                <p>In this video, Wilder Wohns gives examples of how ancestral recombination graphs (ARGs)
stored in <code class="language-plaintext highlighter-rouge">tskit</code> format have the potential to revolutionise population and statistical
genetics. In particular, inferred tree sequences can be used for inferring geographical
locations of human ancestry, and can be used to improve the speed and accuracy of genetic
association analysis.</p>

                            </p>

                        </div>
                        
<!--                            Vimeo embeds don't work with out a code from the video creator, so just link-->
                            <a href="https://www.vimeo.com/759905338/21922bd6c2" class="video-trigger">
                                <span class="visually-hidden">
                                    Play video
                                </span>
                            <div class="video-trigger__image-wrapper">
                                <img alt="video thumbnail" class="img-fluid"
                                     src="https://i.imgur.com/pOF8gGY.png">
                            </div>
                            </a>
                        
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="paper">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Paper</span>
                    </div>
                    <h5 class="card__heading">
                        A unified genealogy of modern and ancient genomes
                    </h5>
                    
                        <div class="wysiwyg">
                            <p>
                                25 February, 2022
                            </p>
                            <p>
                                Science (2022) Wohns <em>et al.</em>
                            </p>
                            <p>
                                doi: <a href="https://doi.org/10.1126/science.abi8264">10.1126/science.abi8264</a>
                            </p>
                            <p>
                                <p>This paper describes using <code class="language-plaintext highlighter-rouge">tskit</code>, <code class="language-plaintext highlighter-rouge">tsinfer</code> and <code class="language-plaintext highlighter-rouge">tsdate</code> to create 
a unified tree sequence of 3601 modern and 8  ancient human genome
sequences compiled from eight datasets. Then estimates
of ancestor geographic location are introduced that
recapitulate key features of human history.</p>

                            </p>
                        </div>
                        <a href="https://doi.org/10.1126/science.abi8264" class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        Read the article
                                    </span>
                        </a>
                    
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="paper">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Paper</span>
                    </div>
                    <h5 class="card__heading">
                        Efficient ancestry and mutation simulation with msprime 1.0
                    </h5>
                    
                        <div class="wysiwyg">
                            <p>
                                13 December, 2021
                            </p>
                            <p>
                                Genetics (2021) Baumdicker <em>et al.</em>
                            </p>
                            <p>
                                doi: <a href="https://doi.org/10.1093/genetics/iyab229">10.1093/genetics/iyab229</a>
                            </p>
                            <p>
                                <p>The accompanying paper to the <code class="language-plaintext highlighter-rouge">msprime</code> 1.0 release, summarising its features
and performance and discussing its development model.</p>

                            </p>
                        </div>
                        <a href="https://doi.org/10.1093/genetics/iyab229" class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        Read the article
                                    </span>
                        </a>
                    
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="tutorial">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Tutorial</span>
                    </div>
                    <h5 class="card__heading">
                        Tskit Terminology and Concepts
                    </h5>
                    
                    
                    
                        <div class="wysiwyg">
                            <p>
                                21 June, 2021
                            </p>
                            <p> <p>This tutorial serves as an introduction to the terminology and concepts in tskit, and its underlying data structures.</p>
 </p>
                        </div>
                        <a
                            
                                href=/tutorials/terminology_and_concepts.html
                            
                           class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        See the tutorial
                                    </span>
                        </a>
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="tutorial">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Tutorial</span>
                    </div>
                    <h5 class="card__heading">
                        Getting started with tskit
                    </h5>
                    
                    
                    
                        <div class="wysiwyg">
                            <p>
                                21 June, 2021
                            </p>
                            <p> <p>You’ve run some simulations or inference methods, and you now have a TreeSequence object; what now? This tutorial is
aimed users who are new to tskit and would like to get some basic tasks completed.</p>
 </p>
                        </div>
                        <a
                            
                                href=/tutorials/getting_started.html
                            
                           class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        See the tutorial
                                    </span>
                        </a>
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="tutorial">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Tutorial</span>
                    </div>
                    <h5 class="card__heading">
                        Analysing Tree Sequences
                    </h5>
                    
                    
                    
                        <div class="wysiwyg">
                            <p>
                                21 June, 2021
                            </p>
                            <p> <p>This tutorial aims to give a quick overview of how the tskit statistics APIs work and how to use them effectively.</p>
 </p>
                        </div>
                        <a
                            
                                href=/tutorials/analysing_tree_sequences.html
                            
                           class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        See the tutorial
                                    </span>
                        </a>
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="tutorial">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Tutorial</span>
                    </div>
                    <h5 class="card__heading">
                        Tables and Editing
                    </h5>
                    
                    
                    
                        <div class="wysiwyg">
                            <p>
                                19 June, 2021
                            </p>
                            <p> <p>The underlying representation of a tree sequence in <code class="language-plaintext highlighter-rouge">tskit</code> is a set of tables. This tutorial shows how to access and
manipulate these tables.</p>
 </p>
                        </div>
                        <a
                            
                                href=/tutorials/tables_and_editing.html
                            
                           class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        See the tutorial
                                    </span>
                        </a>
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="tutorial">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Tutorial</span>
                    </div>
                    <h5 class="card__heading">
                        Analysing Trees
                    </h5>
                    
                    
                    
                        <div class="wysiwyg">
                            <p>
                                19 June, 2021
                            </p>
                            <p> <p>tskit provides single tree traversals, algorithms and phylogenetic statistics, of which this tutorial
gives an overview.</p>
 </p>
                        </div>
                        <a
                            
                                href=/tutorials/analysing_trees.html
                            
                           class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        See the tutorial
                                    </span>
                        </a>
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="tutorial">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Tutorial</span>
                    </div>
                    <h5 class="card__heading">
                        Working with Metadata
                    </h5>
                    
                    
                    
                        <div class="wysiwyg">
                            <p>
                                16 June, 2021
                            </p>
                            <p> <p>This tutorial gives an overview of <code class="language-plaintext highlighter-rouge">tskit</code>’s metadata system. This allows arbitrary, documented metadata to be attached to 
entities in tree sequences.</p>
 </p>
                        </div>
                        <a
                            
                                href=/tutorials/metadata.html
                            
                           class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        See the tutorial
                                    </span>
                        </a>
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="tutorial">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Tutorial</span>
                    </div>
                    <h5 class="card__heading">
                        Do you really need mutations?
                    </h5>
                    
                    
                    
                        <div class="wysiwyg">
                            <p>
                                06 June, 2021
                            </p>
                            <p> <p>In tree sequences, the genetic genealogy exists independently of the mutations that generate genetic variation, and often we are primarily interested in genetic variation because of what it can tell us about those genealogies. This tutorial aims to illustrate when we can leave mutations and genetic variation aside and study the genealogies directly.</p>
 </p>
                        </div>
                        <a
                            
                                href=/tutorials/no_mutations.html
                            
                           class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        See the tutorial
                                    </span>
                        </a>
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="tutorial">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Tutorial</span>
                    </div>
                    <h5 class="card__heading">
                        Visualization
                    </h5>
                    
                    
                    
                        <div class="wysiwyg">
                            <p>
                                29 May, 2021
                            </p>
                            <p> <p>It is often helpful to visualize a single tree — or multiple trees along a tree sequence — together with
sites and mutations. <code class="language-plaintext highlighter-rouge">tskit</code> provides functions to do this, outputting either plain ascii or unicode text,
or the more flexible Scalable Vector Graphics (SVG) format. This tutorial illustrates various examples.</p>
 </p>
                        </div>
                        <a
                            
                                href=/tutorials/viz.html
                            
                           class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        See the tutorial
                                    </span>
                        </a>
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="tutorial">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Tutorial</span>
                    </div>
                    <h5 class="card__heading">
                        Tskit and R
                    </h5>
                    
                    
                    
                        <div class="wysiwyg">
                            <p>
                                11 May, 2021
                            </p>
                            <p> <p>To interface with <code class="language-plaintext highlighter-rouge">tskit</code> in R, we can use the <code class="language-plaintext highlighter-rouge">reticulate</code> R package, which lets you call
Python functions within an R session. In this short tutorial, we’ll go through a couple of
examples to show you how to get started.</p>
 </p>
                        </div>
                        <a
                            
                                href=/tutorials/tskitr.html
                            
                           class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        See the tutorial
                                    </span>
                        </a>
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="tutorial">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Tutorial</span>
                    </div>
                    <h5 class="card__heading">
                        Completing forwards simulations
                    </h5>
                    
                    
                    
                        <div class="wysiwyg">
                            <p>
                                20 January, 2021
                            </p>
                            <p> <p>In this tutorial we show how to combine the best of both forwards and backwards simulation approaches by simulating
the recent past using a forwards-time simulator and then complete the simulation of the ancient past using <code class="language-plaintext highlighter-rouge">msprime</code>.</p>
 </p>
                        </div>
                        <a
                            
                                href=/tutorials/completing_forward_sims.html
                            
                           class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        See the tutorial
                                    </span>
                        </a>
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="tutorial">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Tutorial</span>
                    </div>
                    <h5 class="card__heading">
                        msprime tutorials
                    </h5>
                    
                    
                    
                        <div class="wysiwyg">
                            <p>
                                19 January, 2021
                            </p>
                            <p> <p>A set of tutorials for <code class="language-plaintext highlighter-rouge">msprime</code>. Covering demography, bottlenecks and introgression.</p>
 </p>
                        </div>
                        <a
                            
                                href=/tutorials/advanced_msprime.html
                            
                           class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        See the tutorial
                                    </span>
                        </a>
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="video">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Video</span>
                    </div>
                    <h5 class="card__heading">
                        Inferring the ancestry of everyone
                    </h5>
                    
                    
                        <div class="wysiwyg">
                            <p>
                                02 December, 2020
                            </p>
                            <p>
                                Jerome Kelleher at PopGen Vienna</br>
                                

                            </p>

                        </div>
                        
                            <a href="https://www.youtube.com/watch?v=yrDSLlssoek" class="video-trigger"
                               data-bs-toggle="modal" data-bs-target="#globalModal" data-video-id="yrDSLlssoek"
                               data-video-aspect-ratio="16x9">
                                <span class="visually-hidden">
                                    Play video
                                </span>
                                <div class="video-trigger__image-wrapper">
                                    <img alt="video thumbnail" class="img-fluid"
                                         src="https://img.youtube.com/vi/yrDSLlssoek/maxresdefault.jpg">
                                </div>
                            </a>
                        
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="paper">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Paper</span>
                    </div>
                    <h5 class="card__heading">
                        Efficiently Summarizing Relationships in Large Samples: A General Duality Between Statistics of Genealogies and Genomes
                    </h5>
                    
                        <div class="wysiwyg">
                            <p>
                                01 July, 2020
                            </p>
                            <p>
                                Genetics (2020) Ralph <em>et al.</em>
                            </p>
                            <p>
                                doi: <a href="https://doi.org/10.1534/genetics.120.303253">10.1534/genetics.120.303253</a>
                            </p>
                            <p>
                                <p>This paper shows that we can think about any statistic that works on
sequence data in an equivalent (and more powerful) way in terms of the
underlying trees, and that we can compute these statistics very efficiently.
Read this paper if you would like more technical details on how the
underlying data structures work and an introduction to incremental
tree sequence algorithms.</p>

                            </p>
                        </div>
                        <a href="https://doi.org/10.1534/genetics.120.303253" class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        Read the article
                                    </span>
                        </a>
                    
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="video">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Video</span>
                    </div>
                    <h5 class="card__heading">
                        Population genetics using inferred genealogies and tree sequences
                    </h5>
                    
                    
                        <div class="wysiwyg">
                            <p>
                                17 June, 2020
                            </p>
                            <p>
                                Leo Speidel at Phyloseminar.org</br>
                                <p>A talk at <a href="https://phyloseminar.org">phyloseminar.org</a>,
that reviews the quickly gorwing set of tools that extract information about evolutionary past
from ARGs, focused on <a href="https://myersgroup.github.io/relate/">Relate</a> and <a href="https://tskit.dev/tsinfer/docs/latest">tsinfer</a>.</p>

<!--
*Inferred genealogies are a reconstruction of our shared genetic past; embedded in them is information about key evolutionary events that have shaped the genetic variation observed in present-day individuals. Stored in tree sequence format, tree-based inferences can be extremely powerful and computationally efficient, allowing for joint, genome-wide analyses of large sample sizes. This seminar will aim at giving an overview of the quickly growing set of tools that extract information about our evolutionary past from these trees. We will focus particularly on Relate and tsinfer and discuss tree-based inferences of genetic structure, demographic histories, and natural selection, among others.*
-->

                            </p>

                        </div>
                        
                            <a href="https://www.youtube.com/watch?v=kgSGhoXb2Pg" class="video-trigger"
                               data-bs-toggle="modal" data-bs-target="#globalModal" data-video-id="kgSGhoXb2Pg"
                               data-video-aspect-ratio="16x9">
                                <span class="visually-hidden">
                                    Play video
                                </span>
                                <div class="video-trigger__image-wrapper">
                                    <img alt="video thumbnail" class="img-fluid"
                                         src="https://img.youtube.com/vi/kgSGhoXb2Pg/maxresdefault.jpg">
                                </div>
                            </a>
                        
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="video">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Video</span>
                    </div>
                    <h5 class="card__heading">
                        Tree sequences and inference
                    </h5>
                    
                    
                        <div class="wysiwyg">
                            <p>
                                22 May, 2020
                            </p>
                            <p>
                                Yan Wong at Phyloseminar.org</br>
                                <p>A walk through of the workings of the <a href="https://tskit.dev/software/tsinfer.html">tsinfer</a> algorithm,
a rapid way to infer <code class="language-plaintext highlighter-rouge">tskit</code> tree sequences from existing genetic variation data.</p>

<p>A talk at <a href="https://phyloseminar.org">phyloseminar.org</a>.</p>

                            </p>

                        </div>
                        
                            <a href="https://www.youtube.com/watch?v=yB4uSle66Pw" class="video-trigger"
                               data-bs-toggle="modal" data-bs-target="#globalModal" data-video-id="yB4uSle66Pw"
                               data-video-aspect-ratio="16x9">
                                <span class="visually-hidden">
                                    Play video
                                </span>
                                <div class="video-trigger__image-wrapper">
                                    <img alt="video thumbnail" class="img-fluid"
                                         src="https://img.youtube.com/vi/yB4uSle66Pw/maxresdefault.jpg">
                                </div>
                            </a>
                        
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="video">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Video</span>
                    </div>
                    <h5 class="card__heading">
                        Tree sequence fundamentals
                    </h5>
                    
                    
                        <div class="wysiwyg">
                            <p>
                                08 April, 2020
                            </p>
                            <p>
                                Wilder Wohns at Phyloseminar.org</br>
                                <p>A talk at <a href="https://phyloseminar.org">phyloseminar.org</a>,
that gives an overview of tskit and the related ecosystem (as of 2020).</p>

<!--
*The tree sequence data structure is a concise encoding of whole-genome ancestry and sequence data, with a rapidly maturing software ecosystem. Tree sequences encode the genealogical history of individuals subject to recombination and gene conversion, providing an efficient means of working with phylogenetic networks and other applications with a linear sequence of correlated trees. The tskit (tree sequence toolkit) library is a comprehensive framework for navigating and manipulating tree sequences using Python and C APIs and also hosts a unified interface for efficiently calculating a growing number of summary statistics. The ecosystem developing around this central technology now includes several genome simulators, a highly-scalable method for inferring ancestry from data, tsinfer, and an efficient method for dating ancestral nodes in a tree sequence topology, tsdate. In this primer session, I will introduce tskit and the tree sequence data structure as well as use downloadable Jupyter Notebooks to demonstrate the simulation of genomic datasets with msprime and the calculation of population genetic statistics with tskit.*
-->

                            </p>

                        </div>
                        
                            <a href="https://www.youtube.com/watch?v=YrZTKjLzZY0" class="video-trigger"
                               data-bs-toggle="modal" data-bs-target="#globalModal" data-video-id="YrZTKjLzZY0"
                               data-video-aspect-ratio="16x9">
                                <span class="visually-hidden">
                                    Play video
                                </span>
                                <div class="video-trigger__image-wrapper">
                                    <img alt="video thumbnail" class="img-fluid"
                                         src="https://img.youtube.com/vi/YrZTKjLzZY0/maxresdefault.jpg">
                                </div>
                            </a>
                        
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="paper">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Paper</span>
                    </div>
                    <h5 class="card__heading">
                        Inferring whole-genome histories in large population datasets
                    </h5>
                    
                        <div class="wysiwyg">
                            <p>
                                02 September, 2019
                            </p>
                            <p>
                                Nature Genetics (2019) Kelleher <em>et al.</em>
                            </p>
                            <p>
                                doi: <a href="https://doi.org/10.1038/s41588-019-0483-y">10.1038/s41588-019-0483-y</a>
                            </p>
                            <p>
                                <p>Start here if you’re new to tree sequences. This paper introduces
<a href="/tsinfer.html">tsinfer</a>, the method to infer tree sequence
topologies from genetic variation data.
Please see the <a href="https://www.biorxiv.org/content/10.1101/458067v1">preprint</a>
if you cannot access the Nature Genetics paper.</p>

                            </p>
                        </div>
                        <a href="https://doi.org/10.1038/s41588-019-0483-y" class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        Read the article
                                    </span>
                        </a>
                    
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="video">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Video</span>
                    </div>
                    <h5 class="card__heading">
                        Succinct tree sequences for megasample genomics (47:03)
                    </h5>
                    
                    
                        <div class="wysiwyg">
                            <p>
                                26 April, 2019
                            </p>
                            <p>
                                Jerome Kelleher at MIA</br>
                                

                            </p>

                        </div>
                        
                            <a href="https://www.youtube.com/watch?v=X1GEuQrF1jQ" class="video-trigger"
                               data-bs-toggle="modal" data-bs-target="#globalModal" data-video-id="X1GEuQrF1jQ"
                               data-video-aspect-ratio="16x9">
                                <span class="visually-hidden">
                                    Play video
                                </span>
                                <div class="video-trigger__image-wrapper">
                                    <img alt="video thumbnail" class="img-fluid"
                                         src="https://img.youtube.com/vi/X1GEuQrF1jQ/maxresdefault.jpg">
                                </div>
                            </a>
                        
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="video">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Video</span>
                    </div>
                    <h5 class="card__heading">
                        Introduction to the tree sequence toolchain
                    </h5>
                    
                    
                        <div class="wysiwyg">
                            <p>
                                25 April, 2019
                            </p>
                            <p>
                                Wilder Wohns at MIA Primer</br>
                                

                            </p>

                        </div>
                        
                            <a href="https://www.youtube.com/watch?v=X1GEuQrF1jQ" class="video-trigger"
                               data-bs-toggle="modal" data-bs-target="#globalModal" data-video-id="X1GEuQrF1jQ"
                               data-video-aspect-ratio="16x9">
                                <span class="visually-hidden">
                                    Play video
                                </span>
                                <div class="video-trigger__image-wrapper">
                                    <img alt="video thumbnail" class="img-fluid"
                                         src="https://img.youtube.com/vi/X1GEuQrF1jQ/maxresdefault.jpg">
                                </div>
                            </a>
                        
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="paper">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Paper</span>
                    </div>
                    <h5 class="card__heading">
                        Tree‐sequence recording in SLiM opens new horizons for forward‐time simulation of whole genomes
                    </h5>
                    
                        <div class="wysiwyg">
                            <p>
                                22 November, 2018
                            </p>
                            <p>
                                Molecular Ecology Resources (2019) Haller <em>et al.</em>
                            </p>
                            <p>
                                doi: <a href="https://doi.org/10.1111/1755-0998.12968">10.1111/1755-0998.12968</a>
                            </p>
                            <p>
                                <p>Continuing on from the 2018 PLOS Computational Biology paper,
we discuss here how the tree
sequence recording method was implemented in the powerful
<a href="https://messerlab.org/slim/">SLiM</a> simulator. We
show how some simulations are orders of magnitude more efficient
and examples of the new possibilities that keeping a full
record of the genetic ancestry makes available.</p>

                            </p>
                        </div>
                        <a href="https://doi.org/10.1111/1755-0998.12968" class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        Read the article
                                    </span>
                        </a>
                    
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="paper">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Paper</span>
                    </div>
                    <h5 class="card__heading">
                        Efficient pedigree recording for fast population genetics simulation
                    </h5>
                    
                        <div class="wysiwyg">
                            <p>
                                01 November, 2018
                            </p>
                            <p>
                                PLOS Computational Biology (2018) Kelleher <em>et al.</em>
                            </p>
                            <p>
                                doi: <a href="https://doi.org/10.1371/journal.pcbi.1006581">10.1371/journal.pcbi.1006581</a>
                            </p>
                            <p>
                                <p>Forwards-in-time simulations are very flexible but also usually very CPU
intensive. This paper shows how we used tree sequences to make
forwards-in-time simulations both more efficient and even more flexible.</p>

                            </p>
                        </div>
                        <a href="https://doi.org/10.1371/journal.pcbi.1006581" class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        Read the article
                                    </span>
                        </a>
                    
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="video">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Video</span>
                    </div>
                    <h5 class="card__heading">
                        Simulating, storing & processing genetic variation data for millions of samples
                    </h5>
                    
                    
                        <div class="wysiwyg">
                            <p>
                                26 April, 2017
                            </p>
                            <p>
                                Jerome Kelleher at MIA</br>
                                

                            </p>

                        </div>
                        
                            <a href="https://www.youtube.com/watch?v=MH2b9iU4oUA" class="video-trigger"
                               data-bs-toggle="modal" data-bs-target="#globalModal" data-video-id="MH2b9iU4oUA"
                               data-video-aspect-ratio="16x9">
                                <span class="visually-hidden">
                                    Play video
                                </span>
                                <div class="video-trigger__image-wrapper">
                                    <img alt="video thumbnail" class="img-fluid"
                                         src="https://img.youtube.com/vi/MH2b9iU4oUA/maxresdefault.jpg">
                                </div>
                            </a>
                        
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
            <div class="col-lg-4" data-r-type="paper">
                <!-- START: card -->
                <div class="card card--type-3">
                    <div class="card__metadata">
                        <span class="card__category">Paper</span>
                    </div>
                    <h5 class="card__heading">
                        Efficient Coalescent Simulation and Genealogical Analysis for Large Sample Sizes
                    </h5>
                    
                        <div class="wysiwyg">
                            <p>
                                04 April, 2016
                            </p>
                            <p>
                                PLOS Computational Biology (2016) Kelleher <em>et al.</em>
                            </p>
                            <p>
                                doi: <a href="https://doi.org/10.1371/journal.pcbi.1004842">10.1371/journal.pcbi.1004842</a>
                            </p>
                            <p>
                                <p>This is where the tree sequence data structure
was first described. Here we introduce the
<a href="/software/msprime.html">msprime</a> coalescent simulator and
the core algorithms and data structures that would later
be separated out into <a href="software/tskit.html">tskit</a>.
Read this paper if you would like to find out more about
coalescent simulation, or to understand the core tree sequence
algorithms and theoretical results. <em>Note:</em> much
of the terminology has been updated since this original publication
as the models were generalised.</p>

                            </p>
                        </div>
                        <a href="https://doi.org/10.1371/journal.pcbi.1004842" class="list__item__link link link--type-1 link--type-1--external"
                           target="_blank"
                           rel="noopener noreferrer">
                                    <span class="link__text">
                                        Read the article
                                    </span>
                        </a>
                    
                    
                    

<!--                    <div class="card__metadata">-->
<!--                        <span class="card__tag">Population genetics</span>-->
<!--                    </div>-->
                </div>
                <!-- END: card -->
            </div>
            
        </div>
    </div>

    <!-- START: pagination -->
    <div class="container" style="display: none">
        <div class="row">
            <div class="col">
                <nav class="pagination-wrapper">
                    <ul class="pagination list list--unstyled list--horizontal">
                        <li class="list__item">
                                    <span class="pagination__number pagination__number--prev disabled">
                                        Prev
                                    </span>
                        </li>
                        <li class="list__item">
                                    <span class="pagination__number pagination__number--current">
                                        1
                                    </span>
                        </li>
                        <li class="list__item">
                            <a class="pagination__number" href="#">
                                2
                            </a>
                        </li>
                        <li class="list__item">
                            <a class="pagination__number" href="#">
                                3
                            </a>
                        </li>
                        <li class="list__item">
                            <a class="pagination__number" href="#">
                                4
                            </a>
                        </li>
                        <li class="list__item">
                                    <span class="pagination__dots">
                                        &hellip;
                                    </span>
                        </li>
                        <li class="list__item">
                            <a class="pagination__number" href="#">
                                12
                            </a>
                        </li>
                        <li class="list__item">
                            <a class="pagination__number pagination__number--next" href="#">
                                Next
                            </a>
                        </li>
                    </ul>
                </nav>
            </div>
        </div>
    </div>
    <!-- END: pagination -->

</section>
<!-- END: cards-type-3 -->

    </main>
    <!-- END: content -->

    <!-- START: footer -->
        <footer class="footer">
    <div class="container">
        <div class="row">
            <div class="col-md-5 offset-md-1">
                <a class="footer__logo" href="./">
                    <!-- START: icon-logo -->
                    <svg class="icon icon--logo" xmlns="http://www.w3.org/2000/svg" x="0" y="0"
                         viewBox="0 0 84.8 68">
                        <style>
                            .logo-st0 {
                                fill: #085167
                            }

                            .logo-st1 {
                                fill: #14e2a8
                            }
                        </style>
                        <path class="logo-st0"
                              d="M2 61.4v-8.2H0v-3.4h2v-5h4.8v5h5v3.4h-5v8.2c0 1.9.8 2.9 2.5 2.9.9 0 1.7-.2 2.4-.6v3.5c-1.1.5-2.4.8-3.7.8-4.2 0-6-2.8-6-6.6"></path>
                        <path class="logo-st0"
                              d="M62.7 67.5v-3.4h1.9V53.2h-1.9v-3.4h6.8v14.3h1.9v3.4h-8.7zm1.1-23.3c-.1-1.6 1.2-2.9 2.7-3 1.6-.1 2.9 1.2 3 2.7.1 1.6-1.2 2.9-2.7 3h-.1c-1.5.1-2.8-1-2.9-2.5v-.2"></path>
                        <path class="logo-st0"
                              d="M39.6 67.5v-3.4h1.9V45.4h-1.9V42h6.8v14.4h4l1.7-3.2h-2.5v-3.4h9.6v3.4h-2.6L53.7 58l4.2 6h2.8v3.4h-9.9V64h1.7l-2.8-4.6h-3.4V64h1.9v3.4l-8.6.1z"></path>
                        <path class="logo-st0"
                              d="M23.4 57c4.5 1.3 6.4 2.7 6.4 5.9 0 3.3-2.5 5.1-6.4 5.1-2.6 0-4.3-.7-5.4-1.9l-.1 1.3h-3.4v-6.1h3.7c.7 1.9 2.5 3.2 4.5 3.1 1.3 0 2.3-.5 2.3-1.6s-1-1.5-1.7-1.7l-2.6-.7c-2.8-.7-6.2-1.7-6.2-6 0-3.4 2.4-5.2 6-5.2 1.8 0 3.6.6 5 1.8l.1-1.3H29v5.9h-3.7c-.4-1.7-1.9-2.9-3.6-2.9-1.3 0-2.3.5-2.3 1.7 0 1 .6 1.4 1.8 1.7l2.2.9z"></path>
                        <path class="logo-st0"
                              d="M75 61.4v-8.2h-2v-3.4h2v-5h4.8v5h5v3.4h-5v8.2c0 1.9.8 2.9 2.5 2.9.9 0 1.7-.2 2.4-.6v3.5c-1.1.5-2.4.8-3.7.8-4.2 0-6-2.8-6-6.6"></path>
                        <path class="logo-st1" d="M75 28.8h4.8v12.5H75z"></path>
                        <path class="logo-st1" d="M41.5 28.8h4.8v9.6h-4.8z"></path>
                        <path class="logo-st1" d="M41.5 24h38.3v4.8H41.5z"></path>
                        <path class="logo-st1" d="M58.3 14.4h4.8V24h-4.8z"></path>
                        <path class="logo-st1" d="M2 14.4h4.8v26.9H2z"></path>
                        <path class="logo-st1" d="M2 9.6h61.1v4.8H2z"></path>
                        <path class="logo-st1" d="M34.6 0h4.8v9.6h-4.8z"></path>
                    </svg>
                    <!-- END: icon-logo -->
                    <span class="visually-hidden">
                        tskit
                    </span>
                </a>
            </div>
            <div class="col">
                <ul class="list list--unstyled">
                    <li class="list__item">
                        <a href="https://github.com/tskit-dev/tskit-site/issues/new/choose" class="list__item__link link link--type-1">
                                <span class="link__text">
                                    Report a website issue
                                </span>
                        </a>
                    </li>
                    <li class="list__item">
                        <a href="https://github.com/tskit-dev" class="list__item__link link link--type-1">
                                <span class="link__text">
                                    Github
                                </span>
                        </a>
                    </li>
                    <li class="list__item">
                        <a href="https://github.com/tskit-dev/.github/blob/main/CODE_OF_CONDUCT.md" class="list__item__link link link--type-1">
                                <span class="link__text">
                                    Code of conduct
                                </span>
                        </a>
                    </li>
                    <li class="list__item">
                        <a href="/governance/" class="list__item__link link link--type-1">
                                <span class="link__text">
                                    Governance
                                </span>
                        </a>
                    </li>
                    <li class="list__item">
                        <a href="https://github.com/tskit-dev/tskit-site/blob/main/LICENSE" class="list__item__link link link--type-1">
                                <span class="link__text">
                                    Licence
                                </span>
                        </a>
                    </li>
                    <li class="list__item">
                        <a href="/ecosystem-status/" class="list__item__link link link--type-1">
                                <span class="link__text">
                                    Ecosystem Status
                                </span>
                        </a>
                    </li>
                </ul>
            </div>
        </div>
    </div>
</footer>

    <!-- END: footer -->
    <!-- START: global modal -->
    <div class="modal fade"
         id="globalModal"
         tabindex="-1"
         role="dialog"
         aria-labelledby="globalModalLabel"
         aria-hidden="true"
    >
        <div class="modal-dialog modal-dialog-centered modal-xl" role="document">
            <div class="modal-content">
                <button class="navbar-toggler modal-close" type="button" data-bs-dismiss="modal" aria-label="Close">
                    <span class="navbar-toggler-icon"></span>
                    <span class="navbar-toggler-icon"></span>
                    <span class="navbar-toggler-icon"></span>
                </button>
                <div class="modal-header visually-hidden">
                    <h5 class="modal-title" id="globalModalLabel">
                        Modal content
                    </h5>
                </div>
                <div class="modal-body"></div>
            </div>
        </div>
    </div>
    <!-- END: global modal -->

</div>

<script src="/assets/dist/js/main.min.js"></script>

</body>
</html>
