{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "41222b14-2897-4d52-b972-e83db62071db",
   "metadata": {},
   "source": [
    "<div style=\"color: white; background-color: #03303e; display: flex; align-items: flex-start; margin-bottom: 1em\">\n",
    "    <figure style=\"text-align: center; margin: 0 1em 0.5em 0.5em\">\n",
    "    <img src=\"tskit_logo.svg\" style=\"height: 3em;\" />\n",
    "    <figcaption style=\"font-size: 0.6em; line-height: 1em\">Interactive workbooks</figcaption>\n",
    "</figure>\n",
    "    <h1 style=\"margin: auto\">Tskit-based simulation: a quick introduction</h1></div>\n",
    "This tour uses the <em>tskit</em> <a href=\"https://tskit.dev/tskit/docs/stable/python-api.html\">Python interface</a>. Press \"shift-enter\" (or click the toolbar \"run\" button, &#9654;) to run a notebook cell and move to the next one. Repeat this from the top of the workbook to work gradually through the demonstration code."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19a48cc0-ae22-4840-a9dd-2a73b665f32a",
   "metadata": {},
   "source": [
    "## Backward-time simulation\n",
    "\n",
    "An easy way to simulate a tree sequence is to use the backward-time [_msprime_](https://tskit.dev/msprime/docs/stable/intro.html) coalescent simulator. Below is the code to simulate 5 diploid samples from a total population of a million, but you can easily change the number of samples to e.g. 500_000, which will simulate a million DNA sequences of length $1\\times10^4$ basepairs. Re-running this should only take a few seconds, even on a slow computer (try it!)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad71f199",
   "metadata": {},
   "outputs": [],
   "source": [
    "import msprime\n",
    "\n",
    "ts = msprime.sim_ancestry(samples=5, ploidy=2, population_size=1_000_000, sequence_length=1e4, recombination_rate=1e-8)  # NB ploidy=2 is the default, so can be omitted"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06485d70-1c57-4038-83ff-4abd377a329f",
   "metadata": {},
   "source": [
    "In a notebook, you can `display()` the resulting tree sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4961282f",
   "metadata": {},
   "outputs": [],
   "source": [
    "display(ts)  # actually, if this is the last line in the cell, you can even omit the `display()` part, and simply use `ts`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "619320d1-34aa-42c5-aefc-a199d4701619",
   "metadata": {},
   "source": [
    "By default, the simulator has grouped the sampled haploid genomes (\"sample nodes\") into diploid [individuals](https://tskit.dev/tskit/docs/stable/glossary.html#nodes-genomes-or-individuals):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28e2eb0a-1bcd-4f24-8bee-b095c7974c3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "for ind in ts.individuals():\n",
    "    print(f\"Individual {ind.id} has genomes (nodes) with IDs {ind.nodes}\")\n",
    "    if ind.id > 2:\n",
    "        print(\"...\")\n",
    "        break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "058be451-3ea5-4833-82ce-230d53390585",
   "metadata": {},
   "source": [
    "You can perform analysis on this object, such as calculating MRCAs, and general branch-based statistics, including things like PCA:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a896cb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import pyplot as plt\n",
    "pca_output = ts.pca(2, individuals=range(ts.num_individuals))  # PCA with each individual as a point\n",
    "plt.scatter(*pca_output.factors.T)\n",
    "plt.xlabel(\"Principle component 1\")\n",
    "plt.ylabel(\"Principle component 2\");"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87010aa7-6f8e-49e2-8910-2a7f63d7e773",
   "metadata": {},
   "source": [
    "### Adding mutations to a simulated ancestry\n",
    "\n",
    "Often you want a tree sequence with added [sites](https://tskit.dev/tskit/docs/stable/data-model.html#site-table) and [mutations](https://tskit.dev/tskit/docs/stable/data-model.html#mutation-table), which creates realised genetic diversity (but perhaps you [don't really need mutations?](https://tskit.dev/tutorials/no_mutations.html)). Adding neutral mutations can easily be done using `msprime.sim_mutations()`. The code below adds single basepair mutations (causing SNVs) with equal probability between A, T, G, and C, but many other mutation models are available (see https://tskit.dev/msprime/docs/stable/mutations.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44d1a361-2e47-4c7a-9c90-5106cc08d903",
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_with_mutations = msprime.sim_mutations(ts, rate=1e-8)\n",
    "print(f\"Average genetic diversity is {ts_with_mutations.diversity()}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c94268a-94b0-48d6-9321-05687cbc33f3",
   "metadata": {},
   "source": [
    "## Forward-time simulation\n",
    "\n",
    "Forward-time simulation is much more flexible than backward-time simulation, but requires simulating the entire population. We can't just simulate a small number of genomes from a larger population, as we did with _msprime_ above. \n",
    "\n",
    "Various forward-time simulation frameworks exist. A commonly-used one is [SLiM](https://messerlab.org/slim/), which has its own built-in simulation language, Eidos. It's not possible to run SLiM within a browser-based notebook, so below we simply load in the results of the following Eidos script (this simulates a population of 75 diploids, or 150 haploid genomes, a far cry from the millions in the backward-time simulation):\n",
    "\n",
    "```\n",
    "initialize() {\n",
    "    initializeTreeSeq(timeUnit=\"generations\");  // time units appropriate for simple WF model\n",
    "    initializeMutationRate(1e-8);\n",
    "    initializeMutationType(\"m1\", 0.5, \"e\", 0.001);\n",
    "    initializeGenomicElementType(\"g1\", m1, 1.0);\n",
    "    initializeGenomicElement(g1, 0, 999999);\n",
    "    initializeRecombinationRate(1e-8);\n",
    "}\n",
    "1 early() { sim.addSubpop(\"p1\", 75); }\n",
    "100 late() { sim.treeSeqOutput(\"SLiM_output.trees\"); }\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c518928-5a62-40d4-866d-0488616b4461",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tskit\n",
    "slim_ts = tskit.load(\"data/SLiM_output.trees\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "841ec6fb-1436-4ab2-84f1-d7e97d977f12",
   "metadata": {},
   "source": [
    "Another important feature of forward-time simulations can be seen if we plot out the local \"gene tree\" at the start of the simulated genome:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09f03c28-4130-41b4-8c89-51c59e1224e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "node_label_style = (\n",
    "    \".node > .lab {font-size: 50%}\"\n",
    "    \".leaf > .sym {fill: magenta}\"\n",
    "    \".leaf > .lab {text-anchor: start; transform: rotate(90deg) translate(6px)}\"\n",
    ")\n",
    "\n",
    "slim_ts.first().draw_svg(\n",
    "    size=(1000, 400),\n",
    "    time_scale=\"log_time\",\n",
    "    y_axis=True,\n",
    "    symbol_size=3,\n",
    "    y_ticks=[0, 1, 10, 100],\n",
    "    node_labels={u: u for u in slim_ts.samples()},\n",
    "    style = node_label_style,    \n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "170c2156-aab9-4b0c-a124-21a71bc46275",
   "metadata": {},
   "source": [
    "You can see that the ancestry of the 150 magenta \"sample nodes\" does not extend back past 100 generations (\"ticks\"). At 100 generations ago, there are still 3 separate ancestral lineages present (the local tree has \"multiple roots\"), meaning that the simulation has not fully coalesced. This means we are missing important posts of the ancestry, needed to fully understand the relationships between the sample nodes.\n",
    "\n",
    "## Recapitation\n",
    "\n",
    "We can use _msprime_ to simulate the missing ancestry of an existing forward-time simulation, using the clever trick of \"recapitation\". The easiest way to do this is to use the \"pyslim\" library:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccfe27f3-c899-4521-bb10-e61ca0658c4a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "if sys.platform == 'emscripten':  # only needed for jupyterlite to load pyslim\n",
    "    import micropip\n",
    "    await micropip.install(\"pyslim\")\n",
    "\n",
    "import pyslim\n",
    "\n",
    "recap_ts = pyslim.recapitate(slim_ts, ancestral_Ne=500)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f22102c8-4213-440a-995c-a7a137c524d5",
   "metadata": {},
   "source": [
    "And the resulting tree sequence will have fully coalesced, so that each local tree has a single root. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37d9d7de-bfd3-49d7-8f1d-363622da2405",
   "metadata": {},
   "outputs": [],
   "source": [
    "recap_ts.first().draw_svg(\n",
    "    size=(1000, 400),\n",
    "    time_scale=\"log_time\",\n",
    "    y_axis=True,\n",
    "    symbol_size=3,\n",
    "    y_ticks=[0, 1, 10, 100, 1000],\n",
    "    node_labels={u: u for u in recap_ts.samples()},\n",
    "    style = node_label_style,    \n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "667ec091-4246-4513-a85f-aa61d19e9097",
   "metadata": {},
   "source": [
    "Because by definition, neutral mutations don't affect the genealogy, we can add them after the tree sequence (note that by default, `sim_mutations()` removes the mutations generated by the SLiM simulation):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "021117aa-a70b-4bcd-a955-baa543a6465b",
   "metadata": {},
   "outputs": [],
   "source": [
    "mutated_recap_ts = msprime.sim_mutations(recap_ts, rate=1e-8)\n",
    "print(\"Genetic diversity based on variable sites is\", mutated_recap_ts.diversity())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abd244d2-5717-4dae-8565-f11f38bfc599",
   "metadata": {},
   "source": [
    "## Natural selection\n",
    "\n",
    "Incorporating natural selection into simulations is much easier in forward-time rather than backward-time simulation. There is extensive documentation for doing this in SLiM (chapters 9 and 10 of the SLiM manual). However, there is very limited support for approximating the effect of selective sweeps in backward-time in _msprime_ (see [here](https://tskit.dev/msprime/docs/stable/ancestry.html#sec-ancestry-models-selective-sweeps)), which is demonstrated below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb3d313e-0578-4245-83ba-379d7d0e22b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {\"sequence_length\": 1e6, \"population_size\": 1e3, \"random_seed\": 321}\n",
    "\n",
    "# define hard sweep model\n",
    "start_freq = 1.0 / (2 * params[\"population_size\"])\n",
    "end_freq = 1 - start_freq\n",
    "sweep_model = msprime.SweepGenicSelection(\n",
    "    position=params[\"sequence_length\"]/2, start_frequency=start_freq, end_frequency=end_freq, s=0.25, dt=1e-6,\n",
    ")\n",
    "ts = msprime.sim_ancestry(10, model=[sweep_model, \"hudson\"], recombination_rate=1e-7, **params)\n",
    "mutated_ts = msprime.sim_mutations(ts, rate=1e-7)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a31eb165-1855-4830-9b05-4ac403237810",
   "metadata": {},
   "source": [
    "This simulation should show the classic \"diversity dip\" around the selected region. Note that the _tskit_ Python interface makes extensive use of the [numpy](https://numpy.org) numerical Python library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef742cad-7b62-4571-8b67-184d7e1586d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "windows = np.linspace(0, mutated_ts.sequence_length, 50)\n",
    "plt.stairs(mutated_ts.diversity(windows=windows), windows, baseline=None);"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "277af062-a998-4158-9343-12cc35deb851",
   "metadata": {},
   "source": [
    "## Demography and stdpopsim\n",
    "\n",
    "One of the major influences on a population genomic simulation is the demographic history of the population(s) in question: bottlenecks, expansions, population subdivision, etc. Both _msprime_ and _SLiM_ have ways to specify complex demographies (e.g. see https://tskit.dev/msprime/docs/stable/demography.html).\n",
    "\n",
    "However, the Standard Library of Population Genetic Simulation Models ([stdpopsim](https://popsim-consortium.github.io/stdpopsim-docs/stable/introduction.html)) contains already-tested demographic models for many species, which can be much more convenient. Here's an example of dog evolution:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c5ff99b-4e32-43fc-8654-ca8cf71148f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "if sys.platform == 'emscripten':  # only needed for jupyterlite to load stdpopsim & demesdraw\n",
    "    await micropip.install(\"stdpopsim\")\n",
    "    await micropip.install(\"demesdraw\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70979ff9-f9e7-4595-a4a2-70dc94ba2b8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import stdpopsim\n",
    "\n",
    "species = stdpopsim.get_species(\"CanFam\")  # Dog and wolf\n",
    "model = species.get_demographic_model(\"EarlyWolfAdmixture_6F14\")\n",
    "named_populations={pop.description: pop.name for pop in model.populations}\n",
    "contig = species.get_contig(\"chr1\", left=1e6, right=11e6, mutation_rate=model.mutation_rate)\n",
    "\n",
    "samples = {\n",
    "    named_populations[\"Basenji\"]: 10,\n",
    "    named_populations[\"Dingo\"]: 5,\n",
    "    named_populations[\"Croatian wolf\"]: 5,\n",
    "    named_populations[\"Golden jackal\"]: 5,\n",
    "}\n",
    "engine = stdpopsim.get_engine(\"msprime\")  # If you use \"slim\", you can e.g. add exons under negative selection using a DFE\n",
    "ts = engine.simulate(model, contig, samples)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e93d7692-78ac-4dd3-896b-406ed7307973",
   "metadata": {},
   "source": [
    "The [demesdraw](https://grahamgower.github.io/demesdraw) software can be used to show a pictoral representation of the demographic model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42d85f67-4d4d-4404-ba55-00fa8a1e6703",
   "metadata": {},
   "outputs": [],
   "source": [
    "import demesdraw\n",
    "\n",
    "demesdraw.tubes(model.model.to_demes(),\n",
    "    colours={pop_name: f\"tab:{c}\" for pop_name, c in zip(samples.keys(), (\"blue\", \"orange\", \"green\", \"red\"))},\n",
    "    log_time=True,\n",
    ");"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43c23b74-82ea-4a3a-b675-fba08f2bae11",
   "metadata": {},
   "source": [
    "Now the standard analyses can be carried out with the resulting tree sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ea28842-2942-42be-90b5-0605dd2aedc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_pca_by_pop(ts):\n",
    "    pca_output = ts.pca(2)  # PCA with each haploid genome as a point\n",
    "    for pop in ts.populations():\n",
    "        use = ts.samples(population=pop.id)\n",
    "        if use.any():\n",
    "            plt.scatter(*pca_output.factors[use, :].T, label=pop.metadata[\"description\"])\n",
    "    plt.xlabel(\"PC 1\")\n",
    "    plt.ylabel(\"PC 2\")\n",
    "    plt.legend()\n",
    "\n",
    "plot_pca_by_pop(ts)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33f94c54-28ab-4bbd-8300-32e2848c7781",
   "metadata": {},
   "source": [
    "## Simulating multiple chromosomes (advanced)\n",
    "\n",
    "For most purposes, different chromosomes from the same individual can be simulated independently. However, in recent time (e.g. the most recent 10 generations), there is linkage disequilibrium (LD) between chromosomes. This can be simulated in SLiM 5, or in _msprime_ by simulating a single long chromosome, broken up into randomly recombining pieces, as described [in the _msprime_ docs](https://tskit.dev/msprime/docs/stable/ancestry.html#multiple-chromosomes), from which the example below is largely taken. This example treats the genome as a set of 1Mb chromosomes, joined together by 1 basepair regions in which random recombinational segregation occurs (i.e. at rate = $\\log(2)$). In this case we should use the \"DiscreteTimeWrightFisher\" model for the generations when LD exists between the chromosomes; after that we switch to the standard (\"hudson\") coalescent for efficiency."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f27bc83-0bab-4617-b9bc-c7ebda92258a",
   "metadata": {},
   "outputs": [],
   "source": [
    "r_chrom = 1e-8\n",
    "r_break = np.log(2)\n",
    "chrom_lengths = [10e6, 5e6, 1e6]\n",
    "chrom_start = []\n",
    "chrom_end = []\n",
    "pos = 0\n",
    "for L in chrom_lengths:\n",
    "    chrom_start.append(pos)\n",
    "    pos += L\n",
    "    chrom_end.append(pos)\n",
    "    pos += 1\n",
    "chroms = list(zip(chrom_start, chrom_end))\n",
    "print(\"Chromosomes span these positions\", chroms)\n",
    "rates = [r_chrom, r_break, r_chrom, r_break, r_chrom]\n",
    "rate_map = msprime.RateMap(position=np.array(chroms).flatten(), rate=rates)\n",
    "\n",
    "models = [msprime.DiscreteTimeWrightFisher(duration=20), \"hudson\"]\n",
    "\n",
    "# simulate a single long ('joined together') chromosome.\n",
    "joined_ts = msprime.sim_ancestry(10, population_size=1000, recombination_rate=rate_map, model=models)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb63b5a2-a061-4437-8024-aefc7b2f6532",
   "metadata": {},
   "source": [
    "Alternatively, we can do this within _stdpopsim_. Multi-chromosome simulations are not yet native to _stdpopsim_ 0.3, but we can simply pass the required _msprime_ details, by making a generic contig of the right length, with the right recombination map. Then as above, we switch to the \"hudson\" model after 20 generations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "695cce9d-1f14-44a6-a4cb-2d17c6df6e7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This 3 chromosome simulation may take a little time to run\n",
    "\n",
    "contig = species.get_contig(length=max(rate_map), mutation_rate=model.mutation_rate)\n",
    "base_recomb_rate = contig.recombination_map.rate[0]\n",
    "contig.recombination_map = msprime.RateMap(\n",
    "    position=np.array(chroms).flatten(),\n",
    "    rate=[base_recomb_rate, r_break, base_recomb_rate, r_break, base_recomb_rate],\n",
    ")\n",
    "\n",
    "joined_ts = engine.simulate(\n",
    "    model,\n",
    "    contig,\n",
    "    samples,\n",
    "    msprime_model=\"dtwf\",\n",
    "    msprime_change_model=[(20, \"hudson\")],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75a6cf53-8c91-42b2-b31e-4508bae008d9",
   "metadata": {},
   "source": [
    "If each chromosome needs analysing separately, this single \"concatenated\" chromosome can be separated up into separate tree sequences using\n",
    " [`keep_intervals()`](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.keep_intervals)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6d1277c-9216-46f4-9b83-7ed57ca3fc30",
   "metadata": {},
   "outputs": [],
   "source": [
    "ts_list = [joined_ts.keep_intervals([region], simplify=False).trim() for region in chroms]\n",
    "\n",
    "for i, ts in enumerate(ts_list):\n",
    "    print(f\"chr{i+1} (length {ts.sequence_length} bp): diversity = {ts.diversity()}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a10a5ee-9965-450c-b73c-a72418b39f21",
   "metadata": {},
   "source": [
    "If separate chromosomes already exist, e.g. from running separate simulations in parallel, they can be pasted back together using  [`concatenate()`](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.concatenate).\n",
    "\n",
    "<div class=\"alert alert-block alert-info\"><b>Note:</b> because we removed the 1bp joining regions, the concatenated tree sequence below is of total length 3,000,000 bp, rather than 3,000,002 as in the original <code>joined_ts</code>.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4aae7e04-9ebb-4d7c-9cbc-ca348cf92cba",
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_ts = ts_list[0].concatenate(*ts_list[1:])\n",
    "\n",
    "# NB: in this case, because we didn't simplify the separate tree sequences, they should all share the same node table,\n",
    "# so we could preserve internal node identities in the concatenated tree sequence as follows:\n",
    "\n",
    "node_map = np.arange(ts_list[0].num_nodes)\n",
    "concat_ts = ts_list[0].concatenate(*ts_list[1:], node_mappings=[node_map, node_map])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "339424f5-b8d1-4ded-8194-fc1011cf6dfb",
   "metadata": {},
   "source": [
    "Having a single tree sequence is a simple way of performing analysis over all the chromosomes in the genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bf85577-d4d1-4942-8f5d-a8e632851db8",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_pca_by_pop(concat_ts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cf73640-2d47-493e-9f5b-a1fa93dc7d6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "windows = np.linspace(0, concat_ts.sequence_length, 65)\n",
    "windowed_diversity = concat_ts.diversity(windows=windows)\n",
    "\n",
    "# Now figure out some variables for plotting\n",
    "chrom_starts = np.concatenate(([0.0], np.cumsum(chrom_lengths)[:-1]))\n",
    "chrom_ends = chrom_starts + chrom_lengths\n",
    "\n",
    "# local coordinate within its chromosome\n",
    "region_idx = np.searchsorted(chrom_ends, windows, side='left')\n",
    "local_coords = windows - chrom_starts[region_idx]\n",
    "\n",
    "# choose which ticks to label \n",
    "label_every = 2  # (avoid overcrowding)\n",
    "label_indices = np.arange(0, len(windows), label_every)\n",
    "\n",
    "labels = [f\"{local_coords[i]/1e6:g}\" for i in label_indices]\n",
    "tick_positions = windows[label_indices]\n",
    "\n",
    "plt.figure(figsize=(16, 3))\n",
    "plt.stairs(windowed_diversity, windows, baseline=None)\n",
    "plt.xticks(ticks=tick_positions, labels=labels, fontsize=9)\n",
    "plt.xlabel(\"Local position (Mb)\")\n",
    "\n",
    "# --- background shading ---\n",
    "for i, (start, end) in enumerate(zip(chrom_starts, chrom_ends)):\n",
    "    plt.text((start+end)/2, max(windowed_diversity) / 5, f\"Chr{i+1}\", ha=\"center\")\n",
    "    plt.axvspan(\n",
    "        start, end,\n",
    "        color='gray',\n",
    "        alpha=0.15 if i % 2 == 0 else 0.05,   # alternate shade\n",
    "        zorder=0\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a136934-5939-4c3c-96df-82196c87c5b6",
   "metadata": {},
   "source": [
    "## Further information\n",
    "\n",
    "This notebook can be used interactively, to test the extensive capabilities of _msprime_. Both the _msprime_ and _SLiM_ simulators have extensive documentation, and helpful support communities. See https://tskit.dev/software/msprime.html and https://tskit.dev/software/SLiM.html."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad9aebd4-4ecf-4938-a487-aa5cfbd4ccd9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
