{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "19a48cc0-ae22-4840-a9dd-2a73b665f32a",
   "metadata": {},
   "source": [
    "<div style=\"color: white; background-color: #03303e; display: flex; align-items: flex-start; margin-bottom: 1em\">\n",
    "    <figure style=\"text-align: center; margin: 0 1em 0.5em 0.5em\">\n",
    "    <img src=\"tskit_logo.svg\" style=\"height: 3em;\" />\n",
    "    <figcaption style=\"font-size: 0.6em; line-height: 1em\">Interactive workbooks</figcaption>\n",
    "</figure>\n",
    "    <h1 style=\"margin: auto\">An interactive tour of <em>tskit</em></span></h1></div>\n",
    "This tour uses the <em>tskit</em> <a href=\"https://tskit.dev/tskit/docs/stable/python-api.html\">Python interface</a>. Press \"shift-enter\" (or click the toolbar \"run\" button, &#9654;) to run a notebook cell and move to the next one. Repeat this from the top of the workbook to work gradually through the demonstration code.\n",
    "<div class=\"alert alert-block alert-info\"><b>Note:</b> For static worked examples see our <a href=\"https://tskit.dev/tutorials/\">tutorials site</a>, which includes introductions for <a href=\"https://tskit.dev/tutorials/popgen.html\">population genetics</a> and <a <a href=\"https://tskit.dev/tutorials/phylogen.html\">phylogenetics</a>.</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0969a7e2-ac66-4d98-bf16-af2a5389bd06",
   "metadata": {},
   "source": [
    "## Loading an ARG in \"succinct tree sequence\" format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07cc7710-49f1-4e90-98a9-f0ed6c1f476f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tskit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4961282f",
   "metadata": {},
   "outputs": [],
   "source": [
    "ts = tskit.load(\"data/demo.trees\")  # By convention we use `ts` or `arg` for the object name"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "058be451-3ea5-4833-82ce-230d53390585",
   "metadata": {},
   "source": [
    "In a notebook, we can show a tabular summary of the tree sequence file by displaying it to screen:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a896cb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "ts  # Running this notebook cell should display a tabular summary of the `ts` object"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87010aa7-6f8e-49e2-8910-2a7f63d7e773",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\"><b>Note:</b> the <a href=\"https://tskit.dev/tskit/docs/stable/provenance.html\">provenances</a> listed above (in the last part of the output) show how this tree sequence was originally generated. In this case, close inspection shows it was initially simulated by the command <a href=\"https://tskit.dev/msprime/docs/stable/api.html#msprime.sim_ancestry\"><code>sim_ancestry()</code></a>, provided by <a href=\"https://tskit.dev/msprime/docs/stable/intro.html\"><i>msprime</i></a> version 1.4.0, then simplified, with mutations finally added using the <i>msprime</i> <a href=\"https://tskit.dev/msprime/docs/stable/api.html#msprime.sim_mutations\"><code>sim_mutations()</code></a> function.</div>\n",
    "\n",
    "Each [node](https://tskit.dev/tskit/docs/stable/data-model.html#node-table) in the tree sequence represents a (haploid) genome. *Sample* nodes are the (usually current-day) genomes whose DNA sequences are known. Here there are 80 sample nodes, belonging to 40 (diploid) individuals.\n",
    "\n",
    "As its name suggests, a tree sequence can be interpreted as a sequence of evolutionary (\"gene\") trees along a genome. This is done by linking the sample nodes to ancestral nodes via *edges*. The table above reveals that there are 6979 edges in this tree sequence, defining 1811 trees. Later in this workbook you'll see how to plot one of the local trees.\n",
    "\n",
    "## Mutations and genetic variation\n",
    "\n",
    "*Mutations* on an ARG define the DNA sequence of the samples. These cause sample nodes to differ from each other at a number of *variable sites*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4d4a32c-28aa-45c2-8a5d-dc8cfdb3d49d",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\n",
    "    f\"The mutations in this ARG define {ts.num_sites} sites,\",\n",
    "    f\"for {ts.num_samples} sample genomes,\",\n",
    "    f\"over a total length of {ts.sequence_length} base pairs\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ca91e59-2c3f-446b-8926-ec08a60ce716",
   "metadata": {},
   "outputs": [],
   "source": [
    "for v in ts.variants():\n",
    "    print(\n",
    "        \"For the first variable site, the allelic state for\",\n",
    "        f\"{ts.num_samples} genomes at position {v.site.position} is:\\n\",\n",
    "        v.states(),\n",
    "    )\n",
    "    print(f\"At variable site number # {v.site.id} (at position {v.site.position} basepairs along the chromosome)\")\n",
    "    for state, freq in v.frequencies().items():\n",
    "        print(f\"* the frequency of {state} is {freq}\")\n",
    "    break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c94268a-94b0-48d6-9321-05687cbc33f3",
   "metadata": {},
   "source": [
    "## Basic analysis \n",
    "\n",
    "Genetic analysis using tree sequences is usually incredibly fast, even for huge sample sizes. Here are a few examples of analysing the loaded _tskit_ ARG: feel free to change them and re-run the cells.\n",
    "\n",
    "Documentation for the methods is here: https://tskit.dev/tskit/docs/stable/python-api.html (e.g. see [here](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.allele_frequency_spectrum) for the allele frequency spectrum)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c518928-5a62-40d4-866d-0488616b4461",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the unpolarised allele frequency spectrum\n",
    "afs = ts.allele_frequency_spectrum(polarised=False)\n",
    "print(f\"AFS calculated for {ts.num_samples} genomes of length {ts.sequence_length/1e6:.2f} Mb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1accb0d-cf44-430d-8ae5-534800e72663",
   "metadata": {},
   "outputs": [],
   "source": [
    "# and plot it\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "plt.bar(range(ts.num_samples + 1), afs, color=\"grey\")\n",
    "plt.xlabel(\"Count (frequency) in sample\")\n",
    "plt.title(\"Unpolarised allele frequency spectrum\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47dbd599-86f1-4693-88dc-9c8fd457168b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate and plot genetic diversity in windows along the genome\n",
    "genome_windows = list(range(0, int(ts.sequence_length+1), 20_000))  # or `np.linspace(0, ts.sequence_length, 51)`\n",
    "windowed_diversity = ts.diversity(windows=genome_windows)\n",
    "plt.stairs(windowed_diversity, genome_windows)\n",
    "plt.ylabel(\"Genetic diversity (Ï€)\")\n",
    "plt.xlabel(\"Genome position (bp)\")\n",
    "plt.title(\"Diversity along the genome\");\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc775623-ab88-4cb0-b01e-f1cf440e2fbe",
   "metadata": {},
   "source": [
    "## IDs and underlying data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf85d30e-78ef-47a9-b311-6e119ca70c31",
   "metadata": {},
   "source": [
    "At its heart, a tree sequence is just a collection of tables. For example, genomes are represented by **nodes** in the *nodes table*. Each is referred to by an ID, which is simply its row number in the table. Note that all IDs start with 0, not 1. Here, for instance is the first node (ID: 0) in the nodes table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "580fb9e1-e87b-45e4-9b32-dc4133579c4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "ts.node(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3611b3ae-0f14-4802-a4e4-97565fd86693",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\"><b>Note:</b> The fact that the flag is an odd number indicates that node 0 is a <i>sample node</i>. Nodes that are not samples usually represent ancestral genomes.</div>\n",
    "\n",
    "Here's another table (the _populations_ table, usually much smaller than the nodes table). You can see that simulation defines 7 different populations, with IDs 0 to 6. Above, node 0 was shown as belonging to population 0 (i.e. the population named \"A\" below)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37d9d7de-bfd3-49d7-8f1d-363622da2405",
   "metadata": {},
   "outputs": [],
   "source": [
    "ts.tables.populations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "331f0ada-7af9-4802-aca7-23631ee386db",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\"><b>Note:</b> It is not necessary to define or use populations: they are a <i>tskit</i> convenience for grouping multiple nodes together. In other words, the populations table can be left empty, in which case each node should refererence a \"population\" with ID <a href=\"https://tskit.dev/tskit/docs/stable/python-api.html#tskit.NULL\"><code>tskit.NULL</code> (-1)</a>. Similarly although sample nodes are commonly paired into a diploid <a href=\"https://tskit.dev/tskit/docs/stable/data-model.html#individual-table\">individuals</a>, this is not a strict requirement for a <a href=\"https://tskit.dev/tskit/docs/stable/data-model.html#valid-tree-sequence-requirements\">valid tree sequence</a>.</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52b473a2-656c-4c4c-a826-d5684f1fbf41",
   "metadata": {},
   "source": [
    "## More complex analysis \n",
    "\n",
    "_Tskit_ is designed to allow users to write their own analysis tools, but also has several sophisticated built-in analysis methods.\n",
    "\n",
    "### Principle components analysis\n",
    "\n",
    "Here is an example of efficient principle components analysis (PCA). Each point is one of the 80 genomes, but the _tskit_ implementation scales to millions of genomes. Here we using `mode=\"branch\"`, the default for PCA analysis, which means that we don't even need mutations (realised genetic variation) to look at genetic distances (see [this tutorial](https://tskit.dev/tutorials/no_mutations.html))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "981b617f-8de5-4766-81ac-cac0f22a371c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the first 2 principle components for every haploid genome\n",
    "# For a more conventional PCA of diploid individuals, use the `individuals` argument)\n",
    "\n",
    "pca_output = ts.pca(2, mode=\"branch\", random_seed=42)  # or add `individuals=range(ts.num_individuals)`\n",
    "\n",
    "# Plot the PCA\n",
    "for pop in ts.populations():\n",
    "    use = ts.samples(population=pop.id)  # `use = ts.individual_populations == pop.id` if using individuals\n",
    "    if use.any():\n",
    "        plt.scatter(*pca_output.factors[use, :].T, label=f'Population {pop.metadata[\"name\"]}')\n",
    "plt.xlabel(\"PC 1\")\n",
    "plt.ylabel(\"PC 2\")\n",
    "plt.legend();"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e431618b-68be-4142-8af5-c0cc5b5c6593",
   "metadata": {},
   "source": [
    "### Coalescence rates\n",
    "\n",
    "Built-in tools exist to calculate coalescence and cross-coalescence rates. Below you can see that there is no coalescence between populations A and B more recently than 250 generations ago, suggesting populations A and B split around this time, whereas C and D probably split more recently, about 200 generations ago. This could partially explain why points from C and D partially overlapping on the PCA plot above.\n",
    "\n",
    "Note that the \"inverse instantaneous coalescence rate\" (IICR) is sometimes known as the effective population size or $N_e$, which is often used in population genetics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1ee91c9-3eba-41dc-9a74-03b202405c86",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "num_timebins = 20 \n",
    "time_breaks = np.concatenate(( # NB: first bin must start at sample time & last at infinity\n",
    "    [0], np.logspace(np.log10(35), np.log10(ts.max_time/2), num_timebins - 2), [np.inf]\n",
    "))\n",
    "\n",
    "all_rates = ts.pair_coalescence_rates(time_windows=time_breaks)\n",
    "plt.stairs(all_rates[1:-1], time_breaks[1:-1], label=\"All coalescences\")\n",
    "\n",
    "# Now calculate some cross coalescences\n",
    "samples = {pop.metadata[\"name\"]: ts.samples(pop.id) for pop in ts.populations()}\n",
    "cross_AB_rates = ts.pair_coalescence_rates(sample_sets=[samples[\"A\"], samples[\"B\"]], time_windows=time_breaks)\n",
    "cross_CD_rates = ts.pair_coalescence_rates(sample_sets=[samples[\"C\"], samples[\"D\"]], time_windows=time_breaks)\n",
    "plt.stairs(cross_AB_rates[1:-1], time_breaks[1:-1], label=\"A/B cross coalescences\")\n",
    "plt.stairs(cross_CD_rates[1:-1], time_breaks[1:-1], label=\"C/D cross coalescences\")\n",
    "\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")\n",
    "plt.xlabel(f\"Time ({ts.time_units})\")\n",
    "plt.ylabel(f\"Genome-wide instantaneous coalescence rate\")\n",
    "#plt.legend();"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93a395f2-2fdb-4f2d-b9a0-a763510b79dc",
   "metadata": {},
   "source": [
    "### Windowing in time and space\n",
    "\n",
    "Like all other tskit statistics, we can window coalescence rates along the genome. Below you can see a burst of coalescence (red) at about 170 generations ago around 0.3Mb along the genome, with an absence of coalescence information (grey) within that population at more distant times for that genomic region. This is indicative of a selective sweep at that genomic location (particularly clear because of the cleanliness of simulated data)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "deaa639c-b4d6-460a-a1c0-c8416eb38d50",
   "metadata": {},
   "outputs": [],
   "source": [
    "time_breaks = np.concatenate(( # NB: first bin must start at sample time & last at infinity\n",
    "    [0], np.logspace(np.log10(35), np.log10(ts.max_time/10), num_timebins - 2), [np.inf]\n",
    "))\n",
    "\n",
    "popA_rates = ts.pair_coalescence_rates(time_windows=time_breaks, windows=genome_windows, sample_sets=[samples[\"A\"]])\n",
    "colourmap = plt.get_cmap('jet')\n",
    "colourmap.set_bad(color='grey')\n",
    "im = plt.pcolormesh(genome_windows, time_breaks[1:-1], popA_rates[:,1:-1].T, cmap=colourmap)\n",
    "bar = plt.colorbar(im)\n",
    "bar.ax.set_ylabel('pairwise coalescent density', labelpad=10, rotation=270)\n",
    "plt.yscale(\"log\")\n",
    "plt.xlabel(f\"Genome position\")\n",
    "plt.text(0, 170, \"170\\ngens\", ha=\"right\", va=\"center\", c=\"darkred\")\n",
    "plt.ylabel(f\"Time ({ts.time_units})\")\n",
    "plt.title(\"Local coalescence densities for population A\");"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "657053da-0846-4eb4-ae4b-d7eeee2d8d2f",
   "metadata": {},
   "source": [
    "## Information about a tree sequence's origin (\"provenance\")\n",
    "\n",
    "For reproducability, the [provenance table](https://tskit.dev/tskit/docs/stable/provenance.html) stores information about how a tree sequence was generated. The amount of detail is down to the tool(s) used to make the tree sequence. In this case, _msprime_ was used, which stores extensive provenance information including the simulated [demographic model](https://tskit.dev/msprime/docs/stable/demography.html). This can be plotted using the [demesdraw](https://grahamgower.github.io/demesdraw) software."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "021117aa-a70b-4bcd-a955-baa543a6465b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "if sys.platform == 'emscripten':  # only needed for jupyterlite to load demesdraw\n",
    "    import micropip\n",
    "    await micropip.install(\"demesdraw\")\n",
    "\n",
    "import demesdraw\n",
    "import msprime\n",
    "\n",
    "first_provenance_entry = ts.provenance(0)\n",
    "cmd, parameters = msprime.provenance.parse_provenance(first_provenance_entry, ts)\n",
    "assert cmd == \"sim_ancestry\"  # just check we have the right (zeroth) provenance entry\n",
    "\n",
    "msprime_demography_object = parameters[\"demography\"]\n",
    "demesdraw.tubes(\n",
    "    msprime_demography_object.to_demes(),\n",
    "    colours={\"A\": \"tab:blue\", \"B\": \"tab:orange\", \"C\": \"tab:green\", \"D\": \"tab:red\"},\n",
    "    log_time=False,\n",
    ")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2e63d6c-6881-4860-b226-4ff7dd9dd94f",
   "metadata": {},
   "source": [
    "The demography explains the pattern that was seen in the PCA and the cross-coalescence rates, where populations C and D split more recently than A and B."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6de5e1d0-2e10-4a71-a1ae-6c90bed6c103",
   "metadata": {},
   "source": [
    "## Tree plotting\n",
    "\n",
    "The tree sequence allows fast extraction of ancestral trees along the genome, which can be easily plotted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "904b74ab-8fd2-4d52-8e18-5371e4ead31a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "first_tree = ts.first()\n",
    "first_tree.draw_svg(size=(1000, 250), title=\"Tree at genome position 0\", node_labels={})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccd74ad8-76dc-4836-b3e0-dd3669cc78ba",
   "metadata": {},
   "source": [
    "_Tskit_ tree plots can be decorated using a variety of styles (see the [viz tutorial](https://tskit.dev/tutorials/viz.html)). Here we plot the tree at position 0.3 Mb, to see if population A really does have a cluster of coalescences around 170 generations ago (at the dotted magenta line: it does!)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b81ab065-37bd-41f7-9ed5-2b9c34d4e54a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from matplotlib import colors\n",
    "styles = [f\".leaf.p{p.id}>.sym\" + \"{fill:\" + colors.to_hex(f\"C{p.id}\") + \"}\" for p in ts.populations()]\n",
    "\n",
    "# Making a legend can be done, but involves some hand-positioning of elements\n",
    "legend = '<rect width=\"100\" height=\"100\" x=\"100\" y=\"10\" fill=\"#EEE\" /><text x=\"140\" y=\"25\">Key</text>'\n",
    "legend += \"\".join([  # The legend lines, one for each population.\n",
    "    f'<g transform=\"translate(103, {40 + 18*p.id})\" class=\"leaf p{p.id}\">'  # an SVG group\n",
    "    f'<rect width=\"6\" height=\"6\" class=\"sym\" />'  # Square symbol\n",
    "    f'<text x=\"10\" y=\"7\">Population {p.metadata[\"name\"]}</text></g>'  # Label\n",
    "    for p in ts.populations() if len(ts.samples(p.id)) > 0\n",
    "])\n",
    "\n",
    "ts.at(0.3e6).draw_svg(\n",
    "    size=(1000, 250),\n",
    "    node_labels={},    # Remove all node labels for a clearer viz\n",
    "    mutation_labels={},\n",
    "    style=\"\".join(styles) + \".y-axis .tick:nth-child(2) .grid {stroke: magenta; stroke-dasharray: 4}\",\n",
    "    preamble=legend,\n",
    "    y_axis=True,\n",
    "    y_gridlines=True,\n",
    "    y_ticks=[0, 170, 500, 1000, 1500],\n",
    "    title=\"Tree at genome position 300,000 bp\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39364361-7095-4029-8cc0-f4b825fd307f",
   "metadata": {},
   "source": [
    "## Other resources and analyses\n",
    "This notebook is fully interactive. Feel free to change the analyses above or perform more analysis below. \n",
    "\n",
    "Extensive static tutorial material is available at https://tskit.dev/tutorials/, and there are also other notebooks in this JupyterLite instance: View &rarr; File Browser will reveal them on the right-hand toolbar if they are not shown."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0460fee3-4520-4827-9d63-fafdc491a78b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use this as a playground for trying out tskit\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3323c1ea-62f4-4bf8-a228-6fdc1684d914",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
