{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9baa2d4a-8100-462a-942f-e1f09a0ed419",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "if sys.platform != 'emscripten':\n",
    "    # allow the notebook to be downloaded and run locally too\n",
    "    print(\"This notebook is not running in JupyterLite: you may need to install tskit, tszip, etc.\")\n",
    "else:\n",
    "    import micropip\n",
    "    await micropip.install('tszip')\n",
    "    await micropip.install('pyslim==1.0.4')  # Need an older version for SLiM < 5\n",
    "    await micropip.install('stdpopsim')\n",
    "    await micropip.install('demesdraw')\n",
    "    await micropip.install('jupyterquiz')\n",
    "    await micropip.install('tskit_arg_visualizer')\n",
    "\n",
    "from jupyterquiz import display_quiz\n",
    "from matplotlib import pyplot as plt\n",
    "from WhatIsAnARG_module import Workbook2\n",
    "\n",
    "Workbook2.setup()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16853d10-0fc6-45e5-9ca5-0744a091c38a",
   "metadata": {},
   "source": [
    "### Overview\n",
    "This is the first of 2 workbooks: each is intended to take about 1½-2 hours to complete.\n",
    "\n",
    "<dl>\n",
    "<dt>Workbook 1</dt>\n",
    "    <dd>Graphs and local trees, mutations, genetic variation and statistical calculations: exercises A-J & questions 1-11</dd>\n",
    "<dt>Workbook 2</dt>\n",
    "    <dd>Types of ARG, simplification, and ARG simulation: exercises K-U & questions 12-30</dd>\n",
    "</dl>\n",
    "\n",
    "# Workbook 2: ARG types and simulation\n",
    "\n",
    "### Recap\n",
    "\n",
    "ARGs can be saved in tree sequence format, as a collection of **nodes** and **edges**. This format can also contain **individuals** (simply used to group nodes together), and if genetic variation is to be encoded, **mutations** associated with **sites**.\n",
    "\n",
    "Such ARGs can be analysed using the _tskit_ library, and drawn tree-by-tree using the built-in `.draw_svg()` method, or as a graph by using the [tskit_arg_visualizer](https://github.com/kitchensjn/tskit_arg_visualizer/blob/main/docs/tutorial.md) package, based on the D3js interactive visualization library. We have been using a so-called *full ARG* stored in the `example_ARG.trees` file.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58e554da-ab32-4d3e-b7a8-93da8d83bd31",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tskit\n",
    "import tskit_arg_visualizer as argviz\n",
    "from msprime import NODE_IS_RE_EVENT\n",
    "from tskit import NODE_IS_SAMPLE\n",
    "\n",
    "arg = tskit.load(\"data/example_ARG.trees\")\n",
    "d3arg = argviz.D3ARG.from_ts(arg)\n",
    "d3arg.set_node_styles({u: {\"symbol\": \"d3.symbolSquare\"} for u in arg.samples()})\n",
    "d3arg.set_node_styles({\n",
    "    u: {\"fill\": \"red\" if flags & NODE_IS_RE_EVENT else \"grey\" if flags & NODE_IS_SAMPLE else \"blue\"}\n",
    "    for u, flags in enumerate(arg.nodes_flags)\n",
    "})\n",
    "d3arg.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c72a8430-954f-408b-874f-d3763eb0da1b",
   "metadata": {},
   "source": [
    "## Full ARGs: perfect historical knowledge\n",
    "\n",
    "The full ARG above shows \"perfect knowledge\". Specifically, it records the exact times of two sorts of historical event:\n",
    "* <span style=\"color:blue\">Common ancester (CA) events</span>, caused by two genomes duplicating during a round of mitosis. These are represented by a node with two children.\n",
    "* <span style=\"color:red\">Recombination (RE) events</span>, caused by two genomes combining during a round of meiosis. These are represented by a node with two parents (as we saw, the _msprime_ simulator actually encodes these as a pair of nodes, but we visualise this pair as a single node).\n",
    "\n",
    "However, requiring perfect knowledge can be too restrictive, especially when representing ARGs which have been inferred, imperfectly, from real data. In the next section we explore ARG representations which omit unknowable or undetectable nodes and edges. These *simplified ARGs* retain essentially the same genetic ancestry as the equivalent full ARGs, but can be substantially smaller. This means they are easier to visualise, analyse, infer, and simulate.\n",
    "\n",
    "### Unknowable nodes\n",
    "\n",
    "The easiest example of an unknowable structure in a full ARG is a diamond. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bf18db7-30f3-4f83-adfd-a21770e909c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook2.url + \"Q12.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc7ff115-2c52-46a1-bfb5-710b850c8025",
   "metadata": {},
   "source": [
    "We previously saw that a recombination event within a diamond does not cause a change in topology or branch lengths of the local trees. In fact, nowhere on the ARG can we place a mutation such that this diamond is detectable, because none of the nodes in a diamond help to group samples together. More specifically, none of the diamond nodes are *coalescent* in any of the local trees.\n",
    "\n",
    "As a rule, we can treat the times of all non-coalescent nodes as unknowable. This includes all RE nodes, but also some CA nodes such at the top of diamonds. To find these in our ARG, we can use the `Workbook.node_coalescence_status()` function from the first workbook. We will wrap this in a function called `create_styled_D3ARG()` that creates and colours the ARG as appropriate (you don't need to understand the workings of this function, however).\n",
    "\n",
    "<div class=\"alert alert-block alert-info\"><b>Note:</b> Later on we shall start removing nodes from the ARG, which could change the node IDs. For this reason we will switch to labelling the nodes using metadata. Rather unimaginatively, in their metadata, the sample nodes have assigned the names A-J, which will be maintained even if their IDs change. In real data, metadata labels should be more meaningful, e.g. <code>{'name':'HG00157', 'sex':'Male'}</code> for an individual human from the 1000 Genomes Project.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c2bf5fb-86e0-4ff8-8ae9-7d4c56448c91",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import os\n",
    "from pathlib import Path\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "def metadata_based_labels(arg, use_ids=True):\n",
    "    if use_ids:\n",
    "        return {nd.id: (nd.metadata or {}).get(\"label\", nd.id) for nd in arg.nodes()}\n",
    "    else:\n",
    "        # only make keys in the dictionary if there is a \"label\"\n",
    "        return {nd.id: nd.metadata[\"label\"] for nd in arg.nodes() if nd.metadata and \"label\" in nd.metadata}\n",
    "        \n",
    "\n",
    "def create_styled_D3ARG(arg, x_pos_file=None, node_map=None):\n",
    "    \"\"\"\n",
    "    You don't need to understand this function!\n",
    "\n",
    "    Take a tskit ARG and return a D3ARG, with coloured nodes and with sample nodes\n",
    "    labelled with node.metadata[\"label\"]. If an x_pos_file is given, load X positions\n",
    "    from this file, translating the node IDs via the node_map if provided\n",
    "    \"\"\"\n",
    "    d3arg = argviz.D3ARG.from_ts(arg)\n",
    "    status = np.array(['cyan', 'blue', 'green'])[Workbook2.node_coalescence_status(arg)]\n",
    "    status[arg.samples()] = 'gray'\n",
    "    status[(arg.nodes_flags & NODE_IS_RE_EVENT) != 0] = 'red'\n",
    "    d3arg.set_node_styles({u: {\"fill\": str(colour)} for u, colour in enumerate(status)})\n",
    "    d3arg.set_node_styles({u: {\"symbol\": \"d3.symbolSquare\"} for u in arg.samples()})\n",
    "    d3arg.set_node_labels(metadata_based_labels(arg, use_ids=False))\n",
    "\n",
    "    if x_pos_file is not None:\n",
    "        x_pos = argviz.extract_x_positions_from_json(json.loads(Path(x_pos_file).read_text()))\n",
    "        if node_map is None:\n",
    "            d3arg.set_node_x_positions(x_pos)\n",
    "        else:\n",
    "            d3arg.set_node_x_positions({node_map[u]: x for u, x in x_pos.items() if node_map[u] != tskit.NULL})\n",
    "\n",
    "    return d3arg\n",
    "\n",
    "d3arg = create_styled_D3ARG(arg, x_pos_file=\"data/Xpos.json\")\n",
    "d3arg.draw(height=500)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8e30454-bf7c-496d-a25b-b5a8a01f1ff3",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook2.url + \"Q13.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c36dce03-ee1d-4d31-92a4-2f8973f9bf57",
   "metadata": {},
   "source": [
    "## Simplified ARGs: inheritance between genomes\n",
    "\n",
    "It is possible to construct ARGs without unknowable nodes. Instead of being associated with explicit events, nodes in such an ARG are better imagined as *genomes*, between which genetic information is inherited. These *simplified* ARGs can be produced directly from simulation or inferred from real data. Here, however, we will explore how they relate to full ARGs, via the process of *simplification*.\n",
    "\n",
    "### Simplification\n",
    "\n",
    "Simplification is the act of removing redundant information from an ARG. The _tskit_ library provides a flexible [`.simplify()`](https://tskit.dev/tutorials/simplification.html) method that performs this task. The core idea is to retain only the ancestry of a specified set of \"focal nodes\". Nodes that are not needed to represent this ancestry are removed.\n",
    "\n",
    "\n",
    "#### Partial simplification\n",
    "\n",
    "The code below carries out a \"partial simplification\" of the ARG, detaching all unknowable (non-coalescent) nodes from the ancestry (although we use `filter_nodes=False` to keep them in the stored object).\n",
    "\n",
    "Hover over the genome bar to check that a complete ancestry of the samples is still present, even though various intermediate nodes have been removed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "748f8f4e-385b-4122-b6f1-95de6978aed7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def simplify_non_coalescent_nodes(arg, **kwargs):\n",
    "    # NB: this function can probably be removed when https://github.com/tskit-dev/tskit/issues/2127 is fixed\n",
    "    partial_or_always_coalescent_nodes = np.where(Workbook2.node_coalescence_status(arg) > 0)[0]\n",
    "    focal_nodes = np.concatenate((arg.samples(), partial_or_always_coalescent_nodes))\n",
    "    return arg.simplify(\n",
    "        focal_nodes,  # The set of nodes whose ancestry we want to keep\n",
    "        update_sample_flags=False,  # Usually all focal nodes are turned into samples, but we don't want that here\n",
    "        **kwargs,  # pass on any additional parameters (e.g. \"filter_nodes\") to `simplify`\n",
    "    )\n",
    "\n",
    "part_simp = simplify_non_coalescent_nodes(arg, filter_nodes=False)\n",
    "\n",
    "create_styled_D3ARG(\n",
    "    part_simp,\n",
    "    x_pos_file=\"data/Xpos.json\"\n",
    ").draw(height=400)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a18cc2a-7d19-4aef-bfe5-8fa1b894e97c",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook2.url + \"Q14.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a116dc8-4ba6-45f3-832a-dfa30cac3420",
   "metadata": {},
   "source": [
    "By specifying `filter_nodes=True` (which is the default) we can actually delete the unused nodes completely. However, this will change the node IDs. So that we can plot the nodes in the right place, we'll also return the `node_map` of old to new IDs when simplifying."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b0b800a-9d3d-4417-ba51-268880023af6",
   "metadata": {},
   "outputs": [],
   "source": [
    "part_simplified_arg, node_map = simplify_non_coalescent_nodes(arg, filter_nodes=True, map_nodes=True)\n",
    "create_styled_D3ARG(\n",
    "    part_simplified_arg,\n",
    "    x_pos_file=\"data/Xpos.json\",\n",
    "    node_map=node_map\n",
    ").draw(height=400)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84fd3cdf-50c1-471a-81bd-031d18b0560a",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook2.url + \"Q15.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "873482c4-8e5b-4c00-a120-30c3ceb239c5",
   "metadata": {},
   "source": [
    "Here's what the partially simplified trees look like. As you can see, they still look slightly unusual compared to a standard \"gene tree\", as there are a few locally unary nodes left. These nodes are coalescent in some of the local trees, but not in others. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c25cff93-197d-4cf3-ad33-1c304ff88432",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook2.url + \"Q16.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eee18b2e-b0e5-468c-b721-89a45d915682",
   "metadata": {},
   "source": [
    "<dl class=\"exercise\"><dt>Exercise K</dt>\n",
    "    <dd>Use <code>part_simplified_arg.draw_svg()</code> to plot the local trees.\n",
    "        If you change the height of the plot using <code>size=(1000, 400)</code> the tree will be clearer.\n",
    "        You might also want to specify <code>node_labels=metadata_based_labels(part_simplified_arg)</code> to get nicer labels.\n",
    "    </dd>\n",
    "</dl>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad832b00-daa8-4161-be65-91170c74eb51",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Complete Exercise K here\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc4f4115-f505-4cfe-8232-fa2a15e678af",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook2.url + \"Q17.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43cdacbf-499f-4e80-b478-9090854b1e38",
   "metadata": {},
   "source": [
    "## Full simplification\n",
    "\n",
    "We can go one stage further by fully simplifying so that ancestral nodes are all-coalescent. This modifies edges so that, for example, the edge above node 18 in the first tree goes straight to node 20, rather than going via node 19. This will not change the general topology or branch lengths of the trees, and it also makes the trees simpler. However, it can create more edges (making ARG analysis less efficient), and slightly paradoxically, makes the graph visualization *less* simple:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75144e43-5c45-4dbb-bd69-b1c7050faf02",
   "metadata": {},
   "outputs": [],
   "source": [
    "fully_simplified_arg, node_map = arg.simplify(map_nodes=True)\n",
    "\n",
    "display(fully_simplified_arg.draw_svg(\n",
    "    size=(1000, 400),\n",
    "    node_labels=metadata_based_labels(fully_simplified_arg),\n",
    "    title=\"Fully simplified local trees\",\n",
    "))\n",
    "\n",
    "create_styled_D3ARG(\n",
    "    fully_simplified_arg,\n",
    "    \"data/Xpos.json\",\n",
    "    node_map).draw(height=500, title=\"Fully simplified ARG\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27e9c261-00c8-433a-95d2-fca29418b9a4",
   "metadata": {},
   "source": [
    "As you can see above, the local trees look simpler, but the graph looks more tangled (although every node is now <b style=\"color:green\">all-coalescent</b>). The fully-simplified ARG encodes exactly the same genetic information on the samples, and the same branch lengths in the local trees, but is less genealogically-informative about how adjacent trees relate to each other."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "209271ab-29fd-40ec-9665-d49957663edf",
   "metadata": {},
   "source": [
    "<dl class=\"exercise\"><dt>Exercise L</dt>\n",
    "    <dd>use the <code>.num_edges</code>, <code>.num_nodes</code>, <code>.num_trees</code>, and <code>.num_mutations</code> attributes to print out the number of edges, nodes, trees, and mutations in the <code>part_simplified_arg</code> ARG and the <code>fully_simplified_arg</code> ARG (or you could simply <code>display</code> the summary table of each ARG).</dd>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7697e5e7-e844-4081-ac32-eb46f23fe535",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Complete Exercise L here\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc3f545b-0403-4e13-a76e-13b4ea9467c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook2.url + \"Q18.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81d75407-379b-49e3-9946-4dd7edede763",
   "metadata": {},
   "source": [
    "Because the number of sites and mutations has remained the same, and there is still a complete ancestry for all the samples, the encoded genetic variation and even branch-length calculations are the same between the full ARG and any of the simplified versions. This is easy to demonstrate:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c58479ed-1257-41ed-8567-f47a785d1414",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"    ====== Original (branch π: {arg.diversity(mode='branch'):.2f}) ========    \",\n",
    "      f\"    === Fully simplified (branch π: {fully_simplified_arg.diversity(mode='branch'):.2f}) ===\")\n",
    "for v1, v2 in zip(arg.variants(), fully_simplified_arg.variants()):\n",
    "    print(\n",
    "        f\"pos {int(v1.site.position):>3}\", v1.states(), \"   \",\n",
    "        f\"pos {int(v2.site.position):>3}\", v2.states()\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fa60081-b71d-44ed-b256-6691b4ac20b5",
   "metadata": {},
   "source": [
    "For historical reasons, the fully simplified format is the default output of the _msprime_ and _SLiM_ simulators, even though in some cases it can be slightly less efficient."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9e6b0f9-9314-4c64-8983-4e1397ef4076",
   "metadata": {},
   "source": [
    "### Nodes represent genomes\n",
    "\n",
    "Philosophically, the nodes in a simplified ARG no longer represent historical events, but the genomes that are produced as an *outcome* of those (unknown) events. For example, sample node 6 now has 2 parents, even though it does not itself represent a recombination event. We know a recombination event occurred some time between the time of node 6 and the time of its youngest parent, node 15, but we can't say exactly when."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc6144fb-55ef-4505-8acb-fe18716f4d16",
   "metadata": {},
   "source": [
    "### Simplifying with different focal nodes\n",
    "\n",
    "One of the main uses of `simplify()` is to change which nodes are treated as samples. By default the existing sample nodes are taken, but we can specify a subset of the existing samples to make a smaller ARG, which is the topic of the next subsection.\n",
    "\n",
    "However, for demonstration purposes, the code below (unusually) passes a previously non-sample node to `simplify()`. This is a shortcut that allows us to output the assumed genome of internal nodes in the ancestry: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a059faf0-d633-4b07-83f4-07a017200fa0",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_ids_plus_node_21 = np.append(part_simplified_arg.samples(), 21)  # make a list of sample nodes and add node 21\n",
    "new_arg = part_simplified_arg.simplify(sample_ids_plus_node_21)  # pretend that we have \"sampled\" node 21\n",
    "print(\"Known variation for ancestral node 21\")\n",
    "for v in new_arg.variants():\n",
    "    print(f\"pos {int(v.site.position):>3}:   \", v.states(missing_data_string=\"-\")[new_arg.num_samples - 1])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fac20aa-c9d7-42b2-8459-3c3b6cabe2a1",
   "metadata": {},
   "source": [
    "You can see that the start and end of the genome of node 21 is unknown (its haplotype is truncated)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13ec892f-5fae-4ef4-aedc-01d55c554dff",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook2.url + \"Q19.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d5a5593-340a-4822-960a-ecff62e3ce08",
   "metadata": {},
   "source": [
    "#### Simplifying to reduce the sample size\n",
    "\n",
    "If we choose a subset of the existing samples, we can reduce the ARG to reflect the ancestry of only those samples. Repeated simplification to the current-day samples is a key component of forward simulation."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a7038c8-2cc4-4a3e-9c98-b20b97aa1b70",
   "metadata": {},
   "source": [
    "<dl class=\"exercise\"><dt>Exercise M</dt>\n",
    "    <dd>Simplify the original ARG so that it only shows the ARG relating sample 0 (A)  to sample 6 (F). By default <code>simplify()</code> removes any sites that no longer have any mutations (are \"monomorphic\"). Set <code>filter_sites=False</code> to leave those sites in the resulting tree sequence. Sites with no mutations will have a tickmark on the x-axis without an associated red v-shaped mutation mark.</dd></dl>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aae265ad-0688-4252-aaf1-74b9adb6ed8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Carry out exercise M here\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1155316-65f8-44b9-9e9c-0aa1c1d80c0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook2.url + \"Q20.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef7106e2-4eee-4056-b2a7-225fc7674fd7",
   "metadata": {},
   "source": [
    "### The simplest possible ARG\n",
    "\n",
    "We can even go to the extreme, and simplify an ARG to only 2 genomes (say 0 and 9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b8e41a5-8014-4457-a324-866facb83fb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "arg.simplify([0,9], filter_sites=False).draw_svg()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19ce95d4-9b66-4833-9d76-576b1d8d8710",
   "metadata": {},
   "source": [
    "Although there are only 2 samples (and 3 trees) in this ARG, and the topology of the local trees is always a \"cherry\", there is still some ARG structure, because the time to the MRCA differs as we go along the genome. The ability to infer a simple ARG of this sort can still be extremely powerful: this is the basis behind techniques based on the Pairwise Sequentially Markovian Coalescent (PSMC) (see e.g. [this 2019 review](https://doi.org/10.1002/ece3.5888) or more recent papers by [Schweiger (2023)](https://doi.org/10.1101/gr.277665.123) and [Terhorst (2025)](https://doi.org/10.1038/s41588-025-02323-x))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab50db90-3023-4d76-8258-eeaab1566bcc",
   "metadata": {},
   "source": [
    "## Polytomies and undetectable nodes\n",
    "\n",
    "It is possible to create ARGs in *tskit* that contain **polytomies** (nodes in a local tree with more than 2 children). This can happen under some models of evolution (e.g. the Dirac coalescent, where there is a burst of coalescences at a single timepoint). Such ARGs are also produced by some ARG inference methods (e.g. *tsinfer*) in which nodes that are not associated with a mutation are omitted.\n",
    "\n",
    "We can demonstrate the second sort of ARG by removing \"undetectable\" nodes, often those without associated mutations. This process collapses edges in a way that appears simular to parts of the simplification process:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89bdaed2-69bf-4a0d-9d5b-5ae5a9835026",
   "metadata": {},
   "outputs": [],
   "source": [
    "Workbook2.remove_unsupported_edges(arg).draw_svg(\n",
    "    title=\"ARG, collapsing edges unsupported by mutations\",\n",
    "    size=(1000, 400)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5133fe7a-2bc3-4e48-9c4c-f4b5e34f4a98",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook2.url + \"Q21.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1d6165c-21d9-4cb4-bd13-196cf4de605d",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\"><b>Note:</b> Unlike simplification, creating polytomies will change the branch lengths in the local trees, and hence affect branch-length measures of genetic variation. More generally, it needs special treatment when measuring features of ARGs, or performing statistical tests, which can make such ARGs hard to handle.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb1c109d-34eb-4c46-978c-d8101b1e7d67",
   "metadata": {},
   "source": [
    "## Simulating ARGs\n",
    "\n",
    "There are two basic approaches for simulating ancestry: forward and backward in time. Backward-time simulators such as _msprime_ are usually much faster but less flexible (for instance, it is hard to simulate anything but the simplest form of natural selection). Forward-time simulators such as _SLiM_ are more general, but must track the entire population during simulation, rather than simply following the ancestry of a small set of sample genomes.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6780edff-07bf-454d-b84d-583a81956d67",
   "metadata": {},
   "source": [
    "<dl class=\"exercise\"><dt>Exercise N</dt>\n",
    "    <dd>Tree sequence files often have \"provenance\" data,\n",
    "    which decribes how the file was made. You can see this when you display a\n",
    "tree sequence to the screen in a notebook: the list of provenances appears the bottom of the output.\n",
    "Use <code>display(arg)</code> (or simply <code>arg</code>) to show the ARG summary in the cell below.</dd>\n",
    "</dl>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "698273b7-dcd5-49b3-85cb-7267127d3540",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Carry out exercise N, displaying a summary of the ARG, here\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90cba45d-cd10-4d8a-924b-d4ea93596ecb",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook2.url + \"Q22.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de9b5ff2-b945-4224-99a8-86fc3ef8968e",
   "metadata": {},
   "source": [
    "### _Msprime_: a backward-time simulator\n",
    "\n",
    "_Msprime_ is a fast and flexible backward-time simulator whose core functions are `sim_ancestry` and `sim_mutations`. The `sim_ancestry` command is run first and this creates the ARG structure.\n",
    "\n",
    "Running an _msprime_ simulation in Python is extremely easy. You simply need to call `sim_ancestry` with a number of (assumed diploid) samples. You'll usually want to provide a `sequence_length`, a `population_size` and a `recombination_rate` too:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8508618-ca14-4f4b-a1ca-deeb07fad2ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Name the ARG `ts` (for \"tree sequence\")\n",
    "import msprime\n",
    "\n",
    "ts = msprime.sim_ancestry(10, population_size=1e4, sequence_length=10_000, recombination_rate=2e-8)\n",
    "print(f\"Simulated a simplified ARG of {ts.num_samples} haploid genomes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a01af82a-62b6-40c6-87ce-19c3ccd80311",
   "metadata": {},
   "source": [
    "The command above generates a \"fully simplifed\" ARG of 10 diploid samples (20 haploids) over a 10kB genome, assuming a constant effective population size ($N_e$) of 10,000, and with a recombination rate of $2\\times10^{-8}$ crossover mutations per base pair per generation. As the resulting ARG is stored in tree sequence format, we often name the resulting simulation `ts`. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "174c034b-1350-4d6f-8fc3-e2f7c7582732",
   "metadata": {},
   "source": [
    "#### Demographic models\n",
    "\n",
    "Of course, most populations have not stayed at a constant size over time, and are often modelled as multiple subpopulations. Instead of providing a `population_size` parameter, you can provide _msprime_ with a **demographic model**. This is how we simulated the original full ARG (for the curious, the details are in the `sim_ancestry` provenance entry, under \"parameters\" -> \"demography\"). The following code reads the demographic model from the full ARG's provenance, converts it to the portable [demes](https://popsim-consortium.github.io/demes-spec-docs/main/introduction.html) format, and plots it using the elegant [_demesdraw_](https://grahamgower.github.io/demesdraw/latest/) software:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "107b6596-5083-4d52-a301-0198c833bbb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import demesdraw\n",
    "\n",
    "cmd, parameters = msprime.provenance.parse_provenance(arg.provenance(0), arg)\n",
    "assert cmd == \"sim_ancestry\"  # just check we have the right (zeroth) provenance entry\n",
    "\n",
    "msprime_demography_object = parameters[\"demography\"]\n",
    "demesdraw.tubes(msprime_demography_object.to_demes(), log_time=False);"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56a4d653-3382-45bd-b8a2-d5a56f5d6ff3",
   "metadata": {},
   "source": [
    "<dl class=\"exercise\"><dt>Exercise O</dt>\n",
    "    <dd>The widths on the X axis, indicating the two population sizes, is dominated by a rapid exponential increase in the past few hundred generations. Change <code>log_time</code> from <code>True</code> to <code>False</code> above, to zoom in on more recent times, and re-plot.</dd>\n",
    "</dl>\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55daca94-bafd-4e36-a90d-c06647a256ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook2.url + \"Q23.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "163dda08-6ac4-4187-877f-bcd2ad4dc9ac",
   "metadata": {},
   "source": [
    "### Running a larger simulation\n",
    "\n",
    "We will rerun under the same demography, but simulate twenty times more of the genome (20kb). As we did in the original simulation, we will used the `record_full_arg` option to simulate a full ARG, rather than the simplified ARG that _msprime_ produces by default. Because there are two populations, we will 5 samples from each (for a total of 10 diploid samples), and use the original simulation recombination rate of $1.15\\times10^{-8}$. We'll use a fixed `random_seed` to ensure that everyone gets the same results:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93abb800-3335-4692-94e5-bad0bc8cd6ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "\n",
    "n_diploids=10\n",
    "genome_length=20_000  # bp\n",
    "\n",
    "start_time = datetime.now()\n",
    "larger_arg = msprime.sim_ancestry(\n",
    "    samples={\"AFR\": n_diploids/2, \"EUR\": n_diploids/2},\n",
    "    recombination_rate=1.15e-08,\n",
    "    sequence_length=genome_length,\n",
    "    demography=msprime_demography_object,\n",
    "    record_full_arg=True,\n",
    "    random_seed=321\n",
    ")\n",
    "\n",
    "print(f\"Simulated a full ARG of {larger_arg.num_samples} haploid samples over \"\n",
    "      f\"{larger_arg.sequence_length/1e3}kb in {datetime.now()-start_time} seconds\")\n",
    "print(\n",
    "    f\"ARG takes up {larger_arg.nbytes * 1e-6:.2f} MB, with {larger_arg.num_nodes} nodes \"\n",
    "    f\"and {larger_arg.num_edges} edges, encoding {larger_arg.num_trees} trees\\n\")\n",
    "\n",
    "# Calculate the number of \"unknowable\" nodes\n",
    "non_coal = Workbook2.node_coalescence_status(larger_arg) == 0\n",
    "p = (sum(non_coal)-larger_arg.num_samples)/(len(non_coal) - larger_arg.num_samples)\n",
    "print(f'{p * 100:.2f}% of nodes in this full ARG are non-coalescent (\"unknowable\")')\n",
    "print(f'{np.sum(larger_arg.nodes_flags & msprime.NODE_IS_RE_EVENT > 0)/ larger_arg.num_nodes * 100:.2f}% of nodes are RE nodes')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45781815-cf97-4347-b89d-6ab593977d21",
   "metadata": {},
   "source": [
    "You can see that this slightly larger ARG now has quite a high proportion of unknowable nodes.\n",
    "Plotting this out as a complete ARG is possible, but maybe not very helpful (feel free to change the `edge_type` below to `\"ortho\"` if you think it will help):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27958fcc-40f6-4f89-8116-b9ce72454009",
   "metadata": {},
   "outputs": [],
   "source": [
    "if larger_arg.num_edges > 10000:\n",
    "    raise RuntimeError(\"ARG too big to plot!!!\")\n",
    "d3larger_arg = create_styled_D3ARG(larger_arg)\n",
    "d3larger_arg.draw(\n",
    "    height=800, width=800,\n",
    "    title=f\"ARG of {larger_arg.num_trees} trees\",\n",
    "    y_axis_labels={0:0, 0.5e4:0.5e4, 1e4:1e4, 1.5e4:1.5e4, 2e4:2e4, 2.5e4:2.5e4, 3e4:3e4})\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2ae0fb6-c2d2-447a-bb0d-e76d2ad11a94",
   "metadata": {},
   "source": [
    "As you can see, this ARG is rather too large to visualise, although it can be shrunk considerably by removing the ~70% of non-coalescent nodes (most of which are recombination nodes)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e69a99ec-20ca-4fea-9176-dc6f9f994ead",
   "metadata": {},
   "outputs": [],
   "source": [
    "partial_or_always_coalescent_nodes = np.where(Workbook2.node_coalescence_status(larger_arg) > 0)[0]\n",
    "focal_nodes = np.concatenate((larger_arg.samples(), partial_or_always_coalescent_nodes))\n",
    "part_simp_arg = larger_arg.simplify(\n",
    "    focal_nodes,  # The set of nodes whose ancestry we want to keep\n",
    "    filter_nodes=False,  # Omit nodes from the genealogy, but do not completely remove them in the returned object    \n",
    "    update_sample_flags=False,  # Usually all focal nodes are turned into samples, but we don't want that here\n",
    ")\n",
    "\n",
    "d3part_simp_arg = create_styled_D3ARG(part_simp_arg)\n",
    "d3part_simp_arg.draw(height=1000, width=800, title=\"Simplified larger ARG\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fbc1668-f568-42e0-8243-3162c9afa842",
   "metadata": {},
   "source": [
    "<dl class=\"exercise\"><dt>Exercise P</dt>\n",
    "    <dd>Now repeat the <i>msprime</i> <code>larger_arg</code> simulation above, including all the subsequent <code>print</code> statements in that cell, but edit the <code>n_diploids</code> value to simulate 500 diploids (1000 sample genomes) and the <code>genome_length</code> line to simulate 10 Mb of genome instead. Run the simulation (which could take a minute or two). BE CAREFUL NOT TO PLOT THE RESULTING ARG (that will probably make your computer run out of memory and/or crash the browser).</dd></dl>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a932d608-192a-49a9-acad-ca694cf024e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use this cell for Exercise P, simulating 500 diploids over a 10Mb genome\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2609ee4a-7e3b-4137-a7e5-65edaac1b91f",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook2.url + \"Q24.json\", shuffle_answers=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2f8fe98-56cd-4e59-be44-a93ce03a77c2",
   "metadata": {},
   "source": [
    "Rather than go through the hassle a simulating a full ARG, we can simulate an already-simplified ARG. This is much more efficient:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7f169ff-fa07-4809-8d10-ee6099b719a3",
   "metadata": {},
   "source": [
    "<dl class=\"exercise\"><dt>Exercise Q</dt>\n",
    "    <dd>Go back to the simulation code you just ran, change the <code>record_full_arg</code> parameter to <code>False</code> (or remove it entirely), and re-run the 1000 sample / 10Mb simulation. Rather than simulating the full ARG, this simulates the all-coalescent (\"fully simplified\") ARG, and so should be substantially faster.</dd></dl>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f560a698-a97d-44bf-b722-30beb2866526",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Use this cell for Exercise Q, simulating a 10Mb genome for 500 diploids but with record_full_arg=False\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22796e91-de6a-478a-ad34-94bd906954f3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "display_quiz(Workbook2.url + \"Q25.json\", shuffle_answers=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a117d284-c193-4ef4-8c50-1776f3c6387e",
   "metadata": {},
   "source": [
    "The <code>extend_haplotypes()</code> method can infer missing non-coalescent segments of nodes, and place them back on the ARG. However, this can take some time for larger ARGs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47d9a12f-a0db-4d10-a959-185bc9bad4ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "start_time = datetime.now()  # Could take a few minutes\n",
    "part_coalescent_arg = larger_arg.extend_haplotypes()\n",
    "print(\n",
    "    \"Added suggested non-coalescent regions in\",\n",
    "    (datetime.now()-start_time).seconds,\n",
    "    \"seconds\",\n",
    ")\n",
    "display(part_coalescent_arg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "930d959e-4d74-49a8-9067-6804ff45cf26",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook2.url + \"Q26.json\", shuffle_answers=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b701a2c-7320-4fac-b712-1f96bf48e848",
   "metadata": {},
   "source": [
    "Although it can take some time, it may be worth creating this `part_coalescent_arg`, as a smaller ARG can make many genetic calculations more efficient, and span of ancestral haplotypes, relevant to e.g. the span of segments considered identical-by-descent, is likely to be more accurate.\n",
    "\n",
    "Alternatively, you can retain the exact non-coalescing spans of ancestral haplotypes during an _msprime_ simulation by using `coalescing_segments_only=False`. This is equivalent to partially simplifying the full ARG. However, this is not guaranteed to produce a smaller encoding (it retains extra unknowable information such as the genetic location of masked recombination events)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bb113d6-dd5a-4e10-a820-ae2a09100bcd",
   "metadata": {},
   "source": [
    "## Mutations\n",
    "\n",
    "We can add mutations *after* simulating ancestry using the [`msprime.sim_mutations()`](https://tskit.dev/msprime/docs/stable/mutations.html) function.\n",
    "\n",
    "It is valid to add mutations *after* simulating the ARG as long as they are neutral. One of the main reasons that ARGs are an efficient way to simulate genomes is because most mutations are neutral, and therefore do not need to be tracked during ancestry simulation.\n",
    "\n",
    "The default _msprime_ mutation model only adds single nucleotide changes, with equal probability of mutating between the 4 bases, but other sorts of mutation models [are available](https://tskit.dev/msprime/docs/stable/mutations.html#models)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a534ffb4-a119-4217-aaa6-32ae9d69c94e",
   "metadata": {},
   "outputs": [],
   "source": [
    "if larger_arg.sequence_length != 10_000_000 or larger_arg.num_samples != 1000:\n",
    "    raise RuntimeError(\"You have not simulated the right length ARG: Make sure you run exercise Q correctly\")\n",
    "mutated_large_arg = msprime.sim_mutations(larger_arg, rate=1e-8, random_seed=42)\n",
    "print(f\"Simulated {mutated_large_arg.num_mutations} mutations at {mutated_large_arg.num_sites} sites\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66741561-840a-466e-bb66-e09698c3ad47",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook2.url + \"Q27.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10d0f7da-db2a-4f0f-bbc8-f7bea2540a8a",
   "metadata": {},
   "source": [
    "<dl class=\"exercise\"><dt>Exercise R</dt>\n",
    "    <dd>\n",
    "        Get the times of all these mutations as a big array using <code>mutated_large_arg.mutations_time</code>, take the log (base 10) of this using <code>np.log10</code>, then plot these logged times out using<code>plt.hist</code>. You might want to also set nice tickmarks, e.g. using <code>plt.xticks(np.arange(5), 10**np.arange(5))</code>, and specify a label e.g. via <code>plt.xlabel(f\"Mutation time ({mutated_large_arg.time_units})\")</code>\n",
    "    </dd>\n",
    "</dl>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dcc08c8-ceb3-4897-a38a-801ccf0d96dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot mutation times using this cell\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27adbb00-fd9d-4f71-85bd-bc9327e853f9",
   "metadata": {},
   "source": [
    "## Stdpopsim: easily run verified simulations\n",
    "\n",
    "Even though the demographic model above is relatively simple, it still contains many parameters, and specifying it, or something similar, can be tricky and prone to error. For this reason, rather than using <em>msprime</em> directly, it is often much easier to use the <a href=\"https://popsim-consortium.github.io/stdpopsim-docs/\">Standard Library for Population Genetic Simulation Models</a> (<em>stdpopsim</em>).\n",
    "\n",
    "<img style=\"float:right; margin: 0.5em;\" src=\"https://upload.wikimedia.org/wikipedia/commons/thumb/9/9a/NASA_Joins_Jane_Goodall_to_Conserve_Chimp_Habitats_%28SVS14410%29.jpg/330px-NASA_Joins_Jane_Goodall_to_Conserve_Chimp_Habitats_%28SVS14410%29.jpg\" />\n",
    "<em>Stdpopsim</em> is a set of tried and tested genomic and demographic models for various species. We will demonstrate using the Python API (as documented in the <a href=\"https://popsim-consortium.github.io/stdpopsim-docs/stable/tutorial.html#running-stdpopsim-with-the-python-interface-api\">tutorial documentation</a>). For instance, here is an example of a demographic model of populations in the genus <em>Pan</em> i.e. common chimpanzees and bonobos."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87b5cd4a-c132-4c73-bfc8-260774ed34e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import stdpopsim\n",
    "import demesdraw\n",
    "species = stdpopsim.get_species(\"PanTro\")  # /Pan troglodytes/\n",
    "model = species.get_demographic_model(\"BonoboGhost_4K19\")\n",
    "\n",
    "print(\n",
    "    f'Chosen the demographic model \"{model.id}\" for `{species.name}` ({species.common_name}) '\n",
    "    f\"out of {len(species.demographic_models)} model(s) available:\")\n",
    "for d in species.demographic_models:\n",
    "    print(f\"* {d.id}: {d.long_description}\")\n",
    "\n",
    "colours = {'central': 'tab:blue', 'bonobo': 'tab:orange', 'western': 'tab:green'}\n",
    "demesdraw.tubes(model.model.to_demes(), colours=colours);  # You can ignore the warnings here, see https://github.com/popsim-consortium/stdpopsim/issues/1734"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a51987fa-39ac-4657-8fd8-a33018ba308a",
   "metadata": {},
   "source": [
    "And here's how to tell _stdpopsim_ to run an _msprime_ simulation of the model,  e.g. the first 20 MB of [chromosome 2A](https://en.wikipedia.org/wiki/Chimpanzee_genome_project#Genes_of_the_chromosome_2_fusion_site), from 20 bonobos, 12 central chimpanzees, and 8 western chimpanzees (remember that these are diploid genomes, so that makes 20 haploid sample nodes in total). Mutations will be automatically laid onto the ARG using `msprime.sim_mutations()` (you'll notice that we provide a `mutation_rate` below)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48eb3053-6160-4091-9da7-8f501b2a2901",
   "metadata": {},
   "outputs": [],
   "source": [
    "contig = species.get_contig(\"chr2A\", mutation_rate=model.mutation_rate, right=20e6)  # only simulate 20 Mb\n",
    "samples = {\"bonobo\": 20, \"central\": 12, \"western\": 8}\n",
    "\n",
    "engine = stdpopsim.get_engine(\"msprime\")  # Use msprime as the underlying simulator\n",
    "chimp_arg = engine.simulate(\n",
    "    model,\n",
    "    contig,\n",
    "    samples,\n",
    "    msprime_model=\"dtwf\",\n",
    "    msprime_change_model=[(20, \"hudson\")],\n",
    "    coalescing_segments_only=False,\n",
    "    random_seed=42,\n",
    ")\n",
    "print(f\"Simulated a chimp ARG of {chimp_arg.num_samples} samples with {chimp_arg.num_trees} local trees\")\n",
    "chimp_arg = chimp_arg.trim()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94e474cd-7eb5-4521-a60e-374d92026c15",
   "metadata": {},
   "source": [
    "The `model`, `contig` and `samples` parameters should be obvious. Here we have also used best practice and used the \"Discrete-Time Wright Fisher\" ([`dtwf`](https://tskit.dev/msprime/docs/stable/ancestry.html#sec-ancestry-models-dtwf)) coalescent model for the most recent 20 generations, and the (default) [`hudson`](https://tskit.dev/msprime/docs/stable/ancestry.html#sec-ancestry-models-hudson) model further back in time (this is mainly important for whole genome simulations or when sample sizes appraoch the population size, see [this paper](https://doi.org/10.1371/journal.pgen.1008619)).\n",
    "\n",
    "Finally, any extra options are passed directly to the underlying simulation engine, in this case _msprime_. In particular, we have seen that the `coalescing_segments_only` argument will generate a partially- rather than fully-simplified ARG."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9559653-c777-4ddd-a3a8-2ddc6c52e76f",
   "metadata": {},
   "source": [
    "<dl class=\"exercise\"><dt>Exercise S</dt>\n",
    "    <dd>\n",
    "        The cell below creates a list of positions along the simulated genome that can be used as windows, and a mapping of population name to population ID. In the cell below that, get the samples that correspond to the <i>central</i> population by using <code>chimp_arg.samples(pop_ids['central'])</code>, and calculate the diversity in windows along the genome using <p><code>win_div = chimp_arg.diversity(central_samples, windows=windows)<code></p>You should then be able to plot the diversity using <p><code>plt.stairs(win_div, windows, baseline=None, label='central', color=colours['central'])</code></p>In the same cell, repeat for the <i>bonobo</i>, and <i>western</i> populations, and add <code>plt.legend()</code> to generate a legend.\n",
    "    </dd>\n",
    "</dl>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "734ead7e-c5ed-41c4-aa59-2776711f3229",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This code creates the genome windows and a map from population name to population ID.\n",
    "windows = np.linspace(0, chimp_arg.sequence_length, 11)\n",
    "print(f\"Created {len(windows)} windows of length {(windows[1]-windows[0]) / 1e6} megabases each\")\n",
    "pop_ids = {pop.metadata[\"name\"]: pop.id for pop in chimp_arg.populations()}\n",
    "print(\"Population IDs are:\", pop_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f38f579-51a1-4fba-ad2c-ba559cc4f086",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Complete Exercise S here: plot diversity in windows along the genome for the 'central' population, then add the two other populations\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95f18899-dc53-4ea5-949d-2f2695994f0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook2.url + \"Q28.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1303d529-7e35-48f3-8638-6807dd3783bb",
   "metadata": {},
   "source": [
    "### Coalescence rates\n",
    "\n",
    "A relatively new and useful _tskit_ feature is the ability to estimate coalescence rates (including cross-coalescence rates) through time. The key method here is <code>pair_coalescence_rates()</code> which estimates rates in time windows and also allows different rates to be estimated along the genome. In this case, as we have a neutral simulation, there's not much point using genome windows. If e.g. we expect selection, or want to inspect hybrids, we could easily plot out a heatmap instead."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c1df814-ecf1-462b-a327-d0a977f18cef",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_log_timebins = 20\n",
    "time_breaks = np.logspace(\n",
    "    3,  # start at 10^3 generations (not much info before that)\n",
    "    np.log10(chimp_arg.max_time),\n",
    "    num_log_timebins\n",
    ")\n",
    "plot_breaks = time_breaks.copy()\n",
    "time_breaks[0] = 0  # we require the first bin to start at the sampling time\n",
    "time_breaks[-1] = np.inf  # we require the last bin to go to infinity\n",
    "rates = chimp_arg.pair_coalescence_rates(time_windows=time_breaks)\n",
    "plt.stairs(1/rates, plot_breaks)\n",
    "plt.xscale(\"log\")\n",
    "plt.xlabel(f\"Time ({chimp_arg.time_units})\")\n",
    "plt.ylabel(f\"IICR (inverse instantaneous coal. rate)\");"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb8505a4-fcb9-4b2d-acbf-458d86be1dc6",
   "metadata": {},
   "source": [
    "The \"huge\" IICR (sometimes interpreted as $N_e$) between 1000 and 10,000 generations probably just reflects the fact that these are separate species, and therefore not coalesing much at all. In this case the IICR"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "480bed8e-314e-4d73-90ec-6b873c9119c4",
   "metadata": {},
   "source": [
    "#### Cross coalescence\n",
    "If we want to look at migration *between* populations, we can look at the **cross coalescence rate**. For this we define some sets of samples (the samples in each population: central, bonobo, and western) and only record a pairwise coalescence if one of the pair is from population A and the other is from population B. Since we have 3 populations, there are only 3 possible pairs to look at:\n",
    "\n",
    "* central - bonobo\n",
    "* central - western\n",
    "* bonobo - western\n",
    "\n",
    "We specify which comparisons we want to look at by using indexes into the provided list of sample sets, i.e. in this case `(0, 1)`, `(0, 2)`, and `(1, 2)`. Note that below we plot the coalescence rate (not its inverse)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f8389b5-a13b-4fba-a2f9-d19ea7fcbace",
   "metadata": {},
   "outputs": [],
   "source": [
    "central_samples = chimp_arg.samples(pop_ids['central'])\n",
    "bonobo_samples = chimp_arg.samples(pop_ids['bonobo'])\n",
    "western_samples = chimp_arg.samples(pop_ids['western'])\n",
    "rates = chimp_arg.pair_coalescence_rates(\n",
    "    time_windows=time_breaks,\n",
    "    sample_sets=[central_samples, bonobo_samples, western_samples],\n",
    "    indexes=[(0, 1), (0, 2), (1, 2)]\n",
    ")\n",
    "plt.stairs(rates[0, :], plot_breaks, label=\"central/bonobo\")\n",
    "plt.stairs(rates[1, :], plot_breaks, label=\"central/western\")\n",
    "plt.stairs(rates[2, :], plot_breaks, label=\"bonobo/western\")\n",
    "plt.xscale(\"log\")\n",
    "plt.xlabel(f\"Time ({chimp_arg.time_units})\")\n",
    "plt.ylabel(f\"Estimated instantaneous coalecence rate\")\n",
    "plt.legend();"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90f0d4aa-283f-4bf0-90a2-eef717ea1b58",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook2.url + \"Q29.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54a65659-c5f3-4a61-981f-3a1fa3fed266",
   "metadata": {},
   "source": [
    "### PCAs\n",
    "\n",
    "Another _tskit_ feature is the ability to perform efficient matrix-vector multiplications on the genetic relatedness matrix. This enables a very efficient implementation of genetic principle components analysis, even for huge datasets. Here's an example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94863202-171e-4bd6-bdef-cb239dd1d233",
   "metadata": {},
   "outputs": [],
   "source": [
    "principal_components = chimp_arg.pca(2).factors\n",
    "for pop in chimp_arg.populations():\n",
    "    samples = chimp_arg.samples(population=pop.id)\n",
    "    if len(samples) > 0:\n",
    "        plt.scatter(\n",
    "            principal_components[samples, 0],\n",
    "            principal_components[samples, 1],\n",
    "            c=colours[pop.metadata[\"name\"]],\n",
    "            label=pop.metadata[\"name\"],\n",
    "            alpha=0.5,\n",
    "        )\n",
    "    plt.xlabel(\"PCA 1\")\n",
    "    plt.ylabel(\"PCA 2\")\n",
    "    plt.legend()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee747f06-3f52-4bf1-b1f5-e13426316307",
   "metadata": {},
   "source": [
    "## SLiM simulations\n",
    "\n",
    "If you want to add complex selection or demography to a simulation (which can be done in _stdpopsim_), you will need to use a forward simulator like the SLiM simulation engine.\n",
    "\n",
    "SLiM is a separate program that won't work in the browser, so we can't perform forward-time simulations easily in this workbook. Instead, we'll load up the results of a previous SLiM simulation. Although _tskit_ `.trees` files are quite small, there is also an additional compression program, _tszip_, which specializes in making them even smaller. It's common to use this for saving and loading large ARGs in `.tsz` format, but _tszip_ will also quite happily load normal `.trees` files too, so it's not a bad idea to get use to using it to load any tree sequence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2651eff-5f88-45f3-a5dd-844e6a76652c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tszip\n",
    "# Running SLiM is an advanced topic beyond the scope of this workbook\n",
    "slim_arg = tszip.load(\"data/SLiM_sim.tsz\")\n",
    "\n",
    "print(\n",
    "    f\"Loaded a SLiM simulated ARG with {slim_arg.num_samples} sampled genomes, \" +\n",
    "    f\"{slim_arg.num_trees} trees, \" +\n",
    "    f\"and {slim_arg.num_mutations} mutations.\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f4274bd-b2a9-427f-9374-f5449fb3a403",
   "metadata": {},
   "source": [
    "<dl class=\"exercise\"><dt>Exercise T</dt>\n",
    "    <dd>Get the first tree in the ARG using <code>slim_arg.first()</code>, and then plot it <code>.draw_svg()</code>. As there are a lot of samples, you might also want to set <code>size=(1000, 400)</code> and  (to omit any node labels) <code>node_labels={}</code>.</dd>\n",
    "</dl>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c9ba680-137f-4219-adbb-b7cc03d27110",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Complete Exercise T here (plot the first tree in the slim_arg)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a693f6d-789b-433f-9481-c38a33c52728",
   "metadata": {},
   "source": [
    "## Recapitation\n",
    "\n",
    "You can see that the simulation needed more burn-in, as the lineages have not all coalesced (in _tskit_ parlance, the trees have **multiple roots**). We can add the correct amount of (neutral) burn-in using _msprime_, which is usually much faster than SLiM. This is called [recapitation](https://tskit.dev/pyslim/docs/latest/tutorial.html). Here's a quick introduction.\n",
    "\n",
    "You can carry out recapitation using `msprime.sim_ancestry()`, but there is a convenient wrapper provided by the _pyslim_ Python library:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d64f9ed-0fbc-464f-aa38-75852ea92cb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyslim\n",
    "\n",
    "# ignore the warning the command below generates: see the docs\n",
    "coalesced_arg = pyslim.recapitate(slim_arg, recombination_rate=1e-8, ancestral_Ne=200, random_seed=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59a4e996-a83c-4993-a486-4ad8c438a560",
   "metadata": {},
   "source": [
    "You can now check that the simulation has indeed fully coalesced:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0478332-2970-4c9f-8c95-714358bae5af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check the first tree visually\n",
    "coalesced_arg.first().draw_svg(size=(1000, 400), node_labels={})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca6cc43c-3223-4b4d-bb4f-9e37bea37980",
   "metadata": {},
   "source": [
    "<dl class=\"exercise\"><dt>Exercise U</dt>\n",
    "    <dd>It's all very well checking by eye, but you probably want to make sure that all the trees (not just the first) have coalesced. It's very useful to get into the habit of putting lots of <code>assert</code> statements into your code, to sanity check your logic. An assert statement should always be True: your code will stop otherwise: it's is a great way to avoid mistakes.\n",
    "    Check that all the trees in the <code>coalesced_arg</code> have a single root by iterating through the trees and calling <code>assert tree.num_roots == 1</code> for each tree.\n",
    "    </dd>\n",
    "</dl>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f712d52b-a743-4282-aa2f-caa7b8e16d6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Complete Exercise U here (put in an assert to check that recapitation has worked)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82995d95-6d1d-4df0-8ef7-eacea2cde979",
   "metadata": {},
   "source": [
    "Now that we have a fully coalescent ARG, we can e.g. add neutral mutations, to create realistic genetic variation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1a3cf07-b220-4eb9-ae71-50ccf0277c72",
   "metadata": {},
   "outputs": [],
   "source": [
    "mu = 1e-8\n",
    "mutated_slim_sim = msprime.sim_mutations(coalesced_arg, rate=mu, random_seed=123)\n",
    "print(f\"Added {mutated_slim_sim.num_mutations} mutations to the ARG\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24c3f5cc-9fcd-4824-942e-9fe1fd9013b8",
   "metadata": {},
   "source": [
    "With a fully-coalesced ARG, we can reasonably calculate stats such as windowed genetic diversity (and because we have added mutations, we can calculate site-based as well as branch-based measures):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d9308f2-4034-4078-b36e-0ba0ca61293a",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_windows = np.linspace(0, mutated_slim_sim.sequence_length, 1001)\n",
    "step = genome_windows[1] - genome_windows[0]\n",
    "site_diversity = mutated_slim_sim.diversity(windows=genome_windows)\n",
    "branch_diversity = mutated_slim_sim.diversity(windows=genome_windows, mode=\"branch\")\n",
    "scaled_branch_diversity = branch_diversity * mu\n",
    "plt.title(f\"Genetic diversity in {step / 1000:.0f}kb genome windows\")\n",
    "plt.stairs(site_diversity, genome_windows, baseline=None, ls=\":\", label=\"site-based\")\n",
    "plt.stairs(scaled_branch_diversity, genome_windows, baseline=None, label=\"branch-based\")\n",
    "plt.legend();"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe3b94eb-5fa1-4a7f-9901-eabe5bda4fec",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook2.url + \"Q30.json\", shuffle_answers=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5904c3f5-2315-49bb-a7d4-bb7ceb0844f6",
   "metadata": {},
   "source": [
    "Recapitation can be a game-changer when simulating realistic populations, saving many days of simulation time. \n",
    "\n",
    "# More information\n",
    "\n",
    "That's the end of the workbook. Theere is a lot more official documentation at https://tskit.dev/tskit/docs/, and an ever-growing set of tutorials at https://tskit.dev/tutorials. More help to write tutorials is also most welcome! Good luck on your ARG journey."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
