{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d839c4bf-ceff-4e60-b93b-c87ad7667ef3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "if sys.platform != 'emscripten':\n",
    "    # allow the notebook to be downloaded and run locally too\n",
    "    print(\"This notebook is not running in JupyterLite: you may need to install tskit, tszip, etc.\")\n",
    "else:\n",
    "    import micropip\n",
    "    await micropip.install('jupyterquiz')\n",
    "    await micropip.install('tskit_arg_visualizer')\n",
    "\n",
    "from jupyterquiz import display_quiz\n",
    "from WhatIsAnARG_module import Workbook1\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "404f5dec-3289-4ef5-b532-ae0b2eff15f3",
   "metadata": {},
   "source": [
    "# ARGs and _tskit_: an introduction\n",
    "\n",
    "### Background reading and skills required\n",
    "\n",
    "ARGs provide a principled way of thinking about genetic inheritance with recombination: i.e. recombinant phylogenies. Some recent reviews are [Lewanski at al. (2024)](https://doi.org/10.1371/journal.pgen.1011110) and [Nielsen et al. (2024)](https://doi.org/10.1038/s41576-024-00772-4). To complete this workbook you should be comfortable with basic programming in [Python](https://www.python.org) (including e.g. [list comprehensions](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions) and [dictionaries / dict comprehensions](https://docs.python.org/3/tutorial/datastructures.html#dictionaries)). The workbook also makes substantial use of the \"numerical Python\" library, [`numpy`](https://numpy.org): there is a quickstart tutorial [here](https://numpy.org/devdocs/user/quickstart.html). It may also help if you have some familiarity with the basics of the [Tree Sequence Toolkit (_tskit_)](https://tskit.dev/genetics-research/), and its Python interface, by reading https://tskit.dev/tutorials/what_is.html, https://tskit.dev/tutorials/args.html and https://tskit.dev/tutorials/no_mutations.html.\n",
    "\n",
    "### Overview\n",
    "This is the first of 2 workbooks: each is intended to take about 1½-2 hours to complete.\n",
    "\n",
    "<dl>\n",
    "<dt>Workbook 1</dt>\n",
    "    <dd>Graphs and local trees, mutations, genetic variation and statistical calculations: exercises A-J & questions 1-11</dd>\n",
    "<dt>Workbook 2</dt>\n",
    "    <dd>Types of ARG, simplification, and ARG simulation: exercises K-U & questions 12-30</dd>\n",
    "</dl>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "521cd35b-a4b5-40c8-981c-652aa621832d",
   "metadata": {},
   "outputs": [],
   "source": [
    "Workbook1.setup()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa72ab24-0871-4111-9d3e-6a8c156f1b41",
   "metadata": {},
   "source": [
    "# Workbook 1: ARGs, local trees, and genetic variation\n",
    "\n",
    "## _Tskit_ basics\n",
    "\n",
    "The _tskit_ library allows you to store and analyse ARGs in the *tree sequence* format. This is the library that underlies simulators such as _msprime_ and _SLiM_, and inference tools such as _tsinfer_, _tsdate_, and _SINGER_. The library is extensive and well-documented at https://tskit.dev/tskit/docs/stable/. We will use the [Python interface](https://tskit.dev/tskit/docs/stable/python-api.html), but there are also C and Rust inferfaces, and it is possible to use _tskit_ from [within R]() as well.\n",
    "\n",
    "First, we load a small simulated ARG in tree sequence format. This is a \"full ARG\" which describes the ancestry of a set of 10 sampled genomes and includes nodes that explicitly record recombination events (so-called \"recombination nodes\").\n",
    "\n",
    "It is conventional to store such genealogies in a variable called `arg` or `ts`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdb6e2f4-c0f7-4bf1-a129-9ea9fa4eb57b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tskit\n",
    "arg = tskit.load(\"data/example_ARG.trees\")  # it is conventional to use `arg` or `ts` as the standard variable\n",
    "print(\"A tskit ARG has been loaded into the variable 'arg'\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86833468-8d09-42e4-9797-7132f0f7af18",
   "metadata": {},
   "source": [
    "It is important to distinguish the *sample genomes* in this ARG. Their numerical IDs can be obtained using the `.samples()` method. Often these will be sequential, starting from zero. However, this need not always be the case so it's always best to check:\n",
    "<div class=\"alert alert-block alert-info\"><b>Note:</b> For those coming from <em>R</em>, <em>tskit</em> uses the Python convention of indexing from 0 rather than from 1. So the first node has ID 0, not ID 1</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78dc246a-73a7-4571-b846-247488810954",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"The sampled genomes have IDs:\", arg.samples())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfdf625f-7ad6-485a-aa77-1b865eb7e30d",
   "metadata": {},
   "source": [
    "## ARGs and local trees\n",
    "\n",
    "The [_tskit_arg_visualizer_](https://github.com/kitchensjn/tskit_arg_visualizer) software uses the [D3js library](https://d3js.org) to visualise ARGs and other tree sequences interactively, in a browser or Jupyter notebook. As is conventional, the oldest nodes are drawn at the top, with the youngest, usually at time 0, at the bottom.\n",
    "\n",
    "The visualiser creates a [`D3ARG`](https://github.com/kitchensjn/tskit_arg_visualizer/blob/main/docs/tutorial.md#what-is-a-d3arg) object from the _tskit_ ARG. This object can then be plotted using `.draw()`.\n",
    "\n",
    "<div class=\"alert alert-block alert-info\"><b>Note:</b> You'll see that some nodes in this plot have two IDs. Don't worry about this: as we'll see later it's because the simulator has represented recombination using 2 nodes, which have been overlaid in the visualizer</div>\n",
    "\n",
    "<dl class=\"exercise\"><dt>Exercise A</dt>\n",
    "    <dd>Try tidying-up the plot below by dragging nodes horizontally. You can save the result if you want via the buttons above the graph.</dd>\n",
    "</dl>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b7b2e45-9518-4bbd-a426-7b4eb7bc2557",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tskit_arg_visualizer as argviz\n",
    "\n",
    "d3arg = argviz.D3ARG.from_ts(arg)  # by convention, the argviz object has \"d3\" prepended to the original name\n",
    "d3arg.set_node_styles({u: {\"symbol\": \"d3.symbolSquare\"} for u in arg.samples()}) # Set some viz styles\n",
    "d3arg.draw(edge_type=\"ortho\", width=800);  # draw the D3ARG in the notebook: the semicolon hides the return value of the draw() method"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6473209-6dce-45f6-91b6-a0e2fc96802c",
   "metadata": {},
   "source": [
    "## The importance of sample nodes\n",
    "\n",
    "Each node in the graph above represents a genome. The **square** nodes representing *sampled genomes* (or \"samples\") and the round \"nonsample\" nodes represent *ancestral genomes*. Importantly, ARGs only represent the ancestry of the sample nodes. For example, ancestor 35 would have passed on a whole genome's worth of DNA to node 34, but if you hover over the edge that joins them, you well see that only part of the genome eventually made it into the samples.\n",
    "\n",
    "This means that even if we have the whole genomes of all the sample nodes, the known regions of ancestral genome may not cover the whole sequence length. To put it another way, ancestral genomes may only be partially knowable.\n",
    "\n",
    "<dl class=\"exercise\"><dt>Exercise B</dt>\n",
    "    <dd>Hover over a few of the lines in the graph above to see which inherited regions are inherited along different edges. Then try adding <code>variable_edge_width=True</code> to the <code>.draw()</code> call, to display the widths of edges as proportional to their span. Does this help show the routes through which the majority of the genome has travelled?</dd>\n",
    "</dl> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "134503fa-ba13-44c3-8ac7-957741e1d7ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook1.url + \"Q1.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39bc5010-1fc4-4830-ad01-6e67f88bb7fe",
   "metadata": {},
   "source": [
    "## SPR operations\n",
    "\n",
    "Each left/right transition from one tree to another represents a *recombination breakpoint* where forward in time, two lineages (one from the maternal and one from the paternal genome) were combined via meiosis. Looking backwards in time, this results in one genome (node) having two parents.\n",
    "\n",
    "If we know the recombination nodes, it is easy to see how one tree changes into another along the genome. A good approximation is that the tree on one side of a breakpoint can be transformed into the tree on the other side of the breakpoint via a single tree-editing operation, or SPR. \n",
    "\n",
    "<dl class=\"exercise\"><dt>Exercise C</dt>\n",
    "    <dd>Move your pointer over the genome bar underneath the graph, and try to get a feel for how one local tree (which will be highlighted in green, embedded in the larger ARG) is transformed into the next one.</dd>\n",
    "</dl> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57a422ed-ca10-4d8b-a8cb-22b30dbe7ac1",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook1.url + \"Q2.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1f37537-2dd1-4a2b-93c9-d9ddd651746b",
   "metadata": {},
   "source": [
    "## Left-right ARG traversal: the `.trees()` iterator\n",
    "\n",
    "A fundamental ARG operation, which forms the basis of operations such as calculating genetic statistics, is to move along the genome, obtaining the local trees at each point. In _tskit_ this is done using the [`.trees()`](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.trees) method. Each new tree, and the data associated with it, is formed by making a slight change to the previous tree, making this an efficient operation, even for huge trees of millions of nodes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b8d0caa-f1e6-4f3b-87dd-1a1bc30e446c",
   "metadata": {},
   "outputs": [],
   "source": [
    "for tree in arg.trees():\n",
    "    display(tree)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "628dcc03-d985-4e24-a817-5f8e7ac4cca1",
   "metadata": {},
   "source": [
    "### Visualising local trees\n",
    "\n",
    "By default, `tskit` displays each local tree as a summary table, as above. To draw the tree, you can use the [`.draw_svg()`](https://tskit.dev/tutorials/viz.html#svg-format) method, suitable for small trees of tens or hundreds of nodes each."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "974cd8e4-8db4-41f0-8341-3766a9bb5341",
   "metadata": {},
   "outputs": [],
   "source": [
    "arg.draw_svg(\n",
    "    size=(1200, 400),  # (width, height) in pixels\n",
    "    omit_sites=True,  # Later in the workbook we'll remove this, to reveal the variable sites \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c614741-f2c2-48d3-ad40-de9fed6bb222",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook1.url + \"Q3.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd840a92-7216-4315-82b3-be46949572b2",
   "metadata": {},
   "source": [
    "### Branch length changes\n",
    "Perhaps surprisingly, many recombination events in an ARG do not change the topology of the local trees, but merely change branch lengths (i.e. change the time and identity of the most recent common ancestor, or MRCA, of certain sets of samples). This can be seen most clearly at the top of the graph-based ARG visualization. At 30 000 generations ago there are only two lineages left. Whichever route we take above that through the ARG, will end up joining node 26 and node 23, just at a potentially different time in the past."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc015258-3a45-4604-a7e4-52bb0c0cee59",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook1.url + \"Q4.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b8016f3-49ca-4750-a92c-f55562c5c20d",
   "metadata": {},
   "source": [
    "### \"Identical tree\" changes\n",
    "\n",
    "Some recombination events change neither the topology nor the branch lengths. This is seen when going from the third to the fourth tree. That change only involves node ID 24 switching to ID 25, which represents the recombination event at the bottom of the diamond. A peculiarity of ARGs created using the _msprime_ simulator is that for [technical reasons](https://tskit.dev/msprime/docs/stable/ancestry.html#recombination-events), recombination events are recorded as *two* simultanous recombination nodes (e.g. nodes 24/25), representing the paternal and maternal genomes prior to meiosis. These are visually merged into a single node in the ARG visualizer. Assuming a single crossover event, one node is to the left and one to the right of the breakpoint."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b3e5ac3-3595-42a9-98d0-16996f395c98",
   "metadata": {},
   "source": [
    "<dl class=\"exercise\"><dt>Exercise D</dt>\n",
    "    <dd>Copy the <code>draw_svg()</code> code above into the box below, but plot only the trees between genome position 200 to 800 by using <code>x_lim=(200, 800)</code>, and add a Y axis, using <code>y_axis=True</code>. To make the Y axis ticks look nicer, you could also specify <code>y_ticks={0: \"0\", 20_000: \"20k\", 40_000: \"40k\"}</code>.</dd>\n",
    "</dl>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75f20ce5-96ba-4959-ad45-c849969e95ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Complete Exercise D here\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36f07cea-6e9e-4915-9214-903a8db46e43",
   "metadata": {},
   "source": [
    "## Coalescent and non-coalescent regions\n",
    "\n",
    "Looking at the tree-by-tree plot, it should be clear that some of the nodes in a local tree have one child in some trees, and two children in others. There are even some nodes that have only one child in every tree in which they appear (e.g. node 26). We can classify nodes into:\n",
    "\n",
    "0. **non-coalescent**, sometimes called _always unary_ (i.e. one child in all local trees, e.g. node 26)\n",
    "1. **part-coalescent**, sometimes called _locally unary_ (i.e. one child in some local trees, coalescent in others, e.g. node 18)\n",
    "2. **all-coalescent**, or _never unary_ (always 2 or more children in any local tree in which they are found, e.g. node 19).\n",
    "\n",
    "<div class=\"alert alert-block alert-info\"><b>Note:</b> This classification will turn out to be important in part B, when we learn about <em>simplifying</em> ARGs by collapsing some nodes (particularly \"non-coalescent\" nodes) depending on whether they are detectable.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9603eaa-9a3c-44d9-bdba-ad922ef73bd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook1.url + \"Q5.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "470751f7-82b7-4b4c-a487-8128adb6c36b",
   "metadata": {},
   "source": [
    "The `Workbook.node_coalescence_status()` function, written for this workbook, uses the `.trees()` iterator to construct an array denoting the classification of each node. `returned_array[i]` is **0** if node `i` is non-coalescent, **1** if node `i` is part-coalescent, and **2** if node `i` is all-coalescent. You can check below that this agrees with your understanding:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd53f945-120c-474b-8d62-43144aa144a5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "description = {0: \"Non-coalescent\", 1: \"Part-coalescent\", 2: \"All-coalescent\"}\n",
    "samples = arg.samples()\n",
    "for node_id, status in enumerate(Workbook1.node_coalescence_status(arg)):\n",
    "    extra = \", sample\" if node_id in samples else \"\"\n",
    "    print(f\"Node {node_id} coalescence status: {description[status]} ({status}{extra})\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd136763-1591-4e94-a861-414746c675ed",
   "metadata": {},
   "source": [
    "## Nodes and individuals\n",
    "\n",
    "You can think of each ARG node as representing a (haploid) genome. Humans (and most other eukaryotes) are *diploid*, meaning they contain 2 genomes: a maternal and a paternal one. \n",
    "\n",
    "In _tskit_, this information is stored by allowing a node to be asociated with an individual ID, linked to a separate table of individuals. Usually only sample nodes have positive individual ID. Other nodes have their `individual` ID set to `-1` (also known as `tskit.NULL`); this includes ancestral nodes, as we usually don't know how ancestral genomes were grouped into diploid individuals."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34ceda57-b54a-410c-bf54-fcb05b30ff84",
   "metadata": {},
   "outputs": [],
   "source": [
    "colours = {\n",
    "    tskit.NULL: \"LightGrey\",  # nodes that are not associated with an individual are coloured in grey\n",
    "    0: \"LightGreen\",\n",
    "    1: \"DodgerBlue\", \n",
    "    2: \"MediumVioletRed\",\n",
    "    3: \"Coral\",\n",
    "    4: \"DarkGoldenRod\"\n",
    "}\n",
    "for node in arg.nodes():\n",
    "    individual_id = node.individual\n",
    "    print(\n",
    "        f\"Node {node.id} is associated with individual {individual_id}\",\n",
    "        f\"which we will colour {colours[individual_id]}\" if  individual_id != tskit.NULL else \"\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f9ee5a7-b0ca-4b9f-a5e0-9193e797c563",
   "metadata": {},
   "source": [
    "When visualising the ARG (either as a graph or as a series of trees), we don't usually want to group the two nodes of each individual together, as the maternal and paternal genomes within an individual can have quite different histories. You can try to rearrange the sample nodes by colour, to convince yourself that the two samples from a single individual do not trivially group together.\n",
    "\n",
    "Note that the following code does not use the \"orthogonal\" plot style, but reverts to the default `edge_type` of `\"line\"`, which simply joins nodes by straight lines. As well as colouring the nodes by individual, it colours recombination nodes as black. It also shows not just the *topology* of the ARG, but also the length (in time) of each edge, by plotting the Y axis on a linear timescale."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34dd2283-5c5a-4f4a-8723-6e1370275a1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import msprime\n",
    "\n",
    "d3arg.set_node_styles({\n",
    "    node.id: {\"fill\": colours[node.individual]}\n",
    "    for node in arg.nodes()\n",
    "})\n",
    "d3arg.draw(title=\"Nodes coloured by individual\", y_axis_scale=\"time\", width=800);"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0876be3-5bfb-42ee-884d-c844b4c19743",
   "metadata": {},
   "source": [
    "The ARG was actually simulated using a model of human evolution that reflects the Out of Africa event. As well as having a value denoting the <code>individual</code>, each node also has a value indicating a <code>population</code> to which it belongs.\n",
    "\n",
    "<dl class=\"exercise\"><dt>Exercise E</dt>\n",
    "    <dd>Change the code above to colour by <code>node.population</code> ID rather than <code>node.individual</code> ID. You could also stop colouring the recombination nodes as black if you like.</dd>\n",
    "</dl>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21f0b573-2b5c-4d02-ba7a-58026b4ee64b",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook1.url + \"Q6.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61739bba-c121-461d-928e-033bb5a66420",
   "metadata": {},
   "source": [
    "### Metadata (a side note)\n",
    "\n",
    "For efficiency, most <em>tskit</em> functions refer to nodes, individuals, populations, etc using their numerical IDs. However can be tricky to keep track of e.g. which populations correspond to which IDs. Moreover, these ID can change, for instance if populations or nodes are removed from the tree sequence. So to help keep track of objects, even when their IDs change, <em>tskit</em> allows them to be associated with user-defined *metadata*. There's a <a href=\"https://tskit.dev/tutorials/metadata.html\">tutorial on metadata</a> if you need to know more.\n",
    "\n",
    "In our example ARG, we have added metadata for each population consisting of a name and a fuller description. You can see the this when accessing a population by its ID:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bce6869-3119-43ea-889d-a699f395911a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_id = 0\n",
    "pop = arg.population(pop_id)\n",
    "print(pop)  # or print(pop.metadata) to see just the metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49ae6107-df87-4ba0-81ba-0737111da346",
   "metadata": {},
   "source": [
    "Or you can see the metadata for all populations by looping over them:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "133abfb7-4d07-4d7c-a26d-be55183930ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "for pop in arg.populations():\n",
    "    print(pop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b504e594-b8d6-4819-8e84-41cd5decc0ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sometimes it is helpful to use a mapping of a metadata value to the object ID\n",
    "pop_name_to_id_map = {pop.metadata['name']: pop.id for pop in arg.populations()}\n",
    "\n",
    "# Now use the map to get only those samples from the `AFR` population\n",
    "print(\n",
    "    \"The sample genomes for the AFR population have IDs:\",\n",
    "    arg.samples(population=pop_name_to_id_map[\"AFR\"])\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45a7aa32-6ee9-4056-a170-7e9c235996be",
   "metadata": {},
   "source": [
    "<dl class=\"exercise\"><dt>Exercise F</dt>\n",
    "    <dd>Using a <code>for</code> loop, iterate over <code>arg.individuals()</code> and print our each individual.</dd>\n",
    "</dl>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6b7b70c-7a7b-4bf9-a6e7-4b132b906456",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Complete Exercise F here\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27c9c0bf-eb16-4e81-ac3c-7e14d06f07a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook1.url + \"Q7.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e09238d-b45a-426a-ad26-ed353cb740f4",
   "metadata": {},
   "source": [
    "## Mutations\n",
    "\n",
    "Genetic variation is create by *mutations* on the ARG. So far we have not shown mutations, but we can reveal them by setting `show_mutations` to `True` in the ARG visualizer. If you hover over the orange mutation symbols in the visualization below, their position on the lower genome bar will be revealed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1312cf4d-34b1-48cc-b379-ab7b62bc255e",
   "metadata": {},
   "outputs": [],
   "source": [
    "d3arg.draw(show_mutations=True, y_axis_scale=\"time\", width=800, title=f\"ARG with {arg.num_mutations} mutations\");"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f94dcef1-920f-4930-a283-63608499d5ad",
   "metadata": {},
   "source": [
    "### Encoding sites and mutations in _tskit_\n",
    "\n",
    "In _tskit_, mutations occur at specific *sites*. A site is defined by a *site id*, a *position* and an *ancestral state*. Mutations are defined as an *mutation id*, a *derived state*, an associated *node id* an (optional) *time*, and an associated *site id*. If two mutations have the same site id, those mutations occur at the same position along the genome."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "253e914d-db52-4122-b678-99a55f223e9e",
   "metadata": {},
   "source": [
    "<dl class=\"exercise\"><dt>Exercise G</dt>\n",
    "    <dd>The <code>include_mutation_labels</code> and <code>condense_mutations</code> parameters provide different ways to display mutations on branches. Add <code>label_mutations=True</code> to the <code>.draw()</code> method above, to see its effect.</dd>\n",
    "</dl>\n",
    "\n",
    "<dl class=\"exercise\"><dt>Exercise H</dt>\n",
    "    <dd>In the cell below, use <code>arg.draw_svg()</code> to plot the mutated arg as a series of trees <em>without</em> using the <code>omit_mutations</code> parameter. You will see that mutations are drawn as red crosses on the trees, and also depicted as red arrows on the x axis, at the appropriate site position. Multiple mutations at the same site will have those arrows stacked.</dd>\n",
    "</dl>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8839bd4-66a7-462d-b928-9e798dce6271",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Complete exercise H here\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "227ff283-f93d-46bc-aa16-d631d56a23c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook1.url + \"Q8.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87345d4f-65be-49d4-887d-44eba3d48bf3",
   "metadata": {},
   "source": [
    "## Genetic variation & the `.variants()` iterator\n",
    "\n",
    "The inheritance of mutations through the ARG defines the genotypes at each variable site for each sample. The _tskit_ [`.variants()`](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.variants) method goes through the local trees, decoding the genetic variation at each site by looking at the inheritance of mutations. For efficiency, the genotypes for all the samples at a site are returned as a numerical vector. Each value in the vector denotes the index into a corresponding list of alleles:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a29715a1-4f22-4590-9507-87b2a612f270",
   "metadata": {},
   "outputs": [],
   "source": [
    "for variant in arg.variants():\n",
    "    print(\n",
    "        f\"Genotypes at position {variant.site.position}\",\n",
    "        f\"for the {arg.num_samples} samples \",\n",
    "        f\"are {variant.genotypes},\",\n",
    "        f\"denoting the alleles {variant.alleles}\",\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ddf955b-0b7c-4b4d-980d-e506f8091c77",
   "metadata": {},
   "source": [
    "<dl class=\"exercise\"><dt>Exercise I</dt>\n",
    "    <dd>Change the code above to display the actual allelic states (rather that a numerical value), by using the <a href=\"https://tskit.dev/tskit/docs/stable/python-api.html#tskit.Variant.states\"><code>variant.states()</code></a> method in place of accessing <a href=\"https://tskit.dev/tskit/docs/stable/python-api.html#tskit.Variant.genotypes\"><code>variant.genotypes</code></a> directly. This can be useful for displaying or checking smaller examples, as opposed to numerical analysis of larger datasets.</dd>\n",
    "</dl>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57dbec68-f09c-476e-a235-9837890fd46e",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook1.url + \"Q9.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5d18a64-829f-4e6a-bd05-be683e995826",
   "metadata": {},
   "source": [
    "## Mutations allow ARG inference\n",
    "\n",
    "Mutations on an ARG generate genetic variation. Conversely, genetic variation among a set of samples can be used to *infer* an ARG that could plausibly have generated the samples. To do this, it is often useful to make the \"infinite sites\" approximation, that every mutation occurs at a different position (as we saw above, this is not always true).\n",
    "\n",
    "When we use mutations do infer genealogy, it can be helpful to *polarise* the alleles at a genetic site (figure out which allele is ancestral and which are derived). Then we can reasonably assume that if a set of genomes share the derived allele, they all inherited it from a common ancestor. This allows us to deduce the presence of a common ancestor node in the ARG.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17a0ee9f-125b-4fa9-a8af-93cec1f5f8ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook1.url + \"Q10.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd637a11-3bf9-4724-99c0-95b62dc7fdba",
   "metadata": {},
   "source": [
    "## Provenance\n",
    "\n",
    "Helpfully, _tskit_ tries to store information on how an ARG was generated. Software like _msprime_, whcih was used to generate this ARG, place details of their commands in the [provenance](https://tskit.dev/tskit/docs/stable/provenance.html) table. We'll use this to check on the mutation rate that was used in the simulation, so that we can use it later:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ca88b44-16b0-46d5-9e83-21e3cb6ffd6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json  # Provenance data is stored in JSON format\n",
    "last_provenance = json.loads(arg.provenance(-1).record)\n",
    "assert last_provenance['parameters']['command'] == \"sim_mutations\"  # Check the last operation added mutations via the msprime `sim_mutations` command\n",
    "mutation_rate = last_provenance['parameters']['rate']\n",
    "print(\"The ARG was generated using a mutation rate of\", mutation_rate, \"mutations per base pair per generation\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa360968-fb3e-4abc-a94b-28b3056cc91f",
   "metadata": {},
   "source": [
    "## Population genetic statistics\n",
    "\n",
    "An ARG with mutations completely summarises the genetic variation present in the samples. Moreover, by only looking at differences between local trees, genome-wide statistical calculations can often be performed much more efficiently than naive approaches. Details of the statistics available in _tskit_ (including genetic divergence, Patterson's f statistics, PCA and genetic relatedness, etc) are at [https://tskit.dev/tskit/docs/stable/stats.html](https://tskit.dev/tskit/docs/stable/stats.html). Many of these statistics are summaries of the _allele frequency spectrum_, which itself can be returned by a single call to the [`.allele_frequency_spectrum()`](https://tskit.dev/tskit/docs/stable/stats.html#sec-stats-notes-afs) method:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d049466-552a-4308-a0cf-bfb5755d006c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import pyplot as plt\n",
    "\n",
    "\n",
    "# Whole genome AFS\n",
    "afs = arg.allele_frequency_spectrum(polarised=True)\n",
    "plt.bar(range(arg.num_samples + 1), afs)\n",
    "plt.xlabel(\"Number of samples inheriting a mutation\")\n",
    "plt.ylabel(\"Density\")\n",
    "plt.title(\"Polarised allele frequency spectrum\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e857d4bd-37de-4a44-b36a-8ef25589e23f",
   "metadata": {},
   "source": [
    "<dl class=\"exercise\"><dt>Exercise J</dt>\n",
    "    <dd>Add <code>span_normalise=False</code> to the <code>.allele_frequency_spectrum()</code> call above, and re-run, so that the Y axis plots the actual number of mutations instead dividing that by the sequence length. Does the number of mutations in all the bars add up to the total number in the ARG?</dd>\n",
    "</dl>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c950279-0632-42c7-896f-bda9544b3116",
   "metadata": {},
   "source": [
    "### Windowed statistics\n",
    "\n",
    "Statistics can also be windowed along the genome and also by time. Here we demonstrate a simplest measure, the genetic diversity (sometimes known as $\\pi$), and a more complex one, Wright's Fst."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8047854-99c8-4548-9bf6-259f76f4b648",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Average genetic diversity (π) = {arg.diversity()}\")\n",
    "AFR_samples = arg.samples(population=0)\n",
    "EUR_samples = arg.samples(population=1)\n",
    "\n",
    "print(f\"Average Fst = {arg.Fst([AFR_samples, EUR_samples])}\")\n",
    "\n",
    "# Now show this in windows\n",
    "genome_windows = [0, 250, 500, 750, 1000]\n",
    "windowed_diversity = arg.diversity(windows=genome_windows)\n",
    "windowed_Fst = arg.Fst([AFR_samples, EUR_samples], windows=genome_windows)\n",
    "\n",
    "fig, (ax_top, ax_bottom) = plt.subplots(2, 1, sharex=True)\n",
    "ax_top.stairs(windowed_diversity, genome_windows, baseline=None)\n",
    "ax_top.set_ylabel(\"Genetic diversity\")\n",
    "ax_top.set_title(f\"Genetic statistics in {len(genome_windows) - 1} windows along the genome\")\n",
    "ax_bottom.stairs(windowed_Fst, genome_windows, baseline=None)\n",
    "ax_bottom.set_ylabel(\"AFR/EUR Fst\")\n",
    "ax_bottom.set_xlabel(\"Genome position\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53770b27-71d1-4f7f-8f36-9b06711c0634",
   "metadata": {},
   "source": [
    "### Do you need genetic variation?\n",
    "\n",
    "It is worth noting that if you have the genealogy, then an alternative set of statistics exist which can be less noisy than those based on the raw genetic variation data. See https://tskit.dev/tutorials/no_mutations.html for more details.\n",
    "\n",
    "These \"branch\" mode statistics do not require mutations, and so do not contain noise associated with the mutational process. For instance, as our ARG was created under a simple neutral model, we would expect the genetic statistics along the genome to be constant. And indeed, using `mode=\"branch\"` results in a comparable windowed genome plot, in which the genetic statistics are less noisy than those based on the genetic variation created by mutations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d00b5c27-ecb2-4c87-ad07-2f8a607dca96",
   "metadata": {},
   "outputs": [],
   "source": [
    "branch_Fst = arg.Fst([AFR_samples, EUR_samples], windows=genome_windows, mode=\"branch\")\n",
    "branch_diversity = arg.diversity(windows=genome_windows, mode=\"branch\")\n",
    "scaled_branch_diversity = branch_diversity * mutation_rate\n",
    "\n",
    "\n",
    "fig, (ax_top, ax_bottom) = plt.subplots(2, 1, sharex=True)\n",
    "ax_top.stairs(windowed_diversity, genome_windows, baseline=None, ls=\":\", label=\"site (mutation)-based\")\n",
    "ax_top.stairs(scaled_branch_diversity, genome_windows, baseline=None, label=\"branch-based\")\n",
    "ax_top.set_ylabel(\"Genetic diversity\")\n",
    "ax_top.set_title(f\"Genetic statistics in {len(genome_windows) - 1} windows along the genome\")\n",
    "ax_bottom.stairs(windowed_Fst, genome_windows, baseline=None, ls=\":\", label=\"site (mutation)-based\")\n",
    "ax_bottom.stairs(branch_Fst, genome_windows, baseline=None, label=\"branch-based\")\n",
    "ax_bottom.text(0, 0.13, f'Average branch Fst: {arg.Fst([AFR_samples, EUR_samples], mode=\"branch\"):.4f}')\n",
    "ax_bottom.set_ylabel(\"AFR/EUR Fst\")\n",
    "\n",
    "ax_bottom.set_xlabel(\"Genome position\")\n",
    "ax_top.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55e337a5-2a8a-42fc-ab2c-4185ffcb6ddd",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_quiz(Workbook1.url + \"Q11.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23c37c45-ecbe-4db3-8a93-7d7d86d58c32",
   "metadata": {},
   "source": [
    "We will encounter other summary statistics in future workshops"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fd19d18-0f30-4acb-b225-9834e1e7a09e",
   "metadata": {},
   "source": [
    "## Extra *tskit* details\n",
    "\n",
    "This section aims to give you a deeper understanding of how ARGs are stored within _tskit_, and how to access the raw data underlying a _tskit_ ARG. It is deliberately optional in the sense that there are no exercises or questions, but it should help you to master the _tskit_ library.\n",
    "\n",
    "### Tables\n",
    "\n",
    "Nodes, edges, individuals, populations, sites, and mutations are all stored as *rows* in a set of separate *tskit* **tables**. The ID of an object is simply its row number in the corresponding table. Together the tables make up a [TableCollection](https://tskit.dev//tables_and_editing.html). You can see a summary of the tables by simply displaying a tree sequence in a notebook: this describes the number of rows in each table, and the size of each table. Below you will see that our `arg` has 42 edges, 37 nodes, and 13 mutations at 12 different sites along the genome:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "247cf3af-26d7-4e7b-bea5-6da4495b5a6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "display(arg)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71194ca9-1a67-48e3-bed9-e061a6321c3c",
   "metadata": {},
   "source": [
    "Each of the tables can also be displayed separately, using `arg.tables.nodes`, `arg.tables.edges`, etc. Printing a table or the whole table collection creates a plain-text representation (helpful e.g. if you are not using a notebook):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3453036-fcad-49b3-ba6f-6cec24f08f9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(arg.tables.edges)  # or display(arg.tables.edges) for a prettier HTML version"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81f33a8f-e58a-444b-ae12-37c92dcd9793",
   "metadata": {},
   "source": [
    "It should be reasonably obvious how this works. E.g. edge 0 connects parent node 10 to child node 6 in the part of the genome that spans 0 to 930 bp. For further information see [https://tskit.dev/tskit/docs/stable/data-model.html](https://tskit.dev/tskit/docs/stable/data-model.html), and for a tutorial approach, see [https://tskit.dev/tutorials/tables_and_editing.html](https://tskit.dev/tutorials/tables_and_editing.html).\n",
    "\n",
    "We previously used [`arg.nodes()`](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.nodes), [`arg.individuals()`](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.individuals), and [`arg.populations()`](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.populations) to return Python objects, created by iterating over all the rows in a table. Similarly, methods exist for [`arg.edges()`](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.edges), [`arg.sites()`](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.sites), and [`arg.mutations()`](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.mutations). To access a specific edge, node, site, etc. as a Python object you can also use [`arg.edge(i)`](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.edge), [`arg.node(i)`](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.node), [`arg.site(i)`](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.site), and so on. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d0364b9-4d0c-4be7-886f-19f90d71156a",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(arg.edge(0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5aa40b3-9385-415e-8b53-1dbe9d2cf8bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(arg.node(10))\n",
    "print(arg.node(6))\n",
    "individual_id_for_node_6 = arg.node(6).individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79110ae0-3777-480f-a10a-33bf7687f613",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(arg.individual(individual_id_for_node_6))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63667e7f-4439-42eb-ac01-04b7dd2418f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(arg.mutation(0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef3ca4b7-c86d-4c75-9313-2ee93a4f9f32",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(arg.site(0))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bca42060-a2fb-45b4-9175-ea63560300c8",
   "metadata": {},
   "source": [
    "#### High performance data access\n",
    "\n",
    "Using Python objects is convenient, but can be inefficient for large ARGs. The most performant way to access the underlying data is to use the [efficient column accessors](https://tskit.dev/tskit/docs/stable/python-api.html#efficient-table-column-access), which provide _numpy_ arrays that are a direct view into memory. For example, to find all the site positions along the genome, you can use  `arg.tables.sites.position` (or the shortcut [`arg.sites_position`](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.sites_position)). This is particularly relevant when dealing with ARGs containing large tables (e.g. millions of rows)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35f24619-da2f-407f-b724-584c9439a1c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "arg.tables.sites.position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec64bfaa-0af3-411b-a2c9-42173a9386b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "arg.tables.edges.parent  # Also available via the shorthand arg.edges_parent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6dfdf05e-bc7b-4316-9ed4-60404cb5f8bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "arg.tables.edges.child  # Also available via the shorthand arg.edges_child"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ba5e13f-3a1d-454a-997a-2fc871191fda",
   "metadata": {},
   "source": [
    "and so on. These can be used with _numpy_, the numerical Python library:\n",
    "\n",
    "<!-- For more information about writing high performance algorithms, see https://github.com/tskit-dev/tutorials/issues/151 (not yet written) -->"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "caf9f270-9eee-47b6-841e-98a3b7606981",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "# find the the mean of the base 10 log of the times of all mutation in the arg\n",
    "np.mean(np.log10(arg.tables.mutations.time))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f166586-db39-4281-bd12-8fed99dfe711",
   "metadata": {},
   "source": [
    "#### High performance trees\n",
    "\n",
    "Local trees in a tree sequence are not stored in a table, but iteratively constructed on the fly using the `arg.trees()` method. However, a tree object has a set of [fast array access methods](https://tskit.dev/tskit/docs/stable/python-api.html#array-access) to provide efficient access to tree-based information, such as the [parents of nodes](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.Tree.parent_array) in a tree, the [number of children](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.Tree.num_children_array) of tree nodes, or the [edge above each node](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.Tree.edge_array).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b41204d-9b9f-4bc6-b80a-fe13a44375fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "tree = arg.first()\n",
    "\n",
    "# Simple access to the parent of node 0 in the tree\n",
    "print(\"Parent of node 0 in the first tree is\", tree.parent(0))\n",
    "\n",
    "# High performance tree array access\n",
    "parents = tree.parent_array  # The parent ID of each node in this tree, or tskit.NULL (-1) if the node is not in the tree, or has no parent\n",
    "print(\"\\nParents of all the nodes:\\n\", parents)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c15973a2-754a-4404-8515-acfb5a5ffdad",
   "metadata": {},
   "source": [
    "The `parent_array` is a direct view into memory, which only needs slightly adjusting to encode the parents in the next tree. For large ARGs, this is much more efficient than making a new array for each local tree:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4155fb3-2278-4b9d-9857-3f76aa4c8017",
   "metadata": {},
   "outputs": [],
   "source": [
    "tree.next()\n",
    "print(\"Parents of all the nodes in the next tree:\\n\", parents)  # will be slightly different from the previous time we viewed this array"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "628108dd-1aee-4627-bc3a-dd54a945a23b",
   "metadata": {},
   "source": [
    "*This is the end of Workbook 1. Workbook 2 looks in detail about different sorts of ARG, and how they can be simulated.*"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
