
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Tskit for population genetics &#8212; Tree Sequence Tutorials</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=9c3e77be" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=1ae7504c"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'popgen';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Tskit for phylogenetics" href="phylogen.html" />
    <link rel="prev" title="Developing new tutorials" href="tutorial_development.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <script data-goatcounter="https://tskit.goatcounter.com/count" async src="//gc.zgo.at/count.js"></script>
</head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/tskit_logo.svg" class="logo__image only-light" alt="Tree Sequence Tutorials - Home"/>
    <script>document.write(`<img src="_static/tskit_logo.svg" class="logo__image only-dark" alt="Tree Sequence Tutorials - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="what_is.html">What is a tree sequence?</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Tree sequence basics</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="terminology_and_concepts.html">Terminology &amp; concepts</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting started with <strong class="program">tskit</strong></a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Statistics and Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="analysing_tree_sequences.html"><em>Analysing tree sequences</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="analysing_trees.html">Analysing trees</a></li>
<li class="toctree-l1"><a class="reference internal" href="incremental_algorithms.html"><em>Incremental algorithms</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="counting_topologies.html">Counting topologies</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallelization.html"><em>Parallelization</em></a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Further tskit tutorials</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="tables_and_editing.html">Tables and editing</a></li>
<li class="toctree-l1"><a class="reference internal" href="simplification.html"><em>Simplification</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="viz.html">Visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="metadata.html">Working with Metadata</a></li>
<li class="toctree-l1"><a class="reference internal" href="args.html">ARGs as tree sequences</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Simulation</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="simulation_overview.html">Tree sequences and simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="no_mutations.html">Do you really need mutations?</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="advanced_msprime.html">Advanced <strong class="program">msprime</strong> topics</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="demography.html">Demography</a></li>
<li class="toctree-l2"><a class="reference internal" href="bottlenecks.html">Instantaneous Bottlenecks</a></li>
<li class="toctree-l2"><a class="reference internal" href="introgression.html">Introgression</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="completing_forward_sims.html">Completing forwards simulations</a></li>
<li class="toctree-l1"><a class="reference internal" href="forward_sims.html">Building a forward simulator</a></li>
<li class="toctree-l1"><a class="reference internal" href="more_forward_sims.html"><em>Advanced forward simulations</em></a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Other languages</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="tskitr.html">Tskit and R</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Development</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="tutorial_development.html"><em>Developing new tutorials</em></a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Tskit for ...</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#"><code class="docutils literal notranslate"><span class="pre">Tskit</span></code> for population genetics</a></li>
<li class="toctree-l1"><a class="reference internal" href="phylogen.html"><code class="docutils literal notranslate"><span class="pre">Tskit</span></code> for phylogenetics</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/popgen.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Tskit for population genetics</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#population-genetic-simulation">Population genetic simulation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#population-genetic-inference">Population genetic inference</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#topological-analysis">Topological analysis</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#further-information">Further information</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="tskit-for-population-genetics">
<span id="sec-intro-popgen"></span><h1><code class="docutils literal notranslate"><span class="pre">Tskit</span></code> for population genetics<a class="headerlink" href="#tskit-for-population-genetics" title="Link to this heading">#</a></h1>
<p><a class="reference external" href="https://tskit.dev/tskit/docs/stable/introduction.html#sec-introduction" title="(in Project name not set)"><span class="xref std std-ref">Tskit</span></a>, the tree sequence toolkit, brings the power of
evolutionary trees to the field of population genetics. The
<a class="reference internal" href="what_is.html#sec-what-is"><span class="std std-ref">succinct tree sequence</span></a> format
is designed to store DNA sequences jointly with their ancestral history (the
“genetic genealogy” or <a class="reference internal" href="args.html#sec-args"><span class="std std-ref">ARG</span></a>). Storing population genetic data in this
form enables highly efficient computation and analysis.</p>
<p>The core <code class="docutils literal notranslate"><span class="pre">tskit</span></code> library provides methods for storing genetic data, a flexible
analysis framework, and APIs to build your own efficient population genetic algorithms.
Because of its speed and scalability, <code class="docutils literal notranslate"><span class="pre">tskit</span></code> is well-suited to interactive analysis of
large genomic datasets.</p>
<section id="population-genetic-simulation">
<h2>Population genetic simulation<a class="headerlink" href="#population-genetic-simulation" title="Link to this heading">#</a></h2>
<p>Several simulation tools output tree sequences. Below we use the
standard library for population genetic simulation models
(<a class="reference external" href="https://popsim-consortium.github.io/stdpopsim-docs/">stdpopsim</a>) to generate a model of
<em>Homo sapiens</em>, in which African, Eurasian,
and Asian populations combine to generate a mixed American population. We can use the
<a class="reference external" href="https://pypi.org/project/demesdraw/">demesdraw</a> package to plot a schematic of the
migrations and population size changes that define this model.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">stdpopsim</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">demesdraw</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">matplotlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">pyplot</span> <span class="k">as</span> <span class="n">plt</span>

<span class="n">species</span> <span class="o">=</span> <span class="n">stdpopsim</span><span class="o">.</span><span class="n">get_species</span><span class="p">(</span><span class="s2">&quot;HomSap&quot;</span><span class="p">)</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">species</span><span class="o">.</span><span class="n">get_demographic_model</span><span class="p">(</span><span class="s2">&quot;AmericanAdmixture_4B18&quot;</span><span class="p">)</span>

<span class="c1"># Plot a schematic of the model</span>
<span class="n">demesdraw</span><span class="o">.</span><span class="n">tubes</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">to_demes</span><span class="p">(),</span> <span class="n">ax</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">(),</span> <span class="n">seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">log_time</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/9bc6a9be31bba2451a4236cf5c275c68fe038e1ae38e9ca58afaffdfbd79efec.png" src="_images/9bc6a9be31bba2451a4236cf5c275c68fe038e1ae38e9ca58afaffdfbd79efec.png" />
</div>
</div>
<p>Genomic data in tree sequence format can be generated via the widely-used
<a class="reference external" href="https://tskit.dev/software/msprime.html">msprime</a> simulator. Here we simulate 1
megabase of genome sequence at the start of human chromosome 1 under this model,
together with its evolutionary history. We generate 16 diploid genomes: 4 from each of
the populations in the model. The DNA sequences and their ancestry are stored in a
succinct tree sequence named <code class="docutils literal notranslate"><span class="pre">ts</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">contig</span> <span class="o">=</span> <span class="n">species</span><span class="o">.</span><span class="n">get_contig</span><span class="p">(</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="n">model</span><span class="o">.</span><span class="n">mutation_rate</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="mi">1_000_000</span><span class="p">)</span>
<span class="n">samples</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;AFR&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span> <span class="s2">&quot;EUR&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span> <span class="s2">&quot;ASIA&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span> <span class="s2">&quot;ADMIX&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">}</span> <span class="c1"># 16 diploid samples</span>
<span class="n">engine</span> <span class="o">=</span> <span class="n">stdpopsim</span><span class="o">.</span><span class="n">get_engine</span><span class="p">(</span><span class="s2">&quot;msprime&quot;</span><span class="p">)</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">engine</span><span class="o">.</span><span class="n">simulate</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">contig</span><span class="p">,</span> <span class="n">samples</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">9</span><span class="p">)</span><span class="o">.</span><span class="n">trim</span><span class="p">()</span>  <span class="c1"># trim to first 20kb simulated</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Simulated a tree sequence of </span><span class="si">{</span><span class="n">ts</span><span class="o">.</span><span class="n">num_samples</span><span class="si">}</span><span class="s2"> haploid genomes:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ts</span><span class="o">.</span><span class="n">num_sites</span><span class="si">}</span><span class="s2"> variable sites over </span><span class="si">{</span><span class="n">ts</span><span class="o">.</span><span class="n">sequence_length</span><span class="si">}</span><span class="s2"> base pairs&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Simulated a tree sequence of 32 haploid genomes:
3580 variable sites over 1000000.0 base pairs
</pre></div>
</div>
</div>
</div>
<p>We can now inspect alleles and their frequencies at the variable sites we have simulated
along the genome:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">variants</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;Variable site </span><span class="si">{</span><span class="n">v</span><span class="o">.</span><span class="n">site</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> at position </span><span class="si">{</span><span class="n">v</span><span class="o">.</span><span class="n">site</span><span class="o">.</span><span class="n">position</span><span class="si">}</span><span class="s2"> has allele frequencies&quot;</span><span class="p">,</span>
        <span class="p">{</span><span class="n">state</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">freq</span><span class="si">:</span><span class="s2">.1%</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">state</span><span class="p">,</span> <span class="n">freq</span> <span class="ow">in</span> <span class="n">v</span><span class="o">.</span><span class="n">frequencies</span><span class="p">()</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
    <span class="p">)</span>
    <span class="k">if</span> <span class="n">v</span><span class="o">.</span><span class="n">site</span><span class="o">.</span><span class="n">id</span> <span class="o">&gt;</span> <span class="mi">4</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;...&quot;</span><span class="p">)</span>
        <span class="k">break</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Variable site 0 at position 1014.0 has allele frequencies {&#39;T&#39;: &#39;96.9%&#39;, &#39;G&#39;: &#39;3.1%&#39;}
Variable site 1 at position 2137.0 has allele frequencies {&#39;A&#39;: &#39;84.4%&#39;, &#39;T&#39;: &#39;15.6%&#39;}
Variable site 2 at position 2428.0 has allele frequencies {&#39;T&#39;: &#39;96.9%&#39;, &#39;A&#39;: &#39;3.1%&#39;}
Variable site 3 at position 2558.0 has allele frequencies {&#39;A&#39;: &#39;78.1%&#39;, &#39;G&#39;: &#39;21.9%&#39;}
Variable site 4 at position 3515.0 has allele frequencies {&#39;T&#39;: &#39;56.2%&#39;, &#39;G&#39;: &#39;43.8%&#39;}
Variable site 5 at position 3626.0 has allele frequencies {&#39;T&#39;: &#39;56.2%&#39;, &#39;C&#39;: &#39;43.8%&#39;}
...
</pre></div>
</div>
</div>
</div>
<p>Or we can efficiently grab the genotypes for each sampled genome</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sample ---&gt; &quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">u</span><span class="si">:</span><span class="s2">&gt;2</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">populations</span><span class="p">()</span> <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">population</span><span class="o">=</span><span class="n">p</span><span class="o">.</span><span class="n">id</span><span class="p">)]))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Population |&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">p</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">^</span><span class="si">{</span><span class="mi">3</span><span class="o">*</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">population</span><span class="o">=</span><span class="n">p</span><span class="o">.</span><span class="n">id</span><span class="p">)))</span><span class="o">-</span><span class="mi">1</span><span class="si">}}</span><span class="s2">|&quot;</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">populations</span><span class="p">()]))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;__________ |&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;_&quot;</span> <span class="o">*</span> <span class="p">(</span><span class="mi">3</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">population</span><span class="o">=</span><span class="n">p</span><span class="o">.</span><span class="n">id</span><span class="p">))</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;|&quot;</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">populations</span><span class="p">()]))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;  Position&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">variants</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">site</span><span class="o">.</span><span class="n">position</span><span class="p">)</span><span class="si">:</span><span class="s2">&gt;10</span><span class="si">}</span><span class="s2"> | &quot;</span><span class="p">,</span> <span class="s2">&quot;  &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">states</span><span class="p">()))</span>
    <span class="k">if</span> <span class="n">v</span><span class="o">.</span><span class="n">site</span><span class="o">.</span><span class="n">id</span> <span class="o">&gt;=</span> <span class="mi">30</span><span class="p">:</span> <span class="c1">#  Only show the first 30 sites, for brevity</span>
        <span class="k">break</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Sample ---&gt;   0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31
Population |           AFR          |          EUR          |         ASIA          |         ADMIX         |
__________ | _______________________|_______________________|_______________________|_______________________|
  Position
      1014 |  T  T  T  T  G  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T
      2137 |  A  T  A  A  A  A  A  A  A  A  A  A  A  A  T  T  T  A  A  A  A  A  A  A  A  A  A  T  A  A  A  A
      2428 |  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  A  T  T  T  T  T  T
      2558 |  A  G  A  A  G  A  A  A  A  A  A  A  A  A  G  G  G  A  A  A  A  A  A  A  A  G  A  G  A  A  A  A
      3515 |  T  T  T  T  T  G  T  T  G  G  G  T  G  T  T  T  T  G  G  T  G  G  G  T  T  T  T  T  G  G  G  G
      3626 |  T  T  T  T  T  C  T  T  C  C  C  T  C  T  T  T  T  C  C  T  C  C  C  T  T  T  T  T  C  C  C  C
      3780 |  G  G  G  G  G  G  G  G  G  G  G  G  G  G  G  G  G  G  G  G  G  A  G  G  G  G  G  G  G  G  G  G
      4255 |  A  A  A  A  A  A  A  A  A  A  A  C  A  C  A  A  A  A  A  C  A  A  A  C  C  A  C  A  A  A  A  A
      4440 |  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  C  A  A  A  A  A  A
      4543 |  A  T  A  A  T  T  A  A  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T  T
      4604 |  T  T  T  T  T  T  T  T  T  T  T  C  T  C  T  T  T  T  T  C  T  T  T  C  C  T  T  T  T  T  T  T
      6926 |  G  G  G  G  G  C  G  G  C  C  C  C  C  C  G  C  G  C  C  C  G  G  G  C  C  C  C  G  G  C  C  G
      7188 |  T  A  A  A  A  A  T  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A
      7425 |  G  C  C  C  C  G  G  C  G  G  G  G  G  G  C  G  C  G  G  G  C  C  C  G  G  G  G  C  C  G  G  C
      7828 |  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  T  C  C  C  C  C  C  C  C  C  C
      7883 |  A  A  A  A  A  C  A  A  C  C  C  C  C  C  A  C  A  C  C  C  A  A  A  A  C  A  C  A  A  C  C  A
      8835 |  G  G  G  G  G  T  G  G  T  T  T  T  T  T  G  T  G  T  T  T  G  G  G  G  T  G  T  G  G  T  T  G
      8844 |  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  A  C  C  C  C  C
      8861 |  A  A  A  A  A  A  A  A  A  A  A  G  A  G  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A
      8897 |  C  T  T  T  T  C  C  T  C  C  C  C  C  C  T  C  T  C  C  C  T  T  T  T  C  C  C  T  T  C  C  T
      9037 |  G  G  G  G  G  A  G  G  A  A  A  A  A  A  G  A  G  A  A  A  G  G  G  G  A  G  A  G  G  A  A  G
      9334 |  G  C  C  C  C  G  G  C  G  G  G  G  G  G  C  G  C  G  G  G  C  C  C  C  G  G  G  C  C  G  G  C
      9354 |  G  C  C  C  C  G  G  C  G  G  G  G  G  G  C  G  C  G  G  G  C  C  C  C  G  G  G  C  C  G  G  C
      9363 |  G  C  C  C  C  G  G  C  G  G  G  G  G  G  C  G  C  G  G  G  C  C  C  C  G  G  G  C  C  G  G  C
      9387 |  G  G  G  G  G  G  G  G  G  G  G  G  G  G  G  G  G  G  G  G  C  G  G  G  G  G  G  G  G  G  G  G
      9454 |  T  C  C  C  C  C  T  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C
      9547 |  A  A  A  A  A  A  A  A  A  A  G  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A  A
      9616 |  C  C  A  A  C  C  C  A  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C
      9972 |  C  G  G  G  G  C  C  G  C  C  C  C  C  C  G  C  G  C  C  C  C  C  C  G  C  C  C  G  C  C  C  C
     10011 |  C  C  C  C  C  G  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C  C
     10043 |  T  G  T  T  T  T  T  T  T  T  T  T  T  T  G  T  G  T  T  T  T  T  T  G  T  T  T  G  T  T  T  T
</pre></div>
</div>
</div>
</div>
<p>It is also possible to grab the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.haplotypes">haplotypes</a>
for specific samples (although this is slightly less efficient)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pop_id</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">samples</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">population</span><span class="o">=</span><span class="n">pop_id</span><span class="p">)</span>

<span class="k">for</span> <span class="n">sample_id</span><span class="p">,</span> <span class="n">haplotype</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span> <span class="n">ts</span><span class="o">.</span><span class="n">haplotypes</span><span class="p">(</span><span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">)):</span>
    <span class="n">h</span> <span class="o">=</span> <span class="s2">&quot;.&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">haplotype</span><span class="p">))</span>  <span class="c1"># Add a dot between letters, to clarify they are not  adjacent</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Sample </span><span class="si">{</span><span class="n">sample_id</span><span class="si">:</span><span class="s2">&lt;2</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">ts</span><span class="o">.</span><span class="n">population</span><span class="p">(</span><span class="n">pop_id</span><span class="p">)</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">^5</span><span class="si">}</span><span class="s2">): </span><span class="si">{</span><span class="n">h</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Sample 0  ( AFR ): T.A.T.A.T.T.G.A.A.A.T.G.T.G.C.A.G.C.A.C.G.G.G.G.G.T.A.C.C.C.T.C.G.A.G.A.C.C.C.T.C.T.C.G.G.A.G.T.T.G.C.A.T.T.T.A.G.T.C.G.A.A.T.T.C.T.T.A.A.C.G.T.A.T.C.C.A.G.A.C.G.T.A.T.T.C.T.C.A.G.T.G.C.G.A.G.G.G.T.A.T.C.A.G.T.G.A.T.A.C.A.A.A.C.G.C.G.G.C.G.T.T.T.A.C.T.T.T.C.G.C.T.A.T.C.T.C.G.G.C.A.A.G.T.C.G.C.G.T.G.A.T.G.G.A.T.T.C.T.A.T.A.G.G.T.G.A.A.C.A.G.A.G.T.A.A.C.T.T.A.A.C.G.A.T.C.C.T.C.T.T.G.G.A.C.C.T.G.A.A.G.A.C.C.G.A.A.A.T.T.A.G.C.T.A.T.C.A.G.C.A.T.G.T.A.G.G.A.A.T.C.T.C.C.A.G.G.A.G.C.G.G.G.C.C.C.G.G.A.A.T.C.A.C.A.G.C.G.G.G.T.G.T.T.T.G.G.A.G.C.G.G.A.T.T.C.G.T.T.T.A.C.T.C.C.G.C.T.A.C.G.T.A.C.T.G.A.T.C.C.A.C.G.A.A.T.G.G.C.A.A.C.A.A.G.A.G.T.G.T.G.C.G.C.C.A.T.T.G.G.T.G.G.C.G.A.C.G.T.T.C.C.C.C.G.G.C.T.G.T.G.T.A.G.T.C.C.G.T.G.G.T.T.A.C.T.G.T.C.A.T.A.A.G.G.G.G.A.T.A.C.A.A.G.T.G.T.G.A.A.G.A.C.C.T.A.C.C.A.C.A.G.A.G.A.A.G.G.A.T.C.T.T.A.A.A.T.C.A.A.C.T.T.C.A.C.T.G.T.T.G.A.A.A.C.T.A.G.T.A.C.G.G.T.G.A.A.T.C.A.C.G.C.G.G.C.C.T.T.T.G.A.C.A.G.A.G.G.T.G.C.A.C.C.C.T.A.G.G.T.C.T.C.A.C.G.A.A.C.T.A.C.C.A.A.C.A.T.G.C.A.T.G.C.C.C.T.C.T.A.C.T.G.G.C.T.A.G.C.G.C.T.A.T.G.C.T.G.G.C.G.T.A.T.C.C.G.G.G.A.T.G.A.G.G.T.T.A.G.C.T.C.T.T.C.T.T.T.T.A.A.A.A.A.C.A.A.C.C.T.G.C.T.C.G.C.T.C.G.T.C.T.C.A.A.A.G.G.C.T.C.A.T.G.T.G.A.A.T.T.C.A.C.G.C.T.C.A.C.T.T.A.G.C.A.T.T.G.T.T.G.G.T.T.G.T.T.C.T.G.G.C.C.T.A.A.A.A.A.T.C.C.C.C.T.T.G.A.C.G.A.T.A.C.T.A.A.A.T.T.C.A.T.C.C.T.G.G.A.A.T.A.C.T.A.C.A.A.G.T.C.A.G.A.A.C.T.A.G.T.C.G.T.C.G.G.G.C.G.T.A.T.G.C.C.T.C.T.C.T.A.T.C.G.G.A.C.T.G.A.A.A.T.G.C.A.C.A.C.G.C.C.A.A.T.A.A.T.C.G.T.C.A.A.T.A.C.A.T.A.G.C.G.C.T.G.A.G.C.C.T.C.A.A.T.A.T.T.A.A.A.C.G.C.A.C.T.C.A.A.T.C.T.A.G.C.A.C.T.C.G.A.T.T.A.G.C.C.G.T.G.T.C.G.G.A.G.T.T.T.A.T.G.G.A.G.T.T.A.G.G.T.C.G.A.T.T.A.T.G.A.C.T.G.G.G.A.G.A.A.G.G.C.T.A.G.G.G.C.C.G.T.T.G.T.C.C.T.T.C.A.A.C.A.T.C.G.T.A.A.A.G.C.T.G.G.A.A.G.A.C.C.C.T.C.C.T.C.A.T.C.A.A.C.C.A.G.C.A.G.G.G.A.C.C.G.C.A.C.T.T.C.G.A.T.C.A.C.T.C.T.A.C.G.A.G.G.C.G.T.G.A.A.C.C.G.T.C.C.T.C.C.T.G.C.C.T.G.A.C.C.T.T.A.T.G.C.A.A.T.G.C.C.T.G.T.C.T.T.T.G.G.C.T.G.T.T.A.C.G.A.C.A.A.C.A.C.A.T.G.C.A.G.A.A.C.C.C.T.C.G.C.T.A.T.C.A.A.T.T.T.A.T.A.C.A.T.G.T.T.T.T.G.A.T.A.A.T.G.A.G.C.T.A.C.T.T.T.C.C.A.T.T.A.G.G.T.C.A.G.A.C.G.C.T.T.A.G.T.A.A.C.T.T.G.G.T.G.A.T.A.T.T.G.A.C.C.A.T.C.A.C.T.A.A.C.G.C.C.T.T.A.G.T.A.C.A.G.T.A.G.T.T.G.G.T.C.A.C.T.C.C.G.G.T.C.C.A.A.A.T.C.A.A.T.C.T.T.C.A.T.C.C.G.A.G.C.C.A.G.C.A.C.A.C.T.C.C.G.C.G.T.C.T.T.G.A.A.C.T.T.A.A.C.G.A.T.G.A.C.G.T.C.G.G.T.A.G.G.A.C.C.C.C.C.A.G.C.T.C.C.G.T.T.A.T.G.T.C.A.G.A.C.T.G.C.T.C.G.G.T.C.A.G.G.C.C.A.A.A.T.G.G.A.C.T.A.C.C.T.T.T.T.A.C.C.A.A.C.G.A.A.C.A.C.A.G.G.G.G.C.A.T.T.G.G.G.A.G.C.C.C.T.T.T.A.G.A.C.A.T.C.G.T.G.G.C.G.C.C.C.T.G.C.C.C.A.C.C.G.T.T.T.C.C.T.C.T.A.C.T.T.G.A.G.G.T.G.C.A.A.C.T.A.A.C.G.G.G.T.T.C.G.G.G.G.C.T.A.T.G.C.G.A.C.C.T.C.G.G.C.T.A.G.A.A.A.T.A.T.C.A.T.T.G.C.C.T.G.A.T.C.T.T.C.A.G.C.C.A.A.A.A.G.T.G.A.A.C.G.G.G.C.A.A.C.T.C.C.T.A.T.T.G.A.A.T.C.G.T.G.A.G.G.A.A.C.C.C.C.T.G.G.G.A.C.A.G.A.T.C.A.A.A.A.T.G.T.A.T.T.C.A.A.T.C.A.C.T.C.G.T.G.A.G.T.G.C.G.C.A.C.C.T.T.C.T.C.T.T.A.C.G.C.C.G.A.A.T.T.A.T.C.A.T.G.A.A.G.A.T.G.C.C.A.T.A.G.G.G.G.C.A.C.T.T.G.T.T.A.G.G.G.G.T.G.G.G.A.T.G.C.A.C.A.C.A.T.G.C.G.C.T.G.A.T.A.T.A.A.A.A.A.T.T.A.C.G.A.C.A.G.A.C.A.G.A.C.G.T.C.C.A.G.G.A.A.G.C.T.A.T.A.C.G.A.G.G.G.C.T.G.C.G.T.C.A.T.G.C.A.T.C.C.T.A.G.G.G.G.G.G.G.C.T.A.C.T.A.C.C.A.C.A.T.G.A.T.T.C.C.A.G.C.A.A.A.T.C.G.C.T.A.T.T.G.C.T.C.G.C.G.C.C.A.T.G.T.G.C.T.G.A.T.G.C.G.G.C.C.A.T.G.C.G.T.G.G.A.T.T.A.C.G.A.T.T.A.T.T.C.G.C.A.T.G.G.T.G.T.G.A.T.A.T.A.C.A.A.G.G.T.T.A.G.T.A.G.C.A.A.T.G.T.C.T.A.T.T.G.T.A.G.T.T.A.A.A.C.A.G.G.C.G.T.C.A.C.G.A.C.C.T.A.G.C.G.A.C.T.C.C.C.C.G.A.G.T.A.C.G.G.T.T.G.G.A.C.T.C.G.A.T.G.A.C.C.A.C.T.T.T.T.C.C.A.A.A.A.G.T.A.T.C.G.A.G.A.G.A.G.G.A.C.C.G.T.T.G.A.T.C.G.A.A.C.T.C.G.C.T.A.T.G.T.C.G.A.C.A.A.G.C.T.G.G.G.G.C.C.A.T.G.G.G.C.C.C.T.T.T.A.T.C.G.T.A.A.A.G.G.A.G.T.T.T.C.A.T.A.A.C.T.C.C.C.A.G.C.C.A.C.A.A.C.T.G.G.T.A.C.A.C.T.C.T.A.T.C.T.C.A.C.G.A.T.C.C.G.A.A.C.A.T.C.G.T.T.T.G.G.A.T.A.G.C.G.C.G.C.T.C.C.G.G.C.G.T.A.G.T.C.G.C.A.G.C.A.C.G.A.A.G.T.A.G.T.A.T.A.T.T.A.C.A.T.G.A.T.C.G.A.T.A.G.T.T.T.T.C.T.C.T.C.C.A.A.G.T.C.C.A.T.G.T.A.G.C.T.G.G.G.A.A.A.C.C.T.A.C.T.T.A.C.G.A.C.T.T.C.G.G.C.C.A.A.T.G.C.T.T.C.T.T.C.T.G.G.T.C.A.C.T.A.A.A.G.G.G.G.A.C.C.A.G.T.A.C.G.T.G.G.C.G.C.G.T.C.G.A.C.G.A.C.T.G.T.G.T.G.C.A.C.G.C.G.G.A.T.T.C.T.T.C.T.T.T.C.A.G.A.A.G.T.A.A.G.C.T.G.G.A.G.T.T.C.A.C.G.A.T.G.A.A.C.G.G.G.A.T.G.G.C.G.T.T.C.G.C.T.T.C.C.T.T.C.A.A.T.G.G.C.T.A.T.A.A.G.A.T.G.A.A.C.C.A.A.T.C.C.A.T.T.G.G.G.T.A.T.A.A.C.C.T.T.T.G.G.C.G.C.T.C.T.C.G.C.G.T.A.C.C.G.A.G.G.G.A.A.C.T.T.C.T.A.A.A.A.C.C.T.A.G.T.A.T.C.G.C.C.A.C.C.A.T.C.C.G.A.T.C.C.T.C.G.C.A.T.A.C.T.C.A.A.A.A.A.A.G.A.C.A.T.G.A.C.A.A.C.T.C.C.G.A.G.A.A.C.T.A.G.G.A.T.G.A.T.A.G.T.C.C.C.T.G.G.A.G.G.G.A.A.C.T.A.C.G.G.C.T.G.A.T.T.C.G.A.G.T.A.T.G.C.G.T.A.C.A.C.C.G.A.G.G.T.T.G.G.T.C.A.A.A.G.G.T.G.T.T.T.C.A.T.A.G.T.G.C.T.A.A.T.A.C.C.C.G.A.T.C.A.A.A.T.A.A.T.C.C.T.G.C.G.G.T.C.A.G.A.C.C.A.T.G.C.T.G.T.G.A.C.A.A.A.G.T.A.T.G.A.T.T.T.C.A.T.A.T.C.G.C.G.T.T.G.C.G.T.A.T.T.C.C.A.G.G.A.C.T.T.C.A.C.T.C.A.T.G.A.A.C.C.T.C.C.A.A.T.C.C.G.A.C.T.T.A.A.C.G.C.G.T.T.A.A.C.C.C.T.C.T.A.G.G.C.C.A.G.T.A.A.C.G.C.A.A.C.C.A.G.A.T.C.A.C.G.A.A.A.G.G.A.T.T.G.G.T.T.G.A.C.C.G.A.A.C.A.A.A.T.A.C.G.C.G.G.T.G.G.T.T.C.T.C.C.T.A.A.A.A.T.G.C.A.A.A.A.T.T.A.A.G.A.G.G.T.T.T.G.C.T.A.T.A.T.G.C.C.A.G.T.A.C.G.C.G.C.A.T.C.T.G.A.G.C.G.C.G.T.A.G.C.A.A.T.C.G.C.G.C.A.G.T.C.A.A.A.G.C.C.A.G.A.C.G.G.G.G.G.A.T.G.A.A.A.C.A.G.A.G.T.C.C.A.C.G.T.T.A.C.T.T.T.C.C.C.A.C.G.A.T.C.T.T.A.T.T.A.G.A.T.A.T.A.G.G.A.G.G.A.G.G.A.C.A.T.T.C.G.C.T.G.G.T.T.A.T.C.G.A.A.G.C.T.A.T.C.A.A.A.G.C.G.G.G.T.C.G.G.T.G.T.C.C.G.G.T.T.A.A.C.T.C.A.T.A.C.A.G.C.G.C.G.C.T.G.C.C.T.G.G.T.C.C.C.A.G.C.A.T.T.T.C.T.C.A.A.C.G.T.A.G.G.C.G.C.G.T.C.C.A.T.G.C.A.A.A.T.C.T.C.C.T.T.A.T.G.A.T.T.G.G.T.G.C.T.A.C.A.A.G.T.T.C.A.A.G.A.G.T.G.T.G.G.T.C.C.T.A.C.C.T.C.A.T.T.G.C.G.A.T.T.G.A.G.A.C.T.G.C.A.C.T.T.G.C.A.C.A.T.T.T.T.T.T.T.A.G.T.T.A.C.T.G.T.G.C.G.C.G.A.G.A.G.T.C.A.A.A.G.A.T.A.A.C.G.C.C.C.G.T.C.A.A.G.C.T.C.T.A.T.G.C.T.C.T.T.C.G.C.C.G.G.G.C.G.T.G.A.T.T.C.G.C.T.G.A.A.G.G.G.G.C.C.C.T.A.C.G.A.A.G.A.A.C.T.G.C.A.T.T.C.T.A.C.C.T.T.T.C.C.T.G.A.T.T.A.T.A.T.G.T.G.C.A.T.C.T.T.C.G.C.T.A.A.G.A.G.C.G.C.C.G.T.C.G.C.C.C.A.G.G.A.T.C.A.A.A.C.C.T.G.T.G.A.G.C.C.C.A.C.A.G.C.T.C.C.C.A.A.G.G.G.A.C.T.T.A.C.A.T.T.A.T.G.T.G.G.T.A.G.A.G.T.C.T.C.G.C.G.G.C.C.T.T.T.A.T.C.T.T.T.T.T.A.A.C.C.C.G.C.A.G.G.G.G.T.T.G.C.C.C.T.T.G.C.C.A.G.A.C.G.C.A.A.T.G.G.C.G.T.C.C.C.C.G.G.T.T.G.A.A.G.T.G.A.T.C.A.T.T.A.A.C.T.C.A.T.T.G.G.T.G.C.C.T.T.T.C.C.A.C.G.G.A.C.C.T.G.G.G.T.G.A.C.G.C.C.C.C.T.C.C.A.C.G.T.G.T.A.C.G.T.C.G.C.G.A.C.T.C.A.A.C.G.G.C.G.C.A.T.T.C.A.T.G.A.C.G.G.T.G.T.T.C.G.A.G.T.G.T.A.G.A.C.C.C.G.T.G.A.A.G.T.T.A.G.A.G.A.G.T.G.C.G.C.G.C.C.A.T.A.A.T.G.A.A.G.T.A.T.G.C.A.C.A.A.C.G.G.T.C.T.T.A.T.G.A.T.T.A.A.A.C.T.G.T.G.T.A.G.C.A.G.G.C.G.C.T.T.G.G.A.T.G.A.G.A.C.T.T.T.C.G.A.A.C.C.G.C.C.G.T.A.C.C.T.G.A.C.G.C.C.C.G.C.A.C.C.C.A.A.G.A.A.A.T.T.A.T.G.T.T.C.A.A.T.C.T.T.C.G.G.A.T.C.A.A.A.T.A.A.T.G.C.T.A.T.A.A.C.A.G.A.T.G.G.A.A.T.A.C.C.G.T.C.C.T.G.G.G.T.G.A.A.A.C.G.A.T.G.C.G.A.C.T.G.T.T.G.T.C.C.A.C.T.G.C.A.T.A.C.A.G.C.A.T.A.G.G.C.T.C.G.A.T.A.G.C.T.T.C.C.A.A.A.G.A.T.G.C.T.A.C.A.G.A.G.C.T.T.G.A.A.C.C.G.G.T.C.A.C.T.T.A.C.A.G.G.C.G.G.C.G.C.A.T.A.G.A.C.T.T.A.A.G.T.A.T.A.A.C.C.A.G.C.A.A.G.T.A.C.C.G.C.T.A.A.G.G.C.A.C.G.G.G.C.C.T.G.C.A.A.A.A.C.C.C.T.G.T.G.G.C.T.A.C.A.C.A.A.C.A.A.A.G.G.T.T.A.T.A.G.T.G.A.T.T.G.C
Sample 1  ( AFR ): T.T.T.G.T.T.G.A.A.T.T.G.A.C.C.A.G.C.A.T.G.C.C.C.G.C.A.C.G.C.G.A.G.A.G.A.C.T.C.T.C.T.A.C.T.T.A.G.G.C.G.G.C.A.T.A.T.C.T.G.A.A.T.T.C.C.A.A.C.T.T.T.C.T.T.C.A.C.A.C.G.T.C.T.T.C.T.C.T.A.T.G.T.G.C.G.G.G.T.A.T.C.T.G.G.T.G.G.A.C.A.G.A.C.G.C.G.G.C.G.T.T.G.A.C.T.T.C.C.G.C.T.A.T.C.T.A.G.C.C.A.T.G.T.C.G.C.G.T.G.A.T.G.G.A.T.T.C.G.A.G.A.G.G.T.G.A.A.C.A.G.A.G.T.A.A.C.G.T.A.A.C.A.A.T.C.C.T.C.T.T.G.G.G.G.A.T.C.C.A.G.G.C.C.C.G.T.C.G.G.G.T.C.G.A.T.C.T.G.C.T.T.G.G.C.G.G.A.A.T.C.T.G.C.C.C.G.A.G.A.G.T.C.G.T.C.G.A.A.T.A.C.C.C.A.G.T.G.G.G.A.A.T.T.T.A.C.T.G.C.G.A.A.G.T.C.G.T.T.C.G.G.C.A.G.A.C.T.C.T.G.C.T.C.T.G.G.C.C.C.A.C.G.A.A.T.C.G.C.A.A.A.A.A.G.A.G.T.G.T.G.C.G.C.C.A.T.T.G.G.T.G.G.C.G.A.C.T.T.T.C.A.C.C.G.G.A.T.A.T.G.T.A.G.T.C.A.G.T.G.G.T.T.T.T.T.G.T.A.A.T.G.G.G.C.G.G.A.T.A.C.A.A.G.T.G.G.G.A.T.G.A.G.A.C.A.C.C.A.C.C.G.A.G.G.A.C.G.A.T.C.T.T.A.A.A.T.A.A.A.C.T.T.C.C.C.T.G.T.T.G.A.A.A.G.T.G.G.T.G.C.G.G.G.G.T.A.T.C.A.C.G.C.G.T.C.C.C.T.T.A.A.T.A.G.A.G.A.T.G.T.A.C.C.C.A.A.G.G.T.A.T.C.A.C.G.A.A.A.T.A.C.C.A.A.T.A.T.T.C.A.T.G.G.C.G.T.G.T.A.G.T.T.T.C.T.A.G.C.G.C.G.G.G.G.C.T.G.T.C.G.T.A.C.C.C.G.G.G.G.C.A.A.C.G.T.T.A.G.C.T.C.T.T.C.T.T.T.T.T.A.A.G.A.C.A.A.C.C.C.T.C.T.C.G.C.T.C.G.T.C.T.C.T.T.A.G.C.A.C.C.A.T.G.T.G.A.A.T.T.C.A.A.G.C.T.C.G.T.T.T.A.G.C.G.T.T.T.A.T.G.G.G.T.G.T.T.C.T.G.G.G.C.T.A.A.A.A.G.T.C.C.C.C.T.T.G.A.C.T.A.T.A.C.A.A.A.A.T.C.C.A.T.C.C.G.G.G.A.A.T.A.T.T.C.C.C.A.G.T.C.A.G.T.A.C.G.C.G.T.C.C.C.C.G.C.C.C.G.G.A.G.A.C.C.G.C.G.C.T.C.A.G.G.A.A.C.T.G.G.T.C.A.G.T.A.C.A.C.G.C.C.A.A.T.A.A.T.C.G.T.C.C.A.T.A.C.A.T.A.G.C.G.C.T.G.A.G.T.C.G.C.A.A.T.A.T.A.A.A.A.T.T.C.G.C.T.C.A.A.T.C.T.A.G.C.A.C.T.C.G.A.C.T.A.G.G.C.G.T.G.T.T.A.G.A.G.T.T.T.A.A.G.G.A.T.T.T.A.G.G.T.C.G.A.G.T.A.T.G.A.G.T.T.A.C.T.A.C.A.A.G.T.T.A.T.G.C.G.C.C.T.G.T.T.C.T.T.C.C.G.T.C.A.T.C.G.T.A.A.A.G.C.T.G.G.A.A.G.A.C.C.C.T.C.C.T.C.A.T.C.A.A.C.C.A.G.C.A.G.G.G.A.C.C.G.C.A.C.T.T.C.G.A.T.C.A.C.T.C.G.A.C.G.A.G.G.C.G.T.G.A.A.C.C.G.T.C.C.T.C.C.T.G.C.C.T.G.A.C.C.T.T.A.T.G.C.A.A.T.G.A.G.T.G.T.C.A.G.T.T.T.A.T.G.T.T.A.C.G.A.C.A.A.C.A.C.A.T.G.C.A.G.A.A.C.C.C.T.C.G.C.T.A.T.G.G.T.T.T.T.A.T.A.G.A.T.G.G.T.A.T.A.T.A.T.A.C.G.A.A.T.T.A.C.G.T.G.C.T.G.T.T.A.G.G.T.C.A.G.A.C.G.C.T.T.A.G.T.A.A.C.T.T.G.G.T.G.G.T.T.T.T.G.A.C.G.A.T.C.A.C.A.C.A.C.G.T.C.T.T.G.G.A.A.C.A.G.T.G.G.T.T.G.G.T.C.A.C.T.C.C.G.C.T.C.T.A.A.T.G.T.A.A.T.C.A.T.C.A.T.T.C.G.G.G.C.C.A.G.C.C.C.A.C.T.C.C.G.C.G.T.C.G.T.G.A.A.C.T.A.A.A.T.G.A.T.G.C.C.G.G.C.G.G.T.A.G.G.C.C.C.C.C.C.A.G.C.T.C.C.G.T.T.A.T.G.G.C.A.G.A.C.T.G.C.T.C.G.G.T.C.A.G.G.C.C.C.A.A.T.G.G.A.C.T.A.C.C.T.T.T.T.A.C.T.A.A.C.G.A.T.C.C.C.A.G.G.G.G.C.A.T.T.G.G.G.A.G.C.C.C.C.T.C.A.G.A.C.A.T.C.G.T.G.A.C.G.C.G.C.C.G.C.C.C.A.C.C.G.T.T.T.C.C.T.C.T.A.C.T.T.G.A.G.G.T.G.C.A.A.C.T.A.A.C.G.G.G.T.T.C.G.G.G.G.C.T.A.A.G.C.G.A.G.C.T.C.G.G.C.T.A.C.T.A.A.T.A.T.C.A.T.T.G.C.C.G.G.A.T.C.A.T.C.A.G.C.C.C.A.A.A.G.T.G.A.A.C.G.G.G.C.A.A.T.T.C.C.T.A.T.T.G.A.A.T.A.A.T.G.A.G.G.A.A.C.C.C.C.T.G.G.G.A.C.A.G.A.T.C.C.A.A.A.T.G.T.A.G.T.C.A.A.T.C.A.C.T.C.G.T.G.A.G.A.G.C.A.T.A.C.C.G.G.C.T.C.T.T.A.G.G.C.C.G.A.A.T.T.A.T.C.C.T.C.G.A.G.A.A.C.A.C.C.T.A.G.G.C.A.G.A.G.T.C.C.G.T.G.C.A.T.A.A.A.C.T.A.G.G.C.T.G.C.C.A.C.G.C.G.C.T.G.G.T.A.T.G.A.A.A.A.T.A.A.C.G.A.C.G.G.A.C.A.A.G.C.C.T.C.C.A.G.G.G.C.G.C.T.C.C.T.C.G.A.T.T.G.A.T.C.A.G.A.C.C.G.G.A.A.G.T.C.G.A.G.T.G.C.T.G.G.C.T.A.C.T.A.C.A.A.A.C.T.G.A.T.T.A.C.G.G.C.G.G.G.A.C.G.A.T.A.T.T.G.C.T.C.G.C.G.C.C.A.T.G.T.G.C.T.G.A.T.G.C.G.G.C.T.T.A.G.T.G.C.G.G.A.T.A.A.G.G.G.T.T.A.C.T.A.A.C.A.C.A.C.T.A.A.T.A.A.A.T.A.C.G.A.G.G.T.T.A.G.T.C.T.A.A.A.T.G.T.C.T.G.A.T.A.T.A.G.T.A.A.A.A.C.C.C.G.C.A.G.C.A.G.G.A.C.C.T.A.G.G.G.A.C.T.C.C.C.C.G.A.G.T.A.C.G.G.T.T.G.G.A.C.T.C.G.A.T.G.A.A.C.A.C.T.T.T.T.C.C.A.A.A.G.G.T.T.T.C.G.A.G.A.G.C.G.G.A.C.C.G.T.T.G.A.T.C.G.A.A.G.T.C.G.C.T.A.T.C.T.C.G.T.C.A.A.G.C.A.G.G.G.G.G.C.C.T.G.G.G.T.C.C.T.T.T.A.T.T.G.T.A.A.A.G.G.A.G.T.T.T.C.A.T.A.A.C.T.G.G.C.A.G.C.C.A.C.A.T.C.T.T.G.T.A.C.A.C.T.C.T.A.T.C.A.C.A.C.G.A.G.C.C.A.A.A.C.A.T.C.G.T.T.T.G.G.A.T.A.G.C.G.C.G.C.T.C.C.G.G.G.C.T.A.G.T.C.G.C.G.G.C.G.C.A.A.A.T.C.G.G.T.A.T.A.A.T.T.C.A.T.G.A.T.C.G.A.T.A.G.G.T.T.T.G.T.C.T.A.G.A.A.G.T.C.C.A.T.G.T.A.G.C.T.G.G.T.T.T.G.A.G.T.A.C.T.T.A.C.G.A.G.T.T.C.C.G.G.C.A.A.A.G.C.T.T.C.T.T.C.T.G.G.T.C.A.C.T.A.A.A.G.C.G.A.A.C.C.A.G.T.A.C.G.T.G.G.C.G.A.T.T.A.T.A.C.C.A.C.T.G.T.G.T.G.C.A.C.C.C.G.C.A.C.C.C.G.A.G.T.T.T.C.A.T.T.T.C.T.A.T.G.C.T.G.C.A.G.T.T.C.A.C.G.A.T.C.A.A.C.G.G.G.A.T.G.G.C.G.C.T.C.T.G.A.T.C.C.A.T.A.A.C.T.G.G.C.T.A.T.A.A.G.A.T.G.A.A.C.C.A.A.T.C.C.A.T.T.G.G.G.T.A.C.A.A.C.C.T.G.T.G.C.C.G.T.T.C.T.T.G.C.A.C.C.G.A.G.A.A.G.C.A.A.C.A.T.C.T.A.A.A.A.C.C.T.A.G.T.A.T.C.G.A.C.A.C.T.A.T.C.C.G.A.T.C.C.T.C.G.A.A.T.A.C.T.C.A.A.A.A.A.A.G.G.C.A.T.G.A.C.A.A.C.T.C.C.G.A.G.A.C.C.T.A.G.G.A.T.G.A.T.A.C.T.C.C.C.G.G.G.A.G.G.G.A.A.A.A.A.T.G.G.C.A.C.G.T.T.A.G.A.G.T.A.T.C.G.T.A.A.T.C.C.C.C.A.T.C.A.G.C.G.T.G.A.A.T.A.A.A.C.G.G.T.C.C.A.A.A.T.C.T.T.G.T.T.G.T.A.T.C.A.T.C.C.A.A.T.T.T.A.C.C.T.T.C.T.G.T.C.A.G.A.C.T.A.G.A.C.T.G.T.A.G.T.A.A.A.T.T.A.G.A.A.A.A.T.T.T.T.T.T.A.G.A.G.C.T.G.C.G.G.T.T.G.C.T.C.C.G.A.C.T.T.C.T.C.T.A.T.A.T.G.A.A.C.A.C.C.G.G.G.C.T.A.G.C.T.T.A.T.C.T.C.A.A.C.C.C.G.T.C.G.C.G.G.C.C.C.C.A.C.T.A.T.C.T.C.C.C.C.T.G.G.C.T.C.T.A.A.A.G.A.A.G.A.A.C.G.A.T.G.A.A.A.C.A.A.A.T.A.A.A.T.C.C.G.C.G.G.T.A.G.T.T.C.T.C.C.T.A.A.A.A.T.C.C.C.A.G.A.T.T.A.A.G.A.G.G.T.C.T.G.C.T.A.T.A.T.G.C.C.A.G.T.C.C.C.C.A.C.T.C.C.G.A.A.T.C.C.G.C.T.G.T.C.C.G.C.T.G.C.G.T.G.A.G.T.T.C.T.G.C.C.A.G.C.C.A.G.G.G.G.A.T.G.A.A.A.G.C.T.A.A.T.C.G.A.G.G.T.T.C.C.G.C.T.C.C.T.T.G.T.T.T.G.A.T.A.A.T.T.C.T.T.A.G.T.G.G.A.A.G.A.C.A.A.A.C.T.C.G.C.C.T.T.A.T.T.A.T.T.G.A.A.G.C.T.C.A.T.A.T.A.G.C.T.G.G.T.C.T.G.T.G.T.C.G.G.G.T.T.A.A.C.T.C.A.A.C.C.A.G.C.G.C.G.C.T.G.C.C.T.G.G.T.A.T.C.G.G.C.A.T.A.T.T.T.T.A.A.C.G.T.A.G.G.T.A.A.A.G.T.C.A.T.G.A.A.T.G.T.C.T.C.G.C.T.A.T.G.G.T.T.T.C.A.A.C.T.A.C.A.A.G.T.T.G.A.G.G.A.A.T.G.G.G.G.A.C.C.T.A.C.G.T.C.C.T.T.G.C.G.A.T.T.G.A.G.A.C.T.T.C.A.G.C.T.G.C.A.C.T.C.T.T.T.T.T.T.A.G.C.T.A.A.T.G.T.G.C.G.C.G.A.T.A.G.T.A.A.G.A.G.T.T.A.A.C.G.C.C.C.G.T.C.A.A.G.C.T.C.T.A.T.G.C.T.C.T.T.C.G.C.G.G.G.G.C.G.T.G.A.T.T.C.G.C.T.G.A.A.G.G.G.G.C.C.C.C.A.C.G.A.A.G.A.C.C.T.G.C.A.T.T.C.T.C.C.C.T.T.T.C.C.T.G.A.T.T.A.T.A.T.G.T.G.C.A.T.C.T.T.C.G.C.T.A.A.G.A.G.C.G.C.C.G.T.C.G.C.C.G.A.G.G.A.T.C.A.A.A.C.C.T.G.T.G.A.G.C.C.C.A.C.A.G.C.T.C.C.C.A.A.G.G.G.A.C.T.T.A.C.A.T.T.A.T.G.T.G.G.T.A.G.A.G.T.C.T.C.G.C.G.G.C.C.T.T.T.A.T.C.T.T.T.G.T.A.A.C.C.C.G.C.A.G.G.G.G.T.T.A.C.C.C.T.T.G.C.C.A.T.A.C.G.C.A.A.C.G.A.A.G.T.A.C.C.C.G.G.T.A.T.T.A.G.A.T.A.T.C.A.T.C.A.A.C.G.T.C.T.T.G.A.T.G.C.C.T.T.T.T.C.A.C.G.G.T.C.C.T.G.G.G.T.G.A.A.G.C.C.C.A.T.C.C.A.C.A.T.G.T.A.T.G.T.C.A.A.G.G.C.T.C.A.A.C.G.C.C.G.C.T.T.T.C.C.T.G.A.A.G.G.A.T.T.T.C.T.A.G.T.T.T.A.T.G.C.C.C.G.T.G.A.A.G.T.T.A.G.A.G.A.G.G.C.C.G.T.G.T.G.A.G.A.A.T.G.A.A.G.T.A.T.T.A.A.A.C.A.C.G.G.T.C.T.T.C.T.G.A.T.T.A.T.A.C.A.G.A.G.T.A.G.C.A.G.T.C.A.C.G.T.G.G.A.T.A.A.T.C.C.A.T.T.C.C.A.A.C.C.G.C.G.G.T.A.C.G.T.G.A.C.G.C.C.C.G.C.A.G.C.C.A.A.C.A.A.A.T.T.A.C.G.T.T.C.C.A.G.C.T.T.C.G.T.A.A.C.A.C.A.G.A.A.T.G.C.T.A.T.A.A.C.A.C.A.T.C.G.C.A.T.A.C.C.G.C.C.C.T.G.G.G.T.G.A.A.A.C.G.A.G.G.C.G.A.C.G.G.T.T.G.T.C.G.A.A.T.G.A.C.T.A.C.A.T.C.A.T.A.G.G.C.T.C.C.A.T.A.G.G.T.T.C.C.A.C.G.G.T.T.G.C.C.C.C.G.C.A.G.C.C.T.T.T.C.C.C.C.G.T.C.T.C.C.A.A.A.T.G.A.C.T.C.G.G.C.A.A.T.G.A.C.T.T.A.A.G.T.A.T.A.A.C.C.A.G.C.A.A.G.T.A.C.C.G.A.T.A.A.G.G.C.A.C.G.G.G.C.C.T.G.C.A.A.A.A.C.C.C.T.G.T.A.G.C.G.C.C.G.C.A.T.G.A.C.A.A.A.T.G.C.G.C.C.T.G.A.G.T.T.G
Sample 2  ( AFR ): T.A.T.A.T.T.G.A.A.A.T.G.A.C.C.A.G.C.A.T.G.C.C.C.G.C.A.A.G.C.T.A.G.A.G.A.C.C.C.G.G.A.A.C.T.T.A.G.G.C.G.G.C.A.T.A.T.C.C.G.A.A.T.T.C.C.A.G.C.T.T.T.C.A.T.C.A.C.A.C.G.T.C.T.T.C.T.C.T.G.T.C.C.T.C.G.C.G.T.A.A.C.T.G.G.T.G.G.A.C.C.G.G.A.G.C.G.T.T.A.T.A.G.A.A.C.T.T.C.G.C.T.A.C.C.T.A.G.C.C.A.T.T.T.C.A.C.C.A.G.T.T.G.G.A.A.T.C.G.A.T.T.G.G.C.A.T.A.C.C.G.T.T.C.C.C.T.G.C.A.A.C.G.T.G.G.A.C.T.T.T.A.G.A.C.C.T.G.A.A.T.A.G.C.G.A.A.A.T.T.A.G.C.T.A.T.C.A.G.C.A.T.G.T.A.G.G.A.A.T.C.C.G.C.C.G.G.A.G.A.G.G.G.C.C.C.G.G.A.A.T.C.A.C.A.G.C.G.G.G.T.G.T.T.T.G.G.A.G.C.G.G.A.T.T.C.G.T.T.T.A.G.T.A.C.G.C.T.A.T.G.C.A.C.T.A.A.C.C.C.A.C.G.A.A.T.G.A.C.A.A.A.A.A.G.A.G.T.A.G.G.C.C.C.C.T.T.T.G.C.T.T.G.T.G.A.C.T.T.G.C.A.A.C.G.G.A.T.A.G.G.T.A.T.T.A.A.A.T.G.G.C.T.T.T.T.G.T.A.A.C.A.G.G.C.G.G.A.T.T.C.A.T.G.T.C.G.G.A.A.A.C.C.A.T.A.A.C.T.G.A.C.T.G.A.A.C.G.A.A.T.A.T.A.A.C.A.C.A.T.C.C.T.A.C.C.C.T.G.T.G.G.T.A.G.A.A.G.A.G.C.G.T.T.T.A.G.T.C.T.C.G.A.T.T.C.C.T.T.T.G.A.T.A.G.A.G.G.A.G.C.T.C.C.A.T.A.G.G.T.A.A.C.G.A.G.T.G.A.T.A.C.C.A.A.C.A.T.G.C.A.T.G.C.C.C.T.C.T.A.C.T.G.G.A.T.A.A.C.G.C.T.A.G.T.C.T.G.T.C.G.T.A.T.C.C.G.G.G.A.C.G.A.C.G.T.T.T.A.C.T.C.T.T.C.T.T.T.C.T.A.A.A.A.C.A.A.C.C.T.G.G.T.C.G.C.C.G.A.C.C.T.C.T.T.A.G.C.A.C.C.A.T.G.T.G.A.A.T.T.C.A.A.G.C.T.C.G.C.T.T.A.G.C.G.T.T.T.A.T.G.G.T.T.G.T.T.C.T.G.G.C.C.T.A.A.A.A.G.T.C.C.C.C.T.T.G.A.C.G.A.T.A.C.A.A.A.A.T.T.C.A.T.C.C.T.G.G.A.A.T.A.C.T.A.C.A.A.G.T.C.A.G.A.A.C.T.A.T.T.C.G.T.C.G.G.G.C.G.T.A.G.A.C.C.G.C.G.C.T.C.A.G.G.A.A.C.T.G.G.T.C.A.G.T.A.C.A.T.G.C.C.A.G.T.A.A.T.A.G.G.C.C.A.G.A.C.A.C.A.G.C.C.C.C.G.A.C.C.T.G.C.A.A.G.A.T.T.C.C.T.T.G.C.A.A.T.C.T.A.A.C.G.A.G.C.T.C.T.A.G.A.C.T.A.G.C.C.G.G.T.A.T.G.C.T.G.T.T.T.T.T.G.G.A.G.T.G.A.A.G.T.C.G.A.G.T.A.T.G.A.G.T.T.G.G.A.G.A.A.A.G.T.T.C.T.G.G.C.C.G.T.T.G.T.G.C.T.T.C.G.T.C.A.T.C.G.A.A.A.A.G.C.C.G.G.A.A.G.A.C.T.C.T.C.C.C.T.A.T.C.G.A.C.C.A.G.C.T.G.C.T.A.C.G.G.G.A.C.T.T.C.G.A.T.C.A.G.T.C.G.A.C.G.C.G.G.C.G.T.G.A.A.A.T.C.A.G.G.C.T.A.T.C.A.A.T.G.A.G.C.T.T.A.A.G.T.T.A.G.G.C.G.G.T.T.C.A.G.T.T.G.C.T.G.T.T.A.T.G.C.T.C.A.T.T.G.A.C.G.C.T.G.A.A.C.C.C.T.G.G.C.T.A.T.C.A.A.T.G.T.A.G.A.C.A.T.G.T.T.A.T.G.A.T.A.A.T.G.A.A.C.T.A.C.G.T.T.C.C.A.T.T.A.G.G.T.C.A.G.A.C.C.C.T.T.A.G.T.A.A.C.T.T.G.G.T.G.G.T.T.T.T.G.A.C.G.A.T.C.A.C.A.C.G.C.G.T.C.T.T.G.G.T.A.C.T.G.T.G.G.T.A.G.C.C.C.A.C.C.G.C.C.C.T.C.T.G.A.T.T.T.A.A.G.C.A.G.C.A.T.T.T.G.G.G.A.C.T.G.T.A.T.T.A.T.A.T.G.G.G.T.T.G.T.C.A.A.C.T.A.A.A.C.G.A.T.G.A.C.G.T.A.G.G.T.A.G.G.C.C.G.C.C.C.A.G.C.T.C.C.C.T.T.A.T.G.T.C.A.G.A.C.T.G.C.T.C.G.G.T.C.A.G.G.C.C.A.T.A.T.G.G.A.C.T.A.C.C.T.T.T.T.A.G.C.A.A.C.G.A.A.A.A.C.A.G.G.G.G.C.A.T.T.G.G.G.A.G.C.C.C.T.T.C.A.G.A.C.A.T.C.G.T.G.G.C.G.C.C.C.T.G.C.C.C.A.C.C.G.T.T.T.C.C.T.C.T.A.C.T.T.G.A.G.G.T.G.C.A.A.C.T.A.A.C.G.G.G.T.T.C.G.G.G.G.C.T.A.T.G.C.G.A.C.C.T.C.G.G.C.T.A.G.A.A.A.T.A.T.C.A.T.T.G.C.C.T.G.A.T.C.T.T.C.A.G.C.C.A.A.A.A.G.T.G.A.A.C.G.G.G.C.A.A.C.T.C.C.T.A.T.T.G.A.A.T.C.G.T.G.A.G.G.A.A.C.C.C.C.T.G.G.G.A.C.A.G.A.T.C.A.A.A.A.T.G.T.A.T.T.C.A.A.T.C.A.C.T.C.G.T.G.A.G.A.G.C.A.C.A.C.C.T.T.C.G.C.T.T.A.C.G.C.C.G.A.A.T.T.A.T.C.A.T.G.A.A.G.A.T.G.C.C.A.T.A.G.G.G.G.C.A.C.T.T.G.T.T.A.G.G.G.G.T.G.G.G.A.T.G.C.A.C.A.C.A.T.G.C.G.C.T.G.A.T.A.T.A.A.A.A.A.T.T.A.C.G.A.C.A.G.A.C.A.G.A.C.G.T.C.C.A.G.G.A.A.G.C.T.A.T.A.C.A.A.G.G.G.C.T.G.C.G.T.C.A.T.G.C.T.T.T.C.G.A.G.G.G.C.T.G.G.C.T.A.C.T.A.C.A.A.A.C.T.G.A.T.T.A.C.G.G.C.G.G.G.A.C.G.A.T.A.T.T.G.C.T.C.G.C.G.C.C.A.T.G.C.G.C.T.G.A.T.G.C.G.G.C.T.A.A.G.T.C.C.G.G.A.T.T.C.G.G.A.T.T.A.C.T.A.A.C.A.C.A.C.T.A.A.T.A.A.C.T.G.C.A.A.G.G.T.T.A.G.A.A.G.C.A.A.T.G.T.C.T.G.T.T.G.T.A.T.T.A.A.A.A.C.C.C.G.A.G.G.C.T.G.C.C.C.C.C.A.G.G.G.A.T.A.C.C.G.T.A.A.T.T.C.C.T.T.T.T.A.C.A.C.T.T.G.T.T.A.T.C.C.C.C.T.T.T.T.C.T.A.A.T.A.G.T.T.G.C.G.T.G.A.T.A.G.G.T.C.A.G.T.T.G.A.T.A.G.A.A.C.T.G.G.C.T.A.T.C.T.C.G.T.C.A.A.G.C.T.G.G.G.T.C.A.C.T.G.C.G.C.C.C.G.T.C.T.T.C.G.T.A.A.A.C.A.G.C.A.A.A.A.A.C.A.A.C.T.G.G.C.A.C.A.G.A.C.A.A.C.C.T.C.T.T.C.A.C.T.C.T.C.T.C.T.C.G.C.G.A.T.C.C.G.A.A.C.T.A.C.G.T.G.T.G.G.A.A.A.G.C.G.C.G.C.T.C.C.G.G.C.G.T.A.G.T.C.G.A.A.G.C.G.C.G.A.A.T.C.A.G.T.A.T.A.A.T.A.T.G.T.C.A.T.C.G.A.T.G.G.T.T.T.T.G.T.C.T.C.C.A.A.G.T.C.C.A.T.G.T.A.G.C.A.G.T.G.A.A.A.A.G.A.A.C.A.T.C.C.G.A.G.T.T.C.G.G.G.C.A.C.T.G.G.G.T.C.C.T.C.T.G.G.T.C.A.C.T.A.A.A.T.G.G.G.G.C.C.A.G.T.A.C.T.T.G.G.G.G.C.G.T.C.G.A.C.C.A.T.T.G.T.G.T.G.C.C.A.G.C.G.G.A.T.T.C.T.A.G.T.T.T.A.A.G.A.A.G.T.A.A.G.C.T.G.C.A.C.T.G.C.T.A.T.A.T.G.C.A.C.T.G.G.A.C.G.G.C.A.T.G.C.G.C.T.T.A.C.T.T.A.T.A.C.C.G.C.T.A.T.A.A.C.T.T.G.G.T.G.G.T.A.T.T.T.A.A.T.G.G.G.G.T.T.A.A.C.C.T.G.T.G.G.C.T.C.T.A.T.C.C.C.G.C.C.C.C.A.A.G.G.C.T.A.C.A.T.C.T.A.A.C.A.A.C.G.A.G.T.A.T.C.G.C.C.A.C.C.A.G.C.C.G.A.T.C.C.T.C.G.C.A.T.A.C.T.C.A.A.A.A.G.A.G.A.A.A.C.G.A.C.A.A.C.G.C.C.G.A.G.A.C.C.T.A.G.G.T.T.G.A.T.A.C.T.C.C.C.G.G.G.A.G.G.G.A.A.C.A.T.T.G.C.C.A.C.G.T.A.A.G.A.G.T.A.T.G.G.T.A.A.T.C.C.C.C.A.T.C.A.G.C.G.T.G.A.A.T.A.A.A.C.T.G.T.C.C.A.A.A.T.C.T.T.G.T.T.G.T.A.T.G.A.T.C.C.A.A.A.T.T.A.T.C.T.T.C.G.G.T.C.A.G.A.C.T.A.G.A.C.T.G.T.A.A.T.A.G.A.T.T.A.G.A.T.A.A.T.C.T.T.T.T.A.G.A.C.T.T.G.C.G.T.T.T.T.C.T.C.C.G.A.C.T.T.C.T.C.T.A.A.A.T.G.A.A.C.A.C.C.G.G.G.C.T.A.G.T.T.T.T.T.C.T.C.A.A.C.C.C.G.T.C.G.C.G.G.C.C.C.C.A.C.A.A.T.C.T.C.C.A.C.T.G.G.A.T.C.T.C.A.A.A.A.G.T.A.T.T.T.G.T.T.A.A.C.C.A.T.G.C.A.A.A.T.A.C.A.A.G.G.T.A.G.T.T.C.T.C.C.T.A.A.A.G.T.C.C.A.T.A.A.T.T.A.C.G.A.G.G.T.C.T.G.C.T.A.T.A.T.G.C.C.A.G.T.A.T.G.C.G.C.A.T.C.T.G.A.G.A.G.C.G.C.A.G.C.A.A.T.C.G.C.G.C.A.G.T.C.A.A.A.G.C.C.A.G.C.A.G.G.G.G.G.A.T.G.A.A.A.C.A.G.A.A.T.C.G.A.C.G.T.T.A.C.T.C.T.C.C.C.A.C.T.A.T.C.T.T.G.T.T.A.G.A.G.A.T.A.G.G.A.A.C.C.C.G.A.A.A.T.C.G.G.C.T.T.A.T.T.G.T.T.G.A.C.G.C.T.C.A.T.A.T.A.G.A.T.C.G.T.C.G.G.T.G.T.C.C.G.G.T.T.T.A.G.T.C.A.T.C.C.A.G.C.G.C.A.C.T.G.C.C.T.G.G.T.C.C.A.G.G.C.A.T.A.T.T.T.T.A.A.C.G.T.A.G.G.T.A.A.A.T.C.C.A.T.G.A.G.A.G.T.C.T.C.G.T.T.A.T.G.G.C.T.T.G.A.G.G.C.A.C.C.A.G.T.A.G.T.G.G.A.G.T.T.T.G.G.A.C.C.C.C.G.G.G.G.A.G.T.G.C.G.A.A.T.G.T.G.A.C.T.T.C.A.G.T.T.G.C.A.C.A.T.T.T.T.T.T.T.A.G.T.T.A.A.T.G.T.A.C.G.C.G.A.T.A.T.T.A.A.A.A.G.T.T.A.A.C.G.T.C.C.G.T.C.A.A.C.C.T.G.C.A.T.T.A.G.C.G.G.A.T.T.C.G.G.G.C.G.A.G.C.T.T.C.G.C.T.G.A.A.G.G.G.G.C.C.C.C.A.C.G.C.T.C.A.A.T.T.G.G.A.T.T.C.T.C.C.C.T.T.T.T.C.T.G.A.G.T.A.T.A.G.G.T.G.C.A.T.C.T.T.A.G.C.T.A.A.G.A.G.C.G.T.C.G.A.C.G.C.C.G.A.A.G.A.T.C.A.C.A.C.C.T.G.C.G.A.G.C.C.C.C.C.A.T.C.T.T.A.T.A.A.G.G.G.A.C.T.T.A.C.A.T.T.C.T.G.T.G.G.T.A.G.A.G.T.C.T.C.A.C.G.G.C.C.T.T.T.A.T.C.T.T.T.T.T.A.A.C.C.C.G.C.A.G.G.G.G.T.T.A.C.C.C.T.T.G.C.C.A.G.A.C.G.C.A.A.T.G.G.C.G.T.C.C.C.C.G.G.T.T.T.T.A.G.A.G.A.T.C.C.T.C.A.A.C.G.T.A.T.T.G.A.T.G.C.C.T.T.T.T.C.A.C.G.G.T.T.C.G.G.G.G.T.G.A.A.G.C.C.C.C.T.C.C.A.C.G.T.G.T.A.C.G.T.C.G.A.G.A.C.G.C.A.A.C.G.G.C.G.C.A.T.T.A.A.G.G.A.C.G.G.A.G.T.T.C.G.A.G.T.G.T.A.T.A.C.C.C.G.T.G.A.A.G.T.C.A.G.A.G.A.G.T.G.C.G.C.G.T.G.A.T.A.A.T.G.A.A.G.G.A.T.T.C.A.C.A.A.C.G.G.T.C.T.T.A.T.G.A.T.T.G.A.A.C.A.G.A.G.T.A.G.C.A.G.T.C.A.C.T.T.G.G.A.T.G.A.G.C.A.A.T.T.C.C.A.A.C.C.G.C.G.G.T.A.C.G.T.G.A.C.C.C.C.C.G.C.A.C.C.C.A.A.G.A.A.A.T.T.A.T.G.T.T.C.A.A.T.C.T.T.C.G.G.A.T.C.A.C.A.G.A.A.T.G.C.T.G.T.A.A.C.A.G.A.T.G.G.C.A.T.A.C.C.G.C.C.C.T.G.G.G.T.G.A.A.A.C.G.A.T.G.C.G.A.C.T.G.T.T.G.T.C.C.A.C.T.G.C.A.T.A.C.A.T.C.A.T.A.G.G.C.T.C.C.A.T.A.G.C.A.T.C.C.A.A.G.G.A.G.A.C.T.A.C.A.C.A.G.C.T.A.G.T.A.C.C.C.G.T.C.T.C.T.A.A.C.A.G.G.C.G.G.G.G.C.A.A.A.T.T.C.T.T.A.A.G.T.A.T.A.A.C.C.A.T.C.A.A.G.T.A.C.C.G.C.T.A.A.G.G.C.A.C.G.G.G.C.C.T.G.C.A.A.A.A.C.C.C.T.G.T.A.G.C.T.C.T.A.C.A.T.C.A.A.A.G.A.T.T.A.T.C.C.T.G.A.T.T.T.C
Sample 3  ( AFR ): T.A.T.A.T.T.G.A.A.A.T.G.A.C.C.A.G.C.A.T.G.C.C.C.G.C.A.A.G.C.T.A.G.A.G.A.C.C.C.G.C.A.A.C.T.T.A.G.G.C.G.G.C.A.T.A.T.C.C.G.A.A.T.T.C.C.A.G.C.T.T.T.C.A.T.C.G.C.A.C.G.T.C.T.T.C.T.C.T.G.T.C.C.T.C.G.C.G.T.A.A.A.T.G.G.T.G.G.A.C.C.G.G.A.G.C.G.G.T.A.T.T.G.A.A.C.T.T.C.G.C.T.A.C.C.G.A.G.C.C.A.T.G.T.C.A.C.C.A.G.T.T.G.G.A.A.T.A.G.A.T.A.G.G.T.A.T.T.C.A.T.T.G.C.A.A.C.G.T.T.A.C.G.A.T.C.C.C.C.G.T.G.G.A.C.C.T.G.A.A.G.A.C.A.G.A.A.A.T.T.A.G.C.T.A.T.C.A.G.C.A.T.G.T.A.G.G.A.A.T.C.T.G.C.C.G.G.A.G.A.G.G.G.C.C.C.G.G.A.A.T.C.A.C.A.G.C.G.G.A.T.G.T.T.T.G.G.A.G.C.G.G.A.T.T.C.G.T.T.T.A.G.T.A.C.G.C.T.A.T.G.C.A.C.T.A.A.C.C.C.A.C.T.A.A.T.G.G.C.A.A.A.A.A.G.A.G.T.G.T.G.C.G.C.C.A.T.T.G.G.T.G.G.C.G.A.C.G.T.T.C.C.C.C.G.G.C.T.G.T.G.T.A.G.T.C.C.G.T.G.G.T.T.A.C.C.G.T.C.A.T.A.A.G.G.A.G.A.T.A.C.A.A.G.T.G.T.G.A.A.G.A.C.A.T.A.C.C.A.C.A.G.A.G.A.A.G.G.A.T.C.T.T.A.A.A.T.C.A.A.C.T.T.C.C.C.T.G.T.T.G.A.A.C.C.T.A.G.T.A.C.G.G.T.G.A.A.T.C.A.G.G.C.G.T.C.C.T.A.T.G.A.C.A.G.C.G.G.T.G.C.A.C.C.C.A.A.G.G.T.A.T.C.A.C.A.A.A.A.T.G.A.C.A.G.C.A.T.T.C.A.T.G.G.A.G.A.G.T.A.G.A.T.T.C.T.A.G.C.G.C.T.G.G.T.C.T.G.T.C.G.T.A.T.C.C.G.G.G.A.C.G.A.C.G.T.A.T.A.C.T.C.T.T.C.T.T.T.C.T.A.A.A.A.C.A.A.C.C.T.G.G.T.C.G.C.C.G.A.C.C.T.C.T.T.A.G.C.A.C.C.A.T.G.T.G.A.A.T.T.C.A.A.G.C.T.C.G.C.T.T.A.G.C.G.T.T.T.A.T.G.G.T.T.G.T.T.C.T.G.G.C.C.T.A.A.A.A.G.T.C.C.C.C.T.T.G.A.C.G.A.T.A.C.A.A.A.A.T.T.C.A.T.C.C.T.G.G.A.A.T.A.C.T.A.C.A.A.G.T.C.A.G.A.A.C.T.A.G.T.C.G.T.C.G.G.G.C.G.T.A.G.A.C.C.G.C.G.C.T.C.A.G.G.A.A.C.T.G.G.T.C.A.G.T.A.C.A.T.G.C.C.A.G.T.A.A.T.A.G.G.C.C.A.G.A.C.A.C.C.G.C.C.C.T.G.A.C.C.T.G.C.A.A.G.A.T.T.C.C.T.T.G.C.A.A.T.C.T.A.A.C.G.A.G.C.T.C.T.A.G.A.C.T.A.G.C.C.G.G.T.A.T.G.C.T.G.T.T.T.T.T.G.G.A.G.T.G.A.A.G.T.C.G.A.G.T.A.T.G.A.G.T.T.G.G.A.G.A.A.A.G.T.T.C.T.G.G.C.C.G.T.T.G.T.G.C.T.T.C.G.T.C.A.T.C.G.A.A.A.A.G.C.C.G.G.A.A.G.A.C.T.C.T.C.C.C.T.A.T.C.G.A.C.C.A.G.C.T.G.C.T.A.C.G.G.G.A.G.T.T.C.G.A.T.C.A.G.T.C.G.A.C.G.C.G.G.C.G.T.G.A.A.A.T.C.A.G.G.C.T.A.T.C.A.A.T.C.C.C.C.T.T.G.A.G.C.A.A.T.G.C.G.T.G.T.C.A.G.T.T.T.A.T.G.T.C.A.C.G.A.C.A.A.T.T.G.A.C.G.C.A.G.A.A.C.C.C.T.C.G.C.T.A.T.C.A.A.T.T.T.A.T.T.C.G.T.G.T.T.A.T.G.A.T.A.C.T.C.A.A.C.T.A.C.G.T.T.C.C.A.T.T.A.G.G.T.C.A.G.A.C.G.C.T.T.A.G.T.A.A.C.T.T.G.G.T.C.G.T.T.T.T.G.A.C.G.A.T.C.A.C.A.C.A.G.G.T.T.T.T.A.G.T.A.C.A.G.T.A.G.T.T.G.G.T.C.A.C.T.C.C.G.G.T.C.C.A.A.A.T.T.A.A.T.C.T.T.C.A.T.C.C.G.A.G.C.C.A.G.C.A.C.A.C.T.C.C.G.C.G.T.C.T.T.G.A.A.C.T.T.A.A.C.G.A.T.G.A.C.G.T.C.G.G.T.A.G.G.A.C.C.C.C.C.A.G.C.T.C.C.G.T.T.A.T.T.T.C.G.A.C.C.T.G.C.T.C.G.G.T.C.A.G.G.C.C.C.A.A.T.T.G.C.C.T.A.G.C.T.T.T.T.A.C.T.A.A.C.G.A.A.C.A.C.A.G.G.G.G.C.A.T.T.G.G.G.A.G.T.C.C.C.T.C.A.G.A.C.A.T.C.G.T.G.A.C.G.C.G.C.T.G.C.A.A.A.C.C.G.T.T.T.A.C.T.G.A.G.C.A.T.G.A.C.A.T.G.G.T.G.C.T.A.A.G.A.T.A.T.T.C.G.G.G.G.C.T.A.A.G.C.G.A.C.C.T.C.G.G.C.T.A.C.A.A.A.T.A.T.C.A.T.T.G.C.C.G.G.A.T.C.T.T.C.A.G.C.C.C.A.A.A.G.T.T.A.A.C.G.G.G.C.A.A.C.T.C.C.T.A.T.T.G.A.A.T.A.G.T.G.A.G.G.A.A.C.T.C.C.T.G.G.G.A.C.A.G.A.T.C.C.A.A.A.T.G.T.A.G.T.C.A.A.T.C.A.C.T.C.G.T.G.A.T.A.G.C.A.C.A.C.C.T.T.C.T.C.T.T.A.C.G.A.C.G.A.A.T.T.A.T.C.A.T.G.A.A.G.A.T.G.A.C.A.T.A.G.G.G.G.C.A.C.T.T.G.T.T.A.C.G.G.A.T.G.G.G.A.T.G.C.A.C.A.C.A.T.G.C.G.C.T.G.G.T.A.T.A.A.A.A.A.T.A.G.G.G.A.C.G.G.A.C.A.G.G.C.C.T.A.C.A.T.G.A.C.T.A.A.A.C.A.C.G.A.G.G.G.C.T.G.A.G.T.C.C.G.G.A.A.G.T.C.G.A.G.G.G.G.G.G.G.C.T.A.C.T.A.C.C.A.C.C.T.G.A.T.T.C.C.A.G.C.A.A.A.A.C.G.C.T.A.T.T.G.A.T.C.T.A.G.C.C.A.T.G.T.G.C.T.G.A.T.G.C.G.G.C.C.A.T.G.C.G.T.G.G.A.T.T.A.G.G.A.T.T.A.T.T.C.G.C.A.T.A.G.T.G.T.T.A.A.A.T.G.A.A.A.A.G.A.C.A.G.T.A.G.C.G.A.T.G.T.C.T.A.T.T.G.T.A.G.A.T.A.A.A.C.A.G.G.C.G.T.T.A.C.G.A.C.C.T.A.G.G.A.A.T.A.G.C.G.T.A.A.T.T.A.C.T.G.T.T.A.C.A.C.T.C.G.T.T.A.A.C.C.A.C.T.T.A.T.C.C.A.A.A.A.G.T.T.T.C.G.T.A.C.T.A.G.C.T.T.C.G.T.T.G.A.T.C.G.A.A.C.T.C.G.C.T.A.T.G.T.C.G.A.G.A.A.G.C.T.G.G.G.G.C.C.A.T.C.G.G.C.C.C.T.T.T.A.T.C.G.T.A.A.A.G.G.A.G.T.T.T.C.A.T.A.A.C.T.C.C.C.A.G.C.C.A.C.A.A.C.T.T.G.T.A.C.A.C.T.C.G.A.T.C.T.C.A.C.G.A.T.C.C.G.A.A.C.A.T.A.G.T.T.T.G.G.A.T.A.G.C.G.C.G.C.T.C.C.G.G.G.G.T.C.G.G.C.C.C.A.G.C.G.C.A.A.T.T.C.A.G.T.A.T.A.A.T.A.C.G.T.C.A.A.C.G.A.T.A.G.T.T.T.T.G.T.C.T.C.C.A.A.G.T.C.C.A.G.G.T.A.G.C.T.G.G.G.A.A.A.A.G.A.A.A.A.C.A.C.G.A.G.T.T.C.C.G.G.C.A.A.A.G.C.T.T.T.T.T.C.T.G.G.T.C.A.C.T.A.A.A.G.C.G.A.A.C.C.A.G.T.A.C.G.T.G.G.C.G.A.T.T.A.T.A.C.C.A.C.T.G.T.G.T.G.C.A.C.C.C.G.C.A.C.C.C.G.A.G.T.T.C.C.A.G.A.A.G.T.A.T.A.C.T.A.C.A.G.T.T.C.A.C.T.A.G.G.C.A.C.T.G.G.T.T.G.G.C.G.T.T.C.G.C.T.T.C.C.T.T.A.A.A.T.G.G.C.T.A.T.A.A.G.A.T.G.A.A.C.C.A.A.T.C.C.A.T.T.G.G.G.T.A.T.G.A.T.C.T.G.T.G.G.A.G.C.T.C.T.C.G.C.G.T.A.C.C.G.T.G.G.C.A.A.C.A.T.C.T.A.A.A.A.C.C.T.A.G.T.A.T.C.G.A.C.A.C.T.A.T.C.A.G.A.T.C.C.T.C.G.C.A.T.A.C.T.C.A.A.A.A.A.A.G.G.C.A.T.G.A.G.T.G.G.T.G.C.T.T.T.C.A.A.A.C.C.G.A.T.C.T.T.A.C.C.T.A.C.T.G.A.T.A.G.T.T.A.C.T.A.T.G.G.C.T.G.A.T.T.C.G.A.G.T.A.T.G.C.G.T.A.C.C.C.C.G.A.G.G.T.T.G.G.T.C.A.A.A.G.G.T.G.T.T.T.C.A.T.A.G.T.G.C.T.A.A.T.A.C.C.C.G.A.T.C.A.A.T.T.A.A.T.C.C.T.G.C.G.G.T.C.A.G.A.C.C.A.T.G.C.T.G.T.G.A.C.A.A.A.G.T.A.T.G.A.T.T.T.C.A.T.A.T.C.G.C.G.T.T.G.C.G.T.A.T.T.C.C.A.G.G.A.C.T.T.C.A.C.T.C.A.T.G.A.A.C.C.T.C.C.A.A.T.C.C.G.A.C.T.T.A.A.C.G.C.G.T.T.A.A.C.C.C.G.A.T.A.G.C.T.C.A.G.T.G.T.A.G.C.C.A.A.C.G.G.A.T.C.T.C.A.A.A.A.G.T.A.T.T.T.G.T.T.A.A.C.C.A.T.G.C.A.A.A.T.A.A.A.A.G.G.T.A.G.T.T.C.T.C.C.T.A.A.A.G.T.C.C.A.A.A.A.T.T.A.C.G.A.G.G.T.C.T.G.C.T.A.T.A.T.G.C.C.A.G.T.A.T.G.C.G.C.A.T.C.T.G.A.G.A.G.C.G.C.A.G.C.A.A.T.C.G.C.G.C.A.G.T.C.A.A.A.G.C.C.A.G.C.A.G.G.G.G.G.A.T.G.A.A.A.C.A.G.A.A.T.C.G.A.C.G.T.T.A.C.T.C.T.C.C.C.A.C.T.A.T.C.T.T.G.T.T.A.G.A.G.A.T.A.G.G.A.A.C.C.C.G.A.A.A.T.C.G.G.C.T.T.A.T.T.G.T.T.G.A.C.G.C.T.C.A.T.A.T.A.G.A.T.C.G.T.C.G.G.T.G.T.C.C.G.G.T.T.T.A.G.T.C.A.T.C.T.A.G.C.G.C.A.C.T.G.C.C.T.G.C.T.C.C.C.A.G.C.A.G.T.T.C.T.C.A.A.C.G.T.A.G.G.T.G.C.G.T.C.G.A.T.G.C.A.A.A.T.C.T.C.C.T.T.A.T.G.A.T.A.G.G.T.G.C.T.A.C.A.A.G.T.T.G.A.G.G.G.G.T.T.T.G.G.A.C.C.C.C.G.G.G.C.A.G.T.G.C.G.A.A.T.G.T.G.A.C.T.T.C.A.G.T.T.G.C.A.C.A.T.T.T.T.T.T.T.A.G.T.T.A.A.T.G.T.A.C.G.C.G.A.T.A.T.T.A.A.A.A.G.T.T.A.A.C.G.T.C.C.G.T.C.A.A.C.C.T.G.C.A.T.T.A.G.C.G.G.A.T.T.C.G.G.G.C.G.A.G.C.T.T.C.T.C.T.T.A.A.G.G.C.G.C.A.C.C.C.C.G.C.T.C.A.A.T.T.G.C.A.T.T.C.T.C.C.C.T.T.T.C.C.T.G.A.G.T.A.T.A.G.G.T.G.G.A.T.C.T.T.A.G.C.T.A.A.G.A.G.C.G.T.C.G.A.C.G.C.C.G.A.A.G.A.T.C.G.A.A.C.C.T.G.C.G.A.G.C.C.C.C.C.A.T.A.T.T.C.T.A.A.G.G.G.A.C.A.C.A.C.A.T.T.A.T.G.T.T.G.T.C.G.A.G.G.C.T.T.A.C.G.T.C.C.C.T.T.G.T.A.T.T.G.T.C.A.A.T.C.C.G.C.A.T.T.G.G.G.T.A.C.C.G.A.G.T.T.C.C.G.T.A.G.T.G.C.T.G.A.C.C.T.C.C.C.T.G.C.T.T.G.A.A.G.T.G.A.T.C.A.T.C.A.A.C.T.C.A.T.T.C.G.T.G.C.C.C.T.T.C.C.G.A.G.G.T.C.C.T.C.G.A.T.G.A.A.A.T.A.A.C.C.G.G.A.G.G.C.C.C.C.C.G.A.T.G.A.T.A.G.T.G.A.A.G.G.G.C.G.A.T.T.T.C.A.T.G.A.C.G.G.A.G.T.T.C.G.A.G.T.G.T.A.T.A.C.C.C.G.T.G.A.C.G.T.T.A.G.A.G.A.G.T.G.C.G.C.G.T.G.A.T.A.T.T.G.G.A.G.T.A.T.T.A.C.C.A.A.G.G.T.T.C.T.A.A.T.G.A.T.T.A.T.C.C.A.G.A.G.T.A.G.C.A.G.G.T.A.C.T.T.G.G.A.T.G.A.G.C.A.A.A.T.C.C.A.A.C.C.G.C.G.G.T.A.C.G.T.G.A.C.C.C.C.C.G.C.A.C.C.C.A.A.G.A.A.A.T.T.A.T.G.T.T.C.A.A.T.C.T.T.C.G.G.A.T.C.A.C.A.G.A.A.T.G.C.T.G.T.A.A.C.A.G.A.T.G.G.C.A.T.A.C.C.G.C.C.C.T.G.G.G.T.G.A.A.A.C.G.A.T.G.C.G.A.C.T.G.T.T.G.T.C.C.A.C.T.G.C.A.T.A.C.A.T.C.A.T.A.G.G.C.T.C.C.A.T.A.G.G.T.T.C.C.A.C.G.G.A.T.G.C.C.C.C.G.C.T.A.C.C.T.T.T.C.C.C.C.G.T.C.T.C.C.A.A.A.T.G.A.C.G.C.G.G.C.A.A.T.G.A.A.T.T.A.A.G.T.A.T.A.A.C.C.A.G.C.A.A.G.T.A.C.C.G.C.T.A.A.G.G.C.A.C.G.G.G.C.C.C.G.C.A.A.A.A.C.G.C.T.G.T.A.G.C.G.C.C.G.C.G.T.G.A.C.A.A.A.T.G.C.G.C.C.T.G.A.T.T.T.G
Sample 4  ( AFR ): G.A.T.G.T.T.G.A.A.T.T.G.A.C.C.A.G.C.A.T.G.C.C.C.G.C.A.C.G.C.T.A.G.A.G.A.C.C.C.G.C.A.A.C.T.T.A.G.G.C.G.G.C.A.T.A.T.C.C.G.A.A.T.T.C.C.A.A.C.T.T.T.C.T.T.C.A.C.T.C.G.T.C.T.T.C.T.C.T.G.T.G.C.G.C.G.C.G.T.A.A.C.T.A.G.T.G.T.G.T.C.G.G.A.C.C.G.G.T.A.T.T.G.A.A.C.T.T.C.G.C.A.A.C.C.T.A.G.C.C.G.T.G.T.C.A.C.C.A.G.T.T.G.G.A.A.T.A.G.A.T.A.G.G.T.A.T.T.G.A.G.A.G.T.A.A.C.T.T.A.A.C.G.A.T.C.C.T.C.T.T.G.G.A.C.C.T.G.A.A.G.A.C.C.G.A.A.A.T.T.A.G.C.T.A.T.C.A.G.C.A.T.G.T.A.G.G.A.A.T.C.T.G.C.C.G.G.A.G.A.G.G.G.C.C.G.G.G.A.T.A.C.C.C.A.G.T.G.C.G.A.A.T.T.C.G.G.T.G.C.G.A.A.G.T.G.G.C.T.C.G.G.C.A.C.G.C.A.A.T.A.C.A.C.T.A.A.C.C.C.A.G.G.T.A.T.G.G.A.T.A.A.A.A.T.A.G.T.G.T.C.C.G.C.C.T.T.T.G.G.C.G.G.C.G.A.G.T.T.T.C.A.C.C.C.G.A.T.A.T.G.T.A.G.T.C.A.G.T.G.G.T.G.T.T.T.G.T.A.T.T.G.G.G.C.G.G.A.T.A.C.A.A.G.A.G.G.G.G.T.G.A.G.A.C.A.C.C.A.C.A.G.A.C.G.A.C.G.A.T.C.T.T.A.A.A.T.A.A.A.C.T.T.C.C.G.T.G.T.T.G.A.A.A.G.T.G.G.T.G.C.G.G.G.G.T.A.T.C.A.C.G.C.G.T.C.C.C.T.T.A.A.T.A.G.A.C.A.T.G.T.A.C.C.C.A.A.G.G.T.A.T.C.A.C.G.A.A.A.T.A.C.C.A.A.C.A.T.G.C.A.T.G.C.C.C.T.C.T.A.C.T.G.G.C.T.A.G.C.G.C.T.A.G.G.C.T.G.G.C.G.T.A.T.C.C.G.G.A.A.T.G.A.G.G.T.T.A.G.C.T.C.T.T.C.T.T.T.T.T.A.A.A.G.C.A.A.T.C.T.G.C.T.C.G.C.T.C.G.T.C.T.C.A.A.A.G.G.C.T.C.A.T.G.T.G.A.A.T.T.C.A.C.G.C.T.C.A.C.T.T.A.G.C.A.T.T.G.T.T.G.C.T.T.G.T.T.C.T.G.G.C.C.T.A.A.A.A.A.T.C.C.C.C.T.G.G.A.C.G.A.T.A.C.T.A.A.A.T.T.C.A.T.C.C.T.G.G.A.A.T.A.C.T.A.C.A.A.G.T.C.A.G.A.A.C.T.A.G.T.C.G.T.A.G.G.G.C.G.T.A.T.G.C.C.T.C.T.C.T.C.T.G.G.A.C.C.T.G.A.A.A.T.G.C.T.A.A.T.G.A.C.A.A.T.A.G.T.A.G.T.C.C.T.G.G.A.A.T.A.G.C.G.C.T.G.A.G.C.C.G.C.T.A.T.G.T.T.A.A.A.T.G.C.A.C.T.C.A.A.T.C.G.A.G.C.T.C.T.C.G.A.C.T.A.G.C.C.G.G.T.A.T.G.C.T.G.T.T.G.T.T.G.G.A.G.T.T.A.G.G.T.C.G.A.G.T.A.T.G.A.G.T.T.G.G.A.G.A.A.A.G.T.T.A.T.G.G.C.A.G.C.T.G.A.C.C.T.T.C.G.T.C.A.A.C.G.A.A.G.A.G.C.T.G.G.A.A.G.G.C.T.C.T.C.C.T.C.G.T.G.G.C.C.C.A.C.G.T.G.G.G.A.C.G.G.G.A.C.T.T.C.G.A.T.C.A.G.T.C.G.A.C.G.A.G.A.A.G.T.G.A.C.A.T.G.A.G.G.C.C.A.T.G.A.A.T.C.C.C.C.T.G.G.A.G.C.A.A.T.G.C.G.T.G.G.C.A.G.T.T.T.A.T.G.T.T.A.C.G.A.C.A.A.T.T.G.A.C.G.C.T.G.G.A.C.C.C.G.C.G.C.T.A.G.G.A.A.C.T.C.A.T.A.G.A.T.G.G.T.A.T.G.T.A.T.A.C.G.A.A.T.T.A.A.G.T.T.C.T.A.T.T.A.G.G.T.C.A.G.A.C.G.C.T.T.A.G.T.A.A.A.C.T.G.G.T.G.G.T.A.T.T.G.A.C.C.A.T.G.G.A.A.C.A.C.G.T.C.T.T.G.G.T.A.C.A.G.T.G.G.A.A.G.C.T.C.A.C.C.G.C.C.C.T.G.T.A.C.T.T.T.A.A.T.C.A.G.C.A.T.T.T.G.G.G.A.C.T.G.T.A.T.T.A.T.A.T.T.C.G.T.T.G.T.C.A.C.C.T.A.A.A.C.G.A.T.G.A.C.G.T.C.T.G.T.A.G.G.C.C.C.C.C.C.A.G.C.T.C.C.G.T.G.A.T.T.T.C.G.A.C.C.T.G.C.T.T.G.G.T.C.A.G.G.C.C.A.A.A.T.G.G.A.C.C.A.C.A.T.T.T.T.A.C.T.A.A.C.G.A.A.C.A.T.A.G.G.G.G.C.A.T.T.G.G.G.A.G.C.C.C.C.T.C.C.G.A.C.A.T.T.A.T.C.A.C.G.A.G.C.T.G.C.A.A.G.A.C.T.A.T.C.A.C.T.G.A.A.C.A.T.G.A.G.G.T.G.G.T.G.C.T.A.A.G.A.G.A.T.G.C.G.G.G.G.C.T.A.A.G.C.C.A.C.C.C.C.T.G.T.T.C.C.A.A.A.T.A.T.C.A.A.T.A.C.C.G.G.A.G.C.T.A.C.A.G.C.C.C.A.A.A.T.T.G.A.A.C.G.G.G.C.A.A.C.T.C.C.T.A.T.T.G.A.A.T.A.G.C.G.A.G.G.A.A.C.C.C.C.T.G.G.G.A.C.A.G.A.G.C.C.A.A.A.T.G.T.A.T.T.C.A.A.T.C.A.C.T.C.G.T.G.G.G.A.G.C.A.C.A.T.C.T.T.C.T.C.T.T.A.C.G.C.C.G.A.A.T.T.A.T.C.A.T.G.A.A.G.A.T.G.A.C.A.T.A.G.G.G.G.C.A.C.T.T.G.T.T.A.C.G.G.A.T.G.G.G.A.T.G.C.A.C.A.C.A.T.G.C.G.C.T.G.G.T.A.C.A.A.A.A.A.A.T.A.C.G.A.C.G.G.A.C.A.G.A.C.G.T.C.C.A.G.C.A.A.G.C.T.A.T.A.C.G.A.G.G.G.C.T.G.C.G.T.C.A.T.G.C.A.T.C.C.G.A.G.G.G.C.G.A.C.A.G.G.T.T.G.C.C.G.C.C.T.G.G.T.T.C.C.A.G.G.A.A.A.T.C.G.C.T.A.T.T.G.C.T.C.G.C.G.C.C.A.T.G.T.G.C.T.G.A.C.G.C.A.G.C.T.A.A.G.C.G.T.T.G.A.T.T.A.G.G.A.T.T.A.C.T.A.G.C.A.T.A.G.T.G.T.T.T.A.A.T.G.C.A.A.A.G.A.T.A.G.T.A.G.C.G.A.T.G.T.C.G.A.T.T.G.T.A.G.T.T.A.A.A.C.A.G.G.C.G.T.C.A.C.G.A.C.C.T.A.G.G.A.A.T.A.C.C.G.T.A.A.T.T.A.C.T.G.T.T.A.C.A.C.T.T.G.T.T.A.T.C.C.C.C.T.T.T.T.C.T.A.A.T.A.C.T.T.T.C.G.T.G.C.T.A.G.C.T.C.C.G.T.T.G.A.T.A.G.A.A.C.T.C.G.C.T.A.T.C.T.C.C.T.C.A.A.G.C.T.G.G.G.G.C.C.A.G.G.G.G.C.C.C.T.T.T.A.T.C.A.T.A.G.A.G.G.A.G.T.T.T.C.A.T.A.A.C.T.G.G.C.A.G.C.G.A.C.A.A.C.T.T.G.T.A.A.A.C.T.C.T.A.T.C.T.C.A.C.G.A.T.C.C.G.A.A.C.T.T.C.G.T.T.T.G.C.A.T.G.G.A.G.C.G.C.T.C.C.G.G.C.G.C.A.G.T.T.G.C.A.G.C.G.C.A.A.T.T.C.A.G.T.A.A.A.A.A.A.C.A.T.G.A.T.C.C.A.G.A.G.T.T.T.C.G.T.T.C.C.C.T.A.G.T.C.C.A.T.G.G.C.G.C.T.G.G.G.A.A.G.A.G.A.T.A.A.T.C.C.G.A.G.T.T.G.G.T.G.C.A.A.T.C.C.T.T.C.T.T.C.T.G.G.T.C.A.C.T.G.A.A.G.G.G.G.A.C.C.T.G.T.A.A.G.T.G.G.C.G.C.T.T.A.G.A.C.C.A.C.T.T.C.G.T.G.C.A.A.G.C.G.G.A.T.T.G.T.A.G.T.T.T.C.A.T.A.A.C.T.A.T.G.C.T.G.C.A.G.T.T.C.A.C.G.A.T.G.A.A.C.G.G.G.A.T.G.G.C.G.T.T.C.T.G.T.T.C.C.T.T.A.A.C.T.G.T.C.T.A.T.A.A.G.A.T.G.A.A.C.C.A.A.T.C.C.A.T.T.G.G.G.T.A.C.A.A.C.C.T.G.T.G.G.C.G.C.T.C.T.T.G.C.G.T.A.C.C.G.A.G.G.C.A.A.C.A.T.C.T.A.A.A.T.C.C.T.A.G.T.A.T.C.T.A.C.G.C.T.A.T.C.C.G.A.T.C.C.T.C.G.C.A.T.A.C.T.T.A.A.A.A.A.A.G.G.C.A.T.G.A.C.A.A.C.T.C.C.G.A.G.A.C.C.T.A.G.G.A.T.G.A.C.A.C.T.C.C.C.G.G.G.A.G.G.G.A.A.A.A.A.T.G.G.C.A.C.G.G.T.A.G.A.G.T.G.T.G.G.T.A.T.T.C.C.C.C.A.T.C.A.G.C.G.T.G.A.A.T.A.A.A.C.T.G.T.T.C.A.A.A.T.C.T.T.G.T.T.G.T.A.T.G.A.T.C.C.A.A.T.T.T.A.C.C.T.T.C.T.G.T.C.G.G.A.C.T.A.G.A.C.T.G.T.A.A.T.A.A.A.T.T.A.G.A.A.A.A.T.C.T.T.T.T.A.G.A.G.C.T.G.C.G.T.T.T.G.C.C.A.C.C.A.C.T.T.C.T.T.A.C.A.A.T.G.C.A.C.A.C.C.G.G.T.T.T.G.G.C.T.T.A.A.T.T.C.G.T.T.A.A.C.C.C.G.C.T.A.C.C.C.C.A.C.T.A.T.C.T.C.C.C.C.T.G.G.C.T.C.A.C.G.A.A.A.G.G.A.T.T.G.G.T.T.G.T.C.T.G.A.A.C.A.A.A.T.A.C.A.A.T.C.T.A.T.A.T.C.C.C.C.C.A.A.A.A.T.C.C.C.A.G.A.T.T.A.A.G.T.G.C.A.C.G.A.C.T.G.T.G.T.G.G.C.A.G.C.A.C.C.G.A.C.T.C.C.G.A.A.G.C.C.G.C.T.G.T.G.A.G.C.T.G.C.G.T.G.A.G.T.A.A.T.G.C.C.C.A.C.C.G.G.G.G.G.A.T.G.A.A.T.G.C.T.G.A.T.C.G.C.G.G.T.T.C.G.G.C.T.C.C.T.T.C.T.T.T.G.A.T.A.A.T.T.C.T.T.C.G.A.G.C.C.A.C.A.C.G.A.A.A.T.C.G.G.C.T.T.A.T.C.A.T.T.G.T.A.G.C.T.C.A.T.G.T.A.G.C.G.G.G.T.C.G.G.T.G.T.C.C.G.G.T.T.A.A.C.T.C.A.T.A.C.A.G.C.G.C.G.C.T.G.C.C.T.G.G.T.C.C.C.A.G.C.A.T.A.T.T.T.T.A.A.C.G.T.A.G.G.T.A.A.A.T.C.C.A.T.G.C.A.A.G.T.C.T.C.C.T.T.A.T.C.G.T.T.G.G.T.G.C.T.A.C.A.A.G.T.T.G.A.G.G.A.A.T.G.G.G.G.A.C.T.T.A.C.G.T.C.C.T.T.G.C.G.A.T.T.G.A.G.A.C.C.T.C.A.G.T.T.G.C.A.C.A.T.T.T.T.T.T.G.A.G.T.T.A.A.A.G.T.A.C.G.C.G.C.T.C.G.G.A.A.A.A.C.T.C.T.A.C.G.C.C.C.G.T.C.A.A.G.C.T.C.T.A.T.G.C.T.C.T.T.C.G.C.C.G.G.G.C.G.T.G.A.G.T.C.G.C.T.G.T.A.G.G.G.G.C.C.C.C.A.C.G.A.A.G.A.A.C.T.G.C.A.T.T.C.T.C.C.C.T.T.T.C.C.T.G.A.T.T.A.T.A.T.G.T.G.C.A.T.C.T.A.A.G.C.T.A.A.G.A.G.C.G.T.C.G.T.C.G.C.C.G.A.G.G.A.T.C.A.A.A.C.C.T.G.C.G.A.G.C.C.C.C.C.G.T.C.T.C.C.C.A.A.C.G.G.A.G.A.C.A.C.A.T.T.C.T.G.T.G.G.T.A.G.A.G.T.C.T.C.A.C.G.G.C.C.T.T.T.A.T.C.T.T.T.T.T.A.A.C.C.C.G.C.A.G.G.G.G.T.T.A.C.C.C.T.T.G.C.C.A.G.A.C.G.C.A.A.T.G.G.C.G.T.C.C.C.C.G.G.T.T.G.A.A.G.T.G.A.T.C.A.T.T.A.A.C.T.C.A.T.T.G.G.T.A.C.A.T.T.T.C.C.A.C.G.G.T.C.C.T.G.G.G.T.G.A.A.G.C.C.C.C.T.C.C.A.C.G.T.G.T.A.C.G.T.C.G.A.G.A.C.T.C.A.C.C.G.G.C.G.C.A.T.T.C.A.T.G.A.C.G.G.A.G.T.T.T.G.A.T.T.G.T.C.T.A.C.C.C.G.T.G.A.A.G.T.C.A.G.A.G.A.G.T.G.C.G.C.G.T.G.A.T.A.T.T.G.G.A.G.T.A.T.T.A.C.C.A.A.G.G.T.T.C.T.A.A.T.G.G.T.T.A.T.C.C.A.A.A.G.T.A.G.T.A.G.T.C.A.C.T.T.G.G.A.T.A.A.G.C.C.A.T.T.C.C.A.A.C.C.G.C.G.G.T.A.C.G.T.G.A.C.G.C.A.C.T.C.A.C.C.C.A.A.C.A.A.A.T.T.A.T.A.A.T.C.A.A.T.C.T.T.C.G.G.T.T.C.A.C.A.G.A.A.T.G.C.T.A.T.A.A.C.A.C.A.T.G.G.C.A.G.G.G.G.G.C.C.C.G.T.G.A.A.C.G.A.C.C.A.A.T.A.A.T.A.A.G.G.G.T.G.T.C.C.A.A.A.G.C.C.T.G.G.C.T.C.C.T.G.A.T.G.A.C.C.C.A.T.G.C.A.G.C.C.A.A.G.G.A.T.A.G.T.A.C.G.C.A.G.C.C.T.T.T.A.C.C.C.G.T.C.T.C.C.A.A.A.T.G.A.C.G.C.G.G.C.A.A.T.G.A.C.T.T.A.T.G.T.A.T.A.A.C.C.A.G.C.A.A.G.T.A.T.C.G.C.G.T.T.C.G.C.A.G.C.T.G.A.C.T.G.C.A.C.A.A.C.C.C.T.G.T.G.G.C.T.C.C.A.T.A.A.C.A.A.A.G.A.T.T.A.T.C.C.T.A.A.T.T.T.C
Sample 5  ( AFR ): T.A.T.A.G.C.G.A.A.T.T.C.A.G.C.C.T.C.A.C.A.G.G.G.G.C.A.C.C.G.T.C.G.G.G.A.C.C.C.T.C.T.C.G.G.A.G.T.T.G.C.A.T.T.T.A.G.T.C.G.A.A.T.T.C.C.A.A.C.T.T.T.C.T.T.C.A.C.A.C.A.T.C.A.T.C.T.C.T.G.T.C.C.T.C.G.C.G.T.A.A.C.T.G.G.T.G.G.A.C.A.A.A.C.G.T.G.G.C.G.T.T.T.A.C.T.T.T.C.G.G.T.A.T.C.T.C.G.G.C.A.A.G.T.C.G.C.G.T.G.A.T.G.G.A.T.T.C.G.A.T.A.G.T.T.G.A.A.C.A.G.A.G.T.A.A.C.G.T.A.A.A.G.A.T.C.C.C.C.T.T.G.G.G.G.A.T.G.C.A.G.G.C.C.C.G.T.C.G.G.A.G.C.T.A.A.C.T.G.T.T.T.G.T.A.G.G.A.A.T.C.T.G.C.C.G.T.C.A.A.G.T.G.G.C.C.G.G.A.T.A.C.C.C.A.G.T.C.G.G.A.A.T.T.T.A.G.T.G.C.G.A.A.G.T.C.G.T.T.C.G.G.C.A.G.A.C.T.C.T.G.C.A.C.T.A.A.C.G.C.A.C.G.A.A.T.G.G.C.A.A.C.A.A.G.A.C.T.G.G.G.A.C.T.G.T.C.A.C.G.T.T.G.C.G.T.C.G.T.T.G.C.C.C.G.G.C.T.A.T.G.T.A.G.G.C.C.G.T.G.G.T.T.A.C.T.T.T.C.A.T.A.A.G.C.G.G.A.C.A.G.A.A.G.T.G.T.G.A.A.G.A.C.A.T.A.C.C.A.C.A.G.A.G.A.A.G.G.A.T.C.T.T.A.A.A.T.C.A.A.C.T.T.C.C.C.T.G.T.T.G.A.A.A.C.T.A.G.T.A.C.G.G.T.G.A.A.T.C.T.C.C.C.G.T.C.C.T.T.T.G.A.T.A.G.A.G.A.T.G.C.T.T.C.A.T.A.G.G.T.A.A.C.G.A.G.T.A.A.T.A.C.C.A.A.C.A.C.G.C.A.T.G.C.C.C.T.C.T.A.C.T.G.G.C.G.A.G.C.G.C.T.A.G.G.C.T.G.T.T.G.T.A.C.C.T.G.G.G.A.C.G.A.C.G.G.T.A.G.G.T.G.A.A.G.T.G.G.T.T.G.A.A.A.G.A.A.C.A.T.G.G.A.C.G.C.C.G.A.C.C.T.C.T.T.A.G.C.A.C.C.A.T.G.T.G.G.A.T.T.C.A.A.T.C.T.C.A.C.T.T.A.G.C.A.T.T.G.A.T.C.G.T.G.G.C.A.T.T.G.G.G.C.G.A.A.A.A.G.T.C.G.C.C.T.T.A.A.T.G.A.T.G.C.T.A.A.A.T.T.C.A.T.C.C.T.G.G.A.A.T.G.T.A.C.C.A.A.T.A.C.C.G.A.T.A.T.A.G.T.C.G.T.C.A.G.G.C.T.T.A.G.G.C.C.T.C.T.C.T.C.T.G.T.G.A.C.G.G.A.A.A.T.T.C.A.C.A.C.G.C.C.A.A.T.A.A.T.C.G.T.C.C.A.T.A.C.A.T.A.G.C.G.C.T.G.A.G.C.C.G.C.A.A.T.A.T.T.A.A.A.T.T.C.G.C.T.C.A.A.T.C.T.A.G.C.A.C.T.C.G.A.C.T.A.G.G.C.G.T.G.T.T.A.G.A.G.T.T.T.A.A.G.G.A.G.T.T.A.G.G.T.C.G.A.G.T.A.T.G.A.G.T.T.A.C.T.A.C.A.A.G.T.T.A.T.G.C.G.C.C.T.G.T.T.C.T.T.C.C.G.T.C.C.T.C.G.T.A.A.A.G.C.T.G.G.A.A.C.A.C.C.C.T.C.C.T.C.A.T.C.A.A.C.C.A.G.C.A.G.G.G.A.C.C.G.C.A.C.T.T.C.G.A.T.C.A.C.T.C.G.A.C.G.A.G.G.C.G.T.G.A.A.C.C.G.T.C.C.T.C.C.T.G.C.C.T.G.A.C.C.T.T.A.T.G.C.A.A.T.G.A.G.T.G.T.C.A.G.T.T.T.A.T.G.T.T.A.C.G.A.C.A.A.C.A.C.A.T.G.C.A.G.A.A.C.C.C.T.C.G.C.T.A.T.C.A.A.C.T.C.A.T.A.G.A.T.G.G.T.A.T.G.T.A.T.A.C.G.A.A.T.T.A.A.G.T.T.C.T.A.T.T.T.G.G.T.C.A.G.A.C.G.C.T.T.A.G.T.A.G.C.T.T.G.G.T.G.G.T.T.T.T.G.A.T.G.A.C.C.A.C.A.C.A.C.G.T.C.T.T.G.G.T.A.C.A.G.T.G.G.T.A.G.G.T.C.G.C.T.C.C.G.G.C.C.C.A.A.T.T.T.A.A.T.C.T.T.C.T.T.C.C.G.G.G.C.C.A.G.C.A.C.A.C.T.C.C.G.C.G.G.C.G.T.G.A.A.C.T.A.A.A.C.G.A.T.G.A.C.G.T.C.T.G.T.A.G.G.C.C.C.C.C.C.A.G.C.T.C.C.G.T.T.A.T.T.T.C.G.A.C.C.T.G.C.T.C.G.G.T.C.A.G.G.C.A.C.A.A.T.G.G.A.C.T.A.C.C.T.T.T.C.A.C.T.A.A.C.G.A.A.C.A.C.A.G.G.G.G.C.A.T.G.G.G.G.A.G.C.T.C.C.T.C.A.G.A.C.A.T.C.G.T.G.A.C.G.A.G.G.T.G.T.A.A.A.C.C.T.A.T.T.A.C.T.G.A.A.C.A.T.G.A.C.G.T.G.G.T.G.C.T.A.A.G.A.G.A.T.T.C.G.G.G.G.A.T.C.A.G.C.G.A.C.C.T.C.G.G.C.T.C.C.A.A.G.T.T.T.G.A.T.G.G.C.A.G.G.T.T.C.T.T.C.T.A.T.C.C.G.T.A.G.T.G.A.C.G.T.T.G.T.A.A.C.T.C.C.T.A.T.T.G.A.A.T.A.G.T.G.A.G.G.A.A.C.C.C.C.T.G.G.G.A.C.A.G.A.T.C.C.A.A.A.T.G.T.A.G.T.C.A.A.T.C.A.C.T.C.G.T.G.A.G.A.G.C.A.C.A.C.C.T.T.C.T.C.T.T.A.C.G.C.C.G.A.A.T.T.A.C.C.A.T.G.A.A.G.A.T.G.A.C.A.T.A.G.G.G.G.C.A.C.T.T.G.T.A.A.C.G.G.A.T.G.G.G.A.T.G.C.A.C.A.C.A.T.G.C.G.C.T.G.G.T.A.T.A.A.A.A.A.T.A.A.C.G.G.C.G.G.A.C.A.A.G.C.C.T.C.C.A.G.G.G.C.G.C.T.C.C.T.C.G.A.T.G.G.C.T.G.A.G.T.C.C.G.G.A.A.G.T.C.G.A.G.G.G.G.G.G.G.C.T.A.C.T.A.C.C.A.C.C.T.G.A.T.T.C.C.A.G.C.A.A.A.A.G.G.C.T.T.T.C.G.C.T.C.G.C.G.C.C.A.T.G.T.G.C.T.G.A.T.G.C.A.G.C.T.A.A.G.C.G.T.T.G.A.T.T.A.G.C.A.T.A.A.C.T.A.A.C.A.C.A.C.T.A.A.T.A.A.A.T.A.C.A.A.G.G.T.T.A.G.T.A.G.C.A.A.T.C.T.C.T.G.T.T.G.T.A.G.T.A.A.A.A.C.C.C.G.A.G.G.C.T.G.C.C.C.C.T.A.G.G.G.A.T.A.C.C.G.T.A.A.T.T.A.C.T.G.T.T.A.C.A.C.T.C.G.T.T.A.A.C.C.A.C.T.T.A.G.C.C.A.A.A.A.G.C.T.T.C.G.T.G.C.T.A.G.C.T.T.C.G.T.T.G.A.T.C.G.A.A.C.T.C.G.C.T.A.T.G.T.C.G.T.C.A.A.C.C.T.G.C.G.G.C.C.A.T.G.G.G.C.C.A.T.T.T.A.T.C.G.T.C.A.A.G.G.A.G.T.T.T.C.A.T.T.T.G.T.G.G.C.A.G.A.G.A.C.T.A.C.T.T.G.G.A.C.A.G.T.C.T.A.T.C.T.C.A.C.G.A.T.A.C.G.G.A.C.T.T.C.T.T.T.T.G.G.A.T.A.G.C.G.C.G.C.T.C.C.G.G.C.G.T.C.C.T.C.C.C.A.G.T.G.G.A.G.A.T.C.A.G.T.C.T.A.A.T.A.C.G.T.C.G.T.C.G.A.T.A.G.T.T.T.T.G.T.C.T.C.C.A.A.G.T.C.C.T.T.A.T.C.C.G.T.G.G.G.A.A.G.A.C.T.A.C.T.T.A.C.G.A.G.T.A.C.G.G.C.C.T.A.T.G.C.T.C.C.T.T.C.T.A.G.T.C.A.C.T.A.A.A.G.G.G.G.A.C.C.A.G.C.A.C.T.T.G.G.G.T.C.G.T.C.G.A.C.C.A.T.T.G.T.G.T.G.T.A.A.G.C.G.G.C.T.T.C.T.A.G.T.T.T.A.A.G.A.A.G.T.G.A.G.C.T.G.C.A.C.T.G.C.T.A.T.A.T.G.C.A.C.T.G.G.A.C.G.G.C.A.T.G.C.G.C.T.T.A.C.T.T.A.T.A.T.C.G.C.T.A.T.A.A.G.T.T.G.G.T.G.G.A.A.T.T.T.A.A.T.G.G.G.G.T.T.A.A.C.C.T.G.C.G.G.C.T.C.T.A.T.C.C.C.G.C.C.C.C.A.A.G.C.C.T.A.C.A.T.C.T.A.A.A.A.A.C.G.G.G.A.T.T.G.G.A.C.A.T.T.T.T.G.C.T.A.A.G.C.A.G.G.C.A.G.A.T.C.C.C.T.C.A.A.A.C.G.C.A.T.G.A.C.A.G.C.T.C.C.G.A.G.A.A.C.T.A.G.G.A.T.G.A.T.A.C.T.C.C.C.T.G.G.A.G.G.G.A.A.C.A.A.T.G.G.C.A.C.G.T.T.A.G.A.G.T.A.T.G.G.T.A.A.T.C.C.C.C.G.T.C.A.G.C.A.T.G.A.A.T.A.A.A.C.T.T.T.C.C.A.A.A.T.C.C.T.G.T.T.G.T.A.T.G.A.T.C.C.A.A.T.T.T.A.C.C.T.T.C.T.G.T.C.A.G.A.C.T.A.G.A.C.T.G.T.A.A.T.G.A.A.T.A.C.G.A.A.T.A.T.C.T.T.T.T.C.G.A.G.T.C.G.G.A.T.T.A.T.G.C.A.C.C.A.C.T.T.T.A.C.T.C.A.T.G.G.A.C.C.T.C.C.A.A.T.C.C.G.A.C.T.T.A.A.C.T.C.G.T.T.A.A.C.C.C.G.C.T.A.G.C.T.C.A.G.T.A.T.A.G.A.C.A.A.C.G.G.A.T.C.A.C.G.A.A.A.G.G.A.T.T.G.G.T.T.G.A.C.C.G.A.A.C.A.A.A.T.A.C.G.C.G.G.T.A.G.T.T.C.T.C.C.T.A.A.A.A.T.C.C.A.A.G.C.T.T.A.A.G.A.G.C.T.C.G.A.C.T.A.T.G.T.G.G.C.A.G.T.C.C.C.C.A.C.T.C.C.G.A.A.T.C.C.G.C.T.G.T.C.A.G.C.T.G.C.G.T.G.A.G.T.T.C.T.G.C.C.A.G.C.C.A.G.G.G.G.A.T.G.A.A.A.C.A.G.A.A.T.C.G.A.C.G.T.T.A.C.T.C.T.C.C.C.A.C.G.A.T.C.T.G.A.T.T.A.G.A.T.A.T.A.G.G.A.G.G.A.G.G.A.C.A.T.T.C.G.C.T.G.G.T.T.A.T.C.G.A.A.G.C.T.A.T.C.A.T.A.C.A.T.G.G.G.C.G.G.C.C.T.C.C.A.G.T.T.A.A.C.G.C.T.T.C.C.A.G.C.G.C.G.G.G.G.T.T.A.A.G.T.C.C.C.A.G.T.A.T.T.T.C.A.C.A.A.C.G.T.A.A.G.T.A.A.A.T.C.C.C.T.G.A.A.A.G.T.C.T.C.G.T.T.A.T.G.G.T.T.T.G.A.G.C.T.A.A.A.A.G.C.T.G.A.G.G.A.A.C.G.G.A.A.A.A.C.T.A.C.G.T.C.A.T.T.C.C.G.A.T.T.G.A.G.A.C.T.T.C.G.G.T.T.C.C.A.C.A.T.T.G.T.T.T.T.A.G.T.G.A.A.T.G.T.G.C.C.G.G.C.T.C.G.T.C.A.A.A.C.T.C.T.T.A.G.C.G.T.C.C.C.A.G.G.G.A.C.T.A.T.G.C.T.C.T.G.C.G.C.C.G.T.G.C.G.A.G.C.T.T.G.T.C.G.T.A.A.G.G.C.G.C.A.C.C.C.C.A.C.T.G.G.C.T.G.T.C.A.T.T.C.G.C.C.G.T.T.T.C.C.T.C.A.T.T.C.C.G.T.G.T.C.C.A.T.C.C.T.A.T.A.T.A.G.T.C.T.G.G.T.C.G.T.A.C.C.C.G.G.G.T.A.T.T.A.A.G.C.G.A.G.C.G.A.T.T.G.G.C.A.A.G.C.T.C.C.C.G.A.G.G.T.A.C.T.T.T.C.A.T.A.A.T.G.A.T.C.T.A.G.A.G.T.C.T.C.A.C.G.G.C.C.T.T.T.A.T.C.T.C.T.T.T.A.A.C.C.C.G.C.A.G.G.G.G.T.T.A.C.C.C.T.T.G.C.C.C.G.A.C.G.C.A.A.T.G.G.C.G.T.C.C.C.C.G.G.T.T.G.T.A.C.A.G.A.C.G.A.T.C.A.G.G.G.T.A.G.T.G.A.T.G.C.C.T.A.A.C.C.A.A.A.G.T.T.C.G.G.G.G.T.G.A.A.G.C.C.C.C.T.C.C.T.C.G.T.G.T.A.C.G.T.C.G.A.G.A.C.G.C.A.A.C.G.G.C.G.C.A.T.G.C.A.T.G.A.C.G.G.T.G.T.C.C.G.A.G.T.G.T.A.G.A.C.C.C.G.T.G.A.A.G.T.T.A.G.A.G.A.G.T.G.C.G.C.G.C.C.A.T.A.A.T.G.A.A.G.T.A.T.G.C.A.C.A.A.C.G.G.T.C.T.T.A.T.G.A.T.T.A.A.A.C.A.G.A.G.T.A.G.C.G.G.G.C.G.C.T.T.G.G.A.T.G.A.G.A.C.T.T.T.C.G.A.A.C.C.G.C.C.G.T.A.C.C.T.G.A.C.G.C.C.T.G.C.A.C.C.C.A.A.G.A.A.A.T.T.A.T.G.T.T.C.A.A.T.C.T.T.C.G.G.A.T.C.A.A.A.T.A.A.T.G.C.T.A.T.A.A.C.A.G.A.T.G.G.A.A.T.A.C.C.G.C.C.C.T.G.G.G.T.G.A.A.A.C.G.A.T.G.C.G.A.C.T.G.T.T.G.T.C.C.A.C.T.G.C.A.T.A.C.A.G.C.A.T.A.G.G.C.T.C.C.A.T.A.G.C.T.T.C.C.A.A.A.G.A.T.G.C.T.A.C.A.G.A.G.C.T.T.G.A.A.C.C.G.G.T.C.A.C.T.T.A.C.A.G.G.C.G.G.G.G.G.A.T.A.T.A.C.A.T.G.A.G.T.A.G.T.A.C.C.G.G.C.A.A.G.T.G.C.C.G.C.T.T.T.C.G.C.A.C.C.G.G.C.G.T.G.C.A.A.A.A.C.G.C.T.G.T.A.G.C.G.C.C.G.C.G.T.G.A.C.A.A.A.T.G.C.G.C.C.T.G.A.T.T.T.G
Sample 6  ( AFR ): T.A.T.A.T.T.G.A.A.A.T.G.T.G.C.A.G.C.A.C.G.G.G.G.G.T.A.C.C.C.T.C.G.A.G.A.C.C.C.T.C.T.C.G.G.A.G.T.T.G.C.A.T.T.T.A.G.T.C.G.A.A.T.T.C.C.A.A.C.T.T.C.C.T.T.C.A.C.A.C.G.T.C.T.T.C.T.T.T.G.T.C.C.T.C.G.C.G.T.A.A.C.T.G.G.T.G.G.A.C.A.A.A.C.G.C.G.G.C.G.T.T.T.A.C.T.T.T.C.G.G.T.A.T.C.T.C.G.G.C.A.A.G.T.C.G.C.G.T.G.A.T.G.G.C.T.T.C.G.C.T.A.G.G.T.G.A.A.C.A.G.A.G.T.A.A.C.G.T.A.A.C.G.A.T.C.C.T.C.T.T.G.G.G.G.A.T.C.C.A.G.G.C.C.C.G.T.C.G.G.G.T.C.G.A.T.C.T.G.C.T.T.G.G.C.G.G.A.A.T.C.T.G.C.C.C.G.A.G.A.T.G.G.C.C.C.G.G.G.A.T.C.A.C.A.G.C.G.G.G.T.G.T.T.T.G.G.A.G.C.G.G.C.T.T.C.G.T.T.T.A.G.T.A.C.G.C.T.A.T.G.C.A.C.T.A.A.C.C.C.A.C.G.A.A.T.G.G.A.A.A.A.A.A.G.A.G.T.G.G.G.C.C.C.C.A.T.T.G.G.T.G.G.C.G.A.C.T.T.T.C.A.C.C.G.G.A.T.A.T.G.C.A.G.T.C.A.G.T.G.G.T.T.T.T.T.G.T.A.A.T.G.G.G.C.G.G.A.T.A.C.A.A.G.T.G.T.G.A.A.G.A.C.A.T.A.C.T.A.C.A.G.A.G.A.A.G.G.A.T.C.T.T.A.A.A.T.C.A.A.C.T.T.C.C.C.T.G.T.T.A.A.A.A.C.T.A.G.T.A.C.G.G.T.G.A.A.T.C.A.C.G.C.G.T.C.C.T.T.T.G.A.T.A.G.A.G.A.T.G.T.A.C.C.C.T.A.G.G.T.C.T.C.A.C.G.A.A.C.T.A.C.C.A.A.C.A.T.G.A.A.A.G.C.C.C.T.C.T.A.C.T.G.G.C.T.A.G.C.C.G.T.A.G.G.C.T.G.T.C.G.T.A.T.C.C.G.G.G.A.C.G.A.G.G.T.T.A.G.C.T.C.T.T.C.A.T.T.T.T.A.A.A.A.C.A.A.C.C.T.G.C.T.C.G.C.T.C.G.T.C.T.T.T.T.A.T.G.C.C.A.A.C.A.G.C.A.T.T.T.C.C.A.G.C.A.C.A.C.A.T.A.C.C.G.T.G.T.A.T.G.G.T.T.G.T.A.C.T.G.C.G.C.T.C.A.A.A.G.G.C.G.C.T.T.T.A.A.T.G.A.T.G.C.T.A.A.A.T.T.C.A.T.C.C.T.G.G.A.A.T.A.T.A.C.T.A.A.T.A.G.C.G.A.T.A.T.A.G.T.T.G.T.C.G.G.G.C.T.T.A.G.G.C.G.T.C.T.C.T.C.T.G.G.G.A.C.T.G.A.A.A.T.G.C.A.C.C.C.G.C.T.A.A.T.A.A.T.C.G.T.G.C.A.T.A.C.A.T.A.G.C.C.C.T.G.A.C.C.T.G.C.A.A.G.A.T.T.C.C.T.T.G.C.A.A.T.C.T.A.A.C.G.A.G.C.T.C.T.A.G.A.C.T.A.G.C.C.G.G.T.A.T.G.C.T.G.T.T.T.T.T.G.G.A.G.T.G.A.A.G.G.C.G.A.G.T.A.T.G.A.G.T.T.G.G.A.G.A.A.A.G.T.T.C.T.G.G.C.C.G.T.T.G.T.G.C.T.T.C.G.T.C.A.T.C.G.A.A.A.A.G.C.C.G.G.A.A.G.A.C.T.C.T.T.C.C.T.A.T.C.G.A.C.C.A.G.C.T.G.C.T.A.C.G.G.G.A.C.T.T.C.G.A.T.C.A.G.T.G.G.A.C.G.C.G.G.C.G.T.G.A.A.A.T.C.A.G.G.C.T.A.T.C.A.A.T.C.C.C.C.T.T.G.A.G.C.A.A.T.G.C.G.T.G.T.C.A.G.T.T.T.A.T.G.T.T.A.C.G.A.C.A.A.T.T.G.A.C.G.C.A.G.A.A.C.C.C.T.C.G.C.T.A.T.C.A.A.T.T.T.A.T.T.C.G.T.G.T.T.A.T.G.A.T.A.C.T.C.A.A.C.T.A.C.G.T.T.C.C.A.T.T.A.G.G.T.C.A.G.A.C.G.C.T.T.A.G.T.A.A.C.T.T.C.G.T.C.G.T.T.T.A.G.A.C.G.A.T.C.A.C.A.C.A.C.G.T.C.T.G.G.G.T.A.C.A.G.T.G.G.T.T.G.G.T.C.A.C.T.C.C.G.C.T.C.T.A.A.T.T.T.C.A.T.T.A.T.C.A.T.T.C.G.G.G.C.C.A.G.C.A.C.A.C.T.C.C.G.C.G.T.C.T.T.G.A.A.C.T.T.A.A.C.G.A.G.G.A.A.A.T.C.G.G.T.A.G.G.C.T.C.C.C.G.A.G.C.T.C.C.G.T.T.A.T.T.T.C.G.A.C.C.T.G.C.T.C.G.G.T.C.A.G.G.C.C.C.A.A.T.G.G.A.C.T.A.C.C.T.T.T.T.A.C.T.A.C.C.G.A.A.C.A.C.A.G.G.G.G.C.A.T.T.G.G.G.A.G.C.C.G.C.T.C.A.G.A.G.A.T.C.G.T.G.A.C.G.A.G.C.T.G.C.A.A.A.C.C.T.A.T.T.A.C.C.C.T.A.C.T.T.G.A.G.G.T.G.C.A.A.C.T.A.A.C.G.G.G.T.T.C.G.G.G.G.C.T.A.A.G.C.G.A.C.A.T.C.G.G.C.T.A.C.A.A.A.T.A.T.C.A.T.T.G.C.C.G.G.A.T.C.T.T.C.A.G.C.C.C.A.A.A.G.T.G.A.A.C.G.G.G.C.A.A.C.T.C.C.T.A.T.T.G.A.A.T.A.G.T.G.A.G.G.A.A.C.C.C.C.T.G.G.G.A.C.A.G.A.T.C.C.A.A.G.T.G.T.A.T.T.C.A.A.T.C.A.C.T.C.G.T.G.A.G.A.G.C.A.C.A.C.C.T.G.C.T.T.T.T.A.C.G.C.C.G.A.A.T.T.A.T.C.A.T.C.A.A.G.A.T.G.A.C.A.T.G.G.G.G.G.C.G.C.T.T.G.T.T.A.C.G.G.A.T.G.G.G.A.T.G.C.T.G.C.C.A.T.G.C.G.C.T.G.G.T.A.C.A.A.A.A.A.A.A.G.G.T.A.C.G.G.A.C.A.G.G.C.C.T.A.C.A.G.G.A.C.T.A.T.A.C.A.C.G.A.G.G.G.C.T.G.A.G.T.C.C.G.G.A.A.G.T.C.G.A.G.G.G.G.G.G.G.C.T.A.C.T.A.C.C.A.C.C.T.G.A.G.T.C.C.A.G.C.A.A.A.A.C.G.C.T.A.A.T.G.C.T.C.G.C.T.C.C.A.T.G.T.T.C.T.G.C.T.G.C.G.G.G.C.A.T.G.C.G.T.G.G.A.T.T.A.G.G.A.T.T.A.C.T.A.A.C.A.C.A.C.T.A.A.T.A.A.A.T.A.C.A.A.G.G.T.T.A.G.T.A.G.C.A.A.T.C.T.C.T.G.T.T.G.T.A.G.T.A.C.A.A.C.C.C.G.A.G.G.C.T.G.C.C.C.C.T.A.G.G.G.A.T.A.C.C.G.T.G.T.G.T.A.C.T.G.T.T.G.G.A.C.T.C.G.T.T.A.A.C.C.A.C.T.T.A.T.C.C.A.C.A.A.C.T.T.T.C.G.T.G.C.T.A.G.C.T.C.C.G.T.T.G.A.T.A.G.A.A.C.T.C.G.C.T.A.T.G.C.C.G.A.C.A.A.G.C.T.A.G.G.G.G.C.C.T.G.G.G.T.C.C.T.T.T.A.T.T.G.T.A.A.A.G.G.A.G.T.T.T.C.A.T.A.A.C.T.G.G.C.G.G.C.C.A.C.A.T.C.T.T.G.T.A.C.A.C.T.C.T.A.T.C.A.C.A.C.G.A.T.C.C.A.A.A.C.A.T.C.G.T.T.T.G.G.A.T.A.G.C.G.C.G.C.T.C.C.G.G.G.C.T.A.G.T.C.G.C.G.G.C.G.C.A.A.A.T.C.A.G.T.A.T.A.A.T.A.C.G.C.C.A.T.C.G.A.T.A.G.T.T.T.T.G.T.C.T.C.C.A.T.G.T.A.C.A.T.G.T.A.G.C.T.G.G.T.T.A.G.A.G.T.A.C.T.T.C.C.G.A.G.T.T.C.G.G.G.C.A.A.T.C.C.T.T.C.T.T.C.T.G.G.T.C.A.C.T.A.A.A.G.G.G.G.A.G.C.A.G.T.A.A.G.T.G.G.C.G.C.T.T.A.G.A.C.C.T.C.T.G.C.G.T.G.C.A.A.G.C.G.G.A.T.T.G.T.A.G.T.T.T.C.A.T.A.A.C.T.A.T.G.C.T.G.C.A.G.T.T.C.A.C.G.A.T.G.A.A.T.G.G.G.A.T.G.G.C.G.T.T.C.T.G.A.T.C.C.A.T.A.A.C.T.G.G.C.T.A.T.A.A.G.A.T.G.A.A.C.C.A.A.T.C.C.A.T.T.G.G.G.T.A.C.A.A.C.C.T.G.T.G.C.C.G.T.T.C.T.T.G.C.A.C.C.G.A.G.A.A.G.C.A.T.G.A.C.T.T.G.A.A.A.C.G.T.G.C.A.T.A.G.G.A.C.A.T.T.A.T.C.C.G.A.T.C.C.T.C.G.C.A.T.A.C.T.C.A.A.A.A.A.A.G.A.C.A.T.G.A.C.A.A.C.T.C.G.G.A.G.A.A.C.T.A.G.T.A.T.G.A.T.A.G.T.C.C.C.T.T.G.A.G.A.G.A.T.C.A.A.T.C.G.C.A.C.A.T.T.A.C.T.G.T.A.T.G.G.T.A.A.T.C.C.C.C.A.T.C.A.G.C.G.T.G.A.A.T.A.A.A.C.T.G.T.C.C.A.A.A.T.C.T.T.G.T.T.G.T.A.T.G.G.T.C.C.A.A.T.T.T.A.C.C.T.T.C.T.G.A.C.A.G.A.C.T.A.G.A.C.T.G.T.A.A.T.A.A.A.T.T.A.G.A.A.A.A.T.C.T.T.T.T.A.G.A.G.C.T.G.C.G.T.T.T.G.C.T.C.C.G.A.C.T.T.C.T.C.T.A.A.A.T.G.A.A.C.A.C.C.G.G.G.C.T.A.G.C.T.T.A.T.C.T.C.A.A.C.C.C.G.T.C.G.C.G.G.C.C.C.T.T.C.T.A.T.C.T.C.C.C.C.T.G.G.C.G.C.A.C.G.A.A.A.G.G.A.T.T.G.G.T.T.G.A.C.C.G.A.A.C.A.A.A.T.A.C.G.C.G.G.T.G.G.T.T.C.T.C.C.T.A.A.A.A.T.C.C.A.A.A.A.T.T.A.A.G.A.G.G.T.C.T.G.C.T.A.T.A.T.G.C.C.A.G.T.A.C.G.C.G.C.A.T.C.G.G.A.G.C.G.C.G.T.A.G.C.A.A.T.C.G.C.G.C.A.G.T.C.A.A.A.G.C.C.A.G.C.C.G.G.G.G.G.A.T.G.A.A.A.C.A.G.A.A.T.C.G.A.C.G.T.T.A.C.T.C.T.C.C.C.A.C.T.A.T.C.T.T.G.T.T.A.G.A.G.A.T.A.G.G.A.A.C.A.C.G.A.A.A.G.C.G.G.T.T.T.A.T.T.G.T.T.G.A.A.T.A.T.C.A.C.A.T.G.G.A.T.C.G.T.C.G.G.T.G.T.C.C.G.G.T.C.A.A.C.T.C.A.T.C.C.T.G.A.G.C.G.C.T.G.C.C.T.G.G.T.C.C.C.A.G.C.A.T.T.T.C.T.C.A.C.C.G.T.A.G.G.T.G.C.G.T.C.C.A.T.G.C.A.A.G.T.C.T.C.C.T.T.A.T.G.G.T.T.G.G.T.G.C.T.A.C.A.A.G.T.T.G.A.G.G.A.A.T.G.G.G.G.A.C.T.T.A.C.G.T.C.C.T.T.G.C.G.A.T.T.G.A.G.A.C.C.T.C.A.G.T.T.G.C.A.C.A.T.T.T.T.T.T.G.A.G.T.T.T.A.T.G.A.G.C.G.C.G.A.T.A.G.T.C.T.A.A.G.T.T.A.A.C.G.C.C.C.G.T.C.A.A.G.C.T.C.T.A.T.G.C.T.C.T.T.C.G.C.C.G.G.G.C.G.T.G.A.T.T.C.G.C.T.G.A.A.G.G.G.G.C.C.C.C.A.C.G.A.A.G.A.A.C.T.G.C.A.T.T.C.T.C.C.C.T.T.T.C.C.G.G.A.T.C.A.T.A.T.G.T.G.C.A.T.C.T.T.A.G.C.A.T.A.T.A.T.G.G.T.C.G.T.A.C.C.C.G.G.G.T.A.T.T.A.A.G.C.G.A.C.C.G.A.G.C.C.C.C.C.A.G.C.T.T.C.T.A.A.G.G.G.A.C.T.T.A.C.A.T.T.C.T.G.T.G.G.T.A.G.A.G.T.C.T.C.A.C.G.G.C.C.T.T.T.A.T.C.T.T.T.T.T.A.A.C.C.C.G.C.A.G.G.G.G.T.T.A.C.C.C.T.T.G.C.C.A.G.A.C.G.C.A.A.T.G.G.C.G.T.C.C.C.C.G.G.T.T.G.A.A.G.T.G.A.T.C.A.T.T.A.A.C.T.C.A.T.T.G.G.T.A.C.A.T.T.T.C.C.A.C.G.G.T.C.C.T.G.G.G.T.G.A.A.G.C.C.C.C.T.C.C.A.C.G.T.G.T.A.C.G.T.C.G.A.G.A.C.T.C.A.A.C.G.G.C.G.C.A.T.T.C.A.T.G.A.C.G.G.A.G.T.T.C.G.A.G.C.G.T.A.T.A.C.C.C.G.A.G.G.A.G.T.C.A.G.C.G.A.A.T.G.C.T.C.G.T.G.A.T.A.A.T.G.A.A.G.T.G.T.T.C.A.C.A.C.C.T.G.A.C.T.T.A.T.A.A.G.G.A.T.C.C.A.G.A.G.T.A.G.C.A.G.G.C.A.A.T.T.G.G.A.T.G.A.G.A.C.A.T.T.C.G.A.G.A.C.G.C.G.T.C.A.C.G.G.G.A.C.G.C.C.C.G.C.A.C.C.C.A.A.C.C.A.A.G.T.A.T.G.T.T.C.A.A.T.C.T.T.C.G.G.A.T.C.A.C.A.G.A.A.C.G.C.T.A.A.A.C.G.C.G.C.T.G.G.C.A.G.G.G.G.G.C.C.C.G.T.G.G.A.C.G.A.C.C.A.A.T.A.A.T.A.A.G.G.G.C.T.T.C.C.A.A.A.C.C.C.T.G.G.C.T.C.C.T.G.A.T.G.A.A.C.C.A.T.G.C.A.T.A.C.A.A.G.G.A.T.A.C.T.A.C.A.C.A.G.C.T.T.G.T.A.C.C.C.G.T.C.T.C.T.A.C.C.A.G.G.C.G.G.G.G.C.A.A.A.T.T.C.A.T.G.A.G.C.A.G.A.T.C.C.G.G.C.A.T.C.T.G.C.C.G.C.G.T.T.C.G.C.A.G.C.T.G.A.C.T.G.C.A.C.A.A.C.C.C.T.G.T.G.G.T.T.C.C.A.C.A.T.C.A.A.A.G.G.T.T.A.T.A.G.T.G.A.T.T.G.C
Sample 7  ( AFR ): T.A.T.A.T.T.G.A.A.A.T.G.A.C.C.A.G.C.A.T.G.C.C.C.G.C.A.A.G.C.T.A.G.A.G.A.C.C.C.G.C.A.A.C.T.T.A.G.G.C.G.G.C.A.T.A.T.C.C.G.A.A.T.T.C.C.A.G.C.T.T.T.C.A.T.C.G.C.A.C.G.T.C.T.T.C.T.C.T.G.T.C.C.T.C.G.C.G.T.A.A.A.T.G.G.T.G.G.A.C.C.G.G.A.G.C.G.G.T.A.T.T.G.A.A.C.T.T.C.G.C.T.A.C.C.G.A.G.C.C.A.T.G.T.C.A.C.C.A.G.T.T.G.G.A.A.T.A.G.A.T.A.G.G.T.A.T.T.C.A.T.T.G.C.A.A.C.G.T.T.G.C.G.A.T.C.C.C.C.T.T.G.G.A.C.C.A.G.A.A.G.A.C.C.G.A.A.A.T.T.A.G.C.T.A.T.C.A.G.C.A.T.G.T.A.G.G.A.A.T.C.T.G.C.C.G.G.A.G.A.G.G.G.C.C.C.G.G.A.A.T.C.A.C.A.G.C.G.G.A.T.G.T.T.T.G.G.A.G.C.G.G.A.T.T.C.G.T.T.T.A.G.T.A.C.G.C.T.A.T.G.C.A.C.T.A.A.C.C.C.A.C.T.A.A.T.G.G.C.A.A.A.A.A.G.A.G.T.G.T.G.C.G.C.C.A.T.T.G.G.T.G.G.C.G.A.C.G.T.T.C.C.C.C.G.G.C.T.G.T.G.T.A.G.T.C.C.G.T.G.G.T.T.A.C.C.G.T.C.A.T.A.A.G.G.G.G.A.T.A.C.A.A.G.T.G.T.G.A.A.G.A.C.A.T.A.C.C.A.C.A.G.A.G.A.A.G.G.A.T.C.T.T.A.A.A.T.C.G.A.C.T.T.C.C.C.T.G.T.T.G.A.A.A.C.T.A.G.T.A.C.G.G.T.G.A.A.T.C.A.C.G.C.G.T.C.C.T.T.T.G.A.C.A.G.A.G.G.T.G.C.A.C.C.C.A.A.G.G.T.A.T.G.A.C.G.A.A.A.T.A.C.C.A.A.C.A.T.T.C.A.T.G.G.C.G.T.G.T.A.G.T.T.T.C.T.A.G.C.G.C.T.A.G.G.C.T.T.T.C.A.A.A.C.C.T.G.G.G.A.C.G.G.C.G.G.T.A.G.G.C.C.T.T.C.T.G.T.T.T.A.A.A.A.C.A.C.C.C.T.G.G.T.A.T.A.C.C.G.T.C.T.T.T.T.A.G.G.C.C.A.A.C.A.G.C.A.T.T.T.C.C.A.G.C.A.C.A.C.A.T.A.C.C.G.T.T.T.A.T.G.G.T.T.G.T.A.C.A.G.G.G.C.T.A.A.C.A.G.T.C.G.C.C.G.T.G.G.C.G.T.T.A.A.A.A.A.C.T.C.C.A.T.C.C.G.G.G.A.A.T.A.T.T.C.C.C.A.G.T.C.A.G.T.A.C.G.C.G.T.C.C.C.C.G.C.C.C.G.G.A.G.A.C.C.G.C.G.C.T.C.T.G.G.A.A.C.T.T.A.A.A.T.G.C.A.A.A.T.T.A.C.A.A.C.C.A.T.A.G.T.C.C.T.G.A.A.T.T.A.G.C.G.C.T.G.G.G.C.C.G.C.A.A.T.G.T.T.A.A.A.T.G.C.A.C.T.C.T.A.A.C.G.A.T.C.T.A.T.A.C.A.C.T.A.T.C.C.G.G.T.A.T.G.G.T.G.A.C.T.A.T.G.G.G.G.G.T.C.G.T.T.T.G.A.T.A.G.T.G.A.C.T.T.G.G.A.G.A.A.A.T.T.G.A.T.G.G.C.C.G.T.T.G.T.C.C.G.T.C.A.A.G.A.T.C.C.A.G.A.C.T.G.T.G.A.C.G.G.A.C.T.C.T.C.C.T.C.A.T.C.G.A.G.C.G.G.C.T.G.G.G.A.A.G.C.G.C.C.C.G.C.G.A.T.T.C.C.T.C.G.A.C.C.A.A.G.C.G.C.C.A.A.C.C.G.T.C.C.T.C.C.A.G.C.C.T.G.A.G.C.T.T.A.A.G.T.T.G.T.G.C.G.T.G.T.C.A.G.T.T.G.C.T.G.T.T.A.T.G.C.C.C.A.T.T.G.T.C.A.A.T.G.A.A.C.C.C.T.C.G.T.T.A.T.G.A.T.T.T.T.A.T.A.G.A.T.G.G.T.A.T.G.T.T.A.A.T.G.A.A.C.T.A.C.G.T.T.C.T.A.T.T.A.G.G.T.C.A.G.A.C.G.C.T.T.A.G.T.A.A.A.C.T.G.G.T.G.G.T.A.T.T.G.A.C.C.A.T.G.G.A.A.C.A.C.G.T.C.T.T.G.G.T.A.C.A.G.T.G.G.A.A.G.C.T.C.A.C.C.G.C.C.C.T.G.T.A.C.T.T.T.A.A.T.C.A.G.C.A.T.T.T.G.G.G.A.C.T.G.T.A.T.T.A.T.A.T.T.C.G.T.T.G.T.C.A.A.C.T.A.A.A.C.G.A.T.G.A.C.G.T.C.T.G.T.A.G.G.C.C.C.C.C.C.A.G.C.T.C.C.G.T.G.A.T.T.T.C.G.A.C.C.T.G.C.T.T.G.G.T.C.A.G.G.C.C.A.A.A.T.G.G.A.C.C.A.C.A.T.A.T.T.A.C.T.A.A.C.G.A.A.C.A.C.A.G.G.G.G.C.A.T.T.G.G.G.A.G.C.C.C.C.T.C.C.G.A.C.A.T.T.A.T.C.A.C.G.A.G.C.T.G.C.A.A.G.A.C.T.A.T.C.A.C.T.G.A.A.C.A.T.G.A.G.G.T.G.G.T.G.C.T.A.A.G.A.G.A.T.G.C.G.C.G.G.C.T.A.A.G.C.C.A.C.C.C.C.T.G.C.G.C.C.A.T.A.T.A.T.C.A.T.T.G.T.C.G.G.A.T.G.T.T.C.A.G.C.G.C.A.A.A.G.A.G.G.A.G.G.T.A.C.C.G.C.G.A.C.G.G.C.C.T.C.G.T.A.G.T.G.T.T.G.C.G.A.C.G.T.G.A.A.A.A.T.A.A.C.T.A.C.G.C.A.G.C.G.T.G.T.C.G.A.G.T.C.C.C.T.T.T.A.A.G.A.G.A.A.C.G.C.G.G.G.T.T.C.T.C.A.C.C.C.C.A.G.A.C.T.G.T.G.A.C.C.A.A.G.A.A.G.A.C.C.A.A.G.G.G.G.C.A.G.T.T.G.T.T.A.C.G.G.A.T.G.G.G.A.T.G.C.A.C.A.C.A.T.G.C.G.C.T.G.G.A.A.T.A.A.A.A.A.T.A.A.C.G.A.C.G.G.A.C.A.A.G.C.C.T.C.C.A.G.G.G.C.G.C.T.A.T.A.C.G.A.G.G.G.C.T.G.C.G.T.C.A.T.G.C.A.T.C.C.G.A.G.G.G.C.G.A.C.A.G.G.T.G.G.C.C.G.C.C.T.G.G.T.G.C.G.A.G.C.A.G.A.A.G.G.C.T.T.T.T.C.C.T.C.G.C.G.C.C.A.T.G.T.G.T.T.G.A.T.G.C.G.G.C.C.A.T.G.C.G.T.G.G.A.C.T.A.G.G.A.T.T.A.T.T.C.G.C.A.T.A.G.T.G.T.T.A.A.A.G.G.C.A.C.A.G.A.T.A.G.T.A.G.C.G.A.T.G.A.C.T.A.T.T.G.T.A.G.T.T.A.A.A.C.A.G.G.C.G.T.C.A.C.G.A.C.C.T.A.A.G.A.A.T.A.C.C.G.T.A.A.T.T.C.C.T.T.T.T.A.C.A.C.T.T.G.T.T.A.T.C.C.C.C.T.T.T.T.C.T.A.A.A.A.G.T.T.T.C.G.A.G.A.G.A.G.G.A.C.C.G.T.T.G.A.T.C.G.C.A.C.T.C.G.C.T.A.T.G.T.T.G.A.C.A.A.G.C.T.G.G.A.G.C.C.A.T.G.G.G.C.C.C.T.C.T.A.T.C.G.T.A.A.A.G.G.A.G.T.T.T.C.A.T.A.A.C.T.C.C.C.A.G.C.C.A.C.A.A.C.T.T.G.T.A.C.A.C.T.C.T.A.T.C.T.C.A.C.G.A.T.C.C.G.A.A.C.A.T.C.G.T.T.T.G.G.A.T.A.G.C.G.C.G.C.T.C.C.C.G.C.G.T.A.G.T.C.G.C.A.G.C.A.C.G.A.A.G.T.A.G.A.A.T.A.A.T.A.C.G.C.C.A.T.G.G.A.T.A.G.T.T.T.T.G.T.C.T.C.C.A.T.G.T.A.C.A.T.G.T.A.G.C.T.G.G.G.T.A.G.A.G.T.A.C.T.T.C.C.T.A.G.T.T.C.G.G.G.C.A.A.T.C.C.T.T.C.T.T.C.T.G.G.T.C.A.C.T.A.A.A.G.G.G.G.A.G.C.A.G.T.A.A.G.G.T.G.C.G.C.T.T.A.G.A.C.C.A.C.T.G.T.G.T.G.C.A.A.G.C.G.G.A.T.T.G.T.A.G.T.T.T.C.A.T.A.A.C.T.A.T.G.C.T.G.C.A.G.T.T.C.A.C.G.A.T.G.A.A.C.G.G.G.A.T.G.G.C.G.T.T.C.T.G.A.T.C.C.A.T.A.A.C.T.G.G.C.G.A.T.A.A.G.A.T.G.A.A.C.C.A.A.T.C.C.A.T.T.G.G.G.T.A.C.A.A.C.C.T.G.T.G.C.C.G.T.T.C.T.T.G.C.A.C.C.G.A.G.A.A.G.C.A.T.G.A.C.T.T.G.A.A.A.C.G.T.G.C.A.T.A.G.G.A.C.A.T.T.A.T.C.C.G.A.T.C.C.T.C.G.C.A.T.A.C.T.C.A.A.A.C.A.A.G.A.C.A.T.G.A.C.A.A.C.T.C.C.G.A.G.A.A.C.T.A.G.T.A.T.G.A.T.A.G.T.C.C.C.G.G.G.A.G.G.G.A.A.A.A.A.T.G.G.C.A.C.G.T.T.A.G.A.G.T.A.T.G.G.T.A.A.T.C.C.C.C.A.T.C.A.G.C.G.T.G.A.A.T.A.A.A.C.T.G.T.C.C.A.A.A.T.C.T.C.G.T.T.G.T.A.T.G.A.T.C.C.A.A.A.T.T.A.C.C.T.T.C.G.G.T.C.A.G.G.C.T.A.G.A.C.T.G.T.A.A.T.A.A.A.T.T.A.G.A.A.A.A.T.C.T.T.T.T.A.G.A.G.C.T.G.C.G.T.T.T.G.C.C.A.C.C.A.C.T.T.C.T.T.A.C.A.A.T.G.C.A.C.A.C.C.G.G.T.T.T.G.G.C.T.T.A.A.T.T.C.G.T.T.A.A.C.C.C.G.C.T.A.C.C.C.C.A.C.T.A.T.C.T.C.C.C.C.T.G.G.C.T.A.T.A.A.A.G.A.A.G.A.A.C.G.A.T.T.A.A.A.C.A.A.A.T.A.A.A.G.C.C.G.C.G.G.T.A.G.T.T.C.T.C.C.C.A.A.A.A.T.C.C.C.A.G.A.T.C.A.A.G.T.C.C.T.C.G.A.C.A.A.T.G.T.G.G.C.A.G.C.A.C.C.C.A.C.T.C.C.G.A.A.G.C.C.G.C.T.G.T.C.A.G.C.T.G.G.G.T.G.A.G.T.A.A.T.G.T.C.C.A.C.C.G.C.G.G.G.A.T.G.A.A.A.G.C.T.A.A.T.T.G.A.G.G.T.T.C.C.G.C.T.C.C.T.T.C.T.T.T.G.A.T.A.A.T.T.C.T.T.C.G.A.G.C.C.A.C.A.C.G.T.A.A.T.C.G.G.C.T.T.A.G.T.A.T.T.G.A.A.G.C.T.C.A.C.A.T.A.G.A.T.G.G.T.C.G.G.T.G.T.C.C.G.G.T.T.A.A.C.T.C.A.T.C.C.A.G.C.G.C.G.C.G.G.T.C.T.G.G.T.C.C.C.A.G.C.A.T.A.T.T.T.T.T.A.C.G.T.A.G.G.T.A.A.A.T.C.C.A.T.G.A.A.A.G.A.C.C.C.G.T.T.A.A.G.G.T.T.T.G.A.A.C.T.A.C.A.A.G.T.A.G.A.G.G.G.G.T.T.T.G.G.A.C.C.T.C.G.G.G.C.A.G.T.G.C.C.A.A.T.G.T.G.A.C.T.T.C.A.G.T.T.C.C.A.G.A.T.T.G.T.T.T.T.A.G.T.G.A.A.T.G.T.G.C.C.G.G.C.T.C.G.T.C.A.A.G.G.T.T.A.A.C.G.C.C.C.G.T.T.A.A.G.C.T.C.T.G.T.G.C.T.C.T.T.C.G.C.C.G.G.G.C.G.T.G.A.T.T.C.G.C.T.G.A.A.G.G.G.A.T.C.C.C.A.C.G.A.A.G.A.A.C.T.G.C.A.T.G.C.T.C.C.G.T.T.T.C.C.T.C.A.T.T.A.C.A.T.G.T.G.C.A.T.C.C.T.A.T.A.T.A.G.G.A.T.G.G.T.A.C.T.A.C.C.C.G.G.G.T.A.T.T.A.A.G.C.G.A.G.C.G.T.T.C.G.G.C.C.A.G.C.G.T.C.T.A.A.G.G.G.A.C.T.T.A.C.A.T.T.A.T.G.T.G.G.T.A.G.A.G.T.C.T.C.A.C.C.G.C.C.T.T.T.A.G.C.T.T.T.T.T.A.A.C.C.C.G.C.A.G.G.G.G.T.T.A.C.A.C.T.T.G.C.C.A.G.A.C.G.C.A.A.T.G.G.C.G.T.C.C.C.C.G.G.T.T.T.T.A.G.A.G.A.T.C.C.T.C.A.A.C.G.T.A.T.T.G.A.T.G.C.C.T.T.T.T.C.A.C.G.G.T.T.C.G.G.G.G.T.G.A.A.G.C.C.C.C.T.C.C.A.C.G.T.G.T.A.C.G.T.C.G.A.G.A.C.G.C.A.A.C.G.G.C.G.C.A.T.T.A.A.G.G.A.C.G.G.A.G.T.T.C.G.A.G.T.G.T.A.T.A.C.C.C.G.T.G.A.A.G.T.C.A.G.A.G.A.G.T.G.C.G.C.T.T.G.A.T.A.A.T.G.A.A.G.G.A.T.T.C.A.C.A.A.C.G.G.T.C.T.T.A.T.G.A.T.T.G.A.A.C.A.G.A.G.T.A.G.C.A.G.T.C.A.C.T.T.G.G.A.T.G.A.G.C.A.A.T.T.C.C.A.A.C.T.G.C.G.G.T.A.C.G.T.G.C.C.G.C.C.C.G.C.A.C.T.C.A.A.G.A.A.A.T.T.A.T.G.T.T.C.A.A.G.C.T.T.C.G.T.A.A.C.A.C.A.G.A.A.T.G.C.T.A.A.A.C.G.C.G.C.T.G.G.C.A.G.G.G.G.G.C.C.C.G.T.G.G.A.C.G.A.C.C.A.A.T.A.A.T.A.A.G.G.G.T.G.T.C.C.A.A.A.G.C.C.T.G.G.C.T.C.C.T.G.A.T.G.A.C.C.C.A.T.G.C.A.T.C.C.C.A.G.G.A.T.A.C.T.A.C.A.C.A.G.C.T.T.G.T.A.C.C.C.C.T.C.T.C.T.A.A.C.A.G.G.C.G.G.G.G.C.A.A.A.T.T.C.A.T.G.A.G.C.A.T.A.A.T.C.A.G.C.A.A.G.T.A.C.C.A.C.T.T.T.C.C.C.A.G.C.T.G.A.C.T.G.C.A.C.A.A.C.C.C.T.T.T.A.A.C.T.C.T.A.C.A.T.C.A.A.A.G.A.T.T.A.T.C.C.T.G.A.T.A.T.C
</pre></div>
</div>
</div>
</div>
<p>You can easily obtain the
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.allele_frequency_spectrum" title="(in Project name not set)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.allele_frequency_spectrum()</span></code></a> for the entire region (or for
<a class="reference internal" href="getting_started.html#sec-tskit-getting-started-compute-statistics-windowing"><span class="std std-ref">windowed regions</span></a>)
directly from the tree sequence (i.e. without needing to reconstruct genotypes)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">afs</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">allele_frequency_spectrum</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">num_samples</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">afs</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Allele frequency spectrum&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/8ece7d03d4762e2bcaa6aa892c77c3c7100997c0287fcb253eb1c04df9088e1c.png" src="_images/8ece7d03d4762e2bcaa6aa892c77c3c7100997c0287fcb253eb1c04df9088e1c.png" />
</div>
</div>
<p>Similarly <code class="docutils literal notranslate"><span class="pre">tskit</span></code> allows fast and easy
<a class="reference internal" href="analysing_tree_sequences.html#sec-tutorial-stats"><span class="std std-ref">calculation of statistics</span></a> along the genome. Here is
a plot of windowed <span class="math notranslate nohighlight">\(F_{st}\)</span> between Africans and admixed Americans over this
region of chromosome:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define the samples between which Fst will be calculated</span>
<span class="n">pop_id</span> <span class="o">=</span> <span class="p">{</span><span class="n">p</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]:</span> <span class="n">p</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">populations</span><span class="p">()}</span>
<span class="n">sample_sets</span><span class="o">=</span><span class="p">[</span><span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">pop_id</span><span class="p">[</span><span class="s2">&quot;AFR&quot;</span><span class="p">]),</span> <span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">pop_id</span><span class="p">[</span><span class="s2">&quot;ADMIX&quot;</span><span class="p">])]</span>

<span class="c1"># Do the windowed calculation, using windows of 10 kilobases</span>
<span class="n">windows</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="mi">10_000</span><span class="p">))</span>
<span class="n">F_st</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">Fst</span><span class="p">(</span><span class="n">sample_sets</span><span class="p">,</span> <span class="n">windows</span><span class="o">=</span><span class="n">windows</span><span class="p">)</span>

<span class="c1"># Plot</span>
<span class="n">plt</span><span class="o">.</span><span class="n">stairs</span><span class="p">(</span><span class="n">F_st</span><span class="p">,</span> <span class="n">windows</span><span class="p">,</span> <span class="n">baseline</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;AFR-ADMIX Fst&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Genome position&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/5041104ba6f1dfeb0d3163ad610e4656e05f8a8e6edd73807bafff0e560c4ceb.png" src="_images/5041104ba6f1dfeb0d3163ad610e4656e05f8a8e6edd73807bafff0e560c4ceb.png" />
</div>
</div>
<p>Extracting the genetic tree at a specific genomic location is easy using <code class="docutils literal notranslate"><span class="pre">tskit</span></code>, which
also provides methods to <a class="reference internal" href="viz.html#sec-tskit-viz"><span class="std std-ref">plot</span></a> these trees. Here we
grab the tree at position 10kb, and colour the different populations by
grab the tree at position 10kb, and colour the samples according to their population,
as described in the <a class="reference internal" href="viz.html#sec-tskit-viz-styling"><span class="std std-ref">viz tutorial</span></a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tree</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">at</span><span class="p">(</span><span class="mi">10_000</span><span class="p">)</span>

<span class="n">colours</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">AFR</span><span class="o">=</span><span class="s2">&quot;yellow&quot;</span><span class="p">,</span> <span class="n">EUR</span><span class="o">=</span><span class="s2">&quot;cyan&quot;</span><span class="p">,</span> <span class="n">ASIA</span><span class="o">=</span><span class="s2">&quot;green&quot;</span><span class="p">,</span> <span class="n">ADMIX</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">)</span>
<span class="n">styles</span> <span class="o">=</span> <span class="p">[</span>
    <span class="sa">f</span><span class="s2">&quot;.leaf.p</span><span class="si">{</span><span class="n">pop</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> &gt; .sym </span><span class="se">{{</span><span class="s2">fill: </span><span class="si">{</span><span class="n">colours</span><span class="p">[</span><span class="n">pop</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]]</span><span class="si">}</span><span class="se">}}</span><span class="s2">&quot;</span>
    <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">populations</span><span class="p">()</span>
<span class="p">]</span>

<span class="n">styles</span> <span class="o">+=</span> <span class="p">[</span> <span class="c1"># rotate the population labels, etc</span>
    <span class="s2">&quot;.leaf &gt; .lab {text-anchor: start; transform: rotate(90deg) translate(6px)}&quot;</span><span class="p">,</span>
    <span class="s2">&quot;.leaf &gt; .sym {stroke: black}&quot;</span>
<span class="p">]</span>

<span class="n">labels</span> <span class="o">=</span> <span class="p">{</span> <span class="c1"># Label samples by population</span>
    <span class="n">u</span><span class="p">:</span> <span class="n">ts</span><span class="o">.</span><span class="n">population</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="n">u</span><span class="p">)</span><span class="o">.</span><span class="n">population</span><span class="p">)</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">()</span>
<span class="p">}</span>

<span class="n">tree</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span>
    <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">800</span><span class="p">,</span> <span class="mi">500</span><span class="p">),</span>
    <span class="n">canvas_size</span><span class="o">=</span><span class="p">(</span><span class="mi">800</span><span class="p">,</span> <span class="mi">520</span><span class="p">),</span>
    <span class="n">node_labels</span><span class="o">=</span><span class="n">labels</span><span class="p">,</span>
    <span class="n">style</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">styles</span><span class="p">),</span>
    <span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">y_ticks</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">30_000</span><span class="p">,</span> <span class="mi">10_000</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/20bfa553801f863774ff3f034e21d91887e5dc85388799e3f94fad0dde43bdb8.svg" src="_images/20bfa553801f863774ff3f034e21d91887e5dc85388799e3f94fad0dde43bdb8.svg" />
</div>
</div>
<p>Or we can plot a principal components analysis of the genome, which should reflect
geographical distinctiveness:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">matplotlib.patches</span><span class="w"> </span><span class="kn">import</span> <span class="n">Patch</span>

<span class="c1"># Run the Principal Components Analysis (PCA)</span>
<span class="n">pca_obj</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">num_components</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

<span class="c1"># Plot the PCA &quot;factors&quot;</span>
<span class="n">col_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">colours</span><span class="p">[</span><span class="n">pop</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]]</span> <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">populations</span><span class="p">()]</span>
<span class="n">sample_pop_ids</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">nodes_population</span><span class="p">[</span><span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">()]</span>
<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="o">*</span><span class="n">pca_obj</span><span class="o">.</span><span class="n">factors</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="p">[</span><span class="n">col_list</span><span class="p">[</span><span class="n">p</span><span class="p">]</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">sample_pop_ids</span><span class="p">],</span> <span class="n">edgecolors</span><span class="o">=</span> <span class="s2">&quot;black&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;PCA 1&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;PCA 2&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="p">[</span>
    <span class="n">Patch</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">col_list</span><span class="p">[</span><span class="n">pop</span><span class="o">.</span><span class="n">id</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="n">pop</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">])</span> <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">populations</span><span class="p">()</span>
<span class="p">]);</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/42b8505d6a54e161c8ca179a92ff773bea38e3e2fd0b08e1bb3fbfbd12fe5e47.png" src="_images/42b8505d6a54e161c8ca179a92ff773bea38e3e2fd0b08e1bb3fbfbd12fe5e47.png" />
</div>
</div>
</section>
<section id="population-genetic-inference">
<h2>Population genetic inference<a class="headerlink" href="#population-genetic-inference" title="Link to this heading">#</a></h2>
<p>If, instead of simulations, you want to analyse existing genomic data (for example
stored in a VCF file), you will need to infer a tree sequence from it, using e.g.
<a class="reference external" href="https://tskit.dev/tsinfer/docs/stable/">tsinfer</a>. Here we load an illustrative portion
of an <a class="reference external" href="https://zenodo.org/record/5512994">inferred tree sequence</a>
based on about 7500 public human genomes, including genomes from the
<a class="reference external" href="https://www.internationalgenome.org/data-portal/data-collection/grch38">Thousand Genomes Project</a> and
<a class="reference external" href="https://www.internationalgenome.org/data-portal/data-collection/hgdp">Human Genome Diversity Project</a>.
The genomic region encoded in this tree sequence has been cut down to
span positions 108Mb-110Mb of human chromosome 2, which spans the
<a class="reference external" href="https://en.wikipedia.org/wiki/Ectodysplasin_A_receptor">EDAR</a> gene.</p>
<p>Note that we are using <a class="reference external" href="https://tskit.dev/tszip/docs/latest/python-api.html#tszip.load" title="(in Project name not set)"><code class="docutils literal notranslate"><span class="pre">tszip.load()</span></code></a> to load the file, as this
utility can also read and write compressed tree sequences in <code class="docutils literal notranslate"><span class="pre">.tsz</span></code> format.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">tszip</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">tszip</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;data/unified_genealogy_2q_108Mb-110Mb.tsz&quot;</span><span class="p">)</span>

<span class="c1"># The ts encompasses a region on chr 2 with an interesting SNP (rs3827760) in the EDAR gene</span>
<span class="n">edar_gene_bounds</span> <span class="o">=</span> <span class="p">[</span><span class="mi">108_894_471</span><span class="p">,</span> <span class="mi">108_989_220</span><span class="p">]</span>  <span class="c1"># In Mb from the start of chromosome 2</span>
<span class="n">focal_variant</span> <span class="o">=</span> <span class="p">[</span><span class="n">v</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">variants</span><span class="p">()</span> <span class="k">if</span> <span class="n">v</span><span class="o">.</span><span class="n">site</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ID&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;rs3827760&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;An interesting SNP within the EDAR gene:&quot;</span><span class="p">)</span>
<span class="n">focal_variant</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>An interesting SNP within the EDAR gene:
</pre></div>
</div>
<div class="output text_html">
        <div>
            <style>
                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}
                .tskit-table tbody tr td {padding: 0.5em 0.5em;}
                .tskit-table tbody tr td:first-of-type {text-align: left;}
                .tskit-details-label {vertical-align: top; padding-right:5px;}
                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}
                .tskit-table-set-table {margin: 12px 0 0 12px;}
                details {display: inline-block;}
                summary {cursor: pointer; outline: 0; display: list-item;}
            </style>
            <div class="tskit-table-set">
                <div class="tskit-table-set-table">
                <table class="tskit-table">
                    <thead>
                    <tr>
                        <th style="padding:0;line-height:21px;">
                        <img style="height: 32px;display: inline-block;padding: 3px 5px 3px 0;" src="https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg"/>
                        <a target="_blank" href="https://tskit.dev/tskit/docs/latest/python-api.html#the-variant-class"> Variant </a>
                        </th>
                    </tr>
                    </thead>
                    <tbody>
        
                <tr><td>Site Id</td><td>9 378</td></tr>
                <tr><td>Site Position</td><td>1.0889714e+08</td></tr>
                <tr><td>Number of Nodes</td><td>836</td></tr>
                <tr><td>Number of Alleles</td><td>2</td></tr>
            <tr><td>Nodes with Allele 'A'</td><td>384 (46%)</td></tr>
<tr><td>Nodes with Allele 'G'</td><td>452 (54%)</td></tr>
                <tr><td>Has Missing Data</td><td>False</td></tr>
                <tr><td>Isolated as Missing</td><td>True</td></tr>
            
                    </tbody>
            </table>
            </div>
        </div>
        </div>
        </div></div>
</div>
<p>For simplicity, this tree sequence has been <a class="reference internal" href="simplification.html#sec-simplification"><span class="std std-ref">simplified</span></a> to
include only those samples from the African and East Asian regions. These belong to a
number of populations. The population information, as well as information describing the
variable sites, is stored in tree sequence <a class="reference internal" href="metadata.html#sec-tutorial-metadata"><span class="std std-ref">metadata</span></a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>

<span class="nb">print</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">num_populations</span><span class="p">,</span> <span class="s2">&quot;populations defined in the tree sequence:&quot;</span><span class="p">)</span>

<span class="n">pop_names_regions</span> <span class="o">=</span> <span class="p">[</span>
    <span class="p">[</span><span class="n">p</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;name&quot;</span><span class="p">),</span> <span class="n">p</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;region&quot;</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">population</span><span class="o">=</span><span class="n">p</span><span class="o">.</span><span class="n">id</span><span class="p">))]</span>
    <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">populations</span><span class="p">()</span>
<span class="p">]</span>
<span class="k">with</span> <span class="n">pd</span><span class="o">.</span><span class="n">option_context</span><span class="p">(</span><span class="s1">&#39;display.max_rows&#39;</span><span class="p">,</span> <span class="mi">100</span><span class="p">):</span>
    <span class="n">display</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pop_names_regions</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;region&quot;</span><span class="p">,</span> <span class="s2">&quot;# genomes&quot;</span><span class="p">]))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>67 populations defined in the tree sequence:
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>region</th>
      <th># genomes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>BantuKenya</td>
      <td>AFRICA</td>
      <td>22</td>
    </tr>
    <tr>
      <th>1</th>
      <td>BantuSouthAfrica</td>
      <td>AFRICA</td>
      <td>16</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Biaka</td>
      <td>AFRICA</td>
      <td>44</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Cambodian</td>
      <td>EAST_ASIA</td>
      <td>18</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Dai</td>
      <td>EAST_ASIA</td>
      <td>18</td>
    </tr>
    <tr>
      <th>5</th>
      <td>Daur</td>
      <td>EAST_ASIA</td>
      <td>18</td>
    </tr>
    <tr>
      <th>6</th>
      <td>Han</td>
      <td>EAST_ASIA</td>
      <td>66</td>
    </tr>
    <tr>
      <th>7</th>
      <td>Hezhen</td>
      <td>EAST_ASIA</td>
      <td>18</td>
    </tr>
    <tr>
      <th>8</th>
      <td>Japanese</td>
      <td>EAST_ASIA</td>
      <td>54</td>
    </tr>
    <tr>
      <th>9</th>
      <td>Lahu</td>
      <td>EAST_ASIA</td>
      <td>16</td>
    </tr>
    <tr>
      <th>10</th>
      <td>Mandenka</td>
      <td>AFRICA</td>
      <td>44</td>
    </tr>
    <tr>
      <th>11</th>
      <td>Mbuti</td>
      <td>AFRICA</td>
      <td>26</td>
    </tr>
    <tr>
      <th>12</th>
      <td>Miao</td>
      <td>EAST_ASIA</td>
      <td>20</td>
    </tr>
    <tr>
      <th>13</th>
      <td>Mongolian</td>
      <td>EAST_ASIA</td>
      <td>18</td>
    </tr>
    <tr>
      <th>14</th>
      <td>Naxi</td>
      <td>EAST_ASIA</td>
      <td>16</td>
    </tr>
    <tr>
      <th>15</th>
      <td>NorthernHan</td>
      <td>EAST_ASIA</td>
      <td>20</td>
    </tr>
    <tr>
      <th>16</th>
      <td>Oroqen</td>
      <td>EAST_ASIA</td>
      <td>18</td>
    </tr>
    <tr>
      <th>17</th>
      <td>San</td>
      <td>AFRICA</td>
      <td>12</td>
    </tr>
    <tr>
      <th>18</th>
      <td>She</td>
      <td>EAST_ASIA</td>
      <td>20</td>
    </tr>
    <tr>
      <th>19</th>
      <td>Tu</td>
      <td>EAST_ASIA</td>
      <td>20</td>
    </tr>
    <tr>
      <th>20</th>
      <td>Tujia</td>
      <td>EAST_ASIA</td>
      <td>18</td>
    </tr>
    <tr>
      <th>21</th>
      <td>Xibo</td>
      <td>EAST_ASIA</td>
      <td>18</td>
    </tr>
    <tr>
      <th>22</th>
      <td>Yakut</td>
      <td>EAST_ASIA</td>
      <td>50</td>
    </tr>
    <tr>
      <th>23</th>
      <td>Yi</td>
      <td>EAST_ASIA</td>
      <td>20</td>
    </tr>
    <tr>
      <th>24</th>
      <td>Yoruba</td>
      <td>AFRICA</td>
      <td>44</td>
    </tr>
    <tr>
      <th>25</th>
      <td>Ami</td>
      <td>EastAsia</td>
      <td>4</td>
    </tr>
    <tr>
      <th>26</th>
      <td>Atayal</td>
      <td>EastAsia</td>
      <td>2</td>
    </tr>
    <tr>
      <th>27</th>
      <td>BantuHerero</td>
      <td>Africa</td>
      <td>4</td>
    </tr>
    <tr>
      <th>28</th>
      <td>BantuKenya</td>
      <td>Africa</td>
      <td>4</td>
    </tr>
    <tr>
      <th>29</th>
      <td>BantuTswana</td>
      <td>Africa</td>
      <td>4</td>
    </tr>
    <tr>
      <th>30</th>
      <td>Biaka</td>
      <td>Africa</td>
      <td>4</td>
    </tr>
    <tr>
      <th>31</th>
      <td>Burmese</td>
      <td>EastAsia</td>
      <td>4</td>
    </tr>
    <tr>
      <th>32</th>
      <td>Cambodian</td>
      <td>EastAsia</td>
      <td>4</td>
    </tr>
    <tr>
      <th>33</th>
      <td>Dai</td>
      <td>EastAsia</td>
      <td>8</td>
    </tr>
    <tr>
      <th>34</th>
      <td>Daur</td>
      <td>EastAsia</td>
      <td>2</td>
    </tr>
    <tr>
      <th>35</th>
      <td>Dinka</td>
      <td>Africa</td>
      <td>6</td>
    </tr>
    <tr>
      <th>36</th>
      <td>Esan</td>
      <td>Africa</td>
      <td>4</td>
    </tr>
    <tr>
      <th>37</th>
      <td>Gambian</td>
      <td>Africa</td>
      <td>4</td>
    </tr>
    <tr>
      <th>38</th>
      <td>Han</td>
      <td>EastAsia</td>
      <td>6</td>
    </tr>
    <tr>
      <th>39</th>
      <td>Hezhen</td>
      <td>EastAsia</td>
      <td>4</td>
    </tr>
    <tr>
      <th>40</th>
      <td>Japanese</td>
      <td>EastAsia</td>
      <td>6</td>
    </tr>
    <tr>
      <th>41</th>
      <td>Ju_hoan_North</td>
      <td>Africa</td>
      <td>8</td>
    </tr>
    <tr>
      <th>42</th>
      <td>Khomani_San</td>
      <td>Africa</td>
      <td>4</td>
    </tr>
    <tr>
      <th>43</th>
      <td>Kinh</td>
      <td>EastAsia</td>
      <td>4</td>
    </tr>
    <tr>
      <th>44</th>
      <td>Korean</td>
      <td>EastAsia</td>
      <td>4</td>
    </tr>
    <tr>
      <th>45</th>
      <td>Lahu</td>
      <td>EastAsia</td>
      <td>4</td>
    </tr>
    <tr>
      <th>46</th>
      <td>Luhya</td>
      <td>Africa</td>
      <td>4</td>
    </tr>
    <tr>
      <th>47</th>
      <td>Luo</td>
      <td>Africa</td>
      <td>4</td>
    </tr>
    <tr>
      <th>48</th>
      <td>Mandenka</td>
      <td>Africa</td>
      <td>6</td>
    </tr>
    <tr>
      <th>49</th>
      <td>Masai</td>
      <td>Africa</td>
      <td>4</td>
    </tr>
    <tr>
      <th>50</th>
      <td>Mbuti</td>
      <td>Africa</td>
      <td>8</td>
    </tr>
    <tr>
      <th>51</th>
      <td>Mende</td>
      <td>Africa</td>
      <td>4</td>
    </tr>
    <tr>
      <th>52</th>
      <td>Miao</td>
      <td>EastAsia</td>
      <td>4</td>
    </tr>
    <tr>
      <th>53</th>
      <td>Mozabite</td>
      <td>Africa</td>
      <td>4</td>
    </tr>
    <tr>
      <th>54</th>
      <td>Naxi</td>
      <td>EastAsia</td>
      <td>4</td>
    </tr>
    <tr>
      <th>55</th>
      <td>Oroqen</td>
      <td>EastAsia</td>
      <td>4</td>
    </tr>
    <tr>
      <th>56</th>
      <td>Saharawi</td>
      <td>Africa</td>
      <td>4</td>
    </tr>
    <tr>
      <th>57</th>
      <td>She</td>
      <td>EastAsia</td>
      <td>4</td>
    </tr>
    <tr>
      <th>58</th>
      <td>Somali</td>
      <td>Africa</td>
      <td>2</td>
    </tr>
    <tr>
      <th>59</th>
      <td>Thai</td>
      <td>EastAsia</td>
      <td>4</td>
    </tr>
    <tr>
      <th>60</th>
      <td>Tu</td>
      <td>EastAsia</td>
      <td>4</td>
    </tr>
    <tr>
      <th>61</th>
      <td>Tujia</td>
      <td>EastAsia</td>
      <td>4</td>
    </tr>
    <tr>
      <th>62</th>
      <td>Uygur</td>
      <td>EastAsia</td>
      <td>4</td>
    </tr>
    <tr>
      <th>63</th>
      <td>Xibo</td>
      <td>EastAsia</td>
      <td>4</td>
    </tr>
    <tr>
      <th>64</th>
      <td>Yi</td>
      <td>EastAsia</td>
      <td>4</td>
    </tr>
    <tr>
      <th>65</th>
      <td>Yoruba</td>
      <td>Africa</td>
      <td>6</td>
    </tr>
    <tr>
      <th>66</th>
      <td>Denisovan</td>
      <td>None</td>
      <td>2</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>You can see that there are multiple African and East asian populations, grouped by
region. Here we collect two lists of IDs for the sample
<a class="reference internal" href="terminology_and_concepts.html#sec-terminology-nodes"><span class="std std-ref">nodes</span></a> from the African region and from the East asian
region:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sample_lists</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">n</span><span class="p">,</span> <span class="n">rgns</span> <span class="ow">in</span> <span class="p">{</span><span class="s2">&quot;Africa&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;AFRICA&#39;</span><span class="p">,</span> <span class="s1">&#39;Africa&#39;</span><span class="p">},</span> <span class="s2">&quot;East asia&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;EAST_ASIA&#39;</span><span class="p">,</span> <span class="s1">&#39;EastAsia&#39;</span><span class="p">}}</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">pop_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">p</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">populations</span><span class="p">()</span> <span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;region&quot;</span><span class="p">)</span> <span class="ow">in</span> <span class="n">rgns</span><span class="p">]</span>
    <span class="n">sample_lists</span><span class="p">[</span><span class="n">n</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">u</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">pop_ids</span> <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">population</span><span class="o">=</span><span class="n">p</span><span class="p">)]</span>
</pre></div>
</div>
</div>
</div>
<p>With these lists we can calculate different windowed statistics
(here <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.diversity" title="(in Project name not set)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">genetic</span> <span class="pre">diversity</span></code></a> and
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.Tajimas_D" title="(in Project name not set)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tajima's</span> <span class="pre">D</span></code></a>) within each of these regions:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">edar_ts</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">trim</span><span class="p">()</span>  <span class="c1"># remove regions with no data (changes the coordinate system)</span>
<span class="n">windows</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">edar_ts</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="mi">10_000</span><span class="p">))</span>
<span class="n">data</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;Genetic diversity&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="n">region</span><span class="p">:</span> <span class="n">edar_ts</span><span class="o">.</span><span class="n">diversity</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span> <span class="n">windows</span><span class="o">=</span><span class="n">windows</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">region</span><span class="p">,</span> <span class="n">samples</span> <span class="ow">in</span> <span class="n">sample_lists</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
    <span class="p">},</span>
    <span class="s2">&quot;Tajima&#39;s D&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="n">region</span><span class="p">:</span> <span class="n">edar_ts</span><span class="o">.</span><span class="n">Tajimas_D</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span> <span class="n">windows</span><span class="o">=</span><span class="n">windows</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">region</span><span class="p">,</span> <span class="n">samples</span> <span class="ow">in</span> <span class="n">sample_lists</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
    <span class="p">},</span>  
<span class="p">}</span>

<span class="c1"># Plot the `data`</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
<span class="n">start</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">edges_left</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>  <span class="c1"># the empty amount at the start of the tree sequence</span>

<span class="k">for</span> <span class="p">(</span><span class="n">title</span><span class="p">,</span> <span class="n">plot_data</span><span class="p">),</span> <span class="n">ax</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">axes</span><span class="p">):</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">axvspan</span><span class="p">(</span><span class="n">edar_gene_bounds</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">edar_gene_bounds</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;lightgray&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">focal_variant</span><span class="o">.</span><span class="n">site</span><span class="o">.</span><span class="n">position</span><span class="p">,</span> <span class="n">ls</span><span class="o">=</span><span class="s2">&quot;:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">label</span><span class="p">,</span> <span class="n">stat</span> <span class="ow">in</span> <span class="n">plot_data</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">stairs</span><span class="p">(</span><span class="n">stat</span><span class="p">,</span> <span class="n">windows</span><span class="o">+</span><span class="n">start</span><span class="p">,</span> <span class="n">baseline</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">label</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">edar_gene_bounds</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;EDAR&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/c9f7b0d33b53ad4a5d31b78c382ca0adda447ce685663b47e8bbca75e6eb5ec6.png" src="_images/c9f7b0d33b53ad4a5d31b78c382ca0adda447ce685663b47e8bbca75e6eb5ec6.png" />
</div>
</div>
<p>Other population genetic libraries such as
<a class="reference external" href="https://scikit-allel.readthedocs.io/en/stable/">scikit-allel</a> (which is
<a class="reference internal" href="getting_started.html#sec-tskit-getting-started-exporting-data-allel"><span class="std std-ref">interoperable</span></a> with <code class="docutils literal notranslate"><span class="pre">tskit</span></code>)
could also have been used to produce the plot above. In this case, the advantage of
using tree sequences is simply that they allow these sorts of analysis to
<a class="reference internal" href="what_is.html#plot-incremental-calculation"><span class="std std-ref">scale</span></a> to datasets of millions of whole genomes.</p>
<section id="topological-analysis">
<span id="sec-popgen-topological"></span><h3>Topological analysis<a class="headerlink" href="#topological-analysis" title="Link to this heading">#</a></h3>
<p>As this inferred tree sequence stores (an estimate of) the underlying
genealogy, we can also derive statistics based on genealogical relationships. You
may have noticed that this tree sequence also contains a sample genome based on an ancient
genome, a <a class="reference external" href="https://en.wikipedia.org/wiki/Denisovan">Denisovan</a> individual. We’ll first
simplify the tree sequence to focus on only the Denisovan plus
a common East Asian and a common African population:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Focus on Han, San, and Denisovan</span>
<span class="n">focal</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;Han&quot;</span><span class="p">:</span> <span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">population</span><span class="o">=</span><span class="mi">6</span><span class="p">),</span>
    <span class="s2">&quot;San&quot;</span><span class="p">:</span> <span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">population</span><span class="o">=</span><span class="mi">17</span><span class="p">),</span>
    <span class="s2">&quot;Denisovan&quot;</span><span class="p">:</span> <span class="n">ts</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">population</span><span class="o">=</span><span class="mi">66</span><span class="p">),</span>
<span class="p">}</span>

<span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">nodes</span> <span class="ow">in</span> <span class="n">focal</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>  <span class="c1"># Sanity check that we got the right IDs</span>
    <span class="k">assert</span> <span class="n">ts</span><span class="o">.</span><span class="n">population</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">node</span><span class="p">(</span><span class="n">nodes</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">.</span><span class="n">population</span><span class="p">)</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">name</span>

<span class="c1"># Simplify to just those samples ...</span>
<span class="n">all_focal_samples</span> <span class="o">=</span> <span class="p">[</span><span class="n">u</span> <span class="k">for</span> <span class="n">samples</span> <span class="ow">in</span> <span class="n">focal</span><span class="o">.</span><span class="n">values</span><span class="p">()</span> <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">samples</span><span class="p">]</span>
<span class="n">simplified_ts</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">simplify</span><span class="p">(</span><span class="n">all_focal_samples</span><span class="p">,</span> <span class="n">filter_sites</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1"># ... and find the tree around the rs3827760 SNP</span>
<span class="n">focal_site</span> <span class="o">=</span> <span class="n">simplified_ts</span><span class="o">.</span><span class="n">site</span><span class="p">(</span><span class="n">focal_variant</span><span class="o">.</span><span class="n">site</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">simplified_ts</span><span class="o">.</span><span class="n">at</span><span class="p">(</span><span class="n">focal_site</span><span class="o">.</span><span class="n">position</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>With this smaller number of samples, we can easily plot the tree
at the “rs3827760” SNP:</p>
<div class="cell tag_hide-input docutils container">
<details class="admonition hide above-input">
<summary aria-label="Toggle hidden content">
<p class="collapsed admonition-title">Show code cell source</p>
<p class="expanded admonition-title">Hide code cell source</p>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Make some nice labels, colours, and legend</span>
<span class="n">mutation_labels</span> <span class="o">=</span> <span class="p">{</span><span class="n">m</span><span class="o">.</span><span class="n">id</span><span class="p">:</span> <span class="n">focal_site</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ID&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">focal_site</span><span class="o">.</span><span class="n">mutations</span><span class="p">}</span>
<span class="n">colours</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">San</span><span class="o">=</span><span class="s2">&quot;yellow&quot;</span><span class="p">,</span> <span class="n">Han</span><span class="o">=</span><span class="s2">&quot;green&quot;</span><span class="p">,</span> <span class="n">Denisovan</span><span class="o">=</span><span class="s2">&quot;magenta&quot;</span><span class="p">)</span>
<span class="n">styles</span> <span class="o">=</span> <span class="p">[</span>
    <span class="sa">f</span><span class="s2">&quot;.leaf.p</span><span class="si">{</span><span class="n">pop</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> &gt; .sym </span><span class="se">{{</span><span class="s2">fill: </span><span class="si">{</span><span class="n">colours</span><span class="p">[</span><span class="n">pop</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]]</span><span class="si">}</span><span class="s2">; stroke: grey</span><span class="se">}}</span><span class="s2">&quot;</span>
    <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">simplified_ts</span><span class="o">.</span><span class="n">populations</span><span class="p">()</span>
<span class="p">]</span>
<span class="n">legend</span> <span class="o">=</span> <span class="s1">&#39;&lt;rect width=&quot;125&quot; height=&quot;75&quot; x=&quot;100&quot; y=&quot;30&quot; fill=&quot;transparent&quot; stroke=&quot;grey&quot; /&gt;&#39;</span>
<span class="n">legend</span> <span class="o">+=</span> <span class="s1">&#39;&lt;text x=&quot;120&quot; y=&quot;45&quot; font-weight=&quot;bold&quot;&gt;Populations&lt;/text&gt;&#39;</span>
<span class="c1"># Create the legend lines, one for each population. Setting classes that match those</span>
<span class="c1"># used for normal nodes means that styled colours are auto automatically picked-up.</span>
<span class="n">legend</span> <span class="o">+=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
    <span class="sa">f</span><span class="s1">&#39;&lt;g transform=&quot;translate(105, </span><span class="si">{</span><span class="mi">60</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">15</span><span class="o">*</span><span class="n">p</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s1">)&quot; class=&quot;leaf p</span><span class="si">{</span><span class="n">p</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s1">&quot;&gt;&#39;</span>  <span class="c1"># an SVG group</span>
    <span class="sa">f</span><span class="s1">&#39;&lt;rect width=&quot;6&quot; height=&quot;6&quot; class=&quot;sym&quot; /&gt;&#39;</span>  <span class="c1"># Square symbol</span>
    <span class="sa">f</span><span class="s1">&#39;&lt;text x=&quot;10&quot; y=&quot;7&quot;&gt;</span><span class="si">{</span><span class="n">p</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span> <span class="c1"># Label</span>
    <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="p">(</span><span class="s2">&quot; (&quot;</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">p</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;region&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot; &quot;</span><span class="p">)</span><span class="o">.</span><span class="n">title</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s2">&quot;)&quot;</span><span class="p">)</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="s2">&quot;region&quot;</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">p</span><span class="o">.</span><span class="n">metadata</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s2">&quot;&quot;</span><span class="si">}</span><span class="s1">&lt;/text&gt;&lt;/g&gt;&#39;</span>  
    <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">simplified_ts</span><span class="o">.</span><span class="n">populations</span><span class="p">()</span>
<span class="p">])</span>

<span class="n">tree</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span>
    <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">400</span><span class="p">),</span>
    <span class="n">style</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">styles</span><span class="p">),</span>
    <span class="n">node_labels</span><span class="o">=</span><span class="p">{},</span>
    <span class="n">mutation_labels</span><span class="o">=</span><span class="n">mutation_labels</span><span class="p">,</span>
    <span class="n">preamble</span><span class="o">=</span><span class="n">legend</span><span class="p">,</span>
    <span class="n">title</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Tree of human chromosome 2 at position </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">focal_variant</span><span class="o">.</span><span class="n">site</span><span class="o">.</span><span class="n">position</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="n">y_axis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">y_ticks</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">50_000</span><span class="p">,</span> <span class="mi">10_000</span><span class="p">),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<img alt="_images/08fc2699cf55263096d48d3b06faa0cc3706568ff495e5ba2b025e79e6e7e664.svg" src="_images/08fc2699cf55263096d48d3b06faa0cc3706568ff495e5ba2b025e79e6e7e664.svg" />
</div>
</div>
<p>You can see that the pair of magenta Denisovan genomes in this region tend to be
more closely associated with the East Asian genomes. We can assess that by counting
all the 3-tip topologies in the tree that contain one genome from each population:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">topology_counter</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">count_topologies</span><span class="p">()</span>
<span class="n">embedded_topologies</span> <span class="o">=</span> <span class="n">topology_counter</span><span class="p">[</span><span class="nb">range</span><span class="p">(</span><span class="n">simplified_ts</span><span class="o">.</span><span class="n">num_populations</span><span class="p">)]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell tag_hide-input docutils container">
<details class="admonition hide above-input">
<summary aria-label="Toggle hidden content">
<p class="collapsed admonition-title">Show code cell source</p>
<p class="expanded admonition-title">Hide code cell source</p>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># All the following code is simply to plot the embedded_topologies nicely</span>
<span class="n">all_trees</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">tskit</span><span class="o">.</span><span class="n">all_trees</span><span class="p">(</span><span class="n">simplified_ts</span><span class="o">.</span><span class="n">num_populations</span><span class="p">))</span>
<span class="n">last</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">all_trees</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
<span class="n">svgs</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="n">style</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">styles</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;.sample text.lab {baseline-shift: super; font-size: 0.7em;}&quot;</span>
<span class="n">style</span> <span class="o">=</span> <span class="n">style</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.leaf.p&quot;</span><span class="p">,</span> <span class="s2">&quot;.leaf.n&quot;</span><span class="p">)</span>  <span class="c1"># Hack to map node IDs to population colours</span>
<span class="n">params</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;size&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">160</span><span class="p">,</span> <span class="mi">150</span><span class="p">),</span>
    <span class="s2">&quot;node_labels&quot;</span><span class="p">:</span> <span class="p">{</span><span class="n">pop</span><span class="o">.</span><span class="n">id</span><span class="p">:</span> <span class="n">pop</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">simplified_ts</span><span class="o">.</span><span class="n">populations</span><span class="p">()}</span>
<span class="p">}</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">t</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">all_trees</span><span class="p">):</span>
    <span class="n">rank</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">rank</span><span class="p">()</span>
    <span class="n">count</span> <span class="o">=</span> <span class="n">embedded_topologies</span><span class="p">[</span><span class="n">rank</span><span class="p">]</span>
    <span class="n">params</span><span class="p">[</span><span class="s2">&quot;title&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">count</span><span class="si">}</span><span class="s2"> trees&quot;</span>
    <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="n">last</span><span class="p">:</span>
        <span class="n">svgs</span> <span class="o">+=</span> <span class="n">t</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">root_svg_attributes</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;x&#39;</span><span class="p">:</span> <span class="p">(</span><span class="n">last</span> <span class="o">-</span> <span class="n">i</span><span class="p">)</span> <span class="o">*</span> <span class="mi">150</span><span class="p">},</span> <span class="o">**</span><span class="n">params</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Plot the last svg and stack the previous ones to the right</span>
        <span class="n">display</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">preamble</span><span class="o">=</span><span class="n">svgs</span><span class="p">,</span> <span class="n">canvas_size</span><span class="o">=</span><span class="p">(</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">150</span><span class="p">),</span> <span class="n">style</span><span class="o">=</span><span class="n">style</span><span class="p">,</span> <span class="o">**</span><span class="n">params</span><span class="p">))</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<img alt="_images/1dcb2a86801e60bcb47940d1074f37cf22fb6ca713dffd45be489708cc877410.svg" src="_images/1dcb2a86801e60bcb47940d1074f37cf22fb6ca713dffd45be489708cc877410.svg" />
</div>
</div>
<p>See <a class="reference internal" href="counting_topologies.html#sec-counting-topologies"><span class="std std-ref">Counting topologies</span></a> for an introduction to topological methods in
<code class="docutils literal notranslate"><span class="pre">tskit</span></code>.</p>
</section>
</section>
<section id="further-information">
<h2>Further information<a class="headerlink" href="#further-information" title="Link to this heading">#</a></h2>
<p>This brief introduction is meant as a simple taster. Many other efficient population
genetic <a class="reference internal" href="analysing_tree_sequences.html#sec-analysing-tree-sequences"><span class="std std-ref">analyses</span></a> are possible when you have
genomic data stored as a tree sequence.</p>
<p>The rest of the <a class="reference internal" href="intro.html#sec-intro"><span class="std std-ref">tutorials</span></a> contain a large number of examples which
are relevant to population genetic analysis and research. You can also visit the
<a class="reference external" href="https://tskit.dev/learn/">learning section</a> of the <a class="reference external" href="https://tskit.dev/">tskit website</a>.</p>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="tutorial_development.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title"><em>Developing new tutorials</em></p>
      </div>
    </a>
    <a class="right-next"
       href="phylogen.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><code class="docutils literal notranslate"><span class="pre">Tskit</span></code> for phylogenetics</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#population-genetic-simulation">Population genetic simulation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#population-genetic-inference">Population genetic inference</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#topological-analysis">Topological analysis</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#further-information">Further information</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Tskit Developers
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>