
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Data model &#8212; Tskit manual</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=5349f25f" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=9c3e77be" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="_static/bespoke.css?v=00f581dd" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=1ae7504c"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'data-model';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Metadata" href="metadata.html" />
    <link rel="prev" title="Glossary" href="glossary.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <script data-goatcounter="https://tskit.goatcounter.com/count" async src="//gc.zgo.at/count.js"></script>
</head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="introduction.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo.svg" class="logo__image only-light" alt=""/>
    <script>document.write(`<img src="_static/logo.svg" class="logo__image only-dark" alt=""/>`);</script>
  
  
    <p class="title logo__title">Version 1.0.1</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Concepts</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Data model</a></li>
<li class="toctree-l1"><a class="reference internal" href="metadata.html">Metadata</a></li>
<li class="toctree-l1"><a class="reference internal" href="provenance.html">Provenance</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="stats.html">Statistics</a></li>
<li class="toctree-l1"><a class="reference internal" href="topological-analysis.html">Topological analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="ibd.html">Identity by descent</a></li>
<li class="toctree-l1"><a class="reference internal" href="export.html">Data export</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Interfaces</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="python-api.html">Python API</a></li>
<li class="toctree-l1"><a class="reference internal" href="numba.html">Numba Integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="c-api.html">C API</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">Command line interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="file-formats.html">File formats</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">For developers</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="development.html">Development</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelogs.html">Changelogs</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Miscellaneous</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="citation.html">Citing tskit</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/tskit-dev/tskit" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/tskit-dev/tskit/edit/main/docs/data-model.md" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/tskit-dev/tskit/issues/new?title=Issue%20on%20page%20%2Fdata-model.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/data-model.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Data model</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#table-definitions">Table definitions</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#table-types">Table types</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#node-table">Node Table</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#individual-table">Individual Table</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#edge-table">Edge Table</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#site-table">Site Table</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mutation-table">Mutation Table</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#migration-table">Migration Table</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#population-table">Population Table</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#provenance-table">Provenance Table</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#metadata">Metadata</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#valid-tree-sequence-requirements">Valid tree sequence requirements</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#individual-requirements">Individual requirements</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#node-requirements">Node requirements</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#edge-requirements">Edge requirements</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#site-requirements">Site requirements</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mutation-requirements">Mutation requirements</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#migration-requirements">Migration requirements</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#population-requirements">Population requirements</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#provenance-requirements">Provenance requirements</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#table-indexes">Table indexes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#saving-to-file">Saving to file</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-encoding">Data encoding</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#encoding-ragged-columns">Encoding ragged columns</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#reference-sequence">Reference sequence</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tree-structure">Tree structure</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#quintuply-linked-trees">Quintuply linked trees</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#convenience-arrays">Convenience arrays</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#roots">Roots</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-root-threshold">The root threshold</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-virtual-root">The virtual root</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#isolated-nodes">Isolated nodes</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#isolated-sample-nodes">Isolated sample nodes</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#dead-leaves-and-branches">Dead leaves and branches</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#encoding-genetic-variation">Encoding genetic variation</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#missing-data">Missing data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#possibly-surprising-consequences-of-the-data-model">Possibly surprising consequences of the data model</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#unrelated-material">Unrelated material</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="data-model">
<span id="sec-data-model"></span><h1>Data model<a class="headerlink" href="#data-model" title="Link to this heading">#</a></h1>
<p>The <code class="docutils literal notranslate"><span class="pre">tskit</span></code> library deals with sets of sampled genome sequences through storage
and analysis of their shared genetic ancestry. This genealogical ancestry (sometimes
known as an Ancestral Recombination Graph) is stored concisely in <code class="docutils literal notranslate"><span class="pre">tskit</span></code> in the
“succinct tree sequence” format, which comprises a collection of easy-to-understand
tables. This page documents the structure of the tables and encoding of table data,
as well as the encoding of the correlated genetic trees that can be extracted from
a <code class="docutils literal notranslate"><span class="pre">tskit</span></code> tree sequence.</p>
<p>We begin by defining the the structure of the tables in the <a class="reference internal" href="#sec-table-definitions"><span class="std std-ref">Table definitions</span></a>
section. The <a class="reference internal" href="#sec-data-model-data-encoding"><span class="std std-ref">Data encoding</span></a> section then describe how data is
stored in those tables (also see the <a class="reference internal" href="file-formats.html#sec-file-formats"><span class="std std-ref">File formats</span></a>
chapter). The <a class="reference internal" href="#sec-data-model-tree-structure"><span class="std std-ref">Tree structure</span></a> section then
describes the encoding of the trees that are generated from the <a class="reference internal" href="python-api.html#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a>
and <a class="reference internal" href="python-api.html#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a>. Finally, we describe how genotype data arises from tree
structure, especially how we can incorporate the idea of missing data.</p>
<section id="table-definitions">
<span id="sec-table-definitions"></span><h2>Table definitions<a class="headerlink" href="#table-definitions" title="Link to this heading">#</a></h2>
<section id="table-types">
<span id="sec-table-types-definitions"></span><h3>Table types<a class="headerlink" href="#table-types" title="Link to this heading">#</a></h3>
<p>A tree sequence can be stored in a collection of eight tables:
<a class="reference internal" href="#sec-node-table-definition"><span class="std std-ref">Node</span></a>,
<a class="reference internal" href="#sec-edge-table-definition"><span class="std std-ref">Edge</span></a>,
<a class="reference internal" href="#sec-individual-table-definition"><span class="std std-ref">Individual</span></a>,
<a class="reference internal" href="#sec-site-table-definition"><span class="std std-ref">Site</span></a>,
<a class="reference internal" href="#sec-mutation-table-definition"><span class="std std-ref">Mutation</span></a>,
<a class="reference internal" href="#sec-migration-table-definition"><span class="std std-ref">Migration</span></a>,
<a class="reference internal" href="#sec-population-table-definition"><span class="std std-ref">Population</span></a>, and
<a class="reference internal" href="#sec-provenance-table-definition"><span class="std std-ref">Provenance</span></a>.
The Node and Edge tables store the genealogical
relationships that define the trees, and the Individual table
describes how multiple genomes are grouped within individuals;
the Site and Mutation tables describe where mutations fall
on the trees; the Migration table describes how lineages move across space;
and the Provenance table contains information on where the data came from.
Only Node and Edge tables are necessary to encode the genealogical trees;
Sites and Mutations are optional but necessary to encode polymorphism
(sequence) data; the remainder are optional.
In the following sections we define these components of a tree sequence in
more detail.</p>
<section id="node-table">
<span id="sec-node-table-definition"></span><h4>Node Table<a class="headerlink" href="#node-table" title="Link to this heading">#</a></h4>
<p>A <strong>node</strong> defines a monoploid set of chromosomes (a “genome”) of a specific
individual that was born at some time in the past: the set of
chromosomes inherited from a particular one of the individual’s parents.
(See <a class="reference internal" href="glossary.html#sec-nodes-or-individuals"><span class="std std-ref">Nodes, Genomes, or Individuals?</span></a> for more discussion.)
Every vertex in the marginal trees of a tree sequence corresponds
to exactly one node, and a node may be present in many trees. The
node table contains five columns, of which <code class="docutils literal notranslate"><span class="pre">flags</span></code> and <code class="docutils literal notranslate"><span class="pre">time</span></code> are
mandatory:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head text-left"><p>Column</p></th>
<th class="head"><p>Type</p></th>
<th class="head text-right"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-left"><p>flags</p></td>
<td><p>uint32</p></td>
<td class="text-right"><p>Bitwise flags.</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>time</p></td>
<td><p>double</p></td>
<td class="text-right"><p>Birth time of node.</p></td>
</tr>
<tr class="row-even"><td class="text-left"><p>population</p></td>
<td><p>int32</p></td>
<td class="text-right"><p>Birth population of node.</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>individual</p></td>
<td><p>int32</p></td>
<td class="text-right"><p>The individual the node belongs to.</p></td>
</tr>
<tr class="row-even"><td class="text-left"><p>metadata</p></td>
<td><p>binary</p></td>
<td class="text-right"><p>Node <a class="reference internal" href="#sec-metadata-definition"><span class="std std-ref">Metadata</span></a>.</p></td>
</tr>
</tbody>
</table>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">time</span></code> column records the birth time of the individual in question,
and is a floating point value. Similarly,
the <code class="docutils literal notranslate"><span class="pre">population</span></code> column records the ID of the population where this
individual was born. If not provided, <code class="docutils literal notranslate"><span class="pre">population</span></code> defaults to the
null ID (-1). Otherwise, the population ID must refer to a row in the
<a class="reference internal" href="#sec-population-table-definition"><span class="std std-ref">Population Table</span></a>.
The <code class="docutils literal notranslate"><span class="pre">individual</span></code> column records the ID of the
<a class="reference internal" href="#sec-individual-table-definition"><span class="std std-ref">Individual</span></a>
individual that this node belongs to. If specified, the ID must refer
to a valid individual. If not provided, <code class="docutils literal notranslate"><span class="pre">individual</span></code>
defaults to the null ID (-1).</p>
<p>The <code class="docutils literal notranslate"><span class="pre">flags</span></code> column stores information about a particular node, and
is composed of 32 bitwise boolean values. Currently, the only flag defined
is <code class="docutils literal notranslate"><span class="pre">NODE_IS_SAMPLE</span> <span class="pre">=</span> <span class="pre">1</span></code>, which defines the <em>sample</em> status of nodes. Marking
a particular node as a “sample” means, for example, that the mutational state
of the node will be included in the genotypes produced by
<a class="reference internal" href="python-api.html#tskit.TreeSequence.variants" title="tskit.TreeSequence.variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.variants()</span></code></a>.</p>
<p>Bits 0-15 (inclusive) of the <code class="docutils literal notranslate"><span class="pre">flags</span></code> column are reserved for internal use by
<code class="docutils literal notranslate"><span class="pre">tskit</span></code> and should not be used by applications for anything other
than the purposes documented here. Bits 16-31 (inclusive) are free for applications
to use for any purpose and will not be altered or interpreteted by
<code class="docutils literal notranslate"><span class="pre">tskit</span></code>.</p>
<p>See the <a class="reference internal" href="#sec-node-requirements"><span class="std std-ref">Node requirements</span></a> section for details on the properties
required for a valid set of nodes.</p>
<p>For convenience, the <a class="reference internal" href="file-formats.html#sec-text-file-format"><span class="std std-ref">text format</span></a> for nodes
decomposes the <code class="docutils literal notranslate"><span class="pre">flags</span></code> value into its separate values. Thus, in the
text format we have a column for <code class="docutils literal notranslate"><span class="pre">is_sample</span></code>, which corresponds to the
<code class="docutils literal notranslate"><span class="pre">flags</span></code> column in the underlying table. As more flags values are
defined, these will be added to the text file format.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> column provides a location for client code to store
information about each node. See the <a class="reference internal" href="#sec-metadata-definition"><span class="std std-ref">Metadata</span></a> section for
more details on how metadata columns should be used.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The distinction between <code class="docutils literal notranslate"><span class="pre">flags</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata</span></code> is that flags
holds information about a node that the library understands, whereas
metadata holds information about a node that the library <em>does not</em>
understand. Metadata is for storing auxiliarly information that is
not necessary for the core tree sequence algorithms.</p>
</div>
</section>
<section id="individual-table">
<span id="sec-individual-table-definition"></span><h4>Individual Table<a class="headerlink" href="#individual-table" title="Link to this heading">#</a></h4>
<p>An <strong>individual</strong> defines how nodes (which can be seen
as representing single chromosomes) group together in a polyploid individual.
The individual table contains three columns, of which only <code class="docutils literal notranslate"><span class="pre">flags</span></code> is mandatory.</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head text-left"><p>Column</p></th>
<th class="head"><p>Type</p></th>
<th class="head text-right"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-left"><p>flags</p></td>
<td><p>uint32</p></td>
<td class="text-right"><p>Bitwise flags.</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>location</p></td>
<td><p>double</p></td>
<td class="text-right"><p>Location in arbitrary dimensions.</p></td>
</tr>
<tr class="row-even"><td class="text-left"><p>parents</p></td>
<td><p>int32</p></td>
<td class="text-right"><p>Ids of parent individuals.</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>metadata</p></td>
<td><p>binary</p></td>
<td class="text-right"><p>Individual <a class="reference internal" href="#sec-metadata-definition"><span class="std std-ref">Metadata</span></a>.</p></td>
</tr>
</tbody>
</table>
</div>
<p>See the <a class="reference internal" href="#sec-individual-requirements"><span class="std std-ref">Individual requirements</span></a> section for details on the properties
required for a valid set of individuals.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">flags</span></code> column stores information about a particular individual, and
is composed of 32 bitwise boolean values. Currently, no flags are
defined.</p>
<p>Bits 0-15 (inclusive) of the <code class="docutils literal notranslate"><span class="pre">flags</span></code> column are reserved for internal use by
<code class="docutils literal notranslate"><span class="pre">tskit</span></code> and should not be used by applications for anything other
than the purposes documented here. Bits 16-31 (inclusive) are free for applications
to use for any purpose and will not be altered or interpreteted by
<code class="docutils literal notranslate"><span class="pre">tskit</span></code>.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">location</span></code> column stores the location of an individual in arbitrary
dimensions. This column is <a class="reference internal" href="#sec-encoding-ragged-columns"><span class="std std-ref">ragged</span></a>, and
so different individuals can have locations with different dimensions (i.e.,
one individual may have location <code class="docutils literal notranslate"><span class="pre">[]</span></code> and another <code class="docutils literal notranslate"><span class="pre">[0,</span> <span class="pre">1,</span> <span class="pre">0]</span></code>. This could
therefore be used to store other quantities (e.g., phenotype).</p>
<p>The <code class="docutils literal notranslate"><span class="pre">parents</span></code> column stores the ids of other individuals that are the parents of
an individual. This can be used to store pedigree information for individuals.
This column is <a class="reference internal" href="#sec-encoding-ragged-columns"><span class="std std-ref">ragged</span></a> such that an
individual can have any number of parents.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> column provides a location for client code to store
information about each individual. See the <a class="reference internal" href="#sec-metadata-definition"><span class="std std-ref">Metadata</span></a> section for
more details on how metadata columns should be used.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The distinction between <code class="docutils literal notranslate"><span class="pre">flags</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata</span></code> is that flags
holds information about a individual that the library understands, whereas
metadata holds information about a individual that the library <em>does not</em>
understand. Metadata is for storing auxiliarly information that is
not necessary for the core tree sequence algorithms.</p>
</div>
</section>
<section id="edge-table">
<span id="sec-edge-table-definition"></span><h4>Edge Table<a class="headerlink" href="#edge-table" title="Link to this heading">#</a></h4>
<p>An <strong>edge</strong> defines a parent-child relationship between a pair of nodes
over a specific sequence interval. The edge table contains five columns,
all of which are mandatory except <code class="docutils literal notranslate"><span class="pre">metadata</span></code>:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head text-left"><p>Column</p></th>
<th class="head"><p>Type</p></th>
<th class="head text-right"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-left"><p>left</p></td>
<td><p>double</p></td>
<td class="text-right"><p>Left coordinate of the edge (inclusive).</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>right</p></td>
<td><p>double</p></td>
<td class="text-right"><p>Right coordinate of the edge (exclusive).</p></td>
</tr>
<tr class="row-even"><td class="text-left"><p>parent</p></td>
<td><p>int32</p></td>
<td class="text-right"><p>Parent node ID.</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>child</p></td>
<td><p>int32</p></td>
<td class="text-right"><p>Child node ID.</p></td>
</tr>
<tr class="row-even"><td class="text-left"><p>metadata</p></td>
<td><p>binary</p></td>
<td class="text-right"><p>Node <a class="reference internal" href="#sec-metadata-definition"><span class="std std-ref">Metadata</span></a>.</p></td>
</tr>
</tbody>
</table>
</div>
<p>Each row in an edge table describes a half-open genomic interval <code class="docutils literal notranslate"><span class="pre">[left,</span> <span class="pre">right)</span></code>
over which the <code class="docutils literal notranslate"><span class="pre">child</span></code> inherited from the given <code class="docutils literal notranslate"><span class="pre">parent</span></code>.
The <code class="docutils literal notranslate"><span class="pre">left</span></code> and <code class="docutils literal notranslate"><span class="pre">right</span></code> columns are defined using double precision
floating point values. The <code class="docutils literal notranslate"><span class="pre">parent</span></code> and <code class="docutils literal notranslate"><span class="pre">child</span></code>
columns specify integer IDs in the associated <a class="reference internal" href="#sec-node-table-definition"><span class="std std-ref">Node Table</span></a>.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> column provides a location for client code to store
information about each edge. See the <a class="reference internal" href="#sec-metadata-definition"><span class="std std-ref">Metadata</span></a> section for
more details on how metadata columns should be used.</p>
<p>See the <a class="reference internal" href="#sec-edge-requirements"><span class="std std-ref">Edge requirements</span></a> section for details on the properties
required for a valid set of edges.</p>
</section>
<section id="site-table">
<span id="sec-site-table-definition"></span><h4>Site Table<a class="headerlink" href="#site-table" title="Link to this heading">#</a></h4>
<p>A <strong>site</strong> defines a particular location along the genome in which
we are interested in observing the allelic state. The site table
contains three columns, of which <code class="docutils literal notranslate"><span class="pre">position</span></code> and <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code>
are mandatory.</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head text-left"><p>Column</p></th>
<th class="head"><p>Type</p></th>
<th class="head text-right"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-left"><p>position</p></td>
<td><p>double</p></td>
<td class="text-right"><p>Position of site in genome coordinates.</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>ancestral_state</p></td>
<td><p>text</p></td>
<td class="text-right"><p>The state at the root of the tree.</p></td>
</tr>
<tr class="row-even"><td class="text-left"><p>metadata</p></td>
<td><p>binary</p></td>
<td class="text-right"><p>Site <a class="reference internal" href="#sec-metadata-definition"><span class="std std-ref">Metadata</span></a>.</p></td>
</tr>
</tbody>
</table>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">position</span></code> column is a floating point value defining the location
of the site in question along the genome.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> column specifies the allelic state at the root
of the tree, thus defining the state that nodes inherit if no mutations
intervene. The column stores text character data of arbitrary length.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> column provides a location for client code to store
information about each site. See the <a class="reference internal" href="#sec-metadata-definition"><span class="std std-ref">Metadata</span></a> section for
more details on how metadata columns should be used.</p>
<p>See the <a class="reference internal" href="#sec-site-requirements"><span class="std std-ref">Site requirements</span></a> section for details on the properties
required for a valid set of sites.</p>
</section>
<section id="mutation-table">
<span id="sec-mutation-table-definition"></span><h4>Mutation Table<a class="headerlink" href="#mutation-table" title="Link to this heading">#</a></h4>
<p>A <strong>mutation</strong> defines a change of allelic state on a tree at a particular site.
The mutation table contains five columns, of which <code class="docutils literal notranslate"><span class="pre">site</span></code>, <code class="docutils literal notranslate"><span class="pre">node</span></code> and
<code class="docutils literal notranslate"><span class="pre">derived_state</span></code> are mandatory.</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head text-left"><p>Column</p></th>
<th class="head"><p>Type</p></th>
<th class="head text-right"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-left"><p>site</p></td>
<td><p>int32</p></td>
<td class="text-right"><p>The ID of the site the mutation occurs at.</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>node</p></td>
<td><p>int32</p></td>
<td class="text-right"><p>The node this mutation occurs at.</p></td>
</tr>
<tr class="row-even"><td class="text-left"><p>parent</p></td>
<td><p>int32</p></td>
<td class="text-right"><p>The ID of the parent mutation.</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>time</p></td>
<td><p>double</p></td>
<td class="text-right"><p>Time at which the mutation occurred.</p></td>
</tr>
<tr class="row-even"><td class="text-left"><p>derived_state</p></td>
<td><p>char</p></td>
<td class="text-right"><p>The allelic state resulting from the mutation.</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>metadata</p></td>
<td><p>binary</p></td>
<td class="text-right"><p>Mutation <a class="reference internal" href="#sec-metadata-definition"><span class="std std-ref">Metadata</span></a>.</p></td>
</tr>
</tbody>
</table>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">site</span></code> column is an integer value defining the ID of the
<a class="reference internal" href="#sec-site-table-definition"><span class="std std-ref">site</span></a> at which this mutation occurred.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">node</span></code> column is an integer value defining the ID of the
first <a class="reference internal" href="#sec-node-table-definition"><span class="std std-ref">node</span></a> in the tree below this mutation.</p>
<p>The  <code class="docutils literal notranslate"><span class="pre">time</span></code> column is a double precision floating point value recording how long ago
the mutation happened.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">derived_state</span></code> column specifies the allelic state resulting from the mutation,
thus defining the state that the <code class="docutils literal notranslate"><span class="pre">node</span></code> and any descendant nodes in the
subtree inherit unless further mutations occur. The column stores text
character data of arbitrary length.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">parent</span></code> column is an integer value defining the ID of the mutation whose
allelic state this mutation replaced. If there is no mutation at the
site in question on the path back to root, then this field is set to the
null ID (-1). (The <code class="docutils literal notranslate"><span class="pre">parent</span></code> column is only required in situations
where there are multiple mutations at a given site. For
“infinite sites” mutations, it can be ignored.)</p>
<p>The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> column provides a location for client code to store
information about each site. See the <a class="reference internal" href="#sec-metadata-definition"><span class="std std-ref">Metadata</span></a> section for
more details on how metadata columns should be used.</p>
<p>See the <a class="reference internal" href="#sec-mutation-requirements"><span class="std std-ref">Mutation requirements</span></a> section for details on the properties
required for a valid set of mutations.</p>
</section>
<section id="migration-table">
<span id="sec-migration-table-definition"></span><h4>Migration Table<a class="headerlink" href="#migration-table" title="Link to this heading">#</a></h4>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Encoding migration in the migrations table is a legacy approach
associated with older versions of <code class="docutils literal notranslate"><span class="pre">msprime</span></code>; recording movement between
populations in the migration table is entirely optional, even when related
nodes are assigned to different populations.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The migration table may be entirely removed from the <code class="docutils literal notranslate"><span class="pre">tskit</span></code> data model
in the future. Meanwhile, a number of <code class="docutils literal notranslate"><span class="pre">tskit</span></code> functions, such as
<a class="reference internal" href="python-api.html#tskit.TreeSequence.simplify" title="tskit.TreeSequence.simplify"><code class="xref py py-meth docutils literal notranslate"><span class="pre">simplify()</span></code></a> will raise an error if data exists in
the migrations table.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>The <a class="reference external" href="https://tskit.dev/msprime/docs/stable/ancestry.html#sec-ancestry-record-migrations" title="(in Project name not set)"><span>Migration events</span></a>
sections and the associated discussion of
<a class="reference external" href="https://tskit.dev/msprime/docs/stable/demography.html#sec-demography-migration" title="(in Project name not set)"><span>Migration</span></a> in the <code class="docutils literal notranslate"><span class="pre">msprime</span></code> documentation.</p>
</div>
<p>In simulations, trees can be thought of as spread across space, and it is
helpful for inferring demographic history to record this history.
Migrations are performed by individual ancestors, but might not be tagged by an
individual whose genome is tracked as a <code class="docutils literal notranslate"><span class="pre">node</span></code> (as in a discrete-deme model they are
unlikely to be both a migrant and a most recent common ancestor).  So,
<code class="docutils literal notranslate"><span class="pre">tskit</span></code> can record separately when a segment of ancestry has moved between
populations. This table is not required, even if different nodes come from
different populations.</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head text-left"><p>Column</p></th>
<th class="head"><p>Type</p></th>
<th class="head text-right"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-left"><p>left</p></td>
<td><p>double</p></td>
<td class="text-right"><p>Left coordinate of the migrating segment (inclusive).</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>right</p></td>
<td><p>double</p></td>
<td class="text-right"><p>Right coordinate of the migrating segment (exclusive).</p></td>
</tr>
<tr class="row-even"><td class="text-left"><p>node</p></td>
<td><p>int32</p></td>
<td class="text-right"><p>Node ID.</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>source</p></td>
<td><p>int32</p></td>
<td class="text-right"><p>Source population ID.</p></td>
</tr>
<tr class="row-even"><td class="text-left"><p>dest</p></td>
<td><p>int32</p></td>
<td class="text-right"><p>Destination population ID.</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>time</p></td>
<td><p>double</p></td>
<td class="text-right"><p>Time of migration event.</p></td>
</tr>
<tr class="row-even"><td class="text-left"><p>metadata</p></td>
<td><p>binary</p></td>
<td class="text-right"><p>Migration <a class="reference internal" href="#sec-metadata-definition"><span class="std std-ref">Metadata</span></a>.</p></td>
</tr>
</tbody>
</table>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">left</span></code> and <code class="docutils literal notranslate"><span class="pre">right</span></code> columns are floating point values defining the
half-open segment of genome affected (these need not exactly correspond to
breakpoints between edges). The <code class="docutils literal notranslate"><span class="pre">source</span></code> and <code class="docutils literal notranslate"><span class="pre">dest</span></code> columns record the IDs of
the respective populations (note that by <code class="docutils literal notranslate"><span class="pre">msprime</span></code> convention, “source” and
“destination” are defined in reverse time, see
<a class="reference external" href="https://tskit.dev/msprime/docs/stable/demography.html#sec-demography-direction-of-time" title="(in Project name not set)"><span>Direction of time</span></a>.). The <code class="docutils literal notranslate"><span class="pre">time</span></code> column
holds floating point values recording the time of the event, with migrations
assumed to occur instantaneously. The <code class="docutils literal notranslate"><span class="pre">node</span></code> column records the ID of the child
node of the migrating segment; in consequence the population ID of the <code class="docutils literal notranslate"><span class="pre">node</span></code> will
match the <code class="docutils literal notranslate"><span class="pre">src</span></code> ID (unless sequential migrations affect the same <code class="docutils literal notranslate"><span class="pre">node</span></code>, in which
case it will match the <code class="docutils literal notranslate"><span class="pre">src</span></code> value of the youngest of those migrations).</p>
<p>The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> column provides a location for client code to store
information about each migration. See the <a class="reference internal" href="#sec-metadata-definition"><span class="std std-ref">Metadata</span></a> section for
more details on how metadata columns should be used.</p>
<p>See the <a class="reference internal" href="#sec-migration-requirements"><span class="std std-ref">Migration requirements</span></a> section for details on the properties
required for a valid set of migrations.</p>
</section>
<section id="population-table">
<span id="sec-population-table-definition"></span><h4>Population Table<a class="headerlink" href="#population-table" title="Link to this heading">#</a></h4>
<p>A <strong>population</strong> defines a grouping of individuals that a node can
be said to belong to.</p>
<p>The population table contains one column, <code class="docutils literal notranslate"><span class="pre">metadata</span></code>.</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head text-left"><p>Column</p></th>
<th class="head"><p>Type</p></th>
<th class="head text-right"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-left"><p>metadata</p></td>
<td><p>binary</p></td>
<td class="text-right"><p>Population <a class="reference internal" href="#sec-metadata-definition"><span class="std std-ref">Metadata</span></a>.</p></td>
</tr>
</tbody>
</table>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> column provides a location for client code to store
information about each population. See the <a class="reference internal" href="#sec-metadata-definition"><span class="std std-ref">Metadata</span></a> section for
more details on how metadata columns should be used.</p>
<p>See the <a class="reference internal" href="#sec-population-requirements"><span class="std std-ref">Population requirements</span></a> section for details on the properties
required for a valid set of populations.</p>
</section>
<section id="provenance-table">
<span id="sec-provenance-table-definition"></span><h4>Provenance Table<a class="headerlink" href="#provenance-table" title="Link to this heading">#</a></h4>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head text-left"><p>Column</p></th>
<th class="head"><p>Type</p></th>
<th class="head text-right"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-left"><p>timestamp</p></td>
<td><p>char</p></td>
<td class="text-right"><p>Timestamp in <a class="reference external" href="https://en.wikipedia.org/wiki/ISO_8601">ISO-8601</a> format.</p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p>record</p></td>
<td><p>char</p></td>
<td class="text-right"><p>Provenance record as JSON.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>
<section id="metadata">
<span id="sec-metadata-definition"></span><h3>Metadata<a class="headerlink" href="#metadata" title="Link to this heading">#</a></h3>
<p>Each table (excluding provenance) has a metadata column for storing and passing along
information that tskit does not use or interpret. See <a class="reference internal" href="metadata.html#sec-metadata"><span class="std std-ref">Metadata</span></a> for details.
The metadata columns are <a class="reference internal" href="python-api.html#sec-tables-api-binary-columns"><span class="std std-ref">binary columns</span></a>.</p>
<p>When using the <a class="reference internal" href="file-formats.html#sec-text-file-format"><span class="std std-ref">Text file formats</span></a>, metadata values are written as opaque
text. By default, :meth:<code class="docutils literal notranslate"><span class="pre">TreeSequence.dump_text</span></code> will base64-encode metadata values
that are stored as raw bytes (when <code class="docutils literal notranslate"><span class="pre">base64_metadata=True</span></code>) so that binary data can
be safely printed and exchanged; in this case :func:<code class="docutils literal notranslate"><span class="pre">tskit.load_text</span></code> will base64-decode
the corresponding text fields back to bytes. When metadata has already been decoded
to a structured Python object (for example via a metadata schema), the textual
representation written by :meth:<code class="docutils literal notranslate"><span class="pre">TreeSequence.dump_text</span></code> is the <code class="docutils literal notranslate"><span class="pre">repr</span></code> of that
object, and :func:<code class="docutils literal notranslate"><span class="pre">tskit.load_text</span></code> does not attempt to reconstruct the original
structured value from this representation. For reliable metadata round-tripping,
prefer the native binary tree sequence file format over the text formats.</p>
<p>The tree sequence itself also has metadata stored as a byte array.</p>
</section>
<section id="valid-tree-sequence-requirements">
<span id="sec-valid-tree-sequence-requirements"></span><h3>Valid tree sequence requirements<a class="headerlink" href="#valid-tree-sequence-requirements" title="Link to this heading">#</a></h3>
<p>Arbitrary data can be stored in tables using the classes in the
<a class="reference internal" href="python-api.html#sec-tables-api"><span class="std std-ref">Tables and Table Collections</span></a>. The <a class="reference internal" href="python-api.html#tskit.TableCollection.tree_sequence" title="tskit.TableCollection.tree_sequence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.tree_sequence()</span></code></a> method
can be used to turn such a <a class="reference internal" href="python-api.html#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a> into an immutable
<a class="reference internal" href="python-api.html#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> object, but this requires the tables to
fulfil a specific set of requirements. In this
section we list these requirements, and explain their rationale.
Violations of most of these requirements are detected when the
user attempts to load a tree sequence via <a class="reference internal" href="python-api.html#tskit.load" title="tskit.load"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load()</span></code></a> or
<a class="reference internal" href="python-api.html#tskit.TableCollection.tree_sequence" title="tskit.TableCollection.tree_sequence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.tree_sequence()</span></code></a>, raising an informative
error message. Some more complex requirements may not be detectable at load-time,
and errors may not occur until certain operations are attempted.
These are documented below.</p>
<p>At the tree-sequence level, we require that the coordinate space has a finite,
strictly positive length; that is, the <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> attribute must be a
finite value greater than zero.</p>
<p>The Python API also provides tools that can transform a collection of
tables into a valid collection of tables, so long as they are logically
consistent, see <a class="reference internal" href="python-api.html#sec-tables-api-creating-valid-tree-sequence"><span class="std std-ref">Creating a valid tree sequence</span></a>.</p>
<section id="individual-requirements">
<span id="sec-individual-requirements"></span><h4>Individual requirements<a class="headerlink" href="#individual-requirements" title="Link to this heading">#</a></h4>
<p>Individuals are a basic type in a tree sequence and are not defined with
respect to any other tables. Individuals can have a reference to their parent
individuals, if present these references must be valid or null (-1). An
individual cannot list itself as its own parent.</p>
<p>A valid tree sequence does not require individuals to be sorted in any
particular order, and sorting a set of tables using <a class="reference internal" href="python-api.html#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.sort()</span></code></a>
has no effect on individuals. However, individuals can be optionally sorted
using <a class="reference internal" href="python-api.html#tskit.TableCollection.sort_individuals" title="tskit.TableCollection.sort_individuals"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.sort_individuals()</span></code></a>.</p>
</section>
<section id="node-requirements">
<span id="sec-node-requirements"></span><h4>Node requirements<a class="headerlink" href="#node-requirements" title="Link to this heading">#</a></h4>
<p>Given a valid set of individuals and populations, the requirements for
each node are:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">time</span></code> must be a finite (non-NaN, non-infinite) value;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">population</span></code> must either be null (-1) or refer to a valid population ID;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">individual</span></code> must either be null (-1) or refer to a valid individual ID.</p></li>
</ul>
<p>An ID refers to a zero-indexed row number in the relevant table,
and so is “valid” if is between 0 and one less than the number of rows in the relevant table.</p>
<p>There are no requirements regarding the ordering of nodes with respect to time.</p>
<p>Sorting a set of tables using <a class="reference internal" href="python-api.html#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.sort()</span></code></a>
has no effect on nodes.</p>
</section>
<section id="edge-requirements">
<span id="sec-edge-requirements"></span><h4>Edge requirements<a class="headerlink" href="#edge-requirements" title="Link to this heading">#</a></h4>
<p>Given a valid set of nodes and a sequence length <span class="math notranslate nohighlight">\(L\)</span>, the simple
requirements for each edge are:</p>
<ul class="simple">
<li><p>We must have finite coordinates with <span class="math notranslate nohighlight">\(0 \leq\)</span> <code class="docutils literal notranslate"><span class="pre">left</span></code> <span class="math notranslate nohighlight">\(&lt;\)</span> <code class="docutils literal notranslate"><span class="pre">right</span></code> <span class="math notranslate nohighlight">\(\leq L\)</span>;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">parent</span></code> and <code class="docutils literal notranslate"><span class="pre">child</span></code> must be valid node IDs;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">time[parent]</span></code> &gt; <code class="docutils literal notranslate"><span class="pre">time[child]</span></code>;</p></li>
<li><p>edges must be unique (i.e., no duplicate edges are allowed).</p></li>
</ul>
<p>The first requirement simply ensures that the interval makes sense. The
third requirement ensures that we cannot have loops, since time is
always increasing as we ascend the tree.</p>
<p>To ensure a valid tree sequence there is one further requirement:</p>
<ul class="simple">
<li><p>The set of intervals on which each node is a child must be disjoint.</p></li>
</ul>
<p>This guarantees that we cannot have contradictory edges (i.e.,
where a node <code class="docutils literal notranslate"><span class="pre">a</span></code> is a child of both <code class="docutils literal notranslate"><span class="pre">b</span></code> and <code class="docutils literal notranslate"><span class="pre">c</span></code>), and ensures that
at each point on the sequence we have a well-formed forest of trees.</p>
<p>In the interest of algorithmic efficiency, edges must have the following
sortedness properties:</p>
<ul class="simple">
<li><p>All edges for a given parent must be contiguous;</p></li>
<li><p>Edges must be listed in nondecreasing order of <code class="docutils literal notranslate"><span class="pre">parent</span></code> time;</p></li>
<li><p>Within the edges for a given <code class="docutils literal notranslate"><span class="pre">parent</span></code>, edges must be sorted
first by <code class="docutils literal notranslate"><span class="pre">child</span></code> ID and then by <code class="docutils literal notranslate"><span class="pre">left</span></code> coordinate.</p></li>
</ul>
<p>Violations of these requirements are detected at load time.
The <a class="reference internal" href="python-api.html#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.sort()</span></code></a> method will ensure that these sortedness
properties are fulfilled.</p>
</section>
<section id="site-requirements">
<span id="sec-site-requirements"></span><h4>Site requirements<a class="headerlink" href="#site-requirements" title="Link to this heading">#</a></h4>
<p>Given a valid set of nodes and a sequence length <span class="math notranslate nohighlight">\(L\)</span>, the simple
requirements for a valid set of sites are:</p>
<ul class="simple">
<li><p>We must have a finite coordinate with <span class="math notranslate nohighlight">\(0 \leq\)</span> <code class="docutils literal notranslate"><span class="pre">position</span></code> <span class="math notranslate nohighlight">\(&lt; L\)</span>;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">position</span></code> values must be unique.</p></li>
</ul>
<p>For simplicity and algorithmic efficiency, sites must also:</p>
<ul class="simple">
<li><p>Be sorted in increasing order of <code class="docutils literal notranslate"><span class="pre">position</span></code>.</p></li>
</ul>
<p>Violations of these requirements are detected at load time.
The <a class="reference internal" href="python-api.html#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.sort()</span></code></a> method ensures that sites are sorted
according to these criteria.</p>
</section>
<section id="mutation-requirements">
<span id="sec-mutation-requirements"></span><h4>Mutation requirements<a class="headerlink" href="#mutation-requirements" title="Link to this heading">#</a></h4>
<p>Given a valid set of nodes, edges and sites, the
requirements for a valid set of mutations are:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">site</span></code> must refer to a valid site ID;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">node</span></code> must refer to a valid node ID;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">time</span></code> must either be <code class="docutils literal notranslate"><span class="pre">UNKNOWN_TIME</span></code> (a NAN value which indicates
the time is unknown) or be a finite value which is greater or equal to the
mutation <code class="docutils literal notranslate"><span class="pre">node</span></code>’s <code class="docutils literal notranslate"><span class="pre">time</span></code>, less than the <code class="docutils literal notranslate"><span class="pre">node</span></code> above the mutation’s
<code class="docutils literal notranslate"><span class="pre">time</span></code> and equal to or less than the <code class="docutils literal notranslate"><span class="pre">time</span></code> of the <code class="docutils literal notranslate"><span class="pre">parent</span></code> mutation
if this mutation has one. If one mutation on a site has <code class="docutils literal notranslate"><span class="pre">UNKNOWN_TIME</span></code> then
all mutations at that site must, as a mixture of known and unknown is not valid.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">parent</span></code> must either be the null ID (-1), if the mutation has no parent, or a
valid mutation ID within the current table.</p></li>
</ul>
<p>Furthermore,</p>
<ul class="simple">
<li><p>The <code class="docutils literal notranslate"><span class="pre">parent</span></code> value must be consistent with the topology of the tree at the site
of the mutation, such that a path from the child mutation to the parent mutation
exists without passing through any other mutations at the same site.</p></li>
</ul>
<p>For simplicity and algorithmic efficiency, mutations must also:</p>
<ul class="simple">
<li><p>be sorted by site ID;</p></li>
<li><p>when there are multiple mutations per site, mutations should be ordered by
decreasing time, if known, and parent mutations must occur
<strong>before</strong> their children (i.e. if a mutation with ID <span class="math notranslate nohighlight">\(x\)</span> has
<code class="docutils literal notranslate"><span class="pre">parent</span></code> with ID <span class="math notranslate nohighlight">\(y\)</span>, then we must have <span class="math notranslate nohighlight">\(y &lt; x\)</span>).</p></li>
</ul>
<p>Violations of these sorting requirements are detected at load time.
The <a class="reference internal" href="python-api.html#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.sort()</span></code></a> method ensures that mutations are sorted
according site ID, but does not at present enforce that mutations occur
after their parent mutations.</p>
<p>Silent mutations (i.e., mutations for which the ancestral and derived
states are the same) are allowed.
For example, if we have a site with ancestral state
of “A” and a single mutation with derived state “A”, then this
mutation does not result in any change of state.
(This addition was made in release C_0.99.11.)</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>As <code class="docutils literal notranslate"><span class="pre">tskit.UNKNOWN_TIME</span></code> is implemented as a <code class="docutils literal notranslate"><span class="pre">NaN</span></code> value, tests for equality
will always fail. Use <code class="docutils literal notranslate"><span class="pre">tskit.is_unknown_time</span></code> to detect unknown values.</p>
</div>
</section>
<section id="migration-requirements">
<span id="sec-migration-requirements"></span><h4>Migration requirements<a class="headerlink" href="#migration-requirements" title="Link to this heading">#</a></h4>
<p>Given a valid set of nodes and edges, the requirements for a valid set of
migrations are:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">left</span></code> and <code class="docutils literal notranslate"><span class="pre">right</span></code> must be finite values that lie within the tree sequence
coordinate space (i.e., from 0 to <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code>), with <span class="math notranslate nohighlight">\(0 \leq\)</span>
<code class="docutils literal notranslate"><span class="pre">left</span></code> <span class="math notranslate nohighlight">\(&lt;\)</span> <code class="docutils literal notranslate"><span class="pre">right</span></code> <span class="math notranslate nohighlight">\(\leq L\)</span>;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">node</span></code> must be a valid node ID;</p></li>
<li><p>if population references are checked, <code class="docutils literal notranslate"><span class="pre">source</span></code> and <code class="docutils literal notranslate"><span class="pre">dest</span></code> must be valid
population IDs;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">time</span></code> must be a finite value.</p></li>
</ul>
<p>To enable efficient processing, migrations must also be sorted by
nondecreasing <code class="docutils literal notranslate"><span class="pre">time</span></code> value.</p>
<p>Conceptually, a migration records that a segment of ancestry for the given
<code class="docutils literal notranslate"><span class="pre">node</span></code> moves between populations along the tree. In typical demographic
models we expect:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">time</span></code> to lie strictly between the time of the migrating <code class="docutils literal notranslate"><span class="pre">node</span></code> and the time
of any ancestral node from which that node inherits on the segment
<code class="docutils literal notranslate"><span class="pre">[left,</span> <span class="pre">right)</span></code>;</p></li>
<li><p>the <code class="docutils literal notranslate"><span class="pre">population</span></code> of any such ancestor to match the <code class="docutils literal notranslate"><span class="pre">source</span></code> population,
until another <code class="docutils literal notranslate"><span class="pre">migration</span></code> intervenes.</p></li>
</ul>
<p>These conceptual relationships are not currently validated. It is
the responsibility of code that creates migrations to satisfy them where
required.</p>
<p>Note in particular that there is no requirement that adjacent migration records
should be “squashed”. That is, we can have two records <code class="docutils literal notranslate"><span class="pre">m1</span></code> and <code class="docutils literal notranslate"><span class="pre">m2</span></code>
such that <code class="docutils literal notranslate"><span class="pre">m1.right</span></code> = <code class="docutils literal notranslate"><span class="pre">m2.left</span></code> and with the <code class="docutils literal notranslate"><span class="pre">node</span></code>, <code class="docutils literal notranslate"><span class="pre">source</span></code>,
<code class="docutils literal notranslate"><span class="pre">dest</span></code> and <code class="docutils literal notranslate"><span class="pre">time</span></code> fields equal. This is because such records will usually
represent two independent ancestral segments migrating at the same time, and
as such squashing them into a single record would result in a loss of information.</p>
</section>
<section id="population-requirements">
<span id="sec-population-requirements"></span><h4>Population requirements<a class="headerlink" href="#population-requirements" title="Link to this heading">#</a></h4>
<p>There are no requirements on a population table.</p>
</section>
<section id="provenance-requirements">
<span id="sec-provenance-requirements"></span><h4>Provenance requirements<a class="headerlink" href="#provenance-requirements" title="Link to this heading">#</a></h4>
<p>The <code class="docutils literal notranslate"><span class="pre">timestamp</span></code> column of a provenance table should be in
<a class="reference external" href="https://en.wikipedia.org/wiki/ISO_8601">ISO-8601</a> format.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">record</span></code> column stores a JSON document describing how and where the tree sequence
was produced. For tree sequences generated by tskit and related tools, this JSON is
expected to conform to the :ref:<code class="docutils literal notranslate"><span class="pre">provenance</span> <span class="pre">schema</span> <span class="pre">&lt;sec_provenance_schema&gt;</span></code> described
in <a class="reference internal" href="provenance.html#sec-provenance"><span class="std std-ref">Provenance</span></a>.</p>
</section>
</section>
<section id="table-indexes">
<span id="sec-table-indexes"></span><h3>Table indexes<a class="headerlink" href="#table-indexes" title="Link to this heading">#</a></h3>
<p>To efficiently iterate over the trees in a tree sequence, <code class="docutils literal notranslate"><span class="pre">tskit</span></code> uses
indexes built on the edges. To create a tree sequence from a table collection
the tables must be indexed; the <a class="reference internal" href="python-api.html#tskit.TableCollection.build_index" title="tskit.TableCollection.build_index"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.build_index()</span></code></a> method
can be used to create an index on a table collection if necessary.</p>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>Add more details on what the indexes actually are.</p>
</div>
</section>
<section id="saving-to-file">
<span id="sec-data-model-saving"></span><h3>Saving to file<a class="headerlink" href="#saving-to-file" title="Link to this heading">#</a></h3>
<p>When serializing (e.g. storing a <a class="reference internal" href="python-api.html#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> to disk using
<a class="reference internal" href="python-api.html#tskit.TreeSequence.dump" title="tskit.TreeSequence.dump"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dump</span></code></a>), the underlying tables are stored along with the
indexes, top-level metadata, attributes such as the sequence length and time units, and
the <a class="reference internal" href="#sec-data-model-reference-sequence"><span class="std std-ref">Reference sequence</span></a> if it exists. <a class="reference internal" href="python-api.html#tskit.load" title="tskit.load"><code class="xref py py-func docutils literal notranslate"><span class="pre">Loading</span></code></a> such a
file returns an immutable tree sequence object, with pre-calculated indexes immediately
available. See the <a class="reference internal" href="file-formats.html#sec-tree-sequence-file-format"><span class="std std-ref">Tree sequence file format</span></a> section for more details.</p>
<p>Although data in a raw <a class="reference internal" href="python-api.html#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a> need not conform to the
<a class="reference internal" href="#sec-valid-tree-sequence-requirements"><span class="std std-ref">Valid tree sequence requirements</span></a>, it too can be
<a class="reference internal" href="python-api.html#tskit.TableCollection.dump" title="tskit.TableCollection.dump"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dumped</span></code></a> to a file (with indexes stored if they exist).</p>
</section>
</section>
<section id="data-encoding">
<span id="sec-data-model-data-encoding"></span><h2>Data encoding<a class="headerlink" href="#data-encoding" title="Link to this heading">#</a></h2>
<p>In this section we describe the high-level details of how data is encoded in
tables. Tables store data in a <strong>columnar</strong> manner. In memory, each
table is organised as a number of blocks of contiguous storage, one for
each column. There are many advantages to this approach, but the key
property for us is that allows for very efficient transfer of data
in and out of tables. Rather than inserting data into tables row-by-row
(which can be done in Python
<a class="reference internal" href="python-api.html#sec-tables-api-accessing-table-data"><span class="std std-ref">using the add_row methods</span></a>), it is much
more efficient to add many rows at the same time by providing pointers to blocks of
contiguous memory. By taking this approach, we can work with tables containing
gigabytes of data very efficiently.</p>
<p>For instance, in the <a class="reference internal" href="python-api.html#sec-python-api"><span class="std std-ref">Python API</span></a> we can use the
<a class="reference external" href="https://docs.scipy.org/doc/numpy-1.13.0/reference/arrays.html">numpy Array API</a>
to allow us to define and work with numeric arrays of the required types.
Node IDs, for example, are defined using 32 bit integers. Thus, the
<code class="docutils literal notranslate"><span class="pre">parent</span></code> column of an <a class="reference internal" href="#sec-edge-table-definition"><span class="std std-ref">Edge Table</span></a>’s with <code class="docutils literal notranslate"><span class="pre">n</span></code> rows
is a block <code class="docutils literal notranslate"><span class="pre">4n</span></code> bytes.</p>
<p>This approach is very straightforward for columns in which each row contains
a fixed number of values. However, dealing with columns containing a
<strong>variable</strong> number of values is more problematic.</p>
<section id="encoding-ragged-columns">
<span id="sec-encoding-ragged-columns"></span><h3>Encoding ragged columns<a class="headerlink" href="#encoding-ragged-columns" title="Link to this heading">#</a></h3>
<p>A <strong>ragged</strong> column is a column in which the rows are not of a fixed length.
For example, <a class="reference internal" href="#sec-metadata-definition"><span class="std std-ref">Metadata</span></a> columns contain binary of data of arbitrary
length. To encode such columns in the tables API, we store <strong>two</strong> columns:
one contains the flattened array of data and another stores the <strong>offsets</strong>
of each row into this flattened array. Consider an example:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">tskit</span>

<span class="n">s</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">SiteTable</span><span class="p">()</span>
<span class="n">s</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">)</span>
<span class="n">s</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="n">s</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;TTT&quot;</span><span class="p">)</span>
<span class="n">s</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;G&quot;</span><span class="p">)</span>
<span class="n">s</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
        <div>
            <style scoped="">
                .tskit-table tbody tr th:only-of-type {vertical-align: middle;}
                .tskit-table tbody tr th {vertical-align: top;}
                .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}
                .tskit-table tbody th {padding: 0.5em 0.5em;}
            </style>
            <table border="1" class="tskit-table">
                <thead>
                    <tr>
                        <th>id</th><th>position</th><th>ancestral_state</th><th>metadata</th>
                    </tr>
                </thead>
                <tbody>
                    <tr><td>0</td><td>0</td><td>A</td><td></td></tr>
<tr><td>1</td><td>0</td><td></td><td></td></tr>
<tr><td>2</td><td>0</td><td>TTT</td><td></td></tr>
<tr><td>3</td><td>0</td><td>G</td><td></td></tr>

                </tbody>
            </table>
        </div>
    </div></div>
</div>
<p>In this example we create a <a class="reference internal" href="#sec-site-table-definition"><span class="std std-ref">Site Table</span></a> with four rows,
and then display this table. We can see that the second row has the
empty string as its <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code>, and the third row’s
<code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> is <code class="docutils literal notranslate"><span class="pre">TTT</span></code>. Now let’s print out the columns:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Ancestral state (numerical): &quot;</span><span class="p">,</span> <span class="n">s</span><span class="o">.</span><span class="n">ancestral_state</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Ancestral state (as bytes): &quot;</span><span class="p">,</span> <span class="n">s</span><span class="o">.</span><span class="n">ancestral_state</span><span class="o">.</span><span class="n">tobytes</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Ancestral state offsets: &quot;</span><span class="p">,</span> <span class="n">s</span><span class="o">.</span><span class="n">ancestral_state_offset</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Ancestral state (numerical):  [65 84 84 84 71]
Ancestral state (as bytes):  b&#39;ATTTG&#39;
Ancestral state offsets:  [0 1 1 4 5]
</pre></div>
</div>
</div>
</div>
<p>When we print out the tables <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code>
column, we see that its a numpy array of length 5: this is the
flattened array of <a class="reference external" href="https://en.wikipedia.org/wiki/ASCII">ASCII encoded</a>
values for these rows. When we decode these bytes using the
numpy <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.tobytes.html#numpy.ndarray.tobytes" title="(in NumPy v2.4)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">tobytes</span></code></a> method, we get the string ‘ATTTG’.
This flattened array can now be transferred efficiently in memory like any other column
We then use the <code class="docutils literal notranslate"><span class="pre">ancestral_state_offset</span></code> column to allow us find the individual rows.
For a row <code class="docutils literal notranslate"><span class="pre">j</span></code>:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>ancestral_state[ancestral_state_offset[j]: ancestral_state_offset[j + 1]]
</pre></div>
</div>
<p>gives us the array of bytes for the ancestral state in that row. For example, here is
row 2:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">s</span><span class="o">.</span><span class="n">ancestral_state</span><span class="p">[</span><span class="n">s</span><span class="o">.</span><span class="n">ancestral_state_offset</span><span class="p">[</span><span class="mi">2</span><span class="p">]:</span> <span class="n">s</span><span class="o">.</span><span class="n">ancestral_state_offset</span><span class="p">[</span><span class="mi">3</span><span class="p">]]</span><span class="o">.</span><span class="n">tobytes</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>b&#39;TTT&#39;
</pre></div>
</div>
</div>
</div>
<p>For a table with <code class="docutils literal notranslate"><span class="pre">n</span></code> rows, any offset column must have <code class="docutils literal notranslate"><span class="pre">n</span> <span class="pre">+</span> <span class="pre">1</span></code>
values, the first of which is always <code class="docutils literal notranslate"><span class="pre">0</span></code>. The values in this column must be
nondecreasing, and cannot exceed the length of the ragged column in question.</p>
</section>
</section>
<section id="reference-sequence">
<span id="sec-data-model-reference-sequence"></span><h2>Reference sequence<a class="headerlink" href="#reference-sequence" title="Link to this heading">#</a></h2>
<p>Along with the topology and site information stored in the tskit tree
sequence, we can also optionally store an associated reference sequence.
Reference sequences are flexible, and can consist simply of some
metadata recording which assembly build a tree sequence uses, or
storing the entire sequence itself.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Reference sequence support in tskit is preliminary. Reference sequence
data can be stored and accessed via the C API. Support in the Python
API is limited to usage in <a class="reference internal" href="python-api.html#tskit.TreeSequence.alignments" title="tskit.TreeSequence.alignments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.alignments()</span></code></a> and
related methods, where it provides the default values for nucleotide
positions between <a class="reference internal" href="glossary.html#sec-data-model-definitions-site"><span class="std std-ref">sites</span></a>.</p>
</div>
</section>
<section id="tree-structure">
<span id="sec-data-model-tree-structure"></span><h2>Tree structure<a class="headerlink" href="#tree-structure" title="Link to this heading">#</a></h2>
<section id="quintuply-linked-trees">
<span id="sec-data-model-quintuply-linked-trees"></span><h3>Quintuply linked trees<a class="headerlink" href="#quintuply-linked-trees" title="Link to this heading">#</a></h3>
<p>Tree structure in <code class="docutils literal notranslate"><span class="pre">tskit</span></code> is encoded internally as a “quintuply
linked tree”, a generalisation of the triply linked tree encoding
used by Knuth and others. Nodes are represented by their integer
IDs, and their relationships to other nodes are recorded in the
<code class="docutils literal notranslate"><span class="pre">parent</span></code>, <code class="docutils literal notranslate"><span class="pre">left_child</span></code>, <code class="docutils literal notranslate"><span class="pre">right_child</span></code>, <code class="docutils literal notranslate"><span class="pre">left_sib</span></code> and
<code class="docutils literal notranslate"><span class="pre">right_sib</span></code> arrays. For example, consider the following tree
and its associated arrays:</p>
<div class="cell tag_hide-input docutils container">
<details class="admonition hide above-input">
<summary aria-label="Toggle hidden content">
<p class="collapsed admonition-title">Show code cell source</p>
<p class="expanded admonition-title">Hide code cell source</p>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">io</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">tskit</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">IPython.display</span><span class="w"> </span><span class="kn">import</span> <span class="n">SVG</span>

<span class="n">nodes</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span><span class="se">\</span>
<span class="s2">id      is_sample   time</span>
<span class="s2">0       1           0</span>
<span class="s2">1       1           0</span>
<span class="s2">2       1           0</span>
<span class="s2">3       1           0</span>
<span class="s2">4       1           0</span>
<span class="s2">5       0           1</span>
<span class="s2">6       0           2</span>
<span class="s2">7       0           3</span>
<span class="s2">&quot;&quot;&quot;</span>
<span class="n">edges</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span><span class="se">\</span>
<span class="s2">left    right   parent  child</span>
<span class="s2">0       60      5       4,3</span>
<span class="s2">0       40      6       2</span>
<span class="s2">0       60      6       1,0</span>
<span class="s2">20      40      6       5</span>
<span class="s2">0       20      7       5</span>
<span class="s2">40      60      7       5</span>
<span class="s2">0       60      7       6</span>
<span class="s2">40      60      7       2</span>
<span class="s2">&quot;&quot;&quot;</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">load_text</span><span class="p">(</span>
    <span class="n">nodes</span><span class="o">=</span><span class="n">io</span><span class="o">.</span><span class="n">StringIO</span><span class="p">(</span><span class="n">nodes</span><span class="p">),</span> <span class="n">edges</span><span class="o">=</span><span class="n">io</span><span class="o">.</span><span class="n">StringIO</span><span class="p">(</span><span class="n">edges</span><span class="p">),</span> <span class="n">strict</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>

<span class="n">SVG</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">first</span><span class="p">()</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">time_scale</span><span class="o">=</span><span class="s2">&quot;rank&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<img alt="_images/a061e942a6e4074003539868129d64230e01b5a35c6f5e5cdeea30bd08eb8838.svg" src="_images/a061e942a6e4074003539868129d64230e01b5a35c6f5e5cdeea30bd08eb8838.svg" />
</div>
</div>
<div class="cell tag_hide-input docutils container">
<details class="admonition hide above-input">
<summary aria-label="Toggle hidden content">
<p class="collapsed admonition-title">Show code cell source</p>
<p class="expanded admonition-title">Hide code cell source</p>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">IPython.display</span><span class="w"> </span><span class="kn">import</span> <span class="n">HTML</span>

<span class="k">def</span><span class="w"> </span><span class="nf">html_quintuple_table</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">show_virtual_root</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">show_convenience_arrays</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="n">tree</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">first</span><span class="p">()</span>
    <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;node&quot;</span><span class="p">,</span> <span class="s2">&quot;parent&quot;</span><span class="p">,</span> <span class="s2">&quot;left_child&quot;</span><span class="p">,</span> <span class="s2">&quot;right_child&quot;</span><span class="p">,</span> <span class="s2">&quot;left_sib&quot;</span><span class="p">,</span> <span class="s2">&quot;right_sib&quot;</span><span class="p">]</span>
    <span class="n">convenience_arrays</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;num_children&quot;</span><span class="p">,</span> <span class="s2">&quot;edge&quot;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">show_convenience_arrays</span><span class="p">:</span>
        <span class="n">columns</span> <span class="o">+=</span> <span class="n">convenience_arrays</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:[]</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">columns</span><span class="p">}</span>
    <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">nodes</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">virtual_root</span> <span class="k">if</span> <span class="n">show_virtual_root</span> <span class="k">else</span> <span class="kc">None</span><span class="p">)):</span>
        <span class="k">for</span> <span class="n">colname</span> <span class="ow">in</span> <span class="n">columns</span><span class="p">:</span>
            <span class="n">data</span><span class="p">[</span><span class="n">colname</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">u</span> <span class="k">if</span> <span class="n">colname</span> <span class="o">==</span> <span class="s2">&quot;node&quot;</span> <span class="k">else</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">colname</span><span class="p">)(</span><span class="n">u</span><span class="p">))</span>
    <span class="n">html</span> <span class="o">=</span> <span class="s2">&quot;&lt;tr&gt;&quot;</span>
    <span class="k">for</span> <span class="n">colname</span> <span class="ow">in</span> <span class="n">columns</span><span class="p">:</span>
        <span class="n">html</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;&lt;th&gt;</span><span class="si">{</span><span class="n">colname</span><span class="si">}</span><span class="s2">&lt;/th&gt;&quot;</span>
    <span class="n">html</span> <span class="o">+=</span> <span class="s2">&quot;&lt;/tr&gt;&quot;</span>
    <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;node&quot;</span><span class="p">])):</span>
        <span class="n">html</span> <span class="o">+=</span> <span class="s2">&quot;&lt;tr&gt;&quot;</span> <span class="k">if</span> <span class="n">u</span> <span class="o">&lt;</span> <span class="n">ts</span><span class="o">.</span><span class="n">num_nodes</span> <span class="k">else</span> <span class="s2">&quot;&lt;tr style=&#39;font-style: italic; color:red&#39;&gt;&quot;</span>
        <span class="k">for</span> <span class="n">colname</span> <span class="ow">in</span> <span class="n">columns</span><span class="p">:</span>
            <span class="n">html</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;&lt;td&gt;</span><span class="si">{</span><span class="n">data</span><span class="p">[</span><span class="n">colname</span><span class="p">][</span><span class="n">u</span><span class="p">]</span><span class="si">}</span><span class="s2">&lt;/td&gt;&quot;</span>
        <span class="n">html</span> <span class="o">+=</span> <span class="s2">&quot;&lt;/tr&gt;&quot;</span>
    <span class="k">return</span> <span class="s2">&quot;&lt;table&gt;&quot;</span> <span class="o">+</span> <span class="n">html</span> <span class="o">+</span> <span class="s2">&quot;&lt;/table&gt;&quot;</span>

<span class="n">HTML</span><span class="p">(</span><span class="n">html_quintuple_table</span><span class="p">(</span><span class="n">ts</span><span class="p">))</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<div class="output text_html"><table><tr><th>node</th><th>parent</th><th>left_child</th><th>right_child</th><th>left_sib</th><th>right_sib</th></tr><tr><td>0</td><td>6</td><td>-1</td><td>-1</td><td>-1</td><td>1</td></tr><tr><td>1</td><td>6</td><td>-1</td><td>-1</td><td>0</td><td>2</td></tr><tr><td>2</td><td>6</td><td>-1</td><td>-1</td><td>1</td><td>-1</td></tr><tr><td>3</td><td>5</td><td>-1</td><td>-1</td><td>-1</td><td>4</td></tr><tr><td>4</td><td>5</td><td>-1</td><td>-1</td><td>3</td><td>-1</td></tr><tr><td>5</td><td>7</td><td>3</td><td>4</td><td>-1</td><td>6</td></tr><tr><td>6</td><td>7</td><td>0</td><td>2</td><td>5</td><td>-1</td></tr><tr><td>7</td><td>-1</td><td>5</td><td>6</td><td>-1</td><td>-1</td></tr></table></div></div>
</div>
<p>Each node in the tree corresponds to a row in this table, and
the columns are the individual arrays recording the quintuply linked
structure. Thus, we can see that the parent of nodes <code class="docutils literal notranslate"><span class="pre">0</span></code>, <code class="docutils literal notranslate"><span class="pre">1</span></code>, and <code class="docutils literal notranslate"><span class="pre">2</span></code>
is <code class="docutils literal notranslate"><span class="pre">6</span></code>. Similarly, the left child of <code class="docutils literal notranslate"><span class="pre">6</span></code> is <code class="docutils literal notranslate"><span class="pre">0</span></code> and the
right child of <code class="docutils literal notranslate"><span class="pre">6</span></code> is <code class="docutils literal notranslate"><span class="pre">2</span></code>. The <code class="docutils literal notranslate"><span class="pre">left_sib</span></code> and <code class="docutils literal notranslate"><span class="pre">right_sib</span></code> arrays
then record each nodes sibling on its left or right, respectively;
hence the right sib of <code class="docutils literal notranslate"><span class="pre">0</span></code> is <code class="docutils literal notranslate"><span class="pre">1</span></code>, and the right sib of <code class="docutils literal notranslate"><span class="pre">1</span></code> is <code class="docutils literal notranslate"><span class="pre">2</span></code>.
Thus, sibling information allows us to efficiently support trees
with arbitrary numbers of children. In each of the five pointer arrays,
the null node (-1) is used to indicate the end of a path; thus,
for example, the parent of <code class="docutils literal notranslate"><span class="pre">7</span></code> and left sib of <code class="docutils literal notranslate"><span class="pre">0</span></code> are null.</p>
<p>Please see this <a class="reference internal" href="c-api.html#sec-c-api-examples-tree-traversals"><span class="std std-ref">example</span></a> for
details of how to use the quintuply linked structure in the C API.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For many applications we do not need the quintuply linked trees,
and (for example) the <code class="docutils literal notranslate"><span class="pre">left_sib</span></code> and <code class="docutils literal notranslate"><span class="pre">right_child</span></code> arrays can be
ignored. The reason for using a quintuply instead of triply linked
encoding is that it is not possible to efficiently update the trees
as we move along the sequence without the quintuply linked structure.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The left-to-right ordering of nodes is determined by the order
in which edges are inserted into the tree during iteration along the sequence.
Thus, if we arrive at the same tree by iterating from different directions,
the left-to-right ordering of nodes may be different! The specific
ordering of the children of a node should therefore not be depended on.</p>
</div>
</section>
<section id="convenience-arrays">
<h3>Convenience arrays<a class="headerlink" href="#convenience-arrays" title="Link to this heading">#</a></h3>
<p>Similar to the five arrays representing the <a class="reference internal" href="#sec-data-model-quintuply-linked-trees"><span class="std std-ref">quintuply linked tree</span></a>,
convenience arrays track information on each node in the tree. These arrays are not essential to
represent the trees within a tree sequence. However, they can be useful for
specific algorithms (e.g. when computing tree (im)balance metrics). Two
convenience arrays have been implemented so far:
<a class="reference internal" href="python-api.html#tskit.Tree.num_children_array" title="tskit.Tree.num_children_array"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Tree.num_children_array</span></code></a> and <a class="reference internal" href="python-api.html#tskit.Tree.edge_array" title="tskit.Tree.edge_array"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Tree.edge_array</span></code></a>.</p>
<p>Here is the table above with the convenience arrays also shown:</p>
<div class="cell tag_hide-input docutils container">
<details class="admonition hide above-input">
<summary aria-label="Toggle hidden content">
<p class="collapsed admonition-title">Show code cell source</p>
<p class="expanded admonition-title">Hide code cell source</p>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">HTML</span><span class="p">(</span><span class="n">html_quintuple_table</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">show_convenience_arrays</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<div class="output text_html"><table><tr><th>node</th><th>parent</th><th>left_child</th><th>right_child</th><th>left_sib</th><th>right_sib</th><th>num_children</th><th>edge</th></tr><tr><td>0</td><td>6</td><td>-1</td><td>-1</td><td>-1</td><td>1</td><td>0</td><td>2</td></tr><tr><td>1</td><td>6</td><td>-1</td><td>-1</td><td>0</td><td>2</td><td>0</td><td>3</td></tr><tr><td>2</td><td>6</td><td>-1</td><td>-1</td><td>1</td><td>-1</td><td>0</td><td>4</td></tr><tr><td>3</td><td>5</td><td>-1</td><td>-1</td><td>-1</td><td>4</td><td>0</td><td>0</td></tr><tr><td>4</td><td>5</td><td>-1</td><td>-1</td><td>3</td><td>-1</td><td>0</td><td>1</td></tr><tr><td>5</td><td>7</td><td>3</td><td>4</td><td>-1</td><td>6</td><td>2</td><td>7</td></tr><tr><td>6</td><td>7</td><td>0</td><td>2</td><td>5</td><td>-1</td><td>3</td><td>9</td></tr><tr><td>7</td><td>-1</td><td>5</td><td>6</td><td>-1</td><td>-1</td><td>2</td><td>-1</td></tr></table></div></div>
</div>
</section>
<section id="roots">
<span id="sec-data-model-tree-roots"></span><h3>Roots<a class="headerlink" href="#roots" title="Link to this heading">#</a></h3>
<p>In the <code class="docutils literal notranslate"><span class="pre">tskit</span></code> <a class="reference internal" href="python-api.html#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">trees</span></code></a> we have shown so far, all the sample nodes have
been connected to each other. This means each tree has only a single <a class="reference internal" href="python-api.html#tskit.Tree.root" title="tskit.Tree.root"><code class="xref py py-attr docutils literal notranslate"><span class="pre">root</span></code></a>
(i.e. the oldest node found when tracing a path backwards in time from any sample).
However, a tree can contain <a class="reference internal" href="#sec-data-model-tree-isolated-sample-nodes"><span class="std std-ref">Isolated sample nodes</span></a>
or unconnected topologies, and can therefore have <em>multiple</em> <a class="reference internal" href="python-api.html#tskit.Tree.roots" title="tskit.Tree.roots"><code class="xref py py-attr docutils literal notranslate"><span class="pre">roots</span></code></a>.
Here’s an example, created by deleting the edge joining <code class="docutils literal notranslate"><span class="pre">6</span></code> and <code class="docutils literal notranslate"><span class="pre">7</span></code> in the tree sequence
used above:</p>
<div class="cell tag_hide-input docutils container">
<details class="admonition hide above-input">
<summary aria-label="Toggle hidden content">
<p class="collapsed admonition-title">Show code cell source</p>
<p class="expanded admonition-title">Hide code cell source</p>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tables</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">dump_tables</span><span class="p">()</span>
<span class="n">tables</span><span class="o">.</span><span class="n">edges</span><span class="o">.</span><span class="n">truncate</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">num_edges</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">ts_multiroot</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">()</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts_multiroot</span><span class="o">.</span><span class="n">first</span><span class="p">()</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">time_scale</span><span class="o">=</span><span class="s2">&quot;rank&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<img alt="_images/26348eac8933f5a92fefe5efc28824746e052f084161386fff38ab8addb29ecf.svg" src="_images/26348eac8933f5a92fefe5efc28824746e052f084161386fff38ab8addb29ecf.svg" />
</div>
</div>
<p>In <code class="docutils literal notranslate"><span class="pre">tskit</span></code> terminology, this should <em>not</em> be thought
of as two separate trees, but as a single multi-root “tree”, comprising
two unlinked topologies. This fits with the definition of a tree
in a tree sequence: a tree describes the ancestry of the same
fixed set of sample nodes at a single position in the genome. In the
picture above, <em>both</em> the left and right hand topologies are required
to describe the genealogy of samples 0..4 at this position.</p>
<p>Here’s what the entire tree sequence now looks like:</p>
<div class="cell tag_hide-input docutils container">
<details class="admonition hide above-input">
<summary aria-label="Toggle hidden content">
<p class="collapsed admonition-title">Show code cell source</p>
<p class="expanded admonition-title">Hide code cell source</p>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">SVG</span><span class="p">(</span><span class="n">ts_multiroot</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">time_scale</span><span class="o">=</span><span class="s2">&quot;rank&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<img alt="_images/09d0472eb00362db2818a89e8a134c3d45d87586b1c333bbcae1cde312c37651.svg" src="_images/09d0472eb00362db2818a89e8a134c3d45d87586b1c333bbcae1cde312c37651.svg" />
</div>
</div>
<p>From the terminology above, it can be seen that this tree sequence consists of only
three trees (not five). The first tree, which applies from position 0 to 20, is the one
used in our example. As we saw, removing the edge connecting node 6 to node 7 has
created a tree with 2 roots (and thus 2 unconnected topologies in a single tree).
In contrast, the second tree, from position 20 to 40, has a single root. Finally the
third tree, from position 40 to 60, again has two roots.</p>
<section id="the-root-threshold">
<span id="sec-data-model-tree-root-threshold"></span><h4>The root threshold<a class="headerlink" href="#the-root-threshold" title="Link to this heading">#</a></h4>
<p>The roots of a tree are defined by reference to the
<a class="reference internal" href="glossary.html#sec-data-model-definitions-sample"><span class="std std-ref">sample nodes</span></a>. By default, roots are the unique
endpoints of the paths traced upwards from the sample nodes; equivalently, each root
counts one or more samples among its descendants (or is itself a sample node). This is
the case when the <a class="reference internal" href="python-api.html#tskit.Tree.root_threshold" title="tskit.Tree.root_threshold"><code class="xref py py-attr docutils literal notranslate"><span class="pre">root_threshold</span></code></a> property of a tree is left at its default
value of <code class="docutils literal notranslate"><span class="pre">1</span></code>. If, however, the <code class="docutils literal notranslate"><span class="pre">root_threshold</span></code> is (say) <code class="docutils literal notranslate"><span class="pre">2</span></code>, then a node is
considered a root only if it counts at least two samples among its descendants. Setting
an alternative <code class="docutils literal notranslate"><span class="pre">root_threshold</span></code> value can be used to avoid visiting
<a class="reference internal" href="#sec-data-model-tree-isolated-sample-nodes"><span class="std std-ref">Isolated sample nodes</span></a>, for example when dealing with trees
containing <a class="reference internal" href="#sec-data-model-missing-data"><span class="std std-ref">Missing data</span></a>.</p>
</section>
<section id="the-virtual-root">
<span id="sec-data-model-tree-virtual-root"></span><h4>The virtual root<a class="headerlink" href="#the-virtual-root" title="Link to this heading">#</a></h4>
<p>To access all the <a class="reference internal" href="python-api.html#tskit.Tree.roots" title="tskit.Tree.roots"><code class="xref py py-attr docutils literal notranslate"><span class="pre">roots</span></code></a> in a tree, tskit uses a special additional node
called the <strong>virtual root</strong>. This is primarily a bookkeeping device, and
can normally be ignored: it is not plotted in any visualizations and
does not exist as an independent node in the node table.
However, the virtual root can be useful in certain algorithms because its
children are defined as all the “real” roots in a tree. Hence by
descending downwards from the virtual root, it is possible
to access the entire genealogy at a given site, even in a multi-root
tree. In the quintuply linked tree encoding, the virtual root appears as an
extra element at the end of each of the tree arrays. Here’s the same table
as before but with the virtual root also shown, using red italics to
emphasise that it is not a “real” node:</p>
<div class="cell tag_hide-input docutils container">
<details class="admonition hide above-input">
<summary aria-label="Toggle hidden content">
<p class="collapsed admonition-title">Show code cell source</p>
<p class="expanded admonition-title">Hide code cell source</p>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">HTML</span><span class="p">(</span><span class="n">html_quintuple_table</span><span class="p">(</span><span class="n">ts_multiroot</span><span class="p">,</span> <span class="n">show_virtual_root</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<div class="output text_html"><table><tr><th>node</th><th>parent</th><th>left_child</th><th>right_child</th><th>left_sib</th><th>right_sib</th></tr><tr><td>0</td><td>6</td><td>-1</td><td>-1</td><td>-1</td><td>1</td></tr><tr><td>1</td><td>6</td><td>-1</td><td>-1</td><td>0</td><td>2</td></tr><tr><td>2</td><td>6</td><td>-1</td><td>-1</td><td>1</td><td>-1</td></tr><tr><td>3</td><td>5</td><td>-1</td><td>-1</td><td>-1</td><td>4</td></tr><tr><td>4</td><td>5</td><td>-1</td><td>-1</td><td>3</td><td>-1</td></tr><tr><td>5</td><td>7</td><td>3</td><td>4</td><td>-1</td><td>-1</td></tr><tr><td>6</td><td>-1</td><td>0</td><td>2</td><td>-1</td><td>7</td></tr><tr><td>7</td><td>-1</td><td>5</td><td>5</td><td>6</td><td>-1</td></tr><tr style='font-style: italic; color:red'><td>8</td><td>-1</td><td>6</td><td>7</td><td>-1</td><td>-1</td></tr></table></div></div>
</div>
<p>You can see that the virtual root (node 8) has 6 as its left child and 7
as its right child. Importantly, though, this is an asymmetric
relationship: the parent of the “real” roots 6 and 7 is null
(-1) and <em>not</em> the virtual root. Hence when we ascend up the tree from the
sample nodes to their parents, we stop at the “real” roots, and never
encounter the virtual root.</p>
<p>Because the virtual root can be useful in some algorithms, it can
optionally be returned in traversal orders (see <a class="reference internal" href="python-api.html#tskit.Tree.nodes" title="tskit.Tree.nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.nodes()</span></code></a>).
The following properties apply:</p>
<ul class="simple">
<li><p>All trees in a tree sequence share the same virtual root.</p></li>
<li><p>The virtual root’s ID is always equal to the number of nodes in the tree sequence
(i.e. the length of the node table). However, there is <strong>no corresponding row</strong>
in the node table, and any attempts to access information about the
virtual root via either the tree sequence or tables APIs will fail with
an out-of-bounds error.</p></li>
<li><p>The parent and siblings of the virtual root are null.</p></li>
<li><p>The time of the virtual root is defined as positive infinity (if
accessed via <a class="reference internal" href="python-api.html#tskit.Tree.time" title="tskit.Tree.time"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.time()</span></code></a>). This is useful in defining the
time-based node traversal orderings.</p></li>
<li><p>The virtual root is the parent of no other node—roots do <strong>not</strong>
have parent pointers to the virtual root.</p></li>
</ul>
</section>
</section>
<section id="isolated-nodes">
<span id="sec-data-model-tree-isolated-nodes"></span><h3>Isolated nodes<a class="headerlink" href="#isolated-nodes" title="Link to this heading">#</a></h3>
<p>In a tree, it is possible for a node to have no children and no parent. Such a node is
said to be <em>isolated</em>, meaning that we don’t know anything about its relationships
over a specific genomic interval. This is commonly true for ancestral genomes, which
often have large regions that have
not been inherited by any of the <a class="reference internal" href="glossary.html#sec-data-model-definitions-sample"><span class="std std-ref">sample nodes</span></a>
in the tree sequence, and therefore regions about which we know nothing. This is true,
for example, of node 7 in the middle tree of our previous example, which is why it is
not plotted on that tree:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">display</span><span class="p">(</span><span class="n">SVG</span><span class="p">(</span><span class="n">ts_multiroot</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">time_scale</span><span class="o">=</span><span class="s2">&quot;rank&quot;</span><span class="p">)))</span>
<span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">ts_multiroot</span><span class="o">.</span><span class="n">trees</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span>
        <span class="s2">&quot;Node 7&quot;</span><span class="p">,</span>
        <span class="s2">&quot;is&quot;</span> <span class="k">if</span> <span class="n">tree</span><span class="o">.</span><span class="n">is_isolated</span><span class="p">(</span><span class="mi">7</span><span class="p">)</span> <span class="k">else</span> <span class="s2">&quot;is not&quot;</span><span class="p">,</span>
        <span class="s2">&quot;isolated from position&quot;</span><span class="p">,</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">interval</span><span class="o">.</span><span class="n">left</span><span class="p">,</span>
        <span class="s2">&quot;to&quot;</span><span class="p">,</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">interval</span><span class="o">.</span><span class="n">right</span><span class="p">,</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/09d0472eb00362db2818a89e8a134c3d45d87586b1c333bbcae1cde312c37651.svg" src="_images/09d0472eb00362db2818a89e8a134c3d45d87586b1c333bbcae1cde312c37651.svg" />
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Node 7 is not isolated from position 0.0 to 20.0
Node 7 is isolated from position 20.0 to 40.0
Node 7 is not isolated from position 40.0 to 60.0
</pre></div>
</div>
</div>
</div>
<section id="isolated-sample-nodes">
<span id="sec-data-model-tree-isolated-sample-nodes"></span><h4>Isolated sample nodes<a class="headerlink" href="#isolated-sample-nodes" title="Link to this heading">#</a></h4>
<p>It is also possible for a <a class="reference internal" href="glossary.html#sec-data-model-definitions-sample"><span class="std std-ref">sample node</span></a>
to be isolated. As long as the <a class="reference internal" href="#sec-data-model-tree-root-threshold"><span class="std std-ref">root threshold</span></a>
is set to its default value, an isolated <em>sample</em> node will count as a root, and
therefore be considered as being present on the tree (meaning it will be
returned by the <a class="reference internal" href="python-api.html#tskit.Tree.nodes" title="tskit.Tree.nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.nodes()</span></code></a>
and <a class="reference internal" href="python-api.html#tskit.Tree.samples" title="tskit.Tree.samples"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.samples()</span></code></a> methods). When displaying a tree, isolated samples are shown
unconnected to other nodes. To illustrate, we can remove the edge from node 2 to node 7:</p>
<div class="cell tag_hide-input docutils container">
<details class="admonition hide above-input">
<summary aria-label="Toggle hidden content">
<p class="collapsed admonition-title">Show code cell source</p>
<p class="expanded admonition-title">Hide code cell source</p>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tables</span> <span class="o">=</span> <span class="n">ts_multiroot</span><span class="o">.</span><span class="n">dump_tables</span><span class="p">()</span>
<span class="n">tables</span><span class="o">.</span><span class="n">edges</span><span class="o">.</span><span class="n">set_columns</span><span class="p">(</span>
    <span class="o">**</span><span class="n">tables</span><span class="o">.</span><span class="n">edges</span><span class="p">[(</span><span class="n">tables</span><span class="o">.</span><span class="n">edges</span><span class="o">.</span><span class="n">parent</span> <span class="o">!=</span> <span class="mi">7</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">tables</span><span class="o">.</span><span class="n">edges</span><span class="o">.</span><span class="n">child</span> <span class="o">!=</span> <span class="mi">2</span><span class="p">)]</span><span class="o">.</span><span class="n">asdict</span><span class="p">())</span>
<span class="n">ts_isolated</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">()</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">ts_isolated</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">(</span><span class="n">time_scale</span><span class="o">=</span><span class="s2">&quot;rank&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<img alt="_images/ce854e1cb0f2250816018e7aa007d274f5c0cbf7adde34aa8a0b8c41f1e4ae21.svg" src="_images/ce854e1cb0f2250816018e7aa007d274f5c0cbf7adde34aa8a0b8c41f1e4ae21.svg" />
</div>
</div>
<p>The rightmost tree now contains an isolated sample node (node 2), which counts as
one of the <a class="reference internal" href="#sec-data-model-tree-roots"><span class="std std-ref">Roots</span></a> of the tree. This tree therefore has three
roots, one of which is node 2:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rightmost_tree</span> <span class="o">=</span> <span class="n">ts_isolated</span><span class="o">.</span><span class="n">at_index</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">rightmost_tree</span><span class="o">.</span><span class="n">num_roots</span><span class="p">,</span> <span class="s2">&quot;roots in the rightmost tree, with IDs&quot;</span><span class="p">,</span> <span class="n">rightmost_tree</span><span class="o">.</span><span class="n">roots</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="s2">&quot;IDs of isolated samples in this tree:&quot;</span><span class="p">,</span>
    <span class="p">[</span><span class="n">u</span> <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">rightmost_tree</span><span class="o">.</span><span class="n">samples</span><span class="p">()</span> <span class="k">if</span> <span class="n">rightmost_tree</span><span class="o">.</span><span class="n">is_isolated</span><span class="p">(</span><span class="n">u</span><span class="p">)],</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>3 roots in the rightmost tree, with IDs [2, 6, 7]
IDs of isolated samples in this tree: [2]
</pre></div>
</div>
</div>
</div>
<p>In <code class="docutils literal notranslate"><span class="pre">tskit</span></code>, isolated sample nodes are closely associated with the encoding of
<a class="reference internal" href="#sec-data-model-missing-data"><span class="std std-ref">Missing data</span></a>.</p>
</section>
</section>
<section id="dead-leaves-and-branches">
<span id="sec-data-model-tree-dead-leaves-and-branches"></span><h3>Dead leaves and branches<a class="headerlink" href="#dead-leaves-and-branches" title="Link to this heading">#</a></h3>
<p>In a <code class="docutils literal notranslate"><span class="pre">tskit</span></code> tree, a <em>leaf node</em> is defined as a node without any children. The
implications of this turn out to be slighly unintuitive, and so are worth briefly
documenting here. Firstly, the same node can be a leaf in one tree, and not a leaf
in the next tree along the tree sequence. Secondly all isolated nodes must be leaves
(as by definition they have no children). Thirdly sample nodes need not be leaves
(they could be “internal samples”); likewise leaf nodes need not be samples.</p>
<p>Node 7 in the example above provides a good case study. Note that it is a root node with
at least one child (i.e. not a leaf) in trees 0 and 2; in contrast in tree 1 it is
isolated. Strictly, because it is isolated in tree 1, it is also a leaf node there,
although it is not attached to a root, not a sample, and is therefore not plotted. In
this case, in that tree we can think of node 7 as a “dead leaf” (and we don’t normally
plot dead leaves). In fact, in a large tree sequence of many trees, most ancestral nodes
will be isolated in any given tree, and therefore most nodes in such a tree will be of
this sort. However, these dead leaves are excluded from most calculations on trees,
because algorithms usually traverse the tree by starting at a root and working down,
or by starting at a sample and working up. Hence when we refer to the leaves of a tree,
it is usually shorthand for the leaves <strong>on</strong> the tree (that is, attached via branches,
to one of the the tree roots). Dead leaves are excluded from this definition.</p>
<p>Note that it is also possible to have trees in which there are “dead branches”: that is
sections of topology which are not accessible from a root, and whose tips are all
dead leaves. Although valid, this is a relatively unusual state of affairs, and such
branches are not plotted by the standard <a class="reference external" href="https://tskit.dev/tutorials/viz.html#sec-tskit-viz" title="(in Project name not set)"><span>Visualization</span></a> methods. The
<a class="reference internal" href="python-api.html#tskit.Tree.nodes" title="tskit.Tree.nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.nodes()</span></code></a> method will not, by default, traverse through dead branches, although
it can be made to do so by specifying the ID of a dead node as the root for traversal.</p>
</section>
</section>
<section id="encoding-genetic-variation">
<span id="sec-data-model-genetic-data"></span><h2>Encoding genetic variation<a class="headerlink" href="#encoding-genetic-variation" title="Link to this heading">#</a></h2>
<p>Genetic variation is incorporated into a tree sequence by placing
<a class="reference internal" href="#sec-mutation-table-definition"><span class="std std-ref">mutations</span></a> at
<a class="reference internal" href="#sec-mutation-table-definition"><span class="std std-ref">sites</span></a> along the genome.
The genotypes of the different samples at each site
can be found by using the tree to calculate which mutations are inherited by
the different samples. This is the fundamental basis of how tree sequences
efficiently encode DNA sequences, and is explained in depth elsewhere
(e.g. <a class="reference external" href="https://tskit.dev/tutorials/what_is.html#sec-what-is-dna-data" title="(in Project name not set)"><span class="xref std std-ref">in the tutorials</span></a>).</p>
<p>Below, we discuss some implications of this encoding in more detail, in particular
the way in which it can be used to model missing data.</p>
<section id="missing-data">
<span id="sec-data-model-missing-data"></span><h3>Missing data<a class="headerlink" href="#missing-data" title="Link to this heading">#</a></h3>
<p>If, at a particular genomic position, a node is
<a class="reference internal" href="#sec-data-model-tree-isolated-nodes"><span class="std std-ref">isolated</span></a> <em>and</em> additionally has
no mutations directly above it, its genotype at that position is considered to be
unknown (however, if there is a mutation above an isolated node, it
can be thought of as saying directly what the genotype is, and so renders the
genotype at that position not missing).</p>
<p>By way of illustration, we’ll use the <a class="reference internal" href="python-api.html#tskit.TableCollection.delete_intervals" title="tskit.TableCollection.delete_intervals"><code class="xref py py-meth docutils literal notranslate"><span class="pre">delete_intervals()</span></code></a> method
to remove all knowledge of the ancestry in the
middle portion of the previous example (say from position 15 to 45) sprinkle
on some mutations, and make sure there are sites at every position:</p>
<div class="cell tag_hide-input docutils container">
<details class="admonition hide above-input">
<summary aria-label="Toggle hidden content">
<p class="collapsed admonition-title">Show code cell source</p>
<p class="expanded admonition-title">Hide code cell source</p>
</summary>
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">msprime</span>

<span class="n">tables</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_mutations</span><span class="p">(</span><span class="n">ts_isolated</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">123</span><span class="p">)</span><span class="o">.</span><span class="n">dump_tables</span><span class="p">()</span>
<span class="n">tables</span><span class="o">.</span><span class="n">delete_intervals</span><span class="p">([[</span><span class="mi">15</span><span class="p">,</span> <span class="mi">45</span><span class="p">]],</span> <span class="n">simplify</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">missing_sites</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">setdiff1d</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">tables</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">),</span> <span class="n">tables</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">position</span><span class="p">)</span>
<span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">missing_sites</span><span class="p">:</span>
    <span class="n">tables</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">position</span><span class="o">=</span><span class="n">pos</span><span class="p">,</span> <span class="n">ancestral_state</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">)</span>  <span class="c1"># Add sites at every pos</span>
<span class="n">tables</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
<span class="n">missing_ts</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">()</span>
<span class="n">SVG</span><span class="p">(</span><span class="n">missing_ts</span><span class="o">.</span><span class="n">draw_svg</span><span class="p">())</span>
</pre></div>
</div>
</div>
</details>
<div class="cell_output docutils container">
<img alt="_images/8cbfeb6afab442c86c8805fbf637dbaea817a6937913ab3475134aa9a7bdea8f.svg" src="_images/8cbfeb6afab442c86c8805fbf637dbaea817a6937913ab3475134aa9a7bdea8f.svg" />
</div>
</div>
<p>The middle section of the genome now has no ancestry at all, and therefore for any site
that is in this region, the genotypic state that it is assigned is a special value
<code class="docutils literal notranslate"><span class="pre">tskit.MISSING_DATA</span></code>, or <code class="docutils literal notranslate"><span class="pre">-1</span></code>. The <a class="reference internal" href="python-api.html#tskit.TreeSequence.haplotypes" title="tskit.TreeSequence.haplotypes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">haplotypes()</span></code></a> method, which
outputs the actual allelic state for each sample, defaults to outputting an <code class="docutils literal notranslate"><span class="pre">N</span></code> at
these sites. Therefore where any sample node is isolated, the haplotype will show
an <code class="docutils literal notranslate"><span class="pre">N</span></code>, indicating the DNA sequence is unknown. This will be so not only in the
middle of all of the sample genomes, but also at the right hand end of the genome of
sample 2, as it is an isolated sample node in the rightmost tree:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">h</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">missing_ts</span><span class="o">.</span><span class="n">haplotypes</span><span class="p">()):</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Sample </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">h</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Sample 0: AAAGCCAATGAAATANNNNNNNNNNNNNNNNNNNNNNNNNNNNNNGGAACGAATTGAACT
Sample 1: AGAGACAATAGAATANNNNNNNNNNNNNNNNNNNNNNNNNNNNNNTAAATAACCTGAACT
Sample 2: AGAGACGACAGAATANNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
Sample 3: AGATACGATAGCAAANNNNNNNNNNNNNNNNNNNNNNNNNNNNNNTCTACGAACAAAAAC
Sample 4: AGAGATGATAGCAAANNNNNNNNNNNNNNNNNNNNNNNNNNNNNNTCTACGAACAAAACC
</pre></div>
</div>
</div>
</div>
<p>See the <a class="reference internal" href="python-api.html#tskit.TreeSequence.variants" title="tskit.TreeSequence.variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.variants()</span></code></a> method and <a class="reference internal" href="python-api.html#tskit.Variant" title="tskit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> class for
more information on how missing data is represented in variant data.</p>
</section>
</section>
<section id="possibly-surprising-consequences-of-the-data-model">
<span id="sec-gotchas"></span><h2>Possibly surprising consequences of the data model<a class="headerlink" href="#possibly-surprising-consequences-of-the-data-model" title="Link to this heading">#</a></h2>
<p>This is a section of miscellaneous issues that might trip even an experienced user up,
also known as “gotchas”.
The current examples are quite uncommon, so can be ignored for most purposes,
but the list may be expanded in the future.</p>
<section id="unrelated-material">
<h3>Unrelated material<a class="headerlink" href="#unrelated-material" title="Link to this heading">#</a></h3>
<p>Usually, all parts of a tree sequence are ancestral to at least one sample,
since that’s essentially the definition of a sample: the genomes that
we’re describing the ancestry of.
However, in some cases there will be portions of the tree sequence from which
no samples inherit - notably, the result of a forwards simulation that has
not been simplified.
In fact, if the simulation has not coalesced,
one can have entire portions of some marginal tree that are
unrelated to any of the samples
(for instance, an individual in the initial generation of the simulation
that had no offspring).
This can lead to a gotcha:
the <em>roots</em> of a tree are defined to be only those roots <em>reachable from the samples</em>
(and, furthermore, reachable from at least <code class="docutils literal notranslate"><span class="pre">root_threshold</span></code> samples;
see <a class="reference internal" href="python-api.html#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.trees()</span></code></a>).
So, our unlucky ancestor would not appear in the list of <code class="docutils literal notranslate"><span class="pre">roots</span></code>, even though
if we drew all the relationships provided by the tree sequence,
they’d definitely be a root.
Furthermore, only nodes <em>reachable from a root</em> are included in the
<a class="reference internal" href="python-api.html#tskit.Tree.nodes" title="tskit.Tree.nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.nodes()</span></code></a>. So, if you iterate over all the nodes in each marginal tree,
you won’t see those parts of the tree sequence that are unrelated to the samples.
If you need to get those, too, you could either
work with the <a class="reference internal" href="python-api.html#tskit.TreeSequence.edge_diffs" title="tskit.TreeSequence.edge_diffs"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.edge_diffs()</span></code></a> directly,
or iterate over all nodes (instead of over <a class="reference internal" href="python-api.html#tskit.Tree.nodes" title="tskit.Tree.nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.nodes()</span></code></a>).</p>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="glossary.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Glossary</p>
      </div>
    </a>
    <a class="right-next"
       href="metadata.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Metadata</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#table-definitions">Table definitions</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#table-types">Table types</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#node-table">Node Table</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#individual-table">Individual Table</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#edge-table">Edge Table</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#site-table">Site Table</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mutation-table">Mutation Table</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#migration-table">Migration Table</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#population-table">Population Table</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#provenance-table">Provenance Table</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#metadata">Metadata</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#valid-tree-sequence-requirements">Valid tree sequence requirements</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#individual-requirements">Individual requirements</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#node-requirements">Node requirements</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#edge-requirements">Edge requirements</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#site-requirements">Site requirements</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mutation-requirements">Mutation requirements</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#migration-requirements">Migration requirements</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#population-requirements">Population requirements</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#provenance-requirements">Provenance requirements</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#table-indexes">Table indexes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#saving-to-file">Saving to file</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-encoding">Data encoding</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#encoding-ragged-columns">Encoding ragged columns</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#reference-sequence">Reference sequence</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tree-structure">Tree structure</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#quintuply-linked-trees">Quintuply linked trees</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#convenience-arrays">Convenience arrays</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#roots">Roots</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-root-threshold">The root threshold</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-virtual-root">The virtual root</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#isolated-nodes">Isolated nodes</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#isolated-sample-nodes">Isolated sample nodes</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#dead-leaves-and-branches">Dead leaves and branches</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#encoding-genetic-variation">Encoding genetic variation</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#missing-data">Missing data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#possibly-surprising-consequences-of-the-data-model">Possibly surprising consequences of the data model</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#unrelated-material">Unrelated material</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Tskit Developers
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>