
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Python API &#8212; Tskit manual</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=5349f25f" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=9c3e77be" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="_static/bespoke.css?v=00f581dd" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=1ae7504c"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'python-api';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Numba Integration" href="numba.html" />
    <link rel="prev" title="Data export" href="export.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <script data-goatcounter="https://tskit.goatcounter.com/count" async src="//gc.zgo.at/count.js"></script>
</head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="introduction.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo.svg" class="logo__image only-light" alt=""/>
    <script>document.write(`<img src="_static/logo.svg" class="logo__image only-dark" alt=""/>`);</script>
  
  
    <p class="title logo__title">Version 1.0.1</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Concepts</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="data-model.html">Data model</a></li>
<li class="toctree-l1"><a class="reference internal" href="metadata.html">Metadata</a></li>
<li class="toctree-l1"><a class="reference internal" href="provenance.html">Provenance</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="stats.html">Statistics</a></li>
<li class="toctree-l1"><a class="reference internal" href="topological-analysis.html">Topological analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="ibd.html">Identity by descent</a></li>
<li class="toctree-l1"><a class="reference internal" href="export.html">Data export</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Interfaces</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Python API</a></li>
<li class="toctree-l1"><a class="reference internal" href="numba.html">Numba Integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="c-api.html">C API</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">Command line interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="file-formats.html">File formats</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">For developers</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="development.html">Development</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelogs.html">Changelogs</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Miscellaneous</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="citation.html">Citing tskit</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/tskit-dev/tskit" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/tskit-dev/tskit/edit/main/docs/python-api.md" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/tskit-dev/tskit/issues/new?title=Issue%20on%20page%20%2Fpython-api.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/python-api.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Python API</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#trees-and-tree-sequences">Trees and tree sequences</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#treesequence-api"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code> API</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#general-properties">General properties</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#efficient-table-column-access">Efficient table column access</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-and-saving">Loading and saving</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#obtaining-trees">Obtaining trees</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#obtaining-other-objects">Obtaining other objects</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tree-topology">Tree topology</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#genetic-variation">Genetic variation</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#demography">Demography</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#other">Other</a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tree-sequence-modification">Tree sequence modification</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#sec-python-api-tree-sequences-ibd">Identity by descent</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tables">Tables</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#statistics">Statistics</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#topological-analysis">Topological analysis</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#display">Display</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#export">Export</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tree-api"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code> API</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#sec-python-api-trees-general-properties">General properties</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#creating-new-trees">Creating new trees</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#node-measures">Node measures</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#simple-measures">Simple measures</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#array-access">Array access</a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tree-traversal">Tree traversal</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#sec-python-api-trees-topological-analysis">Topological analysis</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#comparing-trees">Comparing trees</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#balance-imbalance-indices">Balance/imbalance indices</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#sites-and-mutations">Sites and mutations</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#moving-to-other-trees">Moving to other trees</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#id3">Display</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#id4">Export</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tables-and-table-collections">Tables and Table Collections</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tablecollection-api"><code class="docutils literal notranslate"><span class="pre">TableCollection</span></code> API</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#id5">General properties</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#transformation">Transformation</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#modification">Modification</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#creating-a-valid-tree-sequence">Creating a valid tree sequence</a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#miscellaneous-methods">Miscellaneous methods</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#id6">Export</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#table-apis">Table APIs</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#accessing-table-data">Accessing table data</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#text-columns">Text columns</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#binary-columns">Binary columns</a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#table-functions">Table functions</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#metadata-api">Metadata API</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#provenance">Provenance</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#utility-functions">Utility functions</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#reference-documentation">Reference documentation</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#constants">Constants</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NULL"><code class="docutils literal notranslate"><span class="pre">NULL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MISSING_DATA"><code class="docutils literal notranslate"><span class="pre">MISSING_DATA</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NODE_IS_SAMPLE"><code class="docutils literal notranslate"><span class="pre">NODE_IS_SAMPLE</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.FORWARD"><code class="docutils literal notranslate"><span class="pre">FORWARD</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.REVERSE"><code class="docutils literal notranslate"><span class="pre">REVERSE</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ALLELES_01"><code class="docutils literal notranslate"><span class="pre">ALLELES_01</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ALLELES_ACGT"><code class="docutils literal notranslate"><span class="pre">ALLELES_ACGT</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.UNKNOWN_TIME"><code class="docutils literal notranslate"><span class="pre">UNKNOWN_TIME</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TIME_UNITS_UNKNOWN"><code class="docutils literal notranslate"><span class="pre">TIME_UNITS_UNKNOWN</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TIME_UNITS_UNCALIBRATED"><code class="docutils literal notranslate"><span class="pre">TIME_UNITS_UNCALIBRATED</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exceptions">Exceptions</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.DuplicatePositionsError"><code class="docutils literal notranslate"><span class="pre">DuplicatePositionsError</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MetadataEncodingError"><code class="docutils literal notranslate"><span class="pre">MetadataEncodingError</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MetadataSchemaValidationError"><code class="docutils literal notranslate"><span class="pre">MetadataSchemaValidationError</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MetadataValidationError"><code class="docutils literal notranslate"><span class="pre">MetadataValidationError</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceValidationError"><code class="docutils literal notranslate"><span class="pre">ProvenanceValidationError</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#top-level-functions">Top-level functions</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.all_trees"><code class="docutils literal notranslate"><span class="pre">all_trees()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.all_tree_shapes"><code class="docutils literal notranslate"><span class="pre">all_tree_shapes()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.all_tree_labellings"><code class="docutils literal notranslate"><span class="pre">all_tree_labellings()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.is_unknown_time"><code class="docutils literal notranslate"><span class="pre">is_unknown_time()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.load"><code class="docutils literal notranslate"><span class="pre">load()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.load_text"><code class="docutils literal notranslate"><span class="pre">load_text()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.pack_bytes"><code class="docutils literal notranslate"><span class="pre">pack_bytes()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.pack_strings"><code class="docutils literal notranslate"><span class="pre">pack_strings()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.parse_edges"><code class="docutils literal notranslate"><span class="pre">parse_edges()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.parse_individuals"><code class="docutils literal notranslate"><span class="pre">parse_individuals()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.parse_mutations"><code class="docutils literal notranslate"><span class="pre">parse_mutations()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.parse_nodes"><code class="docutils literal notranslate"><span class="pre">parse_nodes()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.parse_populations"><code class="docutils literal notranslate"><span class="pre">parse_populations()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.parse_migrations"><code class="docutils literal notranslate"><span class="pre">parse_migrations()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.parse_sites"><code class="docutils literal notranslate"><span class="pre">parse_sites()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.random_nucleotides"><code class="docutils literal notranslate"><span class="pre">random_nucleotides()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.register_metadata_codec"><code class="docutils literal notranslate"><span class="pre">register_metadata_codec()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.validate_provenance"><code class="docutils literal notranslate"><span class="pre">validate_provenance()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.unpack_bytes"><code class="docutils literal notranslate"><span class="pre">unpack_bytes()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.unpack_strings"><code class="docutils literal notranslate"><span class="pre">unpack_strings()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tree-and-tree-sequence-classes">Tree and tree sequence classes</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-tree-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree"><code class="docutils literal notranslate"><span class="pre">Tree</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.copy"><code class="docutils literal notranslate"><span class="pre">Tree.copy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.tree_sequence"><code class="docutils literal notranslate"><span class="pre">Tree.tree_sequence</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.root_threshold"><code class="docutils literal notranslate"><span class="pre">Tree.root_threshold</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.first"><code class="docutils literal notranslate"><span class="pre">Tree.first()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.last"><code class="docutils literal notranslate"><span class="pre">Tree.last()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.next"><code class="docutils literal notranslate"><span class="pre">Tree.next()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.prev"><code class="docutils literal notranslate"><span class="pre">Tree.prev()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.clear"><code class="docutils literal notranslate"><span class="pre">Tree.clear()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.seek_index"><code class="docutils literal notranslate"><span class="pre">Tree.seek_index()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.seek"><code class="docutils literal notranslate"><span class="pre">Tree.seek()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.rank"><code class="docutils literal notranslate"><span class="pre">Tree.rank()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.unrank"><code class="docutils literal notranslate"><span class="pre">Tree.unrank()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.count_topologies"><code class="docutils literal notranslate"><span class="pre">Tree.count_topologies()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.branch_length"><code class="docutils literal notranslate"><span class="pre">Tree.branch_length()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.total_branch_length"><code class="docutils literal notranslate"><span class="pre">Tree.total_branch_length</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.mrca"><code class="docutils literal notranslate"><span class="pre">Tree.mrca()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.tmrca"><code class="docutils literal notranslate"><span class="pre">Tree.tmrca()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.parent"><code class="docutils literal notranslate"><span class="pre">Tree.parent()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.parent_array"><code class="docutils literal notranslate"><span class="pre">Tree.parent_array</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.ancestors"><code class="docutils literal notranslate"><span class="pre">Tree.ancestors()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.left_child"><code class="docutils literal notranslate"><span class="pre">Tree.left_child()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.left_child_array"><code class="docutils literal notranslate"><span class="pre">Tree.left_child_array</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.right_child"><code class="docutils literal notranslate"><span class="pre">Tree.right_child()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.right_child_array"><code class="docutils literal notranslate"><span class="pre">Tree.right_child_array</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.left_sib"><code class="docutils literal notranslate"><span class="pre">Tree.left_sib()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.left_sib_array"><code class="docutils literal notranslate"><span class="pre">Tree.left_sib_array</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.right_sib"><code class="docutils literal notranslate"><span class="pre">Tree.right_sib()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.right_sib_array"><code class="docutils literal notranslate"><span class="pre">Tree.right_sib_array</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.siblings"><code class="docutils literal notranslate"><span class="pre">Tree.siblings()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.num_children_array"><code class="docutils literal notranslate"><span class="pre">Tree.num_children_array</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.edge"><code class="docutils literal notranslate"><span class="pre">Tree.edge()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.edge_array"><code class="docutils literal notranslate"><span class="pre">Tree.edge_array</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.virtual_root"><code class="docutils literal notranslate"><span class="pre">Tree.virtual_root</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.num_edges"><code class="docutils literal notranslate"><span class="pre">Tree.num_edges</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.left_root"><code class="docutils literal notranslate"><span class="pre">Tree.left_root</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.children"><code class="docutils literal notranslate"><span class="pre">Tree.children()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.time"><code class="docutils literal notranslate"><span class="pre">Tree.time()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.depth"><code class="docutils literal notranslate"><span class="pre">Tree.depth()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.population"><code class="docutils literal notranslate"><span class="pre">Tree.population()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.is_internal"><code class="docutils literal notranslate"><span class="pre">Tree.is_internal()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.is_leaf"><code class="docutils literal notranslate"><span class="pre">Tree.is_leaf()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.is_isolated"><code class="docutils literal notranslate"><span class="pre">Tree.is_isolated()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.is_sample"><code class="docutils literal notranslate"><span class="pre">Tree.is_sample()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.is_descendant"><code class="docutils literal notranslate"><span class="pre">Tree.is_descendant()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.num_nodes"><code class="docutils literal notranslate"><span class="pre">Tree.num_nodes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.num_roots"><code class="docutils literal notranslate"><span class="pre">Tree.num_roots</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.has_single_root"><code class="docutils literal notranslate"><span class="pre">Tree.has_single_root</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.has_multiple_roots"><code class="docutils literal notranslate"><span class="pre">Tree.has_multiple_roots</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.roots"><code class="docutils literal notranslate"><span class="pre">Tree.roots</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.root"><code class="docutils literal notranslate"><span class="pre">Tree.root</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.is_root"><code class="docutils literal notranslate"><span class="pre">Tree.is_root()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.index"><code class="docutils literal notranslate"><span class="pre">Tree.index</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.interval"><code class="docutils literal notranslate"><span class="pre">Tree.interval</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.span"><code class="docutils literal notranslate"><span class="pre">Tree.span</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.mid"><code class="docutils literal notranslate"><span class="pre">Tree.mid</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.draw_text"><code class="docutils literal notranslate"><span class="pre">Tree.draw_text()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.draw_svg"><code class="docutils literal notranslate"><span class="pre">Tree.draw_svg()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.draw"><code class="docutils literal notranslate"><span class="pre">Tree.draw()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.num_mutations"><code class="docutils literal notranslate"><span class="pre">Tree.num_mutations</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.num_sites"><code class="docutils literal notranslate"><span class="pre">Tree.num_sites</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.sites"><code class="docutils literal notranslate"><span class="pre">Tree.sites()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.mutations"><code class="docutils literal notranslate"><span class="pre">Tree.mutations()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.leaves"><code class="docutils literal notranslate"><span class="pre">Tree.leaves()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.samples"><code class="docutils literal notranslate"><span class="pre">Tree.samples()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.num_children"><code class="docutils literal notranslate"><span class="pre">Tree.num_children()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.num_samples"><code class="docutils literal notranslate"><span class="pre">Tree.num_samples()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.num_tracked_samples"><code class="docutils literal notranslate"><span class="pre">Tree.num_tracked_samples()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.preorder"><code class="docutils literal notranslate"><span class="pre">Tree.preorder()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.postorder"><code class="docutils literal notranslate"><span class="pre">Tree.postorder()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.timeasc"><code class="docutils literal notranslate"><span class="pre">Tree.timeasc()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.timedesc"><code class="docutils literal notranslate"><span class="pre">Tree.timedesc()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.nodes"><code class="docutils literal notranslate"><span class="pre">Tree.nodes()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.as_newick"><code class="docutils literal notranslate"><span class="pre">Tree.as_newick()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.newick"><code class="docutils literal notranslate"><span class="pre">Tree.newick()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.as_dict_of_dicts"><code class="docutils literal notranslate"><span class="pre">Tree.as_dict_of_dicts()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.__str__"><code class="docutils literal notranslate"><span class="pre">Tree.__str__()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree._repr_html_"><code class="docutils literal notranslate"><span class="pre">Tree._repr_html_()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.map_mutations"><code class="docutils literal notranslate"><span class="pre">Tree.map_mutations()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.kc_distance"><code class="docutils literal notranslate"><span class="pre">Tree.kc_distance()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.rf_distance"><code class="docutils literal notranslate"><span class="pre">Tree.rf_distance()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.path_length"><code class="docutils literal notranslate"><span class="pre">Tree.path_length()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.distance_between"><code class="docutils literal notranslate"><span class="pre">Tree.distance_between()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.b1_index"><code class="docutils literal notranslate"><span class="pre">Tree.b1_index()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.b2_index"><code class="docutils literal notranslate"><span class="pre">Tree.b2_index()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.colless_index"><code class="docutils literal notranslate"><span class="pre">Tree.colless_index()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.sackin_index"><code class="docutils literal notranslate"><span class="pre">Tree.sackin_index()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.num_lineages"><code class="docutils literal notranslate"><span class="pre">Tree.num_lineages()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.split_polytomies"><code class="docutils literal notranslate"><span class="pre">Tree.split_polytomies()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.generate_star"><code class="docutils literal notranslate"><span class="pre">Tree.generate_star()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.generate_balanced"><code class="docutils literal notranslate"><span class="pre">Tree.generate_balanced()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.generate_comb"><code class="docutils literal notranslate"><span class="pre">Tree.generate_comb()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.generate_random_binary"><code class="docutils literal notranslate"><span class="pre">Tree.generate_random_binary()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-treesequence-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence"><code class="docutils literal notranslate"><span class="pre">TreeSequence</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.equals"><code class="docutils literal notranslate"><span class="pre">TreeSequence.equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.aslist"><code class="docutils literal notranslate"><span class="pre">TreeSequence.aslist()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.dump"><code class="docutils literal notranslate"><span class="pre">TreeSequence.dump()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.reference_sequence"><code class="docutils literal notranslate"><span class="pre">TreeSequence.reference_sequence</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.has_reference_sequence"><code class="docutils literal notranslate"><span class="pre">TreeSequence.has_reference_sequence()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.tables_dict"><code class="docutils literal notranslate"><span class="pre">TreeSequence.tables_dict</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.tables"><code class="docutils literal notranslate"><span class="pre">TreeSequence.tables</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.nbytes"><code class="docutils literal notranslate"><span class="pre">TreeSequence.nbytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.dump_tables"><code class="docutils literal notranslate"><span class="pre">TreeSequence.dump_tables()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.link_ancestors"><code class="docutils literal notranslate"><span class="pre">TreeSequence.link_ancestors()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.dump_text"><code class="docutils literal notranslate"><span class="pre">TreeSequence.dump_text()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.__str__"><code class="docutils literal notranslate"><span class="pre">TreeSequence.__str__()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence._repr_html_"><code class="docutils literal notranslate"><span class="pre">TreeSequence._repr_html_()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.num_samples"><code class="docutils literal notranslate"><span class="pre">TreeSequence.num_samples</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.table_metadata_schemas"><code class="docutils literal notranslate"><span class="pre">TreeSequence.table_metadata_schemas</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.discrete_genome"><code class="docutils literal notranslate"><span class="pre">TreeSequence.discrete_genome</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.discrete_time"><code class="docutils literal notranslate"><span class="pre">TreeSequence.discrete_time</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.min_time"><code class="docutils literal notranslate"><span class="pre">TreeSequence.min_time</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.max_time"><code class="docutils literal notranslate"><span class="pre">TreeSequence.max_time</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.sequence_length"><code class="docutils literal notranslate"><span class="pre">TreeSequence.sequence_length</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.metadata"><code class="docutils literal notranslate"><span class="pre">TreeSequence.metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.metadata_schema"><code class="docutils literal notranslate"><span class="pre">TreeSequence.metadata_schema</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.time_units"><code class="docutils literal notranslate"><span class="pre">TreeSequence.time_units</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.num_edges"><code class="docutils literal notranslate"><span class="pre">TreeSequence.num_edges</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.num_trees"><code class="docutils literal notranslate"><span class="pre">TreeSequence.num_trees</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.num_sites"><code class="docutils literal notranslate"><span class="pre">TreeSequence.num_sites</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.num_mutations"><code class="docutils literal notranslate"><span class="pre">TreeSequence.num_mutations</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.num_individuals"><code class="docutils literal notranslate"><span class="pre">TreeSequence.num_individuals</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.num_nodes"><code class="docutils literal notranslate"><span class="pre">TreeSequence.num_nodes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.num_provenances"><code class="docutils literal notranslate"><span class="pre">TreeSequence.num_provenances</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.num_populations"><code class="docutils literal notranslate"><span class="pre">TreeSequence.num_populations</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.num_migrations"><code class="docutils literal notranslate"><span class="pre">TreeSequence.num_migrations</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.max_root_time"><code class="docutils literal notranslate"><span class="pre">TreeSequence.max_root_time</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.migrations"><code class="docutils literal notranslate"><span class="pre">TreeSequence.migrations()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.individuals"><code class="docutils literal notranslate"><span class="pre">TreeSequence.individuals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.nodes"><code class="docutils literal notranslate"><span class="pre">TreeSequence.nodes()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.edges"><code class="docutils literal notranslate"><span class="pre">TreeSequence.edges()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.edge_diffs"><code class="docutils literal notranslate"><span class="pre">TreeSequence.edge_diffs()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.sites"><code class="docutils literal notranslate"><span class="pre">TreeSequence.sites()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.mutations"><code class="docutils literal notranslate"><span class="pre">TreeSequence.mutations()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.populations"><code class="docutils literal notranslate"><span class="pre">TreeSequence.populations()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.provenances"><code class="docutils literal notranslate"><span class="pre">TreeSequence.provenances()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.breakpoints"><code class="docutils literal notranslate"><span class="pre">TreeSequence.breakpoints()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.at"><code class="docutils literal notranslate"><span class="pre">TreeSequence.at()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.at_index"><code class="docutils literal notranslate"><span class="pre">TreeSequence.at_index()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.first"><code class="docutils literal notranslate"><span class="pre">TreeSequence.first()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.last"><code class="docutils literal notranslate"><span class="pre">TreeSequence.last()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.trees"><code class="docutils literal notranslate"><span class="pre">TreeSequence.trees()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.coiterate"><code class="docutils literal notranslate"><span class="pre">TreeSequence.coiterate()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.haplotypes"><code class="docutils literal notranslate"><span class="pre">TreeSequence.haplotypes()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.variants"><code class="docutils literal notranslate"><span class="pre">TreeSequence.variants()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.genotype_matrix"><code class="docutils literal notranslate"><span class="pre">TreeSequence.genotype_matrix()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.alignments"><code class="docutils literal notranslate"><span class="pre">TreeSequence.alignments()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.individuals_population"><code class="docutils literal notranslate"><span class="pre">TreeSequence.individuals_population</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.individuals_time"><code class="docutils literal notranslate"><span class="pre">TreeSequence.individuals_time</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.individuals_location"><code class="docutils literal notranslate"><span class="pre">TreeSequence.individuals_location</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.individuals_flags"><code class="docutils literal notranslate"><span class="pre">TreeSequence.individuals_flags</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.individuals_metadata"><code class="docutils literal notranslate"><span class="pre">TreeSequence.individuals_metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.individuals_nodes"><code class="docutils literal notranslate"><span class="pre">TreeSequence.individuals_nodes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.nodes_metadata"><code class="docutils literal notranslate"><span class="pre">TreeSequence.nodes_metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.nodes_time"><code class="docutils literal notranslate"><span class="pre">TreeSequence.nodes_time</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.nodes_flags"><code class="docutils literal notranslate"><span class="pre">TreeSequence.nodes_flags</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.nodes_population"><code class="docutils literal notranslate"><span class="pre">TreeSequence.nodes_population</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.nodes_individual"><code class="docutils literal notranslate"><span class="pre">TreeSequence.nodes_individual</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.edges_left"><code class="docutils literal notranslate"><span class="pre">TreeSequence.edges_left</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.edges_right"><code class="docutils literal notranslate"><span class="pre">TreeSequence.edges_right</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.edges_parent"><code class="docutils literal notranslate"><span class="pre">TreeSequence.edges_parent</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.edges_child"><code class="docutils literal notranslate"><span class="pre">TreeSequence.edges_child</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.edges_metadata"><code class="docutils literal notranslate"><span class="pre">TreeSequence.edges_metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.sites_position"><code class="docutils literal notranslate"><span class="pre">TreeSequence.sites_position</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.sites_ancestral_state"><code class="docutils literal notranslate"><span class="pre">TreeSequence.sites_ancestral_state</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.sites_metadata"><code class="docutils literal notranslate"><span class="pre">TreeSequence.sites_metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.mutations_site"><code class="docutils literal notranslate"><span class="pre">TreeSequence.mutations_site</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.mutations_node"><code class="docutils literal notranslate"><span class="pre">TreeSequence.mutations_node</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.mutations_parent"><code class="docutils literal notranslate"><span class="pre">TreeSequence.mutations_parent</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.mutations_time"><code class="docutils literal notranslate"><span class="pre">TreeSequence.mutations_time</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.mutations_derived_state"><code class="docutils literal notranslate"><span class="pre">TreeSequence.mutations_derived_state</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.mutations_metadata"><code class="docutils literal notranslate"><span class="pre">TreeSequence.mutations_metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.mutations_edge"><code class="docutils literal notranslate"><span class="pre">TreeSequence.mutations_edge</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.mutations_inherited_state"><code class="docutils literal notranslate"><span class="pre">TreeSequence.mutations_inherited_state</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.migrations_left"><code class="docutils literal notranslate"><span class="pre">TreeSequence.migrations_left</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.migrations_right"><code class="docutils literal notranslate"><span class="pre">TreeSequence.migrations_right</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.migrations_node"><code class="docutils literal notranslate"><span class="pre">TreeSequence.migrations_node</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.migrations_source"><code class="docutils literal notranslate"><span class="pre">TreeSequence.migrations_source</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.migrations_dest"><code class="docutils literal notranslate"><span class="pre">TreeSequence.migrations_dest</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.migrations_time"><code class="docutils literal notranslate"><span class="pre">TreeSequence.migrations_time</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.migrations_metadata"><code class="docutils literal notranslate"><span class="pre">TreeSequence.migrations_metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.populations_metadata"><code class="docutils literal notranslate"><span class="pre">TreeSequence.populations_metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.indexes_edge_insertion_order"><code class="docutils literal notranslate"><span class="pre">TreeSequence.indexes_edge_insertion_order</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.indexes_edge_removal_order"><code class="docutils literal notranslate"><span class="pre">TreeSequence.indexes_edge_removal_order</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.individual"><code class="docutils literal notranslate"><span class="pre">TreeSequence.individual()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.node"><code class="docutils literal notranslate"><span class="pre">TreeSequence.node()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.edge"><code class="docutils literal notranslate"><span class="pre">TreeSequence.edge()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.migration"><code class="docutils literal notranslate"><span class="pre">TreeSequence.migration()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.mutation"><code class="docutils literal notranslate"><span class="pre">TreeSequence.mutation()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.site"><code class="docutils literal notranslate"><span class="pre">TreeSequence.site()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.population"><code class="docutils literal notranslate"><span class="pre">TreeSequence.population()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.provenance"><code class="docutils literal notranslate"><span class="pre">TreeSequence.provenance()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.samples"><code class="docutils literal notranslate"><span class="pre">TreeSequence.samples()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.as_vcf"><code class="docutils literal notranslate"><span class="pre">TreeSequence.as_vcf()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.write_vcf"><code class="docutils literal notranslate"><span class="pre">TreeSequence.write_vcf()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.write_fasta"><code class="docutils literal notranslate"><span class="pre">TreeSequence.write_fasta()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.as_fasta"><code class="docutils literal notranslate"><span class="pre">TreeSequence.as_fasta()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.write_nexus"><code class="docutils literal notranslate"><span class="pre">TreeSequence.write_nexus()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.as_nexus"><code class="docutils literal notranslate"><span class="pre">TreeSequence.as_nexus()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.to_macs"><code class="docutils literal notranslate"><span class="pre">TreeSequence.to_macs()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.simplify"><code class="docutils literal notranslate"><span class="pre">TreeSequence.simplify()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.delete_sites"><code class="docutils literal notranslate"><span class="pre">TreeSequence.delete_sites()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.delete_intervals"><code class="docutils literal notranslate"><span class="pre">TreeSequence.delete_intervals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.keep_intervals"><code class="docutils literal notranslate"><span class="pre">TreeSequence.keep_intervals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.ltrim"><code class="docutils literal notranslate"><span class="pre">TreeSequence.ltrim()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.rtrim"><code class="docutils literal notranslate"><span class="pre">TreeSequence.rtrim()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.trim"><code class="docutils literal notranslate"><span class="pre">TreeSequence.trim()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.shift"><code class="docutils literal notranslate"><span class="pre">TreeSequence.shift()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.concatenate"><code class="docutils literal notranslate"><span class="pre">TreeSequence.concatenate()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.split_edges"><code class="docutils literal notranslate"><span class="pre">TreeSequence.split_edges()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.decapitate"><code class="docutils literal notranslate"><span class="pre">TreeSequence.decapitate()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.extend_haplotypes"><code class="docutils literal notranslate"><span class="pre">TreeSequence.extend_haplotypes()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.subset"><code class="docutils literal notranslate"><span class="pre">TreeSequence.subset()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.union"><code class="docutils literal notranslate"><span class="pre">TreeSequence.union()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.draw_svg"><code class="docutils literal notranslate"><span class="pre">TreeSequence.draw_svg()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.draw_text"><code class="docutils literal notranslate"><span class="pre">TreeSequence.draw_text()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.general_stat"><code class="docutils literal notranslate"><span class="pre">TreeSequence.general_stat()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.sample_count_stat"><code class="docutils literal notranslate"><span class="pre">TreeSequence.sample_count_stat()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.diversity"><code class="docutils literal notranslate"><span class="pre">TreeSequence.diversity()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.divergence"><code class="docutils literal notranslate"><span class="pre">TreeSequence.divergence()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.divergence_matrix"><code class="docutils literal notranslate"><span class="pre">TreeSequence.divergence_matrix()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.genetic_relatedness"><code class="docutils literal notranslate"><span class="pre">TreeSequence.genetic_relatedness()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.genetic_relatedness_matrix"><code class="docutils literal notranslate"><span class="pre">TreeSequence.genetic_relatedness_matrix()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.genetic_relatedness_weighted"><code class="docutils literal notranslate"><span class="pre">TreeSequence.genetic_relatedness_weighted()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.genetic_relatedness_vector"><code class="docutils literal notranslate"><span class="pre">TreeSequence.genetic_relatedness_vector()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.pca"><code class="docutils literal notranslate"><span class="pre">TreeSequence.pca()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.trait_covariance"><code class="docutils literal notranslate"><span class="pre">TreeSequence.trait_covariance()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.trait_correlation"><code class="docutils literal notranslate"><span class="pre">TreeSequence.trait_correlation()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.trait_regression"><code class="docutils literal notranslate"><span class="pre">TreeSequence.trait_regression()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.trait_linear_model"><code class="docutils literal notranslate"><span class="pre">TreeSequence.trait_linear_model()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.segregating_sites"><code class="docutils literal notranslate"><span class="pre">TreeSequence.segregating_sites()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.allele_frequency_spectrum"><code class="docutils literal notranslate"><span class="pre">TreeSequence.allele_frequency_spectrum()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.Tajimas_D"><code class="docutils literal notranslate"><span class="pre">TreeSequence.Tajimas_D()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.Fst"><code class="docutils literal notranslate"><span class="pre">TreeSequence.Fst()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.Y3"><code class="docutils literal notranslate"><span class="pre">TreeSequence.Y3()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.Y2"><code class="docutils literal notranslate"><span class="pre">TreeSequence.Y2()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.Y1"><code class="docutils literal notranslate"><span class="pre">TreeSequence.Y1()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.f4"><code class="docutils literal notranslate"><span class="pre">TreeSequence.f4()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.f3"><code class="docutils literal notranslate"><span class="pre">TreeSequence.f3()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.f2"><code class="docutils literal notranslate"><span class="pre">TreeSequence.f2()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.mean_descendants"><code class="docutils literal notranslate"><span class="pre">TreeSequence.mean_descendants()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.genealogical_nearest_neighbours"><code class="docutils literal notranslate"><span class="pre">TreeSequence.genealogical_nearest_neighbours()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.kc_distance"><code class="docutils literal notranslate"><span class="pre">TreeSequence.kc_distance()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.count_topologies"><code class="docutils literal notranslate"><span class="pre">TreeSequence.count_topologies()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.ibd_segments"><code class="docutils literal notranslate"><span class="pre">TreeSequence.ibd_segments()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.pair_coalescence_counts"><code class="docutils literal notranslate"><span class="pre">TreeSequence.pair_coalescence_counts()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.pair_coalescence_quantiles"><code class="docutils literal notranslate"><span class="pre">TreeSequence.pair_coalescence_quantiles()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.pair_coalescence_rates"><code class="docutils literal notranslate"><span class="pre">TreeSequence.pair_coalescence_rates()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.impute_unknown_mutations_time"><code class="docutils literal notranslate"><span class="pre">TreeSequence.impute_unknown_mutations_time()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.sample_nodes_by_ploidy"><code class="docutils literal notranslate"><span class="pre">TreeSequence.sample_nodes_by_ploidy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.map_to_vcf_model"><code class="docutils literal notranslate"><span class="pre">TreeSequence.map_to_vcf_model()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.pairwise_diversity"><code class="docutils literal notranslate"><span class="pre">TreeSequence.pairwise_diversity()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#simple-container-classes">Simple container classes</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-individual-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Individual</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Individual"><code class="docutils literal notranslate"><span class="pre">Individual</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Individual.id"><code class="docutils literal notranslate"><span class="pre">Individual.id</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Individual.flags"><code class="docutils literal notranslate"><span class="pre">Individual.flags</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Individual.location"><code class="docutils literal notranslate"><span class="pre">Individual.location</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Individual.parents"><code class="docutils literal notranslate"><span class="pre">Individual.parents</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Individual.nodes"><code class="docutils literal notranslate"><span class="pre">Individual.nodes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Individual.metadata"><code class="docutils literal notranslate"><span class="pre">Individual.metadata</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-node-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Node</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Node"><code class="docutils literal notranslate"><span class="pre">Node</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Node.id"><code class="docutils literal notranslate"><span class="pre">Node.id</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Node.flags"><code class="docutils literal notranslate"><span class="pre">Node.flags</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Node.time"><code class="docutils literal notranslate"><span class="pre">Node.time</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Node.population"><code class="docutils literal notranslate"><span class="pre">Node.population</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Node.individual"><code class="docutils literal notranslate"><span class="pre">Node.individual</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Node.metadata"><code class="docutils literal notranslate"><span class="pre">Node.metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Node.is_sample"><code class="docutils literal notranslate"><span class="pre">Node.is_sample()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-edge-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Edge"><code class="docutils literal notranslate"><span class="pre">Edge</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Edge.id"><code class="docutils literal notranslate"><span class="pre">Edge.id</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Edge.left"><code class="docutils literal notranslate"><span class="pre">Edge.left</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Edge.right"><code class="docutils literal notranslate"><span class="pre">Edge.right</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Edge.parent"><code class="docutils literal notranslate"><span class="pre">Edge.parent</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Edge.child"><code class="docutils literal notranslate"><span class="pre">Edge.child</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Edge.metadata"><code class="docutils literal notranslate"><span class="pre">Edge.metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Edge.span"><code class="docutils literal notranslate"><span class="pre">Edge.span</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Edge.interval"><code class="docutils literal notranslate"><span class="pre">Edge.interval</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-site-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Site</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Site"><code class="docutils literal notranslate"><span class="pre">Site</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Site.id"><code class="docutils literal notranslate"><span class="pre">Site.id</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Site.position"><code class="docutils literal notranslate"><span class="pre">Site.position</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Site.ancestral_state"><code class="docutils literal notranslate"><span class="pre">Site.ancestral_state</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Site.mutations"><code class="docutils literal notranslate"><span class="pre">Site.mutations</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Site.metadata"><code class="docutils literal notranslate"><span class="pre">Site.metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Site.alleles"><code class="docutils literal notranslate"><span class="pre">Site.alleles</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-mutation-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Mutation</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Mutation"><code class="docutils literal notranslate"><span class="pre">Mutation</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Mutation.id"><code class="docutils literal notranslate"><span class="pre">Mutation.id</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Mutation.site"><code class="docutils literal notranslate"><span class="pre">Mutation.site</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Mutation.node"><code class="docutils literal notranslate"><span class="pre">Mutation.node</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Mutation.time"><code class="docutils literal notranslate"><span class="pre">Mutation.time</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Mutation.derived_state"><code class="docutils literal notranslate"><span class="pre">Mutation.derived_state</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Mutation.parent"><code class="docutils literal notranslate"><span class="pre">Mutation.parent</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Mutation.metadata"><code class="docutils literal notranslate"><span class="pre">Mutation.metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Mutation.edge"><code class="docutils literal notranslate"><span class="pre">Mutation.edge</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Mutation.inherited_state"><code class="docutils literal notranslate"><span class="pre">Mutation.inherited_state</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-variant-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant"><code class="docutils literal notranslate"><span class="pre">Variant</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.site"><code class="docutils literal notranslate"><span class="pre">Variant.site</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.alleles"><code class="docutils literal notranslate"><span class="pre">Variant.alleles</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.samples"><code class="docutils literal notranslate"><span class="pre">Variant.samples</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.genotypes"><code class="docutils literal notranslate"><span class="pre">Variant.genotypes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.isolated_as_missing"><code class="docutils literal notranslate"><span class="pre">Variant.isolated_as_missing</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.has_missing_data"><code class="docutils literal notranslate"><span class="pre">Variant.has_missing_data</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.num_missing"><code class="docutils literal notranslate"><span class="pre">Variant.num_missing</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.num_alleles"><code class="docutils literal notranslate"><span class="pre">Variant.num_alleles</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.decode"><code class="docutils literal notranslate"><span class="pre">Variant.decode()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.copy"><code class="docutils literal notranslate"><span class="pre">Variant.copy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.states"><code class="docutils literal notranslate"><span class="pre">Variant.states()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.counts"><code class="docutils literal notranslate"><span class="pre">Variant.counts()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.frequencies"><code class="docutils literal notranslate"><span class="pre">Variant.frequencies()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-migration-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Migration</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Migration"><code class="docutils literal notranslate"><span class="pre">Migration</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Migration.left"><code class="docutils literal notranslate"><span class="pre">Migration.left</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Migration.right"><code class="docutils literal notranslate"><span class="pre">Migration.right</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Migration.node"><code class="docutils literal notranslate"><span class="pre">Migration.node</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Migration.source"><code class="docutils literal notranslate"><span class="pre">Migration.source</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Migration.dest"><code class="docutils literal notranslate"><span class="pre">Migration.dest</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Migration.time"><code class="docutils literal notranslate"><span class="pre">Migration.time</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Migration.metadata"><code class="docutils literal notranslate"><span class="pre">Migration.metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Migration.id"><code class="docutils literal notranslate"><span class="pre">Migration.id</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-population-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Population"><code class="docutils literal notranslate"><span class="pre">Population</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Population.id"><code class="docutils literal notranslate"><span class="pre">Population.id</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Population.metadata"><code class="docutils literal notranslate"><span class="pre">Population.metadata</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-provenance-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Provenance</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Provenance"><code class="docutils literal notranslate"><span class="pre">Provenance</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Provenance.timestamp"><code class="docutils literal notranslate"><span class="pre">Provenance.timestamp</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Provenance.record"><code class="docutils literal notranslate"><span class="pre">Provenance.record</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-interval-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Interval"><code class="docutils literal notranslate"><span class="pre">Interval</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Interval.left"><code class="docutils literal notranslate"><span class="pre">Interval.left</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Interval.right"><code class="docutils literal notranslate"><span class="pre">Interval.right</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Interval.span"><code class="docutils literal notranslate"><span class="pre">Interval.span</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Interval.mid"><code class="docutils literal notranslate"><span class="pre">Interval.mid</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-rank-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Rank</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Rank"><code class="docutils literal notranslate"><span class="pre">Rank</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Rank.shape"><code class="docutils literal notranslate"><span class="pre">Rank.shape</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Rank.label"><code class="docutils literal notranslate"><span class="pre">Rank.label</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tablecollection-and-table-classes">TableCollection and Table classes</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-tablecollection-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection"><code class="docutils literal notranslate"><span class="pre">TableCollection</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.individuals"><code class="docutils literal notranslate"><span class="pre">TableCollection.individuals</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.nodes"><code class="docutils literal notranslate"><span class="pre">TableCollection.nodes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.edges"><code class="docutils literal notranslate"><span class="pre">TableCollection.edges</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.migrations"><code class="docutils literal notranslate"><span class="pre">TableCollection.migrations</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.sites"><code class="docutils literal notranslate"><span class="pre">TableCollection.sites</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.mutations"><code class="docutils literal notranslate"><span class="pre">TableCollection.mutations</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.populations"><code class="docutils literal notranslate"><span class="pre">TableCollection.populations</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.provenances"><code class="docutils literal notranslate"><span class="pre">TableCollection.provenances</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.indexes"><code class="docutils literal notranslate"><span class="pre">TableCollection.indexes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.sequence_length"><code class="docutils literal notranslate"><span class="pre">TableCollection.sequence_length</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.file_uuid"><code class="docutils literal notranslate"><span class="pre">TableCollection.file_uuid</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.time_units"><code class="docutils literal notranslate"><span class="pre">TableCollection.time_units</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.has_reference_sequence"><code class="docutils literal notranslate"><span class="pre">TableCollection.has_reference_sequence()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.reference_sequence"><code class="docutils literal notranslate"><span class="pre">TableCollection.reference_sequence</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.asdict"><code class="docutils literal notranslate"><span class="pre">TableCollection.asdict()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.table_name_map"><code class="docutils literal notranslate"><span class="pre">TableCollection.table_name_map</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.nbytes"><code class="docutils literal notranslate"><span class="pre">TableCollection.nbytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.equals"><code class="docutils literal notranslate"><span class="pre">TableCollection.equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.assert_equals"><code class="docutils literal notranslate"><span class="pre">TableCollection.assert_equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.dump"><code class="docutils literal notranslate"><span class="pre">TableCollection.dump()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.copy"><code class="docutils literal notranslate"><span class="pre">TableCollection.copy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.tree_sequence"><code class="docutils literal notranslate"><span class="pre">TableCollection.tree_sequence()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.simplify"><code class="docutils literal notranslate"><span class="pre">TableCollection.simplify()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.link_ancestors"><code class="docutils literal notranslate"><span class="pre">TableCollection.link_ancestors()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.sort"><code class="docutils literal notranslate"><span class="pre">TableCollection.sort()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.sort_individuals"><code class="docutils literal notranslate"><span class="pre">TableCollection.sort_individuals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.canonicalise"><code class="docutils literal notranslate"><span class="pre">TableCollection.canonicalise()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.compute_mutation_parents"><code class="docutils literal notranslate"><span class="pre">TableCollection.compute_mutation_parents()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.compute_mutation_times"><code class="docutils literal notranslate"><span class="pre">TableCollection.compute_mutation_times()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.deduplicate_sites"><code class="docutils literal notranslate"><span class="pre">TableCollection.deduplicate_sites()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.delete_sites"><code class="docutils literal notranslate"><span class="pre">TableCollection.delete_sites()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.delete_intervals"><code class="docutils literal notranslate"><span class="pre">TableCollection.delete_intervals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.keep_intervals"><code class="docutils literal notranslate"><span class="pre">TableCollection.keep_intervals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.ltrim"><code class="docutils literal notranslate"><span class="pre">TableCollection.ltrim()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.rtrim"><code class="docutils literal notranslate"><span class="pre">TableCollection.rtrim()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.trim"><code class="docutils literal notranslate"><span class="pre">TableCollection.trim()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.shift"><code class="docutils literal notranslate"><span class="pre">TableCollection.shift()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.delete_older"><code class="docutils literal notranslate"><span class="pre">TableCollection.delete_older()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.clear"><code class="docutils literal notranslate"><span class="pre">TableCollection.clear()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.has_index"><code class="docutils literal notranslate"><span class="pre">TableCollection.has_index()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.build_index"><code class="docutils literal notranslate"><span class="pre">TableCollection.build_index()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.drop_index"><code class="docutils literal notranslate"><span class="pre">TableCollection.drop_index()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.subset"><code class="docutils literal notranslate"><span class="pre">TableCollection.subset()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.union"><code class="docutils literal notranslate"><span class="pre">TableCollection.union()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.ibd_segments"><code class="docutils literal notranslate"><span class="pre">TableCollection.ibd_segments()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.metadata"><code class="docutils literal notranslate"><span class="pre">TableCollection.metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.metadata_bytes"><code class="docutils literal notranslate"><span class="pre">TableCollection.metadata_bytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.metadata_schema"><code class="docutils literal notranslate"><span class="pre">TableCollection.metadata_schema</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#individualtable-classes"><code class="xref py py-class docutils literal notranslate"><span class="pre">IndividualTable</span></code> classes</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable"><code class="docutils literal notranslate"><span class="pre">IndividualTable</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.add_row"><code class="docutils literal notranslate"><span class="pre">IndividualTable.add_row()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.set_columns"><code class="docutils literal notranslate"><span class="pre">IndividualTable.set_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.append_columns"><code class="docutils literal notranslate"><span class="pre">IndividualTable.append_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.packset_location"><code class="docutils literal notranslate"><span class="pre">IndividualTable.packset_location()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.packset_parents"><code class="docutils literal notranslate"><span class="pre">IndividualTable.packset_parents()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.keep_rows"><code class="docutils literal notranslate"><span class="pre">IndividualTable.keep_rows()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.__getitem__"><code class="docutils literal notranslate"><span class="pre">IndividualTable.__getitem__()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.append"><code class="docutils literal notranslate"><span class="pre">IndividualTable.append()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.asdict"><code class="docutils literal notranslate"><span class="pre">IndividualTable.asdict()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.assert_equals"><code class="docutils literal notranslate"><span class="pre">IndividualTable.assert_equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.clear"><code class="docutils literal notranslate"><span class="pre">IndividualTable.clear()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.copy"><code class="docutils literal notranslate"><span class="pre">IndividualTable.copy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.drop_metadata"><code class="docutils literal notranslate"><span class="pre">IndividualTable.drop_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.equals"><code class="docutils literal notranslate"><span class="pre">IndividualTable.equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.metadata_schema"><code class="docutils literal notranslate"><span class="pre">IndividualTable.metadata_schema</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.metadata_vector"><code class="docutils literal notranslate"><span class="pre">IndividualTable.metadata_vector()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.nbytes"><code class="docutils literal notranslate"><span class="pre">IndividualTable.nbytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.packset_metadata"><code class="docutils literal notranslate"><span class="pre">IndividualTable.packset_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.truncate"><code class="docutils literal notranslate"><span class="pre">IndividualTable.truncate()</span></code></a></li>
</ul>
</li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#associated-row-class">Associated row class</a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTableRow"><code class="docutils literal notranslate"><span class="pre">IndividualTableRow</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTableRow.flags"><code class="docutils literal notranslate"><span class="pre">IndividualTableRow.flags</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTableRow.location"><code class="docutils literal notranslate"><span class="pre">IndividualTableRow.location</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTableRow.parents"><code class="docutils literal notranslate"><span class="pre">IndividualTableRow.parents</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTableRow.metadata"><code class="docutils literal notranslate"><span class="pre">IndividualTableRow.metadata</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTableRow.asdict"><code class="docutils literal notranslate"><span class="pre">IndividualTableRow.asdict()</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTableRow.replace"><code class="docutils literal notranslate"><span class="pre">IndividualTableRow.replace()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#nodetable-classes"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code> classes</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable"><code class="docutils literal notranslate"><span class="pre">NodeTable</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.add_row"><code class="docutils literal notranslate"><span class="pre">NodeTable.add_row()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.set_columns"><code class="docutils literal notranslate"><span class="pre">NodeTable.set_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.append_columns"><code class="docutils literal notranslate"><span class="pre">NodeTable.append_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.__getitem__"><code class="docutils literal notranslate"><span class="pre">NodeTable.__getitem__()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.append"><code class="docutils literal notranslate"><span class="pre">NodeTable.append()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.asdict"><code class="docutils literal notranslate"><span class="pre">NodeTable.asdict()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.assert_equals"><code class="docutils literal notranslate"><span class="pre">NodeTable.assert_equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.clear"><code class="docutils literal notranslate"><span class="pre">NodeTable.clear()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.copy"><code class="docutils literal notranslate"><span class="pre">NodeTable.copy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.drop_metadata"><code class="docutils literal notranslate"><span class="pre">NodeTable.drop_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.equals"><code class="docutils literal notranslate"><span class="pre">NodeTable.equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.keep_rows"><code class="docutils literal notranslate"><span class="pre">NodeTable.keep_rows()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.metadata_schema"><code class="docutils literal notranslate"><span class="pre">NodeTable.metadata_schema</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.metadata_vector"><code class="docutils literal notranslate"><span class="pre">NodeTable.metadata_vector()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.nbytes"><code class="docutils literal notranslate"><span class="pre">NodeTable.nbytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.packset_metadata"><code class="docutils literal notranslate"><span class="pre">NodeTable.packset_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.truncate"><code class="docutils literal notranslate"><span class="pre">NodeTable.truncate()</span></code></a></li>
</ul>
</li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#id10">Associated row class</a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTableRow"><code class="docutils literal notranslate"><span class="pre">NodeTableRow</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTableRow.flags"><code class="docutils literal notranslate"><span class="pre">NodeTableRow.flags</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTableRow.time"><code class="docutils literal notranslate"><span class="pre">NodeTableRow.time</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTableRow.population"><code class="docutils literal notranslate"><span class="pre">NodeTableRow.population</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTableRow.individual"><code class="docutils literal notranslate"><span class="pre">NodeTableRow.individual</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTableRow.metadata"><code class="docutils literal notranslate"><span class="pre">NodeTableRow.metadata</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTableRow.asdict"><code class="docutils literal notranslate"><span class="pre">NodeTableRow.asdict()</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTableRow.replace"><code class="docutils literal notranslate"><span class="pre">NodeTableRow.replace()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#edgetable-classes"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code> classes</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable"><code class="docutils literal notranslate"><span class="pre">EdgeTable</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.add_row"><code class="docutils literal notranslate"><span class="pre">EdgeTable.add_row()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.set_columns"><code class="docutils literal notranslate"><span class="pre">EdgeTable.set_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.append_columns"><code class="docutils literal notranslate"><span class="pre">EdgeTable.append_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.squash"><code class="docutils literal notranslate"><span class="pre">EdgeTable.squash()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.__getitem__"><code class="docutils literal notranslate"><span class="pre">EdgeTable.__getitem__()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.append"><code class="docutils literal notranslate"><span class="pre">EdgeTable.append()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.asdict"><code class="docutils literal notranslate"><span class="pre">EdgeTable.asdict()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.assert_equals"><code class="docutils literal notranslate"><span class="pre">EdgeTable.assert_equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.clear"><code class="docutils literal notranslate"><span class="pre">EdgeTable.clear()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.copy"><code class="docutils literal notranslate"><span class="pre">EdgeTable.copy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.drop_metadata"><code class="docutils literal notranslate"><span class="pre">EdgeTable.drop_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.equals"><code class="docutils literal notranslate"><span class="pre">EdgeTable.equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.keep_rows"><code class="docutils literal notranslate"><span class="pre">EdgeTable.keep_rows()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.metadata_schema"><code class="docutils literal notranslate"><span class="pre">EdgeTable.metadata_schema</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.metadata_vector"><code class="docutils literal notranslate"><span class="pre">EdgeTable.metadata_vector()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.nbytes"><code class="docutils literal notranslate"><span class="pre">EdgeTable.nbytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.packset_metadata"><code class="docutils literal notranslate"><span class="pre">EdgeTable.packset_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.truncate"><code class="docutils literal notranslate"><span class="pre">EdgeTable.truncate()</span></code></a></li>
</ul>
</li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#id11">Associated row class</a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTableRow"><code class="docutils literal notranslate"><span class="pre">EdgeTableRow</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTableRow.left"><code class="docutils literal notranslate"><span class="pre">EdgeTableRow.left</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTableRow.right"><code class="docutils literal notranslate"><span class="pre">EdgeTableRow.right</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTableRow.parent"><code class="docutils literal notranslate"><span class="pre">EdgeTableRow.parent</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTableRow.child"><code class="docutils literal notranslate"><span class="pre">EdgeTableRow.child</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTableRow.metadata"><code class="docutils literal notranslate"><span class="pre">EdgeTableRow.metadata</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTableRow.asdict"><code class="docutils literal notranslate"><span class="pre">EdgeTableRow.asdict()</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTableRow.replace"><code class="docutils literal notranslate"><span class="pre">EdgeTableRow.replace()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#migrationtable-classes"><code class="xref py py-class docutils literal notranslate"><span class="pre">MigrationTable</span></code> classes</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable"><code class="docutils literal notranslate"><span class="pre">MigrationTable</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.add_row"><code class="docutils literal notranslate"><span class="pre">MigrationTable.add_row()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.set_columns"><code class="docutils literal notranslate"><span class="pre">MigrationTable.set_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.append_columns"><code class="docutils literal notranslate"><span class="pre">MigrationTable.append_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.__getitem__"><code class="docutils literal notranslate"><span class="pre">MigrationTable.__getitem__()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.append"><code class="docutils literal notranslate"><span class="pre">MigrationTable.append()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.asdict"><code class="docutils literal notranslate"><span class="pre">MigrationTable.asdict()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.assert_equals"><code class="docutils literal notranslate"><span class="pre">MigrationTable.assert_equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.clear"><code class="docutils literal notranslate"><span class="pre">MigrationTable.clear()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.copy"><code class="docutils literal notranslate"><span class="pre">MigrationTable.copy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.drop_metadata"><code class="docutils literal notranslate"><span class="pre">MigrationTable.drop_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.equals"><code class="docutils literal notranslate"><span class="pre">MigrationTable.equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.keep_rows"><code class="docutils literal notranslate"><span class="pre">MigrationTable.keep_rows()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.metadata_schema"><code class="docutils literal notranslate"><span class="pre">MigrationTable.metadata_schema</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.metadata_vector"><code class="docutils literal notranslate"><span class="pre">MigrationTable.metadata_vector()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.nbytes"><code class="docutils literal notranslate"><span class="pre">MigrationTable.nbytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.packset_metadata"><code class="docutils literal notranslate"><span class="pre">MigrationTable.packset_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.truncate"><code class="docutils literal notranslate"><span class="pre">MigrationTable.truncate()</span></code></a></li>
</ul>
</li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#id12">Associated row class</a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTableRow"><code class="docutils literal notranslate"><span class="pre">MigrationTableRow</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTableRow.left"><code class="docutils literal notranslate"><span class="pre">MigrationTableRow.left</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTableRow.right"><code class="docutils literal notranslate"><span class="pre">MigrationTableRow.right</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTableRow.node"><code class="docutils literal notranslate"><span class="pre">MigrationTableRow.node</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTableRow.source"><code class="docutils literal notranslate"><span class="pre">MigrationTableRow.source</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTableRow.dest"><code class="docutils literal notranslate"><span class="pre">MigrationTableRow.dest</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTableRow.time"><code class="docutils literal notranslate"><span class="pre">MigrationTableRow.time</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTableRow.metadata"><code class="docutils literal notranslate"><span class="pre">MigrationTableRow.metadata</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTableRow.asdict"><code class="docutils literal notranslate"><span class="pre">MigrationTableRow.asdict()</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTableRow.replace"><code class="docutils literal notranslate"><span class="pre">MigrationTableRow.replace()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#sitetable-classes"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code> classes</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable"><code class="docutils literal notranslate"><span class="pre">SiteTable</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.add_row"><code class="docutils literal notranslate"><span class="pre">SiteTable.add_row()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.set_columns"><code class="docutils literal notranslate"><span class="pre">SiteTable.set_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.append_columns"><code class="docutils literal notranslate"><span class="pre">SiteTable.append_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.packset_ancestral_state"><code class="docutils literal notranslate"><span class="pre">SiteTable.packset_ancestral_state()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.__getitem__"><code class="docutils literal notranslate"><span class="pre">SiteTable.__getitem__()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.append"><code class="docutils literal notranslate"><span class="pre">SiteTable.append()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.asdict"><code class="docutils literal notranslate"><span class="pre">SiteTable.asdict()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.assert_equals"><code class="docutils literal notranslate"><span class="pre">SiteTable.assert_equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.clear"><code class="docutils literal notranslate"><span class="pre">SiteTable.clear()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.copy"><code class="docutils literal notranslate"><span class="pre">SiteTable.copy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.drop_metadata"><code class="docutils literal notranslate"><span class="pre">SiteTable.drop_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.equals"><code class="docutils literal notranslate"><span class="pre">SiteTable.equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.keep_rows"><code class="docutils literal notranslate"><span class="pre">SiteTable.keep_rows()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.metadata_schema"><code class="docutils literal notranslate"><span class="pre">SiteTable.metadata_schema</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.metadata_vector"><code class="docutils literal notranslate"><span class="pre">SiteTable.metadata_vector()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.nbytes"><code class="docutils literal notranslate"><span class="pre">SiteTable.nbytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.packset_metadata"><code class="docutils literal notranslate"><span class="pre">SiteTable.packset_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.truncate"><code class="docutils literal notranslate"><span class="pre">SiteTable.truncate()</span></code></a></li>
</ul>
</li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#id13">Associated row class</a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTableRow"><code class="docutils literal notranslate"><span class="pre">SiteTableRow</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTableRow.position"><code class="docutils literal notranslate"><span class="pre">SiteTableRow.position</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTableRow.ancestral_state"><code class="docutils literal notranslate"><span class="pre">SiteTableRow.ancestral_state</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTableRow.metadata"><code class="docutils literal notranslate"><span class="pre">SiteTableRow.metadata</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTableRow.asdict"><code class="docutils literal notranslate"><span class="pre">SiteTableRow.asdict()</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTableRow.replace"><code class="docutils literal notranslate"><span class="pre">SiteTableRow.replace()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mutationtable-classes"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutationTable</span></code> classes</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable"><code class="docutils literal notranslate"><span class="pre">MutationTable</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.add_row"><code class="docutils literal notranslate"><span class="pre">MutationTable.add_row()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.set_columns"><code class="docutils literal notranslate"><span class="pre">MutationTable.set_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.append_columns"><code class="docutils literal notranslate"><span class="pre">MutationTable.append_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.packset_derived_state"><code class="docutils literal notranslate"><span class="pre">MutationTable.packset_derived_state()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.keep_rows"><code class="docutils literal notranslate"><span class="pre">MutationTable.keep_rows()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.__getitem__"><code class="docutils literal notranslate"><span class="pre">MutationTable.__getitem__()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.append"><code class="docutils literal notranslate"><span class="pre">MutationTable.append()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.asdict"><code class="docutils literal notranslate"><span class="pre">MutationTable.asdict()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.assert_equals"><code class="docutils literal notranslate"><span class="pre">MutationTable.assert_equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.clear"><code class="docutils literal notranslate"><span class="pre">MutationTable.clear()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.copy"><code class="docutils literal notranslate"><span class="pre">MutationTable.copy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.drop_metadata"><code class="docutils literal notranslate"><span class="pre">MutationTable.drop_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.equals"><code class="docutils literal notranslate"><span class="pre">MutationTable.equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.metadata_schema"><code class="docutils literal notranslate"><span class="pre">MutationTable.metadata_schema</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.metadata_vector"><code class="docutils literal notranslate"><span class="pre">MutationTable.metadata_vector()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.nbytes"><code class="docutils literal notranslate"><span class="pre">MutationTable.nbytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.packset_metadata"><code class="docutils literal notranslate"><span class="pre">MutationTable.packset_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.truncate"><code class="docutils literal notranslate"><span class="pre">MutationTable.truncate()</span></code></a></li>
</ul>
</li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#id14">Associated row class</a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTableRow"><code class="docutils literal notranslate"><span class="pre">MutationTableRow</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTableRow.site"><code class="docutils literal notranslate"><span class="pre">MutationTableRow.site</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTableRow.node"><code class="docutils literal notranslate"><span class="pre">MutationTableRow.node</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTableRow.derived_state"><code class="docutils literal notranslate"><span class="pre">MutationTableRow.derived_state</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTableRow.parent"><code class="docutils literal notranslate"><span class="pre">MutationTableRow.parent</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTableRow.metadata"><code class="docutils literal notranslate"><span class="pre">MutationTableRow.metadata</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTableRow.time"><code class="docutils literal notranslate"><span class="pre">MutationTableRow.time</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTableRow.asdict"><code class="docutils literal notranslate"><span class="pre">MutationTableRow.asdict()</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTableRow.replace"><code class="docutils literal notranslate"><span class="pre">MutationTableRow.replace()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#populationtable-classes"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationTable</span></code> classes</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable"><code class="docutils literal notranslate"><span class="pre">PopulationTable</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.add_row"><code class="docutils literal notranslate"><span class="pre">PopulationTable.add_row()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.set_columns"><code class="docutils literal notranslate"><span class="pre">PopulationTable.set_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.append_columns"><code class="docutils literal notranslate"><span class="pre">PopulationTable.append_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.__getitem__"><code class="docutils literal notranslate"><span class="pre">PopulationTable.__getitem__()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.append"><code class="docutils literal notranslate"><span class="pre">PopulationTable.append()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.asdict"><code class="docutils literal notranslate"><span class="pre">PopulationTable.asdict()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.assert_equals"><code class="docutils literal notranslate"><span class="pre">PopulationTable.assert_equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.clear"><code class="docutils literal notranslate"><span class="pre">PopulationTable.clear()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.copy"><code class="docutils literal notranslate"><span class="pre">PopulationTable.copy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.drop_metadata"><code class="docutils literal notranslate"><span class="pre">PopulationTable.drop_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.equals"><code class="docutils literal notranslate"><span class="pre">PopulationTable.equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.keep_rows"><code class="docutils literal notranslate"><span class="pre">PopulationTable.keep_rows()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.metadata_schema"><code class="docutils literal notranslate"><span class="pre">PopulationTable.metadata_schema</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.metadata_vector"><code class="docutils literal notranslate"><span class="pre">PopulationTable.metadata_vector()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.nbytes"><code class="docutils literal notranslate"><span class="pre">PopulationTable.nbytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.packset_metadata"><code class="docutils literal notranslate"><span class="pre">PopulationTable.packset_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.truncate"><code class="docutils literal notranslate"><span class="pre">PopulationTable.truncate()</span></code></a></li>
</ul>
</li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#id15">Associated row class</a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTableRow"><code class="docutils literal notranslate"><span class="pre">PopulationTableRow</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTableRow.metadata"><code class="docutils literal notranslate"><span class="pre">PopulationTableRow.metadata</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTableRow.asdict"><code class="docutils literal notranslate"><span class="pre">PopulationTableRow.asdict()</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTableRow.replace"><code class="docutils literal notranslate"><span class="pre">PopulationTableRow.replace()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#provenancetable-classes"><code class="xref py py-class docutils literal notranslate"><span class="pre">ProvenanceTable</span></code> classes</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.add_row"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.add_row()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.set_columns"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.set_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.append_columns"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.append_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.packset_record"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.packset_record()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.packset_timestamp"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.packset_timestamp()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.equals"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.assert_equals"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.assert_equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.append"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.append()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.asdict"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.asdict()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.clear"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.clear()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.copy"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.copy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.keep_rows"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.keep_rows()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.nbytes"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.nbytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.truncate"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.truncate()</span></code></a></li>
</ul>
</li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#id16">Associated row class</a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTableRow"><code class="docutils literal notranslate"><span class="pre">ProvenanceTableRow</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTableRow.timestamp"><code class="docutils literal notranslate"><span class="pre">ProvenanceTableRow.timestamp</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTableRow.record"><code class="docutils literal notranslate"><span class="pre">ProvenanceTableRow.record</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTableRow.asdict"><code class="docutils literal notranslate"><span class="pre">ProvenanceTableRow.asdict()</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTableRow.replace"><code class="docutils literal notranslate"><span class="pre">ProvenanceTableRow.replace()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sec-python-api-reference-identity">Identity classes</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-identitysegments-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">IdentitySegments</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegments"><code class="docutils literal notranslate"><span class="pre">IdentitySegments</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegments.num_segments"><code class="docutils literal notranslate"><span class="pre">IdentitySegments.num_segments</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegments.num_pairs"><code class="docutils literal notranslate"><span class="pre">IdentitySegments.num_pairs</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegments.total_span"><code class="docutils literal notranslate"><span class="pre">IdentitySegments.total_span</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegments.pairs"><code class="docutils literal notranslate"><span class="pre">IdentitySegments.pairs</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-identitysegmentlist-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">IdentitySegmentList</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegmentList"><code class="docutils literal notranslate"><span class="pre">IdentitySegmentList</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegmentList.total_span"><code class="docutils literal notranslate"><span class="pre">IdentitySegmentList.total_span</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegmentList.left"><code class="docutils literal notranslate"><span class="pre">IdentitySegmentList.left</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegmentList.right"><code class="docutils literal notranslate"><span class="pre">IdentitySegmentList.right</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegmentList.node"><code class="docutils literal notranslate"><span class="pre">IdentitySegmentList.node</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-identitysegment-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">IdentitySegment</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegment"><code class="docutils literal notranslate"><span class="pre">IdentitySegment</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegment.left"><code class="docutils literal notranslate"><span class="pre">IdentitySegment.left</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegment.right"><code class="docutils literal notranslate"><span class="pre">IdentitySegment.right</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegment.node"><code class="docutils literal notranslate"><span class="pre">IdentitySegment.node</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegment.span"><code class="docutils literal notranslate"><span class="pre">IdentitySegment.span</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#miscellaneous-classes">Miscellaneous classes</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-referencesequence-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">ReferenceSequence</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ReferenceSequence"><code class="docutils literal notranslate"><span class="pre">ReferenceSequence</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ReferenceSequence.is_null"><code class="docutils literal notranslate"><span class="pre">ReferenceSequence.is_null()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ReferenceSequence.data"><code class="docutils literal notranslate"><span class="pre">ReferenceSequence.data</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ReferenceSequence.metadata"><code class="docutils literal notranslate"><span class="pre">ReferenceSequence.metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ReferenceSequence.metadata_bytes"><code class="docutils literal notranslate"><span class="pre">ReferenceSequence.metadata_bytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ReferenceSequence.metadata_schema"><code class="docutils literal notranslate"><span class="pre">ReferenceSequence.metadata_schema</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-metadataschema-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">MetadataSchema</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MetadataSchema"><code class="docutils literal notranslate"><span class="pre">MetadataSchema</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MetadataSchema.asdict"><code class="docutils literal notranslate"><span class="pre">MetadataSchema.asdict()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MetadataSchema.validate_and_encode_row"><code class="docutils literal notranslate"><span class="pre">MetadataSchema.validate_and_encode_row()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MetadataSchema.decode_row"><code class="docutils literal notranslate"><span class="pre">MetadataSchema.decode_row()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MetadataSchema.encode_row"><code class="docutils literal notranslate"><span class="pre">MetadataSchema.encode_row()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MetadataSchema.structured_array_from_buffer"><code class="docutils literal notranslate"><span class="pre">MetadataSchema.structured_array_from_buffer()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MetadataSchema.permissive_json"><code class="docutils literal notranslate"><span class="pre">MetadataSchema.permissive_json()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MetadataSchema.null"><code class="docutils literal notranslate"><span class="pre">MetadataSchema.null()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-tablemetadataschemas-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">TableMetadataSchemas</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableMetadataSchemas"><code class="docutils literal notranslate"><span class="pre">TableMetadataSchemas</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableMetadataSchemas.node"><code class="docutils literal notranslate"><span class="pre">TableMetadataSchemas.node</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableMetadataSchemas.edge"><code class="docutils literal notranslate"><span class="pre">TableMetadataSchemas.edge</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableMetadataSchemas.site"><code class="docutils literal notranslate"><span class="pre">TableMetadataSchemas.site</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableMetadataSchemas.mutation"><code class="docutils literal notranslate"><span class="pre">TableMetadataSchemas.mutation</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableMetadataSchemas.migration"><code class="docutils literal notranslate"><span class="pre">TableMetadataSchemas.migration</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableMetadataSchemas.individual"><code class="docutils literal notranslate"><span class="pre">TableMetadataSchemas.individual</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableMetadataSchemas.population"><code class="docutils literal notranslate"><span class="pre">TableMetadataSchemas.population</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-topologycounter-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">TopologyCounter</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TopologyCounter"><code class="docutils literal notranslate"><span class="pre">TopologyCounter</span></code></a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-ldcalculator-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">LdCalculator</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.LdCalculator"><code class="docutils literal notranslate"><span class="pre">LdCalculator</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.LdCalculator.r2"><code class="docutils literal notranslate"><span class="pre">LdCalculator.r2()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.LdCalculator.r2_array"><code class="docutils literal notranslate"><span class="pre">LdCalculator.r2_array()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.LdCalculator.r2_matrix"><code class="docutils literal notranslate"><span class="pre">LdCalculator.r2_matrix()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-tablecollectionindexes-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollectionIndexes</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollectionIndexes"><code class="docutils literal notranslate"><span class="pre">TableCollectionIndexes</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollectionIndexes.asdict"><code class="docutils literal notranslate"><span class="pre">TableCollectionIndexes.asdict()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollectionIndexes.nbytes"><code class="docutils literal notranslate"><span class="pre">TableCollectionIndexes.nbytes</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-svgstring-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">SVGString</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SVGString"><code class="docutils literal notranslate"><span class="pre">SVGString</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SVGString._repr_svg_"><code class="docutils literal notranslate"><span class="pre">SVGString._repr_svg_()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-pcaresult-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">PCAResult</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PCAResult"><code class="docutils literal notranslate"><span class="pre">PCAResult</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PCAResult.factors"><code class="docutils literal notranslate"><span class="pre">PCAResult.factors</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PCAResult.eigenvalues"><code class="docutils literal notranslate"><span class="pre">PCAResult.eigenvalues</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PCAResult.range_sketch"><code class="docutils literal notranslate"><span class="pre">PCAResult.range_sketch</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PCAResult.error_bound"><code class="docutils literal notranslate"><span class="pre">PCAResult.error_bound</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="python-api">
<span id="sec-python-api"></span><h1>Python API<a class="headerlink" href="#python-api" title="Link to this heading">#</a></h1>
<p>This page documents the full tskit Python API. Brief thematic summaries of common
classes and methods are presented first. The <a class="reference internal" href="#sec-python-api-reference"><span class="std std-ref">Reference documentation</span></a> section
at the end then contains full details which aim to be concise, precise and exhaustive.
Note that this may not therefore be the best place to start if you are new
to a particular piece of functionality.</p>
<section id="trees-and-tree-sequences">
<span id="sec-python-api-trees-and-tree-sequences"></span><h2>Trees and tree sequences<a class="headerlink" href="#trees-and-tree-sequences" title="Link to this heading">#</a></h2>
<p>The <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> class represents a sequence of correlated
evolutionary trees along a genome. The <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> class represents a
single tree in this sequence. These classes are the interfaces used to
interact with the trees and mutational information stored in a tree sequence,
for example as returned from a simulation or inferred from a set of DNA
sequences.</p>
<section id="treesequence-api">
<span id="sec-python-api-tree-sequences"></span><h3><a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> API<a class="headerlink" href="#treesequence-api" title="Link to this heading">#</a></h3>
<section id="general-properties">
<span id="sec-python-api-tree-sequences-properties"></span><h4>General properties<a class="headerlink" href="#general-properties" title="Link to this heading">#</a></h4>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.time_units" title="tskit.TreeSequence.time_units"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.time_units</span></code></a></p></td>
<td><p>String describing the units of the time dimension for this TreeSequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.nbytes" title="tskit.TreeSequence.nbytes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.nbytes</span></code></a></p></td>
<td><p>Returns the total number of bytes required to store the data in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.sequence_length" title="tskit.TreeSequence.sequence_length"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.sequence_length</span></code></a></p></td>
<td><p>Returns the sequence length in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.max_root_time" title="tskit.TreeSequence.max_root_time"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.max_root_time</span></code></a></p></td>
<td><p>Returns the time of the oldest root in any of the trees in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.discrete_genome" title="tskit.TreeSequence.discrete_genome"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.discrete_genome</span></code></a></p></td>
<td><p>Returns True if all genome coordinates in this TreeSequence are discrete integer values.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.discrete_time" title="tskit.TreeSequence.discrete_time"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.discrete_time</span></code></a></p></td>
<td><p>Returns True if all time coordinates in this TreeSequence are discrete integer values.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.metadata" title="tskit.TreeSequence.metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.metadata</span></code></a></p></td>
<td><p>The decoded metadata for this TreeSequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.metadata_schema" title="tskit.TreeSequence.metadata_schema"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.metadata_schema</span></code></a></p></td>
<td><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this TreeSequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.reference_sequence" title="tskit.TreeSequence.reference_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.reference_sequence</span></code></a></p></td>
<td><p>The <a class="reference internal" href="#tskit.ReferenceSequence" title="tskit.ReferenceSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReferenceSequence</span></code></a> associated with this <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> if one is defined (see <a class="reference internal" href="#tskit.TreeSequence.has_reference_sequence" title="tskit.TreeSequence.has_reference_sequence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.has_reference_sequence()</span></code></a>), or None otherwise.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="efficient-table-column-access">
<h4>Efficient table column access<a class="headerlink" href="#efficient-table-column-access" title="Link to this heading">#</a></h4>
<p>The <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> class provides access to underlying numerical
data defined in the <a class="reference internal" href="data-model.html#sec-data-model"><span class="std std-ref">data model</span></a> in two ways:</p>
<ol class="arabic simple">
<li><p>Via the <a class="reference internal" href="#tskit.TreeSequence.tables" title="tskit.TreeSequence.tables"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.tables</span></code></a> property and the
<a class="reference internal" href="#sec-tables-api-accessing-table-data"><span class="std std-ref">Tables API</span></a>.
Since version 1.0 this provides a direct, zero-copy, immutable view of the
underlying memory.</p></li>
<li><p>Via a set of properties on the <code class="docutils literal notranslate"><span class="pre">TreeSequence</span></code> class that provide
direct and efficient access to a single array in the underlying memory.</p></li>
</ol>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.individuals_flags" title="tskit.TreeSequence.individuals_flags"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.individuals_flags</span></code></a></p></td>
<td><p>Efficient access to the bitwise <code class="docutils literal notranslate"><span class="pre">flags</span></code> column in the <a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">Individual Table</span></a> as a numpy array (dtype=np.uint32).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.nodes_time" title="tskit.TreeSequence.nodes_time"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.nodes_time</span></code></a></p></td>
<td><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">time</span></code> column in the <a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">Node Table</span></a> as a numpy array (dtype=np.float64).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.nodes_flags" title="tskit.TreeSequence.nodes_flags"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.nodes_flags</span></code></a></p></td>
<td><p>Efficient access to the bitwise <code class="docutils literal notranslate"><span class="pre">flags</span></code> column in the <a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">Node Table</span></a> as a numpy array (dtype=np.uint32).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.nodes_population" title="tskit.TreeSequence.nodes_population"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.nodes_population</span></code></a></p></td>
<td><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">population</span></code> column in the <a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">Node Table</span></a> as a numpy array (dtype=np.int32).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.nodes_individual" title="tskit.TreeSequence.nodes_individual"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.nodes_individual</span></code></a></p></td>
<td><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">individual</span></code> column in the <a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">Node Table</span></a> as a numpy array (dtype=np.int32).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.edges_left" title="tskit.TreeSequence.edges_left"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.edges_left</span></code></a></p></td>
<td><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">left</span></code> column in the <a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">Edge Table</span></a> as a numpy array (dtype=np.float64).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.edges_right" title="tskit.TreeSequence.edges_right"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.edges_right</span></code></a></p></td>
<td><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">right</span></code> column in the <a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">Edge Table</span></a> as a numpy array (dtype=np.float64).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.edges_parent" title="tskit.TreeSequence.edges_parent"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.edges_parent</span></code></a></p></td>
<td><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">parent</span></code> column in the <a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">Edge Table</span></a> as a numpy array (dtype=np.int32).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.edges_child" title="tskit.TreeSequence.edges_child"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.edges_child</span></code></a></p></td>
<td><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">child</span></code> column in the <a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">Edge Table</span></a> as a numpy array (dtype=np.int32).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.sites_position" title="tskit.TreeSequence.sites_position"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.sites_position</span></code></a></p></td>
<td><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">position</span></code> column in the <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">Site Table</span></a> as a numpy array (dtype=np.float64).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.sites_ancestral_state" title="tskit.TreeSequence.sites_ancestral_state"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.sites_ancestral_state</span></code></a></p></td>
<td><p>The <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> column in the <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">Site Table</span></a> as a numpy array (dtype=StringDtype).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.mutations_site" title="tskit.TreeSequence.mutations_site"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.mutations_site</span></code></a></p></td>
<td><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">site</span></code> column in the <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">Mutation Table</span></a> as a numpy array (dtype=np.int32).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.mutations_node" title="tskit.TreeSequence.mutations_node"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.mutations_node</span></code></a></p></td>
<td><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">node</span></code> column in the <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">Mutation Table</span></a> as a numpy array (dtype=np.int32).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.mutations_parent" title="tskit.TreeSequence.mutations_parent"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.mutations_parent</span></code></a></p></td>
<td><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">parent</span></code> column in the <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">Mutation Table</span></a> as a numpy array (dtype=np.int32).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.mutations_time" title="tskit.TreeSequence.mutations_time"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.mutations_time</span></code></a></p></td>
<td><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">time</span></code> column in the <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">Mutation Table</span></a> as a numpy array (dtype=np.float64).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.mutations_derived_state" title="tskit.TreeSequence.mutations_derived_state"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.mutations_derived_state</span></code></a></p></td>
<td><p>Access to the <code class="docutils literal notranslate"><span class="pre">derived_state</span></code> column in the <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">Mutation Table</span></a> as a numpy array (dtype=StringDtype).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.mutations_metadata" title="tskit.TreeSequence.mutations_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.mutations_metadata</span></code></a></p></td>
<td><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> column in the <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">Mutation Table</span></a> as a structured numpy array.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.mutations_edge" title="tskit.TreeSequence.mutations_edge"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.mutations_edge</span></code></a></p></td>
<td><p>Return an array of the ID of the edge each mutation sits on in the tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.mutations_inherited_state" title="tskit.TreeSequence.mutations_inherited_state"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.mutations_inherited_state</span></code></a></p></td>
<td><p>Return an array of the inherited state for each mutation in the tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.migrations_left" title="tskit.TreeSequence.migrations_left"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.migrations_left</span></code></a></p></td>
<td><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">left</span></code> column in the <a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">Migration Table</span></a> as a numpy array (dtype=np.float64).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.migrations_right" title="tskit.TreeSequence.migrations_right"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.migrations_right</span></code></a></p></td>
<td><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">right</span></code> column in the <a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">Migration Table</span></a> as a numpy array (dtype=np.float64).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.migrations_right" title="tskit.TreeSequence.migrations_right"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.migrations_right</span></code></a></p></td>
<td><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">right</span></code> column in the <a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">Migration Table</span></a> as a numpy array (dtype=np.float64).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.migrations_node" title="tskit.TreeSequence.migrations_node"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.migrations_node</span></code></a></p></td>
<td><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">node</span></code> column in the <a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">Migration Table</span></a> as a numpy array (dtype=np.int32).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.migrations_source" title="tskit.TreeSequence.migrations_source"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.migrations_source</span></code></a></p></td>
<td><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">source</span></code> column in the <a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">Migration Table</span></a> as a numpy array (dtype=np.int32).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.migrations_dest" title="tskit.TreeSequence.migrations_dest"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.migrations_dest</span></code></a></p></td>
<td><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">dest</span></code> column in the <a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">Migration Table</span></a> as a numpy array (dtype=np.int32).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.migrations_time" title="tskit.TreeSequence.migrations_time"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.migrations_time</span></code></a></p></td>
<td><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">time</span></code> column in the <a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">Migration Table</span></a> as a numpy array (dtype=np.float64).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.indexes_edge_insertion_order" title="tskit.TreeSequence.indexes_edge_insertion_order"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.indexes_edge_insertion_order</span></code></a></p></td>
<td><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">edge_insertion_order</span></code> column in the <a class="reference internal" href="data-model.html#sec-table-indexes"><span class="std std-ref">Table indexes</span></a> as a numpy array (dtype=np.int32).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.indexes_edge_removal_order" title="tskit.TreeSequence.indexes_edge_removal_order"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.indexes_edge_removal_order</span></code></a></p></td>
<td><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">edge_removal_order</span></code> column in the <a class="reference internal" href="data-model.html#sec-table-indexes"><span class="std std-ref">Table indexes</span></a> as a numpy array (dtype=np.int32).</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="loading-and-saving">
<span id="sec-python-api-tree-sequences-loading-and-saving"></span><h4>Loading and saving<a class="headerlink" href="#loading-and-saving" title="Link to this heading">#</a></h4>
<p>There are several methods for loading data into a <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>
instance. The simplest and most convenient is the use the <a class="reference internal" href="#tskit.load" title="tskit.load"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load()</span></code></a>
function to load a <a class="reference internal" href="file-formats.html#sec-tree-sequence-file-format"><span class="std std-ref">tree sequence file</span></a>. For small
scale data and debugging, it is often convenient to use the <a class="reference internal" href="#tskit.load_text" title="tskit.load_text"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load_text()</span></code></a>
function to read data in the <a class="reference internal" href="file-formats.html#sec-text-file-format"><span class="std std-ref">text file format</span></a>.
The <a class="reference internal" href="#tskit.TableCollection.tree_sequence" title="tskit.TableCollection.tree_sequence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.tree_sequence()</span></code></a> function
efficiently creates a <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> object from a
<a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">collection</span> <span class="pre">of</span> <span class="pre">tables</span></code></a>
using the <a class="reference internal" href="#sec-tables-api"><span class="std std-ref">Tables API</span></a>.</p>
<dl>
<dt>Load a tree sequence</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.load" title="tskit.load"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load</span></code></a>(file,*[,skip_tables,...])</p></td>
<td><p>Return a <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> instance loaded from the specified file object or path.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.load_text" title="tskit.load_text"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_text</span></code></a>(nodes,edges[,sites,mutations,...])</p></td>
<td><p>Return a <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> instance parsed from tabulated text data contained in the specified file-like objects.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.tree_sequence" title="tskit.TableCollection.tree_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.tree_sequence</span></code></a>()</p></td>
<td><p>Returns a <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> instance from the tables defined in this <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>, building the required indexes if they have not yet been created by <a class="reference internal" href="#tskit.TableCollection.build_index" title="tskit.TableCollection.build_index"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_index()</span></code></a>.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
<dt>Save a tree sequence</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.dump" title="tskit.TreeSequence.dump"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.dump</span></code></a>(file_or_path[,...])</p></td>
<td><p>Writes the tree sequence to the specified path or file object.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>Tree sequences with a single simple topology can also be created from scratch by
<a class="reference internal" href="#sec-python-api-trees-creating"><span class="std std-ref">generating</span></a> a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> and accessing its
<a class="reference internal" href="#tskit.Tree.tree_sequence" title="tskit.Tree.tree_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tree_sequence</span></code></a> property.</p>
</div>
</section>
<section id="obtaining-trees">
<span id="sec-python-api-tree-sequences-obtaining-trees"></span><h4>Obtaining trees<a class="headerlink" href="#obtaining-trees" title="Link to this heading">#</a></h4>
<p>The following properties and methods return information about the
<a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">trees</span></code></a> that are generated along a tree sequence.</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.num_trees" title="tskit.TreeSequence.num_trees"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.num_trees</span></code></a></p></td>
<td><p>Returns the number of distinct trees in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.trees</span></code></a>([tracked_samples,...])</p></td>
<td><p>Returns an iterator over the trees in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.breakpoints" title="tskit.TreeSequence.breakpoints"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.breakpoints</span></code></a>([as_array])</p></td>
<td><p>Returns the breakpoints that separate trees along the chromosome, including the two extreme points 0 and L. This is equivalent to::.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.coiterate" title="tskit.TreeSequence.coiterate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.coiterate</span></code></a>(other,**kwargs)</p></td>
<td><p>Returns an iterator over the pairs of trees for each distinct interval in the specified pair of tree sequences.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.first" title="tskit.TreeSequence.first"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.first</span></code></a>(**kwargs)</p></td>
<td><p>Returns the first tree in this <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.last" title="tskit.TreeSequence.last"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.last</span></code></a>(**kwargs)</p></td>
<td><p>Returns the last tree in this <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.aslist" title="tskit.TreeSequence.aslist"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.aslist</span></code></a>(**kwargs)</p></td>
<td><p>Returns the trees in this tree sequence as a list.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.at" title="tskit.TreeSequence.at"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.at</span></code></a>(position,**kwargs)</p></td>
<td><p>Returns the tree covering the specified genomic location.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.at_index" title="tskit.TreeSequence.at_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.at_index</span></code></a>(index,**kwargs)</p></td>
<td><p>Returns the tree at the specified index.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="obtaining-other-objects">
<h4>Obtaining other objects<a class="headerlink" href="#obtaining-other-objects" title="Link to this heading">#</a></h4>
<p id="sec-python-api-tree-sequences-obtaining-other-objects">Various components make up a tree sequence, such as nodes and edges, sites and
mutations, and populations and individuals. These can be counted or converted into
Python objects using the following classes, properties, and methods.</p>
<section id="tree-topology">
<h5>Tree topology<a class="headerlink" href="#tree-topology" title="Link to this heading">#</a></h5>
<dl>
<dt>Nodes</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Node" title="tskit.Node"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Node</span></code></a>(*args[,metadata_decoder])</p></td>
<td><p>A <a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">node</span></a> in a tree sequence, corresponding to a single genome.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.num_nodes" title="tskit.TreeSequence.num_nodes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.num_nodes</span></code></a></p></td>
<td><p>Returns the number of <a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">nodes</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.nodes" title="tskit.TreeSequence.nodes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.nodes</span></code></a>(*[,order])</p></td>
<td><p>Returns an iterable sequence of all the <a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">nodes</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.node" title="tskit.TreeSequence.node"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.node</span></code></a>(id_)</p></td>
<td><p>Returns the <a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">node</span></a> in this tree sequence with the specified ID.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.num_samples" title="tskit.TreeSequence.num_samples"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.num_samples</span></code></a></p></td>
<td><p>Returns the number of sample nodes in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.samples" title="tskit.TreeSequence.samples"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.samples</span></code></a>([population,...])</p></td>
<td><p>Returns an array of the sample node IDs in this tree sequence.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
<dt>Edges</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Edge" title="tskit.Edge"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Edge</span></code></a>(left,right,parent,child[,metadata,...])</p></td>
<td><p>An <a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">edge</span></a> in a tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.num_edges" title="tskit.TreeSequence.num_edges"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.num_edges</span></code></a></p></td>
<td><p>Returns the number of <a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">edges</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.edges" title="tskit.TreeSequence.edges"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.edges</span></code></a>()</p></td>
<td><p>Returns an iterable sequence of all the <a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">edges</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.edge" title="tskit.TreeSequence.edge"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.edge</span></code></a>(id_)</p></td>
<td><p>Returns the <a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">edge</span></a> in this tree sequence with the specified ID.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
</dl>
</section>
<section id="genetic-variation">
<h5>Genetic variation<a class="headerlink" href="#genetic-variation" title="Link to this heading">#</a></h5>
<dl>
<dt>Sites</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Site" title="tskit.Site"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Site</span></code></a>(*args[,metadata_decoder])</p></td>
<td><p>A <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">site</span></a> in a tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.num_sites" title="tskit.TreeSequence.num_sites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.num_sites</span></code></a></p></td>
<td><p>Returns the number of <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">sites</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.sites" title="tskit.TreeSequence.sites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.sites</span></code></a>()</p></td>
<td><p>Returns an iterable sequence of all the <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">sites</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.site" title="tskit.TreeSequence.site"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.site</span></code></a>([id_,position])</p></td>
<td><p>Returns the <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">site</span></a> in this tree sequence with either the specified ID or position.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Variant" title="tskit.Variant"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Variant</span></code></a>(tree_sequence[,samples,...])</p></td>
<td><p>A variant in a tree sequence, describing the observed genetic variation among the specified nodes (by default, the sample nodes) for a given site.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.variants" title="tskit.TreeSequence.variants"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.variants</span></code></a>(*[,samples,...])</p></td>
<td><p>Returns an iterator over the variants between the <code class="docutils literal notranslate"><span class="pre">left</span></code> (inclusive) and <code class="docutils literal notranslate"><span class="pre">right</span></code> (exclusive) genomic positions in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.genotype_matrix" title="tskit.TreeSequence.genotype_matrix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.genotype_matrix</span></code></a>(*[,samples,...])</p></td>
<td><p>Returns an <span class="math notranslate nohighlight">\(m \times n\)</span> numpy array of the genotypes in this tree sequence, where <span class="math notranslate nohighlight">\(m\)</span> is the number of sites and <span class="math notranslate nohighlight">\(n\)</span> is the number of requested nodes (default: the number of sample nodes).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.haplotypes" title="tskit.TreeSequence.haplotypes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.haplotypes</span></code></a>(*[,...])</p></td>
<td><p>Returns an iterator over the strings of haplotypes that result from the trees and mutations in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.alignments" title="tskit.TreeSequence.alignments"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.alignments</span></code></a>(*[,...])</p></td>
<td><p>Returns an iterator over the full sequence alignments for the defined samples in this tree sequence.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
<dt>Mutations</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Mutation" title="tskit.Mutation"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Mutation</span></code></a>(*args[,metadata_decoder])</p></td>
<td><p>A <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">mutation</span></a> in a tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.num_mutations" title="tskit.TreeSequence.num_mutations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.num_mutations</span></code></a></p></td>
<td><p>Returns the number of <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">mutations</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.mutations" title="tskit.TreeSequence.mutations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.mutations</span></code></a>()</p></td>
<td><p>Returns an iterator over all the <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">mutations</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.mutation" title="tskit.TreeSequence.mutation"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.mutation</span></code></a>(id_)</p></td>
<td><p>Returns the <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">mutation</span></a> in this tree sequence with the specified ID.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
</dl>
</section>
<section id="demography">
<h5>Demography<a class="headerlink" href="#demography" title="Link to this heading">#</a></h5>
<dl>
<dt>Populations</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Population" title="tskit.Population"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Population</span></code></a>(*args[,metadata_decoder])</p></td>
<td><p>A <a class="reference internal" href="data-model.html#sec-population-table-definition"><span class="std std-ref">population</span></a> in a tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.num_populations" title="tskit.TreeSequence.num_populations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.num_populations</span></code></a></p></td>
<td><p>Returns the number of <a class="reference internal" href="data-model.html#sec-population-table-definition"><span class="std std-ref">populations</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.populations" title="tskit.TreeSequence.populations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.populations</span></code></a>()</p></td>
<td><p>Returns an iterable sequence of all the <a class="reference internal" href="data-model.html#sec-population-table-definition"><span class="std std-ref">populations</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.population" title="tskit.TreeSequence.population"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.population</span></code></a>(id_)</p></td>
<td><p>Returns the <a class="reference internal" href="data-model.html#sec-population-table-definition"><span class="std std-ref">population</span></a> in this tree sequence with the specified ID.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
<dt>Migrations</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Migration" title="tskit.Migration"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Migration</span></code></a>(*args[,metadata_decoder])</p></td>
<td><p>A <a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">migration</span></a> in a tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.num_migrations" title="tskit.TreeSequence.num_migrations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.num_migrations</span></code></a></p></td>
<td><p>Returns the number of <a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">migrations</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.migrations" title="tskit.TreeSequence.migrations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.migrations</span></code></a>()</p></td>
<td><p>Returns an iterable sequence of all the <a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">migrations</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.migration" title="tskit.TreeSequence.migration"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.migration</span></code></a>(id_)</p></td>
<td><p>Returns the <a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">migration</span></a> in this tree sequence with the specified ID.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
</dl>
</section>
<section id="other">
<h5>Other<a class="headerlink" href="#other" title="Link to this heading">#</a></h5>
<dl>
<dt>Individuals</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Individual" title="tskit.Individual"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Individual</span></code></a>(*args[,tree_sequence])</p></td>
<td><p>An <a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">individual</span></a> in a tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.num_individuals" title="tskit.TreeSequence.num_individuals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.num_individuals</span></code></a></p></td>
<td><p>Returns the number of <a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">individuals</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.individuals" title="tskit.TreeSequence.individuals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.individuals</span></code></a>()</p></td>
<td><p>Returns an iterable sequence of all the <a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">individuals</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.individual" title="tskit.TreeSequence.individual"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.individual</span></code></a>(id_)</p></td>
<td><p>Returns the <a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">individual</span></a> in this tree sequence with the specified ID.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
<dt>Provenance entries (also see <a class="reference internal" href="#sec-python-api-provenance"><span class="std std-ref">Provenance</span></a>)</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Provenance" title="tskit.Provenance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Provenance</span></code></a>(id,timestamp,record)</p></td>
<td><p>A provenance entry in a tree sequence, detailing how this tree sequence was generated, or subsequent operations on it (see <a class="reference internal" href="provenance.html#sec-provenance"><span class="std std-ref">Provenance</span></a>).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.num_provenances" title="tskit.TreeSequence.num_provenances"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.num_provenances</span></code></a></p></td>
<td><p>Returns the number of <a class="reference internal" href="data-model.html#sec-provenance-table-definition"><span class="std std-ref">provenances</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.provenances" title="tskit.TreeSequence.provenances"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.provenances</span></code></a>()</p></td>
<td><p>Returns an iterable sequence of all the <a class="reference internal" href="data-model.html#sec-provenance-table-definition"><span class="std std-ref">provenances</span></a> in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.provenance" title="tskit.TreeSequence.provenance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.provenance</span></code></a>(id_)</p></td>
<td><p>Returns the <a class="reference internal" href="data-model.html#sec-provenance-table-definition"><span class="std std-ref">provenance</span></a> in this tree sequence with the specified ID.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
</dl>
</section>
</section>
<section id="tree-sequence-modification">
<span id="sec-python-api-tree-sequences-modification"></span><h4>Tree sequence modification<a class="headerlink" href="#tree-sequence-modification" title="Link to this heading">#</a></h4>
<p>Although tree sequences are immutable, several methods will taken an existing tree
sequence and return a modifed version. These are thin wrappers around the
<a class="reference internal" href="#sec-tables-api-modification"><span class="std std-ref">identically named methods of a TableCollection</span></a>,
which perform the same actions but modify the <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a> in place.</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.simplify" title="tskit.TreeSequence.simplify"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.simplify</span></code></a>([samples,map_nodes,...])</p></td>
<td><p>Returns a simplified tree sequence that retains only the history of the nodes given in the list <code class="docutils literal notranslate"><span class="pre">samples</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.subset" title="tskit.TreeSequence.subset"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.subset</span></code></a>(nodes[,...])</p></td>
<td><p>Returns a tree sequence containing only information directly referencing the provided list of nodes to retain.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.union" title="tskit.TreeSequence.union"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.union</span></code></a>(other,node_mapping[,...])</p></td>
<td><p>Returns an expanded tree sequence which contains the node-wise union of <code class="docutils literal notranslate"><span class="pre">self</span></code> and <code class="docutils literal notranslate"><span class="pre">other</span></code>, obtained by adding the non-shared portions of <code class="docutils literal notranslate"><span class="pre">other</span></code> onto <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.concatenate" title="tskit.TreeSequence.concatenate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.concatenate</span></code></a>(*args[,...])</p></td>
<td><p>Concatenate a set of tree sequences to the right of this one, by shifting their coordinate systems and adding all edges, sites, mutations, and any additional nodes, individuals, or populations needed for these.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.keep_intervals" title="tskit.TreeSequence.keep_intervals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.keep_intervals</span></code></a>(intervals[,...])</p></td>
<td><p>Returns a copy of this tree sequence which includes only information in the specified list of genomic intervals.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.delete_intervals" title="tskit.TreeSequence.delete_intervals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.delete_intervals</span></code></a>(intervals[,...])</p></td>
<td><p>Returns a copy of this tree sequence for which information in the specified list of genomic intervals has been deleted.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.delete_sites" title="tskit.TreeSequence.delete_sites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.delete_sites</span></code></a>(site_ids[,...])</p></td>
<td><p>Returns a copy of this tree sequence with the specified sites (and their associated mutations) entirely removed.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.trim" title="tskit.TreeSequence.trim"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.trim</span></code></a>([record_provenance])</p></td>
<td><p>Returns a copy of this tree sequence with any empty regions (i.e., those not covered by any edge) on the right and left trimmed away.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.shift" title="tskit.TreeSequence.shift"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.shift</span></code></a>(value[,sequence_length,...])</p></td>
<td><p>Shift the coordinate system (used by edges and sites) of this TableCollection by a given value.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.split_edges" title="tskit.TreeSequence.split_edges"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.split_edges</span></code></a>(time,*[,flags,...])</p></td>
<td><p>Returns a copy of this tree sequence in which we replace any edge <code class="docutils literal notranslate"><span class="pre">(left,</span> <span class="pre">right,</span> <span class="pre">parent,</span> <span class="pre">child)</span></code> in which <code class="docutils literal notranslate"><span class="pre">node_time[child]</span> <span class="pre">&lt;</span> <span class="pre">time</span> <span class="pre">&lt;</span> <span class="pre">node_time[parent]</span></code> with two edges <code class="docutils literal notranslate"><span class="pre">(left,</span> <span class="pre">right,</span> <span class="pre">parent,</span> <span class="pre">u)</span></code> and <code class="docutils literal notranslate"><span class="pre">(left,</span> <span class="pre">right,</span> <span class="pre">u,</span> <span class="pre">child)</span></code>, where <code class="docutils literal notranslate"><span class="pre">u</span></code> is a newly added node for each intersecting edge.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.decapitate" title="tskit.TreeSequence.decapitate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.decapitate</span></code></a>(time,*[,flags,...])</p></td>
<td><p>Delete all edge topology and mutational information at least as old as the specified time from this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.extend_haplotypes" title="tskit.TreeSequence.extend_haplotypes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.extend_haplotypes</span></code></a>([max_iter])</p></td>
<td><p>Returns a new tree sequence in which the span covered by ancestral nodes is &quot;extended&quot; to regions of the genome according to the following rule: If an ancestral segment corresponding to node <cite>n</cite> has ancestor <cite>p</cite> and descendant <cite>c</cite> on some portion of the genome, and on an adjacent segment of genome <cite>p</cite> is still an ancestor of <cite>c</cite>, then <cite>n</cite> is inserted into the path from <cite>p</cite> to <cite>c</cite>.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="sec-python-api-tree-sequences-ibd">
<span id="identity-by-descent"></span><h4>Identity by descent<a class="headerlink" href="#sec-python-api-tree-sequences-ibd" title="Link to this heading">#</a></h4>
<p>The <a class="reference internal" href="#tskit.TreeSequence.ibd_segments" title="tskit.TreeSequence.ibd_segments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.ibd_segments()</span></code></a> method allows us to compute
identity relationships between pairs of samples. See the
<a class="reference internal" href="ibd.html#sec-identity"><span class="std std-ref">Identity by descent</span></a> section for more details and examples
and the <a class="reference internal" href="#sec-python-api-reference-identity"><span class="std std-ref">Identity classes</span></a> section for
API documentation on the associated classes.</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.ibd_segments" title="tskit.TreeSequence.ibd_segments"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.ibd_segments</span></code></a>(*[,within,...])</p></td>
<td><p>Finds pairs of samples that are identical by descent (IBD) and returns the result as an <a class="reference internal" href="#tskit.IdentitySegments" title="tskit.IdentitySegments"><code class="xref py py-class docutils literal notranslate"><span class="pre">IdentitySegments</span></code></a> instance.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="tables">
<span id="sec-python-api-tree-sequences-tables"></span><h4>Tables<a class="headerlink" href="#tables" title="Link to this heading">#</a></h4>
<p>The underlying data in a tree sequence is stored in a
<a class="reference internal" href="#sec-tables-api"><span class="std std-ref">collection of tables</span></a>. The following methods give access
to tables and associated functionality. Since tables can be modified, this
allows tree sequences to be edited: see the <a class="reference external" href="https://tskit.dev/tutorials/tables_and_editing.html#sec-tables" title="(in Project name not set)"><span>Tables and editing</span></a> tutorial for
an introduction.</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.tables" title="tskit.TreeSequence.tables"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.tables</span></code></a></p></td>
<td><p>Returns an immutable view of the tables underlying this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.dump_tables" title="tskit.TreeSequence.dump_tables"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.dump_tables</span></code></a>()</p></td>
<td><p>Returns a modifiable copy of the <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">tables</span></code></a> defining this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.table_metadata_schemas" title="tskit.TreeSequence.table_metadata_schemas"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.table_metadata_schemas</span></code></a></p></td>
<td><p>The set of metadata schemas for the tables in this tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.tables_dict" title="tskit.TreeSequence.tables_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.tables_dict</span></code></a></p></td>
<td><p>Returns a dictionary mapping names to tables in the underlying <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="statistics">
<span id="sec-python-api-tree-sequences-statistics"></span><h4>Statistics<a class="headerlink" href="#statistics" title="Link to this heading">#</a></h4>
<dl>
<dt>Single site</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.allele_frequency_spectrum" title="tskit.TreeSequence.allele_frequency_spectrum"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.allele_frequency_spectrum</span></code></a>([...])</p></td>
<td><p>Computes the allele frequency spectrum (AFS) in windows across the genome for with respect to the specified <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.divergence" title="tskit.TreeSequence.divergence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.divergence</span></code></a>(sample_sets[,...])</p></td>
<td><p>Computes mean genetic divergence between (and within) pairs of sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.diversity" title="tskit.TreeSequence.diversity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.diversity</span></code></a>([sample_sets,...])</p></td>
<td><p>Computes mean genetic diversity (also known as &quot;pi&quot;) in each of the sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.f2" title="tskit.TreeSequence.f2"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.f2</span></code></a>(sample_sets[,indexes,...])</p></td>
<td><p>Computes Patterson's f2 statistic between two groups of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.f3" title="tskit.TreeSequence.f3"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.f3</span></code></a>(sample_sets[,indexes,...])</p></td>
<td><p>Computes Patterson's f3 statistic between three groups of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.f4" title="tskit.TreeSequence.f4"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.f4</span></code></a>(sample_sets[,indexes,...])</p></td>
<td><p>Computes Patterson's f4 statistic between four groups of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.Fst" title="tskit.TreeSequence.Fst"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.Fst</span></code></a>(sample_sets[,indexes,...])</p></td>
<td><p>Computes &quot;windowed&quot; Fst between pairs of sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.genealogical_nearest_neighbours" title="tskit.TreeSequence.genealogical_nearest_neighbours"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.genealogical_nearest_neighbours</span></code></a>(...)</p></td>
<td><p>Return the genealogical nearest neighbours (GNN) proportions for the given focal nodes, with reference to two or more sets of interest, averaged over all trees in the tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.genetic_relatedness" title="tskit.TreeSequence.genetic_relatedness"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.genetic_relatedness</span></code></a>(sample_sets)</p></td>
<td><p>Computes genetic relatedness between (and within) pairs of sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.genetic_relatedness_weighted" title="tskit.TreeSequence.genetic_relatedness_weighted"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.genetic_relatedness_weighted</span></code></a>(W)</p></td>
<td><p>Computes weighted genetic relatedness.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.genetic_relatedness_vector" title="tskit.TreeSequence.genetic_relatedness_vector"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.genetic_relatedness_vector</span></code></a>(W[,...])</p></td>
<td><p>Computes the product of the genetic relatedness matrix and a vector of weights (one per sample).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.genetic_relatedness_matrix" title="tskit.TreeSequence.genetic_relatedness_matrix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.genetic_relatedness_matrix</span></code></a>([...])</p></td>
<td><p>Computes the full matrix of pairwise genetic relatedness values between (and within) pairs of sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.general_stat" title="tskit.TreeSequence.general_stat"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.general_stat</span></code></a>(W,f,output_dim)</p></td>
<td><p>Compute a windowed statistic from weights and a summary function.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.segregating_sites" title="tskit.TreeSequence.segregating_sites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.segregating_sites</span></code></a>([...])</p></td>
<td><p>Computes the density of segregating sites for each of the sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>, and related quantities.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.sample_count_stat" title="tskit.TreeSequence.sample_count_stat"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.sample_count_stat</span></code></a>(sample_sets,...)</p></td>
<td><p>Compute a windowed statistic from sample counts and a summary function.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.mean_descendants" title="tskit.TreeSequence.mean_descendants"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.mean_descendants</span></code></a>(sample_sets)</p></td>
<td><p>Computes for every node the mean number of samples in each of the <cite>sample_sets</cite> that descend from that node, averaged over the portions of the genome for which the node is ancestral to <em>any</em> sample.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.Tajimas_D" title="tskit.TreeSequence.Tajimas_D"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.Tajimas_D</span></code></a>([sample_sets,...])</p></td>
<td><p>Computes Tajima's D of sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> in windows.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.trait_correlation" title="tskit.TreeSequence.trait_correlation"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.trait_correlation</span></code></a>(W[,windows,...])</p></td>
<td><p>Computes the mean squared correlations between each of the columns of <code class="docutils literal notranslate"><span class="pre">W</span></code> (the &quot;phenotypes&quot;) and inheritance along the tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.trait_covariance" title="tskit.TreeSequence.trait_covariance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.trait_covariance</span></code></a>(W[,windows,...])</p></td>
<td><p>Computes the mean squared covariances between each of the columns of <code class="docutils literal notranslate"><span class="pre">W</span></code> (the &quot;phenotypes&quot;) and inheritance along the tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.trait_linear_model" title="tskit.TreeSequence.trait_linear_model"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.trait_linear_model</span></code></a>(W[,Z,...])</p></td>
<td><p>Finds the relationship between trait and genotype after accounting for covariates.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.Y2" title="tskit.TreeSequence.Y2"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.Y2</span></code></a>(sample_sets[,indexes,...])</p></td>
<td><p>Computes the 'Y2' statistic between pairs of sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.Y3" title="tskit.TreeSequence.Y3"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.Y3</span></code></a>(sample_sets[,indexes,...])</p></td>
<td><p>Computes the 'Y' statistic between triples of sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
<dt>Comparative</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.kc_distance" title="tskit.TreeSequence.kc_distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.kc_distance</span></code></a>(other[,lambda_])</p></td>
<td><p>Returns the average <a class="reference internal" href="#tskit.Tree.kc_distance" title="tskit.Tree.kc_distance"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.kc_distance()</span></code></a> between pairs of trees along the sequence whose intervals overlap.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
</dl>
</section>
<section id="topological-analysis">
<span id="sec-python-api-tree-sequences-topological-analysis"></span><h4>Topological analysis<a class="headerlink" href="#topological-analysis" title="Link to this heading">#</a></h4>
<p>The topology of a tree in a tree sequence refers to the relationship among
samples ignoring branch lengths. Functionality as described in
<a class="reference internal" href="topological-analysis.html#sec-topological-analysis"><span class="std std-ref">Topological analysis</span></a> is mainly provided via
<a class="reference internal" href="#sec-python-api-trees-topological-analysis"><span class="std std-ref">methods on trees</span></a>, but more
efficient methods sometimes exist for entire tree sequences:</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.count_topologies" title="tskit.TreeSequence.count_topologies"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.count_topologies</span></code></a>([sample_sets])</p></td>
<td><p>Returns a generator that produces the same distribution of topologies as <a class="reference internal" href="#tskit.Tree.count_topologies" title="tskit.Tree.count_topologies"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.count_topologies()</span></code></a> but sequentially for every tree in a tree sequence.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="display">
<span id="sec-python-api-tree-sequences-display"></span><h4>Display<a class="headerlink" href="#display" title="Link to this heading">#</a></h4>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.draw_svg" title="tskit.TreeSequence.draw_svg"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.draw_svg</span></code></a>([path,size,x_scale,...])</p></td>
<td><p>Return an SVG representation of a tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.draw_text" title="tskit.TreeSequence.draw_text"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.draw_text</span></code></a>(*[,node_labels,...])</p></td>
<td><p>Create a text representation of a tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.__str__" title="tskit.TreeSequence.__str__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.__str__</span></code></a>()</p></td>
<td><p>Return a plain text summary of the contents of a tree sequence</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence._repr_html_" title="tskit.TreeSequence._repr_html_"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence._repr_html_</span></code></a>()</p></td>
<td><p>Return an html summary of a tree sequence.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="export">
<span id="sec-python-api-tree-sequences-export"></span><h4>Export<a class="headerlink" href="#export" title="Link to this heading">#</a></h4>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.as_fasta" title="tskit.TreeSequence.as_fasta"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.as_fasta</span></code></a>(**kwargs)</p></td>
<td><p>Return the result of <a class="reference internal" href="#tskit.TreeSequence.write_fasta" title="tskit.TreeSequence.write_fasta"><code class="xref py py-meth docutils literal notranslate"><span class="pre">write_fasta()</span></code></a> as a string.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.as_nexus" title="tskit.TreeSequence.as_nexus"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.as_nexus</span></code></a>(**kwargs)</p></td>
<td><p>Return the result of <a class="reference internal" href="#tskit.TreeSequence.write_nexus" title="tskit.TreeSequence.write_nexus"><code class="xref py py-meth docutils literal notranslate"><span class="pre">write_nexus()</span></code></a> as a string.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.dump_text" title="tskit.TreeSequence.dump_text"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.dump_text</span></code></a>([nodes,edges,...])</p></td>
<td><p>Writes a text representation of the tables underlying the tree sequence to the specified connections.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.to_macs" title="tskit.TreeSequence.to_macs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.to_macs</span></code></a>()</p></td>
<td><p>Return a <a class="reference external" href="https://github.com/gchen98/macs">macs encoding</a> of this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.write_fasta" title="tskit.TreeSequence.write_fasta"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.write_fasta</span></code></a>(file_or_path,*[,...])</p></td>
<td><p>Writes the <a class="reference internal" href="#tskit.TreeSequence.alignments" title="tskit.TreeSequence.alignments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">alignments()</span></code></a> for this tree sequence to file in <a class="reference external" href="https://en.wikipedia.org/wiki/FASTA_format">FASTA</a> format.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TreeSequence.write_nexus" title="tskit.TreeSequence.write_nexus"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.write_nexus</span></code></a>(file_or_path,*[,...])</p></td>
<td><p>Returns a <a class="reference external" href="https://en.wikipedia.org/wiki/Nexus_file">nexus encoding</a> of this tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TreeSequence.write_vcf" title="tskit.TreeSequence.write_vcf"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TreeSequence.write_vcf</span></code></a>(output[,ploidy,...])</p></td>
<td><p>Convert the genetic variation data in this tree sequence to Variant Call Format and write to the specified file-like object.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>
<section id="tree-api">
<span id="sec-python-api-trees"></span><h3><a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> API<a class="headerlink" href="#tree-api" title="Link to this heading">#</a></h3>
<p>A tree is an instance of the <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> class. These trees cannot exist
independently of the <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> from which they are generated.
Usually, therefore, a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instance is created by
<a class="reference internal" href="#sec-python-api-tree-sequences-obtaining-trees"><span class="std std-ref">Obtaining trees</span></a> from an existing tree
sequence (although it is also possible to generate a new instance of a
<a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> belonging to the same tree sequence using <a class="reference internal" href="#tskit.Tree.copy" title="tskit.Tree.copy"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.copy()</span></code></a>).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For efficiency, each instance of a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> is a state-machine
whose internal state corresponds to one of the trees in the parent tree sequence:
<a class="reference internal" href="#sec-python-api-trees-moving-to"><span class="std std-ref">Moving to other trees</span></a> in the tree sequence does not require a
new instance to be created, but simply the internal state to be changed.</p>
</div>
<section id="sec-python-api-trees-general-properties">
<span id="id1"></span><h4>General properties<a class="headerlink" href="#sec-python-api-trees-general-properties" title="Link to this heading">#</a></h4>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.tree_sequence" title="tskit.Tree.tree_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.tree_sequence</span></code></a></p></td>
<td><p>Returns the tree sequence that this tree is from.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.total_branch_length" title="tskit.Tree.total_branch_length"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.total_branch_length</span></code></a></p></td>
<td><p>Returns the sum of all the branch lengths in this tree (in units of time). This is equivalent to::.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.root_threshold" title="tskit.Tree.root_threshold"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.root_threshold</span></code></a></p></td>
<td><p>Returns the minimum number of samples that a node must be an ancestor of to be considered a potential root.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.virtual_root" title="tskit.Tree.virtual_root"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.virtual_root</span></code></a></p></td>
<td><p>The ID of the virtual root in this tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.num_edges" title="tskit.Tree.num_edges"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.num_edges</span></code></a></p></td>
<td><p>The total number of edges in this tree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.num_roots" title="tskit.Tree.num_roots"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.num_roots</span></code></a></p></td>
<td><p>The number of roots in this tree, as defined in the <a class="reference internal" href="#tskit.Tree.roots" title="tskit.Tree.roots"><code class="xref py py-attr docutils literal notranslate"><span class="pre">roots</span></code></a> attribute.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.has_single_root" title="tskit.Tree.has_single_root"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.has_single_root</span></code></a></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if this tree has a single root, <code class="docutils literal notranslate"><span class="pre">False</span></code> otherwise.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.has_multiple_roots" title="tskit.Tree.has_multiple_roots"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.has_multiple_roots</span></code></a></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if this tree has more than one root, <code class="docutils literal notranslate"><span class="pre">False</span></code> otherwise.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.root" title="tskit.Tree.root"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.root</span></code></a></p></td>
<td><p>The root of this tree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.roots" title="tskit.Tree.roots"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.roots</span></code></a></p></td>
<td><p>The list of roots in this tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.index" title="tskit.Tree.index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.index</span></code></a></p></td>
<td><p>Returns the index this tree occupies in the parent tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.interval" title="tskit.Tree.interval"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.interval</span></code></a></p></td>
<td><p>Returns the coordinates of the genomic interval that this tree represents the history of.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.span" title="tskit.Tree.span"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.span</span></code></a></p></td>
<td><p>Returns the genomic distance that this tree spans.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="creating-new-trees">
<span id="sec-python-api-trees-creating"></span><h4>Creating new trees<a class="headerlink" href="#creating-new-trees" title="Link to this heading">#</a></h4>
<p>It is sometimes useful to create an entirely new tree sequence consisting
of just a single tree (a one-tree sequence). The follow methods create such an
object and return a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instance corresponding to that tree.
The new tree sequence to which the tree belongs is available through the
<a class="reference internal" href="#tskit.Tree.tree_sequence" title="tskit.Tree.tree_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tree_sequence</span></code></a> property.</p>
<dl>
<dt>Creating a new tree</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.generate_balanced" title="tskit.Tree.generate_balanced"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.generate_balanced</span></code></a>(num_leaves,*[,...])</p></td>
<td><p>Generate a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> with the specified number of leaves that is maximally balanced.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.generate_comb" title="tskit.Tree.generate_comb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.generate_comb</span></code></a>(num_leaves,*[,span,...])</p></td>
<td><p>Generate a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> in which all internal nodes have two children and the left child is a leaf.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.generate_random_binary" title="tskit.Tree.generate_random_binary"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.generate_random_binary</span></code></a>(num_leaves,*[,...])</p></td>
<td><p>Generate a random binary <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> with <span class="math notranslate nohighlight">\(n\)</span> = <code class="docutils literal notranslate"><span class="pre">num_leaves</span></code> leaves with an equal probability of returning any topology and leaf label permutation among the <span class="math notranslate nohighlight">\((2n - 3)! / (2^{n - 2} (n - 2)!)\)</span> leaf-labelled binary trees.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.generate_star" title="tskit.Tree.generate_star"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.generate_star</span></code></a>(num_leaves,*[,span,...])</p></td>
<td><p>Generate a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> whose leaf nodes all have the same parent (i.e., a &quot;star&quot; tree).</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
<dt>Creating a new tree from an existing tree</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.split_polytomies" title="tskit.Tree.split_polytomies"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.split_polytomies</span></code></a>(*[,epsilon,method,...])</p></td>
<td><p>Return a new <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> where extra nodes and edges have been inserted so that any any node <code class="docutils literal notranslate"><span class="pre">u</span></code> with greater than 2 children --- a multifurcation or &quot;polytomy&quot; --- is resolved into successive bifurcations.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#tskit.Tree.unrank" title="tskit.Tree.unrank"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.unrank()</span></code></a> for creating a new one-tree sequence from its
<a class="reference internal" href="#sec-python-api-trees-topological-analysis"><span class="std std-ref">topological rank</span></a>.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Several of these methods are <a class="reference external" href="https://docs.python.org/3/library/functions.html#staticmethod" title="(in Python v3.14)"><code class="xref py py-func docutils literal notranslate"><span class="pre">static</span></code></a>, so should
be called e.g. as <code class="docutils literal notranslate"><span class="pre">tskit.Tree.generate_balanced(4)</span></code> rather than used on
a specific <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instance.</p>
</div>
</section>
<section id="node-measures">
<span id="sec-python-api-trees-node-measures"></span><h4>Node measures<a class="headerlink" href="#node-measures" title="Link to this heading">#</a></h4>
<p>Often it is useful to access information pertinant to a specific node or set of nodes
but which might also change from tree to tree in the tree sequence. Examples include
the encoding of the tree via <code class="docutils literal notranslate"><span class="pre">parent</span></code>, <code class="docutils literal notranslate"><span class="pre">left_child</span></code>, etc.
(see <a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">Tree structure</span></a>), the number of samples under a node,
or the most recent common ancestor (MRCA) of two nodes. This sort of information is
available via simple and high performance <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> methods</p>
<section id="simple-measures">
<span id="sec-python-api-trees-node-measures-simple"></span><h5>Simple measures<a class="headerlink" href="#simple-measures" title="Link to this heading">#</a></h5>
<p>These return a simple number, or (usually) short list of numbers relevant to a specific
node or limited set of nodes.</p>
<dl>
<dt>Node information</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.is_sample" title="tskit.Tree.is_sample"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.is_sample</span></code></a>(u)</p></td>
<td><p>Returns True if the specified node is a sample.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.is_isolated" title="tskit.Tree.is_isolated"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.is_isolated</span></code></a>(u)</p></td>
<td><p>Returns True if the specified node is isolated in this tree: that is it has no parents and no children (note that all isolated nodes in the tree are therefore also <a class="reference internal" href="#tskit.Tree.is_leaf" title="tskit.Tree.is_leaf"><code class="xref py py-meth docutils literal notranslate"><span class="pre">leaves</span></code></a>).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.is_leaf" title="tskit.Tree.is_leaf"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.is_leaf</span></code></a>(u)</p></td>
<td><p>Returns True if the specified node is a leaf.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.is_internal" title="tskit.Tree.is_internal"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.is_internal</span></code></a>(u)</p></td>
<td><p>Returns True if the specified node is not a leaf.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.parent" title="tskit.Tree.parent"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.parent</span></code></a>(u)</p></td>
<td><p>Returns the parent of the specified node.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.num_children" title="tskit.Tree.num_children"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.num_children</span></code></a>(u)</p></td>
<td><p>Returns the number of children of the specified node (i.e., <code class="docutils literal notranslate"><span class="pre">len(tree.children(u))</span></code>)</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.time" title="tskit.Tree.time"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.time</span></code></a>(u)</p></td>
<td><p>Returns the time of the specified node.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.branch_length" title="tskit.Tree.branch_length"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.branch_length</span></code></a>(u)</p></td>
<td><p>Returns the length of the branch (in units of time) joining the specified node to its parent. This is equivalent to::.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.depth" title="tskit.Tree.depth"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.depth</span></code></a>(u)</p></td>
<td><p>Returns the number of nodes on the path from <code class="docutils literal notranslate"><span class="pre">u</span></code> to a root, not including <code class="docutils literal notranslate"><span class="pre">u</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.population" title="tskit.Tree.population"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.population</span></code></a>(u)</p></td>
<td><p>Returns the population associated with the specified node.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.right_sib" title="tskit.Tree.right_sib"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.right_sib</span></code></a>(u)</p></td>
<td><p>Returns the sibling node to the right of u, or <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> if u does not have a right sibling.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.left_sib" title="tskit.Tree.left_sib"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.left_sib</span></code></a>(u)</p></td>
<td><p>Returns the sibling node to the left of u, or <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> if u does not have a left sibling.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.right_child" title="tskit.Tree.right_child"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.right_child</span></code></a>(u)</p></td>
<td><p>Returns the rightmost child of the specified node.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.left_child" title="tskit.Tree.left_child"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.left_child</span></code></a>(u)</p></td>
<td><p>Returns the leftmost child of the specified node.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.children" title="tskit.Tree.children"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.children</span></code></a>(u)</p></td>
<td><p>Returns the children of the specified node <code class="docutils literal notranslate"><span class="pre">u</span></code> as a tuple of integer node IDs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.edge" title="tskit.Tree.edge"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.edge</span></code></a>(u)</p></td>
<td><p>Returns the id of the edge encoding the relationship between <code class="docutils literal notranslate"><span class="pre">u</span></code> and its parent, or <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> if <code class="docutils literal notranslate"><span class="pre">u</span></code> is a root, virtual root or is not a node in the current tree.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
<dt>Descendant nodes</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.leaves" title="tskit.Tree.leaves"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.leaves</span></code></a>([u])</p></td>
<td><p>Returns an iterator over all the leaves in this tree that descend from the specified node.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.samples" title="tskit.Tree.samples"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.samples</span></code></a>([u])</p></td>
<td><p>Returns an iterator over the numerical IDs of all the sample nodes in this tree that are underneath the node with ID <code class="docutils literal notranslate"><span class="pre">u</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.num_samples" title="tskit.Tree.num_samples"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.num_samples</span></code></a>([u])</p></td>
<td><p>Returns the number of sample nodes in this tree underneath the specified node (including the node itself).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.num_tracked_samples" title="tskit.Tree.num_tracked_samples"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.num_tracked_samples</span></code></a>([u])</p></td>
<td><p>Returns the number of samples in the set specified in the <code class="docutils literal notranslate"><span class="pre">tracked_samples</span></code> parameter of the <a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.trees()</span></code></a> method underneath the specified node.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
</dl>
<p>Note that <a class="reference internal" href="#tskit.Tree.num_samples" title="tskit.Tree.num_samples"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.num_samples()</span></code></a> provides an efficient way to count samples under a node.
However, samples and leaves are not always equivalent: some samples may be internal nodes,
some leaves may not be samples (in unsimplified tree sequences), and the same node can be
a leaf in one tree but internal in another. While <code class="docutils literal notranslate"><span class="pre">tree.num_samples()</span></code> often equals the
leaf count (particularly in simplified tree sequences without internal samples), a strict
leaf count requires tree traversal, e.g. via <code class="docutils literal notranslate"><span class="pre">num_leaves</span> <span class="pre">=</span> <span class="pre">len(list(tree.leaves()))</span></code>.</p>
<dl>
<dt>Multiple nodes</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.is_descendant" title="tskit.Tree.is_descendant"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.is_descendant</span></code></a>(u,v)</p></td>
<td><p>Returns True if the specified node u is a descendant of node v and False otherwise.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.mrca" title="tskit.Tree.mrca"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.mrca</span></code></a>(*args)</p></td>
<td><p>Returns the most recent common ancestor of the specified nodes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.tmrca" title="tskit.Tree.tmrca"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.tmrca</span></code></a>(*args)</p></td>
<td><p>Returns the time of the most recent common ancestor of the specified nodes. This is equivalent to::.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
</dl>
</section>
<section id="array-access">
<span id="sec-python-api-trees-node-measures-array"></span><h5>Array access<a class="headerlink" href="#array-access" title="Link to this heading">#</a></h5>
<p>These all return a numpy array whose length corresponds to
the total number of nodes in the tree sequence. They provide direct access
to the underlying memory structures, and are thus very efficient, providing a
high performance interface which can be used in conjunction with the equivalent
<a class="reference internal" href="#sec-python-api-trees-traversal"><span class="std std-ref">traversal methods</span></a>.</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.parent_array" title="tskit.Tree.parent_array"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.parent_array</span></code></a></p></td>
<td><p>A numpy array (dtype=np.int32) encoding the parent of each node in this tree, such that <code class="docutils literal notranslate"><span class="pre">tree.parent_array[u]</span> <span class="pre">==</span> <span class="pre">tree.parent(u)</span></code> for all <code class="docutils literal notranslate"><span class="pre">0</span> <span class="pre">&lt;=</span> <span class="pre">u</span> <span class="pre">&lt;=</span> <span class="pre">ts.num_nodes</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.left_child_array" title="tskit.Tree.left_child_array"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.left_child_array</span></code></a></p></td>
<td><p>A numpy array (dtype=np.int32) encoding the left child of each node in this tree, such that <code class="docutils literal notranslate"><span class="pre">tree.left_child_array[u]</span> <span class="pre">==</span> <span class="pre">tree.left_child(u)</span></code> for all <code class="docutils literal notranslate"><span class="pre">0</span> <span class="pre">&lt;=</span> <span class="pre">u</span> <span class="pre">&lt;=</span> <span class="pre">ts.num_nodes</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.right_child_array" title="tskit.Tree.right_child_array"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.right_child_array</span></code></a></p></td>
<td><p>A numpy array (dtype=np.int32) encoding the right child of each node in this tree, such that <code class="docutils literal notranslate"><span class="pre">tree.right_child_array[u]</span> <span class="pre">==</span> <span class="pre">tree.right_child(u)</span></code> for all <code class="docutils literal notranslate"><span class="pre">0</span> <span class="pre">&lt;=</span> <span class="pre">u</span> <span class="pre">&lt;=</span> <span class="pre">ts.num_nodes</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.left_sib_array" title="tskit.Tree.left_sib_array"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.left_sib_array</span></code></a></p></td>
<td><p>A numpy array (dtype=np.int32) encoding the left sib of each node in this tree, such that <code class="docutils literal notranslate"><span class="pre">tree.left_sib_array[u]</span> <span class="pre">==</span> <span class="pre">tree.left_sib(u)</span></code> for all <code class="docutils literal notranslate"><span class="pre">0</span> <span class="pre">&lt;=</span> <span class="pre">u</span> <span class="pre">&lt;=</span> <span class="pre">ts.num_nodes</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.right_sib_array" title="tskit.Tree.right_sib_array"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.right_sib_array</span></code></a></p></td>
<td><p>A numpy array (dtype=np.int32) encoding the right sib of each node in this tree, such that <code class="docutils literal notranslate"><span class="pre">tree.right_sib_array[u]</span> <span class="pre">==</span> <span class="pre">tree.right_sib(u)</span></code> for all <code class="docutils literal notranslate"><span class="pre">0</span> <span class="pre">&lt;=</span> <span class="pre">u</span> <span class="pre">&lt;=</span> <span class="pre">ts.num_nodes</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.num_children_array" title="tskit.Tree.num_children_array"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.num_children_array</span></code></a></p></td>
<td><p>A numpy array (dtype=np.int32) encoding the number of children of each node in this tree, such that <code class="docutils literal notranslate"><span class="pre">tree.num_children_array[u]</span> <span class="pre">==</span> <span class="pre">tree.num_children(u)</span></code> for all <code class="docutils literal notranslate"><span class="pre">0</span> <span class="pre">&lt;=</span> <span class="pre">u</span> <span class="pre">&lt;=</span> <span class="pre">ts.num_nodes</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.edge_array" title="tskit.Tree.edge_array"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.edge_array</span></code></a></p></td>
<td><p>A numpy array (dtype=np.int32) of edge ids encoding the relationship between the child node <code class="docutils literal notranslate"><span class="pre">u</span></code> and its parent, such that <code class="docutils literal notranslate"><span class="pre">tree.edge_array[u]</span> <span class="pre">==</span> <span class="pre">tree.edge(u)</span></code> for all <code class="docutils literal notranslate"><span class="pre">0</span> <span class="pre">&lt;=</span> <span class="pre">u</span> <span class="pre">&lt;=</span> <span class="pre">ts.num_nodes</span></code>.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>
<section id="tree-traversal">
<span id="sec-python-api-trees-traversal"></span><h4>Tree traversal<a class="headerlink" href="#tree-traversal" title="Link to this heading">#</a></h4>
<p>Moving around within a tree usually involves visiting the tree nodes in some sort of
order. Often, given a particular order, it is convenient to iterate over each node
using the <a class="reference internal" href="#tskit.Tree.nodes" title="tskit.Tree.nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.nodes()</span></code></a> method. However, for high performance algorithms, it
may be more convenient to access the node indices for a particular order as
an array, and use this, for example, to index into one of the node arrays (see
<a class="reference internal" href="topological-analysis.html#sec-topological-analysis-traversal"><span class="std std-ref">Visiting nodes</span></a>). Note that the most efficient of these
methods is <a class="reference internal" href="#tskit.Tree.preorder" title="tskit.Tree.preorder"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.preorder()</span></code></a>.</p>
<dl>
<dt>Iterator access</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.nodes" title="tskit.Tree.nodes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.nodes</span></code></a>([root,order])</p></td>
<td><p>Returns an iterator over the node IDs reachable from the specified node in this tree in the specified traversal order.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.ancestors" title="tskit.Tree.ancestors"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.ancestors</span></code></a>(u)</p></td>
<td><p>Returns an iterator over the ancestors of node <code class="docutils literal notranslate"><span class="pre">u</span></code> in this tree (i.e. the chain of parents from <code class="docutils literal notranslate"><span class="pre">u</span></code> to the root).</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
<dt>Array access</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.postorder" title="tskit.Tree.postorder"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.postorder</span></code></a>([u])</p></td>
<td><p>Returns a numpy array of node ids in <a class="reference external" href="https://en.wikipedia.org/wiki/Tree_traversal##Post-order_(LRN)">postorder</a>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.preorder" title="tskit.Tree.preorder"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.preorder</span></code></a>([u])</p></td>
<td><p>Returns a numpy array of node ids in <a class="reference external" href="https://en.wikipedia.org/wiki/Tree_traversal#Pre-order_(NLR)">preorder</a>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.timeasc" title="tskit.Tree.timeasc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.timeasc</span></code></a>([u])</p></td>
<td><p>Returns a numpy array of node ids.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.timedesc" title="tskit.Tree.timedesc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.timedesc</span></code></a>([u])</p></td>
<td><p>Returns a numpy array of node ids.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
</dl>
</section>
<section id="sec-python-api-trees-topological-analysis">
<span id="id2"></span><h4>Topological analysis<a class="headerlink" href="#sec-python-api-trees-topological-analysis" title="Link to this heading">#</a></h4>
<p>The topology of a tree refers to the simple relationship among samples
(i.e. ignoring branch lengths), see <a class="reference internal" href="topological-analysis.html#sec-combinatorics"><span class="std std-ref">Identifying and counting topologies</span></a> for more details. These
methods provide ways to enumerate and count tree topologies.</p>
<p>Briefly, the position of a tree in the enumeration <code class="docutils literal notranslate"><span class="pre">all_trees</span></code> can be obtained using
the trees <a class="reference internal" href="#tskit.Tree.rank" title="tskit.Tree.rank"><code class="xref py py-meth docutils literal notranslate"><span class="pre">rank()</span></code></a> method. Inversely, a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> can be constructed
from a position in the enumeration with <a class="reference internal" href="#tskit.Tree.unrank" title="tskit.Tree.unrank"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.unrank()</span></code></a>.</p>
<dl>
<dt>Methods of a tree</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.rank" title="tskit.Tree.rank"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.rank</span></code></a>()</p></td>
<td><p>Produce the rank of this tree in the enumeration of all leaf-labelled trees of n leaves.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.count_topologies" title="tskit.Tree.count_topologies"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.count_topologies</span></code></a>([sample_sets])</p></td>
<td><p>Calculates the distribution of embedded topologies for every combination of the sample sets in <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
<dt>Functions and static methods</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.unrank" title="tskit.Tree.unrank"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.unrank</span></code></a>(num_leaves,rank,*[,span,...])</p></td>
<td><p>Reconstruct the tree of the given <code class="docutils literal notranslate"><span class="pre">rank</span></code> (see <a class="reference internal" href="#tskit.Tree.rank" title="tskit.Tree.rank"><code class="xref py py-meth docutils literal notranslate"><span class="pre">tskit.Tree.rank()</span></code></a>) with <code class="docutils literal notranslate"><span class="pre">num_leaves</span></code> leaves.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.all_tree_shapes" title="tskit.all_tree_shapes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">all_tree_shapes</span></code></a>(num_leaves[,span])</p></td>
<td><p>Generates all unique shapes of trees with <code class="docutils literal notranslate"><span class="pre">num_leaves</span></code> leaves.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.all_tree_labellings" title="tskit.all_tree_labellings"><code class="xref py py-obj docutils literal notranslate"><span class="pre">all_tree_labellings</span></code></a>(tree[,span])</p></td>
<td><p>Generates all unique labellings of the leaves of a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.Tree</span></code></a>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.all_trees" title="tskit.all_trees"><code class="xref py py-obj docutils literal notranslate"><span class="pre">all_trees</span></code></a>(num_leaves[,span])</p></td>
<td><p>Generates all unique leaf-labelled trees with <code class="docutils literal notranslate"><span class="pre">num_leaves</span></code> leaves.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
</dl>
</section>
<section id="comparing-trees">
<span id="sec-python-api-trees-comparing"></span><h4>Comparing trees<a class="headerlink" href="#comparing-trees" title="Link to this heading">#</a></h4>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.kc_distance" title="tskit.Tree.kc_distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.kc_distance</span></code></a>(other[,lambda_])</p></td>
<td><p>Returns the Kendall-Colijn distance between the specified pair of trees.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.rf_distance" title="tskit.Tree.rf_distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.rf_distance</span></code></a>(other)</p></td>
<td><p>Returns the (unweighted) Robinson-Foulds distance between the specified pair of trees, where corresponding samples between the two trees are identified by node ID.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="balance-imbalance-indices">
<span id="sec-python-api-trees-balance"></span><h4>Balance/imbalance indices<a class="headerlink" href="#balance-imbalance-indices" title="Link to this heading">#</a></h4>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.colless_index" title="tskit.Tree.colless_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.colless_index</span></code></a>()</p></td>
<td><p>Returns the <a class="reference external" href="https://treebalance.wordpress.com/colless-index/">Colless imbalance index</a> for this tree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.sackin_index" title="tskit.Tree.sackin_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.sackin_index</span></code></a>()</p></td>
<td><p>Returns the <a class="reference external" href="https://treebalance.wordpress.com/sackin-index/">Sackin imbalance index</a> for this tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.b1_index" title="tskit.Tree.b1_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.b1_index</span></code></a>()</p></td>
<td><p>Returns the <a class="reference external" href="https://treebalance.wordpress.com/b-index/">B1 balance index</a> for this tree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.b2_index" title="tskit.Tree.b2_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.b2_index</span></code></a>([base])</p></td>
<td><p>Returns the <a class="reference external" href="https://treebalance.wordpress.com/b-index/">B2 balance index</a> this tree.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="sites-and-mutations">
<span id="sec-python-api-trees-sites-mutations"></span><h4>Sites and mutations<a class="headerlink" href="#sites-and-mutations" title="Link to this heading">#</a></h4>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.sites" title="tskit.Tree.sites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.sites</span></code></a>()</p></td>
<td><p>Returns an iterator over all the <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">sites</span></a> in this tree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.num_sites" title="tskit.Tree.num_sites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.num_sites</span></code></a></p></td>
<td><p>Returns the number of sites on this tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.mutations" title="tskit.Tree.mutations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.mutations</span></code></a>()</p></td>
<td><p>Returns an iterator over all the <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">mutations</span></a> in this tree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.num_mutations" title="tskit.Tree.num_mutations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.num_mutations</span></code></a></p></td>
<td><p>Returns the total number of mutations across all sites on this tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.map_mutations" title="tskit.Tree.map_mutations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.map_mutations</span></code></a>(genotypes,alleles[,...])</p></td>
<td><p>Given observations for the samples in this tree described by the specified set of genotypes and alleles, return a parsimonious set of state transitions explaining these observations.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="moving-to-other-trees">
<span id="sec-python-api-trees-moving-to"></span><h4>Moving to other trees<a class="headerlink" href="#moving-to-other-trees" title="Link to this heading">#</a></h4>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.next" title="tskit.Tree.next"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.next</span></code></a>()</p></td>
<td><p>Seeks to the next tree in the sequence. If the tree is in the initial null state we seek to the first tree (equivalent to calling <a class="reference internal" href="#tskit.Tree.first" title="tskit.Tree.first"><code class="xref py py-meth docutils literal notranslate"><span class="pre">first()</span></code></a>). Calling <code class="docutils literal notranslate"><span class="pre">next</span></code> on the last tree in the sequence results in the tree being cleared back into the null initial state (equivalent to calling <a class="reference internal" href="#tskit.Tree.clear" title="tskit.Tree.clear"><code class="xref py py-meth docutils literal notranslate"><span class="pre">clear()</span></code></a>). The return value of the function indicates whether the tree is in a non-null state, and can be used to loop over the trees::.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.prev" title="tskit.Tree.prev"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.prev</span></code></a>()</p></td>
<td><p>Seeks to the previous tree in the sequence. If the tree is in the initial null state we seek to the last tree (equivalent to calling <a class="reference internal" href="#tskit.Tree.last" title="tskit.Tree.last"><code class="xref py py-meth docutils literal notranslate"><span class="pre">last()</span></code></a>). Calling <code class="docutils literal notranslate"><span class="pre">prev</span></code> on the first tree in the sequence results in the tree being cleared back into the null initial state (equivalent to calling <a class="reference internal" href="#tskit.Tree.clear" title="tskit.Tree.clear"><code class="xref py py-meth docutils literal notranslate"><span class="pre">clear()</span></code></a>). The return value of the function indicates whether the tree is in a non-null state, and can be used to loop over the trees::.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.first" title="tskit.Tree.first"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.first</span></code></a>()</p></td>
<td><p>Seeks to the first tree in the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.last" title="tskit.Tree.last"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.last</span></code></a>()</p></td>
<td><p>Seeks to the last tree in the sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.seek" title="tskit.Tree.seek"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.seek</span></code></a>(position[,skip])</p></td>
<td><p>Sets the state to represent the tree that covers the specified position in the parent tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.seek_index" title="tskit.Tree.seek_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.seek_index</span></code></a>(index[,skip])</p></td>
<td><p>Sets the state to represent the tree at the specified index in the parent tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.clear" title="tskit.Tree.clear"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.clear</span></code></a>()</p></td>
<td><p>Resets this tree back to the initial null state.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="id3">
<h4>Display<a class="headerlink" href="#id3" title="Link to this heading">#</a></h4>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.draw_svg" title="tskit.Tree.draw_svg"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.draw_svg</span></code></a>([path,size,time_scale,...])</p></td>
<td><p>Return an SVG representation of a single tree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.draw_text" title="tskit.Tree.draw_text"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.draw_text</span></code></a>([orientation,node_labels,...])</p></td>
<td><p>Create a text representation of a tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.__str__" title="tskit.Tree.__str__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.__str__</span></code></a>()</p></td>
<td><p>Return a plain text summary of a tree in a tree sequence</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree._repr_html_" title="tskit.Tree._repr_html_"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree._repr_html_</span></code></a>()</p></td>
<td><p>Return an html summary of a tree in a tree sequence.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="id4">
<h4>Export<a class="headerlink" href="#id4" title="Link to this heading">#</a></h4>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.Tree.as_dict_of_dicts" title="tskit.Tree.as_dict_of_dicts"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.as_dict_of_dicts</span></code></a>()</p></td>
<td><p>Convert tree to dict of dicts for conversion to a <a class="reference external" href="https://networkx.github.io/documentation/stable/reference/classes/digraph.html">networkx graph</a>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.Tree.as_newick" title="tskit.Tree.as_newick"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Tree.as_newick</span></code></a>(*[,root,precision,...])</p></td>
<td><p>Returns a <a class="reference external" href="https://en.wikipedia.org/wiki/Newick_format">newick encoding</a> of this tree. For example, a binary tree with 3 leaves generated by <a class="reference internal" href="#tskit.Tree.generate_balanced" title="tskit.Tree.generate_balanced"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.generate_balanced(3)</span></code></a> encodes as::.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>
</section>
<section id="tables-and-table-collections">
<span id="sec-tables-api"></span><h2>Tables and Table Collections<a class="headerlink" href="#tables-and-table-collections" title="Link to this heading">#</a></h2>
<p>The information required to construct a tree sequence is stored in a collection
of <em>tables</em>, each defining a different aspect of the structure of a tree
sequence. These tables are described individually in
<a class="reference internal" href="#sec-tables-api-table"><span class="std std-ref">the next section</span></a>. However, these are interrelated,
and so many operations work
on the entire collection of tables, known as a <em>table collection</em>.</p>
<section id="tablecollection-api">
<span id="sec-tables-api-table-collection"></span><h3><code class="docutils literal notranslate"><span class="pre">TableCollection</span></code> API<a class="headerlink" href="#tablecollection-api" title="Link to this heading">#</a></h3>
<p>The <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a> and <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> classes are
deeply related. A <code class="docutils literal notranslate"><span class="pre">TreeSequence</span></code> instance is based on the information
encoded in a <code class="docutils literal notranslate"><span class="pre">TableCollection</span></code>. Tree sequences are <strong>immutable</strong>, and
provide methods for obtaining trees from the sequence. A <code class="docutils literal notranslate"><span class="pre">TableCollection</span></code>
is <strong>mutable</strong>, and does not have any methods for obtaining trees.
The <code class="docutils literal notranslate"><span class="pre">TableCollection</span></code> class thus allows creation and modification of
tree sequences (see the <a class="reference external" href="https://tskit.dev/tutorials/tables_and_editing.html#sec-tables" title="(in Project name not set)"><span>Tables and editing</span></a> tutorial).</p>
<section id="id5">
<h4>General properties<a class="headerlink" href="#id5" title="Link to this heading">#</a></h4>
<p>Specific <a class="reference internal" href="#sec-tables-api-table"><span class="std std-ref">tables</span></a> in the <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>
are be accessed using the plural version of their name, so that, for instance, the
individual table can be accessed using <code class="docutils literal notranslate"><span class="pre">table_collection.individuals</span></code>. A table
collection also has other properties containing, for example, number of bytes taken
to store it and the top-level metadata associated with the tree sequence as a whole.</p>
<dl>
<dt>Table access</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.individuals" title="tskit.TableCollection.individuals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.individuals</span></code></a></p></td>
<td><p>The <a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">Individual Table</span></a> in this collection.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.nodes" title="tskit.TableCollection.nodes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.nodes</span></code></a></p></td>
<td><p>The <a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">Node Table</span></a> in this collection.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.edges" title="tskit.TableCollection.edges"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.edges</span></code></a></p></td>
<td><p>The <a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">Edge Table</span></a> in this collection.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.migrations" title="tskit.TableCollection.migrations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.migrations</span></code></a></p></td>
<td><p>The <a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">Migration Table</span></a> in this collection</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.sites" title="tskit.TableCollection.sites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.sites</span></code></a></p></td>
<td><p>The <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">Site Table</span></a> in this collection.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.mutations" title="tskit.TableCollection.mutations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.mutations</span></code></a></p></td>
<td><p>The <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">Mutation Table</span></a> in this collection.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.populations" title="tskit.TableCollection.populations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.populations</span></code></a></p></td>
<td><p>The <a class="reference internal" href="data-model.html#sec-population-table-definition"><span class="std std-ref">Population Table</span></a> in this collection.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.provenances" title="tskit.TableCollection.provenances"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.provenances</span></code></a></p></td>
<td><p>The <a class="reference internal" href="data-model.html#sec-provenance-table-definition"><span class="std std-ref">Provenance Table</span></a> in this collection.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
<dt>Other properties</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.file_uuid" title="tskit.TableCollection.file_uuid"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.file_uuid</span></code></a></p></td>
<td><p>The UUID for the file this TableCollection is derived from, or None if not derived from a file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.indexes" title="tskit.TableCollection.indexes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.indexes</span></code></a></p></td>
<td><p>The edge insertion and removal indexes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.nbytes" title="tskit.TableCollection.nbytes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.nbytes</span></code></a></p></td>
<td><p>Returns the total number of bytes required to store the data in this table collection.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.table_name_map" title="tskit.TableCollection.table_name_map"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.table_name_map</span></code></a></p></td>
<td><p>Returns a dictionary mapping table names to the corresponding table instances.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.metadata" title="tskit.TableCollection.metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.metadata</span></code></a></p></td>
<td><p>The decoded metadata for this object.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.metadata_bytes" title="tskit.TableCollection.metadata_bytes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.metadata_bytes</span></code></a></p></td>
<td><p>The raw bytes of metadata for this TableCollection</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.metadata_schema" title="tskit.TableCollection.metadata_schema"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.metadata_schema</span></code></a></p></td>
<td><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this object.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.sequence_length" title="tskit.TableCollection.sequence_length"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.sequence_length</span></code></a></p></td>
<td><p>The sequence length defining the coordinate space.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.time_units" title="tskit.TableCollection.time_units"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.time_units</span></code></a></p></td>
<td><p>The units used for the time dimension of this TableCollection</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
</dl>
</section>
<section id="transformation">
<span id="sec-tables-api-transformation"></span><h4>Transformation<a class="headerlink" href="#transformation" title="Link to this heading">#</a></h4>
<p>These methods act in-place to transform the contents of a <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>,
either by modifying the underlying tables (removing, editing, or adding to them) or
by adjusting the table collection so that it meets the
<a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">Valid tree sequence requirements</span></a>.</p>
<section id="modification">
<span id="sec-tables-api-modification"></span><h5>Modification<a class="headerlink" href="#modification" title="Link to this heading">#</a></h5>
<p>These methods modify the data stored in a <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>. They also have
<a class="reference internal" href="#sec-python-api-tree-sequences-modification"><span class="std std-ref">equivalant TreeSequence versions</span></a>
(unlike the methods described below those do <em>not</em> operate in place, but rather act in
a functional way, returning a new tree sequence while leaving the original unchanged).</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.clear" title="tskit.TableCollection.clear"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.clear</span></code></a>([clear_provenance,...])</p></td>
<td><p>Remove all rows of the data tables, optionally remove provenance, metadata schemas and ts-level metadata.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.simplify" title="tskit.TableCollection.simplify"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.simplify</span></code></a>([samples,...])</p></td>
<td><p>Simplifies the tables in place to retain only the information necessary to reconstruct the tree sequence describing the given <code class="docutils literal notranslate"><span class="pre">samples</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.subset" title="tskit.TableCollection.subset"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.subset</span></code></a>(nodes[,...])</p></td>
<td><p>Modifies the tables in place to contain only the entries referring to the provided list of node IDs, with nodes reordered according to the order they appear in the list.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.delete_intervals" title="tskit.TableCollection.delete_intervals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.delete_intervals</span></code></a>(intervals)</p></td>
<td><p>Delete all information from this set of tables which lies <em>within</em> the specified list of genomic intervals.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.keep_intervals" title="tskit.TableCollection.keep_intervals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.keep_intervals</span></code></a>(intervals[,...])</p></td>
<td><p>Delete all information from this set of tables which lies <em>outside</em> the specified list of genomic intervals.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.delete_sites" title="tskit.TableCollection.delete_sites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.delete_sites</span></code></a>(site_ids[,...])</p></td>
<td><p>Remove the specified sites entirely from the sites and mutations tables in this collection.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.trim" title="tskit.TableCollection.trim"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.trim</span></code></a>([record_provenance])</p></td>
<td><p>Trim away any empty regions on the right and left of the tree sequence encoded by these tables.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.shift" title="tskit.TableCollection.shift"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.shift</span></code></a>(value,*[,...])</p></td>
<td><p>Shift the coordinate system (used by edges, sites, and migrations) of this TableCollection by a given value.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.union" title="tskit.TableCollection.union"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.union</span></code></a>(other,node_mapping[,...])</p></td>
<td><p>Modifies the table collection in place by adding the non-shared portions of <code class="docutils literal notranslate"><span class="pre">other</span></code> to itself.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.delete_older" title="tskit.TableCollection.delete_older"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.delete_older</span></code></a>(time)</p></td>
<td><p>Deletes edge, mutation and migration information at least as old as the specified time.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="creating-a-valid-tree-sequence">
<span id="sec-tables-api-creating-valid-tree-sequence"></span><h5>Creating a valid tree sequence<a class="headerlink" href="#creating-a-valid-tree-sequence" title="Link to this heading">#</a></h5>
<p>These methods can be used to help reorganise or rationalise the
<a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a> so that it is in the form
<a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">required</span></a> for
it to be <a class="reference internal" href="#tskit.TableCollection.tree_sequence" title="tskit.TableCollection.tree_sequence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">converted</span></code></a>
into a <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>. This may require sorting the tables,
ensuring they are logically consistent, and adding <a class="reference internal" href="data-model.html#sec-table-indexes"><span class="std std-ref">Table indexes</span></a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>These methods are not guaranteed to make valid a <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a> which is
logically inconsistent, for example if multiple edges have the same child at a
given position on the genome or if non-existent node IDs are referenced.</p>
</div>
<dl>
<dt>Sorting</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.sort</span></code></a>([edge_start,...])</p></td>
<td><p>Sorts the tables in place.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.sort_individuals" title="tskit.TableCollection.sort_individuals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.sort_individuals</span></code></a>()</p></td>
<td><p>Sorts the individual table in place, so that parents come before children, and the parent column is remapped as required.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.canonicalise" title="tskit.TableCollection.canonicalise"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.canonicalise</span></code></a>([...])</p></td>
<td><p>This puts the tables in <em>canonical</em> form, imposing a stricter order on the tables than <a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">required</span></a> for a valid tree sequence.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
<dt>Logical consistency</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.compute_mutation_parents" title="tskit.TableCollection.compute_mutation_parents"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.compute_mutation_parents</span></code></a>()</p></td>
<td><p>Modifies the tables in place, computing the <code class="docutils literal notranslate"><span class="pre">parent</span></code> column of the mutation table.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.compute_mutation_times" title="tskit.TableCollection.compute_mutation_times"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.compute_mutation_times</span></code></a>()</p></td>
<td><p>Modifies the tables in place, computing valid values for the <code class="docutils literal notranslate"><span class="pre">time</span></code> column of the mutation table.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.deduplicate_sites" title="tskit.TableCollection.deduplicate_sites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.deduplicate_sites</span></code></a>()</p></td>
<td><p>Modifies the tables in place, removing entries in the site table with duplicate <code class="docutils literal notranslate"><span class="pre">position</span></code> (and keeping only the <em>first</em> entry for each site), and renumbering the <code class="docutils literal notranslate"><span class="pre">site</span></code> column of the mutation table appropriately.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
<dt>Indexing</dt><dd><div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.has_index" title="tskit.TableCollection.has_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.has_index</span></code></a>()</p></td>
<td><p>Returns True if this TableCollection is indexed.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.build_index" title="tskit.TableCollection.build_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.build_index</span></code></a>()</p></td>
<td><p>Builds an index on this TableCollection.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.drop_index" title="tskit.TableCollection.drop_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.drop_index</span></code></a>()</p></td>
<td><p>Drops any indexes present on this table collection.</p></td>
</tr>
</tbody>
</table>
</div>
</dd>
</dl>
</section>
</section>
<section id="miscellaneous-methods">
<h4>Miscellaneous methods<a class="headerlink" href="#miscellaneous-methods" title="Link to this heading">#</a></h4>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.copy" title="tskit.TableCollection.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.copy</span></code></a>()</p></td>
<td><p>Returns a deep copy of this TableCollection.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.equals" title="tskit.TableCollection.equals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.equals</span></code></a>(other,*[,...])</p></td>
<td><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.link_ancestors" title="tskit.TableCollection.link_ancestors"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.link_ancestors</span></code></a>(samples,...)</p></td>
<td><p>Returns an <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a> instance describing a subset of the genealogical relationships between the nodes in <code class="docutils literal notranslate"><span class="pre">samples</span></code> and <code class="docutils literal notranslate"><span class="pre">ancestors</span></code>.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="id6">
<h4>Export<a class="headerlink" href="#id6" title="Link to this heading">#</a></h4>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.TableCollection.tree_sequence" title="tskit.TableCollection.tree_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.tree_sequence</span></code></a>()</p></td>
<td><p>Returns a <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> instance from the tables defined in this <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>, building the required indexes if they have not yet been created by <a class="reference internal" href="#tskit.TableCollection.build_index" title="tskit.TableCollection.build_index"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_index()</span></code></a>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.TableCollection.dump" title="tskit.TableCollection.dump"><code class="xref py py-obj docutils literal notranslate"><span class="pre">TableCollection.dump</span></code></a>(file_or_path)</p></td>
<td><p>Writes the table collection to the specified path or file object.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>
<section id="table-apis">
<span id="sec-tables-api-table"></span><h3>Table APIs<a class="headerlink" href="#table-apis" title="Link to this heading">#</a></h3>
<p>Here we outline the table classes and the common methods and variables available for
each. For description and definition of each tables meaning
and use, see <a class="reference internal" href="data-model.html#sec-table-definitions"><span class="std std-ref">the table definitions</span></a>.</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.IndividualTable" title="tskit.IndividualTable"><code class="xref py py-obj docutils literal notranslate"><span class="pre">IndividualTable</span></code></a>([max_rows_increment,ll_table])</p></td>
<td><p>A table defining the individuals in a tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-obj docutils literal notranslate"><span class="pre">NodeTable</span></code></a>([max_rows_increment,ll_table])</p></td>
<td><p>A table defining the nodes in a tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-obj docutils literal notranslate"><span class="pre">EdgeTable</span></code></a>([max_rows_increment,ll_table])</p></td>
<td><p>A table defining the edges in a tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.MigrationTable" title="tskit.MigrationTable"><code class="xref py py-obj docutils literal notranslate"><span class="pre">MigrationTable</span></code></a>([max_rows_increment,ll_table])</p></td>
<td><p>A table defining the migrations in a tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><code class="xref py py-obj docutils literal notranslate"><span class="pre">SiteTable</span></code></a>([max_rows_increment,ll_table])</p></td>
<td><p>A table defining the sites in a tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.MutationTable" title="tskit.MutationTable"><code class="xref py py-obj docutils literal notranslate"><span class="pre">MutationTable</span></code></a>([max_rows_increment,ll_table])</p></td>
<td><p>A table defining the mutations in a tree sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.PopulationTable" title="tskit.PopulationTable"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PopulationTable</span></code></a>([max_rows_increment,ll_table])</p></td>
<td><p>A table defining the populations referred to in a tree sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.ProvenanceTable" title="tskit.ProvenanceTable"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ProvenanceTable</span></code></a>([max_rows_increment,ll_table])</p></td>
<td><p>A table recording the provenance (i.e., history) of this table, so that the origin of the underlying data and sequence of subsequent operations can be traced.</p></td>
</tr>
</tbody>
</table>
</div>
<section id="accessing-table-data">
<span id="sec-tables-api-accessing-table-data"></span><h4>Accessing table data<a class="headerlink" href="#accessing-table-data" title="Link to this heading">#</a></h4>
<p>The tables API provides an efficient way of working
with and interchanging <a class="reference internal" href="data-model.html#sec-data-model"><span class="std std-ref">tree sequence data</span></a>. Each table class
(e.g, <a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a>, <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a>, <a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code></a>) has a specific set
of columns with fixed types, and a set of methods for setting and getting the data
in these columns. The number of rows in the table <code class="docutils literal notranslate"><span class="pre">t</span></code> is given by <code class="docutils literal notranslate"><span class="pre">len(t)</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">tskit</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">EdgeTable</span><span class="p">()</span>
<span class="n">t</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">child</span><span class="o">=</span><span class="mi">11</span><span class="p">)</span>
<span class="n">t</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="mi">9</span><span class="p">,</span> <span class="n">child</span><span class="o">=</span><span class="mi">11</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The table contains&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">t</span><span class="p">),</span> <span class="s2">&quot;rows&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>The table contains 2 rows

idleftrightparentchildmetadata

0    0    1    10   11        
1    1    2     9   11        

</pre></div>
</div>
</div>
</div>
<p>Each table supports accessing the data either by row or column. To access the data in
a <em>column</em>, we can use standard attribute access which will
return a copy of the column data as a numpy array:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">t</span><span class="o">.</span><span class="n">left</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([0., 1.])
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">t</span><span class="o">.</span><span class="n">parent</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([10,  9], dtype=int32)
</pre></div>
</div>
</div>
</div>
<p>To access the data in a <em>row</em>, say row number <code class="docutils literal notranslate"><span class="pre">j</span></code> in table <code class="docutils literal notranslate"><span class="pre">t</span></code>, simply use <code class="docutils literal notranslate"><span class="pre">t[j]</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">t</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>EdgeTableRow(left=0.0, right=1.0, parent=10, child=11, metadata=b&#39;&#39;)
</pre></div>
</div>
</div>
</div>
<p>This also works as expected with negative <code class="docutils literal notranslate"><span class="pre">j</span></code>, counting rows from the end of the table</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">t</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>EdgeTableRow(left=1.0, right=2.0, parent=9, child=11, metadata=b&#39;&#39;)
</pre></div>
</div>
</div>
</div>
<p>The returned row has attributes allowing contents to be accessed by name, e.g.
<code class="docutils literal notranslate"><span class="pre">site_table[0].position</span></code>, <code class="docutils literal notranslate"><span class="pre">site_table[0].ancestral_state</span></code>, <code class="docutils literal notranslate"><span class="pre">site_table[0].metadata</span></code>
etc.:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">t</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">right</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2.0
</pre></div>
</div>
</div>
</div>
<p>Row attributes cannot be modified directly. Instead, the <code class="docutils literal notranslate"><span class="pre">replace</span></code> method of a row
object can be used to create a new row with one or more changed column
values, which can then be used to replace the original. For example:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">t</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">t</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">child</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
idleftrightparentchildmetadata

0    0    1    10   11        
1    1    3     9    4        

</pre></div>
</div>
</div>
</div>
<p>Tables also support the <a class="reference external" href="https://docs.python.org/3/library/pickle.html#module-pickle" title="(in Python v3.14)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pickle</span></code></a> protocol, and so can be easily serialised and
deserialised. This can be useful, for example, when performing parallel computations
using the <a class="reference external" href="https://docs.python.org/3/library/multiprocessing.html#module-multiprocessing" title="(in Python v3.14)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">multiprocessing</span></code></a> module (however, pickling will not be as efficient
as storing tables in the native <a class="reference internal" href="file-formats.html#sec-tree-sequence-file-format"><span class="std std-ref">format</span></a>).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pickle</span>
<span class="n">serialised</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
<span class="n">t2</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">serialised</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">t2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
idleftrightparentchildmetadata

0    0    1    10   11        
1    1    3     9    4        

</pre></div>
</div>
</div>
</div>
<p>Tables support the equality operator <code class="docutils literal notranslate"><span class="pre">==</span></code> based on the data
held in the columns:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">t</span> <span class="o">==</span> <span class="n">t2</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>1
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">t</span> <span class="ow">is</span> <span class="n">t2</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>False
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">t2</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">t2</span><span class="p">)</span>
<span class="n">t</span> <span class="o">==</span> <span class="n">t2</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
idleftrightparentchildmetadata

0    0    1    10   11        
1    1    3     9    4        
2    0    1     2    3        

</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0
</pre></div>
</div>
</div>
</div>
<div class="admonition-todo admonition" id="id7">
<p class="admonition-title">Todo</p>
<p>Move some or all of these examples into a suitable alternative chapter.</p>
</div>
<section id="text-columns">
<span id="sec-tables-api-text-columns"></span><h5>Text columns<a class="headerlink" href="#text-columns" title="Link to this heading">#</a></h5>
<p>As described in the <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>, working with
variable length columns is somewhat more involved. Columns
encoding text data store the <strong>encoded bytes</strong> of the flattened
strings, and the offsets into this column in two separate
arrays.</p>
<p>Consider the following example:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">t</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">SiteTable</span><span class="p">()</span>
<span class="n">t</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">)</span>
<span class="n">t</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;BB&quot;</span><span class="p">)</span>
<span class="n">t</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="n">t</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;CCC&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
idpositionancestral_statemetadata

0        0              A        
1        1             BB        
2        2                       
3        3            CCC        


SiteTableRow(position=0.0, ancestral_state=&#39;A&#39;, metadata=b&#39;&#39;)
SiteTableRow(position=1.0, ancestral_state=&#39;BB&#39;, metadata=b&#39;&#39;)
SiteTableRow(position=2.0, ancestral_state=&#39;&#39;, metadata=b&#39;&#39;)
SiteTableRow(position=3.0, ancestral_state=&#39;CCC&#39;, metadata=b&#39;&#39;)
</pre></div>
</div>
</div>
</div>
<p>Here we create a <a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code></a> and add four rows, each with a different
<code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code>. We can then access this information from each
row in a straightforward manner. Working with columns of text data
is a little trickier, however:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">ancestral_state</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">ancestral_state_offset</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[65 66 66 67 67 67]
[0 1 3 3 6]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tskit</span><span class="o">.</span><span class="n">unpack_strings</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">ancestral_state</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">ancestral_state_offset</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&#39;A&#39;, &#39;BB&#39;, &#39;&#39;, &#39;CCC&#39;]
</pre></div>
</div>
</div>
</div>
<p>Here, the <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> array is the UTF8 encoded bytes of the flattened
strings, and the <code class="docutils literal notranslate"><span class="pre">ancestral_state_offset</span></code> is the offset into this array
for each row. The <a class="reference internal" href="#tskit.unpack_strings" title="tskit.unpack_strings"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.unpack_strings()</span></code></a> function, however, is a convient
way to recover the original strings from this encoding. We can also use the
<a class="reference internal" href="#tskit.pack_strings" title="tskit.pack_strings"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.pack_strings()</span></code></a> to insert data using this approach:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">a</span><span class="p">,</span> <span class="n">off</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">pack_strings</span><span class="p">([</span><span class="s2">&quot;0&quot;</span><span class="p">,</span> <span class="s2">&quot;12&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">])</span>
<span class="n">t</span><span class="o">.</span><span class="n">set_columns</span><span class="p">(</span><span class="n">position</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="n">ancestral_state</span><span class="o">=</span><span class="n">a</span><span class="p">,</span> <span class="n">ancestral_state_offset</span><span class="o">=</span><span class="n">off</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
idpositionancestral_statemetadata

0        0              0        
1        1             12        
2        2                       

</pre></div>
</div>
</div>
</div>
<p>When inserting many rows with standard infinite sites mutations (i.e.,
ancestral state is 0), it is more efficient to construct the
numpy arrays directly than to create a list of strings and use
<a class="reference internal" href="#tskit.pack_strings" title="tskit.pack_strings"><code class="xref py py-func docutils literal notranslate"><span class="pre">pack_strings()</span></code></a>. When doing this, it is important to note that
it is the <strong>encoded</strong> byte values that are stored; by default, we
use UTF8 (which corresponds to ASCII for simple printable characters).:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="n">t_s</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">SiteTable</span><span class="p">()</span>
<span class="n">m</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">a</span> <span class="o">=</span> <span class="nb">ord</span><span class="p">(</span><span class="s2">&quot;0&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int8</span><span class="p">)</span>
<span class="n">off</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">m</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">uint32</span><span class="p">)</span>
<span class="n">t_s</span><span class="o">.</span><span class="n">set_columns</span><span class="p">(</span><span class="n">position</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">m</span><span class="p">),</span> <span class="n">ancestral_state</span><span class="o">=</span><span class="n">a</span><span class="p">,</span> <span class="n">ancestral_state_offset</span><span class="o">=</span><span class="n">off</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">t_s</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;ancestral state data&quot;</span><span class="p">,</span> <span class="n">t_s</span><span class="o">.</span><span class="n">ancestral_state</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;ancestral state offsets&quot;</span><span class="p">,</span> <span class="n">t_s</span><span class="o">.</span><span class="n">ancestral_state_offset</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
idpositionancestral_statemetadata

0        0              0        
1        1              0        
2        2              0        
3        3              0        
4        4              0        
5        5              0        
6        6              0        
7        7              0        
8        8              0        
9        9              0        


ancestral state data [48 48 48 48 48 48 48 48 48 48]
ancestral state offsets [ 0  1  2  3  4  5  6  7  8  9 10]
</pre></div>
</div>
</div>
</div>
<p>In the mutation table, the derived state of each mutation can be handled similarly:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">t_m</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">MutationTable</span><span class="p">()</span>
<span class="n">site</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
<span class="n">d</span><span class="p">,</span> <span class="n">off</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">pack_strings</span><span class="p">([</span><span class="s2">&quot;1&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="n">m</span><span class="p">)</span>
<span class="n">node</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
<span class="n">t_m</span><span class="o">.</span><span class="n">set_columns</span><span class="p">(</span><span class="n">site</span><span class="o">=</span><span class="n">site</span><span class="p">,</span> <span class="n">node</span><span class="o">=</span><span class="n">node</span><span class="p">,</span> <span class="n">derived_state</span><span class="o">=</span><span class="n">d</span><span class="p">,</span> <span class="n">derived_state_offset</span><span class="o">=</span><span class="n">off</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">t_m</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
idsitenodetimederived_stateparentmetadata

0    0   0 nan            1    -1        
1    1   0 nan            1    -1        
2    2   0 nan            1    -1        
3    3   0 nan            1    -1        
4    4   0 nan            1    -1        
5    5   0 nan            1    -1        
6    6   0 nan            1    -1        
7    7   0 nan            1    -1        
8    8   0 nan            1    -1        
9    9   0 nan            1    -1        

</pre></div>
</div>
</div>
</div>
<div class="admonition-todo admonition" id="id8">
<p class="admonition-title">Todo</p>
<p>Move some or all of these examples into a suitable alternative chapter.</p>
</div>
</section>
<section id="binary-columns">
<span id="sec-tables-api-binary-columns"></span><h5>Binary columns<a class="headerlink" href="#binary-columns" title="Link to this heading">#</a></h5>
<p>Columns storing binary data take the same approach as
<a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> to encoding
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">variable length data</span></a>.
The difference between the two is only raw <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bytes</span></code></a> values are accepted: no
character encoding or decoding is done on the data. Consider the following example
where a table has no <code class="docutils literal notranslate"><span class="pre">metadata_schema</span></code> such that arbitrary bytes can be stored and
no automatic encoding or decoding of objects is performed by the Python API and we can
store and retrieve raw <code class="docutils literal notranslate"><span class="pre">bytes</span></code>. (See <a class="reference internal" href="metadata.html#sec-metadata"><span class="std std-ref">Metadata</span></a> for details):</p>
<p>Below, we add two rows to a <a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a>, with different
<a class="reference internal" href="data-model.html#sec-metadata-definition"><span class="std std-ref">metadata</span></a>. The first row contains a simple
byte string, and the second contains a Python dictionary serialised using
<a class="reference external" href="https://docs.python.org/3/library/pickle.html#module-pickle" title="(in Python v3.14)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pickle</span></code></a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">t</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NodeTable</span><span class="p">()</span>
<span class="n">t</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">metadata</span><span class="o">=</span><span class="sa">b</span><span class="s2">&quot;these are raw bytes&quot;</span><span class="p">)</span>
<span class="n">t</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">metadata</span><span class="o">=</span><span class="n">pickle</span><span class="o">.</span><span class="n">dumps</span><span class="p">({</span><span class="s2">&quot;x&quot;</span><span class="p">:</span> <span class="mf">1.1</span><span class="p">}))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
idtimeflagspopulationindividualmetadata                                

0    0    0        -1        -1                  b&#39;these are raw bytes&#39;
1    0    0        -1        -1b&#39;\x80\x04\x95\x11\x00\x00\x00\x00\x0...

</pre></div>
</div>
</div>
</div>
<p>Note that the pickled dictionary is encoded in 24 bytes containing unprintable
characters. It appears to be unrelated to the original contents, because the binary
data is <a class="reference external" href="https://en.wikipedia.org/wiki/Base64">base64 encoded</a> to ensure that it is
print-safe (and doesnt break your terminal). (See the
<a class="reference internal" href="data-model.html#sec-metadata-definition"><span class="std std-ref">Metadata</span></a> section for more information on the
use of base64 encoding.).</p>
<p>We can access the metadata in a row (e.g., <code class="docutils literal notranslate"><span class="pre">t[0].metadata</span></code>) which returns a Python
bytes object containing precisely the bytes that were inserted.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">metadata</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">metadata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>b&#39;these are raw bytes&#39;
b&#39;\x80\x04\x95\x11\x00\x00\x00\x00\x00\x00\x00}\x94\x8c\x01x\x94G?\xf1\x99\x99\x99\x99\x99\x9as.&#39;
</pre></div>
</div>
</div>
</div>
<p>The metadata containing the pickled dictionary can be unpickled using
<a class="reference external" href="https://docs.python.org/3/library/pickle.html#pickle.loads" title="(in Python v3.14)"><code class="xref py py-func docutils literal notranslate"><span class="pre">pickle.loads()</span></code></a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">pickle</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">t</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">metadata</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{&#39;x&#39;: 1.1}
</pre></div>
</div>
</div>
</div>
<p>As previously, the <code class="docutils literal notranslate"><span class="pre">replace</span></code> method can be used to change the metadata,
by overwriting an existing row with an updated one:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">t</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">t</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">metadata</span><span class="o">=</span><span class="sa">b</span><span class="s2">&quot;different raw bytes&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
idtimeflagspopulationindividualmetadata                                

0    0    0        -1        -1                  b&#39;different raw bytes&#39;
1    0    0        -1        -1b&#39;\x80\x04\x95\x11\x00\x00\x00\x00\x0...

</pre></div>
</div>
</div>
</div>
<p>Finally, when we print the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> column, we see the raw byte values
encoded as signed integers. As for <a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a>,
the <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> column encodes the offsets into this array. So, we
see that the first metadata value is 9 bytes long and the second is 24.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">metadata</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">metadata_offset</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[ 100  105  102  102  101  114  101  110  116   32  114   97  119   32
   98  121  116  101  115 -128    4 -107   17    0    0    0    0    0
    0    0  125 -108 -116    1  120 -108   71   63  -15 -103 -103 -103
 -103 -103 -102  115   46]
[ 0 19 47]
</pre></div>
</div>
</div>
</div>
<p>The <a class="reference internal" href="#tskit.pack_bytes" title="tskit.pack_bytes"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.pack_bytes()</span></code></a> and <a class="reference internal" href="#tskit.unpack_bytes" title="tskit.unpack_bytes"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.unpack_bytes()</span></code></a> functions are
also useful for encoding data in these columns.</p>
<div class="admonition-todo admonition" id="id9">
<p class="admonition-title">Todo</p>
<p>Move some or all of these examples into a suitable alternative chapter.</p>
</div>
</section>
</section>
<section id="table-functions">
<h4>Table functions<a class="headerlink" href="#table-functions" title="Link to this heading">#</a></h4>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.parse_nodes" title="tskit.parse_nodes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">parse_nodes</span></code></a>(source[,strict,encoding,...])</p></td>
<td><p>Parse the specified file-like object containing a whitespace delimited description of a node table and returns the corresponding <a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a> instance.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.parse_edges" title="tskit.parse_edges"><code class="xref py py-obj docutils literal notranslate"><span class="pre">parse_edges</span></code></a>(source[,strict,table,...])</p></td>
<td><p>Parse the specified file-like object containing a whitespace delimited description of a edge table and returns the corresponding <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a> instance.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.parse_sites" title="tskit.parse_sites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">parse_sites</span></code></a>(source[,strict,encoding,...])</p></td>
<td><p>Parse the specified file-like object containing a whitespace delimited description of a site table and returns the corresponding <a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code></a> instance.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.parse_mutations" title="tskit.parse_mutations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">parse_mutations</span></code></a>(source[,strict,encoding,...])</p></td>
<td><p>Parse the specified file-like object containing a whitespace delimited description of a mutation table and returns the corresponding <a class="reference internal" href="#tskit.MutationTable" title="tskit.MutationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutationTable</span></code></a> instance.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.parse_individuals" title="tskit.parse_individuals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">parse_individuals</span></code></a>(source[,strict,...])</p></td>
<td><p>Parse the specified file-like object containing a whitespace delimited description of an individual table and returns the corresponding <a class="reference internal" href="#tskit.IndividualTable" title="tskit.IndividualTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">IndividualTable</span></code></a> instance.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.parse_populations" title="tskit.parse_populations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">parse_populations</span></code></a>(source[,strict,...])</p></td>
<td><p>Parse the specified file-like object containing a whitespace delimited description of a population table and returns the corresponding <a class="reference internal" href="#tskit.PopulationTable" title="tskit.PopulationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationTable</span></code></a> instance.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.parse_migrations" title="tskit.parse_migrations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">parse_migrations</span></code></a>(source[,strict,encoding,...])</p></td>
<td><p>Parse the specified file-like object containing a whitespace delimited description of a migration table and returns the corresponding <a class="reference internal" href="#tskit.MigrationTable" title="tskit.MigrationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MigrationTable</span></code></a> instance.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.pack_strings" title="tskit.pack_strings"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pack_strings</span></code></a>(strings[,encoding])</p></td>
<td><p>Packs the specified list of strings into a flattened numpy array of 8 bit integers and corresponding offsets using the specified text encoding.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.unpack_strings" title="tskit.unpack_strings"><code class="xref py py-obj docutils literal notranslate"><span class="pre">unpack_strings</span></code></a>(packed,offset[,encoding])</p></td>
<td><p>Unpacks a list of strings from the specified numpy arrays of packed byte data and corresponding offsets using the specified text encoding.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.pack_bytes" title="tskit.pack_bytes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pack_bytes</span></code></a>(data)</p></td>
<td><p>Packs the specified list of bytes into a flattened numpy array of 8 bit integers and corresponding offsets.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.unpack_bytes" title="tskit.unpack_bytes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">unpack_bytes</span></code></a>(packed,offset)</p></td>
<td><p>Unpacks a list of bytes from the specified numpy arrays of packed byte data and corresponding offsets.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>
</section>
<section id="metadata-api">
<span id="sec-python-api-metadata"></span><h2>Metadata API<a class="headerlink" href="#metadata-api" title="Link to this heading">#</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> module provides validation, encoding and decoding of metadata
using a schema. See <a class="reference internal" href="metadata.html#sec-metadata"><span class="std std-ref">Metadata</span></a>, <a class="reference internal" href="metadata.html#sec-metadata-api-overview"><span class="std std-ref">Python Metadata API Overview</span></a> and
<a class="reference external" href="https://tskit.dev/tutorials/metadata.html#sec-tutorial-metadata" title="(in Project name not set)"><span>Working with Metadata</span></a>.</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-obj docutils literal notranslate"><span class="pre">MetadataSchema</span></code></a>(schema)</p></td>
<td><p>Class for validating, encoding and decoding metadata.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.register_metadata_codec" title="tskit.register_metadata_codec"><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_metadata_codec</span></code></a>(codec_cls,codec_id)</p></td>
<td><p>Register a metadata codec class.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>Refer to the top level metadata-related properties of TreeSequences and TableCollections,
such as <a class="reference internal" href="#tskit.TreeSequence.metadata" title="tskit.TreeSequence.metadata"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.metadata</span></code></a> and <a class="reference internal" href="#tskit.TreeSequence.metadata_schema" title="tskit.TreeSequence.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.metadata_schema</span></code></a>. Also the
metadata fields of
<a class="reference internal" href="#sec-python-api-tree-sequences-obtaining-other-objects"><span class="std std-ref">objects accessed</span></a> through
the <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> API.</p>
</div>
</section>
<section id="provenance">
<span id="sec-python-api-provenance"></span><h2>Provenance<a class="headerlink" href="#provenance" title="Link to this heading">#</a></h2>
<p>We provide some preliminary support for validating JSON documents against the
<a class="reference internal" href="provenance.html#sec-provenance"><span class="std std-ref">provenance schema</span></a>. Programmatic access to provenance
information is planned for future versions.</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.validate_provenance" title="tskit.validate_provenance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">validate_provenance</span></code></a>(provenance)</p></td>
<td><p>Validates the specified dict-like object against the tskit <a class="reference internal" href="provenance.html#sec-provenance"><span class="std std-ref">provenance schema</span></a>.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="utility-functions">
<span id="sec-utility-api"></span><h2>Utility functions<a class="headerlink" href="#utility-functions" title="Link to this heading">#</a></h2>
<p>Miscellaneous top-level utility functions.</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#tskit.is_unknown_time" title="tskit.is_unknown_time"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_unknown_time</span></code></a>(time)</p></td>
<td><p>As the default unknown mutation time (<a class="reference internal" href="#tskit.UNKNOWN_TIME" title="tskit.UNKNOWN_TIME"><code class="xref py py-const docutils literal notranslate"><span class="pre">UNKNOWN_TIME</span></code></a>) is a specific NAN value, equality always fails (A NAN value is not equal to itself by definition).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#tskit.random_nucleotides" title="tskit.random_nucleotides"><code class="xref py py-obj docutils literal notranslate"><span class="pre">random_nucleotides</span></code></a>(length,*[,seed])</p></td>
<td><p>Returns a random string of nucleotides of the specified length.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="reference-documentation">
<span id="sec-python-api-reference"></span><h2>Reference documentation<a class="headerlink" href="#reference-documentation" title="Link to this heading">#</a></h2>
<section id="constants">
<span id="sec-python-api-constants"></span><h3>Constants<a class="headerlink" href="#constants" title="Link to this heading">#</a></h3>
<p>The following constants are used throughout the <code class="docutils literal notranslate"><span class="pre">tskit</span></code> API.</p>
<dl class="py data" id="module-tskit">
<dt class="sig sig-object py" id="tskit.NULL">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">NULL</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">-1</span></em><a class="headerlink" href="#tskit.NULL" title="Link to this definition">#</a></dt>
<dd><p>Special reserved value representing a null ID.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="tskit.MISSING_DATA">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">MISSING_DATA</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">-1</span></em><a class="headerlink" href="#tskit.MISSING_DATA" title="Link to this definition">#</a></dt>
<dd><p>Special value representing missing data in a genotype array</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="tskit.NODE_IS_SAMPLE">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">NODE_IS_SAMPLE</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">1</span></em><a class="headerlink" href="#tskit.NODE_IS_SAMPLE" title="Link to this definition">#</a></dt>
<dd><p>Node flag value indicating that it is a sample.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="tskit.FORWARD">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">FORWARD</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">1</span></em><a class="headerlink" href="#tskit.FORWARD" title="Link to this definition">#</a></dt>
<dd><p>Constant representing the forward direction of travel (i.e.,
increasing genomic coordinate values).</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="tskit.REVERSE">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">REVERSE</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">-1</span></em><a class="headerlink" href="#tskit.REVERSE" title="Link to this definition">#</a></dt>
<dd><p>Constant representing the reverse direction of travel (i.e.,
decreasing genomic coordinate values).</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="tskit.ALLELES_01">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">ALLELES_01</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('0',</span> <span class="pre">'1')</span></em><a class="headerlink" href="#tskit.ALLELES_01" title="Link to this definition">#</a></dt>
<dd><p>The allele mapping where the strings 0 and 1 map to genotype
values 0 and 1.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="tskit.ALLELES_ACGT">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">ALLELES_ACGT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('A',</span> <span class="pre">'C',</span> <span class="pre">'G',</span> <span class="pre">'T')</span></em><a class="headerlink" href="#tskit.ALLELES_ACGT" title="Link to this definition">#</a></dt>
<dd><p>The allele mapping where the four nucleotides A, C, G and T map to
the genotype integers 0, 1, 2, and 3, respectively.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="tskit.UNKNOWN_TIME">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">UNKNOWN_TIME</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">nan</span></em><a class="headerlink" href="#tskit.UNKNOWN_TIME" title="Link to this definition">#</a></dt>
<dd><p>Special NAN value used to indicate unknown mutation times. Since this is a
NAN value, you cannot use <cite>==</cite> to test for it. Use <a class="reference internal" href="#tskit.is_unknown_time" title="tskit.is_unknown_time"><code class="xref py py-func docutils literal notranslate"><span class="pre">is_unknown_time()</span></code></a> instead.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="tskit.TIME_UNITS_UNKNOWN">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">TIME_UNITS_UNKNOWN</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'unknown'</span></em><a class="headerlink" href="#tskit.TIME_UNITS_UNKNOWN" title="Link to this definition">#</a></dt>
<dd><p>Default value of ts.time_units</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="tskit.TIME_UNITS_UNCALIBRATED">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">TIME_UNITS_UNCALIBRATED</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'uncalibrated'</span></em><a class="headerlink" href="#tskit.TIME_UNITS_UNCALIBRATED" title="Link to this definition">#</a></dt>
<dd><p>ts.time_units value when dimension is uncalibrated</p>
</dd></dl>

</section>
<section id="exceptions">
<span id="sec-python-api-exceptions"></span><h3>Exceptions<a class="headerlink" href="#exceptions" title="Link to this heading">#</a></h3>
<dl class="py exception">
<dt class="sig sig-object py" id="tskit.DuplicatePositionsError">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">DuplicatePositionsError</span></span><a class="reference internal" href="_modules/tskit/exceptions.html#DuplicatePositionsError"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.DuplicatePositionsError" title="Link to this definition">#</a></dt>
<dd><p>Duplicate positions in the list of sites.</p>
</dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="tskit.MetadataEncodingError">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">MetadataEncodingError</span></span><a class="reference internal" href="_modules/tskit/exceptions.html#MetadataEncodingError"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MetadataEncodingError" title="Link to this definition">#</a></dt>
<dd><p>A metadata object was of a type that could not be encoded</p>
</dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="tskit.MetadataSchemaValidationError">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">MetadataSchemaValidationError</span></span><a class="reference internal" href="_modules/tskit/exceptions.html#MetadataSchemaValidationError"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MetadataSchemaValidationError" title="Link to this definition">#</a></dt>
<dd><p>A metadata schema object did not validate against the metaschema.</p>
</dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="tskit.MetadataValidationError">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">MetadataValidationError</span></span><a class="reference internal" href="_modules/tskit/exceptions.html#MetadataValidationError"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MetadataValidationError" title="Link to this definition">#</a></dt>
<dd><p>A metadata object did not validate against the provenance schema.</p>
</dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="tskit.ProvenanceValidationError">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">ProvenanceValidationError</span></span><a class="reference internal" href="_modules/tskit/exceptions.html#ProvenanceValidationError"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.ProvenanceValidationError" title="Link to this definition">#</a></dt>
<dd><p>A JSON document did not validate against the provenance schema.</p>
</dd></dl>

</section>
<section id="top-level-functions">
<span id="sec-python-api-functions"></span><h3>Top-level functions<a class="headerlink" href="#top-level-functions" title="Link to this heading">#</a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="tskit.all_trees">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">all_trees</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_leaves</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/combinatorics.html#all_trees"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.all_trees" title="Link to this definition">#</a></dt>
<dd><p>Generates all unique leaf-labelled trees with <code class="docutils literal notranslate"><span class="pre">num_leaves</span></code>
leaves. See <a class="reference internal" href="topological-analysis.html#sec-combinatorics"><span class="std std-ref">Identifying and counting topologies</span></a> on the details of this
enumeration. The leaf labels are selected from the set
<code class="docutils literal notranslate"><span class="pre">[0,</span> <span class="pre">num_leaves)</span></code>. The times and labels on internal nodes are
chosen arbitrarily.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_leaves</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The number of leaves of the tree to generate.</p></li>
<li><p><strong>span</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The genomic span of each returned tree.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">tskit.Tree</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tskit.all_tree_shapes">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">all_tree_shapes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_leaves</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/combinatorics.html#all_tree_shapes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.all_tree_shapes" title="Link to this definition">#</a></dt>
<dd><p>Generates all unique shapes of trees with <code class="docutils literal notranslate"><span class="pre">num_leaves</span></code> leaves.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_leaves</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The number of leaves of the tree to generate.</p></li>
<li><p><strong>span</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The genomic span of each returned tree.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">tskit.Tree</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tskit.all_tree_labellings">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">all_tree_labellings</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/combinatorics.html#all_tree_labellings"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.all_tree_labellings" title="Link to this definition">#</a></dt>
<dd><p>Generates all unique labellings of the leaves of a
<a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.Tree</span></code></a>. Leaves are labelled from the set
<code class="docutils literal notranslate"><span class="pre">[0,</span> <span class="pre">n)</span></code> where <code class="docutils literal notranslate"><span class="pre">n</span></code> is the number of leaves of <code class="docutils literal notranslate"><span class="pre">tree</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree</strong> (<a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><em>tskit.Tree</em></a>)  The tree used to generate
labelled trees of the same shape.</p></li>
<li><p><strong>span</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The genomic span of each returned tree.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">tskit.Tree</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tskit.is_unknown_time">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">is_unknown_time</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">time</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/util.html#is_unknown_time"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.is_unknown_time" title="Link to this definition">#</a></dt>
<dd><p>As the default unknown mutation time (<a class="reference internal" href="#tskit.UNKNOWN_TIME" title="tskit.UNKNOWN_TIME"><code class="xref py py-const docutils literal notranslate"><span class="pre">UNKNOWN_TIME</span></code></a>) is a specific NAN value,
equality always fails (A NAN value is not equal to itself by definition).
This method compares the bitfield such that unknown times can be detected. Either
single floats can be passed or lists/arrays.</p>
<p>Note that NANs are a set of floating-point values. <cite>tskit.UNKNOWN_TIME</cite> is a specific
value in this set. <cite>np.nan</cite> is a differing value, but both are NAN.
See <a class="reference external" href="https://en.wikipedia.org/wiki/NaN">https://en.wikipedia.org/wiki/NaN</a></p>
<p>This function only returns true for <code class="docutils literal notranslate"><span class="pre">tskit.is_unknown_time(tskit.UNKNOWN_TIME)</span></code>
and will return false for <code class="docutils literal notranslate"><span class="pre">tskit.is_unknown_time(np.nan)</span></code> or any other NAN or
non-NAN value.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a><em> | </em><em>array-like</em>)  Value or array to check.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A single boolean or array of booleans the same shape as <code class="docutils literal notranslate"><span class="pre">time</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a> | <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tskit.load">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">load</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skip_tables</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skip_reference_sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#load"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.load" title="Link to this definition">#</a></dt>
<dd><p>Return a <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> instance loaded from the specified file object or
path. The file must be in the
<a class="reference internal" href="file-formats.html#sec-tree-sequence-file-format"><span class="std std-ref">tree sequence file format</span></a>
produced by the <a class="reference internal" href="#tskit.TreeSequence.dump" title="tskit.TreeSequence.dump"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.dump()</span></code></a> method.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>With any of the <code class="docutils literal notranslate"><span class="pre">skip_tables</span></code> or <code class="docutils literal notranslate"><span class="pre">skip_reference_sequence</span></code>
options set, it is not possible to load data from a non-seekable stream
(e.g. a socket or STDIN) of multiple tree sequences using consecutive
calls to <a class="reference internal" href="#tskit.load" title="tskit.load"><code class="xref py py-meth docutils literal notranslate"><span class="pre">tskit.load()</span></code></a>.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The file object or path of the <code class="docutils literal notranslate"><span class="pre">.trees</span></code> file containing the
tree sequence we wish to load.</p></li>
<li><p><strong>skip_tables</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, no tables are read from the <code class="docutils literal notranslate"><span class="pre">.trees</span></code>
file and only the top-level information is populated in the tree
sequence object.</p></li>
<li><p><strong>skip_reference_sequence</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, the tree sequence is read
without loading its reference sequence.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The tree sequence object containing the information
stored in the specified file path.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tskit.load_text">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">load_text</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nodes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edges</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individuals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">populations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">migrations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">encoding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'utf8'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base64_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#load_text"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.load_text" title="Link to this definition">#</a></dt>
<dd><p>Return a <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> instance parsed from tabulated text data
contained in the specified file-like objects. The format
for these files is documented in the <a class="reference internal" href="file-formats.html#sec-text-file-format"><span class="std std-ref">Text file formats</span></a> section,
and is produced by the <a class="reference internal" href="#tskit.TreeSequence.dump_text" title="tskit.TreeSequence.dump_text"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.dump_text()</span></code></a> method. Further
properties required for an input tree sequence are described in the
<a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">Valid tree sequence requirements</span></a> section. This method is intended as a
convenient interface for importing external data into tskit; the binary
file format using by <a class="reference internal" href="#tskit.load" title="tskit.load"><code class="xref py py-meth docutils literal notranslate"><span class="pre">tskit.load()</span></code></a> is many times more efficient than
this text format.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">nodes</span></code> and <code class="docutils literal notranslate"><span class="pre">edges</span></code> parameters are mandatory and must be file-like
objects containing text with whitespace delimited columns,  parsable by
<a class="reference internal" href="#tskit.parse_nodes" title="tskit.parse_nodes"><code class="xref py py-func docutils literal notranslate"><span class="pre">parse_nodes()</span></code></a> and <a class="reference internal" href="#tskit.parse_edges" title="tskit.parse_edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">parse_edges()</span></code></a>, respectively. <code class="docutils literal notranslate"><span class="pre">sites</span></code>,
<code class="docutils literal notranslate"><span class="pre">individuals</span></code>, <code class="docutils literal notranslate"><span class="pre">populations</span></code>, <code class="docutils literal notranslate"><span class="pre">mutations</span></code>, and <code class="docutils literal notranslate"><span class="pre">migrations</span></code> are optional,
and must be parsable by <a class="reference internal" href="#tskit.parse_sites" title="tskit.parse_sites"><code class="xref py py-func docutils literal notranslate"><span class="pre">parse_sites()</span></code></a>, <a class="reference internal" href="#tskit.parse_individuals" title="tskit.parse_individuals"><code class="xref py py-func docutils literal notranslate"><span class="pre">parse_individuals()</span></code></a>,
<a class="reference internal" href="#tskit.parse_populations" title="tskit.parse_populations"><code class="xref py py-func docutils literal notranslate"><span class="pre">parse_populations()</span></code></a>, <a class="reference internal" href="#tskit.parse_mutations" title="tskit.parse_mutations"><code class="xref py py-func docutils literal notranslate"><span class="pre">parse_mutations()</span></code></a>, and <a class="reference internal" href="#tskit.parse_migrations" title="tskit.parse_migrations"><code class="xref py py-func docutils literal notranslate"><span class="pre">parse_migrations()</span></code></a>,
respectively. For convenience, if the node table refers to populations,
but the <code class="docutils literal notranslate"><span class="pre">populations</span></code> parameter is not provided, a minimal set of rows are
added to the population table, so that a valid tree sequence can be returned.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> parameter determines the
<a class="reference internal" href="#tskit.TreeSequence.sequence_length" title="tskit.TreeSequence.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.sequence_length</span></code></a> of the returned tree sequence. If it
is 0 or not specified, the value is taken to be the maximum right
coordinate of the input edges. This parameter is useful in degenerate
situations (such as when there are zero edges), but can usually be ignored.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">strict</span></code> parameter controls the field delimiting algorithm that
is used. If <code class="docutils literal notranslate"><span class="pre">strict</span></code> is True (the default), we require exactly one
tab character separating each field. If <code class="docutils literal notranslate"><span class="pre">strict</span></code> is False, a more relaxed
whitespace delimiting algorithm is used, such that any run of whitespace
is regarded as a field separator. In most situations, <code class="docutils literal notranslate"><span class="pre">strict=False</span></code>
is more convenient, but it can lead to error in certain situations. For
example, if a deletion is encoded in the mutation table this will not
be parseable when <code class="docutils literal notranslate"><span class="pre">strict=False</span></code>.</p>
<p>After parsing the tables, <a class="reference internal" href="#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.sort()</span></code></a> is called to ensure that
the loaded tables satisfy the tree sequence <a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">ordering requirements</span></a>. Note that this may result in the
IDs of various entities changing from their positions in the input file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nodes</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.14)"><em>io.TextIOBase</em></a>)  The file-like object containing text describing a
<a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a>.</p></li>
<li><p><strong>edges</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.14)"><em>io.TextIOBase</em></a>)  The file-like object containing text
describing an <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a>.</p></li>
<li><p><strong>sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.14)"><em>io.TextIOBase</em></a>)  The file-like object containing text describing a
<a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code></a>.</p></li>
<li><p><strong>mutations</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.14)"><em>io.TextIOBase</em></a>)  The file-like object containing text
describing a <a class="reference internal" href="#tskit.MutationTable" title="tskit.MutationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutationTable</span></code></a>.</p></li>
<li><p><strong>individuals</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.14)"><em>io.TextIOBase</em></a>)  The file-like object containing text
describing a <a class="reference internal" href="#tskit.IndividualTable" title="tskit.IndividualTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">IndividualTable</span></code></a>.</p></li>
<li><p><strong>populations</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.14)"><em>io.TextIOBase</em></a>)  The file-like object containing text
describing a <a class="reference internal" href="#tskit.PopulationTable" title="tskit.PopulationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationTable</span></code></a>.</p></li>
<li><p><strong>migrations</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.14)"><em>io.TextIOBase</em></a>)  The file-like object containing text
describing a <a class="reference internal" href="#tskit.MigrationTable" title="tskit.MigrationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MigrationTable</span></code></a>.</p></li>
<li><p><strong>sequence_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The sequence length of the returned tree sequence. If
not supplied or zero this will be inferred from the set of edges.</p></li>
<li><p><strong>strict</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, require strict tab delimiting (default). If
False, a relaxed whitespace splitting algorithm is used.</p></li>
<li><p><strong>encoding</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  Encoding used for text representation.</p></li>
<li><p><strong>base64_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, metadata is encoded using Base64
encoding; otherwise, as plain text.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The tree sequence object containing the information
stored in the specified file paths.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tskit.pack_bytes">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">pack_bytes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/util.html#pack_bytes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.pack_bytes" title="Link to this definition">#</a></dt>
<dd><p>Packs the specified list of bytes into a flattened numpy array of 8 bit integers
and corresponding offsets. See <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> for details
of this encoding.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>data</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.14)"><em>bytes</em></a><em>]</em>)  The list of bytes values to encode.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The tuple (packed, offset) of numpy arrays representing the flattened
input data and offsets.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a> (dtype=np.int8), <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a> (dtype=np.uint32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tskit.pack_strings">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">pack_strings</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">strings</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">encoding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'utf8'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/util.html#pack_strings"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.pack_strings" title="Link to this definition">#</a></dt>
<dd><p>Packs the specified list of strings into a flattened numpy array of 8 bit integers
and corresponding offsets using the specified text encoding.
See <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> for details of this encoding of
columns of variable length data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>]</em>)  The list of strings to encode.</p></li>
<li><p><strong>encoding</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The text encoding to use when converting string data
to bytes. See the <a class="reference external" href="https://docs.python.org/3/library/codecs.html#module-codecs" title="(in Python v3.14)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">codecs</span></code></a> module for information on available
string encodings.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The tuple (packed, offset) of numpy arrays representing the flattened
input data and offsets.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a> (dtype=np.int8), <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a> (dtype=np.uint32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tskit.parse_edges">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">parse_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">encoding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'utf8'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base64_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#parse_edges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.parse_edges" title="Link to this definition">#</a></dt>
<dd><p>Parse the specified file-like object containing a whitespace delimited
description of a edge table and returns the corresponding <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a>
instance. See the <a class="reference internal" href="file-formats.html#sec-edge-text-format"><span class="std std-ref">edge text format</span></a> section
for the details of the required format and the
<a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">edge table definition</span></a> section for the
required properties of the contents.</p>
<p>See <a class="reference internal" href="#tskit.load_text" title="tskit.load_text"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load_text()</span></code></a> for a detailed explanation of the <code class="docutils literal notranslate"><span class="pre">strict</span></code> parameter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.14)"><em>io.TextIOBase</em></a>)  The file-like object containing the text.</p></li>
<li><p><strong>strict</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, require strict tab delimiting (default). If
False, a relaxed whitespace splitting algorithm is used.</p></li>
<li><p><strong>table</strong> (<a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><em>EdgeTable</em></a>)  If specified, write the edges into this table. If
not, create a new <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a> instance and return.</p></li>
<li><p><strong>encoding</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  Encoding used for text representation.</p></li>
<li><p><strong>base64_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, metadata is encoded using Base64
encoding; otherwise, as plain text.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tskit.parse_individuals">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">parse_individuals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">encoding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'utf8'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base64_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#parse_individuals"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.parse_individuals" title="Link to this definition">#</a></dt>
<dd><p>Parse the specified file-like object containing a whitespace delimited
description of an individual table and returns the corresponding
<a class="reference internal" href="#tskit.IndividualTable" title="tskit.IndividualTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">IndividualTable</span></code></a> instance. See the <a class="reference internal" href="file-formats.html#sec-individual-text-format"><span class="std std-ref">individual text format</span></a> section for the details of the required
format and the <a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">individual table definition</span></a> section for the required properties of
the contents.</p>
<p>See <a class="reference internal" href="#tskit.load_text" title="tskit.load_text"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load_text()</span></code></a> for a detailed explanation of the <code class="docutils literal notranslate"><span class="pre">strict</span></code>
parameter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.14)"><em>io.TextIOBase</em></a>)  The file-like object containing the text.</p></li>
<li><p><strong>strict</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, require strict tab delimiting (default). If
False, a relaxed whitespace splitting algorithm is used.</p></li>
<li><p><strong>encoding</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  Encoding used for text representation.</p></li>
<li><p><strong>base64_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, metadata is encoded using Base64
encoding; otherwise, as plain text.</p></li>
<li><p><strong>table</strong> (<a class="reference internal" href="#tskit.IndividualTable" title="tskit.IndividualTable"><em>IndividualTable</em></a>)  If specified write into this table. If not,
create a new <a class="reference internal" href="#tskit.IndividualTable" title="tskit.IndividualTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">IndividualTable</span></code></a> instance.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tskit.parse_mutations">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">parse_mutations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">encoding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'utf8'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base64_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#parse_mutations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.parse_mutations" title="Link to this definition">#</a></dt>
<dd><p>Parse the specified file-like object containing a whitespace delimited
description of a mutation table and returns the corresponding <a class="reference internal" href="#tskit.MutationTable" title="tskit.MutationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutationTable</span></code></a>
instance. See the <a class="reference internal" href="file-formats.html#sec-mutation-text-format"><span class="std std-ref">mutation text format</span></a> section
for the details of the required format and the
<a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">mutation table definition</span></a> section for the
required properties of the contents. Note that if the <code class="docutils literal notranslate"><span class="pre">time</span></code> column is missing its
entries are filled with <code class="docutils literal notranslate"><span class="pre">UNKNOWN_TIME</span></code>.</p>
<p>See <a class="reference internal" href="#tskit.load_text" title="tskit.load_text"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load_text()</span></code></a> for a detailed explanation of the <code class="docutils literal notranslate"><span class="pre">strict</span></code>
parameter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.14)"><em>io.TextIOBase</em></a>)  The file-like object containing the text.</p></li>
<li><p><strong>strict</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, require strict tab delimiting (default). If
False, a relaxed whitespace splitting algorithm is used.</p></li>
<li><p><strong>encoding</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  Encoding used for text representation.</p></li>
<li><p><strong>base64_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, metadata is encoded using Base64
encoding; otherwise, as plain text.</p></li>
<li><p><strong>table</strong> (<a class="reference internal" href="#tskit.MutationTable" title="tskit.MutationTable"><em>MutationTable</em></a>)  If specified, write mutations into this table.
If not, create a new <a class="reference internal" href="#tskit.MutationTable" title="tskit.MutationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutationTable</span></code></a> instance.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tskit.parse_nodes">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">parse_nodes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">encoding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'utf8'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base64_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#parse_nodes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.parse_nodes" title="Link to this definition">#</a></dt>
<dd><p>Parse the specified file-like object containing a whitespace delimited
description of a node table and returns the corresponding <a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a>
instance. See the <a class="reference internal" href="file-formats.html#sec-node-text-format"><span class="std std-ref">node text format</span></a> section
for the details of the required format and the
<a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">node table definition</span></a> section for the
required properties of the contents.</p>
<p>See <a class="reference internal" href="#tskit.load_text" title="tskit.load_text"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load_text()</span></code></a> for a detailed explanation of the <code class="docutils literal notranslate"><span class="pre">strict</span></code>
parameter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.14)"><em>io.TextIOBase</em></a>)  The file-like object containing the text.</p></li>
<li><p><strong>strict</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, require strict tab delimiting (default). If
False, a relaxed whitespace splitting algorithm is used.</p></li>
<li><p><strong>encoding</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  Encoding used for text representation.</p></li>
<li><p><strong>base64_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, metadata is encoded using Base64
encoding; otherwise, as plain text.</p></li>
<li><p><strong>table</strong> (<a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><em>NodeTable</em></a>)  If specified write into this table. If not,
create a new <a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a> instance.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tskit.parse_populations">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">parse_populations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">encoding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'utf8'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base64_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#parse_populations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.parse_populations" title="Link to this definition">#</a></dt>
<dd><p>Parse the specified file-like object containing a whitespace delimited
description of a population table and returns the corresponding
<a class="reference internal" href="#tskit.PopulationTable" title="tskit.PopulationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationTable</span></code></a> instance. See the <a class="reference internal" href="file-formats.html#sec-population-text-format"><span class="std std-ref">population text format</span></a> section for the details of the required
format and the <a class="reference internal" href="data-model.html#sec-population-table-definition"><span class="std std-ref">population table definition</span></a> section for the required properties of
the contents.</p>
<p>See <a class="reference internal" href="#tskit.load_text" title="tskit.load_text"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load_text()</span></code></a> for a detailed explanation of the <code class="docutils literal notranslate"><span class="pre">strict</span></code>
parameter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.14)"><em>io.TextIOBase</em></a>)  The file-like object containing the text.</p></li>
<li><p><strong>strict</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, require strict tab delimiting (default). If
False, a relaxed whitespace splitting algorithm is used.</p></li>
<li><p><strong>encoding</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  Encoding used for text representation.</p></li>
<li><p><strong>base64_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, metadata is encoded using Base64
encoding; otherwise, as plain text.</p></li>
<li><p><strong>table</strong> (<a class="reference internal" href="#tskit.PopulationTable" title="tskit.PopulationTable"><em>PopulationTable</em></a>)  If specified write into this table. If not,
create a new <a class="reference internal" href="#tskit.PopulationTable" title="tskit.PopulationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationTable</span></code></a> instance.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tskit.parse_migrations">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">parse_migrations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">encoding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'utf8'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base64_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#parse_migrations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.parse_migrations" title="Link to this definition">#</a></dt>
<dd><p>Parse the specified file-like object containing a whitespace delimited
description of a migration table and returns the corresponding
<a class="reference internal" href="#tskit.MigrationTable" title="tskit.MigrationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MigrationTable</span></code></a> instance.</p>
<p>See the <a class="reference internal" href="file-formats.html#sec-migration-text-format"><span class="std std-ref">migration text format</span></a> section
for the details of the required format and the
<a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">migration table definition</span></a> section
for the required properties of the contents. Note that if the <code class="docutils literal notranslate"><span class="pre">time</span></code> column
is missing its entries are filled with <a class="reference internal" href="#tskit.UNKNOWN_TIME" title="tskit.UNKNOWN_TIME"><code class="xref py py-data docutils literal notranslate"><span class="pre">UNKNOWN_TIME</span></code></a>.</p>
<p>See <a class="reference internal" href="#tskit.load_text" title="tskit.load_text"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load_text()</span></code></a> for a detailed explanation of the <code class="docutils literal notranslate"><span class="pre">strict</span></code>
parameter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.14)"><em>io.TextIOBase</em></a>)  The file-like object containing the text.</p></li>
<li><p><strong>strict</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, require strict tab delimiting (default). If
False, a relaxed whitespace splitting algorithm is used.</p></li>
<li><p><strong>encoding</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  Encoding used for text representation.</p></li>
<li><p><strong>base64_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, metadata is encoded using Base64
encoding; otherwise, as plain text.</p></li>
<li><p><strong>table</strong> (<a class="reference internal" href="#tskit.MigrationTable" title="tskit.MigrationTable"><em>MigrationTable</em></a>)  If specified, write migrations into this table.
If not, create a new <a class="reference internal" href="#tskit.MigrationTable" title="tskit.MigrationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MigrationTable</span></code></a> instance.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tskit.parse_sites">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">parse_sites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">encoding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'utf8'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base64_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#parse_sites"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.parse_sites" title="Link to this definition">#</a></dt>
<dd><p>Parse the specified file-like object containing a whitespace delimited
description of a site table and returns the corresponding <a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code></a>
instance. See the <a class="reference internal" href="file-formats.html#sec-site-text-format"><span class="std std-ref">site text format</span></a> section
for the details of the required format and the
<a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">site table definition</span></a> section for the
required properties of the contents.</p>
<p>See <a class="reference internal" href="#tskit.load_text" title="tskit.load_text"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load_text()</span></code></a> for a detailed explanation of the <code class="docutils literal notranslate"><span class="pre">strict</span></code>
parameter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.14)"><em>io.TextIOBase</em></a>)  The file-like object containing the text.</p></li>
<li><p><strong>strict</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, require strict tab delimiting (default). If
False, a relaxed whitespace splitting algorithm is used.</p></li>
<li><p><strong>encoding</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  Encoding used for text representation.</p></li>
<li><p><strong>base64_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, metadata is encoded using Base64
encoding; otherwise, as plain text.</p></li>
<li><p><strong>table</strong> (<a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><em>SiteTable</em></a>)  If specified write site into this table. If not,
create a new <a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code></a> instance.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tskit.random_nucleotides">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">random_nucleotides</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/util.html#random_nucleotides"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.random_nucleotides" title="Link to this definition">#</a></dt>
<dd><p>Returns a random string of nucleotides of the specified length. Characters
are drawn uniformly from the alphabet ACTG.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The length of the random sequence.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A string of the specified length consisting of random nucleotide
characters.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tskit.register_metadata_codec">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">register_metadata_codec</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">codec_cls</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">codec_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/metadata.html#register_metadata_codec"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.register_metadata_codec" title="Link to this definition">#</a></dt>
<dd><p>Register a metadata codec class.
This function maintains a mapping from metadata codec identifiers used in schemas
to codec classes. When a codec class is registered, it will replace any class
previously registered under the same codec identifier, if present.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>codec_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  String to use to refer to the codec in the schema.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.14)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tskit.validate_provenance">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">validate_provenance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">provenance</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/provenance.html#validate_provenance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.validate_provenance" title="Link to this definition">#</a></dt>
<dd><p>Validates the specified dict-like object against the tskit
<a class="reference internal" href="provenance.html#sec-provenance"><span class="std std-ref">provenance schema</span></a>. If the input does
not represent a valid instance of the schema an exception is
raised.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a>)  The dictionary representing a JSON document
to be validated against the schema.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.ProvenanceValidationError" title="tskit.ProvenanceValidationError"><strong>ProvenanceValidationError</strong></a>  if the schema is not valid.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tskit.unpack_bytes">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">unpack_bytes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">packed</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">offset</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/util.html#unpack_bytes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.unpack_bytes" title="Link to this definition">#</a></dt>
<dd><p>Unpacks a list of bytes from the specified numpy arrays of packed byte
data and corresponding offsets. See <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> for details
of this encoding.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>packed</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a>)  The flattened array of byte values.</p></li>
<li><p><strong>offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a>)  The array of offsets into the <code class="docutils literal notranslate"><span class="pre">packed</span></code> array.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The list of bytes values unpacked from the parameter arrays.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.14)">bytes</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tskit.unpack_strings">
<span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">unpack_strings</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">packed</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">offset</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">encoding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'utf8'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/util.html#unpack_strings"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.unpack_strings" title="Link to this definition">#</a></dt>
<dd><p>Unpacks a list of strings from the specified numpy arrays of packed byte
data and corresponding offsets using the specified text encoding.
See <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> for details of this encoding of
columns of variable length data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>packed</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a>)  The flattened array of byte values.</p></li>
<li><p><strong>offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a>)  The array of offsets into the <code class="docutils literal notranslate"><span class="pre">packed</span></code> array.</p></li>
<li><p><strong>encoding</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The text encoding to use when converting string data
to bytes. See the <a class="reference external" href="https://docs.python.org/3/library/codecs.html#module-codecs" title="(in Python v3.14)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">codecs</span></code></a> module for information on available
string encodings.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The list of strings unpacked from the parameter arrays.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)">list</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a>]</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="tree-and-tree-sequence-classes">
<h3>Tree and tree sequence classes<a class="headerlink" href="#tree-and-tree-sequence-classes" title="Link to this heading">#</a></h3>
<section id="the-tree-class">
<h4>The <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> class<a class="headerlink" href="#the-tree-class" title="Link to this heading">#</a></h4>
<p>Also see the <a class="reference internal" href="#sec-python-api-trees"><span class="std std-ref">Tree API</span></a> summary.</p>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.Tree">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">Tree</span></span><a class="reference internal" href="_modules/tskit/trees.html#Tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree" title="Link to this definition">#</a></dt>
<dd><p>A single tree in a <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>. Please see the
<a class="reference external" href="https://tskit.dev/tutorials/getting_started.html#sec-processing-trees" title="(in Project name not set)"><span>Processing trees</span></a> section for information
on how efficiently access trees sequentially or obtain a list
of individual trees in a tree sequence.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">sample_lists</span></code> parameter controls the features that are enabled
for this tree. If <code class="docutils literal notranslate"><span class="pre">sample_lists</span></code> is True a more efficient algorithm is
used in the <a class="reference internal" href="#tskit.Tree.samples" title="tskit.Tree.samples"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.samples()</span></code></a> method.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">tracked_samples</span></code> parameter can be used to efficiently count the
number of samples in a given set that exist in a particular subtree
using the <a class="reference internal" href="#tskit.Tree.num_tracked_samples" title="tskit.Tree.num_tracked_samples"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.num_tracked_samples()</span></code></a> method.</p>
<p>The <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> class is a state-machine which has a state
corresponding to each of the trees in the parent tree sequence. We
transition between these states by using the seek functions like
<a class="reference internal" href="#tskit.Tree.first" title="tskit.Tree.first"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.first()</span></code></a>, <a class="reference internal" href="#tskit.Tree.last" title="tskit.Tree.last"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.last()</span></code></a>, <a class="reference internal" href="#tskit.Tree.seek" title="tskit.Tree.seek"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.seek()</span></code></a> and
<a class="reference internal" href="#tskit.Tree.seek_index" title="tskit.Tree.seek_index"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.seek_index()</span></code></a>. There is one more state, the so-called null
or cleared state. This is the state that a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> is in
immediately after initialisation;  it has an index of -1, and no edges. We
can also enter the null state by calling <a class="reference internal" href="#tskit.Tree.next" title="tskit.Tree.next"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.next()</span></code></a> on the last
tree in a sequence, calling <a class="reference internal" href="#tskit.Tree.prev" title="tskit.Tree.prev"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.prev()</span></code></a> on the first tree in a
sequence or calling calling the <a class="reference internal" href="#tskit.Tree.clear" title="tskit.Tree.clear"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.clear()</span></code></a> method at any time.</p>
<p>The high-level TreeSequence seeking and iterations methods (e.g,
<a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.trees()</span></code></a>) are built on these low-level state-machine
seek operations. We recommend these higher level operations for most
users.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree_sequence</strong> (<a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><em>TreeSequence</em></a>)  The parent tree sequence.</p></li>
<li><p><strong>tracked_samples</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  The list of samples to be tracked and
counted using the <a class="reference internal" href="#tskit.Tree.num_tracked_samples" title="tskit.Tree.num_tracked_samples"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.num_tracked_samples()</span></code></a> method.</p></li>
<li><p><strong>sample_lists</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, provide more efficient access
to the samples beneath a given node using the
<a class="reference internal" href="#tskit.Tree.samples" title="tskit.Tree.samples"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.samples()</span></code></a> method.</p></li>
<li><p><strong>root_threshold</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The minimum number of samples that a node
must be ancestral to for it to be in the list of roots. By default
this is 1, so that isolated samples (representing missing data)
are roots. To efficiently restrict the roots of the tree to
those subtending meaningful topology, set this to 2. This value
is only relevant when trees have multiple roots.</p></li>
<li><p><strong>sample_counts</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Deprecated since 0.2.4.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.copy" title="Link to this definition">#</a></dt>
<dd><p>Returns a deep copy of this tree. The returned tree will have identical state
to this tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A copy of this tree.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">Tree</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.tree_sequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">tree_sequence</span></span><a class="headerlink" href="#tskit.Tree.tree_sequence" title="Link to this definition">#</a></dt>
<dd><p>Returns the tree sequence that this tree is from.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The parent tree sequence for this tree.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.root_threshold">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">root_threshold</span></span><a class="headerlink" href="#tskit.Tree.root_threshold" title="Link to this definition">#</a></dt>
<dd><p>Returns the minimum number of samples that a node must be an ancestor
of to be considered a potential root. This can be set, for example, when
calling the <a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.trees()</span></code></a> iterator.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The root threshold.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.first">
<span class="sig-name descname"><span class="pre">first</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.first"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.first" title="Link to this definition">#</a></dt>
<dd><p>Seeks to the first tree in the sequence. This can be called whether
the tree is in the null state or not.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.last">
<span class="sig-name descname"><span class="pre">last</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.last"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.last" title="Link to this definition">#</a></dt>
<dd><p>Seeks to the last tree in the sequence. This can be called whether
the tree is in the null state or not.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.next">
<span class="sig-name descname"><span class="pre">next</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.next"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.next" title="Link to this definition">#</a></dt>
<dd><p>Seeks to the next tree in the sequence. If the tree is in the initial
null state we seek to the first tree (equivalent to calling <a class="reference internal" href="#tskit.Tree.first" title="tskit.Tree.first"><code class="xref py py-meth docutils literal notranslate"><span class="pre">first()</span></code></a>).
Calling <code class="docutils literal notranslate"><span class="pre">next</span></code> on the last tree in the sequence results in the tree
being cleared back into the null initial state (equivalent to calling
<a class="reference internal" href="#tskit.Tree.clear" title="tskit.Tree.clear"><code class="xref py py-meth docutils literal notranslate"><span class="pre">clear()</span></code></a>). The return value of the function indicates whether the
tree is in a non-null state, and can be used to loop over the trees:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Iterate over the trees from left-to-right</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">Tree</span><span class="p">(</span><span class="n">tree_sequence</span><span class="p">)</span>
<span class="k">while</span> <span class="n">tree</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>
    <span class="c1"># Do something with the tree.</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="c1"># tree is now back in the null state.</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>True if the tree has been transformed into one of the trees
in the sequence; False if the tree has been transformed into the
null state.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.prev">
<span class="sig-name descname"><span class="pre">prev</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.prev"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.prev" title="Link to this definition">#</a></dt>
<dd><p>Seeks to the previous tree in the sequence. If the tree is in the initial
null state we seek to the last tree (equivalent to calling <a class="reference internal" href="#tskit.Tree.last" title="tskit.Tree.last"><code class="xref py py-meth docutils literal notranslate"><span class="pre">last()</span></code></a>).
Calling <code class="docutils literal notranslate"><span class="pre">prev</span></code> on the first tree in the sequence results in the tree
being cleared back into the null initial state (equivalent to calling
<a class="reference internal" href="#tskit.Tree.clear" title="tskit.Tree.clear"><code class="xref py py-meth docutils literal notranslate"><span class="pre">clear()</span></code></a>). The return value of the function indicates whether the
tree is in a non-null state, and can be used to loop over the trees:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Iterate over the trees from right-to-left</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">Tree</span><span class="p">(</span><span class="n">tree_sequence</span><span class="p">)</span>
<span class="k">while</span> <span class="n">tree</span><span class="o">.</span><span class="n">prev</span><span class="p">()</span>
    <span class="c1"># Do something with the tree.</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="c1"># tree is now back in the null state.</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>True if the tree has been transformed into one of the trees
in the sequence; False if the tree has been transformed into the
null state.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.clear"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.clear" title="Link to this definition">#</a></dt>
<dd><p>Resets this tree back to the initial null state. Calling this method
on a tree already in the null state has no effect.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.seek_index">
<span class="sig-name descname"><span class="pre">seek_index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skip</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.seek_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.seek_index" title="Link to this definition">#</a></dt>
<dd><p>Sets the state to represent the tree at the specified
index in the parent tree sequence. Negative indexes following the
standard Python conventions are allowed, i.e., <code class="docutils literal notranslate"><span class="pre">index=-1</span></code> will
seek to the last tree in the sequence.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The current implementation of this operation is linear in the number of
trees, so may be inefficient for large tree sequences. See
<a class="reference external" href="https://github.com/tskit-dev/tskit/issues/684">this issue</a> for more
information.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The tree index to seek to.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#IndexError" title="(in Python v3.14)"><strong>IndexError</strong></a>  If an index outside the acceptable range is provided.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.seek">
<span class="sig-name descname"><span class="pre">seek</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">position</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skip</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.seek"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.seek" title="Link to this definition">#</a></dt>
<dd><p>Sets the state to represent the tree that covers the specified
position in the parent tree sequence. After a successful return
of this method we have <code class="docutils literal notranslate"><span class="pre">tree.interval.left</span></code> &lt;= <code class="docutils literal notranslate"><span class="pre">position</span></code>
&lt; <code class="docutils literal notranslate"><span class="pre">tree.interval.right</span></code>.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The current implementation of this operation is linear in the number of
trees, so may be inefficient for large tree sequences. See
<a class="reference external" href="https://github.com/tskit-dev/tskit/issues/684">this issue</a> for more
information.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>position</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The position along the sequence length to
seek to.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.14)"><strong>ValueError</strong></a>  If <code class="docutils literal notranslate"><span class="pre">position</span></code> is less than 0 or <code class="docutils literal notranslate"><span class="pre">position</span></code> is greater
than or equal to
<a class="reference internal" href="#tskit.TreeSequence.sequence_length" title="tskit.TreeSequence.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.sequence_length</span></code></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.rank">
<span class="sig-name descname"><span class="pre">rank</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.rank"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.rank" title="Link to this definition">#</a></dt>
<dd><p>Produce the rank of this tree in the enumeration of all leaf-labelled
trees of n leaves. See the <a class="reference internal" href="topological-analysis.html#sec-tree-ranks"><span class="std std-ref">Interpreting Tree Ranks</span></a> section for
details on ranking and unranking trees.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.14)"><strong>ValueError</strong></a>  If the tree has multiple roots.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#tskit.Rank" title="tskit.combinatorics.Rank"><code class="xref py py-class docutils literal notranslate"><span class="pre">Rank</span></code></a></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.unrank">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">unrank</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_leaves</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rank</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">branch_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.unrank"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.unrank" title="Link to this definition">#</a></dt>
<dd><p>Reconstruct the tree of the given <code class="docutils literal notranslate"><span class="pre">rank</span></code>
(see <a class="reference internal" href="#tskit.Tree.rank" title="tskit.Tree.rank"><code class="xref py py-meth docutils literal notranslate"><span class="pre">tskit.Tree.rank()</span></code></a>) with <code class="docutils literal notranslate"><span class="pre">num_leaves</span></code> leaves.
The labels and times of internal nodes are assigned by a postorder
traversal of the nodes, such that the time of each internal node
is the maximum time of its children plus the specified <code class="docutils literal notranslate"><span class="pre">branch_length</span></code>.
The time of each leaf is 0.</p>
<p>See the <a class="reference internal" href="topological-analysis.html#sec-tree-ranks"><span class="std std-ref">Interpreting Tree Ranks</span></a> section for details on ranking and
unranking trees and what constitutes valid ranks.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_leaves</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The number of leaves of the tree to generate.</p></li>
<li><p><strong>rank</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.14)"><em>tuple</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>)</em>)  The rank of the tree to generate.</p></li>
<li><p><strong>span</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The genomic span of the returned tree. The tree will cover
the interval <span class="math notranslate nohighlight">\([0, \text{span})\)</span> and the <a class="reference internal" href="#tskit.Tree.tree_sequence" title="tskit.Tree.tree_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tree_sequence</span></code></a>
from which the tree is taken will have its
<a class="reference internal" href="#tskit.TreeSequence.sequence_length" title="tskit.TreeSequence.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">sequence_length</span></code></a> equal to <code class="docutils literal notranslate"><span class="pre">span</span></code>.</p></li>
<li><p><strong>branch_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The minimum length of a branch in this tree.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.14)"><strong>ValueError</strong></a>  If the given rank is out of bounds for trees
with <code class="docutils literal notranslate"><span class="pre">num_leaves</span></code> leaves.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#tskit.Tree" title="tskit.trees.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.count_topologies">
<span class="sig-name descname"><span class="pre">count_topologies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.count_topologies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.count_topologies" title="Link to this definition">#</a></dt>
<dd><p>Calculates the distribution of embedded topologies for every combination
of the sample sets in <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>. <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> defaults to all
samples in the tree grouped by population.</p>
<p><code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> need not include all samples but must be pairwise disjoint.</p>
<p>The returned object is a <a class="reference internal" href="#tskit.TopologyCounter" title="tskit.TopologyCounter"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TopologyCounter</span></code></a> that contains
counts of topologies per combination of sample sets. For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">topology_counter</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">count_topologies</span><span class="p">()</span>
<span class="n">rank</span><span class="p">,</span> <span class="n">count</span> <span class="o">=</span> <span class="n">topology_counter</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">most_common</span><span class="p">(</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
<p>produces the most common tree topology, with populations 0, 1
and 2 as its tips, according to the genealogies of those
populations samples in this tree.</p>
<p>The counts for each topology in the <a class="reference internal" href="#tskit.TopologyCounter" title="tskit.TopologyCounter"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TopologyCounter</span></code></a>
are absolute counts that we would get if we were to select all
combinations of samples from the relevant sample sets.
For sample sets <span class="math notranslate nohighlight">\([s_0, s_1, ..., s_n]\)</span>, the total number of
topologies for those sample sets is equal to
<span class="math notranslate nohighlight">\(|s_0| * |s_1| * ... * |s_n|\)</span>, so the counts in the counter
<code class="docutils literal notranslate"><span class="pre">topology_counter[0,</span> <span class="pre">1,</span> <span class="pre">...,</span> <span class="pre">n]</span></code> should sum to
<span class="math notranslate nohighlight">\(|s_0| * |s_1| * ... * |s_n|\)</span>.</p>
<p>To convert the topology counts to probabilities, divide by the total
possible number of sample combinations from the sample sets in question:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">set_sizes</span> <span class="o">=</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">sample_set</span><span class="p">)</span> <span class="k">for</span> <span class="n">sample_set</span> <span class="ow">in</span> <span class="n">sample_sets</span><span class="p">]</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">count</span> <span class="o">/</span> <span class="p">(</span><span class="n">set_sizes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">set_sizes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">set_sizes</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The interface for this method is preliminary and may be subject to
backwards incompatible changes in the near future.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.
Defaults to all samples grouped by population.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.14)"><strong>ValueError</strong></a>  If nodes in <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> are invalid or are
internal samples.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#tskit.TopologyCounter" title="tskit.combinatorics.TopologyCounter"><code class="xref py py-class docutils literal notranslate"><span class="pre">TopologyCounter</span></code></a></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.branch_length">
<span class="sig-name descname"><span class="pre">branch_length</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.branch_length"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.branch_length" title="Link to this definition">#</a></dt>
<dd><p>Returns the length of the branch (in units of time) joining the
specified node to its parent. This is equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">tree</span><span class="o">.</span><span class="n">time</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">parent</span><span class="p">(</span><span class="n">u</span><span class="p">))</span> <span class="o">-</span> <span class="n">tree</span><span class="o">.</span><span class="n">time</span><span class="p">(</span><span class="n">u</span><span class="p">)</span>
</pre></div>
</div>
<p>The branch length for a node that has no parent (e.g., a root) is
defined as zero.</p>
<p>Note that this is not related to the property <cite>.length</cite> which
is a deprecated alias for the genomic <a class="reference internal" href="#tskit.Tree.span" title="tskit.Tree.span"><code class="xref py py-attr docutils literal notranslate"><span class="pre">span</span></code></a> covered by a tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The branch length from u to its parent.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.total_branch_length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">total_branch_length</span></span><a class="headerlink" href="#tskit.Tree.total_branch_length" title="Link to this definition">#</a></dt>
<dd><p>Returns the sum of all the branch lengths in this tree (in
units of time). This is equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">sum</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">branch_length</span><span class="p">(</span><span class="n">u</span><span class="p">)</span> <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">nodes</span><span class="p">())</span>
</pre></div>
</div>
<p>Note that the branch lengths for root nodes are defined as zero.</p>
<p>As this is defined by a traversal of the tree, technically we
return the sum of all branch lengths that are reachable from
roots. Thus, this is the total length of all branches that are connected
to at least one sample. This distinction is only important
in tree sequences that contain dead branches, i.e., those
that define topology that is not connected to a tree root
(see <a class="reference internal" href="data-model.html#sec-data-model-tree-dead-leaves-and-branches"><span class="std std-ref">Dead leaves and branches</span></a>)</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The sum of lengths of branches in this tree.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.mrca">
<span class="sig-name descname"><span class="pre">mrca</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.mrca"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.mrca" title="Link to this definition">#</a></dt>
<dd><p>Returns the most recent common ancestor of the specified nodes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>*args</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  input node IDs, at least 2 arguments are required.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The node ID of the most recent common ancestor of the
input nodes, or <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> if the nodes do not share
a common ancestor in the tree.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.tmrca">
<span class="sig-name descname"><span class="pre">tmrca</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.tmrca"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.tmrca" title="Link to this definition">#</a></dt>
<dd><p>Returns the time of the most recent common ancestor of the specified
nodes. This is equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">tree</span><span class="o">.</span><span class="n">time</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">mrca</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">))</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you are using this method to calculate average tmrca values along the
genome between pairs of sample nodes, for efficiency reasons you should
instead consider the <code class="docutils literal notranslate"><span class="pre">mode=&quot;branch&quot;</span></code> option of the
<a class="reference internal" href="#tskit.TreeSequence.divergence" title="tskit.TreeSequence.divergence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.divergence()</span></code></a> or <a class="reference internal" href="#tskit.TreeSequence.diversity" title="tskit.TreeSequence.diversity"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.diversity()</span></code></a> methods.
Since these calculate the average branch length between pairs of sample
nodes, for samples at time 0 the resulting statistics will be exactly
twice the tmrca value.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>*args</strong>  input node IDs, at least 2 arguments are required.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The time of the most recent common ancestor of all the nodes.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)">float</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.14)"><strong>ValueError</strong></a>  If the nodes do not share a single common ancestor in this
tree (i.e., if <code class="docutils literal notranslate"><span class="pre">tree.mrca(*args)</span> <span class="pre">==</span> <span class="pre">tskit.NULL</span></code>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.parent">
<span class="sig-name descname"><span class="pre">parent</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.parent"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.parent" title="Link to this definition">#</a></dt>
<dd><p>Returns the parent of the specified node. Returns
<a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> if u is a root or is not a node in
the current tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The parent of u.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.parent_array">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">parent_array</span></span><a class="headerlink" href="#tskit.Tree.parent_array" title="Link to this definition">#</a></dt>
<dd><p>A numpy array (dtype=np.int32) encoding the parent of each node
in this tree, such that <code class="docutils literal notranslate"><span class="pre">tree.parent_array[u]</span> <span class="pre">==</span> <span class="pre">tree.parent(u)</span></code>
for all <code class="docutils literal notranslate"><span class="pre">0</span> <span class="pre">&lt;=</span> <span class="pre">u</span> <span class="pre">&lt;=</span> <span class="pre">ts.num_nodes</span></code>. See the <a class="reference internal" href="#tskit.Tree.parent" title="tskit.Tree.parent"><code class="xref py py-meth docutils literal notranslate"><span class="pre">parent()</span></code></a>
method for details on the semantics of tree parents and the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">Tree structure</span></a> section for information on the
quintuply linked tree encoding.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The length of these arrays is
equal to the number of nodes in the tree sequence plus 1, with the
final element corresponding to the trees <a class="reference internal" href="#tskit.Tree.virtual_root" title="tskit.Tree.virtual_root"><code class="xref py py-meth docutils literal notranslate"><span class="pre">virtual_root()</span></code></a>.
Please see the <a class="reference internal" href="data-model.html#sec-data-model-tree-roots"><span class="std std-ref">tree roots</span></a> section
for more details.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This is a high-performance interface which
provides zero-copy access to memory used in the C library.
As a consequence, the values stored in this array will change as
the Tree state is modified as we move along the tree sequence. See the
<a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> documentation for more details. Therefore, if you want to
compare arrays representing different trees along the sequence, you must
take <strong>copies</strong> of the arrays.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.ancestors">
<span class="sig-name descname"><span class="pre">ancestors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.ancestors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.ancestors" title="Link to this definition">#</a></dt>
<dd><p>Returns an iterator over the ancestors of node <code class="docutils literal notranslate"><span class="pre">u</span></code> in this tree
(i.e. the chain of parents from <code class="docutils literal notranslate"><span class="pre">u</span></code> to the root).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.left_child">
<span class="sig-name descname"><span class="pre">left_child</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.left_child"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.left_child" title="Link to this definition">#</a></dt>
<dd><p>Returns the leftmost child of the specified node. Returns
<a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> if u is a leaf or is not a node in
the current tree. The left-to-right ordering of children
is arbitrary and should not be depended on; see the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">data model</span></a> section
for details.</p>
<p>This is a low-level method giving access to the quintuply linked
tree structure in memory; the <a class="reference internal" href="#tskit.Tree.children" title="tskit.Tree.children"><code class="xref py py-meth docutils literal notranslate"><span class="pre">children()</span></code></a> method is a more
convenient way to obtain the children of a given node.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The leftmost child of u.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.left_child_array">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">left_child_array</span></span><a class="headerlink" href="#tskit.Tree.left_child_array" title="Link to this definition">#</a></dt>
<dd><p>A numpy array (dtype=np.int32) encoding the left child of each node
in this tree, such that <code class="docutils literal notranslate"><span class="pre">tree.left_child_array[u]</span> <span class="pre">==</span> <span class="pre">tree.left_child(u)</span></code>
for all <code class="docutils literal notranslate"><span class="pre">0</span> <span class="pre">&lt;=</span> <span class="pre">u</span> <span class="pre">&lt;=</span> <span class="pre">ts.num_nodes</span></code>. See the <a class="reference internal" href="#tskit.Tree.left_child" title="tskit.Tree.left_child"><code class="xref py py-meth docutils literal notranslate"><span class="pre">left_child()</span></code></a>
method for details on the semantics of tree left_child and the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">Tree structure</span></a> section for information on the
quintuply linked tree encoding.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The length of these arrays is
equal to the number of nodes in the tree sequence plus 1, with the
final element corresponding to the trees <a class="reference internal" href="#tskit.Tree.virtual_root" title="tskit.Tree.virtual_root"><code class="xref py py-meth docutils literal notranslate"><span class="pre">virtual_root()</span></code></a>.
Please see the <a class="reference internal" href="data-model.html#sec-data-model-tree-roots"><span class="std std-ref">tree roots</span></a> section
for more details.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This is a high-performance interface which
provides zero-copy access to memory used in the C library.
As a consequence, the values stored in this array will change as
the Tree state is modified as we move along the tree sequence. See the
<a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> documentation for more details. Therefore, if you want to
compare arrays representing different trees along the sequence, you must
take <strong>copies</strong> of the arrays.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.right_child">
<span class="sig-name descname"><span class="pre">right_child</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.right_child"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.right_child" title="Link to this definition">#</a></dt>
<dd><p>Returns the rightmost child of the specified node. Returns
<a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> if u is a leaf or is not a node in
the current tree. The left-to-right ordering of children
is arbitrary and should not be depended on; see the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">data model</span></a> section
for details.</p>
<p>This is a low-level method giving access to the quintuply linked
tree structure in memory; the <a class="reference internal" href="#tskit.Tree.children" title="tskit.Tree.children"><code class="xref py py-meth docutils literal notranslate"><span class="pre">children()</span></code></a> method is a more
convenient way to obtain the children of a given node.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The rightmost child of u.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.right_child_array">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">right_child_array</span></span><a class="headerlink" href="#tskit.Tree.right_child_array" title="Link to this definition">#</a></dt>
<dd><p>A numpy array (dtype=np.int32) encoding the right child of each node
in this tree, such that <code class="docutils literal notranslate"><span class="pre">tree.right_child_array[u]</span> <span class="pre">==</span> <span class="pre">tree.right_child(u)</span></code>
for all <code class="docutils literal notranslate"><span class="pre">0</span> <span class="pre">&lt;=</span> <span class="pre">u</span> <span class="pre">&lt;=</span> <span class="pre">ts.num_nodes</span></code>. See the <a class="reference internal" href="#tskit.Tree.right_child" title="tskit.Tree.right_child"><code class="xref py py-meth docutils literal notranslate"><span class="pre">right_child()</span></code></a>
method for details on the semantics of tree right_child and the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">Tree structure</span></a> section for information on the
quintuply linked tree encoding.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The length of these arrays is
equal to the number of nodes in the tree sequence plus 1, with the
final element corresponding to the trees <a class="reference internal" href="#tskit.Tree.virtual_root" title="tskit.Tree.virtual_root"><code class="xref py py-meth docutils literal notranslate"><span class="pre">virtual_root()</span></code></a>.
Please see the <a class="reference internal" href="data-model.html#sec-data-model-tree-roots"><span class="std std-ref">tree roots</span></a> section
for more details.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This is a high-performance interface which
provides zero-copy access to memory used in the C library.
As a consequence, the values stored in this array will change as
the Tree state is modified as we move along the tree sequence. See the
<a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> documentation for more details. Therefore, if you want to
compare arrays representing different trees along the sequence, you must
take <strong>copies</strong> of the arrays.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.left_sib">
<span class="sig-name descname"><span class="pre">left_sib</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.left_sib"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.left_sib" title="Link to this definition">#</a></dt>
<dd><p>Returns the sibling node to the left of u, or <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a>
if u does not have a left sibling.
The left-to-right ordering of children
is arbitrary and should not be depended on; see the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">data model</span></a> section
for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The sibling node to the left of u.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.left_sib_array">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">left_sib_array</span></span><a class="headerlink" href="#tskit.Tree.left_sib_array" title="Link to this definition">#</a></dt>
<dd><p>A numpy array (dtype=np.int32) encoding the left sib of each node
in this tree, such that <code class="docutils literal notranslate"><span class="pre">tree.left_sib_array[u]</span> <span class="pre">==</span> <span class="pre">tree.left_sib(u)</span></code>
for all <code class="docutils literal notranslate"><span class="pre">0</span> <span class="pre">&lt;=</span> <span class="pre">u</span> <span class="pre">&lt;=</span> <span class="pre">ts.num_nodes</span></code>. See the <a class="reference internal" href="#tskit.Tree.left_sib" title="tskit.Tree.left_sib"><code class="xref py py-meth docutils literal notranslate"><span class="pre">left_sib()</span></code></a>
method for details on the semantics of tree left_sib and the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">Tree structure</span></a> section for information on the
quintuply linked tree encoding.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The length of these arrays is
equal to the number of nodes in the tree sequence plus 1, with the
final element corresponding to the trees <a class="reference internal" href="#tskit.Tree.virtual_root" title="tskit.Tree.virtual_root"><code class="xref py py-meth docutils literal notranslate"><span class="pre">virtual_root()</span></code></a>.
Please see the <a class="reference internal" href="data-model.html#sec-data-model-tree-roots"><span class="std std-ref">tree roots</span></a> section
for more details.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This is a high-performance interface which
provides zero-copy access to memory used in the C library.
As a consequence, the values stored in this array will change as
the Tree state is modified as we move along the tree sequence. See the
<a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> documentation for more details. Therefore, if you want to
compare arrays representing different trees along the sequence, you must
take <strong>copies</strong> of the arrays.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.right_sib">
<span class="sig-name descname"><span class="pre">right_sib</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.right_sib"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.right_sib" title="Link to this definition">#</a></dt>
<dd><p>Returns the sibling node to the right of u, or <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a>
if u does not have a right sibling.
The left-to-right ordering of children
is arbitrary and should not be depended on; see the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">data model</span></a> section
for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The sibling node to the right of u.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.right_sib_array">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">right_sib_array</span></span><a class="headerlink" href="#tskit.Tree.right_sib_array" title="Link to this definition">#</a></dt>
<dd><p>A numpy array (dtype=np.int32) encoding the right sib of each node
in this tree, such that <code class="docutils literal notranslate"><span class="pre">tree.right_sib_array[u]</span> <span class="pre">==</span> <span class="pre">tree.right_sib(u)</span></code>
for all <code class="docutils literal notranslate"><span class="pre">0</span> <span class="pre">&lt;=</span> <span class="pre">u</span> <span class="pre">&lt;=</span> <span class="pre">ts.num_nodes</span></code>. See the <a class="reference internal" href="#tskit.Tree.right_sib" title="tskit.Tree.right_sib"><code class="xref py py-meth docutils literal notranslate"><span class="pre">right_sib()</span></code></a>
method for details on the semantics of tree right_sib and the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">Tree structure</span></a> section for information on the
quintuply linked tree encoding.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The length of these arrays is
equal to the number of nodes in the tree sequence plus 1, with the
final element corresponding to the trees <a class="reference internal" href="#tskit.Tree.virtual_root" title="tskit.Tree.virtual_root"><code class="xref py py-meth docutils literal notranslate"><span class="pre">virtual_root()</span></code></a>.
Please see the <a class="reference internal" href="data-model.html#sec-data-model-tree-roots"><span class="std std-ref">tree roots</span></a> section
for more details.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This is a high-performance interface which
provides zero-copy access to memory used in the C library.
As a consequence, the values stored in this array will change as
the Tree state is modified as we move along the tree sequence. See the
<a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> documentation for more details. Therefore, if you want to
compare arrays representing different trees along the sequence, you must
take <strong>copies</strong> of the arrays.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.siblings">
<span class="sig-name descname"><span class="pre">siblings</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.siblings"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.siblings" title="Link to this definition">#</a></dt>
<dd><p>Returns the sibling(s) of the specified node <code class="docutils literal notranslate"><span class="pre">u</span></code> as a tuple of integer
node IDs. If <code class="docutils literal notranslate"><span class="pre">u</span></code> has no siblings or is not a node in the current tree,
returns an empty tuple. If <code class="docutils literal notranslate"><span class="pre">u</span></code> is the root of a single-root tree,
returns an empty tuple; if <code class="docutils literal notranslate"><span class="pre">u</span></code> is the root of a multi-root tree,
returns the other roots (note all the roots are related by the virtual root).
If <code class="docutils literal notranslate"><span class="pre">u</span></code> is the virtual root (which has no siblings), returns an empty tuple.
If <code class="docutils literal notranslate"><span class="pre">u</span></code> is an isolated node, whether it has siblings or not depends on
whether it is a sample or non-sample node; if it is a sample node,
returns the root(s) of the tree, otherwise, returns an empty tuple.
The ordering of siblings  is arbitrary and should not be depended on;
see the <a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">data model</span></a> section for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The siblings of <code class="docutils literal notranslate"><span class="pre">u</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.14)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.num_children_array">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_children_array</span></span><a class="headerlink" href="#tskit.Tree.num_children_array" title="Link to this definition">#</a></dt>
<dd><p>A numpy array (dtype=np.int32) encoding the number of children of
each node in this tree, such that
<code class="docutils literal notranslate"><span class="pre">tree.num_children_array[u]</span> <span class="pre">==</span> <span class="pre">tree.num_children(u)</span></code> for all
<code class="docutils literal notranslate"><span class="pre">0</span> <span class="pre">&lt;=</span> <span class="pre">u</span> <span class="pre">&lt;=</span> <span class="pre">ts.num_nodes</span></code>. See the <a class="reference internal" href="#tskit.Tree.num_children" title="tskit.Tree.num_children"><code class="xref py py-meth docutils literal notranslate"><span class="pre">num_children()</span></code></a>
method for details on the semantics of tree num_children and the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">Tree structure</span></a> section for information on the
quintuply linked tree encoding.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The length of these arrays is
equal to the number of nodes in the tree sequence plus 1, with the
final element corresponding to the trees <a class="reference internal" href="#tskit.Tree.virtual_root" title="tskit.Tree.virtual_root"><code class="xref py py-meth docutils literal notranslate"><span class="pre">virtual_root()</span></code></a>.
Please see the <a class="reference internal" href="data-model.html#sec-data-model-tree-roots"><span class="std std-ref">tree roots</span></a> section
for more details.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This is a high-performance interface which
provides zero-copy access to memory used in the C library.
As a consequence, the values stored in this array will change as
the Tree state is modified as we move along the tree sequence. See the
<a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> documentation for more details. Therefore, if you want to
compare arrays representing different trees along the sequence, you must
take <strong>copies</strong> of the arrays.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.edge">
<span class="sig-name descname"><span class="pre">edge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.edge"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.edge" title="Link to this definition">#</a></dt>
<dd><p>Returns the id of the edge encoding the relationship between <code class="docutils literal notranslate"><span class="pre">u</span></code>
and its parent, or <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> if <code class="docutils literal notranslate"><span class="pre">u</span></code> is a root, virtual root
or is not a node in the current tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Id of edge connecting u to its parent.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.edge_array">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">edge_array</span></span><a class="headerlink" href="#tskit.Tree.edge_array" title="Link to this definition">#</a></dt>
<dd><p>A numpy array (dtype=np.int32) of edge ids encoding the relationship
between the child node <code class="docutils literal notranslate"><span class="pre">u</span></code> and its parent, such that
<code class="docutils literal notranslate"><span class="pre">tree.edge_array[u]</span> <span class="pre">==</span> <span class="pre">tree.edge(u)</span></code> for all
<code class="docutils literal notranslate"><span class="pre">0</span> <span class="pre">&lt;=</span> <span class="pre">u</span> <span class="pre">&lt;=</span> <span class="pre">ts.num_nodes</span></code>. See the <a class="reference internal" href="#tskit.Tree.edge" title="tskit.Tree.edge"><code class="xref py py-meth docutils literal notranslate"><span class="pre">edge()</span></code></a>
method for details on the semantics of tree edge and the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">Tree structure</span></a> section for information on the
quintuply linked tree encoding.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The length of these arrays is
equal to the number of nodes in the tree sequence plus 1, with the
final element corresponding to the trees <a class="reference internal" href="#tskit.Tree.virtual_root" title="tskit.Tree.virtual_root"><code class="xref py py-meth docutils literal notranslate"><span class="pre">virtual_root()</span></code></a>.
Please see the <a class="reference internal" href="data-model.html#sec-data-model-tree-roots"><span class="std std-ref">tree roots</span></a> section
for more details.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This is a high-performance interface which
provides zero-copy access to memory used in the C library.
As a consequence, the values stored in this array will change as
the Tree state is modified as we move along the tree sequence. See the
<a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> documentation for more details. Therefore, if you want to
compare arrays representing different trees along the sequence, you must
take <strong>copies</strong> of the arrays.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.virtual_root">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">virtual_root</span></span><a class="headerlink" href="#tskit.Tree.virtual_root" title="Link to this definition">#</a></dt>
<dd><p>The ID of the virtual root in this tree. This is equal to
<a class="reference internal" href="#tskit.TreeSequence.num_nodes" title="tskit.TreeSequence.num_nodes"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.num_nodes</span></code></a>.</p>
<p>Please see the <a class="reference internal" href="data-model.html#sec-data-model-tree-roots"><span class="std std-ref">tree roots</span></a>
section for more details.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.num_edges">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_edges</span></span><a class="headerlink" href="#tskit.Tree.num_edges" title="Link to this definition">#</a></dt>
<dd><p>The total number of edges in this tree. This is equal to the
number of tree sequence edges that intersect with this trees
genomic interval.</p>
<p>Note that this may be greater than the number of branches that
are reachable from the trees roots, since we can have topology
that is not associated with any samples.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.left_root">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">left_root</span></span><a class="headerlink" href="#tskit.Tree.left_root" title="Link to this definition">#</a></dt>
<dd><p>The leftmost root in this tree. If there are multiple roots
in this tree, they are siblings of this node, and so we can
use <a class="reference internal" href="#tskit.Tree.right_sib" title="tskit.Tree.right_sib"><code class="xref py py-meth docutils literal notranslate"><span class="pre">right_sib()</span></code></a> to iterate over all roots:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">u</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">left_root</span>
<span class="k">while</span> <span class="n">u</span> <span class="o">!=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Root:&quot;</span><span class="p">,</span> <span class="n">u</span><span class="p">)</span>
    <span class="n">u</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">right_sib</span><span class="p">(</span><span class="n">u</span><span class="p">)</span>
</pre></div>
</div>
<p>The left-to-right ordering of roots is arbitrary and should
not be depended on; see the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">data model</span></a>
section for details.</p>
<p>This is a low-level method giving access to the quintuply linked
tree structure in memory; the <a class="reference internal" href="#tskit.Tree.roots" title="tskit.Tree.roots"><code class="xref py py-attr docutils literal notranslate"><span class="pre">roots</span></code></a> attribute is a more
convenient way to obtain the roots of a tree. If you are assuming
that there is a single root in the tree you should use the
<a class="reference internal" href="#tskit.Tree.root" title="tskit.Tree.root"><code class="xref py py-attr docutils literal notranslate"><span class="pre">root</span></code></a> property.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Do not use this property if you are assuming that there
is a single root in trees that are being processed. The
<a class="reference internal" href="#tskit.Tree.root" title="tskit.Tree.root"><code class="xref py py-attr docutils literal notranslate"><span class="pre">root</span></code></a> property should be used in this case, as it will
raise an error when multiple roots exists.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.children">
<span class="sig-name descname"><span class="pre">children</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.children"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.children" title="Link to this definition">#</a></dt>
<dd><p>Returns the children of the specified node <code class="docutils literal notranslate"><span class="pre">u</span></code> as a tuple of integer node IDs.
If <code class="docutils literal notranslate"><span class="pre">u</span></code> is a leaf, return the empty tuple. The ordering of children
is arbitrary and should not be depended on; see the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">data model</span></a> section
for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The children of <code class="docutils literal notranslate"><span class="pre">u</span></code> as a tuple of integers</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.14)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.time">
<span class="sig-name descname"><span class="pre">time</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.time"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.time" title="Link to this definition">#</a></dt>
<dd><p>Returns the time of the specified node. This is equivalently
to <code class="docutils literal notranslate"><span class="pre">tree.tree_sequence.node(u).time</span></code> except for the special
case of the trees <a class="reference internal" href="data-model.html#sec-data-model-tree-roots"><span class="std std-ref">virtual root</span></a>,
which is defined as positive infinity.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The time of u.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.depth">
<span class="sig-name descname"><span class="pre">depth</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.depth"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.depth" title="Link to this definition">#</a></dt>
<dd><p>Returns the number of nodes on the path from <code class="docutils literal notranslate"><span class="pre">u</span></code> to a
root, not including <code class="docutils literal notranslate"><span class="pre">u</span></code>. Thus, the depth of a root is
zero.</p>
<p>As a special case, the depth of the <a class="reference internal" href="data-model.html#sec-data-model-tree-roots"><span class="std std-ref">virtual root</span></a> is defined as -1.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The depth of u.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.population">
<span class="sig-name descname"><span class="pre">population</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.population"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.population" title="Link to this definition">#</a></dt>
<dd><p>Returns the population associated with the specified node.
Equivalent to <code class="docutils literal notranslate"><span class="pre">tree.tree_sequence.node(u).population</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The ID of the population associated with node u.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.is_internal">
<span class="sig-name descname"><span class="pre">is_internal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.is_internal"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.is_internal" title="Link to this definition">#</a></dt>
<dd><p>Returns True if the specified node is not a leaf. A node is internal
if it has one or more children in the current tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if u is not a leaf node.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.is_leaf">
<span class="sig-name descname"><span class="pre">is_leaf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.is_leaf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.is_leaf" title="Link to this definition">#</a></dt>
<dd><p>Returns True if the specified node is a leaf. A node <span class="math notranslate nohighlight">\(u\)</span> is a
leaf if it has zero children.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><span class="math notranslate nohighlight">\(u\)</span> can be any node in the entire tree sequence, including ones
which are not connected via branches to a root node of the tree (and which
are therefore not conventionally considered part of the tree). Indeed, if
there are many trees in the tree sequence, it is common for the majority of
non-sample nodes to be <a class="reference internal" href="#tskit.Tree.is_isolated" title="tskit.Tree.is_isolated"><code class="xref py py-meth docutils literal notranslate"><span class="pre">isolated</span></code></a> in any one
tree. By the definition above, this method will return <code class="docutils literal notranslate"><span class="pre">True</span></code> for such
a tree when a node of this sort is specified. Such nodes can be thought of
as dead leaves, see <a class="reference internal" href="data-model.html#sec-data-model-tree-dead-leaves-and-branches"><span class="std std-ref">Dead leaves and branches</span></a>.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if u is a leaf node.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.is_isolated">
<span class="sig-name descname"><span class="pre">is_isolated</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.is_isolated"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.is_isolated" title="Link to this definition">#</a></dt>
<dd><p>Returns True if the specified node is isolated in this tree: that is
it has no parents and no children (note that all isolated nodes in the tree
are therefore also <a class="reference internal" href="#tskit.Tree.is_leaf" title="tskit.Tree.is_leaf"><code class="xref py py-meth docutils literal notranslate"><span class="pre">leaves</span></code></a>). Sample nodes that are isolated
and have no mutations above them are used to represent
<a class="reference internal" href="data-model.html#sec-data-model-missing-data"><span class="std std-ref">missing data</span></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if u is an isolated node.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.is_sample">
<span class="sig-name descname"><span class="pre">is_sample</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.is_sample"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.is_sample" title="Link to this definition">#</a></dt>
<dd><p>Returns True if the specified node is a sample. A node <span class="math notranslate nohighlight">\(u\)</span> is a
sample if it has been marked as a sample in the parent tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if u is a sample.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.is_descendant">
<span class="sig-name descname"><span class="pre">is_descendant</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.is_descendant"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.is_descendant" title="Link to this definition">#</a></dt>
<dd><p>Returns True if the specified node u is a descendant of node v and False
otherwise. A node <span class="math notranslate nohighlight">\(u\)</span> is a descendant of another node <span class="math notranslate nohighlight">\(v\)</span> if
<span class="math notranslate nohighlight">\(v\)</span> is on the path from <span class="math notranslate nohighlight">\(u\)</span> to root. A node is considered
to be a descendant of itself, so <code class="docutils literal notranslate"><span class="pre">tree.is_descendant(u,</span> <span class="pre">u)</span></code> will be
True for any valid node.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The descendant node.</p></li>
<li><p><strong>v</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The ancestral node.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if u is a descendant of v.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.14)"><strong>ValueError</strong></a>  If u or v are not valid node IDs.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.num_nodes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_nodes</span></span><a class="headerlink" href="#tskit.Tree.num_nodes" title="Link to this definition">#</a></dt>
<dd><p>Returns the number of nodes in the <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> this tree is in.
Equivalent to <code class="docutils literal notranslate"><span class="pre">tree.tree_sequence.num_nodes</span></code>.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 0.4: </span>Use <a class="reference internal" href="#tskit.TreeSequence.num_nodes" title="tskit.TreeSequence.num_nodes"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Tree.tree_sequence.num_nodes</span></code></a> if you want
the number of nodes in the entire tree sequence, or
<code class="docutils literal notranslate"><span class="pre">len(tree.preorder())</span></code> to find the number of nodes that are
reachable from all roots in this tree.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.num_roots">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_roots</span></span><a class="headerlink" href="#tskit.Tree.num_roots" title="Link to this definition">#</a></dt>
<dd><p>The number of roots in this tree, as defined in the <a class="reference internal" href="#tskit.Tree.roots" title="tskit.Tree.roots"><code class="xref py py-attr docutils literal notranslate"><span class="pre">roots</span></code></a>
attribute.</p>
<p>Only requires O(number of roots) time.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.has_single_root">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">has_single_root</span></span><a class="headerlink" href="#tskit.Tree.has_single_root" title="Link to this definition">#</a></dt>
<dd><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if this tree has a single root, <code class="docutils literal notranslate"><span class="pre">False</span></code> otherwise.
Equivalent to tree.num_roots == 1. This is a O(1) operation.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.has_multiple_roots">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">has_multiple_roots</span></span><a class="headerlink" href="#tskit.Tree.has_multiple_roots" title="Link to this definition">#</a></dt>
<dd><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if this tree has more than one root, <code class="docutils literal notranslate"><span class="pre">False</span></code> otherwise.
Equivalent to tree.num_roots &gt; 1. This is a O(1) operation.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.roots">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">roots</span></span><a class="headerlink" href="#tskit.Tree.roots" title="Link to this definition">#</a></dt>
<dd><p>The list of roots in this tree. A root is defined as a unique endpoint of the
paths starting at samples, subject to the condition that it is connected to at
least <a class="reference internal" href="#tskit.Tree.root_threshold" title="tskit.Tree.root_threshold"><code class="xref py py-attr docutils literal notranslate"><span class="pre">root_threshold</span></code></a> samples. We can define the set of roots as follows:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">roots</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
<span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">tree_sequence</span><span class="o">.</span><span class="n">samples</span><span class="p">():</span>
    <span class="k">while</span> <span class="n">tree</span><span class="o">.</span><span class="n">parent</span><span class="p">(</span><span class="n">u</span><span class="p">)</span> <span class="o">!=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span><span class="p">:</span>
        <span class="n">u</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">parent</span><span class="p">(</span><span class="n">u</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">tree</span><span class="o">.</span><span class="n">num_samples</span><span class="p">(</span><span class="n">u</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">tree</span><span class="o">.</span><span class="n">root_threshold</span><span class="p">:</span>
        <span class="n">roots</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">u</span><span class="p">)</span>
<span class="c1"># roots is now the set of all roots in this tree.</span>
<span class="k">assert</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">roots</span><span class="p">)</span> <span class="o">==</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">roots</span><span class="p">)</span>
</pre></div>
</div>
<p>The roots of the tree are returned in a list, in no particular order.</p>
<p>Only requires O(number of roots) time.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In trees with large amounts of <a class="reference internal" href="data-model.html#sec-data-model-missing-data"><span class="std std-ref">Missing data</span></a>,
for example where a region of the genome lacks any ancestral information,
there can be a very large number of roots, potentially all the samples
in the tree sequence.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The list of roots in this tree.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)">list</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.root">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">root</span></span><a class="headerlink" href="#tskit.Tree.root" title="Link to this definition">#</a></dt>
<dd><p>The root of this tree. If the tree contains multiple roots, a ValueError is
raised indicating that the <a class="reference internal" href="#tskit.Tree.roots" title="tskit.Tree.roots"><code class="xref py py-attr docutils literal notranslate"><span class="pre">roots</span></code></a> attribute should be used instead.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The root node.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.14)"><strong>ValueError</strong></a>  if this tree contains more than one root.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.is_root">
<span class="sig-name descname"><span class="pre">is_root</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.is_root"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.is_root" title="Link to this definition">#</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">True</span></code> if the specified node is a root in this tree (see
<a class="reference internal" href="#tskit.Tree.roots" title="tskit.Tree.roots"><code class="xref py py-attr docutils literal notranslate"><span class="pre">roots</span></code></a> for the definition of a root). This is exactly equivalent to
finding the node ID in <a class="reference internal" href="#tskit.Tree.roots" title="tskit.Tree.roots"><code class="xref py py-attr docutils literal notranslate"><span class="pre">roots</span></code></a>, but is more efficient for trees
with large numbers of roots, such as in regions with extensive
<a class="reference internal" href="data-model.html#sec-data-model-missing-data"><span class="std std-ref">Missing data</span></a>.  Note that <code class="docutils literal notranslate"><span class="pre">False</span></code> is returned for all
other nodes, including <a class="reference internal" href="data-model.html#sec-data-model-tree-isolated-nodes"><span class="std std-ref">isolated</span></a>
non-sample nodes which are not found in the topology of the current tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node of interest.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if u is a root.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.index">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">index</span></span><a class="headerlink" href="#tskit.Tree.index" title="Link to this definition">#</a></dt>
<dd><p>Returns the index this tree occupies in the parent tree sequence.
This index is zero based, so the first tree in the sequence has index 0.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The index of this tree.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.interval">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">interval</span></span><a class="headerlink" href="#tskit.Tree.interval" title="Link to this definition">#</a></dt>
<dd><p>Returns the coordinates of the genomic interval that this tree
represents the history of. The interval is returned as a tuple
<span class="math notranslate nohighlight">\((l, r)\)</span> and is a half-open interval such that the left
coordinate is inclusive and the right coordinate is exclusive. This
tree therefore applies to all genomic locations <span class="math notranslate nohighlight">\(x\)</span> such that
<span class="math notranslate nohighlight">\(l \leq x &lt; r\)</span>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A named tuple (l, r) representing the left-most (inclusive)
and right-most (exclusive) coordinates of the genomic region
covered by this tree. The coordinates can be accessed by index
(<code class="docutils literal notranslate"><span class="pre">0</span></code> or <code class="docutils literal notranslate"><span class="pre">1</span></code>) or equivalently by name (<code class="docutils literal notranslate"><span class="pre">.left</span></code> or <code class="docutils literal notranslate"><span class="pre">.right</span></code>)</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.Interval" title="tskit.Interval">Interval</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.span">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">span</span></span><a class="headerlink" href="#tskit.Tree.span" title="Link to this definition">#</a></dt>
<dd><p>Returns the genomic distance that this tree spans.
This is defined as <span class="math notranslate nohighlight">\(r - l\)</span>, where <span class="math notranslate nohighlight">\((l, r)\)</span> is the genomic
interval returned by <a class="reference internal" href="#tskit.Tree.interval" title="tskit.Tree.interval"><code class="xref py py-attr docutils literal notranslate"><span class="pre">interval</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The genomic distance covered by this tree.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.mid">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mid</span></span><a class="headerlink" href="#tskit.Tree.mid" title="Link to this definition">#</a></dt>
<dd><p>Returns the midpoint of the genomic interval that this tree represents
the history of. This is defined as <span class="math notranslate nohighlight">\((l + (r - l) / 2)\)</span>, where
<span class="math notranslate nohighlight">\((l, r)\)</span> is the genomic interval returned by
<a class="reference internal" href="#tskit.Tree.interval" title="tskit.Tree.interval"><code class="xref py py-attr docutils literal notranslate"><span class="pre">interval</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The genomic distance covered by this tree.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.draw_text">
<span class="sig-name descname"><span class="pre">draw_text</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_time</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_ascii</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.draw_text"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.draw_text" title="Link to this definition">#</a></dt>
<dd><p>Create a text representation of a tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>orientation</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  one of <code class="docutils literal notranslate"><span class="pre">&quot;top&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;left&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;bottom&quot;</span></code>, or
<code class="docutils literal notranslate"><span class="pre">&quot;right&quot;</span></code>, specifying the margin on which the root is placed. Specifying
<code class="docutils literal notranslate"><span class="pre">&quot;left&quot;</span></code> or <code class="docutils literal notranslate"><span class="pre">&quot;right&quot;</span></code> will lead to time being shown on the x axis (i.e.
a horizontal tree. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) use the standard coalescent
arrangement of a vertical tree with recent nodes at the bottom of the plot
and older nodes above.</p></li>
<li><p><strong>node_labels</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a>)  If specified, show custom labels for the nodes
that are present in the map. Any nodes not specified in the map will
not have a node label.</p></li>
<li><p><strong>max_time</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  If equal to <code class="docutils literal notranslate"><span class="pre">&quot;tree&quot;</span></code> (the default), the maximum time
is set to be that of the oldest root in the tree. If equal to <code class="docutils literal notranslate"><span class="pre">&quot;ts&quot;</span></code> the
maximum time is set to be the time of the oldest root in the tree
sequence; this is useful when drawing trees from the same tree sequence as it
ensures that node heights are consistent.</p></li>
<li><p><strong>use_ascii</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If <code class="docutils literal notranslate"><span class="pre">False</span></code> (default) then use unicode
<a class="reference external" href="https://en.wikipedia.org/wiki/Box-drawing_character">box drawing characters</a>
to render the tree. If <code class="docutils literal notranslate"><span class="pre">True</span></code>, use plain ascii characters, which look
cruder but are less susceptible to misalignment or font substitution.
Alternatively, if you are having alignment problems with Unicode, you can try
out the solution documented <a class="reference external" href="https://github.com/tskit-dev/tskit/issues/189#issuecomment-499114811">here</a>.</p></li>
<li><p><strong>order</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The left-to-right ordering of child nodes in the drawn tree.
This can be either: <code class="docutils literal notranslate"><span class="pre">&quot;minlex&quot;</span></code>, which minimises the differences
between adjacent trees (see also the <code class="docutils literal notranslate"><span class="pre">&quot;minlex_postorder&quot;</span></code> traversal
order for the <a class="reference internal" href="#tskit.Tree.nodes" title="tskit.Tree.nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">nodes()</span></code></a> method); or <code class="docutils literal notranslate"><span class="pre">&quot;tree&quot;</span></code> which draws trees
in the left-to-right order defined by the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">quintuply linked tree structure</span></a>.
If not specified or None, this defaults to <code class="docutils literal notranslate"><span class="pre">&quot;minlex&quot;</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A text representation of a tree.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.draw_svg">
<span class="sig-name descname"><span class="pre">draw_svg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tree_height_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_time</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_time</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_tree_height</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_titles</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation_titles</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">root_svg_attributes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">style</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_root_branch</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">symbol_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_regions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_ticks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_gridlines</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">all_edge_mutations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">omit_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">canvas_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">preamble</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.draw_svg"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.draw_svg" title="Link to this definition">#</a></dt>
<dd><p>Return an SVG representation of a single tree. By default, numeric
labels are drawn beside nodes and mutations: these can be altered using the
<code class="docutils literal notranslate"><span class="pre">node_labels</span></code> and <code class="docutils literal notranslate"><span class="pre">mutation_labels</span></code> parameters. See the
<a class="reference external" href="https://tskit.dev/tutorials/viz.html#sec-tskit-viz" title="(in Project name not set)"><span class="xref std std-ref">visualization tutorial</span></a> for more details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The path to the file to write the output. If None, do not
write to file.</p></li>
<li><p><strong>size</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.14)"><em>tuple</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>)</em>)  A tuple of (width, height) specifying a target
drawing size in abstract user units (usually interpreted as pixels on
initial display). Components of the drawing will be scaled so that the total
plot including labels etc. normally fits onto a canvas of this size (see
<code class="docutils literal notranslate"><span class="pre">canvas_size</span></code> below). If <code class="docutils literal notranslate"><span class="pre">None</span></code>, pick a size appropriate for a tree
with a reasonably small number (i.e. tens) of samples. Default: <code class="docutils literal notranslate"><span class="pre">None</span></code></p></li>
<li><p><strong>time_scale</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  Control how height values for nodes are computed.
If this is equal to <code class="docutils literal notranslate"><span class="pre">&quot;time&quot;</span></code> (the default), node heights are proportional
to their time values. If this is equal to <code class="docutils literal notranslate"><span class="pre">&quot;log_time&quot;</span></code>, node heights are
proportional to their log(time) values. If it is equal to <code class="docutils literal notranslate"><span class="pre">&quot;rank&quot;</span></code>, node
heights are spaced equally according to their ranked times.</p></li>
<li><p><strong>tree_height_scale</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  Deprecated alias for time_scale. (Deprecated in
0.3.6)</p></li>
<li><p><strong>title</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A title string to be included in the SVG output. If <code class="docutils literal notranslate"><span class="pre">None</span></code>
(default) no title is shown, which gives more vertical space for the tree.</p></li>
<li><p><strong>max_time</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>,</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The maximum plotted time value in the current
scaling system (see <code class="docutils literal notranslate"><span class="pre">time_scale</span></code>). Can be either a string or a
numeric value. If equal to <code class="docutils literal notranslate"><span class="pre">&quot;tree&quot;</span></code> (the default), the maximum time
is set to be that of the oldest root in the tree. If equal to <code class="docutils literal notranslate"><span class="pre">&quot;ts&quot;</span></code> the
maximum time is set to be the time of the oldest root in the tree
sequence; this is useful when drawing trees from the same tree sequence as it
ensures that node heights are consistent. If a numeric value, this is used as
the maximum plotted time by which to scale other nodes.</p></li>
<li><p><strong>min_time</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>,</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The minimum plotted time value in the current
scaling system (see <code class="docutils literal notranslate"><span class="pre">time_scale</span></code>). Can be either a string or a
numeric value. If equal to <code class="docutils literal notranslate"><span class="pre">&quot;tree&quot;</span></code> (the default), the minimum time
is set to be that of the youngest node in the tree. If equal to <code class="docutils literal notranslate"><span class="pre">&quot;ts&quot;</span></code> the
minimum time is set to be the time of the youngest node in the tree
sequence; this is useful when drawing trees from the same tree sequence as it
ensures that node heights are consistent. If a numeric value, this is used as
the minimum plotted time.</p></li>
<li><p><strong>max_tree_height</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>,</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  Deprecated alias for max_time. (Deprecated in
0.3.6)</p></li>
<li><p><strong>node_labels</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>)</em>)  If specified, show custom labels for the nodes
(specified by ID) that are present in this map; any nodes not present will
not have a label. To use a metadata key, for example, use
<code class="docutils literal notranslate"><span class="pre">node_labels={node.id:</span> <span class="pre">node.metadata[&quot;key&quot;]</span> <span class="pre">for</span> <span class="pre">node</span> <span class="pre">in</span> <span class="pre">ts.nodes()}</span></code>.</p></li>
<li><p><strong>mutation_labels</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>)</em>)  If specified, show custom labels for the
mutations (specified by ID) that are present in the map; any mutations
not present will not have a label.</p></li>
<li><p><strong>node_titles</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>)</em>)  If specified, add a <code class="docutils literal notranslate"><span class="pre">&lt;title&gt;</span></code> string to
symbols for each node (specified by ID) present in this map. SVG visualizers
such as web browsers will commonly display this string on mousing over the
node symbol.</p></li>
<li><p><strong>mutation_titles</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>)</em>)  If specified, add a <code class="docutils literal notranslate"><span class="pre">&lt;title&gt;</span></code> string to
symbols for each mutation (specified by ID) present in this map. SVG
visualizers such as web browsers will commonly display this string on
mousing over the mutation symbol in the tree and (if show) on the x axis.</p></li>
<li><p><strong>root_svg_attributes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a>)  Additional attributes, such as an id, that will
be embedded in the root <code class="docutils literal notranslate"><span class="pre">&lt;svg&gt;</span></code> tag of the generated drawing.</p></li>
<li><p><strong>style</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A
<a class="reference external" href="https://www.w3.org/TR/CSS22/syndata.html">css style string</a> that will be
included in the <code class="docutils literal notranslate"><span class="pre">&lt;style&gt;</span></code> tag of the generated svg.</p></li>
<li><p><strong>order</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The left-to-right ordering of child nodes in the drawn tree.
This can be either: <code class="docutils literal notranslate"><span class="pre">&quot;minlex&quot;</span></code>, which minimises the differences
between adjacent trees (see also the <code class="docutils literal notranslate"><span class="pre">&quot;minlex_postorder&quot;</span></code> traversal
order for the <a class="reference internal" href="#tskit.Tree.nodes" title="tskit.Tree.nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">nodes()</span></code></a> method); or <code class="docutils literal notranslate"><span class="pre">&quot;tree&quot;</span></code> which draws trees
in the left-to-right order defined by the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">quintuply linked tree structure</span></a>.
If not specified or None, this defaults to <code class="docutils literal notranslate"><span class="pre">&quot;minlex&quot;</span></code>.</p></li>
<li><p><strong>force_root_branch</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If <code class="docutils literal notranslate"><span class="pre">True</span></code> always plot a branch (edge) above the
root(s) in the tree. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) then only plot such root branches
if there is a mutation above a root of the tree.</p></li>
<li><p><strong>symbol_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  Change the default size of the node and mutation
plotting symbols. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) use a standard size.</p></li>
<li><p><strong>x_axis</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Should the plot have an X axis line, showing the start and
end position of this tree along the genome. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) do not
plot an X axis.</p></li>
<li><p><strong>x_label</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  Place a label under the plot. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) and
there is an X axis, create and place an appropriate label.</p></li>
<li><p><strong>x_regions</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a>)  A dictionary mapping (left, right) tuples to names. This
draws a box, labelled with the name, on the X axis between the left and
right positions, and can be used for annotating genomic regions (e.g.
genes) on the X axis. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) do not plot any regions.</p></li>
<li><p><strong>y_axis</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Should the plot have an Y axis line, showing time (or
ranked node time if <code class="docutils literal notranslate"><span class="pre">time_scale=&quot;rank&quot;</span></code>). If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default)
do not plot a Y axis.</p></li>
<li><p><strong>y_label</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  Place a label to the left of the plot. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default)
and there is a Y axis,  create and place an appropriate label.</p></li>
<li><p><strong>y_ticks</strong> (<em>Union</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a><em>]</em>)  A list of Y values at which to plot
tickmarks, or a dictionary mapping Y values to labels (<code class="docutils literal notranslate"><span class="pre">[]</span></code> gives no
tickmarks). If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), plot one tickmark for each unique node
value. Note that if <code class="docutils literal notranslate"><span class="pre">time_scale=&quot;rank&quot;</span></code>, the Y values refer to the
zero-based rank of the plotted nodes, rather than the node time itself.</p></li>
<li><p><strong>y_gridlines</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to plot horizontal lines behind the tree
at each y tickmark.</p></li>
<li><p><strong>all_edge_mutations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  The edge on which a mutation occurs may span
multiple trees. If <code class="docutils literal notranslate"><span class="pre">False</span></code> or <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) mutations are only drawn
on an edge if their site position exists within the genomic interval covered
by this tree. If <code class="docutils literal notranslate"><span class="pre">True</span></code>, all mutations on each edge of the tree are drawn,
even if their genomic position is to the left or right of the tree
itself. Note that this means that independent drawings of different trees
from the same tree sequence may share some plotted mutations.</p></li>
<li><p><strong>omit_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, omit sites and mutations from the drawing.
Default: False</p></li>
<li><p><strong>canvas_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.14)"><em>tuple</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>)</em>)  The (width, height) of the SVG canvas.
This will change the SVG width and height without rescaling graphical
elements, allowing extra room e.g. for unusually long labels. If <code class="docutils literal notranslate"><span class="pre">None</span></code>
take the canvas size to be the same as the target drawing size (see
<code class="docutils literal notranslate"><span class="pre">size</span></code>, above). Default: None</p></li>
<li><p><strong>preamble</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  SVG commands to be included at the start of the returned
object, immediately after the opening tag. These can include custom svg
elements such as legends or annotations or even entire <code class="docutils literal notranslate"><span class="pre">&lt;svg&gt;</span></code> elements.
The preamble is not checked for validity, so it is up to the user to
ensure that it is valid SVG. Default: None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An SVG representation of a tree.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.SVGString" title="tskit.SVGString">SVGString</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.draw">
<span class="sig-name descname"><span class="pre">draw</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">height</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_colours</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation_colours</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_colours</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tree_height_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_time</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_time</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_tree_height</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">omit_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.draw"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.draw" title="Link to this definition">#</a></dt>
<dd><p>Returns a drawing of this tree.</p>
<p>When working in a Jupyter notebook, use the <code class="docutils literal notranslate"><span class="pre">IPython.display.SVG</span></code>
function to display the SVG output from this function inline in the notebook:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">SVG</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">draw</span><span class="p">())</span>
</pre></div>
</div>
<p>The unicode format uses unicode <a class="reference external" href="https://en.wikipedia.org/wiki/Box-drawing_character">box drawing characters</a> to render the tree.
This allows rendered trees to be printed out to the terminal:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>print(tree.draw(format=&quot;unicode&quot;))
  6

   5
 
   4
  
3 0 1 2
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">node_labels</span></code> argument allows the user to specify custom labels
for nodes, or no labels at all:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>print(tree.draw(format=&quot;unicode&quot;, node_labels={}))
  

   
 
   
  
   
</pre></div>
</div>
<p>Note: in some environments such as Jupyter notebooks with Windows or Mac,
users have observed that the Unicode box drawings can be misaligned. In
these cases, we recommend using the SVG or ASCII display formats instead.
If you have a strong preference for aligned Unicode, you can try out the
solution documented
<a class="reference external" href="https://github.com/tskit-dev/tskit/issues/189#issuecomment-499114811">here</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The path to the file to write the output. If None, do not
write to file.</p></li>
<li><p><strong>width</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The width of the image in pixels. If not specified, either
defaults to the minimum size required to depict the tree (text formats)
or 200 pixels.</p></li>
<li><p><strong>height</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The height of the image in pixels. If not specified, either
defaults to the minimum size required to depict the tree (text formats)
or 200 pixels.</p></li>
<li><p><strong>node_labels</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a>)  If specified, show custom labels for the nodes
that are present in the map. Any nodes not specified in the map will
not have a node label.</p></li>
<li><p><strong>node_colours</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a>)  If specified, show custom colours for the nodes
given in the map. Any nodes not specified in the map will take the default
colour; a value of <code class="docutils literal notranslate"><span class="pre">None</span></code> is treated as transparent and hence the node
symbol is not plotted. (Only supported in the SVG format.)</p></li>
<li><p><strong>mutation_labels</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a>)  If specified, show custom labels for the mutations
(specified by ID) that are present in the map. Any mutations not in the map
will not have a label. (Showing mutations is currently only supported in the
SVG format)</p></li>
<li><p><strong>mutation_colours</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a>)  If specified, show custom colours for the mutations
given in the map (specified by ID). As for <code class="docutils literal notranslate"><span class="pre">node_colours</span></code>, mutations not
present in the map take the default colour, and those mapping to <code class="docutils literal notranslate"><span class="pre">None</span></code>
are not drawn. (Only supported in the SVG format.)</p></li>
<li><p><strong>format</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The format of the returned image. Currently supported
are svg, ascii and unicode. Note that the <a class="reference internal" href="#tskit.Tree.draw_svg" title="tskit.Tree.draw_svg"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.draw_svg()</span></code></a>
method provides more comprehensive functionality for creating SVGs.</p></li>
<li><p><strong>edge_colours</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a>)  If specified, show custom colours for the edge
joining each node in the map to its parent. As for <code class="docutils literal notranslate"><span class="pre">node_colours</span></code>,
unspecified edges take the default colour, and <code class="docutils literal notranslate"><span class="pre">None</span></code> values result in the
edge being omitted. (Only supported in the SVG format.)</p></li>
<li><p><strong>time_scale</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  Control how height values for nodes are computed.
If this is equal to <code class="docutils literal notranslate"><span class="pre">&quot;time&quot;</span></code>, node heights are proportional to their time
values. If this is equal to <code class="docutils literal notranslate"><span class="pre">&quot;log_time&quot;</span></code>, node heights are proportional to
their log(time) values. If it is equal to <code class="docutils literal notranslate"><span class="pre">&quot;rank&quot;</span></code>, node heights are spaced
equally according to their ranked times. For SVG output the default is
time-scale whereas for text output the default is rank-scale.
Time scaling is not currently supported for text output.</p></li>
<li><p><strong>tree_height_scale</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  Deprecated alias for time_scale. (Deprecated in
0.3.6)</p></li>
<li><p><strong>max_time</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>,</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The maximum time value in the current
scaling system (see <code class="docutils literal notranslate"><span class="pre">time_scale</span></code>). Can be either a string or a
numeric value. If equal to <code class="docutils literal notranslate"><span class="pre">&quot;tree&quot;</span></code>, the maximum time is set to be
that of the oldest root in the tree. If equal to <code class="docutils literal notranslate"><span class="pre">&quot;ts&quot;</span></code> the maximum
time is set to be the time of the oldest root in the tree sequence;
this is useful when drawing trees from the same tree sequence as it ensures
that node heights are consistent. If a numeric value, this is used as the
maximum time by which to scale other nodes. This parameter
is not currently supported for text output.</p></li>
<li><p><strong>min_time</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>,</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The minimum time value in the current
scaling system (see <code class="docutils literal notranslate"><span class="pre">time_scale</span></code>). Can be either a string or a
numeric value. If equal to <code class="docutils literal notranslate"><span class="pre">&quot;tree&quot;</span></code>, the minimum time is set to be
that of the youngest node in the tree. If equal to <code class="docutils literal notranslate"><span class="pre">&quot;ts&quot;</span></code> the minimum
time is set to be the time of the youngest node in the tree sequence;
this is useful when drawing trees from the same tree sequence as it ensures
that node heights are consistent. If a numeric value, this is used as the
minimum time to display. This parameter is not currently supported for text
output.</p></li>
<li><p><strong>max_tree_height</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  Deprecated alias for max_time. (Deprecated in
0.3.6)</p></li>
<li><p><strong>order</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The left-to-right ordering of child nodes in the drawn tree.
This can be either: <code class="docutils literal notranslate"><span class="pre">&quot;minlex&quot;</span></code>, which minimises the differences
between adjacent trees (see also the <code class="docutils literal notranslate"><span class="pre">&quot;minlex_postorder&quot;</span></code> traversal
order for the <a class="reference internal" href="#tskit.Tree.nodes" title="tskit.Tree.nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">nodes()</span></code></a> method); or <code class="docutils literal notranslate"><span class="pre">&quot;tree&quot;</span></code> which draws trees
in the left-to-right order defined by the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">quintuply linked tree structure</span></a>.
If not specified or None, this defaults to <code class="docutils literal notranslate"><span class="pre">&quot;minlex&quot;</span></code>.</p></li>
<li><p><strong>omit_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, omit sites and mutations from the drawing
(only relevant to the SVG format). Default: False</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A representation of this tree in the requested format.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.num_mutations">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_mutations</span></span><a class="headerlink" href="#tskit.Tree.num_mutations" title="Link to this definition">#</a></dt>
<dd><p>Returns the total number of mutations across all sites on this tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The total number of mutations over all sites on this tree.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Tree.num_sites">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_sites</span></span><a class="headerlink" href="#tskit.Tree.num_sites" title="Link to this definition">#</a></dt>
<dd><p>Returns the number of sites on this tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of sites on this tree.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.sites">
<span class="sig-name descname"><span class="pre">sites</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.sites"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.sites" title="Link to this definition">#</a></dt>
<dd><p>Returns an iterator over all the <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">sites</span></a>
in this tree. Sites are returned in order of increasing ID
(and also position). See the <a class="reference internal" href="#tskit.Site" title="tskit.Site"><code class="xref py py-class docutils literal notranslate"><span class="pre">Site</span></code></a> class for details on
the available fields for each site.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An iterator over all sites in this tree.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.mutations">
<span class="sig-name descname"><span class="pre">mutations</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.mutations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.mutations" title="Link to this definition">#</a></dt>
<dd><p>Returns an iterator over all the
<a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">mutations</span></a> in this tree.
Mutations are returned in their
<a class="reference internal" href="data-model.html#sec-mutation-requirements"><span class="std std-ref">order in the mutations table</span></a>,
that is, by nondecreasing site ID, and within a site, by decreasing
mutation time with parent mutations before their children.
See the <a class="reference internal" href="#tskit.Mutation" title="tskit.Mutation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mutation</span></code></a> class for details on the available fields for
each mutation.</p>
<p>The returned iterator is equivalent to iterating over all sites
and all mutations in each site, i.e.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">sites</span><span class="p">():</span>
    <span class="k">for</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">site</span><span class="o">.</span><span class="n">mutations</span><span class="p">:</span>
        <span class="k">yield</span> <span class="n">mutation</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An iterator over all <a class="reference internal" href="#tskit.Mutation" title="tskit.Mutation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mutation</span></code></a> objects in this tree.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>iter(<a class="reference internal" href="#tskit.Mutation" title="tskit.Mutation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mutation</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.leaves">
<span class="sig-name descname"><span class="pre">leaves</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.leaves"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.leaves" title="Link to this definition">#</a></dt>
<dd><p>Returns an iterator over all the leaves in this tree that descend from
the specified node. If <span class="math notranslate nohighlight">\(u\)</span>  is not specified, return all leaves on
the tree (i.e. all leaves reachable from the tree root(s), see note below).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><span class="math notranslate nohighlight">\(u\)</span> can be any node in the entire tree sequence, including ones
which are not connected via branches to a root node of the tree. If
called on such a node, the iterator will return dead leaves
(see <a class="reference internal" href="data-model.html#sec-data-model-tree-dead-leaves-and-branches"><span class="std std-ref">Dead leaves and branches</span></a>) which cannot
be reached from a root of this tree. However, dead leaves will never be
returned if <span class="math notranslate nohighlight">\(u\)</span> is left unspecified.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An iterator over all leaves in the subtree rooted at u.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable" title="(in Python v3.14)">collections.abc.Iterable</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.samples">
<span class="sig-name descname"><span class="pre">samples</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.samples"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.samples" title="Link to this definition">#</a></dt>
<dd><p>Returns an iterator over the numerical IDs of all the sample nodes in
this tree that are underneath the node with ID <code class="docutils literal notranslate"><span class="pre">u</span></code>. If <code class="docutils literal notranslate"><span class="pre">u</span></code> is a sample,
it is included in the returned iterator. If <code class="docutils literal notranslate"><span class="pre">u</span></code> is not a sample, it is
possible for the returned iterator to be empty, for example if <code class="docutils literal notranslate"><span class="pre">u</span></code> is an
<a class="reference internal" href="#tskit.Tree.is_isolated" title="tskit.Tree.is_isolated"><code class="xref py py-meth docutils literal notranslate"><span class="pre">isolated</span></code></a> node that is not part of the the current
topology. If u is not specified, return all sample node IDs in the tree
(equivalent to all the sample node IDs in the tree sequence).</p>
<p>If the <a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.trees()</span></code></a> method is called with
<code class="docutils literal notranslate"><span class="pre">sample_lists=True</span></code>, this method uses an efficient algorithm to find
the sample nodes. If not, a simple traversal based method is used.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The iterator is <em>not</em> guaranteed to return the sample node IDs in
numerical or any other particular order.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An iterator over all sample node IDs in the subtree rooted at u.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable" title="(in Python v3.14)">collections.abc.Iterable</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.num_children">
<span class="sig-name descname"><span class="pre">num_children</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.num_children"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.num_children" title="Link to this definition">#</a></dt>
<dd><p>Returns the number of children of the specified
node (i.e., <code class="docutils literal notranslate"><span class="pre">len(tree.children(u))</span></code>)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The number of immediate children of the node u in this tree.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.num_samples">
<span class="sig-name descname"><span class="pre">num_samples</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.num_samples"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.num_samples" title="Link to this definition">#</a></dt>
<dd><p>Returns the number of sample nodes in this tree underneath the specified
node (including the node itself). If u is not specified return
the total number of samples in the tree.</p>
<p>This is a constant time operation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The number of samples in the subtree rooted at u.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.num_tracked_samples">
<span class="sig-name descname"><span class="pre">num_tracked_samples</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.num_tracked_samples"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.num_tracked_samples" title="Link to this definition">#</a></dt>
<dd><p>Returns the number of samples in the set specified in the
<code class="docutils literal notranslate"><span class="pre">tracked_samples</span></code> parameter of the <a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.trees()</span></code></a> method
underneath the specified node. If the input node is not specified,
return the total number of tracked samples in the tree.</p>
<p>This is a constant time operation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The number of samples within the set of tracked samples in
the subtree rooted at u.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.preorder">
<span class="sig-name descname"><span class="pre">preorder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.preorder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.preorder" title="Link to this definition">#</a></dt>
<dd><p>Returns a numpy array of node ids in <a class="reference external" href="https://en.wikipedia.org/wiki/Tree_traversal#Pre-order_(NLR)">preorder</a>. If the node u
is specified the traversal is rooted at this node (and it will be the first
element in the returned array). Otherwise, all nodes reachable from the tree
roots will be returned. See <a class="reference external" href="https://tskit.dev/tutorials/analysing_trees.html#sec-analysing-trees-traversals" title="(in Project name not set)"><span>Tree traversals</span></a> for
examples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  If specified, return all nodes in the subtree rooted at u
(including u) in traversal order.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Array of node ids</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a> (dtype=np.int32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.postorder">
<span class="sig-name descname"><span class="pre">postorder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.postorder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.postorder" title="Link to this definition">#</a></dt>
<dd><p>Returns a numpy array of node ids in <a class="reference external" href="https://en.wikipedia.org/wiki/Tree_traversal##Post-order_(LRN)">postorder</a>. If the node u
is specified the traversal is rooted at this node (and it will be the last
element in the returned array). Otherwise, all nodes reachable from the tree
roots will be returned. See <a class="reference external" href="https://tskit.dev/tutorials/analysing_trees.html#sec-analysing-trees-traversals" title="(in Project name not set)"><span>Tree traversals</span></a> for
examples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  If specified, return all nodes in the subtree rooted at u
(including u) in traversal order.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Array of node ids</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a> (dtype=np.int32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.timeasc">
<span class="sig-name descname"><span class="pre">timeasc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.timeasc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.timeasc" title="Link to this definition">#</a></dt>
<dd><p>Returns a numpy array of node ids. Starting at <cite>u</cite>, returns the reachable
descendant nodes in order of increasing time (most recent first), falling back
to increasing ID if times are equal. Also see
<a class="reference external" href="https://tskit.dev/tutorials/analysing_trees.html#sec-analysing-trees-traversals" title="(in Project name not set)"><span>Tree traversals</span></a> for examples of how to use
traversals.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  If specified, return all nodes in the subtree rooted at u
(including u) in traversal order.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Array of node ids</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a> (dtype=np.int32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.timedesc">
<span class="sig-name descname"><span class="pre">timedesc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.timedesc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.timedesc" title="Link to this definition">#</a></dt>
<dd><p>Returns a numpy array of node ids. Starting at <cite>u</cite>, returns the reachable
descendant nodes in order of decreasing time (least recent first), falling back
to decreasing ID if times are equal. Also see
<a class="reference external" href="https://tskit.dev/tutorials/analysing_trees.html#sec-analysing-trees-traversals" title="(in Project name not set)"><span>Tree traversals</span></a> for examples of how to use
traversals.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  If specified, return all nodes in the subtree rooted at u
(including u) in traversal order.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Array of node ids</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a> (dtype=np.int32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.nodes">
<span class="sig-name descname"><span class="pre">nodes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">root</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'preorder'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.nodes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.nodes" title="Link to this definition">#</a></dt>
<dd><p>Returns an iterator over the node IDs reachable from the specified node in this
tree in the specified traversal order.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Unlike the <a class="reference internal" href="#tskit.TreeSequence.nodes" title="tskit.TreeSequence.nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.nodes()</span></code></a> method, this iterator produces
integer node IDs, not <a class="reference internal" href="#tskit.Node" title="tskit.Node"><code class="xref py py-class docutils literal notranslate"><span class="pre">Node</span></code></a> objects.</p>
</div>
<p>If the <code class="docutils literal notranslate"><span class="pre">root</span></code> parameter is not provided or <code class="docutils literal notranslate"><span class="pre">None</span></code>, iterate over all
nodes reachable from the roots (see <a class="reference internal" href="#tskit.Tree.roots" title="tskit.Tree.roots"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Tree.roots</span></code></a> for details
on which nodes are considered roots). If the <code class="docutils literal notranslate"><span class="pre">root</span></code> parameter
is provided, only the nodes in the subtree rooted at this node
(including the specified node) will be iterated over. If the
<a class="reference internal" href="#tskit.Tree.virtual_root" title="tskit.Tree.virtual_root"><code class="xref py py-attr docutils literal notranslate"><span class="pre">virtual_root</span></code></a> is specified as the traversal root, it will
be included in the traversed nodes in the appropriate position
for the given ordering. (See the
<a class="reference internal" href="data-model.html#sec-data-model-tree-virtual-root"><span class="std std-ref">tree roots</span></a> section for more
information on the virtual root.)</p>
<p>The <code class="docutils literal notranslate"><span class="pre">order</span></code> parameter defines the order in which tree nodes are visited
in the iteration (also see the <a class="reference external" href="https://tskit.dev/tutorials/analysing_trees.html#sec-analysing-trees-traversals" title="(in Project name not set)"><span>Tree traversals</span></a> section
in the <a class="reference external" href="https://tskit.dev/tutorials">tutorials</a>). The available orders are:</p>
<ul class="simple">
<li><p>preorder: starting at root, yield the current node, then recurse
and do a preorder on each child of the current node. See also <a class="reference external" href="https://en.wikipedia.org/wiki/Tree_traversal#Pre-order_(NLR)">Wikipedia</a>.</p></li>
<li><p>inorder: starting at root, assuming binary trees, recurse and do
an inorder on the first child, then yield the current node, then
recurse and do an inorder on the second child. In the case of <code class="docutils literal notranslate"><span class="pre">n</span></code>
child nodes (not necessarily 2), the first <code class="docutils literal notranslate"><span class="pre">n</span> <span class="pre">//</span> <span class="pre">2</span></code> children are
visited in the first stage, and the remaining <code class="docutils literal notranslate"><span class="pre">n</span> <span class="pre">-</span> <span class="pre">n</span> <span class="pre">//</span> <span class="pre">2</span></code> children
are visited in the second stage. See also <a class="reference external" href="https://en.wikipedia.org/wiki/Tree_traversal#In-order_(LNR)">Wikipedia</a>.</p></li>
<li><p>postorder: starting at root, recurse and do a postorder on each
child of the current node, then yield the current node. See also
<a class="reference external" href="https://en.wikipedia.org/wiki/Tree_traversal#Post-order_(LRN)">Wikipedia</a>.</p></li>
<li><p>levelorder (breadthfirst): visit the nodes under root (including
the root) in increasing order of their depth from root. See also
<a class="reference external" href="https://en.wikipedia.org/wiki/Tree_traversal#Breadth-first_search_/_level_order">Wikipedia</a>.</p></li>
<li><p>timeasc: visits the nodes in order of increasing time, falling back to
increasing ID if times are equal.</p></li>
<li><p>timedesc: visits the nodes in order of decreasing time, falling back to
decreasing ID if times are equal.</p></li>
<li><p>minlex_postorder: a usual postorder has ambiguity in the order in
which children of a node are visited. We constrain this by outputting
a postorder such that the leaves visited, when their IDs are
listed out, have minimum <a class="reference external" href="https://en.wikipedia.org/wiki/Lexicographical_order">lexicographic order</a> out of all valid
traversals. This traversal is useful for drawing multiple trees of
a <code class="docutils literal notranslate"><span class="pre">TreeSequence</span></code>, as it leads to more consistency between adjacent
trees. Note that internal non-leaf nodes are not counted in
assessing the lexicographic order.</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>root</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The root of the subtree we are traversing.</p></li>
<li><p><strong>order</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The traversal ordering. Currently preorder,
inorder, postorder, levelorder (breadthfirst), timeasc and
timedesc and minlex_postorder are supported.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An iterator over the node IDs in the tree in some traversal order.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable" title="(in Python v3.14)">collections.abc.Iterable</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.as_newick">
<span class="sig-name descname"><span class="pre">as_newick</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">root</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">precision</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_branch_lengths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.as_newick"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.as_newick" title="Link to this definition">#</a></dt>
<dd><p>Returns a <a class="reference external" href="https://en.wikipedia.org/wiki/Newick_format">newick encoding</a> of this tree.
For example, a binary tree with 3 leaves generated by
<a class="reference internal" href="#tskit.Tree.generate_balanced" title="tskit.Tree.generate_balanced"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.generate_balanced(3)</span></code></a>
encodes as:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">(</span><span class="n">n0</span><span class="p">:</span><span class="mi">2</span><span class="p">,(</span><span class="n">n1</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span><span class="n">n2</span><span class="p">:</span><span class="mi">1</span><span class="p">):</span><span class="mi">1</span><span class="p">);</span>
</pre></div>
</div>
<p>By default <a class="reference internal" href="glossary.html#sec-data-model-definitions"><span class="std std-ref">sample nodes</span></a> are
labelled using the form <code class="docutils literal notranslate"><span class="pre">f&quot;n{node_id}&quot;</span></code>, i.e. the sample nodes
ID prefixed with the string <code class="docutils literal notranslate"><span class="pre">&quot;n&quot;</span></code>. Node labels can be specified
explicitly using the <code class="docutils literal notranslate"><span class="pre">node_labels</span></code> argument, which is a mapping from
integer node IDs to the corresponding string label. If a node is not
present in the mapping, no label is associated with the node in
output.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Node labels are <strong>not</strong> Newick escaped, so care must be taken
to provide labels that will not break the encoding.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Specifying a <code class="docutils literal notranslate"><span class="pre">node_labels</span></code> dictionary or setting
<code class="docutils literal notranslate"><span class="pre">include_branch_lengths=False</span></code> results in a less efficient
method being used to generate the newick output. The performance
difference can be substantial for large trees.</p>
</div>
<p>By default, branch lengths are printed out with sufficient precision
for them to be recovered exactly in double precision (although note
that this does not necessarily mean that we can precisely recover the
corresponding node times, since branch lengths are obtained by
subtraction). If all times on the tree sequence are discrete, then
branch lengths are printed as integers. Otherwise, branch lengths are
printed with 17 digits of precision (i.e., <code class="docutils literal notranslate"><span class="pre">&quot;%.17f&quot;</span></code> in
printf-notation).</p>
<p>The precision for branch lengths can be specified using the <code class="docutils literal notranslate"><span class="pre">precision</span></code>
argument. Branch lengths can be omitted entirely by setting
<code class="docutils literal notranslate"><span class="pre">include_branch_lengths=False</span></code>.</p>
<p>If the <code class="docutils literal notranslate"><span class="pre">root</span></code> argument is specified, we return the newick encoding of
the specified subtree, otherwise the full tree is returned. If the tree
has <a class="reference internal" href="data-model.html#sec-data-model-tree-roots"><span class="std std-ref">multiple roots</span></a> and a root node
is not explicitly specified, we raise a <code class="docutils literal notranslate"><span class="pre">ValueError</span></code>. This is because
most libraries and downstream software consider a newick string that
contains multiple disconnected subtrees an error, and it is therefore
best to consider how such topologies should be interchanged on a
case-by-base basis. A list of the newick strings for each root can be
obtained by <code class="docutils literal notranslate"><span class="pre">[tree.as_newick(root=root)</span> <span class="pre">for</span> <span class="pre">root</span> <span class="pre">in</span> <span class="pre">tree.roots]</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>precision</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The numerical precision with which branch lengths are
printed. If not specified or None default to 0 if the tree sequence
contains only integer node times, or 17 otherwise.</p></li>
<li><p><strong>root</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  If specified, return the tree rooted at this node.</p></li>
<li><p><strong>node_labels</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a>)  If specified, show custom labels for the nodes
that are present in the map. Any nodes not specified in the map will
not have a node label.</p></li>
<li><p><strong>include_branch_lengths</strong>  If True (default), output branch lengths in the
Newick string. If False, only output the topology, without branch lengths.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A newick representation of this tree.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.newick">
<span class="sig-name descname"><span class="pre">newick</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">precision</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">14</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">root</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_branch_lengths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.newick"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.newick" title="Link to this definition">#</a></dt>
<dd><div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method is deprecated and may be removed in future
versions of tskit. Please use the <a class="reference internal" href="#tskit.Tree.as_newick" title="tskit.Tree.as_newick"><code class="xref py py-meth docutils literal notranslate"><span class="pre">as_newick()</span></code></a> method
in new code.</p>
</div>
<p>This method is a deprecated version of the <a class="reference internal" href="#tskit.Tree.as_newick" title="tskit.Tree.as_newick"><code class="xref py py-meth docutils literal notranslate"><span class="pre">as_newick()</span></code></a> method.
Functionality is equivalent, except for the default node labels.</p>
<p>By default, <em>leaf</em> nodes are labelled with their numerical ID + 1,
and internal nodes are not labelled. This default strategy was originally
used to mimic the output of the <code class="docutils literal notranslate"><span class="pre">ms</span></code> simulator. However, the choice
of labelling leaf nodes rather than samples is problematic, and this
behaviour is only retained to avoid breaking existing code which may
rely on it.</p>
<p>Other parameters behave as documented in the <a class="reference internal" href="#tskit.Tree.as_newick" title="tskit.Tree.as_newick"><code class="xref py py-meth docutils literal notranslate"><span class="pre">as_newick()</span></code></a> method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>precision</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The numerical precision with which branch lengths are
printed. Defaults to 14.</p></li>
<li><p><strong>root</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  If specified, return the tree rooted at this node.</p></li>
<li><p><strong>node_labels</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a>)  If specified, show custom labels for the nodes
that are present in the map. Any nodes not specified in the map will
not have a node label.</p></li>
<li><p><strong>include_branch_lengths</strong>  If True (default), output branch lengths in the
Newick string. If False, only output the topology, without branch lengths.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A newick representation of this tree.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.as_dict_of_dicts">
<span class="sig-name descname"><span class="pre">as_dict_of_dicts</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.as_dict_of_dicts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.as_dict_of_dicts" title="Link to this definition">#</a></dt>
<dd><p>Convert tree to dict of dicts for conversion to a
<a class="reference external" href="https://networkx.github.io/documentation/stable/reference/classes/digraph.html">networkx graph</a>.</p>
<p>For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">networkx</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">nx</span>
<span class="n">nx</span><span class="o">.</span><span class="n">DiGraph</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">as_dict_of_dicts</span><span class="p">())</span>
<span class="c1"># undirected graphs work as well</span>
<span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">as_dict_of_dicts</span><span class="p">())</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary of dictionaries of dictionaries where the first key
is the source, the second key is the target of an edge, and the
third key is an edge annotation. At this point the only annotation
is branch_length, the length of the branch (in units of time).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.__str__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.__str__" title="Link to this definition">#</a></dt>
<dd><p>Return a plain text summary of a tree in a tree sequence</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree._repr_html_">
<span class="sig-name descname"><span class="pre">_repr_html_</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree._repr_html_"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree._repr_html_" title="Link to this definition">#</a></dt>
<dd><p>Return an html summary of a tree in a tree sequence. Called by jupyter
notebooks to render trees</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.map_mutations">
<span class="sig-name descname"><span class="pre">map_mutations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genotypes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alleles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ancestral_state</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.map_mutations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.map_mutations" title="Link to this definition">#</a></dt>
<dd><p>Given observations for the samples in this tree described by the specified
set of genotypes and alleles, return a parsimonious set of state transitions
explaining these observations. The genotypes array is interpreted as indexes
into the alleles list in the same manner as described in the
<a class="reference internal" href="#tskit.TreeSequence.variants" title="tskit.TreeSequence.variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.variants()</span></code></a> method. Thus, if sample <code class="docutils literal notranslate"><span class="pre">j</span></code> carries the
allele at index <code class="docutils literal notranslate"><span class="pre">k</span></code>, then we have <code class="docutils literal notranslate"><span class="pre">genotypes[j]</span> <span class="pre">=</span> <span class="pre">k</span></code>.
Missing observations can be specified for a sample using the value
<code class="docutils literal notranslate"><span class="pre">tskit.MISSING_DATA</span></code> (-1), in which case the state at this sample does not
influence the ancestral state or the position of mutations returned. At least
one non-missing observation must be provided. A maximum of 64 alleles are
supported.</p>
<p>The current implementation uses the Hartigan parsimony algorithm to determine
the minimum number of state transitions required to explain the data. In this
model, transitions between any of the non-missing states is equally likely.</p>
<p>The returned values correspond directly to the data model for describing
variation at sites using mutations. See the <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">Site Table</span></a>
and <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">Mutation Table</span></a> definitions for details and background.</p>
<p>The state reconstruction is returned as two-tuple, <code class="docutils literal notranslate"><span class="pre">(ancestral_state,</span>
<span class="pre">mutations)</span></code>, where <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> is the allele assigned to the
tree root(s) and <code class="docutils literal notranslate"><span class="pre">mutations</span></code> is a list of <a class="reference internal" href="#tskit.Mutation" title="tskit.Mutation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mutation</span></code></a> objects,
ordered as <a class="reference internal" href="data-model.html#sec-mutation-requirements"><span class="std std-ref">required in a mutation table</span></a>.
For each mutation, <code class="docutils literal notranslate"><span class="pre">derived_state</span></code> is the new state after this mutation and
<code class="docutils literal notranslate"><span class="pre">node</span></code> is the tree node immediately beneath the mutation (if there are unary
nodes between two branch points, hence multiple nodes above which the
mutation could be parsimoniously placed, the oldest node is used). The
<code class="docutils literal notranslate"><span class="pre">parent</span></code> property contains the index in the returned list of the previous
mutation on the path to root, or <code class="docutils literal notranslate"><span class="pre">tskit.NULL</span></code>
if there are no previous mutations (see the <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">Mutation Table</span></a>
for more information on the concept of mutation parents). All other attributes
of the <a class="reference internal" href="#tskit.Mutation" title="tskit.Mutation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mutation</span></code></a> object are undefined and should not be used.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Sample states observed as missing in the input <code class="docutils literal notranslate"><span class="pre">genotypes</span></code> need
not correspond to samples whose nodes are actually missing (i.e.,
<a class="reference internal" href="data-model.html#sec-data-model-missing-data"><span class="std std-ref">isolated</span></a>) in the tree. In this
case, mapping the mutations returned by this method onto the tree
will result in these missing observations being imputed to the
most parsimonious state.</p>
</div>
<p>Because the <code class="docutils literal notranslate"><span class="pre">parent</span></code> in the returned list of mutations refers to the index
in that list, if you are adding mutations to an existing tree sequence, you
will need to maintain a map of list IDs to the newly added mutations, for
instance:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">last_tree</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">last</span><span class="p">()</span>
<span class="n">anc_state</span><span class="p">,</span> <span class="n">parsimonious_muts</span> <span class="o">=</span> <span class="n">last_tree</span><span class="o">.</span><span class="n">map_mutations</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;T&quot;</span><span class="p">))</span>
<span class="c1"># Edit the tree sequence, see the &quot;Tables and Editing&quot; tutorial</span>
<span class="n">tables</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">dump_tables</span><span class="p">()</span>
<span class="c1"># add a new site at the end of ts, assumes there isn&#39;t one there already</span>
<span class="n">site_id</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">anc_state</span><span class="p">)</span>

<span class="n">mut_id_map</span> <span class="o">=</span> <span class="p">{</span><span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span><span class="p">:</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span><span class="p">}</span>  <span class="c1"># don&#39;t change if parent id is -1</span>
<span class="k">for</span> <span class="n">list_id</span><span class="p">,</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">parsimonious_muts</span><span class="p">):</span>
    <span class="n">mut_id_map</span><span class="p">[</span><span class="n">list_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
        <span class="n">mutation</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">site</span><span class="o">=</span><span class="n">site_id</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="n">mut_id_map</span><span class="p">[</span><span class="n">mutation</span><span class="o">.</span><span class="n">parent</span><span class="p">]))</span>
<span class="n">tables</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>  <span class="c1"># Redundant here, but needed if the site is not the last one</span>
<span class="n">new_ts</span> <span class="o">=</span> <span class="n">tables</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">()</span>
</pre></div>
</div>
<p>See the <a class="reference external" href="https://tskit.dev/tutorials/analysing_trees.html#sec-analysing-trees-parsimony" title="(in Project name not set)"><span>Parsimony</span></a> section in the tutorial
for further examples of how to use this method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genotypes</strong> (<em>array_like</em>)  The input observations for the samples in this tree.</p></li>
<li><p><strong>alleles</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.14)"><em>tuple</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>)</em>)  The alleles for the specified <code class="docutils literal notranslate"><span class="pre">genotypes</span></code>. Each
positive value in the <code class="docutils literal notranslate"><span class="pre">genotypes</span></code> array is treated as an index into this
list of alleles.</p></li>
<li><p><strong>ancestral_state</strong> (<em>Union</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>]</em>)  A fixed ancestral state, specified either as a
non-negative integer less than the number of alleles, or a string which
must be one of the <code class="docutils literal notranslate"><span class="pre">alleles</span></code> provided above. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) then
an ancestral state is chosen arbitrarily from among those that provide
the most parsimonious placement of mutations. Note that if the ancestral
state is specified, the placement of mutations may not be as parsimonious
as that which could be achieved by leaving the ancestral state unspecified;
additionally it may lead to mutations being placed above the root node(s) of
the tree (for example if all the samples have a genotype of 1 but the
ancestral state is fixed to be 0).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The inferred ancestral state and list of mutations on this tree
that encode the specified observations.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)">list</a>(<a class="reference internal" href="#tskit.Mutation" title="tskit.Mutation">tskit.Mutation</a>))</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.kc_distance">
<span class="sig-name descname"><span class="pre">kc_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lambda_</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.kc_distance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.kc_distance" title="Link to this definition">#</a></dt>
<dd><p>Returns the Kendall-Colijn distance between the specified pair of trees.
The <code class="docutils literal notranslate"><span class="pre">lambda_</span></code> parameter  determines the relative weight of topology
vs branch lengths in calculating the distance. If <code class="docutils literal notranslate"><span class="pre">lambda_</span></code> is 0
(the default) we only consider topology, and if it is 1 we only
consider branch lengths. See <a class="reference external" href="https://academic.oup.com/mbe/article/33/10/2735/2925548">Kendall &amp; Colijn (2016)</a> for details.</p>
<p>The trees we are comparing to must have identical lists of sample
nodes (i.e., the same IDs in the same order). The metric operates on
samples, not leaves, so internal samples are treated identically to
sample tips. Subtrees with no samples do not contribute to the metric.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><em>Tree</em></a>)  The other tree to compare to.</p></li>
<li><p><strong>lambda</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The KC metric lambda parameter determining the
relative weight of topology and branch length.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The computed KC distance between this tree and other.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.rf_distance">
<span class="sig-name descname"><span class="pre">rf_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.rf_distance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.rf_distance" title="Link to this definition">#</a></dt>
<dd><p>Returns the (unweighted) Robinson-Foulds distance between the specified pair
of trees, where corresponding samples between the two trees are identified by
node ID. The Robinson-Foulds distance (also known as the symmetric difference)
is defined as the number of bipartitions that are present in one tree but
not the other (see
<a class="reference external" href="https://doi.org/10.1016/0025-5564(81)90043-2">Robinson &amp; Foulds (1981)</a>).
This method returns the unnormalised RF distance: if the
trees are strictly bifurcating, i.e. binary, the value can be
normalised by dividing by the maximum, which is $2n-4$ for two rooted
trees of $n$ samples (however, if the trees contain polytomies, the maximum
RF distance is less easily defined).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The RF distance can be sensitive to small changes in topology: in some
cases, changing the position of a single leaf can result in the maximum
RF distance. Therefore even if adjacent trees in a tree sequence differ
by a single subtree-prune-and-regraft operation, the RF distance
between them can be large.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>other</strong> (<a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><em>Tree</em></a>)  The other tree to compare to. Trees are treated as rooted.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The unweighted Robinson-Foulds distance between this tree and <code class="docutils literal notranslate"><span class="pre">other</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.14)"><strong>ValueError</strong></a>  If either tree has multiple roots, or the trees have
different sample nodes.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.path_length">
<span class="sig-name descname"><span class="pre">path_length</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.path_length"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.path_length" title="Link to this definition">#</a></dt>
<dd><p>Returns the number of edges on the path in this tree between the two nodes.
If the two nodes have a most recent common ancestor, then this is defined as
<code class="docutils literal notranslate"><span class="pre">tree.depth(u)</span> <span class="pre">+</span> <span class="pre">tree.depth(v)</span> <span class="pre">-</span> <span class="pre">2</span> <span class="pre">*</span> <span class="pre">tree.depth(tree.mrca(u,</span> <span class="pre">v))</span></code>. If the nodes
do not have an MRCA (i.e., they are in disconnected subtrees) the path length
is infinity.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This counts the number of hops between two nodes. To find the branch
length distance between them, in units of time (i.e. the sum of edge lengths
that separate two nodes) use the <a class="reference internal" href="#tskit.Tree.distance_between" title="tskit.Tree.distance_between"><code class="xref py py-meth docutils literal notranslate"><span class="pre">distance_between()</span></code></a> method instead.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>See also the <a class="reference internal" href="#tskit.Tree.depth" title="tskit.Tree.depth"><code class="xref py py-meth docutils literal notranslate"><span class="pre">depth()</span></code></a> method</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The first node for path length computation.</p></li>
<li><p><strong>v</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The second node for path length computation.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The number of edges between the two nodes.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.distance_between">
<span class="sig-name descname"><span class="pre">distance_between</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">u</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.distance_between"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.distance_between" title="Link to this definition">#</a></dt>
<dd><p>Returns the total distance between two nodes in the tree, expressed as
the sum of branch lengths from both nodes to their most recent common ancestor.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>u</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The first node for path length computation.</p></li>
<li><p><strong>v</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The second node for path length computation.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The distance between the two nodes, the sum of branch lengths .</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.b1_index">
<span class="sig-name descname"><span class="pre">b1_index</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.b1_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.b1_index" title="Link to this definition">#</a></dt>
<dd><p>Returns the
<a class="reference external" href="https://treebalance.wordpress.com/b-index/">B1 balance index</a>
for this tree. This is defined as the inverse of the sum of all
longest paths to leaves for each node besides roots.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>See <a class="reference external" href="https://www.jstor.org/stable/2992186">Shao and Sokal (1990)</a> for details.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The B1 balance index.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.b2_index">
<span class="sig-name descname"><span class="pre">b2_index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">base</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.b2_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.b2_index" title="Link to this definition">#</a></dt>
<dd><p>Returns the
<a class="reference external" href="https://treebalance.wordpress.com/b-index/">B2 balance index</a>
this tree.
This is defined as the Shannon entropy of the probability
distribution to reach leaves assuming a random walk
from a root. The default base is 10, following Shao and Sokal (1990).</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>See <a class="reference external" href="https://www.jstor.org/stable/2992186">Shao and Sokal (1990)</a> for details.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>base</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The base used for the logarithm in the
Shannon entropy computation.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The B2 balance index.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.colless_index">
<span class="sig-name descname"><span class="pre">colless_index</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.colless_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.colless_index" title="Link to this definition">#</a></dt>
<dd><p>Returns the
<a class="reference external" href="https://treebalance.wordpress.com/colless-index/">Colless imbalance index</a>
for this tree. This is defined as the sum of all differences between
number of leaves subtended by the left and right child of each node.
The Colless index is undefined for non-binary trees and trees with
multiple roots. This method will raise a LibraryError if the tree is
not singly-rooted and binary.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>See <a class="reference external" href="https://www.jstor.org/stable/2992186">Shao and Sokal (1990)</a> for details.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The Colless imbalance index.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.sackin_index">
<span class="sig-name descname"><span class="pre">sackin_index</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.sackin_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.sackin_index" title="Link to this definition">#</a></dt>
<dd><p>Returns the
<a class="reference external" href="https://treebalance.wordpress.com/sackin-index/">Sackin imbalance index</a>
for this tree. This is defined as the sum of the depths of all leaves
in the tree. Equivalent to <code class="docutils literal notranslate"><span class="pre">sum(tree.depth(u)</span> <span class="pre">for</span> <span class="pre">u</span> <span class="pre">in</span>
<span class="pre">tree.leaves())</span></code></p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>See <a class="reference external" href="https://www.jstor.org/stable/2992186">Shao and Sokal (1990)</a> for details.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The Sackin imbalance index.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.num_lineages">
<span class="sig-name descname"><span class="pre">num_lineages</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.num_lineages"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.num_lineages" title="Link to this definition">#</a></dt>
<dd><p>Returns the number of lineages present in this tree at time <code class="docutils literal notranslate"><span class="pre">t</span></code>. This
is defined as the number of branches in this tree (reachable from the
samples) that intersect with <code class="docutils literal notranslate"><span class="pre">t</span></code>. Thus, <code class="docutils literal notranslate"><span class="pre">tree.num_lineages(t)</span></code>
is equal to 0 for any <code class="docutils literal notranslate"><span class="pre">t</span></code> greater than or equal to the time of
the root in a singly-rooted tree.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Note that this definition means that if a (non root) node
with three children has time <code class="docutils literal notranslate"><span class="pre">t</span></code>, then it will count as one lineage,
not three.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>t</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The time to count lineages at.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The number of lineages in the tree at time t.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.split_polytomies">
<span class="sig-name descname"><span class="pre">split_polytomies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epsilon</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.split_polytomies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.split_polytomies" title="Link to this definition">#</a></dt>
<dd><p>Return a new <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> where extra nodes and edges have been inserted
so that any any node <code class="docutils literal notranslate"><span class="pre">u</span></code> with greater than 2 children  a multifurcation
or polytomy  is resolved into successive bifurcations. New nodes are
inserted at times fractionally less than than the time of node <code class="docutils literal notranslate"><span class="pre">u</span></code>.
Times are allocated to different levels of the tree, such that any newly
inserted sibling nodes will have the same time.</p>
<p>By default, the times of the newly generated children of a particular
node are the minimum representable distance in floating point arithmetic
from their parents (using the <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.nextafter.html">nextafter</a>
function). Thus, the generated branches have the shortest possible nonzero
length. A fixed branch length between inserted nodes and their parents
can also be specified by using the <code class="docutils literal notranslate"><span class="pre">epsilon</span></code> parameter.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>A tree sequence <a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">requires</span></a> that
parents be older than children and that mutations are younger than the
parent of the edge on which they lie. If a fixed <code class="docutils literal notranslate"><span class="pre">epsilon</span></code> is specifed
and is not small enough compared to the distance between a polytomy and
its oldest child (or oldest child mutation) these requirements may not
be met. In this case an error will be raised.</p>
</div>
<p>If the <code class="docutils literal notranslate"><span class="pre">method</span></code> is <code class="docutils literal notranslate"><span class="pre">&quot;random&quot;</span></code> (currently the only option, and the default
when no method is specified), then for a node with <span class="math notranslate nohighlight">\(n\)</span> children, the
<span class="math notranslate nohighlight">\((2n - 3)! / (2^(n - 2) (n - 2!))\)</span> possible binary trees with equal
probability.</p>
<p>The returned <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> will have the same genomic span as this tree,
and node IDs will be conserved (that is, node <code class="docutils literal notranslate"><span class="pre">u</span></code> in this tree will
be the same node in the returned tree). The returned tree is derived from a
tree sequence that contains only one non-degenerate tree, that is, where
edges cover only the interval spanned by this tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>epsilon</strong>  If specified, the fixed branch length between inserted
nodes and their parents. If None (the default), the minimal possible
nonzero branch length is generated for each node.</p></li>
<li><p><strong>method</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The method used to break polytomies. Currently only random
is supported, which can also be specified by <code class="docutils literal notranslate"><span class="pre">method=None</span></code>
(Default: <code class="docutils literal notranslate"><span class="pre">None</span></code>).</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, add details of this operation to the
provenance information of the returned tree sequence. (Default: True).</p></li>
<li><p><strong>random_seed</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The random seed. If this is None, a random seed will
be automatically generated. Valid random seeds must be between 1 and
<span class="math notranslate nohighlight">\(2^32  1\)</span>.</p></li>
<li><p><strong>**kwargs</strong>  Further arguments used as parameters when constructing the
returned <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>. For example
<code class="docutils literal notranslate"><span class="pre">tree.split_polytomies(sample_lists=True)</span></code> will
return a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> created with <code class="docutils literal notranslate"><span class="pre">sample_lists=True</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A new tree with polytomies split into random bifurcations.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">tskit.Tree</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.generate_star">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">generate_star</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_leaves</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">branch_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.generate_star"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.generate_star" title="Link to this definition">#</a></dt>
<dd><p>Generate a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> whose leaf nodes all have the same parent (i.e.,
a star tree). The leaf nodes are all at time 0 and are marked as sample nodes.</p>
<p>The tree produced by this method is identical to
<code class="docutils literal notranslate"><span class="pre">tskit.Tree.unrank(n,</span> <span class="pre">(0,</span> <span class="pre">0))</span></code>, but generated more efficiently for large <code class="docutils literal notranslate"><span class="pre">n</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_leaves</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The number of leaf nodes in the returned tree (must be
2 or greater).</p></li>
<li><p><strong>span</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The span of the tree, and therefore the
<a class="reference internal" href="#tskit.TreeSequence.sequence_length" title="tskit.TreeSequence.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">sequence_length</span></code></a> of the <a class="reference internal" href="#tskit.Tree.tree_sequence" title="tskit.Tree.tree_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tree_sequence</span></code></a>
property of the returned <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>.</p></li>
<li><p><strong>branch_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The length of every branch in the tree (equivalent
to the time of the root node).</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, add details of this operation to the
provenance information of the returned tree sequence. (Default: True).</p></li>
<li><p><strong>**kwargs</strong>  Further arguments used as parameters when constructing the
returned <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>. For example
<code class="docutils literal notranslate"><span class="pre">tskit.Tree.generate_star(sample_lists=True)</span></code> will
return a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> created with <code class="docutils literal notranslate"><span class="pre">sample_lists=True</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A star-shaped tree. Its corresponding <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> is available
via the <a class="reference internal" href="#tskit.Tree.tree_sequence" title="tskit.Tree.tree_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tree_sequence</span></code></a> attribute.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">Tree</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.generate_balanced">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">generate_balanced</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_leaves</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">arity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">branch_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.generate_balanced"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.generate_balanced" title="Link to this definition">#</a></dt>
<dd><p>Generate a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> with the specified number of leaves that is maximally
balanced. By default, the tree returned is binary, such that for each
node that subtends <span class="math notranslate nohighlight">\(n\)</span> leaves, the left child will subtend
<span class="math notranslate nohighlight">\(\lfloor{n / 2}\rfloor\)</span> leaves and the right child the
remainder. Balanced trees with higher arity can also generated using the
<code class="docutils literal notranslate"><span class="pre">arity</span></code> parameter, where the leaves subtending a node are distributed
among its children analogously.</p>
<p>In the returned tree, the leaf nodes are all at time 0, marked as samples,
and labelled 0 to n from left-to-right. Internal node IDs are assigned
sequentially from n in a postorder traversal, and the time of an internal
node is the maximum time of its children plus the specified <code class="docutils literal notranslate"><span class="pre">branch_length</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_leaves</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The number of leaf nodes in the returned tree (must be
be 2 or greater).</p></li>
<li><p><strong>arity</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The maximum number of children a node can have in the returned
tree.</p></li>
<li><p><strong>span</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The span of the tree, and therefore the
<a class="reference internal" href="#tskit.TreeSequence.sequence_length" title="tskit.TreeSequence.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">sequence_length</span></code></a> of the <a class="reference internal" href="#tskit.Tree.tree_sequence" title="tskit.Tree.tree_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tree_sequence</span></code></a>
property of the returned <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>.</p></li>
<li><p><strong>branch_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The minimum length of a branch in the tree (see
above for details on how internal node times are assigned).</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, add details of this operation to the
provenance information of the returned tree sequence. (Default: True).</p></li>
<li><p><strong>**kwargs</strong>  Further arguments used as parameters when constructing the
returned <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>. For example
<code class="docutils literal notranslate"><span class="pre">tskit.Tree.generate_balanced(sample_lists=True)</span></code> will
return a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> created with <code class="docutils literal notranslate"><span class="pre">sample_lists=True</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A balanced tree. Its corresponding <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> is available
via the <a class="reference internal" href="#tskit.Tree.tree_sequence" title="tskit.Tree.tree_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tree_sequence</span></code></a> attribute.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">Tree</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.generate_comb">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">generate_comb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_leaves</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">branch_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.generate_comb"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.generate_comb" title="Link to this definition">#</a></dt>
<dd><p>Generate a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> in which all internal nodes have two children
and the left child is a leaf. This is a comb, ladder or pectinate
phylogeny, and also known as a <a class="reference external" href="https://en.wikipedia.org/wiki/Caterpillar_tree">caterpillar tree</a>.</p>
<p>The leaf nodes are all at time 0, marked as samples,
and labelled 0 to n from left-to-right. Internal node IDs are assigned
sequentially from n as we ascend the tree, and the time of an internal
node is the maximum time of its children plus the specified <code class="docutils literal notranslate"><span class="pre">branch_length</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_leaves</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The number of leaf nodes in the returned tree (must be
2 or greater).</p></li>
<li><p><strong>span</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The span of the tree, and therefore the
<a class="reference internal" href="#tskit.TreeSequence.sequence_length" title="tskit.TreeSequence.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">sequence_length</span></code></a> of the <a class="reference internal" href="#tskit.Tree.tree_sequence" title="tskit.Tree.tree_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tree_sequence</span></code></a>
property of the returned <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>.</p></li>
<li><p><strong>branch_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The branch length between each internal node; the
root node is therefore placed at time <code class="docutils literal notranslate"><span class="pre">branch_length</span> <span class="pre">*</span> <span class="pre">(num_leaves</span> <span class="pre">-</span> <span class="pre">1)</span></code>.</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, add details of this operation to the
provenance information of the returned tree sequence. (Default: True).</p></li>
<li><p><strong>**kwargs</strong>  Further arguments used as parameters when constructing the
returned <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>. For example
<code class="docutils literal notranslate"><span class="pre">tskit.Tree.generate_comb(sample_lists=True)</span></code> will
return a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> created with <code class="docutils literal notranslate"><span class="pre">sample_lists=True</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A comb-shaped bifurcating tree. Its corresponding <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>
is available via the <a class="reference internal" href="#tskit.Tree.tree_sequence" title="tskit.Tree.tree_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tree_sequence</span></code></a> attribute.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">Tree</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Tree.generate_random_binary">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">generate_random_binary</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_leaves</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">branch_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Tree.generate_random_binary"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Tree.generate_random_binary" title="Link to this definition">#</a></dt>
<dd><p>Generate a random binary <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> with <span class="math notranslate nohighlight">\(n\)</span> = <code class="docutils literal notranslate"><span class="pre">num_leaves</span></code>
leaves with an equal probability of returning any topology and
leaf label permutation among the <span class="math notranslate nohighlight">\((2n - 3)! / (2^{n - 2} (n - 2)!)\)</span>
leaf-labelled binary trees.</p>
<p>The leaf nodes are marked as samples, labelled 0 to n, and placed at
time 0. Internal node IDs are assigned sequentially from n as we ascend
the tree, and the time of an internal node is the maximum time of its
children plus the specified <code class="docutils literal notranslate"><span class="pre">branch_length</span></code>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The returned tree has not been created under any explicit model of
evolution. In order to simulate such trees, additional software
such as <cite>msprime &lt;https://github.com/tskit-dev/msprime&gt;`</cite> is required.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_leaves</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The number of leaf nodes in the returned tree (must
be 2 or greater).</p></li>
<li><p><strong>span</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The span of the tree, and therefore the
<a class="reference internal" href="#tskit.TreeSequence.sequence_length" title="tskit.TreeSequence.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">sequence_length</span></code></a> of the <a class="reference internal" href="#tskit.Tree.tree_sequence" title="tskit.Tree.tree_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tree_sequence</span></code></a>
property of the returned <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>.</p></li>
<li><p><strong>branch_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The minimum time between parent and child nodes.</p></li>
<li><p><strong>random_seed</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The random seed. If this is None, a random seed will
be automatically generated. Valid random seeds must be between 1 and
<span class="math notranslate nohighlight">\(2^32  1\)</span>.</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, add details of this operation to the
provenance information of the returned tree sequence. (Default: True).</p></li>
<li><p><strong>**kwargs</strong>  Further arguments used as parameters when constructing the
returned <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>. For example
<code class="docutils literal notranslate"><span class="pre">tskit.Tree.generate_comb(sample_lists=True)</span></code> will
return a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> created with <code class="docutils literal notranslate"><span class="pre">sample_lists=True</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A random binary tree. Its corresponding <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> is
available via the <a class="reference internal" href="#tskit.Tree.tree_sequence" title="tskit.Tree.tree_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tree_sequence</span></code></a> attribute.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">Tree</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="the-treesequence-class">
<h4>The <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> class<a class="headerlink" href="#the-treesequence-class" title="Link to this heading">#</a></h4>
<p>Also see the <a class="reference internal" href="#sec-python-api-tree-sequences"><span class="std std-ref">TreeSequence API</span></a> summary.</p>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.TreeSequence">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">TreeSequence</span></span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence" title="Link to this definition">#</a></dt>
<dd><p>A single tree sequence, as defined by the <a class="reference internal" href="data-model.html#sec-data-model"><span class="std std-ref">data model</span></a>.
A TreeSequence instance can be created from a set of
<a class="reference internal" href="data-model.html#sec-table-definitions"><span class="std std-ref">tables</span></a> using
<a class="reference internal" href="#tskit.TableCollection.tree_sequence" title="tskit.TableCollection.tree_sequence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.tree_sequence()</span></code></a>, or loaded from a set of text files
using <a class="reference internal" href="#tskit.load_text" title="tskit.load_text"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load_text()</span></code></a>, or loaded from a native binary file using
<a class="reference internal" href="#tskit.load" title="tskit.load"><code class="xref py py-func docutils literal notranslate"><span class="pre">tskit.load()</span></code></a>.</p>
<p>TreeSequences are immutable. To change the data held in a particular
tree sequence, first get the table information as a <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>
instance (using <a class="reference internal" href="#tskit.TreeSequence.dump_tables" title="tskit.TreeSequence.dump_tables"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dump_tables()</span></code></a>), edit those tables using the
<a class="reference internal" href="#sec-tables-api"><span class="std std-ref">tables api</span></a>, and create a new tree sequence using
<a class="reference internal" href="#tskit.TableCollection.tree_sequence" title="tskit.TableCollection.tree_sequence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.tree_sequence()</span></code></a>.</p>
<p>The <a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trees()</span></code></a> method iterates over all trees in a tree sequence, and
the <a class="reference internal" href="#tskit.TreeSequence.variants" title="tskit.TreeSequence.variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">variants()</span></code></a> method iterates over all sites and their genotypes.</p>
<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.equals">
<span class="sig-name descname"><span class="pre">equals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_ts_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_timestamps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_tables</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_reference_sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.equals"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.equals" title="Link to this definition">#</a></dt>
<dd><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal. Uses the underlying table
equality, see <a class="reference internal" href="#tskit.TableCollection.equals" title="tskit.TableCollection.equals"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.equals()</span></code></a> for details and options.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.aslist">
<span class="sig-name descname"><span class="pre">aslist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.aslist"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.aslist" title="Link to this definition">#</a></dt>
<dd><p>Returns the trees in this tree sequence as a list. Each tree is
represented by a different instance of <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>. As such, this
method is inefficient and may use a large amount of memory, and should
not be used when performance is a consideration. The <a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trees()</span></code></a>
method is the recommended way to efficiently iterate over the trees
in a tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>**kwargs</strong>  Further arguments used as parameters when constructing the
returned trees. For example <code class="docutils literal notranslate"><span class="pre">ts.aslist(sample_lists=True)</span></code> will result
in a list of <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> instances created with <code class="docutils literal notranslate"><span class="pre">sample_lists=True</span></code>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of the trees in this tree sequence.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)">list</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.dump">
<span class="sig-name descname"><span class="pre">dump</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_or_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zlib_compression</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.dump"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.dump" title="Link to this definition">#</a></dt>
<dd><p>Writes the tree sequence to the specified path or file object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_or_path</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The file object or path to write the TreeSequence to.</p></li>
<li><p><strong>zlib_compression</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  This parameter is deprecated and ignored.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.reference_sequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reference_sequence</span></span><a class="headerlink" href="#tskit.TreeSequence.reference_sequence" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="#tskit.ReferenceSequence" title="tskit.ReferenceSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReferenceSequence</span></code></a> associated with this <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>
if one is defined (see <a class="reference internal" href="#tskit.TreeSequence.has_reference_sequence" title="tskit.TreeSequence.has_reference_sequence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.has_reference_sequence()</span></code></a>),
or None otherwise.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.has_reference_sequence">
<span class="sig-name descname"><span class="pre">has_reference_sequence</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.has_reference_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.has_reference_sequence" title="Link to this definition">#</a></dt>
<dd><p>Returns True if this <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> has an associated
<a class="reference internal" href="data-model.html#sec-data-model-reference-sequence"><span class="std std-ref">reference sequence</span></a>.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.tables_dict">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">tables_dict</span></span><a class="headerlink" href="#tskit.TreeSequence.tables_dict" title="Link to this definition">#</a></dt>
<dd><p>Returns a dictionary mapping names to tables in the
underlying <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>. Equivalent to calling
<code class="docutils literal notranslate"><span class="pre">ts.tables.table_name_map</span></code>.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.tables">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">tables</span></span><a class="headerlink" href="#tskit.TreeSequence.tables" title="Link to this definition">#</a></dt>
<dd><p>Returns an immutable view of the tables underlying this tree sequence.</p>
<p>This view shares the same data as the TreeSequence (zero-copy).
Use <a class="reference internal" href="#tskit.TreeSequence.dump_tables" title="tskit.TreeSequence.dump_tables"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dump_tables()</span></code></a> for a modifiable copy.</p>
<p>Note that if tskit was built with Numpy 1, this method acts as
<a class="reference internal" href="#tskit.TreeSequence.dump_tables" title="tskit.TreeSequence.dump_tables"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dump_tables()</span></code></a> and returns a mutable TableCollection.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An immutable view of the TableCollection underlying this tree sequence.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.nbytes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nbytes</span></span><a class="headerlink" href="#tskit.TreeSequence.nbytes" title="Link to this definition">#</a></dt>
<dd><p>Returns the total number of bytes required to store the data
in this tree sequence. Note that this may not be equal to
the actual memory footprint.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.dump_tables">
<span class="sig-name descname"><span class="pre">dump_tables</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.dump_tables"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.dump_tables" title="Link to this definition">#</a></dt>
<dd><p>Returns a modifiable copy of the <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">tables</span></code></a> defining
this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a> containing all tables underlying
the tree sequence.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection">TableCollection</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.link_ancestors">
<span class="sig-name descname"><span class="pre">link_ancestors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">samples</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ancestors</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.link_ancestors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.link_ancestors" title="Link to this definition">#</a></dt>
<dd><p>Equivalent to <a class="reference internal" href="#tskit.TableCollection.link_ancestors" title="tskit.TableCollection.link_ancestors"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.link_ancestors()</span></code></a>; see that method for full
documentation and parameter semantics.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samples</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>]</em>)  Node IDs to retain as samples.</p></li>
<li><p><strong>ancestors</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>]</em>)  Node IDs to treat as ancestors.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An <a class="reference internal" href="#tskit.EdgeTable" title="tskit.tables.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">tables.EdgeTable</span></code></a> containing the genealogical links between
the supplied <code class="docutils literal notranslate"><span class="pre">samples</span></code> and <code class="docutils literal notranslate"><span class="pre">ancestors</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.EdgeTable" title="tskit.tables.EdgeTable">tables.EdgeTable</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.dump_text">
<span class="sig-name descname"><span class="pre">dump_text</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individuals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">populations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">migrations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">provenances</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">precision</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">6</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">encoding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'utf8'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base64_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.dump_text"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.dump_text" title="Link to this definition">#</a></dt>
<dd><p>Writes a text representation of the tables underlying the tree sequence
to the specified connections.</p>
<p>If Base64 encoding is not used, then metadata will be saved directly, possibly
resulting in errors reading the tables back in if metadata includes whitespace.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nodes</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.14)"><em>io.TextIOBase</em></a>)  The file-like object (having a .write() method) to
write the NodeTable to.</p></li>
<li><p><strong>edges</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.14)"><em>io.TextIOBase</em></a>)  The file-like object to write the EdgeTable to.</p></li>
<li><p><strong>sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.14)"><em>io.TextIOBase</em></a>)  The file-like object to write the SiteTable to.</p></li>
<li><p><strong>mutations</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.14)"><em>io.TextIOBase</em></a>)  The file-like object to write the
MutationTable to.</p></li>
<li><p><strong>individuals</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.14)"><em>io.TextIOBase</em></a>)  The file-like object to write the
IndividualTable to.</p></li>
<li><p><strong>populations</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.14)"><em>io.TextIOBase</em></a>)  The file-like object to write the
PopulationTable to.</p></li>
<li><p><strong>migrations</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.14)"><em>io.TextIOBase</em></a>)  The file-like object to write the
MigrationTable to.</p></li>
<li><p><strong>provenances</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.TextIOBase" title="(in Python v3.14)"><em>io.TextIOBase</em></a>)  The file-like object to write the
ProvenanceTable to.</p></li>
<li><p><strong>precision</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The number of digits of precision.</p></li>
<li><p><strong>encoding</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  Encoding used for text representation.</p></li>
<li><p><strong>base64_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Only used if a schema is not present on each table
being dumped. If True, metadata is encoded using Base64
encoding; otherwise, as plain text.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.__str__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.__str__" title="Link to this definition">#</a></dt>
<dd><p>Return a plain text summary of the contents of a tree sequence</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence._repr_html_">
<span class="sig-name descname"><span class="pre">_repr_html_</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence._repr_html_"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence._repr_html_" title="Link to this definition">#</a></dt>
<dd><p>Return an html summary of a tree sequence. Called by jupyter notebooks
to render a TreeSequence.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.num_samples">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_samples</span></span><a class="headerlink" href="#tskit.TreeSequence.num_samples" title="Link to this definition">#</a></dt>
<dd><p>Returns the number of sample nodes in this tree sequence. This is also the
number of sample nodes in each tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of sample nodes in this tree sequence.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.table_metadata_schemas">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">table_metadata_schemas</span></span><a class="headerlink" href="#tskit.TreeSequence.table_metadata_schemas" title="Link to this definition">#</a></dt>
<dd><p>The set of metadata schemas for the tables in this tree sequence.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.discrete_genome">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">discrete_genome</span></span><a class="headerlink" href="#tskit.TreeSequence.discrete_genome" title="Link to this definition">#</a></dt>
<dd><p>Returns True if all genome coordinates in this TreeSequence are
discrete integer values. This is true iff all the following are true:</p>
<ul class="simple">
<li><p>The sequence length is discrete</p></li>
<li><p>All site positions are discrete</p></li>
<li><p>All left and right edge coordinates are discrete</p></li>
<li><p>All migration left and right coordinates are discrete</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>True if this TreeSequence uses discrete genome coordinates.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.discrete_time">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">discrete_time</span></span><a class="headerlink" href="#tskit.TreeSequence.discrete_time" title="Link to this definition">#</a></dt>
<dd><p>Returns True if all time coordinates in this TreeSequence are
discrete integer values. This is true iff all the following are true:</p>
<ul class="simple">
<li><p>All node times are discrete</p></li>
<li><p>All mutation times are discrete</p></li>
<li><p>All migration times are discrete</p></li>
</ul>
<p>Note that <code class="docutils literal notranslate"><span class="pre">tskit.UNKNOWN_TIME</span></code> counts as discrete.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>True if this TreeSequence uses discrete time coordinates.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.min_time">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">min_time</span></span><a class="headerlink" href="#tskit.TreeSequence.min_time" title="Link to this definition">#</a></dt>
<dd><p>Returns the min time in this tree sequence. This is the minimum
of the node times and mutation times.</p>
<p>Note that mutation times with the value <code class="docutils literal notranslate"><span class="pre">tskit.UNKNOWN_TIME</span></code>
are ignored.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The min time of the nodes and mutations in this tree sequence.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.max_time">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_time</span></span><a class="headerlink" href="#tskit.TreeSequence.max_time" title="Link to this definition">#</a></dt>
<dd><p>Returns the max time in this tree sequence. This is the maximum
of the node times and mutation times.</p>
<p>Note that mutation times with the value <code class="docutils literal notranslate"><span class="pre">tskit.UNKNOWN_TIME</span></code>
are ignored.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The max time of the nodes and mutations in this tree sequence.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.sequence_length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence_length</span></span><a class="headerlink" href="#tskit.TreeSequence.sequence_length" title="Link to this definition">#</a></dt>
<dd><p>Returns the sequence length in this tree sequence. This defines the
genomic scale over which tree coordinates are defined. Given a
tree sequence with a sequence length <span class="math notranslate nohighlight">\(L\)</span>, the constituent
trees will be defined over the half-closed interval
<span class="math notranslate nohighlight">\([0, L)\)</span>. Each tree then covers some subset of this
interval  see <a class="reference internal" href="#tskit.Tree.interval" title="tskit.Tree.interval"><code class="xref py py-attr docutils literal notranslate"><span class="pre">tskit.Tree.interval</span></code></a> for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The length of the sequence in this tree sequence in bases.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.metadata">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">metadata</span></span><a class="headerlink" href="#tskit.TreeSequence.metadata" title="Link to this definition">#</a></dt>
<dd><p>The decoded metadata for this TreeSequence.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.metadata_schema">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">metadata_schema</span></span><a class="headerlink" href="#tskit.TreeSequence.metadata_schema" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this TreeSequence.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.time_units">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">time_units</span></span><a class="headerlink" href="#tskit.TreeSequence.time_units" title="Link to this definition">#</a></dt>
<dd><p>String describing the units of the time dimension for this TreeSequence.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.num_edges">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_edges</span></span><a class="headerlink" href="#tskit.TreeSequence.num_edges" title="Link to this definition">#</a></dt>
<dd><p>Returns the number of <a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">edges</span></a> in this
tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of edges in this tree sequence.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.num_trees">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_trees</span></span><a class="headerlink" href="#tskit.TreeSequence.num_trees" title="Link to this definition">#</a></dt>
<dd><p>Returns the number of distinct trees in this tree sequence. This
is equal to the number of trees returned by the <a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trees()</span></code></a>
method.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of trees in this tree sequence.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.num_sites">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_sites</span></span><a class="headerlink" href="#tskit.TreeSequence.num_sites" title="Link to this definition">#</a></dt>
<dd><p>Returns the number of <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">sites</span></a> in
this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of sites in this tree sequence.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.num_mutations">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_mutations</span></span><a class="headerlink" href="#tskit.TreeSequence.num_mutations" title="Link to this definition">#</a></dt>
<dd><p>Returns the number of <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">mutations</span></a>
in this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of mutations in this tree sequence.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.num_individuals">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_individuals</span></span><a class="headerlink" href="#tskit.TreeSequence.num_individuals" title="Link to this definition">#</a></dt>
<dd><p>Returns the number of <a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">individuals</span></a> in
this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of individuals in this tree sequence.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.num_nodes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_nodes</span></span><a class="headerlink" href="#tskit.TreeSequence.num_nodes" title="Link to this definition">#</a></dt>
<dd><p>Returns the number of <a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">nodes</span></a> in
this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of nodes in this tree sequence.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.num_provenances">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_provenances</span></span><a class="headerlink" href="#tskit.TreeSequence.num_provenances" title="Link to this definition">#</a></dt>
<dd><p>Returns the number of <a class="reference internal" href="data-model.html#sec-provenance-table-definition"><span class="std std-ref">provenances</span></a>
in this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of provenances in this tree sequence.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.num_populations">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_populations</span></span><a class="headerlink" href="#tskit.TreeSequence.num_populations" title="Link to this definition">#</a></dt>
<dd><p>Returns the number of <a class="reference internal" href="data-model.html#sec-population-table-definition"><span class="std std-ref">populations</span></a>
in this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of populations in this tree sequence.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.num_migrations">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_migrations</span></span><a class="headerlink" href="#tskit.TreeSequence.num_migrations" title="Link to this definition">#</a></dt>
<dd><p>Returns the number of <a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">migrations</span></a>
in this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of migrations in this tree sequence.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.max_root_time">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_root_time</span></span><a class="headerlink" href="#tskit.TreeSequence.max_root_time" title="Link to this definition">#</a></dt>
<dd><p>Returns the time of the oldest root in any of the trees in this tree sequence.
This is usually equal to <code class="docutils literal notranslate"><span class="pre">np.max(ts.tables.nodes.time)</span></code> but may not be
since there can be non-sample nodes that are not present in any tree. Note that
isolated samples are also defined as roots (so there can be a max_root_time
even in a tree sequence with no edges).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The maximum time of a root in this tree sequence.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)">float</a></p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.14)"><strong>ValueError</strong></a>  If there are no samples in the tree, and hence no roots (as
roots are defined by the ends of the upward paths from the set of samples).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.migrations">
<span class="sig-name descname"><span class="pre">migrations</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.migrations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.migrations" title="Link to this definition">#</a></dt>
<dd><p>Returns an iterable sequence of all the
<a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">migrations</span></a> in this tree sequence.</p>
<p>Migrations are returned in nondecreasing order of the <code class="docutils literal notranslate"><span class="pre">time</span></code> value.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An iterable sequence of all migrations.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>Sequence(<a class="reference internal" href="#tskit.Migration" title="tskit.Migration"><code class="xref py py-class docutils literal notranslate"><span class="pre">Migration</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.individuals">
<span class="sig-name descname"><span class="pre">individuals</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.individuals"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.individuals" title="Link to this definition">#</a></dt>
<dd><p>Returns an iterable sequence of all the
<a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">individuals</span></a> in this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An iterable sequence of all individuals.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>Sequence(<a class="reference internal" href="#tskit.Individual" title="tskit.Individual"><code class="xref py py-class docutils literal notranslate"><span class="pre">Individual</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.nodes">
<span class="sig-name descname"><span class="pre">nodes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.nodes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.nodes" title="Link to this definition">#</a></dt>
<dd><p>Returns an iterable sequence of all the <a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">nodes</span></a>
in this tree sequence.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Although node ids are commonly ordered by node time, this is not a
formal tree sequence requirement. If you wish to iterate over nodes in
time order, you should therefore use <code class="docutils literal notranslate"><span class="pre">order=&quot;timeasc&quot;</span></code> (and wrap the
resulting sequence in the standard Python <a class="reference external" href="https://docs.python.org/3/library/functions.html#reversed" title="(in Python v3.14)"><code class="docutils literal notranslate"><span class="pre">reversed()</span></code></a> function
if you wish to iterate over older nodes before younger ones)</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>order</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The order in which the nodes should be returned: must be
one of id (default) or timeasc (ascending order of time, then by
ascending node id, matching the first two ordering requirements of
parent nodes in a <a class="reference internal" href="#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">sorted</span></code></a> edge table).</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An iterable sequence of all nodes.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Sequence(<a class="reference internal" href="#tskit.Node" title="tskit.Node"><code class="xref py py-class docutils literal notranslate"><span class="pre">Node</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.edges">
<span class="sig-name descname"><span class="pre">edges</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.edges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.edges" title="Link to this definition">#</a></dt>
<dd><p>Returns an iterable sequence of all the <a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">edges</span></a>
in this tree sequence. Edges are returned in the order required
for a <a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">valid tree sequence</span></a>. So,
edges are guaranteed to be ordered such that (a) all parents with a
given ID are contiguous; (b) edges are returned in non-decreasing
order of parent time ago; (c) within the edges for a given parent, edges
are sorted first by child ID and then by left coordinate.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An iterable sequence of all edges.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>Sequence(<a class="reference internal" href="#tskit.Edge" title="tskit.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.edge_diffs">
<span class="sig-name descname"><span class="pre">edge_diffs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">include_terminal</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">direction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.edge_diffs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.edge_diffs" title="Link to this definition">#</a></dt>
<dd><p>Returns an iterator over all the <a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">edges</span></a> that
are inserted and removed to build the trees as we move from left-to-right along
the tree sequence. Each iteration yields a named tuple consisting of 3 values,
<code class="docutils literal notranslate"><span class="pre">(interval,</span> <span class="pre">edges_out,</span> <span class="pre">edges_in)</span></code>. The first value, <code class="docutils literal notranslate"><span class="pre">interval</span></code>, is the
genomic interval <code class="docutils literal notranslate"><span class="pre">(left,</span> <span class="pre">right)</span></code> covered by the incoming tree
(see <a class="reference internal" href="#tskit.Tree.interval" title="tskit.Tree.interval"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Tree.interval</span></code></a>). The second, <code class="docutils literal notranslate"><span class="pre">edges_out</span></code> is a list of the edges
that were just-removed to create the tree covering the interval
(hence <code class="docutils literal notranslate"><span class="pre">edges_out</span></code> will always be empty for the first tree). The last value,
<code class="docutils literal notranslate"><span class="pre">edges_in</span></code>, is a list of edges that were just
inserted to construct the tree covering the current interval.</p>
<p>The edges returned within each <code class="docutils literal notranslate"><span class="pre">edges_in</span></code> list are ordered by ascending
time of the parent node, then ascending parent id, then ascending child id.
The edges within each <code class="docutils literal notranslate"><span class="pre">edges_out</span></code> list are the reverse order (e.g.
descending parent time, parent id, then child_id). This means that within
each list, edges with the same parent appear consecutively.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">direction</span></code> argument can be used to control whether diffs are produced
in the forward (left-to-right, increasing genome coordinate value)
or reverse (right-to-left, decreasing genome coordinate value) direction.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>include_terminal</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If False (default), the iterator terminates
after the final interval in the tree sequence (i.e., it does not
report a final removal of all remaining edges), and the number
of iterations will be equal to the number of trees in the tree
sequence. If True, an additional iteration takes place, with the last
<code class="docutils literal notranslate"><span class="pre">edges_out</span></code> value reporting all the edges contained in the final
tree (with both <code class="docutils literal notranslate"><span class="pre">left</span></code> and <code class="docutils literal notranslate"><span class="pre">right</span></code> equal to the sequence length).</p></li>
<li><p><strong>direction</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The direction of travel along the sequence for
diffs. Must be one of <a class="reference internal" href="#tskit.FORWARD" title="tskit.FORWARD"><code class="xref py py-data docutils literal notranslate"><span class="pre">FORWARD</span></code></a> or <a class="reference internal" href="#tskit.REVERSE" title="tskit.REVERSE"><code class="xref py py-data docutils literal notranslate"><span class="pre">REVERSE</span></code></a>.
(Default: <a class="reference internal" href="#tskit.FORWARD" title="tskit.FORWARD"><code class="xref py py-data docutils literal notranslate"><span class="pre">FORWARD</span></code></a>).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An iterator over the (interval, edges_out, edges_in) tuples. This
is a named tuple, so the 3 values can be accessed by position
(e.g. <code class="docutils literal notranslate"><span class="pre">returned_tuple[0]</span></code>) or name (e.g. <code class="docutils literal notranslate"><span class="pre">returned_tuple.interval</span></code>).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">collections.abc.Iterable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.sites">
<span class="sig-name descname"><span class="pre">sites</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.sites"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.sites" title="Link to this definition">#</a></dt>
<dd><p>Returns an iterable sequence of all the <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">sites</span></a>
in this tree sequence. Sites are returned in order of increasing ID
(and also position). See the <a class="reference internal" href="#tskit.Site" title="tskit.Site"><code class="xref py py-class docutils literal notranslate"><span class="pre">Site</span></code></a> class for details on
the available fields for each site.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An iterable sequence of all sites.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>Sequence(<a class="reference internal" href="#tskit.Site" title="tskit.Site"><code class="xref py py-class docutils literal notranslate"><span class="pre">Site</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.mutations">
<span class="sig-name descname"><span class="pre">mutations</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.mutations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.mutations" title="Link to this definition">#</a></dt>
<dd><p>Returns an iterator over all the
<a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">mutations</span></a> in this tree sequence.
Mutations are returned in order of nondecreasing site ID.
See the <a class="reference internal" href="#tskit.Mutation" title="tskit.Mutation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mutation</span></code></a> class for details on the available fields for
each mutation.</p>
<p>The returned iterator is equivalent to iterating over all sites
and all mutations in each site, i.e.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">tree_sequence</span><span class="o">.</span><span class="n">sites</span><span class="p">():</span>
    <span class="k">for</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">site</span><span class="o">.</span><span class="n">mutations</span><span class="p">:</span>
        <span class="k">yield</span> <span class="n">mutation</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An iterator over all mutations in this tree sequence.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>iter(<a class="reference internal" href="#tskit.Mutation" title="tskit.Mutation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mutation</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.populations">
<span class="sig-name descname"><span class="pre">populations</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.populations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.populations" title="Link to this definition">#</a></dt>
<dd><p>Returns an iterable sequence of all the
<a class="reference internal" href="data-model.html#sec-population-table-definition"><span class="std std-ref">populations</span></a> in this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An iterable sequence of all populations.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>Sequence(<a class="reference internal" href="#tskit.Population" title="tskit.Population"><code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.provenances">
<span class="sig-name descname"><span class="pre">provenances</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.provenances"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.provenances" title="Link to this definition">#</a></dt>
<dd><p>Returns an iterable sequence of all the
<a class="reference internal" href="data-model.html#sec-provenance-table-definition"><span class="std std-ref">provenances</span></a> in this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An iterable sequence of all provenances.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>Sequence(<a class="reference internal" href="#tskit.Provenance" title="tskit.Provenance"><code class="xref py py-class docutils literal notranslate"><span class="pre">Provenance</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.breakpoints">
<span class="sig-name descname"><span class="pre">breakpoints</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">as_array</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.breakpoints"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.breakpoints" title="Link to this definition">#</a></dt>
<dd><p>Returns the breakpoints that separate trees along the chromosome, including the
two extreme points 0 and L. This is equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">iter</span><span class="p">([</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">t</span><span class="o">.</span><span class="n">interval</span><span class="o">.</span><span class="n">right</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">trees</span><span class="p">()])</span>
</pre></div>
</div>
<p>By default we return an iterator over the breakpoints as Python float objects;
if <code class="docutils literal notranslate"><span class="pre">as_array</span></code> is True we return them as a numpy array.</p>
<p>Note that the <code class="docutils literal notranslate"><span class="pre">as_array</span></code> form will be more efficient and convenient in most
cases; the default iterator behaviour is mainly kept to ensure compatibility
with existing code.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>as_array</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, return the breakpoints as a numpy array.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The breakpoints defined by the tree intervals along the sequence.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable" title="(in Python v3.14)">collections.abc.Iterable</a> or <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.at">
<span class="sig-name descname"><span class="pre">at</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">position</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.at"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.at" title="Link to this definition">#</a></dt>
<dd><p>Returns the tree covering the specified genomic location. The returned tree
will have <code class="docutils literal notranslate"><span class="pre">tree.interval.left</span></code> &lt;= <code class="docutils literal notranslate"><span class="pre">position</span></code> &lt; <code class="docutils literal notranslate"><span class="pre">tree.interval.right</span></code>.
See also <a class="reference internal" href="#tskit.Tree.seek" title="tskit.Tree.seek"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.seek()</span></code></a>.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The current implementation of this operation is linear in the number of
trees, so may be inefficient for large tree sequences. See
<a class="reference external" href="https://github.com/tskit-dev/tskit/issues/684">this issue</a> for more
information.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>position</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  A genomic location.</p></li>
<li><p><strong>**kwargs</strong>  Further arguments used as parameters when constructing the
returned <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>. For example <code class="docutils literal notranslate"><span class="pre">ts.at(2.5,</span> <span class="pre">sample_lists=True)</span></code> will
result in a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> created with <code class="docutils literal notranslate"><span class="pre">sample_lists=True</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A new instance of <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> positioned to cover the specified
genomic location.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">Tree</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.at_index">
<span class="sig-name descname"><span class="pre">at_index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.at_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.at_index" title="Link to this definition">#</a></dt>
<dd><p>Returns the tree at the specified index. See also <a class="reference internal" href="#tskit.Tree.seek_index" title="tskit.Tree.seek_index"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.seek_index()</span></code></a>.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The current implementation of this operation is linear in the number of
trees, so may be inefficient for large tree sequences. See
<a class="reference external" href="https://github.com/tskit-dev/tskit/issues/684">this issue</a> for more
information.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The index of the required tree.</p></li>
<li><p><strong>**kwargs</strong>  Further arguments used as parameters when constructing the
returned <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>. For example <code class="docutils literal notranslate"><span class="pre">ts.at_index(4,</span> <span class="pre">sample_lists=True)</span></code>
will result in a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> created with <code class="docutils literal notranslate"><span class="pre">sample_lists=True</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A new instance of <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> positioned at the specified index.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree">Tree</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.first">
<span class="sig-name descname"><span class="pre">first</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.first"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.first" title="Link to this definition">#</a></dt>
<dd><p>Returns the first tree in this <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>. To iterate over all
trees in the sequence, use the <a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trees()</span></code></a> method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>**kwargs</strong>  Further arguments used as parameters when constructing the
returned <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>. For example <code class="docutils literal notranslate"><span class="pre">ts.first(sample_lists=True)</span></code> will
result in a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> created with <code class="docutils literal notranslate"><span class="pre">sample_lists=True</span></code>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The first tree in this tree sequence.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.last">
<span class="sig-name descname"><span class="pre">last</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.last"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.last" title="Link to this definition">#</a></dt>
<dd><p>Returns the last tree in this <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>. To iterate over all
trees in the sequence, use the <a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trees()</span></code></a> method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>**kwargs</strong>  Further arguments used as parameters when constructing the
returned <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>. For example <code class="docutils literal notranslate"><span class="pre">ts.first(sample_lists=True)</span></code> will
result in a <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> created with <code class="docutils literal notranslate"><span class="pre">sample_lists=True</span></code>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The last tree in this tree sequence.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.trees">
<span class="sig-name descname"><span class="pre">trees</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tracked_samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_lists</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">root_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_counts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tracked_leaves</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">leaf_counts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">leaf_lists</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.trees"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.trees" title="Link to this definition">#</a></dt>
<dd><p>Returns an iterator over the trees in this tree sequence. Each value
returned in this iterator is an instance of <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>. Upon
successful termination of the iterator, the tree will be in the
cleared null state.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">sample_lists</span></code> and <code class="docutils literal notranslate"><span class="pre">tracked_samples</span></code> parameters are passed
to the <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a> constructor, and control
the options that are set in the returned tree instance.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Do not store the results of this iterator in a list!
For performance reasons, the same underlying object is used
for every tree returned which will most likely lead to unexpected
behaviour. If you wish to obtain a list of trees in a tree sequence
please use <code class="docutils literal notranslate"><span class="pre">ts.aslist()</span></code> instead.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tracked_samples</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  The list of samples to be tracked and
counted using the <a class="reference internal" href="#tskit.Tree.num_tracked_samples" title="tskit.Tree.num_tracked_samples"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.num_tracked_samples()</span></code></a> method.</p></li>
<li><p><strong>sample_lists</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, provide more efficient access
to the samples beneath a given node using the
<a class="reference internal" href="#tskit.Tree.samples" title="tskit.Tree.samples"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.samples()</span></code></a> method.</p></li>
<li><p><strong>root_threshold</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The minimum number of samples that a node
must be ancestral to for it to be in the list of roots. By default
this is 1, so that isolated samples (representing missing data)
are roots. To efficiently restrict the roots of the tree to
those subtending meaningful topology, set this to 2. This value
is only relevant when trees have multiple roots.</p></li>
<li><p><strong>sample_counts</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Deprecated since 0.2.4.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An iterator over the Trees in this tree sequence.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>collections.abc.Iterable, <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.coiterate">
<span class="sig-name descname"><span class="pre">coiterate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.coiterate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.coiterate" title="Link to this definition">#</a></dt>
<dd><p>Returns an iterator over the pairs of trees for each distinct
interval in the specified pair of tree sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><em>TreeSequence</em></a>)  The other tree sequence from which to take trees. The
sequence length must be the same as the current tree sequence.</p></li>
<li><p><strong>**kwargs</strong>  Further named arguments that will be passed to the
<a class="reference internal" href="#tskit.TreeSequence.trees" title="tskit.TreeSequence.trees"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trees()</span></code></a> method when constructing the returned trees.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An iterator returning successive tuples of the form
<code class="docutils literal notranslate"><span class="pre">(interval,</span> <span class="pre">tree_self,</span> <span class="pre">tree_other)</span></code>. For example, the first item returned
will consist of an tuple of the initial interval, the first tree of the
current tree sequence, and the first tree of the <code class="docutils literal notranslate"><span class="pre">other</span></code> tree sequence;
the <code class="docutils literal notranslate"><span class="pre">.left</span></code> attribute of the initial interval will be 0 and the <code class="docutils literal notranslate"><span class="pre">.right</span></code>
attribute will be the smallest non-zero breakpoint of the 2 tree sequences.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>iter(<a class="reference internal" href="#tskit.Interval" title="tskit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a>, <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>, <a class="reference internal" href="#tskit.Tree" title="tskit.Tree"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.haplotypes">
<span class="sig-name descname"><span class="pre">haplotypes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isolated_as_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">missing_data_character</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">impute_missing_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.haplotypes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.haplotypes" title="Link to this definition">#</a></dt>
<dd><p>Returns an iterator over the strings of haplotypes that result from
the trees and mutations in this tree sequence. Each haplotype string
is guaranteed to be of the same length. A tree sequence with
<span class="math notranslate nohighlight">\(n\)</span> requested nodes (default: the number of sample nodes) and with
<span class="math notranslate nohighlight">\(s\)</span> sites lying between <code class="docutils literal notranslate"><span class="pre">left</span></code> and <code class="docutils literal notranslate"><span class="pre">right</span></code> will return a total
of <span class="math notranslate nohighlight">\(n\)</span> strings of <span class="math notranslate nohighlight">\(s\)</span> alleles concatenated together, where an allele
consists of a single ascii character (tree sequences that include alleles
which are not a single character in length, or where the character is
non-ascii, will raise an error). The first string returned is the
haplotype for the first requested node, and so on.</p>
<p>The alleles at each site must be represented by single byte characters,
(i.e., variants must be single nucleotide polymorphisms, or SNPs), hence
the strings returned will all be of length <span class="math notranslate nohighlight">\(s\)</span>. If the <code class="docutils literal notranslate"><span class="pre">left</span></code>
position is less than or equal to the position of the first site, for a
haplotype <code class="docutils literal notranslate"><span class="pre">h</span></code>, the value of <code class="docutils literal notranslate"><span class="pre">h[j]</span></code> will therefore be the observed
allelic state at site <code class="docutils literal notranslate"><span class="pre">j</span></code>.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">isolated_as_missing</span></code> is True (the default), isolated nodes without
mutations directly above them (whether samples or non-samples) will be treated as
<a class="reference internal" href="data-model.html#sec-data-model-missing-data"><span class="std std-ref">missing data</span></a> and will be
represented in the string by the <code class="docutils literal notranslate"><span class="pre">missing_data_character</span></code>. If
instead it is set to False, missing data will be assigned the ancestral state
(unless they have mutations directly above them, in which case they will take
the most recent derived mutational state for that node). This was the default
behaviour in versions prior to 0.2.0. Prior to 0.3.0 the <cite>impute_missing_data</cite>
argument controlled this behaviour.</p>
<p>It is also possible to provide <strong>non-sample</strong> nodes via the <code class="docutils literal notranslate"><span class="pre">samples</span></code>
argument if you wish to output haplotypes for (e.g.) internal nodes.
See also the <a class="reference internal" href="#tskit.TreeSequence.variants" title="tskit.TreeSequence.variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">variants()</span></code></a> iterator for site-centric access
to genotypes for the requested nodes.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>For large datasets, this method can consume a <strong>very large</strong> amount of
memory! To output all the sample data, it is more efficient to iterate
over sites rather than over samples.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An iterator over the haplotype strings for the samples in
this tree sequence.</p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>isolated_as_missing</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, the allele assigned to
missing samples (i.e., isolated samples without mutations) is
the <code class="docutils literal notranslate"><span class="pre">missing_data_character</span></code>. If False,
missing samples will be assigned the ancestral state.
Default: True.</p></li>
<li><p><strong>missing_data_character</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A single ascii character that will
be used to represent missing data.
If any normal allele contains this character, an error is raised.
Default: N.</p></li>
<li><p><strong>samples</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>]</em>)  The node IDs for which to output haplotypes. If
<code class="docutils literal notranslate"><span class="pre">None</span></code> (default), return haplotypes for all the sample nodes in the tree
sequence, in the order given by the <a class="reference internal" href="#tskit.TreeSequence.samples" title="tskit.TreeSequence.samples"><code class="xref py py-meth docutils literal notranslate"><span class="pre">samples()</span></code></a> method. Non-sample
nodes may also be provided.</p></li>
<li><p><strong>left</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  Haplotype strings will start with the first site at or after
this genomic position. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) start at the first site.</p></li>
<li><p><strong>right</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  Haplotype strings will end with the last site before this
position. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) assume <code class="docutils literal notranslate"><span class="pre">right</span></code> is the sequence length
(i.e. the last character in the string will be the last site in the tree
sequence).</p></li>
<li><p><strong>impute_missing_data</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  <em>Deprecated in 0.3.0. Use ``isolated_as_missing``, but inverting value.
Will be removed in a future version</em></p></li>
</ul>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable" title="(in Python v3.14)">collections.abc.Iterable</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#TypeError" title="(in Python v3.14)"><strong>TypeError</strong></a>  if the <code class="docutils literal notranslate"><span class="pre">missing_data_character</span></code> or any of the alleles
at a site are not a single ascii character.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.14)"><strong>ValueError</strong></a>  if the <code class="docutils literal notranslate"><span class="pre">missing_data_character</span></code> exists in one of the
alleles</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.variants">
<span class="sig-name descname"><span class="pre">variants</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isolated_as_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alleles</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">impute_missing_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.variants"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.variants" title="Link to this definition">#</a></dt>
<dd><p>Returns an iterator over the variants between the <code class="docutils literal notranslate"><span class="pre">left</span></code> (inclusive)
and <code class="docutils literal notranslate"><span class="pre">right</span></code> (exclusive) genomic positions in this tree sequence. Each
returned <a class="reference internal" href="#tskit.Variant" title="tskit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> object has a site, a list of possible allelic
states and an array of genotypes for the specified <code class="docutils literal notranslate"><span class="pre">samples</span></code>. The
<code class="docutils literal notranslate"><span class="pre">genotypes</span></code> value is a numpy array containing indexes into the
<code class="docutils literal notranslate"><span class="pre">alleles</span></code> list. By default, this list is generated automatically for
each site such that the first entry, <code class="docutils literal notranslate"><span class="pre">alleles[0]</span></code>, is the ancestral
state and subsequent alleles are listed in no
particular order. This means that the encoding of alleles in
terms of genotype values can vary from site-to-site, which is
sometimes inconvenient. It is possible to specify a fixed mapping
from allele strings to genotype values using the <code class="docutils literal notranslate"><span class="pre">alleles</span></code>
parameter. For example, if we set <code class="docutils literal notranslate"><span class="pre">alleles=(&quot;A&quot;,</span> <span class="pre">&quot;C&quot;,</span> <span class="pre">&quot;G&quot;,</span> <span class="pre">&quot;T&quot;)</span></code>,
this will map allele A to 0, C to 1 and so on (the
<a class="reference internal" href="#tskit.ALLELES_ACGT" title="tskit.ALLELES_ACGT"><code class="xref py py-data docutils literal notranslate"><span class="pre">ALLELES_ACGT</span></code></a> constant provides a shortcut for this
common mapping).</p>
<p>By default, genotypes are generated for all samples. The <code class="docutils literal notranslate"><span class="pre">samples</span></code>
parameter allows us to specify the nodes for which genotypes are
generated; output order of genotypes in the returned variants
corresponds to the order of the samples in this list. It is also
possible to provide <strong>non-sample</strong> nodes as an argument here, if you
wish to generate genotypes for (e.g.) internal nodes. Missingness is
detected for any requested node (sample or non-sample) when
<code class="docutils literal notranslate"><span class="pre">isolated_as_missing</span></code> is True: if a node is isolated at a site (i.e.,
has no parent and no children in the marginal tree) and has no mutation
above it at that site, its genotype will be reported as
<a class="reference internal" href="#tskit.MISSING_DATA" title="tskit.MISSING_DATA"><code class="xref py py-data docutils literal notranslate"><span class="pre">MISSING_DATA</span></code></a> (-1). If <code class="docutils literal notranslate"><span class="pre">isolated_as_missing</span></code> is False, such
nodes are assigned the sites ancestral allele index.</p>
<p>If isolated samples are present at a given site without mutations above them,
they are interpreted by default as
<a class="reference internal" href="data-model.html#sec-data-model-missing-data"><span class="std std-ref">missing data</span></a>, and the genotypes array
will contain a special value <a class="reference internal" href="#tskit.MISSING_DATA" title="tskit.MISSING_DATA"><code class="xref py py-data docutils literal notranslate"><span class="pre">MISSING_DATA</span></code></a> (-1) to identify them
while the <code class="docutils literal notranslate"><span class="pre">alleles</span></code> tuple will end with the value <code class="docutils literal notranslate"><span class="pre">None</span></code> (note that this
will be the case whether or not we specify a fixed mapping using the
<code class="docutils literal notranslate"><span class="pre">alleles</span></code> parameter; see the <a class="reference internal" href="#tskit.Variant" title="tskit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> class for more details).
Alternatively, if <code class="docutils literal notranslate"><span class="pre">isolated_as_missing</span></code> is set to to False, such isolated
samples will not be treated as missing, and instead assigned the ancestral
state (this was the default behaviour in versions prior to 0.2.0). Prior to
0.3.0 the <cite>impute_missing_data</cite> argument controlled this behaviour.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samples</strong> (<em>array_like</em>)  An array of node IDs for which to generate
genotypes, or None for all sample nodes. Default: None.</p></li>
<li><p><strong>isolated_as_missing</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, the genotype value assigned to
missing samples (i.e., isolated samples without mutations) is
<a class="reference internal" href="#tskit.MISSING_DATA" title="tskit.MISSING_DATA"><code class="xref py py-data docutils literal notranslate"><span class="pre">MISSING_DATA</span></code></a> (-1). If False, missing samples will be
assigned the allele index for the ancestral state.
Default: True.</p></li>
<li><p><strong>alleles</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.14)"><em>tuple</em></a>)  A tuple of strings defining the encoding of
alleles as integer genotype values. At least one allele must be provided.
If duplicate alleles are provided, output genotypes will always be
encoded as the first occurrence of the allele. If None (the default),
the alleles are encoded as they are encountered during genotype
generation.</p></li>
<li><p><strong>impute_missing_data</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  <em>Deprecated in 0.3.0. Use ``isolated_as_missing``, but inverting value.
Will be removed in a future version</em></p></li>
<li><p><strong>copy</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If False re-use the same Variant object for each site such that any
references held to it are overwritten when the next site is visited.
If True return a fresh <a class="reference internal" href="#tskit.Variant" title="tskit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> for each site. Default: True.</p></li>
<li><p><strong>left</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  Start with the first site at or after
this genomic position. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) start at the first site.</p></li>
<li><p><strong>right</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  End with the last site before this position. If <code class="docutils literal notranslate"><span class="pre">None</span></code>
(default) assume <code class="docutils literal notranslate"><span class="pre">right</span></code> is the sequence length, so that the last
variant corresponds to the last site in the tree sequence.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An iterator over all variants in this tree sequence.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>iter(<a class="reference internal" href="#tskit.Variant" title="tskit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.genotype_matrix">
<span class="sig-name descname"><span class="pre">genotype_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isolated_as_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alleles</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">impute_missing_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.genotype_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.genotype_matrix" title="Link to this definition">#</a></dt>
<dd><p>Returns an <span class="math notranslate nohighlight">\(m \times n\)</span> numpy array of the genotypes in this
tree sequence, where <span class="math notranslate nohighlight">\(m\)</span> is the number of sites and <span class="math notranslate nohighlight">\(n\)</span>
is the number of requested nodes (default: the number of sample nodes).
The genotypes are the indexes into the array of <code class="docutils literal notranslate"><span class="pre">alleles</span></code>, as
described for the <a class="reference internal" href="#tskit.Variant" title="tskit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> class.</p>
<p>It is possible to provide <strong>non-sample</strong> nodes via the <code class="docutils literal notranslate"><span class="pre">samples</span></code>
argument if you wish to generate genotypes for (e.g.) internal nodes.
Missingness is detected for any requested node (sample or non-sample)
when <code class="docutils literal notranslate"><span class="pre">isolated_as_missing</span></code> is True: if a node is isolated at a site
(i.e., has no parent and no children in the marginal tree) and has no
mutation above it at that site, its genotype will be reported as
<a class="reference internal" href="#tskit.MISSING_DATA" title="tskit.MISSING_DATA"><code class="xref py py-data docutils literal notranslate"><span class="pre">MISSING_DATA</span></code></a> (-1).</p>
<p>Such nodes are treated as missing data by default. If
<code class="docutils literal notranslate"><span class="pre">isolated_as_missing</span></code> is set to False, they will not be treated as
missing, and will instead be assigned the ancestral state. This was the
default behaviour in versions prior to 0.2.0. Prior to 0.3.0 the
<code class="docutils literal notranslate"><span class="pre">impute_missing_data</span></code> argument controlled this behaviour.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method can consume a <strong>very large</strong> amount of memory! If
all genotypes are not needed at once, it is usually better to
access them sequentially using the <a class="reference internal" href="#tskit.TreeSequence.variants" title="tskit.TreeSequence.variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">variants()</span></code></a> iterator.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samples</strong> (<em>array_like</em>)  An array of node IDs for which to generate
genotypes. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), generate genotypes for all sample
nodes. Non-sample nodes may also be provided, in which case genotypes
will be generated for those nodes too.</p></li>
<li><p><strong>isolated_as_missing</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, the genotype value assigned to
isolated nodes without mutations (samples or non-samples) is
<a class="reference internal" href="#tskit.MISSING_DATA" title="tskit.MISSING_DATA"><code class="xref py py-data docutils literal notranslate"><span class="pre">MISSING_DATA</span></code></a> (-1). If False, such nodes will be
assigned the allele index for the ancestral state.
Default: True.</p></li>
<li><p><strong>alleles</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.14)"><em>tuple</em></a>)  A tuple of strings describing the encoding of
alleles to genotype values. At least one allele must be provided.
If duplicate alleles are provided, output genotypes will always be
encoded as the first occurrence of the allele. If None (the default),
the alleles are encoded as they are encountered during genotype
generation.</p></li>
<li><p><strong>impute_missing_data</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  <em>Deprecated in 0.3.0. Use ``isolated_as_missing``, but inverting value.
Will be removed in a future version</em></p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The full matrix of genotypes.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a> (dtype=np.int32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.alignments">
<span class="sig-name descname"><span class="pre">alignments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">missing_data_character</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isolated_as_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.alignments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.alignments" title="Link to this definition">#</a></dt>
<dd><p>Returns an iterator over the full sequence alignments for the defined samples
in this tree sequence. Each yielded alignment <code class="docutils literal notranslate"><span class="pre">a</span></code> is a string of length
<code class="docutils literal notranslate"><span class="pre">L</span></code> where the first character is the genomic sequence at the <code class="docutils literal notranslate"><span class="pre">start</span></code>
position in the genome (defaulting to 0) and the last character is the
genomic sequence one position before the <code class="docutils literal notranslate"><span class="pre">stop</span></code> value (defaulting to the
<a class="reference internal" href="#tskit.TreeSequence.sequence_length" title="tskit.TreeSequence.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">sequence_length</span></code></a> of this tree sequence, which must have
<a class="reference internal" href="#tskit.TreeSequence.discrete_genome" title="tskit.TreeSequence.discrete_genome"><code class="xref py py-attr docutils literal notranslate"><span class="pre">discrete_genome</span></code></a> equal to True). By default <code class="docutils literal notranslate"><span class="pre">L</span></code> is therefore equal
to the <a class="reference internal" href="#tskit.TreeSequence.sequence_length" title="tskit.TreeSequence.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">sequence_length</span></code></a>, and <code class="docutils literal notranslate"><span class="pre">a[j]</span></code> is the nucleotide value at
genomic position <code class="docutils literal notranslate"><span class="pre">j</span></code>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This is inherently a <strong>zero-based</strong> representation of the sequence
coordinate space. Care will be needed when interacting with other
libraries and upstream coordinate spaces.</p>
</div>
<p>The <a class="reference internal" href="glossary.html#sec-data-model-definitions-site"><span class="std std-ref">sites</span></a> in a tree sequence will
usually only define the variation for a subset of the <code class="docutils literal notranslate"><span class="pre">L</span></code> nucleotide
positions along the genome, and the remaining positions are filled using
a <a class="reference internal" href="data-model.html#sec-data-model-reference-sequence"><span class="std std-ref">reference sequence</span></a>.
The reference sequence data is defined either via the
<code class="docutils literal notranslate"><span class="pre">reference_sequence</span></code> parameter to this method, or embedded within
with the tree sequence itself via the <a class="reference internal" href="#tskit.TreeSequence.reference_sequence" title="tskit.TreeSequence.reference_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.reference_sequence</span></code></a>.</p>
<p>Site information from the tree sequence takes precedence over the reference
sequence so that, for example, at a site with no mutations all samples
will have the sites ancestral state.</p>
<p>The reference sequence bases are determined in the following way:</p>
<ul class="simple">
<li><p>If the <code class="docutils literal notranslate"><span class="pre">reference_sequence</span></code> parameter is supplied this will be
used, regardless of whether the tree sequence has an embedded
reference sequence.</p></li>
<li><p>Otherwise, if the tree sequence has an embedded reference sequence,
this will be used.</p></li>
<li><p>If the <code class="docutils literal notranslate"><span class="pre">reference_sequence</span></code> parameter is not specified and
there is no embedded reference sequence, <code class="docutils literal notranslate"><span class="pre">L</span></code> copies of the
<code class="docutils literal notranslate"><span class="pre">missing_data_character</span></code> (which defaults to N) are used
instead.</p></li>
</ul>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The <a class="reference internal" href="#tskit.ReferenceSequence" title="tskit.ReferenceSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReferenceSequence</span></code></a> API is preliminary and
some behaviours may change in the future. In particular, a
tree sequence is currently regarded as having an embedded reference
sequence even if it only has some metadata defined. In this case
the <code class="docutils literal notranslate"><span class="pre">reference_sequence</span></code> parameter will need to be explicitly set.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Two common options for setting a reference sequence are:</p>
<ul class="simple">
<li><p>Mark them as missing data, by setting
<code class="docutils literal notranslate"><span class="pre">reference_sequence=&quot;N&quot;</span> <span class="pre">*</span> <span class="pre">int(ts.sequence_length)</span></code></p></li>
<li><p>Fill the gaps with random nucleotides, by setting
<code class="docutils literal notranslate"><span class="pre">reference_sequence=tskit.random_nucleotides(ts.sequence_length)</span></code>.
See the <a class="reference internal" href="#tskit.random_nucleotides" title="tskit.random_nucleotides"><code class="xref py py-func docutils literal notranslate"><span class="pre">random_nucleotides()</span></code></a> function for more information.</p></li>
</ul>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Insertions and deletions are not currently supported and
the alleles at each site must be represented by
single byte characters, (i.e., variants must be single nucleotide
polymorphisms, or SNPs).</p>
</div>
<p>Missing data handling</p>
<ul class="simple">
<li><p>If <code class="docutils literal notranslate"><span class="pre">isolated_as_missing=True</span></code> (default), nodes that are isolated
(no parent and no children) are rendered as the missing character across
each tree interval. At site positions, the per-site allele overrides the
missing character; if a genotype is missing (<code class="docutils literal notranslate"><span class="pre">-1</span></code>), the missing
character is retained.</p></li>
<li><p>If <code class="docutils literal notranslate"><span class="pre">isolated_as_missing=False</span></code>, no missing overlay is applied. At sites,
genotypes are decoded as usual; at non-sites, bases come from the
reference sequence.</p></li>
</ul>
<p>See also the <a class="reference internal" href="#tskit.TreeSequence.variants" title="tskit.TreeSequence.variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">variants()</span></code></a> iterator for site-centric access
to sample genotypes and <a class="reference internal" href="#tskit.TreeSequence.haplotypes" title="tskit.TreeSequence.haplotypes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">haplotypes()</span></code></a> for access to sample sequences
at just the sites in the tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>reference_sequence</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The reference sequence to fill in
gaps between sites in the alignments. If provided, it must be a
string of length equal to <a class="reference internal" href="#tskit.TreeSequence.sequence_length" title="tskit.TreeSequence.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">sequence_length</span></code></a>; the sequence is
sliced internally to the requested <code class="docutils literal notranslate"><span class="pre">[left,</span> <span class="pre">right)</span></code> interval.</p></li>
<li><p><strong>missing_data_character</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A single ascii character that will
be used to represent missing data.
If any normal allele contains this character, an error is raised.
Default: N.</p></li>
<li><p><strong>isolated_as_missing</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, treat isolated nodes as missing
across the covered tree intervals (see above). If None (default), this
is treated as True.</p></li>
<li><p><strong>samples</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>]</em>)  The nodes for which to output alignments. If
<code class="docutils literal notranslate"><span class="pre">None</span></code> (default), return alignments for all sample nodes in the order
given by the <a class="reference internal" href="#tskit.TreeSequence.samples" title="tskit.TreeSequence.samples"><code class="xref py py-meth docutils literal notranslate"><span class="pre">samples()</span></code></a> method. Non-sample nodes are also supported
and will be decoded at sites in the same way as samples.</p></li>
<li><p><strong>left</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  Alignments will start at this genomic position. If <code class="docutils literal notranslate"><span class="pre">None</span></code>
(default) alignments start at 0.</p></li>
<li><p><strong>right</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  Alignments will stop before this genomic position.
If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) alignments will continue until the end of the
tree sequence.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An iterator over the alignment strings for specified samples in
this tree sequence, in the order given in <code class="docutils literal notranslate"><span class="pre">samples</span></code>. Each string has
length <code class="docutils literal notranslate"><span class="pre">L</span> <span class="pre">=</span> <span class="pre">right</span> <span class="pre">-</span> <span class="pre">left</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable" title="(in Python v3.14)">collections.abc.Iterable</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.14)"><strong>ValueError</strong></a>  if any genome coordinate in this tree sequence is not
discrete, or if the <code class="docutils literal notranslate"><span class="pre">reference_sequence</span></code> is not of the correct length.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#TypeError" title="(in Python v3.14)"><strong>TypeError</strong></a>  if any of the alleles at a site are not a
single ascii character.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.individuals_population">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">individuals_population</span></span><a class="headerlink" href="#tskit.TreeSequence.individuals_population" title="Link to this definition">#</a></dt>
<dd><p>Returns the length-<code class="docutils literal notranslate"><span class="pre">num_individuals</span></code> array containing, for each
individual, the <code class="docutils literal notranslate"><span class="pre">population</span></code> attribute of their nodes, or
<code class="docutils literal notranslate"><span class="pre">tskit.NULL</span></code> for individuals with no nodes. Errors if any individual
has nodes with inconsistent non-NULL populations.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.individuals_time">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">individuals_time</span></span><a class="headerlink" href="#tskit.TreeSequence.individuals_time" title="Link to this definition">#</a></dt>
<dd><p>Returns the length-<code class="docutils literal notranslate"><span class="pre">num_individuals</span></code> array containing, for each
individual, the <code class="docutils literal notranslate"><span class="pre">time</span></code> attribute of their nodes or <code class="docutils literal notranslate"><span class="pre">np.nan</span></code> for
individuals with no nodes. Errors if any individual has nodes with
inconsistent times.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.individuals_location">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">individuals_location</span></span><a class="headerlink" href="#tskit.TreeSequence.individuals_location" title="Link to this definition">#</a></dt>
<dd><p>Convenience method returning the <code class="docutils literal notranslate"><span class="pre">num_individuals</span> <span class="pre">x</span> <span class="pre">n</span></code> array
whose row k-th row contains the <code class="docutils literal notranslate"><span class="pre">location</span></code> property of the k-th
individual. The method only works if all individuals locations
have the same length (which is <code class="docutils literal notranslate"><span class="pre">n</span></code>), and errors otherwise.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.individuals_flags">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">individuals_flags</span></span><a class="headerlink" href="#tskit.TreeSequence.individuals_flags" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the bitwise <code class="docutils literal notranslate"><span class="pre">flags</span></code> column in the
<a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">Individual Table</span></a> as a numpy array (dtype=np.uint32).
Equivalent to <code class="docutils literal notranslate"><span class="pre">ts.tables.individuals.flags</span></code> (but avoiding the full copy
of the table data that accessing <code class="docutils literal notranslate"><span class="pre">ts.tables</span></code> currently entails).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.individuals_metadata">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">individuals_metadata</span></span><a class="headerlink" href="#tskit.TreeSequence.individuals_metadata" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> column in the
<a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">Individual Table</span></a> as a structured numpy array.
The returned dtype will depend on the metadata schema used. Only a subset
of <cite>struct</cite> metadata schemas are supported.
See <a class="reference internal" href="metadata.html#sec-structured-array-metadata"><span class="std std-ref">Structured array metadata</span></a> for more information.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.individuals_nodes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">individuals_nodes</span></span><a class="headerlink" href="#tskit.TreeSequence.individuals_nodes" title="Link to this definition">#</a></dt>
<dd><p>Return an array of node IDs for each individual in the tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Array of shape (num_individuals, max_ploidy) containing node IDs.
Values of -1 indicate unused slots for individuals with ploidy
less than the maximum.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a> (dtype=np.int32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.nodes_metadata">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nodes_metadata</span></span><a class="headerlink" href="#tskit.TreeSequence.nodes_metadata" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> column in the
<a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">Node Table</span></a> as a structured numpy array.
The returned dtype will depend on the metadata schema used. Only a subset
of <cite>struct</cite> metadata schemas are supported.
See <a class="reference internal" href="metadata.html#sec-structured-array-metadata"><span class="std std-ref">Structured array metadata</span></a> for more information.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.nodes_time">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nodes_time</span></span><a class="headerlink" href="#tskit.TreeSequence.nodes_time" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">time</span></code> column in the
<a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">Node Table</span></a> as a numpy array (dtype=np.float64).
Equivalent to <code class="docutils literal notranslate"><span class="pre">ts.tables.nodes.time</span></code> (but avoiding the full copy
of the table data that accessing <code class="docutils literal notranslate"><span class="pre">ts.tables</span></code> currently entails).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.nodes_flags">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nodes_flags</span></span><a class="headerlink" href="#tskit.TreeSequence.nodes_flags" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the bitwise <code class="docutils literal notranslate"><span class="pre">flags</span></code> column in the
<a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">Node Table</span></a> as a numpy array (dtype=np.uint32).
Equivalent to <code class="docutils literal notranslate"><span class="pre">ts.tables.nodes.flags</span></code> (but avoiding the full copy
of the table data that accessing <code class="docutils literal notranslate"><span class="pre">ts.tables</span></code> currently entails).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.nodes_population">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nodes_population</span></span><a class="headerlink" href="#tskit.TreeSequence.nodes_population" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">population</span></code> column in the
<a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">Node Table</span></a> as a numpy array (dtype=np.int32).
Equivalent to <code class="docutils literal notranslate"><span class="pre">ts.tables.nodes.population</span></code> (but avoiding the full copy
of the table data that accessing <code class="docutils literal notranslate"><span class="pre">ts.tables</span></code> currently entails).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.nodes_individual">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nodes_individual</span></span><a class="headerlink" href="#tskit.TreeSequence.nodes_individual" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">individual</span></code> column in the
<a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">Node Table</span></a> as a numpy array (dtype=np.int32).
Equivalent to <code class="docutils literal notranslate"><span class="pre">ts.tables.nodes.individual</span></code> (but avoiding the full copy
of the table data that accessing <code class="docutils literal notranslate"><span class="pre">ts.tables</span></code> currently entails).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.edges_left">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">edges_left</span></span><a class="headerlink" href="#tskit.TreeSequence.edges_left" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">left</span></code> column in the
<a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">Edge Table</span></a> as a numpy array (dtype=np.float64).
Equivalent to <code class="docutils literal notranslate"><span class="pre">ts.tables.edges.left</span></code> (but avoiding the full copy
of the table data that accessing <code class="docutils literal notranslate"><span class="pre">ts.tables</span></code> currently entails).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.edges_right">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">edges_right</span></span><a class="headerlink" href="#tskit.TreeSequence.edges_right" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">right</span></code> column in the
<a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">Edge Table</span></a> as a numpy array (dtype=np.float64).
Equivalent to <code class="docutils literal notranslate"><span class="pre">ts.tables.edges.right</span></code> (but avoiding the full copy
of the table data that accessing <code class="docutils literal notranslate"><span class="pre">ts.tables</span></code> currently entails).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.edges_parent">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">edges_parent</span></span><a class="headerlink" href="#tskit.TreeSequence.edges_parent" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">parent</span></code> column in the
<a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">Edge Table</span></a> as a numpy array (dtype=np.int32).
Equivalent to <code class="docutils literal notranslate"><span class="pre">ts.tables.edges.parent</span></code> (but avoiding the full copy
of the table data that accessing <code class="docutils literal notranslate"><span class="pre">ts.tables</span></code> currently entails).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.edges_child">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">edges_child</span></span><a class="headerlink" href="#tskit.TreeSequence.edges_child" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">child</span></code> column in the
<a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">Edge Table</span></a> as a numpy array (dtype=np.int32).
Equivalent to <code class="docutils literal notranslate"><span class="pre">ts.tables.edges.child</span></code> (but avoiding the full copy
of the table data that accessing <code class="docutils literal notranslate"><span class="pre">ts.tables</span></code> currently entails).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.edges_metadata">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">edges_metadata</span></span><a class="headerlink" href="#tskit.TreeSequence.edges_metadata" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> column in the
<a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">Edge Table</span></a> as a structured numpy array.
The returned dtype will depend on the metadata schema used. Only a subset
of <cite>struct</cite> metadata schemas are supported.
See <a class="reference internal" href="metadata.html#sec-structured-array-metadata"><span class="std std-ref">Structured array metadata</span></a> for more information.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.sites_position">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sites_position</span></span><a class="headerlink" href="#tskit.TreeSequence.sites_position" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">position</span></code> column in the
<a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">Site Table</span></a> as a numpy array (dtype=np.float64).
Equivalent to <code class="docutils literal notranslate"><span class="pre">ts.tables.sites.position</span></code> (but avoiding the full copy
of the table data that accessing <code class="docutils literal notranslate"><span class="pre">ts.tables</span></code> currently entails).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.sites_ancestral_state">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sites_ancestral_state</span></span><a class="headerlink" href="#tskit.TreeSequence.sites_ancestral_state" title="Link to this definition">#</a></dt>
<dd><p>The <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> column in the
<a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">Site Table</span></a> as a numpy array (dtype=StringDtype).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.sites_metadata">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sites_metadata</span></span><a class="headerlink" href="#tskit.TreeSequence.sites_metadata" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> column in the
<a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">Site Table</span></a> as a structured numpy array.
The returned dtype will depend on the metadata schema used. Only a subset
of <cite>struct</cite> metadata schemas are supported.
See <a class="reference internal" href="metadata.html#sec-structured-array-metadata"><span class="std std-ref">Structured array metadata</span></a> for more information.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.mutations_site">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mutations_site</span></span><a class="headerlink" href="#tskit.TreeSequence.mutations_site" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">site</span></code> column in the
<a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">Mutation Table</span></a> as a numpy array (dtype=np.int32).
Equivalent to <code class="docutils literal notranslate"><span class="pre">ts.tables.mutations.site</span></code> (but avoiding the full copy
of the table data that accessing <code class="docutils literal notranslate"><span class="pre">ts.tables</span></code> currently entails).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>To efficently get an array of the number of mutations per site, you
can use <code class="docutils literal notranslate"><span class="pre">np.bincount(ts.mutations_site,</span> <span class="pre">minlength=ts.num_sites)</span></code>.</p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.mutations_node">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mutations_node</span></span><a class="headerlink" href="#tskit.TreeSequence.mutations_node" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">node</span></code> column in the
<a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">Mutation Table</span></a> as a numpy array (dtype=np.int32).
Equivalent to <code class="docutils literal notranslate"><span class="pre">ts.tables.mutations.node</span></code> (but avoiding the full copy
of the table data that accessing <code class="docutils literal notranslate"><span class="pre">ts.tables</span></code> currently entails).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.mutations_parent">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mutations_parent</span></span><a class="headerlink" href="#tskit.TreeSequence.mutations_parent" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">parent</span></code> column in the
<a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">Mutation Table</span></a> as a numpy array (dtype=np.int32).
Equivalent to <code class="docutils literal notranslate"><span class="pre">ts.tables.mutations.parent</span></code> (but avoiding the full copy
of the table data that accessing <code class="docutils literal notranslate"><span class="pre">ts.tables</span></code> currently entails).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.mutations_time">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mutations_time</span></span><a class="headerlink" href="#tskit.TreeSequence.mutations_time" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">time</span></code> column in the
<a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">Mutation Table</span></a> as a numpy array (dtype=np.float64).
Equivalent to <code class="docutils literal notranslate"><span class="pre">ts.tables.mutations.time</span></code> (but avoiding the full copy
of the table data that accessing <code class="docutils literal notranslate"><span class="pre">ts.tables</span></code> currently entails).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.mutations_derived_state">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mutations_derived_state</span></span><a class="headerlink" href="#tskit.TreeSequence.mutations_derived_state" title="Link to this definition">#</a></dt>
<dd><p>Access to the <code class="docutils literal notranslate"><span class="pre">derived_state</span></code> column in the
<a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">Mutation Table</span></a> as a numpy array (dtype=StringDtype).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.mutations_metadata">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mutations_metadata</span></span><a class="headerlink" href="#tskit.TreeSequence.mutations_metadata" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> column in the
<a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">Mutation Table</span></a> as a structured numpy array.
The returned dtype will depend on the metadata schema used. Only a subset
of <cite>struct</cite> metadata schemas are supported.
See <a class="reference internal" href="metadata.html#sec-structured-array-metadata"><span class="std std-ref">Structured array metadata</span></a> for more information.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.mutations_edge">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mutations_edge</span></span><a class="headerlink" href="#tskit.TreeSequence.mutations_edge" title="Link to this definition">#</a></dt>
<dd><p>Return an array of the ID of the edge each mutation sits on in the tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Array of shape (num_mutations,) containing edge IDs.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a> (dtype=np.int32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.mutations_inherited_state">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mutations_inherited_state</span></span><a class="headerlink" href="#tskit.TreeSequence.mutations_inherited_state" title="Link to this definition">#</a></dt>
<dd><p>Return an array of the inherited state for each mutation in the tree sequence.</p>
<p>The inherited state for a mutation is the state that existed at the site
before the mutation occurred. This is either the ancestral state of the site
(if the mutation has no parent) or the derived state of the mutations
parent mutation (if it has a parent).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Array of shape (num_mutations,) containing inherited states.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.migrations_left">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">migrations_left</span></span><a class="headerlink" href="#tskit.TreeSequence.migrations_left" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">left</span></code> column in the
<a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">Migration Table</span></a> as a numpy array (dtype=np.float64).
Equivalent to <code class="docutils literal notranslate"><span class="pre">ts.tables.migrations.left</span></code> (but avoiding the full copy
of the table data that accessing <code class="docutils literal notranslate"><span class="pre">ts.tables</span></code> currently entails).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.migrations_right">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">migrations_right</span></span><a class="headerlink" href="#tskit.TreeSequence.migrations_right" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">right</span></code> column in the
<a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">Migration Table</span></a> as a numpy array (dtype=np.float64).
Equivalent to <code class="docutils literal notranslate"><span class="pre">ts.tables.migrations.right</span></code> (but avoiding the full copy
of the table data that accessing <code class="docutils literal notranslate"><span class="pre">ts.tables</span></code> currently entails).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.migrations_node">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">migrations_node</span></span><a class="headerlink" href="#tskit.TreeSequence.migrations_node" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">node</span></code> column in the
<a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">Migration Table</span></a> as a numpy array (dtype=np.int32).
Equivalent to <code class="docutils literal notranslate"><span class="pre">ts.tables.migrations.node</span></code> (but avoiding the full copy
of the table data that accessing <code class="docutils literal notranslate"><span class="pre">ts.tables</span></code> currently entails).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.migrations_source">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">migrations_source</span></span><a class="headerlink" href="#tskit.TreeSequence.migrations_source" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">source</span></code> column in the
<a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">Migration Table</span></a> as a numpy array (dtype=np.int32).
Equivalent to <code class="docutils literal notranslate"><span class="pre">ts.tables.migrations.source</span></code> (but avoiding the full copy
of the table data that accessing <code class="docutils literal notranslate"><span class="pre">ts.tables</span></code> currently entails).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.migrations_dest">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">migrations_dest</span></span><a class="headerlink" href="#tskit.TreeSequence.migrations_dest" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">dest</span></code> column in the
<a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">Migration Table</span></a> as a numpy array (dtype=np.int32).
Equivalent to <code class="docutils literal notranslate"><span class="pre">ts.tables.migrations.dest</span></code> (but avoiding the full copy
of the table data that accessing <code class="docutils literal notranslate"><span class="pre">ts.tables</span></code> currently entails).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.migrations_time">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">migrations_time</span></span><a class="headerlink" href="#tskit.TreeSequence.migrations_time" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">time</span></code> column in the
<a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">Migration Table</span></a> as a numpy array (dtype=np.float64).
Equivalent to <code class="docutils literal notranslate"><span class="pre">ts.tables.migrations.time</span></code> (but avoiding the full copy
of the table data that accessing <code class="docutils literal notranslate"><span class="pre">ts.tables</span></code> currently entails).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.migrations_metadata">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">migrations_metadata</span></span><a class="headerlink" href="#tskit.TreeSequence.migrations_metadata" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> column in the
<a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">Migration Table</span></a> as a structured numpy array.
The returned dtype will depend on the metadata schema used. Only a subset
of <cite>struct</cite> metadata schemas are supported.
See <a class="reference internal" href="metadata.html#sec-structured-array-metadata"><span class="std std-ref">Structured array metadata</span></a> for more information.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.populations_metadata">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">populations_metadata</span></span><a class="headerlink" href="#tskit.TreeSequence.populations_metadata" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> column in the
<a class="reference internal" href="data-model.html#sec-population-table-definition"><span class="std std-ref">Population Table</span></a> as a structured numpy array.
The returned dtype will depend on the metadata schema used. Only a subset
of <cite>struct</cite> metadata schemas are supported.
See <a class="reference internal" href="metadata.html#sec-structured-array-metadata"><span class="std std-ref">Structured array metadata</span></a> for more information.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.indexes_edge_insertion_order">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">indexes_edge_insertion_order</span></span><a class="headerlink" href="#tskit.TreeSequence.indexes_edge_insertion_order" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">edge_insertion_order</span></code> column in the
<a class="reference internal" href="data-model.html#sec-table-indexes"><span class="std std-ref">Table indexes</span></a> as a numpy array (dtype=np.int32).
Equivalent to <code class="docutils literal notranslate"><span class="pre">ts.tables.indexes.edge_insertion_order</span></code> (but avoiding
the full copy of the table data that accessing <code class="docutils literal notranslate"><span class="pre">ts.tables</span></code>
currently entails).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TreeSequence.indexes_edge_removal_order">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">indexes_edge_removal_order</span></span><a class="headerlink" href="#tskit.TreeSequence.indexes_edge_removal_order" title="Link to this definition">#</a></dt>
<dd><p>Efficient access to the <code class="docutils literal notranslate"><span class="pre">edge_removal_order</span></code> column in the
<a class="reference internal" href="data-model.html#sec-table-indexes"><span class="std std-ref">Table indexes</span></a> as a numpy array (dtype=np.int32).
Equivalent to <code class="docutils literal notranslate"><span class="pre">ts.tables.indexes.edge_removal_order</span></code> (but avoiding
the full copy of the table data that accessing <code class="docutils literal notranslate"><span class="pre">ts.tables</span></code>
currently entails).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.individual">
<span class="sig-name descname"><span class="pre">individual</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id_</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.individual"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.individual" title="Link to this definition">#</a></dt>
<dd><p>Returns the <a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">individual</span></a>
in this tree sequence with the specified ID.  As with python lists, negative
IDs can be used to index backwards from the last individual.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Individual" title="tskit.Individual"><code class="xref py py-class docutils literal notranslate"><span class="pre">Individual</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.node">
<span class="sig-name descname"><span class="pre">node</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id_</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.node"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.node" title="Link to this definition">#</a></dt>
<dd><p>Returns the <a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">node</span></a> in this tree sequence
with the specified ID. As with python lists, negative IDs can be used to
index backwards from the last node.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Node" title="tskit.Node"><code class="xref py py-class docutils literal notranslate"><span class="pre">Node</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.edge">
<span class="sig-name descname"><span class="pre">edge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id_</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.edge"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.edge" title="Link to this definition">#</a></dt>
<dd><p>Returns the <a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">edge</span></a> in this tree sequence
with the specified ID. As with python lists, negative IDs can be used to
index backwards from the last edge.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Edge" title="tskit.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.migration">
<span class="sig-name descname"><span class="pre">migration</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id_</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.migration"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.migration" title="Link to this definition">#</a></dt>
<dd><p>Returns the <a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">migration</span></a> in this tree
sequence with the specified ID. As with python lists, negative IDs can be
used to index backwards from the last migration.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Migration" title="tskit.Migration"><code class="xref py py-class docutils literal notranslate"><span class="pre">Migration</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.mutation">
<span class="sig-name descname"><span class="pre">mutation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id_</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.mutation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.mutation" title="Link to this definition">#</a></dt>
<dd><p>Returns the <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">mutation</span></a> in this tree sequence
with the specified ID. As with python lists, negative IDs can be used to
index backwards from the last mutation.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Mutation" title="tskit.Mutation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mutation</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.site">
<span class="sig-name descname"><span class="pre">site</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id_</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.site"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.site" title="Link to this definition">#</a></dt>
<dd><p>Returns the <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">site</span></a> in this tree sequence
with either the specified ID or position. As with python lists, negative IDs
can be used to index backwards from the last site.</p>
<p>When position is specified instead of site ID, a binary search is done
on the list of positions of the sites to try to find a site
with the user-specified position.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Site" title="tskit.Site"><code class="xref py py-class docutils literal notranslate"><span class="pre">Site</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.population">
<span class="sig-name descname"><span class="pre">population</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id_</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.population"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.population" title="Link to this definition">#</a></dt>
<dd><p>Returns the <a class="reference internal" href="data-model.html#sec-population-table-definition"><span class="std std-ref">population</span></a>
in this tree sequence with the specified ID.  As with python lists, negative
IDs can be used to index backwards from the last population.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.Population" title="tskit.Population"><code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.provenance">
<span class="sig-name descname"><span class="pre">provenance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id_</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.provenance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.provenance" title="Link to this definition">#</a></dt>
<dd><p>Returns the <a class="reference internal" href="data-model.html#sec-provenance-table-definition"><span class="std std-ref">provenance</span></a>
in this tree sequence with the specified ID.  As with python lists,
negative IDs can be used to index backwards from the last provenance.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.samples">
<span class="sig-name descname"><span class="pre">samples</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">population</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">population_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.samples"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.samples" title="Link to this definition">#</a></dt>
<dd><p>Returns an array of the sample node IDs in this tree sequence. If
<cite>population</cite> is specified, only return sample IDs from that population.
It is also possible to restrict samples by time using the parameter
<cite>time</cite>. If <cite>time</cite> is a numeric value, only return sample IDs whose node
time is approximately equal to the specified time. If <cite>time</cite> is a pair
of values of the form <cite>(min_time, max_time)</cite>, only return sample IDs
whose node time <cite>t</cite> is in this interval such that <cite>min_time &lt;= t &lt; max_time</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>population</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The population of interest. If None, do not
filter samples by population.</p></li>
<li><p><strong>population_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  Deprecated alias for <code class="docutils literal notranslate"><span class="pre">population</span></code>.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a><em>,</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.14)"><em>tuple</em></a>)  The time or time interval of interest. If
None, do not filter samples by time.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A numpy array of the node IDs for the samples of interest,
listed in numerical order.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a> (dtype=np.int32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.as_vcf">
<span class="sig-name descname"><span class="pre">as_vcf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.as_vcf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.as_vcf" title="Link to this definition">#</a></dt>
<dd><p>Return the result of <a class="reference internal" href="#tskit.TreeSequence.write_vcf" title="tskit.TreeSequence.write_vcf"><code class="xref py py-meth docutils literal notranslate"><span class="pre">write_vcf()</span></code></a> as a string.
Keyword parameters are as defined in <a class="reference internal" href="#tskit.TreeSequence.write_vcf" title="tskit.TreeSequence.write_vcf"><code class="xref py py-meth docutils literal notranslate"><span class="pre">write_vcf()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A VCF encoding of the variants in this tree sequence as a string.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.write_vcf">
<span class="sig-name descname"><span class="pre">write_vcf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ploidy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contig_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'1'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individuals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individual_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_transform</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">site_mask</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_mask</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isolated_as_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_position_zero</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_non_sample_nodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.write_vcf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.write_vcf" title="Link to this definition">#</a></dt>
<dd><p>Convert the genetic variation data in this tree sequence to Variant
Call Format and write to the specified file-like object.</p>
<p>Multiploid samples in the output VCF are generated either using
individual information in the data model (see
<a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">Individual Table</span></a>), or by combining genotypes for
adjacent sample nodes using the <code class="docutils literal notranslate"><span class="pre">ploidy</span></code> argument. See the
<a class="reference internal" href="export.html#sec-export-vcf-constructing-gt"><span class="std std-ref">Constructing GT values</span></a> section for more details
and examples.</p>
<p>If individuals are defined in the
data model (see <a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">Individual Table</span></a>), the genotypes
for each of the individuals nodes are combined into a phased
multiploid values at each site. By default, all individuals are
included with their sample nodes, individuals with no nodes are
omitted. The <code class="docutils literal notranslate"><span class="pre">include_non_sample_nodes</span></code> argument can be used to
included non-sample nodes in the output VCF.</p>
<p>Subsets or permutations of the sample individuals may be specified
using the <code class="docutils literal notranslate"><span class="pre">individuals</span></code> argument.</p>
<p>Mixed-sample individuals (e.g., those associated with one node
that is a sample and another that is not) in the data model will
only have the sample nodes output by default. However, non-sample
nodes can be included using the <code class="docutils literal notranslate"><span class="pre">include_non_sample_nodes</span></code> argument.</p>
<p>If there are no individuals in the tree sequence,
synthetic individuals are created by combining adjacent samples, and
the number of samples combined is equal to the <code class="docutils literal notranslate"><span class="pre">ploidy</span></code> value (1 by
default). For example, if we have a <code class="docutils literal notranslate"><span class="pre">ploidy</span></code> of 2 and 6 sample nodes,
then we will have 3 diploid samples in the VCF, consisting of the
combined genotypes for samples [0, 1], [2, 3] and [4, 5]. If we had
genotypes 011110 at a particular variant, then we would output the
diploid genotypes 0|1, 1|1 and 1|0 in VCF.</p>
<p>Each individual in the output is identified by a string; these are the
VCF sample names. By default, these are of the form <code class="docutils literal notranslate"><span class="pre">tsk_0</span></code>,
<code class="docutils literal notranslate"><span class="pre">tsk_1</span></code> etc, up to the number of individuals, but can be manually
specified using the <code class="docutils literal notranslate"><span class="pre">individual_names</span></code> argument. We do not check
for duplicates in this array, or perform any checks to ensure that
the output VCF is well-formed.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The default individual names (VCF sample IDs) are always of
the form <code class="docutils literal notranslate"><span class="pre">tsk_0</span></code>, <code class="docutils literal notranslate"><span class="pre">tsk_1</span></code>, , <code class="docutils literal notranslate"><span class="pre">tsk_{N</span> <span class="pre">-</span> <span class="pre">1}</span></code>, where
N is the number of individuals we output. These numbers
are <strong>not</strong> necessarily the individual IDs.</p>
</div>
<p>The REF value in the output VCF is the ancestral allele for a site
and ALT values are the remaining alleles. It is important to note,
therefore, that for real data this means that the REF value for a given
site <strong>may not</strong> be equal to the reference allele. We also do not
check that the alleles result in a valid VCFfor example, it is possible
to use the tab character as an allele, leading to a broken VCF.</p>
<p>The ID value in the output VCF file is the integer ID of the
corresponding <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">site</span></a> (<code class="docutils literal notranslate"><span class="pre">site.id</span></code>).
These ID values can be utilized to match the contents of the VCF file
to the sites in the tree sequence object.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Older code often uses the <code class="docutils literal notranslate"><span class="pre">ploidy=2</span></code> argument, because old
versions of msprime did not output individual data. Specifying
individuals in the tree sequence is more robust, and since tree
sequences now  typically contain individuals (e.g., as produced by
<code class="docutils literal notranslate"><span class="pre">msprime.sim_ancestry(</span> <span class="pre">)</span></code>), this is not necessary, and the
<code class="docutils literal notranslate"><span class="pre">ploidy</span></code> argument can safely be removed as part of the process
of updating from the msprime 0.x legacy API.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>output</strong> (<a class="reference external" href="https://docs.python.org/3/library/io.html#io.IOBase" title="(in Python v3.14)"><em>io.IOBase</em></a>)  The file-like object to write the VCF output.</p></li>
<li><p><strong>ploidy</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The ploidy of the individuals to be written to
VCF. This sample size must be evenly divisible by ploidy. Cannot be
used if there is individual data in the tree sequence.</p></li>
<li><p><strong>contig_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The value of the CHROM column in the output VCF.</p></li>
<li><p><strong>individuals</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>)</em>)  A list containing the individual IDs to
corresponding to the VCF samples. Defaults to all individuals
associated with sample nodes in the tree sequence.
See the {ref}`sec_export_vcf_constructing_gt` section for more
details and examples.</p></li>
<li><p><strong>individual_names</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>)</em>)  A list of string names to identify
individual columns in the VCF. In VCF nomenclature, these are the
sample IDs. If specified, this must be a list of strings of
length equal to the number of individuals to be output. Note that
we do not check the form of these strings in any way, so that is
is possible to output malformed VCF (for example, by embedding a
tab character within on of the names). The default is to output
<code class="docutils literal notranslate"><span class="pre">tsk_j</span></code> for the jth individual.
See the <a class="reference internal" href="export.html#sec-export-vcf-individual-names"><span class="std std-ref">Individual names</span></a> for examples
and more information.</p></li>
<li><p><strong>position_transform</strong>  A callable that transforms the
site position values into integer valued coordinates suitable for
VCF. The function takes a single positional parameter x and must
return an integer numpy array the same dimension as x. By default,
this is set to <code class="docutils literal notranslate"><span class="pre">numpy.round()</span></code> which will round values to the
nearest integer. If the string legacy is provided here, the
pre 0.2.0 legacy behaviour of rounding values to the nearest integer
(starting from 1) and avoiding the output of identical positions
by incrementing is used.
See the <a class="reference internal" href="export.html#sec-export-vcf-modifying-coordinates"><span class="std std-ref">Modifying coordinates</span></a> for examples
and more information.</p></li>
<li><p><strong>site_mask</strong>  A numpy boolean array (or something convertable to
a numpy boolean array) with num_sites elements, used to mask out
sites in the output. If  <code class="docutils literal notranslate"><span class="pre">site_mask[j]</span></code> is True, then this
site (i.e., the line in the VCF file) will be omitted.
See the <a class="reference internal" href="export.html#sec-export-vcf-masking-output"><span class="std std-ref">Masking output</span></a> for examples
and more information.</p></li>
<li><p><strong>sample_mask</strong>  A numpy boolean array (or something convertable to
a numpy boolean array) with num_samples elements, or a callable
that returns such an array, such that if
<code class="docutils literal notranslate"><span class="pre">sample_mask[j]</span></code> is True, then the genotype for sample <code class="docutils literal notranslate"><span class="pre">j</span></code>
will be marked as missing using a .. If <code class="docutils literal notranslate"><span class="pre">sample_mask</span></code> is a
callable, it must take a single argument and return a boolean
numpy array. This function will be called for each (unmasked) site
with the corresponding <a class="reference internal" href="#tskit.Variant" title="tskit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> object, allowing
for dynamic masks to be generated.
See the <a class="reference internal" href="export.html#sec-export-vcf-masking-output"><span class="std std-ref">Masking output</span></a> for examples
and more information.</p></li>
<li><p><strong>isolated_as_missing</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, the genotype value assigned to
missing samples (i.e., isolated samples without mutations) is .
If False, missing samples will be assigned the ancestral allele.
See <a class="reference internal" href="#tskit.TreeSequence.variants" title="tskit.TreeSequence.variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">variants()</span></code></a> for more information. Default: True.</p></li>
<li><p><strong>allow_position_zero</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True allow sites with position zero to be
output to the VCF, otherwise if one is present an error will be raised.
The VCF spec does not allow for sites at position 0. However, in practise
many tools will be fine with this. Default: False.</p></li>
<li><p><strong>include_non_sample_nodes</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, include non-sample nodes
in the output VCF. By default, only sample nodes are included.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.write_fasta">
<span class="sig-name descname"><span class="pre">write_fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_or_path</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wrap_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">60</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">missing_data_character</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isolated_as_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.write_fasta"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.write_fasta" title="Link to this definition">#</a></dt>
<dd><p>Writes the <a class="reference internal" href="#tskit.TreeSequence.alignments" title="tskit.TreeSequence.alignments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">alignments()</span></code></a> for this tree sequence to file in
<a class="reference external" href="https://en.wikipedia.org/wiki/FASTA_format">FASTA</a> format.
Please see the <a class="reference internal" href="#tskit.TreeSequence.alignments" title="tskit.TreeSequence.alignments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">alignments()</span></code></a> method for details on how
reference sequences are handled.</p>
<p>Alignments are returned for the
<a class="reference internal" href="glossary.html#sec-data-model-definitions"><span class="std std-ref">sample nodes</span></a> in this tree
sequence, and a sample with node id <code class="docutils literal notranslate"><span class="pre">u</span></code> is given the label
<code class="docutils literal notranslate"><span class="pre">f&quot;n{u}&quot;</span></code>, following the same convention as the
<a class="reference internal" href="#tskit.TreeSequence.write_nexus" title="tskit.TreeSequence.write_nexus"><code class="xref py py-meth docutils literal notranslate"><span class="pre">write_nexus()</span></code></a> and <a class="reference internal" href="#tskit.Tree.as_newick" title="tskit.Tree.as_newick"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.as_newick()</span></code></a> methods.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">wrap_width</span></code> parameter controls the maximum width of lines
of sequence data in the output. By default this is 60
characters in accordance with fasta standard outputs. To turn off
line-wrapping of sequences, set <code class="docutils literal notranslate"><span class="pre">wrap_width</span></code> = 0.</p>
<p>Example usage:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span><span class="o">.</span><span class="n">write_fasta</span><span class="p">(</span><span class="s2">&quot;output.fa&quot;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_or_path</strong>  The file object or path to write the output.
Paths can be either strings or <a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.14)"><code class="docutils literal notranslate"><span class="pre">pathlib.Path</span></code></a> objects.</p></li>
<li><p><strong>wrap_width</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The number of sequence
characters to include on each line in the fasta file, before wrapping
to the next line for each sequence, or 0 to turn off line wrapping.
(Default=60).</p></li>
<li><p><strong>reference_sequence</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  As for the <a class="reference internal" href="#tskit.TreeSequence.alignments" title="tskit.TreeSequence.alignments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">alignments()</span></code></a> method.</p></li>
<li><p><strong>missing_data_character</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  As for the <a class="reference internal" href="#tskit.TreeSequence.alignments" title="tskit.TreeSequence.alignments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">alignments()</span></code></a> method.</p></li>
<li><p><strong>isolated_as_missing</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  As for the <a class="reference internal" href="#tskit.TreeSequence.alignments" title="tskit.TreeSequence.alignments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">alignments()</span></code></a> method.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.as_fasta">
<span class="sig-name descname"><span class="pre">as_fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.as_fasta"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.as_fasta" title="Link to this definition">#</a></dt>
<dd><p>Return the result of <a class="reference internal" href="#tskit.TreeSequence.write_fasta" title="tskit.TreeSequence.write_fasta"><code class="xref py py-meth docutils literal notranslate"><span class="pre">write_fasta()</span></code></a> as a string.
Keyword parameters are as defined in <a class="reference internal" href="#tskit.TreeSequence.write_fasta" title="tskit.TreeSequence.write_fasta"><code class="xref py py-meth docutils literal notranslate"><span class="pre">write_fasta()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A FASTA encoding of the alignments in this tree sequence as a string.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.write_nexus">
<span class="sig-name descname"><span class="pre">write_nexus</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_or_path</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">precision</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_trees</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_alignments</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">missing_data_character</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isolated_as_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.write_nexus"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.write_nexus" title="Link to this definition">#</a></dt>
<dd><p>Returns a <a class="reference external" href="https://en.wikipedia.org/wiki/Nexus_file">nexus encoding</a>
of this tree sequence. By default, tree topologies are included
in the output, and sequence data alignments are included by default
if this tree sequence has discrete genome coordinates and one or
more sites. Inclusion of these sections can be controlled manually
using the <code class="docutils literal notranslate"><span class="pre">include_trees</span></code> and <code class="docutils literal notranslate"><span class="pre">include_alignments</span></code> parameters.</p>
<p>Tree topologies and branch lengths are listed
sequentially in the TREES block and the spatial location of each tree
encoded within the tree name labels. Specifically, a tree spanning
the interval <span class="math notranslate nohighlight">\([x, y)`\)</span> is given the name <code class="docutils literal notranslate"><span class="pre">f&quot;t{x}^{y}&quot;</span></code>
(See below for a description of the precision at which these spatial
coordinates are printed out).</p>
<p>The <a class="reference internal" href="glossary.html#sec-data-model-definitions"><span class="std std-ref">sample nodes</span></a> in this tree
sequence are regarded as taxa, and a sample with node id <code class="docutils literal notranslate"><span class="pre">u</span></code>
is given the label <code class="docutils literal notranslate"><span class="pre">f&quot;n{u}&quot;</span></code>, following the same convention
as the <a class="reference internal" href="#tskit.Tree.as_newick" title="tskit.Tree.as_newick"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.as_newick()</span></code></a> method.</p>
<p>By default, genome positions are printed out with with sufficient
precision for them to be recovered exactly in double precision.
If the tree sequence is defined on a <a class="reference internal" href="#tskit.TreeSequence.discrete_genome" title="tskit.TreeSequence.discrete_genome"><code class="xref py py-attr docutils literal notranslate"><span class="pre">discrete_genome</span></code></a>,
then positions are written out as integers. Otherwise, 17 digits
of precision is used. Branch length precision defaults are handled
in the same way as <a class="reference internal" href="#tskit.Tree.as_newick" title="tskit.Tree.as_newick"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.as_newick()</span></code></a>.</p>
<p>If the <code class="docutils literal notranslate"><span class="pre">precision</span></code> argument is provided, genome positions and
branch lengths are printed out with this many digits of precision.</p>
<p>For example, here is the nexus encoding of a simple tree sequence
with integer times and genome coordinates with three samples
and two trees:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#NEXUS</span>
<span class="n">BEGIN</span> <span class="n">TAXA</span><span class="p">;</span>
  <span class="n">DIMENSIONS</span> <span class="n">NTAX</span><span class="o">=</span><span class="mi">3</span><span class="p">;</span>
  <span class="n">TAXLABELS</span> <span class="n">n0</span> <span class="n">n1</span> <span class="n">n2</span><span class="p">;</span>
<span class="n">END</span><span class="p">;</span>
<span class="n">BEGIN</span> <span class="n">TREES</span><span class="p">;</span>
  <span class="n">TREE</span> <span class="n">t0</span><span class="o">^</span><span class="mi">2</span> <span class="o">=</span> <span class="p">[</span><span class="o">&amp;</span><span class="n">R</span><span class="p">]</span> <span class="p">(</span><span class="n">n0</span><span class="p">:</span><span class="mi">3</span><span class="p">,(</span><span class="n">n1</span><span class="p">:</span><span class="mi">2</span><span class="p">,</span><span class="n">n2</span><span class="p">:</span><span class="mi">2</span><span class="p">):</span><span class="mi">1</span><span class="p">);</span>
  <span class="n">TREE</span> <span class="n">t2</span><span class="o">^</span><span class="mi">10</span> <span class="o">=</span> <span class="p">[</span><span class="o">&amp;</span><span class="n">R</span><span class="p">]</span> <span class="p">(</span><span class="n">n1</span><span class="p">:</span><span class="mi">2</span><span class="p">,(</span><span class="n">n0</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span><span class="n">n2</span><span class="p">:</span><span class="mi">1</span><span class="p">):</span><span class="mi">1</span><span class="p">);</span>
<span class="n">END</span><span class="p">;</span>
</pre></div>
</div>
<p>If sequence data <a class="reference internal" href="#tskit.TreeSequence.alignments" title="tskit.TreeSequence.alignments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">alignments()</span></code></a> are defined for this tree sequence
and there is at least one site present, sequence alignment data will also
be included by default (this can be suppressed by setting
<code class="docutils literal notranslate"><span class="pre">include_alignments=False</span></code>). For example, this tree sequence has
a sequence length of 10, two variable sites and no
<a class="reference internal" href="data-model.html#sec-data-model-reference-sequence"><span class="std std-ref">reference sequence</span></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>#NEXUS
BEGIN TAXA;
  DIMENSIONS NTAX=3;
  TAXLABELS n0 n1 n2;
END;
BEGIN DATA;
  DIMENSIONS NCHAR=10;
  FORMAT DATATYPE=DNA MISSING=?;
  MATRIX
    n0 ??G??????T
    n1 ??A??????C
    n2 ??A??????C
  ;
END;
BEGIN TREES;
  TREE t0^10 = [&amp;R] (n0:2,(n1:1,n2:1):1);
END;
</pre></div>
</div>
<p>Please see the <a class="reference internal" href="#tskit.TreeSequence.alignments" title="tskit.TreeSequence.alignments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">alignments()</span></code></a> method for details on how
reference sequences are handled.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Note the default <code class="docutils literal notranslate"><span class="pre">missing_data_character</span></code> for this method
is ? rather then N, in keeping with common conventions for
nexus data. This can be changed using the <code class="docutils literal notranslate"><span class="pre">missing_data_character</span></code>
parameter.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p><a class="reference internal" href="data-model.html#sec-data-model-missing-data"><span class="std std-ref">Missing data</span></a>
is not supported for encoding tree topology information
as our convention of using trees with multiple roots
is not often supported by newick parsers. Thus, the method
will raise a ValueError if we try to output trees with
multiple roots.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>precision</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The numerical precision with which branch lengths
and tree positions are printed.</p></li>
<li><p><strong>include_trees</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  True if the tree topology information should
be included; False otherwise (default=True).</p></li>
<li><p><strong>include_alignments</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  True if the sequence data alignment information
should be included; False otherwise (default=True if sequence alignments
are well-defined and the tree sequence contains at least one site).</p></li>
<li><p><strong>reference_sequence</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  As for the <a class="reference internal" href="#tskit.TreeSequence.alignments" title="tskit.TreeSequence.alignments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">alignments()</span></code></a> method.</p></li>
<li><p><strong>missing_data_character</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  As for the <a class="reference internal" href="#tskit.TreeSequence.alignments" title="tskit.TreeSequence.alignments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">alignments()</span></code></a> method,
but defaults to ?.</p></li>
<li><p><strong>isolated_as_missing</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  As for the <a class="reference internal" href="#tskit.TreeSequence.alignments" title="tskit.TreeSequence.alignments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">alignments()</span></code></a> method.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A nexus representation of this <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.as_nexus">
<span class="sig-name descname"><span class="pre">as_nexus</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.as_nexus"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.as_nexus" title="Link to this definition">#</a></dt>
<dd><p>Return the result of <a class="reference internal" href="#tskit.TreeSequence.write_nexus" title="tskit.TreeSequence.write_nexus"><code class="xref py py-meth docutils literal notranslate"><span class="pre">write_nexus()</span></code></a> as a string.
Keyword parameters are as defined in <a class="reference internal" href="#tskit.TreeSequence.write_nexus" title="tskit.TreeSequence.write_nexus"><code class="xref py py-meth docutils literal notranslate"><span class="pre">write_nexus()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A nexus encoding of the alignments in this tree sequence as a string.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.to_macs">
<span class="sig-name descname"><span class="pre">to_macs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.to_macs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.to_macs" title="Link to this definition">#</a></dt>
<dd><p>Return a <a class="reference external" href="https://github.com/gchen98/macs">macs encoding</a>
of this tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The macs representation of this TreeSequence as a string.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.simplify">
<span class="sig-name descname"><span class="pre">simplify</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">map_nodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reduce_to_site_topology</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_populations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_individuals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_nodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">update_sample_flags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_unary</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_unary_in_individuals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_input_roots</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_zero_mutation_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.simplify"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.simplify" title="Link to this definition">#</a></dt>
<dd><p>Returns a simplified tree sequence that retains only the history of
the nodes given in the list <code class="docutils literal notranslate"><span class="pre">samples</span></code>. If <code class="docutils literal notranslate"><span class="pre">map_nodes</span></code> is true,
also return a numpy array whose <code class="docutils literal notranslate"><span class="pre">u</span></code>-th element is the ID of the node
in the simplified tree sequence that corresponds to node <code class="docutils literal notranslate"><span class="pre">u</span></code> in the
original tree sequence, or <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> (-1) if <code class="docutils literal notranslate"><span class="pre">u</span></code> is no longer
present in the simplified tree sequence.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you wish to simplify a set of tables that do not satisfy all
requirements for building a TreeSequence, then use
<a class="reference internal" href="#tskit.TableCollection.simplify" title="tskit.TableCollection.simplify"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.simplify()</span></code></a>.</p>
</div>
<p>If the <code class="docutils literal notranslate"><span class="pre">reduce_to_site_topology</span></code> parameter is True, the returned tree
sequence will contain only topological information that is necessary to
represent the trees that contain sites. If there are zero sites in this
tree sequence, this will result in an output tree sequence with zero edges.
When the number of sites is greater than zero, every tree in the output
tree sequence will contain at least one site. For a given site, the
topology of the tree containing that site will be identical
(up to node ID remapping) to the topology of the corresponding tree
in the input tree sequence.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">filter_populations</span></code>, <code class="docutils literal notranslate"><span class="pre">filter_individuals</span></code>, <code class="docutils literal notranslate"><span class="pre">filter_sites</span></code>, or
<code class="docutils literal notranslate"><span class="pre">filter_nodes</span></code> is True, any of the corresponding objects that are not
referenced elsewhere are filtered out. As this is the default behaviour,
it is important to realise IDs for these objects may change through
simplification. By setting these parameters to False, however, the
corresponding tables can be preserved without changes.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">filter_nodes</span></code> is False, then the output node table will be
unchanged except for updating the sample status of nodes and any ID
remappings caused by filtering individuals and populations (if the
<code class="docutils literal notranslate"><span class="pre">filter_individuals</span></code> and <code class="docutils literal notranslate"><span class="pre">filter_populations</span></code> options are enabled).
Nodes that are in the specified list of <code class="docutils literal notranslate"><span class="pre">samples</span></code> will be marked as
samples in the output, and nodes that are currently marked as samples
in the node table but not in the specified list of <code class="docutils literal notranslate"><span class="pre">samples</span></code> will
have their <a class="reference internal" href="#tskit.NODE_IS_SAMPLE" title="tskit.NODE_IS_SAMPLE"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NODE_IS_SAMPLE</span></code></a> flag cleared. Note also that
the order of the <code class="docutils literal notranslate"><span class="pre">samples</span></code> list is not meaningful when
<code class="docutils literal notranslate"><span class="pre">filter_nodes</span></code> is False. In this case, the returned node mapping is
always the identity mapping, such that <code class="docutils literal notranslate"><span class="pre">a[u]</span> <span class="pre">==</span> <span class="pre">u</span></code> for all nodes.</p>
<p>Setting the <code class="docutils literal notranslate"><span class="pre">update_sample_flags</span></code> parameter to False disables the
automatic sample status update of nodes (described above) from
occuring, making it the responsibility of calling code to keep track of
the ultimate sample status of nodes. This is an advanced option, mostly
of use when combined with the <code class="docutils literal notranslate"><span class="pre">filter_nodes=False</span></code>,
<code class="docutils literal notranslate"><span class="pre">filter_populations=False</span></code> and <code class="docutils literal notranslate"><span class="pre">filter_individuals=False</span></code> options,
which then guarantees that the node table will not be altered by
simplification.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samples</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>]</em>)  A list of node IDs to retain as samples. They
need not be nodes marked as samples in the original tree sequence, but
will constitute the entire set of samples in the returned tree sequence.
If not specified or None, use all nodes marked with the IS_SAMPLE flag.
The list may be provided as a numpy array (or array-like) object
(dtype=np.int32).</p></li>
<li><p><strong>map_nodes</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, return a tuple containing the resulting
tree sequence and a numpy array mapping node IDs in the current tree
sequence to their corresponding node IDs in the returned tree sequence.
If False (the default), return only the tree sequence object itself.</p></li>
<li><p><strong>reduce_to_site_topology</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to reduce the topology down
to the trees that are present at sites. (Default: False)</p></li>
<li><p><strong>filter_populations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, remove any populations that are
not referenced by nodes after simplification; new population IDs are
allocated sequentially from zero. If False, the population table will
not be altered in any way. (Default: None, treated as True)</p></li>
<li><p><strong>filter_individuals</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, remove any individuals that are
not referenced by nodes after simplification; new individual IDs are
allocated sequentially from zero. If False, the individual table will
not be altered in any way. (Default: None, treated as True)</p></li>
<li><p><strong>filter_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, remove any sites that are
not referenced by mutations after simplification; new site IDs are
allocated sequentially from zero. If False, the site table will not
be altered in any way. (Default: None, treated as True)</p></li>
<li><p><strong>filter_nodes</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, remove any nodes that are
not referenced by edges after simplification. If False, the only
potential change to the node table may be to change the node flags
(if <code class="docutils literal notranslate"><span class="pre">samples</span></code> is specified and different from the existing samples).
(Default: None, treated as True)</p></li>
<li><p><strong>update_sample_flags</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, update node flags to so that
nodes in the specified list of samples have the NODE_IS_SAMPLE
flag after simplification, and nodes that are not in this list
do not. (Default: None, treated as True)</p></li>
<li><p><strong>keep_unary</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, preserve unary nodes (i.e., nodes with
exactly one child) that exist on the path from samples to root.
(Default: False)</p></li>
<li><p><strong>keep_unary_in_individuals</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, preserve unary nodes
that exist on the path from samples to root, but only if they are
associated with an individual in the individuals table. Cannot be
specified at the same time as <code class="docutils literal notranslate"><span class="pre">keep_unary</span></code>. (Default: <code class="docutils literal notranslate"><span class="pre">None</span></code>,
equivalent to False)</p></li>
<li><p><strong>keep_input_roots</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to retain history ancestral to the
MRCA of the samples. If <code class="docutils literal notranslate"><span class="pre">False</span></code>, no topology older than the MRCAs of the
samples will be included. If <code class="docutils literal notranslate"><span class="pre">True</span></code> the roots of all trees in the returned
tree sequence will be the same roots as in the original tree sequence.
(Default: False)</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, record details of this call to
simplify in the returned tree sequences provenance information
(Default: True).</p></li>
<li><p><strong>filter_zero_mutation_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Deprecated alias for <code class="docutils literal notranslate"><span class="pre">filter_sites</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The simplified tree sequence, or (if <code class="docutils literal notranslate"><span class="pre">map_nodes</span></code> is True)
a tuple consisting of the simplified tree sequence and a numpy array
mapping source node IDs to their corresponding IDs in the new tree
sequence.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence">tskit.TreeSequence</a> or (<a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence">tskit.TreeSequence</a>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.delete_sites">
<span class="sig-name descname"><span class="pre">delete_sites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">site_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.delete_sites"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.delete_sites" title="Link to this definition">#</a></dt>
<dd><p>Returns a copy of this tree sequence with the specified sites (and their
associated mutations) entirely removed. The site IDs do not need to be in any
particular order, and specifying the same ID multiple times does not have any
effect (i.e., calling <code class="docutils literal notranslate"><span class="pre">tree_sequence.delete_sites([0,</span> <span class="pre">1,</span> <span class="pre">1])</span></code> has the same
effect as calling <code class="docutils literal notranslate"><span class="pre">tree_sequence.delete_sites([0,</span> <span class="pre">1])</span></code>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>To remove only the mutations associated with a site, but keep the site
itself, use the <a class="reference internal" href="#tskit.MutationTable.keep_rows" title="tskit.MutationTable.keep_rows"><code class="xref py py-meth docutils literal notranslate"><span class="pre">MutationTable.keep_rows()</span></code></a> method.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>site_ids</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>]</em>)  A list of site IDs specifying the sites to remove.</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If <code class="docutils literal notranslate"><span class="pre">True</span></code>, add details of this operation to the
provenance information of the returned tree sequence. (Default: <code class="docutils literal notranslate"><span class="pre">True</span></code>).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.delete_intervals">
<span class="sig-name descname"><span class="pre">delete_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">simplify</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.delete_intervals"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.delete_intervals" title="Link to this definition">#</a></dt>
<dd><p>Returns a copy of this tree sequence for which information in the
specified list of genomic intervals has been deleted. Edges spanning these
intervals are truncated or deleted, and sites and mutations falling within
them are discarded. Note that it is the information in the intervals that
is deleted, not the intervals themselves, so in particular, all samples
will be isolated in the deleted intervals.</p>
<p>Note that node IDs may change as a result of this operation,
as by default <a class="reference internal" href="#tskit.TreeSequence.simplify" title="tskit.TreeSequence.simplify"><code class="xref py py-meth docutils literal notranslate"><span class="pre">simplify()</span></code></a> is called on the returned tree sequence
to remove redundant nodes. If you wish to map node IDs onto the same
nodes before and after this method has been called, specify <code class="docutils literal notranslate"><span class="pre">simplify=False</span></code>.</p>
<p>See also <a class="reference internal" href="#tskit.TreeSequence.keep_intervals" title="tskit.TreeSequence.keep_intervals"><code class="xref py py-meth docutils literal notranslate"><span class="pre">keep_intervals()</span></code></a>, <a class="reference internal" href="#tskit.TreeSequence.ltrim" title="tskit.TreeSequence.ltrim"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ltrim()</span></code></a>, <a class="reference internal" href="#tskit.TreeSequence.rtrim" title="tskit.TreeSequence.rtrim"><code class="xref py py-meth docutils literal notranslate"><span class="pre">rtrim()</span></code></a>, and
<a class="reference internal" href="data-model.html#sec-data-model-missing-data"><span class="std std-ref">missing data</span></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> (<em>array_like</em>)  A list (start, end) pairs describing the
genomic intervals to delete. Intervals must be non-overlapping and
in increasing order. The list of intervals must be interpretable as a
2D numpy array with shape (N, 2), where N is the number of intervals.</p></li>
<li><p><strong>simplify</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, return a simplified tree sequence where nodes
no longer used are discarded. (Default: True).</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If <code class="docutils literal notranslate"><span class="pre">True</span></code>, add details of this operation to the
provenance information of the returned tree sequence. (Default: <code class="docutils literal notranslate"><span class="pre">True</span></code>).</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence">tskit.TreeSequence</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.keep_intervals">
<span class="sig-name descname"><span class="pre">keep_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">simplify</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.keep_intervals"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.keep_intervals" title="Link to this definition">#</a></dt>
<dd><p>Returns a copy of this tree sequence which includes only information in
the specified list of genomic intervals. Edges are truncated to lie within
these intervals, and sites and mutations falling outside these intervals
are discarded.  Note that it is the information outside the intervals that
is deleted, not the intervals themselves, so in particular, all samples
will be isolated outside of the retained intervals.</p>
<p>Note that node IDs may change as a result of this operation,
as by default <a class="reference internal" href="#tskit.TreeSequence.simplify" title="tskit.TreeSequence.simplify"><code class="xref py py-meth docutils literal notranslate"><span class="pre">simplify()</span></code></a> is called on the returned tree sequence
to remove redundant nodes. If you wish to map node IDs onto the same
nodes before and after this method has been called, specify <code class="docutils literal notranslate"><span class="pre">simplify=False</span></code>.</p>
<p>See also <a class="reference internal" href="#tskit.TreeSequence.keep_intervals" title="tskit.TreeSequence.keep_intervals"><code class="xref py py-meth docutils literal notranslate"><span class="pre">keep_intervals()</span></code></a>, <a class="reference internal" href="#tskit.TreeSequence.ltrim" title="tskit.TreeSequence.ltrim"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ltrim()</span></code></a>, <a class="reference internal" href="#tskit.TreeSequence.rtrim" title="tskit.TreeSequence.rtrim"><code class="xref py py-meth docutils literal notranslate"><span class="pre">rtrim()</span></code></a>, and
<a class="reference internal" href="data-model.html#sec-data-model-missing-data"><span class="std std-ref">missing data</span></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> (<em>array_like</em>)  A list (start, end) pairs describing the
genomic intervals to keep. Intervals must be non-overlapping and
in increasing order. The list of intervals must be interpretable as a
2D numpy array with shape (N, 2), where N is the number of intervals.</p></li>
<li><p><strong>simplify</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, return a simplified tree sequence where nodes
no longer used are discarded. (Default: True).</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, add details of this operation to the
provenance information of the returned tree sequence.
(Default: True).</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence">tskit.TreeSequence</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.ltrim">
<span class="sig-name descname"><span class="pre">ltrim</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.ltrim"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.ltrim" title="Link to this definition">#</a></dt>
<dd><p>Returns a copy of this tree sequence with a potentially changed coordinate
system, such that empty regions (i.e., those not covered by any edge) at the
start of the tree sequence are trimmed away, and the leftmost edge starts at
position 0. This affects the reported position of sites and
edges. Additionally, sites and their associated mutations to the left of
the new zero point are thrown away.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, add details of this operation to the
provenance information of the returned tree sequence. (Default: True).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.rtrim">
<span class="sig-name descname"><span class="pre">rtrim</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.rtrim"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.rtrim" title="Link to this definition">#</a></dt>
<dd><p>Returns a copy of this tree sequence with the <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> property reset
so that the sequence ends at the end of the rightmost edge. Additionally, sites
and their associated mutations at positions greater than the new
<code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> are thrown away.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, add details of this operation to the
provenance information of the returned tree sequence. (Default: True).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.trim">
<span class="sig-name descname"><span class="pre">trim</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.trim"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.trim" title="Link to this definition">#</a></dt>
<dd><p>Returns a copy of this tree sequence with any empty regions (i.e., those not
covered by any edge) on the right and left trimmed away. This may reset both the
coordinate system and the <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> property. It is functionally
equivalent to <a class="reference internal" href="#tskit.TreeSequence.rtrim" title="tskit.TreeSequence.rtrim"><code class="xref py py-meth docutils literal notranslate"><span class="pre">rtrim()</span></code></a> followed by <a class="reference internal" href="#tskit.TreeSequence.ltrim" title="tskit.TreeSequence.ltrim"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ltrim()</span></code></a>. Sites and their
associated mutations in the empty regions are thrown away.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, add details of this operation to the
provenance information of the returned tree sequence. (Default: True).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.shift">
<span class="sig-name descname"><span class="pre">shift</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.shift"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.shift" title="Link to this definition">#</a></dt>
<dd><p>Shift the coordinate system (used by edges and sites) of this TableCollection by
a given value. Positive values shift the coordinate system to the right, negative
values to the left. The sequence length of the tree sequence will be changed by
<code class="docutils literal notranslate"><span class="pre">value</span></code>, unless <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> is given, in which case this will be used
for the new sequence length.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>By setting <code class="docutils literal notranslate"><span class="pre">value=0</span></code>, this method will simply return a tree sequence
with a new sequence length.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>value</strong>  The amount by which to shift the coordinate system.</p></li>
<li><p><strong>sequence_length</strong>  The new sequence length of the tree sequence. If
<code class="docutils literal notranslate"><span class="pre">None</span></code> (default) add <code class="docutils literal notranslate"><span class="pre">value</span></code> to the sequence length.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.14)"><strong>ValueError</strong></a>  If the new coordinate system is invalid (e.g., if
shifting the coordinate system results in negative coordinates).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.concatenate">
<span class="sig-name descname"><span class="pre">concatenate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_mappings</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_populations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.concatenate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.concatenate" title="Link to this definition">#</a></dt>
<dd><p>Concatenate a set of tree sequences to the right of this one, by shifting
their coordinate systems and adding all edges, sites, mutations, and
any additional nodes, individuals, or populations needed for these.
Concretely, to concatenate an <code class="docutils literal notranslate"><span class="pre">other</span></code> tree sequence to <code class="docutils literal notranslate"><span class="pre">self</span></code>, the value
of <code class="docutils literal notranslate"><span class="pre">self.sequence_length</span></code> is added to all genomic coordinates in <code class="docutils literal notranslate"><span class="pre">other</span></code>,
and then the concatenated tree sequence  will contain all edges, sites, and
mutations in both. Which nodes in <code class="docutils literal notranslate"><span class="pre">other</span></code> are treated as new, and hence
added as well, is controlled by <code class="docutils literal notranslate"><span class="pre">node_mappings</span></code>. Any individuals to which
new nodes belong are added as well.</p>
<p>The method uses <a class="reference internal" href="#tskit.TreeSequence.shift" title="tskit.TreeSequence.shift"><code class="xref py py-meth docutils literal notranslate"><span class="pre">shift()</span></code></a> followed by <a class="reference internal" href="#tskit.TreeSequence.union" title="tskit.TreeSequence.union"><code class="xref py py-meth docutils literal notranslate"><span class="pre">union()</span></code></a>, with
<code class="docutils literal notranslate"><span class="pre">all_mutations=True</span></code>, <code class="docutils literal notranslate"><span class="pre">all_edges=True</span></code>, and <code class="docutils literal notranslate"><span class="pre">check_shared_equality=False</span></code>.</p>
<p>By default, the samples in current and input tree sequences are assumed to
refer to the same nodes, and are matched based on the numerical order of
sample node IDs; all other nodes are assumed to be new. This can be
changed by providing explicit <code class="docutils literal notranslate"><span class="pre">node_mappings</span></code> for each input tree sequence
(see below).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>To add gaps between the concatenated tree sequences, use <a class="reference internal" href="#tskit.TreeSequence.shift" title="tskit.TreeSequence.shift"><code class="xref py py-meth docutils literal notranslate"><span class="pre">shift()</span></code></a>
or to remove gaps, use <a class="reference internal" href="#tskit.TreeSequence.trim" title="tskit.TreeSequence.trim"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trim()</span></code></a> before concatenating.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>*args</strong> (<a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><em>TreeSequence</em></a>)  A list of other tree sequences to append to
the right of this one.</p></li>
<li><p><strong>node_mappings</strong> (<em>Union</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>, </em><em>None</em><em>]</em>)  A list of node mappings for each
input tree sequence in <code class="docutils literal notranslate"><span class="pre">args</span></code>. Each should either be an array of
integers of the same length as the number of nodes in the equivalent
input tree sequence (see <a class="reference internal" href="#tskit.TreeSequence.union" title="tskit.TreeSequence.union"><code class="xref py py-meth docutils literal notranslate"><span class="pre">union()</span></code></a> for details), or
<code class="docutils literal notranslate"><span class="pre">None</span></code>. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, only sample nodes are mapped to each other.
Default: <code class="docutils literal notranslate"><span class="pre">None</span></code>, treated as <code class="docutils literal notranslate"><span class="pre">[None]</span> <span class="pre">*</span> <span class="pre">len(args)</span></code>.</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True (default), record details of this
call to <code class="docutils literal notranslate"><span class="pre">concatenate</span></code> in the returned tree sequences provenance
information (Default: True).</p></li>
<li><p><strong>add_populations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True (default), nodes new to <code class="docutils literal notranslate"><span class="pre">self</span></code> will
be assigned new population IDs (see <a class="reference internal" href="#tskit.TreeSequence.union" title="tskit.TreeSequence.union"><code class="xref py py-meth docutils literal notranslate"><span class="pre">union()</span></code></a>)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.split_edges">
<span class="sig-name descname"><span class="pre">split_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">time</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">population</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.split_edges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.split_edges" title="Link to this definition">#</a></dt>
<dd><p>Returns a copy of this tree sequence in which we replace any
edge <code class="docutils literal notranslate"><span class="pre">(left,</span> <span class="pre">right,</span> <span class="pre">parent,</span> <span class="pre">child)</span></code> in which
<code class="docutils literal notranslate"><span class="pre">node_time[child]</span> <span class="pre">&lt;</span> <span class="pre">time</span> <span class="pre">&lt;</span> <span class="pre">node_time[parent]</span></code> with two edges
<code class="docutils literal notranslate"><span class="pre">(left,</span> <span class="pre">right,</span> <span class="pre">parent,</span> <span class="pre">u)</span></code> and <code class="docutils literal notranslate"><span class="pre">(left,</span> <span class="pre">right,</span> <span class="pre">u,</span> <span class="pre">child)</span></code>,
where <code class="docutils literal notranslate"><span class="pre">u</span></code> is a newly added node for each intersecting edge.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">metadata</span></code>, <code class="docutils literal notranslate"><span class="pre">flags</span></code>, or <code class="docutils literal notranslate"><span class="pre">population</span></code> are specified, newly
added nodes will be assigned these values. Otherwise, default values
will be used. The default metadata is an empty dictionary if a metadata
schema is defined for the node table, and is an empty byte string
otherwise. The default population for the new node is
<a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a>. Newly added have a default <code class="docutils literal notranslate"><span class="pre">flags</span></code> value of 0.</p>
<p>Any metadata associated with a split edge will be copied to the new edge.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method currently does not support migrations
and a error will be raised if the migration table is not
empty. Future versions may take migrations that intersect with the
edge into account when determining the default population
assignments for new nodes.</p>
</div>
<p>Any mutations lying on the edge whose time is &gt;= <code class="docutils literal notranslate"><span class="pre">time</span></code> will have
their node value set to <code class="docutils literal notranslate"><span class="pre">u</span></code>. Note that the time of the mutation is
defined as the time of the child node if the mutations time is
unknown.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The cutoff time.</p></li>
<li><p><strong>flags</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The flags value for newly-inserted nodes. (Default = 0)</p></li>
<li><p><strong>population</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The population value for newly inserted nodes.
Defaults to <code class="docutils literal notranslate"><span class="pre">tskit.NULL</span></code> if not specified.</p></li>
<li><p><strong>metadata</strong>  The metadata for any newly inserted nodes. See
<a class="reference internal" href="#tskit.NodeTable.add_row" title="tskit.NodeTable.add_row"><code class="xref py py-meth docutils literal notranslate"><span class="pre">NodeTable.add_row()</span></code></a> for details on how default metadata
is produced for a given schema (or none).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A copy of this tree sequence with edges split at the specified time.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence">tskit.TreeSequence</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.decapitate">
<span class="sig-name descname"><span class="pre">decapitate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">time</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">population</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.decapitate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.decapitate" title="Link to this definition">#</a></dt>
<dd><p>Delete all edge topology and mutational information at least as old
as the specified time from this tree sequence.</p>
<p>Removes all edges in which the time of the child is &gt;= the specified
time <code class="docutils literal notranslate"><span class="pre">t</span></code>, and breaks edges that intersect with <code class="docutils literal notranslate"><span class="pre">t</span></code>. For each edge
intersecting with <code class="docutils literal notranslate"><span class="pre">t</span></code> we create a new node with time equal to <code class="docutils literal notranslate"><span class="pre">t</span></code>,
and set the parent of the edge to this new node. The node table
is not altered in any other way. Newly added nodes have values
for <code class="docutils literal notranslate"><span class="pre">flags</span></code>, <code class="docutils literal notranslate"><span class="pre">population</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata</span></code> controlled by parameters
to this function in the same way as <a class="reference internal" href="#tskit.TreeSequence.split_edges" title="tskit.TreeSequence.split_edges"><code class="xref py py-meth docutils literal notranslate"><span class="pre">split_edges()</span></code></a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Note that each edge is treated independently, so that even if two
edges that are broken by this operation share the same parent and
child nodes, there will be two different new parent nodes inserted.</p>
</div>
<p>Any mutation whose time is &gt;= <code class="docutils literal notranslate"><span class="pre">t</span></code> will be removed. A mutations time
is its associated <code class="docutils literal notranslate"><span class="pre">time</span></code> value, or the time of its node if the
mutations time was marked as unknown (<a class="reference internal" href="#tskit.UNKNOWN_TIME" title="tskit.UNKNOWN_TIME"><code class="xref py py-data docutils literal notranslate"><span class="pre">UNKNOWN_TIME</span></code></a>).</p>
<p>Migrations are not supported, and a LibraryError will be raised if
called on a tree sequence containing migration information.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>This method is implemented using the <a class="reference internal" href="#tskit.TreeSequence.split_edges" title="tskit.TreeSequence.split_edges"><code class="xref py py-meth docutils literal notranslate"><span class="pre">split_edges()</span></code></a>
and <a class="reference internal" href="#tskit.TableCollection.delete_older" title="tskit.TableCollection.delete_older"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.delete_older()</span></code></a> functions.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The cutoff time.</p></li>
<li><p><strong>flags</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The flags value for newly-inserted nodes. (Default = 0)</p></li>
<li><p><strong>population</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The population value for newly inserted nodes.
Defaults to <code class="docutils literal notranslate"><span class="pre">tskit.NULL</span></code> if not specified.</p></li>
<li><p><strong>metadata</strong>  The metadata for any newly inserted nodes. See
<a class="reference internal" href="#tskit.NodeTable.add_row" title="tskit.NodeTable.add_row"><code class="xref py py-meth docutils literal notranslate"><span class="pre">NodeTable.add_row()</span></code></a> for details on how default metadata
is produced for a given schema (or none).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A copy of this tree sequence with edges split at the specified time.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence">tskit.TreeSequence</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.extend_haplotypes">
<span class="sig-name descname"><span class="pre">extend_haplotypes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">max_iter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.extend_haplotypes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.extend_haplotypes" title="Link to this definition">#</a></dt>
<dd><p>Returns a new tree sequence in which the span covered by ancestral nodes
is extended to regions of the genome according to the following rule:
If an ancestral segment corresponding to node <cite>n</cite> has ancestor <cite>p</cite> and
descendant <cite>c</cite> on some portion of the genome, and on an adjacent segment of
genome <cite>p</cite> is still an ancestor of <cite>c</cite>, then <cite>n</cite> is inserted into the
path from <cite>p</cite> to <cite>c</cite>. For instance, if <cite>p</cite> is the parent of <cite>n</cite> and <cite>n</cite>
is the parent of <cite>c</cite>, then the span of the edges from <cite>p</cite> to <cite>n</cite> and
<cite>n</cite> to <cite>c</cite> are extended, and the span of the edge from <cite>p</cite> to <cite>c</cite> is
reduced. Thus, the ancestral haplotype represented by <cite>n</cite> is extended
to a longer span of the genome. However, any edges whose child node is
a sample are not modified. See
<a class="reference external" href="https://doi.org/10.1093/genetics/iyaf198">Fritze et al. (2025)</a>
for more details.</p>
<p>Since some edges may be removed entirely, this process usually reduces
the number of edges in the tree sequence.</p>
<p>The method works by iterating over the genome to look for paths that can
be extended in this way; the maximum number of such iterations is
controlled by <code class="docutils literal notranslate"><span class="pre">max_iter</span></code>.</p>
<p>The rationale is that we know that <cite>n</cite> carries a portion of the segment
of ancestral genome inherited by <cite>c</cite> from <cite>p</cite>, and so likely carries
the <em>entire</em> inherited segment (since the implication otherwise would
be that distinct recombined segments were passed down separately from
<cite>p</cite> to <cite>c</cite>).</p>
<p>In the example above, if there was a mutation on the node above <cite>c</cite>
older than the time of <cite>n</cite> in the span into which <cite>n</cite> was extended,
then the mutation will now occur above <cite>n</cite>. So, this operation may change
mutations nodes (but will not affect genotypes).  This is only
unambiguous if the mutations time is known, so the method requires
known mutation times.  See <a class="reference internal" href="#tskit.TreeSequence.impute_unknown_mutations_time" title="tskit.TreeSequence.impute_unknown_mutations_time"><code class="xref py py-meth docutils literal notranslate"><span class="pre">impute_unknown_mutations_time()</span></code></a> if
mutation times are not known.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The method will not affect the marginal trees (so, if the original tree
sequence was simplified, then following up with <cite>simplify</cite> will recover
the original tree sequence, possibly with edges in a different order).
It will also not affect the genotype matrix, or any of the tables other
than the edge table or the node column in the mutation table.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>max_iter</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The maximum number of iterations over the tree
sequence. Defaults to 10.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A new tree sequence with unary nodes extended.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence">tskit.TreeSequence</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.subset">
<span class="sig-name descname"><span class="pre">subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nodes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reorder_populations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_unreferenced</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.subset"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.subset" title="Link to this definition">#</a></dt>
<dd><p>Returns a tree sequence containing only information directly
referencing the provided list of nodes to retain.  The result will
retain only the nodes whose IDs are listed in <code class="docutils literal notranslate"><span class="pre">nodes</span></code>, only edges for
which both parent and child are in <code class="docutils literal notranslate"><span class="pre">nodes`</span></code>, only mutations whose
node is in <code class="docutils literal notranslate"><span class="pre">nodes</span></code>, and only individuals that are referred to by one
of the retained nodes.  Note that this does <em>not</em> retain
the ancestry of these nodes - for that, see <a class="reference internal" href="#tskit.TreeSequence.simplify" title="tskit.TreeSequence.simplify"><code class="xref py py-meth docutils literal notranslate"><span class="pre">simplify()</span></code></a>.</p>
<p>This has the side effect that it may change the order of the nodes,
populations, individuals, and migrations in the tree sequence. Nodes
in the new tree sequence will be in the order provided in <code class="docutils literal notranslate"><span class="pre">nodes</span></code>.
Populations will be ordered in ascending order of the lowest ID of
the nodes that refer to them. Individuals will be not only ordered
so that <a class="reference internal" href="#tskit.Individual.parents" title="tskit.Individual.parents"><code class="xref py py-attr docutils literal notranslate"><span class="pre">parents</span></code></a> come before children (see
<a class="reference internal" href="#tskit.TableCollection.sort_individuals" title="tskit.TableCollection.sort_individuals"><code class="xref py py-meth docutils literal notranslate"><span class="pre">sort_individuals()</span></code></a>) but in addition
will be secondarily sorted in ascending order of the lowest ID of
their referring nodes. (However, <code class="docutils literal notranslate"><span class="pre">reorder_populations</span></code> may be set
to <code class="docutils literal notranslate"><span class="pre">False</span></code> to keep the population table unchanged.)</p>
<p>By default, the method removes all individuals and populations not
referenced by any nodes, and all sites not referenced by any mutations.
To retain these unreferenced individuals, populations, and sites, pass
<code class="docutils literal notranslate"><span class="pre">remove_unreferenced=False</span></code>. If this is done, the site table will
remain unchanged, unreferenced individuals will appear at the end of
the individuals table (and in their original order), and unreferenced
populations will appear at the end of the population table (unless
<code class="docutils literal notranslate"><span class="pre">reorder_populations=False</span></code>).</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#tskit.TreeSequence.keep_intervals" title="tskit.TreeSequence.keep_intervals"><code class="xref py py-meth docutils literal notranslate"><span class="pre">keep_intervals()</span></code></a> for subsetting a given portion of the genome;
<a class="reference internal" href="#tskit.TreeSequence.simplify" title="tskit.TreeSequence.simplify"><code class="xref py py-meth docutils literal notranslate"><span class="pre">simplify()</span></code></a> for retaining the ancestry of a subset of nodes.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nodes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  The list of nodes for which to retain information. This
may be a numpy array (or array-like) object (dtype=np.int32).</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to record a provenance entry
in the provenance table for this operation.</p></li>
<li><p><strong>reorder_populations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to reorder populations
(default: True).  If False, the population table will not be altered in
any way.</p></li>
<li><p><strong>remove_unreferenced</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether sites, individuals, and populations
that are not referred to by any retained entries in the tables should
be removed (default: True). See the description for details.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence">tskit.TreeSequence</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.union">
<span class="sig-name descname"><span class="pre">union</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_mapping</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_shared_equality</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_populations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">all_edges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">all_mutations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.union"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.union" title="Link to this definition">#</a></dt>
<dd><p>Returns an expanded tree sequence which contains the node-wise union of
<code class="docutils literal notranslate"><span class="pre">self</span></code> and <code class="docutils literal notranslate"><span class="pre">other</span></code>, obtained by adding the non-shared portions of
<code class="docutils literal notranslate"><span class="pre">other</span></code> onto <code class="docutils literal notranslate"><span class="pre">self</span></code>. The shared portions are specified using a
map that specifies which nodes in <code class="docutils literal notranslate"><span class="pre">other</span></code> are equivalent to those in
<code class="docutils literal notranslate"><span class="pre">self</span></code>: the <code class="docutils literal notranslate"><span class="pre">node_mapping</span></code> argument should be an array of length
equal to the number of nodes in <code class="docutils literal notranslate"><span class="pre">other</span></code> and whose entries are the ID
of the matching node in <code class="docutils literal notranslate"><span class="pre">self</span></code>, or <code class="docutils literal notranslate"><span class="pre">tskit.NULL</span></code> if there is no
matching node. Those nodes in <code class="docutils literal notranslate"><span class="pre">other</span></code> that map to <code class="docutils literal notranslate"><span class="pre">tskit.NULL</span></code> will
be added to <code class="docutils literal notranslate"><span class="pre">self</span></code>, along with:</p>
<ol class="arabic simple">
<li><p>Individuals whose nodes are new to <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p></li>
<li><p>Edges whose parent or child are new to <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p></li>
<li><p>Mutations whose nodes are new to <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p></li>
<li><p>Sites whose positions are not present in the site positions in
<code class="docutils literal notranslate"><span class="pre">self</span></code>, if the site contains a newly added mutation.</p></li>
</ol>
<p>This can be thought of as a node-wise union: for instance, it can not
be used to add new edges between two nodes already in <code class="docutils literal notranslate"><span class="pre">self</span></code> or new
mutations above nodes already in <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p>
<p>By default, with <code class="docutils literal notranslate"><span class="pre">add_populations=True</span></code>, populations of all newly added
nodes are assumed to be new populations, and added to the end of the
population table as well. This is appropriate if all nodes to be added
are from distinct populations not already in <code class="docutils literal notranslate"><span class="pre">self</span></code> and ordering of
populations is not important. On the other hand, if
<code class="docutils literal notranslate"><span class="pre">add_populations=False</span></code> then no new populations are added, so any
populations referred to in <code class="docutils literal notranslate"><span class="pre">other</span></code> must already exist in <code class="docutils literal notranslate"><span class="pre">self</span></code>.
If some new nodes are in populations already in <code class="docutils literal notranslate"><span class="pre">self</span></code> but other new
nodes are in entirely new populations, then you must set up the
population table first, and then union with <code class="docutils literal notranslate"><span class="pre">add_populations=False</span></code>.</p>
<p>This method makes sense if the shared portions of the tree sequences
are equal; the option <code class="docutils literal notranslate"><span class="pre">check_shared_equality</span></code> performs a consistency
check that this is true. If this check is disabled, it is very easy to
produce nonsensical results via subtle inconsistencies.</p>
<p>The behavior above can be changed by <code class="docutils literal notranslate"><span class="pre">all_edges</span></code> and <code class="docutils literal notranslate"><span class="pre">all_mutations</span></code>.
If <code class="docutils literal notranslate"><span class="pre">all_edges</span></code> is True, then all edges in <code class="docutils literal notranslate"><span class="pre">other</span></code> are added to
<code class="docutils literal notranslate"><span class="pre">self</span></code>, instead of only edges adjacent to added nodes. If
<code class="docutils literal notranslate"><span class="pre">all_mutations</span></code> is True, then similarly all mutations in <code class="docutils literal notranslate"><span class="pre">other</span></code>
are added (not just those on added nodes); furthermore, all sites
at positions without a site already present are added to <code class="docutils literal notranslate"><span class="pre">self</span></code>.
The intended use case for these options is a disjoint union,
where for instance the two tree sequences contain information about
disjoint segments of the genome (see <a class="reference internal" href="#tskit.TreeSequence.concatenate" title="tskit.TreeSequence.concatenate"><code class="xref py py-meth docutils literal notranslate"><span class="pre">concatenate()</span></code></a>).
For some such applications it may be necessary to set
<code class="docutils literal notranslate"><span class="pre">check_shared_equality=False</span></code>: for instance, if <code class="docutils literal notranslate"><span class="pre">other</span></code> has
an identical copy of the node table but no edges, then
<code class="docutils literal notranslate"><span class="pre">all_mutations=True,</span> <span class="pre">check_shared_equality=False</span></code> can be used
to add mutations to <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>If an equivalent node is specified in <code class="docutils literal notranslate"><span class="pre">other</span></code>, the
version in <code class="docutils literal notranslate"><span class="pre">self</span></code> is used without checking the node
properties are the same. Similarly, if the same site position
is present in both <code class="docutils literal notranslate"><span class="pre">self</span></code> and <code class="docutils literal notranslate"><span class="pre">other</span></code>, the version in
<code class="docutils literal notranslate"><span class="pre">self</span></code> is used without checking that site properties are
the same. In these cases metadata and e.g. node times or ancestral
states in <code class="docutils literal notranslate"><span class="pre">other</span></code> are simply ignored.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This operation also sorts the resulting tables, so the resulting
tree sequence may not be equal to <code class="docutils literal notranslate"><span class="pre">self</span></code> even if nothing new
was added (although it would differ only in ordering of the tables).</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><em>TreeSequence</em></a>)  Another tree sequence.</p></li>
<li><p><strong>node_mapping</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An array of node IDs that relate nodes in
<code class="docutils literal notranslate"><span class="pre">other</span></code> to nodes in <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p></li>
<li><p><strong>all_edges</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, then all edges in <code class="docutils literal notranslate"><span class="pre">other</span></code> are added
to <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p></li>
<li><p><strong>all_mutations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, then all mutations and sites in
<code class="docutils literal notranslate"><span class="pre">other</span></code> are added to <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p></li>
<li><p><strong>check_shared_equality</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, the shared portions of the
tree sequences will be checked for equality. It does so by
running <a class="reference internal" href="#tskit.TreeSequence.subset" title="tskit.TreeSequence.subset"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.subset()</span></code></a> on both <code class="docutils literal notranslate"><span class="pre">self</span></code> and <code class="docutils literal notranslate"><span class="pre">other</span></code>
for the equivalent nodes specified in <code class="docutils literal notranslate"><span class="pre">node_mapping</span></code>, and then
checking for equality of the subsets.</p></li>
<li><p><strong>add_populations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, nodes new to <code class="docutils literal notranslate"><span class="pre">self</span></code> will be
assigned new population IDs.</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to record a provenance entry
in the provenance table for this operation.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The union of the two tree sequences.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence">tskit.TreeSequence</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>tskit.LibraryError</strong>  If the resulting tree sequence is invalid
(for instance, a node is specified to have two distinct
parents on the same interval)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.draw_svg">
<span class="sig-name descname"><span class="pre">draw_svg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tree_height_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_titles</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation_titles</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">root_svg_attributes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">style</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_root_branch</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">symbol_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_lim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_regions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_ticks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_gridlines</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">omit_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">canvas_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_num_trees</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">preamble</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.draw_svg"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.draw_svg" title="Link to this definition">#</a></dt>
<dd><p>Return an SVG representation of a tree sequence. See the
<a class="reference external" href="https://tskit.dev/tutorials/viz.html#sec-tskit-viz" title="(in Project name not set)"><span class="xref std std-ref">visualization tutorial</span></a> for more details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The path to the file to write the output. If None, do not write
to file.</p></li>
<li><p><strong>size</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.14)"><em>tuple</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>)</em>)  A tuple of (width, height) specifying a target
drawing size in abstract user units (usually interpreted as pixels on
initial display). Components of the drawing will be scaled so that the total
plot including labels etc. normally fits onto a canvas of this size (see
<code class="docutils literal notranslate"><span class="pre">canvas_size</span></code> below). If <code class="docutils literal notranslate"><span class="pre">None</span></code>, chose values such that each tree is
drawn at a size appropriate for a reasonably small set of samples (this will
nevertheless result in a very wide drawing if there are many trees to
display). Default: <code class="docutils literal notranslate"><span class="pre">None</span></code></p></li>
<li><p><strong>x_scale</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  Control how the X axis is drawn. If physical (the default)
the axis scales linearly with physical distance along the sequence,
background shading is used to indicate the position of the trees along the
X axis, and sites (with associated mutations) are marked at the
appropriate physical position on axis line. If treewise, each axis tick
corresponds to a tree boundary, which are positioned evenly along the axis,
so that the X axis is of variable scale, no background scaling is required,
and site positions are not marked on the axis.</p></li>
<li><p><strong>time_scale</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  Control how height values for nodes are computed.
If this is equal to <code class="docutils literal notranslate"><span class="pre">&quot;time&quot;</span></code>, node heights are proportional to their time
values (this is the default). If this is equal to <code class="docutils literal notranslate"><span class="pre">&quot;log_time&quot;</span></code>, node
heights are proportional to their log(time) values. If it is equal to
<code class="docutils literal notranslate"><span class="pre">&quot;rank&quot;</span></code>, node heights are spaced equally according to their ranked times.</p></li>
<li><p><strong>tree_height_scale</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  Deprecated alias for time_scale. (Deprecated in
0.3.6)</p></li>
<li><p><strong>title</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A title string to be included in the SVG output. If <code class="docutils literal notranslate"><span class="pre">None</span></code>
(default) no title is shown, which gives more vertical space for the tree.</p></li>
<li><p><strong>node_labels</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>)</em>)  If specified, show custom labels for the nodes
(specified by ID) that are present in this map; any nodes not present will
not have a label.</p></li>
<li><p><strong>mutation_labels</strong>  If specified, show custom labels for the
mutations (specified by ID) that are present in the map; any mutations
not present will not have a label.</p></li>
<li><p><strong>node_titles</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>)</em>)  If specified, add a <code class="docutils literal notranslate"><span class="pre">&lt;title&gt;</span></code> string to
symbols for each node (specified by ID) present in this map. SVG visualizers
such as web browsers will commonly display this string on mousing over
node symbol.</p></li>
<li><p><strong>mutation_titles</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>)</em>)  If specified, add a <code class="docutils literal notranslate"><span class="pre">&lt;title&gt;</span></code> string to
symbols for each mutation (specified by ID) present in this map. SVG
visualizers such as web browsers will commonly display this string on
mousing over the mutation symbol in the tree and (if show) on the x axis.</p></li>
<li><p><strong>root_svg_attributes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a>)  Additional attributes, such as an id, that will
be embedded in the root <code class="docutils literal notranslate"><span class="pre">&lt;svg&gt;</span></code> tag of the generated drawing.</p></li>
<li><p><strong>style</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A <a class="reference external" href="https://www.w3.org/TR/CSS21/syndata.htm">css string</a>
that will be included in the <code class="docutils literal notranslate"><span class="pre">&lt;style&gt;</span></code> tag of the generated svg.</p></li>
<li><p><strong>order</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The left-to-right ordering of child nodes in each drawn tree.
This can be either: <code class="docutils literal notranslate"><span class="pre">&quot;minlex&quot;</span></code>, which minimises the differences
between adjacent trees (see also the <code class="docutils literal notranslate"><span class="pre">&quot;minlex_postorder&quot;</span></code> traversal
order for the <a class="reference internal" href="#tskit.Tree.nodes" title="tskit.Tree.nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.nodes()</span></code></a> method); or <code class="docutils literal notranslate"><span class="pre">&quot;tree&quot;</span></code> which draws trees
in the left-to-right order defined by the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">quintuply linked tree structure</span></a>.
If not specified or None, this defaults to <code class="docutils literal notranslate"><span class="pre">&quot;minlex&quot;</span></code>.</p></li>
<li><p><strong>force_root_branch</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If <code class="docutils literal notranslate"><span class="pre">True</span></code> plot a branch (edge) above every tree
root in the tree sequence. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) then only plot such
root branches if any root in the tree sequence has a mutation above it.</p></li>
<li><p><strong>symbol_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  Change the default size of the node and mutation
plotting symbols. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) use a standard size.</p></li>
<li><p><strong>x_axis</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Should the plot have an X axis line, showing the positions
of trees along the genome. The scale used is determined by the <code class="docutils literal notranslate"><span class="pre">x_scale</span></code>
parameter. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) plot an X axis.</p></li>
<li><p><strong>x_label</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  Place a label under the plot. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) and
there is an X axis, create and place an appropriate label.</p></li>
<li><p><strong>x_lim</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of size two giving the genomic positions between which
trees should be plotted. If the first is <code class="docutils literal notranslate"><span class="pre">None</span></code>, then plot from the first
non-empty region of the tree sequence. If the second is <code class="docutils literal notranslate"><span class="pre">None</span></code>, then plot
up to the end of the last non-empty region of the tree sequence. The default
value <code class="docutils literal notranslate"><span class="pre">x_lim=None</span></code> is shorthand for the list [<code class="docutils literal notranslate"><span class="pre">None</span></code>, <code class="docutils literal notranslate"><span class="pre">None</span></code>]. If
numerical values are given, then regions outside the interval have all
information discarded: this means that mutations outside the interval will
not be shown. To force display of the entire tree sequence, including empty
flanking regions, specify <code class="docutils literal notranslate"><span class="pre">x_lim=[0,</span> <span class="pre">ts.sequence_length]</span></code>.</p></li>
<li><p><strong>x_regions</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a>)  A dictionary mapping (left, right) tuples to names. This
draws a box, labelled with the name, on the X axis between the left and
right positions, and can be used for annotating genomic regions (e.g.
genes) on the X axis. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) do not plot any regions.</p></li>
<li><p><strong>y_axis</strong> (<em>Union</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>]</em>)  Should the plot have an Y axis line, showing
time. If <code class="docutils literal notranslate"><span class="pre">False</span></code> do not plot a Y axis. If <code class="docutils literal notranslate"><span class="pre">True</span></code>, plot the Y axis on
left hand side of the plot. Can also take the strings <code class="docutils literal notranslate"><span class="pre">&quot;left&quot;</span></code> or
<code class="docutils literal notranslate"><span class="pre">&quot;right&quot;</span></code>, specifying the side of the plot on which to plot the Y axis.
Default: <code class="docutils literal notranslate"><span class="pre">None</span></code>, treated as <code class="docutils literal notranslate"><span class="pre">False</span></code>.</p></li>
<li><p><strong>y_label</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  Place a label to the left of the plot. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default)
and there is a Y axis, create and place an appropriate label.</p></li>
<li><p><strong>y_ticks</strong> (<em>Union</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a><em>]</em>)  A list of Y values at which to plot
tickmarks, or a dictionary mapping Y values to labels (<code class="docutils literal notranslate"><span class="pre">[]</span></code> gives no
tickmarks). If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), plot one tickmark for each unique node
value. Note that if <code class="docutils literal notranslate"><span class="pre">time_scale=&quot;rank&quot;</span></code>, the Y values refer to the
zero-based rank of the plotted nodes, rather than the node time itself.</p></li>
<li><p><strong>y_gridlines</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to plot horizontal lines behind the tree
at each y tickmark.</p></li>
<li><p><strong>omit_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, omit sites and mutations from the drawing.
Default: False</p></li>
<li><p><strong>canvas_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.14)"><em>tuple</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>)</em>)  The (width, height) of the SVG canvas.
This will change the SVG width and height without rescaling graphical
elements, allowing extra room e.g. for unusually long labels. If <code class="docutils literal notranslate"><span class="pre">None</span></code>
take the canvas size to be the same as the target drawing size (see
<code class="docutils literal notranslate"><span class="pre">size</span></code>, above). Default: None</p></li>
<li><p><strong>max_num_trees</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The maximum number of trees to plot. If there are
more trees than this in the tree sequence, the middle trees will be skipped
from the plot and a message XX trees skipped displayed in their place.
If <code class="docutils literal notranslate"><span class="pre">None</span></code>, all the trees will be plotted: this can produce a very wide
plot if there are many trees in the tree sequence. Default: None</p></li>
<li><p><strong>preamble</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  SVG commands to be included at the start of the returned
object, immediately after the opening tag. These can include custom svg
elements such as legends or annotations or even entire <code class="docutils literal notranslate"><span class="pre">&lt;svg&gt;</span></code> elements.
The preamble is not checked for validity, so it is up to the user to
ensure that it is valid SVG. Default: None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An SVG representation of a tree sequence.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.SVGString" title="tskit.SVGString">SVGString</a></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Technically, x_lim[0] specifies a <em>minimum</em> value for the start of the X
axis, and x_lim[1] specifies a <em>maximum</em> value for the end. This is only
relevant if the tree sequence contains empty regions with no edges or
mutations. In this case if x_lim[0] lies strictly within an empty region
(i.e., <code class="docutils literal notranslate"><span class="pre">empty_tree.interval.left</span> <span class="pre">&lt;</span> <span class="pre">x_lim[0]</span> <span class="pre">&lt;</span> <span class="pre">empty_tree.interval.right</span></code>)
then that tree will not be plotted on the left hand side, and the X axis
will start at <code class="docutils literal notranslate"><span class="pre">empty_tree.interval.right</span></code>. Similarly, if x_lim[1] lies
strictly within an empty region then that tree will not be plotted on the
right hand side, and the X axis will end at <code class="docutils literal notranslate"><span class="pre">empty_tree.interval.left</span></code></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.draw_text">
<span class="sig-name descname"><span class="pre">draw_text</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_ascii</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_label_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_label_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.draw_text"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.draw_text" title="Link to this definition">#</a></dt>
<dd><p>Create a text representation of a tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>node_labels</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a>)  If specified, show custom labels for the nodes
that are present in the map. Any nodes not specified in the map will
not have a node label.</p></li>
<li><p><strong>use_ascii</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If <code class="docutils literal notranslate"><span class="pre">False</span></code> (default) then use unicode
<a class="reference external" href="https://en.wikipedia.org/wiki/Box-drawing_character">box drawing characters</a>
to render the tree. If <code class="docutils literal notranslate"><span class="pre">True</span></code>, use plain ascii characters, which look
cruder but are less susceptible to misalignment or font substitution.
Alternatively, if you are having alignment problems with Unicode, you can try
out the solution documented <a class="reference external" href="https://github.com/tskit-dev/tskit/issues/189#issuecomment-499114811">here</a>.</p></li>
<li><p><strong>time_label_format</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A python format string specifying the format (e.g.
number of decimal places or significant figures) used to print the numerical
time values on the time axis. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, this defaults to <code class="docutils literal notranslate"><span class="pre">&quot;{:.2f}&quot;</span></code>.</p></li>
<li><p><strong>position_label_format</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A python format string specifying the format
(e.g. number of decimal places or significant figures) used to print genomic
positions. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, this defaults to <code class="docutils literal notranslate"><span class="pre">&quot;{:.2f}&quot;</span></code>.</p></li>
<li><p><strong>order</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The left-to-right ordering of child nodes in the drawn tree.
This can be either: <code class="docutils literal notranslate"><span class="pre">&quot;minlex&quot;</span></code>, which minimises the differences
between adjacent trees (see also the <code class="docutils literal notranslate"><span class="pre">&quot;minlex_postorder&quot;</span></code> traversal
order for the <a class="reference internal" href="#tskit.Tree.nodes" title="tskit.Tree.nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.nodes()</span></code></a> method); or <code class="docutils literal notranslate"><span class="pre">&quot;tree&quot;</span></code> which draws trees
in the left-to-right order defined by the
<a class="reference internal" href="data-model.html#sec-data-model-tree-structure"><span class="std std-ref">quintuply linked tree structure</span></a>.
If not specified or None, this defaults to <code class="docutils literal notranslate"><span class="pre">&quot;minlex&quot;</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A text representation of a tree sequence.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.general_stat">
<span class="sig-name descname"><span class="pre">general_stat</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">W</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dim</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polarised</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.general_stat"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.general_stat" title="Link to this definition">#</a></dt>
<dd><p>Compute a windowed statistic from weights and a summary function.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.
On each tree, this
propagates the weights <code class="docutils literal notranslate"><span class="pre">W</span></code> up the tree, so that the weight of each
node is the sum of the weights of all samples at or below the node.
Then the summary function <code class="docutils literal notranslate"><span class="pre">f</span></code> is applied to the weights, giving a
summary for each node in each tree. How this is then aggregated depends
on <code class="docutils literal notranslate"><span class="pre">mode</span></code>:</p>
<dl class="simple">
<dt>site</dt><dd><p>Adds together the total summary value across all alleles in each window.</p>
</dd>
<dt>branch</dt><dd><p>Adds together the summary value for each node, multiplied by the
length of the branch above the node and the span of the tree.</p>
</dd>
<dt>node</dt><dd><p>Returns each nodes summary value added across trees and multiplied
by the span of the tree.</p>
</dd>
</dl>
<p>Both the weights and the summary can be multidimensional: if <code class="docutils literal notranslate"><span class="pre">W</span></code> has <code class="docutils literal notranslate"><span class="pre">k</span></code>
columns, and <code class="docutils literal notranslate"><span class="pre">f</span></code> takes a <code class="docutils literal notranslate"><span class="pre">k</span></code>-vector and returns an <code class="docutils literal notranslate"><span class="pre">m</span></code>-vector,
then the output will be <code class="docutils literal notranslate"><span class="pre">m</span></code>-dimensional for each node or window (depending
on mode).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The summary function <code class="docutils literal notranslate"><span class="pre">f</span></code> should return zero when given both 0 and
the total weight (i.e., <code class="docutils literal notranslate"><span class="pre">f(0)</span> <span class="pre">=</span> <span class="pre">0</span></code> and <code class="docutils literal notranslate"><span class="pre">f(np.sum(W,</span> <span class="pre">axis=0))</span> <span class="pre">=</span> <span class="pre">0</span></code>),
unless <code class="docutils literal notranslate"><span class="pre">strict=False</span></code>.  This is necessary for the statistic to be
unaffected by parts of the tree sequence ancestral to none or all
of the samples, respectively.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>W</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a>)  An array of values with one row for each sample and one
column for each weight.</p></li>
<li><p><strong>f</strong>  A function that takes a one-dimensional array of length
equal to the number of columns of <code class="docutils literal notranslate"><span class="pre">W</span></code> and returns a one-dimensional
array.</p></li>
<li><p><strong>output_dim</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The length of <code class="docutils literal notranslate"><span class="pre">f</span></code>s return value.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>polarised</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to leave the ancestral state out of computations:
see <a class="reference internal" href="stats.html#sec-stats"><span class="std std-ref">Statistics</span></a> for more details.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of the statistic to be computed
(defaults to site).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to divide the result by the span of the
window (defaults to True).</p></li>
<li><p><strong>strict</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to check that f(0) and f(total weight) are zero.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.sample_count_stat">
<span class="sig-name descname"><span class="pre">sample_count_stat</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dim</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polarised</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.sample_count_stat"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.sample_count_stat" title="Link to this definition">#</a></dt>
<dd><p>Compute a windowed statistic from sample counts and a summary function.
This is a wrapper around <a class="reference internal" href="#tskit.TreeSequence.general_stat" title="tskit.TreeSequence.general_stat"><code class="xref py py-meth docutils literal notranslate"><span class="pre">general_stat()</span></code></a> for the common case in
which the weights are all either 1 or 0, i.e., functions of the joint
allele frequency spectrum.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-sample-sets"><span class="std std-ref">sample sets</span></a>,
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.
If <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> is a list of <code class="docutils literal notranslate"><span class="pre">k</span></code> sets of samples, then
<code class="docutils literal notranslate"><span class="pre">f</span></code> should be a function that takes an argument of length <code class="docutils literal notranslate"><span class="pre">k</span></code> and
returns a one-dimensional array. The <code class="docutils literal notranslate"><span class="pre">j</span></code>-th element of the argument
to <code class="docutils literal notranslate"><span class="pre">f</span></code> will be the number of samples in <code class="docutils literal notranslate"><span class="pre">sample_sets[j]</span></code> that lie
below the node that <code class="docutils literal notranslate"><span class="pre">f</span></code> is being evaluated for. See
<a class="reference internal" href="#tskit.TreeSequence.general_stat" title="tskit.TreeSequence.general_stat"><code class="xref py py-meth docutils literal notranslate"><span class="pre">general_stat()</span></code></a>  for more details.</p>
<p>Here is a contrived example: suppose that <code class="docutils literal notranslate"><span class="pre">A</span></code> and <code class="docutils literal notranslate"><span class="pre">B</span></code> are two sets
of samples with <code class="docutils literal notranslate"><span class="pre">nA</span></code> and <code class="docutils literal notranslate"><span class="pre">nB</span></code> elements, respectively. Passing these
as sample sets will give <code class="docutils literal notranslate"><span class="pre">f</span></code> an argument of length two, giving the number
of samples in <code class="docutils literal notranslate"><span class="pre">A</span></code> and <code class="docutils literal notranslate"><span class="pre">B</span></code> below the node in question. So, if we define</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">f</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
    <span class="n">pA</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">/</span> <span class="n">nA</span>
    <span class="n">pB</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">/</span> <span class="n">nB</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">pA</span> <span class="o">*</span> <span class="n">pB</span><span class="p">])</span>
</pre></div>
</div>
<p>then if all sites are biallelic,</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span><span class="o">.</span><span class="n">sample_count_stat</span><span class="p">([</span><span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">],</span> <span class="n">f</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">windows</span><span class="o">=</span><span class="s2">&quot;sites&quot;</span><span class="p">,</span> <span class="n">polarised</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;site&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>would compute, for each site, the product of the derived allele
frequencies in the two sample sets, in a (num sites, 1) array.  If
instead <code class="docutils literal notranslate"><span class="pre">f</span></code> returns <code class="docutils literal notranslate"><span class="pre">np.array([pA,</span> <span class="pre">pB,</span> <span class="pre">pA</span> <span class="pre">*</span> <span class="pre">pB])</span></code>, then the
output would be a (num sites, 3) array, with the first two columns
giving the allele frequencies in <code class="docutils literal notranslate"><span class="pre">A</span></code> and <code class="docutils literal notranslate"><span class="pre">B</span></code>, respectively.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The summary function <code class="docutils literal notranslate"><span class="pre">f</span></code> should return zero when given both 0 and
the sample size (i.e., <code class="docutils literal notranslate"><span class="pre">f(0)</span> <span class="pre">=</span> <span class="pre">0</span></code> and
<code class="docutils literal notranslate"><span class="pre">f(np.array([len(x)</span> <span class="pre">for</span> <span class="pre">x</span> <span class="pre">in</span> <span class="pre">sample_sets]))</span> <span class="pre">=</span> <span class="pre">0</span></code>).  This is
necessary for the statistic to be unaffected by parts of the tree
sequence ancestral to none or all of the samples, respectively.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>f</strong>  A function that takes a one-dimensional array of length
equal to the number of sample sets and returns a one-dimensional array.</p></li>
<li><p><strong>output_dim</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The length of <code class="docutils literal notranslate"><span class="pre">f</span></code>s return value.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>polarised</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to leave the ancestral state out of computations:
see <a class="reference internal" href="stats.html#sec-stats"><span class="std std-ref">Statistics</span></a> for more details.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of the statistic to be computed
(defaults to site).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to divide the result by the span of the
window (defaults to True).</p></li>
<li><p><strong>strict</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to check that f(0) and f(total weight) are zero.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.diversity">
<span class="sig-name descname"><span class="pre">diversity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'site'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.diversity"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.diversity" title="Link to this definition">#</a></dt>
<dd><p>Computes mean genetic diversity (also known as pi) in each of the
sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.  The statistic is also known as
sample heterozygosity; a common citation for the definition is
<a class="reference external" href="https://doi.org/10.1073/pnas.76.10.5269">Nei and Li (1979)</a>
(equation 22), so it is sometimes called called Neis pi
(but also sometimes Tajimas pi).</p>
<p>Please see the <a class="reference internal" href="stats.html#sec-stats-sample-sets-one-way"><span class="std std-ref">one-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> argument is interpreted
and how it interacts with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.</p>
<p>Note that this quantity can also be computed by the
<a class="reference internal" href="#tskit.TreeSequence.divergence" title="tskit.TreeSequence.divergence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">divergence</span></code></a> method.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>:</p>
<dl class="simple">
<dt>site</dt><dd><p>Mean pairwise genetic diversity: the average over all n choose 2 pairs of
sample nodes, of the density of sites at
which the two carry different alleles, per unit of chromosome length.</p>
</dd>
<dt>branch</dt><dd><p>Mean distance in the tree: the average across over all n choose 2 pairs of
sample nodes and locations in the window, of the mean distance in
the tree between the two samples (in units of time).</p>
</dd>
<dt>node</dt><dd><p>For each node, the proportion of genome on which the node is an ancestor to
only one of a pair of sample nodes from the sample set, averaged
over over all n choose 2 pairs of sample nodes.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of lists of Node IDs, specifying the
groups of nodes for which the statistic is computed. If any of the
sample sets contain only a single node, the returned diversity will be
NaN. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), average over all n choose 2 pairs of distinct
sample nodes in the tree sequence.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of the statistic to be computed
(defaults to site).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A numpy array whose length is equal to the number of sample sets.
If there is one sample set and windows=None, a numpy scalar is returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.divergence">
<span class="sig-name descname"><span class="pre">divergence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indexes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'site'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.divergence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.divergence" title="Link to this definition">#</a></dt>
<dd><p>Computes mean genetic divergence between (and within) pairs of
sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
This is the average number of differences, usually referred to as dxy;
a common citation for this definition is Nei and Li (1979), who called it
<span class="math notranslate nohighlight">\(\pi_{XY}\)</span>. Note that the mean pairwise nucleotide diversity of a
sample set to itself (computed by passing an index of the form (j,j))
is its <a class="reference internal" href="#tskit.TreeSequence.diversity" title="tskit.TreeSequence.diversity"><code class="xref py py-meth docutils literal notranslate"><span class="pre">diversity</span></code></a> (see the note below).</p>
<p>Operates on <code class="docutils literal notranslate"><span class="pre">k</span> <span class="pre">=</span> <span class="pre">2</span></code> sample sets at a time; please see the
<a class="reference internal" href="stats.html#sec-stats-sample-sets-multi-way"><span class="std std-ref">multi-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> and <code class="docutils literal notranslate"><span class="pre">indexes</span></code> arguments are
interpreted and how they interact with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>To avoid unexpected results, sample sets should be nonoverlapping,
since comparisons of individuals to themselves are not removed when computing
divergence between distinct sample sets. (However, specifying an index
<code class="docutils literal notranslate"><span class="pre">(j,</span> <span class="pre">j)</span></code> computes the <a class="reference internal" href="#tskit.TreeSequence.diversity" title="tskit.TreeSequence.diversity"><code class="xref py py-meth docutils literal notranslate"><span class="pre">diversity</span></code></a>
of <code class="docutils literal notranslate"><span class="pre">sample_set[j]</span></code>, which removes self comparisons to provide
an unbiased estimate.)</p>
</div>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>:</p>
<dl class="simple">
<dt>site</dt><dd><p>Mean pairwise genetic divergence: the average across every possible pair of
chromosomes (one from each sample set), of the density of sites at which
the two carry different alleles, per unit of chromosome length.</p>
</dd>
<dt>branch</dt><dd><p>Mean distance in the tree: the average across every possible pair of
chromosomes (one from each sample set) and locations in the window, of
the mean distance in the tree between the two samples (in units of time).</p>
</dd>
<dt>node</dt><dd><p>For each node, the proportion of genome on which the node is an ancestor to
only one of a pair of chromosomes from the sample set, averaged
over all possible pairs.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of 2-tuples, or None.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of the statistic to be computed
(defaults to site).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).
If there is one pair of sample sets and windows=None, a numpy scalar is
returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.divergence_matrix">
<span class="sig-name descname"><span class="pre">divergence_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.divergence_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.divergence_matrix" title="Link to this definition">#</a></dt>
<dd><p>Finds the matrix of pairwise <a class="reference internal" href="#tskit.TreeSequence.divergence" title="tskit.TreeSequence.divergence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">divergence()</span></code></a> values between groups
of sample nodes. Returns a numpy array indexed by (window,
sample_set, sample_set): the [k,i,j]th value of the result gives the
mean divergence between pairs of samples from the i-th and j-th
sample sets in the k-th window. As for <a class="reference internal" href="#tskit.TreeSequence.divergence" title="tskit.TreeSequence.divergence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">divergence()</span></code></a>,
diagonal entries are corrected so that the
value gives the mean divergence for <em>distinct</em> samples,
and so diagonal entries are given by the <a class="reference internal" href="#tskit.TreeSequence.diversity" title="tskit.TreeSequence.diversity"><code class="xref py py-meth docutils literal notranslate"><span class="pre">diversity()</span></code></a> of that
sample set.  For this reason, if an element of <cite>sample_sets</cite> has only
one element, the corresponding <a class="reference internal" href="#tskit.TreeSequence.diversity" title="tskit.TreeSequence.diversity"><code class="xref py py-meth docutils literal notranslate"><span class="pre">diversity()</span></code></a> will be NaN.
However, this method will place a value of 0 in the diagonal instead of NaN
in such cases; otherwise, this is equivalent to computing values with
<cite>meth</cite>:.divergence`.
However, this is (usually) more efficient than computing many
pairwise values using the <cite>indexes</cite> argument to <a class="reference internal" href="#tskit.TreeSequence.divergence" title="tskit.TreeSequence.divergence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">divergence()</span></code></a>,
so see <a class="reference internal" href="#tskit.TreeSequence.divergence" title="tskit.TreeSequence.divergence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">divergence()</span></code></a> for a description of what exactly is computed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of sets of IDs of samples.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  The breakpoints of the windows (including start
and end, so has one more entry than number of windows).</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of the statistic to be computed
(defaults to site; the other option is branch).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An array indexed by (window, sample_set, sample_set), or if windows is
<cite>None</cite>, an array indexed by (sample_set, sample_set).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.genetic_relatedness">
<span class="sig-name descname"><span class="pre">genetic_relatedness</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indexes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'site'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polarised</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proportion</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">centre</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.genetic_relatedness"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.genetic_relatedness" title="Link to this definition">#</a></dt>
<dd><p>Computes genetic relatedness between (and within) pairs of
sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
Operates on <code class="docutils literal notranslate"><span class="pre">k</span> <span class="pre">=</span> <span class="pre">2</span></code> sample sets at a time; please see the
<a class="reference internal" href="stats.html#sec-stats-sample-sets-multi-way"><span class="std std-ref">multi-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> and <code class="docutils literal notranslate"><span class="pre">indexes</span></code> arguments are
interpreted and how they interact with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
<a class="reference internal" href="stats.html#sec-stats-polarisation"><span class="std std-ref">polarised</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>:</p>
<dl class="simple">
<dt>site</dt><dd><p>Frequency of pairwise allelic matches in the window between two
sample sets relative to the rest of the sample sets. To be precise,
let <cite>m(u,v)</cite> denote the total number of alleles shared between
nodes <cite>u</cite> and <cite>v</cite>, and let <cite>m(I,J)</cite> be the average of <cite>m(u,v)</cite> over
all nodes <cite>u</cite> in sample set <cite>I</cite> and <cite>v</cite> in sample set <cite>J</cite>. Let <cite>S</cite>
and <cite>T</cite> be independently chosen sample sets. Then, for sample sets
<cite>I</cite> and <cite>J</cite>, this computes <cite>E[m(I,J) - m(I,S) - m(J,T) + m(S,T)]</cite>
if centre=True (the default), or <cite>E[m(I,J)]</cite> if centre=False.
This can also be seen as the covariance of a quantitative trait
determined by additive contributions from the genomes in each
sample set. Let each derived allele be associated with an effect
drawn from a <cite>N(0,1)</cite> distribution, and let the trait value of a
sample be the sum of its allele effects. Then, this computes
the covariance between the average trait values of two sample sets.
For example, to compute covariance between the traits of diploid
individuals, each sample set would be the pair of genomes of each
individual, with the trait being the average of the two genomes.
If <code class="docutils literal notranslate"><span class="pre">proportion=True</span></code>, this then corresponds to <span class="math notranslate nohighlight">\(K_{c0}\)</span> in
<a class="reference external" href="https://www.nature.com/articles/nrg3821">Speed &amp; Balding (2014)</a>,
multiplied by four (see below).</p>
</dd>
<dt>branch</dt><dd><p>Average area of branches in the window ancestral to pairs of samples
in two sample sets relative to the rest of the sample sets. To be
precise, let <cite>B(u,v)</cite> denote the total area of all branches
ancestral to nodes <cite>u</cite> and <cite>v</cite>, and let <cite>B(I,J)</cite> be the average of
<cite>B(u,v)</cite> over all nodes <cite>u</cite> in sample set <cite>I</cite> and <cite>v</cite> in sample set
<cite>J</cite>. Let <cite>S</cite> and <cite>T</cite> be two independently chosen sample sets. Then
for sample sets <cite>I</cite> and <cite>J</cite>, this computes
<cite>E[B(I,J) - B(I,S) - B(J,T) + B(S,T)]</cite> if centre=True (the default),
or <cite>E[B(I,J)]</cite> if centre=False.</p>
</dd>
<dt>node</dt><dd><p>For each node, the proportion of the window over which pairs of
samples in two sample sets are descendants, relative to the rest of
the sample sets. To be precise, for each node <cite>n</cite>, let <cite>N(u,v)</cite>
denote the proportion of the window over which samples <cite>u</cite> and <cite>v</cite>
are descendants of <cite>n</cite>, and let and let <cite>N(I,J)</cite> be the average of
<cite>N(u,v)</cite> over all nodes <cite>u</cite> in sample set <cite>I</cite> and <cite>v</cite> in sample set
<cite>J</cite>. Let <cite>S</cite> and <cite>T</cite> be two independently chosen sample sets. Then
for sample sets <cite>I</cite> and <cite>J</cite>, this computes
<cite>E[N(I,J) - N(I,S) - N(J,T) + N(S,T)]</cite> if centre=True (the default),
or <cite>E[N(I,J)]</cite> if centre=False.</p>
</dd>
</dl>
<p><em>Note:</em> The default for this statistic - unlike most other statistics - is
<code class="docutils literal notranslate"><span class="pre">polarised=True</span></code>. Using the default value <code class="docutils literal notranslate"><span class="pre">centre=True</span></code>, setting
<code class="docutils literal notranslate"><span class="pre">polarised=False</span></code> will only multiply the result by a factor of two
for branch-mode, or site-mode if all sites are biallelic. (With
multiallelic sites the difference is more complicated.) The uncentred
and unpolarised value is probably not what you are looking for: for
instance, the unpolarised, uncentred site statistic between two samples
counts the number of alleles inherited by both <em>and</em> the number of
alleles inherited by neither of the two samples.</p>
<p><em>Note:</em> Some authors
(see <a class="reference external" href="https://www.nature.com/articles/nrg3821">Speed &amp; Balding (2014)</a>)
compute relatedness between <cite>I</cite> and <cite>J</cite> as the total number of all pairwise
allelic matches between <cite>I</cite> and <cite>J</cite>, rather than the frequency,
which would define <cite>m(I,J)</cite> as the sum of <cite>m(u,v)</cite> rather than the average
in the definition of site relatedness above. If every sample set is the
samples of a <span class="math notranslate nohighlight">\(k\)</span>-ploid individual, this would simply multiply the
result by <span class="math notranslate nohighlight">\(k^2\)</span>. However, this definition would make the result not
useful as a summary statistic of typical relatedness for larger sample
sets.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of 2-tuples, or None.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of the statistic to be computed
(defaults to site).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to divide the result by the span of the
window (defaults to True). Has no effect if <code class="docutils literal notranslate"><span class="pre">proportion</span></code> is True.</p></li>
<li><p><strong>proportion</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Defaults to True.  Whether to divide the result by
<a class="reference internal" href="#tskit.TreeSequence.segregating_sites" title="tskit.TreeSequence.segregating_sites"><code class="xref py py-meth docutils literal notranslate"><span class="pre">segregating_sites()</span></code></a>, called with the same <code class="docutils literal notranslate"><span class="pre">windows</span></code>,
<code class="docutils literal notranslate"><span class="pre">mode</span></code>, and <code class="docutils literal notranslate"><span class="pre">span_normalise</span></code>. Note that this counts sites
that are segregating between <em>any</em> of the samples of <em>any</em> of the
sample sets (rather than segregating between all of the samples of
the tree sequence).</p></li>
<li><p><strong>polarised</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to leave the ancestral state out of computations:
see <a class="reference internal" href="stats.html#sec-stats"><span class="std std-ref">Statistics</span></a> for more details. Defaults to True.</p></li>
<li><p><strong>centre</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Defaults to True. Whether to centre the result, as
described above (the usual definition is centred).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).
If there is one pair of sample sets and windows=None, a numpy scalar is
returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.genetic_relatedness_matrix">
<span class="sig-name descname"><span class="pre">genetic_relatedness_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.genetic_relatedness_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.genetic_relatedness_matrix" title="Link to this definition">#</a></dt>
<dd><p>Computes the full matrix of pairwise genetic relatedness values
between (and within) pairs of sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
Returns a numpy array indexed by (window, sample_set, sample_set):
the [k,i,j]th value of the result gives the
genetic relatedness between pairs of samples from the i-th and j-th
sample sets in the k-th window.
This is (usually) more efficient than computing many pairwise
values using the <cite>indexes</cite> argument to <a class="reference internal" href="#tskit.TreeSequence.genetic_relatedness" title="tskit.TreeSequence.genetic_relatedness"><code class="xref py py-meth docutils literal notranslate"><span class="pre">genetic_relatedness()</span></code></a>.
Specifically, this computes <a class="reference internal" href="#tskit.TreeSequence.genetic_relatedness" title="tskit.TreeSequence.genetic_relatedness"><code class="xref py py-meth docutils literal notranslate"><span class="pre">genetic_relatedness()</span></code></a> with
<code class="docutils literal notranslate"><span class="pre">centre=True</span></code> and <code class="docutils literal notranslate"><span class="pre">proportion=False</span></code> (with caveats, see below).</p>
<p><em>Warning:</em> in some cases, this does not compute exactly the same thing as
<a class="reference internal" href="#tskit.TreeSequence.genetic_relatedness" title="tskit.TreeSequence.genetic_relatedness"><code class="xref py py-meth docutils literal notranslate"><span class="pre">genetic_relatedness()</span></code></a>: see below for more details.</p>
<p>If <cite>mode=branch</cite>, then the value obtained is the same as that from
<a class="reference internal" href="#tskit.TreeSequence.genetic_relatedness" title="tskit.TreeSequence.genetic_relatedness"><code class="xref py py-meth docutils literal notranslate"><span class="pre">genetic_relatedness()</span></code></a>, using the options <cite>centre=True</cite> and
<cite>proportion=False</cite>. The same is true if <cite>mode=site</cite> and all sites have
at most one mutation.</p>
<p>However, if some sites have more than one mutation, the value may differ
from that given by <a class="reference internal" href="#tskit.TreeSequence.genetic_relatedness" title="tskit.TreeSequence.genetic_relatedness"><code class="xref py py-meth docutils literal notranslate"><span class="pre">genetic_relatedness()</span></code></a>:, although if the proportion
of such sites is small, the difference will be small.
The reason is that this function (for efficiency) computes relatedness
using <a class="reference internal" href="#tskit.TreeSequence.divergence_matrix" title="tskit.TreeSequence.divergence_matrix"><code class="xref py py-meth docutils literal notranslate"><span class="pre">divergence_matrix()</span></code></a> and the following relationship.
Relatedness measures the number of <em>shared</em> alleles (or branches),
while divergence measures the number of <em>non-shared</em> alleles (or branches).
Let <span class="math notranslate nohighlight">\(T_i\)</span> be the total distance from sample <span class="math notranslate nohighlight">\(i\)</span> up to the root;
then if <span class="math notranslate nohighlight">\(D_{ij}\)</span> is the branch-mode divergence between <span class="math notranslate nohighlight">\(i\)</span> and
<span class="math notranslate nohighlight">\(j\)</span> and <span class="math notranslate nohighlight">\(R_{ij}\)</span> is the branch-mode relatedness between <span class="math notranslate nohighlight">\(i\)</span>
and <span class="math notranslate nohighlight">\(j\)</span>, then <span class="math notranslate nohighlight">\(T_i + T_j = D_{ij} + 2 R_{ij}.\)</span>
So, for any samples <span class="math notranslate nohighlight">\(I\)</span>, <span class="math notranslate nohighlight">\(J\)</span>, <span class="math notranslate nohighlight">\(S\)</span>, <span class="math notranslate nohighlight">\(T\)</span>
(that may now be random choices),
<span class="math notranslate nohighlight">\(R_{IJ}-R_{IS}-R_{JT}+R_{ST} = (D_{IJ}-D_{IS}-D_{JT}+D_{ST})/ (-2)\)</span>.
This is exactly what we want for (centered) relatedness.
However, this relationship does not necessarily hold for <cite>mode=site</cite>:
it does hold if we can treat number of differing alleles as distances
on the tree, but this is not necessarily the case in the presence of
multiple mutations.</p>
<p>Another note regarding the above relationship between <span class="math notranslate nohighlight">\(R\)</span> and <span class="math notranslate nohighlight">\(D\)</span>
is that <a class="reference internal" href="#tskit.TreeSequence.divergence" title="tskit.TreeSequence.divergence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">divergence()</span></code></a> of a sample set to itself does not include
the self comparisons (so as to provide an unbiased estimator of a
population quantity), while the usual definition of genetic relatedness
<em>does</em> include such comparisons (to provide, for instance, an appropriate
value for prospective results beginning with only a given set of
individuals). So, diagonal entries in the relatedness matrix returned here
are obtained from <a class="reference internal" href="#tskit.TreeSequence.divergence_matrix" title="tskit.TreeSequence.divergence_matrix"><code class="xref py py-meth docutils literal notranslate"><span class="pre">divergence_matrix()</span></code></a> after first correcting
diagonals to include these self comparisons.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of the statistic to be computed
(defaults to site).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An array indexed by (window, sample_set, sample_set), or if windows is
<cite>None</cite>, an array indexed by (sample_set, sample_set).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.genetic_relatedness_weighted">
<span class="sig-name descname"><span class="pre">genetic_relatedness_weighted</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">W</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indexes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'site'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polarised</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">centre</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.genetic_relatedness_weighted"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.genetic_relatedness_weighted" title="Link to this definition">#</a></dt>
<dd><p>Computes weighted genetic relatedness. If the <span class="math notranslate nohighlight">\(k\)</span> th pair of indices
is (i, j) then the <span class="math notranslate nohighlight">\(k\)</span> th column of output will be
<span class="math notranslate nohighlight">\(\sum_{a,b} W_{ai} W_{bj} C_{ab}\)</span>,
where <span class="math notranslate nohighlight">\(W\)</span> is the matrix of weights, and <span class="math notranslate nohighlight">\(C_{ab}\)</span> is the
<a class="reference internal" href="#tskit.TreeSequence.genetic_relatedness" title="tskit.TreeSequence.genetic_relatedness"><code class="xref py py-meth docutils literal notranslate"><span class="pre">genetic_relatedness</span></code></a> between sample
a and sample b, summing over all pairs of samples in the tree sequence.</p>
<p><em>Note:</em> the genetic relatedness matrix <span class="math notranslate nohighlight">\(C\)</span> here is as returned by
<a class="reference internal" href="#tskit.TreeSequence.genetic_relatedness" title="tskit.TreeSequence.genetic_relatedness"><code class="xref py py-meth docutils literal notranslate"><span class="pre">genetic_relatedness()</span></code></a>, rather than by <a class="reference internal" href="#tskit.TreeSequence.genetic_relatedness_matrix" title="tskit.TreeSequence.genetic_relatedness_matrix"><code class="xref py py-meth docutils literal notranslate"><span class="pre">genetic_relatedness_matrix()</span></code></a>
(see the latters documentation for the difference).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>W</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a>)  An array of values with one row for each sample node and
one column for each set of weights.</p></li>
<li><p><strong>indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of 2-tuples, or None (default). Note that if
indexes = None, then W must have exactly two columns and this is equivalent
to indexes = [(0,1)].</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of the statistic to be computed
(defaults to site).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to divide the result by the span of the
window (defaults to True).</p></li>
<li><p><strong>polarised</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to leave the ancestral state out of computations:
see <a class="reference internal" href="stats.html#sec-stats"><span class="std std-ref">Statistics</span></a> for more details. Defaults to True.</p></li>
<li><p><strong>centre</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Defaults to True. Whether to centre the result, as
described above (the usual definition is centred).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.genetic_relatedness_vector">
<span class="sig-name descname"><span class="pre">genetic_relatedness_vector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">W</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'site'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">centre</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.genetic_relatedness_vector"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.genetic_relatedness_vector" title="Link to this definition">#</a></dt>
<dd><p>Computes the product of the genetic relatedness matrix and a vector of weights
(one per sample). The output is a (num windows) x (num samples) x (num weights)
array whose <span class="math notranslate nohighlight">\((w,i,j)\)</span>-th element is <span class="math notranslate nohighlight">\(\sum_{b} W_{bj} C_{ib}\)</span>,
where <span class="math notranslate nohighlight">\(W\)</span> is the matrix of weights, and <span class="math notranslate nohighlight">\(C_{ab}\)</span> is the
<a class="reference internal" href="#tskit.TreeSequence.genetic_relatedness" title="tskit.TreeSequence.genetic_relatedness"><code class="xref py py-meth docutils literal notranslate"><span class="pre">genetic_relatedness</span></code></a> between sample
<cite>a</cite> and sample <cite>b</cite> in window <cite>w</cite>, and the sum is over all samples in the tree
sequence.  Like other statistics, if windows is None, the first dimension in
the output is dropped.</p>
<p>The relatedness used here corresponds to <cite>polarised=True</cite>; no unpolarised option
is available for this method.</p>
<p>Optionally, you may provide a list of focal nodes that modifies the behavior
as follows. If <cite>nodes</cite> is a list of <cite>n</cite> node IDs (that do not need to be
samples), then the output will have dimension (num windows) x n x (num weights),
and the matrix <span class="math notranslate nohighlight">\(C\)</span> used in the definition above is the rectangular matrix
with <span class="math notranslate nohighlight">\(C_{ij}\)</span> the relatedness between <cite>nodes[i]</cite> and <cite>samples[j]</cite>. This
can only be used with <cite>centre=False</cite>; if relatedness between uncentred nodes
and centred samples is desired, then simply subtract column means from <cite>W</cite> first.
The default is <cite>nodes=None</cite>, which is equivalent to setting <cite>nodes</cite> equal to
<cite>ts.samples()</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>W</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a>)  An array of values with one row for each sample node and
one column for each set of weights.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of the statistic to be computed
(defaults to site).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to divide the result by the span of the
window (defaults to True).</p></li>
<li><p><strong>centre</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to use the <em>centred</em> relatedness matrix or not:
see <a class="reference internal" href="#tskit.TreeSequence.genetic_relatedness" title="tskit.TreeSequence.genetic_relatedness"><code class="xref py py-meth docutils literal notranslate"><span class="pre">genetic_relatedness</span></code></a>.</p></li>
<li><p><strong>nodes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  Optionally, a list of focal nodes as described above
(default: None).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num samples, num weights),
or (num samples, num weights) if windows is None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.pca">
<span class="sig-name descname"><span class="pre">pca</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_components</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individuals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'branch'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">centre</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_oversamples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">range_sketch</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.pca"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.pca" title="Link to this definition">#</a></dt>
<dd><p>Performs principal component analysis (PCA) for a given set of samples or
individuals (default: all samples). The principal components are the
eigenvectors of the genetic relatedness matrix, which are obtained by a
randomized singular value decomposition (rSVD) algorithm.</p>
<p>Concretely, take <span class="math notranslate nohighlight">\(M\)</span> as the matrix of non-span-normalised
genetic relatedness values, for instance obtained by
setting <span class="math notranslate nohighlight">\(M_{ij}\)</span> to be the <a class="reference internal" href="#tskit.TreeSequence.genetic_relatedness" title="tskit.TreeSequence.genetic_relatedness"><code class="xref py py-meth docutils literal notranslate"><span class="pre">genetic_relatedness()</span></code></a>
between sample <span class="math notranslate nohighlight">\(i\)</span> and sample <span class="math notranslate nohighlight">\(j\)</span> with the specified <code class="docutils literal notranslate"><span class="pre">mode</span></code>,
<code class="docutils literal notranslate"><span class="pre">proportion=False</span></code> and <code class="docutils literal notranslate"><span class="pre">span_normalise=False</span></code>. Then by default this
returns the top <code class="docutils literal notranslate"><span class="pre">num_components</span></code> eigenvectors of <span class="math notranslate nohighlight">\(M\)</span>, so that
<code class="docutils literal notranslate"><span class="pre">output.factors[i,k]</span></code> is the position of sample <cite>i</cite> on the <cite>k</cite> th PC.
If <code class="docutils literal notranslate"><span class="pre">samples</span></code> or <code class="docutils literal notranslate"><span class="pre">individuals</span></code> are provided, then this does the same
thing, except with <span class="math notranslate nohighlight">\(M_{ij}\)</span> either the relatedness between
<code class="docutils literal notranslate"><span class="pre">samples[i]</span></code> and <code class="docutils literal notranslate"><span class="pre">samples[j]</span></code> or the average relatedness between the
nodes of <code class="docutils literal notranslate"><span class="pre">individuals[i]</span></code> and <code class="docutils literal notranslate"><span class="pre">individuals[j]</span></code>, respectively.
Factors are normalized to have norm 1, i.e.,
<code class="docutils literal notranslate"><span class="pre">output.factors[:,k]</span> <span class="pre">**</span> <span class="pre">2).sum()</span> <span class="pre">==</span> <span class="pre">1)</span></code> for any <code class="docutils literal notranslate"><span class="pre">k</span></code>.</p>
<p>The parameters <code class="docutils literal notranslate"><span class="pre">centre</span></code> and <code class="docutils literal notranslate"><span class="pre">mode</span></code> are passed to
<a class="reference internal" href="#tskit.TreeSequence.genetic_relatedness" title="tskit.TreeSequence.genetic_relatedness"><code class="xref py py-meth docutils literal notranslate"><span class="pre">genetic_relatedness()</span></code></a>: the default <code class="docutils literal notranslate"><span class="pre">centre=True</span></code> results
in factors whose elements sum to zero; <code class="docutils literal notranslate"><span class="pre">mode</span></code> currently only supports the
<code class="docutils literal notranslate"><span class="pre">&quot;branch&quot;</span></code> setting. If <code class="docutils literal notranslate"><span class="pre">windows</span></code> are provided then PCA is carried out
separately in each genomic window. If <code class="docutils literal notranslate"><span class="pre">time_windows</span></code> is provided, then genetic
relatedness is measured using only ancestral material within the given time
window (see <a class="reference internal" href="#tskit.TreeSequence.decapitate" title="tskit.TreeSequence.decapitate"><code class="xref py py-meth docutils literal notranslate"><span class="pre">decapitate</span></code></a> for how this is
defined).</p>
<p>So that the method scales to large tree sequences, the underlying method
relies on a randomized SVD algorithm, using
<a class="reference internal" href="#tskit.TreeSequence.genetic_relatedness_vector" title="tskit.TreeSequence.genetic_relatedness_vector"><code class="xref py py-meth docutils literal notranslate"><span class="pre">genetic_relatedness_vector</span></code></a>).
Larger values of <code class="docutils literal notranslate"><span class="pre">num_iterations</span></code> and
<code class="docutils literal notranslate"><span class="pre">num_oversamples</span></code> should produce better approximations to the true eigenvalues,
at the cost of greater compute times and/or memory usage. The method relies on
constructing <code class="docutils literal notranslate"><span class="pre">range_sketch</span></code>, a low-dimensional approximation to the range
of <span class="math notranslate nohighlight">\(M\)</span>, so that the result of a previous call to <code class="docutils literal notranslate"><span class="pre">pca()</span></code> may be passed
in.</p>
<p>To check for convergence, compare
<code class="docutils literal notranslate"><span class="pre">pc1</span> <span class="pre">=</span> <span class="pre">ts.pca()</span></code> and <code class="docutils literal notranslate"><span class="pre">pc2</span> <span class="pre">=</span> <span class="pre">ts.pca(range_sketch=pc1.range_sketch)</span></code>; the
difference between <code class="docutils literal notranslate"><span class="pre">pc1.factors</span></code> and <code class="docutils literal notranslate"><span class="pre">pc2.factors</span></code> provides a
diagnostic of the convergence of the algorithm (i.e., if they are close
then it has likely converged). Alternatively, the output value of <code class="docutils literal notranslate"><span class="pre">error_bound</span></code>
gives an approximate upper bound for the spectral norm of the difference
between <span class="math notranslate nohighlight">\(M\)</span> and the projection of <span class="math notranslate nohighlight">\(M\)</span> into the space spanned by
the columns of <code class="docutils literal notranslate"><span class="pre">range_sketch</span></code>.
Algorithms are based on Algorithms 8
and 9 in Martinsson and Tropp, <a class="reference external" href="https://arxiv.org/pdf/2002.01387">https://arxiv.org/pdf/2002.01387</a> .</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_components</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  Number of principal components to return.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the windows
to compute the statistic in (default: the entire genome).</p></li>
<li><p><strong>samples</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a>)  Samples to perform PCA with (default: all samples).</p></li>
<li><p><strong>individuals</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a>)  Individuals to perform PCA with. Cannot specify
both <code class="docutils literal notranslate"><span class="pre">samples</span></code> and <code class="docutils literal notranslate"><span class="pre">individuals</span></code>.</p></li>
<li><p><strong>time_windows</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a>)  The time interval on which to apply PCA:
currently, this must be either None (default, covers all time)
or a single interval.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of relatedness to be computed
(defaults to branch; see
<a class="reference internal" href="#tskit.TreeSequence.genetic_relatedness_vector" title="tskit.TreeSequence.genetic_relatedness_vector"><code class="xref py py-meth docutils literal notranslate"><span class="pre">genetic_relatedness_vector</span></code></a>).</p></li>
<li><p><strong>centre</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to centre the genetic relatedness matrix.</p></li>
<li><p><strong>num_iterations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  Number of power iterations used in the range finding
algorithm.</p></li>
<li><p><strong>num_oversamples</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  Number of additional test vectors (default: 10).
Cannot specify along with range_sketch.</p></li>
<li><p><strong>random_seed</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The random seed. If this is None, a random seed will
be automatically generated. Valid random seeds are between 1 and
<span class="math notranslate nohighlight">\(2^32  1\)</span>. Only used if <cite>range_sketch</cite> is not provided.</p></li>
<li><p><strong>range_sketch</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a>)  Sketch matrix for each window. Default is
randomly generated; cannot specify along with num_oversamples.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A <a class="reference internal" href="#tskit.PCAResult" title="tskit.PCAResult"><code class="xref py py-class docutils literal notranslate"><span class="pre">PCAResult</span></code></a> object, containing estimated principal components,
eigenvalues, and other information:
the principal component loadings are in PCAResult.factors
and the principal values are in PCAResult.eigenvalues.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.trait_covariance">
<span class="sig-name descname"><span class="pre">trait_covariance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">W</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'site'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.trait_covariance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.trait_covariance" title="Link to this definition">#</a></dt>
<dd><p>Computes the mean squared covariances between each of the columns of <code class="docutils literal notranslate"><span class="pre">W</span></code>
(the phenotypes) and inheritance along the tree sequence.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.
Operates on all samples in the tree sequence.</p>
<p>Concretely, if <cite>g</cite> is a binary vector that indicates inheritance from an allele,
branch, or node and <cite>w</cite> is a column of W, normalised to have mean zero,
then the covariance of <cite>g</cite> and <cite>w</cite> is <span class="math notranslate nohighlight">\(\sum_i g_i w_i\)</span>, the sum of the
weights corresponding to entries of <cite>g</cite> that are <cite>1</cite>. Since weights sum to
zero, this is also equal to the sum of weights whose entries of <cite>g</cite> are 0.
So, <span class="math notranslate nohighlight">\(cov(g,w)^2 = ((\sum_i g_i w_i)^2 + (\sum_i (1-g_i) w_i)^2)/2\)</span>.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>:</p>
<dl class="simple">
<dt>site</dt><dd><p>The sum of squared covariances between presence/absence of each allele and
phenotypes, divided by length of the window (if <code class="docutils literal notranslate"><span class="pre">span_normalise=True</span></code>).
This is computed as sum_a (sum(w[a])^2 / 2), where
w is a column of W with the average subtracted off,
and w[a] is the sum of all entries of w corresponding to samples
carrying allele a, and the first sum is over all alleles.</p>
</dd>
<dt>branch</dt><dd><p>The sum of squared covariances between the split induced by each branch and
phenotypes, multiplied by branch length, averaged across trees in
the window. This is computed as above: a branch with total weight
w[b] below b contributes (branch length) * w[b]^2 to the total
value for a tree. (Since the sum of w is zero, the total weight
below b and not below b are equal, canceling the factor of 2
above.)</p>
</dd>
<dt>node</dt><dd><p>For each node, the squared covariance between the property of
inheriting from this node and phenotypes, computed as in branch.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>W</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a>)  An array of values with one row for each sample and one
column for each phenotype.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of the statistic to be computed
(defaults to site).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).
If windows=None and W is a single column, a numpy scalar is returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.trait_correlation">
<span class="sig-name descname"><span class="pre">trait_correlation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">W</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'site'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.trait_correlation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.trait_correlation" title="Link to this definition">#</a></dt>
<dd><p>Computes the mean squared correlations between each of the columns of <code class="docutils literal notranslate"><span class="pre">W</span></code>
(the phenotypes) and inheritance along the tree sequence.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.
Operates on all samples in the tree sequence.</p>
<p>This is computed as squared covariance in
<a class="reference internal" href="#tskit.TreeSequence.trait_covariance" title="tskit.TreeSequence.trait_covariance"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trait_covariance</span></code></a>,
but divided by <span class="math notranslate nohighlight">\(p (1-p)\)</span>, where <cite>p</cite> is the proportion of samples
inheriting from the allele, branch, or node in question.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>:</p>
<dl class="simple">
<dt>site</dt><dd><p>The sum of squared correlations between presence/absence of each allele and
phenotypes, divided by length of the window (if <code class="docutils literal notranslate"><span class="pre">span_normalise=True</span></code>).
This is computed as the
<a class="reference internal" href="#tskit.TreeSequence.trait_covariance" title="tskit.TreeSequence.trait_covariance"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trait_covariance</span></code></a>
divided by the variance of the relevant column of W
and by <span class="math notranslate nohighlight">\(p * (1 - p)\)</span>, where <span class="math notranslate nohighlight">\(p\)</span> is the allele frequency.</p>
</dd>
<dt>branch</dt><dd><p>The sum of squared correlations between the split induced by each branch and
phenotypes, multiplied by branch length, averaged across trees in
the window. This is computed as the
<a class="reference internal" href="#tskit.TreeSequence.trait_covariance" title="tskit.TreeSequence.trait_covariance"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trait_covariance</span></code></a>,
divided by the variance of the column of w
and by <span class="math notranslate nohighlight">\(p * (1 - p)\)</span>, where <span class="math notranslate nohighlight">\(p\)</span> is the proportion of
the samples lying below the branch.</p>
</dd>
<dt>node</dt><dd><p>For each node, the squared correlation between the property of
inheriting from this node and phenotypes, computed as in branch.</p>
</dd>
</dl>
<p>Note that above we divide by the <strong>sample</strong> variance, which for a
vector x of length n is <code class="docutils literal notranslate"><span class="pre">np.var(x)</span> <span class="pre">*</span> <span class="pre">n</span> <span class="pre">/</span> <span class="pre">(n-1)</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>W</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a>)  An array of values with one row for each sample and one
column for each phenotype. Each column must have positive standard
deviation.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of the statistic to be computed
(defaults to site).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).
If windows=None and W is a single column, a numpy scalar is returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.trait_regression">
<span class="sig-name descname"><span class="pre">trait_regression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.trait_regression"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.trait_regression" title="Link to this definition">#</a></dt>
<dd><p>Deprecated synonym for
<a class="reference internal" href="#tskit.TreeSequence.trait_linear_model" title="tskit.TreeSequence.trait_linear_model"><code class="xref py py-meth docutils literal notranslate"><span class="pre">trait_linear_model</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.trait_linear_model">
<span class="sig-name descname"><span class="pre">trait_linear_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">W</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Z</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'site'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.trait_linear_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.trait_linear_model" title="Link to this definition">#</a></dt>
<dd><p>Finds the relationship between trait and genotype after accounting for
covariates.  Concretely, for each trait w (i.e., each column of W),
this does a least-squares fit of the linear model <span class="math notranslate nohighlight">\(w \sim g + Z\)</span>,
where <span class="math notranslate nohighlight">\(g\)</span> is inheritance in the tree sequence (e.g., genotype)
and the columns of <span class="math notranslate nohighlight">\(Z\)</span> are covariates, and returns the squared
coefficient of <span class="math notranslate nohighlight">\(g\)</span> in this linear model.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.
Operates on all samples in the tree sequence.</p>
<p>To do this, if <cite>g</cite> is a binary vector that indicates inheritance from an allele,
branch, or node and <cite>w</cite> is a column of W, there are <span class="math notranslate nohighlight">\(k\)</span> columns of
<span class="math notranslate nohighlight">\(Z\)</span>, and the <span class="math notranslate nohighlight">\(k+2\)</span>-vector <span class="math notranslate nohighlight">\(b\)</span> minimises
<span class="math notranslate nohighlight">\(\sum_i (w_i - b_0 - b_1 g_i - b_2 z_{2,i} - ... b_{k+2} z_{k+2,i})^2\)</span>
then this returns the number <span class="math notranslate nohighlight">\(b_1^2\)</span>. If <span class="math notranslate nohighlight">\(g\)</span> lies in the linear span
of the columns of <span class="math notranslate nohighlight">\(Z\)</span>, then <span class="math notranslate nohighlight">\(b_1\)</span> is set to 0. To fit the
linear model without covariates (only the intercept), set <cite>Z = None</cite>.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>:</p>
<dl class="simple">
<dt>site</dt><dd><p>Computes the sum of <span class="math notranslate nohighlight">\(b_1^2/2\)</span> for each allele in the window,
as above with <span class="math notranslate nohighlight">\(g\)</span> indicating presence/absence of the allele,
then divided by the length of the window if <code class="docutils literal notranslate"><span class="pre">span_normalise=True</span></code>.
(For biallelic loci, this number is the same for both alleles, and so summing
over each cancels the factor of two.)</p>
</dd>
<dt>branch</dt><dd><p>The squared coefficient <span class="math notranslate nohighlight">\(b_1^2\)</span>, computed for the split induced by each
branch (i.e., with <span class="math notranslate nohighlight">\(g\)</span> indicating inheritance from that branch),
multiplied by branch length and tree span, summed over all trees
in the window, and divided by the length of the window if
<code class="docutils literal notranslate"><span class="pre">span_normalise=True</span></code>.</p>
</dd>
<dt>node</dt><dd><p>For each node, the squared coefficient <span class="math notranslate nohighlight">\(b_1^2\)</span>, computed for
the property of inheriting from this node, as in branch.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>W</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a>)  An array of values with one row for each sample and one
column for each phenotype.</p></li>
<li><p><strong>Z</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a>)  An array of values with one row for each sample and one
column for each covariate, or <cite>None</cite>. Columns of <cite>Z</cite> must be linearly
independent.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of the statistic to be computed
(defaults to site).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).
If windows=None and W is a single column, a numpy scalar is returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.segregating_sites">
<span class="sig-name descname"><span class="pre">segregating_sites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'site'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.segregating_sites"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.segregating_sites" title="Link to this definition">#</a></dt>
<dd><p>Computes the density of segregating sites for each of the sets of nodes
from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>, and related quantities.
Please see the <a class="reference internal" href="stats.html#sec-stats-sample-sets-one-way"><span class="std std-ref">one-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> argument is interpreted
and how it interacts with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>, <a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>. For a sample set <code class="docutils literal notranslate"><span class="pre">A</span></code>, computes:</p>
<dl class="simple">
<dt>site</dt><dd><p>The sum over sites of [the number of alleles found in <code class="docutils literal notranslate"><span class="pre">A</span></code> at each site
minus one], per unit of chromosome length.
If all sites have at most two alleles in <code class="docutils literal notranslate"><span class="pre">A</span></code>,
this is the density of segregating/polymorphic sites in <code class="docutils literal notranslate"><span class="pre">A</span></code>
(since the minus one reduces the sum for monoallelic sites).
For sites with more than two alleles, the sum is increased by
one for each additional allele segregating in <code class="docutils literal notranslate"><span class="pre">A</span></code>.
To get the <strong>number</strong> of segregating alleles in <code class="docutils literal notranslate"><span class="pre">A</span></code>,
use <code class="docutils literal notranslate"><span class="pre">span_normalise=False</span></code>.</p>
</dd>
<dt>branch</dt><dd><p>The total length of all branches in the tree subtended by the samples in
<code class="docutils literal notranslate"><span class="pre">A</span></code>, averaged across the window.</p>
</dd>
<dt>node</dt><dd><p>The proportion of the window on which the node is ancestral to some,
but not all, of the samples in <code class="docutils literal notranslate"><span class="pre">A</span></code>.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of the statistic to be computed
(defaults to site).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).
If there is one sample set and windows=None, a numpy scalar is returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.allele_frequency_spectrum">
<span class="sig-name descname"><span class="pre">allele_frequency_spectrum</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'site'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">polarised</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.allele_frequency_spectrum"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.allele_frequency_spectrum" title="Link to this definition">#</a></dt>
<dd><p>Computes the allele frequency spectrum (AFS) in windows across the genome for
with respect to the specified <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-sample-sets"><span class="std std-ref">sample sets</span></a>,
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
<a class="reference internal" href="stats.html#sec-stats-polarisation"><span class="std std-ref">polarised</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.
and see <a class="reference external" href="https://tskit.dev/tutorials/analysing_tree_sequences.html#sec-tutorial-afs" title="(in Project name not set)"><span>Allele frequency spectra</span></a> for examples of how to use this method.</p>
<p>Similar to other windowed stats, the first dimension in the returned array
corresponds to windows, such that <code class="docutils literal notranslate"><span class="pre">result[i]</span></code> is the AFS in the ith
window. The AFS in each window is a k-dimensional numpy array, where k is
the number of input sample sets, such that <code class="docutils literal notranslate"><span class="pre">result[i,</span> <span class="pre">j0,</span> <span class="pre">j1,</span> <span class="pre">...]</span></code> is the
value associated with frequency <code class="docutils literal notranslate"><span class="pre">j0</span></code> in <code class="docutils literal notranslate"><span class="pre">sample_sets[0]</span></code>, <code class="docutils literal notranslate"><span class="pre">j1</span></code> in
<code class="docutils literal notranslate"><span class="pre">sample_sets[1]</span></code>, etc, in window <code class="docutils literal notranslate"><span class="pre">i</span></code>. From here, we will assume that
<code class="docutils literal notranslate"><span class="pre">afs</span></code> corresponds to the result in a single window, i.e.,
<code class="docutils literal notranslate"><span class="pre">afs</span> <span class="pre">=</span> <span class="pre">result[i]</span></code>.</p>
<p>If a single sample set is specified, the allele frequency spectrum within
this set is returned, such that <code class="docutils literal notranslate"><span class="pre">afs[j]</span></code> is the value associated with
frequency <code class="docutils literal notranslate"><span class="pre">j</span></code>. Thus, singletons are counted in <code class="docutils literal notranslate"><span class="pre">afs[1]</span></code>, doubletons in
<code class="docutils literal notranslate"><span class="pre">afs[2]</span></code>, and so on. The zeroth entry counts alleles or branches not
seen in the samples but that are polymorphic among the rest of the samples
of the tree sequence; likewise, the last entry counts alleles fixed in
the sample set but polymorphic in the entire set of samples. Please see
the <a class="reference external" href="https://tskit.dev/tutorials/analysing_tree_sequences.html#sec-tutorial-afs-zeroth-entry" title="(in Project name not set)"><span>Zeroth and final entries in the AFS</span></a> for an illustration.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Please note that singletons are <strong>not</strong> counted in the initial
entry in each AFS array (i.e., <code class="docutils literal notranslate"><span class="pre">afs[0]</span></code>), but in <code class="docutils literal notranslate"><span class="pre">afs[1]</span></code>.</p>
</div>
<p>If <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> is None (the default), the allele frequency spectrum
for all samples in the tree sequence is returned. For convenience, if
there is only a single sample set, the outer list may be omitted (so that,
unlike other statistics, <code class="docutils literal notranslate"><span class="pre">sample_sets=[0,1,2]</span></code> is equivalent to
<code class="docutils literal notranslate"><span class="pre">sample_sets=[[0,1,2]]</span></code>).</p>
<p>If more than one sample set is specified, the <strong>joint</strong> allele frequency
spectrum within windows is returned. For example, if we set
<code class="docutils literal notranslate"><span class="pre">sample_sets</span> <span class="pre">=</span> <span class="pre">[S0,</span> <span class="pre">S1]</span></code>, then afs[1, 2] counts the number of sites that
are singletons within S0 and doubletons in S1. The dimensions of the
output array will be <code class="docutils literal notranslate"><span class="pre">[num_windows]</span> <span class="pre">+</span> <span class="pre">[1</span> <span class="pre">+</span> <span class="pre">len(S)</span> <span class="pre">for</span> <span class="pre">S</span> <span class="pre">in</span> <span class="pre">sample_sets]</span></code>.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">polarised</span></code> is False (the default) the AFS will be <em>folded</em>, so that
the counts do not depend on knowing which allele is ancestral. If folded,
the frequency spectrum for a single sample set <code class="docutils literal notranslate"><span class="pre">S</span></code> has <code class="docutils literal notranslate"><span class="pre">afs[j]</span> <span class="pre">=</span> <span class="pre">0</span></code> for
all <code class="docutils literal notranslate"><span class="pre">j</span> <span class="pre">&gt;</span> <span class="pre">len(S)</span> <span class="pre">/</span> <span class="pre">2</span></code>, so that alleles at frequency <code class="docutils literal notranslate"><span class="pre">j</span></code> and <code class="docutils literal notranslate"><span class="pre">len(S)</span> <span class="pre">-</span> <span class="pre">j</span></code>
both add to the same entry. If there is more than one sample set, the
returned array is lower triangular in a similar way. For more details,
especially about handling of multiallelic sites, see <a class="reference internal" href="stats.html#sec-stats-notes-afs"><span class="std std-ref">Allele frequency spectrum</span></a>.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>:</p>
<dl class="simple">
<dt>site</dt><dd><p>The number of alleles at a given frequency within the specified sample
sets for each window, per unit of sequence length. To obtain the total
number of alleles, set <code class="docutils literal notranslate"><span class="pre">span_normalise</span></code> to False.</p>
</dd>
<dt>branch</dt><dd><p>The total length of branches in the trees subtended by subsets of the
specified sample sets, per unit of sequence length. To obtain the
total, set <code class="docutils literal notranslate"><span class="pre">span_normalise</span></code> to False.</p>
</dd>
<dt>node</dt><dd><p>Not supported for this method (raises a ValueError).</p>
</dd>
</dl>
<p>For example, suppose that <cite>S0</cite> is a list of 5 sample IDs, and <cite>S1</cite> is
a list of 3 other sample IDs. Then <cite>afs = ts.allele_frequency_spectrum([S0, S1],
mode=site, span_normalise=False)</cite> will be a 5x3 numpy array, and if
there are six alleles that are present in only one sample of <cite>S0</cite> but
two samples of <cite>S1</cite>, then <cite>afs[1,2]</cite> will be equal to 6.  Similarly,
<cite>branch_afs = ts.allele_frequency_spectrum([S0, S1], mode=branch,
span_normalise=False)</cite> will also be a 5x3 array, and <cite>branch_afs[1,2]</cite>
will be the total area (i.e., length times span) of all branches that
are above exactly one sample of <cite>S0</cite> and two samples of <cite>S1</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of lists of Node IDs, specifying the
groups of samples to compute the joint allele frequency.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between windows
along the genome.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of the statistic to be computed
(defaults to site).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A (k + 1) dimensional numpy array, where k is the number of sample
sets specified.
If there is one sample set and windows=None, a 1 dimensional array is
returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.Tajimas_D">
<span class="sig-name descname"><span class="pre">Tajimas_D</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'site'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.Tajimas_D"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.Tajimas_D" title="Link to this definition">#</a></dt>
<dd><p>Computes Tajimas D of sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> in windows.
Please see the <a class="reference internal" href="stats.html#sec-stats-sample-sets-one-way"><span class="std std-ref">one-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> argument is interpreted
and how it interacts with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>, <a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.
Operates on <code class="docutils literal notranslate"><span class="pre">k</span> <span class="pre">=</span> <span class="pre">1</span></code> sample sets at a
time. For a sample set <code class="docutils literal notranslate"><span class="pre">X</span></code> of <code class="docutils literal notranslate"><span class="pre">n</span></code> nodes, if and <code class="docutils literal notranslate"><span class="pre">T</span></code> is the mean
number of pairwise differing sites in <code class="docutils literal notranslate"><span class="pre">X</span></code> and <code class="docutils literal notranslate"><span class="pre">S</span></code> is the number of
sites segregating in <code class="docutils literal notranslate"><span class="pre">X</span></code> (computed with <a class="reference internal" href="#tskit.TreeSequence.diversity" title="tskit.TreeSequence.diversity"><code class="xref py py-meth docutils literal notranslate"><span class="pre">diversity</span></code></a> and <a class="reference internal" href="#tskit.TreeSequence.segregating_sites" title="tskit.TreeSequence.segregating_sites"><code class="xref py py-meth docutils literal notranslate"><span class="pre">segregating</span> <span class="pre">sites</span></code></a>, respectively, both not span
normalised), then Tajimas D is</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">D</span> <span class="o">=</span> <span class="p">(</span><span class="n">T</span> <span class="o">-</span> <span class="n">S</span> <span class="o">/</span> <span class="n">h</span><span class="p">)</span> <span class="o">/</span> <span class="n">sqrt</span><span class="p">(</span><span class="n">a</span> <span class="o">*</span> <span class="n">S</span> <span class="o">+</span> <span class="p">(</span><span class="n">b</span> <span class="o">/</span> <span class="n">c</span><span class="p">)</span> <span class="o">*</span> <span class="n">S</span> <span class="o">*</span> <span class="p">(</span><span class="n">S</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span>
<span class="n">h</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">+</span> <span class="o">...</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="n">n</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">g</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">/</span> <span class="mi">2</span><span class="o">**</span><span class="mi">2</span> <span class="o">+</span> <span class="o">...</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="n">n</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span>
<span class="n">a</span> <span class="o">=</span> <span class="p">(</span><span class="n">n</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="mi">3</span> <span class="o">*</span> <span class="p">(</span><span class="n">n</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="n">h</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">h</span><span class="o">**</span><span class="mi">2</span>
<span class="n">b</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">*</span> <span class="p">(</span><span class="n">n</span><span class="o">**</span><span class="mi">2</span> <span class="o">+</span> <span class="n">n</span> <span class="o">+</span> <span class="mi">3</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="mi">9</span> <span class="o">*</span> <span class="n">n</span> <span class="o">*</span> <span class="p">(</span><span class="n">n</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span> <span class="o">-</span> <span class="p">(</span><span class="n">n</span> <span class="o">+</span> <span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">h</span> <span class="o">*</span> <span class="n">n</span><span class="p">)</span> <span class="o">+</span> <span class="n">g</span> <span class="o">/</span> <span class="n">h</span><span class="o">**</span><span class="mi">2</span>
<span class="n">c</span> <span class="o">=</span> <span class="n">h</span><span class="o">**</span><span class="mi">2</span> <span class="o">+</span> <span class="n">g</span>
</pre></div>
</div>
<p>What is computed for diversity and segregating sites depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>;
see those functions for more details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of 2-tuples, or None.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of the statistic to be computed
(defaults to site).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).
If there is one sample set and windows=None, a numpy scalar is returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.Fst">
<span class="sig-name descname"><span class="pre">Fst</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indexes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'site'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.Fst"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.Fst" title="Link to this definition">#</a></dt>
<dd><p>Computes windowed Fst between pairs of sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
Operates on <code class="docutils literal notranslate"><span class="pre">k</span> <span class="pre">=</span> <span class="pre">2</span></code> sample sets at a time; please see the
<a class="reference internal" href="stats.html#sec-stats-sample-sets-multi-way"><span class="std std-ref">multi-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> and <code class="docutils literal notranslate"><span class="pre">indexes</span></code> arguments are
interpreted and how they interact with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.</p>
<p>For sample sets <code class="docutils literal notranslate"><span class="pre">X</span></code> and <code class="docutils literal notranslate"><span class="pre">Y</span></code>, if <code class="docutils literal notranslate"><span class="pre">d(X,</span> <span class="pre">Y)</span></code> is the
<a class="reference internal" href="#tskit.TreeSequence.divergence" title="tskit.TreeSequence.divergence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">divergence</span></code></a>
between <code class="docutils literal notranslate"><span class="pre">X</span></code> and <code class="docutils literal notranslate"><span class="pre">Y</span></code>, and <code class="docutils literal notranslate"><span class="pre">d(X)</span></code> is the
<a class="reference internal" href="#tskit.TreeSequence.diversity" title="tskit.TreeSequence.diversity"><code class="xref py py-meth docutils literal notranslate"><span class="pre">diversity</span></code></a> of <code class="docutils literal notranslate"><span class="pre">X</span></code>, then what is
computed is</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">Fst</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="mi">2</span> <span class="o">*</span> <span class="p">(</span><span class="n">d</span><span class="p">(</span><span class="n">X</span><span class="p">)</span> <span class="o">+</span> <span class="n">d</span><span class="p">(</span><span class="n">Y</span><span class="p">))</span> <span class="o">/</span> <span class="p">(</span><span class="n">d</span><span class="p">(</span><span class="n">X</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">d</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">Y</span><span class="p">)</span> <span class="o">+</span> <span class="n">d</span><span class="p">(</span><span class="n">Y</span><span class="p">))</span>
</pre></div>
</div>
<p>What is computed for diversity and divergence depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>;
see those functions for more details.</p>
<p>For <code class="docutils literal notranslate"><span class="pre">mode='site'</span></code>, this definition of Fst appears as equation (6) in
<a class="reference external" href="https://doi.org/10.1017/S0016672300029827">Slatkin (1991)</a>, and
is also found as equation (9) in
<a class="reference external" href="https://doi.org/10.1073/pnas.70.12.3321">Nei (1973)</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of 2-tuples.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of the statistic to be computed
(defaults to site).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).
If there is one pair of sample sets and windows=None, a numpy scalar is
returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.Y3">
<span class="sig-name descname"><span class="pre">Y3</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indexes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'site'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.Y3"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.Y3" title="Link to this definition">#</a></dt>
<dd><p>Computes the Y statistic between triples of sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
Operates on <code class="docutils literal notranslate"><span class="pre">k</span> <span class="pre">=</span> <span class="pre">3</span></code> sample sets at a time; please see the
<a class="reference internal" href="stats.html#sec-stats-sample-sets-multi-way"><span class="std std-ref">multi-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> and <code class="docutils literal notranslate"><span class="pre">indexes</span></code> arguments are
interpreted and how they interact with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>. Each is an average across every
combination of trios of samples <code class="docutils literal notranslate"><span class="pre">(a,</span> <span class="pre">b,</span> <span class="pre">c)</span></code>, one chosen from each sample set:</p>
<dl class="simple">
<dt>site</dt><dd><p>The average density of sites at which <code class="docutils literal notranslate"><span class="pre">a</span></code> differs from <code class="docutils literal notranslate"><span class="pre">b</span></code> and
<code class="docutils literal notranslate"><span class="pre">c</span></code>, per unit of chromosome length.</p>
</dd>
<dt>branch</dt><dd><p>The average length of all branches that separate <code class="docutils literal notranslate"><span class="pre">a</span></code> from <code class="docutils literal notranslate"><span class="pre">b</span></code>
and <code class="docutils literal notranslate"><span class="pre">c</span></code> (in units of time).</p>
</dd>
<dt>node</dt><dd><p>For each node, the average proportion of the window on which <code class="docutils literal notranslate"><span class="pre">a</span></code>
inherits from that node but <code class="docutils literal notranslate"><span class="pre">b</span></code> and <code class="docutils literal notranslate"><span class="pre">c</span></code> do not, or vice-versa.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of 3-tuples, or None.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of the statistic to be computed
(defaults to site).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).
If there is one triple of sample sets and windows=None, a numpy scalar is
returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.Y2">
<span class="sig-name descname"><span class="pre">Y2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indexes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'site'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.Y2"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.Y2" title="Link to this definition">#</a></dt>
<dd><p>Computes the Y2 statistic between pairs of sets of nodes from <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
Operates on <code class="docutils literal notranslate"><span class="pre">k</span> <span class="pre">=</span> <span class="pre">2</span></code> sample sets at a time; please see the
<a class="reference internal" href="stats.html#sec-stats-sample-sets-multi-way"><span class="std std-ref">multi-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> and <code class="docutils literal notranslate"><span class="pre">indexes</span></code> arguments are
interpreted and how they interact with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>. Each is computed exactly as
<code class="docutils literal notranslate"><span class="pre">Y3</span></code>, except that the average is across every possible trio of samples
<code class="docutils literal notranslate"><span class="pre">(a,</span> <span class="pre">b1,</span> <span class="pre">b2)</span></code>, where <code class="docutils literal notranslate"><span class="pre">a</span></code> is chosen from the first sample set, and
<code class="docutils literal notranslate"><span class="pre">b1,</span> <span class="pre">b2</span></code> are chosen (without replacement) from the second sample set.
See <a class="reference internal" href="#tskit.TreeSequence.Y3" title="tskit.TreeSequence.Y3"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Y3</span></code></a> for more details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of 2-tuples, or None.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of the statistic to be computed
(defaults to site).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).
If there is one pair of sample sets and windows=None, a numpy scalar is
returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.Y1">
<span class="sig-name descname"><span class="pre">Y1</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'site'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.Y1"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.Y1" title="Link to this definition">#</a></dt>
<dd><p>Computes the Y1 statistic within each of the sets of nodes given by
<code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
Please see the <a class="reference internal" href="stats.html#sec-stats-sample-sets-one-way"><span class="std std-ref">one-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> argument is interpreted
and how it interacts with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>, <a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.
Operates on <code class="docutils literal notranslate"><span class="pre">k</span> <span class="pre">=</span> <span class="pre">1</span></code> sample set at a time.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>. Each is computed exactly as
<code class="docutils literal notranslate"><span class="pre">Y3</span></code>, except that the average is across every possible trio of samples
samples <code class="docutils literal notranslate"><span class="pre">(a1,</span> <span class="pre">a2,</span> <span class="pre">a3)</span></code> all chosen without replacement from the same
sample set. See <a class="reference internal" href="#tskit.TreeSequence.Y3" title="tskit.TreeSequence.Y3"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Y3</span></code></a> for more details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of the statistic to be computed
(defaults to site).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).
If there is one sample set and windows=None, a numpy scalar is returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.f4">
<span class="sig-name descname"><span class="pre">f4</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indexes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'site'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.f4"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.f4" title="Link to this definition">#</a></dt>
<dd><p>Computes Pattersons f4 statistic between four groups of nodes from
<code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
Operates on <code class="docutils literal notranslate"><span class="pre">k</span> <span class="pre">=</span> <span class="pre">4</span></code> sample sets at a time; please see the
<a class="reference internal" href="stats.html#sec-stats-sample-sets-multi-way"><span class="std std-ref">multi-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> and <code class="docutils literal notranslate"><span class="pre">indexes</span></code> arguments are
interpreted and how they interact with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>. Each is an average across every possible
combination of four samples <code class="docutils literal notranslate"><span class="pre">(a,</span> <span class="pre">b;</span> <span class="pre">c,</span> <span class="pre">d)</span></code>, one chosen from each sample set:</p>
<dl class="simple">
<dt>site</dt><dd><p>The average density of sites at which <code class="docutils literal notranslate"><span class="pre">a</span></code> and <code class="docutils literal notranslate"><span class="pre">c</span></code> agree but
differs from <code class="docutils literal notranslate"><span class="pre">b</span></code> and <code class="docutils literal notranslate"><span class="pre">d</span></code>, minus the average density of sites at
which <code class="docutils literal notranslate"><span class="pre">a</span></code> and <code class="docutils literal notranslate"><span class="pre">d</span></code> agree but differs from <code class="docutils literal notranslate"><span class="pre">b</span></code> and <code class="docutils literal notranslate"><span class="pre">c</span></code>, per
unit of chromosome length.</p>
</dd>
<dt>branch</dt><dd><p>The average length of all branches that separate <code class="docutils literal notranslate"><span class="pre">a</span></code> and <code class="docutils literal notranslate"><span class="pre">c</span></code>
from <code class="docutils literal notranslate"><span class="pre">b</span></code> and <code class="docutils literal notranslate"><span class="pre">d</span></code>, minus the average length of all branches that
separate <code class="docutils literal notranslate"><span class="pre">a</span></code> and <code class="docutils literal notranslate"><span class="pre">d</span></code> from <code class="docutils literal notranslate"><span class="pre">b</span></code> and <code class="docutils literal notranslate"><span class="pre">c</span></code> (in units of time).</p>
</dd>
<dt>node</dt><dd><p>For each node, the average proportion of the window on which <code class="docutils literal notranslate"><span class="pre">a</span></code> and <code class="docutils literal notranslate"><span class="pre">c</span></code>
inherit from that node but <code class="docutils literal notranslate"><span class="pre">b</span></code> and <code class="docutils literal notranslate"><span class="pre">d</span></code> do not, or vice-versa,
minus the average proportion of the window on which <code class="docutils literal notranslate"><span class="pre">a</span></code> and <code class="docutils literal notranslate"><span class="pre">d</span></code>
inherit from that node but <code class="docutils literal notranslate"><span class="pre">b</span></code> and <code class="docutils literal notranslate"><span class="pre">c</span></code> do not, or vice-versa.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of 4-tuples, or None.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of the statistic to be computed
(defaults to site).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).
If there are four sample sets and windows=None, a numpy scalar is returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.f3">
<span class="sig-name descname"><span class="pre">f3</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indexes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'site'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.f3"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.f3" title="Link to this definition">#</a></dt>
<dd><p>Computes Pattersons f3 statistic between three groups of nodes from
<code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
Note that the order of the arguments of f3 differs across the literature:
here, <code class="docutils literal notranslate"><span class="pre">f3([A,</span> <span class="pre">B,</span> <span class="pre">C])</span></code> for sample sets <code class="docutils literal notranslate"><span class="pre">A</span></code>, <code class="docutils literal notranslate"><span class="pre">B</span></code>, and <code class="docutils literal notranslate"><span class="pre">C</span></code>
will estimate
<span class="math notranslate nohighlight">\(f_3(A; B, C) = \mathbb{E}[(p_A - p_B) (p_A - p_C)]\)</span>,
where <span class="math notranslate nohighlight">\(p_A\)</span> is the allele frequency in <code class="docutils literal notranslate"><span class="pre">A</span></code>.
When used as a test for admixture, the putatively admixed population
is usually placed as population <code class="docutils literal notranslate"><span class="pre">A</span></code> (see
<a class="reference external" href="https://doi.org/10.1534/genetics.115.183913">Peter (2016)</a>
for more discussion).</p>
<p>Operates on <code class="docutils literal notranslate"><span class="pre">k</span> <span class="pre">=</span> <span class="pre">3</span></code> sample sets at a time; please see the
<a class="reference internal" href="stats.html#sec-stats-sample-sets-multi-way"><span class="std std-ref">multi-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> and <code class="docutils literal notranslate"><span class="pre">indexes</span></code> arguments are
interpreted and how they interact with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>. Each works exactly as
<a class="reference internal" href="#tskit.TreeSequence.f4" title="tskit.TreeSequence.f4"><code class="xref py py-meth docutils literal notranslate"><span class="pre">f4</span></code></a>, except the average is across every possible
combination of four samples <code class="docutils literal notranslate"><span class="pre">(a1,</span> <span class="pre">b;</span> <span class="pre">a2,</span> <span class="pre">c)</span></code> where <cite>a1</cite> and <cite>a2</cite> have both
been chosen (without replacement) from the first sample set. See
<a class="reference internal" href="#tskit.TreeSequence.f4" title="tskit.TreeSequence.f4"><code class="xref py py-meth docutils literal notranslate"><span class="pre">f4</span></code></a> for more details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of 3-tuples, or None.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of the statistic to be computed
(defaults to site).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).
If there are three sample sets and windows=None, a numpy scalar is returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.f2">
<span class="sig-name descname"><span class="pre">f2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indexes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'site'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.f2"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.f2" title="Link to this definition">#</a></dt>
<dd><p>Computes Pattersons f2 statistic between two groups of nodes from
<code class="docutils literal notranslate"><span class="pre">sample_sets</span></code>.
Operates on <code class="docutils literal notranslate"><span class="pre">k</span> <span class="pre">=</span> <span class="pre">2</span></code> sample sets at a time; please see the
<a class="reference internal" href="stats.html#sec-stats-sample-sets-multi-way"><span class="std std-ref">multi-way statistics</span></a>
section for details on how the <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> and <code class="docutils literal notranslate"><span class="pre">indexes</span></code> arguments are
interpreted and how they interact with the dimensions of the output array.
See the <a class="reference internal" href="stats.html#sec-stats-interface"><span class="std std-ref">statistics interface</span></a> section for details on
<a class="reference internal" href="stats.html#sec-stats-windows"><span class="std std-ref">windows</span></a>,
<a class="reference internal" href="stats.html#sec-stats-mode"><span class="std std-ref">mode</span></a>,
<a class="reference internal" href="stats.html#sec-stats-span-normalise"><span class="std std-ref">span normalise</span></a>,
and <a class="reference internal" href="stats.html#sec-stats-output-format"><span class="std std-ref">return value</span></a>.</p>
<p>What is computed depends on <code class="docutils literal notranslate"><span class="pre">mode</span></code>. Each works exactly as
<a class="reference internal" href="#tskit.TreeSequence.f4" title="tskit.TreeSequence.f4"><code class="xref py py-meth docutils literal notranslate"><span class="pre">f4</span></code></a>, except the average is across every possible
combination of four samples <code class="docutils literal notranslate"><span class="pre">(a1,</span> <span class="pre">b1;</span> <span class="pre">a2,</span> <span class="pre">b2)</span></code> where <cite>a1</cite> and <cite>a2</cite> have
both been chosen (without replacement) from the first sample set, and <code class="docutils literal notranslate"><span class="pre">b1</span></code>
and <code class="docutils literal notranslate"><span class="pre">b2</span></code> have both been chosen (without replacement) from the second
sample set. See <a class="reference internal" href="#tskit.TreeSequence.f4" title="tskit.TreeSequence.f4"><code class="xref py py-meth docutils literal notranslate"><span class="pre">f4</span></code></a> for more details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p></li>
<li><p><strong>indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of 2-tuples, or None.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the windows
to compute the statistic in.</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string giving the type of the statistic to be computed
(defaults to site).</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to divide the result by the span of the
window (defaults to True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A ndarray with shape equal to (num windows, num statistics).
If there is one pair of sample sets and windows=None, a numpy scalar is
returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.mean_descendants">
<span class="sig-name descname"><span class="pre">mean_descendants</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.mean_descendants"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.mean_descendants" title="Link to this definition">#</a></dt>
<dd><p>Computes for every node the mean number of samples in each of the
<cite>sample_sets</cite> that descend from that node, averaged over the
portions of the genome for which the node is ancestral to <em>any</em> sample.
The output is an array, <cite>C[node, j]</cite>, which reports the total span of
all genomes in <cite>sample_sets[j]</cite> that inherit from <cite>node</cite>, divided by
the total span of the genome on which <cite>node</cite> is an ancestor to any
sample in the tree sequence.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The interface for this method is preliminary and may be subject to
backwards incompatible changes in the near future. The long-term stable
API for this method will be consistent with other <a class="reference internal" href="stats.html#sec-stats"><span class="std std-ref">Statistics</span></a>.
In particular, the normalization by proportion of the genome that <cite>node</cite>
is an ancestor to anyone may not be the default behaviour in the future.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of lists of node IDs.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An array with dimensions (number of nodes in the tree sequence,
number of reference sets)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.genealogical_nearest_neighbours">
<span class="sig-name descname"><span class="pre">genealogical_nearest_neighbours</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">focal</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.genealogical_nearest_neighbours"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.genealogical_nearest_neighbours" title="Link to this definition">#</a></dt>
<dd><p>Return the genealogical nearest neighbours (GNN) proportions for the given
focal nodes, with reference to two or more sets of interest, averaged over all
trees in the tree sequence.</p>
<p>The GNN proportions for a focal node in a single tree are given by first finding
the most recent common ancestral node <span class="math notranslate nohighlight">\(a\)</span> between the focal node and any
other node present in the reference sets. The GNN proportion for a specific
reference set, <span class="math notranslate nohighlight">\(S\)</span> is the number of nodes in <span class="math notranslate nohighlight">\(S\)</span> that descend from
<span class="math notranslate nohighlight">\(a\)</span>, as a proportion of the total number of descendant nodes in any of the
reference sets.</p>
<p>For example, consider a case with 2 sample sets, <span class="math notranslate nohighlight">\(S_1\)</span> and <span class="math notranslate nohighlight">\(S_2\)</span>.
For a given tree, <span class="math notranslate nohighlight">\(a\)</span> is the node that includes at least one descendant in
<span class="math notranslate nohighlight">\(S_1\)</span> or <span class="math notranslate nohighlight">\(S_2\)</span> (not including the focal node). If the descendants of
<span class="math notranslate nohighlight">\(a\)</span> include some nodes in <span class="math notranslate nohighlight">\(S_1\)</span> but no nodes in <span class="math notranslate nohighlight">\(S_2\)</span>, then the
GNN proportions for that tree will be 100% <span class="math notranslate nohighlight">\(S_1\)</span> and 0% <span class="math notranslate nohighlight">\(S_2\)</span>, or
<span class="math notranslate nohighlight">\([1.0, 0.0]\)</span>.</p>
<p>For a given focal node, the GNN proportions returned by this function are an
average of the GNNs for each tree, weighted by the genomic distance spanned by
that tree.</p>
<p>For an precise mathematical definition of GNN, see <a class="reference external" href="https://doi.org/10.1101/458067">https://doi.org/10.1101/458067</a></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The reference sets need not include all the samples, hence the most
recent common ancestral node of the reference sets, <span class="math notranslate nohighlight">\(a\)</span>, need not be
the immediate ancestor of the focal node. If the reference sets only comprise
sequences from relatively distant individuals, the GNN statistic may end up
as a measure of comparatively distant ancestry, even for tree sequences that
contain many closely related individuals.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The interface for this method is preliminary and may be subject to
backwards incompatible changes in the near future. The long-term stable
API for this method will be consistent with other <a class="reference internal" href="stats.html#sec-stats"><span class="std std-ref">Statistics</span></a>.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>focal</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of <span class="math notranslate nohighlight">\(n\)</span> nodes whose GNNs should be calculated.</p></li>
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of <span class="math notranslate nohighlight">\(m\)</span> lists of node IDs.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An <span class="math notranslate nohighlight">\(n\)</span>  by <span class="math notranslate nohighlight">\(m\)</span> array of focal nodes by GNN proportions.
Every focal node corresponds to a row. The numbers in each
row corresponding to the GNN proportion for each of the passed-in reference
sets. Rows therefore sum to one.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.kc_distance">
<span class="sig-name descname"><span class="pre">kc_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lambda_</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.kc_distance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.kc_distance" title="Link to this definition">#</a></dt>
<dd><p>Returns the average <a class="reference internal" href="#tskit.Tree.kc_distance" title="tskit.Tree.kc_distance"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.kc_distance()</span></code></a> between pairs of trees along
the sequence whose intervals overlap. The average is weighted by the
fraction of the sequence on which each pair of trees overlap.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><em>TreeSequence</em></a>)  The other tree sequence to compare to.</p></li>
<li><p><strong>lambda</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The KC metric lambda parameter determining the
relative weight of topology and branch length.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The computed KC distance between this tree sequence and other.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.count_topologies">
<span class="sig-name descname"><span class="pre">count_topologies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.count_topologies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.count_topologies" title="Link to this definition">#</a></dt>
<dd><p>Returns a generator that produces the same distribution of topologies as
<a class="reference internal" href="#tskit.Tree.count_topologies" title="tskit.Tree.count_topologies"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.count_topologies()</span></code></a> but sequentially for every tree in a tree
sequence. For use on a tree sequence this method is much faster than
computing the result independently per tree.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The interface for this method is preliminary and may be subject to
backwards incompatible changes in the near future.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>iter(<a class="reference internal" href="#tskit.TopologyCounter" title="tskit.TopologyCounter"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TopologyCounter</span></code></a>)</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.14)"><strong>ValueError</strong></a>  If nodes in <code class="docutils literal notranslate"><span class="pre">sample_sets</span></code> are invalid or are
internal samples.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.ibd_segments">
<span class="sig-name descname"><span class="pre">ibd_segments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">within</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">between</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_time</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_span</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">store_pairs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">store_segments</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.ibd_segments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.ibd_segments" title="Link to this definition">#</a></dt>
<dd><p>Finds pairs of samples that are identical by descent (IBD) and returns
the result as an <a class="reference internal" href="#tskit.IdentitySegments" title="tskit.IdentitySegments"><code class="xref py py-class docutils literal notranslate"><span class="pre">IdentitySegments</span></code></a> instance. The information
stored in this object is controlled by the <code class="docutils literal notranslate"><span class="pre">store_pairs</span></code> and
<code class="docutils literal notranslate"><span class="pre">store_segments</span></code> parameters. By default only total counts and other
statistics of the IBD segments are stored (i.e.,
<code class="docutils literal notranslate"><span class="pre">store_pairs=False</span></code>), since storing pairs and segments has a
substantial CPU and memory overhead. Please see the
<a class="reference internal" href="ibd.html#sec-identity"><span class="std std-ref">Identity by descent</span></a> section for more details on how to access the
information stored in the <a class="reference internal" href="#tskit.IdentitySegments" title="tskit.IdentitySegments"><code class="xref py py-class docutils literal notranslate"><span class="pre">IdentitySegments</span></code></a>.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">within</span></code> is specified, only IBD segments for pairs of nodes within
that set will be recorded. If <code class="docutils literal notranslate"><span class="pre">between</span></code> is specified, only IBD
segments from pairs that are in one or other of the specified sample
sets will be reported. Note that <code class="docutils literal notranslate"><span class="pre">within</span></code> and <code class="docutils literal notranslate"><span class="pre">between</span></code> are
mutually exclusive.</p>
<p>A pair of nodes <code class="docutils literal notranslate"><span class="pre">(u,</span> <span class="pre">v)</span></code> has an IBD segment with a left and right
coordinate <code class="docutils literal notranslate"><span class="pre">[left,</span> <span class="pre">right)</span></code> and ancestral node <code class="docutils literal notranslate"><span class="pre">a</span></code> iff the most
recent common ancestor of the segment <code class="docutils literal notranslate"><span class="pre">[left,</span> <span class="pre">right)</span></code> in nodes <code class="docutils literal notranslate"><span class="pre">u</span></code>
and <code class="docutils literal notranslate"><span class="pre">v</span></code> is <code class="docutils literal notranslate"><span class="pre">a</span></code>, and the segment has been inherited along the same
genealogical path (ie. it has not been broken by recombination). The
segments returned are the longest possible ones.</p>
<p>Note that this definition is purely genealogical  allelic states
<em>are not</em> considered here. If used without time or length thresholds, the
segments returned for a given pair will partition the span of the contig
represented by the tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>within</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of node IDs defining set of nodes that
we find IBD segments for. If not specified, this defaults to
all samples in the tree sequence.</p></li>
<li><p><strong>between</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>]</em>)  A list of lists of sample node IDs. Given
two sample sets A and B, only IBD segments will be returned such
that one of the samples is an element of A and the other is
an element of B. Cannot be specified with <code class="docutils literal notranslate"><span class="pre">within</span></code>.</p></li>
<li><p><strong>max_time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  Only segments inherited from common
ancestors whose node times are more recent than the specified time
will be returned. Specifying a maximum time is strongly recommended when
working with large tree sequences.</p></li>
<li><p><strong>min_span</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  Only segments in which the difference between
the right and left genome coordinates (i.e., the span of the
segment) is greater than this value will be included. (Default=0)</p></li>
<li><p><strong>store_pairs</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True store information separately for each
pair of samples <code class="docutils literal notranslate"><span class="pre">(a,</span> <span class="pre">b)</span></code> that are found to be IBD. Otherwise
store summary information about all sample pairs. (Default=False)</p></li>
<li><p><strong>store_segments</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True store each IBD segment
<code class="docutils literal notranslate"><span class="pre">(left,</span> <span class="pre">right,</span> <span class="pre">c)</span></code> and associate it with the corresponding
sample pair <code class="docutils literal notranslate"><span class="pre">(a,</span> <span class="pre">b)</span></code>. If True, implies <code class="docutils literal notranslate"><span class="pre">store_pairs</span></code>.
(Default=False).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An <a class="reference internal" href="#tskit.IdentitySegments" title="tskit.IdentitySegments"><code class="xref py py-class docutils literal notranslate"><span class="pre">IdentitySegments</span></code></a> object containing the recorded
IBD information.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.IdentitySegments" title="tskit.IdentitySegments">IdentitySegments</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.pair_coalescence_counts">
<span class="sig-name descname"><span class="pre">pair_coalescence_counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indexes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">span_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pair_normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'nodes'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.pair_coalescence_counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.pair_coalescence_counts" title="Link to this definition">#</a></dt>
<dd><p>Calculate the number of coalescing sample pairs per node, summed over
trees and weighted by tree span.</p>
<p>The number of coalescing pairs may be calculated within or between the
non-overlapping lists of samples contained in <cite>sample_sets</cite>. In the
latter case, pairs are counted if they have exactly one member in each
of two sample sets. If <cite>sample_sets</cite> is omitted, a single set
containing all samples is assumed.</p>
<p>The argument <cite>indexes</cite> may be used to specify which pairs of sample
sets to compute the statistic between, and in what order. If
<cite>indexes=None</cite>, then <cite>indexes</cite> is assumed to equal <cite>[(0,0)]</cite> for a
single sample set and <cite>[(0,1)]</cite> for two sample sets. For more than two
sample sets, <cite>indexes</cite> must be explicitly passed.</p>
<p>The argument <cite>time_windows</cite> may be used to count coalescence
events within time intervals (if an array of breakpoints is supplied)
rather than for individual nodes (the default).</p>
<p>The output array has dimension <cite>(windows, indexes, nodes)</cite> with
dimensions dropped when the corresponding argument is set to None.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with, or None.</p></li>
<li><p><strong>indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of 2-tuples, or None.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the
sequence windows to compute the statistic in, or None.</p></li>
<li><p><strong>span_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to divide the result by the span of
non-missing sequence in the window (defaults to True).</p></li>
<li><p><strong>pair_normalise</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to divide the result by the total
number of pairs for a given index (defaults to False).</p></li>
<li><p><strong>time_windows</strong>  Either a string nodes or an increasing
list of breakpoints between time intervals.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.pair_coalescence_quantiles">
<span class="sig-name descname"><span class="pre">pair_coalescence_quantiles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">quantiles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indexes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.pair_coalescence_quantiles"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.pair_coalescence_quantiles" title="Link to this definition">#</a></dt>
<dd><p>Estimate quantiles of pair coalescence times by inverting the empirical
CDF. This is equivalent to the inverted_cdf method of
<cite>numpy.quantile</cite> applied to node times, with weights proportional to
the number of coalescing pairs per node (averaged over trees, see
<cite>TreeSequence.pair_coalescence_counts</cite>).</p>
<p>Quantiles of pair coalescence times may be calculated within or
between the non-overlapping lists of samples contained in <cite>sample_sets</cite>. In
the latter case, pairs are counted if they have exactly one member in each
of two sample sets. If <cite>sample_sets</cite> is omitted, a single set containing
all samples is assumed.</p>
<p>The argument <cite>indexes</cite> may be used to specify which pairs of sample sets to
compute coalescences between, and in what order. If <cite>indexes=None</cite>, then
<cite>indexes</cite> is assumed to equal <cite>[(0,0)]</cite> for a single sample set and
<cite>[(0,1)]</cite> for two sample sets. For more than two sample sets, <cite>indexes</cite>
must be explicitly passed.</p>
<p>The output array has dimension <cite>(windows, indexes, quantiles)</cite> with
dimensions dropped when the corresponding argument is set to None.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>quantiles</strong>  A list of increasing breakpoints between [0, 1].</p></li>
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with, or None.</p></li>
<li><p><strong>indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of 2-tuples, or None.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the
sequence windows to compute the statistic in, or None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.pair_coalescence_rates">
<span class="sig-name descname"><span class="pre">pair_coalescence_rates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">time_windows</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_sets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indexes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.pair_coalescence_rates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.pair_coalescence_rates" title="Link to this definition">#</a></dt>
<dd><p>Estimate the rate at which pairs of samples coalesce within time
windows, using the empirical cumulative distribution function (ecdf) of
pair coalescence times.  Assuming that pair coalescence events follow a
nonhomogeneous Poisson process, the empirical rate for a time window
<span class="math notranslate nohighlight">\([a, b)\)</span> where <span class="math notranslate nohighlight">\(ecdf(b) &lt; 1\)</span> is,</p>
<p>..math:</p>
<blockquote>
<div><p>log(1 - frac{ecdf(b) - ecdf(a)}{1 - ecdf(a)}) / (a - b)</p>
</div></blockquote>
<p>If the last coalescence event is within <span class="math notranslate nohighlight">\([a, b)\)</span>, so that
<span class="math notranslate nohighlight">\(ecdf(b) = 1\)</span>, then an estimate of the empirical rate is</p>
<p>..math:</p>
<blockquote>
<div><p>(mathbb{E}[t | t &gt; a] - a)^{-1}</p>
</div></blockquote>
<p>where <span class="math notranslate nohighlight">\(\mathbb{E}[t | t &lt; a]\)</span> is the average pair coalescence time
conditional on coalescence after the start of the last epoch.</p>
<p>The first breakpoint in <cite>time_windows</cite> must start at the age of the
samples, and the last must end at infinity. In the output array, any
time windows where all pairs have coalesced by start of the window will
contain <cite>NaN</cite> values.</p>
<p>Pair coalescence rates may be calculated within or between the
non-overlapping lists of samples contained in <cite>sample_sets</cite>. In the
latter case, pairs are counted if they have exactly one member in each
of two sample sets. If <cite>sample_sets</cite> is omitted, a single group
containing all samples is assumed.</p>
<p>The argument <cite>indexes</cite> may be used to specify which pairs of sample
sets to compute the statistic between, and in what order. If
<cite>indexes=None</cite>, then <cite>indexes</cite> is assumed to equal <cite>[(0,0)]</cite> for a
single sample set and <cite>[(0,1)]</cite> for two sample sets. For more than two
sample sets, <cite>indexes</cite> must be explicitly passed.</p>
<p>The output array has dimension <cite>(windows, indexes, time_windows)</cite> with
dimensions dropped when the corresponding argument is set to None.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>time_windows</strong>  An increasing list of breakpoints between time
intervals, starting at the age of the samples and ending at
infinity.</p></li>
<li><p><strong>sample_sets</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of lists of Node IDs, specifying the
groups of nodes to compute the statistic with, or None.</p></li>
<li><p><strong>indexes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of 2-tuples, or None.</p></li>
<li><p><strong>windows</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An increasing list of breakpoints between the
sequence windows to compute the statistic in, or None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.impute_unknown_mutations_time">
<span class="sig-name descname"><span class="pre">impute_unknown_mutations_time</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.impute_unknown_mutations_time"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.impute_unknown_mutations_time" title="Link to this definition">#</a></dt>
<dd><p>Returns an array of mutation times, where any unknown times are
imputed from the times of associated nodes. Not to be confused with
<a class="reference internal" href="#tskit.TableCollection.compute_mutation_times" title="tskit.TableCollection.compute_mutation_times"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.compute_mutation_times()</span></code></a>, which modifies the
<code class="docutils literal notranslate"><span class="pre">time</span></code> column of the mutations table in place.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>method</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The method used to impute the unknown mutation times.
Currently only min is supported, which uses the time of the node
below the mutation as the mutation time. The min method can also
be specified by <code class="docutils literal notranslate"><span class="pre">method=None</span></code> (Default: <code class="docutils literal notranslate"><span class="pre">None</span></code>).</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An array of length equal to the number of mutations in the
tree sequence.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.sample_nodes_by_ploidy">
<span class="sig-name descname"><span class="pre">sample_nodes_by_ploidy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ploidy</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.sample_nodes_by_ploidy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.sample_nodes_by_ploidy" title="Link to this definition">#</a></dt>
<dd><p>Returns an 2D array of node IDs, where each row has length <cite>ploidy</cite>.
This is useful when individuals are not defined in the tree sequence
so <cite>TreeSequence.individuals_nodes</cite> cannot be used. The samples are
placed in the array in the order which they are found in the node
table. The number of sample nodes must be a multiple of ploidy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ploidy</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The number of samples per individual.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A 2D array of node IDs, where each row has length <cite>ploidy</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.map_to_vcf_model">
<span class="sig-name descname"><span class="pre">map_to_vcf_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">individuals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ploidy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name_metadata_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individual_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_non_sample_nodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_transform</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contig_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isolated_as_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.map_to_vcf_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.map_to_vcf_model" title="Link to this definition">#</a></dt>
<dd><p>Maps the sample nodes in this tree sequence to a representation suitable for
VCF output, using the individuals if present.</p>
<p>Creates a VcfModelMapping object that contains a nodes-to-individual
mapping as a 2D array of (individuals, nodes), the individual names and VCF
compatible site positions and contig length. The
mapping is created by first checking if the tree sequence contains individuals.
If it does, the mapping is created using the individuals in the tree sequence.
By default only the sample nodes of the individuals are included in the mapping,
unless <code class="docutils literal notranslate"><span class="pre">include_non_sample_nodes</span></code> is set to True, in which case all nodes
belonging to the individuals are included. Any individuals without any nodes
will have no nodes in their row of the mapping, being essentially of zero ploidy.
If no individuals are present, the mapping is created using only the sample nodes
and the specified ploidy.</p>
<p>As the tskit data model allows non-integer positions, site positions and contig
length are transformed to integer values suitable for VCF output. The
transformation is done using the <code class="docutils literal notranslate"><span class="pre">position_transform</span></code> function, which must
return an integer numpy array the same dimension as the input. By default,
this is set to <code class="docutils literal notranslate"><span class="pre">numpy.round()</span></code> which will round values to the nearest integer.</p>
<p>If neither <code class="docutils literal notranslate"><span class="pre">name_metadata_key</span></code> nor <code class="docutils literal notranslate"><span class="pre">individual_names</span></code> is specified, the
individual names are set to <code class="docutils literal notranslate"><span class="pre">&quot;tsk_{individual_id}&quot;</span></code> for each individual. If
no individuals are present, the individual names are set to <code class="docutils literal notranslate"><span class="pre">&quot;tsk_{i}&quot;</span></code> with
<code class="docutils literal notranslate"><span class="pre">0</span> <span class="pre">&lt;=</span> <span class="pre">i</span> <span class="pre">&lt;</span> <span class="pre">num_sample_nodes/ploidy</span></code>.</p>
<p>A warning is emitted if any sample nodes do not have an individual ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>individuals</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  Specific individual IDs to include in the VCF. If not
specified and the tree sequence contains individuals, all individuals are
included that are associated with least one sample node (or at least one of
any node if <code class="docutils literal notranslate"><span class="pre">include_non_sample_nodes</span></code> is True), and the mapping arrays
will be in ascending order of the ID of the individual in the tree sequence.</p></li>
<li><p><strong>ploidy</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The ploidy, or number of nodes per individual. Only used when
the tree sequence does not contain individuals. Cannot be used if the tree
sequence contains individuals. Defaults to 1 if not specified.</p></li>
<li><p><strong>name_metadata_key</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The key in the individual metadata to use
for individual names. Cannot be specified simultaneously with
individual_names.</p></li>
<li><p><strong>individual_names</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  The names to use for each individual. Cannot
be specified simultaneously with name_metadata_key.</p></li>
<li><p><strong>include_non_sample_nodes</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, include all nodes belonging to
the individuals in the mapping. If False, only include sample nodes.
Defaults to False.</p></li>
<li><p><strong>position_transform</strong>  A callable that transforms the
site position values into integer valued coordinates suitable for
VCF. The function takes a single positional parameter x and must
return an integer numpy array the same dimension as x. By default,
this is set to <code class="docutils literal notranslate"><span class="pre">numpy.round()</span></code> which will round values to the
nearest integer. If the string legacy is provided here, the
pre 0.2.0 legacy behaviour of rounding values to the nearest integer
(starting from 1) and avoiding the output of identical positions
by incrementing is used.
See the <a class="reference internal" href="export.html#sec-export-vcf-modifying-coordinates"><span class="std std-ref">Modifying coordinates</span></a> for examples
and more information.</p></li>
<li><p><strong>contig_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The ID of the contig to use in the VCF output.
Defaults to 1 if not specified.</p></li>
<li><p><strong>isolated_as_missing</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, isolated samples without mutations
will be considered as missing data in the VCF output. If False, these samples
will have the ancestral state in the VCF output.
Default: True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A VcfModelMapping containing the node-to-individual mapping,
individual names, transformed positions, and transformed contig length.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.14)"><strong>ValueError</strong></a>  If both name_metadata_key and individual_names are specified,
if ploidy is specified when individuals are present, if an invalid individual
ID is specified, if a specified individual has no nodes, or if the number of
individuals doesnt match the number of names.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TreeSequence.pairwise_diversity">
<span class="sig-name descname"><span class="pre">pairwise_diversity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TreeSequence.pairwise_diversity"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TreeSequence.pairwise_diversity" title="Link to this definition">#</a></dt>
<dd><p>Returns the pairwise nucleotide site diversity, the average number of sites
that differ between a every possible pair of distinct samples.  If <cite>samples</cite> is
specified, calculate the diversity within this set.</p>
<blockquote>
<div><div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 0.2.0: </span>please use <a class="reference internal" href="#tskit.TreeSequence.diversity" title="tskit.TreeSequence.diversity"><code class="xref py py-meth docutils literal notranslate"><span class="pre">diversity()</span></code></a> instead. Since version 0.2.0 the error
semantics have also changed slightly. It is no longer an error
when there is one sample and a tskit.LibraryError is raised
when non-sample IDs are provided rather than a ValueError. It is
also no longer an error to compute pairwise diversity at sites
with multiple mutations.</p>
</div>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>samples</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  The set of samples within which we calculate
the diversity. If None, calculate diversity within the entire sample.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The pairwise nucleotide site diversity.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)">float</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
</section>
<section id="simple-container-classes">
<h3>Simple container classes<a class="headerlink" href="#simple-container-classes" title="Link to this heading">#</a></h3>
<section id="the-individual-class">
<h4>The <a class="reference internal" href="#tskit.Individual" title="tskit.Individual"><code class="xref py py-class docutils literal notranslate"><span class="pre">Individual</span></code></a> class<a class="headerlink" href="#the-individual-class" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.Individual">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">Individual</span></span><a class="reference internal" href="_modules/tskit/trees.html#Individual"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Individual" title="Link to this definition">#</a></dt>
<dd><p>An <a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">individual</span></a> in a tree sequence.
Since nodes correspond to genomes, individuals are associated with a collection
of nodes (e.g., two nodes per diploid). See <a class="reference internal" href="glossary.html#sec-nodes-or-individuals"><span class="std std-ref">Nodes, Genomes, or Individuals?</span></a>
for more discussion of this distinction.</p>
<p>Modifying the attributes in this class will have <strong>no effect</strong> on the
underlying tree sequence data.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Individual.id">
<span class="sig-name descname"><span class="pre">id</span></span><a class="headerlink" href="#tskit.Individual.id" title="Link to this definition">#</a></dt>
<dd><p>The integer ID of this individual. Varies from 0 to
<a class="reference internal" href="#tskit.TreeSequence.num_individuals" title="tskit.TreeSequence.num_individuals"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.num_individuals</span></code></a> - 1.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Individual.flags">
<span class="sig-name descname"><span class="pre">flags</span></span><a class="headerlink" href="#tskit.Individual.flags" title="Link to this definition">#</a></dt>
<dd><p>The bitwise flags for this individual.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Individual.location">
<span class="sig-name descname"><span class="pre">location</span></span><a class="headerlink" href="#tskit.Individual.location" title="Link to this definition">#</a></dt>
<dd><p>The spatial location of this individual as a numpy array. The location is an empty
array if no spatial location is defined.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Individual.parents">
<span class="sig-name descname"><span class="pre">parents</span></span><a class="headerlink" href="#tskit.Individual.parents" title="Link to this definition">#</a></dt>
<dd><p>The parent individual ids of this individual as a numpy array. The parents is an
empty array if no parents are defined.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Individual.nodes">
<span class="sig-name descname"><span class="pre">nodes</span></span><a class="headerlink" href="#tskit.Individual.nodes" title="Link to this definition">#</a></dt>
<dd><p>The IDs of the nodes that are associated with this individual as
a numpy array (dtype=np.int32). If no nodes are associated with the
individual this array will be empty.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Individual.metadata">
<span class="sig-name descname"><span class="pre">metadata</span></span><a class="headerlink" href="#tskit.Individual.metadata" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="data-model.html#sec-metadata-definition"><span class="std std-ref">metadata</span></a>
for this individual, decoded if a schema applies.</p>
</dd></dl>

</dd></dl>

</section>
<section id="the-node-class">
<h4>The <a class="reference internal" href="#tskit.Node" title="tskit.Node"><code class="xref py py-class docutils literal notranslate"><span class="pre">Node</span></code></a> class<a class="headerlink" href="#the-node-class" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.Node">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">Node</span></span><a class="reference internal" href="_modules/tskit/trees.html#Node"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Node" title="Link to this definition">#</a></dt>
<dd><p>A <a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">node</span></a> in a tree sequence, corresponding
to a single genome. The <code class="docutils literal notranslate"><span class="pre">time</span></code> and <code class="docutils literal notranslate"><span class="pre">population</span></code> are attributes of the
<code class="docutils literal notranslate"><span class="pre">Node</span></code>, rather than the <code class="docutils literal notranslate"><span class="pre">Individual</span></code>, as discussed in
<a class="reference internal" href="glossary.html#sec-nodes-or-individuals"><span class="std std-ref">Nodes, Genomes, or Individuals?</span></a>.</p>
<p>Modifying the attributes in this class will have <strong>no effect</strong> on the
underlying tree sequence data.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Node.id">
<span class="sig-name descname"><span class="pre">id</span></span><a class="headerlink" href="#tskit.Node.id" title="Link to this definition">#</a></dt>
<dd><p>The integer ID of this node. Varies from 0 to <a class="reference internal" href="#tskit.TreeSequence.num_nodes" title="tskit.TreeSequence.num_nodes"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.num_nodes</span></code></a> - 1.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Node.flags">
<span class="sig-name descname"><span class="pre">flags</span></span><a class="headerlink" href="#tskit.Node.flags" title="Link to this definition">#</a></dt>
<dd><p>The bitwise flags for this node.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Node.time">
<span class="sig-name descname"><span class="pre">time</span></span><a class="headerlink" href="#tskit.Node.time" title="Link to this definition">#</a></dt>
<dd><p>The birth time of this node.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Node.population">
<span class="sig-name descname"><span class="pre">population</span></span><a class="headerlink" href="#tskit.Node.population" title="Link to this definition">#</a></dt>
<dd><p>The integer ID of the population that this node was born in.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Node.individual">
<span class="sig-name descname"><span class="pre">individual</span></span><a class="headerlink" href="#tskit.Node.individual" title="Link to this definition">#</a></dt>
<dd><p>The integer ID of the individual that this node was a part of.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Node.metadata">
<span class="sig-name descname"><span class="pre">metadata</span></span><a class="headerlink" href="#tskit.Node.metadata" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="data-model.html#sec-metadata-definition"><span class="std std-ref">metadata</span></a> for this node, decoded if a schema
applies.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Node.is_sample">
<span class="sig-name descname"><span class="pre">is_sample</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Node.is_sample"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Node.is_sample" title="Link to this definition">#</a></dt>
<dd><p>Returns True if this node is a sample. This value is derived from the
<code class="docutils literal notranslate"><span class="pre">flag</span></code> variable.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="the-edge-class">
<h4>The <a class="reference internal" href="#tskit.Edge" title="tskit.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> class<a class="headerlink" href="#the-edge-class" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.Edge">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">Edge</span></span><a class="reference internal" href="_modules/tskit/trees.html#Edge"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Edge" title="Link to this definition">#</a></dt>
<dd><p>An <a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">edge</span></a> in a tree sequence.</p>
<p>Modifying the attributes in this class will have <strong>no effect</strong> on the
underlying tree sequence data.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Edge.id">
<span class="sig-name descname"><span class="pre">id</span></span><a class="headerlink" href="#tskit.Edge.id" title="Link to this definition">#</a></dt>
<dd><p>The integer ID of this edge. Varies from 0 to
<a class="reference internal" href="#tskit.TreeSequence.num_edges" title="tskit.TreeSequence.num_edges"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.num_edges</span></code></a> - 1.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Edge.left">
<span class="sig-name descname"><span class="pre">left</span></span><a class="headerlink" href="#tskit.Edge.left" title="Link to this definition">#</a></dt>
<dd><p>The left coordinate of this edge.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Edge.right">
<span class="sig-name descname"><span class="pre">right</span></span><a class="headerlink" href="#tskit.Edge.right" title="Link to this definition">#</a></dt>
<dd><p>The right coordinate of this edge.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Edge.parent">
<span class="sig-name descname"><span class="pre">parent</span></span><a class="headerlink" href="#tskit.Edge.parent" title="Link to this definition">#</a></dt>
<dd><p>The integer ID of the parent node for this edge.
To obtain further information about a node with a given ID, use
<a class="reference internal" href="#tskit.TreeSequence.node" title="tskit.TreeSequence.node"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.node()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Edge.child">
<span class="sig-name descname"><span class="pre">child</span></span><a class="headerlink" href="#tskit.Edge.child" title="Link to this definition">#</a></dt>
<dd><p>The integer ID of the child node for this edge.
To obtain further information about a node with a given ID, use
<a class="reference internal" href="#tskit.TreeSequence.node" title="tskit.TreeSequence.node"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.node()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Edge.metadata">
<span class="sig-name descname"><span class="pre">metadata</span></span><a class="headerlink" href="#tskit.Edge.metadata" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="data-model.html#sec-metadata-definition"><span class="std std-ref">metadata</span></a> for this edge, decoded if a schema
applies.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Edge.span">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">span</span></span><a class="headerlink" href="#tskit.Edge.span" title="Link to this definition">#</a></dt>
<dd><p>Returns the span of this edge, i.e., the right position minus the left position</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The span of this edge.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Edge.interval">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">interval</span></span><a class="headerlink" href="#tskit.Edge.interval" title="Link to this definition">#</a></dt>
<dd><p>Returns the left and right positions of this edge as an <a class="reference internal" href="#tskit.Interval" title="tskit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The interval covered by this edge.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.Interval" title="tskit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="the-site-class">
<h4>The <a class="reference internal" href="#tskit.Site" title="tskit.Site"><code class="xref py py-class docutils literal notranslate"><span class="pre">Site</span></code></a> class<a class="headerlink" href="#the-site-class" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.Site">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">Site</span></span><a class="reference internal" href="_modules/tskit/trees.html#Site"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Site" title="Link to this definition">#</a></dt>
<dd><p>A <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">site</span></a> in a tree sequence.</p>
<p>Modifying the attributes in this class will have <strong>no effect</strong> on the
underlying tree sequence data.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Site.id">
<span class="sig-name descname"><span class="pre">id</span></span><a class="headerlink" href="#tskit.Site.id" title="Link to this definition">#</a></dt>
<dd><p>The integer ID of this site. Varies from 0 to <a class="reference internal" href="#tskit.TreeSequence.num_sites" title="tskit.TreeSequence.num_sites"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.num_sites</span></code></a> - 1.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Site.position">
<span class="sig-name descname"><span class="pre">position</span></span><a class="headerlink" href="#tskit.Site.position" title="Link to this definition">#</a></dt>
<dd><p>The floating point location of this site in genome coordinates.
Ranges from 0 (inclusive) to <a class="reference internal" href="#tskit.TreeSequence.sequence_length" title="tskit.TreeSequence.sequence_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.sequence_length</span></code></a> (exclusive).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Site.ancestral_state">
<span class="sig-name descname"><span class="pre">ancestral_state</span></span><a class="headerlink" href="#tskit.Site.ancestral_state" title="Link to this definition">#</a></dt>
<dd><p>The ancestral state at this site (i.e., the state inherited by nodes, unless
mutations occur).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Site.mutations">
<span class="sig-name descname"><span class="pre">mutations</span></span><a class="headerlink" href="#tskit.Site.mutations" title="Link to this definition">#</a></dt>
<dd><p>The list of mutations at this site. Mutations within a site are returned in the</p>
<p>order they are specified in the underlying <a class="reference internal" href="#tskit.MutationTable" title="tskit.MutationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutationTable</span></code></a>. For canonical
(i.e., valid) tables, this means ancestral mutations precede their descendants, so
older mutations (as defined by the canonical mutation ordering; see
<a class="reference internal" href="data-model.html#sec-mutation-requirements"><span class="std std-ref">Mutation requirements</span></a>) appear before younger ones.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Site.metadata">
<span class="sig-name descname"><span class="pre">metadata</span></span><a class="headerlink" href="#tskit.Site.metadata" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="data-model.html#sec-metadata-definition"><span class="std std-ref">metadata</span></a> for this site, decoded if a schema
applies.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Site.alleles">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">alleles</span></span><a class="headerlink" href="#tskit.Site.alleles" title="Link to this definition">#</a></dt>
<dd><p>Return the set of all the alleles defined at this site</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This deliberately returns an (unordered) <em>set</em> of the possible allelic
states (as defined by the sites ancestral allele and its associated
mutations). If you wish to obtain an (ordered) <em>list</em> of alleles, for
example to translate the numeric genotypes at a site into allelic states,
you should instead use <code class="docutils literal notranslate"><span class="pre">.alleles</span></code> attribute of the <a class="reference internal" href="#tskit.Variant" title="tskit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> class,
which unlike this attribute includes <code class="docutils literal notranslate"><span class="pre">None</span></code> as a state when there is
missing data at a site.</p>
</div>
</dd></dl>

</dd></dl>

</section>
<section id="the-mutation-class">
<h4>The <a class="reference internal" href="#tskit.Mutation" title="tskit.Mutation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mutation</span></code></a> class<a class="headerlink" href="#the-mutation-class" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.Mutation">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">Mutation</span></span><a class="reference internal" href="_modules/tskit/trees.html#Mutation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Mutation" title="Link to this definition">#</a></dt>
<dd><p>A <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">mutation</span></a> in a tree sequence.</p>
<p>Modifying the attributes in this class will have <strong>no effect</strong> on the
underlying tree sequence data.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Mutation.id">
<span class="sig-name descname"><span class="pre">id</span></span><a class="headerlink" href="#tskit.Mutation.id" title="Link to this definition">#</a></dt>
<dd><p>The integer ID of this mutation. Varies from 0 to
<a class="reference internal" href="#tskit.TreeSequence.num_mutations" title="tskit.TreeSequence.num_mutations"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.num_mutations</span></code></a> - 1.</p>
<p>Modifying the attributes in this class will have <strong>no effect</strong> on the
underlying tree sequence data.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Mutation.site">
<span class="sig-name descname"><span class="pre">site</span></span><a class="headerlink" href="#tskit.Mutation.site" title="Link to this definition">#</a></dt>
<dd><p>The integer ID of the site that this mutation occurs at. To obtain
further information about a site with a given ID use <a class="reference internal" href="#tskit.TreeSequence.site" title="tskit.TreeSequence.site"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.site()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Mutation.node">
<span class="sig-name descname"><span class="pre">node</span></span><a class="headerlink" href="#tskit.Mutation.node" title="Link to this definition">#</a></dt>
<dd><p>The integer ID of the first node that inherits this mutation.
To obtain further information about a node with a given ID, use
<a class="reference internal" href="#tskit.TreeSequence.node" title="tskit.TreeSequence.node"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.node()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Mutation.time">
<span class="sig-name descname"><span class="pre">time</span></span><a class="headerlink" href="#tskit.Mutation.time" title="Link to this definition">#</a></dt>
<dd><p>The occurrence time of this mutation.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Mutation.derived_state">
<span class="sig-name descname"><span class="pre">derived_state</span></span><a class="headerlink" href="#tskit.Mutation.derived_state" title="Link to this definition">#</a></dt>
<dd><p>The derived state for this mutation. This is the state
inherited by nodes in the subtree rooted at this mutations node, unless
another mutation occurs.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Mutation.parent">
<span class="sig-name descname"><span class="pre">parent</span></span><a class="headerlink" href="#tskit.Mutation.parent" title="Link to this definition">#</a></dt>
<dd><p>The integer ID of this mutations parent mutation. When multiple
mutations occur at a site along a path in the tree, mutations must
record the mutation that is immediately above them. If the mutation does
not have a parent, this is equal to the <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">NULL</span></code></a> (-1).
To obtain further information about a mutation with a given ID, use
<a class="reference internal" href="#tskit.TreeSequence.mutation" title="tskit.TreeSequence.mutation"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.mutation()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Mutation.metadata">
<span class="sig-name descname"><span class="pre">metadata</span></span><a class="headerlink" href="#tskit.Mutation.metadata" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="data-model.html#sec-metadata-definition"><span class="std std-ref">metadata</span></a> for this mutation, decoded if a schema
applies.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Mutation.edge">
<span class="sig-name descname"><span class="pre">edge</span></span><a class="headerlink" href="#tskit.Mutation.edge" title="Link to this definition">#</a></dt>
<dd><p>The ID of the edge that this mutation is on.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Mutation.inherited_state">
<span class="sig-name descname"><span class="pre">inherited_state</span></span><a class="headerlink" href="#tskit.Mutation.inherited_state" title="Link to this definition">#</a></dt>
<dd><p>The inherited state for this mutation. This is the state that existed at the site
before this mutation occurred. This is either the ancestral state of the site
(if the mutation has no parent) or the derived state of the mutations
parent mutation (if it has a parent).</p>
</dd></dl>

</dd></dl>

</section>
<section id="the-variant-class">
<h4>The <a class="reference internal" href="#tskit.Variant" title="tskit.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a> class<a class="headerlink" href="#the-variant-class" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.Variant">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">Variant</span></span><a class="reference internal" href="_modules/tskit/genotypes.html#Variant"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Variant" title="Link to this definition">#</a></dt>
<dd><p>A variant in a tree sequence, describing the observed genetic variation
among the specified nodes (by default, the sample nodes) for a given site.
A variant consists of (a) a tuple of <strong>alleles</strong> listing the potential
allelic states which the requested nodes at this site can possess; (b) an
array of <strong>genotypes</strong> mapping node IDs to the observed alleles; (c) a
reference to the <a class="reference internal" href="#tskit.Site" title="tskit.Site"><code class="xref py py-class docutils literal notranslate"><span class="pre">Site</span></code></a> at which the Variant has been decoded; and
(d) an array of <strong>samples</strong> giving the node ID to which each element of the
genotypes array corresponds.</p>
<p>After creation a Variant is not yet decoded, and has no genotypes.
To decode a Variant, call the <a class="reference internal" href="#tskit.Variant.decode" title="tskit.Variant.decode"><code class="xref py py-meth docutils literal notranslate"><span class="pre">decode()</span></code></a> method. The Variant class will then
use a Tree, internal to the Variant, to seek to the position of the site and
decode the genotypes at that site. It is therefore much more efficient to visit
sites in sequential genomic order, either in a forwards or backwards direction,
than to do so randomly.</p>
<p>Each element in the <code class="docutils literal notranslate"><span class="pre">alleles</span></code> tuple is a string, representing an
observed allelic state that may be seen at this site. The <code class="docutils literal notranslate"><span class="pre">alleles</span></code> tuple,
which is guaranteed not to contain any duplicates, is generated in one of two
ways. The first (and default) way is for <code class="docutils literal notranslate"><span class="pre">tskit</span></code> to generate the encoding on
the fly while generating genotypes. In this case, the first element of this
tuple is guaranteed to be the same as the sites <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> value.
Note that allelic values may be listed that are not referred to by any
samples. For example, if we have a site that is fixed for the derived state
(i.e., we have a mutation over the tree root), all genotypes will be 1, but
the alleles list will be equal to <code class="docutils literal notranslate"><span class="pre">('0',</span> <span class="pre">'1')</span></code>. Other than the
ancestral state being the first allele, the alleles are listed in
no particular order, and the ordering should not be relied upon
(but see the notes on missing data below).</p>
<p>The second way is for the user to define the mapping between
genotype values and allelic state strings using the
<code class="docutils literal notranslate"><span class="pre">alleles</span></code> parameter to the <a class="reference internal" href="#tskit.TreeSequence.variants" title="tskit.TreeSequence.variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.variants()</span></code></a> method.
In this case, there is no indication of which allele is the ancestral state,
as the ordering is determined by the user.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">genotypes</span></code> represent the observed allelic states for each requested
node, such that <code class="docutils literal notranslate"><span class="pre">var.alleles[var.genotypes[j]]</span></code> gives the string allele
for the node at index <code class="docutils literal notranslate"><span class="pre">j</span></code> (i.e., for <code class="docutils literal notranslate"><span class="pre">variant.samples[j]</span></code>). Thus, the
elements of the genotypes array are
indexes into the <code class="docutils literal notranslate"><span class="pre">alleles</span></code> list. The genotypes are provided in this
way via a numpy numeric array to enable efficient calculations. To obtain a
(less efficient) array of allele strings for each node, you can use e.g.
<code class="docutils literal notranslate"><span class="pre">np.asarray(variant.alleles)[variant.genotypes]</span></code>.</p>
<p>When <a class="reference internal" href="data-model.html#sec-data-model-missing-data"><span class="std std-ref">missing data</span></a> is present at a given
site, the property <code class="docutils literal notranslate"><span class="pre">has_missing_data</span></code> will be True, at least one element
of the <code class="docutils literal notranslate"><span class="pre">genotypes</span></code> array will be equal to <code class="docutils literal notranslate"><span class="pre">tskit.MISSING_DATA</span></code>, and the
last element of the <code class="docutils literal notranslate"><span class="pre">alleles</span></code> array will be <code class="docutils literal notranslate"><span class="pre">None</span></code>. Note that in this
case <code class="docutils literal notranslate"><span class="pre">variant.num_alleles</span></code> will <strong>not</strong> be equal to
<code class="docutils literal notranslate"><span class="pre">len(variant.alleles)</span></code>. The rationale for adding <code class="docutils literal notranslate"><span class="pre">None</span></code> to the end of
the <code class="docutils literal notranslate"><span class="pre">alleles</span></code> list is to help code that does not handle missing data
correctly fail early rather than introducing subtle and hard-to-find bugs.
As <code class="docutils literal notranslate"><span class="pre">tskit.MISSING_DATA</span></code> is equal to -1, code that decodes genotypes into
allelic values without taking missing data into account would otherwise
incorrectly output the last allele in the list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree_sequence</strong> (<a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><em>TreeSequence</em></a>)  The tree sequence to which this variant
belongs.</p></li>
<li><p><strong>samples</strong> (<em>array_like</em>)  An array of node IDs for which to generate
genotypes, or <code class="docutils literal notranslate"><span class="pre">None</span></code> for all sample nodes. Non-sample nodes may also
be provided to generate genotypes for internal nodes. Default: <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>isolated_as_missing</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, the genotype value assigned to
isolated nodes without mutations (samples or non-samples) is
<a class="reference internal" href="#tskit.MISSING_DATA" title="tskit.MISSING_DATA"><code class="xref py py-data docutils literal notranslate"><span class="pre">MISSING_DATA</span></code></a> (-1). If False, such nodes will be
assigned the allele index for the ancestral state.
Default: True.</p></li>
<li><p><strong>alleles</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.14)"><em>tuple</em></a>)  A tuple of strings defining the encoding of
alleles as integer genotype values. At least one allele must be provided.
If duplicate alleles are provided, output genotypes will always be
encoded as the first occurrence of the allele. If None (the default),
the alleles are encoded as they are encountered during genotype
generation.</p></li>
</ul>
</dd>
</dl>
<dl class="py property">
<dt class="sig sig-object py" id="tskit.Variant.site">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">site</span></span><a class="headerlink" href="#tskit.Variant.site" title="Link to this definition">#</a></dt>
<dd><p>The Site object for the site at which this variant has been decoded.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Variant.alleles">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">alleles</span></span><a class="headerlink" href="#tskit.Variant.alleles" title="Link to this definition">#</a></dt>
<dd><p>A tuple of the allelic values which nodes can possess at the current
site. Unless an encoding of alleles is specified when creating this
variant instance, the first element of this tuple is always the sites
ancestral state.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Variant.samples">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">samples</span></span><a class="headerlink" href="#tskit.Variant.samples" title="Link to this definition">#</a></dt>
<dd><p>A numpy array of the node ids whose genotypes will be returned
by the <a class="reference internal" href="#tskit.Variant.genotypes" title="tskit.Variant.genotypes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">genotypes()</span></code></a> method.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Variant.genotypes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">genotypes</span></span><a class="headerlink" href="#tskit.Variant.genotypes" title="Link to this definition">#</a></dt>
<dd><p>An array of indexes into the list <code class="docutils literal notranslate"><span class="pre">alleles</span></code>, giving the
state of each requested node at the current site.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Variant.isolated_as_missing">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">isolated_as_missing</span></span><a class="headerlink" href="#tskit.Variant.isolated_as_missing" title="Link to this definition">#</a></dt>
<dd><p>True if isolated nodes are decoded to missing data. If False, isolated
nodes are decoded to the ancestral state.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Variant.has_missing_data">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">has_missing_data</span></span><a class="headerlink" href="#tskit.Variant.has_missing_data" title="Link to this definition">#</a></dt>
<dd><p>True if there is missing data for any of the
requested nodes at the current site.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Variant.num_missing">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_missing</span></span><a class="headerlink" href="#tskit.Variant.num_missing" title="Link to this definition">#</a></dt>
<dd><p>The number of requested nodes with missing data at this site.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Variant.num_alleles">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_alleles</span></span><a class="headerlink" href="#tskit.Variant.num_alleles" title="Link to this definition">#</a></dt>
<dd><p>The number of distinct alleles at this site. Note that this may
not be the same as the number of distinct values in the genotypes
array: firstly missing data is not counted as an allele, and secondly,
the site may contain mutations to alternative allele states (which are
counted in the number of alleles) without the mutation being inherited
by any of the requested nodes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Variant.decode">
<span class="sig-name descname"><span class="pre">decode</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">site_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/genotypes.html#Variant.decode"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Variant.decode" title="Link to this definition">#</a></dt>
<dd><p>Decode the variant at the given site, setting the site ID, genotypes and
alleles to those of the site and samples of this Variant.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>site_id</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The ID of the site to decode. This must be a valid site ID.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.14)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Variant.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/genotypes.html#Variant.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Variant.copy" title="Link to this definition">#</a></dt>
<dd><p>Create a copy of this Variant. Note that calling <a class="reference internal" href="#tskit.Variant.decode" title="tskit.Variant.decode"><code class="xref py py-meth docutils literal notranslate"><span class="pre">decode()</span></code></a> on the
copy will fail as it does not take a copy of the internal tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#tskit.Variant" title="tskit.genotypes.Variant"><code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The copy of this Variant.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Variant.states">
<span class="sig-name descname"><span class="pre">states</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">missing_data_string</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/genotypes.html#Variant.states"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Variant.states" title="Link to this definition">#</a></dt>
<dd><p>Returns the allelic states at this site as an array of strings.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Using this method is inefficient compared to working with the
underlying integer representation of genotypes as returned by
the <a class="reference internal" href="#tskit.Variant.genotypes" title="tskit.Variant.genotypes"><code class="xref py py-attr docutils literal notranslate"><span class="pre">genotypes</span></code></a> property.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>missing_data_string</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string that will be used to represent missing
data. If any normal allele contains this character, an error is raised.
Default: <cite>None</cite>, treated as <cite>N</cite>.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a></span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An numpy array of strings of length <code class="docutils literal notranslate"><span class="pre">num_sites</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Variant.counts">
<span class="sig-name descname"><span class="pre">counts</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/genotypes.html#Variant.counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Variant.counts" title="Link to this definition">#</a></dt>
<dd><p>Returns a <a class="reference external" href="https://docs.python.org/3/library/collections.html#collections.Counter" title="(in Python v3.14)"><code class="docutils literal notranslate"><span class="pre">collections.Counter</span></code></a> object providing counts for each
possible <a class="reference internal" href="#tskit.Variant.alleles" title="tskit.Variant.alleles"><code class="xref py py-attr docutils literal notranslate"><span class="pre">allele</span></code></a> at this site: i.e. the number of
samples possessing that allele among the set of samples specified when creating
this Variant (by default, this is all the sample nodes in the tree sequence).
Missing data is represented by an allelic state of <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Counter" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Counter</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a> | <a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.14)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A counter of the number of samples associated with each allele.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.Variant.frequencies">
<span class="sig-name descname"><span class="pre">frequencies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">remove_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/genotypes.html#Variant.frequencies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Variant.frequencies" title="Link to this definition">#</a></dt>
<dd><p>Return a dictionary mapping each possible <a class="reference internal" href="#tskit.Variant.alleles" title="tskit.Variant.alleles"><code class="xref py py-attr docutils literal notranslate"><span class="pre">allele</span></code></a>
at this site to the frequency of that allele: i.e. the number of samples
with that allele divided by the total number of samples, among the set of
samples specified when creating this Variant (by default, this is all the
sample nodes in the tree sequence). Note, therefore, that if a restricted set
of samples was specified on creation, the allele frequencies returned here
will <em>not</em> be the global allele frequencies in the whole tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>remove_missing</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, only samples with non-missing data will
be counted in the total number of samples used to calculate the frequency,
and no information on the frequency of missing data is returned. Otherwise
(default), samples with missing data are included when calculating
frequencies.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>]</span></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A dictionary mapping allelic states to the frequency of each allele
among the samples</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="the-migration-class">
<h4>The <a class="reference internal" href="#tskit.Migration" title="tskit.Migration"><code class="xref py py-class docutils literal notranslate"><span class="pre">Migration</span></code></a> class<a class="headerlink" href="#the-migration-class" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.Migration">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">Migration</span></span><a class="reference internal" href="_modules/tskit/trees.html#Migration"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Migration" title="Link to this definition">#</a></dt>
<dd><p>A <a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">migration</span></a> in a tree sequence.</p>
<p>Modifying the attributes in this class will have <strong>no effect</strong> on the
underlying tree sequence data.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Migration.left">
<span class="sig-name descname"><span class="pre">left</span></span><a class="headerlink" href="#tskit.Migration.left" title="Link to this definition">#</a></dt>
<dd><p>The left end of the genomic interval covered by this
migration (inclusive).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Migration.right">
<span class="sig-name descname"><span class="pre">right</span></span><a class="headerlink" href="#tskit.Migration.right" title="Link to this definition">#</a></dt>
<dd><p>The right end of the genomic interval covered by this migration
(exclusive).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Migration.node">
<span class="sig-name descname"><span class="pre">node</span></span><a class="headerlink" href="#tskit.Migration.node" title="Link to this definition">#</a></dt>
<dd><p>The integer ID of the node involved in this migration event.
To obtain further information about a node with a given ID, use
<a class="reference internal" href="#tskit.TreeSequence.node" title="tskit.TreeSequence.node"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.node()</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Migration.source">
<span class="sig-name descname"><span class="pre">source</span></span><a class="headerlink" href="#tskit.Migration.source" title="Link to this definition">#</a></dt>
<dd><p>The source population ID.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Migration.dest">
<span class="sig-name descname"><span class="pre">dest</span></span><a class="headerlink" href="#tskit.Migration.dest" title="Link to this definition">#</a></dt>
<dd><p>The destination population ID.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Migration.time">
<span class="sig-name descname"><span class="pre">time</span></span><a class="headerlink" href="#tskit.Migration.time" title="Link to this definition">#</a></dt>
<dd><p>The time at which this migration occurred at.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Migration.metadata">
<span class="sig-name descname"><span class="pre">metadata</span></span><a class="headerlink" href="#tskit.Migration.metadata" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="data-model.html#sec-metadata-definition"><span class="std std-ref">metadata</span></a> for this migration, decoded if a schema
applies.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Migration.id">
<span class="sig-name descname"><span class="pre">id</span></span><a class="headerlink" href="#tskit.Migration.id" title="Link to this definition">#</a></dt>
<dd><p>The integer ID of this migration. Varies from 0 to
<a class="reference internal" href="#tskit.TreeSequence.num_migrations" title="tskit.TreeSequence.num_migrations"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.num_migrations</span></code></a> - 1.</p>
</dd></dl>

</dd></dl>

</section>
<section id="the-population-class">
<h4>The <a class="reference internal" href="#tskit.Population" title="tskit.Population"><code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code></a> class<a class="headerlink" href="#the-population-class" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.Population">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">Population</span></span><a class="reference internal" href="_modules/tskit/trees.html#Population"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Population" title="Link to this definition">#</a></dt>
<dd><p>A <a class="reference internal" href="data-model.html#sec-population-table-definition"><span class="std std-ref">population</span></a> in a tree sequence.</p>
<p>Modifying the attributes in this class will have <strong>no effect</strong> on the
underlying tree sequence data.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Population.id">
<span class="sig-name descname"><span class="pre">id</span></span><a class="headerlink" href="#tskit.Population.id" title="Link to this definition">#</a></dt>
<dd><p>The integer ID of this population. Varies from 0 to
<a class="reference internal" href="#tskit.TreeSequence.num_populations" title="tskit.TreeSequence.num_populations"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.num_populations</span></code></a> - 1.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Population.metadata">
<span class="sig-name descname"><span class="pre">metadata</span></span><a class="headerlink" href="#tskit.Population.metadata" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="data-model.html#sec-metadata-definition"><span class="std std-ref">metadata</span></a> for this population, decoded if a
schema applies.</p>
</dd></dl>

</dd></dl>

</section>
<section id="the-provenance-class">
<h4>The <a class="reference internal" href="#tskit.Provenance" title="tskit.Provenance"><code class="xref py py-class docutils literal notranslate"><span class="pre">Provenance</span></code></a> class<a class="headerlink" href="#the-provenance-class" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.Provenance">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">Provenance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timestamp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#Provenance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Provenance" title="Link to this definition">#</a></dt>
<dd><p>A provenance entry in a tree sequence, detailing how this tree
sequence was generated, or subsequent operations on it (see <a class="reference internal" href="provenance.html#sec-provenance"><span class="std std-ref">Provenance</span></a>).</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Provenance.timestamp">
<span class="sig-name descname"><span class="pre">timestamp</span></span><a class="headerlink" href="#tskit.Provenance.timestamp" title="Link to this definition">#</a></dt>
<dd><p>The time that this entry was made</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Provenance.record">
<span class="sig-name descname"><span class="pre">record</span></span><a class="headerlink" href="#tskit.Provenance.record" title="Link to this definition">#</a></dt>
<dd><p>A JSON string giving details of the provenance (see <a class="reference internal" href="provenance.html#sec-provenance-example"><span class="std std-ref">Example</span></a>
for an example JSON string)</p>
</dd></dl>

</dd></dl>

</section>
<section id="the-interval-class">
<h4>The <a class="reference internal" href="#tskit.Interval" title="tskit.Interval"><code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code></a> class<a class="headerlink" href="#the-interval-class" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.Interval">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">Interval</span></span><a class="reference internal" href="_modules/tskit/trees.html#Interval"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Interval" title="Link to this definition">#</a></dt>
<dd><p>A tuple of 2 numbers, <code class="docutils literal notranslate"><span class="pre">[left,</span> <span class="pre">right)</span></code>, defining an interval over the genome.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Interval.left">
<span class="sig-name descname"><span class="pre">left</span></span><a class="headerlink" href="#tskit.Interval.left" title="Link to this definition">#</a></dt>
<dd><p>The left hand end of the interval. By convention this value is included
in the interval</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Interval.right">
<span class="sig-name descname"><span class="pre">right</span></span><a class="headerlink" href="#tskit.Interval.right" title="Link to this definition">#</a></dt>
<dd><p>The right hand end of the interval. By convention this value is <em>not</em>
included in the interval, i.e., the interval is half-open.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Interval.span">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">span</span></span><a class="headerlink" href="#tskit.Interval.span" title="Link to this definition">#</a></dt>
<dd><p>The span of the genome covered by this interval, simply <code class="docutils literal notranslate"><span class="pre">right-left</span></code>.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.Interval.mid">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mid</span></span><a class="headerlink" href="#tskit.Interval.mid" title="Link to this definition">#</a></dt>
<dd><p>The middle point of this interval, simply <code class="docutils literal notranslate"><span class="pre">left+(right-left)/2</span></code>.</p>
</dd></dl>

</dd></dl>

</section>
<section id="the-rank-class">
<h4>The <a class="reference internal" href="#tskit.Rank" title="tskit.Rank"><code class="xref py py-class docutils literal notranslate"><span class="pre">Rank</span></code></a> class<a class="headerlink" href="#the-rank-class" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.Rank">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">Rank</span></span><a class="reference internal" href="_modules/tskit/combinatorics.html#Rank"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.Rank" title="Link to this definition">#</a></dt>
<dd><p>A tuple of 2 numbers, <code class="docutils literal notranslate"><span class="pre">(shape,</span> <span class="pre">label)</span></code>, together defining a unique
topology for a labeled tree. See <a class="reference internal" href="topological-analysis.html#sec-combinatorics"><span class="std std-ref">Identifying and counting topologies</span></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Rank.shape">
<span class="sig-name descname"><span class="pre">shape</span></span><a class="headerlink" href="#tskit.Rank.shape" title="Link to this definition">#</a></dt>
<dd><p>A non-negative integer representing the (unlabelled) topology of a tree with a
defined number of tips.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.Rank.label">
<span class="sig-name descname"><span class="pre">label</span></span><a class="headerlink" href="#tskit.Rank.label" title="Link to this definition">#</a></dt>
<dd><p>A non-negative integer representing the order of labels for a given tree topology.</p>
</dd></dl>

</dd></dl>

</section>
</section>
<section id="tablecollection-and-table-classes">
<h3>TableCollection and Table classes<a class="headerlink" href="#tablecollection-and-table-classes" title="Link to this heading">#</a></h3>
<section id="the-tablecollection-class">
<h4>The <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a> class<a class="headerlink" href="#the-tablecollection-class" title="Link to this heading">#</a></h4>
<p>Also see the <a class="reference internal" href="#sec-tables-api-table-collection"><span class="std std-ref">TableCollection API</span></a> summary.</p>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.TableCollection">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">TableCollection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ll_tables</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection" title="Link to this definition">#</a></dt>
<dd><p>A collection of mutable tables defining a tree sequence. See the
<a class="reference internal" href="data-model.html#sec-data-model"><span class="std std-ref">Data model</span></a> section for definition on the various tables
and how they together define a <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>. Arbitrary
data can be stored in a TableCollection, but there are certain
<a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">requirements</span></a> that must be
satisfied for these tables to be interpreted as a tree sequence.</p>
<p>To obtain an immutable <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> instance corresponding to the
current state of a <code class="docutils literal notranslate"><span class="pre">TableCollection</span></code>, please use the <a class="reference internal" href="#tskit.TableCollection.tree_sequence" title="tskit.TableCollection.tree_sequence"><code class="xref py py-meth docutils literal notranslate"><span class="pre">tree_sequence()</span></code></a>
method.</p>
<dl class="py property">
<dt class="sig sig-object py" id="tskit.TableCollection.individuals">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">individuals</span></span><a class="headerlink" href="#tskit.TableCollection.individuals" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="data-model.html#sec-individual-table-definition"><span class="std std-ref">Individual Table</span></a> in this collection.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TableCollection.nodes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nodes</span></span><a class="headerlink" href="#tskit.TableCollection.nodes" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">Node Table</span></a> in this collection.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TableCollection.edges">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">edges</span></span><a class="headerlink" href="#tskit.TableCollection.edges" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">Edge Table</span></a> in this collection.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TableCollection.migrations">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">migrations</span></span><a class="headerlink" href="#tskit.TableCollection.migrations" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">Migration Table</span></a> in this collection</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TableCollection.sites">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sites</span></span><a class="headerlink" href="#tskit.TableCollection.sites" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">Site Table</span></a> in this collection.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TableCollection.mutations">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mutations</span></span><a class="headerlink" href="#tskit.TableCollection.mutations" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">Mutation Table</span></a> in this collection.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TableCollection.populations">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">populations</span></span><a class="headerlink" href="#tskit.TableCollection.populations" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="data-model.html#sec-population-table-definition"><span class="std std-ref">Population Table</span></a> in this collection.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TableCollection.provenances">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">provenances</span></span><a class="headerlink" href="#tskit.TableCollection.provenances" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="data-model.html#sec-provenance-table-definition"><span class="std std-ref">Provenance Table</span></a> in this collection.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TableCollection.indexes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">indexes</span></span><a class="headerlink" href="#tskit.TableCollection.indexes" title="Link to this definition">#</a></dt>
<dd><p>The edge insertion and removal indexes.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TableCollection.sequence_length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence_length</span></span><a class="headerlink" href="#tskit.TableCollection.sequence_length" title="Link to this definition">#</a></dt>
<dd><p>The sequence length defining the coordinate space.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TableCollection.file_uuid">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">file_uuid</span></span><a class="headerlink" href="#tskit.TableCollection.file_uuid" title="Link to this definition">#</a></dt>
<dd><p>The UUID for the file this TableCollection is derived
from, or None if not derived from a file.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TableCollection.time_units">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">time_units</span></span><a class="headerlink" href="#tskit.TableCollection.time_units" title="Link to this definition">#</a></dt>
<dd><p>The units used for the time dimension of this TableCollection</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.has_reference_sequence">
<span class="sig-name descname"><span class="pre">has_reference_sequence</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.has_reference_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.has_reference_sequence" title="Link to this definition">#</a></dt>
<dd><p>Returns True if this <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a> has an associated
<a class="reference internal" href="data-model.html#sec-data-model-reference-sequence"><span class="std std-ref">reference sequence</span></a>.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TableCollection.reference_sequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reference_sequence</span></span><a class="headerlink" href="#tskit.TableCollection.reference_sequence" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="#tskit.ReferenceSequence" title="tskit.ReferenceSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReferenceSequence</span></code></a> associated with this <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Note that the behaviour of this attribute differs from
<a class="reference internal" href="#tskit.TreeSequence.reference_sequence" title="tskit.TreeSequence.reference_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TreeSequence.reference_sequence</span></code></a> in that we return a valid
instance of <a class="reference internal" href="#tskit.ReferenceSequence" title="tskit.ReferenceSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReferenceSequence</span></code></a> even when
<a class="reference internal" href="#tskit.TableCollection.has_reference_sequence" title="tskit.TableCollection.has_reference_sequence"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TableCollection.has_reference_sequence</span></code></a> is False. This is
to allow us to update the state of the reference sequence.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">force_offset_64</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.asdict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.asdict" title="Link to this definition">#</a></dt>
<dd><p>Returns the nested dictionary representation of this TableCollection
used for interchange.</p>
<p>Note: the semantics of this method changed at tskit 0.1.0. Previously a
map of table names to the tables themselves was returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>force_offset_64</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, all offset columns will have dtype
np.uint64. If False (the default) the offset array columns will have
a dtype of either np.uint32 or np.uint64, depending on the size of the
corresponding data array.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The dictionary representation of this table collection.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)">dict</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TableCollection.table_name_map">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">table_name_map</span></span><a class="headerlink" href="#tskit.TableCollection.table_name_map" title="Link to this definition">#</a></dt>
<dd><p>Returns a dictionary mapping table names to the corresponding
table instances. For example, the returned dictionary will contain the
key edges that maps to an <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a> instance.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TableCollection.nbytes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nbytes</span></span><a class="headerlink" href="#tskit.TableCollection.nbytes" title="Link to this definition">#</a></dt>
<dd><p>Returns the total number of bytes required to store the data
in this table collection. Note that this may not be equal to
the actual memory footprint.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.equals">
<span class="sig-name descname"><span class="pre">equals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_ts_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_timestamps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_tables</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_reference_sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.equals"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.equals" title="Link to this definition">#</a></dt>
<dd><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal. By default, two table
collections are considered equal if their</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> properties are identical;</p></li>
<li><p>top-level tree sequence metadata and metadata schemas are
byte-wise identical;</p></li>
<li><p>constituent tables are byte-wise identical.</p></li>
</ul>
<p>Some of the requirements in this definition can be relaxed using the
parameters, which can be used to remove certain parts of the data model
from the comparison.</p>
<p>Table indexes are not considered in the equality comparison.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><em>TableCollection</em></a>)  Another table collection.</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True <em>all</em> metadata and metadata schemas
will be excluded from the comparison. This includes the top-level
tree sequence and constituent table metadata (default=False).</p></li>
<li><p><strong>ignore_ts_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True the top-level tree sequence
metadata and metadata schemas will be excluded from the comparison.
If <code class="docutils literal notranslate"><span class="pre">ignore_metadata</span></code> is True, this parameter has no effect.</p></li>
<li><p><strong>ignore_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True the provenance tables are
not included in the comparison.</p></li>
<li><p><strong>ignore_timestamps</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True the provenance timestamp column
is ignored in the comparison. If <code class="docutils literal notranslate"><span class="pre">ignore_provenance</span></code> is True, this
parameter has no effect.</p></li>
<li><p><strong>ignore_tables</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True no tables are included in the
comparison, thus comparing only the top-level information.</p></li>
<li><p><strong>ignore_reference_sequence</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True the reference sequence
is not included in the comparison.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if other is equal to this table collection; False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.assert_equals">
<span class="sig-name descname"><span class="pre">assert_equals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_ts_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_timestamps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_tables</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_reference_sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.assert_equals"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.assert_equals" title="Link to this definition">#</a></dt>
<dd><p>Raise an AssertionError for the first found difference between
this and another table collection. Note that table indexes are not checked.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong>  Another table collection (TableCollection or
ImmutableTableCollection).</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True <em>all</em> metadata and metadata schemas
will be excluded from the comparison. This includes the top-level
tree sequence and constituent table metadata (default=False).</p></li>
<li><p><strong>ignore_ts_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True the top-level tree sequence
metadata and metadata schemas will be excluded from the comparison.
If <code class="docutils literal notranslate"><span class="pre">ignore_metadata</span></code> is True, this parameter has no effect.</p></li>
<li><p><strong>ignore_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True the provenance tables are
not included in the comparison.</p></li>
<li><p><strong>ignore_timestamps</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True the provenance timestamp column
is ignored in the comparison. If <code class="docutils literal notranslate"><span class="pre">ignore_provenance</span></code> is True, this
parameter has no effect.</p></li>
<li><p><strong>ignore_tables</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True no tables are included in the
comparison, thus comparing only the top-level information.</p></li>
<li><p><strong>ignore_reference_sequence</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True the reference sequence
is not included in the comparison.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.dump">
<span class="sig-name descname"><span class="pre">dump</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_or_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.dump"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.dump" title="Link to this definition">#</a></dt>
<dd><p>Writes the table collection to the specified path or file object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>file_or_path</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The file object or path to write the TreeSequence to.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.copy" title="Link to this definition">#</a></dt>
<dd><p>Returns a deep copy of this TableCollection.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A deep copy of this TableCollection.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection">tskit.TableCollection</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.tree_sequence">
<span class="sig-name descname"><span class="pre">tree_sequence</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.tree_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.tree_sequence" title="Link to this definition">#</a></dt>
<dd><p>Returns a <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> instance from the tables defined in this
<a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>, building the required indexes if they have not yet
been created by <a class="reference internal" href="#tskit.TableCollection.build_index" title="tskit.TableCollection.build_index"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_index()</span></code></a>. If the table collection does not meet
the <a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">Valid tree sequence requirements</span></a>, for example if the tables
are not correctly sorted or if they cannot be interpreted as a tree sequence,
an exception is raised. Note that in the former case, the <a class="reference internal" href="#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">sort()</span></code></a>
method may be used to ensure that sorting requirements are met.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a> instance reflecting the structures
defined in this set of tables.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence">tskit.TreeSequence</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.simplify">
<span class="sig-name descname"><span class="pre">simplify</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reduce_to_site_topology</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_populations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_individuals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_nodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">update_sample_flags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_unary</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_unary_in_individuals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_input_roots</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_zero_mutation_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.simplify"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.simplify" title="Link to this definition">#</a></dt>
<dd><p>Simplifies the tables in place to retain only the information necessary
to reconstruct the tree sequence describing the given <code class="docutils literal notranslate"><span class="pre">samples</span></code>.
If <code class="docutils literal notranslate"><span class="pre">filter_nodes</span></code> is True (the default), this can change the ID of
the nodes, so that the node <code class="docutils literal notranslate"><span class="pre">samples[k]</span></code> will have ID <code class="docutils literal notranslate"><span class="pre">k</span></code> in the
result, resulting in a NodeTable where only the first <code class="docutils literal notranslate"><span class="pre">len(samples)</span></code>
nodes are marked as samples. The mapping from node IDs in the current
set of tables to their equivalent values in the simplified tables is
returned as a numpy array. If an array <code class="docutils literal notranslate"><span class="pre">a</span></code> is returned by this
function and <code class="docutils literal notranslate"><span class="pre">u</span></code> is the ID of a node in the input table, then
<code class="docutils literal notranslate"><span class="pre">a[u]</span></code> is the ID of this node in the output table. For any node <code class="docutils literal notranslate"><span class="pre">u</span></code>
that is not mapped into the output tables, this mapping will equal
<code class="docutils literal notranslate"><span class="pre">tskit.NULL</span></code> (<code class="docutils literal notranslate"><span class="pre">-1</span></code>).</p>
<p>Tables operated on by this function must: be sorted (see
<a class="reference internal" href="#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.sort()</span></code></a>), have children be born strictly after their
parents, and the intervals on which any node is a child must be
disjoint. Other than this the tables need not satisfy remaining
requirements to specify a valid tree sequence (but the resulting tables
will).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>To invert the returned <code class="docutils literal notranslate"><span class="pre">node_map</span></code>, that is, to obtain a reverse
mapping from the node ID in the output table to the node ID in
the input table, you can use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">rev_map</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">node_map</span><span class="p">,</span> <span class="n">shape</span><span class="o">=</span><span class="n">simplified_ts</span><span class="o">.</span><span class="n">num_nodes</span><span class="p">)</span>
<span class="n">kept</span> <span class="o">=</span> <span class="n">node_map</span> <span class="o">!=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span>
<span class="n">rev_map</span><span class="p">[</span><span class="n">node_map</span><span class="p">[</span><span class="n">kept</span><span class="p">]]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">node_map</span><span class="p">))[</span><span class="n">kept</span><span class="p">]</span>
</pre></div>
</div>
<p>In this case, no elements of the <code class="docutils literal notranslate"><span class="pre">rev_map</span></code> array will be set to
<code class="docutils literal notranslate"><span class="pre">tskit.NULL</span></code>.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>This is identical to <a class="reference internal" href="#tskit.TreeSequence.simplify" title="tskit.TreeSequence.simplify"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.simplify()</span></code></a> but acts <em>in place</em> to
alter the data in this <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>. Please see the
<a class="reference internal" href="#tskit.TreeSequence.simplify" title="tskit.TreeSequence.simplify"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.simplify()</span></code></a> method for a description of the remaining
parameters.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samples</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>]</em>)  A list of node IDs to retain as samples. They
need not be nodes marked as samples in the original tree sequence, but
will constitute the entire set of samples in the returned tree sequence.
If not specified or None, use all nodes marked with the IS_SAMPLE flag.
The list may be provided as a numpy array (or array-like) object
(dtype=np.int32).</p></li>
<li><p><strong>reduce_to_site_topology</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to reduce the topology down
to the trees that are present at sites. (Default: False).</p></li>
<li><p><strong>filter_populations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, remove any populations that are
not referenced by nodes after simplification; new population IDs are
allocated sequentially from zero. If False, the population table will
not be altered in any way. (Default: None, treated as True)</p></li>
<li><p><strong>filter_individuals</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, remove any individuals that are
not referenced by nodes after simplification; new individual IDs are
allocated sequentially from zero. If False, the individual table will
not be altered in any way. (Default: None, treated as True)</p></li>
<li><p><strong>filter_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, remove any sites that are
not referenced by mutations after simplification; new site IDs are
allocated sequentially from zero. If False, the site table will not
be altered in any way. (Default: None, treated as True)</p></li>
<li><p><strong>filter_nodes</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, remove any nodes that are
not referenced by edges after simplification. If False, the only
potential change to the node table may be to change the node flags
(if <code class="docutils literal notranslate"><span class="pre">samples</span></code> is specified and different from the existing samples).
(Default: None, treated as True)</p></li>
<li><p><strong>update_sample_flags</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, update node flags to so that
nodes in the specified list of samples have the NODE_IS_SAMPLE
flag after simplification, and nodes that are not in this list
do not. (Default: None, treated as True)</p></li>
<li><p><strong>keep_unary</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, preserve unary nodes (i.e. nodes with
exactly one child) that exist on the path from samples to root.
(Default: False)</p></li>
<li><p><strong>keep_unary_in_individuals</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, preserve unary nodes
that exist on the path from samples to root, but only if they are
associated with an individual in the individuals table. Cannot be
specified at the same time as <code class="docutils literal notranslate"><span class="pre">keep_unary</span></code>. (Default: <code class="docutils literal notranslate"><span class="pre">None</span></code>,
equivalent to False)</p></li>
<li><p><strong>keep_input_roots</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to retain history ancestral to the
MRCA of the samples. If <code class="docutils literal notranslate"><span class="pre">False</span></code>, no topology older than the MRCAs of the
samples will be included. If <code class="docutils literal notranslate"><span class="pre">True</span></code> the roots of all trees in the returned
tree sequence will be the same roots as in the original tree sequence.
(Default: False)</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, record details of this call to
simplify in the returned tree sequences provenance information
(Default: True).</p></li>
<li><p><strong>filter_zero_mutation_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Deprecated alias for <code class="docutils literal notranslate"><span class="pre">filter_sites</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A numpy array mapping node IDs in the input tables to their
corresponding node IDs in the output tables.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a> (dtype=np.int32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.link_ancestors">
<span class="sig-name descname"><span class="pre">link_ancestors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">samples</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ancestors</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.link_ancestors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.link_ancestors" title="Link to this definition">#</a></dt>
<dd><p>Returns an <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a> instance describing a subset of the genealogical
relationships between the nodes in <code class="docutils literal notranslate"><span class="pre">samples</span></code> and <code class="docutils literal notranslate"><span class="pre">ancestors</span></code>.</p>
<p>Each row <code class="docutils literal notranslate"><span class="pre">parent,</span> <span class="pre">child,</span> <span class="pre">left,</span> <span class="pre">right</span></code> in the output table indicates that
<code class="docutils literal notranslate"><span class="pre">child</span></code> has inherited the segment <code class="docutils literal notranslate"><span class="pre">[left,</span> <span class="pre">right)</span></code> from <code class="docutils literal notranslate"><span class="pre">parent</span></code> more
recently than from any other node in these lists.</p>
<p>In particular, suppose <code class="docutils literal notranslate"><span class="pre">samples</span></code> is a list of nodes such that <code class="docutils literal notranslate"><span class="pre">time</span></code> is 0
for each node, and <code class="docutils literal notranslate"><span class="pre">ancestors</span></code> is a list of nodes such that <code class="docutils literal notranslate"><span class="pre">time</span></code> is
greater than 0.0 for each node. Then each row of the output table will show
an interval <code class="docutils literal notranslate"><span class="pre">[left,</span> <span class="pre">right)</span></code> over which a node in <code class="docutils literal notranslate"><span class="pre">samples</span></code> has inherited
most recently from a node in <code class="docutils literal notranslate"><span class="pre">ancestors</span></code>, or an interval over which one of
these <code class="docutils literal notranslate"><span class="pre">ancestors</span></code> has inherited most recently from another node in
<code class="docutils literal notranslate"><span class="pre">ancestors</span></code>.</p>
<p>The following table shows which <code class="docutils literal notranslate"><span class="pre">parent-&gt;child</span></code> pairs will be shown in the
output of <code class="docutils literal notranslate"><span class="pre">link_ancestors</span></code>.
A node is a relevant descendant on a given interval if it also appears somewhere
in the <code class="docutils literal notranslate"><span class="pre">parent</span></code> column of the outputted table.</p>
<div class="pst-scrollable-table-container"><table class="table">
<tbody>
<tr class="row-odd"><td><p>Type of relationship</p></td>
<td><p>Shown in output of <code class="docutils literal notranslate"><span class="pre">link_ancestors</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">ancestor-&gt;sample</span></code></p></td>
<td><p>Always</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">ancestor1-&gt;ancestor2</span></code></p></td>
<td><p>Only if <code class="docutils literal notranslate"><span class="pre">ancestor2</span></code> has a relevant descendant</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">sample1-&gt;sample2</span></code></p></td>
<td><p>Always</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">sample-&gt;ancestor</span></code></p></td>
<td><p>Only if <code class="docutils literal notranslate"><span class="pre">ancestor</span></code> has a relevant descendant</p></td>
</tr>
</tbody>
</table>
</div>
<p>The difference between <code class="docutils literal notranslate"><span class="pre">samples</span></code> and <code class="docutils literal notranslate"><span class="pre">ancestors</span></code> is that information about
the ancestors of a node in <code class="docutils literal notranslate"><span class="pre">ancestors</span></code> will only be retained if it also has a
relevant descendant, while information about the ancestors of a node in
<code class="docutils literal notranslate"><span class="pre">samples</span></code> will always be retained.
The node IDs in <code class="docutils literal notranslate"><span class="pre">parent</span></code> and <code class="docutils literal notranslate"><span class="pre">child</span></code> refer to the IDs in the node table
of the inputted tree sequence.</p>
<p>The supplied nodes must be non-empty lists of the node IDs in the tree sequence:
in particular, they do not have to be <em>samples</em> of the tree sequence. The lists
of <code class="docutils literal notranslate"><span class="pre">samples</span></code> and <code class="docutils literal notranslate"><span class="pre">ancestors</span></code> may overlap, although adding a node from
<code class="docutils literal notranslate"><span class="pre">samples</span></code> to <code class="docutils literal notranslate"><span class="pre">ancestors</span></code> will not change the output. So, setting <code class="docutils literal notranslate"><span class="pre">samples</span></code>
and <code class="docutils literal notranslate"><span class="pre">ancestors</span></code> to the same list of nodes will find all genealogical
relationships within this list.</p>
<p>If none of the nodes in <code class="docutils literal notranslate"><span class="pre">ancestors</span></code> or <code class="docutils literal notranslate"><span class="pre">samples</span></code> are ancestral to <code class="docutils literal notranslate"><span class="pre">samples</span></code>
anywhere in the tree sequence, an empty table will be returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samples</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>]</em>)  A list of node IDs to retain as samples.</p></li>
<li><p><strong>ancestors</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>]</em>)  A list of node IDs to use as ancestors.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a> instance displaying relationships between
the <cite>samples</cite> and <cite>ancestors</cite>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.sort">
<span class="sig-name descname"><span class="pre">sort</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">edge_start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">site_start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation_start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.sort"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.sort" title="Link to this definition">#</a></dt>
<dd><p>Sorts the tables in place. This ensures that all tree sequence ordering
requirements listed in the
<a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">Valid tree sequence requirements</span></a> section are met, as long
as each site has at most one mutation (see below).</p>
<p>If the <code class="docutils literal notranslate"><span class="pre">edge_start</span></code> parameter is provided, this specifies the index
in the edge table where sorting should start. Only rows with index
greater than or equal to <code class="docutils literal notranslate"><span class="pre">edge_start</span></code> are sorted; rows before this index
are not affected. This parameter is provided to allow for efficient sorting
when the user knows that the edges up to a given index are already sorted.</p>
<p>If both <code class="docutils literal notranslate"><span class="pre">site_start</span></code> and <code class="docutils literal notranslate"><span class="pre">mutation_start</span></code> are equal to the number of rows
in their retrospective tables then neither is sorted. Note that a partial
non-sorting is not possible, and both or neither must be skipped.</p>
<p>The node, individual, population and provenance tables are not affected
by this method.</p>
<p>Edges are sorted as follows:</p>
<ul class="simple">
<li><p>time of parent, then</p></li>
<li><p>parent node ID, then</p></li>
<li><p>child node ID, then</p></li>
<li><p>left endpoint.</p></li>
</ul>
<p>Note that this sorting order exceeds the
<a class="reference internal" href="data-model.html#sec-edge-requirements"><span class="std std-ref">edge sorting requirements</span></a> for a valid
tree sequence. For a valid tree sequence, we require that all edges for a
given parent ID are adjacent, but we do not require that they be listed in
sorted order.</p>
<p>Sites are sorted by position, and sites with the same position retain
their relative ordering.</p>
<p>Mutations are sorted by site, then time (if known), then the mutations
nodes time, then number of descendant mutations (ensuring that parent
mutations occur before children), then node, then original order in the
tables.</p>
<p>Migrations are sorted by <code class="docutils literal notranslate"><span class="pre">time</span></code>, <code class="docutils literal notranslate"><span class="pre">source</span></code>, <code class="docutils literal notranslate"><span class="pre">dest</span></code>, <code class="docutils literal notranslate"><span class="pre">left</span></code> and
<code class="docutils literal notranslate"><span class="pre">node</span></code> values. This defines a total sort order, such that any permutation
of a valid migration table will be sorted into the same output order.
Note that this sorting order exceeds the
<a class="reference internal" href="data-model.html#sec-migration-requirements"><span class="std std-ref">migration sorting requirements</span></a> for a
valid tree sequence, which only requires that migrations are sorted by
time value.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>edge_start</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The index in the edge table where sorting starts
(default=0; must be &lt;= len(edges)).</p></li>
<li><p><strong>site_start</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The index in the site table where sorting starts
(default=0; must be one of [0, len(sites)]).</p></li>
<li><p><strong>mutation_start</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The index in the mutation table where sorting starts
(default=0; must be one of [0, len(mutations)]).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.sort_individuals">
<span class="sig-name descname"><span class="pre">sort_individuals</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.sort_individuals"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.sort_individuals" title="Link to this definition">#</a></dt>
<dd><p>Sorts the individual table in place, so that parents come before children,
and the parent column is remapped as required. Node references to individuals
are also updated. This is a stricter order than is required for a valid tree
sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.canonicalise">
<span class="sig-name descname"><span class="pre">canonicalise</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">remove_unreferenced</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.canonicalise"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.canonicalise" title="Link to this definition">#</a></dt>
<dd><p>This puts the tables in <em>canonical</em> form, imposing a stricter order on the
tables than <a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">required</span></a> for
a valid tree sequence. In particular, the population table is sorted to
place populations with the lowest node IDs first, and the individual table
is sorted firstly as in <a class="reference internal" href="#tskit.TableCollection.sort_individuals" title="tskit.TableCollection.sort_individuals"><code class="xref py py-meth docutils literal notranslate"><span class="pre">sort_individuals()</span></code></a> and secondarily
by the lowest ID of the nodes that refer to each individual
(see <a class="reference internal" href="#tskit.TreeSequence.subset" title="tskit.TreeSequence.subset"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.subset()</span></code></a>). The remaining tables are sorted
as in <a class="reference internal" href="#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">sort()</span></code></a>, with the modification that mutations are sorted by
site, then time (if known), then the mutations nodes time, then number
of descendant mutations (ensuring that parent mutations occur before
children), then node, then original order in the tables. This ensures
that any two tables with the same information
and node order should be identical after canonical sorting (note
that no canonical order exists for the node table).</p>
<p>By default, the method removes sites, individuals, and populations that
are not referenced (by mutations and nodes, respectively). If you wish
to keep these, pass <code class="docutils literal notranslate"><span class="pre">remove_unreferenced=False</span></code>, but note that
unreferenced individuals and populations are put at the end of the tables
in their original order.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">sort()</span></code></a> for sorting edges, mutations, and sites, and
<a class="reference internal" href="#tskit.TableCollection.subset" title="tskit.TableCollection.subset"><code class="xref py py-meth docutils literal notranslate"><span class="pre">subset()</span></code></a> for reordering nodes, individuals, and populations.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>remove_unreferenced</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to remove unreferenced sites,
individuals, and populations (default=True).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.compute_mutation_parents">
<span class="sig-name descname"><span class="pre">compute_mutation_parents</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.compute_mutation_parents"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.compute_mutation_parents" title="Link to this definition">#</a></dt>
<dd><p>Modifies the tables in place, computing the <code class="docutils literal notranslate"><span class="pre">parent</span></code> column of the
mutation table. For this to work, the node and edge tables must be
valid, and the site and mutation tables must be sorted (see
<a class="reference internal" href="#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.sort()</span></code></a>).  This will produce an error if mutations
are not sorted (i.e., if a mutation appears before its mutation parent)
<em>unless</em> the two mutations occur on the same branch, and have unknown times
in which case there is no way to detect the error.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">parent</span></code> of a given mutation is the ID of the next mutation
encountered traversing the tree upwards from that mutation, or
<code class="docutils literal notranslate"><span class="pre">NULL</span></code> if there is no such mutation.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.compute_mutation_times">
<span class="sig-name descname"><span class="pre">compute_mutation_times</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.compute_mutation_times"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.compute_mutation_times" title="Link to this definition">#</a></dt>
<dd><p>Modifies the tables in place, computing valid values for the <code class="docutils literal notranslate"><span class="pre">time</span></code> column of
the mutation table. For this to work, the node and edge tables must be
valid, and the site and mutation tables must be sorted and indexed(see
<a class="reference internal" href="#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.sort()</span></code></a> and <a class="reference internal" href="#tskit.TableCollection.build_index" title="tskit.TableCollection.build_index"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.build_index()</span></code></a>).</p>
<p>For a single mutation on an edge at a site, the <code class="docutils literal notranslate"><span class="pre">time</span></code> assigned to a mutation
by this method is the mid-point between the times of the nodes above and below
the mutation. In the case where there is more than one mutation on an edge for
a site, the times are evenly spread along the edge. For mutations that are
above a root node, the time of the root node is assigned.</p>
<p>The mutation table will be sorted if the new times mean that the original order
is no longer valid.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.deduplicate_sites">
<span class="sig-name descname"><span class="pre">deduplicate_sites</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.deduplicate_sites"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.deduplicate_sites" title="Link to this definition">#</a></dt>
<dd><p>Modifies the tables in place, removing entries in the site table with
duplicate <code class="docutils literal notranslate"><span class="pre">position</span></code> (and keeping only the <em>first</em> entry for each
site), and renumbering the <code class="docutils literal notranslate"><span class="pre">site</span></code> column of the mutation table
appropriately.  This requires the site table to be sorted by position.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method does not sort the tables afterwards, so
mutations may no longer be sorted by time.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.delete_sites">
<span class="sig-name descname"><span class="pre">delete_sites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">site_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.delete_sites"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.delete_sites" title="Link to this definition">#</a></dt>
<dd><p>Remove the specified sites entirely from the sites and mutations tables in this
collection. This is identical to <a class="reference internal" href="#tskit.TreeSequence.delete_sites" title="tskit.TreeSequence.delete_sites"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.delete_sites()</span></code></a> but acts
<em>in place</em> to alter the data in this <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>site_ids</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em>]</em>)  A list of site IDs specifying the sites to remove.</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If <code class="docutils literal notranslate"><span class="pre">True</span></code>, add details of this operation
to the provenance table in this TableCollection. (Default: <code class="docutils literal notranslate"><span class="pre">True</span></code>).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.delete_intervals">
<span class="sig-name descname"><span class="pre">delete_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">simplify</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.delete_intervals"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.delete_intervals" title="Link to this definition">#</a></dt>
<dd><p>Delete all information from this set of tables which lies <em>within</em> the
specified list of genomic intervals. This is identical to
<a class="reference internal" href="#tskit.TreeSequence.delete_intervals" title="tskit.TreeSequence.delete_intervals"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.delete_intervals()</span></code></a> but acts <em>in place</em> to alter
the data in this <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> (<em>array_like</em>)  A list (start, end) pairs describing the
genomic intervals to delete. Intervals must be non-overlapping and
in increasing order. The list of intervals must be interpretable as a
2D numpy array with shape (N, 2), where N is the number of intervals.</p></li>
<li><p><strong>simplify</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, run simplify on the tables so that nodes
no longer used are discarded. (Default: True).</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If <code class="docutils literal notranslate"><span class="pre">True</span></code>, add details of this operation
to the provenance table in this TableCollection. (Default: <code class="docutils literal notranslate"><span class="pre">True</span></code>).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.keep_intervals">
<span class="sig-name descname"><span class="pre">keep_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">simplify</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.keep_intervals"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.keep_intervals" title="Link to this definition">#</a></dt>
<dd><p>Delete all information from this set of tables which lies <em>outside</em> the
specified list of genomic intervals. This is identical to
<a class="reference internal" href="#tskit.TreeSequence.keep_intervals" title="tskit.TreeSequence.keep_intervals"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.keep_intervals()</span></code></a> but acts <em>in place</em> to alter
the data in this <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> (<em>array_like</em>)  A list (start, end) pairs describing the
genomic intervals to keep. Intervals must be non-overlapping and
in increasing order. The list of intervals must be interpretable as a
2D numpy array with shape (N, 2), where N is the number of intervals.</p></li>
<li><p><strong>simplify</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, run simplify on the tables so that nodes
no longer used are discarded. Must be <code class="docutils literal notranslate"><span class="pre">False</span></code> if input tree sequence
includes migrations. (Default: True).</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If <code class="docutils literal notranslate"><span class="pre">True</span></code>, add details of this operation
to the provenance table in this TableCollection. (Default: <code class="docutils literal notranslate"><span class="pre">True</span></code>).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.ltrim">
<span class="sig-name descname"><span class="pre">ltrim</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.ltrim"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.ltrim" title="Link to this definition">#</a></dt>
<dd><p>Reset the coordinate system used in these tables, changing the left and right
genomic positions in the edge table such that the leftmost edge now starts at
position 0. This is identical to <a class="reference internal" href="#tskit.TreeSequence.ltrim" title="tskit.TreeSequence.ltrim"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.ltrim()</span></code></a> but acts <em>in place</em>
to alter the data in this <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If <code class="docutils literal notranslate"><span class="pre">True</span></code>, add details of this operation
to the provenance table in this TableCollection. (Default: <code class="docutils literal notranslate"><span class="pre">True</span></code>).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.rtrim">
<span class="sig-name descname"><span class="pre">rtrim</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.rtrim"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.rtrim" title="Link to this definition">#</a></dt>
<dd><p>Reset the <code class="docutils literal notranslate"><span class="pre">sequence_length</span></code> property so that the sequence ends at the end of
the last edge. This is identical to <a class="reference internal" href="#tskit.TreeSequence.rtrim" title="tskit.TreeSequence.rtrim"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.rtrim()</span></code></a> but acts
<em>in place</em> to alter the data in this <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If <code class="docutils literal notranslate"><span class="pre">True</span></code>, add details of this operation
to the provenance table in this TableCollection. (Default: <code class="docutils literal notranslate"><span class="pre">True</span></code>).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.trim">
<span class="sig-name descname"><span class="pre">trim</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.trim"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.trim" title="Link to this definition">#</a></dt>
<dd><p>Trim away any empty regions on the right and left of the tree sequence encoded by
these tables. This is identical to <a class="reference internal" href="#tskit.TreeSequence.trim" title="tskit.TreeSequence.trim"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.trim()</span></code></a> but acts <em>in place</em>
to alter the data in this <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If <code class="docutils literal notranslate"><span class="pre">True</span></code>, add details of this operation
to the provenance table in this TableCollection. (Default: <code class="docutils literal notranslate"><span class="pre">True</span></code>).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.shift">
<span class="sig-name descname"><span class="pre">shift</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.shift"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.shift" title="Link to this definition">#</a></dt>
<dd><p>Shift the coordinate system (used by edges, sites, and migrations) of this
TableCollection by a given value. This is identical to <a class="reference internal" href="#tskit.TreeSequence.shift" title="tskit.TreeSequence.shift"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.shift()</span></code></a>
but acts <em>in place</em> to alter the data in this <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>No attempt is made to check that the new coordinate system or sequence length
is valid: if you wish to do this, use {meth}`TreeSequence.shift` instead.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>value</strong>  The amount by which to shift the coordinate system.</p></li>
<li><p><strong>sequence_length</strong>  The new sequence length of the tree sequence. If
<code class="docutils literal notranslate"><span class="pre">None</span></code> (default) add <cite>value</cite> to the sequence length.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.delete_older">
<span class="sig-name descname"><span class="pre">delete_older</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">time</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.delete_older"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.delete_older" title="Link to this definition">#</a></dt>
<dd><p>Deletes edge, mutation and migration information at least as old as
the specified time.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>This method is similar to the higher-level
<a class="reference internal" href="#tskit.TreeSequence.decapitate" title="tskit.TreeSequence.decapitate"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.decapitate()</span></code></a> method, which also splits
edges that intersect with the given time.
<a class="reference internal" href="#tskit.TreeSequence.decapitate" title="tskit.TreeSequence.decapitate"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.decapitate()</span></code></a>
is more useful for most purposes, and may be what
you need instead of this method!</p>
</div>
<p>For the purposes of this method, an edge covers the times from the
child node up until the <em>parent</em> node, so that any any edge with parent
node time &gt; <code class="docutils literal notranslate"><span class="pre">time</span></code> will be removed.</p>
<p>Any mutation whose time is &gt;= <code class="docutils literal notranslate"><span class="pre">time</span></code> will be removed. A mutations time
is its associated <code class="docutils literal notranslate"><span class="pre">time</span></code> value, or the time of its node if the
mutations time was marked as unknown (<a class="reference internal" href="#tskit.UNKNOWN_TIME" title="tskit.UNKNOWN_TIME"><code class="xref py py-data docutils literal notranslate"><span class="pre">UNKNOWN_TIME</span></code></a>).</p>
<p>Any migration with time &gt;= <code class="docutils literal notranslate"><span class="pre">time</span></code> will be removed.</p>
<p>The node table is not affected by this operation.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This method does not have any specific sorting requirements
and will maintain mutation parent mappings.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The cutoff time.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">clear_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clear_metadata_schemas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clear_ts_metadata_and_schema</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.clear"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.clear" title="Link to this definition">#</a></dt>
<dd><p>Remove all rows of the data tables, optionally remove provenance, metadata
schemas and ts-level metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>clear_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If <code class="docutils literal notranslate"><span class="pre">True</span></code>, remove all rows of the provenance
table. (Default: <code class="docutils literal notranslate"><span class="pre">False</span></code>).</p></li>
<li><p><strong>clear_metadata_schemas</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If <code class="docutils literal notranslate"><span class="pre">True</span></code>, clear the table metadata
schemas. (Default: <code class="docutils literal notranslate"><span class="pre">False</span></code>).</p></li>
<li><p><strong>clear_ts_metadata_and_schema</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If <code class="docutils literal notranslate"><span class="pre">True</span></code>, clear the tree-sequence
level metadata and schema (Default: <code class="docutils literal notranslate"><span class="pre">False</span></code>).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.has_index">
<span class="sig-name descname"><span class="pre">has_index</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.has_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.has_index" title="Link to this definition">#</a></dt>
<dd><p>Returns True if this TableCollection is indexed. See <a class="reference internal" href="data-model.html#sec-table-indexes"><span class="std std-ref">Table indexes</span></a>
for information on indexes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.build_index">
<span class="sig-name descname"><span class="pre">build_index</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.build_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.build_index" title="Link to this definition">#</a></dt>
<dd><p>Builds an index on this TableCollection. Any existing indexes are automatically
dropped.  See <a class="reference internal" href="data-model.html#sec-table-indexes"><span class="std std-ref">Table indexes</span></a> for information on indexes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.drop_index">
<span class="sig-name descname"><span class="pre">drop_index</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.drop_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.drop_index" title="Link to this definition">#</a></dt>
<dd><p>Drops any indexes present on this table collection. If the tables are not
currently indexed this method has no effect.  See <a class="reference internal" href="data-model.html#sec-table-indexes"><span class="std std-ref">Table indexes</span></a>
for information on indexes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.subset">
<span class="sig-name descname"><span class="pre">subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nodes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reorder_populations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_unreferenced</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.subset"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.subset" title="Link to this definition">#</a></dt>
<dd><p>Modifies the tables in place to contain only the entries referring to
the provided list of node IDs, with nodes reordered according to the
order they appear in the list. Other tables are <a class="reference internal" href="#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">sorted</span></code></a>
to conform to the <a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">Valid tree sequence requirements</span></a>, and
additionally sorted as described in the documentation for the equivalent
tree sequence method <a class="reference internal" href="#tskit.TreeSequence.subset" title="tskit.TreeSequence.subset"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.subset()</span></code></a>: please see this for more
detail.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nodes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  The list of nodes for which to retain information. This
may be a numpy array (or array-like) object (dtype=np.int32).</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to record a provenance entry
in the provenance table for this operation.</p></li>
<li><p><strong>reorder_populations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to reorder the population table
(default: True).  If False, the population table will not be altered
in any way.</p></li>
<li><p><strong>remove_unreferenced</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether sites, individuals, and populations
that are not referred to by any retained entries in the tables should
be removed (default: True). See the description for details.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.union">
<span class="sig-name descname"><span class="pre">union</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_mapping</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_shared_equality</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_populations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">all_edges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">all_mutations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.union"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.union" title="Link to this definition">#</a></dt>
<dd><p>Modifies the table collection in place by adding the non-shared
portions of <code class="docutils literal notranslate"><span class="pre">other</span></code> to itself. To perform the node-wise union,
the method relies on a <code class="docutils literal notranslate"><span class="pre">node_mapping</span></code> array, that maps nodes in
<code class="docutils literal notranslate"><span class="pre">other</span></code> to its equivalent node in <code class="docutils literal notranslate"><span class="pre">self</span></code> or <code class="docutils literal notranslate"><span class="pre">tskit.NULL</span></code> if
the node is exclusive to <code class="docutils literal notranslate"><span class="pre">other</span></code>. See <a class="reference internal" href="#tskit.TreeSequence.union" title="tskit.TreeSequence.union"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.union()</span></code></a> for a more
detailed description.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><em>TableCollection</em></a>)  Another table collection.</p></li>
<li><p><strong>node_mapping</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  An array of node IDs that relate nodes in
<code class="docutils literal notranslate"><span class="pre">other</span></code> to nodes in <code class="docutils literal notranslate"><span class="pre">self</span></code>: the k-th element of <code class="docutils literal notranslate"><span class="pre">node_mapping</span></code>
should be the index of the equivalent node in <code class="docutils literal notranslate"><span class="pre">self</span></code>, or
<code class="docutils literal notranslate"><span class="pre">tskit.NULL</span></code> if the node is not present in <code class="docutils literal notranslate"><span class="pre">self</span></code> (in which case it
will be added to self).</p></li>
<li><p><strong>check_shared_equality</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, the shared portions of the
table collections will be checked for equality.</p></li>
<li><p><strong>add_populations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, nodes new to <code class="docutils literal notranslate"><span class="pre">self</span></code> will be
assigned new population IDs.</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  Whether to record a provenance entry
in the provenance table for this operation.</p></li>
<li><p><strong>all_edges</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, then all edges in <code class="docutils literal notranslate"><span class="pre">other</span></code> are added
to <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p></li>
<li><p><strong>all_mutations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True, then all mutations in <code class="docutils literal notranslate"><span class="pre">other</span></code> are added
to <code class="docutils literal notranslate"><span class="pre">self</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollection.ibd_segments">
<span class="sig-name descname"><span class="pre">ibd_segments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">within</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">between</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_time</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_span</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">store_pairs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">store_segments</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollection.ibd_segments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollection.ibd_segments" title="Link to this definition">#</a></dt>
<dd><p>Equivalent to the <a class="reference internal" href="#tskit.TreeSequence.ibd_segments" title="tskit.TreeSequence.ibd_segments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.ibd_segments()</span></code></a> method; please see its
documentation for more details, and use this method only if you specifically need
to work with a <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a> object.</p>
<p>This method has the same data requirements as
<a class="reference internal" href="#tskit.TableCollection.simplify" title="tskit.TableCollection.simplify"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.simplify()</span></code></a>. In particular, the tables in the collection
have <a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">required</span></a> sorting orders.
To enforce this, you can call <a class="reference internal" href="#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TableCollection.sort()</span></code></a> before using this
method. If the edge table contains any edges with identical
parents and children over adjacent genomic intervals, any IBD intervals
underneath the edges will also be split across the breakpoint(s). To prevent this
behaviour in this situation, use <a class="reference internal" href="#tskit.EdgeTable.squash" title="tskit.EdgeTable.squash"><code class="xref py py-meth docutils literal notranslate"><span class="pre">EdgeTable.squash()</span></code></a> beforehand.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>within</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  As for the <a class="reference internal" href="#tskit.TreeSequence.ibd_segments" title="tskit.TreeSequence.ibd_segments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.ibd_segments()</span></code></a> method.</p></li>
<li><p><strong>between</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>]</em>)  As for the <a class="reference internal" href="#tskit.TreeSequence.ibd_segments" title="tskit.TreeSequence.ibd_segments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.ibd_segments()</span></code></a> method.</p></li>
<li><p><strong>max_time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  As for the <a class="reference internal" href="#tskit.TreeSequence.ibd_segments" title="tskit.TreeSequence.ibd_segments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.ibd_segments()</span></code></a> method.</p></li>
<li><p><strong>min_span</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  As for the <a class="reference internal" href="#tskit.TreeSequence.ibd_segments" title="tskit.TreeSequence.ibd_segments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.ibd_segments()</span></code></a> method.</p></li>
<li><p><strong>store_pairs</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  As for the <a class="reference internal" href="#tskit.TreeSequence.ibd_segments" title="tskit.TreeSequence.ibd_segments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.ibd_segments()</span></code></a> method.</p></li>
<li><p><strong>store_segments</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  As for the <a class="reference internal" href="#tskit.TreeSequence.ibd_segments" title="tskit.TreeSequence.ibd_segments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.ibd_segments()</span></code></a> method.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An <a class="reference internal" href="#tskit.IdentitySegments" title="tskit.IdentitySegments"><code class="xref py py-class docutils literal notranslate"><span class="pre">IdentitySegments</span></code></a> object containing the recorded
IBD information.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tskit.IdentitySegments" title="tskit.IdentitySegments">IdentitySegments</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TableCollection.metadata">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">metadata</span></span><a class="headerlink" href="#tskit.TableCollection.metadata" title="Link to this definition">#</a></dt>
<dd><p>The decoded metadata for this object.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TableCollection.metadata_bytes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">metadata_bytes</span></span><a class="headerlink" href="#tskit.TableCollection.metadata_bytes" title="Link to this definition">#</a></dt>
<dd><p>The raw bytes of metadata for this TableCollection</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TableCollection.metadata_schema">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">metadata_schema</span></span><a class="headerlink" href="#tskit.TableCollection.metadata_schema" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this object.</p>
</dd></dl>

</dd></dl>

</section>
<section id="individualtable-classes">
<h4><a class="reference internal" href="#tskit.IndividualTable" title="tskit.IndividualTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">IndividualTable</span></code></a> classes<a class="headerlink" href="#individualtable-classes" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.IndividualTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">IndividualTable</span></span><a class="reference internal" href="_modules/tskit/tables.html#IndividualTable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.IndividualTable" title="Link to this definition">#</a></dt>
<dd><p>A table defining the individuals in a tree sequence. Note that although
each Individual has associated nodes, reference to these is not stored in
the individual table, but rather reference to the individual is stored for
each node in the <a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a>.  This is similar to the way in which
the relationship between sites and mutations is modelled.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The numpy arrays returned by table attribute accesses are copies
of the underlying data. In particular, this means that editing
individual values in the arrays will not change the table data
Instead, you should set entire columns or rows at once
(see <a class="reference internal" href="#sec-tables-api-accessing-table-data"><span class="std std-ref">Accessing table data</span></a>).</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Variables<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>flags</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>)  The array of flags values.</p></li>
<li><p><strong>location</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The flattened array of floating point location values. See
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> for more details.</p></li>
<li><p><strong>location_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>)  The array of offsets into the location column. See
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> for more details.</p></li>
<li><p><strong>parents</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The flattened array of parent individual ids. See
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> for more details.</p></li>
<li><p><strong>parents_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>)  The array of offsets into the parents column. See
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> for more details.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened array of binary metadata values. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>)  The array of offsets into the metadata column. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_schema</strong> (<a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><em>tskit.MetadataSchema</em></a>)  The metadata schema for this tables metadata column</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tskit.IndividualTable.add_row">
<span class="sig-name descname"><span class="pre">add_row</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">flags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">location</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parents</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#IndividualTable.add_row"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.IndividualTable.add_row" title="Link to this definition">#</a></dt>
<dd><p>Adds a new row to this <a class="reference internal" href="#tskit.IndividualTable" title="tskit.IndividualTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">IndividualTable</span></code></a> and returns the ID of the
corresponding individual. Metadata, if specified, will be validated and encoded
according to the tables
<a class="reference internal" href="#tskit.IndividualTable.metadata_schema" title="tskit.IndividualTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>flags</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The bitwise flags for the new node.</p></li>
<li><p><strong>location</strong> (<em>array-like</em>)  A list of numeric values or one-dimensional numpy
array describing the location of this individual. If not specified
or None, a zero-dimensional location is stored.</p></li>
<li><p><strong>parents</strong> (<em>array-like</em>)  A list or array of ids of parent individuals. If not
specified an empty array is stored.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><em>object</em></a>)  Any object that is valid metadata for the tables schema.
Defaults to the default metadata value for the tables schema. This is
typically <code class="docutils literal notranslate"><span class="pre">{}</span></code>. For no schema, <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The ID of the newly added individual.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.IndividualTable.set_columns">
<span class="sig-name descname"><span class="pre">set_columns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">flags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">location</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">location_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parents</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parents_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_schema</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#IndividualTable.set_columns"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.IndividualTable.set_columns" title="Link to this definition">#</a></dt>
<dd><p>Sets the values for each column in this <a class="reference internal" href="#tskit.IndividualTable" title="tskit.IndividualTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">IndividualTable</span></code></a> using the
values in the specified arrays. Overwrites existing data in all the table
columns.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">flags</span></code> array is mandatory and defines the number of individuals
the table will contain.
The <code class="docutils literal notranslate"><span class="pre">location</span></code> and <code class="docutils literal notranslate"><span class="pre">location_offset</span></code> parameters must be supplied
together, and meet the requirements for <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>.
The <code class="docutils literal notranslate"><span class="pre">parents</span></code> and <code class="docutils literal notranslate"><span class="pre">parents_offset</span></code> parameters must be supplied
together, and meet the requirements for <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>.
The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied
together, and meet the requirements for <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>.
See <a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information and
<a class="reference external" href="https://tskit.dev/tutorials/metadata.html#sec-tutorial-metadata-bulk" title="(in Project name not set)"><span>Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>flags</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>)  The bitwise flags for each individual. Required.</p></li>
<li><p><strong>location</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The flattened location array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">location_offset</span></code>. If not specified or None, an empty location
value is stored for each individual.</p></li>
<li><p><strong>location_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">location</span></code> array.</p></li>
<li><p><strong>parents</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The flattened parents array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">parents_offset</span></code>. If not specified or None, an empty parents array
is stored for each individual.</p></li>
<li><p><strong>parents_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">parents</span></code> array.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each individual.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
<li><p><strong>metadata_schema</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The encoded metadata schema. If None (default)
do not overwrite the exising schema. Note that a schema will need to be
encoded as a string, e.g. via <code class="docutils literal notranslate"><span class="pre">repr(new_metadata_schema)</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.IndividualTable.append_columns">
<span class="sig-name descname"><span class="pre">append_columns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">flags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">location</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">location_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parents</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parents_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#IndividualTable.append_columns"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.IndividualTable.append_columns" title="Link to this definition">#</a></dt>
<dd><p>Appends the specified arrays to the end of the columns in this
<a class="reference internal" href="#tskit.IndividualTable" title="tskit.IndividualTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">IndividualTable</span></code></a>. This allows many new rows to be added at once.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">flags</span></code> array is mandatory and defines the number of
extra individuals to add to the table.
The <code class="docutils literal notranslate"><span class="pre">parents</span></code> and <code class="docutils literal notranslate"><span class="pre">parents_offset</span></code> parameters must be supplied
together, and meet the requirements for <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>.
The <code class="docutils literal notranslate"><span class="pre">location</span></code> and <code class="docutils literal notranslate"><span class="pre">location_offset</span></code> parameters must be supplied
together, and meet the requirements for <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>.
The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied
together, and meet the requirements for <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>.
See <a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information and
<a class="reference external" href="https://tskit.dev/tutorials/metadata.html#sec-tutorial-metadata-bulk" title="(in Project name not set)"><span>Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>flags</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>)  The bitwise flags for each individual. Required.</p></li>
<li><p><strong>location</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The flattened location array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">location_offset</span></code>. If not specified or None, an empty location
value is stored for each individual.</p></li>
<li><p><strong>location_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">location</span></code> array.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each individual.</p></li>
<li><p><strong>parents</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The flattened parents array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">parents_offset</span></code>. If not specified or None, an empty parents array
is stored for each individual.</p></li>
<li><p><strong>parents_offset</strong>  The offsets into the <code class="docutils literal notranslate"><span class="pre">parents</span></code> array.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.IndividualTable.packset_location">
<span class="sig-name descname"><span class="pre">packset_location</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locations</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#IndividualTable.packset_location"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.IndividualTable.packset_location" title="Link to this definition">#</a></dt>
<dd><p>Packs the specified list of location values and updates the <code class="docutils literal notranslate"><span class="pre">location</span></code>
and <code class="docutils literal notranslate"><span class="pre">location_offset</span></code> columns. The length of the locations array
must be equal to the number of rows in the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>locations</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of locations interpreted as numpy float64
arrays.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.IndividualTable.packset_parents">
<span class="sig-name descname"><span class="pre">packset_parents</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parents</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#IndividualTable.packset_parents"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.IndividualTable.packset_parents" title="Link to this definition">#</a></dt>
<dd><p>Packs the specified list of parent values and updates the <code class="docutils literal notranslate"><span class="pre">parent</span></code>
and <code class="docutils literal notranslate"><span class="pre">parent_offset</span></code> columns. The length of the parents array
must be equal to the number of rows in the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>parents</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of list of parent ids, interpreted as numpy int32
arrays.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.IndividualTable.keep_rows">
<span class="sig-name descname"><span class="pre">keep_rows</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">keep</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#IndividualTable.keep_rows"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.IndividualTable.keep_rows" title="Link to this definition">#</a></dt>
<dd><p>Updates this table in-place according to the specified boolean
array, and returns the resulting mapping from old to new row IDs.
For each row <code class="docutils literal notranslate"><span class="pre">j</span></code>, if <code class="docutils literal notranslate"><span class="pre">keep[j]</span></code> is True, that row will be
retained in the output; otherwise, the row will be deleted.
Rows are retained in their original ordering.</p>
<p>The returned <code class="docutils literal notranslate"><span class="pre">id_map</span></code> is an array of the same length as
this table before the operation, such that <code class="docutils literal notranslate"><span class="pre">id_map[j]</span> <span class="pre">=</span> <span class="pre">-1</span></code>
(<a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a>) if row <code class="docutils literal notranslate"><span class="pre">j</span></code> was deleted, and <code class="docutils literal notranslate"><span class="pre">id_map[j]</span></code>
is the new ID of that row, otherwise.</p>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>This needs some examples to link to. See
<a class="github reference external" href="https://github.com/tskit-dev/tskit/issues/2708">tskit-dev/tskit#2708</a></p>
</div>
<p>The values in the <code class="docutils literal notranslate"><span class="pre">parents</span></code> column are updated according to this
map, so that reference integrity within the table is maintained.
As a consequence of this, the values in the <code class="docutils literal notranslate"><span class="pre">parents</span></code> column
for kept rows are bounds-checked and an error raised if they
are not valid. Rows that are deleted are not checked for
parent ID integrity.</p>
<p>If an attempt is made to delete rows that are referred to by
the <code class="docutils literal notranslate"><span class="pre">parents</span></code> column of rows that are retained, an error
is raised.</p>
<p>These error conditions are checked before any alterations to
the table are made.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>keep</strong> (<em>array-like</em>)  The rows to keep as a boolean array. Must
be the same length as the table, and convertible to a numpy
array of dtype bool.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The mapping between old and new row IDs as a numpy
array (dtype int32).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a> (dtype=np.int32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.IndividualTable.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.IndividualTable.__getitem__" title="Link to this definition">#</a></dt>
<dd><p>If passed an integer, return the specified row of this table, decoding metadata
if it is present. Supports negative indexing, e.g. <code class="docutils literal notranslate"><span class="pre">table[-5]</span></code>.
If passed a slice, iterable or array return a new table containing the specified
rows. Similar to numpy fancy indexing, if the array or iterables contains
booleans then the index acts as a mask, returning those rows for which the mask
is True. Note that as the result is a new table, the row ids will change as tskit
row ids are row indexes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong>  the index of a desired row, a slice of the desired rows, an
iterable or array of the desired row numbers, or a boolean array to use as
a mask.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.IndividualTable.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.IndividualTable.append" title="Link to this definition">#</a></dt>
<dd><p>Adds a new row to this table and returns the ID of the new row. Metadata, if
specified, will be validated and encoded according to the tables
<a class="reference internal" href="#tskit.IndividualTable.metadata_schema" title="tskit.IndividualTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>row</strong> (<em>row-like</em>)  An object that has attributes corresponding to the
properties of the new row. Both the objects returned from <code class="docutils literal notranslate"><span class="pre">table[i]</span></code> and
e.g. <code class="docutils literal notranslate"><span class="pre">ts.individual(i)</span></code> work for this purpose, along with any other
object with the correct attributes.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The index of the newly added row.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.IndividualTable.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.IndividualTable.asdict" title="Link to this definition">#</a></dt>
<dd><p>Returns a dictionary mapping the names of the columns in this table
to the corresponding numpy arrays.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.IndividualTable.assert_equals">
<span class="sig-name descname"><span class="pre">assert_equals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.IndividualTable.assert_equals" title="Link to this definition">#</a></dt>
<dd><p>Raise an AssertionError for the first found difference between
this and another table of the same type.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong>  Another table instance</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True exclude metadata and metadata schemas
from the comparison.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.IndividualTable.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.IndividualTable.clear" title="Link to this definition">#</a></dt>
<dd><p>Deletes all rows in this table.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.IndividualTable.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.IndividualTable.copy" title="Link to this definition">#</a></dt>
<dd><p>Returns a deep copy of this table</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.IndividualTable.drop_metadata">
<span class="sig-name descname"><span class="pre">drop_metadata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_schema</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.IndividualTable.drop_metadata" title="Link to this definition">#</a></dt>
<dd><p>Drops all metadata in this table. By default, the schema is also cleared,
except if <code class="docutils literal notranslate"><span class="pre">keep_schema</span></code> is True.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>keep_schema</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  True if the current schema should be kept intact.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.IndividualTable.equals">
<span class="sig-name descname"><span class="pre">equals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.IndividualTable.equals" title="Link to this definition">#</a></dt>
<dd><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal. By default, two tables
are considered equal if their columns and metadata schemas are
byte-for-byte identical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong>  Another table instance</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True exclude metadata and metadata schemas
from the comparison.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if other is equal to this table; False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.IndividualTable.metadata_schema">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">metadata_schema</span></span><a class="headerlink" href="#tskit.IndividualTable.metadata_schema" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this table.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.IndividualTable.metadata_vector">
<span class="sig-name descname"><span class="pre">metadata_vector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_value=&lt;object</span> <span class="pre">object&gt;</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.IndividualTable.metadata_vector" title="Link to this definition">#</a></dt>
<dd><p>Returns a numpy array of metadata values obtained by extracting <code class="docutils literal notranslate"><span class="pre">key</span></code>
from each metadata entry, and using <code class="docutils literal notranslate"><span class="pre">default_value</span></code> if the key is
not present. <code class="docutils literal notranslate"><span class="pre">key</span></code> may be a list, in which case nested values are returned.
For instance, <code class="docutils literal notranslate"><span class="pre">key</span> <span class="pre">=</span> <span class="pre">[&quot;a&quot;,</span> <span class="pre">&quot;x&quot;]</span></code> will return an array of
<code class="docutils literal notranslate"><span class="pre">row.metadata[&quot;a&quot;][&quot;x&quot;]</span></code> values, iterated over rows in this table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The name, or a list of names, of metadata entries.</p></li>
<li><p><strong>dtype</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The dtype of the result (can usually be omitted).</p></li>
<li><p><strong>default_value</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><em>object</em></a>)  The value to be inserted if the metadata key
is not present. Note that for numeric columns, a default value of None
will result in a non-numeric array. The default behaviour is to raise
<code class="docutils literal notranslate"><span class="pre">KeyError</span></code> on missing entries.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.IndividualTable.nbytes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nbytes</span></span><a class="headerlink" href="#tskit.IndividualTable.nbytes" title="Link to this definition">#</a></dt>
<dd><p>Returns the total number of bytes required to store the data
in this table. Note that this may not be equal to
the actual memory footprint.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.IndividualTable.packset_metadata">
<span class="sig-name descname"><span class="pre">packset_metadata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metadatas</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.IndividualTable.packset_metadata" title="Link to this definition">#</a></dt>
<dd><p>Packs the specified list of metadata values and updates the <code class="docutils literal notranslate"><span class="pre">metadata</span></code>
and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> columns. The length of the metadatas array
must be equal to the number of rows in the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>metadatas</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of metadata bytes values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.IndividualTable.truncate">
<span class="sig-name descname"><span class="pre">truncate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_rows</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.IndividualTable.truncate" title="Link to this definition">#</a></dt>
<dd><p>Truncates this table so that the only the first <code class="docutils literal notranslate"><span class="pre">num_rows</span></code> are retained.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>num_rows</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The number of rows to retain in this table.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<section id="associated-row-class">
<h5>Associated row class<a class="headerlink" href="#associated-row-class" title="Link to this heading">#</a></h5>
<p>A row returned from an <a class="reference internal" href="#tskit.IndividualTable" title="tskit.IndividualTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">IndividualTable</span></code></a> is an instance of the following
basic class, where each attribute matches an identically named attribute in the
<a class="reference internal" href="#tskit.Individual" title="tskit.Individual"><code class="xref py py-class docutils literal notranslate"><span class="pre">Individual</span></code></a> class.</p>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.IndividualTableRow">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">IndividualTableRow</span></span><a class="reference internal" href="_modules/tskit/tables.html#IndividualTableRow"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.IndividualTableRow" title="Link to this definition">#</a></dt>
<dd><p>A row in an <a class="reference internal" href="#tskit.IndividualTable" title="tskit.IndividualTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">IndividualTable</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.IndividualTableRow.flags">
<span class="sig-name descname"><span class="pre">flags</span></span><a class="headerlink" href="#tskit.IndividualTableRow.flags" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Individual.flags" title="tskit.Individual.flags"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Individual.flags</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.IndividualTableRow.location">
<span class="sig-name descname"><span class="pre">location</span></span><a class="headerlink" href="#tskit.IndividualTableRow.location" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Individual.location" title="tskit.Individual.location"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Individual.location</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.IndividualTableRow.parents">
<span class="sig-name descname"><span class="pre">parents</span></span><a class="headerlink" href="#tskit.IndividualTableRow.parents" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Individual.parents" title="tskit.Individual.parents"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Individual.parents</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.IndividualTableRow.metadata">
<span class="sig-name descname"><span class="pre">metadata</span></span><a class="headerlink" href="#tskit.IndividualTableRow.metadata" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Individual.metadata" title="tskit.Individual.metadata"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Individual.metadata</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.IndividualTableRow.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.IndividualTableRow.asdict" title="Link to this definition">#</a></dt>
<dd><p>Return a new dict which maps field names to their corresponding values
in this dataclass.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.IndividualTableRow.replace">
<span class="sig-name descname"><span class="pre">replace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.IndividualTableRow.replace" title="Link to this definition">#</a></dt>
<dd><p>Return a new instance of this dataclass, with the specified attributes
overwritten by new values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new instance of the same type</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
</section>
<section id="nodetable-classes">
<h4><a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a> classes<a class="headerlink" href="#nodetable-classes" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.NodeTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">NodeTable</span></span><a class="reference internal" href="_modules/tskit/tables.html#NodeTable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.NodeTable" title="Link to this definition">#</a></dt>
<dd><p>A table defining the nodes in a tree sequence. See the
<a class="reference internal" href="data-model.html#sec-node-table-definition"><span class="std std-ref">definitions</span></a> for details on the columns
in this table and the
<a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">tree sequence requirements</span></a> section
for the properties needed for a node table to be a part of a valid tree sequence.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The numpy arrays returned by table attribute accesses are copies
of the underlying data. In particular, this means that editing
individual values in the arrays will not change the table data
Instead, you should set entire columns or rows at once
(see <a class="reference internal" href="#sec-tables-api-accessing-table-data"><span class="std std-ref">Accessing table data</span></a>).</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Variables<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>time</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The array of time values.</p></li>
<li><p><strong>flags</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>)  The array of flags values.</p></li>
<li><p><strong>population</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The array of population IDs.</p></li>
<li><p><strong>individual</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The array of individual IDs that each node belongs to.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened array of binary metadata values. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>)  The array of offsets into the metadata column. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_schema</strong> (<a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><em>tskit.MetadataSchema</em></a>)  The metadata schema for this tables metadata column</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tskit.NodeTable.add_row">
<span class="sig-name descname"><span class="pre">add_row</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">flags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">population</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individual</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#NodeTable.add_row"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.NodeTable.add_row" title="Link to this definition">#</a></dt>
<dd><p>Adds a new row to this <a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a> and returns the ID of the
corresponding node. Metadata, if specified, will be validated and encoded
according to the tables
<a class="reference internal" href="#tskit.NodeTable.metadata_schema" title="tskit.NodeTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>flags</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The bitwise flags for the new node.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The birth time for the new node.</p></li>
<li><p><strong>population</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The ID of the population in which the new node was born.
Defaults to <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a>.</p></li>
<li><p><strong>individual</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The ID of the individual in which the new node was born.
Defaults to <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a>.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><em>object</em></a>)  Any object that is valid metadata for the tables schema.
Defaults to the default metadata value for the tables schema. This is
typically <code class="docutils literal notranslate"><span class="pre">{}</span></code>. For no schema, <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The ID of the newly added node.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.NodeTable.set_columns">
<span class="sig-name descname"><span class="pre">set_columns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">flags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">population</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individual</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_schema</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#NodeTable.set_columns"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.NodeTable.set_columns" title="Link to this definition">#</a></dt>
<dd><p>Sets the values for each column in this <a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a> using the values in
the specified arrays. Overwrites existing data in all the table columns.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">flags</span></code>, <code class="docutils literal notranslate"><span class="pre">time</span></code> and <code class="docutils literal notranslate"><span class="pre">population</span></code> arrays must all be of the same length,
which is equal to the number of nodes the table will contain. The
<code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied together, and
meet the requirements for <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>.
See <a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information and
<a class="reference external" href="https://tskit.dev/tutorials/metadata.html#sec-tutorial-metadata-bulk" title="(in Project name not set)"><span>Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>flags</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>)  The bitwise flags for each node. Required.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The time values for each node. Required.</p></li>
<li><p><strong>population</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The population values for each node. If not specified
or None, the <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> value is stored for each node.</p></li>
<li><p><strong>individual</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The individual values for each node. If not specified
or None, the <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> value is stored for each node.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each node.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
<li><p><strong>metadata_schema</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The encoded metadata schema. If None (default)
do not overwrite the exising schema. Note that a schema will need to be
encoded as a string, e.g. via <code class="docutils literal notranslate"><span class="pre">repr(new_metadata_schema)</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.NodeTable.append_columns">
<span class="sig-name descname"><span class="pre">append_columns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">flags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">population</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individual</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#NodeTable.append_columns"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.NodeTable.append_columns" title="Link to this definition">#</a></dt>
<dd><p>Appends the specified arrays to the end of the columns in this
<a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a>. This allows many new rows to be added at once.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">flags</span></code>, <code class="docutils literal notranslate"><span class="pre">time</span></code> and <code class="docutils literal notranslate"><span class="pre">population</span></code> arrays must all be of the same length,
which is equal to the number of nodes that will be added to the table. The
<code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied together, and
meet the requirements for <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>.
See <a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information and
<a class="reference external" href="https://tskit.dev/tutorials/metadata.html#sec-tutorial-metadata-bulk" title="(in Project name not set)"><span>Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>flags</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>)  The bitwise flags for each node. Required.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The time values for each node. Required.</p></li>
<li><p><strong>population</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The population values for each node. If not specified
or None, the <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> value is stored for each node.</p></li>
<li><p><strong>individual</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The individual values for each node. If not specified
or None, the <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a> value is stored for each node.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each node.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.NodeTable.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.NodeTable.__getitem__" title="Link to this definition">#</a></dt>
<dd><p>If passed an integer, return the specified row of this table, decoding metadata
if it is present. Supports negative indexing, e.g. <code class="docutils literal notranslate"><span class="pre">table[-5]</span></code>.
If passed a slice, iterable or array return a new table containing the specified
rows. Similar to numpy fancy indexing, if the array or iterables contains
booleans then the index acts as a mask, returning those rows for which the mask
is True. Note that as the result is a new table, the row ids will change as tskit
row ids are row indexes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong>  the index of a desired row, a slice of the desired rows, an
iterable or array of the desired row numbers, or a boolean array to use as
a mask.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.NodeTable.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.NodeTable.append" title="Link to this definition">#</a></dt>
<dd><p>Adds a new row to this table and returns the ID of the new row. Metadata, if
specified, will be validated and encoded according to the tables
<a class="reference internal" href="#tskit.IndividualTable.metadata_schema" title="tskit.IndividualTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>row</strong> (<em>row-like</em>)  An object that has attributes corresponding to the
properties of the new row. Both the objects returned from <code class="docutils literal notranslate"><span class="pre">table[i]</span></code> and
e.g. <code class="docutils literal notranslate"><span class="pre">ts.individual(i)</span></code> work for this purpose, along with any other
object with the correct attributes.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The index of the newly added row.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.NodeTable.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.NodeTable.asdict" title="Link to this definition">#</a></dt>
<dd><p>Returns a dictionary mapping the names of the columns in this table
to the corresponding numpy arrays.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.NodeTable.assert_equals">
<span class="sig-name descname"><span class="pre">assert_equals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.NodeTable.assert_equals" title="Link to this definition">#</a></dt>
<dd><p>Raise an AssertionError for the first found difference between
this and another table of the same type.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong>  Another table instance</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True exclude metadata and metadata schemas
from the comparison.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.NodeTable.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.NodeTable.clear" title="Link to this definition">#</a></dt>
<dd><p>Deletes all rows in this table.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.NodeTable.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.NodeTable.copy" title="Link to this definition">#</a></dt>
<dd><p>Returns a deep copy of this table</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.NodeTable.drop_metadata">
<span class="sig-name descname"><span class="pre">drop_metadata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_schema</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.NodeTable.drop_metadata" title="Link to this definition">#</a></dt>
<dd><p>Drops all metadata in this table. By default, the schema is also cleared,
except if <code class="docutils literal notranslate"><span class="pre">keep_schema</span></code> is True.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>keep_schema</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  True if the current schema should be kept intact.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.NodeTable.equals">
<span class="sig-name descname"><span class="pre">equals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.NodeTable.equals" title="Link to this definition">#</a></dt>
<dd><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal. By default, two tables
are considered equal if their columns and metadata schemas are
byte-for-byte identical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong>  Another table instance</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True exclude metadata and metadata schemas
from the comparison.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if other is equal to this table; False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.NodeTable.keep_rows">
<span class="sig-name descname"><span class="pre">keep_rows</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">keep</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.NodeTable.keep_rows" title="Link to this definition">#</a></dt>
<dd><p>Updates this table in-place according to the specified boolean
array, and returns the resulting mapping from old to new row IDs.
For each row <code class="docutils literal notranslate"><span class="pre">j</span></code>, if <code class="docutils literal notranslate"><span class="pre">keep[j]</span></code> is True, that row will be
retained in the output; otherwise, the row will be deleted.
Rows are retained in their original ordering.</p>
<p>The returned <code class="docutils literal notranslate"><span class="pre">id_map</span></code> is an array of the same length as
this table before the operation, such that <code class="docutils literal notranslate"><span class="pre">id_map[j]</span> <span class="pre">=</span> <span class="pre">-1</span></code>
(<a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a>) if row <code class="docutils literal notranslate"><span class="pre">j</span></code> was deleted, and <code class="docutils literal notranslate"><span class="pre">id_map[j]</span></code>
is the new ID of that row, otherwise.</p>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>This needs some examples to link to. See
<a class="github reference external" href="https://github.com/tskit-dev/tskit/issues/2708">tskit-dev/tskit#2708</a></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>keep</strong> (<em>array-like</em>)  The rows to keep as a boolean array. Must
be the same length as the table, and convertible to a numpy
array of dtype bool.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The mapping between old and new row IDs as a numpy
array (dtype int32).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a> (dtype=np.int32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.NodeTable.metadata_schema">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">metadata_schema</span></span><a class="headerlink" href="#tskit.NodeTable.metadata_schema" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this table.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.NodeTable.metadata_vector">
<span class="sig-name descname"><span class="pre">metadata_vector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_value=&lt;object</span> <span class="pre">object&gt;</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.NodeTable.metadata_vector" title="Link to this definition">#</a></dt>
<dd><p>Returns a numpy array of metadata values obtained by extracting <code class="docutils literal notranslate"><span class="pre">key</span></code>
from each metadata entry, and using <code class="docutils literal notranslate"><span class="pre">default_value</span></code> if the key is
not present. <code class="docutils literal notranslate"><span class="pre">key</span></code> may be a list, in which case nested values are returned.
For instance, <code class="docutils literal notranslate"><span class="pre">key</span> <span class="pre">=</span> <span class="pre">[&quot;a&quot;,</span> <span class="pre">&quot;x&quot;]</span></code> will return an array of
<code class="docutils literal notranslate"><span class="pre">row.metadata[&quot;a&quot;][&quot;x&quot;]</span></code> values, iterated over rows in this table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The name, or a list of names, of metadata entries.</p></li>
<li><p><strong>dtype</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The dtype of the result (can usually be omitted).</p></li>
<li><p><strong>default_value</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><em>object</em></a>)  The value to be inserted if the metadata key
is not present. Note that for numeric columns, a default value of None
will result in a non-numeric array. The default behaviour is to raise
<code class="docutils literal notranslate"><span class="pre">KeyError</span></code> on missing entries.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.NodeTable.nbytes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nbytes</span></span><a class="headerlink" href="#tskit.NodeTable.nbytes" title="Link to this definition">#</a></dt>
<dd><p>Returns the total number of bytes required to store the data
in this table. Note that this may not be equal to
the actual memory footprint.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.NodeTable.packset_metadata">
<span class="sig-name descname"><span class="pre">packset_metadata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metadatas</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.NodeTable.packset_metadata" title="Link to this definition">#</a></dt>
<dd><p>Packs the specified list of metadata values and updates the <code class="docutils literal notranslate"><span class="pre">metadata</span></code>
and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> columns. The length of the metadatas array
must be equal to the number of rows in the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>metadatas</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of metadata bytes values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.NodeTable.truncate">
<span class="sig-name descname"><span class="pre">truncate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_rows</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.NodeTable.truncate" title="Link to this definition">#</a></dt>
<dd><p>Truncates this table so that the only the first <code class="docutils literal notranslate"><span class="pre">num_rows</span></code> are retained.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>num_rows</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The number of rows to retain in this table.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<section id="id10">
<h5>Associated row class<a class="headerlink" href="#id10" title="Link to this heading">#</a></h5>
<p>A row returned from a <a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a> is an instance of the following
basic class, where each attribute matches an identically named attribute in the
<a class="reference internal" href="#tskit.Node" title="tskit.Node"><code class="xref py py-class docutils literal notranslate"><span class="pre">Node</span></code></a> class.</p>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.NodeTableRow">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">NodeTableRow</span></span><a class="reference internal" href="_modules/tskit/tables.html#NodeTableRow"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.NodeTableRow" title="Link to this definition">#</a></dt>
<dd><p>A row in a <a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.NodeTableRow.flags">
<span class="sig-name descname"><span class="pre">flags</span></span><a class="headerlink" href="#tskit.NodeTableRow.flags" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Node.flags" title="tskit.Node.flags"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Node.flags</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.NodeTableRow.time">
<span class="sig-name descname"><span class="pre">time</span></span><a class="headerlink" href="#tskit.NodeTableRow.time" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Node.time" title="tskit.Node.time"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Node.time</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.NodeTableRow.population">
<span class="sig-name descname"><span class="pre">population</span></span><a class="headerlink" href="#tskit.NodeTableRow.population" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Node.population" title="tskit.Node.population"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Node.population</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.NodeTableRow.individual">
<span class="sig-name descname"><span class="pre">individual</span></span><a class="headerlink" href="#tskit.NodeTableRow.individual" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Node.individual" title="tskit.Node.individual"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Node.individual</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.NodeTableRow.metadata">
<span class="sig-name descname"><span class="pre">metadata</span></span><a class="headerlink" href="#tskit.NodeTableRow.metadata" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Node.metadata" title="tskit.Node.metadata"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Node.metadata</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.NodeTableRow.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.NodeTableRow.asdict" title="Link to this definition">#</a></dt>
<dd><p>Return a new dict which maps field names to their corresponding values
in this dataclass.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.NodeTableRow.replace">
<span class="sig-name descname"><span class="pre">replace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.NodeTableRow.replace" title="Link to this definition">#</a></dt>
<dd><p>Return a new instance of this dataclass, with the specified attributes
overwritten by new values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new instance of the same type</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
</section>
<section id="edgetable-classes">
<h4><a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a> classes<a class="headerlink" href="#edgetable-classes" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.EdgeTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">EdgeTable</span></span><a class="reference internal" href="_modules/tskit/tables.html#EdgeTable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.EdgeTable" title="Link to this definition">#</a></dt>
<dd><p>A table defining the edges in a tree sequence. See the
<a class="reference internal" href="data-model.html#sec-edge-table-definition"><span class="std std-ref">definitions</span></a> for details on the columns
in this table and the
<a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">tree sequence requirements</span></a> section
for the properties needed for an edge table to be a part of a valid tree sequence.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The numpy arrays returned by table attribute accesses are copies
of the underlying data. In particular, this means that editing
individual values in the arrays will not change the table data
Instead, you should set entire columns or rows at once
(see <a class="reference internal" href="#sec-tables-api-accessing-table-data"><span class="std std-ref">Accessing table data</span></a>).</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Variables<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>left</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The array of left coordinates.</p></li>
<li><p><strong>right</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The array of right coordinates.</p></li>
<li><p><strong>parent</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The array of parent node IDs.</p></li>
<li><p><strong>child</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The array of child node IDs.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened array of binary metadata values. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>)  The array of offsets into the metadata column. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_schema</strong> (<a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><em>tskit.MetadataSchema</em></a>)  The metadata schema for this tables metadata column</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tskit.EdgeTable.add_row">
<span class="sig-name descname"><span class="pre">add_row</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">left</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">child</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#EdgeTable.add_row"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.EdgeTable.add_row" title="Link to this definition">#</a></dt>
<dd><p>Adds a new row to this <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a> and returns the ID of the
corresponding edge. Metadata, if specified, will be validated and encoded
according to the tables
<a class="reference internal" href="#tskit.EdgeTable.metadata_schema" title="tskit.EdgeTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>left</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The left coordinate (inclusive).</p></li>
<li><p><strong>right</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The right coordinate (exclusive).</p></li>
<li><p><strong>parent</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The ID of parent node.</p></li>
<li><p><strong>child</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The ID of child node.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><em>object</em></a>)  Any object that is valid metadata for the tables schema.
Defaults to the default metadata value for the tables schema. This is
typically <code class="docutils literal notranslate"><span class="pre">{}</span></code>. For no schema, <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The ID of the newly added edge.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.EdgeTable.set_columns">
<span class="sig-name descname"><span class="pre">set_columns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">child</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_schema</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#EdgeTable.set_columns"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.EdgeTable.set_columns" title="Link to this definition">#</a></dt>
<dd><p>Sets the values for each column in this <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a> using the values
in the specified arrays. Overwrites existing data in all the table columns.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">left</span></code>, <code class="docutils literal notranslate"><span class="pre">right</span></code>, <code class="docutils literal notranslate"><span class="pre">parent</span></code> and <code class="docutils literal notranslate"><span class="pre">child</span></code> parameters are mandatory,
and must be numpy arrays of the same length (which is equal to the number of
edges the table will contain).
The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied together,
and meet the requirements for <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>.
See <a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information and
<a class="reference external" href="https://tskit.dev/tutorials/metadata.html#sec-tutorial-metadata-bulk" title="(in Project name not set)"><span>Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>left</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The left coordinates (inclusive).</p></li>
<li><p><strong>right</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The right coordinates (exclusive).</p></li>
<li><p><strong>parent</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The parent node IDs.</p></li>
<li><p><strong>child</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The child node IDs.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each node.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
<li><p><strong>metadata_schema</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The encoded metadata schema. If None (default)
do not overwrite the exising schema. Note that a schema will need to be
encoded as a string, e.g. via <code class="docutils literal notranslate"><span class="pre">repr(new_metadata_schema)</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.EdgeTable.append_columns">
<span class="sig-name descname"><span class="pre">append_columns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">left</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">child</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#EdgeTable.append_columns"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.EdgeTable.append_columns" title="Link to this definition">#</a></dt>
<dd><p>Appends the specified arrays to the end of the columns of this
<a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a>. This allows many new rows to be added at once.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">left</span></code>, <code class="docutils literal notranslate"><span class="pre">right</span></code>, <code class="docutils literal notranslate"><span class="pre">parent</span></code> and <code class="docutils literal notranslate"><span class="pre">child</span></code> parameters are mandatory,
and must be numpy arrays of the same length (which is equal to the number of
additional edges to add to the table). The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> and
<code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied together, and
meet the requirements for <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>.
See <a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information and
<a class="reference external" href="https://tskit.dev/tutorials/metadata.html#sec-tutorial-metadata-bulk" title="(in Project name not set)"><span>Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>left</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The left coordinates (inclusive).</p></li>
<li><p><strong>right</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The right coordinates (exclusive).</p></li>
<li><p><strong>parent</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The parent node IDs.</p></li>
<li><p><strong>child</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The child node IDs.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each node.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.EdgeTable.squash">
<span class="sig-name descname"><span class="pre">squash</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#EdgeTable.squash"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.EdgeTable.squash" title="Link to this definition">#</a></dt>
<dd><p>Sorts, then condenses the table into the smallest possible number of rows by
combining any adjacent edges.
A pair of edges is said to be <cite>adjacent</cite> if they have the same parent and child
nodes, and if the left coordinate of one of the edges is equal to the right
coordinate of the other edge.
The <code class="docutils literal notranslate"><span class="pre">squash</span></code> method modifies an <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a> in place so that any set of
adjacent edges is replaced by a single edge.
The new edge will have the same parent and child node, a left coordinate
equal to the smallest left coordinate in the set, and a right coordinate
equal to the largest right coordinate in the set.
The new edge table will be sorted in the order (P, C, L, R): if the node table
is ordered by increasing node time, as is common, this order will meet the
<a class="reference internal" href="data-model.html#sec-edge-requirements"><span class="std std-ref">Edge requirements</span></a> for a valid tree sequence, otherwise you will need
to call <a class="reference internal" href="#tskit.TableCollection.sort" title="tskit.TableCollection.sort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">sort()</span></code></a> on the entire <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Note that this method will fail if any edges have non-empty metadata.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.EdgeTable.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.EdgeTable.__getitem__" title="Link to this definition">#</a></dt>
<dd><p>If passed an integer, return the specified row of this table, decoding metadata
if it is present. Supports negative indexing, e.g. <code class="docutils literal notranslate"><span class="pre">table[-5]</span></code>.
If passed a slice, iterable or array return a new table containing the specified
rows. Similar to numpy fancy indexing, if the array or iterables contains
booleans then the index acts as a mask, returning those rows for which the mask
is True. Note that as the result is a new table, the row ids will change as tskit
row ids are row indexes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong>  the index of a desired row, a slice of the desired rows, an
iterable or array of the desired row numbers, or a boolean array to use as
a mask.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.EdgeTable.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.EdgeTable.append" title="Link to this definition">#</a></dt>
<dd><p>Adds a new row to this table and returns the ID of the new row. Metadata, if
specified, will be validated and encoded according to the tables
<a class="reference internal" href="#tskit.IndividualTable.metadata_schema" title="tskit.IndividualTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>row</strong> (<em>row-like</em>)  An object that has attributes corresponding to the
properties of the new row. Both the objects returned from <code class="docutils literal notranslate"><span class="pre">table[i]</span></code> and
e.g. <code class="docutils literal notranslate"><span class="pre">ts.individual(i)</span></code> work for this purpose, along with any other
object with the correct attributes.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The index of the newly added row.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.EdgeTable.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.EdgeTable.asdict" title="Link to this definition">#</a></dt>
<dd><p>Returns a dictionary mapping the names of the columns in this table
to the corresponding numpy arrays.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.EdgeTable.assert_equals">
<span class="sig-name descname"><span class="pre">assert_equals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.EdgeTable.assert_equals" title="Link to this definition">#</a></dt>
<dd><p>Raise an AssertionError for the first found difference between
this and another table of the same type.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong>  Another table instance</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True exclude metadata and metadata schemas
from the comparison.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.EdgeTable.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.EdgeTable.clear" title="Link to this definition">#</a></dt>
<dd><p>Deletes all rows in this table.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.EdgeTable.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.EdgeTable.copy" title="Link to this definition">#</a></dt>
<dd><p>Returns a deep copy of this table</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.EdgeTable.drop_metadata">
<span class="sig-name descname"><span class="pre">drop_metadata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_schema</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.EdgeTable.drop_metadata" title="Link to this definition">#</a></dt>
<dd><p>Drops all metadata in this table. By default, the schema is also cleared,
except if <code class="docutils literal notranslate"><span class="pre">keep_schema</span></code> is True.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>keep_schema</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  True if the current schema should be kept intact.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.EdgeTable.equals">
<span class="sig-name descname"><span class="pre">equals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.EdgeTable.equals" title="Link to this definition">#</a></dt>
<dd><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal. By default, two tables
are considered equal if their columns and metadata schemas are
byte-for-byte identical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong>  Another table instance</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True exclude metadata and metadata schemas
from the comparison.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if other is equal to this table; False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.EdgeTable.keep_rows">
<span class="sig-name descname"><span class="pre">keep_rows</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">keep</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.EdgeTable.keep_rows" title="Link to this definition">#</a></dt>
<dd><p>Updates this table in-place according to the specified boolean
array, and returns the resulting mapping from old to new row IDs.
For each row <code class="docutils literal notranslate"><span class="pre">j</span></code>, if <code class="docutils literal notranslate"><span class="pre">keep[j]</span></code> is True, that row will be
retained in the output; otherwise, the row will be deleted.
Rows are retained in their original ordering.</p>
<p>The returned <code class="docutils literal notranslate"><span class="pre">id_map</span></code> is an array of the same length as
this table before the operation, such that <code class="docutils literal notranslate"><span class="pre">id_map[j]</span> <span class="pre">=</span> <span class="pre">-1</span></code>
(<a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a>) if row <code class="docutils literal notranslate"><span class="pre">j</span></code> was deleted, and <code class="docutils literal notranslate"><span class="pre">id_map[j]</span></code>
is the new ID of that row, otherwise.</p>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>This needs some examples to link to. See
<a class="github reference external" href="https://github.com/tskit-dev/tskit/issues/2708">tskit-dev/tskit#2708</a></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>keep</strong> (<em>array-like</em>)  The rows to keep as a boolean array. Must
be the same length as the table, and convertible to a numpy
array of dtype bool.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The mapping between old and new row IDs as a numpy
array (dtype int32).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a> (dtype=np.int32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.EdgeTable.metadata_schema">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">metadata_schema</span></span><a class="headerlink" href="#tskit.EdgeTable.metadata_schema" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this table.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.EdgeTable.metadata_vector">
<span class="sig-name descname"><span class="pre">metadata_vector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_value=&lt;object</span> <span class="pre">object&gt;</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.EdgeTable.metadata_vector" title="Link to this definition">#</a></dt>
<dd><p>Returns a numpy array of metadata values obtained by extracting <code class="docutils literal notranslate"><span class="pre">key</span></code>
from each metadata entry, and using <code class="docutils literal notranslate"><span class="pre">default_value</span></code> if the key is
not present. <code class="docutils literal notranslate"><span class="pre">key</span></code> may be a list, in which case nested values are returned.
For instance, <code class="docutils literal notranslate"><span class="pre">key</span> <span class="pre">=</span> <span class="pre">[&quot;a&quot;,</span> <span class="pre">&quot;x&quot;]</span></code> will return an array of
<code class="docutils literal notranslate"><span class="pre">row.metadata[&quot;a&quot;][&quot;x&quot;]</span></code> values, iterated over rows in this table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The name, or a list of names, of metadata entries.</p></li>
<li><p><strong>dtype</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The dtype of the result (can usually be omitted).</p></li>
<li><p><strong>default_value</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><em>object</em></a>)  The value to be inserted if the metadata key
is not present. Note that for numeric columns, a default value of None
will result in a non-numeric array. The default behaviour is to raise
<code class="docutils literal notranslate"><span class="pre">KeyError</span></code> on missing entries.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.EdgeTable.nbytes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nbytes</span></span><a class="headerlink" href="#tskit.EdgeTable.nbytes" title="Link to this definition">#</a></dt>
<dd><p>Returns the total number of bytes required to store the data
in this table. Note that this may not be equal to
the actual memory footprint.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.EdgeTable.packset_metadata">
<span class="sig-name descname"><span class="pre">packset_metadata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metadatas</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.EdgeTable.packset_metadata" title="Link to this definition">#</a></dt>
<dd><p>Packs the specified list of metadata values and updates the <code class="docutils literal notranslate"><span class="pre">metadata</span></code>
and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> columns. The length of the metadatas array
must be equal to the number of rows in the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>metadatas</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of metadata bytes values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.EdgeTable.truncate">
<span class="sig-name descname"><span class="pre">truncate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_rows</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.EdgeTable.truncate" title="Link to this definition">#</a></dt>
<dd><p>Truncates this table so that the only the first <code class="docutils literal notranslate"><span class="pre">num_rows</span></code> are retained.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>num_rows</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The number of rows to retain in this table.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<section id="id11">
<h5>Associated row class<a class="headerlink" href="#id11" title="Link to this heading">#</a></h5>
<p>A row returned from an <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a> is an instance of the following
basic class, where each attribute matches an identically named attribute in the
<a class="reference internal" href="#tskit.Edge" title="tskit.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> class.</p>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.EdgeTableRow">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">EdgeTableRow</span></span><a class="reference internal" href="_modules/tskit/tables.html#EdgeTableRow"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.EdgeTableRow" title="Link to this definition">#</a></dt>
<dd><p>A row in an <a class="reference internal" href="#tskit.EdgeTable" title="tskit.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.EdgeTableRow.left">
<span class="sig-name descname"><span class="pre">left</span></span><a class="headerlink" href="#tskit.EdgeTableRow.left" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Edge.left" title="tskit.Edge.left"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Edge.left</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.EdgeTableRow.right">
<span class="sig-name descname"><span class="pre">right</span></span><a class="headerlink" href="#tskit.EdgeTableRow.right" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Edge.right" title="tskit.Edge.right"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Edge.right</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.EdgeTableRow.parent">
<span class="sig-name descname"><span class="pre">parent</span></span><a class="headerlink" href="#tskit.EdgeTableRow.parent" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Edge.parent" title="tskit.Edge.parent"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Edge.parent</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.EdgeTableRow.child">
<span class="sig-name descname"><span class="pre">child</span></span><a class="headerlink" href="#tskit.EdgeTableRow.child" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Edge.child" title="tskit.Edge.child"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Edge.child</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.EdgeTableRow.metadata">
<span class="sig-name descname"><span class="pre">metadata</span></span><a class="headerlink" href="#tskit.EdgeTableRow.metadata" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Edge.metadata" title="tskit.Edge.metadata"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Edge.metadata</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.EdgeTableRow.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.EdgeTableRow.asdict" title="Link to this definition">#</a></dt>
<dd><p>Return a new dict which maps field names to their corresponding values
in this dataclass.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.EdgeTableRow.replace">
<span class="sig-name descname"><span class="pre">replace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.EdgeTableRow.replace" title="Link to this definition">#</a></dt>
<dd><p>Return a new instance of this dataclass, with the specified attributes
overwritten by new values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new instance of the same type</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
</section>
<section id="migrationtable-classes">
<h4><a class="reference internal" href="#tskit.MigrationTable" title="tskit.MigrationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MigrationTable</span></code></a> classes<a class="headerlink" href="#migrationtable-classes" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.MigrationTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">MigrationTable</span></span><a class="reference internal" href="_modules/tskit/tables.html#MigrationTable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MigrationTable" title="Link to this definition">#</a></dt>
<dd><p>A table defining the migrations in a tree sequence. See the
<a class="reference internal" href="data-model.html#sec-migration-table-definition"><span class="std std-ref">definitions</span></a> for details on the columns
in this table and the
<a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">tree sequence requirements</span></a> section
for the properties needed for a migration table to be a part of a valid tree
sequence.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The numpy arrays returned by table attribute accesses are copies
of the underlying data. In particular, this means that editing
individual values in the arrays will not change the table data
Instead, you should set entire columns or rows at once
(see <a class="reference internal" href="#sec-tables-api-accessing-table-data"><span class="std std-ref">Accessing table data</span></a>).</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Variables<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>left</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The array of left coordinates.</p></li>
<li><p><strong>right</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The array of right coordinates.</p></li>
<li><p><strong>node</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The array of node IDs.</p></li>
<li><p><strong>source</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The array of source population IDs.</p></li>
<li><p><strong>dest</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The array of destination population IDs.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The array of time values.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened array of binary metadata values. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>)  The array of offsets into the metadata column. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_schema</strong> (<a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><em>tskit.MetadataSchema</em></a>)  The metadata schema for this tables metadata column</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tskit.MigrationTable.add_row">
<span class="sig-name descname"><span class="pre">add_row</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">left</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dest</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#MigrationTable.add_row"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MigrationTable.add_row" title="Link to this definition">#</a></dt>
<dd><p>Adds a new row to this <a class="reference internal" href="#tskit.MigrationTable" title="tskit.MigrationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MigrationTable</span></code></a> and returns the ID of the
corresponding migration. Metadata, if specified, will be validated and encoded
according to the tables
<a class="reference internal" href="#tskit.MigrationTable.metadata_schema" title="tskit.MigrationTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>left</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The left coordinate (inclusive).</p></li>
<li><p><strong>right</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The right coordinate (exclusive).</p></li>
<li><p><strong>node</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The node ID.</p></li>
<li><p><strong>source</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The ID of the source population.</p></li>
<li><p><strong>dest</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The ID of the destination population.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The time of the migration event.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><em>object</em></a>)  Any object that is valid metadata for the tables schema.
Defaults to the default metadata value for the tables schema. This is
typically <code class="docutils literal notranslate"><span class="pre">{}</span></code>. For no schema, <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The ID of the newly added migration.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MigrationTable.set_columns">
<span class="sig-name descname"><span class="pre">set_columns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dest</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_schema</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#MigrationTable.set_columns"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MigrationTable.set_columns" title="Link to this definition">#</a></dt>
<dd><p>Sets the values for each column in this <a class="reference internal" href="#tskit.MigrationTable" title="tskit.MigrationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MigrationTable</span></code></a> using the values
in the specified arrays. Overwrites existing data in all the table columns.</p>
<p>All parameters except <code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> and are mandatory,
and must be numpy arrays of the same length (which is equal to the number of
migrations the table will contain).
The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied together,
and meet the requirements for <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>.
See <a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information and
<a class="reference external" href="https://tskit.dev/tutorials/metadata.html#sec-tutorial-metadata-bulk" title="(in Project name not set)"><span>Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>left</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The left coordinates (inclusive).</p></li>
<li><p><strong>right</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The right coordinates (exclusive).</p></li>
<li><p><strong>node</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The node IDs.</p></li>
<li><p><strong>source</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The source population IDs.</p></li>
<li><p><strong>dest</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The destination population IDs.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int64</em>)  The time of each migration.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each migration.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
<li><p><strong>metadata_schema</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The encoded metadata schema. If None (default)
do not overwrite the exising schema. Note that a schema will need to be
encoded as a string, e.g. via <code class="docutils literal notranslate"><span class="pre">repr(new_metadata_schema)</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MigrationTable.append_columns">
<span class="sig-name descname"><span class="pre">append_columns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">left</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dest</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#MigrationTable.append_columns"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MigrationTable.append_columns" title="Link to this definition">#</a></dt>
<dd><p>Appends the specified arrays to the end of the columns of this
<a class="reference internal" href="#tskit.MigrationTable" title="tskit.MigrationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MigrationTable</span></code></a>. This allows many new rows to be added at once.</p>
<p>All parameters except <code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> and are mandatory,
and must be numpy arrays of the same length (which is equal to the number of
additional migrations to add to the table). The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> and
<code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied together, and
meet the requirements for <a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a>.
See <a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information and
<a class="reference external" href="https://tskit.dev/tutorials/metadata.html#sec-tutorial-metadata-bulk" title="(in Project name not set)"><span>Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>left</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The left coordinates (inclusive).</p></li>
<li><p><strong>right</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The right coordinates (exclusive).</p></li>
<li><p><strong>node</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The node IDs.</p></li>
<li><p><strong>source</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The source population IDs.</p></li>
<li><p><strong>dest</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The destination population IDs.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int64</em>)  The time of each migration.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each migration.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MigrationTable.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MigrationTable.__getitem__" title="Link to this definition">#</a></dt>
<dd><p>If passed an integer, return the specified row of this table, decoding metadata
if it is present. Supports negative indexing, e.g. <code class="docutils literal notranslate"><span class="pre">table[-5]</span></code>.
If passed a slice, iterable or array return a new table containing the specified
rows. Similar to numpy fancy indexing, if the array or iterables contains
booleans then the index acts as a mask, returning those rows for which the mask
is True. Note that as the result is a new table, the row ids will change as tskit
row ids are row indexes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong>  the index of a desired row, a slice of the desired rows, an
iterable or array of the desired row numbers, or a boolean array to use as
a mask.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MigrationTable.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MigrationTable.append" title="Link to this definition">#</a></dt>
<dd><p>Adds a new row to this table and returns the ID of the new row. Metadata, if
specified, will be validated and encoded according to the tables
<a class="reference internal" href="#tskit.IndividualTable.metadata_schema" title="tskit.IndividualTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>row</strong> (<em>row-like</em>)  An object that has attributes corresponding to the
properties of the new row. Both the objects returned from <code class="docutils literal notranslate"><span class="pre">table[i]</span></code> and
e.g. <code class="docutils literal notranslate"><span class="pre">ts.individual(i)</span></code> work for this purpose, along with any other
object with the correct attributes.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The index of the newly added row.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MigrationTable.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MigrationTable.asdict" title="Link to this definition">#</a></dt>
<dd><p>Returns a dictionary mapping the names of the columns in this table
to the corresponding numpy arrays.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MigrationTable.assert_equals">
<span class="sig-name descname"><span class="pre">assert_equals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MigrationTable.assert_equals" title="Link to this definition">#</a></dt>
<dd><p>Raise an AssertionError for the first found difference between
this and another table of the same type.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong>  Another table instance</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True exclude metadata and metadata schemas
from the comparison.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MigrationTable.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MigrationTable.clear" title="Link to this definition">#</a></dt>
<dd><p>Deletes all rows in this table.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MigrationTable.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MigrationTable.copy" title="Link to this definition">#</a></dt>
<dd><p>Returns a deep copy of this table</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MigrationTable.drop_metadata">
<span class="sig-name descname"><span class="pre">drop_metadata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_schema</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MigrationTable.drop_metadata" title="Link to this definition">#</a></dt>
<dd><p>Drops all metadata in this table. By default, the schema is also cleared,
except if <code class="docutils literal notranslate"><span class="pre">keep_schema</span></code> is True.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>keep_schema</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  True if the current schema should be kept intact.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MigrationTable.equals">
<span class="sig-name descname"><span class="pre">equals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MigrationTable.equals" title="Link to this definition">#</a></dt>
<dd><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal. By default, two tables
are considered equal if their columns and metadata schemas are
byte-for-byte identical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong>  Another table instance</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True exclude metadata and metadata schemas
from the comparison.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if other is equal to this table; False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MigrationTable.keep_rows">
<span class="sig-name descname"><span class="pre">keep_rows</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">keep</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MigrationTable.keep_rows" title="Link to this definition">#</a></dt>
<dd><p>Updates this table in-place according to the specified boolean
array, and returns the resulting mapping from old to new row IDs.
For each row <code class="docutils literal notranslate"><span class="pre">j</span></code>, if <code class="docutils literal notranslate"><span class="pre">keep[j]</span></code> is True, that row will be
retained in the output; otherwise, the row will be deleted.
Rows are retained in their original ordering.</p>
<p>The returned <code class="docutils literal notranslate"><span class="pre">id_map</span></code> is an array of the same length as
this table before the operation, such that <code class="docutils literal notranslate"><span class="pre">id_map[j]</span> <span class="pre">=</span> <span class="pre">-1</span></code>
(<a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a>) if row <code class="docutils literal notranslate"><span class="pre">j</span></code> was deleted, and <code class="docutils literal notranslate"><span class="pre">id_map[j]</span></code>
is the new ID of that row, otherwise.</p>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>This needs some examples to link to. See
<a class="github reference external" href="https://github.com/tskit-dev/tskit/issues/2708">tskit-dev/tskit#2708</a></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>keep</strong> (<em>array-like</em>)  The rows to keep as a boolean array. Must
be the same length as the table, and convertible to a numpy
array of dtype bool.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The mapping between old and new row IDs as a numpy
array (dtype int32).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a> (dtype=np.int32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.MigrationTable.metadata_schema">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">metadata_schema</span></span><a class="headerlink" href="#tskit.MigrationTable.metadata_schema" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this table.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MigrationTable.metadata_vector">
<span class="sig-name descname"><span class="pre">metadata_vector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_value=&lt;object</span> <span class="pre">object&gt;</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MigrationTable.metadata_vector" title="Link to this definition">#</a></dt>
<dd><p>Returns a numpy array of metadata values obtained by extracting <code class="docutils literal notranslate"><span class="pre">key</span></code>
from each metadata entry, and using <code class="docutils literal notranslate"><span class="pre">default_value</span></code> if the key is
not present. <code class="docutils literal notranslate"><span class="pre">key</span></code> may be a list, in which case nested values are returned.
For instance, <code class="docutils literal notranslate"><span class="pre">key</span> <span class="pre">=</span> <span class="pre">[&quot;a&quot;,</span> <span class="pre">&quot;x&quot;]</span></code> will return an array of
<code class="docutils literal notranslate"><span class="pre">row.metadata[&quot;a&quot;][&quot;x&quot;]</span></code> values, iterated over rows in this table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The name, or a list of names, of metadata entries.</p></li>
<li><p><strong>dtype</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The dtype of the result (can usually be omitted).</p></li>
<li><p><strong>default_value</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><em>object</em></a>)  The value to be inserted if the metadata key
is not present. Note that for numeric columns, a default value of None
will result in a non-numeric array. The default behaviour is to raise
<code class="docutils literal notranslate"><span class="pre">KeyError</span></code> on missing entries.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.MigrationTable.nbytes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nbytes</span></span><a class="headerlink" href="#tskit.MigrationTable.nbytes" title="Link to this definition">#</a></dt>
<dd><p>Returns the total number of bytes required to store the data
in this table. Note that this may not be equal to
the actual memory footprint.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MigrationTable.packset_metadata">
<span class="sig-name descname"><span class="pre">packset_metadata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metadatas</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MigrationTable.packset_metadata" title="Link to this definition">#</a></dt>
<dd><p>Packs the specified list of metadata values and updates the <code class="docutils literal notranslate"><span class="pre">metadata</span></code>
and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> columns. The length of the metadatas array
must be equal to the number of rows in the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>metadatas</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of metadata bytes values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MigrationTable.truncate">
<span class="sig-name descname"><span class="pre">truncate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_rows</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MigrationTable.truncate" title="Link to this definition">#</a></dt>
<dd><p>Truncates this table so that the only the first <code class="docutils literal notranslate"><span class="pre">num_rows</span></code> are retained.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>num_rows</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The number of rows to retain in this table.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<section id="id12">
<h5>Associated row class<a class="headerlink" href="#id12" title="Link to this heading">#</a></h5>
<p>A row returned from a <a class="reference internal" href="#tskit.MigrationTable" title="tskit.MigrationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MigrationTable</span></code></a> is an instance of the following
basic class, where each attribute matches an identically named attribute in the
<a class="reference internal" href="#tskit.Migration" title="tskit.Migration"><code class="xref py py-class docutils literal notranslate"><span class="pre">Migration</span></code></a> class.</p>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.MigrationTableRow">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">MigrationTableRow</span></span><a class="reference internal" href="_modules/tskit/tables.html#MigrationTableRow"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MigrationTableRow" title="Link to this definition">#</a></dt>
<dd><p>A row in a <a class="reference internal" href="#tskit.MigrationTable" title="tskit.MigrationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MigrationTable</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.MigrationTableRow.left">
<span class="sig-name descname"><span class="pre">left</span></span><a class="headerlink" href="#tskit.MigrationTableRow.left" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Migration.left" title="tskit.Migration.left"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Migration.left</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.MigrationTableRow.right">
<span class="sig-name descname"><span class="pre">right</span></span><a class="headerlink" href="#tskit.MigrationTableRow.right" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Migration.right" title="tskit.Migration.right"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Migration.right</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.MigrationTableRow.node">
<span class="sig-name descname"><span class="pre">node</span></span><a class="headerlink" href="#tskit.MigrationTableRow.node" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Migration.node" title="tskit.Migration.node"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Migration.node</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.MigrationTableRow.source">
<span class="sig-name descname"><span class="pre">source</span></span><a class="headerlink" href="#tskit.MigrationTableRow.source" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Migration.source" title="tskit.Migration.source"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Migration.source</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.MigrationTableRow.dest">
<span class="sig-name descname"><span class="pre">dest</span></span><a class="headerlink" href="#tskit.MigrationTableRow.dest" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Migration.dest" title="tskit.Migration.dest"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Migration.dest</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.MigrationTableRow.time">
<span class="sig-name descname"><span class="pre">time</span></span><a class="headerlink" href="#tskit.MigrationTableRow.time" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Migration.time" title="tskit.Migration.time"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Migration.time</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.MigrationTableRow.metadata">
<span class="sig-name descname"><span class="pre">metadata</span></span><a class="headerlink" href="#tskit.MigrationTableRow.metadata" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Migration.metadata" title="tskit.Migration.metadata"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Migration.metadata</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MigrationTableRow.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MigrationTableRow.asdict" title="Link to this definition">#</a></dt>
<dd><p>Return a new dict which maps field names to their corresponding values
in this dataclass.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MigrationTableRow.replace">
<span class="sig-name descname"><span class="pre">replace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MigrationTableRow.replace" title="Link to this definition">#</a></dt>
<dd><p>Return a new instance of this dataclass, with the specified attributes
overwritten by new values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new instance of the same type</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
</section>
<section id="sitetable-classes">
<h4><a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code></a> classes<a class="headerlink" href="#sitetable-classes" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.SiteTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">SiteTable</span></span><a class="reference internal" href="_modules/tskit/tables.html#SiteTable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.SiteTable" title="Link to this definition">#</a></dt>
<dd><p>A table defining the sites in a tree sequence. See the
<a class="reference internal" href="data-model.html#sec-site-table-definition"><span class="std std-ref">definitions</span></a> for details on the columns
in this table and the
<a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">tree sequence requirements</span></a> section
for the properties needed for a site table to be a part of a valid tree
sequence.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The numpy arrays returned by table attribute accesses are copies
of the underlying data. In particular, this means that editing
individual values in the arrays will not change the table data
Instead, you should set entire columns or rows at once
(see <a class="reference internal" href="#sec-tables-api-accessing-table-data"><span class="std std-ref">Accessing table data</span></a>).</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Variables<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>position</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The array of site position coordinates.</p></li>
<li><p><strong>ancestral_state</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened array of ancestral state strings.
See <a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more details.</p></li>
<li><p><strong>ancestral_state_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>)  The offsets of rows in the ancestral_state
array. See <a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more details.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened array of binary metadata values. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>)  The array of offsets into the metadata column. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_schema</strong> (<a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><em>tskit.MetadataSchema</em></a>)  The metadata schema for this tables metadata column</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tskit.SiteTable.add_row">
<span class="sig-name descname"><span class="pre">add_row</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">position</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ancestral_state</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#SiteTable.add_row"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.SiteTable.add_row" title="Link to this definition">#</a></dt>
<dd><p>Adds a new row to this <a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code></a> and returns the ID of the
corresponding site. Metadata, if specified, will be validated and encoded
according to the tables
<a class="reference internal" href="#tskit.SiteTable.metadata_schema" title="tskit.SiteTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>position</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The position of this site in genome coordinates.</p></li>
<li><p><strong>ancestral_state</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The state of this site at the root of the tree.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><em>object</em></a>)  Any object that is valid metadata for the tables schema.
Defaults to the default metadata value for the tables schema. This is
typically <code class="docutils literal notranslate"><span class="pre">{}</span></code>. For no schema, <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The ID of the newly added site.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.SiteTable.set_columns">
<span class="sig-name descname"><span class="pre">set_columns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">position</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ancestral_state</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ancestral_state_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_schema</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#SiteTable.set_columns"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.SiteTable.set_columns" title="Link to this definition">#</a></dt>
<dd><p>Sets the values for each column in this <a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code></a> using the values
in the specified arrays. Overwrites existing data in all the table columns.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">position</span></code>, <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> and <code class="docutils literal notranslate"><span class="pre">ancestral_state_offset</span></code>
parameters are mandatory, and must be 1D numpy arrays. The length
of the <code class="docutils literal notranslate"><span class="pre">position</span></code> array determines the number of rows in table.
The <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> and <code class="docutils literal notranslate"><span class="pre">ancestral_state_offset</span></code> parameters must
be supplied together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more information). The
<code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied
together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information) and
<a class="reference external" href="https://tskit.dev/tutorials/metadata.html#sec-tutorial-metadata-bulk" title="(in Project name not set)"><span>Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>position</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The position of each site in genome coordinates.</p></li>
<li><p><strong>ancestral_state</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened ancestral_state array. Required.</p></li>
<li><p><strong>ancestral_state_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> array.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each node.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
<li><p><strong>metadata_schema</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The encoded metadata schema. If None (default)
do not overwrite the exising schema. Note that a schema will need to be
encoded as a string, e.g. via <code class="docutils literal notranslate"><span class="pre">repr(new_metadata_schema)</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.SiteTable.append_columns">
<span class="sig-name descname"><span class="pre">append_columns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">position</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ancestral_state</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ancestral_state_offset</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#SiteTable.append_columns"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.SiteTable.append_columns" title="Link to this definition">#</a></dt>
<dd><p>Appends the specified arrays to the end of the columns of this
<a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code></a>. This allows many new rows to be added at once.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">position</span></code>, <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> and <code class="docutils literal notranslate"><span class="pre">ancestral_state_offset</span></code>
parameters are mandatory, and must be 1D numpy arrays. The length
of the <code class="docutils literal notranslate"><span class="pre">position</span></code> array determines the number of additional rows
to add the table.
The <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> and <code class="docutils literal notranslate"><span class="pre">ancestral_state_offset</span></code> parameters must
be supplied together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more information). The
<code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied
together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information) and
<a class="reference external" href="https://tskit.dev/tutorials/metadata.html#sec-tutorial-metadata-bulk" title="(in Project name not set)"><span>Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>position</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The position of each site in genome coordinates.</p></li>
<li><p><strong>ancestral_state</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened ancestral_state array. Required.</p></li>
<li><p><strong>ancestral_state_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> array.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each node.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.SiteTable.packset_ancestral_state">
<span class="sig-name descname"><span class="pre">packset_ancestral_state</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ancestral_states</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#SiteTable.packset_ancestral_state"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.SiteTable.packset_ancestral_state" title="Link to this definition">#</a></dt>
<dd><p>Packs the specified list of ancestral_state values and updates the
<code class="docutils literal notranslate"><span class="pre">ancestral_state</span></code> and <code class="docutils literal notranslate"><span class="pre">ancestral_state_offset</span></code> columns. The length
of the ancestral_states array must be equal to the number of rows in
the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ancestral_states</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>)</em>)  A list of string ancestral state values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.SiteTable.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.SiteTable.__getitem__" title="Link to this definition">#</a></dt>
<dd><p>If passed an integer, return the specified row of this table, decoding metadata
if it is present. Supports negative indexing, e.g. <code class="docutils literal notranslate"><span class="pre">table[-5]</span></code>.
If passed a slice, iterable or array return a new table containing the specified
rows. Similar to numpy fancy indexing, if the array or iterables contains
booleans then the index acts as a mask, returning those rows for which the mask
is True. Note that as the result is a new table, the row ids will change as tskit
row ids are row indexes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong>  the index of a desired row, a slice of the desired rows, an
iterable or array of the desired row numbers, or a boolean array to use as
a mask.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.SiteTable.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.SiteTable.append" title="Link to this definition">#</a></dt>
<dd><p>Adds a new row to this table and returns the ID of the new row. Metadata, if
specified, will be validated and encoded according to the tables
<a class="reference internal" href="#tskit.IndividualTable.metadata_schema" title="tskit.IndividualTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>row</strong> (<em>row-like</em>)  An object that has attributes corresponding to the
properties of the new row. Both the objects returned from <code class="docutils literal notranslate"><span class="pre">table[i]</span></code> and
e.g. <code class="docutils literal notranslate"><span class="pre">ts.individual(i)</span></code> work for this purpose, along with any other
object with the correct attributes.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The index of the newly added row.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.SiteTable.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.SiteTable.asdict" title="Link to this definition">#</a></dt>
<dd><p>Returns a dictionary mapping the names of the columns in this table
to the corresponding numpy arrays.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.SiteTable.assert_equals">
<span class="sig-name descname"><span class="pre">assert_equals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.SiteTable.assert_equals" title="Link to this definition">#</a></dt>
<dd><p>Raise an AssertionError for the first found difference between
this and another table of the same type.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong>  Another table instance</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True exclude metadata and metadata schemas
from the comparison.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.SiteTable.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.SiteTable.clear" title="Link to this definition">#</a></dt>
<dd><p>Deletes all rows in this table.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.SiteTable.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.SiteTable.copy" title="Link to this definition">#</a></dt>
<dd><p>Returns a deep copy of this table</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.SiteTable.drop_metadata">
<span class="sig-name descname"><span class="pre">drop_metadata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_schema</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.SiteTable.drop_metadata" title="Link to this definition">#</a></dt>
<dd><p>Drops all metadata in this table. By default, the schema is also cleared,
except if <code class="docutils literal notranslate"><span class="pre">keep_schema</span></code> is True.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>keep_schema</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  True if the current schema should be kept intact.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.SiteTable.equals">
<span class="sig-name descname"><span class="pre">equals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.SiteTable.equals" title="Link to this definition">#</a></dt>
<dd><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal. By default, two tables
are considered equal if their columns and metadata schemas are
byte-for-byte identical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong>  Another table instance</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True exclude metadata and metadata schemas
from the comparison.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if other is equal to this table; False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.SiteTable.keep_rows">
<span class="sig-name descname"><span class="pre">keep_rows</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">keep</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.SiteTable.keep_rows" title="Link to this definition">#</a></dt>
<dd><p>Updates this table in-place according to the specified boolean
array, and returns the resulting mapping from old to new row IDs.
For each row <code class="docutils literal notranslate"><span class="pre">j</span></code>, if <code class="docutils literal notranslate"><span class="pre">keep[j]</span></code> is True, that row will be
retained in the output; otherwise, the row will be deleted.
Rows are retained in their original ordering.</p>
<p>The returned <code class="docutils literal notranslate"><span class="pre">id_map</span></code> is an array of the same length as
this table before the operation, such that <code class="docutils literal notranslate"><span class="pre">id_map[j]</span> <span class="pre">=</span> <span class="pre">-1</span></code>
(<a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a>) if row <code class="docutils literal notranslate"><span class="pre">j</span></code> was deleted, and <code class="docutils literal notranslate"><span class="pre">id_map[j]</span></code>
is the new ID of that row, otherwise.</p>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>This needs some examples to link to. See
<a class="github reference external" href="https://github.com/tskit-dev/tskit/issues/2708">tskit-dev/tskit#2708</a></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>keep</strong> (<em>array-like</em>)  The rows to keep as a boolean array. Must
be the same length as the table, and convertible to a numpy
array of dtype bool.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The mapping between old and new row IDs as a numpy
array (dtype int32).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a> (dtype=np.int32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.SiteTable.metadata_schema">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">metadata_schema</span></span><a class="headerlink" href="#tskit.SiteTable.metadata_schema" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this table.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.SiteTable.metadata_vector">
<span class="sig-name descname"><span class="pre">metadata_vector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_value=&lt;object</span> <span class="pre">object&gt;</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.SiteTable.metadata_vector" title="Link to this definition">#</a></dt>
<dd><p>Returns a numpy array of metadata values obtained by extracting <code class="docutils literal notranslate"><span class="pre">key</span></code>
from each metadata entry, and using <code class="docutils literal notranslate"><span class="pre">default_value</span></code> if the key is
not present. <code class="docutils literal notranslate"><span class="pre">key</span></code> may be a list, in which case nested values are returned.
For instance, <code class="docutils literal notranslate"><span class="pre">key</span> <span class="pre">=</span> <span class="pre">[&quot;a&quot;,</span> <span class="pre">&quot;x&quot;]</span></code> will return an array of
<code class="docutils literal notranslate"><span class="pre">row.metadata[&quot;a&quot;][&quot;x&quot;]</span></code> values, iterated over rows in this table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The name, or a list of names, of metadata entries.</p></li>
<li><p><strong>dtype</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The dtype of the result (can usually be omitted).</p></li>
<li><p><strong>default_value</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><em>object</em></a>)  The value to be inserted if the metadata key
is not present. Note that for numeric columns, a default value of None
will result in a non-numeric array. The default behaviour is to raise
<code class="docutils literal notranslate"><span class="pre">KeyError</span></code> on missing entries.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.SiteTable.nbytes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nbytes</span></span><a class="headerlink" href="#tskit.SiteTable.nbytes" title="Link to this definition">#</a></dt>
<dd><p>Returns the total number of bytes required to store the data
in this table. Note that this may not be equal to
the actual memory footprint.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.SiteTable.packset_metadata">
<span class="sig-name descname"><span class="pre">packset_metadata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metadatas</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.SiteTable.packset_metadata" title="Link to this definition">#</a></dt>
<dd><p>Packs the specified list of metadata values and updates the <code class="docutils literal notranslate"><span class="pre">metadata</span></code>
and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> columns. The length of the metadatas array
must be equal to the number of rows in the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>metadatas</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of metadata bytes values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.SiteTable.truncate">
<span class="sig-name descname"><span class="pre">truncate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_rows</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.SiteTable.truncate" title="Link to this definition">#</a></dt>
<dd><p>Truncates this table so that the only the first <code class="docutils literal notranslate"><span class="pre">num_rows</span></code> are retained.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>num_rows</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The number of rows to retain in this table.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<section id="id13">
<h5>Associated row class<a class="headerlink" href="#id13" title="Link to this heading">#</a></h5>
<p>A row returned from a <a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code></a> is an instance of the following
basic class, where each attribute matches an identically named attribute in the
<a class="reference internal" href="#tskit.Site" title="tskit.Site"><code class="xref py py-class docutils literal notranslate"><span class="pre">Site</span></code></a> class.</p>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.SiteTableRow">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">SiteTableRow</span></span><a class="reference internal" href="_modules/tskit/tables.html#SiteTableRow"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.SiteTableRow" title="Link to this definition">#</a></dt>
<dd><p>A row in a <a class="reference internal" href="#tskit.SiteTable" title="tskit.SiteTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.SiteTableRow.position">
<span class="sig-name descname"><span class="pre">position</span></span><a class="headerlink" href="#tskit.SiteTableRow.position" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Site.position" title="tskit.Site.position"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Site.position</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.SiteTableRow.ancestral_state">
<span class="sig-name descname"><span class="pre">ancestral_state</span></span><a class="headerlink" href="#tskit.SiteTableRow.ancestral_state" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Site.ancestral_state" title="tskit.Site.ancestral_state"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Site.ancestral_state</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.SiteTableRow.metadata">
<span class="sig-name descname"><span class="pre">metadata</span></span><a class="headerlink" href="#tskit.SiteTableRow.metadata" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Site.metadata" title="tskit.Site.metadata"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Site.metadata</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.SiteTableRow.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.SiteTableRow.asdict" title="Link to this definition">#</a></dt>
<dd><p>Return a new dict which maps field names to their corresponding values
in this dataclass.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.SiteTableRow.replace">
<span class="sig-name descname"><span class="pre">replace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.SiteTableRow.replace" title="Link to this definition">#</a></dt>
<dd><p>Return a new instance of this dataclass, with the specified attributes
overwritten by new values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new instance of the same type</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
</section>
<section id="mutationtable-classes">
<h4><a class="reference internal" href="#tskit.MutationTable" title="tskit.MutationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutationTable</span></code></a> classes<a class="headerlink" href="#mutationtable-classes" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.MutationTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">MutationTable</span></span><a class="reference internal" href="_modules/tskit/tables.html#MutationTable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MutationTable" title="Link to this definition">#</a></dt>
<dd><p>A table defining the mutations in a tree sequence. See the
<a class="reference internal" href="data-model.html#sec-mutation-table-definition"><span class="std std-ref">definitions</span></a> for details on the columns
in this table and the
<a class="reference internal" href="data-model.html#sec-valid-tree-sequence-requirements"><span class="std std-ref">tree sequence requirements</span></a> section
for the properties needed for a mutation table to be a part of a valid tree
sequence.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The numpy arrays returned by table attribute accesses are copies
of the underlying data. In particular, this means that editing
individual values in the arrays will not change the table data
Instead, you should set entire columns or rows at once
(see <a class="reference internal" href="#sec-tables-api-accessing-table-data"><span class="std std-ref">Accessing table data</span></a>).</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Variables<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>site</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The array of site IDs.</p></li>
<li><p><strong>node</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The array of node IDs.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The array of time values.</p></li>
<li><p><strong>derived_state</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened array of derived state strings.
See <a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more details.</p></li>
<li><p><strong>derived_state_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>)  The offsets of rows in the derived_state
array. See <a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more details.</p></li>
<li><p><strong>parent</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The array of parent mutation IDs.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened array of binary metadata values. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>)  The array of offsets into the metadata column. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_schema</strong> (<a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><em>tskit.MetadataSchema</em></a>)  The metadata schema for this tables metadata column</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tskit.MutationTable.add_row">
<span class="sig-name descname"><span class="pre">add_row</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">site</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">derived_state</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#MutationTable.add_row"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MutationTable.add_row" title="Link to this definition">#</a></dt>
<dd><p>Adds a new row to this <a class="reference internal" href="#tskit.MutationTable" title="tskit.MutationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutationTable</span></code></a> and returns the ID of the
corresponding mutation. Metadata, if specified, will be validated and encoded
according to the tables
<a class="reference internal" href="#tskit.MutationTable.metadata_schema" title="tskit.MutationTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>site</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The ID of the site that this mutation occurs at.</p></li>
<li><p><strong>node</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The ID of the first node inheriting this mutation.</p></li>
<li><p><strong>derived_state</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The state of the site at this mutations node.</p></li>
<li><p><strong>parent</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The ID of the parent mutation. If not specified,
defaults to <a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-attr docutils literal notranslate"><span class="pre">NULL</span></code></a>.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><em>object</em></a>)  Any object that is valid metadata for the tables schema.
Defaults to the default metadata value for the tables schema. This is
typically <code class="docutils literal notranslate"><span class="pre">{}</span></code>. For no schema, <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The occurrence time for the new mutation. If not specified,
defaults to <code class="docutils literal notranslate"><span class="pre">UNKNOWN_TIME</span></code>, indicating the time is unknown.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The ID of the newly added mutation.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MutationTable.set_columns">
<span class="sig-name descname"><span class="pre">set_columns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">site</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">derived_state</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">derived_state_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_schema</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#MutationTable.set_columns"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MutationTable.set_columns" title="Link to this definition">#</a></dt>
<dd><p>Sets the values for each column in this <a class="reference internal" href="#tskit.MutationTable" title="tskit.MutationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutationTable</span></code></a> using the values
in the specified arrays. Overwrites existing data in all the the table columns.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">site</span></code>, <code class="docutils literal notranslate"><span class="pre">node</span></code>, <code class="docutils literal notranslate"><span class="pre">derived_state</span></code> and <code class="docutils literal notranslate"><span class="pre">derived_state_offset</span></code>
parameters are mandatory, and must be 1D numpy arrays. The
<code class="docutils literal notranslate"><span class="pre">site</span></code> and <code class="docutils literal notranslate"><span class="pre">node</span></code> (also <code class="docutils literal notranslate"><span class="pre">parent</span></code> and <code class="docutils literal notranslate"><span class="pre">time</span></code>, if supplied) arrays
must be of equal length, and determine the number of rows in the table.
The <code class="docutils literal notranslate"><span class="pre">derived_state</span></code> and <code class="docutils literal notranslate"><span class="pre">derived_state_offset</span></code> parameters must
be supplied together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more information). The
<code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied
together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information) and
<a class="reference external" href="https://tskit.dev/tutorials/metadata.html#sec-tutorial-metadata-bulk" title="(in Project name not set)"><span>Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>site</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The ID of the site each mutation occurs at.</p></li>
<li><p><strong>node</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The ID of the node each mutation is associated with.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The time values for each mutation.</p></li>
<li><p><strong>derived_state</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened derived_state array. Required.</p></li>
<li><p><strong>derived_state_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">derived_state</span></code> array.</p></li>
<li><p><strong>parent</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The ID of the parent mutation for each mutation.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each node.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
<li><p><strong>metadata_schema</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The encoded metadata schema. If None (default)
do not overwrite the exising schema. Note that a schema will need to be
encoded as a string, e.g. via <code class="docutils literal notranslate"><span class="pre">repr(new_metadata_schema)</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MutationTable.append_columns">
<span class="sig-name descname"><span class="pre">append_columns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">site</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">derived_state</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">derived_state_offset</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#MutationTable.append_columns"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MutationTable.append_columns" title="Link to this definition">#</a></dt>
<dd><p>Appends the specified arrays to the end of the columns of this
<a class="reference internal" href="#tskit.MutationTable" title="tskit.MutationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutationTable</span></code></a>. This allows many new rows to be added at once.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">site</span></code>, <code class="docutils literal notranslate"><span class="pre">node</span></code>, <code class="docutils literal notranslate"><span class="pre">derived_state</span></code> and <code class="docutils literal notranslate"><span class="pre">derived_state_offset</span></code>
parameters are mandatory, and must be 1D numpy arrays. The
<code class="docutils literal notranslate"><span class="pre">site</span></code> and <code class="docutils literal notranslate"><span class="pre">node</span></code> (also <code class="docutils literal notranslate"><span class="pre">time</span></code> and <code class="docutils literal notranslate"><span class="pre">parent</span></code>, if supplied) arrays
must be of equal length, and determine the number of additional
rows to add to the table.
The <code class="docutils literal notranslate"><span class="pre">derived_state</span></code> and <code class="docutils literal notranslate"><span class="pre">derived_state_offset</span></code> parameters must
be supplied together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more information). The
<code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied
together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information) and
<a class="reference external" href="https://tskit.dev/tutorials/metadata.html#sec-tutorial-metadata-bulk" title="(in Project name not set)"><span>Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>site</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The ID of the site each mutation occurs at.</p></li>
<li><p><strong>node</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The ID of the node each mutation is associated with.</p></li>
<li><p><strong>time</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.float64</em>)  The time values for each mutation.</p></li>
<li><p><strong>derived_state</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened derived_state array. Required.</p></li>
<li><p><strong>derived_state_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">derived_state</span></code> array.</p></li>
<li><p><strong>parent</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int32</em>)  The ID of the parent mutation for each mutation.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each node.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MutationTable.packset_derived_state">
<span class="sig-name descname"><span class="pre">packset_derived_state</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">derived_states</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#MutationTable.packset_derived_state"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MutationTable.packset_derived_state" title="Link to this definition">#</a></dt>
<dd><p>Packs the specified list of derived_state values and updates the
<code class="docutils literal notranslate"><span class="pre">derived_state</span></code> and <code class="docutils literal notranslate"><span class="pre">derived_state_offset</span></code> columns. The length
of the derived_states array must be equal to the number of rows in
the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>derived_states</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>)</em>)  A list of string derived state values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MutationTable.keep_rows">
<span class="sig-name descname"><span class="pre">keep_rows</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">keep</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#MutationTable.keep_rows"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MutationTable.keep_rows" title="Link to this definition">#</a></dt>
<dd><p>Updates this table in-place according to the specified boolean
array, and returns the resulting mapping from old to new row IDs.
For each row <code class="docutils literal notranslate"><span class="pre">j</span></code>, if <code class="docutils literal notranslate"><span class="pre">keep[j]</span></code> is True, that row will be
retained in the output; otherwise, the row will be deleted.
Rows are retained in their original ordering.</p>
<p>The returned <code class="docutils literal notranslate"><span class="pre">id_map</span></code> is an array of the same length as
this table before the operation, such that <code class="docutils literal notranslate"><span class="pre">id_map[j]</span> <span class="pre">=</span> <span class="pre">-1</span></code>
(<a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a>) if row <code class="docutils literal notranslate"><span class="pre">j</span></code> was deleted, and <code class="docutils literal notranslate"><span class="pre">id_map[j]</span></code>
is the new ID of that row, otherwise.</p>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>This needs some examples to link to. See
<a class="github reference external" href="https://github.com/tskit-dev/tskit/issues/2708">tskit-dev/tskit#2708</a></p>
</div>
<p>The values in the <code class="docutils literal notranslate"><span class="pre">parent</span></code> column are updated according to this
map, so that reference integrity within the table is maintained.
As a consequence of this, the values in the <code class="docutils literal notranslate"><span class="pre">parent</span></code> column
for kept rows are bounds-checked and an error raised if they
are not valid. Rows that are deleted are not checked for
parent ID integrity.</p>
<p>If an attempt is made to delete rows that are referred to by
the <code class="docutils literal notranslate"><span class="pre">parent</span></code> column of rows that are retained, an error
is raised.</p>
<p>These error conditions are checked before any alterations to
the table are made.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>keep</strong> (<em>array-like</em>)  The rows to keep as a boolean array. Must
be the same length as the table, and convertible to a numpy
array of dtype bool.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The mapping between old and new row IDs as a numpy
array (dtype int32).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a> (dtype=np.int32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MutationTable.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MutationTable.__getitem__" title="Link to this definition">#</a></dt>
<dd><p>If passed an integer, return the specified row of this table, decoding metadata
if it is present. Supports negative indexing, e.g. <code class="docutils literal notranslate"><span class="pre">table[-5]</span></code>.
If passed a slice, iterable or array return a new table containing the specified
rows. Similar to numpy fancy indexing, if the array or iterables contains
booleans then the index acts as a mask, returning those rows for which the mask
is True. Note that as the result is a new table, the row ids will change as tskit
row ids are row indexes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong>  the index of a desired row, a slice of the desired rows, an
iterable or array of the desired row numbers, or a boolean array to use as
a mask.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MutationTable.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MutationTable.append" title="Link to this definition">#</a></dt>
<dd><p>Adds a new row to this table and returns the ID of the new row. Metadata, if
specified, will be validated and encoded according to the tables
<a class="reference internal" href="#tskit.IndividualTable.metadata_schema" title="tskit.IndividualTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>row</strong> (<em>row-like</em>)  An object that has attributes corresponding to the
properties of the new row. Both the objects returned from <code class="docutils literal notranslate"><span class="pre">table[i]</span></code> and
e.g. <code class="docutils literal notranslate"><span class="pre">ts.individual(i)</span></code> work for this purpose, along with any other
object with the correct attributes.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The index of the newly added row.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MutationTable.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MutationTable.asdict" title="Link to this definition">#</a></dt>
<dd><p>Returns a dictionary mapping the names of the columns in this table
to the corresponding numpy arrays.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MutationTable.assert_equals">
<span class="sig-name descname"><span class="pre">assert_equals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MutationTable.assert_equals" title="Link to this definition">#</a></dt>
<dd><p>Raise an AssertionError for the first found difference between
this and another table of the same type.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong>  Another table instance</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True exclude metadata and metadata schemas
from the comparison.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MutationTable.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MutationTable.clear" title="Link to this definition">#</a></dt>
<dd><p>Deletes all rows in this table.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MutationTable.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MutationTable.copy" title="Link to this definition">#</a></dt>
<dd><p>Returns a deep copy of this table</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MutationTable.drop_metadata">
<span class="sig-name descname"><span class="pre">drop_metadata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_schema</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MutationTable.drop_metadata" title="Link to this definition">#</a></dt>
<dd><p>Drops all metadata in this table. By default, the schema is also cleared,
except if <code class="docutils literal notranslate"><span class="pre">keep_schema</span></code> is True.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>keep_schema</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  True if the current schema should be kept intact.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MutationTable.equals">
<span class="sig-name descname"><span class="pre">equals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MutationTable.equals" title="Link to this definition">#</a></dt>
<dd><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal. By default, two tables
are considered equal if their columns and metadata schemas are
byte-for-byte identical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong>  Another table instance</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True exclude metadata and metadata schemas
from the comparison.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if other is equal to this table; False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.MutationTable.metadata_schema">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">metadata_schema</span></span><a class="headerlink" href="#tskit.MutationTable.metadata_schema" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this table.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MutationTable.metadata_vector">
<span class="sig-name descname"><span class="pre">metadata_vector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_value=&lt;object</span> <span class="pre">object&gt;</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MutationTable.metadata_vector" title="Link to this definition">#</a></dt>
<dd><p>Returns a numpy array of metadata values obtained by extracting <code class="docutils literal notranslate"><span class="pre">key</span></code>
from each metadata entry, and using <code class="docutils literal notranslate"><span class="pre">default_value</span></code> if the key is
not present. <code class="docutils literal notranslate"><span class="pre">key</span></code> may be a list, in which case nested values are returned.
For instance, <code class="docutils literal notranslate"><span class="pre">key</span> <span class="pre">=</span> <span class="pre">[&quot;a&quot;,</span> <span class="pre">&quot;x&quot;]</span></code> will return an array of
<code class="docutils literal notranslate"><span class="pre">row.metadata[&quot;a&quot;][&quot;x&quot;]</span></code> values, iterated over rows in this table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The name, or a list of names, of metadata entries.</p></li>
<li><p><strong>dtype</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The dtype of the result (can usually be omitted).</p></li>
<li><p><strong>default_value</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><em>object</em></a>)  The value to be inserted if the metadata key
is not present. Note that for numeric columns, a default value of None
will result in a non-numeric array. The default behaviour is to raise
<code class="docutils literal notranslate"><span class="pre">KeyError</span></code> on missing entries.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.MutationTable.nbytes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nbytes</span></span><a class="headerlink" href="#tskit.MutationTable.nbytes" title="Link to this definition">#</a></dt>
<dd><p>Returns the total number of bytes required to store the data
in this table. Note that this may not be equal to
the actual memory footprint.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MutationTable.packset_metadata">
<span class="sig-name descname"><span class="pre">packset_metadata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metadatas</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MutationTable.packset_metadata" title="Link to this definition">#</a></dt>
<dd><p>Packs the specified list of metadata values and updates the <code class="docutils literal notranslate"><span class="pre">metadata</span></code>
and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> columns. The length of the metadatas array
must be equal to the number of rows in the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>metadatas</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of metadata bytes values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MutationTable.truncate">
<span class="sig-name descname"><span class="pre">truncate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_rows</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MutationTable.truncate" title="Link to this definition">#</a></dt>
<dd><p>Truncates this table so that the only the first <code class="docutils literal notranslate"><span class="pre">num_rows</span></code> are retained.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>num_rows</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The number of rows to retain in this table.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<section id="id14">
<h5>Associated row class<a class="headerlink" href="#id14" title="Link to this heading">#</a></h5>
<p>A row returned from a <a class="reference internal" href="#tskit.MutationTable" title="tskit.MutationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutationTable</span></code></a> is an instance of the following
basic class, where each attribute matches an identically named attribute in the
<a class="reference internal" href="#tskit.Mutation" title="tskit.Mutation"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mutation</span></code></a> class.</p>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.MutationTableRow">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">MutationTableRow</span></span><a class="reference internal" href="_modules/tskit/tables.html#MutationTableRow"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MutationTableRow" title="Link to this definition">#</a></dt>
<dd><p>A row in a <a class="reference internal" href="#tskit.MutationTable" title="tskit.MutationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutationTable</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.MutationTableRow.site">
<span class="sig-name descname"><span class="pre">site</span></span><a class="headerlink" href="#tskit.MutationTableRow.site" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Mutation.site" title="tskit.Mutation.site"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Mutation.site</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.MutationTableRow.node">
<span class="sig-name descname"><span class="pre">node</span></span><a class="headerlink" href="#tskit.MutationTableRow.node" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Mutation.node" title="tskit.Mutation.node"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Mutation.node</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.MutationTableRow.derived_state">
<span class="sig-name descname"><span class="pre">derived_state</span></span><a class="headerlink" href="#tskit.MutationTableRow.derived_state" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Mutation.derived_state" title="tskit.Mutation.derived_state"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Mutation.derived_state</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.MutationTableRow.parent">
<span class="sig-name descname"><span class="pre">parent</span></span><a class="headerlink" href="#tskit.MutationTableRow.parent" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Mutation.parent" title="tskit.Mutation.parent"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Mutation.parent</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.MutationTableRow.metadata">
<span class="sig-name descname"><span class="pre">metadata</span></span><a class="headerlink" href="#tskit.MutationTableRow.metadata" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Mutation.metadata" title="tskit.Mutation.metadata"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Mutation.metadata</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.MutationTableRow.time">
<span class="sig-name descname"><span class="pre">time</span></span><a class="headerlink" href="#tskit.MutationTableRow.time" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Mutation.time" title="tskit.Mutation.time"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Mutation.time</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MutationTableRow.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MutationTableRow.asdict" title="Link to this definition">#</a></dt>
<dd><p>Return a new dict which maps field names to their corresponding values
in this dataclass.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MutationTableRow.replace">
<span class="sig-name descname"><span class="pre">replace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.MutationTableRow.replace" title="Link to this definition">#</a></dt>
<dd><p>Return a new instance of this dataclass, with the specified attributes
overwritten by new values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new instance of the same type</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
</section>
<section id="populationtable-classes">
<h4><a class="reference internal" href="#tskit.PopulationTable" title="tskit.PopulationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationTable</span></code></a> classes<a class="headerlink" href="#populationtable-classes" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.PopulationTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">PopulationTable</span></span><a class="reference internal" href="_modules/tskit/tables.html#PopulationTable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.PopulationTable" title="Link to this definition">#</a></dt>
<dd><p>A table defining the populations referred to in a tree sequence.
The PopulationTable stores metadata for populations that may be referred to
in the NodeTable and MigrationTable.  Note that although nodes
may be associated with populations, this association is stored in
the <a class="reference internal" href="#tskit.NodeTable" title="tskit.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code></a>: only metadata on each population is stored
in the population table.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The numpy arrays returned by table attribute accesses are copies
of the underlying data. In particular, this means that editing
individual values in the arrays will not change the table data
Instead, you should set entire columns or rows at once
(see <a class="reference internal" href="#sec-tables-api-accessing-table-data"><span class="std std-ref">Accessing table data</span></a>).</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Variables<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened array of binary metadata values. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>)  The array of offsets into the metadata column. See
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more details.</p></li>
<li><p><strong>metadata_schema</strong> (<a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><em>tskit.MetadataSchema</em></a>)  The metadata schema for this tables metadata column</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tskit.PopulationTable.add_row">
<span class="sig-name descname"><span class="pre">add_row</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#PopulationTable.add_row"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.PopulationTable.add_row" title="Link to this definition">#</a></dt>
<dd><p>Adds a new row to this <a class="reference internal" href="#tskit.PopulationTable" title="tskit.PopulationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationTable</span></code></a> and returns the ID of the
corresponding population. Metadata, if specified, will be validated and encoded
according to the tables
<a class="reference internal" href="#tskit.PopulationTable.metadata_schema" title="tskit.PopulationTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><em>object</em></a>)  Any object that is valid metadata for the tables schema.
Defaults to the default metadata value for the tables schema. This is
typically <code class="docutils literal notranslate"><span class="pre">{}</span></code>. For no schema, <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The ID of the newly added population.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.PopulationTable.set_columns">
<span class="sig-name descname"><span class="pre">set_columns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_schema</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#PopulationTable.set_columns"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.PopulationTable.set_columns" title="Link to this definition">#</a></dt>
<dd><p>Sets the values for each column in this <a class="reference internal" href="#tskit.PopulationTable" title="tskit.PopulationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationTable</span></code></a> using the
values in the specified arrays. Overwrites existing data in all the table
columns.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied
together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information) and
<a class="reference external" href="https://tskit.dev/tutorials/metadata.html#sec-tutorial-metadata-bulk" title="(in Project name not set)"><span>Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each node.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
<li><p><strong>metadata_schema</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The encoded metadata schema. If None (default)
do not overwrite the exising schema. Note that a schema will need to be
encoded as a string, e.g. via <code class="docutils literal notranslate"><span class="pre">repr(new_metadata_schema)</span></code>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.PopulationTable.append_columns">
<span class="sig-name descname"><span class="pre">append_columns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#PopulationTable.append_columns"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.PopulationTable.append_columns" title="Link to this definition">#</a></dt>
<dd><p>Appends the specified arrays to the end of the columns of this
<a class="reference internal" href="#tskit.PopulationTable" title="tskit.PopulationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationTable</span></code></a>. This allows many new rows to be added at once.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> parameters must be supplied
together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information) and
<a class="reference external" href="https://tskit.dev/tutorials/metadata.html#sec-tutorial-metadata-bulk" title="(in Project name not set)"><span>Metadata for bulk table methods</span></a> for an example of how to prepare metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>metadata</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened metadata array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code>. If not specified or None, an empty metadata
value is stored for each node.</p></li>
<li><p><strong>metadata_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> array.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.PopulationTable.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.PopulationTable.__getitem__" title="Link to this definition">#</a></dt>
<dd><p>If passed an integer, return the specified row of this table, decoding metadata
if it is present. Supports negative indexing, e.g. <code class="docutils literal notranslate"><span class="pre">table[-5]</span></code>.
If passed a slice, iterable or array return a new table containing the specified
rows. Similar to numpy fancy indexing, if the array or iterables contains
booleans then the index acts as a mask, returning those rows for which the mask
is True. Note that as the result is a new table, the row ids will change as tskit
row ids are row indexes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>index</strong>  the index of a desired row, a slice of the desired rows, an
iterable or array of the desired row numbers, or a boolean array to use as
a mask.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.PopulationTable.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.PopulationTable.append" title="Link to this definition">#</a></dt>
<dd><p>Adds a new row to this table and returns the ID of the new row. Metadata, if
specified, will be validated and encoded according to the tables
<a class="reference internal" href="#tskit.IndividualTable.metadata_schema" title="tskit.IndividualTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>row</strong> (<em>row-like</em>)  An object that has attributes corresponding to the
properties of the new row. Both the objects returned from <code class="docutils literal notranslate"><span class="pre">table[i]</span></code> and
e.g. <code class="docutils literal notranslate"><span class="pre">ts.individual(i)</span></code> work for this purpose, along with any other
object with the correct attributes.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The index of the newly added row.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.PopulationTable.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.PopulationTable.asdict" title="Link to this definition">#</a></dt>
<dd><p>Returns a dictionary mapping the names of the columns in this table
to the corresponding numpy arrays.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.PopulationTable.assert_equals">
<span class="sig-name descname"><span class="pre">assert_equals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.PopulationTable.assert_equals" title="Link to this definition">#</a></dt>
<dd><p>Raise an AssertionError for the first found difference between
this and another table of the same type.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong>  Another table instance</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True exclude metadata and metadata schemas
from the comparison.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.PopulationTable.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.PopulationTable.clear" title="Link to this definition">#</a></dt>
<dd><p>Deletes all rows in this table.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.PopulationTable.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.PopulationTable.copy" title="Link to this definition">#</a></dt>
<dd><p>Returns a deep copy of this table</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.PopulationTable.drop_metadata">
<span class="sig-name descname"><span class="pre">drop_metadata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_schema</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.PopulationTable.drop_metadata" title="Link to this definition">#</a></dt>
<dd><p>Drops all metadata in this table. By default, the schema is also cleared,
except if <code class="docutils literal notranslate"><span class="pre">keep_schema</span></code> is True.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>keep_schema</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  True if the current schema should be kept intact.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.PopulationTable.equals">
<span class="sig-name descname"><span class="pre">equals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.PopulationTable.equals" title="Link to this definition">#</a></dt>
<dd><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal. By default, two tables
are considered equal if their columns and metadata schemas are
byte-for-byte identical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong>  Another table instance</p></li>
<li><p><strong>ignore_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True exclude metadata and metadata schemas
from the comparison.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if other is equal to this table; False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.PopulationTable.keep_rows">
<span class="sig-name descname"><span class="pre">keep_rows</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">keep</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.PopulationTable.keep_rows" title="Link to this definition">#</a></dt>
<dd><p>Updates this table in-place according to the specified boolean
array, and returns the resulting mapping from old to new row IDs.
For each row <code class="docutils literal notranslate"><span class="pre">j</span></code>, if <code class="docutils literal notranslate"><span class="pre">keep[j]</span></code> is True, that row will be
retained in the output; otherwise, the row will be deleted.
Rows are retained in their original ordering.</p>
<p>The returned <code class="docutils literal notranslate"><span class="pre">id_map</span></code> is an array of the same length as
this table before the operation, such that <code class="docutils literal notranslate"><span class="pre">id_map[j]</span> <span class="pre">=</span> <span class="pre">-1</span></code>
(<a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a>) if row <code class="docutils literal notranslate"><span class="pre">j</span></code> was deleted, and <code class="docutils literal notranslate"><span class="pre">id_map[j]</span></code>
is the new ID of that row, otherwise.</p>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>This needs some examples to link to. See
<a class="github reference external" href="https://github.com/tskit-dev/tskit/issues/2708">tskit-dev/tskit#2708</a></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>keep</strong> (<em>array-like</em>)  The rows to keep as a boolean array. Must
be the same length as the table, and convertible to a numpy
array of dtype bool.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The mapping between old and new row IDs as a numpy
array (dtype int32).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a> (dtype=np.int32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.PopulationTable.metadata_schema">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">metadata_schema</span></span><a class="headerlink" href="#tskit.PopulationTable.metadata_schema" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this table.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.PopulationTable.metadata_vector">
<span class="sig-name descname"><span class="pre">metadata_vector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_value=&lt;object</span> <span class="pre">object&gt;</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.PopulationTable.metadata_vector" title="Link to this definition">#</a></dt>
<dd><p>Returns a numpy array of metadata values obtained by extracting <code class="docutils literal notranslate"><span class="pre">key</span></code>
from each metadata entry, and using <code class="docutils literal notranslate"><span class="pre">default_value</span></code> if the key is
not present. <code class="docutils literal notranslate"><span class="pre">key</span></code> may be a list, in which case nested values are returned.
For instance, <code class="docutils literal notranslate"><span class="pre">key</span> <span class="pre">=</span> <span class="pre">[&quot;a&quot;,</span> <span class="pre">&quot;x&quot;]</span></code> will return an array of
<code class="docutils literal notranslate"><span class="pre">row.metadata[&quot;a&quot;][&quot;x&quot;]</span></code> values, iterated over rows in this table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The name, or a list of names, of metadata entries.</p></li>
<li><p><strong>dtype</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  The dtype of the result (can usually be omitted).</p></li>
<li><p><strong>default_value</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.14)"><em>object</em></a>)  The value to be inserted if the metadata key
is not present. Note that for numeric columns, a default value of None
will result in a non-numeric array. The default behaviour is to raise
<code class="docutils literal notranslate"><span class="pre">KeyError</span></code> on missing entries.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.PopulationTable.nbytes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nbytes</span></span><a class="headerlink" href="#tskit.PopulationTable.nbytes" title="Link to this definition">#</a></dt>
<dd><p>Returns the total number of bytes required to store the data
in this table. Note that this may not be equal to
the actual memory footprint.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.PopulationTable.packset_metadata">
<span class="sig-name descname"><span class="pre">packset_metadata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metadatas</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.PopulationTable.packset_metadata" title="Link to this definition">#</a></dt>
<dd><p>Packs the specified list of metadata values and updates the <code class="docutils literal notranslate"><span class="pre">metadata</span></code>
and <code class="docutils literal notranslate"><span class="pre">metadata_offset</span></code> columns. The length of the metadatas array
must be equal to the number of rows in the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>metadatas</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>)  A list of metadata bytes values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.PopulationTable.truncate">
<span class="sig-name descname"><span class="pre">truncate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_rows</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.PopulationTable.truncate" title="Link to this definition">#</a></dt>
<dd><p>Truncates this table so that the only the first <code class="docutils literal notranslate"><span class="pre">num_rows</span></code> are retained.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>num_rows</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The number of rows to retain in this table.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<section id="id15">
<h5>Associated row class<a class="headerlink" href="#id15" title="Link to this heading">#</a></h5>
<p>A row returned from a <a class="reference internal" href="#tskit.PopulationTable" title="tskit.PopulationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationTable</span></code></a> is an instance of the following
basic class, where each attribute matches an identically named attribute in the
<a class="reference internal" href="#tskit.Population" title="tskit.Population"><code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code></a> class.</p>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.PopulationTableRow">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">PopulationTableRow</span></span><a class="reference internal" href="_modules/tskit/tables.html#PopulationTableRow"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.PopulationTableRow" title="Link to this definition">#</a></dt>
<dd><p>A row in a <a class="reference internal" href="#tskit.PopulationTable" title="tskit.PopulationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationTable</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.PopulationTableRow.metadata">
<span class="sig-name descname"><span class="pre">metadata</span></span><a class="headerlink" href="#tskit.PopulationTableRow.metadata" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Population.metadata" title="tskit.Population.metadata"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Population.metadata</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.PopulationTableRow.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.PopulationTableRow.asdict" title="Link to this definition">#</a></dt>
<dd><p>Return a new dict which maps field names to their corresponding values
in this dataclass.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.PopulationTableRow.replace">
<span class="sig-name descname"><span class="pre">replace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.PopulationTableRow.replace" title="Link to this definition">#</a></dt>
<dd><p>Return a new instance of this dataclass, with the specified attributes
overwritten by new values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new instance of the same type</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
</section>
<section id="provenancetable-classes">
<h4><a class="reference internal" href="#tskit.ProvenanceTable" title="tskit.ProvenanceTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">ProvenanceTable</span></code></a> classes<a class="headerlink" href="#provenancetable-classes" title="Link to this heading">#</a></h4>
<p>Also see the <a class="reference internal" href="provenance.html#sec-provenance"><span class="std std-ref">Provenance</span></a> and
<a class="reference internal" href="#sec-python-api-provenance"><span class="std std-ref">provenance API methods</span></a>.</p>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.ProvenanceTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">ProvenanceTable</span></span><a class="reference internal" href="_modules/tskit/tables.html#ProvenanceTable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.ProvenanceTable" title="Link to this definition">#</a></dt>
<dd><p>A table recording the provenance (i.e., history) of this table, so that the
origin of the underlying data and sequence of subsequent operations can be
traced. Each row contains a record string (recommended format: JSON) and
a timestamp.</p>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>The format of the <cite>record</cite> field will be more precisely specified in
the future.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Variables<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>record</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened array containing the record strings.
<a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more details.</p></li>
<li><p><strong>record_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>)  The array of offsets into the record column. See
<a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more details.</p></li>
<li><p><strong>timestamp</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened array containing the timestamp strings.
<a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more details.</p></li>
<li><p><strong>timestamp_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32</em>)  The array of offsets into the timestamp column. See
<a class="reference internal" href="#sec-tables-api-text-columns"><span class="std std-ref">Text columns</span></a> for more details.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tskit.ProvenanceTable.add_row">
<span class="sig-name descname"><span class="pre">add_row</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timestamp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#ProvenanceTable.add_row"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.ProvenanceTable.add_row" title="Link to this definition">#</a></dt>
<dd><p>Adds a new row to this ProvenanceTable consisting of the specified record and
timestamp. If timestamp is not specified, it is automatically generated from
the current time.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>record</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A provenance record, describing the parameters and
environment used to generate the current set of tables.</p></li>
<li><p><strong>timestamp</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>)  A string timestamp. This should be in ISO8601 form.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.ProvenanceTable.set_columns">
<span class="sig-name descname"><span class="pre">set_columns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">timestamp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timestamp_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#ProvenanceTable.set_columns"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.ProvenanceTable.set_columns" title="Link to this definition">#</a></dt>
<dd><p>Sets the values for each column in this <a class="reference internal" href="#tskit.ProvenanceTable" title="tskit.ProvenanceTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">ProvenanceTable</span></code></a> using the
values in the specified arrays. Overwrites existing data in all the table
columns.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">timestamp</span></code> and <code class="docutils literal notranslate"><span class="pre">timestamp_offset</span></code> parameters must be supplied
together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information). Likewise
for the <code class="docutils literal notranslate"><span class="pre">record</span></code> and <code class="docutils literal notranslate"><span class="pre">record_offset</span></code> columns</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>timestamp</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened timestamp array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">timestamp_offset</span></code>. If not specified or None, an empty timestamp
value is stored for each node.</p></li>
<li><p><strong>timestamp_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">timestamp</span></code> array.</p></li>
<li><p><strong>record</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened record array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">record_offset</span></code>. If not specified or None, an empty record
value is stored for each node.</p></li>
<li><p><strong>record_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">record</span></code> array.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.ProvenanceTable.append_columns">
<span class="sig-name descname"><span class="pre">append_columns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">timestamp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timestamp_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#ProvenanceTable.append_columns"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.ProvenanceTable.append_columns" title="Link to this definition">#</a></dt>
<dd><p>Appends the specified arrays to the end of the columns of this
<a class="reference internal" href="#tskit.ProvenanceTable" title="tskit.ProvenanceTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">ProvenanceTable</span></code></a>. This allows many new rows to be added at once.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">timestamp</span></code> and <code class="docutils literal notranslate"><span class="pre">timestamp_offset</span></code> parameters must be supplied
together, and meet the requirements for
<a class="reference internal" href="data-model.html#sec-encoding-ragged-columns"><span class="std std-ref">Encoding ragged columns</span></a> (see
<a class="reference internal" href="#sec-tables-api-binary-columns"><span class="std std-ref">Binary columns</span></a> for more information). Likewise
for the <code class="docutils literal notranslate"><span class="pre">record</span></code> and <code class="docutils literal notranslate"><span class="pre">record_offset</span></code> columns</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>timestamp</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened timestamp array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">timestamp_offset</span></code>. If not specified or None, an empty timestamp
value is stored for each node.</p></li>
<li><p><strong>timestamp_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">timestamp</span></code> array.</p></li>
<li><p><strong>record</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.int8</em>)  The flattened record array. Must be specified along
with <code class="docutils literal notranslate"><span class="pre">record_offset</span></code>. If not specified or None, an empty record
value is stored for each node.</p></li>
<li><p><strong>record_offset</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a><em>, </em><em>dtype=np.uint32.</em>)  The offsets into the <code class="docutils literal notranslate"><span class="pre">record</span></code> array.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.ProvenanceTable.packset_record">
<span class="sig-name descname"><span class="pre">packset_record</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">records</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#ProvenanceTable.packset_record"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.ProvenanceTable.packset_record" title="Link to this definition">#</a></dt>
<dd><p>Packs the specified list of record values and updates the
<code class="docutils literal notranslate"><span class="pre">record</span></code> and <code class="docutils literal notranslate"><span class="pre">record_offset</span></code> columns. The length
of the records array must be equal to the number of rows in
the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>records</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>)</em>)  A list of string record values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.ProvenanceTable.packset_timestamp">
<span class="sig-name descname"><span class="pre">packset_timestamp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">timestamps</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#ProvenanceTable.packset_timestamp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.ProvenanceTable.packset_timestamp" title="Link to this definition">#</a></dt>
<dd><p>Packs the specified list of timestamp values and updates the
<code class="docutils literal notranslate"><span class="pre">timestamp</span></code> and <code class="docutils literal notranslate"><span class="pre">timestamp_offset</span></code> columns. The length
of the timestamps array must be equal to the number of rows in
the table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>timestamps</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a><em>)</em>)  A list of string timestamp values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.ProvenanceTable.equals">
<span class="sig-name descname"><span class="pre">equals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_timestamps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#ProvenanceTable.equals"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.ProvenanceTable.equals" title="Link to this definition">#</a></dt>
<dd><p>Returns True if  <cite>self</cite> and <cite>other</cite> are equal. By default, two provenance
tables are considered equal if their columns are byte-for-byte identical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong>  Another provenance table instance</p></li>
<li><p><strong>ignore_timestamps</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True exclude the timestamp column
from the comparison.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if other is equal to this provenance table; False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.ProvenanceTable.assert_equals">
<span class="sig-name descname"><span class="pre">assert_equals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_timestamps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#ProvenanceTable.assert_equals"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.ProvenanceTable.assert_equals" title="Link to this definition">#</a></dt>
<dd><p>Raise an AssertionError for the first found difference between
this and another provenance table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong>  Another provenance table instance</p></li>
<li><p><strong>ignore_timestamps</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>)  If True exclude the timestamp column
from the comparison.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.ProvenanceTable.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.ProvenanceTable.append" title="Link to this definition">#</a></dt>
<dd><p>Adds a new row to this table and returns the ID of the new row. Metadata, if
specified, will be validated and encoded according to the tables
<a class="reference internal" href="#tskit.IndividualTable.metadata_schema" title="tskit.IndividualTable.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>row</strong> (<em>row-like</em>)  An object that has attributes corresponding to the
properties of the new row. Both the objects returned from <code class="docutils literal notranslate"><span class="pre">table[i]</span></code> and
e.g. <code class="docutils literal notranslate"><span class="pre">ts.individual(i)</span></code> work for this purpose, along with any other
object with the correct attributes.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The index of the newly added row.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.ProvenanceTable.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.ProvenanceTable.asdict" title="Link to this definition">#</a></dt>
<dd><p>Returns a dictionary mapping the names of the columns in this table
to the corresponding numpy arrays.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.ProvenanceTable.clear">
<span class="sig-name descname"><span class="pre">clear</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.ProvenanceTable.clear" title="Link to this definition">#</a></dt>
<dd><p>Deletes all rows in this table.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.ProvenanceTable.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tskit.ProvenanceTable.copy" title="Link to this definition">#</a></dt>
<dd><p>Returns a deep copy of this table</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.ProvenanceTable.keep_rows">
<span class="sig-name descname"><span class="pre">keep_rows</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">keep</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.ProvenanceTable.keep_rows" title="Link to this definition">#</a></dt>
<dd><p>Updates this table in-place according to the specified boolean
array, and returns the resulting mapping from old to new row IDs.
For each row <code class="docutils literal notranslate"><span class="pre">j</span></code>, if <code class="docutils literal notranslate"><span class="pre">keep[j]</span></code> is True, that row will be
retained in the output; otherwise, the row will be deleted.
Rows are retained in their original ordering.</p>
<p>The returned <code class="docutils literal notranslate"><span class="pre">id_map</span></code> is an array of the same length as
this table before the operation, such that <code class="docutils literal notranslate"><span class="pre">id_map[j]</span> <span class="pre">=</span> <span class="pre">-1</span></code>
(<a class="reference internal" href="#tskit.NULL" title="tskit.NULL"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.NULL</span></code></a>) if row <code class="docutils literal notranslate"><span class="pre">j</span></code> was deleted, and <code class="docutils literal notranslate"><span class="pre">id_map[j]</span></code>
is the new ID of that row, otherwise.</p>
<div class="admonition-todo admonition" id="id2">
<p class="admonition-title">Todo</p>
<p>This needs some examples to link to. See
<a class="github reference external" href="https://github.com/tskit-dev/tskit/issues/2708">tskit-dev/tskit#2708</a></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>keep</strong> (<em>array-like</em>)  The rows to keep as a boolean array. Must
be the same length as the table, and convertible to a numpy
array of dtype bool.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The mapping between old and new row IDs as a numpy
array (dtype int32).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a> (dtype=np.int32)</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.ProvenanceTable.nbytes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nbytes</span></span><a class="headerlink" href="#tskit.ProvenanceTable.nbytes" title="Link to this definition">#</a></dt>
<dd><p>Returns the total number of bytes required to store the data
in this table. Note that this may not be equal to
the actual memory footprint.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.ProvenanceTable.truncate">
<span class="sig-name descname"><span class="pre">truncate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_rows</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.ProvenanceTable.truncate" title="Link to this definition">#</a></dt>
<dd><p>Truncates this table so that the only the first <code class="docutils literal notranslate"><span class="pre">num_rows</span></code> are retained.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>num_rows</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The number of rows to retain in this table.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<section id="id16">
<h5>Associated row class<a class="headerlink" href="#id16" title="Link to this heading">#</a></h5>
<p>A row returned from a <a class="reference internal" href="#tskit.ProvenanceTable" title="tskit.ProvenanceTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">ProvenanceTable</span></code></a> is an instance of the following
basic class, where each attribute matches an identically named attribute in the
<a class="reference internal" href="#tskit.Provenance" title="tskit.Provenance"><code class="xref py py-class docutils literal notranslate"><span class="pre">Provenance</span></code></a> class.</p>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.ProvenanceTableRow">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">ProvenanceTableRow</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">timestamp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#ProvenanceTableRow"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.ProvenanceTableRow" title="Link to this definition">#</a></dt>
<dd><p>A row in a <a class="reference internal" href="#tskit.ProvenanceTable" title="tskit.ProvenanceTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">ProvenanceTable</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.ProvenanceTableRow.timestamp">
<span class="sig-name descname"><span class="pre">timestamp</span></span><a class="headerlink" href="#tskit.ProvenanceTableRow.timestamp" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Provenance.timestamp" title="tskit.Provenance.timestamp"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Provenance.timestamp</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.ProvenanceTableRow.record">
<span class="sig-name descname"><span class="pre">record</span></span><a class="headerlink" href="#tskit.ProvenanceTableRow.record" title="Link to this definition">#</a></dt>
<dd><p>See <a class="reference internal" href="#tskit.Provenance.record" title="tskit.Provenance.record"><code class="xref py py-attr docutils literal notranslate"><span class="pre">Provenance.record</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.ProvenanceTableRow.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.ProvenanceTableRow.asdict" title="Link to this definition">#</a></dt>
<dd><p>Return a new dict which maps field names to their corresponding values
in this dataclass.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.ProvenanceTableRow.replace">
<span class="sig-name descname"><span class="pre">replace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tskit.ProvenanceTableRow.replace" title="Link to this definition">#</a></dt>
<dd><p>Return a new instance of this dataclass, with the specified attributes
overwritten by new values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A new instance of the same type</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
</section>
</section>
<section id="sec-python-api-reference-identity">
<span id="identity-classes"></span><h3>Identity classes<a class="headerlink" href="#sec-python-api-reference-identity" title="Link to this heading">#</a></h3>
<p>The classes documented in this section are associated with summarising
identity relationships between pairs of samples. See the <a class="reference internal" href="ibd.html#sec-identity"><span class="std std-ref">Identity by descent</span></a>
section for more details and examples.</p>
<section id="the-identitysegments-class">
<h4>The <a class="reference internal" href="#tskit.IdentitySegments" title="tskit.IdentitySegments"><code class="xref py py-class docutils literal notranslate"><span class="pre">IdentitySegments</span></code></a> class<a class="headerlink" href="#the-identitysegments-class" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.IdentitySegments">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">IdentitySegments</span></span><a class="reference internal" href="_modules/tskit/tables.html#IdentitySegments"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.IdentitySegments" title="Link to this definition">#</a></dt>
<dd><p>A class summarising and optionally storing the segments of identity
by descent returned by <a class="reference internal" href="#tskit.TreeSequence.ibd_segments" title="tskit.TreeSequence.ibd_segments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.ibd_segments()</span></code></a>. See the
<a class="reference internal" href="ibd.html#sec-identity"><span class="std std-ref">Identity by descent</span></a> for more information and examples.</p>
<p>Along with the documented methods and attributes, the class supports
the Python mapping protocol, and can be regarded as a dictionary
mapping sample node pair tuples to the corresponding
<a class="reference internal" href="#tskit.IdentitySegmentList" title="tskit.IdentitySegmentList"><code class="xref py py-class docutils literal notranslate"><span class="pre">IdentitySegmentList</span></code></a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>It is important to note that the facilities available
for a given instance of this class are determined by the
<code class="docutils literal notranslate"><span class="pre">store_pairs</span></code> and <code class="docutils literal notranslate"><span class="pre">store_segments</span></code> arguments provided to
<a class="reference internal" href="#tskit.TreeSequence.ibd_segments" title="tskit.TreeSequence.ibd_segments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.ibd_segments()</span></code></a>. For example, attempting
to access per-sample pair information (such as indexing with
<code class="docutils literal notranslate"><span class="pre">[(a,</span> <span class="pre">b)]</span></code>, iterating over the mapping, or accessing
<a class="reference internal" href="#tskit.IdentitySegments.pairs" title="tskit.IdentitySegments.pairs"><code class="xref py py-attr docutils literal notranslate"><span class="pre">IdentitySegments.pairs</span></code></a>) if <code class="docutils literal notranslate"><span class="pre">store_pairs</span></code> is False will
result in an <code class="docutils literal notranslate"><span class="pre">IdentityPairsNotStoredError</span></code> being raised.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This class should not be instantiated directly.</p>
</div>
<dl class="py property">
<dt class="sig sig-object py" id="tskit.IdentitySegments.num_segments">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_segments</span></span><a class="headerlink" href="#tskit.IdentitySegments.num_segments" title="Link to this definition">#</a></dt>
<dd><p>The total number of identity segments found.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.IdentitySegments.num_pairs">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_pairs</span></span><a class="headerlink" href="#tskit.IdentitySegments.num_pairs" title="Link to this definition">#</a></dt>
<dd><p>The total number of distinct sample pairs for which identity
segments were found. (Only available when <code class="docutils literal notranslate"><span class="pre">store_pairs</span></code> or
<code class="docutils literal notranslate"><span class="pre">store_segments</span></code> is specified).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.IdentitySegments.total_span">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">total_span</span></span><a class="headerlink" href="#tskit.IdentitySegments.total_span" title="Link to this definition">#</a></dt>
<dd><p>The total genomic sequence length spanned by all identity
segments that were found.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.IdentitySegments.pairs">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">pairs</span></span><a class="headerlink" href="#tskit.IdentitySegments.pairs" title="Link to this definition">#</a></dt>
<dd><p>A numpy array with shape <code class="docutils literal notranslate"><span class="pre">(segs.num_pairs,</span> <span class="pre">2)</span></code> and dtype=np.int32
containing the sample pairs for which IBD segments were found.</p>
</dd></dl>

</dd></dl>

</section>
<section id="the-identitysegmentlist-class">
<h4>The <a class="reference internal" href="#tskit.IdentitySegmentList" title="tskit.IdentitySegmentList"><code class="xref py py-class docutils literal notranslate"><span class="pre">IdentitySegmentList</span></code></a> class<a class="headerlink" href="#the-identitysegmentlist-class" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.IdentitySegmentList">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">IdentitySegmentList</span></span><a class="reference internal" href="_modules/tskit/tables.html#IdentitySegmentList"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.IdentitySegmentList" title="Link to this definition">#</a></dt>
<dd><p>A summary of identity-by-descent segments for some pair of samples in a
<a class="reference internal" href="#tskit.IdentitySegments" title="tskit.IdentitySegments"><code class="xref py py-class docutils literal notranslate"><span class="pre">IdentitySegments</span></code></a> result. If the <code class="docutils literal notranslate"><span class="pre">store_segments</span></code> argument
has been specified to <a class="reference internal" href="#tskit.TreeSequence.ibd_segments" title="tskit.TreeSequence.ibd_segments"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TreeSequence.ibd_segments()</span></code></a>, this class
can be treated as a sequence of <a class="reference internal" href="#tskit.IdentitySegment" title="tskit.IdentitySegment"><code class="xref py py-class docutils literal notranslate"><span class="pre">IdentitySegment</span></code></a> objects.</p>
<p>Access to the segment data via numpy arrays is also available via
the <a class="reference internal" href="#tskit.IdentitySegmentList.left" title="tskit.IdentitySegmentList.left"><code class="xref py py-attr docutils literal notranslate"><span class="pre">IdentitySegmentList.left</span></code></a>, <a class="reference internal" href="#tskit.IdentitySegmentList.right" title="tskit.IdentitySegmentList.right"><code class="xref py py-attr docutils literal notranslate"><span class="pre">IdentitySegmentList.right</span></code></a>
and <a class="reference internal" href="#tskit.IdentitySegmentList.node" title="tskit.IdentitySegmentList.node"><code class="xref py py-attr docutils literal notranslate"><span class="pre">IdentitySegmentList.node</span></code></a> attributes.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">store_segments</span></code> is False, only the overall summary values
such as <a class="reference internal" href="#tskit.IdentitySegmentList.total_span" title="tskit.IdentitySegmentList.total_span"><code class="xref py py-attr docutils literal notranslate"><span class="pre">IdentitySegmentList.total_span</span></code></a> and <code class="docutils literal notranslate"><span class="pre">len()</span></code> are
available. Attempting to iterate over the list or access per-segment
arrays (<code class="docutils literal notranslate"><span class="pre">left</span></code>, <code class="docutils literal notranslate"><span class="pre">right</span></code>, or <code class="docutils literal notranslate"><span class="pre">node</span></code>) in this case will raise an
<code class="docutils literal notranslate"><span class="pre">IdentitySegmentsNotStoredError</span></code>.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The order of segments within an IdentitySegmentList is
arbitrary and may change in the future</p>
</div>
<dl class="py property">
<dt class="sig sig-object py" id="tskit.IdentitySegmentList.total_span">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">total_span</span></span><a class="headerlink" href="#tskit.IdentitySegmentList.total_span" title="Link to this definition">#</a></dt>
<dd><p>The total genomic span covered by segments in this list. Equal to
<code class="docutils literal notranslate"><span class="pre">sum(seg.span</span> <span class="pre">for</span> <span class="pre">seg</span> <span class="pre">in</span> <span class="pre">seglst)</span></code>.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.IdentitySegmentList.left">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">left</span></span><a class="headerlink" href="#tskit.IdentitySegmentList.left" title="Link to this definition">#</a></dt>
<dd><p>A numpy array (dtype=np.float64) of the <code class="docutils literal notranslate"><span class="pre">left</span></code> coordinates of segments.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.IdentitySegmentList.right">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">right</span></span><a class="headerlink" href="#tskit.IdentitySegmentList.right" title="Link to this definition">#</a></dt>
<dd><p>A numpy array (dtype=np.float64) of the <code class="docutils literal notranslate"><span class="pre">right</span></code> coordinates of segments.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.IdentitySegmentList.node">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">node</span></span><a class="headerlink" href="#tskit.IdentitySegmentList.node" title="Link to this definition">#</a></dt>
<dd><p>A numpy array (dtype=np.int32) of the MRCA node IDs in segments.</p>
</dd></dl>

</dd></dl>

</section>
<section id="the-identitysegment-class">
<h4>The <a class="reference internal" href="#tskit.IdentitySegment" title="tskit.IdentitySegment"><code class="xref py py-class docutils literal notranslate"><span class="pre">IdentitySegment</span></code></a> class<a class="headerlink" href="#the-identitysegment-class" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.IdentitySegment">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">IdentitySegment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">left</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#IdentitySegment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.IdentitySegment" title="Link to this definition">#</a></dt>
<dd><p>A single segment of identity by descent spanning a genomic interval
for a specific ancestor node.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.IdentitySegment.left">
<span class="sig-name descname"><span class="pre">left</span></span><a class="headerlink" href="#tskit.IdentitySegment.left" title="Link to this definition">#</a></dt>
<dd><p>The left genomic coordinate (inclusive).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.IdentitySegment.right">
<span class="sig-name descname"><span class="pre">right</span></span><a class="headerlink" href="#tskit.IdentitySegment.right" title="Link to this definition">#</a></dt>
<dd><p>The right genomic coordinate (exclusive).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.IdentitySegment.node">
<span class="sig-name descname"><span class="pre">node</span></span><a class="headerlink" href="#tskit.IdentitySegment.node" title="Link to this definition">#</a></dt>
<dd><p>The ID of the most recent common ancestor node.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.IdentitySegment.span">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">span</span></span><a class="headerlink" href="#tskit.IdentitySegment.span" title="Link to this definition">#</a></dt>
<dd><p>The length of the genomic region spanned by this identity segment.</p>
</dd></dl>

</dd></dl>

</section>
</section>
<section id="miscellaneous-classes">
<h3>Miscellaneous classes<a class="headerlink" href="#miscellaneous-classes" title="Link to this heading">#</a></h3>
<section id="the-referencesequence-class">
<h4>The <a class="reference internal" href="#tskit.ReferenceSequence" title="tskit.ReferenceSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReferenceSequence</span></code></a> class<a class="headerlink" href="#the-referencesequence-class" title="Link to this heading">#</a></h4>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>Add a top-level summary section that we can link to from here.</p>
</div>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.ReferenceSequence">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">ReferenceSequence</span></span><a class="reference internal" href="_modules/tskit/tables.html#ReferenceSequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.ReferenceSequence" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="data-model.html#sec-data-model-reference-sequence"><span class="std std-ref">reference sequence</span></a> associated
with a given <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a> or <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>.</p>
<p>Metadata concerning reference sequences can be described using the
<a class="reference internal" href="#tskit.ReferenceSequence.metadata_schema" title="tskit.ReferenceSequence.metadata_schema"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata_schema</span></code></a> and stored in the <a class="reference internal" href="#tskit.ReferenceSequence.metadata" title="tskit.ReferenceSequence.metadata"><code class="xref py py-attr docutils literal notranslate"><span class="pre">metadata</span></code></a> attribute.
See the <a class="reference internal" href="metadata.html#sec-metadata-examples-reference-sequence"><span class="std std-ref">examples</span></a> for
idiomatic usage.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This API is preliminary and currently only supports accessing
reference sequence information via the <code class="docutils literal notranslate"><span class="pre">.data</span></code> attribute. Future versions
will also enable transparent fetching of known reference sequences
from a URL (see <a class="github reference external" href="https://github.com/tskit-dev/tskit/issues/2022">tskit-dev/tskit#2022</a>).</p>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="tskit.ReferenceSequence.is_null">
<span class="sig-name descname"><span class="pre">is_null</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#ReferenceSequence.is_null"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.ReferenceSequence.is_null" title="Link to this definition">#</a></dt>
<dd><p>Returns True if this <a class="reference internal" href="#tskit.ReferenceSequence" title="tskit.ReferenceSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReferenceSequence</span></code></a> is null, i.e.,
all fields are empty.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.ReferenceSequence.data">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">data</span></span><a class="headerlink" href="#tskit.ReferenceSequence.data" title="Link to this definition">#</a></dt>
<dd><p>The string encoding of the reference sequence such that <code class="docutils literal notranslate"><span class="pre">data[j]</span></code>
represents the reference nucleotide at base <code class="docutils literal notranslate"><span class="pre">j</span></code>. If this reference
sequence is writable, the value can be assigned, e.g.
<code class="docutils literal notranslate"><span class="pre">tables.reference_sequence.data</span> <span class="pre">=</span> <span class="pre">&quot;ACGT&quot;</span></code></p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.ReferenceSequence.metadata">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">metadata</span></span><a class="headerlink" href="#tskit.ReferenceSequence.metadata" title="Link to this definition">#</a></dt>
<dd><p>The decoded metadata for this object.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.ReferenceSequence.metadata_bytes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">metadata_bytes</span></span><a class="headerlink" href="#tskit.ReferenceSequence.metadata_bytes" title="Link to this definition">#</a></dt>
<dd><p>The raw bytes of metadata for this TableCollection</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.ReferenceSequence.metadata_schema">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">metadata_schema</span></span><a class="headerlink" href="#tskit.ReferenceSequence.metadata_schema" title="Link to this definition">#</a></dt>
<dd><p>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.MetadataSchema</span></code></a> for this object.</p>
</dd></dl>

</dd></dl>

</section>
<section id="the-metadataschema-class">
<h4>The <a class="reference internal" href="#tskit.MetadataSchema" title="tskit.MetadataSchema"><code class="xref py py-class docutils literal notranslate"><span class="pre">MetadataSchema</span></code></a> class<a class="headerlink" href="#the-metadataschema-class" title="Link to this heading">#</a></h4>
<p>Also see the <a class="reference internal" href="#sec-python-api-metadata"><span class="std std-ref">Metadata API</span></a> summary.</p>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.MetadataSchema">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">MetadataSchema</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">schema</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/metadata.html#MetadataSchema"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MetadataSchema" title="Link to this definition">#</a></dt>
<dd><p>Class for validating, encoding and decoding metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>schema</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a>)  A dict containing a valid JSONSchema object.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tskit.MetadataSchema.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/metadata.html#MetadataSchema.asdict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MetadataSchema.asdict" title="Link to this definition">#</a></dt>
<dd><p>Returns a dict representation of this schema. One possible use of this is to
modify this dict and then pass it to the <code class="docutils literal notranslate"><span class="pre">MetadataSchema</span></code> constructor to create
a similar schema.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Mapping" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mapping</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>] | <a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.14)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MetadataSchema.validate_and_encode_row">
<span class="sig-name descname"><span class="pre">validate_and_encode_row</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/metadata.html#MetadataSchema.validate_and_encode_row"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MetadataSchema.validate_and_encode_row" title="Link to this definition">#</a></dt>
<dd><p>Validate a row (dict) of metadata against this schema and return the encoded
representation (bytes) using the codec specified in the schema.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bytes</span></code></a></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MetadataSchema.decode_row">
<span class="sig-name descname"><span class="pre">decode_row</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/metadata.html#MetadataSchema.decode_row"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MetadataSchema.decode_row" title="Link to this definition">#</a></dt>
<dd><p>Decode an encoded row (bytes) of metadata, using the codec specifed in the schema
and return a python dict. Note that no validation of the metadata against the
schema is performed.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MetadataSchema.encode_row">
<span class="sig-name descname"><span class="pre">encode_row</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/metadata.html#MetadataSchema.encode_row"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MetadataSchema.encode_row" title="Link to this definition">#</a></dt>
<dd><p>Encode a row (dict) of metadata to its binary representation (bytes)
using the codec specified in the schema. Note that unlike
<a class="reference internal" href="#tskit.MetadataSchema.validate_and_encode_row" title="tskit.MetadataSchema.validate_and_encode_row"><code class="xref py py-meth docutils literal notranslate"><span class="pre">validate_and_encode_row()</span></code></a> no validation against the schema is performed.
This should only be used for performance if a validation check is not needed.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bytes</span></code></a></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MetadataSchema.structured_array_from_buffer">
<span class="sig-name descname"><span class="pre">structured_array_from_buffer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">buffer</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/metadata.html#MetadataSchema.structured_array_from_buffer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MetadataSchema.structured_array_from_buffer" title="Link to this definition">#</a></dt>
<dd><p>Convert a buffer of metadata into a structured NumPy array.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.14)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MetadataSchema.permissive_json">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">permissive_json</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/metadata.html#MetadataSchema.permissive_json"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MetadataSchema.permissive_json" title="Link to this definition">#</a></dt>
<dd><p>The simplest, permissive JSON schema. Only specifies the JSON codec and has
no constraints on the properties.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.MetadataSchema.null">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">null</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/metadata.html#MetadataSchema.null"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.MetadataSchema.null" title="Link to this definition">#</a></dt>
<dd><p>The null schema which defines no properties and results in raw bytes
being returned on accessing metadata column.</p>
</dd></dl>

</dd></dl>

</section>
<section id="the-tablemetadataschemas-class">
<h4>The <a class="reference internal" href="#tskit.TableMetadataSchemas" title="tskit.TableMetadataSchemas"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableMetadataSchemas</span></code></a> class<a class="headerlink" href="#the-tablemetadataschemas-class" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.TableMetadataSchemas">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">TableMetadataSchemas</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">node</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">site</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">migration</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individual</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">population</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#TableMetadataSchemas"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableMetadataSchemas" title="Link to this definition">#</a></dt>
<dd><p>Convenience class for returning the schemas of all the tables in a tree sequence.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.TableMetadataSchemas.node">
<span class="sig-name descname"><span class="pre">node</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#tskit.TableMetadataSchemas.node" title="Link to this definition">#</a></dt>
<dd><p>The metadata schema of the node table.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.TableMetadataSchemas.edge">
<span class="sig-name descname"><span class="pre">edge</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#tskit.TableMetadataSchemas.edge" title="Link to this definition">#</a></dt>
<dd><p>The metadata schema of the edge table.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.TableMetadataSchemas.site">
<span class="sig-name descname"><span class="pre">site</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#tskit.TableMetadataSchemas.site" title="Link to this definition">#</a></dt>
<dd><p>The metadata schema of the site table.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.TableMetadataSchemas.mutation">
<span class="sig-name descname"><span class="pre">mutation</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#tskit.TableMetadataSchemas.mutation" title="Link to this definition">#</a></dt>
<dd><p>The metadata schema of the mutation table.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.TableMetadataSchemas.migration">
<span class="sig-name descname"><span class="pre">migration</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#tskit.TableMetadataSchemas.migration" title="Link to this definition">#</a></dt>
<dd><p>The metadata schema of the migration table.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.TableMetadataSchemas.individual">
<span class="sig-name descname"><span class="pre">individual</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#tskit.TableMetadataSchemas.individual" title="Link to this definition">#</a></dt>
<dd><p>The metadata schema of the individual table.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.TableMetadataSchemas.population">
<span class="sig-name descname"><span class="pre">population</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#tskit.TableMetadataSchemas.population" title="Link to this definition">#</a></dt>
<dd><p>The metadata schema of the population table.</p>
</dd></dl>

</dd></dl>

</section>
<section id="the-topologycounter-class">
<h4>The <a class="reference internal" href="#tskit.TopologyCounter" title="tskit.TopologyCounter"><code class="xref py py-class docutils literal notranslate"><span class="pre">TopologyCounter</span></code></a> class<a class="headerlink" href="#the-topologycounter-class" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.TopologyCounter">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">TopologyCounter</span></span><a class="reference internal" href="_modules/tskit/combinatorics.html#TopologyCounter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TopologyCounter" title="Link to this definition">#</a></dt>
<dd><p>Contains the distributions of embedded topologies for every combination
of the sample sets used to generate the <code class="docutils literal notranslate"><span class="pre">TopologyCounter</span></code>. It is
indexable by a combination of sample set indexes and returns a
<code class="docutils literal notranslate"><span class="pre">collections.Counter</span></code> whose keys are topology ranks
(see <a class="reference internal" href="topological-analysis.html#sec-tree-ranks"><span class="std std-ref">Interpreting Tree Ranks</span></a>). See <a class="reference internal" href="#tskit.Tree.count_topologies" title="tskit.Tree.count_topologies"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Tree.count_topologies()</span></code></a> for more
detail on how this structure is used.</p>
</dd></dl>

</section>
<section id="the-ldcalculator-class">
<h4>The <a class="reference internal" href="#tskit.LdCalculator" title="tskit.LdCalculator"><code class="xref py py-class docutils literal notranslate"><span class="pre">LdCalculator</span></code></a> class<a class="headerlink" href="#the-ldcalculator-class" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.LdCalculator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">LdCalculator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree_sequence</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/stats.html#LdCalculator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.LdCalculator" title="Link to this definition">#</a></dt>
<dd><p>Class for calculating <a class="reference external" href="https://en.wikipedia.org/wiki/Linkage_disequilibrium">linkage disequilibrium</a> coefficients
between pairs of sites in a <a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This interface is deprecated and a replacement is planned.
Please see <a class="github reference external" href="https://github.com/tskit-dev/tskit/issues/1900">tskit-dev/tskit#1900</a> for
more information. Note also that the current implementation is
quite limited (see warning below).</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This class does not currently support sites that have more than one
mutation. Using it on such a tree sequence will raise a LibraryError with
an Only infinite sites mutations supported message.</p>
<p>Silent mutations are also not supported and will result in a LibraryError.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>tree_sequence</strong> (<a class="reference internal" href="#tskit.TreeSequence" title="tskit.TreeSequence"><em>TreeSequence</em></a>)  The tree sequence of interest.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tskit.LdCalculator.r2">
<span class="sig-name descname"><span class="pre">r2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">a</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">b</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/stats.html#LdCalculator.r2"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.LdCalculator.r2" title="Link to this definition">#</a></dt>
<dd><p>Returns the value of the <span class="math notranslate nohighlight">\(r^2\)</span> statistic between the pair of
sites at the specified indexes. This method is <em>not</em> an efficient
method for computing large numbers of pairwise LD values; please use either
<a class="reference internal" href="#tskit.LdCalculator.r2_array" title="tskit.LdCalculator.r2_array"><code class="xref py py-meth docutils literal notranslate"><span class="pre">r2_array()</span></code></a> or <a class="reference internal" href="#tskit.LdCalculator.r2_matrix" title="tskit.LdCalculator.r2_matrix"><code class="xref py py-meth docutils literal notranslate"><span class="pre">r2_matrix()</span></code></a> for this purpose.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>a</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The index of the first site.</p></li>
<li><p><strong>b</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The index of the second site.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The value of <span class="math notranslate nohighlight">\(r^2\)</span> between the sites at indexes
<code class="docutils literal notranslate"><span class="pre">a</span></code> and <code class="docutils literal notranslate"><span class="pre">b</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.LdCalculator.r2_array">
<span class="sig-name descname"><span class="pre">r2_array</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">a</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">direction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_mutations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_distance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/stats.html#LdCalculator.r2_array"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.LdCalculator.r2_array" title="Link to this definition">#</a></dt>
<dd><p>Returns the value of the <span class="math notranslate nohighlight">\(r^2\)</span> statistic between the focal
site at index <span class="math notranslate nohighlight">\(a\)</span> and a set of other sites. The method
operates by starting at the focal site and iterating over adjacent
sites (in either the forward or backwards direction) until either a
maximum number of other sites have been considered (using the
<code class="docutils literal notranslate"><span class="pre">max_sites</span></code> parameter), a maximum distance in sequence
coordinates has been reached (using the <code class="docutils literal notranslate"><span class="pre">max_distance</span></code> parameter) or
the start/end of the sequence has been reached. For every site
<span class="math notranslate nohighlight">\(b\)</span> considered, we then insert the value of <span class="math notranslate nohighlight">\(r^2\)</span> between
<span class="math notranslate nohighlight">\(a\)</span> and <span class="math notranslate nohighlight">\(b\)</span> at the corresponding index in an array, and
return the entire array. If the returned array is <span class="math notranslate nohighlight">\(x\)</span> and
<code class="docutils literal notranslate"><span class="pre">direction</span></code> is <a class="reference internal" href="#tskit.FORWARD" title="tskit.FORWARD"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.FORWARD</span></code></a> then <span class="math notranslate nohighlight">\(x[0]\)</span> is the
value of the statistic for <span class="math notranslate nohighlight">\(a\)</span> and <span class="math notranslate nohighlight">\(a + 1\)</span>, <span class="math notranslate nohighlight">\(x[1]\)</span>
the value for <span class="math notranslate nohighlight">\(a\)</span> and <span class="math notranslate nohighlight">\(a + 2\)</span>, etc. Similarly, if
<code class="docutils literal notranslate"><span class="pre">direction</span></code> is <a class="reference internal" href="#tskit.REVERSE" title="tskit.REVERSE"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.REVERSE</span></code></a> then <span class="math notranslate nohighlight">\(x[0]\)</span> is the
value of the statistic for <span class="math notranslate nohighlight">\(a\)</span> and <span class="math notranslate nohighlight">\(a - 1\)</span>, <span class="math notranslate nohighlight">\(x[1]\)</span>
the value for <span class="math notranslate nohighlight">\(a\)</span> and <span class="math notranslate nohighlight">\(a - 2\)</span>, etc.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>a</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The index of the focal sites.</p></li>
<li><p><strong>direction</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The direction in which to travel when
examining other sites. Must be either
<a class="reference internal" href="#tskit.FORWARD" title="tskit.FORWARD"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.FORWARD</span></code></a> or <a class="reference internal" href="#tskit.REVERSE" title="tskit.REVERSE"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.REVERSE</span></code></a>. Defaults
to <a class="reference internal" href="#tskit.FORWARD" title="tskit.FORWARD"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.FORWARD</span></code></a>.</p></li>
<li><p><strong>max_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  The maximum number of sites to return
<span class="math notranslate nohighlight">\(r^2\)</span> values for. Defaults to as many sites as
possible.</p></li>
<li><p><strong>max_mutations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>)  Deprecated synonym for max_sites.</p></li>
<li><p><strong>max_distance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>)  The maximum absolute distance between
the focal sites and those for which <span class="math notranslate nohighlight">\(r^2\)</span> values
are returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An array of double precision floating point values
representing the <span class="math notranslate nohighlight">\(r^2\)</span> values for sites in the
specified direction.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tskit.LdCalculator.r2_matrix">
<span class="sig-name descname"><span class="pre">r2_matrix</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/stats.html#LdCalculator.r2_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.LdCalculator.r2_matrix" title="Link to this definition">#</a></dt>
<dd><p>Returns the complete <span class="math notranslate nohighlight">\(m \times m\)</span> matrix of pairwise
<span class="math notranslate nohighlight">\(r^2\)</span> values in a tree sequence with <span class="math notranslate nohighlight">\(m\)</span> sites.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An 2 dimensional square array of double precision
floating point values representing the <span class="math notranslate nohighlight">\(r^2\)</span> values for
all pairs of sites.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="the-tablecollectionindexes-class">
<h4>The <a class="reference internal" href="#tskit.TableCollectionIndexes" title="tskit.TableCollectionIndexes"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollectionIndexes</span></code></a> class<a class="headerlink" href="#the-tablecollectionindexes-class" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.TableCollectionIndexes">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">TableCollectionIndexes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">edge_insertion_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_removal_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollectionIndexes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollectionIndexes" title="Link to this definition">#</a></dt>
<dd><p>A class encapsulating the indexes of a <a class="reference internal" href="#tskit.TableCollection" title="tskit.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="tskit.TableCollectionIndexes.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/tables.html#TableCollectionIndexes.asdict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.TableCollectionIndexes.asdict" title="Link to this definition">#</a></dt>
<dd><p>Return a new dict which maps field names to their corresponding values
in this dataclass.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tskit.TableCollectionIndexes.nbytes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nbytes</span></span><a class="headerlink" href="#tskit.TableCollectionIndexes.nbytes" title="Link to this definition">#</a></dt>
<dd><p>The number of bytes taken by the indexes</p>
</dd></dl>

</dd></dl>

</section>
<section id="the-svgstring-class">
<h4>The <a class="reference internal" href="#tskit.SVGString" title="tskit.SVGString"><code class="xref py py-class docutils literal notranslate"><span class="pre">SVGString</span></code></a> class<a class="headerlink" href="#the-svgstring-class" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.SVGString">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">SVGString</span></span><a class="reference internal" href="_modules/tskit/drawing.html#SVGString"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.SVGString" title="Link to this definition">#</a></dt>
<dd><p>A string containing an SVG representation</p>
<dl class="py method">
<dt class="sig sig-object py" id="tskit.SVGString._repr_svg_">
<span class="sig-name descname"><span class="pre">_repr_svg_</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/drawing.html#SVGString._repr_svg_"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.SVGString._repr_svg_" title="Link to this definition">#</a></dt>
<dd><p>Simply return the SVG string: called by jupyter notebooks to render trees.</p>
</dd></dl>

</dd></dl>

</section>
<section id="the-pcaresult-class">
<h4>The <a class="reference internal" href="#tskit.PCAResult" title="tskit.PCAResult"><code class="xref py py-class docutils literal notranslate"><span class="pre">PCAResult</span></code></a> class<a class="headerlink" href="#the-pcaresult-class" title="Link to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="tskit.PCAResult">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tskit.</span></span><span class="sig-name descname"><span class="pre">PCAResult</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">factors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eigenvalues</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">range_sketch</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">error_bound</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tskit/trees.html#PCAResult"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tskit.PCAResult" title="Link to this definition">#</a></dt>
<dd><p>The result of a call to TreeSequence.pca() capturing the output values
and algorithm convergence details.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.PCAResult.factors">
<span class="sig-name descname"><span class="pre">factors</span></span><a class="headerlink" href="#tskit.PCAResult.factors" title="Link to this definition">#</a></dt>
<dd><p>The principal component factors (or scores).
Columns are orthogonal, with one entry per sample
or individual (see <a class="reference internal" href="#tskit.TreeSequence.pca" title="tskit.TreeSequence.pca"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pca</span></code></a>).
This is the same as the loadings because the GRM is symmetric.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.PCAResult.eigenvalues">
<span class="sig-name descname"><span class="pre">eigenvalues</span></span><a class="headerlink" href="#tskit.PCAResult.eigenvalues" title="Link to this definition">#</a></dt>
<dd><p>Eigenvalues of the genetic relatedness matrix.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.PCAResult.range_sketch">
<span class="sig-name descname"><span class="pre">range_sketch</span></span><a class="headerlink" href="#tskit.PCAResult.range_sketch" title="Link to this definition">#</a></dt>
<dd><p>Range sketch matrix. Can be used as an input for
<a class="reference internal" href="#tskit.TreeSequence.pca" title="tskit.TreeSequence.pca"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pca</span></code></a> option to further improve precision.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tskit.PCAResult.error_bound">
<span class="sig-name descname"><span class="pre">error_bound</span></span><a class="headerlink" href="#tskit.PCAResult.error_bound" title="Link to this definition">#</a></dt>
<dd><p>An estimate of the error resulting from the randomized algorithm (experimental).
Eigenvalues should be correct to within (roughly) this additive factor,
and factors should be correct to within (roughly) this factor divided by the
next-largest eigenvalue in the Euclidean norm. These estimates are obtained from
a bound on the expected L2 operator norm between the true GRM and its
low-dimensional approximation, from equation 1.11 in
<a class="reference external" href="https://arxiv.org/pdf/0909.4061">https://arxiv.org/pdf/0909.4061</a> .</p>
</dd></dl>

</dd></dl>

</section>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="export.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Data export</p>
      </div>
    </a>
    <a class="right-next"
       href="numba.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Numba Integration</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#trees-and-tree-sequences">Trees and tree sequences</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#treesequence-api"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code> API</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#general-properties">General properties</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#efficient-table-column-access">Efficient table column access</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-and-saving">Loading and saving</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#obtaining-trees">Obtaining trees</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#obtaining-other-objects">Obtaining other objects</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tree-topology">Tree topology</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#genetic-variation">Genetic variation</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#demography">Demography</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#other">Other</a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tree-sequence-modification">Tree sequence modification</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#sec-python-api-tree-sequences-ibd">Identity by descent</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tables">Tables</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#statistics">Statistics</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#topological-analysis">Topological analysis</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#display">Display</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#export">Export</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tree-api"><code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code> API</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#sec-python-api-trees-general-properties">General properties</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#creating-new-trees">Creating new trees</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#node-measures">Node measures</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#simple-measures">Simple measures</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#array-access">Array access</a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tree-traversal">Tree traversal</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#sec-python-api-trees-topological-analysis">Topological analysis</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#comparing-trees">Comparing trees</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#balance-imbalance-indices">Balance/imbalance indices</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#sites-and-mutations">Sites and mutations</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#moving-to-other-trees">Moving to other trees</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#id3">Display</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#id4">Export</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tables-and-table-collections">Tables and Table Collections</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tablecollection-api"><code class="docutils literal notranslate"><span class="pre">TableCollection</span></code> API</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#id5">General properties</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#transformation">Transformation</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#modification">Modification</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#creating-a-valid-tree-sequence">Creating a valid tree sequence</a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#miscellaneous-methods">Miscellaneous methods</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#id6">Export</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#table-apis">Table APIs</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#accessing-table-data">Accessing table data</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#text-columns">Text columns</a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#binary-columns">Binary columns</a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#table-functions">Table functions</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#metadata-api">Metadata API</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#provenance">Provenance</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#utility-functions">Utility functions</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#reference-documentation">Reference documentation</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#constants">Constants</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NULL"><code class="docutils literal notranslate"><span class="pre">NULL</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MISSING_DATA"><code class="docutils literal notranslate"><span class="pre">MISSING_DATA</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NODE_IS_SAMPLE"><code class="docutils literal notranslate"><span class="pre">NODE_IS_SAMPLE</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.FORWARD"><code class="docutils literal notranslate"><span class="pre">FORWARD</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.REVERSE"><code class="docutils literal notranslate"><span class="pre">REVERSE</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ALLELES_01"><code class="docutils literal notranslate"><span class="pre">ALLELES_01</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ALLELES_ACGT"><code class="docutils literal notranslate"><span class="pre">ALLELES_ACGT</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.UNKNOWN_TIME"><code class="docutils literal notranslate"><span class="pre">UNKNOWN_TIME</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TIME_UNITS_UNKNOWN"><code class="docutils literal notranslate"><span class="pre">TIME_UNITS_UNKNOWN</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TIME_UNITS_UNCALIBRATED"><code class="docutils literal notranslate"><span class="pre">TIME_UNITS_UNCALIBRATED</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exceptions">Exceptions</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.DuplicatePositionsError"><code class="docutils literal notranslate"><span class="pre">DuplicatePositionsError</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MetadataEncodingError"><code class="docutils literal notranslate"><span class="pre">MetadataEncodingError</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MetadataSchemaValidationError"><code class="docutils literal notranslate"><span class="pre">MetadataSchemaValidationError</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MetadataValidationError"><code class="docutils literal notranslate"><span class="pre">MetadataValidationError</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceValidationError"><code class="docutils literal notranslate"><span class="pre">ProvenanceValidationError</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#top-level-functions">Top-level functions</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.all_trees"><code class="docutils literal notranslate"><span class="pre">all_trees()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.all_tree_shapes"><code class="docutils literal notranslate"><span class="pre">all_tree_shapes()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.all_tree_labellings"><code class="docutils literal notranslate"><span class="pre">all_tree_labellings()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.is_unknown_time"><code class="docutils literal notranslate"><span class="pre">is_unknown_time()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.load"><code class="docutils literal notranslate"><span class="pre">load()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.load_text"><code class="docutils literal notranslate"><span class="pre">load_text()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.pack_bytes"><code class="docutils literal notranslate"><span class="pre">pack_bytes()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.pack_strings"><code class="docutils literal notranslate"><span class="pre">pack_strings()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.parse_edges"><code class="docutils literal notranslate"><span class="pre">parse_edges()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.parse_individuals"><code class="docutils literal notranslate"><span class="pre">parse_individuals()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.parse_mutations"><code class="docutils literal notranslate"><span class="pre">parse_mutations()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.parse_nodes"><code class="docutils literal notranslate"><span class="pre">parse_nodes()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.parse_populations"><code class="docutils literal notranslate"><span class="pre">parse_populations()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.parse_migrations"><code class="docutils literal notranslate"><span class="pre">parse_migrations()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.parse_sites"><code class="docutils literal notranslate"><span class="pre">parse_sites()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.random_nucleotides"><code class="docutils literal notranslate"><span class="pre">random_nucleotides()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.register_metadata_codec"><code class="docutils literal notranslate"><span class="pre">register_metadata_codec()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.validate_provenance"><code class="docutils literal notranslate"><span class="pre">validate_provenance()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.unpack_bytes"><code class="docutils literal notranslate"><span class="pre">unpack_bytes()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.unpack_strings"><code class="docutils literal notranslate"><span class="pre">unpack_strings()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tree-and-tree-sequence-classes">Tree and tree sequence classes</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-tree-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Tree</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree"><code class="docutils literal notranslate"><span class="pre">Tree</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.copy"><code class="docutils literal notranslate"><span class="pre">Tree.copy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.tree_sequence"><code class="docutils literal notranslate"><span class="pre">Tree.tree_sequence</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.root_threshold"><code class="docutils literal notranslate"><span class="pre">Tree.root_threshold</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.first"><code class="docutils literal notranslate"><span class="pre">Tree.first()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.last"><code class="docutils literal notranslate"><span class="pre">Tree.last()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.next"><code class="docutils literal notranslate"><span class="pre">Tree.next()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.prev"><code class="docutils literal notranslate"><span class="pre">Tree.prev()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.clear"><code class="docutils literal notranslate"><span class="pre">Tree.clear()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.seek_index"><code class="docutils literal notranslate"><span class="pre">Tree.seek_index()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.seek"><code class="docutils literal notranslate"><span class="pre">Tree.seek()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.rank"><code class="docutils literal notranslate"><span class="pre">Tree.rank()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.unrank"><code class="docutils literal notranslate"><span class="pre">Tree.unrank()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.count_topologies"><code class="docutils literal notranslate"><span class="pre">Tree.count_topologies()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.branch_length"><code class="docutils literal notranslate"><span class="pre">Tree.branch_length()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.total_branch_length"><code class="docutils literal notranslate"><span class="pre">Tree.total_branch_length</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.mrca"><code class="docutils literal notranslate"><span class="pre">Tree.mrca()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.tmrca"><code class="docutils literal notranslate"><span class="pre">Tree.tmrca()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.parent"><code class="docutils literal notranslate"><span class="pre">Tree.parent()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.parent_array"><code class="docutils literal notranslate"><span class="pre">Tree.parent_array</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.ancestors"><code class="docutils literal notranslate"><span class="pre">Tree.ancestors()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.left_child"><code class="docutils literal notranslate"><span class="pre">Tree.left_child()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.left_child_array"><code class="docutils literal notranslate"><span class="pre">Tree.left_child_array</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.right_child"><code class="docutils literal notranslate"><span class="pre">Tree.right_child()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.right_child_array"><code class="docutils literal notranslate"><span class="pre">Tree.right_child_array</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.left_sib"><code class="docutils literal notranslate"><span class="pre">Tree.left_sib()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.left_sib_array"><code class="docutils literal notranslate"><span class="pre">Tree.left_sib_array</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.right_sib"><code class="docutils literal notranslate"><span class="pre">Tree.right_sib()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.right_sib_array"><code class="docutils literal notranslate"><span class="pre">Tree.right_sib_array</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.siblings"><code class="docutils literal notranslate"><span class="pre">Tree.siblings()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.num_children_array"><code class="docutils literal notranslate"><span class="pre">Tree.num_children_array</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.edge"><code class="docutils literal notranslate"><span class="pre">Tree.edge()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.edge_array"><code class="docutils literal notranslate"><span class="pre">Tree.edge_array</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.virtual_root"><code class="docutils literal notranslate"><span class="pre">Tree.virtual_root</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.num_edges"><code class="docutils literal notranslate"><span class="pre">Tree.num_edges</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.left_root"><code class="docutils literal notranslate"><span class="pre">Tree.left_root</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.children"><code class="docutils literal notranslate"><span class="pre">Tree.children()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.time"><code class="docutils literal notranslate"><span class="pre">Tree.time()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.depth"><code class="docutils literal notranslate"><span class="pre">Tree.depth()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.population"><code class="docutils literal notranslate"><span class="pre">Tree.population()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.is_internal"><code class="docutils literal notranslate"><span class="pre">Tree.is_internal()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.is_leaf"><code class="docutils literal notranslate"><span class="pre">Tree.is_leaf()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.is_isolated"><code class="docutils literal notranslate"><span class="pre">Tree.is_isolated()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.is_sample"><code class="docutils literal notranslate"><span class="pre">Tree.is_sample()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.is_descendant"><code class="docutils literal notranslate"><span class="pre">Tree.is_descendant()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.num_nodes"><code class="docutils literal notranslate"><span class="pre">Tree.num_nodes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.num_roots"><code class="docutils literal notranslate"><span class="pre">Tree.num_roots</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.has_single_root"><code class="docutils literal notranslate"><span class="pre">Tree.has_single_root</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.has_multiple_roots"><code class="docutils literal notranslate"><span class="pre">Tree.has_multiple_roots</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.roots"><code class="docutils literal notranslate"><span class="pre">Tree.roots</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.root"><code class="docutils literal notranslate"><span class="pre">Tree.root</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.is_root"><code class="docutils literal notranslate"><span class="pre">Tree.is_root()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.index"><code class="docutils literal notranslate"><span class="pre">Tree.index</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.interval"><code class="docutils literal notranslate"><span class="pre">Tree.interval</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.span"><code class="docutils literal notranslate"><span class="pre">Tree.span</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.mid"><code class="docutils literal notranslate"><span class="pre">Tree.mid</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.draw_text"><code class="docutils literal notranslate"><span class="pre">Tree.draw_text()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.draw_svg"><code class="docutils literal notranslate"><span class="pre">Tree.draw_svg()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.draw"><code class="docutils literal notranslate"><span class="pre">Tree.draw()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.num_mutations"><code class="docutils literal notranslate"><span class="pre">Tree.num_mutations</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.num_sites"><code class="docutils literal notranslate"><span class="pre">Tree.num_sites</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.sites"><code class="docutils literal notranslate"><span class="pre">Tree.sites()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.mutations"><code class="docutils literal notranslate"><span class="pre">Tree.mutations()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.leaves"><code class="docutils literal notranslate"><span class="pre">Tree.leaves()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.samples"><code class="docutils literal notranslate"><span class="pre">Tree.samples()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.num_children"><code class="docutils literal notranslate"><span class="pre">Tree.num_children()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.num_samples"><code class="docutils literal notranslate"><span class="pre">Tree.num_samples()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.num_tracked_samples"><code class="docutils literal notranslate"><span class="pre">Tree.num_tracked_samples()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.preorder"><code class="docutils literal notranslate"><span class="pre">Tree.preorder()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.postorder"><code class="docutils literal notranslate"><span class="pre">Tree.postorder()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.timeasc"><code class="docutils literal notranslate"><span class="pre">Tree.timeasc()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.timedesc"><code class="docutils literal notranslate"><span class="pre">Tree.timedesc()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.nodes"><code class="docutils literal notranslate"><span class="pre">Tree.nodes()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.as_newick"><code class="docutils literal notranslate"><span class="pre">Tree.as_newick()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.newick"><code class="docutils literal notranslate"><span class="pre">Tree.newick()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.as_dict_of_dicts"><code class="docutils literal notranslate"><span class="pre">Tree.as_dict_of_dicts()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.__str__"><code class="docutils literal notranslate"><span class="pre">Tree.__str__()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree._repr_html_"><code class="docutils literal notranslate"><span class="pre">Tree._repr_html_()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.map_mutations"><code class="docutils literal notranslate"><span class="pre">Tree.map_mutations()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.kc_distance"><code class="docutils literal notranslate"><span class="pre">Tree.kc_distance()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.rf_distance"><code class="docutils literal notranslate"><span class="pre">Tree.rf_distance()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.path_length"><code class="docutils literal notranslate"><span class="pre">Tree.path_length()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.distance_between"><code class="docutils literal notranslate"><span class="pre">Tree.distance_between()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.b1_index"><code class="docutils literal notranslate"><span class="pre">Tree.b1_index()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.b2_index"><code class="docutils literal notranslate"><span class="pre">Tree.b2_index()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.colless_index"><code class="docutils literal notranslate"><span class="pre">Tree.colless_index()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.sackin_index"><code class="docutils literal notranslate"><span class="pre">Tree.sackin_index()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.num_lineages"><code class="docutils literal notranslate"><span class="pre">Tree.num_lineages()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.split_polytomies"><code class="docutils literal notranslate"><span class="pre">Tree.split_polytomies()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.generate_star"><code class="docutils literal notranslate"><span class="pre">Tree.generate_star()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.generate_balanced"><code class="docutils literal notranslate"><span class="pre">Tree.generate_balanced()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.generate_comb"><code class="docutils literal notranslate"><span class="pre">Tree.generate_comb()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Tree.generate_random_binary"><code class="docutils literal notranslate"><span class="pre">Tree.generate_random_binary()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-treesequence-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence"><code class="docutils literal notranslate"><span class="pre">TreeSequence</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.equals"><code class="docutils literal notranslate"><span class="pre">TreeSequence.equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.aslist"><code class="docutils literal notranslate"><span class="pre">TreeSequence.aslist()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.dump"><code class="docutils literal notranslate"><span class="pre">TreeSequence.dump()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.reference_sequence"><code class="docutils literal notranslate"><span class="pre">TreeSequence.reference_sequence</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.has_reference_sequence"><code class="docutils literal notranslate"><span class="pre">TreeSequence.has_reference_sequence()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.tables_dict"><code class="docutils literal notranslate"><span class="pre">TreeSequence.tables_dict</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.tables"><code class="docutils literal notranslate"><span class="pre">TreeSequence.tables</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.nbytes"><code class="docutils literal notranslate"><span class="pre">TreeSequence.nbytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.dump_tables"><code class="docutils literal notranslate"><span class="pre">TreeSequence.dump_tables()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.link_ancestors"><code class="docutils literal notranslate"><span class="pre">TreeSequence.link_ancestors()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.dump_text"><code class="docutils literal notranslate"><span class="pre">TreeSequence.dump_text()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.__str__"><code class="docutils literal notranslate"><span class="pre">TreeSequence.__str__()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence._repr_html_"><code class="docutils literal notranslate"><span class="pre">TreeSequence._repr_html_()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.num_samples"><code class="docutils literal notranslate"><span class="pre">TreeSequence.num_samples</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.table_metadata_schemas"><code class="docutils literal notranslate"><span class="pre">TreeSequence.table_metadata_schemas</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.discrete_genome"><code class="docutils literal notranslate"><span class="pre">TreeSequence.discrete_genome</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.discrete_time"><code class="docutils literal notranslate"><span class="pre">TreeSequence.discrete_time</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.min_time"><code class="docutils literal notranslate"><span class="pre">TreeSequence.min_time</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.max_time"><code class="docutils literal notranslate"><span class="pre">TreeSequence.max_time</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.sequence_length"><code class="docutils literal notranslate"><span class="pre">TreeSequence.sequence_length</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.metadata"><code class="docutils literal notranslate"><span class="pre">TreeSequence.metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.metadata_schema"><code class="docutils literal notranslate"><span class="pre">TreeSequence.metadata_schema</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.time_units"><code class="docutils literal notranslate"><span class="pre">TreeSequence.time_units</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.num_edges"><code class="docutils literal notranslate"><span class="pre">TreeSequence.num_edges</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.num_trees"><code class="docutils literal notranslate"><span class="pre">TreeSequence.num_trees</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.num_sites"><code class="docutils literal notranslate"><span class="pre">TreeSequence.num_sites</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.num_mutations"><code class="docutils literal notranslate"><span class="pre">TreeSequence.num_mutations</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.num_individuals"><code class="docutils literal notranslate"><span class="pre">TreeSequence.num_individuals</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.num_nodes"><code class="docutils literal notranslate"><span class="pre">TreeSequence.num_nodes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.num_provenances"><code class="docutils literal notranslate"><span class="pre">TreeSequence.num_provenances</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.num_populations"><code class="docutils literal notranslate"><span class="pre">TreeSequence.num_populations</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.num_migrations"><code class="docutils literal notranslate"><span class="pre">TreeSequence.num_migrations</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.max_root_time"><code class="docutils literal notranslate"><span class="pre">TreeSequence.max_root_time</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.migrations"><code class="docutils literal notranslate"><span class="pre">TreeSequence.migrations()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.individuals"><code class="docutils literal notranslate"><span class="pre">TreeSequence.individuals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.nodes"><code class="docutils literal notranslate"><span class="pre">TreeSequence.nodes()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.edges"><code class="docutils literal notranslate"><span class="pre">TreeSequence.edges()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.edge_diffs"><code class="docutils literal notranslate"><span class="pre">TreeSequence.edge_diffs()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.sites"><code class="docutils literal notranslate"><span class="pre">TreeSequence.sites()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.mutations"><code class="docutils literal notranslate"><span class="pre">TreeSequence.mutations()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.populations"><code class="docutils literal notranslate"><span class="pre">TreeSequence.populations()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.provenances"><code class="docutils literal notranslate"><span class="pre">TreeSequence.provenances()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.breakpoints"><code class="docutils literal notranslate"><span class="pre">TreeSequence.breakpoints()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.at"><code class="docutils literal notranslate"><span class="pre">TreeSequence.at()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.at_index"><code class="docutils literal notranslate"><span class="pre">TreeSequence.at_index()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.first"><code class="docutils literal notranslate"><span class="pre">TreeSequence.first()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.last"><code class="docutils literal notranslate"><span class="pre">TreeSequence.last()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.trees"><code class="docutils literal notranslate"><span class="pre">TreeSequence.trees()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.coiterate"><code class="docutils literal notranslate"><span class="pre">TreeSequence.coiterate()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.haplotypes"><code class="docutils literal notranslate"><span class="pre">TreeSequence.haplotypes()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.variants"><code class="docutils literal notranslate"><span class="pre">TreeSequence.variants()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.genotype_matrix"><code class="docutils literal notranslate"><span class="pre">TreeSequence.genotype_matrix()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.alignments"><code class="docutils literal notranslate"><span class="pre">TreeSequence.alignments()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.individuals_population"><code class="docutils literal notranslate"><span class="pre">TreeSequence.individuals_population</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.individuals_time"><code class="docutils literal notranslate"><span class="pre">TreeSequence.individuals_time</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.individuals_location"><code class="docutils literal notranslate"><span class="pre">TreeSequence.individuals_location</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.individuals_flags"><code class="docutils literal notranslate"><span class="pre">TreeSequence.individuals_flags</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.individuals_metadata"><code class="docutils literal notranslate"><span class="pre">TreeSequence.individuals_metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.individuals_nodes"><code class="docutils literal notranslate"><span class="pre">TreeSequence.individuals_nodes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.nodes_metadata"><code class="docutils literal notranslate"><span class="pre">TreeSequence.nodes_metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.nodes_time"><code class="docutils literal notranslate"><span class="pre">TreeSequence.nodes_time</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.nodes_flags"><code class="docutils literal notranslate"><span class="pre">TreeSequence.nodes_flags</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.nodes_population"><code class="docutils literal notranslate"><span class="pre">TreeSequence.nodes_population</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.nodes_individual"><code class="docutils literal notranslate"><span class="pre">TreeSequence.nodes_individual</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.edges_left"><code class="docutils literal notranslate"><span class="pre">TreeSequence.edges_left</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.edges_right"><code class="docutils literal notranslate"><span class="pre">TreeSequence.edges_right</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.edges_parent"><code class="docutils literal notranslate"><span class="pre">TreeSequence.edges_parent</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.edges_child"><code class="docutils literal notranslate"><span class="pre">TreeSequence.edges_child</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.edges_metadata"><code class="docutils literal notranslate"><span class="pre">TreeSequence.edges_metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.sites_position"><code class="docutils literal notranslate"><span class="pre">TreeSequence.sites_position</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.sites_ancestral_state"><code class="docutils literal notranslate"><span class="pre">TreeSequence.sites_ancestral_state</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.sites_metadata"><code class="docutils literal notranslate"><span class="pre">TreeSequence.sites_metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.mutations_site"><code class="docutils literal notranslate"><span class="pre">TreeSequence.mutations_site</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.mutations_node"><code class="docutils literal notranslate"><span class="pre">TreeSequence.mutations_node</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.mutations_parent"><code class="docutils literal notranslate"><span class="pre">TreeSequence.mutations_parent</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.mutations_time"><code class="docutils literal notranslate"><span class="pre">TreeSequence.mutations_time</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.mutations_derived_state"><code class="docutils literal notranslate"><span class="pre">TreeSequence.mutations_derived_state</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.mutations_metadata"><code class="docutils literal notranslate"><span class="pre">TreeSequence.mutations_metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.mutations_edge"><code class="docutils literal notranslate"><span class="pre">TreeSequence.mutations_edge</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.mutations_inherited_state"><code class="docutils literal notranslate"><span class="pre">TreeSequence.mutations_inherited_state</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.migrations_left"><code class="docutils literal notranslate"><span class="pre">TreeSequence.migrations_left</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.migrations_right"><code class="docutils literal notranslate"><span class="pre">TreeSequence.migrations_right</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.migrations_node"><code class="docutils literal notranslate"><span class="pre">TreeSequence.migrations_node</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.migrations_source"><code class="docutils literal notranslate"><span class="pre">TreeSequence.migrations_source</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.migrations_dest"><code class="docutils literal notranslate"><span class="pre">TreeSequence.migrations_dest</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.migrations_time"><code class="docutils literal notranslate"><span class="pre">TreeSequence.migrations_time</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.migrations_metadata"><code class="docutils literal notranslate"><span class="pre">TreeSequence.migrations_metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.populations_metadata"><code class="docutils literal notranslate"><span class="pre">TreeSequence.populations_metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.indexes_edge_insertion_order"><code class="docutils literal notranslate"><span class="pre">TreeSequence.indexes_edge_insertion_order</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.indexes_edge_removal_order"><code class="docutils literal notranslate"><span class="pre">TreeSequence.indexes_edge_removal_order</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.individual"><code class="docutils literal notranslate"><span class="pre">TreeSequence.individual()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.node"><code class="docutils literal notranslate"><span class="pre">TreeSequence.node()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.edge"><code class="docutils literal notranslate"><span class="pre">TreeSequence.edge()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.migration"><code class="docutils literal notranslate"><span class="pre">TreeSequence.migration()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.mutation"><code class="docutils literal notranslate"><span class="pre">TreeSequence.mutation()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.site"><code class="docutils literal notranslate"><span class="pre">TreeSequence.site()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.population"><code class="docutils literal notranslate"><span class="pre">TreeSequence.population()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.provenance"><code class="docutils literal notranslate"><span class="pre">TreeSequence.provenance()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.samples"><code class="docutils literal notranslate"><span class="pre">TreeSequence.samples()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.as_vcf"><code class="docutils literal notranslate"><span class="pre">TreeSequence.as_vcf()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.write_vcf"><code class="docutils literal notranslate"><span class="pre">TreeSequence.write_vcf()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.write_fasta"><code class="docutils literal notranslate"><span class="pre">TreeSequence.write_fasta()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.as_fasta"><code class="docutils literal notranslate"><span class="pre">TreeSequence.as_fasta()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.write_nexus"><code class="docutils literal notranslate"><span class="pre">TreeSequence.write_nexus()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.as_nexus"><code class="docutils literal notranslate"><span class="pre">TreeSequence.as_nexus()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.to_macs"><code class="docutils literal notranslate"><span class="pre">TreeSequence.to_macs()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.simplify"><code class="docutils literal notranslate"><span class="pre">TreeSequence.simplify()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.delete_sites"><code class="docutils literal notranslate"><span class="pre">TreeSequence.delete_sites()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.delete_intervals"><code class="docutils literal notranslate"><span class="pre">TreeSequence.delete_intervals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.keep_intervals"><code class="docutils literal notranslate"><span class="pre">TreeSequence.keep_intervals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.ltrim"><code class="docutils literal notranslate"><span class="pre">TreeSequence.ltrim()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.rtrim"><code class="docutils literal notranslate"><span class="pre">TreeSequence.rtrim()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.trim"><code class="docutils literal notranslate"><span class="pre">TreeSequence.trim()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.shift"><code class="docutils literal notranslate"><span class="pre">TreeSequence.shift()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.concatenate"><code class="docutils literal notranslate"><span class="pre">TreeSequence.concatenate()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.split_edges"><code class="docutils literal notranslate"><span class="pre">TreeSequence.split_edges()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.decapitate"><code class="docutils literal notranslate"><span class="pre">TreeSequence.decapitate()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.extend_haplotypes"><code class="docutils literal notranslate"><span class="pre">TreeSequence.extend_haplotypes()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.subset"><code class="docutils literal notranslate"><span class="pre">TreeSequence.subset()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.union"><code class="docutils literal notranslate"><span class="pre">TreeSequence.union()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.draw_svg"><code class="docutils literal notranslate"><span class="pre">TreeSequence.draw_svg()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.draw_text"><code class="docutils literal notranslate"><span class="pre">TreeSequence.draw_text()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.general_stat"><code class="docutils literal notranslate"><span class="pre">TreeSequence.general_stat()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.sample_count_stat"><code class="docutils literal notranslate"><span class="pre">TreeSequence.sample_count_stat()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.diversity"><code class="docutils literal notranslate"><span class="pre">TreeSequence.diversity()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.divergence"><code class="docutils literal notranslate"><span class="pre">TreeSequence.divergence()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.divergence_matrix"><code class="docutils literal notranslate"><span class="pre">TreeSequence.divergence_matrix()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.genetic_relatedness"><code class="docutils literal notranslate"><span class="pre">TreeSequence.genetic_relatedness()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.genetic_relatedness_matrix"><code class="docutils literal notranslate"><span class="pre">TreeSequence.genetic_relatedness_matrix()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.genetic_relatedness_weighted"><code class="docutils literal notranslate"><span class="pre">TreeSequence.genetic_relatedness_weighted()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.genetic_relatedness_vector"><code class="docutils literal notranslate"><span class="pre">TreeSequence.genetic_relatedness_vector()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.pca"><code class="docutils literal notranslate"><span class="pre">TreeSequence.pca()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.trait_covariance"><code class="docutils literal notranslate"><span class="pre">TreeSequence.trait_covariance()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.trait_correlation"><code class="docutils literal notranslate"><span class="pre">TreeSequence.trait_correlation()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.trait_regression"><code class="docutils literal notranslate"><span class="pre">TreeSequence.trait_regression()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.trait_linear_model"><code class="docutils literal notranslate"><span class="pre">TreeSequence.trait_linear_model()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.segregating_sites"><code class="docutils literal notranslate"><span class="pre">TreeSequence.segregating_sites()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.allele_frequency_spectrum"><code class="docutils literal notranslate"><span class="pre">TreeSequence.allele_frequency_spectrum()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.Tajimas_D"><code class="docutils literal notranslate"><span class="pre">TreeSequence.Tajimas_D()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.Fst"><code class="docutils literal notranslate"><span class="pre">TreeSequence.Fst()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.Y3"><code class="docutils literal notranslate"><span class="pre">TreeSequence.Y3()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.Y2"><code class="docutils literal notranslate"><span class="pre">TreeSequence.Y2()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.Y1"><code class="docutils literal notranslate"><span class="pre">TreeSequence.Y1()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.f4"><code class="docutils literal notranslate"><span class="pre">TreeSequence.f4()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.f3"><code class="docutils literal notranslate"><span class="pre">TreeSequence.f3()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.f2"><code class="docutils literal notranslate"><span class="pre">TreeSequence.f2()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.mean_descendants"><code class="docutils literal notranslate"><span class="pre">TreeSequence.mean_descendants()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.genealogical_nearest_neighbours"><code class="docutils literal notranslate"><span class="pre">TreeSequence.genealogical_nearest_neighbours()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.kc_distance"><code class="docutils literal notranslate"><span class="pre">TreeSequence.kc_distance()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.count_topologies"><code class="docutils literal notranslate"><span class="pre">TreeSequence.count_topologies()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.ibd_segments"><code class="docutils literal notranslate"><span class="pre">TreeSequence.ibd_segments()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.pair_coalescence_counts"><code class="docutils literal notranslate"><span class="pre">TreeSequence.pair_coalescence_counts()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.pair_coalescence_quantiles"><code class="docutils literal notranslate"><span class="pre">TreeSequence.pair_coalescence_quantiles()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.pair_coalescence_rates"><code class="docutils literal notranslate"><span class="pre">TreeSequence.pair_coalescence_rates()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.impute_unknown_mutations_time"><code class="docutils literal notranslate"><span class="pre">TreeSequence.impute_unknown_mutations_time()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.sample_nodes_by_ploidy"><code class="docutils literal notranslate"><span class="pre">TreeSequence.sample_nodes_by_ploidy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.map_to_vcf_model"><code class="docutils literal notranslate"><span class="pre">TreeSequence.map_to_vcf_model()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TreeSequence.pairwise_diversity"><code class="docutils literal notranslate"><span class="pre">TreeSequence.pairwise_diversity()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#simple-container-classes">Simple container classes</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-individual-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Individual</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Individual"><code class="docutils literal notranslate"><span class="pre">Individual</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Individual.id"><code class="docutils literal notranslate"><span class="pre">Individual.id</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Individual.flags"><code class="docutils literal notranslate"><span class="pre">Individual.flags</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Individual.location"><code class="docutils literal notranslate"><span class="pre">Individual.location</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Individual.parents"><code class="docutils literal notranslate"><span class="pre">Individual.parents</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Individual.nodes"><code class="docutils literal notranslate"><span class="pre">Individual.nodes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Individual.metadata"><code class="docutils literal notranslate"><span class="pre">Individual.metadata</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-node-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Node</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Node"><code class="docutils literal notranslate"><span class="pre">Node</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Node.id"><code class="docutils literal notranslate"><span class="pre">Node.id</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Node.flags"><code class="docutils literal notranslate"><span class="pre">Node.flags</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Node.time"><code class="docutils literal notranslate"><span class="pre">Node.time</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Node.population"><code class="docutils literal notranslate"><span class="pre">Node.population</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Node.individual"><code class="docutils literal notranslate"><span class="pre">Node.individual</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Node.metadata"><code class="docutils literal notranslate"><span class="pre">Node.metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Node.is_sample"><code class="docutils literal notranslate"><span class="pre">Node.is_sample()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-edge-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Edge"><code class="docutils literal notranslate"><span class="pre">Edge</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Edge.id"><code class="docutils literal notranslate"><span class="pre">Edge.id</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Edge.left"><code class="docutils literal notranslate"><span class="pre">Edge.left</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Edge.right"><code class="docutils literal notranslate"><span class="pre">Edge.right</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Edge.parent"><code class="docutils literal notranslate"><span class="pre">Edge.parent</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Edge.child"><code class="docutils literal notranslate"><span class="pre">Edge.child</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Edge.metadata"><code class="docutils literal notranslate"><span class="pre">Edge.metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Edge.span"><code class="docutils literal notranslate"><span class="pre">Edge.span</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Edge.interval"><code class="docutils literal notranslate"><span class="pre">Edge.interval</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-site-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Site</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Site"><code class="docutils literal notranslate"><span class="pre">Site</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Site.id"><code class="docutils literal notranslate"><span class="pre">Site.id</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Site.position"><code class="docutils literal notranslate"><span class="pre">Site.position</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Site.ancestral_state"><code class="docutils literal notranslate"><span class="pre">Site.ancestral_state</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Site.mutations"><code class="docutils literal notranslate"><span class="pre">Site.mutations</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Site.metadata"><code class="docutils literal notranslate"><span class="pre">Site.metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Site.alleles"><code class="docutils literal notranslate"><span class="pre">Site.alleles</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-mutation-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Mutation</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Mutation"><code class="docutils literal notranslate"><span class="pre">Mutation</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Mutation.id"><code class="docutils literal notranslate"><span class="pre">Mutation.id</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Mutation.site"><code class="docutils literal notranslate"><span class="pre">Mutation.site</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Mutation.node"><code class="docutils literal notranslate"><span class="pre">Mutation.node</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Mutation.time"><code class="docutils literal notranslate"><span class="pre">Mutation.time</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Mutation.derived_state"><code class="docutils literal notranslate"><span class="pre">Mutation.derived_state</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Mutation.parent"><code class="docutils literal notranslate"><span class="pre">Mutation.parent</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Mutation.metadata"><code class="docutils literal notranslate"><span class="pre">Mutation.metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Mutation.edge"><code class="docutils literal notranslate"><span class="pre">Mutation.edge</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Mutation.inherited_state"><code class="docutils literal notranslate"><span class="pre">Mutation.inherited_state</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-variant-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Variant</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant"><code class="docutils literal notranslate"><span class="pre">Variant</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.site"><code class="docutils literal notranslate"><span class="pre">Variant.site</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.alleles"><code class="docutils literal notranslate"><span class="pre">Variant.alleles</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.samples"><code class="docutils literal notranslate"><span class="pre">Variant.samples</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.genotypes"><code class="docutils literal notranslate"><span class="pre">Variant.genotypes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.isolated_as_missing"><code class="docutils literal notranslate"><span class="pre">Variant.isolated_as_missing</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.has_missing_data"><code class="docutils literal notranslate"><span class="pre">Variant.has_missing_data</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.num_missing"><code class="docutils literal notranslate"><span class="pre">Variant.num_missing</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.num_alleles"><code class="docutils literal notranslate"><span class="pre">Variant.num_alleles</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.decode"><code class="docutils literal notranslate"><span class="pre">Variant.decode()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.copy"><code class="docutils literal notranslate"><span class="pre">Variant.copy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.states"><code class="docutils literal notranslate"><span class="pre">Variant.states()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.counts"><code class="docutils literal notranslate"><span class="pre">Variant.counts()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Variant.frequencies"><code class="docutils literal notranslate"><span class="pre">Variant.frequencies()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-migration-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Migration</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Migration"><code class="docutils literal notranslate"><span class="pre">Migration</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Migration.left"><code class="docutils literal notranslate"><span class="pre">Migration.left</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Migration.right"><code class="docutils literal notranslate"><span class="pre">Migration.right</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Migration.node"><code class="docutils literal notranslate"><span class="pre">Migration.node</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Migration.source"><code class="docutils literal notranslate"><span class="pre">Migration.source</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Migration.dest"><code class="docutils literal notranslate"><span class="pre">Migration.dest</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Migration.time"><code class="docutils literal notranslate"><span class="pre">Migration.time</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Migration.metadata"><code class="docutils literal notranslate"><span class="pre">Migration.metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Migration.id"><code class="docutils literal notranslate"><span class="pre">Migration.id</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-population-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Population</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Population"><code class="docutils literal notranslate"><span class="pre">Population</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Population.id"><code class="docutils literal notranslate"><span class="pre">Population.id</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Population.metadata"><code class="docutils literal notranslate"><span class="pre">Population.metadata</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-provenance-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Provenance</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Provenance"><code class="docutils literal notranslate"><span class="pre">Provenance</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Provenance.timestamp"><code class="docutils literal notranslate"><span class="pre">Provenance.timestamp</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Provenance.record"><code class="docutils literal notranslate"><span class="pre">Provenance.record</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-interval-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Interval</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Interval"><code class="docutils literal notranslate"><span class="pre">Interval</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Interval.left"><code class="docutils literal notranslate"><span class="pre">Interval.left</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Interval.right"><code class="docutils literal notranslate"><span class="pre">Interval.right</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Interval.span"><code class="docutils literal notranslate"><span class="pre">Interval.span</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Interval.mid"><code class="docutils literal notranslate"><span class="pre">Interval.mid</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-rank-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">Rank</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Rank"><code class="docutils literal notranslate"><span class="pre">Rank</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Rank.shape"><code class="docutils literal notranslate"><span class="pre">Rank.shape</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.Rank.label"><code class="docutils literal notranslate"><span class="pre">Rank.label</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tablecollection-and-table-classes">TableCollection and Table classes</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-tablecollection-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollection</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection"><code class="docutils literal notranslate"><span class="pre">TableCollection</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.individuals"><code class="docutils literal notranslate"><span class="pre">TableCollection.individuals</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.nodes"><code class="docutils literal notranslate"><span class="pre">TableCollection.nodes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.edges"><code class="docutils literal notranslate"><span class="pre">TableCollection.edges</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.migrations"><code class="docutils literal notranslate"><span class="pre">TableCollection.migrations</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.sites"><code class="docutils literal notranslate"><span class="pre">TableCollection.sites</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.mutations"><code class="docutils literal notranslate"><span class="pre">TableCollection.mutations</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.populations"><code class="docutils literal notranslate"><span class="pre">TableCollection.populations</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.provenances"><code class="docutils literal notranslate"><span class="pre">TableCollection.provenances</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.indexes"><code class="docutils literal notranslate"><span class="pre">TableCollection.indexes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.sequence_length"><code class="docutils literal notranslate"><span class="pre">TableCollection.sequence_length</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.file_uuid"><code class="docutils literal notranslate"><span class="pre">TableCollection.file_uuid</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.time_units"><code class="docutils literal notranslate"><span class="pre">TableCollection.time_units</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.has_reference_sequence"><code class="docutils literal notranslate"><span class="pre">TableCollection.has_reference_sequence()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.reference_sequence"><code class="docutils literal notranslate"><span class="pre">TableCollection.reference_sequence</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.asdict"><code class="docutils literal notranslate"><span class="pre">TableCollection.asdict()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.table_name_map"><code class="docutils literal notranslate"><span class="pre">TableCollection.table_name_map</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.nbytes"><code class="docutils literal notranslate"><span class="pre">TableCollection.nbytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.equals"><code class="docutils literal notranslate"><span class="pre">TableCollection.equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.assert_equals"><code class="docutils literal notranslate"><span class="pre">TableCollection.assert_equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.dump"><code class="docutils literal notranslate"><span class="pre">TableCollection.dump()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.copy"><code class="docutils literal notranslate"><span class="pre">TableCollection.copy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.tree_sequence"><code class="docutils literal notranslate"><span class="pre">TableCollection.tree_sequence()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.simplify"><code class="docutils literal notranslate"><span class="pre">TableCollection.simplify()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.link_ancestors"><code class="docutils literal notranslate"><span class="pre">TableCollection.link_ancestors()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.sort"><code class="docutils literal notranslate"><span class="pre">TableCollection.sort()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.sort_individuals"><code class="docutils literal notranslate"><span class="pre">TableCollection.sort_individuals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.canonicalise"><code class="docutils literal notranslate"><span class="pre">TableCollection.canonicalise()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.compute_mutation_parents"><code class="docutils literal notranslate"><span class="pre">TableCollection.compute_mutation_parents()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.compute_mutation_times"><code class="docutils literal notranslate"><span class="pre">TableCollection.compute_mutation_times()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.deduplicate_sites"><code class="docutils literal notranslate"><span class="pre">TableCollection.deduplicate_sites()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.delete_sites"><code class="docutils literal notranslate"><span class="pre">TableCollection.delete_sites()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.delete_intervals"><code class="docutils literal notranslate"><span class="pre">TableCollection.delete_intervals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.keep_intervals"><code class="docutils literal notranslate"><span class="pre">TableCollection.keep_intervals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.ltrim"><code class="docutils literal notranslate"><span class="pre">TableCollection.ltrim()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.rtrim"><code class="docutils literal notranslate"><span class="pre">TableCollection.rtrim()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.trim"><code class="docutils literal notranslate"><span class="pre">TableCollection.trim()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.shift"><code class="docutils literal notranslate"><span class="pre">TableCollection.shift()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.delete_older"><code class="docutils literal notranslate"><span class="pre">TableCollection.delete_older()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.clear"><code class="docutils literal notranslate"><span class="pre">TableCollection.clear()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.has_index"><code class="docutils literal notranslate"><span class="pre">TableCollection.has_index()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.build_index"><code class="docutils literal notranslate"><span class="pre">TableCollection.build_index()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.drop_index"><code class="docutils literal notranslate"><span class="pre">TableCollection.drop_index()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.subset"><code class="docutils literal notranslate"><span class="pre">TableCollection.subset()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.union"><code class="docutils literal notranslate"><span class="pre">TableCollection.union()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.ibd_segments"><code class="docutils literal notranslate"><span class="pre">TableCollection.ibd_segments()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.metadata"><code class="docutils literal notranslate"><span class="pre">TableCollection.metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.metadata_bytes"><code class="docutils literal notranslate"><span class="pre">TableCollection.metadata_bytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollection.metadata_schema"><code class="docutils literal notranslate"><span class="pre">TableCollection.metadata_schema</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#individualtable-classes"><code class="xref py py-class docutils literal notranslate"><span class="pre">IndividualTable</span></code> classes</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable"><code class="docutils literal notranslate"><span class="pre">IndividualTable</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.add_row"><code class="docutils literal notranslate"><span class="pre">IndividualTable.add_row()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.set_columns"><code class="docutils literal notranslate"><span class="pre">IndividualTable.set_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.append_columns"><code class="docutils literal notranslate"><span class="pre">IndividualTable.append_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.packset_location"><code class="docutils literal notranslate"><span class="pre">IndividualTable.packset_location()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.packset_parents"><code class="docutils literal notranslate"><span class="pre">IndividualTable.packset_parents()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.keep_rows"><code class="docutils literal notranslate"><span class="pre">IndividualTable.keep_rows()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.__getitem__"><code class="docutils literal notranslate"><span class="pre">IndividualTable.__getitem__()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.append"><code class="docutils literal notranslate"><span class="pre">IndividualTable.append()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.asdict"><code class="docutils literal notranslate"><span class="pre">IndividualTable.asdict()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.assert_equals"><code class="docutils literal notranslate"><span class="pre">IndividualTable.assert_equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.clear"><code class="docutils literal notranslate"><span class="pre">IndividualTable.clear()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.copy"><code class="docutils literal notranslate"><span class="pre">IndividualTable.copy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.drop_metadata"><code class="docutils literal notranslate"><span class="pre">IndividualTable.drop_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.equals"><code class="docutils literal notranslate"><span class="pre">IndividualTable.equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.metadata_schema"><code class="docutils literal notranslate"><span class="pre">IndividualTable.metadata_schema</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.metadata_vector"><code class="docutils literal notranslate"><span class="pre">IndividualTable.metadata_vector()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.nbytes"><code class="docutils literal notranslate"><span class="pre">IndividualTable.nbytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.packset_metadata"><code class="docutils literal notranslate"><span class="pre">IndividualTable.packset_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTable.truncate"><code class="docutils literal notranslate"><span class="pre">IndividualTable.truncate()</span></code></a></li>
</ul>
</li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#associated-row-class">Associated row class</a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTableRow"><code class="docutils literal notranslate"><span class="pre">IndividualTableRow</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTableRow.flags"><code class="docutils literal notranslate"><span class="pre">IndividualTableRow.flags</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTableRow.location"><code class="docutils literal notranslate"><span class="pre">IndividualTableRow.location</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTableRow.parents"><code class="docutils literal notranslate"><span class="pre">IndividualTableRow.parents</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTableRow.metadata"><code class="docutils literal notranslate"><span class="pre">IndividualTableRow.metadata</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTableRow.asdict"><code class="docutils literal notranslate"><span class="pre">IndividualTableRow.asdict()</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IndividualTableRow.replace"><code class="docutils literal notranslate"><span class="pre">IndividualTableRow.replace()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#nodetable-classes"><code class="xref py py-class docutils literal notranslate"><span class="pre">NodeTable</span></code> classes</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable"><code class="docutils literal notranslate"><span class="pre">NodeTable</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.add_row"><code class="docutils literal notranslate"><span class="pre">NodeTable.add_row()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.set_columns"><code class="docutils literal notranslate"><span class="pre">NodeTable.set_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.append_columns"><code class="docutils literal notranslate"><span class="pre">NodeTable.append_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.__getitem__"><code class="docutils literal notranslate"><span class="pre">NodeTable.__getitem__()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.append"><code class="docutils literal notranslate"><span class="pre">NodeTable.append()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.asdict"><code class="docutils literal notranslate"><span class="pre">NodeTable.asdict()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.assert_equals"><code class="docutils literal notranslate"><span class="pre">NodeTable.assert_equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.clear"><code class="docutils literal notranslate"><span class="pre">NodeTable.clear()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.copy"><code class="docutils literal notranslate"><span class="pre">NodeTable.copy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.drop_metadata"><code class="docutils literal notranslate"><span class="pre">NodeTable.drop_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.equals"><code class="docutils literal notranslate"><span class="pre">NodeTable.equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.keep_rows"><code class="docutils literal notranslate"><span class="pre">NodeTable.keep_rows()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.metadata_schema"><code class="docutils literal notranslate"><span class="pre">NodeTable.metadata_schema</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.metadata_vector"><code class="docutils literal notranslate"><span class="pre">NodeTable.metadata_vector()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.nbytes"><code class="docutils literal notranslate"><span class="pre">NodeTable.nbytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.packset_metadata"><code class="docutils literal notranslate"><span class="pre">NodeTable.packset_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTable.truncate"><code class="docutils literal notranslate"><span class="pre">NodeTable.truncate()</span></code></a></li>
</ul>
</li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#id10">Associated row class</a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTableRow"><code class="docutils literal notranslate"><span class="pre">NodeTableRow</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTableRow.flags"><code class="docutils literal notranslate"><span class="pre">NodeTableRow.flags</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTableRow.time"><code class="docutils literal notranslate"><span class="pre">NodeTableRow.time</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTableRow.population"><code class="docutils literal notranslate"><span class="pre">NodeTableRow.population</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTableRow.individual"><code class="docutils literal notranslate"><span class="pre">NodeTableRow.individual</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTableRow.metadata"><code class="docutils literal notranslate"><span class="pre">NodeTableRow.metadata</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTableRow.asdict"><code class="docutils literal notranslate"><span class="pre">NodeTableRow.asdict()</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.NodeTableRow.replace"><code class="docutils literal notranslate"><span class="pre">NodeTableRow.replace()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#edgetable-classes"><code class="xref py py-class docutils literal notranslate"><span class="pre">EdgeTable</span></code> classes</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable"><code class="docutils literal notranslate"><span class="pre">EdgeTable</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.add_row"><code class="docutils literal notranslate"><span class="pre">EdgeTable.add_row()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.set_columns"><code class="docutils literal notranslate"><span class="pre">EdgeTable.set_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.append_columns"><code class="docutils literal notranslate"><span class="pre">EdgeTable.append_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.squash"><code class="docutils literal notranslate"><span class="pre">EdgeTable.squash()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.__getitem__"><code class="docutils literal notranslate"><span class="pre">EdgeTable.__getitem__()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.append"><code class="docutils literal notranslate"><span class="pre">EdgeTable.append()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.asdict"><code class="docutils literal notranslate"><span class="pre">EdgeTable.asdict()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.assert_equals"><code class="docutils literal notranslate"><span class="pre">EdgeTable.assert_equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.clear"><code class="docutils literal notranslate"><span class="pre">EdgeTable.clear()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.copy"><code class="docutils literal notranslate"><span class="pre">EdgeTable.copy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.drop_metadata"><code class="docutils literal notranslate"><span class="pre">EdgeTable.drop_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.equals"><code class="docutils literal notranslate"><span class="pre">EdgeTable.equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.keep_rows"><code class="docutils literal notranslate"><span class="pre">EdgeTable.keep_rows()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.metadata_schema"><code class="docutils literal notranslate"><span class="pre">EdgeTable.metadata_schema</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.metadata_vector"><code class="docutils literal notranslate"><span class="pre">EdgeTable.metadata_vector()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.nbytes"><code class="docutils literal notranslate"><span class="pre">EdgeTable.nbytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.packset_metadata"><code class="docutils literal notranslate"><span class="pre">EdgeTable.packset_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTable.truncate"><code class="docutils literal notranslate"><span class="pre">EdgeTable.truncate()</span></code></a></li>
</ul>
</li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#id11">Associated row class</a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTableRow"><code class="docutils literal notranslate"><span class="pre">EdgeTableRow</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTableRow.left"><code class="docutils literal notranslate"><span class="pre">EdgeTableRow.left</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTableRow.right"><code class="docutils literal notranslate"><span class="pre">EdgeTableRow.right</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTableRow.parent"><code class="docutils literal notranslate"><span class="pre">EdgeTableRow.parent</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTableRow.child"><code class="docutils literal notranslate"><span class="pre">EdgeTableRow.child</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTableRow.metadata"><code class="docutils literal notranslate"><span class="pre">EdgeTableRow.metadata</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTableRow.asdict"><code class="docutils literal notranslate"><span class="pre">EdgeTableRow.asdict()</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.EdgeTableRow.replace"><code class="docutils literal notranslate"><span class="pre">EdgeTableRow.replace()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#migrationtable-classes"><code class="xref py py-class docutils literal notranslate"><span class="pre">MigrationTable</span></code> classes</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable"><code class="docutils literal notranslate"><span class="pre">MigrationTable</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.add_row"><code class="docutils literal notranslate"><span class="pre">MigrationTable.add_row()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.set_columns"><code class="docutils literal notranslate"><span class="pre">MigrationTable.set_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.append_columns"><code class="docutils literal notranslate"><span class="pre">MigrationTable.append_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.__getitem__"><code class="docutils literal notranslate"><span class="pre">MigrationTable.__getitem__()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.append"><code class="docutils literal notranslate"><span class="pre">MigrationTable.append()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.asdict"><code class="docutils literal notranslate"><span class="pre">MigrationTable.asdict()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.assert_equals"><code class="docutils literal notranslate"><span class="pre">MigrationTable.assert_equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.clear"><code class="docutils literal notranslate"><span class="pre">MigrationTable.clear()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.copy"><code class="docutils literal notranslate"><span class="pre">MigrationTable.copy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.drop_metadata"><code class="docutils literal notranslate"><span class="pre">MigrationTable.drop_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.equals"><code class="docutils literal notranslate"><span class="pre">MigrationTable.equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.keep_rows"><code class="docutils literal notranslate"><span class="pre">MigrationTable.keep_rows()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.metadata_schema"><code class="docutils literal notranslate"><span class="pre">MigrationTable.metadata_schema</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.metadata_vector"><code class="docutils literal notranslate"><span class="pre">MigrationTable.metadata_vector()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.nbytes"><code class="docutils literal notranslate"><span class="pre">MigrationTable.nbytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.packset_metadata"><code class="docutils literal notranslate"><span class="pre">MigrationTable.packset_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTable.truncate"><code class="docutils literal notranslate"><span class="pre">MigrationTable.truncate()</span></code></a></li>
</ul>
</li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#id12">Associated row class</a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTableRow"><code class="docutils literal notranslate"><span class="pre">MigrationTableRow</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTableRow.left"><code class="docutils literal notranslate"><span class="pre">MigrationTableRow.left</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTableRow.right"><code class="docutils literal notranslate"><span class="pre">MigrationTableRow.right</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTableRow.node"><code class="docutils literal notranslate"><span class="pre">MigrationTableRow.node</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTableRow.source"><code class="docutils literal notranslate"><span class="pre">MigrationTableRow.source</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTableRow.dest"><code class="docutils literal notranslate"><span class="pre">MigrationTableRow.dest</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTableRow.time"><code class="docutils literal notranslate"><span class="pre">MigrationTableRow.time</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTableRow.metadata"><code class="docutils literal notranslate"><span class="pre">MigrationTableRow.metadata</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTableRow.asdict"><code class="docutils literal notranslate"><span class="pre">MigrationTableRow.asdict()</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MigrationTableRow.replace"><code class="docutils literal notranslate"><span class="pre">MigrationTableRow.replace()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#sitetable-classes"><code class="xref py py-class docutils literal notranslate"><span class="pre">SiteTable</span></code> classes</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable"><code class="docutils literal notranslate"><span class="pre">SiteTable</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.add_row"><code class="docutils literal notranslate"><span class="pre">SiteTable.add_row()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.set_columns"><code class="docutils literal notranslate"><span class="pre">SiteTable.set_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.append_columns"><code class="docutils literal notranslate"><span class="pre">SiteTable.append_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.packset_ancestral_state"><code class="docutils literal notranslate"><span class="pre">SiteTable.packset_ancestral_state()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.__getitem__"><code class="docutils literal notranslate"><span class="pre">SiteTable.__getitem__()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.append"><code class="docutils literal notranslate"><span class="pre">SiteTable.append()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.asdict"><code class="docutils literal notranslate"><span class="pre">SiteTable.asdict()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.assert_equals"><code class="docutils literal notranslate"><span class="pre">SiteTable.assert_equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.clear"><code class="docutils literal notranslate"><span class="pre">SiteTable.clear()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.copy"><code class="docutils literal notranslate"><span class="pre">SiteTable.copy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.drop_metadata"><code class="docutils literal notranslate"><span class="pre">SiteTable.drop_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.equals"><code class="docutils literal notranslate"><span class="pre">SiteTable.equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.keep_rows"><code class="docutils literal notranslate"><span class="pre">SiteTable.keep_rows()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.metadata_schema"><code class="docutils literal notranslate"><span class="pre">SiteTable.metadata_schema</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.metadata_vector"><code class="docutils literal notranslate"><span class="pre">SiteTable.metadata_vector()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.nbytes"><code class="docutils literal notranslate"><span class="pre">SiteTable.nbytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.packset_metadata"><code class="docutils literal notranslate"><span class="pre">SiteTable.packset_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTable.truncate"><code class="docutils literal notranslate"><span class="pre">SiteTable.truncate()</span></code></a></li>
</ul>
</li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#id13">Associated row class</a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTableRow"><code class="docutils literal notranslate"><span class="pre">SiteTableRow</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTableRow.position"><code class="docutils literal notranslate"><span class="pre">SiteTableRow.position</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTableRow.ancestral_state"><code class="docutils literal notranslate"><span class="pre">SiteTableRow.ancestral_state</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTableRow.metadata"><code class="docutils literal notranslate"><span class="pre">SiteTableRow.metadata</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTableRow.asdict"><code class="docutils literal notranslate"><span class="pre">SiteTableRow.asdict()</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SiteTableRow.replace"><code class="docutils literal notranslate"><span class="pre">SiteTableRow.replace()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mutationtable-classes"><code class="xref py py-class docutils literal notranslate"><span class="pre">MutationTable</span></code> classes</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable"><code class="docutils literal notranslate"><span class="pre">MutationTable</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.add_row"><code class="docutils literal notranslate"><span class="pre">MutationTable.add_row()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.set_columns"><code class="docutils literal notranslate"><span class="pre">MutationTable.set_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.append_columns"><code class="docutils literal notranslate"><span class="pre">MutationTable.append_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.packset_derived_state"><code class="docutils literal notranslate"><span class="pre">MutationTable.packset_derived_state()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.keep_rows"><code class="docutils literal notranslate"><span class="pre">MutationTable.keep_rows()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.__getitem__"><code class="docutils literal notranslate"><span class="pre">MutationTable.__getitem__()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.append"><code class="docutils literal notranslate"><span class="pre">MutationTable.append()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.asdict"><code class="docutils literal notranslate"><span class="pre">MutationTable.asdict()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.assert_equals"><code class="docutils literal notranslate"><span class="pre">MutationTable.assert_equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.clear"><code class="docutils literal notranslate"><span class="pre">MutationTable.clear()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.copy"><code class="docutils literal notranslate"><span class="pre">MutationTable.copy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.drop_metadata"><code class="docutils literal notranslate"><span class="pre">MutationTable.drop_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.equals"><code class="docutils literal notranslate"><span class="pre">MutationTable.equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.metadata_schema"><code class="docutils literal notranslate"><span class="pre">MutationTable.metadata_schema</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.metadata_vector"><code class="docutils literal notranslate"><span class="pre">MutationTable.metadata_vector()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.nbytes"><code class="docutils literal notranslate"><span class="pre">MutationTable.nbytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.packset_metadata"><code class="docutils literal notranslate"><span class="pre">MutationTable.packset_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTable.truncate"><code class="docutils literal notranslate"><span class="pre">MutationTable.truncate()</span></code></a></li>
</ul>
</li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#id14">Associated row class</a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTableRow"><code class="docutils literal notranslate"><span class="pre">MutationTableRow</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTableRow.site"><code class="docutils literal notranslate"><span class="pre">MutationTableRow.site</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTableRow.node"><code class="docutils literal notranslate"><span class="pre">MutationTableRow.node</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTableRow.derived_state"><code class="docutils literal notranslate"><span class="pre">MutationTableRow.derived_state</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTableRow.parent"><code class="docutils literal notranslate"><span class="pre">MutationTableRow.parent</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTableRow.metadata"><code class="docutils literal notranslate"><span class="pre">MutationTableRow.metadata</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTableRow.time"><code class="docutils literal notranslate"><span class="pre">MutationTableRow.time</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTableRow.asdict"><code class="docutils literal notranslate"><span class="pre">MutationTableRow.asdict()</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MutationTableRow.replace"><code class="docutils literal notranslate"><span class="pre">MutationTableRow.replace()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#populationtable-classes"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationTable</span></code> classes</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable"><code class="docutils literal notranslate"><span class="pre">PopulationTable</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.add_row"><code class="docutils literal notranslate"><span class="pre">PopulationTable.add_row()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.set_columns"><code class="docutils literal notranslate"><span class="pre">PopulationTable.set_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.append_columns"><code class="docutils literal notranslate"><span class="pre">PopulationTable.append_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.__getitem__"><code class="docutils literal notranslate"><span class="pre">PopulationTable.__getitem__()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.append"><code class="docutils literal notranslate"><span class="pre">PopulationTable.append()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.asdict"><code class="docutils literal notranslate"><span class="pre">PopulationTable.asdict()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.assert_equals"><code class="docutils literal notranslate"><span class="pre">PopulationTable.assert_equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.clear"><code class="docutils literal notranslate"><span class="pre">PopulationTable.clear()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.copy"><code class="docutils literal notranslate"><span class="pre">PopulationTable.copy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.drop_metadata"><code class="docutils literal notranslate"><span class="pre">PopulationTable.drop_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.equals"><code class="docutils literal notranslate"><span class="pre">PopulationTable.equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.keep_rows"><code class="docutils literal notranslate"><span class="pre">PopulationTable.keep_rows()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.metadata_schema"><code class="docutils literal notranslate"><span class="pre">PopulationTable.metadata_schema</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.metadata_vector"><code class="docutils literal notranslate"><span class="pre">PopulationTable.metadata_vector()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.nbytes"><code class="docutils literal notranslate"><span class="pre">PopulationTable.nbytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.packset_metadata"><code class="docutils literal notranslate"><span class="pre">PopulationTable.packset_metadata()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTable.truncate"><code class="docutils literal notranslate"><span class="pre">PopulationTable.truncate()</span></code></a></li>
</ul>
</li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#id15">Associated row class</a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTableRow"><code class="docutils literal notranslate"><span class="pre">PopulationTableRow</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTableRow.metadata"><code class="docutils literal notranslate"><span class="pre">PopulationTableRow.metadata</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTableRow.asdict"><code class="docutils literal notranslate"><span class="pre">PopulationTableRow.asdict()</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PopulationTableRow.replace"><code class="docutils literal notranslate"><span class="pre">PopulationTableRow.replace()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#provenancetable-classes"><code class="xref py py-class docutils literal notranslate"><span class="pre">ProvenanceTable</span></code> classes</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.add_row"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.add_row()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.set_columns"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.set_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.append_columns"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.append_columns()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.packset_record"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.packset_record()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.packset_timestamp"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.packset_timestamp()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.equals"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.assert_equals"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.assert_equals()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.append"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.append()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.asdict"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.asdict()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.clear"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.clear()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.copy"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.copy()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.keep_rows"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.keep_rows()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.nbytes"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.nbytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTable.truncate"><code class="docutils literal notranslate"><span class="pre">ProvenanceTable.truncate()</span></code></a></li>
</ul>
</li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#id16">Associated row class</a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTableRow"><code class="docutils literal notranslate"><span class="pre">ProvenanceTableRow</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTableRow.timestamp"><code class="docutils literal notranslate"><span class="pre">ProvenanceTableRow.timestamp</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTableRow.record"><code class="docutils literal notranslate"><span class="pre">ProvenanceTableRow.record</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTableRow.asdict"><code class="docutils literal notranslate"><span class="pre">ProvenanceTableRow.asdict()</span></code></a></li>
<li class="toc-h7 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ProvenanceTableRow.replace"><code class="docutils literal notranslate"><span class="pre">ProvenanceTableRow.replace()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sec-python-api-reference-identity">Identity classes</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-identitysegments-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">IdentitySegments</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegments"><code class="docutils literal notranslate"><span class="pre">IdentitySegments</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegments.num_segments"><code class="docutils literal notranslate"><span class="pre">IdentitySegments.num_segments</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegments.num_pairs"><code class="docutils literal notranslate"><span class="pre">IdentitySegments.num_pairs</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegments.total_span"><code class="docutils literal notranslate"><span class="pre">IdentitySegments.total_span</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegments.pairs"><code class="docutils literal notranslate"><span class="pre">IdentitySegments.pairs</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-identitysegmentlist-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">IdentitySegmentList</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegmentList"><code class="docutils literal notranslate"><span class="pre">IdentitySegmentList</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegmentList.total_span"><code class="docutils literal notranslate"><span class="pre">IdentitySegmentList.total_span</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegmentList.left"><code class="docutils literal notranslate"><span class="pre">IdentitySegmentList.left</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegmentList.right"><code class="docutils literal notranslate"><span class="pre">IdentitySegmentList.right</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegmentList.node"><code class="docutils literal notranslate"><span class="pre">IdentitySegmentList.node</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-identitysegment-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">IdentitySegment</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegment"><code class="docutils literal notranslate"><span class="pre">IdentitySegment</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegment.left"><code class="docutils literal notranslate"><span class="pre">IdentitySegment.left</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegment.right"><code class="docutils literal notranslate"><span class="pre">IdentitySegment.right</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegment.node"><code class="docutils literal notranslate"><span class="pre">IdentitySegment.node</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.IdentitySegment.span"><code class="docutils literal notranslate"><span class="pre">IdentitySegment.span</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#miscellaneous-classes">Miscellaneous classes</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-referencesequence-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">ReferenceSequence</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ReferenceSequence"><code class="docutils literal notranslate"><span class="pre">ReferenceSequence</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ReferenceSequence.is_null"><code class="docutils literal notranslate"><span class="pre">ReferenceSequence.is_null()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ReferenceSequence.data"><code class="docutils literal notranslate"><span class="pre">ReferenceSequence.data</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ReferenceSequence.metadata"><code class="docutils literal notranslate"><span class="pre">ReferenceSequence.metadata</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ReferenceSequence.metadata_bytes"><code class="docutils literal notranslate"><span class="pre">ReferenceSequence.metadata_bytes</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.ReferenceSequence.metadata_schema"><code class="docutils literal notranslate"><span class="pre">ReferenceSequence.metadata_schema</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-metadataschema-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">MetadataSchema</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MetadataSchema"><code class="docutils literal notranslate"><span class="pre">MetadataSchema</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MetadataSchema.asdict"><code class="docutils literal notranslate"><span class="pre">MetadataSchema.asdict()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MetadataSchema.validate_and_encode_row"><code class="docutils literal notranslate"><span class="pre">MetadataSchema.validate_and_encode_row()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MetadataSchema.decode_row"><code class="docutils literal notranslate"><span class="pre">MetadataSchema.decode_row()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MetadataSchema.encode_row"><code class="docutils literal notranslate"><span class="pre">MetadataSchema.encode_row()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MetadataSchema.structured_array_from_buffer"><code class="docutils literal notranslate"><span class="pre">MetadataSchema.structured_array_from_buffer()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MetadataSchema.permissive_json"><code class="docutils literal notranslate"><span class="pre">MetadataSchema.permissive_json()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.MetadataSchema.null"><code class="docutils literal notranslate"><span class="pre">MetadataSchema.null()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-tablemetadataschemas-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">TableMetadataSchemas</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableMetadataSchemas"><code class="docutils literal notranslate"><span class="pre">TableMetadataSchemas</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableMetadataSchemas.node"><code class="docutils literal notranslate"><span class="pre">TableMetadataSchemas.node</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableMetadataSchemas.edge"><code class="docutils literal notranslate"><span class="pre">TableMetadataSchemas.edge</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableMetadataSchemas.site"><code class="docutils literal notranslate"><span class="pre">TableMetadataSchemas.site</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableMetadataSchemas.mutation"><code class="docutils literal notranslate"><span class="pre">TableMetadataSchemas.mutation</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableMetadataSchemas.migration"><code class="docutils literal notranslate"><span class="pre">TableMetadataSchemas.migration</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableMetadataSchemas.individual"><code class="docutils literal notranslate"><span class="pre">TableMetadataSchemas.individual</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableMetadataSchemas.population"><code class="docutils literal notranslate"><span class="pre">TableMetadataSchemas.population</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-topologycounter-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">TopologyCounter</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TopologyCounter"><code class="docutils literal notranslate"><span class="pre">TopologyCounter</span></code></a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-ldcalculator-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">LdCalculator</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.LdCalculator"><code class="docutils literal notranslate"><span class="pre">LdCalculator</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.LdCalculator.r2"><code class="docutils literal notranslate"><span class="pre">LdCalculator.r2()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.LdCalculator.r2_array"><code class="docutils literal notranslate"><span class="pre">LdCalculator.r2_array()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.LdCalculator.r2_matrix"><code class="docutils literal notranslate"><span class="pre">LdCalculator.r2_matrix()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-tablecollectionindexes-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">TableCollectionIndexes</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollectionIndexes"><code class="docutils literal notranslate"><span class="pre">TableCollectionIndexes</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollectionIndexes.asdict"><code class="docutils literal notranslate"><span class="pre">TableCollectionIndexes.asdict()</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.TableCollectionIndexes.nbytes"><code class="docutils literal notranslate"><span class="pre">TableCollectionIndexes.nbytes</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-svgstring-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">SVGString</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SVGString"><code class="docutils literal notranslate"><span class="pre">SVGString</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.SVGString._repr_svg_"><code class="docutils literal notranslate"><span class="pre">SVGString._repr_svg_()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-pcaresult-class">The <code class="xref py py-class docutils literal notranslate"><span class="pre">PCAResult</span></code> class</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PCAResult"><code class="docutils literal notranslate"><span class="pre">PCAResult</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PCAResult.factors"><code class="docutils literal notranslate"><span class="pre">PCAResult.factors</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PCAResult.eigenvalues"><code class="docutils literal notranslate"><span class="pre">PCAResult.eigenvalues</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PCAResult.range_sketch"><code class="docutils literal notranslate"><span class="pre">PCAResult.range_sketch</span></code></a></li>
<li class="toc-h6 nav-item toc-entry"><a class="reference internal nav-link" href="#tskit.PCAResult.error_bound"><code class="docutils literal notranslate"><span class="pre">PCAResult.error_bound</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Tskit Developers
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
       Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>