
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>API Documentation &#8212; Tsinfer manual</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=5349f25f" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'api';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Command line interface" href="cli.html" />
    <link rel="prev" title="Large Scale Inference" href="large_scale.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <script data-goatcounter="https://tskit.goatcounter.com/count" async src="//gc.zgo.at/count.js"></script>
</head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/tsinfer_logo.svg" class="logo__image only-light" alt="Tsinfer manual - Home"/>
    <script>document.write(`<img src="_static/tsinfer_logo.svg" class="logo__image only-dark" alt="Tsinfer manual - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Installation</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Usage</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Inference</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="inference.html">Inference overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="large_scale.html">Large Scale Inference</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Interfaces</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">API Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">Command line interface</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">File Formats</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="file_formats.html">File formats</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Miscellaneous</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="development.html">Developer documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="CITATION.html">Citing tsinfer</a></li>
<li class="toctree-l1"><a class="reference internal" href="CHANGELOG.html">Changelog</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/api.rst" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.rst</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>API Documentation</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#variant-data">Variant data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.VariantData"><code class="docutils literal notranslate"><span class="pre">VariantData</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.add_ancestral_state_array"><code class="docutils literal notranslate"><span class="pre">add_ancestral_state_array()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#ancestor-data">Ancestor data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.load"><code class="docutils literal notranslate"><span class="pre">load()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData"><code class="docutils literal notranslate"><span class="pre">AncestorData</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.data_equal"><code class="docutils literal notranslate"><span class="pre">AncestorData.data_equal()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.sequence_length"><code class="docutils literal notranslate"><span class="pre">AncestorData.sequence_length</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.num_sites"><code class="docutils literal notranslate"><span class="pre">AncestorData.num_sites</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.sites_position"><code class="docutils literal notranslate"><span class="pre">AncestorData.sites_position</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.add_provenance"><code class="docutils literal notranslate"><span class="pre">AncestorData.add_provenance()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.arrays"><code class="docutils literal notranslate"><span class="pre">AncestorData.arrays()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.clear_provenances"><code class="docutils literal notranslate"><span class="pre">AncestorData.clear_provenances()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.close"><code class="docutils literal notranslate"><span class="pre">AncestorData.close()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.copy"><code class="docutils literal notranslate"><span class="pre">AncestorData.copy()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.file_size"><code class="docutils literal notranslate"><span class="pre">AncestorData.file_size</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.info"><code class="docutils literal notranslate"><span class="pre">AncestorData.info</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.provenances"><code class="docutils literal notranslate"><span class="pre">AncestorData.provenances()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.record_provenance"><code class="docutils literal notranslate"><span class="pre">AncestorData.record_provenance()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.ancestors_length"><code class="docutils literal notranslate"><span class="pre">AncestorData.ancestors_length</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.insert_proxy_samples"><code class="docutils literal notranslate"><span class="pre">AncestorData.insert_proxy_samples()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.truncate_ancestors"><code class="docutils literal notranslate"><span class="pre">AncestorData.truncate_ancestors()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.add_ancestor"><code class="docutils literal notranslate"><span class="pre">AncestorData.add_ancestor()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.finalise"><code class="docutils literal notranslate"><span class="pre">AncestorData.finalise()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.ancestor"><code class="docutils literal notranslate"><span class="pre">AncestorData.ancestor()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.ancestors"><code class="docutils literal notranslate"><span class="pre">AncestorData.ancestors()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#running-inference">Running inference</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.infer"><code class="docutils literal notranslate"><span class="pre">infer()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.generate_ancestors"><code class="docutils literal notranslate"><span class="pre">generate_ancestors()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.GenotypeEncoding"><code class="docutils literal notranslate"><span class="pre">GenotypeEncoding</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.GenotypeEncoding.EIGHT_BIT"><code class="docutils literal notranslate"><span class="pre">GenotypeEncoding.EIGHT_BIT</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.GenotypeEncoding.ONE_BIT"><code class="docutils literal notranslate"><span class="pre">GenotypeEncoding.ONE_BIT</span></code></a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.match_ancestors"><code class="docutils literal notranslate"><span class="pre">match_ancestors()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.match_samples"><code class="docutils literal notranslate"><span class="pre">match_samples()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.augment_ancestors"><code class="docutils literal notranslate"><span class="pre">augment_ancestors()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.post_process"><code class="docutils literal notranslate"><span class="pre">post_process()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#batched-inference">Batched inference</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.match_ancestors_batch_init"><code class="docutils literal notranslate"><span class="pre">match_ancestors_batch_init()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.match_ancestors_batch_groups"><code class="docutils literal notranslate"><span class="pre">match_ancestors_batch_groups()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.match_ancestors_batch_group_partition"><code class="docutils literal notranslate"><span class="pre">match_ancestors_batch_group_partition()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.match_ancestors_batch_group_finalise"><code class="docutils literal notranslate"><span class="pre">match_ancestors_batch_group_finalise()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.match_ancestors_batch_finalise"><code class="docutils literal notranslate"><span class="pre">match_ancestors_batch_finalise()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.match_samples_batch_init"><code class="docutils literal notranslate"><span class="pre">match_samples_batch_init()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.match_samples_batch_partition"><code class="docutils literal notranslate"><span class="pre">match_samples_batch_partition()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.match_samples_batch_finalise"><code class="docutils literal notranslate"><span class="pre">match_samples_batch_finalise()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#container-classes">Container classes</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.Variant"><code class="docutils literal notranslate"><span class="pre">Variant</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.Site"><code class="docutils literal notranslate"><span class="pre">Site</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exceptions">Exceptions</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.FileFormatError"><code class="docutils literal notranslate"><span class="pre">FileFormatError</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="api-documentation">
<span id="sec-api"></span><h1>API Documentation<a class="headerlink" href="#api-documentation" title="Link to this heading">#</a></h1>
<section id="variant-data">
<span id="sec-api-file-formats"></span><h2>Variant data<a class="headerlink" href="#variant-data" title="Link to this heading">#</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="tsinfer.VariantData">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tsinfer.</span></span><span class="sig-name descname"><span class="pre">VariantData</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path_or_zarr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ancestral_state</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_mask</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">site_mask</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sites_time</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individuals_time</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individuals_location</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individuals_population</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individuals_flags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/formats.html#VariantData"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.VariantData" title="Link to this definition">#</a></dt>
<dd><p>Class representing input variant data used for inference. This is
mostly a thin wrapper for a Zarr dataset storing information in
the VCF Zarr (.vcz) format, plus information specifing the ancestral allele
and (optional) data masks. It then provides various derived properties and
methods for accessing the data in a form suitable for inference.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In the VariantData object, “samples” refer to the individuals in the dataset,
each of which can be of arbitrary ploidy. This is in contrast to <code class="docutils literal notranslate"><span class="pre">tskit</span></code>,
in which each <em>haploid genome</em> is treated as a separate “sample”. For example
in a diploid dataset, the inferred tree sequence returned at the end of
the inference process will have <code class="docutils literal notranslate"><span class="pre">inferred_ts.num_samples</span></code> equal to double
the number returned by <code class="docutils literal notranslate"><span class="pre">VariantData.num_samples</span></code>.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path_or_zarr</strong> (<em>Union</em><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><a class="reference external" href="https://zarr.readthedocs.io/en/stable/api/zarr/index.html#zarr.Group" title="(in zarr v3.1.2)"><em>zarr.Group</em></a><em>)</em>) – The input dataset in
<a class="reference external" href="https://github.com/sgkit-dev/vcf-zarr-spec">VCF Zarr</a> format.
This can either a path to the Zarr dataset saved on disk, or the
Zarr object itself.</p></li>
<li><p><strong>ancestral_state</strong> (<em>Union</em><em>(</em><em>array</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>)</em>) – A numpy array of strings specifying
the ancestral states (alleles) used in inference. This must be the same length
as the number of unmasked sites in the dataset. Alternatively, a single string
can be provided, giving the name of an array in the input dataset which contains
the ancestral states. Unknown ancestral states can be specified using “N”.
Any ancestral states which do not match any of the known alleles at that site,
will be tallied, and a warning issued summarizing the unknown ancestral states.
Note that allelic states are case-sensitive in tsinfer.</p></li>
<li><p><strong>sample_mask</strong> (<em>Union</em><em>(</em><em>array</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>)</em>) – A numpy array of booleans specifying which
samples to mask out (exclude) from the dataset. Alternatively, a string
can be provided, giving the name of an array in the input dataset which contains
the sample mask. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), all samples are included.</p></li>
<li><p><strong>site_mask</strong> (<em>Union</em><em>(</em><em>array</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>)</em>) – A numpy array of booleans specifying which
sites to mask out (exclude) from the dataset. Alternatively, a string
can be provided, giving the name of an array in the input dataset which contains
the site mask. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), all sites are included.</p></li>
<li><p><strong>sites_time</strong> (<em>Union</em><em>(</em><em>array</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>)</em>) – A numpy array of floats specifying the relative
time of occurrence of the mutation to the derived state at each site. This must
be of the same length as the number of unmasked sites. Alternatively, a
string can be provided, giving the name of an array in the input dataset
which contains the site times. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), the frequency of the
derived allele is used as a proxy for the time of occurrence: this is usually a
reasonable approximation to the relative order of ancestors used for inference.
Time values are ignored for sites not used in inference, such as singletons,
sites with more than two alleles, or sites with an unknown ancestral state.</p></li>
<li><p><strong>individuals_time</strong> (<em>Union</em><em>(</em><em>array</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>)</em>) – A numpy array of floats specifying
the time of each individual in the dataset. This must be the same length
as the number of unmasked individuals. Alternatively, a string can be provided,
giving the name of an array in the input dataset which contains the individual
times. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), individuals are assumed to have
<code class="docutils literal notranslate"><span class="pre">tskit.UNKNOWN_TIME</span></code>.</p></li>
<li><p><strong>individuals_location</strong> (<em>Union</em><em>(</em><em>array</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>)</em>) – A numpy array specifying
the location of each individual in the dataset. This must be the same length
as the number of unmasked individuals. Alternatively, a string can be provided,
giving the name of an array in the input dataset which contains the individual
locations. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), individuals are assumed to have empty
location arrays.</p></li>
<li><p><strong>individuals_population</strong> (<em>Union</em><em>(</em><em>array</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>)</em>) – A numpy array of integers specifying
the population of each individual in the dataset. This must be the same length
as the number of unmasked individuals. Alternatively, a string can be provided,
giving the name of an array in the input dataset which contains the individual
populations. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), individuals are assumed to have
<code class="docutils literal notranslate"><span class="pre">tskit.NULL</span></code> as their population.</p></li>
<li><p><strong>individuals_flags</strong> (<em>Union</em><em>(</em><em>array</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>)</em>) – A numpy array of integers specifying
the flags of each individual in the dataset. This must be the same length
as the number of unmasked individuals. Alternatively, a string can be provided,
giving the name of an array in the input dataset which contains the individual
flags. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), individuals are assumed to have flags set to 0.</p></li>
<li><p><strong>sequence_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – An integer specifying the resulting <cite>sequence_length</cite>
attribute of the output tree sequence. If not specified the <cite>contig_length</cite>
attribute from the undelying zarr store for the contig of the selected variants.
is used. If that is not present then the maximum position plus one of the used
variants is used.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tsinfer.add_ancestral_state_array">
<span class="sig-prename descclassname"><span class="pre">tsinfer.</span></span><span class="sig-name descname"><span class="pre">add_ancestral_state_array</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">zarr_group</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fasta_string</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">array_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ancestral_state'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/formats.html#add_ancestral_state_array"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.add_ancestral_state_array" title="Link to this definition">#</a></dt>
<dd><p>Add an ancestral state array to a zarr group from a string of nucleotides
representing the ancestral sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>zarr_group</strong> (<a class="reference external" href="https://zarr.readthedocs.io/en/stable/api/zarr/index.html#zarr.Group" title="(in zarr v3.1.2)"><em>zarr.Group</em></a>) – A zarr group to add the ancestral state array to.
This should be a VCF Zarr format group containing a “variant_position” array.</p></li>
<li><p><strong>fasta_string</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – A string containing the ancestral sequence,
from e.g. FASTA.</p></li>
<li><p><strong>array_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – The name of the array to create in the zarr group.
Default is “ancestral_state”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The newly inserted ancestral state array.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://zarr.readthedocs.io/en/stable/api/zarr/index.html#zarr.Array" title="(in zarr v3.1.2)">zarr.Array</a></p>
</dd>
</dl>
</dd></dl>

</section>
<section id="ancestor-data">
<h2>Ancestor data<a class="headerlink" href="#ancestor-data" title="Link to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="tsinfer.load">
<span class="sig-prename descclassname"><span class="pre">tsinfer.</span></span><span class="sig-name descname"><span class="pre">load</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/formats.html#load"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.load" title="Link to this definition">#</a></dt>
<dd><p>Loads a tsinfer <a class="reference internal" href="#tsinfer.AncestorData" title="tsinfer.AncestorData"><code class="xref py py-class docutils literal notranslate"><span class="pre">AncestorData</span></code></a> file from
the specified path. If the file is format not recognised a
<a class="reference internal" href="#tsinfer.FileFormatError" title="tsinfer.FileFormatError"><code class="xref py py-class docutils literal notranslate"><span class="pre">FileFormatError</span></code></a> will be thrown.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>path</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – The path of the file we wish to load.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The corresponding <a class="reference internal" href="#tsinfer.AncestorData" title="tsinfer.AncestorData"><code class="xref py py-class docutils literal notranslate"><span class="pre">AncestorData</span></code></a>
instance opened in read only mode.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tsinfer.AncestorData" title="tsinfer.AncestorData"><code class="xref py py-class docutils literal notranslate"><span class="pre">AncestorData</span></code></a>.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#tsinfer.FileFormatError" title="tsinfer.FileFormatError"><code class="xref py py-class docutils literal notranslate"><span class="pre">FileFormatError</span></code></a> if the file cannot be read.</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="tsinfer.AncestorData">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tsinfer.</span></span><span class="sig-name descname"><span class="pre">AncestorData</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">position</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence_length</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_flush_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compressor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunk_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1024</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_file_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/formats.html#AncestorData"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.AncestorData" title="Link to this definition">#</a></dt>
<dd><p>Class representing the stored ancestor data produced by
<a class="reference internal" href="#tsinfer.generate_ancestors" title="tsinfer.generate_ancestors"><code class="xref py py-func docutils literal notranslate"><span class="pre">generate_ancestors()</span></code></a>. See the ancestor data file format
<a class="reference internal" href="file_formats.html#sec-file-formats-ancestors"><span class="std std-ref">specifications</span></a> for details on the structure
of this file.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If a <code class="docutils literal notranslate"><span class="pre">path</span></code> is specified, the <code class="docutils literal notranslate"><span class="pre">max_file_size</span></code> option puts an
upper limit on the possible size of the created file. On non-Windows
systems, space for this file is not immediately allocated but just
“reserved” using sparse file systems. However, on Windows systems
the file is allocated immediately, so <code class="docutils literal notranslate"><span class="pre">max_file_size</span></code> takes a smaller
default value, to avoid allocating very large files for no reason.
Users who wish to run large inferences on Windows may therefore need to
explictly set an appropriate <code class="docutils literal notranslate"><span class="pre">max_file_size</span></code>. Note that the
<code class="docutils literal notranslate"><span class="pre">max_file_size</span></code> is only used while the file is being built: one the
file has been finalised, it is shrunk to its minimum size.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>position</strong> (<em>arraylike</em>) – Integer array of the site positions of the ancestors.
All values should be &gt;0 and the array should be monotonically increasing.</p></li>
<li><p><strong>sequence_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a>) – Total length of the sequence, site positions must
be less than this value.</p></li>
<li><p><strong>path</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – The path of the file to store the ancestor data. If None,
the information is stored in memory and not persistent.</p></li>
<li><p><strong>num_flush_threads</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – The number of background threads to use
for compressing data and flushing to disc. If &lt;= 0, do not spawn
any threads but use a synchronous algorithm instead. Default=0.</p></li>
<li><p><strong>compressor</strong> (<a class="reference external" href="https://numcodecs.readthedocs.io/en/stable/abc.html#numcodecs.abc.Codec" title="(in numcodecs v0.16.2)"><em>numcodecs.abc.Codec</em></a>) – A <a class="reference external" href="https://numcodecs.readthedocs.io/en/stable/abc.html#numcodecs.abc.Codec" title="(in numcodecs v0.16.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numcodecs.abc.Codec</span></code></a>
instance to use for compressing data. Any codec may be used, but
problems may occur with very large datasets on certain codecs as
they cannot compress buffers &gt;2GB. If None, do not use any compression.
Default=:class:<cite>numcodecs.zstd.Zstd</cite>.</p></li>
<li><p><strong>chunk_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – The chunk size used for
<a class="reference external" href="http://zarr.readthedocs.io/">zarr arrays</a> in the sample dimension. This
affects compression level and algorithm performance. Default=1024.</p></li>
<li><p><strong>chunk_size_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – <p>The chunk size used for the genotype
<a class="reference external" href="http://zarr.readthedocs.io/">zarr arrays</a> in the sites dimension. This affects
compression level and algorithm performance. Default=16384.</p>
</p></li>
<li><p><strong>max_file_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – If a file is being used to store this data, set
a maximum size in bytes for the stored file. If None, the default
value of 1GiB is used on Windows and 1TiB on other
platforms (see above for details).</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="tsinfer.AncestorData.data_equal">
<span class="sig-name descname"><span class="pre">data_equal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/formats.html#AncestorData.data_equal"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.AncestorData.data_equal" title="Link to this definition">#</a></dt>
<dd><p>Returns True if all the data attributes of this input file and the
specified input file are equal. This compares every attribute.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tsinfer.AncestorData.sequence_length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence_length</span></span><a class="headerlink" href="#tsinfer.AncestorData.sequence_length" title="Link to this definition">#</a></dt>
<dd><p>Returns the sequence length.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tsinfer.AncestorData.num_sites">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_sites</span></span><a class="headerlink" href="#tsinfer.AncestorData.num_sites" title="Link to this definition">#</a></dt>
<dd><p>The number of inference sites used to generate the ancestors</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tsinfer.AncestorData.sites_position">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sites_position</span></span><a class="headerlink" href="#tsinfer.AncestorData.sites_position" title="Link to this definition">#</a></dt>
<dd><p>The positions of the inference sites used to generate the ancestors</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tsinfer.AncestorData.add_provenance">
<span class="sig-name descname"><span class="pre">add_provenance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">timestamp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tsinfer.AncestorData.add_provenance" title="Link to this definition">#</a></dt>
<dd><p>Adds a new provenance record with the specified timestamp and record.
Timestamps should ISO8601 formatted, and record is some JSON encodable
object.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tsinfer.AncestorData.arrays">
<span class="sig-name descname"><span class="pre">arrays</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tsinfer.AncestorData.arrays" title="Link to this definition">#</a></dt>
<dd><p>Returns a list of all the zarr arrays in this DataContainer.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tsinfer.AncestorData.clear_provenances">
<span class="sig-name descname"><span class="pre">clear_provenances</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tsinfer.AncestorData.clear_provenances" title="Link to this definition">#</a></dt>
<dd><p>Clear all provenances in this instance</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tsinfer.AncestorData.close">
<span class="sig-name descname"><span class="pre">close</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tsinfer.AncestorData.close" title="Link to this definition">#</a></dt>
<dd><p>Close this DataContainer. Any read or write operations attempted
after calling this will fail.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tsinfer.AncestorData.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_file_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tsinfer.AncestorData.copy" title="Link to this definition">#</a></dt>
<dd><p>Returns a copy of this DataContainer opened in ‘edit’ mode. If path
is specified, this must not be equal to the path of the current
data container.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tsinfer.AncestorData.file_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">file_size</span></span><a class="headerlink" href="#tsinfer.AncestorData.file_size" title="Link to this definition">#</a></dt>
<dd><p>Returns the size of the underlying file, or -1 if we do not have a
file associated.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tsinfer.AncestorData.info">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">info</span></span><a class="headerlink" href="#tsinfer.AncestorData.info" title="Link to this definition">#</a></dt>
<dd><p>Returns a string containing the zarr info for each array.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tsinfer.AncestorData.provenances">
<span class="sig-name descname"><span class="pre">provenances</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#tsinfer.AncestorData.provenances" title="Link to this definition">#</a></dt>
<dd><p>Returns an iterator over the (timestamp, record) pairs representing
the provenances for this data container.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tsinfer.AncestorData.record_provenance">
<span class="sig-name descname"><span class="pre">record_provenance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">command</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tsinfer.AncestorData.record_provenance" title="Link to this definition">#</a></dt>
<dd><p>Records the provenance information for this file using the
tskit provenances schema.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="tsinfer.AncestorData.ancestors_length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ancestors_length</span></span><a class="headerlink" href="#tsinfer.AncestorData.ancestors_length" title="Link to this definition">#</a></dt>
<dd><p>Returns the length of ancestors in physical coordinates.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tsinfer.AncestorData.insert_proxy_samples">
<span class="sig-name descname"><span class="pre">insert_proxy_samples</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant_data</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epsilon</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_mutation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/formats.html#AncestorData.insert_proxy_samples"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.AncestorData.insert_proxy_samples" title="Link to this definition">#</a></dt>
<dd><p>Take a set of samples from a {class}`VariantData` instance and create additional
“proxy sample ancestors” from them, returning a new <a class="reference internal" href="#tsinfer.AncestorData" title="tsinfer.AncestorData"><code class="xref py py-class docutils literal notranslate"><span class="pre">AncestorData</span></code></a>
instance including both the current ancestors and the additional ancestors
at the appropriate time points.</p>
<p>A <em>proxy sample ancestor</em> is an ancestor based upon a known sample. At
sites used in the full inference process, the haplotype of this ancestor
is identical to that of the sample on which it is based. The time of the
ancestor is taken to be a fraction <code class="docutils literal notranslate"><span class="pre">epsilon</span></code> older than the sample on
which it is based.</p>
<p>A common use of this function is to provide ancestral nodes for anchoring
historical samples at the correct time when matching them into a tree
sequence during the <a class="reference internal" href="#tsinfer.match_samples" title="tsinfer.match_samples"><code class="xref py py-func docutils literal notranslate"><span class="pre">tsinfer.match_samples()</span></code></a> stage of inference.
For this reason, by default, the samples chosen from <code class="docutils literal notranslate"><span class="pre">sample_data</span></code>
are those associated with historical (i.e. non-contemporary)
<a class="reference internal" href="inference.html#sec-inference-data-model-individual"><span class="std std-ref">individuals</span></a>. This can be
altered by using the <code class="docutils literal notranslate"><span class="pre">sample_ids</span></code> parameter.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The proxy sample ancestors inserted here will correspond to extra nodes
in the inferred tree sequence. At sites which are not used in the full
inference process (e.g. sites unique to a single historical sample),
these proxy sample ancestor nodes may have a different genotype from
their corresponding sample.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variant_data</strong> (<a class="reference internal" href="#tsinfer.VariantData" title="tsinfer.VariantData"><em>VariantData</em></a>) – The <cite>VariantData</cite> instance
from which to select the samples used to create extra ancestors.</p></li>
<li><p><strong>sample_ids</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>)</em>) – A list of sample ids in the <code class="docutils literal notranslate"><span class="pre">variant_data</span></code>
instance that will be selected to create the extra ancestors. If
<code class="docutils literal notranslate"><span class="pre">None</span></code> (default) select all the historical samples, i.e. those
associated with an <a class="reference internal" href="inference.html#sec-inference-data-model-individual"><span class="std std-ref">Individual</span></a> whose
time is greater than zero. The order of ids is ignored, as are
duplicate ids.</p></li>
<li><p><strong>epsilon</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a><em>)</em>) – An list of small time increments
determining how much older each proxy sample ancestor is than the
corresponding sample listed in <code class="docutils literal notranslate"><span class="pre">sample_ids</span></code>. A single value is also
allowed, in which case it is used as the time increment for all selected
proxy sample ancestors. If None (default) find <span class="math notranslate nohighlight">\({\delta}t\)</span>, the
smallest time difference between between the sample times and the next
oldest ancestor in the current <a class="reference internal" href="#tsinfer.AncestorData" title="tsinfer.AncestorData"><code class="xref py py-class docutils literal notranslate"><span class="pre">AncestorData</span></code></a> instance, setting
<code class="docutils literal notranslate"><span class="pre">epsilon</span></code> = <span class="math notranslate nohighlight">\({\delta}t / 100\)</span> (or, if all selected samples
are at least as old as the oldest ancestor, take <span class="math notranslate nohighlight">\({\delta}t\)</span>
to be the smallest non-zero time difference between existing ancestors).</p></li>
<li><p><strong>allow_mutation</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> (the default), any site in a proxy
sample ancestor that has a derived allele must have a pre-existing
mutation in an older (non-proxy) ancestor, otherwise an error is raised.
Alternatively, if <code class="docutils literal notranslate"><span class="pre">allow_mutation</span></code> is <code class="docutils literal notranslate"><span class="pre">True</span></code>, proxy ancestors can
contain a de-novo mutation at a site that also has a mutation elsewhere
(i.e. breaking the infinite sites assumption), allowing them to possess
derived alleles at sites where there are no pre-existing mutations in
older ancestors.</p></li>
<li><p><strong>**kwargs</strong> – Further arguments passed to the constructor when creating
the new <a class="reference internal" href="#tsinfer.AncestorData" title="tsinfer.AncestorData"><code class="xref py py-class docutils literal notranslate"><span class="pre">AncestorData</span></code></a> instance which will be returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A new <a class="reference internal" href="#tsinfer.AncestorData" title="tsinfer.AncestorData"><code class="xref py py-class docutils literal notranslate"><span class="pre">AncestorData</span></code></a> object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tsinfer.AncestorData" title="tsinfer.AncestorData">AncestorData</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tsinfer.AncestorData.truncate_ancestors">
<span class="sig-name descname"><span class="pre">truncate_ancestors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lower_time_bound</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">upper_time_bound</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">length_multiplier</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">buffer_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/formats.html#AncestorData.truncate_ancestors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.AncestorData.truncate_ancestors" title="Link to this definition">#</a></dt>
<dd><p>Truncates the length of ancestors above a given time and returns a new
<a class="reference internal" href="#tsinfer.AncestorData" title="tsinfer.AncestorData"><code class="xref py py-class docutils literal notranslate"><span class="pre">AncestorData</span></code></a> instance.</p>
<p>Given a set of haplotypes H such that <code class="docutils literal notranslate"><span class="pre">lower_time_bound</span></code> &lt;= <code class="docutils literal notranslate"><span class="pre">h.time</span></code> &lt;
<code class="docutils literal notranslate"><span class="pre">upper_time_bound</span></code>, we let <code class="docutils literal notranslate"><span class="pre">max_len</span> <span class="pre">=</span> <span class="pre">length_multiplier</span> <span class="pre">*</span> <span class="pre">max(max(h.length)</span>
<span class="pre">for</span> <span class="pre">h</span> <span class="pre">in</span> <span class="pre">H)</span></code>. Then, we truncate all haplotypes containing at least one focal
site where <code class="docutils literal notranslate"><span class="pre">h.time</span> <span class="pre">&gt;=</span> <span class="pre">upper</span></code>, ensuring these haplotypes extend no further than
half of <code class="docutils literal notranslate"><span class="pre">max_len</span></code> to the either side of the leftmost and
rightmost focal sites of the ancestral haplotype. Note that ancestors above
<code class="docutils literal notranslate"><span class="pre">upper_time_bound</span></code> may still be longer than <code class="docutils literal notranslate"><span class="pre">max_len</span></code> if the ancestor
contains greater than 2 focal sites.</p>
<p>This function should be used when <a class="reference internal" href="#tsinfer.generate_ancestors" title="tsinfer.generate_ancestors"><code class="xref py py-func docutils literal notranslate"><span class="pre">tsinfer.generate_ancestors()</span></code></a> generates
old ancestors which are very long, as these can significantly slow down matching
time. Older ancestors should generally be shorter than younger ancestors, so
truncating the lengths of older ancestors has negligible effect on inference
accuracy.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Please ensure that the time values provided to <code class="docutils literal notranslate"><span class="pre">lower_time_bound</span></code> and
<code class="docutils literal notranslate"><span class="pre">upper_time_bound</span></code> match the units used in the <a class="reference internal" href="#tsinfer.AncestorData" title="tsinfer.AncestorData"><code class="xref py py-class docutils literal notranslate"><span class="pre">AncestorData</span></code></a>
file, i.e. if your ancestors do not have site times specified,
<code class="docutils literal notranslate"><span class="pre">upper_time_bound</span></code> should be between 0 and 1.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>lower_time_bound</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a>) – Defines the lower bound (inclusive) of the half
open interval where we search for a truncation value.</p></li>
<li><p><strong>upper_time_bound</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a>) – Defines the upper bound (exclusive) of the half
open interval where we search for a truncation value. The truncation value
is the length of the longest haplotype in this interval multiplied by
<code class="docutils literal notranslate"><span class="pre">length_multiplier</span></code>. The length of ancestors as old or older than
<code class="docutils literal notranslate"><span class="pre">upper_time_bound</span></code> will be truncated using this value.</p></li>
<li><p><strong>length_multiplier</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a>) – A multiplier for the length of the longest
ancestor in the half-open interval between <code class="docutils literal notranslate"><span class="pre">lower_time_bound</span></code> (inclusive)
and <code class="docutils literal notranslate"><span class="pre">uppper_time_bound</span></code> (exclusive), i.e.
if the longest ancestor in the interval is 1 megabase, a
<code class="docutils literal notranslate"><span class="pre">length_multiplier</span></code> of 2 creates a maximum length of 2 megabases.</p></li>
<li><p><strong>buffer_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – The number of changed ancestors to buffer before
writing to disk.</p></li>
<li><p><strong>**kwargs</strong> – Further arguments passed to the <a class="reference internal" href="#tsinfer.AncestorData.copy" title="tsinfer.AncestorData.copy"><code class="xref py py-func docutils literal notranslate"><span class="pre">AncestorData.copy()</span></code></a>
when creating the new <a class="reference internal" href="#tsinfer.AncestorData" title="tsinfer.AncestorData"><code class="xref py py-class docutils literal notranslate"><span class="pre">AncestorData</span></code></a> instance which will be returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A new <a class="reference internal" href="#tsinfer.AncestorData" title="tsinfer.AncestorData"><code class="xref py py-class docutils literal notranslate"><span class="pre">AncestorData</span></code></a> object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tsinfer.AncestorData" title="tsinfer.AncestorData">AncestorData</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tsinfer.AncestorData.add_ancestor">
<span class="sig-name descname"><span class="pre">add_ancestor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">focal_sites</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">haplotype</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/formats.html#AncestorData.add_ancestor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.AncestorData.add_ancestor" title="Link to this definition">#</a></dt>
<dd><p>Adds an ancestor with the specified haplotype, with ancestral material over the
interval [start:end], that is associated with the specified timepoint and has new
mutations at the specified list of focal sites. Ancestors should be added in time
order, with the oldest first. The id of the added ancestor is returned.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tsinfer.AncestorData.finalise">
<span class="sig-name descname"><span class="pre">finalise</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/formats.html#AncestorData.finalise"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.AncestorData.finalise" title="Link to this definition">#</a></dt>
<dd><p>Ensures that the state of the data is flushed and writes the
provenance for the current operation. The specified ‘command’ is used
to fill the corresponding entry in the provenance dictionary.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tsinfer.AncestorData.ancestor">
<span class="sig-name descname"><span class="pre">ancestor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id_</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/formats.html#AncestorData.ancestor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.AncestorData.ancestor" title="Link to this definition">#</a></dt>
<dd><p>Returns the ancestor with the specified ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><cite>Ancestor</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tsinfer.AncestorData.ancestors">
<span class="sig-name descname"><span class="pre">ancestors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">indexes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/formats.html#AncestorData.ancestors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.AncestorData.ancestors" title="Link to this definition">#</a></dt>
<dd><p>Returns an iterator over all the ancestors. If indexes is provided, it should
be a sorted list of indexes giving a subset of ancestors to return.
For efficiency, the indexes should be a numpy integer array.</p>
</dd></dl>

</dd></dl>

<div class="admonition-todo admonition" id="id2">
<p class="admonition-title">Todo</p>
<ol class="arabic simple">
<li><p>Add documentation for the data attributes in read-mode.</p></li>
</ol>
</div>
<section id="running-inference">
<span id="sec-api-file-inference"></span><h3>Running inference<a class="headerlink" href="#running-inference" title="Link to this heading">#</a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="tsinfer.infer">
<span class="sig-prename descclassname"><span class="pre">tsinfer.</span></span><span class="sig-name descname"><span class="pre">infer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant_data</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recombination_rate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mismatch_ratio</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_compression</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_positions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">post_process</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/inference.html#infer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.infer" title="Link to this definition">#</a></dt>
<dd><p>Runs the full <a class="reference internal" href="inference.html#sec-inference"><span class="std std-ref">inference pipeline</span></a> on the specified
<a class="reference internal" href="#tsinfer.VariantData" title="tsinfer.VariantData"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantData</span></code></a> instance and returns the inferred
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a>.  See
<a class="reference internal" href="inference.html#sec-inference-match-ancestors-and-samples"><span class="std std-ref">matching ancestors &amp; samples</span></a>
in the documentation for details of <code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code>, <code class="docutils literal notranslate"><span class="pre">mismatch_ratio</span></code>
and <code class="docutils literal notranslate"><span class="pre">path_compression</span></code>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For finer grained control over inference, for example to set mismatch
ratios when matching ancestors as well as when matching samples, run
<a class="reference internal" href="#tsinfer.generate_ancestors" title="tsinfer.generate_ancestors"><code class="xref py py-func docutils literal notranslate"><span class="pre">tsinfer.generate_ancestors()</span></code></a>, <a class="reference internal" href="#tsinfer.match_ancestors" title="tsinfer.match_ancestors"><code class="xref py py-func docutils literal notranslate"><span class="pre">tsinfer.match_ancestors()</span></code></a> and
<a class="reference internal" href="#tsinfer.match_samples" title="tsinfer.match_samples"><code class="xref py py-func docutils literal notranslate"><span class="pre">tsinfer.match_samples()</span></code></a> separately.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variant_data</strong> (<a class="reference internal" href="#tsinfer.VariantData" title="tsinfer.VariantData"><em>VariantData</em></a>) – The input <a class="reference internal" href="#tsinfer.VariantData" title="tsinfer.VariantData"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantData</span></code></a> instance
representing the observed data that we wish to make inferences from.</p></li>
<li><p><strong>recombination_rate</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a><em>, </em><a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.RateMap" title="(in Project name not set)"><em>msprime.RateMap</em></a>) – Either a floating point value giving a constant rate
<span class="math notranslate nohighlight">\(\rho\)</span> per unit length of genome, or an <a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.RateMap" title="(in Project name not set)"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.RateMap</span></code></a>
object. This is used to calculate the probability of recombination between
adjacent sites in the match_samples stage. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, all matching
conflicts are resolved by recombination and all inference sites will have
a single mutation (equivalent to mismatch_ratio near zero).</p></li>
<li><p><strong>mismatch_ratio</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a>) – The probability of a mismatch relative to the median
probability of recombination between adjacent sites: can only be used if a
recombination rate has been set (default: <code class="docutils literal notranslate"><span class="pre">None</span></code> treated as 1 if
<code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code> is set). This is only applied in the match_samples stage.</p></li>
<li><p><strong>path_compression</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – Whether to merge edges that share identical
paths (essentially taking advantage of shared recombination breakpoints).</p></li>
<li><p><strong>post_process</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – Whether to run the <a class="reference internal" href="#tsinfer.post_process" title="tsinfer.post_process"><code class="xref py py-func docutils literal notranslate"><span class="pre">post_process()</span></code></a> method on the
the tree sequence which, among other things, removes ancestral material that
does not end up in the current samples (if not specified, defaults to <code class="docutils literal notranslate"><span class="pre">True</span></code>)</p></li>
<li><p><strong>exclude_positions</strong> (<em>array_like</em>) – A list of site positions to exclude
for full inference. Sites with these positions will not be used to generate
ancestors, and not used during the copying process. Any such sites that
exist in the sample data file will be included in the trees after the
main inference process using parsimony. The list does not need to be
in to be in any particular order, and can include site positions that
are not present in the sample data file.</p></li>
<li><p><strong>num_threads</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – The number of worker threads to use in parallelised
sections of the algorithm. If &lt;= 0, do not spawn any threads and
use simpler sequential algorithms (default).</p></li>
<li><p><strong>simplify</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – When post_processing, only simplify the tree sequence.
deprecated but retained for backwards compatibility (default: <code class="docutils literal notranslate"><span class="pre">None</span></code>).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a> object inferred from the
input sample data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)">tskit.TreeSequence</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tsinfer.generate_ancestors">
<span class="sig-prename descclassname"><span class="pre">tsinfer.</span></span><span class="sig-name descname"><span class="pre">generate_ancestors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant_data</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_positions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genotype_encoding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mmap_temp_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/inference.html#generate_ancestors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.generate_ancestors" title="Link to this definition">#</a></dt>
<dd><p>Runs the ancestor generation <a class="reference internal" href="inference.html#sec-inference-generate-ancestors"><span class="std std-ref">algorithm</span></a>
on the specified <a class="reference internal" href="#tsinfer.VariantData" title="tsinfer.VariantData"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantData</span></code></a> instance and returns the resulting
<a class="reference internal" href="#tsinfer.AncestorData" title="tsinfer.AncestorData"><code class="xref py py-class docutils literal notranslate"><span class="pre">AncestorData</span></code></a> instance. If you wish to store generated ancestors
persistently on file you must pass the <code class="docutils literal notranslate"><span class="pre">path</span></code> keyword argument to this
function. For example,</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">ancestor_data</span> <span class="o">=</span> <span class="n">tsinfer</span><span class="o">.</span><span class="n">generate_ancestors</span><span class="p">(</span><span class="n">variant_data</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="s2">&quot;mydata.ancestors&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Other keyword arguments are passed to the <a class="reference internal" href="#tsinfer.AncestorData" title="tsinfer.AncestorData"><code class="xref py py-class docutils literal notranslate"><span class="pre">AncestorData</span></code></a> constructor,
which may be used to control the storage properties of the generated file.</p>
<p>Ancestor generation involves loading the entire genotype matrix into
memory, by default using one byte per haploid genotype, which can be
prohibitively large when working with sample sizes of 100,000 or more.
There are two options to help mitigate memory usage. The
<code class="docutils literal notranslate"><span class="pre">genotype_encoding</span></code> parameter allows the user to specify a more compact
encoding scheme, which reduces storage space for datasets with small
numbers of alleles. Currently, the <a class="reference internal" href="#tsinfer.GenotypeEncoding.ONE_BIT" title="tsinfer.GenotypeEncoding.ONE_BIT"><code class="xref py py-attr docutils literal notranslate"><span class="pre">GenotypeEncoding.ONE_BIT</span></code></a>
encoding is supported, which provides 8-fold compression of biallelic,
non-missing data. An error is raised if an encoding that does not support
the range of values present in a given dataset is provided.</p>
<p>The second option for reducing the RAM footprint of this function is to
use the <code class="docutils literal notranslate"><span class="pre">mmap_temp_dir</span></code> parameter. This allows the genotype data to be
cached on file, transparently using the operating system’s virtual memory
subsystem to swap in and out the data. This can work well if the encoded
genotype matrix <em>almost</em> fits in RAM and fast local storage is available.
However, if the size of the encoded genotype matrix is more than, say,
twice the available RAM it is unlikely that this function will complete
in a reasonable time-frame. A temporary file is created in the specified
<code class="docutils literal notranslate"><span class="pre">mmap_temp_dir</span></code>, which is automatically deleted when the function
completes.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The <code class="docutils literal notranslate"><span class="pre">mmap_temp_dir</span></code> option is a silent no-op on Windows!</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variant_data</strong> (<a class="reference internal" href="#tsinfer.VariantData" title="tsinfer.VariantData"><em>VariantData</em></a>) – The <a class="reference internal" href="#tsinfer.VariantData" title="tsinfer.VariantData"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantData</span></code></a> instance that we are
generating putative ancestors from.</p></li>
<li><p><strong>path</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – The path of the file to store the ancestor data. If None,
the information is stored in memory and not persistent.</p></li>
<li><p><strong>exclude_positions</strong> (<em>array_like</em>) – A list of site positions to exclude
for full inference. Sites with these positions will not be used to generate
ancestors, and not used during the copying process. The list does not
need be in any particular order.</p></li>
<li><p><strong>num_threads</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – The number of worker threads to use. If &lt; 1, use a
simpler synchronous algorithm.</p></li>
<li><p><strong>genotype_encoding</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – The encoding to use for genotype data internally
when generating ancestors. See the <a class="reference internal" href="#tsinfer.GenotypeEncoding" title="tsinfer.GenotypeEncoding"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenotypeEncoding</span></code></a> class for
the available options. Defaults to one-byte per genotype.</p></li>
<li><p><strong>mmap_temp_dir</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – The directory within which to create the
temporary backing file when using mmaped memory for bulk genotype
storage. If None (the default) allocate memory directly using the
standard mechanism. This is an advanced option, usually only relevant
when working with very large datasets (see above for more information).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The inferred ancestors stored in an <a class="reference internal" href="#tsinfer.AncestorData" title="tsinfer.AncestorData"><code class="xref py py-class docutils literal notranslate"><span class="pre">AncestorData</span></code></a> instance.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tsinfer.AncestorData" title="tsinfer.AncestorData">AncestorData</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="tsinfer.GenotypeEncoding">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tsinfer.</span></span><span class="sig-name descname"><span class="pre">GenotypeEncoding</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">module</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qualname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">boundary</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/constants.html#GenotypeEncoding"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.GenotypeEncoding" title="Link to this definition">#</a></dt>
<dd><p>The encoding scheme used to store genotypes.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="tsinfer.GenotypeEncoding.EIGHT_BIT">
<span class="sig-name descname"><span class="pre">EIGHT_BIT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">0</span></em><a class="headerlink" href="#tsinfer.GenotypeEncoding.EIGHT_BIT" title="Link to this definition">#</a></dt>
<dd><p>The default approach of using one-byte per genotype. Supports up to 127 alleles
and missing data.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="tsinfer.GenotypeEncoding.ONE_BIT">
<span class="sig-name descname"><span class="pre">ONE_BIT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">1</span></em><a class="headerlink" href="#tsinfer.GenotypeEncoding.ONE_BIT" title="Link to this definition">#</a></dt>
<dd><p>Encode binary genotype data using a single bit.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tsinfer.match_ancestors">
<span class="sig-prename descclassname"><span class="pre">tsinfer.</span></span><span class="sig-name descname"><span class="pre">match_ancestors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ancestor_data</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recombination_rate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mismatch_ratio</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_compression</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/inference.html#match_ancestors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.match_ancestors" title="Link to this definition">#</a></dt>
<dd><p>Run the ancestor matching <a class="reference internal" href="inference.html#sec-inference-match-ancestors"><span class="std std-ref">algorithm</span></a>
on the specified <a class="reference internal" href="#tsinfer.VariantData" title="tsinfer.VariantData"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantData</span></code></a> and <a class="reference internal" href="#tsinfer.AncestorData" title="tsinfer.AncestorData"><code class="xref py py-class docutils literal notranslate"><span class="pre">AncestorData</span></code></a> instances,
returning the resulting <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a> representing the
complete ancestry of the putative ancestors. See
<a class="reference internal" href="inference.html#sec-inference-match-ancestors-and-samples"><span class="std std-ref">matching ancestors &amp; samples</span></a>
in the documentation for details of <code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code>, <code class="docutils literal notranslate"><span class="pre">mismatch_ratio</span></code>
and <code class="docutils literal notranslate"><span class="pre">path_compression</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variant_data</strong> (<a class="reference internal" href="#tsinfer.VariantData" title="tsinfer.VariantData"><em>VariantData</em></a>) – The <a class="reference internal" href="#tsinfer.VariantData" title="tsinfer.VariantData"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantData</span></code></a> instance
representing the input data.</p></li>
<li><p><strong>ancestor_data</strong> (<a class="reference internal" href="#tsinfer.AncestorData" title="tsinfer.AncestorData"><em>AncestorData</em></a>) – The <a class="reference internal" href="#tsinfer.AncestorData" title="tsinfer.AncestorData"><code class="xref py py-class docutils literal notranslate"><span class="pre">AncestorData</span></code></a> instance
representing the set of ancestral haplotypes for which we are finding
a history.</p></li>
<li><p><strong>recombination_rate</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a><em>, </em><a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.RateMap" title="(in Project name not set)"><em>msprime.RateMap</em></a>) – Either a floating point value giving a constant rate
<span class="math notranslate nohighlight">\(\rho\)</span> per unit length of genome, or an <a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.RateMap" title="(in Project name not set)"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.RateMap</span></code></a>
object. This is used to calculate the probability of recombination between
adjacent sites. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, all matching conflicts are resolved by
recombination and all inference sites will have a single mutation
(equivalent to mismatch_ratio near zero)</p></li>
<li><p><strong>mismatch_ratio</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a>) – The probability of a mismatch relative to the median
probability of recombination between adjacent sites: can only be used if a
recombination rate has been set (default: <code class="docutils literal notranslate"><span class="pre">None</span></code> treated as 1 if
<code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code> is set).</p></li>
<li><p><strong>path_compression</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – Whether to merge edges that share identical
paths (essentially taking advantage of shared recombination breakpoints).</p></li>
<li><p><strong>num_threads</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – The number of match worker threads to use. If
this is &lt;= 0 then a simpler sequential algorithm is used (default).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The ancestors tree sequence representing the inferred history
of the set of ancestors.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)">tskit.TreeSequence</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tsinfer.match_samples">
<span class="sig-prename descclassname"><span class="pre">tsinfer.</span></span><span class="sig-name descname"><span class="pre">match_samples</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ancestors_ts</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recombination_rate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mismatch_ratio</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_compression</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">post_process</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indexes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_sample_times</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/inference.html#match_samples"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.match_samples" title="Link to this definition">#</a></dt>
<dd><p>Runs the sample matching <a class="reference internal" href="inference.html#sec-inference-match-samples"><span class="std std-ref">algorithm</span></a>
on the specified <a class="reference internal" href="#tsinfer.VariantData" title="tsinfer.VariantData"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantData</span></code></a> instance and ancestors tree sequence,
returning the final <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a> instance containing
the full inferred history for all samples and sites. See
<a class="reference internal" href="inference.html#sec-inference-match-ancestors-and-samples"><span class="std std-ref">matching ancestors &amp; samples</span></a>
in the documentation for details of <code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code>, <code class="docutils literal notranslate"><span class="pre">mismatch_ratio</span></code>
and <code class="docutils literal notranslate"><span class="pre">path_compression</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variant_data</strong> (<a class="reference internal" href="#tsinfer.VariantData" title="tsinfer.VariantData"><em>VariantData</em></a>) – The <a class="reference internal" href="#tsinfer.VariantData" title="tsinfer.VariantData"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantData</span></code></a> instance
representing the input data.</p></li>
<li><p><strong>ancestors_ts</strong> (<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)"><em>tskit.TreeSequence</em></a>) – The
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a> instance representing the inferred
history among ancestral ancestral haplotypes.</p></li>
<li><p><strong>recombination_rate</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a><em>, </em><a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.RateMap" title="(in Project name not set)"><em>msprime.RateMap</em></a>) – Either a floating point value giving a constant rate
<span class="math notranslate nohighlight">\(\rho\)</span> per unit length of genome, or an <a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.RateMap" title="(in Project name not set)"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.RateMap</span></code></a>
object. This is used to calculate the probability of recombination between
adjacent sites. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, all matching conflicts are resolved by
recombination and all inference sites will have a single mutation
(equivalent to mismatch_ratio near zero)</p></li>
<li><p><strong>mismatch_ratio</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a>) – The probability of a mismatch relative to the median
probability of recombination between adjacent sites: can only be used if a
recombination rate has been set (default: <code class="docutils literal notranslate"><span class="pre">None</span></code> treated as 1 if
<code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code> is set).</p></li>
<li><p><strong>path_compression</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – Whether to merge edges that share identical
paths (essentially taking advantage of shared recombination breakpoints).</p></li>
<li><p><strong>indexes</strong> (<em>array_like</em>) – An array of indexes into the variant_data file of
the samples to match (in increasing order) or None for all samples.</p></li>
<li><p><strong>post_process</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – Whether to run the <a class="reference internal" href="#tsinfer.post_process" title="tsinfer.post_process"><code class="xref py py-func docutils literal notranslate"><span class="pre">post_process()</span></code></a> method on the
the tree sequence which, among other things, removes ancestral material that
does not end up in the current samples (if not specified, defaults to <code class="docutils literal notranslate"><span class="pre">True</span></code>)</p></li>
<li><p><strong>force_sample_times</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – After matching, should an attempt be made to
adjust the time of “historical samples” (those associated with an individual
having a non-zero time) such that the sample nodes in the tree sequence
appear at the time of the individual with which they are associated.</p></li>
<li><p><strong>num_threads</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – The number of match worker threads to use. If
this is &lt;= 0 then a simpler sequential algorithm is used (default).</p></li>
<li><p><strong>simplify</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – Treated as an alias for <code class="docutils literal notranslate"><span class="pre">post_process</span></code>, deprecated but
currently retained for backwards compatibility if set to <code class="docutils literal notranslate"><span class="pre">False</span></code>.</p></li>
<li><p><strong>overlay_non_inference_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If True, sites that were included in
the selected sites, but were not used for inference, will be added to the
tree sequence by mapping their mutations over the inferred topology.
Defaults to True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The tree sequence representing the inferred history
of the sample.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)">tskit.TreeSequence</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tsinfer.augment_ancestors">
<span class="sig-prename descclassname"><span class="pre">tsinfer.</span></span><span class="sig-name descname"><span class="pre">augment_ancestors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ancestors_ts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indexes</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recombination_rate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mismatch_ratio</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_compression</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/inference.html#augment_ancestors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.augment_ancestors" title="Link to this definition">#</a></dt>
<dd><p>Runs the sample matching <a class="reference internal" href="inference.html#sec-inference-match-samples"><span class="std std-ref">algorithm</span></a>
on the specified <a class="reference internal" href="#tsinfer.VariantData" title="tsinfer.VariantData"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantData</span></code></a> instance and ancestors tree sequence,
for the specified subset of sample indexes, returning the
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a> instance including these samples. This
tree sequence can then be used as an ancestors tree sequence for subsequent
matching against all samples.  See
<a class="reference internal" href="inference.html#sec-inference-match-ancestors-and-samples"><span class="std std-ref">matching ancestors &amp; samples</span></a>
in the documentation for details of <code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code>, <code class="docutils literal notranslate"><span class="pre">mismatch_ratio</span></code>
and <code class="docutils literal notranslate"><span class="pre">path_compression</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variant_data</strong> (<a class="reference internal" href="#tsinfer.VariantData" title="tsinfer.VariantData"><em>VariantData</em></a>) – The <a class="reference internal" href="#tsinfer.VariantData" title="tsinfer.VariantData"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantData</span></code></a> instance
representing the input data.</p></li>
<li><p><strong>ancestors_ts</strong> (<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)"><em>tskit.TreeSequence</em></a>) – The
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a> instance representing the inferred
history among ancestral ancestral haplotypes.</p></li>
<li><p><strong>indexes</strong> (<em>array</em>) – The sample indexes to insert into the ancestors
tree sequence, in increasing order.</p></li>
<li><p><strong>recombination_rate</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a><em>, </em><a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.RateMap" title="(in Project name not set)"><em>msprime.RateMap</em></a>) – Either a floating point value giving a constant rate
<span class="math notranslate nohighlight">\(\rho\)</span> per unit length of genome, or an <a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.RateMap" title="(in Project name not set)"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.RateMap</span></code></a>
object. This is used to calculate the probability of recombination between
adjacent sites. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, all matching conflicts are resolved by
recombination and all inference sites will have a single mutation
(equivalent to mismatch_ratio near zero)</p></li>
<li><p><strong>mismatch_ratio</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a>) – The probability of a mismatch relative to the median
probability of recombination between adjacent sites: can only be used if a
recombination rate has been set (default: <code class="docutils literal notranslate"><span class="pre">None</span></code> treated as 1 if
<code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code> is set).</p></li>
<li><p><strong>path_compression</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – Whether to merge edges that share identical
paths (essentially taking advantage of shared recombination breakpoints).</p></li>
<li><p><strong>num_threads</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – The number of match worker threads to use. If
this is &lt;= 0 then a simpler sequential algorithm is used (default).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The specified ancestors tree sequence augmented with copying
paths for the specified sample.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)">tskit.TreeSequence</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tsinfer.post_process">
<span class="sig-prename descclassname"><span class="pre">tsinfer.</span></span><span class="sig-name descname"><span class="pre">post_process</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ts</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">split_ultimate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">erase_flanks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/inference.html#post_process"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.post_process" title="Link to this definition">#</a></dt>
<dd><p>Post-process a tsinferred tree sequence into a more conventional form. This is
the function run by default on the final tree sequence output by
<a class="reference internal" href="#tsinfer.match_samples" title="tsinfer.match_samples"><code class="xref py py-func docutils literal notranslate"><span class="pre">match_samples()</span></code></a>. It involves the following 4 steps:</p>
<ol class="arabic simple">
<li><p>If the oldest node is connected to a single child via an edge that spans the
entire tree sequence, this oldest node is removed, so that its child becomes
the new root (this step is undertaken to remove the “virtual-root-like node”
which is added to ancestor tree sequences to enable matching).</p></li>
<li><p>If the oldest node is removed in the first step and the new root spans the
entire genome, it is treated as the “ultimate ancestor” and (unless
<code class="docutils literal notranslate"><span class="pre">split_ultimate</span></code> is <code class="docutils literal notranslate"><span class="pre">False</span></code>) the node is split into multiple coexisiting
nodes with the splits occurring whenever the children of the ultimate ancestor
change. The rationale is that tsinfer creates a single ancestral haplotype with
all inference sites in the ancestral state: this is, however, unlikely to
represent a single ancestor in the past. If the tree sequence is then dated,
the fact that ultimate ancestor is split into separate nodes allows these nodes
to be dated to different times.</p></li>
<li><p>Often, extensive regions of genome exist before the first defined site and after
the last defined site. Since no data exists in these sections of the genome, post
processing by default erases the inferred topology in these regions. However,
if <code class="docutils literal notranslate"><span class="pre">erase_flanks</span></code> is False, the flanking regions at the start and end will be
assigned the same topology as inferred at the first and last site respectively.</p></li>
<li><p>The sample nodes are reordered such that they are the first nodes listed in the
node table,  removing tree nodes and edges that are not on a path between the
root and any of the samples (by applying the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.simplify" title="(in Project name not set)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">simplify()</span></code></a>
method with <code class="docutils literal notranslate"><span class="pre">keep_unary</span></code> set to True but <code class="docutils literal notranslate"><span class="pre">filter_sites</span></code>,
<code class="docutils literal notranslate"><span class="pre">filter_populations</span></code> and <code class="docutils literal notranslate"><span class="pre">filter_individuals</span></code> set to False).</p></li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>split_ultimate</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> (default) and the oldest node is the only
parent to a single “ultimate ancestor” node, attempt to split this node into
many separate nodes (see above). If <code class="docutils literal notranslate"><span class="pre">False</span></code> do not attempt to identify or
split an ultimate ancestor node.</p></li>
<li><p><strong>erase_flanks</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> (default), keep only the
inferred topology between the first and last sites. If <code class="docutils literal notranslate"><span class="pre">False</span></code>,
output regions of topology inferred before the first site and after
the last site.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The post-processed tree sequence.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)">tskit.TreeSequence</a></p>
</dd>
</dl>
</dd></dl>

</section>
<section id="batched-inference">
<h3>Batched inference<a class="headerlink" href="#batched-inference" title="Link to this heading">#</a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="tsinfer.match_ancestors_batch_init">
<span class="sig-prename descclassname"><span class="pre">tsinfer.</span></span><span class="sig-name descname"><span class="pre">match_ancestors_batch_init</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">work_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variant_data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ancestral_state</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ancestor_data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_work_per_job</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_num_partitions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_mask</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">site_mask</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recombination_rate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mismatch_ratio</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_compression</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recombination</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mismatch</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">precision</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">engine</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'C'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extended_checks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_units</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/inference.html#match_ancestors_batch_init"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.match_ancestors_batch_init" title="Link to this definition">#</a></dt>
<dd><p>match_ancestors_batch_init(work_dir, variant_data_path, ancestral_state,
ancestor_data_path, min_work_per_job, *, max_num_partitions=None,
sample_mask=None, site_mask=None, recombination_rate=None, mismatch_ratio=None,
path_compression=True)</p>
<p>Initialise a batched ancestor matching job. This function is used to
prepare a working directory for running a batched ancestor matching job. The
job is split into groups of ancestors, with each group further split into
partitions of ancestors if necessary. <cite>work_dir</cite> is created and details
are written to <cite>metadata.json</cite> in <cite>work_dir</cite>. The job can then be run
using <a class="reference internal" href="#tsinfer.match_ancestors_batch_groups" title="tsinfer.match_ancestors_batch_groups"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_groups()</span></code></a> and
<a class="reference internal" href="#tsinfer.match_ancestors_batch_group_partition" title="tsinfer.match_ancestors_batch_group_partition"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_group_partition()</span></code></a> then finally
<a class="reference internal" href="#tsinfer.match_ancestors_batch_group_finalise" title="tsinfer.match_ancestors_batch_group_finalise"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_group_finalise()</span></code></a>. See
<a class="reference internal" href="large_scale.html#sec-large-scale"><span class="std std-ref">large scale inference</span></a> for more details about how these
methods work together. See <a class="reference internal" href="#tsinfer.match_ancestors" title="tsinfer.match_ancestors"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors()</span></code></a> for details on
ancestor matching.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>work_dir</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – The directory in which to store the working files.</p></li>
<li><p><strong>variant_data_path</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – <p>The input dataset in
<a class="reference external" href="https://github.com/sgkit-dev/vcf-zarr-spec">VCF Zarr</a> format.
Path to the Zarr dataset saved on disk. See <a class="reference internal" href="#tsinfer.VariantData" title="tsinfer.VariantData"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantData</span></code></a>.</p>
</p></li>
<li><p><strong>ancestral_state</strong> (<em>Union</em><em>(</em><em>array</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>)</em>) – A numpy array of strings specifying
the ancestral states (alleles) used in inference. This must be the same length
as the number of unmasked sites in the dataset. Alternatively, a single string
can be provided, giving the name of an array in the input dataset which contains
the ancestral states. Unknown ancestral states can be specified using “N”.
Any ancestral states which do not match any of the known alleles at that site,
will be tallied, and a warning issued summarizing the unknown ancestral states.</p></li>
<li><p><strong>ancestor_data_path</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – The path to the file containing the ancestors
generated by <a class="reference internal" href="#tsinfer.generate_ancestors" title="tsinfer.generate_ancestors"><code class="xref py py-meth docutils literal notranslate"><span class="pre">generate_ancestors()</span></code></a>.</p></li>
<li><p><strong>min_work_per_job</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – The minimum amount of work (as a count of genotypes) to
allocate to a single parallel job. If the amount of work in a group of ancestors
exceeds this level it will be broken up into parallel partitions, subject to
the constraint of <cite>max_num_partitions</cite>.</p></li>
<li><p><strong>max_num_partitions</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – The maximum number of partitions to split a group of
ancestors into. Useful for limiting the number of jobs in a workflow to
avoid job overhead. Defaults to 1000.</p></li>
<li><p><strong>sample_mask</strong> (<em>Union</em><em>(</em><em>array</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>)</em>) – A numpy array of booleans specifying which
samples to mask out (exclude) from the dataset. Alternatively, a string
can be provided, giving the name of an array in the input dataset which contains
the sample mask. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), all samples are included.</p></li>
<li><p><strong>site_mask</strong> (<em>Union</em><em>(</em><em>array</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>)</em>) – A numpy array of booleans specifying which
sites to mask out (exclude) from the dataset. Alternatively, a string
can be provided, giving the name of an array in the input dataset which contains
the site mask. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), all sites are included.</p></li>
<li><p><strong>sequence_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – An integer specifying the resulting <cite>sequence_length</cite>
attribute of the output tree sequence. If not specified the <cite>contig_length</cite>
attribute from the undelying zarr store for the contig of the selected variants.
is used. If that is not present then the maximum position plus one of the used
variants is used.</p></li>
<li><p><strong>recombination_rate</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a><em>, </em><a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.RateMap" title="(in Project name not set)"><em>msprime.RateMap</em></a>) – Either a floating point value giving a constant rate
<span class="math notranslate nohighlight">\(\rho\)</span> per unit length of genome, or an <a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.RateMap" title="(in Project name not set)"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.RateMap</span></code></a>
object. This is used to calculate the probability of recombination between
adjacent sites. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, all matching conflicts are resolved by
recombination and all inference sites will have a single mutation
(equivalent to mismatch_ratio near zero)</p></li>
<li><p><strong>mismatch_ratio</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a>) – The probability of a mismatch relative to the median
probability of recombination between adjacent sites: can only be used if a
recombination rate has been set (default: <code class="docutils literal notranslate"><span class="pre">None</span></code> treated as 1 if
<code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code> is set).</p></li>
<li><p><strong>path_compression</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – Whether to merge edges that share identical
paths (essentially taking advantage of shared recombination breakpoints).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary of the job metadata, as written to <cite>metadata.json</cite>
in <cite>work_dir</cite>. <cite>ancestor_grouping</cite> in this dict contains the grouping
of ancestors into groups and should be used to guide calling
<a class="reference internal" href="#tsinfer.match_ancestors_batch_groups" title="tsinfer.match_ancestors_batch_groups"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_groups()</span></code></a> and
<a class="reference internal" href="#tsinfer.match_ancestors_batch_group_partition" title="tsinfer.match_ancestors_batch_group_partition"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_group_partition()</span></code></a>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tsinfer.match_ancestors_batch_groups">
<span class="sig-prename descclassname"><span class="pre">tsinfer.</span></span><span class="sig-name descname"><span class="pre">match_ancestors_batch_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">work_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_index_start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_index_end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/inference.html#match_ancestors_batch_groups"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.match_ancestors_batch_groups" title="Link to this definition">#</a></dt>
<dd><p>match_ancestors_batch_groups(work_dir, group_index_start,
group_index_end, num_threads=0)</p>
<p>Match a set of ancestor groups from <cite>group_index_start`(inclusive) to
`group_index_end`(exclusive) in a batched ancestor matching job. See
:ref:`large scale inference&lt;sec_large_scale&gt;</cite> for more details.</p>
<p>A tree sequence file for <cite>group_index_start - 1</cite> must exist in <cite>work_dir</cite>, unless
<cite>group_index_start</cite> is 0. After matching the tree sequence for <cite>group_index_end - 1</cite>
is written to <cite>work_dir</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>work_dir</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – The working directory for the batch job, as written by
<a class="reference internal" href="#tsinfer.match_ancestors_batch_init" title="tsinfer.match_ancestors_batch_init"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_init()</span></code></a>.</p></li>
<li><p><strong>group_index_start</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – The first group index to match.</p></li>
<li><p><strong>group_index_end</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – The group index to stop matching at.</p></li>
<li><p><strong>num_threads</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – The number of worker threads to use. If this is &lt;= 1 then
match sequentially.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The tree sequence representing the inferred ancestors for the last group
matched</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)">tskit.TreeSequence</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tsinfer.match_ancestors_batch_group_partition">
<span class="sig-prename descclassname"><span class="pre">tsinfer.</span></span><span class="sig-name descname"><span class="pre">match_ancestors_batch_group_partition</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">work_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">partition_index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/inference.html#match_ancestors_batch_group_partition"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.match_ancestors_batch_group_partition" title="Link to this definition">#</a></dt>
<dd><p>Match a single partition of ancestors from a group in a batched ancestor matching
job. See <a class="reference internal" href="large_scale.html#sec-large-scale"><span class="std std-ref">large scale inference</span></a> for more details. The
tree sequence for the group before must exist in <cite>work_dir</cite>. After matching the
results for the partition are written to <cite>work_dir</cite>. Once all partitions for a
group have been matched, the group can be finalised using
<a class="reference internal" href="#tsinfer.match_ancestors_batch_group_finalise" title="tsinfer.match_ancestors_batch_group_finalise"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_group_finalise()</span></code></a>. The number of partitions in a
group is recorded in <cite>metadata.json</cite> in the work dir under the
<cite>ancestor_grouping</cite> key. This method uses a single thread.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>work_dir</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – The working directory for the batch job, as written by
<a class="reference internal" href="#tsinfer.match_ancestors_batch_init" title="tsinfer.match_ancestors_batch_init"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_init()</span></code></a>.</p></li>
<li><p><strong>group_index</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – The group index that contains the partition to match.</p></li>
<li><p><strong>partition_index</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – The partition index to match. Must be less than the
number of partitions in the batch job metadata for this group.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tsinfer.match_ancestors_batch_group_finalise">
<span class="sig-prename descclassname"><span class="pre">tsinfer.</span></span><span class="sig-name descname"><span class="pre">match_ancestors_batch_group_finalise</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">work_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/inference.html#match_ancestors_batch_group_finalise"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.match_ancestors_batch_group_finalise" title="Link to this definition">#</a></dt>
<dd><p>Finalise a group of partitioned ancestors in a batched ancestor matching job. See
<a class="reference internal" href="large_scale.html#sec-large-scale"><span class="std std-ref">large scale inference</span></a> for more details. The tree sequence
for the group before must exist in <cite>work_dir</cite>, along with the results for all
partitions in this group. Writes the tree sequence for the group to <cite>work_dir</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>work_dir</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – The working directory for the batch job, as written by
<a class="reference internal" href="#tsinfer.match_ancestors_batch_init" title="tsinfer.match_ancestors_batch_init"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_init()</span></code></a>.</p></li>
<li><p><strong>group_index</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – The group index to finalise.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The tree sequence representing the inferred ancestors for the group</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)">tskit.TreeSequence</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tsinfer.match_ancestors_batch_finalise">
<span class="sig-prename descclassname"><span class="pre">tsinfer.</span></span><span class="sig-name descname"><span class="pre">match_ancestors_batch_finalise</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">work_dir</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/inference.html#match_ancestors_batch_finalise"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.match_ancestors_batch_finalise" title="Link to this definition">#</a></dt>
<dd><p>Finalise a batched ancestor matching job. This method should be called after all
groups have been matched, either by <a class="reference internal" href="#tsinfer.match_ancestors_batch_groups" title="tsinfer.match_ancestors_batch_groups"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_groups()</span></code></a> or
<a class="reference internal" href="#tsinfer.match_ancestors_batch_group_finalise" title="tsinfer.match_ancestors_batch_group_finalise"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_group_finalise()</span></code></a>. Returns the final ancestors
tree sequence for the batch job. <cite>work_dir</cite> is retained and not deleted.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>work_dir</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – The working directory for the batch job, as written by
<a class="reference internal" href="#tsinfer.match_ancestors_batch_init" title="tsinfer.match_ancestors_batch_init"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_init()</span></code></a>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The tree sequence representing the inferred ancestors for the batch job</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)">tskit.TreeSequence</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tsinfer.match_samples_batch_init">
<span class="sig-prename descclassname"><span class="pre">tsinfer.</span></span><span class="sig-name descname"><span class="pre">match_samples_batch_init</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">work_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variant_data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ancestral_state</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ancestor_ts_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_work_per_job</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_mask</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">site_mask</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sites_time</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individuals_time</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individuals_location</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individuals_population</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individuals_flags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recombination_rate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mismatch_ratio</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_compression</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indexes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">post_process</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_sample_times</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlay_non_inference_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recombination</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mismatch</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">precision</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extended_checks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">engine</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'C'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/inference.html#match_samples_batch_init"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.match_samples_batch_init" title="Link to this definition">#</a></dt>
<dd><p>match_samples_batch_init(work_dir, variant_data_path, ancestral_state,
ancestor_ts_path, min_work_per_job, *,
sample_mask=None, site_mask=None, recombination_rate=None, mismatch_ratio=None,
path_compression=True, indexes=None, post_process=None, force_sample_times=False)</p>
<p>Initialise a batched sample matching job. Creates <cite>work_dir</cite> and writes job
details to <cite>metadata.json</cite>. The job can then be run using parallel calls to
<a class="reference internal" href="#tsinfer.match_samples_batch_partition" title="tsinfer.match_samples_batch_partition"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_samples_batch_partition()</span></code></a> and once those are complete
finally <a class="reference internal" href="#tsinfer.match_samples_batch_finalise" title="tsinfer.match_samples_batch_finalise"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_samples_batch_finalise()</span></code></a>.</p>
<p>The <cite>num_partitions</cite> key in the metadata dict contains the number of partitions
that need to be processed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>work_dir</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – The directory in which to store the working files.</p></li>
<li><p><strong>variant_data_path</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – <p>The input dataset in
<a class="reference external" href="https://github.com/sgkit-dev/vcf-zarr-spec">VCF Zarr</a> format.
Path to the Zarr dataset saved on disk. See <a class="reference internal" href="#tsinfer.VariantData" title="tsinfer.VariantData"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantData</span></code></a>.</p>
</p></li>
<li><p><strong>ancestral_state</strong> (<em>Union</em><em>(</em><em>array</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>)</em>) – A numpy array of strings specifying
the ancestral states (alleles) used in inference. This must be the same
length as the number of unmasked sites in the dataset. Alternatively, a
single string can be provided, giving the name of an array in the input
dataset which contains the ancestral states. Unknown ancestral states can
be specified using “N”. Any ancestral states which do not match any of the
known alleles at that site, will be tallied, and a warning issued
summarizing the unknown ancestral states.</p></li>
<li><p><strong>ancestor_ts_path</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – The path to the tree sequence file containing the
ancestors generated by <a class="reference internal" href="#tsinfer.match_ancestors_batch_finalise" title="tsinfer.match_ancestors_batch_finalise"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors_batch_finalise()</span></code></a>, or
<a class="reference internal" href="#tsinfer.match_ancestors" title="tsinfer.match_ancestors"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_ancestors()</span></code></a>.</p></li>
<li><p><strong>min_work_per_job</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – The minimum amount of work (as a count of
genotypes) to allocate to a single parallel job. If the amount of work in
a group of samples exceeds this level it will be broken up into parallel
partitions, subject to the constraint of <cite>max_num_partitions</cite>.</p></li>
<li><p><strong>sample_mask</strong> (<em>Union</em><em>(</em><em>array</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>)</em>) – A numpy array of booleans specifying
which samples to mask out (exclude) from the dataset. Alternatively, a
string can be provided, giving the name of an array in the input dataset
which contains the sample mask. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), all samples are
included.</p></li>
<li><p><strong>site_mask</strong> (<em>Union</em><em>(</em><em>array</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>)</em>) – A numpy array of booleans specifying which
sites to mask out (exclude) from the dataset. Alternatively, a string can
be provided, giving the name of an array in the input dataset which
contains the site mask. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), all sites are included.</p></li>
<li><p><strong>sites_time</strong> (<em>Union</em><em>(</em><em>array</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>)</em>) – A numpy array of floats specifying the relative
time of occurrence of the mutation to the derived state at each site. This must
be of the same length as the number of unmasked sites. Alternatively, a
string can be provided, giving the name of an array in the input dataset
which contains the site times. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), the frequency of the
derived allele is used as a proxy for the time of occurrence: this is usually a
reasonable approximation to the relative order of ancestors used for inference.
Time values are ignored for sites not used in inference, such as singletons,
sites with more than two alleles, or sites with an unknown ancestral state.</p></li>
<li><p><strong>individuals_time</strong> (<em>Union</em><em>(</em><em>array</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>)</em>) – A numpy array of floats specifying
the time of each individual in the dataset. This must be the same length
as the number of unmasked individuals. Alternatively, a string can be provided,
giving the name of an array in the input dataset which contains the individual
times. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), individuals are assumed to have
<code class="docutils literal notranslate"><span class="pre">tskit.UNKNOWN_TIME</span></code>.</p></li>
<li><p><strong>individuals_location</strong> (<em>Union</em><em>(</em><em>array</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>)</em>) – A numpy array specifying
the location of each individual in the dataset. This must be the same length
as the number of unmasked individuals. Alternatively, a string can be provided,
giving the name of an array in the input dataset which contains the individual
locations. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), individuals are assumed to have empty
location arrays.</p></li>
<li><p><strong>individuals_population</strong> (<em>Union</em><em>(</em><em>array</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>)</em>) – A numpy array of integers specifying
the population of each individual in the dataset. This must be the same length
as the number of unmasked individuals. Alternatively, a string can be provided,
giving the name of an array in the input dataset which contains the individual
populations. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), individuals are assumed to have
<code class="docutils literal notranslate"><span class="pre">tskit.NULL</span></code> as their population.</p></li>
<li><p><strong>individuals_flags</strong> (<em>Union</em><em>(</em><em>array</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>)</em>) – A numpy array of integers specifying
the flags of each individual in the dataset. This must be the same length
as the number of unmasked individuals. Alternatively, a string can be provided,
giving the name of an array in the input dataset which contains the individual
flags. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), individuals are assumed to have flags set to 0.</p></li>
<li><p><strong>sequence_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – An integer specifying the resulting <cite>sequence_length</cite>
attribute of the output tree sequence. If not specified the <cite>contig_length</cite>
attribute from the undelying zarr store for the contig of the selected variants.
is used. If that is not present then the maximum position plus one of the used
variants is used.</p></li>
<li><p><strong>recombination_rate</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a><em>, </em><a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.RateMap" title="(in Project name not set)"><em>msprime.RateMap</em></a>) – Either a floating point value giving a constant
rate <span class="math notranslate nohighlight">\(\rho\)</span> per unit length of genome, or an
<a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.RateMap" title="(in Project name not set)"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.RateMap</span></code></a> object. This is used to calculate the
probability of recombination between adjacent sites. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, all
matching conflicts are resolved by recombination and all inference sites
will have a single mutation (equivalent to mismatch_ratio near zero)</p></li>
<li><p><strong>mismatch_ratio</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a>) – The probability of a mismatch relative to the
median probability of recombination between adjacent sites: can only be
used if a recombination rate has been set (default: <code class="docutils literal notranslate"><span class="pre">None</span></code> treated as 1
if <code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code> is set).</p></li>
<li><p><strong>path_compression</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – Whether to merge edges that share identical paths
(essentially taking advantage of shared recombination breakpoints).</p></li>
<li><p><strong>indexes</strong> (<em>arraylike</em>) – The sample indexes to match. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), all
samples are matched.</p></li>
<li><p><strong>post_process</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – Whether to run the <a class="reference internal" href="#tsinfer.post_process" title="tsinfer.post_process"><code class="xref py py-func docutils literal notranslate"><span class="pre">post_process()</span></code></a> method on
the the tree sequence which, among other things, removes ancestral
material that does not end up in the current samples (if not specified,
defaults to <code class="docutils literal notranslate"><span class="pre">True</span></code>)</p></li>
<li><p><strong>force_sample_times</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – After matching, should an attempt be made to
adjust the time of “historical samples” (those associated with an
individual having a non-zero time) such that the sample nodes in the tree
sequence appear at the time of the individual with which they are
associated.</p></li>
<li><p><strong>overlay_non_inference_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a>) – If True, sites that were included in
the selected sites, but were not used for inference, will be added to the
tree sequence by mapping their mutations over the inferred topology.
Defaults to True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary of the job metadata, as written to <cite>metadata.json</cite> in
<cite>work_dir</cite>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tsinfer.match_samples_batch_partition">
<span class="sig-prename descclassname"><span class="pre">tsinfer.</span></span><span class="sig-name descname"><span class="pre">match_samples_batch_partition</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">work_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">partition_index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/inference.html#match_samples_batch_partition"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.match_samples_batch_partition" title="Link to this definition">#</a></dt>
<dd><p>Match a single partition of samples in a batched sample matching job. See
<a class="reference internal" href="large_scale.html#sec-large-scale"><span class="std std-ref">large scale inference</span></a> for more details. Match data
for the partition is written to <cite>work_dir</cite>. Uses a single thread to perform
matching.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>work_dir</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – The working directory for the batch job, as written by
<a class="reference internal" href="#tsinfer.match_samples_batch_init" title="tsinfer.match_samples_batch_init"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_samples_batch_init()</span></code></a>.</p></li>
<li><p><strong>partition_index</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a>) – The partition index to match. Must be less than
the number of partitions in the batch job metadata key <cite>num_partitions</cite>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tsinfer.match_samples_batch_finalise">
<span class="sig-prename descclassname"><span class="pre">tsinfer.</span></span><span class="sig-name descname"><span class="pre">match_samples_batch_finalise</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">work_dir</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/inference.html#match_samples_batch_finalise"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.match_samples_batch_finalise" title="Link to this definition">#</a></dt>
<dd><p>Finalise a batched sample matching job. This method should be called after all
partitions have been matched by <a class="reference internal" href="#tsinfer.match_samples_batch_partition" title="tsinfer.match_samples_batch_partition"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_samples_batch_partition()</span></code></a>. Returns
the final tree sequence for the batch job. <cite>work_dir</cite> is retained and not deleted.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>work_dir</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – The working directory for the batch job, as written by
<a class="reference internal" href="#tsinfer.match_samples_batch_init" title="tsinfer.match_samples_batch_init"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match_samples_batch_init()</span></code></a>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The tree sequence representing the inferred history of the samples.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)">tskit.TreeSequence</a></p>
</dd>
</dl>
</dd></dl>

</section>
<section id="container-classes">
<h3>Container classes<a class="headerlink" href="#container-classes" title="Link to this heading">#</a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="tsinfer.Variant">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tsinfer.</span></span><span class="sig-name descname"><span class="pre">Variant</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">site</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genotypes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alleles</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/formats.html#Variant"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.Variant" title="Link to this definition">#</a></dt>
<dd><p>A single variant. Mirrors the definition in tskit.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="tsinfer.Site">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tsinfer.</span></span><span class="sig-name descname"><span class="pre">Site</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ancestral_allele</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alleles</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsinfer/formats.html#Site"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.Site" title="Link to this definition">#</a></dt>
<dd><p>A single site. Mirrors the definition in tskit with some additional fields.</p>
</dd></dl>

</section>
<section id="exceptions">
<h3>Exceptions<a class="headerlink" href="#exceptions" title="Link to this heading">#</a></h3>
<dl class="py exception">
<dt class="sig sig-object py" id="tsinfer.FileFormatError">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tsinfer.</span></span><span class="sig-name descname"><span class="pre">FileFormatError</span></span><a class="reference internal" href="_modules/tsinfer/exceptions.html#FileFormatError"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsinfer.FileFormatError" title="Link to this definition">#</a></dt>
<dd><p>Exception raised when a malformed file is encountered.</p>
</dd></dl>

</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="large_scale.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Large Scale Inference</p>
      </div>
    </a>
    <a class="right-next"
       href="cli.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Command line interface</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#variant-data">Variant data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.VariantData"><code class="docutils literal notranslate"><span class="pre">VariantData</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.add_ancestral_state_array"><code class="docutils literal notranslate"><span class="pre">add_ancestral_state_array()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#ancestor-data">Ancestor data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.load"><code class="docutils literal notranslate"><span class="pre">load()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData"><code class="docutils literal notranslate"><span class="pre">AncestorData</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.data_equal"><code class="docutils literal notranslate"><span class="pre">AncestorData.data_equal()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.sequence_length"><code class="docutils literal notranslate"><span class="pre">AncestorData.sequence_length</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.num_sites"><code class="docutils literal notranslate"><span class="pre">AncestorData.num_sites</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.sites_position"><code class="docutils literal notranslate"><span class="pre">AncestorData.sites_position</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.add_provenance"><code class="docutils literal notranslate"><span class="pre">AncestorData.add_provenance()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.arrays"><code class="docutils literal notranslate"><span class="pre">AncestorData.arrays()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.clear_provenances"><code class="docutils literal notranslate"><span class="pre">AncestorData.clear_provenances()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.close"><code class="docutils literal notranslate"><span class="pre">AncestorData.close()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.copy"><code class="docutils literal notranslate"><span class="pre">AncestorData.copy()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.file_size"><code class="docutils literal notranslate"><span class="pre">AncestorData.file_size</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.info"><code class="docutils literal notranslate"><span class="pre">AncestorData.info</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.provenances"><code class="docutils literal notranslate"><span class="pre">AncestorData.provenances()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.record_provenance"><code class="docutils literal notranslate"><span class="pre">AncestorData.record_provenance()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.ancestors_length"><code class="docutils literal notranslate"><span class="pre">AncestorData.ancestors_length</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.insert_proxy_samples"><code class="docutils literal notranslate"><span class="pre">AncestorData.insert_proxy_samples()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.truncate_ancestors"><code class="docutils literal notranslate"><span class="pre">AncestorData.truncate_ancestors()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.add_ancestor"><code class="docutils literal notranslate"><span class="pre">AncestorData.add_ancestor()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.finalise"><code class="docutils literal notranslate"><span class="pre">AncestorData.finalise()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.ancestor"><code class="docutils literal notranslate"><span class="pre">AncestorData.ancestor()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.AncestorData.ancestors"><code class="docutils literal notranslate"><span class="pre">AncestorData.ancestors()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#running-inference">Running inference</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.infer"><code class="docutils literal notranslate"><span class="pre">infer()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.generate_ancestors"><code class="docutils literal notranslate"><span class="pre">generate_ancestors()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.GenotypeEncoding"><code class="docutils literal notranslate"><span class="pre">GenotypeEncoding</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.GenotypeEncoding.EIGHT_BIT"><code class="docutils literal notranslate"><span class="pre">GenotypeEncoding.EIGHT_BIT</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.GenotypeEncoding.ONE_BIT"><code class="docutils literal notranslate"><span class="pre">GenotypeEncoding.ONE_BIT</span></code></a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.match_ancestors"><code class="docutils literal notranslate"><span class="pre">match_ancestors()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.match_samples"><code class="docutils literal notranslate"><span class="pre">match_samples()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.augment_ancestors"><code class="docutils literal notranslate"><span class="pre">augment_ancestors()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.post_process"><code class="docutils literal notranslate"><span class="pre">post_process()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#batched-inference">Batched inference</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.match_ancestors_batch_init"><code class="docutils literal notranslate"><span class="pre">match_ancestors_batch_init()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.match_ancestors_batch_groups"><code class="docutils literal notranslate"><span class="pre">match_ancestors_batch_groups()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.match_ancestors_batch_group_partition"><code class="docutils literal notranslate"><span class="pre">match_ancestors_batch_group_partition()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.match_ancestors_batch_group_finalise"><code class="docutils literal notranslate"><span class="pre">match_ancestors_batch_group_finalise()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.match_ancestors_batch_finalise"><code class="docutils literal notranslate"><span class="pre">match_ancestors_batch_finalise()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.match_samples_batch_init"><code class="docutils literal notranslate"><span class="pre">match_samples_batch_init()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.match_samples_batch_partition"><code class="docutils literal notranslate"><span class="pre">match_samples_batch_partition()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.match_samples_batch_finalise"><code class="docutils literal notranslate"><span class="pre">match_samples_batch_finalise()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#container-classes">Container classes</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.Variant"><code class="docutils literal notranslate"><span class="pre">Variant</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.Site"><code class="docutils literal notranslate"><span class="pre">Site</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exceptions">Exceptions</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsinfer.FileFormatError"><code class="docutils literal notranslate"><span class="pre">FileFormatError</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Tskit Developers
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2018.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>