
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Inference overview &#8212; Tsinfer manual</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=5349f25f" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'inference';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Large Scale Inference" href="large_scale.html" />
    <link rel="prev" title="Usage" href="usage.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <script data-goatcounter="https://tskit.goatcounter.com/count" async src="//gc.zgo.at/count.js"></script>
</head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/tsinfer_logo.svg" class="logo__image only-light" alt="Tsinfer manual - Home"/>
    <script>document.write(`<img src="_static/tsinfer_logo.svg" class="logo__image only-dark" alt="Tsinfer manual - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Installation</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Usage</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Inference</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Inference overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="large_scale.html">Large Scale Inference</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Interfaces</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="api.html">API Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">Command line interface</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">File Formats</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="file_formats.html">File formats</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Miscellaneous</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="development.html">Developer documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="CITATION.html">Citing tsinfer</a></li>
<li class="toctree-l1"><a class="reference internal" href="CHANGELOG.html">Changelog</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/inference.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Inference overview</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-requirements">Data requirements</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-model">Data model</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#individual">Individual</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sample">Sample</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#population">Population</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#metadata">Metadata</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#import-samples-data">Import samples data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-ancestors">Generate ancestors</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#matching-ancestors-samples">Matching ancestors &amp; samples</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#recombination-and-mismatch">Recombination and mismatch</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#path-compression">Path compression</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#match-ancestors">Match ancestors</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#match-samples">Match samples</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="inference-overview">
<span id="sec-inference"></span><h1>Inference overview<a class="headerlink" href="#inference-overview" title="Link to this heading">#</a></h1>
<p>The process of inferring a tree sequence from variation data is split into a
number of different steps. We first review the
<a class="reference internal" href="#sec-inference-data-requirements"><span class="std std-ref">requirements</span></a> for input data
and discuss how such data is <a class="reference internal" href="#sec-inference-import-samples"><span class="std std-ref">imported</span></a>
into <code class="docutils literal notranslate"><span class="pre">tsinfer</span></code>’s sample data format. We then outline the three
basic steps for inference:
<a class="reference internal" href="#sec-inference-generate-ancestors"><span class="std std-ref">generating ancestors</span></a>,
<a class="reference internal" href="#sec-inference-match-ancestors"><span class="std std-ref">matching ancestors</span></a> and
<a class="reference internal" href="#sec-inference-match-samples"><span class="std std-ref">matching samples</span></a>.</p>
<section id="data-requirements">
<span id="sec-inference-data-requirements"></span><h2>Data requirements<a class="headerlink" href="#data-requirements" title="Link to this heading">#</a></h2>
<p>Input haplotype data for tskit must satisfy the following requirements:</p>
<ul class="simple">
<li><p>Data must be <em>phased</em>.</p></li>
<li><p>For each site used for inference we must know the <em>ancestral state</em>. Note that
this is not necessarily the same as the REF column from VCF, and ancestral states
must be computed using some other method. Sites with unknown ancestral states can be
specified, but are not used for inference; instead, the ancestral state is imputed.</p></li>
<li><p>Only biallelic sites can be used for inference.</p></li>
<li><p>Missing data can be included in the haplotypes using the value
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.MISSING_DATA" title="(in Project name not set)"><code class="xref py py-data docutils literal notranslate"><span class="pre">tskit.MISSING_DATA</span></code></a> (-1). The inferred tree sequence will have values
imputed for the haplotypes at these sites.</p></li>
</ul>
</section>
<section id="data-model">
<span id="sec-inference-data-model"></span><h2>Data model<a class="headerlink" href="#data-model" title="Link to this heading">#</a></h2>
<p>The data model for <code class="docutils literal notranslate"><span class="pre">tsinfer</span></code> is tightly integrated with
<code class="docutils literal notranslate"><span class="pre">tskit</span></code>’s <a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-data-model" title="(in Project name not set)"><span class="xref std std-ref">data model</span></a>
and uses the same concepts throughout. The intermediate file formats and APIs
described here provide a bridge between this model and existing data sources. For
convenience, we provide a brief description of concepts needed for importing
data into <code class="docutils literal notranslate"><span class="pre">tsinfer</span></code> here. Please see the
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/">tskit documentation</a>
for more detailed information.</p>
<section id="individual">
<span id="sec-inference-data-model-individual"></span><h3>Individual<a class="headerlink" href="#individual" title="Link to this heading">#</a></h3>
<p>In <code class="docutils literal notranslate"><span class="pre">tsinfer</span></code> an individual defines one of the subjects for which we have
genotype data. Individuals may have arbitrary ploidy levels (i.e. haploid,
diploid, tetraploid, etc.). Different individuals within a dataset can have
different ploidy levels. Note that this is equivalent to a “sample” in the
VCF Zarr input, whereas in tsinfer a “sample” is a single sequence within an
individual.</p>
<p>The tree sequence that you infer from your data will contain sample nodes
(i.e. genomes) that are linked to these individuals: the <code class="docutils literal notranslate"><span class="pre">tskit</span></code>
documentation provides more detail on the
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/glossary.html#sec-nodes-or-individuals" title="(in Project name not set)"><span class="xref std std-ref">distinction between individuals and the genomes they contain</span></a>.</p>
</section>
<section id="sample">
<span id="sec-inference-data-model-sample"></span><h3>Sample<a class="headerlink" href="#sample" title="Link to this heading">#</a></h3>
<p>Each individual is composed of one or more <code class="docutils literal notranslate"><span class="pre">sample</span></code> genomes, depending on their
ploidy. If an individual is diploid they have two samples, one each for the
maternal and paternal chromosome copies. More generally, a <code class="docutils literal notranslate"><span class="pre">node</span></code> refers
to a maternal or paternal chromosome that is either a sample, or an
ancestor of our samples.</p>
</section>
<section id="population">
<span id="sec-inference-data-model-population"></span><h3>Population<a class="headerlink" href="#population" title="Link to this heading">#</a></h3>
<p>A population is some grouping of individuals. Populations principally
exist to allow us define metadata for groups of individuals (although
technically, population IDs are associated with samples).
Populations are specified by an <code class="docutils literal notranslate"><span class="pre">individuals_population</span></code> array in the
VCF zarr input, which is an index for each individual to the population
into an accompanying <code class="docutils literal notranslate"><span class="pre">populations_metadata</span></code> array.</p>
</section>
<section id="metadata">
<span id="sec-inference-data-model-metadata"></span><h3>Metadata<a class="headerlink" href="#metadata" title="Link to this heading">#</a></h3>
<p>Metadata can be associated with populations and individuals in <code class="docutils literal notranslate"><span class="pre">tsinfer</span></code>,
which results in this information being available in the final tree
sequence. Metadata allows us to incorporate extra information
that is not part of the basic data model; for example, we can record
the upstream ID of a given individual and their family relationships
with other individuals.</p>
<p>In <code class="docutils literal notranslate"><span class="pre">tsinfer</span></code>, metadata can be stored by providing a JSON encodable
mapping. This information is then stored as JSON, and embedded in the
final tree sequence object and can be recovered using the <code class="docutils literal notranslate"><span class="pre">tskit</span></code>
APIs. Metadata arrays are read from the VCF zarr input.</p>
</section>
</section>
<section id="import-samples-data">
<span id="sec-inference-import-samples"></span><h2>Import samples data<a class="headerlink" href="#import-samples-data" title="Link to this heading">#</a></h2>
<p>In <code class="docutils literal notranslate"><span class="pre">tsinfer</span></code> we make several passes through the input sample haplotypes
in order to construct ancestors and to find copying paths for samples. To
do this efficiently we use <a class="reference external" href="https://github.com/sgkit-dev/vcf-zarr-spec">VCF Zarr</a>
as input. This provides very fast access to genotype data, compressed using cutting-edge
<a class="reference external" href="http://numcodecs.readthedocs.io">compression methods</a>. The input sample haplotypes and related metadata are a fraction of the size of a compressed VCF and can be processed efficiently.</p>
<p>VCF can be converted to VCF Zarr by the (bio2zarr)[https://sgkit-dev.github.io/bio2zarr]
package. See <a class="reference internal" href="usage.html#sec-usage"><span class="std std-ref">Usage</span></a> for examples.</p>
</section>
<section id="generate-ancestors">
<span id="sec-inference-generate-ancestors"></span><h2>Generate ancestors<a class="headerlink" href="#generate-ancestors" title="Link to this heading">#</a></h2>
<p>The first step in a <code class="docutils literal notranslate"><span class="pre">tsinfer</span></code> inference process is to generate a large
number of potential ancestors and to store these in an
<a class="reference internal" href="file_formats.html#sec-file-formats-ancestors"><span class="std std-ref">ancestors file</span></a>. The ancestors
file conventionally ends with <code class="docutils literal notranslate"><span class="pre">.ancestors</span></code>.</p>
<p>The ancestor generation algorithm is described in the Methods section of
<a class="reference external" href="https://doi.org/10.1038/s41588-019-0483-y">the original tsinfer paper</a>.</p>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>Describe the ancestor generation algorithm in more detail here</p>
</div>
</section>
<section id="matching-ancestors-samples">
<span id="sec-inference-match-ancestors-and-samples"></span><h2>Matching ancestors &amp; samples<a class="headerlink" href="#matching-ancestors-samples" title="Link to this heading">#</a></h2>
<p>After we have generated a set of potential ancestors and stored them in
an <a class="reference internal" href="file_formats.html#sec-file-formats-ancestors"><span class="std std-ref">ancestors file</span></a>, we then
run two matching steps. First we match the ancestors against each
other to generate an “ancestors tree sequence”, then we match the samples
against this ancestors tree sequence to generate the final result.</p>
<p>In both matching stages, we can set parameters that adjust the
behaviour of the matching algorithm, in particular the <code class="docutils literal notranslate"><span class="pre">path_compression</span></code>
setting, and the <code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code> and <code class="docutils literal notranslate"><span class="pre">mismatch_ratio</span></code> parameters.
The latter two only need to be specified if you wish to allow multiple
mutations to occur at a single site (i.e. breaking the infinite sites model
of mutation). Note, however, that multiple mutations are useful not only to
show true recurrent or back mutations in the evolutionary history of a
site, but also to represent errors in sequencing etc. which cause the
distribution of variation to fit poorly to the marginal tree at that site.
Hence, if there is error in your dataset, you may wish to experiment with
these settings to obtain optimal results.</p>
<section id="recombination-and-mismatch">
<span id="sec-inference-recombination-and-mismatch"></span><h3>Recombination and mismatch<a class="headerlink" href="#recombination-and-mismatch" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code> parameter is either a floating point value giving a
single rate (<span class="math notranslate nohighlight">\(\rho\)</span>) per unit length of genome, used to calculate the
genetic distance between adjacent sites, or an <a class="reference external" href="https://tskit.dev/msprime/docs/stable/api.html#msprime.RateMap" title="(in Project name not set)"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.RateMap</span></code></a> object
(whose <code class="docutils literal notranslate"><span class="pre">.get_cumulative_mass</span></code> method provides the genetic distances instead).
The genetic distances are then used to derive an array of probabilities of
recombination between adjacent sites, (<span class="math notranslate nohighlight">\(r\)</span>) used when assessing the relative
likelihood that a mismatch (and hence an extra mutation) may be responsible
for some patterns of variation at a site.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">mismatch_ratio</span></code> parameter is only relevant if a recombination rate has
been provided. It is used to adjust the balance of recombination to multiple
mutations at a site. More specifically, a single probability of mismatch is
used for all sites, which is calculated from the median genetic distance between
inference sites, such that for conventional (small) distances, the probability of
a mismatch is approximately equal to the probability of recombination multiplied
by the <code class="docutils literal notranslate"><span class="pre">mismatch_ratio</span></code>. In other words, a mismatch ratio of 2 makes a recurrent
mutation two times more likely than a recombination event to explain why an
otherwise closely matching ancestral haplotype does not match at a particular site.
Setting a high <code class="docutils literal notranslate"><span class="pre">mismatch_ratio</span></code> therefore results in tree sequences with more
recurrent mutations and fewer recombinations (and edges). Setting a low value
results in tree sequences with more recombination events and edges, and fewer
mutations. In the limit, as the mismatch_ratio tends to zero, only one mutation
will be inferred per variable site. This is the default behaviour if no
<code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code> is given or if there is only one inference site.
Alternatively, if <code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code> is set, <code class="docutils literal notranslate"><span class="pre">mismatch_ratio</span></code> defaults to
1, which has been shown to give reasonable results in simulated inference of
human-like data with error.  As a rough guide, such simulations recommend
mismatch ratios between 1e-3 and 1e3.</p>
</section>
<section id="path-compression">
<span id="sec-inference-path-compression"></span><h3>Path compression<a class="headerlink" href="#path-compression" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">path_compression</span></code> setting is used to further minimise recombination events
by looking for <em>shared recombination breakpoints</em>. A shared
breakpoint exists if a set of children share a breakpoint in the same position,
and they also have identical parents to the left of the breakpoint and identical
parents to the right of the breakpoint. Rather than supposing that these
children experienced multiple identical recombination events in parallel, we can
reduce the number of ancestral recombination events by postulating a “synthetic
ancestor” with this breakpoint, existing at a slightly older point
in time, from whom all the children are descended at this genomic position. We
call the algorithm used to implement this addition to the ancestral copying
paths, “path compression”.</p>
</section>
</section>
<section id="match-ancestors">
<span id="sec-inference-match-ancestors"></span><h2>Match ancestors<a class="headerlink" href="#match-ancestors" title="Link to this heading">#</a></h2>
<p>Matching ancestors is dependent on the time allocated to each ancestor; an
ancestor can only copy from any older ancestor. For each ancestor,
we find the most likely path through older ancestors: that is the path that
maximises the product of the probabilities of recombination and mismatch
over all sites.</p>
<div class="admonition-todo admonition" id="id2">
<p class="admonition-title">Todo</p>
<p>Schematic of the ancestors copying process.</p>
</div>
<p>The copying path for each ancestor then describes its ancestry at every
point in the sequence: from a genealogical perspective, we know its
parent node. This information is encoded precisely as an
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-edge-table-definition" title="(in Project name not set)"><span class="xref std std-ref">edge</span></a> in a <a class="reference external" href="https://tskit.dev/tskit/docs/stable/data-model.html#sec-data-model" title="(in Project name not set)"><span class="xref std std-ref">tree sequence</span></a>.
Thus, we refer to the output of this step as the “ancestors tree sequence”,
which is conventionally stored in a file ending with <code class="docutils literal notranslate"><span class="pre">.ancestors.trees</span></code>.</p>
</section>
<section id="match-samples">
<span id="sec-inference-match-samples"></span><h2>Match samples<a class="headerlink" href="#match-samples" title="Link to this heading">#</a></h2>
<p>The final phase of a <code class="docutils literal notranslate"><span class="pre">tsinfer</span></code> inference consists of a number steps:</p>
<ol class="arabic simple">
<li><p>The first (and usually most time-consuming) is to find copying paths
for our sample haplotypes through the ancestors. Each copying path
corresponds to a set of tree sequence edges in precisely the same
way as for ancestors, and the path compression algorithm can be equally
applied here.</p></li>
<li><p>As we only use a subset of the available sites for inference
(excluding by default any sites that are fixed, singletons, or where the ancestral
state is unknown) we then place mutations on the inferred trees in order to
represent the information at these sites. This is done using
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.Tree.map_mutations" title="(in Project name not set)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">tskit.Tree.map_mutations()</span></code></a>.</p></li>
<li><p>Post-process the final tree sequence to
a. modify the oldest ancestors in the tree sequence, in particular, by
taking the single “grand MRCA” containing entirely ancestral states
that <code class="docutils literal notranslate"><span class="pre">tsinfer</span></code> uses to provide a baseline ancestor for the matching
algorithm, and splitting it into separate ancestral genomic fragments.
b. remove topology at positions before the first site and one genomic unit
after the last site, to avoid extrapolating genealogical patterns into
regions for which we have no data.
c. remove ancestral paths that do not lead to any of the samples by
<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.simplify" title="(in Project name not set)"><code class="xref py py-meth docutils literal notranslate"><span class="pre">simplifying</span></code></a> the final output. When
simplifying, we keep non-branching (“unary”) nodes, as they indicate
ancestors which we have actively inferred, and
for technical reasons keeping unary ancestors can also lead to better
compression. Note that this means that not every internal node in the
inferred tree sequence will correspond to a coalescent event.</p></li>
</ol>
<div class="admonition-todo admonition" id="id3">
<p class="admonition-title">Todo</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>1. Describe path compression here and above in the ancestors
   section
2. Describe the structure of the output tree sequences; how the
   nodes are mapped, what the time values mean, etc.
</pre></div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="usage.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Usage</p>
      </div>
    </a>
    <a class="right-next"
       href="large_scale.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Large Scale Inference</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-requirements">Data requirements</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-model">Data model</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#individual">Individual</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sample">Sample</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#population">Population</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#metadata">Metadata</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#import-samples-data">Import samples data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-ancestors">Generate ancestors</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#matching-ancestors-samples">Matching ancestors &amp; samples</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#recombination-and-mismatch">Recombination and mismatch</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#path-compression">Path compression</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#match-ancestors">Match ancestors</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#match-samples">Match samples</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Tskit Developers
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2018.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>