
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Python API &#8212; Tsdate manual</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=5349f25f" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=9c3e77be" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=1ae7504c"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'python-api';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Command line interface" href="cli.html" />
    <link rel="prev" title="APIs" href="api.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <script data-goatcounter="https://tskit.goatcounter.com/count" async src="//gc.zgo.at/count.js"></script>
</head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/tsdate_logo.svg" class="logo__image only-light" alt=""/>
    <script>document.write(`<img src="_static/tsdate_logo.svg" class="logo__image only-dark" alt=""/>`);</script>
  
  
    <p class="title logo__title">Version 0.2.5</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="usage.html">Usage</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="methods.html">Methods</a></li>
<li class="toctree-l2"><a class="reference internal" href="real-data.html">Real data</a></li>
<li class="toctree-l2"><a class="reference internal" href="popsize.html">Population size</a></li>
<li class="toctree-l2"><a class="reference internal" href="historical_samples.html">Historical (Ancient) Samples</a></li>
<li class="toctree-l2"><a class="reference internal" href="priors.html">More on priors (old)</a></li>
</ul>
</details></li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="api.html">APIs</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Python API</a></li>


<li class="toctree-l2"><a class="reference internal" href="cli.html">Command line interface</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="citation.html">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="bibliography.html">References</a></li>
<li class="toctree-l1"><a class="reference internal" href="CHANGELOG.html">Changelog</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/python-api.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Python API</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Python API</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#running-tsdate">Running tsdate</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.date"><code class="docutils literal notranslate"><span class="pre">date()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.core.estimation_methods"><code class="docutils literal notranslate"><span class="pre">estimation_methods</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.variational_gamma"><code class="docutils literal notranslate"><span class="pre">variational_gamma()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.inside_outside"><code class="docutils literal notranslate"><span class="pre">inside_outside()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.maximization"><code class="docutils literal notranslate"><span class="pre">maximization()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#underlying-fit-objects">Underlying fit objects</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.discrete.BeliefPropagation"><code class="docutils literal notranslate"><span class="pre">BeliefPropagation</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.discrete.BeliefPropagation.node_posteriors"><code class="docutils literal notranslate"><span class="pre">BeliefPropagation.node_posteriors()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.variational.ExpectationPropagation"><code class="docutils literal notranslate"><span class="pre">ExpectationPropagation</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.variational.ExpectationPropagation.node_posteriors"><code class="docutils literal notranslate"><span class="pre">ExpectationPropagation.node_posteriors()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.variational.ExpectationPropagation.mutation_posteriors"><code class="docutils literal notranslate"><span class="pre">ExpectationPropagation.mutation_posteriors()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#prior-and-time-discretisation-options">Prior and Time Discretisation Options</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.build_prior_grid"><code class="docutils literal notranslate"><span class="pre">build_prior_grid()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.build_parameter_grid"><code class="docutils literal notranslate"><span class="pre">build_parameter_grid()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.node_time_class.NodeTimeValues"><code class="docutils literal notranslate"><span class="pre">NodeTimeValues</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.prior.DEFAULT_APPROX_PRIOR_SIZE"><code class="docutils literal notranslate"><span class="pre">DEFAULT_APPROX_PRIOR_SIZE</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#variable-population-sizes">Variable population sizes</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.demography.PopulationSizeHistory"><code class="docutils literal notranslate"><span class="pre">PopulationSizeHistory</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#preprocessing-tree-sequences">Preprocessing Tree Sequences</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.preprocess_ts"><code class="docutils literal notranslate"><span class="pre">preprocess_ts()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.util.split_disjoint_nodes"><code class="docutils literal notranslate"><span class="pre">split_disjoint_nodes()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#functions-for-inferring-tree-sequences-with-historical-samples">Functions for Inferring Tree Sequences with Historical Samples</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.sites_time_from_ts"><code class="docutils literal notranslate"><span class="pre">sites_time_from_ts()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.add_sampledata_times"><code class="docutils literal notranslate"><span class="pre">add_sampledata_times()</span></code></a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#module-tsdate">Constants</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.NODE_IS_HISTORICAL_SAMPLE"><code class="docutils literal notranslate"><span class="pre">NODE_IS_HISTORICAL_SAMPLE</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.NODE_SPLIT_BY_PREPROCESS"><code class="docutils literal notranslate"><span class="pre">NODE_SPLIT_BY_PREPROCESS</span></code></a></li>
</ul>
</li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="python-api">
<span id="sec-python-api"></span><h1>Python API<a class="headerlink" href="#python-api" title="Link to this heading">#</a></h1>
<p>This page provides formal documentation for the <em>tsdate</em> Python API.</p>
<section id="running-tsdate">
<h2>Running tsdate<a class="headerlink" href="#running-tsdate" title="Link to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="tsdate.date">
<span class="sig-prename descclassname"><span class="pre">tsdate.</span></span><span class="sig-name descname"><span class="pre">date</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree_sequence</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation_rate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recombination_rate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_units</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">constr_iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_branch_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">set_metadata</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_fit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_likelihood</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_unary</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">progress</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsdate/core.html#date"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsdate.date" title="Link to this definition">#</a></dt>
<dd><p>Infer dates for nodes in a genealogical graph (or <a class="reference external" href="https://tskit.dev/tutorials/args.html#sec-args" title="(in Project name not set)"><span class="xref std std-ref">ARG</span></a>)
stored in the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/introduction.html#sec-introduction" title="(in Project name not set)"><span class="xref std std-ref">succinct tree sequence</span></a> format.
New times are assigned to nodes using the estimation algorithm specified by
<code class="docutils literal notranslate"><span class="pre">method</span></code> (see note below). A <code class="docutils literal notranslate"><span class="pre">mutation_rate</span></code> must be given (the recombination_rate
parameter, implementing a recombination clock, is unsupported at this
time). Times associated with mutations and times associated
with non-fixed (non-sample) nodes are overwritten. For example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">mu</span> <span class="o">=</span> <span class="mf">1e-8</span>
<span class="n">new_ts</span> <span class="o">=</span> <span class="n">tsdate</span><span class="o">.</span><span class="n">date</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="n">mu</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This is a wrapper for the named functions that are listed in
<a class="reference internal" href="#tsdate.core.estimation_methods" title="tsdate.core.estimation_methods"><code class="xref py py-data docutils literal notranslate"><span class="pre">estimation_methods</span></code></a>. Details and specific parameters for
each estimation method are given in the documentation for those functions.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree_sequence</strong> (<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)"><em>TreeSequence</em></a>) – The input tree sequence to be dated (for
example one with <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TIME_UNITS_UNCALIBRATED" title="(in Project name not set)"><code class="xref py py-data docutils literal notranslate"><span class="pre">uncalibrated</span></code></a> node times).</p></li>
<li><p><strong>mutation_rate</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>) – The estimated mutation rate per unit of genome per
unit time (see individual methods)</p></li>
<li><p><strong>recombination_rate</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>) – The estimated recombination rate per unit of genome
per unit time. If provided, the dating algorithm will use a recombination rate
clock to help estimate node dates. Default: <code class="docutils literal notranslate"><span class="pre">None</span></code> (not currently implemented)</p></li>
<li><p><strong>time_units</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>) – The time units used by the <code class="docutils literal notranslate"><span class="pre">mutation_rate</span></code> and
<code class="docutils literal notranslate"><span class="pre">recombination_rate</span></code> values, and stored in the <code class="docutils literal notranslate"><span class="pre">time_units</span></code> attribute of the
output tree sequence. If the conditional coalescent prior is used,
then this is also applies to the value of <code class="docutils literal notranslate"><span class="pre">population_size</span></code>, which in
standard coalescent theory is measured in generations. Therefore if you
wish to use mutation and recombination rates measured in (say) years,
and are using the conditional coalescent prior, the <code class="docutils literal notranslate"><span class="pre">population_size</span></code>
value which you provide must be scaled by multiplying by the number of
years per generation. If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), assume <code class="docutils literal notranslate"><span class="pre">&quot;generations&quot;</span></code>.</p></li>
<li><p><strong>method</strong> (<em>string</em>) – What estimation method to use. See
<a class="reference internal" href="#tsdate.core.estimation_methods" title="tsdate.core.estimation_methods"><code class="xref py py-data docutils literal notranslate"><span class="pre">estimation_methods</span></code></a> for possible values.
If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) the “variational_gamma” method is currently chosen.</p></li>
<li><p><strong>constr_iterations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>) – The maximum number of constrained least
squares iterations to use prior to forcing positive branch lengths.
Default: None, treated as 0.</p></li>
<li><p><strong>min_branch_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>) – The minimum distance separating parent and
child ages in the returned tree sequence. Default: None, treated as 1e-8</p></li>
<li><p><strong>set_metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – Should unconstrained times be stored in table metadata,
in the form of <code class="docutils literal notranslate"><span class="pre">&quot;mn&quot;</span></code> (mean) and <code class="docutils literal notranslate"><span class="pre">&quot;vr&quot;</span></code> (variance) fields? If <code class="docutils literal notranslate"><span class="pre">False</span></code>,
do not store metadata.  If <code class="docutils literal notranslate"><span class="pre">True</span></code>, force metadata to be set (if no schema
is set or the schema is incompatible, clear existing metadata in the relevant
tables and set a new schema). If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), only set metadata if
the existing schema allows (this may overwrite existing <code class="docutils literal notranslate"><span class="pre">&quot;mn&quot;</span></code> and <code class="docutils literal notranslate"><span class="pre">&quot;vr&quot;</span></code>
fields) or if existing metadata is empty, otherwise issue a warning.</p></li>
<li><p><strong>return_fit</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, instead of just a dated tree sequence,
return a tuple of <code class="docutils literal notranslate"><span class="pre">(dated_ts,</span> <span class="pre">fit)</span></code>. Default: None, treated as False.</p></li>
<li><p><strong>return_likelihood</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, return the log marginal likelihood
from the inside algorithm in addition to the dated tree sequence. If
<code class="docutils literal notranslate"><span class="pre">return_fit</span></code> is also <code class="docutils literal notranslate"><span class="pre">True</span></code>, then the marginal likelihood
will be the last element of the tuple. Default: None, treated as False.</p></li>
<li><p><strong>allow_unary</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – Allow nodes that are “locally unary” (i.e. have only
one child in one or more local trees). Default: None, treated as False.</p></li>
<li><p><strong>progress</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – Show a progress bar. Default: None, treated as False.</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – Should the tsdate command be appended to the
provenence information in the returned tree sequence?
Default: None, treated as True.</p></li>
<li><p><strong>**kwargs</strong> – Other keyword arguments specific to the
<a class="reference internal" href="#tsdate.core.estimation_methods" title="tsdate.core.estimation_methods"><code class="xref py py-data docutils literal notranslate"><span class="pre">estimation</span> <span class="pre">method</span></code></a> used. These are
documented in those specific functions.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A copy of the input tree sequence but with updated node times, or (if
<code class="docutils literal notranslate"><span class="pre">return_fit</span></code> or <code class="docutils literal notranslate"><span class="pre">return_likelihood</span></code> is True) a tuple of that
tree sequence plus a fit object and/or the
marginal likelihood given the mutations on the tree sequence.</p>
</dd>
</dl>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="tsdate.core.estimation_methods">
<span class="sig-prename descclassname"><span class="pre">tsdate.core.</span></span><span class="sig-name descname"><span class="pre">estimation_methods</span></span><a class="headerlink" href="#tsdate.core.estimation_methods" title="Link to this definition">#</a></dt>
<dd><p>The names of available estimation methods, each mapped to a function to carry
out the appropriate method. Names can be passed as strings to the
<a class="reference internal" href="#tsdate.date" title="tsdate.date"><code class="xref py py-func docutils literal notranslate"><span class="pre">date()</span></code></a> function, or each named function can be called directly:</p>
<ul class="simple">
<li><p><a class="reference internal" href="#tsdate.variational_gamma" title="tsdate.variational_gamma"><code class="xref py py-func docutils literal notranslate"><span class="pre">tsdate.variational_gamma()</span></code></a>: variational approximation, empirically most accurate.</p></li>
<li><p><a class="reference internal" href="#tsdate.inside_outside" title="tsdate.inside_outside"><code class="xref py py-func docutils literal notranslate"><span class="pre">tsdate.inside_outside()</span></code></a>: empirically better, theoretically problematic.</p></li>
<li><p><a class="reference internal" href="#tsdate.maximization" title="tsdate.maximization"><code class="xref py py-func docutils literal notranslate"><span class="pre">tsdate.maximization()</span></code></a>: worse empirically, especially with gamma approximated
priors, but theoretically robust</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tsdate.variational_gamma">
<span class="sig-prename descclassname"><span class="pre">tsdate.</span></span><span class="sig-name descname"><span class="pre">variational_gamma</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree_sequence</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation_rate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rescaling_intervals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rescaling_iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">match_segregating_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsdate/core.html#variational_gamma"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsdate.variational_gamma" title="Link to this definition">#</a></dt>
<dd><p>Infer dates for nodes in a tree sequence using expectation propagation,
which approximates the marginal posterior distribution of a given node’s
age with a gamma distribution. Convergence to the correct posterior moments
is obtained by updating the distributions for node dates using several rounds
of iteration. For example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">new_ts</span> <span class="o">=</span> <span class="n">tsdate</span><span class="o">.</span><span class="n">variational_gamma</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span> <span class="n">max_iterations</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
<p>A piecewise-constant uniform distribution is used as a prior for each
node, that is updated via expectation maximization in each iteration.
Node-specific priors are not currently supported.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree_sequence</strong> (<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)"><em>TreeSequence</em></a>) – The input tree sequence to be dated.</p></li>
<li><p><strong>mutation_rate</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>) – The estimated mutation rate per unit of genome per
unit time.</p></li>
<li><p><strong>max_iterations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>) – The number of iterations used in the expectation
propagation algorithm. Default: None, treated as 25.</p></li>
<li><p><strong>rescaling_intervals</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>) – For time rescaling, the number of time
intervals within which to estimate a rescaling parameter. Setting this to zero
means that rescaling is not performed. Default <code class="docutils literal notranslate"><span class="pre">None</span></code>, treated as 1000.</p></li>
<li><p><strong>rescaling_iterations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>) – The number of iterations for time rescaling.
Setting this to zero means that rescaling is not performed. Default
<code class="docutils literal notranslate"><span class="pre">None</span></code>, treated as 5.</p></li>
<li><p><strong>match_segregating_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, then time is rescaled
such that branch- and site-mode segregating sites are approximately equal.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, time is rescaled such that branch- and site-mode root-to-leaf
length are approximately equal, which gives unbiased estimates when there
are polytomies. Default <code class="docutils literal notranslate"><span class="pre">False</span></code>.</p></li>
<li><p><strong>**kwargs</strong> – Other keyword arguments as described in the <a class="reference internal" href="#tsdate.date" title="tsdate.date"><code class="xref py py-func docutils literal notranslate"><span class="pre">date()</span></code></a> wrapper
function, including <code class="docutils literal notranslate"><span class="pre">time_units</span></code>, <code class="docutils literal notranslate"><span class="pre">progress</span></code>, <code class="docutils literal notranslate"><span class="pre">allow_unary</span></code> and
<code class="docutils literal notranslate"><span class="pre">record_provenance</span></code>. The arguments <code class="docutils literal notranslate"><span class="pre">return_fit</span></code> and <code class="docutils literal notranslate"><span class="pre">return_likelihood</span></code>
can be used to return additional information (see below).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>ts</strong> (<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>) – a copy of the input tree sequence with
updated node times based on the posterior mean, corrected where necessary to
ensure that parents are strictly older than all their children by an amount
given by the <code class="docutils literal notranslate"><span class="pre">min_branch_length</span></code> parameter.</p></li>
<li><p><strong>fit</strong> (<a class="reference internal" href="#tsdate.variational.ExpectationPropagation" title="tsdate.variational.ExpectationPropagation"><code class="xref py py-class docutils literal notranslate"><span class="pre">ExpectationPropagation</span></code></a>) – (Only returned
if <code class="docutils literal notranslate"><span class="pre">return_fit</span></code> is <code class="docutils literal notranslate"><span class="pre">True</span></code>). The underlying object used to run the dating
inference. This can then be queried e.g. using
<a class="reference internal" href="#tsdate.variational.ExpectationPropagation.node_posteriors" title="tsdate.variational.ExpectationPropagation.node_posteriors"><code class="xref py py-meth docutils literal notranslate"><span class="pre">node_posteriors()</span></code></a></p></li>
<li><p><strong>marginal_likelihood</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>) – (Only returned if
<code class="docutils literal notranslate"><span class="pre">return_likelihood</span></code> is <code class="docutils literal notranslate"><span class="pre">True</span></code>) The marginal likelihood of
the mutation data given the inferred node times. Not currently
implemented for this method (set to <code class="docutils literal notranslate"><span class="pre">None</span></code>)</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tsdate.inside_outside">
<span class="sig-prename descclassname"><span class="pre">tsdate.</span></span><span class="sig-name descname"><span class="pre">inside_outside</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree_sequence</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation_rate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">population_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">priors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outside_standardize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_oldest_root</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">probability_space</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsdate/core.html#inside_outside"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsdate.inside_outside" title="Link to this definition">#</a></dt>
<dd><p>Infer dates for nodes in a genealogical graph using the “inside outside” algorithm.
This approximates the marginal posterior distribution of a node’s age using an
atomic discretization of time (e.g. point masses at particular timepoints).</p>
<p>Currently, this estimation method comprises a single “inside” followed by a similar
“outside” step. The inside step passes backwards in time from the samples to the
roots of the graph,taking account of the distributions of times of each node’s child
(and if a <code class="docutils literal notranslate"><span class="pre">mutation_rate</span></code> is given, the the number of mutations on each edge).
The outside step passes forwards in time from the roots, incorporating the time
distributions for each node’s parents. If there are (undirected) cycles in the
underlying graph, this method does not provide a theoretically exact estimate
of the marginal posterior distribution of node ages, but in practice it
results in an accurate approximation.</p>
<p>For example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">new_ts</span> <span class="o">=</span> <span class="n">tsdate</span><span class="o">.</span><span class="n">inside_outside</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span> <span class="n">population_size</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The prior parameters for each node-to-be-dated take the form of probabilities
for each node at a set of discrete timepoints. If the <code class="docutils literal notranslate"><span class="pre">priors</span></code> parameter is
used, it must specify an object constructed using <a class="reference internal" href="#tsdate.build_prior_grid" title="tsdate.build_prior_grid"><code class="xref py py-func docutils literal notranslate"><span class="pre">build_prior_grid()</span></code></a>
(this can be used to define the number and position of the timepoints).
If <code class="docutils literal notranslate"><span class="pre">priors</span></code> is not used, <code class="docutils literal notranslate"><span class="pre">population_size</span></code> must be provided,
which is used to create a default prior derived from the conditional coalescent
(tilted according to population size and weighted by the genomic
span over which a node has a given number of descendant samples). This default
prior assumes the nodes to be dated are all the non-sample nodes in the input
tree sequence, and that they are contemporaneous.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree_sequence</strong> (<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)"><em>TreeSequence</em></a>) – The input tree sequence to be dated.</p></li>
<li><p><strong>mutation_rate</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>) – The estimated mutation rate per unit of genome per
unit time. If provided, the dating algorithm will use a mutation rate clock to
help estimate node dates. Default: <code class="docutils literal notranslate"><span class="pre">None</span></code></p></li>
<li><p><strong>population_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a><em> or </em><a class="reference internal" href="#tsdate.demography.PopulationSizeHistory" title="tsdate.demography.PopulationSizeHistory"><em>PopulationSizeHistory</em></a>) – The estimated
(diploid) effective population size used to construct the (default) conditional
coalescent prior. For a population with constant size, this can be given as a
single value (for example, as commonly estimated by the observed genetic
diversity of the sample divided by four-times the expected mutation rate).
Alternatively, for a population with time-varying size, this can be given
directly as a <a class="reference internal" href="#tsdate.demography.PopulationSizeHistory" title="tsdate.demography.PopulationSizeHistory"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationSizeHistory</span></code></a> object or a parameter
dictionary passed to initialise a <a class="reference internal" href="#tsdate.demography.PopulationSizeHistory" title="tsdate.demography.PopulationSizeHistory"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationSizeHistory</span></code></a>
object. The <code class="docutils literal notranslate"><span class="pre">population_size</span></code> parameter is only used when <code class="docutils literal notranslate"><span class="pre">priors</span></code> is
<code class="docutils literal notranslate"><span class="pre">None</span></code>. Conversely, if <code class="docutils literal notranslate"><span class="pre">priors</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>, no <code class="docutils literal notranslate"><span class="pre">population_size</span></code>
value should be specified.</p></li>
<li><p><strong>priors</strong> (<a class="reference internal" href="#tsdate.node_time_class.NodeTimeValues" title="tsdate.node_time_class.NodeTimeValues"><em>tsdate.node_time_class.NodeTimeValues</em></a>) – NodeTimeValues object containing
the prior parameters for each node-to-be-dated. Note that different estimation
methods may require different types of prior, as described in the documentation
for each estimation method.</p></li>
<li><p><strong>eps</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>) – The error factor in time difference calculations. Default: None,
treated as 1e-8.</p></li>
<li><p><strong>num_threads</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>) – The number of threads to use when precalculating likelihoods.
A simpler unthreaded algorithm is used unless this is &gt;= 1. Default: None</p></li>
<li><p><strong>outside_standardize</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – Should the likelihoods be standardized during the
outside step? This can help to avoid numerical under/overflow. Using
unstandardized values is mostly useful for testing (e.g. to obtain, in the
outside step, the total functional value for each node).
Default: None, treated as True.</p></li>
<li><p><strong>ignore_oldest_root</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – Should the oldest root in the tree sequence be
ignored in the outside algorithm (if <code class="docutils literal notranslate"><span class="pre">&quot;inside_outside&quot;</span></code> is used as the method).
Ignoring outside root can provide greater stability when dating tree sequences
inferred from real data, in particular if all local trees are assumed to coalesce
in a single “grand MRCA”, as in older versions of <code class="docutils literal notranslate"><span class="pre">tsinfer</span></code>.
Default: None, treated as False.</p></li>
<li><p><strong>probability_space</strong> (<em>string</em>) – Should the internal algorithm save
probabilities in “logarithmic” (slower, less liable to to overflow) or
“linear” space (fast, may overflow). Default: “logarithmic”</p></li>
<li><p><strong>**kwargs</strong> – Other keyword arguments as described in the <a class="reference internal" href="#tsdate.date" title="tsdate.date"><code class="xref py py-func docutils literal notranslate"><span class="pre">date()</span></code></a> wrapper
function, notably <code class="docutils literal notranslate"><span class="pre">mutation_rate</span></code>, and <code class="docutils literal notranslate"><span class="pre">population_size</span></code> or <code class="docutils literal notranslate"><span class="pre">priors</span></code>.
Further arguments include <code class="docutils literal notranslate"><span class="pre">time_units</span></code>, <code class="docutils literal notranslate"><span class="pre">progress</span></code>, <code class="docutils literal notranslate"><span class="pre">allow_unary</span></code> and
<code class="docutils literal notranslate"><span class="pre">record_provenance</span></code>. The additional arguments <code class="docutils literal notranslate"><span class="pre">return_fit</span></code> and
<code class="docutils literal notranslate"><span class="pre">return_likelihood</span></code> can be used to return additional information (see below).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>ts</strong> (<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>) – a copy of the input tree sequence with
updated node times based on the posterior mean, corrected where necessary to
ensure that parents are strictly older than all their children by an amount
given by the <code class="docutils literal notranslate"><span class="pre">min_branch_length</span></code> parameter.</p></li>
<li><p><strong>fit</strong> (<a class="reference internal" href="#tsdate.discrete.BeliefPropagation" title="tsdate.discrete.BeliefPropagation"><code class="xref py py-class docutils literal notranslate"><span class="pre">BeliefPropagation</span></code></a>) – (Only returned if
<code class="docutils literal notranslate"><span class="pre">return_fit</span></code> is <code class="docutils literal notranslate"><span class="pre">True</span></code>) The underlying object used to run the dating
inference. This can then be queried e.g. using
<a class="reference internal" href="#tsdate.discrete.BeliefPropagation.node_posteriors" title="tsdate.discrete.BeliefPropagation.node_posteriors"><code class="xref py py-meth docutils literal notranslate"><span class="pre">node_posteriors()</span></code></a></p></li>
<li><p><strong>marginal_likelihood</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>) – (Only returned if
<code class="docutils literal notranslate"><span class="pre">return_likelihood</span></code> is <code class="docutils literal notranslate"><span class="pre">True</span></code>) The marginal likelihood of
the mutation data given the inferred node times.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tsdate.maximization">
<span class="sig-prename descclassname"><span class="pre">tsdate.</span></span><span class="sig-name descname"><span class="pre">maximization</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree_sequence</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation_rate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">population_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">priors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">probability_space</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsdate/core.html#maximization"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsdate.maximization" title="Link to this definition">#</a></dt>
<dd><p>Infer dates for nodes in a genealogical graph using the “outside maximization”
algorithm. This approximates the marginal posterior distribution of a node’s
age using an atomic discretization of time (e.g. point masses at particular
timepoints).</p>
<p>This estimation method comprises a single “inside” step followed by an
“outside maximization” step. The inside step passes backwards in time from the
samples to the roots of the graph,taking account of the distributions of times of
each node’s child (and if a <code class="docutils literal notranslate"><span class="pre">mutation_rate</span></code> is given, the the number of mutations
on each edge). The outside maximization step passes forwards in time from the roots,
updating each node’s time on the basis of the most likely timepoint for
each parent of that node. This provides a reasonable point estimate for node times,
but does not generate a true posterior time distribution.</p>
<p>For example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">new_ts</span> <span class="o">=</span> <span class="n">tsdate</span><span class="o">.</span><span class="n">maximization</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span> <span class="n">population_size</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The prior parameters for each node-to-be-dated take the form of probabilities
for each node at a set of discrete timepoints. If the <code class="docutils literal notranslate"><span class="pre">priors</span></code> parameter is
used, it must specify an object constructed using <a class="reference internal" href="#tsdate.build_prior_grid" title="tsdate.build_prior_grid"><code class="xref py py-func docutils literal notranslate"><span class="pre">build_prior_grid()</span></code></a>
(this can be used to define the number and position of the timepoints).
If <code class="docutils literal notranslate"><span class="pre">priors</span></code> is not used, <code class="docutils literal notranslate"><span class="pre">population_size</span></code> must be provided,
which is used to create a default prior derived from the conditional coalescent
(tilted according to population size and weighted by the genomic
span over which a node has a given number of descendant samples). This default
prior assumes the nodes to be dated are all the non-sample nodes in the input
tree sequence, and that they are contemporaneous.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree_sequence</strong> (<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)"><em>TreeSequence</em></a>) – The input tree sequence to be dated.</p></li>
<li><p><strong>mutation_rate</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>) – The estimated mutation rate per unit of genome per
unit time. If provided, the dating algorithm will use a mutation rate clock to
help estimate node dates. Default: <code class="docutils literal notranslate"><span class="pre">None</span></code></p></li>
<li><p><strong>population_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a><em> or </em><a class="reference internal" href="#tsdate.demography.PopulationSizeHistory" title="tsdate.demography.PopulationSizeHistory"><em>PopulationSizeHistory</em></a>) – The estimated
(diploid) effective population size used to construct the (default) conditional
coalescent prior. For a population with constant size, this can be given as a
single value (for example, as commonly estimated by the observed genetic
diversity of the sample divided by four-times the expected mutation rate).
Alternatively, for a population with time-varying size, this can be given
directly as a <a class="reference internal" href="#tsdate.demography.PopulationSizeHistory" title="tsdate.demography.PopulationSizeHistory"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationSizeHistory</span></code></a> object or a parameter
dictionary passed to initialise a <a class="reference internal" href="#tsdate.demography.PopulationSizeHistory" title="tsdate.demography.PopulationSizeHistory"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationSizeHistory</span></code></a>
object. The <code class="docutils literal notranslate"><span class="pre">population_size</span></code> parameter is only used when <code class="docutils literal notranslate"><span class="pre">priors</span></code> is
<code class="docutils literal notranslate"><span class="pre">None</span></code>. Conversely, if <code class="docutils literal notranslate"><span class="pre">priors</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>, no <code class="docutils literal notranslate"><span class="pre">population_size</span></code>
value should be specified.</p></li>
<li><p><strong>priors</strong> (<a class="reference internal" href="#tsdate.node_time_class.NodeTimeValues" title="tsdate.node_time_class.NodeTimeValues"><em>tsdate.node_time_class.NodeTimeValues</em></a>) – NodeTimeValues object containing
the prior parameters for each node-to-be-dated. Note that different estimation
methods may require different types of prior, as described in the documentation
for each estimation method.</p></li>
<li><p><strong>eps</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>) – The error factor in time difference calculations. Default: None,
treated as 1e-8.</p></li>
<li><p><strong>num_threads</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>) – The number of threads to use when precalculating likelihoods.
A simpler unthreaded algorithm is used unless this is &gt;= 1. Default: None</p></li>
<li><p><strong>probability_space</strong> (<em>string</em>) – Should the internal algorithm save
probabilities in “logarithmic” (slower, less liable to to overflow) or
“linear” space (fast, may overflow). Default: None treated as”logarithmic”</p></li>
<li><p><strong>**kwargs</strong> – Other keyword arguments as described in the <a class="reference internal" href="#tsdate.date" title="tsdate.date"><code class="xref py py-func docutils literal notranslate"><span class="pre">date()</span></code></a> wrapper
function, notably <code class="docutils literal notranslate"><span class="pre">mutation_rate</span></code>, and <code class="docutils literal notranslate"><span class="pre">population_size</span></code> or <code class="docutils literal notranslate"><span class="pre">priors</span></code>.
Further arguments include <code class="docutils literal notranslate"><span class="pre">time_units</span></code>, <code class="docutils literal notranslate"><span class="pre">progress</span></code>, <code class="docutils literal notranslate"><span class="pre">allow_unary</span></code> and
<code class="docutils literal notranslate"><span class="pre">record_provenance</span></code>.  The additional arguments <code class="docutils literal notranslate"><span class="pre">return_fit</span></code> and
<code class="docutils literal notranslate"><span class="pre">return_likelihood</span></code> can be used to return additional information (see below).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>ts</strong> (<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)"><code class="xref py py-class docutils literal notranslate"><span class="pre">TreeSequence</span></code></a>) – a copy of the input tree sequence with
updated node times based on the posterior mean, corrected where necessary to
ensure that parents are strictly older than all their children by an amount
given by the <code class="docutils literal notranslate"><span class="pre">min_branch_length</span></code> parameter.</p></li>
<li><p><strong>marginal_likelihood</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>) – (Only returned if
<code class="docutils literal notranslate"><span class="pre">return_likelihood</span></code> is <code class="docutils literal notranslate"><span class="pre">True</span></code>) The marginal likelihood of
the mutation data given the inferred node times.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="underlying-fit-objects">
<h2>Underlying fit objects<a class="headerlink" href="#underlying-fit-objects" title="Link to this heading">#</a></h2>
<p>Instances of the classes below are returned by setting <code class="docutils literal notranslate"><span class="pre">return_fit=True</span></code>
when dating. The fits can be inspected to obtain more detailed results than
may be present in the returned tree sequence and its metadata. The classes
are not intended to be instantiated directly.</p>
<dl class="py class">
<dt class="sig sig-object py" id="tsdate.discrete.BeliefPropagation">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tsdate.discrete.</span></span><span class="sig-name descname"><span class="pre">BeliefPropagation</span></span><a class="reference internal" href="_modules/tsdate/discrete.html#BeliefPropagation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsdate.discrete.BeliefPropagation" title="Link to this definition">#</a></dt>
<dd><p>The class that encapsulates running exact belief propagation models,
in particular the discrete-time inside and outside algorithms.</p>
<dl class="py method">
<dt class="sig sig-object py" id="tsdate.discrete.BeliefPropagation.node_posteriors">
<span class="sig-name descname"><span class="pre">node_posteriors</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsdate/discrete.html#BeliefPropagation.node_posteriors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsdate.discrete.BeliefPropagation.node_posteriors" title="Link to this definition">#</a></dt>
<dd><p>Return the distribution of posterior node times as a structured array. The
returned value can be e.g. read into <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> for further analysis.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <code class="docutils literal notranslate"><span class="pre">outside_maximization</span></code> method does not provide node time posteriors.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The distribution of posterior node times as a structured array with
columns as timepoints. Row <code class="docutils literal notranslate"><span class="pre">i</span></code> corresponds to the probabilities of
node <code class="docutils literal notranslate"><span class="pre">i</span></code> lying at each timepoint. Nodes with fixed times are set to
<code class="docutils literal notranslate"><span class="pre">np.nan</span></code> for the entire row.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="tsdate.variational.ExpectationPropagation">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tsdate.variational.</span></span><span class="sig-name descname"><span class="pre">ExpectationPropagation</span></span><a class="reference internal" href="_modules/tsdate/variational.html#ExpectationPropagation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsdate.variational.ExpectationPropagation" title="Link to this definition">#</a></dt>
<dd><p>The class that encapsulates running the variational gamma approach to
tsdate fitting. This contains the Expectation propagation (EP) algorithm
to infer approximate marginal distributions for node ages.</p>
<p>The probability model has the form,</p>
<div class="math notranslate nohighlight">
\[\prod_{i \in \mathcal{N}} f(t_i | \theta_i)
\prod_{(i,j) \in \mathcal{E}} g(y_ij | t_i - t_j)\]</div>
<p>where <span class="math notranslate nohighlight">\(f(.)\)</span> is a prior distribution on node ages with parameters
<span class="math notranslate nohighlight">\(\\theta\)</span> and <span class="math notranslate nohighlight">\(g(.)\)</span> are Poisson likelihoods per edge. The
EP approximation to the posterior has the form,</p>
<div class="math notranslate nohighlight">
\[\prod_{i \in \mathcal{N}} q(t_i | \eta_i)
\prod_{(i,j) \in \mathcal{E}} q(t_i | \gamma_{ij}) q(t_j | \kappa_{ij})\]</div>
<p>where <span class="math notranslate nohighlight">\(q(.)\)</span> are pseudo-gamma distributions (termed ‘factors’), and
<span class="math notranslate nohighlight">\(\eta, \gamma, \kappa\)</span> are variational parameters that reflect to
prior, inside (leaf-to-root), and outside (root-to-edge) information.</p>
<p>Thus, the EP approximation results in gamma-distribution marginals.  The
factors <span class="math notranslate nohighlight">\(q(.)\)</span> do not need to be valid distributions (e.g. the
shape/rate parameters may be negative), as long as the marginals are valid
distributions.  For details on how the variational parameters are
optimized, see Minka (2002) “Expectation Propagation for Approximate
Bayesian Inference”</p>
<dl class="py method">
<dt class="sig sig-object py" id="tsdate.variational.ExpectationPropagation.node_posteriors">
<span class="sig-name descname"><span class="pre">node_posteriors</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsdate/variational.html#ExpectationPropagation.node_posteriors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsdate.variational.ExpectationPropagation.node_posteriors" title="Link to this definition">#</a></dt>
<dd><p>Return parameters specifying the inferred posterior distribution of node
times which can be e.g. read into a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> for further analysis.
The mean times are not strictly constrained by topology, so unlike the
<code class="docutils literal notranslate"><span class="pre">nodes_time</span></code> attribute of a tree sequence, the mean time of a parent node
may occasionally be less than that of one of its children.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The distribution of posterior node times as a structured array of
mean and variance. Row <code class="docutils literal notranslate"><span class="pre">i</span></code> gives the mean and variance of inferred
node times for node <code class="docutils literal notranslate"><span class="pre">i</span></code>.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="tsdate.variational.ExpectationPropagation.mutation_posteriors">
<span class="sig-name descname"><span class="pre">mutation_posteriors</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsdate/variational.html#ExpectationPropagation.mutation_posteriors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsdate.variational.ExpectationPropagation.mutation_posteriors" title="Link to this definition">#</a></dt>
<dd><p>Returns parameters specifying the inferred posterior distribution of mutation
times which can be e.g. read into a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> for further analysis.
These are calculated as the midpoint distribution of the posterior node time
distributions of the node above and below the mutation. Note that this means
it is possible for a mean mutation time not to lie between the mean values of
its parent and child nodes.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For unphased singletons, the posterior mutation time is integrated over
the two possible haploid genomes on which the singleton could be placed,
accounting for the relative branch lengths above each genome.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The distribution of posterior mutation times as a structured array
of mean and variance. Row <code class="docutils literal notranslate"><span class="pre">i</span></code> gives the mean and variance of inferred
mutations times for mutation <code class="docutils literal notranslate"><span class="pre">i</span></code>.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="prior-and-time-discretisation-options">
<h2>Prior and Time Discretisation Options<a class="headerlink" href="#prior-and-time-discretisation-options" title="Link to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="tsdate.build_prior_grid">
<span class="sig-prename descclassname"><span class="pre">tsdate.</span></span><span class="sig-name descname"><span class="pre">build_prior_grid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree_sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">population_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timepoints</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">approximate_priors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">approx_prior_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prior_distribution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'lognorm'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">progress</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_unary</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tsdate.build_prior_grid" title="Link to this definition">#</a></dt>
<dd><p>Using the conditional coalescent, calculate the prior distribution for the age of
each node, given the number of contemporaneous samples below it, and
the discretised time slices at which to evaluate node age.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree_sequence</strong> (<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)"><em>tskit.TreeSequence</em></a>) – The input <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a>,
treated as undated.</p></li>
<li><p><strong>population_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a><em> or </em><a class="reference internal" href="#tsdate.demography.PopulationSizeHistory" title="tsdate.demography.PopulationSizeHistory"><em>demography.PopulationSizeHistory</em></a>) – The estimated
(diploid) effective population size used to construct the prior.
For a population with constant size, this can be given as a single
value. For a population with time-varying size, this can be given directly as
a <a class="reference internal" href="#tsdate.demography.PopulationSizeHistory" title="tsdate.demography.PopulationSizeHistory"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationSizeHistory</span></code></a> object or a parameter dictionary
passed to initialise a <a class="reference internal" href="#tsdate.demography.PopulationSizeHistory" title="tsdate.demography.PopulationSizeHistory"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationSizeHistory</span></code></a> object.
Using standard (unscaled) values for <code class="docutils literal notranslate"><span class="pre">population_size</span></code> results in a prior where
times are measured in generations.</p></li>
<li><p><strong>timepoints</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a><em> or </em><em>array_like</em>) – The number of quantiles used to create the
time slices, or manually-specified time slices as a numpy array. Default: 20</p></li>
<li><p><strong>approximate_priors</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – Whether to use a precalculated approximation to the
treewise conditional coalescent prior if there are large numbers of sample tips.
If an approximate prior has not been precalculated, tsdate will do so and cache
the result. Default: False</p></li>
<li><p><strong>approx_prior_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>) – Number of samples above which a precalculated prior is
used. Only valid if <code class="docutils literal notranslate"><span class="pre">approximate_priors</span></code> is True. Default: <code class="docutils literal notranslate"><span class="pre">None</span></code>, treated as
<a class="reference internal" href="#tsdate.prior.DEFAULT_APPROX_PRIOR_SIZE" title="tsdate.prior.DEFAULT_APPROX_PRIOR_SIZE"><code class="xref py py-data docutils literal notranslate"><span class="pre">DEFAULT_APPROX_PRIOR_SIZE</span></code></a> if
<code class="docutils literal notranslate"><span class="pre">approximate_priors</span></code> is True.</p></li>
<li><p><strong>prior_distr</strong> (<em>string</em>) – What distribution to use to approximate the conditional
coalescent prior. Can be “lognorm” for the lognormal distribution (generally a
better fit, but slightly slower to calculate) or “gamma” for the gamma
distribution (slightly faster, but a poorer fit for recent nodes). Default:
“lognorm”</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A prior object to pass to <a class="reference internal" href="#tsdate.date" title="tsdate.date"><code class="xref py py-func docutils literal notranslate"><span class="pre">date()</span></code></a> and similar functions containing
prior values for inference and a discretised time grid</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#tsdate.node_time_class.NodeTimeValues" title="tsdate.node_time_class.NodeTimeValues">node_time_class.NodeTimeValues</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tsdate.build_parameter_grid">
<span class="sig-prename descclassname"><span class="pre">tsdate.</span></span><span class="sig-name descname"><span class="pre">build_parameter_grid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree_sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">population_size</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">approximate_priors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">approx_prior_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">progress</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_unary</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#tsdate.build_parameter_grid" title="Link to this definition">#</a></dt>
<dd><p>Using the conditional coalescent, calculate the prior distribution for the age of
each node, given the number of contemporaneous samples below it, and
return parameters (shape and rate of gamma) in a grid</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree_sequence</strong> (<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)"><em>tskit.TreeSequence</em></a>) – The input tree sequence, treated as
undated.</p></li>
<li><p><strong>population_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>) – The estimated (diploid) effective population
size: must be specified. May be a single value, or a two-column array with
epoch breakpoints and effective population sizes. Using standard (unscaled)
values for <code class="docutils literal notranslate"><span class="pre">population_size</span></code> results in a prior where times are measured
in generations.</p></li>
<li><p><strong>approximate_priors</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – Whether to use a precalculated approximation to the
treewise conditional coalescent prior if there are large numbers of sample tips.
If an approximate prior has not been precalculated, tsdate will do so and cache
the result. Default: False</p></li>
<li><p><strong>approx_prior_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>) – Number of samples above which a precalculated prior is
used. Only valid if <code class="docutils literal notranslate"><span class="pre">approximate_priors</span></code> is True. Default: <code class="docutils literal notranslate"><span class="pre">None</span></code>, treated as
<a class="reference internal" href="#tsdate.prior.DEFAULT_APPROX_PRIOR_SIZE" title="tsdate.prior.DEFAULT_APPROX_PRIOR_SIZE"><code class="xref py py-data docutils literal notranslate"><span class="pre">DEFAULT_APPROX_PRIOR_SIZE</span></code></a> if <code class="docutils literal notranslate"><span class="pre">approximate_priors</span></code> is True.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#tsdate.node_time_class.NodeTimeValues" title="tsdate.node_time_class.NodeTimeValues">node_time_class.NodeTimeValues</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="tsdate.node_time_class.NodeTimeValues">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tsdate.node_time_class.</span></span><span class="sig-name descname"><span class="pre">NodeTimeValues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_nodes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nonfixed_nodes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timepoints</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fill_value=nan</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype=&lt;class</span> <span class="pre">'numpy.float64'&gt;</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsdate/node_time_class.html#NodeTimeValues"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsdate.node_time_class.NodeTimeValues" title="Link to this definition">#</a></dt>
<dd><p>A class to store times or discretised distributions of times for node ids. For nodes
with fixed times, only a single time value needs to be stored. For non-fixed nodes,
an array of either len(timepoints) probabilties or a set of (gamma) distribution
parameters is required.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This class is not intended to be used directly by users and may be subject
to change of name or internal structure in future versions. For details on
how to create a <code class="docutils literal notranslate"><span class="pre">NodeTimeValues</span></code> object to be used as a prior, see
<a class="reference internal" href="priors.html#sec-priors"><span class="std std-ref">More on priors (old)</span></a>.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Variables<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_nodes</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>) – The number of nodes that will be stored in this object</p></li>
<li><p><strong>nonfixed_nodes</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a>) – a (possibly empty) numpy array of unique positive node ids each
of which must be less than num_nodes. Each will have an array of grid_size
associated with it. All others (up to num_nodes) will be associated with a single
scalar value instead.</p></li>
<li><p><strong>timepoints</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)"><em>numpy.ndarray</em></a>) – Array of time points</p></li>
<li><p><strong>fill_value</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>) – What should we fill the data arrays with to start with</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="tsdate.prior.DEFAULT_APPROX_PRIOR_SIZE">
<span class="sig-prename descclassname"><span class="pre">tsdate.prior.</span></span><span class="sig-name descname"><span class="pre">DEFAULT_APPROX_PRIOR_SIZE</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">10000</span></em><a class="headerlink" href="#tsdate.prior.DEFAULT_APPROX_PRIOR_SIZE" title="Link to this definition">#</a></dt>
<dd><p>The default value for <cite>approx_prior_size</cite> (see <a class="reference internal" href="#tsdate.build_prior_grid" title="tsdate.build_prior_grid"><code class="xref py py-func docutils literal notranslate"><span class="pre">build_prior_grid()</span></code></a> and
<a class="reference internal" href="#tsdate.build_parameter_grid" title="tsdate.build_parameter_grid"><code class="xref py py-func docutils literal notranslate"><span class="pre">build_parameter_grid()</span></code></a>)</p>
</dd></dl>

</section>
<section id="variable-population-sizes">
<h2>Variable population sizes<a class="headerlink" href="#variable-population-sizes" title="Link to this heading">#</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="tsdate.demography.PopulationSizeHistory">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">tsdate.demography.</span></span><span class="sig-name descname"><span class="pre">PopulationSizeHistory</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">population_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_breaks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsdate/demography.html#PopulationSizeHistory"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsdate.demography.PopulationSizeHistory" title="Link to this definition">#</a></dt>
<dd><p>Stores a piecewise constant population size history and tranforms time from
a natural (generational) scale to a coalescent one.</p>
</dd></dl>

</section>
<section id="preprocessing-tree-sequences">
<h2>Preprocessing Tree Sequences<a class="headerlink" href="#preprocessing-tree-sequences" title="Link to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="tsdate.preprocess_ts">
<span class="sig-prename descclassname"><span class="pre">tsdate.</span></span><span class="sig-name descname"><span class="pre">preprocess_ts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree_sequence</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">minimum_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">erase_flanks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delete_intervals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">split_disjoint</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_populations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_individuals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_telomeres</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsdate/util.html#preprocess_ts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsdate.preprocess_ts" title="Link to this definition">#</a></dt>
<dd><p>Function to prepare tree sequences for dating by modifying the tree sequence
to increase the accuracy of dating. This can involve removing data-poor regions,
removing locally-unary segments of nodes via simplification, and splitting
discontinuous nodes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree_sequence</strong> (<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)"><em>tskit.TreeSequence</em></a>) – The input tree sequence
to be preprocessed.</p></li>
<li><p><strong>minimum_gap</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>) – The minimum gap between sites to remove from the tree
sequence. Default: <code class="docutils literal notranslate"><span class="pre">None</span></code> treated as <code class="docutils literal notranslate"><span class="pre">1000000</span></code>. Removed regions are recorded
in the provenance of the resulting tree sequence.</p></li>
<li><p><strong>erase_flanks</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – Should all material before the first site and after the
last site be removed, regardless of the length. Default: <code class="docutils literal notranslate"><span class="pre">None</span></code> treated as
<code class="docutils literal notranslate"><span class="pre">True</span></code></p></li>
<li><p><strong>delete_intervals</strong> (<em>array_like</em>) – A list (start, end) pairs describing the
genomic intervals (gaps) to delete. This is usually left as <code class="docutils literal notranslate"><span class="pre">None</span></code>
(the default) in which case <code class="docutils literal notranslate"><span class="pre">minimum_gap</span></code> and <code class="docutils literal notranslate"><span class="pre">erase_flanks</span></code> are used
to determine the gaps to remove, and the calculated intervals are recorded in
the provenance of the resulting tree sequence.</p></li>
<li><p><strong>split_disjoint</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – Run the {func}`split_disjoint_nodes` function
on the returned tree sequence, breaking any disjoint node into nodes that can
be dated separately (Default: <code class="docutils literal notranslate"><span class="pre">None</span></code> treated as <code class="docutils literal notranslate"><span class="pre">True</span></code>).</p></li>
<li><p><strong>filter_populations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – parameter passed to the
{meth}`tskit.TreeSequence.simplify` command. Unlike calling that command
directly, this defaults to <code class="docutils literal notranslate"><span class="pre">False</span></code>, such that all populations in the tree
sequence are kept.</p></li>
<li><p><strong>filter_individuals</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – parameter passed to the
{meth}`tskit.TreeSequence.simplify` command. Unlike calling that command
directly, this defaults to <code class="docutils literal notranslate"><span class="pre">False</span></code>, such
that all individuals in the tree sequence are kept.</p></li>
<li><p><strong>filter_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – parameter passed to the
{meth}`tskit.TreeSequence.simplify` command. Unlike calling that command
directly, this defaults to <code class="docutils literal notranslate"><span class="pre">False</span></code>, such
that all sites in the tree sequence are kept.</p></li>
<li><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, record details of this call to
simplify in the returned tree sequence’s provenance information
(Default: <code class="docutils literal notranslate"><span class="pre">None</span></code> treated as <code class="docutils literal notranslate"><span class="pre">True</span></code>).</p></li>
<li><p><strong>remove_telomeres</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – Deprecated alias for <code class="docutils literal notranslate"><span class="pre">erase_flanks</span></code>.</p></li>
<li><p><strong>**kwargs</strong> – All further keyword arguments are passed to the
{meth}`tskit.TreeSequence.simplify` command.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tree sequence with gaps removed.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)">tskit.TreeSequence</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tsdate.util.split_disjoint_nodes">
<span class="sig-prename descclassname"><span class="pre">tsdate.util.</span></span><span class="sig-name descname"><span class="pre">split_disjoint_nodes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ts</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_provenance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsdate/util.html#split_disjoint_nodes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsdate.util.split_disjoint_nodes" title="Link to this definition">#</a></dt>
<dd><p>For each non-sample node, split regions separated by gaps into distinct
nodes, returning a tree sequence with potentially duplicated nodes.</p>
<p>Where there are multiple disconnected regions, the leftmost one is assigned
the ID of the original node, and the remainder are assigned new node IDs.
Population, flags, individual, time, and metadata are all copied into the
new nodes. Nodes that have been split will be flagged with
<code class="docutils literal notranslate"><span class="pre">tsdate.NODE_SPLIT_BY_PREPROCESS</span></code>. The metadata of these nodes will also be
updated with an <cite>unsplit_node_id</cite> field giving the node ID in the input tree
sequence to which they correspond. If this metadata cannot be set, a warning
is emitted.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>record_provenance</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, record details of this call in the
returned tree sequence’s provenance information (Default: <code class="docutils literal notranslate"><span class="pre">None</span></code> treated
as <code class="docutils literal notranslate"><span class="pre">True</span></code>).</p>
</dd>
</dl>
</dd></dl>

</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="functions-for-inferring-tree-sequences-with-historical-samples">
<h1>Functions for Inferring Tree Sequences with Historical Samples<a class="headerlink" href="#functions-for-inferring-tree-sequences-with-historical-samples" title="Link to this heading">#</a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="tsdate.sites_time_from_ts">
<span class="sig-prename descclassname"><span class="pre">tsdate.</span></span><span class="sig-name descname"><span class="pre">sites_time_from_ts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tree_sequence</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unconstrained</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_selection</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'child'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_time</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsdate/util.html#sites_time_from_ts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsdate.sites_time_from_ts" title="Link to this definition">#</a></dt>
<dd><p>Returns an estimated “time” for each site. This is the estimated age of the oldest
MRCA which possesses a derived variant at that site, and is useful for performing
(re)inference of a tree sequence. It is calculated from the ages of nodes, with the
appropriate nodes identified by the position of mutations in the trees.</p>
<p>If node times in the tree sequence have been estimated by <code class="docutils literal notranslate"><span class="pre">tsdate</span></code> using the
inside-outside algorithm, then as well as a time in the tree sequence, nodes will
store additional time estimates that have not been explictly constrained by the
tree topology. By default, this function tries to use these “unconstrained” times,
although this is likely to fail (with a warning) on tree sequences that have not
been processed by <code class="docutils literal notranslate"><span class="pre">tsdate</span></code>: in this case the standard node times can be used by
setting <code class="docutils literal notranslate"><span class="pre">unconstrained=False</span></code>.</p>
<p>The concept of a site time is meaningless for non-variable sites, and
so the returned time for these sites is <code class="docutils literal notranslate"><span class="pre">np.nan</span></code> (note that this is not exactly
the same as tskit.UNKNOWN_TIME, which marks sites that could have a meaningful time
but whose time estimate is unknown).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree_sequence</strong> (<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Project name not set)"><em>tskit.TreeSequence</em></a>) – The input tree sequence.</p></li>
<li><p><strong>unconstrained</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – Use estimated node times which have not been constrained
by tree topology. If <code class="docutils literal notranslate"><span class="pre">True</span></code> (default), this requires a tree sequence which has
been dated using the <code class="docutils literal notranslate"><span class="pre">tsdate</span></code> inside-outside algorithm. If this is not the
case, specify <code class="docutils literal notranslate"><span class="pre">False</span></code> to use the standard tree sequence node times.</p></li>
<li><p><strong>node_selection</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>) – <p>Defines how site times are calculated from the age of
the upper and lower nodes that bound each mutation at the site. Options are
“child”, “parent”, “arithmetic” or “geometric”, with the following meanings</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">'child'</span></code> (default): the site time is the age of the oldest node
<em>below</em> each mutation at the site</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">'parent'</span></code>: the site time is the age of the oldest node <em>above</em> each
mutation at the site</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">'arithmetic'</span></code>: the arithmetic mean of the ages of the node above and the
node below each mutation is calculated; the site time is the oldest
of these means.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">'geometric'</span></code>: the geometric mean of the ages of the node above and the
node below each mutation is calculated; the site time is the oldest
of these means</p></li>
</ul>
</p></li>
<li><p><strong>min_time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.14)"><em>float</em></a>) – A site time of zero implies that no MRCA in the past
possessed the derived variant, so the variant cannot be used for inferring
relationships between the samples. To allow all variants to be potentially
available for inference, if a site time would otherwise be calculated as zero
(for example, where the <code class="docutils literal notranslate"><span class="pre">mutation_age</span></code> parameter is “child” or “geometric”
and all mutations at a site are associated with leaf nodes), a minimum site
greater than 0 is recommended. By default this is set to 1, which is generally
reasonable for times measured in generations or years, although it is also
fine to set this to a small epsilon value.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Array of length tree_sequence.num_sites with estimated time of each site</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.4)">numpy.ndarray</a>(dtype=np.float64)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="tsdate.add_sampledata_times">
<span class="sig-prename descclassname"><span class="pre">tsdate.</span></span><span class="sig-name descname"><span class="pre">add_sampledata_times</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">samples</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sites_time</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/tsdate/util.html#add_sampledata_times"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#tsdate.add_sampledata_times" title="Link to this definition">#</a></dt>
<dd><p>Return a tsinfer.SampleData file with estimated times associated with sites.
Ensures that each site’s time is at least as old as the oldest historic sample
carrying a derived allele at that site.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>samples</strong> – A tsinfer SampleData object to
add site times to. Any historic individuals in this SampleData file are used to
constrain site times.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A copy of the input sample data with site times added</p>
</dd>
</dl>
</dd></dl>

</section>
<section class="tex2jax_ignore mathjax_ignore" id="module-tsdate">
<span id="constants"></span><h1>Constants<a class="headerlink" href="#module-tsdate" title="Link to this heading">#</a></h1>
<dl class="py data">
<dt class="sig sig-object py" id="tsdate.NODE_IS_HISTORICAL_SAMPLE">
<span class="sig-prename descclassname"><span class="pre">tsdate.</span></span><span class="sig-name descname"><span class="pre">NODE_IS_HISTORICAL_SAMPLE</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">1048576</span></em><a class="headerlink" href="#tsdate.NODE_IS_HISTORICAL_SAMPLE" title="Link to this definition">#</a></dt>
<dd><p>Node flag value indicating that this is a non-contemporary sample node</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="tsdate.NODE_SPLIT_BY_PREPROCESS">
<span class="sig-prename descclassname"><span class="pre">tsdate.</span></span><span class="sig-name descname"><span class="pre">NODE_SPLIT_BY_PREPROCESS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">1073741824</span></em><a class="headerlink" href="#tsdate.NODE_SPLIT_BY_PREPROCESS" title="Link to this definition">#</a></dt>
<dd><p>Node flag value indicating that this was a disjoint node that was then split</p>
</dd></dl>

</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="api.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">APIs</p>
      </div>
    </a>
    <a class="right-next"
       href="cli.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Command line interface</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Python API</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#running-tsdate">Running tsdate</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.date"><code class="docutils literal notranslate"><span class="pre">date()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.core.estimation_methods"><code class="docutils literal notranslate"><span class="pre">estimation_methods</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.variational_gamma"><code class="docutils literal notranslate"><span class="pre">variational_gamma()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.inside_outside"><code class="docutils literal notranslate"><span class="pre">inside_outside()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.maximization"><code class="docutils literal notranslate"><span class="pre">maximization()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#underlying-fit-objects">Underlying fit objects</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.discrete.BeliefPropagation"><code class="docutils literal notranslate"><span class="pre">BeliefPropagation</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.discrete.BeliefPropagation.node_posteriors"><code class="docutils literal notranslate"><span class="pre">BeliefPropagation.node_posteriors()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.variational.ExpectationPropagation"><code class="docutils literal notranslate"><span class="pre">ExpectationPropagation</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.variational.ExpectationPropagation.node_posteriors"><code class="docutils literal notranslate"><span class="pre">ExpectationPropagation.node_posteriors()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.variational.ExpectationPropagation.mutation_posteriors"><code class="docutils literal notranslate"><span class="pre">ExpectationPropagation.mutation_posteriors()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#prior-and-time-discretisation-options">Prior and Time Discretisation Options</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.build_prior_grid"><code class="docutils literal notranslate"><span class="pre">build_prior_grid()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.build_parameter_grid"><code class="docutils literal notranslate"><span class="pre">build_parameter_grid()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.node_time_class.NodeTimeValues"><code class="docutils literal notranslate"><span class="pre">NodeTimeValues</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.prior.DEFAULT_APPROX_PRIOR_SIZE"><code class="docutils literal notranslate"><span class="pre">DEFAULT_APPROX_PRIOR_SIZE</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#variable-population-sizes">Variable population sizes</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.demography.PopulationSizeHistory"><code class="docutils literal notranslate"><span class="pre">PopulationSizeHistory</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#preprocessing-tree-sequences">Preprocessing Tree Sequences</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.preprocess_ts"><code class="docutils literal notranslate"><span class="pre">preprocess_ts()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.util.split_disjoint_nodes"><code class="docutils literal notranslate"><span class="pre">split_disjoint_nodes()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#functions-for-inferring-tree-sequences-with-historical-samples">Functions for Inferring Tree Sequences with Historical Samples</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.sites_time_from_ts"><code class="docutils literal notranslate"><span class="pre">sites_time_from_ts()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.add_sampledata_times"><code class="docutils literal notranslate"><span class="pre">add_sampledata_times()</span></code></a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#module-tsdate">Constants</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.NODE_IS_HISTORICAL_SAMPLE"><code class="docutils literal notranslate"><span class="pre">NODE_IS_HISTORICAL_SAMPLE</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tsdate.NODE_SPLIT_BY_PREPROCESS"><code class="docutils literal notranslate"><span class="pre">NODE_SPLIT_BY_PREPROCESS</span></code></a></li>
</ul>
</li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Tskit Developers
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>